--- net/minecraft/world/level/levelgen/feature/configurations/RangeDecoratorConfiguration.java
+++ net/minecraft/world/level/levelgen/feature/configurations/RangeDecoratorConfiguration.java
@@ -1,26 +1,19 @@
 package net.minecraft.world.level.levelgen.feature.configurations;
 
-import com.mojang.datafixers.util.Function3;
 import com.mojang.serialization.Codec;
 import com.mojang.serialization.codecs.RecordCodecBuilder;
+import java.util.function.Function;
+import net.minecraft.world.level.levelgen.heightproviders.HeightProvider;
 
-public class RangeDecoratorConfiguration implements DecoratorConfiguration {
+public class RangeDecoratorConfiguration implements DecoratorConfiguration, FeatureConfiguration {
    public static final Codec<RangeDecoratorConfiguration> CODEC = RecordCodecBuilder.create((var0) -> {
-      return var0.group(Codec.INT.fieldOf("bottom_offset").orElse(0).forGetter((var0x) -> {
-         return var0x.bottomOffset;
-      }), Codec.INT.fieldOf("top_offset").orElse(0).forGetter((var0x) -> {
-         return var0x.topOffset;
-      }), Codec.INT.fieldOf("maximum").orElse(0).forGetter((var0x) -> {
-         return var0x.maximum;
-      })).apply(var0, (Function3)(RangeDecoratorConfiguration::new));
+      return var0.group(HeightProvider.CODEC.fieldOf("height").forGetter((var0x) -> {
+         return var0x.height;
+      })).apply(var0, (Function)(RangeDecoratorConfiguration::new));
    });
-   public final int bottomOffset;
-   public final int topOffset;
-   public final int maximum;
+   public final HeightProvider height;
 
-   public RangeDecoratorConfiguration(int var1, int var2, int var3) {
-      this.bottomOffset = var1;
-      this.topOffset = var2;
-      this.maximum = var3;
+   public RangeDecoratorConfiguration(HeightProvider var1) {
+      this.height = var1;
    }
 }
