--- net/minecraft/world/level/levelgen/feature/configurations/SimpleRandomFeatureConfiguration.java
+++ net/minecraft/world/level/levelgen/feature/configurations/SimpleRandomFeatureConfiguration.java
@@ -2,10 +2,11 @@
 
 import com.mojang.serialization.Codec;
 import java.util.List;
 import java.util.function.Supplier;
 import java.util.stream.Stream;
+import net.minecraft.util.ExtraCodecs;
 import net.minecraft.world.level.levelgen.feature.ConfiguredFeature;
 
 public class SimpleRandomFeatureConfiguration implements FeatureConfiguration {
    public static final Codec<SimpleRandomFeatureConfiguration> CODEC;
    public final List<Supplier<ConfiguredFeature<?, ?>>> features;
@@ -19,10 +20,10 @@
          return ((ConfiguredFeature)var0.get()).getFeatures();
       });
    }
 
    static {
-      CODEC = ConfiguredFeature.LIST_CODEC.fieldOf("features").xmap(SimpleRandomFeatureConfiguration::new, (var0) -> {
+      CODEC = ExtraCodecs.nonEmptyList(ConfiguredFeature.LIST_CODEC).fieldOf("features").xmap(SimpleRandomFeatureConfiguration::new, (var0) -> {
          return var0.features;
       }).codec();
    }
 }
