--- net/minecraft/world/level/levelgen/carver/CarverConfiguration.java
+++ net/minecraft/world/level/levelgen/carver/CarverConfiguration.java
@@ -1,5 +1,27 @@
 package net.minecraft.world.level.levelgen.carver;
 
-public interface CarverConfiguration {
-   NoneCarverConfiguration NONE = NoneCarverConfiguration.INSTANCE;
+import com.mojang.serialization.Codec;
+import com.mojang.serialization.codecs.RecordCodecBuilder;
+import net.minecraft.world.level.levelgen.feature.configurations.ProbabilityFeatureConfiguration;
+
+public class CarverConfiguration extends ProbabilityFeatureConfiguration {
+   public static final Codec<CarverConfiguration> CODEC = RecordCodecBuilder.create((var0) -> {
+      return var0.group(Codec.floatRange(0.0F, 1.0F).fieldOf("probability").forGetter((var0x) -> {
+         return var0x.probability;
+      }), CarverDebugSettings.CODEC.optionalFieldOf("debug_settings", CarverDebugSettings.DEFAULT).forGetter(CarverConfiguration::getDebugSettings)).apply(var0, CarverConfiguration::new);
+   });
+   private final CarverDebugSettings debugSettings;
+
+   public CarverConfiguration(float var1, CarverDebugSettings var2) {
+      super(var1);
+      this.debugSettings = var2;
+   }
+
+   public CarverConfiguration(float var1) {
+      this(var1, CarverDebugSettings.DEFAULT);
+   }
+
+   public CarverDebugSettings getDebugSettings() {
+      return this.debugSettings;
+   }
 }
