--- net/minecraft/world/level/levelgen/carver/CarvingContext.java
+++ net/minecraft/world/level/levelgen/carver/CarvingContext.java
@@ -1,0 +1,20 @@
+package net.minecraft.world.level.levelgen.carver;
+
+import net.minecraft.world.level.chunk.ChunkGenerator;
+import net.minecraft.world.level.levelgen.WorldGenerationContext;
+
+public class CarvingContext implements WorldGenerationContext {
+   private final ChunkGenerator generator;
+
+   public CarvingContext(ChunkGenerator var1) {
+      this.generator = var1;
+   }
+
+   public int getMinGenY() {
+      return this.generator.getMinY();
+   }
+
+   public int getGenDepth() {
+      return this.generator.getGenDepth();
+   }
+}
