--- net/minecraft/world/level/levelgen/carver/NoneCarverConfiguration.java
+++ net/minecraft/world/level/levelgen/carver/NoneCarverConfiguration.java
@@ -1,13 +1,0 @@
-package net.minecraft.world.level.levelgen.carver;
-
-import com.mojang.serialization.Codec;
-
-public class NoneCarverConfiguration implements CarverConfiguration {
-   public static final Codec<NoneCarverConfiguration> CODEC = Codec.unit(() -> {
-      return INSTANCE;
-   });
-   public static final NoneCarverConfiguration INSTANCE = new NoneCarverConfiguration();
-
-   public NoneCarverConfiguration() {
-   }
-}
