--- net/minecraft/commands/arguments/AngleArgument.java
+++ net/minecraft/commands/arguments/AngleArgument.java
@@ -13,10 +13,11 @@
 import net.minecraft.util.Mth;
 
 public class AngleArgument implements ArgumentType<AngleArgument.SingleAngle> {
    private static final Collection<String> EXAMPLES = Arrays.asList("0", "~", "~-5");
    public static final SimpleCommandExceptionType ERROR_NOT_COMPLETE = new SimpleCommandExceptionType(new TranslatableComponent("argument.angle.incomplete"));
+   public static final SimpleCommandExceptionType ERROR_INVALID_ANGLE = new SimpleCommandExceptionType(new TranslatableComponent("argument.angle.invalid"));
 
    public AngleArgument() {
    }
 
    public static AngleArgument angle() {
@@ -31,11 +32,15 @@
       if (!var1.canRead()) {
          throw ERROR_NOT_COMPLETE.createWithContext(var1);
       } else {
          boolean var2 = WorldCoordinate.isRelative(var1);
          float var3 = var1.canRead() && var1.peek() != ' ' ? var1.readFloat() : 0.0F;
-         return new AngleArgument.SingleAngle(var3, var2);
+         if (!Float.isNaN(var3) && !Float.isInfinite(var3)) {
+            return new AngleArgument.SingleAngle(var3, var2);
+         } else {
+            throw ERROR_INVALID_ANGLE.createWithContext(var1);
+         }
       }
    }
 
    public Collection<String> getExamples() {
       return EXAMPLES;
