--- net/minecraft/commands/arguments/GameProfileArgument.java
+++ net/minecraft/commands/arguments/GameProfileArgument.java
@@ -87,10 +87,15 @@
 
    public Collection<String> getExamples() {
       return EXAMPLES;
    }
 
+   @FunctionalInterface
+   public interface Result {
+      Collection<GameProfile> getNames(CommandSourceStack var1) throws CommandSyntaxException;
+   }
+
    public static class SelectorResult implements GameProfileArgument.Result {
       private final EntitySelector selector;
 
       public SelectorResult(EntitySelector var1) {
          this.selector = var1;
@@ -111,11 +116,6 @@
 
             return var3;
          }
       }
    }
-
-   @FunctionalInterface
-   public interface Result {
-      Collection<GameProfile> getNames(CommandSourceStack var1) throws CommandSyntaxException;
-   }
 }
