--- net/minecraft/commands/arguments/RangeArgument.java
+++ net/minecraft/commands/arguments/RangeArgument.java
@@ -22,10 +22,14 @@
       private static final Collection<String> EXAMPLES = Arrays.asList("0..5.2", "0", "-5.4", "-100.76..", "..100");
 
       public Floats() {
       }
 
+      public static MinMaxBounds.Floats getRange(CommandContext<CommandSourceStack> var0, String var1) {
+         return (MinMaxBounds.Floats)var0.getArgument(var1, MinMaxBounds.Floats.class);
+      }
+
       public MinMaxBounds.Floats parse(StringReader var1) throws CommandSyntaxException {
          return MinMaxBounds.Floats.fromReader(var1);
       }
 
       public Collection<String> getExamples() {
