--- net/minecraft/commands/arguments/RangeArgument.java
+++ net/minecraft/commands/arguments/RangeArgument.java
@@ -16,37 +16,41 @@
 
    static RangeArgument.Floats floatRange() {
       return new RangeArgument.Floats();
    }
 
-   public static class Floats implements RangeArgument<MinMaxBounds.Floats> {
-      private static final Collection<String> EXAMPLES = Arrays.asList("0..5.2", "0", "-5.4", "-100.76..", "..100");
+   public static class Ints implements RangeArgument<MinMaxBounds.Ints> {
+      private static final Collection<String> EXAMPLES = Arrays.asList("0..5", "0", "-5", "-100..", "..100");
 
-      public Floats() {
+      public Ints() {
       }
 
-      public MinMaxBounds.Floats parse(StringReader var1) throws CommandSyntaxException {
-         return MinMaxBounds.Floats.fromReader(var1);
+      public static MinMaxBounds.Ints getRange(CommandContext<CommandSourceStack> var0, String var1) {
+         return (MinMaxBounds.Ints)var0.getArgument(var1, MinMaxBounds.Ints.class);
       }
 
+      public MinMaxBounds.Ints parse(StringReader var1) throws CommandSyntaxException {
+         return MinMaxBounds.Ints.fromReader(var1);
+      }
+
       public Collection<String> getExamples() {
          return EXAMPLES;
       }
    }
 
-   public static class Ints implements RangeArgument<MinMaxBounds.Ints> {
-      private static final Collection<String> EXAMPLES = Arrays.asList("0..5", "0", "-5", "-100..", "..100");
+   public static class Floats implements RangeArgument<MinMaxBounds.Doubles> {
+      private static final Collection<String> EXAMPLES = Arrays.asList("0..5.2", "0", "-5.4", "-100.76..", "..100");
 
-      public Ints() {
+      public Floats() {
       }
 
-      public static MinMaxBounds.Ints getRange(CommandContext<CommandSourceStack> var0, String var1) {
-         return (MinMaxBounds.Ints)var0.getArgument(var1, MinMaxBounds.Ints.class);
+      public static MinMaxBounds.Doubles getRange(CommandContext<CommandSourceStack> var0, String var1) {
+         return (MinMaxBounds.Doubles)var0.getArgument(var1, MinMaxBounds.Doubles.class);
       }
 
-      public MinMaxBounds.Ints parse(StringReader var1) throws CommandSyntaxException {
-         return MinMaxBounds.Ints.fromReader(var1);
+      public MinMaxBounds.Doubles parse(StringReader var1) throws CommandSyntaxException {
+         return MinMaxBounds.Doubles.fromReader(var1);
       }
 
       public Collection<String> getExamples() {
          return EXAMPLES;
       }
