--- net/minecraft/commands/arguments/ResourceLocationArgument.java
+++ net/minecraft/commands/arguments/ResourceLocationArgument.java
@@ -13,11 +13,13 @@
 import net.minecraft.network.chat.TranslatableComponent;
 import net.minecraft.resources.ResourceLocation;
 import net.minecraft.world.entity.ai.attributes.Attribute;
 import net.minecraft.world.item.crafting.Recipe;
 import net.minecraft.world.item.crafting.RecipeManager;
+import net.minecraft.world.level.storage.loot.ItemModifierManager;
 import net.minecraft.world.level.storage.loot.PredicateManager;
+import net.minecraft.world.level.storage.loot.functions.LootItemFunction;
 import net.minecraft.world.level.storage.loot.predicates.LootItemCondition;
 
 public class ResourceLocationArgument implements ArgumentType<ResourceLocation> {
    private static final Collection<String> EXAMPLES = Arrays.asList("foo", "foo:bar", "012");
    private static final DynamicCommandExceptionType ERROR_UNKNOWN_ADVANCEMENT = new DynamicCommandExceptionType((var0) -> {
@@ -30,10 +32,13 @@
       return new TranslatableComponent("predicate.unknown", new Object[]{var0});
    });
    private static final DynamicCommandExceptionType ERROR_UNKNOWN_ATTRIBUTE = new DynamicCommandExceptionType((var0) -> {
       return new TranslatableComponent("attribute.unknown", new Object[]{var0});
    });
+   private static final DynamicCommandExceptionType ERROR_UNKNOWN_ITEM_MODIFIER = new DynamicCommandExceptionType((var0) -> {
+      return new TranslatableComponent("item_modifier.unknown", new Object[]{var0});
+   });
 
    public ResourceLocationArgument() {
    }
 
    public static ResourceLocationArgument id() {
@@ -67,10 +72,21 @@
       } else {
          return var4;
       }
    }
 
+   public static LootItemFunction getItemModifier(CommandContext<CommandSourceStack> var0, String var1) throws CommandSyntaxException {
+      ResourceLocation var2 = (ResourceLocation)var0.getArgument(var1, ResourceLocation.class);
+      ItemModifierManager var3 = ((CommandSourceStack)var0.getSource()).getServer().getItemModifierManager();
+      LootItemFunction var4 = var3.get(var2);
+      if (var4 == null) {
+         throw ERROR_UNKNOWN_ITEM_MODIFIER.create(var2);
+      } else {
+         return var4;
+      }
+   }
+
    public static Attribute getAttribute(CommandContext<CommandSourceStack> var0, String var1) throws CommandSyntaxException {
       ResourceLocation var2 = (ResourceLocation)var0.getArgument(var1, ResourceLocation.class);
       return (Attribute)Registry.ATTRIBUTE.getOptional(var2).orElseThrow(() -> {
          return ERROR_UNKNOWN_ATTRIBUTE.create(var2);
       });
