--- net/minecraft/commands/arguments/SlotArgument.java
+++ net/minecraft/commands/arguments/SlotArgument.java
@@ -47,17 +47,17 @@
 
       for(var1 = 0; var1 < 15; ++var1) {
          var0.put("horse." + var1, 500 + var1);
       }
 
-      var0.put("weapon", 98);
-      var0.put("weapon.mainhand", 98);
-      var0.put("weapon.offhand", 99);
-      var0.put("armor.head", 100 + EquipmentSlot.HEAD.getIndex());
-      var0.put("armor.chest", 100 + EquipmentSlot.CHEST.getIndex());
-      var0.put("armor.legs", 100 + EquipmentSlot.LEGS.getIndex());
-      var0.put("armor.feet", 100 + EquipmentSlot.FEET.getIndex());
+      var0.put("weapon", EquipmentSlot.MAINHAND.getIndex(98));
+      var0.put("weapon.mainhand", EquipmentSlot.MAINHAND.getIndex(98));
+      var0.put("weapon.offhand", EquipmentSlot.OFFHAND.getIndex(98));
+      var0.put("armor.head", EquipmentSlot.HEAD.getIndex(100));
+      var0.put("armor.chest", EquipmentSlot.CHEST.getIndex(100));
+      var0.put("armor.legs", EquipmentSlot.LEGS.getIndex(100));
+      var0.put("armor.feet", EquipmentSlot.FEET.getIndex(100));
       var0.put("horse.saddle", 400);
       var0.put("horse.armor", 401);
       var0.put("horse.chest", 499);
    });
 
