--- net/minecraft/world/level/BlockGetter.java
+++ net/minecraft/world/level/BlockGetter.java
@@ -14,11 +14,11 @@
 import net.minecraft.world.phys.AABB;
 import net.minecraft.world.phys.BlockHitResult;
 import net.minecraft.world.phys.Vec3;
 import net.minecraft.world.phys.shapes.VoxelShape;
 
-public interface BlockGetter {
+public interface BlockGetter extends LevelHeightAccessor {
    @Nullable
    BlockEntity getBlockEntity(BlockPos var1);
 
    BlockState getBlockState(BlockPos var1);
 
@@ -30,14 +30,10 @@
 
    default int getMaxLightLevel() {
       return 15;
    }
 
-   default int getMaxBuildHeight() {
-      return 256;
-   }
-
    default Stream<BlockState> getBlockStates(AABB var1) {
       return BlockPos.betweenClosedStream(var1).map(this::getBlockState);
    }
 
    default BlockHitResult clip(ClipContext var1) {
