--- net/minecraft/world/level/ChunkPos.java
+++ net/minecraft/world/level/ChunkPos.java
@@ -4,10 +4,11 @@
 import java.util.function.Consumer;
 import java.util.stream.Stream;
 import java.util.stream.StreamSupport;
 import javax.annotation.Nullable;
 import net.minecraft.core.BlockPos;
+import net.minecraft.core.SectionPos;
 
 public class ChunkPos {
    public static final long INVALID_CHUNK_POS = asLong(1875016, 1875016);
    public final int x;
    public final int z;
@@ -16,12 +17,12 @@
       this.x = var1;
       this.z = var2;
    }
 
    public ChunkPos(BlockPos var1) {
-      this.x = var1.getX() >> 4;
-      this.z = var1.getZ() >> 4;
+      this.x = SectionPos.blockToSectionCoord(var1.getX());
+      this.z = SectionPos.blockToSectionCoord(var1.getZ());
    }
 
    public ChunkPos(long var1) {
       this.x = (int)var1;
       this.z = (int)(var1 >> 32);
@@ -59,23 +60,23 @@
          return this.x == var2.x && this.z == var2.z;
       }
    }
 
    public int getMinBlockX() {
-      return this.x << 4;
+      return SectionPos.sectionToBlockCoord(this.x);
    }
 
    public int getMinBlockZ() {
-      return this.z << 4;
+      return SectionPos.sectionToBlockCoord(this.z);
    }
 
    public int getMaxBlockX() {
-      return (this.x << 4) + 15;
+      return SectionPos.sectionToBlockCoord(this.x, 15);
    }
 
    public int getMaxBlockZ() {
-      return (this.z << 4) + 15;
+      return SectionPos.sectionToBlockCoord(this.z, 15);
    }
 
    public int getRegionX() {
       return this.x >> 5;
    }
