--- net/minecraft/world/level/ChunkTickList.java
+++ net/minecraft/world/level/ChunkTickList.java
@@ -76,10 +76,14 @@
       this.ticks.forEach((var1x) -> {
          var1.scheduleTick(var1x.pos, var1x.type, var1x.delay, var1x.priority);
       });
    }
 
+   public int size() {
+      return this.ticks.size();
+   }
+
    static class ScheduledTick<T> {
       private final T type;
       public final BlockPos pos;
       public final int delay;
       public final TickPriority priority;
