--- net/minecraft/world/level/CollisionSpliterator.java
+++ net/minecraft/world/level/CollisionSpliterator.java
@@ -5,10 +5,11 @@
 import java.util.function.BiPredicate;
 import java.util.function.Consumer;
 import javax.annotation.Nullable;
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.Cursor3D;
+import net.minecraft.core.SectionPos;
 import net.minecraft.util.Mth;
 import net.minecraft.world.entity.Entity;
 import net.minecraft.world.level.block.Blocks;
 import net.minecraft.world.level.block.state.BlockState;
 import net.minecraft.world.level.border.WorldBorder;
@@ -104,12 +105,12 @@
       }
    }
 
    @Nullable
    private BlockGetter getChunk(int var1, int var2) {
-      int var3 = var1 >> 4;
-      int var4 = var2 >> 4;
+      int var3 = SectionPos.blockToSectionCoord(var1);
+      int var4 = SectionPos.blockToSectionCoord(var2);
       return this.collisionGetter.getChunkForCollisions(var3, var4);
    }
 
    boolean worldBorderCheck(Consumer<? super VoxelShape> var1) {
       Objects.requireNonNull(this.source);
