--- net/minecraft/world/level/CollisionGetter.java
+++ net/minecraft/world/level/CollisionGetter.java
@@ -63,9 +63,13 @@
 
    default Stream<VoxelShape> getBlockCollisions(@Nullable Entity var1, AABB var2) {
       return StreamSupport.stream(new CollisionSpliterator(this, var1, var2), false);
    }
 
+   default boolean noBlockCollision(@Nullable Entity var1, AABB var2, BiPredicate<BlockState, BlockPos> var3) {
+      return this.getBlockCollisions(var1, var2, var3).allMatch(VoxelShape::isEmpty);
+   }
+
    default Stream<VoxelShape> getBlockCollisions(@Nullable Entity var1, AABB var2, BiPredicate<BlockState, BlockPos> var3) {
       return StreamSupport.stream(new CollisionSpliterator(this, var1, var2, var3), false);
    }
 }
