--- net/minecraft/world/level/GameType.java
+++ net/minecraft/world/level/GameType.java
@@ -1,38 +1,48 @@
 package net.minecraft.world.level;
 
+import javax.annotation.Nullable;
 import net.minecraft.network.chat.Component;
 import net.minecraft.network.chat.TranslatableComponent;
 import net.minecraft.world.entity.player.Abilities;
 
 public enum GameType {
-   NOT_SET(-1, ""),
    SURVIVAL(0, "survival"),
    CREATIVE(1, "creative"),
    ADVENTURE(2, "adventure"),
    SPECTATOR(3, "spectator");
 
+   public static final GameType DEFAULT_MODE = SURVIVAL;
+   private static final int NOT_SET = -1;
    private final int id;
    private final String name;
+   private final Component shortName;
+   private final Component longName;
 
    private GameType(int var3, String var4) {
       this.id = var3;
       this.name = var4;
+      this.shortName = new TranslatableComponent("selectWorld.gameMode." + var4);
+      this.longName = new TranslatableComponent("gameMode." + var4);
    }
 
    public int getId() {
       return this.id;
    }
 
    public String getName() {
       return this.name;
    }
 
-   public Component getDisplayName() {
-      return new TranslatableComponent("gameMode." + this.name);
+   public Component getLongDisplayName() {
+      return this.longName;
    }
 
+   public Component getShortDisplayName() {
+      return this.shortName;
+   }
+
    public void updatePlayerAbilities(Abilities var1) {
       if (this == CREATIVE) {
          var1.mayfly = true;
          var1.instabuild = true;
          var1.invulnerable = true;
@@ -62,11 +72,11 @@
    public boolean isSurvival() {
       return this == SURVIVAL || this == ADVENTURE;
    }
 
    public static GameType byId(int var0) {
-      return byId(var0, SURVIVAL);
+      return byId(var0, DEFAULT_MODE);
    }
 
    public static GameType byId(int var0, GameType var1) {
       GameType[] var2 = values();
       int var3 = var2.length;
@@ -96,6 +106,15 @@
          }
       }
 
       return var1;
    }
+
+   public static int getNullableId(@Nullable GameType var0) {
+      return var0 != null ? var0.id : -1;
+   }
+
+   @Nullable
+   public static GameType byNullableId(int var0) {
+      return var0 == -1 ? null : byId(var0);
+   }
 }
