--- net/minecraft/world/level/LevelAccessor.java
+++ net/minecraft/world/level/LevelAccessor.java
@@ -46,11 +46,11 @@
 
    void addParticle(ParticleOptions var1, double var2, double var4, double var6, double var8, double var10, double var12);
 
    void levelEvent(@Nullable Player var1, int var2, BlockPos var3, int var4);
 
-   default int getHeight() {
+   default int getLogicalHeight() {
       return this.dimensionType().logicalHeight();
    }
 
    default void levelEvent(int var1, BlockPos var2, int var3) {
       this.levelEvent((Player)null, var1, var2, var3);
