--- net/minecraft/world/level/SpawnData.java
+++ net/minecraft/world/level/SpawnData.java
@@ -1,12 +1,14 @@
 package net.minecraft.world.level;
 
 import net.minecraft.nbt.CompoundTag;
 import net.minecraft.resources.ResourceLocation;
-import net.minecraft.util.WeighedRandom;
+import net.minecraft.util.random.WeightedEntry;
 
-public class SpawnData extends WeighedRandom.WeighedRandomItem {
+public class SpawnData extends WeightedEntry.IntrusiveBase {
+   public static final int DEFAULT_WEIGHT = 1;
+   public static final String DEFAULT_TYPE = "minecraft:pig";
    private final CompoundTag tag;
 
    public SpawnData() {
       super(1);
       this.tag = new CompoundTag();
@@ -30,11 +32,11 @@
    }
 
    public CompoundTag save() {
       CompoundTag var1 = new CompoundTag();
       var1.put("Entity", this.tag);
-      var1.putInt("Weight", this.weight);
+      var1.putInt("Weight", this.getWeight().asInt());
       return var1;
    }
 
    public CompoundTag getTag() {
       return this.tag;
