--- net/minecraft/client/searchtree/ReloadableSearchTree.java
+++ net/minecraft/client/searchtree/ReloadableSearchTree.java
@@ -45,11 +45,11 @@
          List var6 = this.tree.search(var4);
          return Lists.newArrayList(new ReloadableIdSearchTree.IntersectionIterator(var3.iterator(), new ReloadableSearchTree.MergingUniqueIterator(var5.iterator(), var6.iterator(), this::comparePosition), this::comparePosition));
       }
    }
 
-   static class MergingUniqueIterator<T> extends AbstractIterator<T> {
+   private static class MergingUniqueIterator<T> extends AbstractIterator<T> {
       private final PeekingIterator<T> firstIterator;
       private final PeekingIterator<T> secondIterator;
       private final Comparator<T> orderT;
 
       public MergingUniqueIterator(Iterator<T> var1, Iterator<T> var2, Comparator<T> var3) {
