--- net/minecraft/client/searchtree/ReloadableIdSearchTree.java
+++ net/minecraft/client/searchtree/ReloadableIdSearchTree.java
@@ -71,11 +71,11 @@
          List var5 = this.pathTree.search(var4);
          return Lists.newArrayList(new ReloadableIdSearchTree.IntersectionIterator(var3.iterator(), var5.iterator(), this::comparePosition));
       }
    }
 
-   public static class IntersectionIterator<T> extends AbstractIterator<T> {
+   protected static class IntersectionIterator<T> extends AbstractIterator<T> {
       private final PeekingIterator<T> firstIterator;
       private final PeekingIterator<T> secondIterator;
       private final Comparator<T> orderT;
 
       public IntersectionIterator(Iterator<T> var1, Iterator<T> var2, Comparator<T> var3) {
