--- net/minecraft/tags/ItemTags.java
+++ net/minecraft/tags/ItemTags.java
@@ -44,13 +44,16 @@
    public static final Tag.Named<Item> FENCES = bind("fences");
    public static final Tag.Named<Item> TALL_FLOWERS = bind("tall_flowers");
    public static final Tag.Named<Item> FLOWERS = bind("flowers");
    public static final Tag.Named<Item> PIGLIN_REPELLENTS = bind("piglin_repellents");
    public static final Tag.Named<Item> PIGLIN_LOVED = bind("piglin_loved");
+   public static final Tag.Named<Item> IGNORED_BY_PIGLIN_BABIES = bind("ignored_by_piglin_babies");
+   public static final Tag.Named<Item> PIGLIN_FOOD = bind("piglin_food");
    public static final Tag.Named<Item> GOLD_ORES = bind("gold_ores");
    public static final Tag.Named<Item> NON_FLAMMABLE_WOOD = bind("non_flammable_wood");
    public static final Tag.Named<Item> SOUL_FIRE_BASE_BLOCKS = bind("soul_fire_base_blocks");
+   public static final Tag.Named<Item> CANDLES = bind("candles");
    public static final Tag.Named<Item> BOATS = bind("boats");
    public static final Tag.Named<Item> FISHES = bind("fishes");
    public static final Tag.Named<Item> SIGNS = bind("signs");
    public static final Tag.Named<Item> MUSIC_DISCS = bind("music_discs");
    public static final Tag.Named<Item> CREEPER_DROP_MUSIC_DISCS = bind("creeper_drop_music_discs");
