--- com/mojang/brigadier/arguments/BoolArgumentType.java
+++ com/mojang/brigadier/arguments/BoolArgumentType.java
@@ -26,15 +26,15 @@
    public Boolean parse(StringReader var1) throws CommandSyntaxException {
       return var1.readBoolean();
    }
 
    public <S> CompletableFuture<Suggestions> listSuggestions(CommandContext<S> var1, SuggestionsBuilder var2) {
-      if ("true".startsWith(var2.getRemaining().toLowerCase())) {
+      if ("true".startsWith(var2.getRemainingLowerCase())) {
          var2.suggest("true");
       }
 
-      if ("false".startsWith(var2.getRemaining().toLowerCase())) {
+      if ("false".startsWith(var2.getRemainingLowerCase())) {
          var2.suggest("false");
       }
 
       return var2.buildFuture();
    }
