--- net/minecraft/recipebook/ServerPlaceRecipe.java
+++ net/minecraft/recipebook/ServerPlaceRecipe.java
@@ -31,23 +31,23 @@
       this.menu = var1;
    }
 
    public void recipeClicked(ServerPlayer var1, @Nullable Recipe<C> var2, boolean var3) {
       if (var2 != null && var1.getRecipeBook().contains(var2)) {
-         this.inventory = var1.inventory;
+         this.inventory = var1.getInventory();
          if (this.testClearGrid() || var1.isCreative()) {
             this.stackedContents.clear();
-            var1.inventory.fillStackedContents(this.stackedContents);
+            var1.getInventory().fillStackedContents(this.stackedContents);
             this.menu.fillCraftSlotsStackedContents(this.stackedContents);
             if (this.stackedContents.canCraft(var2, (IntList)null)) {
                this.handleRecipeClicked(var2, var3);
             } else {
                this.clearGrid();
                var1.connection.send(new ClientboundPlaceGhostRecipePacket(var1.containerMenu.containerId, var2));
             }
 
-            var1.inventory.setChanged();
+            var1.getInventory().setChanged();
          }
       }
    }
 
    protected void clearGrid() {
