--- net/minecraft/world/level/levelgen/surfacebuilders/FrozenOceanSurfaceBuilder.java
+++ net/minecraft/world/level/levelgen/surfacebuilders/FrozenOceanSurfaceBuilder.java
@@ -25,99 +25,99 @@
 
    public FrozenOceanSurfaceBuilder(Codec<SurfaceBuilderBaseConfiguration> var1) {
       super(var1);
    }
 
-   public void apply(Random var1, ChunkAccess var2, Biome var3, int var4, int var5, int var6, double var7, BlockState var9, BlockState var10, int var11, long var12, SurfaceBuilderBaseConfiguration var14) {
-      double var15 = 0.0D;
-      double var17 = 0.0D;
-      BlockPos.MutableBlockPos var19 = new BlockPos.MutableBlockPos();
-      float var20 = var3.getTemperature(var19.set(var4, 63, var5));
-      double var21 = Math.min(Math.abs(var7), this.icebergNoise.getValue((double)var4 * 0.1D, (double)var5 * 0.1D, false) * 15.0D);
-      if (var21 > 1.8D) {
-         double var23 = 0.09765625D;
-         double var25 = Math.abs(this.icebergRoofNoise.getValue((double)var4 * 0.09765625D, (double)var5 * 0.09765625D, false));
-         var15 = var21 * var21 * 1.2D;
-         double var27 = Math.ceil(var25 * 40.0D) + 14.0D;
-         if (var15 > var27) {
-            var15 = var27;
+   public void apply(Random var1, ChunkAccess var2, Biome var3, int var4, int var5, int var6, double var7, BlockState var9, BlockState var10, int var11, int var12, long var13, SurfaceBuilderBaseConfiguration var15) {
+      double var16 = 0.0D;
+      double var18 = 0.0D;
+      BlockPos.MutableBlockPos var20 = new BlockPos.MutableBlockPos();
+      float var21 = var3.getTemperature(var20.set(var4, 63, var5));
+      double var22 = Math.min(Math.abs(var7), this.icebergNoise.getValue((double)var4 * 0.1D, (double)var5 * 0.1D, false) * 15.0D);
+      if (var22 > 1.8D) {
+         double var24 = 0.09765625D;
+         double var26 = Math.abs(this.icebergRoofNoise.getValue((double)var4 * 0.09765625D, (double)var5 * 0.09765625D, false));
+         var16 = var22 * var22 * 1.2D;
+         double var28 = Math.ceil(var26 * 40.0D) + 14.0D;
+         if (var16 > var28) {
+            var16 = var28;
          }
 
-         if (var20 > 0.1F) {
-            var15 -= 2.0D;
+         if (var21 > 0.1F) {
+            var16 -= 2.0D;
          }
 
-         if (var15 > 2.0D) {
-            var17 = (double)var11 - var15 - 7.0D;
-            var15 += (double)var11;
+         if (var16 > 2.0D) {
+            var18 = (double)var11 - var16 - 7.0D;
+            var16 += (double)var11;
          } else {
-            var15 = 0.0D;
+            var16 = 0.0D;
          }
       }
 
-      int var37 = var4 & 15;
-      int var24 = var5 & 15;
-      SurfaceBuilderConfiguration var38 = var3.getGenerationSettings().getSurfaceBuilderConfig();
-      BlockState var26 = var38.getUnderMaterial();
-      BlockState var39 = var38.getTopMaterial();
-      BlockState var28 = var26;
-      BlockState var29 = var39;
-      int var30 = (int)(var7 / 3.0D + 3.0D + var1.nextDouble() * 0.25D);
-      int var31 = -1;
-      int var32 = 0;
-      int var33 = 2 + var1.nextInt(4);
-      int var34 = var11 + 18 + var1.nextInt(10);
+      int var38 = var4 & 15;
+      int var25 = var5 & 15;
+      SurfaceBuilderConfiguration var39 = var3.getGenerationSettings().getSurfaceBuilderConfig();
+      BlockState var27 = var39.getUnderMaterial();
+      BlockState var40 = var39.getTopMaterial();
+      BlockState var29 = var27;
+      BlockState var30 = var40;
+      int var31 = (int)(var7 / 3.0D + 3.0D + var1.nextDouble() * 0.25D);
+      int var32 = -1;
+      int var33 = 0;
+      int var34 = 2 + var1.nextInt(4);
+      int var35 = var11 + 18 + var1.nextInt(10);
 
-      for(int var35 = Math.max(var6, (int)var15 + 1); var35 >= 0; --var35) {
-         var19.set(var37, var35, var24);
-         if (var2.getBlockState(var19).isAir() && var35 < (int)var15 && var1.nextDouble() > 0.01D) {
-            var2.setBlockState(var19, PACKED_ICE, false);
-         } else if (var2.getBlockState(var19).getMaterial() == Material.WATER && var35 > (int)var17 && var35 < var11 && var17 != 0.0D && var1.nextDouble() > 0.15D) {
-            var2.setBlockState(var19, PACKED_ICE, false);
+      for(int var36 = Math.max(var6, (int)var16 + 1); var36 >= var12; --var36) {
+         var20.set(var38, var36, var25);
+         if (var2.getBlockState(var20).isAir() && var36 < (int)var16 && var1.nextDouble() > 0.01D) {
+            var2.setBlockState(var20, PACKED_ICE, false);
+         } else if (var2.getBlockState(var20).getMaterial() == Material.WATER && var36 > (int)var18 && var36 < var11 && var18 != 0.0D && var1.nextDouble() > 0.15D) {
+            var2.setBlockState(var20, PACKED_ICE, false);
          }
 
-         BlockState var36 = var2.getBlockState(var19);
-         if (var36.isAir()) {
-            var31 = -1;
-         } else if (!var36.is(var9.getBlock())) {
-            if (var36.is(Blocks.PACKED_ICE) && var32 <= var33 && var35 > var34) {
-               var2.setBlockState(var19, SNOW_BLOCK, false);
-               ++var32;
+         BlockState var37 = var2.getBlockState(var20);
+         if (var37.isAir()) {
+            var32 = -1;
+         } else if (!var37.is(var9.getBlock())) {
+            if (var37.is(Blocks.PACKED_ICE) && var33 <= var34 && var36 > var35) {
+               var2.setBlockState(var20, SNOW_BLOCK, false);
+               ++var33;
             }
-         } else if (var31 == -1) {
-            if (var30 <= 0) {
-               var29 = AIR;
-               var28 = var9;
-            } else if (var35 >= var11 - 4 && var35 <= var11 + 1) {
-               var29 = var39;
-               var28 = var26;
+         } else if (var32 == -1) {
+            if (var31 <= 0) {
+               var30 = AIR;
+               var29 = var9;
+            } else if (var36 >= var11 - 4 && var36 <= var11 + 1) {
+               var30 = var40;
+               var29 = var27;
             }
 
-            if (var35 < var11 && (var29 == null || var29.isAir())) {
-               if (var3.getTemperature(var19.set(var4, var35, var5)) < 0.15F) {
-                  var29 = ICE;
+            if (var36 < var11 && (var30 == null || var30.isAir())) {
+               if (var3.getTemperature(var20.set(var4, var36, var5)) < 0.15F) {
+                  var30 = ICE;
                } else {
-                  var29 = var10;
+                  var30 = var10;
                }
             }
 
-            var31 = var30;
-            if (var35 >= var11 - 1) {
-               var2.setBlockState(var19, var29, false);
-            } else if (var35 < var11 - 7 - var30) {
-               var29 = AIR;
-               var28 = var9;
-               var2.setBlockState(var19, GRAVEL, false);
+            var32 = var31;
+            if (var36 >= var11 - 1) {
+               var2.setBlockState(var20, var30, false);
+            } else if (var36 < var11 - 7 - var31) {
+               var30 = AIR;
+               var29 = var9;
+               var2.setBlockState(var20, GRAVEL, false);
             } else {
-               var2.setBlockState(var19, var28, false);
+               var2.setBlockState(var20, var29, false);
             }
-         } else if (var31 > 0) {
-            --var31;
-            var2.setBlockState(var19, var28, false);
-            if (var31 == 0 && var28.is(Blocks.SAND) && var30 > 1) {
-               var31 = var1.nextInt(4) + Math.max(0, var35 - 63);
-               var28 = var28.is(Blocks.RED_SAND) ? Blocks.RED_SANDSTONE.defaultBlockState() : Blocks.SANDSTONE.defaultBlockState();
+         } else if (var32 > 0) {
+            --var32;
+            var2.setBlockState(var20, var29, false);
+            if (var32 == 0 && var29.is(Blocks.SAND) && var31 > 1) {
+               var32 = var1.nextInt(4) + Math.max(0, var36 - 63);
+               var29 = var29.is(Blocks.RED_SAND) ? Blocks.RED_SANDSTONE.defaultBlockState() : Blocks.SANDSTONE.defaultBlockState();
             }
          }
       }
 
    }
