--- net/minecraft/world/level/levelgen/surfacebuilders/NetherSurfaceBuilder.java
+++ net/minecraft/world/level/levelgen/surfacebuilders/NetherSurfaceBuilder.java
@@ -20,61 +20,61 @@
 
    public NetherSurfaceBuilder(Codec<SurfaceBuilderBaseConfiguration> var1) {
       super(var1);
    }
 
-   public void apply(Random var1, ChunkAccess var2, Biome var3, int var4, int var5, int var6, double var7, BlockState var9, BlockState var10, int var11, long var12, SurfaceBuilderBaseConfiguration var14) {
-      int var15 = var11;
-      int var16 = var4 & 15;
-      int var17 = var5 & 15;
-      double var18 = 0.03125D;
-      boolean var20 = this.decorationNoise.getValue((double)var4 * 0.03125D, (double)var5 * 0.03125D, 0.0D) * 75.0D + var1.nextDouble() > 0.0D;
-      boolean var21 = this.decorationNoise.getValue((double)var4 * 0.03125D, 109.0D, (double)var5 * 0.03125D) * 75.0D + var1.nextDouble() > 0.0D;
-      int var22 = (int)(var7 / 3.0D + 3.0D + var1.nextDouble() * 0.25D);
-      BlockPos.MutableBlockPos var23 = new BlockPos.MutableBlockPos();
-      int var24 = -1;
-      BlockState var25 = var14.getTopMaterial();
-      BlockState var26 = var14.getUnderMaterial();
+   public void apply(Random var1, ChunkAccess var2, Biome var3, int var4, int var5, int var6, double var7, BlockState var9, BlockState var10, int var11, int var12, long var13, SurfaceBuilderBaseConfiguration var15) {
+      int var16 = var11;
+      int var17 = var4 & 15;
+      int var18 = var5 & 15;
+      double var19 = 0.03125D;
+      boolean var21 = this.decorationNoise.getValue((double)var4 * 0.03125D, (double)var5 * 0.03125D, 0.0D) * 75.0D + var1.nextDouble() > 0.0D;
+      boolean var22 = this.decorationNoise.getValue((double)var4 * 0.03125D, 109.0D, (double)var5 * 0.03125D) * 75.0D + var1.nextDouble() > 0.0D;
+      int var23 = (int)(var7 / 3.0D + 3.0D + var1.nextDouble() * 0.25D);
+      BlockPos.MutableBlockPos var24 = new BlockPos.MutableBlockPos();
+      int var25 = -1;
+      BlockState var26 = var15.getTopMaterial();
+      BlockState var27 = var15.getUnderMaterial();
 
-      for(int var27 = 127; var27 >= 0; --var27) {
-         var23.set(var16, var27, var17);
-         BlockState var28 = var2.getBlockState(var23);
-         if (var28.isAir()) {
-            var24 = -1;
-         } else if (var28.is(var9.getBlock())) {
-            if (var24 == -1) {
-               boolean var29 = false;
-               if (var22 <= 0) {
-                  var29 = true;
-                  var26 = var14.getUnderMaterial();
-               } else if (var27 >= var15 - 4 && var27 <= var15 + 1) {
-                  var25 = var14.getTopMaterial();
-                  var26 = var14.getUnderMaterial();
-                  if (var21) {
-                     var25 = GRAVEL;
-                     var26 = var14.getUnderMaterial();
+      for(int var28 = 127; var28 >= var12; --var28) {
+         var24.set(var17, var28, var18);
+         BlockState var29 = var2.getBlockState(var24);
+         if (var29.isAir()) {
+            var25 = -1;
+         } else if (var29.is(var9.getBlock())) {
+            if (var25 == -1) {
+               boolean var30 = false;
+               if (var23 <= 0) {
+                  var30 = true;
+                  var27 = var15.getUnderMaterial();
+               } else if (var28 >= var16 - 4 && var28 <= var16 + 1) {
+                  var26 = var15.getTopMaterial();
+                  var27 = var15.getUnderMaterial();
+                  if (var22) {
+                     var26 = GRAVEL;
+                     var27 = var15.getUnderMaterial();
                   }
 
-                  if (var20) {
-                     var25 = SOUL_SAND;
+                  if (var21) {
                      var26 = SOUL_SAND;
+                     var27 = SOUL_SAND;
                   }
                }
 
-               if (var27 < var15 && var29) {
-                  var25 = var10;
+               if (var28 < var16 && var30) {
+                  var26 = var10;
                }
 
-               var24 = var22;
-               if (var27 >= var15 - 1) {
-                  var2.setBlockState(var23, var25, false);
+               var25 = var23;
+               if (var28 >= var16 - 1) {
+                  var2.setBlockState(var24, var26, false);
                } else {
-                  var2.setBlockState(var23, var26, false);
+                  var2.setBlockState(var24, var27, false);
                }
-            } else if (var24 > 0) {
-               --var24;
-               var2.setBlockState(var23, var26, false);
+            } else if (var25 > 0) {
+               --var25;
+               var2.setBlockState(var24, var27, false);
             }
          }
       }
 
    }
