--- com/mojang/authlib/yggdrasil/YggdrasilGameProfileRepository.java
+++ com/mojang/authlib/yggdrasil/YggdrasilGameProfileRepository.java
@@ -55,27 +55,27 @@
             boolean var9 = false;
 
             try {
                ProfileSearchResultsResponse var23 = (ProfileSearchResultsResponse)this.authenticationService.makeRequest(HttpAuthenticationService.constantURL(this.searchPageUrl + var2.getName().toLowerCase()), var21, ProfileSearchResultsResponse.class);
                var22 = 0;
-               LOGGER.debug("Page {} returned {} results, parsing", 0, var23.getProfiles().length);
+               LOGGER.debug((String)"Page {} returned {} results, parsing", (int)0, (Object)var23.getProfiles().length);
                HashSet var24 = Sets.newHashSet((Iterable)var21);
                GameProfile[] var25 = var23.getProfiles();
                int var13 = var25.length;
 
                for(int var14 = 0; var14 < var13; ++var14) {
                   GameProfile var15 = var25[var14];
-                  LOGGER.debug("Successfully looked up profile {}", var15);
+                  LOGGER.debug((String)"Successfully looked up profile {}", (Object)var15);
                   var24.remove(var15.getName().toLowerCase());
                   var3.onProfileLookupSucceeded(var15);
                }
 
                Iterator var26 = var24.iterator();
 
                while(var26.hasNext()) {
                   String var27 = (String)var26.next();
-                  LOGGER.debug("Couldn't find profile {}", var27);
+                  LOGGER.debug((String)"Couldn't find profile {}", (Object)var27);
                   var3.onProfileLookupFailed(new GameProfile((UUID)null, var27), new ProfileNotFoundException("Server did not find the requested profile"));
                }
 
                try {
                   Thread.sleep(100L);
@@ -87,11 +87,11 @@
                if (var22 == 3) {
                   Iterator var11 = var21.iterator();
 
                   while(var11.hasNext()) {
                      String var12 = (String)var11.next();
-                     LOGGER.debug("Couldn't find profile {} because of a server error", var12);
+                     LOGGER.debug((String)"Couldn't find profile {} because of a server error", (Object)var12);
                      var3.onProfileLookupFailed(new GameProfile((UUID)null, var12), var10);
                   }
                } else {
                   try {
                      Thread.sleep(750L);
