--- com/mojang/authlib/yggdrasil/YggdrasilAuthenticationService.java
+++ com/mojang/authlib/yggdrasil/YggdrasilAuthenticationService.java
@@ -115,11 +115,11 @@
          throw new AuthenticationUnavailableException("Cannot contact authentication server", var7);
       }
    }
 
    public YggdrasilSocialInteractionsService createSocialInteractionsService(String var1) throws AuthenticationException {
-      return new YggdrasilSocialInteractionsService(this, var1, this.environment);
+      return new YggdrasilSocialInteractionsService(var1, this.getProxy(), this.environment);
    }
 
    private static class GameProfileSerializer implements JsonSerializer<GameProfile>, JsonDeserializer<GameProfile> {
       private GameProfileSerializer() {
       }
