--- com/mojang/authlib/yggdrasil/YggdrasilUserAuthentication.java
+++ com/mojang/authlib/yggdrasil/YggdrasilUserAuthentication.java
@@ -43,11 +43,11 @@
 
    public YggdrasilUserAuthentication(YggdrasilAuthenticationService var1, String var2, Agent var3, Environment var4) {
       super(var1);
       this.clientToken = var2;
       this.agent = var3;
-      LOGGER.info("Environment: " + var4.getName(), ". AuthHost: " + var4.getAuthHost());
+      LOGGER.info((String)("Environment: " + var4.getName()), (Object)(". AuthHost: " + var4.getAuthHost()));
       this.routeAuthenticate = HttpAuthenticationService.constantURL(var4.getAuthHost() + "/authenticate");
       this.routeRefresh = HttpAuthenticationService.constantURL(var4.getAuthHost() + "/refresh");
       this.routeValidate = HttpAuthenticationService.constantURL(var4.getAuthHost() + "/validate");
       this.routeInvalidate = HttpAuthenticationService.constantURL(var4.getAuthHost() + "/invalidate");
       this.routeSignout = HttpAuthenticationService.constantURL(var4.getAuthHost() + "/signout");
