--- net/minecraft/client/gui/screens/packs/package-info.java
+++ net/minecraft/client/gui/screens/packs/package-info.java
@@ -1,0 +1,12 @@
+package net.minecraft.client.gui.screens.packs;
+
+import com.mojang.blaze3d.FieldsAreNonnullByDefault;
+import javax.annotation.ParametersAreNonnullByDefault;
+import net.minecraft.MethodsReturnNonnullByDefault;
+
+// $FF: synthetic class
+@ParametersAreNonnullByDefault
+@MethodsReturnNonnullByDefault
+@FieldsAreNonnullByDefault
+interface package-info {
+}
