--- net/minecraft/client/gui/screens/packs/TransferableSelectionList.java
+++ net/minecraft/client/gui/screens/packs/TransferableSelectionList.java
@@ -46,11 +46,11 @@
    protected int getScrollbarPosition() {
       return this.x1 - 6;
    }
 
    public static class PackEntry extends ObjectSelectionList.Entry<TransferableSelectionList.PackEntry> {
-      private TransferableSelectionList parent;
+      private final TransferableSelectionList parent;
       protected final Minecraft minecraft;
       protected final Screen screen;
       private final PackSelectionModel.Entry pack;
       private final FormattedCharSequence nameDisplayCache;
       private final MultiLineLabel descriptionDisplayCache;
