--- com/mojang/realmsclient/util/task/CloseServerTask.java
+++ com/mojang/realmsclient/util/task/CloseServerTask.java
@@ -35,11 +35,11 @@
          } catch (RetryCallException var4) {
             if (this.aborted()) {
                return;
             }
 
-            pause(var4.delaySeconds);
+            pause((long)var4.delaySeconds);
          } catch (Exception var5) {
             if (this.aborted()) {
                return;
             }
 
