--- com/mojang/realmsclient/util/task/ConnectTask.java
+++ com/mojang/realmsclient/util/task/ConnectTask.java
@@ -2,10 +2,11 @@
 
 import com.mojang.realmsclient.dto.RealmsServer;
 import com.mojang.realmsclient.dto.RealmsServerAddress;
 import net.minecraft.client.Minecraft;
 import net.minecraft.client.gui.screens.Screen;
+import net.minecraft.client.multiplayer.resolver.ServerAddress;
 import net.minecraft.network.chat.TranslatableComponent;
 import net.minecraft.realms.RealmsConnect;
 
 public class ConnectTask extends LongRunningTask {
    private final RealmsConnect realmsConnect;
@@ -18,12 +19,11 @@
       this.realmsConnect = new RealmsConnect(var1);
    }
 
    public void run() {
       this.setTitle(new TranslatableComponent("mco.connect.connecting"));
-      net.minecraft.realms.RealmsServerAddress var1 = net.minecraft.realms.RealmsServerAddress.parseString(this.address.address);
-      this.realmsConnect.connect(this.server, var1.getHost(), var1.getPort());
+      this.realmsConnect.connect(this.server, ServerAddress.parseString(this.address.address));
    }
 
    public void abortTask() {
       this.realmsConnect.abort();
       Minecraft.getInstance().getClientPackSource().clearServerPack();
