--- com/mojang/realmsclient/util/task/DownloadTask.java
+++ com/mojang/realmsclient/util/task/DownloadTask.java
@@ -32,11 +32,11 @@
             if (this.aborted()) {
                return;
             }
 
             WorldDownload var3 = var1.requestDownloadInfo(this.worldId, this.slot);
-            pause(1);
+            pause(1L);
             if (this.aborted()) {
                return;
             }
 
             setScreen(new RealmsDownloadLatestWorldScreen(this.lastScreen, var3, this.downloadName, (var0) -> {
@@ -45,11 +45,11 @@
          } catch (RetryCallException var4) {
             if (this.aborted()) {
                return;
             }
 
-            pause(var4.delaySeconds);
+            pause((long)var4.delaySeconds);
             ++var2;
          } catch (RealmsServiceException var5) {
             if (this.aborted()) {
                return;
             }
