--- com/mojang/realmsclient/util/task/LongRunningTask.java
+++ com/mojang/realmsclient/util/task/LongRunningTask.java
@@ -7,21 +7,23 @@
 import net.minecraft.network.chat.Component;
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
 
 public abstract class LongRunningTask implements ErrorCallback, Runnable {
+   protected static final int NUMBER_OF_RETRIES = 25;
    public static final Logger LOGGER = LogManager.getLogger();
    protected RealmsLongRunningMcoTaskScreen longRunningMcoTaskScreen;
 
    public LongRunningTask() {
    }
 
-   protected static void pause(int var0) {
+   protected static void pause(long var0) {
       try {
-         Thread.sleep((long)(var0 * 1000));
-      } catch (InterruptedException var2) {
-         LOGGER.error("", var2);
+         Thread.sleep(var0 * 1000L);
+      } catch (InterruptedException var3) {
+         Thread.currentThread().interrupt();
+         LOGGER.error("", var3);
       }
 
    }
 
    public static void setScreen(Screen var0) {
