--- com/mojang/realmsclient/util/task/OpenServerTask.java
+++ com/mojang/realmsclient/util/task/OpenServerTask.java
@@ -48,11 +48,11 @@
          } catch (RetryCallException var4) {
             if (this.aborted()) {
                return;
             }
 
-            pause(var4.delaySeconds);
+            pause((long)var4.delaySeconds);
          } catch (Exception var5) {
             if (this.aborted()) {
                return;
             }
 
