--- com/mojang/realmsclient/util/task/RestoreTask.java
+++ com/mojang/realmsclient/util/task/RestoreTask.java
@@ -29,11 +29,11 @@
             if (this.aborted()) {
                return;
             }
 
             var1.restoreWorld(this.worldId, this.backup.backupId);
-            pause(1);
+            pause(1L);
             if (this.aborted()) {
                return;
             }
 
             setScreen(this.lastScreen.getNewScreen());
@@ -41,11 +41,11 @@
          } catch (RetryCallException var4) {
             if (this.aborted()) {
                return;
             }
 
-            pause(var4.delaySeconds);
+            pause((long)var4.delaySeconds);
             ++var2;
          } catch (RealmsServiceException var5) {
             if (this.aborted()) {
                return;
             }
