--- com/mojang/realmsclient/util/task/SwitchSlotTask.java
+++ com/mojang/realmsclient/util/task/SwitchSlotTask.java
@@ -32,11 +32,11 @@
          } catch (RetryCallException var4) {
             if (this.aborted()) {
                return;
             }
 
-            pause(var4.delaySeconds);
+            pause((long)var4.delaySeconds);
          } catch (Exception var5) {
             if (this.aborted()) {
                return;
             }
 
