--- net/minecraft/world/entity/ai/village/poi/PoiSection.java
+++ net/minecraft/world/entity/ai/village/poi/PoiSection.java
@@ -7,20 +7,21 @@
 import com.mojang.serialization.codecs.RecordCodecBuilder;
 import it.unimi.dsi.fastutil.shorts.Short2ObjectMap;
 import it.unimi.dsi.fastutil.shorts.Short2ObjectOpenHashMap;
 import java.util.List;
 import java.util.Map;
+import java.util.Objects;
 import java.util.Optional;
 import java.util.Set;
 import java.util.function.BiConsumer;
 import java.util.function.Consumer;
 import java.util.function.Predicate;
 import java.util.stream.Stream;
-import net.minecraft.SharedConstants;
 import net.minecraft.Util;
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.SectionPos;
+import net.minecraft.util.VisibleForDebug;
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
 import org.apache.logging.log4j.util.Supplier;
 
 public class PoiSection {
@@ -37,11 +38,11 @@
          }), PoiRecord.codec(var0).listOf().fieldOf("Records").forGetter((var0x) -> {
             return ImmutableList.copyOf(var0x.records.values());
          })).apply(var1, PoiSection::new);
       });
       Logger var10002 = LOGGER;
-      var10002.getClass();
+      Objects.requireNonNull(var10002);
       return var10000.orElseGet(Util.prefix("Failed to read POI section: ", var10002::error), () -> {
          return new PoiSection(var0, false, ImmutableList.of());
       });
    }
 
@@ -57,11 +58,11 @@
       var3.forEach(this::add);
    }
 
    public Stream<PoiRecord> getRecords(Predicate<PoiType> var1, PoiManager.Occupancy var2) {
       return this.byType.entrySet().stream().filter((var1x) -> {
-         return var1.test(var1x.getKey());
+         return var1.test((PoiType)var1x.getKey());
       }).flatMap((var0) -> {
          return ((Set)var0.getValue()).stream();
       }).filter(var2.getTest());
    }
 
@@ -85,16 +86,11 @@
       if (var5 != null) {
          if (var3.equals(var5.getPoiType())) {
             return false;
          }
 
-         String var6 = "POI data mismatch: already registered at " + var2;
-         if (SharedConstants.IS_RUNNING_IN_IDE) {
-            throw (IllegalStateException)Util.pauseInIde(new IllegalStateException(var6));
-         }
-
-         LOGGER.error(var6);
+         Util.logAndPauseIfInIde("POI data mismatch: already registered at " + var2);
       }
 
       this.records.put(var4, var1);
       ((Set)this.byType.computeIfAbsent(var3, (var0) -> {
          return Sets.newHashSet();
@@ -103,18 +99,30 @@
    }
 
    public void remove(BlockPos var1) {
       PoiRecord var2 = (PoiRecord)this.records.remove(SectionPos.sectionRelativePos(var1));
       if (var2 == null) {
-         LOGGER.error("POI data mismatch: never registered at " + var1);
+         LOGGER.error("POI data mismatch: never registered at {}", var1);
       } else {
          ((Set)this.byType.get(var2.getPoiType())).remove(var2);
-         LOGGER.debug("Removed POI of type {} @ {}", new Supplier[]{var2::getPoiType, var2::getPos});
+         Logger var10000 = LOGGER;
+         Supplier[] var10002 = new Supplier[2];
+         Objects.requireNonNull(var2);
+         var10002[0] = var2::getPoiType;
+         Objects.requireNonNull(var2);
+         var10002[1] = var2::getPos;
+         var10000.debug("Removed POI of type {} @ {}", var10002);
          this.setDirty.run();
       }
    }
 
+   @Deprecated
+   @VisibleForDebug
+   public int getFreeTickets(BlockPos var1) {
+      return (Integer)this.getPoiRecord(var1).map(PoiRecord::getFreeTickets).orElse(0);
+   }
+
    public boolean release(BlockPos var1) {
       PoiRecord var2 = (PoiRecord)this.records.get(SectionPos.sectionRelativePos(var1));
       if (var2 == null) {
          throw (IllegalStateException)Util.pauseInIde(new IllegalStateException("POI never registered at " + var1));
       } else {
@@ -123,21 +131,21 @@
          return var3;
       }
    }
 
    public boolean exists(BlockPos var1, Predicate<PoiType> var2) {
-      short var3 = SectionPos.sectionRelativePos(var1);
-      PoiRecord var4 = (PoiRecord)this.records.get(var3);
-      return var4 != null && var2.test(var4.getPoiType());
+      return this.getType(var1).filter(var2).isPresent();
    }
 
    public Optional<PoiType> getType(BlockPos var1) {
-      short var2 = SectionPos.sectionRelativePos(var1);
-      PoiRecord var3 = (PoiRecord)this.records.get(var2);
-      return var3 != null ? Optional.of(var3.getPoiType()) : Optional.empty();
+      return this.getPoiRecord(var1).map(PoiRecord::getPoiType);
    }
 
+   private Optional<PoiRecord> getPoiRecord(BlockPos var1) {
+      return Optional.ofNullable((PoiRecord)this.records.get(SectionPos.sectionRelativePos(var1)));
+   }
+
    public void refresh(Consumer<BiConsumer<BlockPos, PoiType>> var1) {
       if (!this.isValid) {
          Short2ObjectOpenHashMap var2 = new Short2ObjectOpenHashMap(this.records);
          this.clear();
          var1.accept((var2x, var3) -> {
