--- net/minecraft/client/renderer/debug/BeeDebugRenderer.java
+++ net/minecraft/client/renderer/debug/BeeDebugRenderer.java
@@ -48,20 +48,18 @@
    public void addOrUpdateBeeInfo(BeeDebugRenderer.BeeInfo var1) {
       this.beeInfosPerEntity.put(var1.uuid, var1);
    }
 
    public void render(PoseStack var1, MultiBufferSource var2, double var3, double var5, double var7) {
-      RenderSystem.pushMatrix();
       RenderSystem.enableBlend();
       RenderSystem.defaultBlendFunc();
       RenderSystem.disableTexture();
       this.clearRemovedHives();
       this.clearRemovedBees();
       this.doRender();
       RenderSystem.enableTexture();
       RenderSystem.disableBlend();
-      RenderSystem.popMatrix();
       if (!this.minecraft.player.isSpectator()) {
          this.updateLastLookedAtUuid();
       }
 
    }
