--- net/minecraft/client/renderer/debug/BrainDebugRenderer.java
+++ net/minecraft/client/renderer/debug/BrainDebugRenderer.java
@@ -28,10 +28,42 @@
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
 
 public class BrainDebugRenderer implements DebugRenderer.SimpleDebugRenderer {
    private static final Logger LOGGER = LogManager.getLogger();
+   private static final boolean SHOW_NAME_FOR_ALL = true;
+   private static final boolean SHOW_PROFESSION_FOR_ALL = false;
+   private static final boolean SHOW_BEHAVIORS_FOR_ALL = false;
+   private static final boolean SHOW_ACTIVITIES_FOR_ALL = false;
+   private static final boolean SHOW_INVENTORY_FOR_ALL = false;
+   private static final boolean SHOW_GOSSIPS_FOR_ALL = false;
+   private static final boolean SHOW_PATH_FOR_ALL = false;
+   private static final boolean SHOW_HEALTH_FOR_ALL = false;
+   private static final boolean SHOW_WANTS_GOLEM_FOR_ALL = true;
+   private static final boolean SHOW_NAME_FOR_SELECTED = true;
+   private static final boolean SHOW_PROFESSION_FOR_SELECTED = true;
+   private static final boolean SHOW_BEHAVIORS_FOR_SELECTED = true;
+   private static final boolean SHOW_ACTIVITIES_FOR_SELECTED = true;
+   private static final boolean SHOW_MEMORIES_FOR_SELECTED = true;
+   private static final boolean SHOW_INVENTORY_FOR_SELECTED = true;
+   private static final boolean SHOW_GOSSIPS_FOR_SELECTED = true;
+   private static final boolean SHOW_PATH_FOR_SELECTED = true;
+   private static final boolean SHOW_HEALTH_FOR_SELECTED = true;
+   private static final boolean SHOW_WANTS_GOLEM_FOR_SELECTED = true;
+   private static final boolean SHOW_POI_INFO = true;
+   private static final int MAX_RENDER_DIST_FOR_BRAIN_INFO = 30;
+   private static final int MAX_RENDER_DIST_FOR_POI_INFO = 30;
+   private static final int MAX_TARGETING_DIST = 8;
+   private static final float TEXT_SCALE = 0.02F;
+   private static final int WHITE = -1;
+   private static final int YELLOW = -256;
+   private static final int CYAN = -16711681;
+   private static final int GREEN = -16711936;
+   private static final int GRAY = -3355444;
+   private static final int PINK = -98404;
+   private static final int RED = -65536;
+   private static final int ORANGE = -23296;
    private final Minecraft minecraft;
    private final Map<BlockPos, BrainDebugRenderer.PoiInfo> pois = Maps.newHashMap();
    private final Map<UUID, BrainDebugRenderer.BrainDump> brainDumpsPerEntity = Maps.newHashMap();
    @Nullable
    private UUID lastLookedAtUuid;
@@ -55,40 +87,44 @@
    }
 
    public void setFreeTicketCount(BlockPos var1, int var2) {
       BrainDebugRenderer.PoiInfo var3 = (BrainDebugRenderer.PoiInfo)this.pois.get(var1);
       if (var3 == null) {
-         LOGGER.warn("Strange, setFreeTicketCount was called for an unknown POI: " + var1);
+         LOGGER.warn("Strange, setFreeTicketCount was called for an unknown POI: {}", var1);
       } else {
          var3.freeTicketCount = var2;
       }
    }
 
    public void addOrUpdateBrainDump(BrainDebugRenderer.BrainDump var1) {
       this.brainDumpsPerEntity.put(var1.uuid, var1);
    }
 
+   public void removeBrainDump(int var1) {
+      this.brainDumpsPerEntity.values().removeIf((var1x) -> {
+         return var1x.id == var1;
+      });
+   }
+
    public void render(PoseStack var1, MultiBufferSource var2, double var3, double var5, double var7) {
-      RenderSystem.pushMatrix();
       RenderSystem.enableBlend();
       RenderSystem.defaultBlendFunc();
       RenderSystem.disableTexture();
       this.clearRemovedEntities();
       this.doRender(var3, var5, var7);
       RenderSystem.enableTexture();
       RenderSystem.disableBlend();
-      RenderSystem.popMatrix();
       if (!this.minecraft.player.isSpectator()) {
          this.updateLastLookedAtUuid();
       }
 
    }
 
    private void clearRemovedEntities() {
       this.brainDumpsPerEntity.entrySet().removeIf((var1) -> {
          Entity var2 = this.minecraft.level.getEntity(((BrainDebugRenderer.BrainDump)var1.getValue()).id);
-         return var2 == null || var2.removed;
+         return var2 == null || var2.isRemoved();
       });
    }
 
    private void doRender(double var1, double var3, double var5) {
       BlockPos var7 = new BlockPos(var1, var3, var5);
@@ -131,29 +167,29 @@
    private void renderGhostPoi(BlockPos var1, List<String> var2) {
       float var3 = 0.05F;
       RenderSystem.enableBlend();
       RenderSystem.defaultBlendFunc();
       DebugRenderer.renderFilledBox(var1, 0.05F, 0.2F, 0.2F, 1.0F, 0.3F);
-      renderTextOverPos("" + var2, var1, 0, -256);
+      renderTextOverPos(var2.makeConcatWithConstants<invokedynamic>(var2), var1, 0, -256);
       renderTextOverPos("Ghost POI", var1, 1, -65536);
    }
 
    private void renderPoiInfo(BrainDebugRenderer.PoiInfo var1) {
       byte var2 = 0;
       Set var3 = this.getTicketHolderNames(var1);
       if (var3.size() < 4) {
          renderTextOverPoi("Owners: " + var3, var1, var2, -256);
       } else {
-         renderTextOverPoi("" + var3.size() + " ticket holders", var1, var2, -256);
+         renderTextOverPoi(var3.size() + " ticket holders", var1, var2, -256);
       }
 
       int var5 = var2 + 1;
       Set var4 = this.getPotentialTicketHolderNames(var1);
       if (var4.size() < 4) {
          renderTextOverPoi("Candidates: " + var4, var1, var5, -23296);
       } else {
-         renderTextOverPoi("" + var4.size() + " potential owners", var1, var5, -23296);
+         renderTextOverPoi(var4.size() + " potential owners", var1, var5, -23296);
       }
 
       ++var5;
       renderTextOverPoi("Free tickets: " + var1.freeTicketCount, var1, var5, -256);
       ++var5;
@@ -177,11 +213,13 @@
          ++var12;
       }
 
       if (var8) {
          int var10 = var1.health < var1.maxHealth ? -23296 : -1;
-         renderTextOverMob(var1.pos, var12, "health: " + String.format("%.1f", var1.health) + " / " + String.format("%.1f", var1.maxHealth), var10, 0.02F);
+         Position var10000 = var1.pos;
+         String var10002 = String.format("%.1f", var1.health);
+         renderTextOverMob(var10000, var12, "health: " + var10002 + " / " + String.format("%.1f", var1.maxHealth), var10, 0.02F);
          ++var12;
       }
 
       if (var8 && !var1.inventory.equals("")) {
          renderTextOverMob(var1.pos, var12, var1.inventory, -98404, 0.02F);
@@ -314,10 +352,22 @@
       DebugRenderer.getTargetedEntity(this.minecraft.getCameraEntity(), 8).ifPresent((var1) -> {
          this.lastLookedAtUuid = var1.getUUID();
       });
    }
 
+   public static class PoiInfo {
+      public final BlockPos pos;
+      public String type;
+      public int freeTicketCount;
+
+      public PoiInfo(BlockPos var1, String var2, int var3) {
+         this.pos = var1;
+         this.type = var2;
+         this.freeTicketCount = var3;
+      }
+   }
+
    public static class BrainDump {
       public final UUID uuid;
       public final int id;
       public final String name;
       public final String profession;
@@ -347,32 +397,20 @@
          this.inventory = var9;
          this.path = var10;
          this.wantsGolem = var11;
       }
 
-      private boolean hasPoi(BlockPos var1) {
+      boolean hasPoi(BlockPos var1) {
          Stream var10000 = this.pois.stream();
-         var1.getClass();
+         Objects.requireNonNull(var1);
          return var10000.anyMatch(var1::equals);
       }
 
-      private boolean hasPotentialPoi(BlockPos var1) {
+      boolean hasPotentialPoi(BlockPos var1) {
          return this.potentialPois.contains(var1);
       }
 
       public UUID getUuid() {
          return this.uuid;
       }
    }
-
-   public static class PoiInfo {
-      public final BlockPos pos;
-      public String type;
-      public int freeTicketCount;
-
-      public PoiInfo(BlockPos var1, String var2, int var3) {
-         this.pos = var1;
-         this.type = var2;
-         this.freeTicketCount = var3;
-      }
-   }
 }
