--- net/minecraft/client/renderer/debug/CaveDebugRenderer.java
+++ net/minecraft/client/renderer/debug/CaveDebugRenderer.java
@@ -5,10 +5,11 @@
 import com.mojang.blaze3d.systems.RenderSystem;
 import com.mojang.blaze3d.vertex.BufferBuilder;
 import com.mojang.blaze3d.vertex.DefaultVertexFormat;
 import com.mojang.blaze3d.vertex.PoseStack;
 import com.mojang.blaze3d.vertex.Tesselator;
+import com.mojang.blaze3d.vertex.VertexFormat;
 import java.util.Iterator;
 import java.util.List;
 import java.util.Map;
 import java.util.Map.Entry;
 import net.minecraft.client.renderer.LevelRenderer;
@@ -38,11 +39,11 @@
       RenderSystem.defaultBlendFunc();
       RenderSystem.disableTexture();
       BlockPos var9 = new BlockPos(var3, 0.0D, var7);
       Tesselator var10 = Tesselator.getInstance();
       BufferBuilder var11 = var10.getBuilder();
-      var11.begin(5, DefaultVertexFormat.POSITION_COLOR);
+      var11.begin(VertexFormat.Mode.TRIANGLE_STRIP, DefaultVertexFormat.POSITION_COLOR);
       Iterator var12 = this.tunnelsList.entrySet().iterator();
 
       while(var12.hasNext()) {
          Entry var13 = (Entry)var12.next();
          BlockPos var14 = (BlockPos)var13.getKey();
