--- net/minecraft/client/renderer/debug/ChunkDebugRenderer.java
+++ net/minecraft/client/renderer/debug/ChunkDebugRenderer.java
@@ -13,18 +13,19 @@
 import net.minecraft.client.Minecraft;
 import net.minecraft.client.multiplayer.ClientChunkCache;
 import net.minecraft.client.multiplayer.ClientLevel;
 import net.minecraft.client.renderer.MultiBufferSource;
 import net.minecraft.client.server.IntegratedServer;
+import net.minecraft.core.SectionPos;
 import net.minecraft.resources.ResourceKey;
 import net.minecraft.server.level.ServerChunkCache;
 import net.minecraft.server.level.ServerLevel;
 import net.minecraft.world.level.ChunkPos;
 import net.minecraft.world.level.chunk.LevelChunk;
 
 public class ChunkDebugRenderer implements DebugRenderer.SimpleDebugRenderer {
-   private final Minecraft minecraft;
+   final Minecraft minecraft;
    private double lastUpdateTime = Double.MIN_VALUE;
    private final int radius = 12;
    @Nullable
    private ChunkDebugRenderer.ChunkData data;
 
@@ -67,11 +68,11 @@
             String[] var20 = var18;
             int var21 = var18.length;
 
             for(int var22 = 0; var22 < var21; ++var22) {
                String var23 = var20[var22];
-               DebugRenderer.renderFloatingText(var23, (double)((var16.x << 4) + 8), var12 + (double)var19, (double)((var16.z << 4) + 8), -1, 0.15F);
+               DebugRenderer.renderFloatingText(var23, (double)SectionPos.sectionToBlockCoord(var16.x, 8), var12 + (double)var19, (double)SectionPos.sectionToBlockCoord(var16.z, 8), -1, 0.15F);
                var19 -= 2;
             }
          }
 
          RenderSystem.depthMask(true);
@@ -79,19 +80,19 @@
          RenderSystem.disableBlend();
       }
 
    }
 
-   final class ChunkData {
-      private final Map<ChunkPos, String> clientData;
-      private final CompletableFuture<Map<ChunkPos, String>> serverData;
+   private final class ChunkData {
+      final Map<ChunkPos, String> clientData;
+      final CompletableFuture<Map<ChunkPos, String>> serverData;
 
-      private ChunkData(IntegratedServer var2, double var3, double var5) {
+      ChunkData(IntegratedServer var2, double var3, double var5) {
          ClientLevel var7 = ChunkDebugRenderer.this.minecraft.level;
          ResourceKey var8 = var7.dimension();
-         int var9 = (int)var3 >> 4;
-         int var10 = (int)var5 >> 4;
+         int var9 = SectionPos.posToSectionCoord(var3);
+         int var10 = SectionPos.posToSectionCoord(var5);
          Builder var11 = ImmutableMap.builder();
          ClientChunkCache var12 = var7.getChunkSource();
 
          for(int var13 = var9 - 12; var13 <= var9 + 12; ++var13) {
             for(int var14 = var10 - 12; var14 <= var10 + 12; ++var14) {
@@ -120,11 +121,12 @@
                ServerChunkCache var7 = var5.getChunkSource();
 
                for(int var8x = var9 - 12; var8x <= var9 + 12; ++var8x) {
                   for(int var9x = var10 - 12; var9x <= var10 + 12; ++var9x) {
                      ChunkPos var10x = new ChunkPos(var8x, var9x);
-                     var6.put(var10x, "Server: " + var7.getChunkDebugData(var10x));
+                     String var10002 = var7.getChunkDebugData(var10x);
+                     var6.put(var10x, "Server: " + var10002);
                   }
                }
 
                return var6.build();
             }
