--- net/minecraft/client/renderer/debug/ChunkBorderRenderer.java
+++ net/minecraft/client/renderer/debug/ChunkBorderRenderer.java
@@ -3,109 +3,108 @@
 import com.mojang.blaze3d.systems.RenderSystem;
 import com.mojang.blaze3d.vertex.BufferBuilder;
 import com.mojang.blaze3d.vertex.DefaultVertexFormat;
 import com.mojang.blaze3d.vertex.PoseStack;
 import com.mojang.blaze3d.vertex.Tesselator;
+import com.mojang.blaze3d.vertex.VertexFormat;
 import net.minecraft.client.Minecraft;
 import net.minecraft.client.renderer.MultiBufferSource;
 import net.minecraft.world.entity.Entity;
+import net.minecraft.world.level.ChunkPos;
 
 public class ChunkBorderRenderer implements DebugRenderer.SimpleDebugRenderer {
    private final Minecraft minecraft;
 
    public ChunkBorderRenderer(Minecraft var1) {
       this.minecraft = var1;
    }
 
    public void render(PoseStack var1, MultiBufferSource var2, double var3, double var5, double var7) {
       RenderSystem.enableDepthTest();
-      RenderSystem.shadeModel(7425);
-      RenderSystem.enableAlphaTest();
-      RenderSystem.defaultAlphaFunc();
       Entity var9 = this.minecraft.gameRenderer.getMainCamera().getEntity();
       Tesselator var10 = Tesselator.getInstance();
       BufferBuilder var11 = var10.getBuilder();
-      double var12 = 0.0D - var5;
-      double var14 = 256.0D - var5;
+      double var12 = (double)this.minecraft.level.getMinBuildHeight() - var5;
+      double var14 = (double)this.minecraft.level.getMaxBuildHeight() - var5;
       RenderSystem.disableTexture();
       RenderSystem.disableBlend();
-      double var16 = (double)(var9.xChunk << 4) - var3;
-      double var18 = (double)(var9.zChunk << 4) - var7;
+      ChunkPos var16 = var9.chunkPosition();
+      double var17 = (double)var16.getMinBlockX() - var3;
+      double var19 = (double)var16.getMinBlockZ() - var7;
       RenderSystem.lineWidth(1.0F);
-      var11.begin(3, DefaultVertexFormat.POSITION_COLOR);
+      var11.begin(VertexFormat.Mode.DEBUG_LINE_STRIP, DefaultVertexFormat.POSITION_COLOR);
 
-      int var20;
       int var21;
-      for(var20 = -16; var20 <= 32; var20 += 16) {
-         for(var21 = -16; var21 <= 32; var21 += 16) {
-            var11.vertex(var16 + (double)var20, var12, var18 + (double)var21).color(1.0F, 0.0F, 0.0F, 0.0F).endVertex();
-            var11.vertex(var16 + (double)var20, var12, var18 + (double)var21).color(1.0F, 0.0F, 0.0F, 0.5F).endVertex();
-            var11.vertex(var16 + (double)var20, var14, var18 + (double)var21).color(1.0F, 0.0F, 0.0F, 0.5F).endVertex();
-            var11.vertex(var16 + (double)var20, var14, var18 + (double)var21).color(1.0F, 0.0F, 0.0F, 0.0F).endVertex();
+      int var22;
+      for(var21 = -16; var21 <= 32; var21 += 16) {
+         for(var22 = -16; var22 <= 32; var22 += 16) {
+            var11.vertex(var17 + (double)var21, var12, var19 + (double)var22).color(1.0F, 0.0F, 0.0F, 0.0F).endVertex();
+            var11.vertex(var17 + (double)var21, var12, var19 + (double)var22).color(1.0F, 0.0F, 0.0F, 0.5F).endVertex();
+            var11.vertex(var17 + (double)var21, var14, var19 + (double)var22).color(1.0F, 0.0F, 0.0F, 0.5F).endVertex();
+            var11.vertex(var17 + (double)var21, var14, var19 + (double)var22).color(1.0F, 0.0F, 0.0F, 0.0F).endVertex();
          }
       }
 
-      for(var20 = 2; var20 < 16; var20 += 2) {
-         var11.vertex(var16 + (double)var20, var12, var18).color(1.0F, 1.0F, 0.0F, 0.0F).endVertex();
-         var11.vertex(var16 + (double)var20, var12, var18).color(1.0F, 1.0F, 0.0F, 1.0F).endVertex();
-         var11.vertex(var16 + (double)var20, var14, var18).color(1.0F, 1.0F, 0.0F, 1.0F).endVertex();
-         var11.vertex(var16 + (double)var20, var14, var18).color(1.0F, 1.0F, 0.0F, 0.0F).endVertex();
-         var11.vertex(var16 + (double)var20, var12, var18 + 16.0D).color(1.0F, 1.0F, 0.0F, 0.0F).endVertex();
-         var11.vertex(var16 + (double)var20, var12, var18 + 16.0D).color(1.0F, 1.0F, 0.0F, 1.0F).endVertex();
-         var11.vertex(var16 + (double)var20, var14, var18 + 16.0D).color(1.0F, 1.0F, 0.0F, 1.0F).endVertex();
-         var11.vertex(var16 + (double)var20, var14, var18 + 16.0D).color(1.0F, 1.0F, 0.0F, 0.0F).endVertex();
+      for(var21 = 2; var21 < 16; var21 += 2) {
+         var11.vertex(var17 + (double)var21, var12, var19).color(1.0F, 1.0F, 0.0F, 0.0F).endVertex();
+         var11.vertex(var17 + (double)var21, var12, var19).color(1.0F, 1.0F, 0.0F, 1.0F).endVertex();
+         var11.vertex(var17 + (double)var21, var14, var19).color(1.0F, 1.0F, 0.0F, 1.0F).endVertex();
+         var11.vertex(var17 + (double)var21, var14, var19).color(1.0F, 1.0F, 0.0F, 0.0F).endVertex();
+         var11.vertex(var17 + (double)var21, var12, var19 + 16.0D).color(1.0F, 1.0F, 0.0F, 0.0F).endVertex();
+         var11.vertex(var17 + (double)var21, var12, var19 + 16.0D).color(1.0F, 1.0F, 0.0F, 1.0F).endVertex();
+         var11.vertex(var17 + (double)var21, var14, var19 + 16.0D).color(1.0F, 1.0F, 0.0F, 1.0F).endVertex();
+         var11.vertex(var17 + (double)var21, var14, var19 + 16.0D).color(1.0F, 1.0F, 0.0F, 0.0F).endVertex();
       }
 
-      for(var20 = 2; var20 < 16; var20 += 2) {
-         var11.vertex(var16, var12, var18 + (double)var20).color(1.0F, 1.0F, 0.0F, 0.0F).endVertex();
-         var11.vertex(var16, var12, var18 + (double)var20).color(1.0F, 1.0F, 0.0F, 1.0F).endVertex();
-         var11.vertex(var16, var14, var18 + (double)var20).color(1.0F, 1.0F, 0.0F, 1.0F).endVertex();
-         var11.vertex(var16, var14, var18 + (double)var20).color(1.0F, 1.0F, 0.0F, 0.0F).endVertex();
-         var11.vertex(var16 + 16.0D, var12, var18 + (double)var20).color(1.0F, 1.0F, 0.0F, 0.0F).endVertex();
-         var11.vertex(var16 + 16.0D, var12, var18 + (double)var20).color(1.0F, 1.0F, 0.0F, 1.0F).endVertex();
-         var11.vertex(var16 + 16.0D, var14, var18 + (double)var20).color(1.0F, 1.0F, 0.0F, 1.0F).endVertex();
-         var11.vertex(var16 + 16.0D, var14, var18 + (double)var20).color(1.0F, 1.0F, 0.0F, 0.0F).endVertex();
+      for(var21 = 2; var21 < 16; var21 += 2) {
+         var11.vertex(var17, var12, var19 + (double)var21).color(1.0F, 1.0F, 0.0F, 0.0F).endVertex();
+         var11.vertex(var17, var12, var19 + (double)var21).color(1.0F, 1.0F, 0.0F, 1.0F).endVertex();
+         var11.vertex(var17, var14, var19 + (double)var21).color(1.0F, 1.0F, 0.0F, 1.0F).endVertex();
+         var11.vertex(var17, var14, var19 + (double)var21).color(1.0F, 1.0F, 0.0F, 0.0F).endVertex();
+         var11.vertex(var17 + 16.0D, var12, var19 + (double)var21).color(1.0F, 1.0F, 0.0F, 0.0F).endVertex();
+         var11.vertex(var17 + 16.0D, var12, var19 + (double)var21).color(1.0F, 1.0F, 0.0F, 1.0F).endVertex();
+         var11.vertex(var17 + 16.0D, var14, var19 + (double)var21).color(1.0F, 1.0F, 0.0F, 1.0F).endVertex();
+         var11.vertex(var17 + 16.0D, var14, var19 + (double)var21).color(1.0F, 1.0F, 0.0F, 0.0F).endVertex();
       }
 
-      double var23;
-      for(var20 = 0; var20 <= 256; var20 += 2) {
-         var23 = (double)var20 - var5;
-         var11.vertex(var16, var23, var18).color(1.0F, 1.0F, 0.0F, 0.0F).endVertex();
-         var11.vertex(var16, var23, var18).color(1.0F, 1.0F, 0.0F, 1.0F).endVertex();
-         var11.vertex(var16, var23, var18 + 16.0D).color(1.0F, 1.0F, 0.0F, 1.0F).endVertex();
-         var11.vertex(var16 + 16.0D, var23, var18 + 16.0D).color(1.0F, 1.0F, 0.0F, 1.0F).endVertex();
-         var11.vertex(var16 + 16.0D, var23, var18).color(1.0F, 1.0F, 0.0F, 1.0F).endVertex();
-         var11.vertex(var16, var23, var18).color(1.0F, 1.0F, 0.0F, 1.0F).endVertex();
-         var11.vertex(var16, var23, var18).color(1.0F, 1.0F, 0.0F, 0.0F).endVertex();
+      double var24;
+      for(var21 = this.minecraft.level.getMinBuildHeight(); var21 <= this.minecraft.level.getMaxBuildHeight(); var21 += 2) {
+         var24 = (double)var21 - var5;
+         var11.vertex(var17, var24, var19).color(1.0F, 1.0F, 0.0F, 0.0F).endVertex();
+         var11.vertex(var17, var24, var19).color(1.0F, 1.0F, 0.0F, 1.0F).endVertex();
+         var11.vertex(var17, var24, var19 + 16.0D).color(1.0F, 1.0F, 0.0F, 1.0F).endVertex();
+         var11.vertex(var17 + 16.0D, var24, var19 + 16.0D).color(1.0F, 1.0F, 0.0F, 1.0F).endVertex();
+         var11.vertex(var17 + 16.0D, var24, var19).color(1.0F, 1.0F, 0.0F, 1.0F).endVertex();
+         var11.vertex(var17, var24, var19).color(1.0F, 1.0F, 0.0F, 1.0F).endVertex();
+         var11.vertex(var17, var24, var19).color(1.0F, 1.0F, 0.0F, 0.0F).endVertex();
       }
 
       var10.end();
       RenderSystem.lineWidth(2.0F);
-      var11.begin(3, DefaultVertexFormat.POSITION_COLOR);
+      var11.begin(VertexFormat.Mode.DEBUG_LINE_STRIP, DefaultVertexFormat.POSITION_COLOR);
 
-      for(var20 = 0; var20 <= 16; var20 += 16) {
-         for(var21 = 0; var21 <= 16; var21 += 16) {
-            var11.vertex(var16 + (double)var20, var12, var18 + (double)var21).color(0.25F, 0.25F, 1.0F, 0.0F).endVertex();
-            var11.vertex(var16 + (double)var20, var12, var18 + (double)var21).color(0.25F, 0.25F, 1.0F, 1.0F).endVertex();
-            var11.vertex(var16 + (double)var20, var14, var18 + (double)var21).color(0.25F, 0.25F, 1.0F, 1.0F).endVertex();
-            var11.vertex(var16 + (double)var20, var14, var18 + (double)var21).color(0.25F, 0.25F, 1.0F, 0.0F).endVertex();
+      for(var21 = 0; var21 <= 16; var21 += 16) {
+         for(var22 = 0; var22 <= 16; var22 += 16) {
+            var11.vertex(var17 + (double)var21, var12, var19 + (double)var22).color(0.25F, 0.25F, 1.0F, 0.0F).endVertex();
+            var11.vertex(var17 + (double)var21, var12, var19 + (double)var22).color(0.25F, 0.25F, 1.0F, 1.0F).endVertex();
+            var11.vertex(var17 + (double)var21, var14, var19 + (double)var22).color(0.25F, 0.25F, 1.0F, 1.0F).endVertex();
+            var11.vertex(var17 + (double)var21, var14, var19 + (double)var22).color(0.25F, 0.25F, 1.0F, 0.0F).endVertex();
          }
       }
 
-      for(var20 = 0; var20 <= 256; var20 += 16) {
-         var23 = (double)var20 - var5;
-         var11.vertex(var16, var23, var18).color(0.25F, 0.25F, 1.0F, 0.0F).endVertex();
-         var11.vertex(var16, var23, var18).color(0.25F, 0.25F, 1.0F, 1.0F).endVertex();
-         var11.vertex(var16, var23, var18 + 16.0D).color(0.25F, 0.25F, 1.0F, 1.0F).endVertex();
-         var11.vertex(var16 + 16.0D, var23, var18 + 16.0D).color(0.25F, 0.25F, 1.0F, 1.0F).endVertex();
-         var11.vertex(var16 + 16.0D, var23, var18).color(0.25F, 0.25F, 1.0F, 1.0F).endVertex();
-         var11.vertex(var16, var23, var18).color(0.25F, 0.25F, 1.0F, 1.0F).endVertex();
-         var11.vertex(var16, var23, var18).color(0.25F, 0.25F, 1.0F, 0.0F).endVertex();
+      for(var21 = this.minecraft.level.getMinBuildHeight(); var21 <= this.minecraft.level.getMaxBuildHeight(); var21 += 16) {
+         var24 = (double)var21 - var5;
+         var11.vertex(var17, var24, var19).color(0.25F, 0.25F, 1.0F, 0.0F).endVertex();
+         var11.vertex(var17, var24, var19).color(0.25F, 0.25F, 1.0F, 1.0F).endVertex();
+         var11.vertex(var17, var24, var19 + 16.0D).color(0.25F, 0.25F, 1.0F, 1.0F).endVertex();
+         var11.vertex(var17 + 16.0D, var24, var19 + 16.0D).color(0.25F, 0.25F, 1.0F, 1.0F).endVertex();
+         var11.vertex(var17 + 16.0D, var24, var19).color(0.25F, 0.25F, 1.0F, 1.0F).endVertex();
+         var11.vertex(var17, var24, var19).color(0.25F, 0.25F, 1.0F, 1.0F).endVertex();
+         var11.vertex(var17, var24, var19).color(0.25F, 0.25F, 1.0F, 0.0F).endVertex();
       }
 
       var10.end();
       RenderSystem.lineWidth(1.0F);
       RenderSystem.enableBlend();
       RenderSystem.enableTexture();
-      RenderSystem.shadeModel(7424);
    }
 }
