--- net/minecraft/client/renderer/debug/GameTestDebugRenderer.java
+++ net/minecraft/client/renderer/debug/GameTestDebugRenderer.java
@@ -8,10 +8,11 @@
 import net.minecraft.Util;
 import net.minecraft.client.renderer.MultiBufferSource;
 import net.minecraft.core.BlockPos;
 
 public class GameTestDebugRenderer implements DebugRenderer.SimpleDebugRenderer {
+   private static final float PADDING = 0.02F;
    private final Map<BlockPos, GameTestDebugRenderer.Marker> markers = Maps.newHashMap();
 
    public GameTestDebugRenderer() {
    }
 
@@ -30,14 +31,13 @@
       });
       this.markers.forEach(this::renderMarker);
    }
 
    private void renderMarker(BlockPos var1, GameTestDebugRenderer.Marker var2) {
-      RenderSystem.pushMatrix();
       RenderSystem.enableBlend();
       RenderSystem.blendFuncSeparate(GlStateManager.SourceFactor.SRC_ALPHA, GlStateManager.DestFactor.ONE_MINUS_SRC_ALPHA, GlStateManager.SourceFactor.ONE, GlStateManager.DestFactor.ZERO);
-      RenderSystem.color4f(0.0F, 1.0F, 0.0F, 0.75F);
+      RenderSystem.setShaderColor(0.0F, 1.0F, 0.0F, 0.75F);
       RenderSystem.disableTexture();
       DebugRenderer.renderFilledBox(var1, 0.02F, var2.getR(), var2.getG(), var2.getB(), var2.getA());
       if (!var2.text.isEmpty()) {
          double var3 = (double)var1.getX() + 0.5D;
          double var5 = (double)var1.getY() + 1.2D;
@@ -45,11 +45,10 @@
          DebugRenderer.renderFloatingText(var2.text, var3, var5, var7, -1, 0.01F, true, 0.0F, true);
       }
 
       RenderSystem.enableTexture();
       RenderSystem.disableBlend();
-      RenderSystem.popMatrix();
    }
 
    static class Marker {
       public int color;
       public String text;
