--- net/minecraft/client/renderer/debug/GoalSelectorDebugRenderer.java
+++ net/minecraft/client/renderer/debug/GoalSelectorDebugRenderer.java
@@ -9,10 +9,11 @@
 import net.minecraft.client.Minecraft;
 import net.minecraft.client.renderer.MultiBufferSource;
 import net.minecraft.core.BlockPos;
 
 public class GoalSelectorDebugRenderer implements DebugRenderer.SimpleDebugRenderer {
+   private static final int MAX_RENDER_DIST = 160;
    private final Minecraft minecraft;
    private final Map<Integer, List<GoalSelectorDebugRenderer.DebugGoal>> goalSelectors = Maps.newHashMap();
 
    public void clear() {
       this.goalSelectors.clear();
@@ -20,17 +21,20 @@
 
    public void addGoalSelector(int var1, List<GoalSelectorDebugRenderer.DebugGoal> var2) {
       this.goalSelectors.put(var1, var2);
    }
 
+   public void removeGoalSelector(int var1) {
+      this.goalSelectors.remove(var1);
+   }
+
    public GoalSelectorDebugRenderer(Minecraft var1) {
       this.minecraft = var1;
    }
 
    public void render(PoseStack var1, MultiBufferSource var2, double var3, double var5, double var7) {
       Camera var9 = this.minecraft.gameRenderer.getMainCamera();
-      RenderSystem.pushMatrix();
       RenderSystem.enableBlend();
       RenderSystem.defaultBlendFunc();
       RenderSystem.disableTexture();
       BlockPos var10 = new BlockPos(var9.getPosition().x, 0.0D, var9.getPosition().z);
       this.goalSelectors.forEach((var1x, var2x) -> {
@@ -46,11 +50,10 @@
          }
 
       });
       RenderSystem.enableDepthTest();
       RenderSystem.enableTexture();
-      RenderSystem.popMatrix();
    }
 
    public static class DebugGoal {
       public final BlockPos pos;
       public final int priority;
