--- net/minecraft/client/renderer/debug/LightDebugRenderer.java
+++ net/minecraft/client/renderer/debug/LightDebugRenderer.java
@@ -19,11 +19,10 @@
       this.minecraft = var1;
    }
 
    public void render(PoseStack var1, MultiBufferSource var2, double var3, double var5, double var7) {
       ClientLevel var9 = this.minecraft.level;
-      RenderSystem.pushMatrix();
       RenderSystem.enableBlend();
       RenderSystem.defaultBlendFunc();
       RenderSystem.disableTexture();
       BlockPos var10 = new BlockPos(var3, var5, var7);
       LongOpenHashSet var11 = new LongOpenHashSet();
@@ -34,17 +33,16 @@
          int var14 = var9.getBrightness(LightLayer.SKY, var13);
          float var15 = (float)(15 - var14) / 15.0F * 0.5F + 0.16F;
          int var16 = Mth.hsvToRgb(var15, 0.9F, 0.9F);
          long var17 = SectionPos.blockToSection(var13.asLong());
          if (var11.add(var17)) {
-            DebugRenderer.renderFloatingText(var9.getChunkSource().getLightEngine().getDebugData(LightLayer.SKY, SectionPos.of(var17)), (double)(SectionPos.x(var17) * 16 + 8), (double)(SectionPos.y(var17) * 16 + 8), (double)(SectionPos.z(var17) * 16 + 8), 16711680, 0.3F);
+            DebugRenderer.renderFloatingText(var9.getChunkSource().getLightEngine().getDebugData(LightLayer.SKY, SectionPos.of(var17)), (double)SectionPos.sectionToBlockCoord(SectionPos.x(var17), 8), (double)SectionPos.sectionToBlockCoord(SectionPos.y(var17), 8), (double)SectionPos.sectionToBlockCoord(SectionPos.z(var17), 8), 16711680, 0.3F);
          }
 
          if (var14 != 15) {
             DebugRenderer.renderFloatingText(String.valueOf(var14), (double)var13.getX() + 0.5D, (double)var13.getY() + 0.25D, (double)var13.getZ() + 0.5D, var16);
          }
       }
 
       RenderSystem.enableTexture();
-      RenderSystem.popMatrix();
    }
 }
