--- net/minecraft/client/renderer/debug/RaidDebugRenderer.java
+++ net/minecraft/client/renderer/debug/RaidDebugRenderer.java
@@ -8,10 +8,12 @@
 import net.minecraft.client.Minecraft;
 import net.minecraft.client.renderer.MultiBufferSource;
 import net.minecraft.core.BlockPos;
 
 public class RaidDebugRenderer implements DebugRenderer.SimpleDebugRenderer {
+   private static final int MAX_RENDER_DIST = 160;
+   private static final float TEXT_SCALE = 0.04F;
    private final Minecraft minecraft;
    private Collection<BlockPos> raidCenters = Lists.newArrayList();
 
    public RaidDebugRenderer(Minecraft var1) {
       this.minecraft = var1;
