--- net/minecraft/client/renderer/debug/StructureRenderer.java
+++ net/minecraft/client/renderer/debug/StructureRenderer.java
@@ -20,10 +20,11 @@
 public class StructureRenderer implements DebugRenderer.SimpleDebugRenderer {
    private final Minecraft minecraft;
    private final Map<DimensionType, Map<String, BoundingBox>> postMainBoxes = Maps.newIdentityHashMap();
    private final Map<DimensionType, Map<String, BoundingBox>> postPiecesBoxes = Maps.newIdentityHashMap();
    private final Map<DimensionType, Map<String, Boolean>> startPiecesMap = Maps.newIdentityHashMap();
+   private static final int MAX_RENDER_DIST = 500;
 
    public StructureRenderer(Minecraft var1) {
       this.minecraft = var1;
    }
 
@@ -38,11 +39,11 @@
          var14 = ((Map)this.postMainBoxes.get(var11)).values().iterator();
 
          while(var14.hasNext()) {
             BoundingBox var15 = (BoundingBox)var14.next();
             if (var12.closerThan(var15.getCenter(), 500.0D)) {
-               LevelRenderer.renderLineBox(var1, var13, (double)var15.x0 - var3, (double)var15.y0 - var5, (double)var15.z0 - var7, (double)(var15.x1 + 1) - var3, (double)(var15.y1 + 1) - var5, (double)(var15.z1 + 1) - var7, 1.0F, 1.0F, 1.0F, 1.0F, 1.0F, 1.0F, 1.0F);
+               LevelRenderer.renderLineBox(var1, var13, (double)var15.minX() - var3, (double)var15.minY() - var5, (double)var15.minZ() - var7, (double)(var15.maxX() + 1) - var3, (double)(var15.maxY() + 1) - var5, (double)(var15.maxZ() + 1) - var7, 1.0F, 1.0F, 1.0F, 1.0F, 1.0F, 1.0F, 1.0F);
             }
          }
       }
 
       if (this.postPiecesBoxes.containsKey(var11)) {
@@ -53,13 +54,13 @@
             String var16 = (String)var19.getKey();
             BoundingBox var17 = (BoundingBox)var19.getValue();
             Boolean var18 = (Boolean)((Map)this.startPiecesMap.get(var11)).get(var16);
             if (var12.closerThan(var17.getCenter(), 500.0D)) {
                if (var18) {
-                  LevelRenderer.renderLineBox(var1, var13, (double)var17.x0 - var3, (double)var17.y0 - var5, (double)var17.z0 - var7, (double)(var17.x1 + 1) - var3, (double)(var17.y1 + 1) - var5, (double)(var17.z1 + 1) - var7, 0.0F, 1.0F, 0.0F, 1.0F, 0.0F, 1.0F, 0.0F);
+                  LevelRenderer.renderLineBox(var1, var13, (double)var17.minX() - var3, (double)var17.minY() - var5, (double)var17.minZ() - var7, (double)(var17.maxX() + 1) - var3, (double)(var17.maxY() + 1) - var5, (double)(var17.maxZ() + 1) - var7, 0.0F, 1.0F, 0.0F, 1.0F, 0.0F, 1.0F, 0.0F);
                } else {
-                  LevelRenderer.renderLineBox(var1, var13, (double)var17.x0 - var3, (double)var17.y0 - var5, (double)var17.z0 - var7, (double)(var17.x1 + 1) - var3, (double)(var17.y1 + 1) - var5, (double)(var17.z1 + 1) - var7, 0.0F, 0.0F, 1.0F, 1.0F, 0.0F, 0.0F, 1.0F);
+                  LevelRenderer.renderLineBox(var1, var13, (double)var17.minX() - var3, (double)var17.minY() - var5, (double)var17.minZ() - var7, (double)(var17.maxX() + 1) - var3, (double)(var17.maxY() + 1) - var5, (double)(var17.maxZ() + 1) - var7, 0.0F, 0.0F, 1.0F, 1.0F, 0.0F, 0.0F, 1.0F);
                }
             }
          }
       }
 
