--- net/minecraft/client/renderer/debug/VillageSectionsDebugRenderer.java
+++ net/minecraft/client/renderer/debug/VillageSectionsDebugRenderer.java
@@ -7,10 +7,11 @@
 import net.minecraft.client.renderer.MultiBufferSource;
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.SectionPos;
 
 public class VillageSectionsDebugRenderer implements DebugRenderer.SimpleDebugRenderer {
+   private static final int MAX_RENDER_DIST_FOR_VILLAGE_SECTIONS = 60;
    private final Set<SectionPos> villageSections = Sets.newHashSet();
 
    VillageSectionsDebugRenderer() {
    }
 
@@ -25,18 +26,16 @@
    public void setNotVillageSection(SectionPos var1) {
       this.villageSections.remove(var1);
    }
 
    public void render(PoseStack var1, MultiBufferSource var2, double var3, double var5, double var7) {
-      RenderSystem.pushMatrix();
       RenderSystem.enableBlend();
       RenderSystem.defaultBlendFunc();
       RenderSystem.disableTexture();
       this.doRender(var3, var5, var7);
       RenderSystem.enableTexture();
       RenderSystem.disableBlend();
-      RenderSystem.popMatrix();
    }
 
    private void doRender(double var1, double var3, double var5) {
       BlockPos var7 = new BlockPos(var1, var3, var5);
       this.villageSections.forEach((var1x) -> {
