--- net/minecraft/client/renderer/debug/WaterDebugRenderer.java
+++ net/minecraft/client/renderer/debug/WaterDebugRenderer.java
@@ -21,11 +21,11 @@
    public void render(PoseStack var1, MultiBufferSource var2, double var3, double var5, double var7) {
       BlockPos var9 = this.minecraft.player.blockPosition();
       Level var10 = this.minecraft.player.level;
       RenderSystem.enableBlend();
       RenderSystem.defaultBlendFunc();
-      RenderSystem.color4f(0.0F, 1.0F, 0.0F, 0.75F);
+      RenderSystem.setShaderColor(0.0F, 1.0F, 0.0F, 0.75F);
       RenderSystem.disableTexture();
       RenderSystem.lineWidth(6.0F);
       Iterator var11 = BlockPos.betweenClosed(var9.offset(-10, -10, -10), var9.offset(10, 10, 10)).iterator();
 
       BlockPos var12;
