--- net/minecraft/client/renderer/debug/WorldGenAttemptRenderer.java
+++ net/minecraft/client/renderer/debug/WorldGenAttemptRenderer.java
@@ -4,10 +4,11 @@
 import com.mojang.blaze3d.systems.RenderSystem;
 import com.mojang.blaze3d.vertex.BufferBuilder;
 import com.mojang.blaze3d.vertex.DefaultVertexFormat;
 import com.mojang.blaze3d.vertex.PoseStack;
 import com.mojang.blaze3d.vertex.Tesselator;
+import com.mojang.blaze3d.vertex.VertexFormat;
 import java.util.List;
 import net.minecraft.client.renderer.LevelRenderer;
 import net.minecraft.client.renderer.MultiBufferSource;
 import net.minecraft.core.BlockPos;
 
@@ -36,11 +37,11 @@
       RenderSystem.enableBlend();
       RenderSystem.defaultBlendFunc();
       RenderSystem.disableTexture();
       Tesselator var9 = Tesselator.getInstance();
       BufferBuilder var10 = var9.getBuilder();
-      var10.begin(5, DefaultVertexFormat.POSITION_COLOR);
+      var10.begin(VertexFormat.Mode.TRIANGLE_STRIP, DefaultVertexFormat.POSITION_COLOR);
 
       for(int var11 = 0; var11 < this.toRender.size(); ++var11) {
          BlockPos var12 = (BlockPos)this.toRender.get(var11);
          Float var13 = (Float)this.scales.get(var11);
          float var14 = var13 / 2.0F;
