--- net/minecraft/util/datafix/schemas/V1483.java
+++ net/minecraft/util/datafix/schemas/V1483.java
@@ -10,9 +10,9 @@
       super(var1, var2);
    }
 
    public Map<String, Supplier<TypeTemplate>> registerEntities(Schema var1) {
       Map var2 = super.registerEntities(var1);
-      var2.put("minecraft:pufferfish", var2.remove("minecraft:puffer_fish"));
+      var2.put("minecraft:pufferfish", (Supplier)var2.remove("minecraft:puffer_fish"));
       return var2;
    }
 }
