--- net/minecraft/util/datafix/schemas/V704.java
+++ net/minecraft/util/datafix/schemas/V704.java
@@ -26,10 +26,11 @@
       var0.put("minecraft:jukebox", "minecraft:jukebox");
       var0.put("minecraft:dispenser", "minecraft:dispenser");
       var0.put("minecraft:dropper", "minecraft:dropper");
       var0.put("minecraft:sign", "minecraft:sign");
       var0.put("minecraft:mob_spawner", "minecraft:mob_spawner");
+      var0.put("minecraft:spawner", "minecraft:mob_spawner");
       var0.put("minecraft:noteblock", "minecraft:noteblock");
       var0.put("minecraft:brewing_stand", "minecraft:brewing_stand");
       var0.put("minecraft:enhanting_table", "minecraft:enchanting_table");
       var0.put("minecraft:command_block", "minecraft:command_block");
       var0.put("minecraft:beacon", "minecraft:beacon");
