--- net/minecraft/util/datafix/schemas/V99.java
+++ net/minecraft/util/datafix/schemas/V99.java
@@ -8,19 +8,18 @@
 import com.mojang.datafixers.types.templates.Hook.HookFunction;
 import com.mojang.serialization.Dynamic;
 import com.mojang.serialization.DynamicOps;
 import java.util.HashMap;
 import java.util.Map;
-import java.util.Objects;
 import java.util.function.Supplier;
 import net.minecraft.util.datafix.fixes.References;
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
 
 public class V99 extends Schema {
    private static final Logger LOGGER = LogManager.getLogger();
-   private static final Map<String, String> ITEM_TO_BLOCKENTITY = (Map)DataFixUtils.make(Maps.newHashMap(), (var0) -> {
+   static final Map<String, String> ITEM_TO_BLOCKENTITY = (Map)DataFixUtils.make(Maps.newHashMap(), (var0) -> {
       var0.put("minecraft:furnace", "Furnace");
       var0.put("minecraft:lit_furnace", "Furnace");
       var0.put("minecraft:chest", "Chest");
       var0.put("minecraft:trapped_chest", "Chest");
       var0.put("minecraft:ender_chest", "EnderChest");
@@ -246,11 +245,11 @@
       });
       var1.registerType(true, References.ENTITY, () -> {
          return DSL.taggedChoiceLazy("id", DSL.string(), var2);
       });
       var1.registerType(true, References.ITEM_STACK, () -> {
-         return DSL.hook(DSL.optionalFields("id", DSL.or(DSL.constType(DSL.intType()), References.ITEM_NAME.in(var1)), "tag", DSL.optionalFields("EntityTag", References.ENTITY_TREE.in(var1), "BlockEntityTag", References.BLOCK_ENTITY.in(var1), "CanDestroy", DSL.list(References.BLOCK_NAME.in(var1)), "CanPlaceOn", DSL.list(References.BLOCK_NAME.in(var1)))), ADD_NAMES, HookFunction.IDENTITY);
+         return DSL.hook(DSL.optionalFields("id", DSL.or(DSL.constType(DSL.intType()), References.ITEM_NAME.in(var1)), "tag", DSL.optionalFields("EntityTag", References.ENTITY_TREE.in(var1), "BlockEntityTag", References.BLOCK_ENTITY.in(var1), "CanDestroy", DSL.list(References.BLOCK_NAME.in(var1)), "CanPlaceOn", DSL.list(References.BLOCK_NAME.in(var1)), "Items", DSL.list(References.ITEM_STACK.in(var1)))), ADD_NAMES, HookFunction.IDENTITY);
       });
       var1.registerType(false, References.OPTIONS, DSL::remainder);
       var1.registerType(false, References.BLOCK_NAME, () -> {
          return DSL.or(DSL.constType(DSL.intType()), DSL.constType(NamespacedSchema.namespacedString()));
       });
@@ -265,25 +264,31 @@
       var1.registerType(false, References.OBJECTIVE, DSL::remainder);
       var1.registerType(false, References.TEAM, DSL::remainder);
       var1.registerType(true, References.UNTAGGED_SPAWNER, DSL::remainder);
       var1.registerType(false, References.POI_CHUNK, DSL::remainder);
       var1.registerType(true, References.WORLD_GEN_SETTINGS, DSL::remainder);
+      var1.registerType(false, References.ENTITY_CHUNK, () -> {
+         return DSL.optionalFields("Entities", DSL.list(References.ENTITY_TREE.in(var1)));
+      });
    }
 
    protected static <T> T addNames(Dynamic<T> var0, Map<String, String> var1, String var2) {
       return var0.update("tag", (var3) -> {
          return var3.update("BlockEntityTag", (var2x) -> {
-            String var3 = var0.get("id").asString("");
-            String var4 = (String)var1.get(NamespacedSchema.ensureNamespaced(var3));
-            if (var4 == null) {
+            String var3 = (String)var0.get("id").asString().result().map(NamespacedSchema::ensureNamespaced).orElse("minecraft:air");
+            if (!"minecraft:air".equals(var3)) {
+               String var4 = (String)var1.get(var3);
+               if (var4 != null) {
+                  return var2x.set("id", var0.createString(var4));
+               }
+
                LOGGER.warn("Unable to resolve BlockEntity for ItemStack: {}", var3);
-               return var2x;
-            } else {
-               return var2x.set("id", var0.createString(var4));
             }
+
+            return var2x;
          }).update("EntityTag", (var2x) -> {
             String var3 = var0.get("id").asString("");
-            return Objects.equals(NamespacedSchema.ensureNamespaced(var3), "minecraft:armor_stand") ? var2x.set("id", var0.createString(var2)) : var2x;
+            return "minecraft:armor_stand".equals(NamespacedSchema.ensureNamespaced(var3)) ? var2x.set("id", var0.createString(var2)) : var2x;
          });
       }).getValue();
    }
 }
