--- net/minecraft/world/item/context/UseOnContext.java
+++ net/minecraft/world/item/context/UseOnContext.java
@@ -74,8 +74,8 @@
    public boolean isSecondaryUseActive() {
       return this.player != null && this.player.isSecondaryUseActive();
    }
 
    public float getRotation() {
-      return this.player == null ? 0.0F : this.player.yRot;
+      return this.player == null ? 0.0F : this.player.getYRot();
    }
 }
