--- net/minecraft/server/packs/repository/PackCompatibility.java
+++ net/minecraft/server/packs/repository/PackCompatibility.java
@@ -2,10 +2,12 @@
 
 import net.minecraft.ChatFormatting;
 import net.minecraft.SharedConstants;
 import net.minecraft.network.chat.Component;
 import net.minecraft.network.chat.TranslatableComponent;
+import net.minecraft.server.packs.PackType;
+import net.minecraft.server.packs.metadata.pack.PackMetadataSection;
 
 public enum PackCompatibility {
    TOO_OLD("old"),
    TOO_NEW("new"),
    COMPATIBLE("compatible");
@@ -20,13 +22,18 @@
 
    public boolean isCompatible() {
       return this == COMPATIBLE;
    }
 
-   public static PackCompatibility forFormat(int var0) {
-      if (var0 < SharedConstants.getCurrentVersion().getPackVersion()) {
+   public static PackCompatibility forFormat(int var0, PackType var1) {
+      int var2 = var1.getVersion(SharedConstants.getCurrentVersion());
+      if (var0 < var2) {
          return TOO_OLD;
       } else {
-         return var0 > SharedConstants.getCurrentVersion().getPackVersion() ? TOO_NEW : COMPATIBLE;
+         return var0 > var2 ? TOO_NEW : COMPATIBLE;
       }
    }
+
+   public static PackCompatibility forMetadata(PackMetadataSection var0, PackType var1) {
+      return forFormat(var0.getPackFormat(), var1);
+   }
 }
