--- net/minecraft/server/packs/repository/package-info.java
+++ net/minecraft/server/packs/repository/package-info.java
@@ -1,0 +1,10 @@
+package net.minecraft.server.packs.repository;
+
+import javax.annotation.ParametersAreNonnullByDefault;
+import net.minecraft.MethodsReturnNonnullByDefault;
+
+// $FF: synthetic class
+@ParametersAreNonnullByDefault
+@MethodsReturnNonnullByDefault
+interface package-info {
+}
