--- net/minecraft/client/gui/screens/recipebook/GhostRecipe.java
+++ net/minecraft/client/gui/screens/recipebook/GhostRecipe.java
@@ -96,9 +96,9 @@
          return this.y;
       }
 
       public ItemStack getItem() {
          ItemStack[] var1 = this.ingredient.getItems();
-         return var1[Mth.floor(GhostRecipe.this.time / 30.0F) % var1.length];
+         return var1.length == 0 ? ItemStack.EMPTY : var1[Mth.floor(GhostRecipe.this.time / 30.0F) % var1.length];
       }
    }
 }
