--- net/minecraft/client/gui/screens/recipebook/RecipeBookPage.java
+++ net/minecraft/client/gui/screens/recipebook/RecipeBookPage.java
@@ -9,10 +9,11 @@
 import net.minecraft.client.gui.components.StateSwitchingButton;
 import net.minecraft.stats.RecipeBook;
 import net.minecraft.world.item.crafting.Recipe;
 
 public class RecipeBookPage {
+   public static final int ITEMS_PER_PAGE = 20;
    private final List<RecipeButton> buttons = Lists.newArrayListWithCapacity(20);
    private RecipeButton hoveredButton;
    private final OverlayRecipeComponent overlay = new OverlayRecipeComponent();
    private Minecraft minecraft;
    private final List<RecipeShownListener> showListeners = Lists.newArrayList();
