--- net/minecraft/client/gui/screens/recipebook/RecipeBookComponent.java
+++ net/minecraft/client/gui/screens/recipebook/RecipeBookComponent.java
@@ -67,11 +67,11 @@
       this.width = var1;
       this.height = var2;
       this.menu = var5;
       var3.player.containerMenu = var5;
       this.book = var3.player.getRecipeBook();
-      this.timesInventoryChanged = var3.player.inventory.getTimesChanged();
+      this.timesInventoryChanged = var3.player.getInventory().getTimesChanged();
       if (this.isVisible()) {
          this.initVisuals(var4);
       }
 
       var3.keyboardHandler.setSendRepeatsToGui(true);
@@ -80,11 +80,11 @@
    public void initVisuals(boolean var1) {
       this.xOffset = var1 ? 0 : 86;
       int var2 = (this.width - 147) / 2 - this.xOffset;
       int var3 = (this.height - 166) / 2;
       this.stackedContents.clear();
-      this.minecraft.player.inventory.fillStackedContents(this.stackedContents);
+      this.minecraft.player.getInventory().fillStackedContents(this.stackedContents);
       this.menu.fillCraftSlotsStackedContents(this.stackedContents);
       String var4 = this.searchBox != null ? this.searchBox.getValue() : "";
       Font var10003 = this.minecraft.font;
       int var10004 = var2 + 25;
       int var10005 = var3 + 14;
@@ -229,22 +229,22 @@
       }
    }
 
    public void tick() {
       if (this.isVisible()) {
-         if (this.timesInventoryChanged != this.minecraft.player.inventory.getTimesChanged()) {
+         if (this.timesInventoryChanged != this.minecraft.player.getInventory().getTimesChanged()) {
             this.updateStackedContents();
-            this.timesInventoryChanged = this.minecraft.player.inventory.getTimesChanged();
+            this.timesInventoryChanged = this.minecraft.player.getInventory().getTimesChanged();
          }
 
          this.searchBox.tick();
       }
    }
 
    private void updateStackedContents() {
       this.stackedContents.clear();
-      this.minecraft.player.inventory.fillStackedContents(this.stackedContents);
+      this.minecraft.player.getInventory().fillStackedContents(this.stackedContents);
       this.menu.fillCraftSlotsStackedContents(this.stackedContents);
       this.updateCollections(false);
    }
 
    public void render(PoseStack var1, int var2, int var3, float var4) {
