--- net/minecraft/client/gui/screens/recipebook/RecipeBookPage.java
+++ net/minecraft/client/gui/screens/recipebook/RecipeBookPage.java
@@ -2,17 +2,20 @@
 
 import com.google.common.collect.Lists;
 import com.mojang.blaze3d.vertex.PoseStack;
 import java.util.Iterator;
 import java.util.List;
+import java.util.function.Consumer;
 import javax.annotation.Nullable;
 import net.minecraft.client.Minecraft;
+import net.minecraft.client.gui.components.AbstractWidget;
 import net.minecraft.client.gui.components.StateSwitchingButton;
 import net.minecraft.stats.RecipeBook;
 import net.minecraft.world.item.crafting.Recipe;
 
 public class RecipeBookPage {
+   public static final int ITEMS_PER_PAGE = 20;
    private final List<RecipeButton> buttons = Lists.newArrayListWithCapacity(20);
    private RecipeButton hoveredButton;
    private final OverlayRecipeComponent overlay = new OverlayRecipeComponent();
    private Minecraft minecraft;
    private final List<RecipeShownListener> showListeners = Lists.newArrayList();
@@ -83,11 +86,12 @@
       this.backButton.visible = this.totalPages > 1 && this.currentPage > 0;
    }
 
    public void render(PoseStack var1, int var2, int var3, int var4, int var5, float var6) {
       if (this.totalPages > 1) {
-         String var7 = this.currentPage + 1 + "/" + this.totalPages;
+         int var10000 = this.currentPage + 1;
+         String var7 = var10000 + "/" + this.totalPages;
          int var8 = this.minecraft.font.width(var7);
          this.minecraft.font.draw(var1, (String)var7, (float)(var2 - var8 / 2 + 73), (float)(var3 + 141), -1);
       }
 
       this.hoveredButton = null;
@@ -185,6 +189,12 @@
    }
 
    public RecipeBook getRecipeBook() {
       return this.recipeBook;
    }
+
+   protected void listButtons(Consumer<AbstractWidget> var1) {
+      var1.accept(this.forwardButton);
+      var1.accept(this.backButton);
+      this.buttons.forEach(var1);
+   }
 }
