--- net/minecraft/client/gui/screens/recipebook/RecipeBookTabButton.java
+++ net/minecraft/client/gui/screens/recipebook/RecipeBookTabButton.java
@@ -6,10 +6,11 @@
 import java.util.List;
 import net.minecraft.client.ClientRecipeBook;
 import net.minecraft.client.Minecraft;
 import net.minecraft.client.RecipeBookCategories;
 import net.minecraft.client.gui.components.StateSwitchingButton;
+import net.minecraft.client.renderer.GameRenderer;
 import net.minecraft.client.renderer.entity.ItemRenderer;
 import net.minecraft.world.inventory.RecipeBookMenu;
 import net.minecraft.world.item.ItemStack;
 import net.minecraft.world.item.crafting.Recipe;
 
@@ -46,18 +47,19 @@
    }
 
    public void renderButton(PoseStack var1, int var2, int var3, float var4) {
       if (this.animationTime > 0.0F) {
          float var5 = 1.0F + 0.1F * (float)Math.sin((double)(this.animationTime / 15.0F * 3.1415927F));
-         RenderSystem.pushMatrix();
-         RenderSystem.translatef((float)(this.x + 8), (float)(this.y + 12), 0.0F);
-         RenderSystem.scalef(1.0F, var5, 1.0F);
-         RenderSystem.translatef((float)(-(this.x + 8)), (float)(-(this.y + 12)), 0.0F);
+         var1.pushPose();
+         var1.translate((double)(this.x + 8), (double)(this.y + 12), 0.0D);
+         var1.scale(1.0F, var5, 1.0F);
+         var1.translate((double)(-(this.x + 8)), (double)(-(this.y + 12)), 0.0D);
       }
 
       Minecraft var9 = Minecraft.getInstance();
-      var9.getTextureManager().bind(this.resourceLocation);
+      RenderSystem.setShader(GameRenderer::getPositionTexShader);
+      RenderSystem.setShaderTexture(0, this.resourceLocation);
       RenderSystem.disableDepthTest();
       int var6 = this.xTexStart;
       int var7 = this.yTexStart;
       if (this.isStateTriggered) {
          var6 += this.xDiffTex;
@@ -70,16 +72,16 @@
       int var8 = this.x;
       if (this.isStateTriggered) {
          var8 -= 2;
       }
 
-      RenderSystem.color4f(1.0F, 1.0F, 1.0F, 1.0F);
+      RenderSystem.setShaderColor(1.0F, 1.0F, 1.0F, 1.0F);
       this.blit(var1, var8, this.y, var6, var7, this.width, this.height);
       RenderSystem.enableDepthTest();
       this.renderIcon(var9.getItemRenderer());
       if (this.animationTime > 0.0F) {
-         RenderSystem.popMatrix();
+         var1.popPose();
          this.animationTime -= var4;
       }
 
    }
 
