--- net/minecraft/commands/Commands.java
+++ net/minecraft/commands/Commands.java
@@ -257,11 +257,11 @@
          var1.sendFailure((new TranslatableComponent("command.failed")).withStyle((var1x) -> {
             return var1x.withHoverEvent(new HoverEvent(HoverEvent.Action.SHOW_TEXT, var5));
          }));
          if (SharedConstants.IS_RUNNING_IN_IDE) {
             var1.sendFailure(new TextComponent(Util.describeError(var15)));
-            LOGGER.error((String)("'" + var2 + "' threw an exception"), (Throwable)var15);
+            LOGGER.error((String)"'{}' threw an exception", (Object)var2, (Object)var15);
          }
 
          var19 = 0;
       } finally {
          var1.getServer().getProfiler().pop();
