--- net/minecraft/commands/SharedSuggestionProvider.java
+++ net/minecraft/commands/SharedSuggestionProvider.java
@@ -234,10 +234,25 @@
       }
 
       return var1.buildFuture();
    }
 
+   static <T> CompletableFuture<Suggestions> suggest(Iterable<T> var0, SuggestionsBuilder var1, Function<T, String> var2, Function<T, Message> var3) {
+      String var4 = var1.getRemaining().toLowerCase(Locale.ROOT);
+      Iterator var5 = var0.iterator();
+
+      while(var5.hasNext()) {
+         Object var6 = var5.next();
+         String var7 = (String)var2.apply(var6);
+         if (matchesSubStr(var4, var7.toLowerCase(Locale.ROOT))) {
+            var1.suggest(var7, (Message)var3.apply(var6));
+         }
+      }
+
+      return var1.buildFuture();
+   }
+
    static boolean matchesSubStr(String var0, String var1) {
       for(int var2 = 0; !var1.startsWith(var0, var2); ++var2) {
          var2 = var1.indexOf(95, var2);
          if (var2 < 0) {
             return false;
