--- net/minecraft/world/level/pathfinder/PathFinder.java
+++ net/minecraft/world/level/pathfinder/PathFinder.java
@@ -16,13 +16,15 @@
 import net.minecraft.core.BlockPos;
 import net.minecraft.world.entity.Mob;
 import net.minecraft.world.level.PathNavigationRegion;
 
 public class PathFinder {
+   private static final float FUDGING = 1.5F;
    private final Node[] neighbors = new Node[32];
    private final int maxVisitedNodes;
    private final NodeEvaluator nodeEvaluator;
+   private static final boolean DEBUG = false;
    private final BinaryHeap openSet = new BinaryHeap();
 
    public PathFinder(NodeEvaluator var1, int var2) {
       this.nodeEvaluator = var1;
       this.maxVisitedNodes = var2;
@@ -116,11 +118,11 @@
       float var3 = Float.MAX_VALUE;
 
       float var6;
       for(Iterator var4 = var2.iterator(); var4.hasNext(); var3 = Math.min(var6, var3)) {
          Target var5 = (Target)var4.next();
-         var6 = var1.distanceTo(var5);
+         var6 = var1.distanceTo((Node)var5);
          var5.updateBest(var6, var1);
       }
 
       return var3;
    }
