--- net/minecraft/world/level/pathfinder/Target.java
+++ net/minecraft/world/level/pathfinder/Target.java
@@ -29,10 +29,14 @@
 
    public void setReached() {
       this.reached = true;
    }
 
+   public boolean isReached() {
+      return this.reached;
+   }
+
    public static Target createFromStream(FriendlyByteBuf var0) {
       Target var1 = new Target(var0.readInt(), var0.readInt(), var0.readInt());
       var1.walkedDistance = var0.readFloat();
       var1.costMalus = var0.readFloat();
       var1.closed = var0.readBoolean();
