--- net/minecraft/world/level/levelgen/feature/trunkplacers/FancyTrunkPlacer.java
+++ net/minecraft/world/level/levelgen/feature/trunkplacers/FancyTrunkPlacer.java
@@ -36,14 +36,11 @@
 
    public List<FoliagePlacer.FoliageAttachment> placeTrunk(LevelSimulatedRW var1, Random var2, int var3, BlockPos var4, Set<BlockPos> var5, BoundingBox var6, TreeConfiguration var7) {
       boolean var8 = true;
       int var9 = var3 + 2;
       int var10 = Mth.floor((double)var9 * 0.618D);
-      if (!var7.fromSapling) {
-         setDirtAt(var1, var4.below());
-      }
-
+      setDirtAt(var1, var2, var4.below(), var7);
       double var11 = 1.0D;
       int var13 = Math.min(1, Mth.floor(1.382D + Math.pow(1.0D * (double)var9 / 13.0D, 2.0D)));
       int var14 = var4.getY() + var10;
       int var15 = var9 - 5;
       ArrayList var16 = Lists.newArrayList();
