--- net/minecraft/world/level/levelgen/feature/trunkplacers/TrunkPlacerType.java
+++ net/minecraft/world/level/levelgen/feature/trunkplacers/TrunkPlacerType.java
@@ -8,10 +8,11 @@
    public static final TrunkPlacerType<ForkingTrunkPlacer> FORKING_TRUNK_PLACER;
    public static final TrunkPlacerType<GiantTrunkPlacer> GIANT_TRUNK_PLACER;
    public static final TrunkPlacerType<MegaJungleTrunkPlacer> MEGA_JUNGLE_TRUNK_PLACER;
    public static final TrunkPlacerType<DarkOakTrunkPlacer> DARK_OAK_TRUNK_PLACER;
    public static final TrunkPlacerType<FancyTrunkPlacer> FANCY_TRUNK_PLACER;
+   public static final TrunkPlacerType<BendingTrunkPlacer> BENDING_TRUNK_PLACER;
    private final Codec<P> codec;
 
    private static <P extends TrunkPlacer> TrunkPlacerType<P> register(String var0, Codec<P> var1) {
       return (TrunkPlacerType)Registry.register(Registry.TRUNK_PLACER_TYPES, (String)var0, new TrunkPlacerType(var1));
    }
@@ -29,7 +30,8 @@
       FORKING_TRUNK_PLACER = register("forking_trunk_placer", ForkingTrunkPlacer.CODEC);
       GIANT_TRUNK_PLACER = register("giant_trunk_placer", GiantTrunkPlacer.CODEC);
       MEGA_JUNGLE_TRUNK_PLACER = register("mega_jungle_trunk_placer", MegaJungleTrunkPlacer.CODEC);
       DARK_OAK_TRUNK_PLACER = register("dark_oak_trunk_placer", DarkOakTrunkPlacer.CODEC);
       FANCY_TRUNK_PLACER = register("fancy_trunk_placer", FancyTrunkPlacer.CODEC);
+      BENDING_TRUNK_PLACER = register("bending_trunk_placer", BendingTrunkPlacer.CODEC);
    }
 }
