--- net/minecraft/world/level/levelgen/feature/trunkplacers/TrunkPlacer.java
+++ net/minecraft/world/level/levelgen/feature/trunkplacers/TrunkPlacer.java
@@ -9,11 +9,10 @@
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.Registry;
 import net.minecraft.world.level.LevelSimulatedRW;
 import net.minecraft.world.level.LevelSimulatedReader;
 import net.minecraft.world.level.LevelWriter;
-import net.minecraft.world.level.block.Block;
 import net.minecraft.world.level.block.Blocks;
 import net.minecraft.world.level.block.state.BlockState;
 import net.minecraft.world.level.levelgen.feature.Feature;
 import net.minecraft.world.level.levelgen.feature.TreeFeature;
 import net.minecraft.world.level.levelgen.feature.configurations.TreeConfiguration;
@@ -55,12 +54,11 @@
       var3.expand(new BoundingBox(var1, var1));
    }
 
    private static boolean isDirt(LevelSimulatedReader var0, BlockPos var1) {
       return var0.isStateAtPosition(var1, (var0x) -> {
-         Block var1 = var0x.getBlock();
-         return Feature.isDirt(var1) && !var0x.is(Blocks.GRASS_BLOCK) && !var0x.is(Blocks.MYCELIUM);
+         return Feature.isDirt(var0x) && !var0x.is(Blocks.GRASS_BLOCK) && !var0x.is(Blocks.MYCELIUM);
       });
    }
 
    protected static void setDirtAt(LevelSimulatedRW var0, BlockPos var1) {
       if (!isDirt(var0, var1)) {
