--- net/minecraft/server/rcon/NetworkDataOutputStream.java
+++ net/minecraft/server/rcon/NetworkDataOutputStream.java
@@ -28,10 +28,18 @@
 
    public void writeShort(short var1) throws IOException {
       this.dataOutputStream.writeShort(Short.reverseBytes(var1));
    }
 
+   public void writeInt(int var1) throws IOException {
+      this.dataOutputStream.writeInt(Integer.reverseBytes(var1));
+   }
+
+   public void writeFloat(float var1) throws IOException {
+      this.dataOutputStream.writeInt(Integer.reverseBytes(Float.floatToIntBits(var1)));
+   }
+
    public byte[] toByteArray() {
       return this.outputStream.toByteArray();
    }
 
    public void reset() {
