--- net/minecraft/world/entity/AreaEffectCloud.java
+++ net/minecraft/world/entity/AreaEffectCloud.java
@@ -203,11 +203,11 @@
                }
             }
          }
       } else {
          if (this.tickCount >= this.waitTime + this.duration) {
-            this.remove();
+            this.discard();
             return;
          }
 
          boolean var16 = this.tickCount < this.waitTime;
          if (var1 != var16) {
@@ -219,11 +219,11 @@
          }
 
          if (this.radiusPerTick != 0.0F) {
             var2 += this.radiusPerTick;
             if (var2 < 0.5F) {
-               this.remove();
+               this.discard();
                return;
             }
 
             this.setRadius(var2);
          }
@@ -286,21 +286,21 @@
                      }
 
                      if (this.radiusOnUse != 0.0F) {
                         var2 += this.radiusOnUse;
                         if (var2 < 0.5F) {
-                           this.remove();
+                           this.discard();
                            return;
                         }
 
                         this.setRadius(var2);
                      }
 
                      if (this.durationOnUse != 0) {
                         this.duration += this.durationOnUse;
                         if (this.duration <= 0) {
-                           this.remove();
+                           this.discard();
                            return;
                         }
                      }
                   }
                }
