--- net/minecraft/world/entity/AreaEffectCloud.java
+++ net/minecraft/world/entity/AreaEffectCloud.java
@@ -8,11 +8,10 @@
 import java.util.Collection;
 import java.util.Iterator;
 import java.util.List;
 import java.util.Map;
 import java.util.UUID;
-import java.util.Map.Entry;
 import javax.annotation.Nullable;
 import net.minecraft.commands.arguments.ParticleArgument;
 import net.minecraft.core.Registry;
 import net.minecraft.core.particles.ParticleOptions;
 import net.minecraft.core.particles.ParticleTypes;
@@ -34,25 +33,29 @@
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
 
 public class AreaEffectCloud extends Entity {
    private static final Logger LOGGER = LogManager.getLogger();
+   private static final int TIME_BETWEEN_APPLICATIONS = 5;
    private static final EntityDataAccessor<Float> DATA_RADIUS;
    private static final EntityDataAccessor<Integer> DATA_COLOR;
    private static final EntityDataAccessor<Boolean> DATA_WAITING;
    private static final EntityDataAccessor<ParticleOptions> DATA_PARTICLE;
+   private static final float MAX_RADIUS = 32.0F;
    private Potion potion;
    private final List<MobEffectInstance> effects;
    private final Map<Entity, Integer> victims;
    private int duration;
    private int waitTime;
    private int reapplicationDelay;
    private boolean fixedColor;
    private int durationOnUse;
    private float radiusOnUse;
    private float radiusPerTick;
+   @Nullable
    private LivingEntity owner;
+   @Nullable
    private UUID ownerUUID;
 
    public AreaEffectCloud(EntityType<? extends AreaEffectCloud> var1, Level var2) {
       super(var1, var2);
       this.potion = Potions.EMPTY;
@@ -77,11 +80,11 @@
       this.getEntityData().define(DATA_PARTICLE, ParticleTypes.ENTITY_EFFECT);
    }
 
    public void setRadius(float var1) {
       if (!this.level.isClientSide) {
-         this.getEntityData().set(DATA_RADIUS, var1);
+         this.getEntityData().set(DATA_RADIUS, Mth.clamp(var1, 0.0F, 32.0F));
       }
 
    }
 
    public void refreshDimensions() {
@@ -157,150 +160,143 @@
    public void tick() {
       super.tick();
       boolean var1 = this.isWaiting();
       float var2 = this.getRadius();
       if (this.level.isClientSide) {
+         if (var1 && this.random.nextBoolean()) {
+            return;
+         }
+
          ParticleOptions var3 = this.getParticle();
-         float var6;
-         float var7;
-         float var8;
-         int var10;
-         int var11;
-         int var12;
+         int var4;
+         float var5;
          if (var1) {
-            if (this.random.nextBoolean()) {
-               for(int var4 = 0; var4 < 2; ++var4) {
-                  float var5 = this.random.nextFloat() * 6.2831855F;
-                  var6 = Mth.sqrt(this.random.nextFloat()) * 0.2F;
-                  var7 = Mth.cos(var5) * var6;
-                  var8 = Mth.sin(var5) * var6;
-                  if (var3.getType() == ParticleTypes.ENTITY_EFFECT) {
-                     int var9 = this.random.nextBoolean() ? 16777215 : this.getColor();
-                     var10 = var9 >> 16 & 255;
-                     var11 = var9 >> 8 & 255;
-                     var12 = var9 & 255;
-                     this.level.addAlwaysVisibleParticle(var3, this.getX() + (double)var7, this.getY(), this.getZ() + (double)var8, (double)((float)var10 / 255.0F), (double)((float)var11 / 255.0F), (double)((float)var12 / 255.0F));
-                  } else {
-                     this.level.addAlwaysVisibleParticle(var3, this.getX() + (double)var7, this.getY(), this.getZ() + (double)var8, 0.0D, 0.0D, 0.0D);
-                  }
-               }
-            }
+            var4 = 2;
+            var5 = 0.2F;
          } else {
-            float var17 = 3.1415927F * var2 * var2;
+            var4 = Mth.ceil(3.1415927F * var2 * var2);
+            var5 = var2;
+         }
 
-            for(int var19 = 0; (float)var19 < var17; ++var19) {
-               var6 = this.random.nextFloat() * 6.2831855F;
-               var7 = Mth.sqrt(this.random.nextFloat()) * var2;
-               var8 = Mth.cos(var6) * var7;
-               float var28 = Mth.sin(var6) * var7;
-               if (var3.getType() == ParticleTypes.ENTITY_EFFECT) {
-                  var10 = this.getColor();
-                  var11 = var10 >> 16 & 255;
-                  var12 = var10 >> 8 & 255;
-                  int var13 = var10 & 255;
-                  this.level.addAlwaysVisibleParticle(var3, this.getX() + (double)var8, this.getY(), this.getZ() + (double)var28, (double)((float)var11 / 255.0F), (double)((float)var12 / 255.0F), (double)((float)var13 / 255.0F));
+         for(int var6 = 0; var6 < var4; ++var6) {
+            float var7 = this.random.nextFloat() * 6.2831855F;
+            float var8 = Mth.sqrt(this.random.nextFloat()) * var5;
+            double var9 = this.getX() + (double)(Mth.cos(var7) * var8);
+            double var11 = this.getY();
+            double var13 = this.getZ() + (double)(Mth.sin(var7) * var8);
+            double var15;
+            double var17;
+            double var19;
+            if (var3.getType() != ParticleTypes.ENTITY_EFFECT) {
+               if (var1) {
+                  var15 = 0.0D;
+                  var17 = 0.0D;
+                  var19 = 0.0D;
                } else {
-                  this.level.addAlwaysVisibleParticle(var3, this.getX() + (double)var8, this.getY(), this.getZ() + (double)var28, (0.5D - this.random.nextDouble()) * 0.15D, 0.009999999776482582D, (0.5D - this.random.nextDouble()) * 0.15D);
+                  var15 = (0.5D - this.random.nextDouble()) * 0.15D;
+                  var17 = 0.009999999776482582D;
+                  var19 = (0.5D - this.random.nextDouble()) * 0.15D;
                }
+            } else {
+               int var21 = var1 && this.random.nextBoolean() ? 16777215 : this.getColor();
+               var15 = (double)((float)(var21 >> 16 & 255) / 255.0F);
+               var17 = (double)((float)(var21 >> 8 & 255) / 255.0F);
+               var19 = (double)((float)(var21 & 255) / 255.0F);
             }
+
+            this.level.addAlwaysVisibleParticle(var3, var9, var11, var13, var15, var17, var19);
          }
       } else {
          if (this.tickCount >= this.waitTime + this.duration) {
-            this.remove();
+            this.discard();
             return;
          }
 
-         boolean var16 = this.tickCount < this.waitTime;
-         if (var1 != var16) {
-            this.setWaiting(var16);
+         boolean var22 = this.tickCount < this.waitTime;
+         if (var1 != var22) {
+            this.setWaiting(var22);
          }
 
-         if (var16) {
+         if (var22) {
             return;
          }
 
          if (this.radiusPerTick != 0.0F) {
             var2 += this.radiusPerTick;
             if (var2 < 0.5F) {
-               this.remove();
+               this.discard();
                return;
             }
 
             this.setRadius(var2);
          }
 
          if (this.tickCount % 5 == 0) {
-            Iterator var18 = this.victims.entrySet().iterator();
+            this.victims.entrySet().removeIf((var1x) -> {
+               return this.tickCount >= (Integer)var1x.getValue();
+            });
+            ArrayList var23 = Lists.newArrayList();
+            Iterator var24 = this.potion.getEffects().iterator();
 
-            while(var18.hasNext()) {
-               Entry var21 = (Entry)var18.next();
-               if (this.tickCount >= (Integer)var21.getValue()) {
-                  var18.remove();
-               }
+            while(var24.hasNext()) {
+               MobEffectInstance var26 = (MobEffectInstance)var24.next();
+               var23.add(new MobEffectInstance(var26.getEffect(), var26.getDuration() / 4, var26.getAmplifier(), var26.isAmbient(), var26.isVisible()));
             }
 
-            ArrayList var20 = Lists.newArrayList();
-            Iterator var22 = this.potion.getEffects().iterator();
-
-            while(var22.hasNext()) {
-               MobEffectInstance var23 = (MobEffectInstance)var22.next();
-               var20.add(new MobEffectInstance(var23.getEffect(), var23.getDuration() / 4, var23.getAmplifier(), var23.isAmbient(), var23.isVisible()));
-            }
-
-            var20.addAll(this.effects);
-            if (var20.isEmpty()) {
+            var23.addAll(this.effects);
+            if (var23.isEmpty()) {
                this.victims.clear();
             } else {
-               List var24 = this.level.getEntitiesOfClass(LivingEntity.class, this.getBoundingBox());
-               if (!var24.isEmpty()) {
-                  Iterator var25 = var24.iterator();
+               List var25 = this.level.getEntitiesOfClass(LivingEntity.class, this.getBoundingBox());
+               if (!var25.isEmpty()) {
+                  Iterator var27 = var25.iterator();
 
                   while(true) {
-                     LivingEntity var26;
-                     double var30;
+                     double var12;
+                     LivingEntity var28;
                      do {
                         do {
                            do {
-                              if (!var25.hasNext()) {
+                              if (!var27.hasNext()) {
                                  return;
                               }
 
-                              var26 = (LivingEntity)var25.next();
-                           } while(this.victims.containsKey(var26));
-                        } while(!var26.isAffectedByPotions());
+                              var28 = (LivingEntity)var27.next();
+                           } while(this.victims.containsKey(var28));
+                        } while(!var28.isAffectedByPotions());
 
-                        double var27 = var26.getX() - this.getX();
-                        double var29 = var26.getZ() - this.getZ();
-                        var30 = var27 * var27 + var29 * var29;
-                     } while(var30 > (double)(var2 * var2));
+                        double var29 = var28.getX() - this.getX();
+                        double var10 = var28.getZ() - this.getZ();
+                        var12 = var29 * var29 + var10 * var10;
+                     } while(var12 > (double)(var2 * var2));
 
-                     this.victims.put(var26, this.tickCount + this.reapplicationDelay);
-                     Iterator var14 = var20.iterator();
+                     this.victims.put(var28, this.tickCount + this.reapplicationDelay);
+                     Iterator var14 = var23.iterator();
 
                      while(var14.hasNext()) {
-                        MobEffectInstance var15 = (MobEffectInstance)var14.next();
-                        if (var15.getEffect().isInstantenous()) {
-                           var15.getEffect().applyInstantenousEffect(this, this.getOwner(), var26, var15.getAmplifier(), 0.5D);
+                        MobEffectInstance var30 = (MobEffectInstance)var14.next();
+                        if (var30.getEffect().isInstantenous()) {
+                           var30.getEffect().applyInstantenousEffect(this, this.getOwner(), var28, var30.getAmplifier(), 0.5D);
                         } else {
-                           var26.addEffect(new MobEffectInstance(var15));
+                           var28.addEffect(new MobEffectInstance(var30), this);
                         }
                      }
 
                      if (this.radiusOnUse != 0.0F) {
                         var2 += this.radiusOnUse;
                         if (var2 < 0.5F) {
-                           this.remove();
+                           this.discard();
                            return;
                         }
 
                         this.setRadius(var2);
                      }
 
                      if (this.durationOnUse != 0) {
                         this.duration += this.durationOnUse;
                         if (this.duration <= 0) {
-                           this.remove();
+                           this.discard();
                            return;
                         }
                      }
                   }
                }
@@ -308,18 +304,38 @@
          }
       }
 
    }
 
+   public float getRadiusOnUse() {
+      return this.radiusOnUse;
+   }
+
    public void setRadiusOnUse(float var1) {
       this.radiusOnUse = var1;
    }
 
+   public float getRadiusPerTick() {
+      return this.radiusPerTick;
+   }
+
    public void setRadiusPerTick(float var1) {
       this.radiusPerTick = var1;
    }
 
+   public int getDurationOnUse() {
+      return this.durationOnUse;
+   }
+
+   public void setDurationOnUse(int var1) {
+      this.durationOnUse = var1;
+   }
+
+   public int getWaitTime() {
+      return this.waitTime;
+   }
+
    public void setWaitTime(int var1) {
       this.waitTime = var1;
    }
 
    public void setOwner(@Nullable LivingEntity var1) {
@@ -398,11 +414,11 @@
 
       if (this.fixedColor) {
          var1.putInt("Color", this.getColor());
       }
 
-      if (this.potion != Potions.EMPTY && this.potion != null) {
+      if (this.potion != Potions.EMPTY) {
          var1.putString("Potion", Registry.POTION.getKey(this.potion).toString());
       }
 
       if (!this.effects.isEmpty()) {
          ListTag var2 = new ListTag();
@@ -424,10 +440,14 @@
       }
 
       super.onSyncedDataUpdated(var1);
    }
 
+   public Potion getPotion() {
+      return this.potion;
+   }
+
    public PushReaction getPistonPushReaction() {
       return PushReaction.IGNORE;
    }
 
    public Packet<?> getAddEntityPacket() {
