--- net/minecraft/world/entity/EntitySelector.java
+++ net/minecraft/world/entity/EntitySelector.java
@@ -9,11 +9,13 @@
 import net.minecraft.world.item.ItemStack;
 import net.minecraft.world.scores.Team;
 
 public final class EntitySelector {
    public static final Predicate<Entity> ENTITY_STILL_ALIVE = Entity::isAlive;
-   public static final Predicate<LivingEntity> LIVING_ENTITY_STILL_ALIVE = LivingEntity::isAlive;
+   public static final Predicate<Entity> LIVING_ENTITY_STILL_ALIVE = (var0) -> {
+      return var0.isAlive() && var0 instanceof LivingEntity;
+   };
    public static final Predicate<Entity> ENTITY_NOT_BEING_RIDDEN = (var0) -> {
       return var0.isAlive() && !var0.isVehicle() && !var0.isPassenger();
    };
    public static final Predicate<Entity> CONTAINER_ENTITY_SELECTOR = (var0) -> {
       return var0 instanceof Container && var0.isAlive();
