--- net/minecraft/world/entity/HumanoidArm.java
+++ net/minecraft/world/entity/HumanoidArm.java
@@ -11,9 +11,17 @@
 
    private HumanoidArm(Component var3) {
       this.name = var3;
    }
 
+   public HumanoidArm getOpposite() {
+      return this == LEFT ? RIGHT : LEFT;
+   }
+
    public String toString() {
       return this.name.getString();
    }
+
+   public Component getName() {
+      return this.name;
+   }
 }
