--- net/minecraft/world/entity/ItemSteerable.java
+++ net/minecraft/world/entity/ItemSteerable.java
@@ -13,11 +13,11 @@
 
    default boolean travel(Mob var1, ItemBasedSteering var2, Vec3 var3) {
       if (!var1.isAlive()) {
          return false;
       } else {
-         Entity var4 = var1.getPassengers().isEmpty() ? null : (Entity)var1.getPassengers().get(0);
+         Entity var4 = var1.getFirstPassenger();
          if (var1.isVehicle() && var1.canBeControlledByRider() && var4 instanceof Player) {
             var1.yRot = var4.yRot;
             var1.yRotO = var1.yRot;
             var1.xRot = var4.xRot * 0.5F;
             var1.setRot(var1.yRot, var1.xRot);
