--- net/minecraft/world/entity/TamableAnimal.java
+++ net/minecraft/world/entity/TamableAnimal.java
@@ -3,10 +3,12 @@
 import java.util.Optional;
 import java.util.UUID;
 import javax.annotation.Nullable;
 import net.minecraft.Util;
 import net.minecraft.advancements.CriteriaTriggers;
+import net.minecraft.core.particles.ParticleTypes;
+import net.minecraft.core.particles.SimpleParticleType;
 import net.minecraft.nbt.CompoundTag;
 import net.minecraft.network.syncher.EntityDataAccessor;
 import net.minecraft.network.syncher.EntityDataSerializers;
 import net.minecraft.network.syncher.SynchedEntityData;
 import net.minecraft.server.level.ServerPlayer;
@@ -16,11 +18,11 @@
 import net.minecraft.world.entity.player.Player;
 import net.minecraft.world.level.GameRules;
 import net.minecraft.world.level.Level;
 import net.minecraft.world.scores.Team;
 
-public abstract class TamableAnimal extends Animal {
+public abstract class TamableAnimal extends Animal implements OwnableEntity {
    protected static final EntityDataAccessor<Byte> DATA_FLAGS_ID;
    protected static final EntityDataAccessor<Optional<UUID>> DATA_OWNERUUID_ID;
    private boolean orderedToSit;
 
    protected TamableAnimal(EntityType<? extends TamableAnimal> var1, Level var2) {
@@ -68,10 +70,36 @@
 
    public boolean canBeLeashed(Player var1) {
       return !this.isLeashed();
    }
 
+   protected void spawnTamingParticles(boolean var1) {
+      SimpleParticleType var2 = ParticleTypes.HEART;
+      if (!var1) {
+         var2 = ParticleTypes.SMOKE;
+      }
+
+      for(int var3 = 0; var3 < 7; ++var3) {
+         double var4 = this.random.nextGaussian() * 0.02D;
+         double var6 = this.random.nextGaussian() * 0.02D;
+         double var8 = this.random.nextGaussian() * 0.02D;
+         this.level.addParticle(var2, this.getRandomX(1.0D), this.getRandomY() + 0.5D, this.getRandomZ(1.0D), var4, var6, var8);
+      }
+
+   }
+
+   public void handleEntityEvent(byte var1) {
+      if (var1 == 7) {
+         this.spawnTamingParticles(true);
+      } else if (var1 == 6) {
+         this.spawnTamingParticles(false);
+      } else {
+         super.handleEntityEvent(var1);
+      }
+
+   }
+
    public boolean isTame() {
       return ((Byte)this.entityData.get(DATA_FLAGS_ID) & 4) != 0;
    }
 
    public void setTame(boolean var1) {
