--- net/minecraft/world/level/saveddata/maps/MapIndex.java
+++ net/minecraft/world/level/saveddata/maps/MapIndex.java
@@ -10,25 +10,25 @@
 
 public class MapIndex extends SavedData {
    private final Object2IntMap<String> usedAuxIds = new Object2IntOpenHashMap();
 
    public MapIndex() {
-      super("idcounts");
       this.usedAuxIds.defaultReturnValue(-1);
    }
 
-   public void load(CompoundTag var1) {
-      this.usedAuxIds.clear();
-      Iterator var2 = var1.getAllKeys().iterator();
+   public static MapIndex load(CompoundTag var0) {
+      MapIndex var1 = new MapIndex();
+      Iterator var2 = var0.getAllKeys().iterator();
 
       while(var2.hasNext()) {
          String var3 = (String)var2.next();
-         if (var1.contains(var3, 99)) {
-            this.usedAuxIds.put(var3, var1.getInt(var3));
+         if (var0.contains(var3, 99)) {
+            var1.usedAuxIds.put(var3, var0.getInt(var3));
          }
       }
 
+      return var1;
    }
 
    public CompoundTag save(CompoundTag var1) {
       ObjectIterator var2 = this.usedAuxIds.object2IntEntrySet().iterator();
 
