--- net/minecraft/server/network/ServerConnectionListener.java
+++ net/minecraft/server/network/ServerConnectionListener.java
@@ -72,11 +72,11 @@
             var5 = SERVER_EVENT_GROUP;
             LOGGER.info("Using default channel type");
          }
 
          this.channels.add(((ServerBootstrap)((ServerBootstrap)(new ServerBootstrap()).channel(var4)).childHandler(new ChannelInitializer<Channel>() {
-            protected void initChannel(Channel var1) throws Exception {
+            protected void initChannel(Channel var1) {
                try {
                   var1.config().setOption(ChannelOption.TCP_NODELAY, true);
                } catch (ChannelException var4) {
                }
 
@@ -93,11 +93,11 @@
 
    public SocketAddress startMemoryChannel() {
       ChannelFuture var1;
       synchronized(this.channels) {
          var1 = ((ServerBootstrap)((ServerBootstrap)(new ServerBootstrap()).channel(LocalServerChannel.class)).childHandler(new ChannelInitializer<Channel>() {
-            protected void initChannel(Channel var1) throws Exception {
+            protected void initChannel(Channel var1) {
                Connection var2 = new Connection(PacketFlow.SERVERBOUND);
                var2.setListener(new MemoryServerHandshakePacketListenerImpl(ServerConnectionListener.this.server, var2));
                ServerConnectionListener.this.connections.add(var2);
                var1.pipeline().addLast("packet_handler", var2);
             }
