--- net/minecraft/util/datafix/fixes/BedBlockEntityInjecter.java
+++ net/minecraft/util/datafix/fixes/BedBlockEntityInjecter.java
@@ -13,10 +13,11 @@
 import com.mojang.datafixers.util.Pair;
 import com.mojang.serialization.Dynamic;
 import java.util.ArrayList;
 import java.util.HashMap;
 import java.util.Iterator;
+import java.util.Objects;
 import java.util.function.Function;
 import java.util.stream.Stream;
 
 public class BedBlockEntityInjecter extends DataFix {
    public BedBlockEntityInjecter(Schema var1, boolean var2) {
@@ -57,11 +58,11 @@
             int var12 = var11.get("Y").asInt(0);
             Stream var13 = var11.get("Blocks").asStream().map((var0) -> {
                return var0.asInt(0);
             });
             int var14 = 0;
-            var13.getClass();
+            Objects.requireNonNull(var13);
 
             for(Iterator var15 = (var13::iterator).iterator(); var15.hasNext(); ++var14) {
                int var16 = (Integer)var15.next();
                if (416 == (var16 & 255) << 4) {
                   int var17 = var14 & 15;
