--- net/minecraft/util/datafix/fixes/EntityMinecartIdentifiersFix.java
+++ net/minecraft/util/datafix/fixes/EntityMinecartIdentifiersFix.java
@@ -7,10 +7,11 @@
 import com.mojang.datafixers.Typed;
 import com.mojang.datafixers.schemas.Schema;
 import com.mojang.datafixers.types.Type;
 import com.mojang.datafixers.types.templates.TaggedChoice;
 import com.mojang.datafixers.util.Pair;
+import com.mojang.serialization.DataResult;
 import com.mojang.serialization.Dynamic;
 import java.util.List;
 import java.util.Objects;
 
 public class EntityMinecartIdentifiersFix extends DataFix {
@@ -36,11 +37,11 @@
                   var6 = (String)MINECART_BY_ID.get(var7);
                } else {
                   var6 = "MinecartRideable";
                }
 
-               return Pair.of(var6, var4.write().map((var2xx) -> {
+               return Pair.of(var6, (DataResult)var4.write().map((var2xx) -> {
                   return ((Type)var2.types().get(var6)).read(var2xx);
                }).result().orElseThrow(() -> {
                   return new IllegalStateException("Could not read the new minecart.");
                }));
             }
