--- net/minecraft/util/datafix/fixes/EntityHorseSplitFix.java
+++ net/minecraft/util/datafix/fixes/EntityHorseSplitFix.java
@@ -38,12 +38,12 @@
          }
 
          var3.remove("Type");
          Type var6 = (Type)this.getOutputSchema().findChoiceType(References.ENTITY).types().get(var4);
          DataResult var10001 = var2.write();
-         var6.getClass();
-         return Pair.of(var4, ((Pair)var10001.flatMap(var6::readTyped).result().orElseThrow(() -> {
+         Objects.requireNonNull(var6);
+         return Pair.of(var4, (Typed)((Pair)var10001.flatMap(var6::readTyped).result().orElseThrow(() -> {
             return new IllegalStateException("Could not parse the new horse");
          })).getFirst());
       } else {
          return Pair.of(var1, var2);
       }
