--- net/minecraft/util/datafix/fixes/EntityRenameFix.java
+++ net/minecraft/util/datafix/fixes/EntityRenameFix.java
@@ -27,11 +27,11 @@
             Pair var7 = this.fix(var5, this.getEntity(var4.getSecond(), var3, var6));
             Type var8 = (Type)var2.types().get(var7.getFirst());
             if (!var8.equals(((Typed)var7.getSecond()).getType(), true, true)) {
                throw new IllegalStateException(String.format("Dynamic type check failed: %s not equal to %s", var8, ((Typed)var7.getSecond()).getType()));
             } else {
-               return Pair.of(var7.getFirst(), ((Typed)var7.getSecond()).getValue());
+               return Pair.of((String)var7.getFirst(), ((Typed)var7.getSecond()).getValue());
             }
          };
       });
    }
 
