--- net/minecraft/util/datafix/fixes/EntityZombieVillagerTypeFix.java
+++ net/minecraft/util/datafix/fixes/EntityZombieVillagerTypeFix.java
@@ -5,10 +5,11 @@
 import com.mojang.datafixers.schemas.Schema;
 import com.mojang.serialization.Dynamic;
 import java.util.Random;
 
 public class EntityZombieVillagerTypeFix extends NamedEntityFix {
+   private static final int PROFESSION_MAX = 6;
    private static final Random RANDOM = new Random();
 
    public EntityZombieVillagerTypeFix(Schema var1, boolean var2) {
       super(var1, var2, "EntityZombieVillagerTypeFix", References.ENTITY, "Zombie");
    }
