--- net/minecraft/util/datafix/fixes/ItemLoreFix.java
+++ net/minecraft/util/datafix/fixes/ItemLoreFix.java
@@ -7,10 +7,11 @@
 import com.mojang.datafixers.TypeRewriteRule;
 import com.mojang.datafixers.schemas.Schema;
 import com.mojang.datafixers.types.Type;
 import com.mojang.serialization.DataResult;
 import com.mojang.serialization.Dynamic;
+import java.util.Objects;
 import java.util.stream.Stream;
 import net.minecraft.network.chat.Component;
 import net.minecraft.network.chat.TextComponent;
 
 public class ItemLoreFix extends DataFix {
@@ -25,11 +26,11 @@
          return var1x.updateTyped(var2, (var0) -> {
             return var0.update(DSL.remainderFinder(), (var0x) -> {
                return var0x.update("display", (var0) -> {
                   return var0.update("Lore", (var0x) -> {
                      DataResult var10000 = var0x.asStreamOpt().map(ItemLoreFix::fixLoreList);
-                     var0x.getClass();
+                     Objects.requireNonNull(var0x);
                      return (Dynamic)DataFixUtils.orElse(var10000.map(var0x::createList).result(), var0x);
                   });
                });
             });
          });
@@ -37,11 +38,11 @@
    }
 
    private static <T> Stream<Dynamic<T>> fixLoreList(Stream<Dynamic<T>> var0) {
       return var0.map((var0x) -> {
          DataResult var10000 = var0x.asString().map(ItemLoreFix::fixLoreEntry);
-         var0x.getClass();
+         Objects.requireNonNull(var0x);
          return (Dynamic)DataFixUtils.orElse(var10000.map(var0x::createString).result(), var0x);
       });
    }
 
    private static String fixLoreEntry(String var0) {
