--- net/minecraft/util/datafix/fixes/OptionsKeyTranslationFix.java
+++ net/minecraft/util/datafix/fixes/OptionsKeyTranslationFix.java
@@ -20,15 +20,15 @@
             return (Dynamic)var0x.getMapValues().map((var1) -> {
                return var0x.createMap((Map)var1.entrySet().stream().map((var1x) -> {
                   if (((Dynamic)var1x.getKey()).asString("").startsWith("key_")) {
                      String var2 = ((Dynamic)var1x.getValue()).asString("");
                      if (!var2.startsWith("key.mouse") && !var2.startsWith("scancode.")) {
-                        return Pair.of(var1x.getKey(), var0x.createString("key.keyboard." + var2.substring("key.".length())));
+                        return Pair.of((Dynamic)var1x.getKey(), var0x.createString("key.keyboard." + var2.substring("key.".length())));
                      }
                   }
 
-                  return Pair.of(var1x.getKey(), var1x.getValue());
+                  return Pair.of((Dynamic)var1x.getKey(), (Dynamic)var1x.getValue());
                }).collect(Collectors.toMap(Pair::getFirst, Pair::getSecond)));
             }).result().orElse(var0x);
          });
       });
    }
