--- net/minecraft/util/datafix/fixes/OptionsKeyLwjgl3Fix.java
+++ net/minecraft/util/datafix/fixes/OptionsKeyLwjgl3Fix.java
@@ -11,10 +11,11 @@
 import it.unimi.dsi.fastutil.ints.Int2ObjectOpenHashMap;
 import java.util.Map;
 import java.util.stream.Collectors;
 
 public class OptionsKeyLwjgl3Fix extends DataFix {
+   public static final String KEY_UNKNOWN = "key.unknown";
    private static final Int2ObjectMap<String> MAP = (Int2ObjectMap)DataFixUtils.make(new Int2ObjectOpenHashMap(), (var0) -> {
       var0.put(0, "key.unknown");
       var0.put(11, "key.0");
       var0.put(2, "key.1");
       var0.put(3, "key.2");
