--- net/minecraft/util/datafix/fixes/References.java
+++ net/minecraft/util/datafix/fixes/References.java
@@ -31,10 +31,13 @@
       return "advancements";
    };
    public static final DSL.TypeReference POI_CHUNK = () -> {
       return "poi_chunk";
    };
+   public static final DSL.TypeReference ENTITY_CHUNK = () -> {
+      return "entity_chunk";
+   };
    public static final DSL.TypeReference BLOCK_ENTITY = () -> {
       return "block_entity";
    };
    public static final DSL.TypeReference ITEM_STACK = () -> {
       return "item_stack";
@@ -76,6 +79,9 @@
       return "biome";
    };
    public static final DSL.TypeReference WORLD_GEN_SETTINGS = () -> {
       return "world_gen_settings";
    };
+
+   public References() {
+   }
 }
