--- net/minecraft/util/datafix/fixes/SavedDataVillageCropFix.java
+++ net/minecraft/util/datafix/fixes/SavedDataVillageCropFix.java
@@ -3,10 +3,11 @@
 import com.mojang.datafixers.DataFix;
 import com.mojang.datafixers.TypeRewriteRule;
 import com.mojang.datafixers.schemas.Schema;
 import com.mojang.serialization.DataResult;
 import com.mojang.serialization.Dynamic;
+import java.util.Objects;
 import java.util.stream.Stream;
 
 public class SavedDataVillageCropFix extends DataFix {
    public SavedDataVillageCropFix(Schema var1, boolean var2) {
       super(var1, var2);
@@ -20,11 +21,11 @@
       return var1.update("Children", SavedDataVillageCropFix::updateChildren);
    }
 
    private static <T> Dynamic<T> updateChildren(Dynamic<T> var0) {
       DataResult var10000 = var0.asStreamOpt().map(SavedDataVillageCropFix::updateChildren);
-      var0.getClass();
+      Objects.requireNonNull(var0);
       return (Dynamic)var10000.map(var0::createList).result().orElse(var0);
    }
 
    private static Stream<? extends Dynamic<?>> updateChildren(Stream<? extends Dynamic<?>> var0) {
       return var0.map((var0x) -> {
