--- net/minecraft/util/datafix/fixes/SimpleEntityRenameFix.java
+++ net/minecraft/util/datafix/fixes/SimpleEntityRenameFix.java
@@ -11,10 +11,10 @@
       super(var1, var2, var3);
    }
 
    protected Pair<String, Typed<?>> fix(String var1, Typed<?> var2) {
       Pair var3 = this.getNewNameAndTag(var1, (Dynamic)var2.getOrCreate(DSL.remainderFinder()));
-      return Pair.of(var3.getFirst(), var2.set(DSL.remainderFinder(), var3.getSecond()));
+      return Pair.of((String)var3.getFirst(), var2.set(DSL.remainderFinder(), (Object)((Dynamic)var3.getSecond())));
    }
 
    protected abstract Pair<String, Dynamic<?>> getNewNameAndTag(String var1, Dynamic<?> var2);
 }
