--- net/minecraft/util/datafix/fixes/TrappedChestBlockEntityFix.java
+++ net/minecraft/util/datafix/fixes/TrappedChestBlockEntityFix.java
@@ -19,10 +19,12 @@
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
 
 public class TrappedChestBlockEntityFix extends DataFix {
    private static final Logger LOGGER = LogManager.getLogger();
+   private static final int SIZE = 4096;
+   private static final short SIZE_BITS = 12;
 
    public TrappedChestBlockEntityFix(Schema var1, boolean var2) {
       super(var1, var2);
    }
 
