--- net/minecraft/util/datafix/fixes/VillagerRebuildLevelAndXpFix.java
+++ net/minecraft/util/datafix/fixes/VillagerRebuildLevelAndXpFix.java
@@ -11,14 +11,15 @@
 import com.mojang.serialization.Dynamic;
 import java.util.Optional;
 import net.minecraft.util.Mth;
 
 public class VillagerRebuildLevelAndXpFix extends DataFix {
+   private static final int TRADES_PER_LEVEL = 2;
    private static final int[] LEVEL_XP_THRESHOLDS = new int[]{0, 10, 50, 100, 150};
 
    public static int getMinXpPerLevel(int var0) {
-      return LEVEL_XP_THRESHOLDS[Mth.clamp(var0 - 1, 0, LEVEL_XP_THRESHOLDS.length - 1)];
+      return LEVEL_XP_THRESHOLDS[Mth.clamp((int)(var0 - 1), (int)0, (int)(LEVEL_XP_THRESHOLDS.length - 1))];
    }
 
    public VillagerRebuildLevelAndXpFix(Schema var1, boolean var2) {
       super(var1, var2);
    }
@@ -40,11 +41,11 @@
                int var7x = (Integer)var3x.getOptionalTyped(var3).flatMap((var1) -> {
                   return var1.getOptionalTyped(var5);
                }).map((var1) -> {
                   return var1.getAllTyped(var7).size();
                }).orElse(0);
-               var5x = Mth.clamp(var7x / 2, 1, 5);
+               var5x = Mth.clamp((int)(var7x / 2), (int)1, (int)5);
                if (var5x > 1) {
                   var6 = addLevel(var3x, var5x);
                }
             }
 
