--- net/minecraft/server/gui/MinecraftServerGui.java
+++ net/minecraft/server/gui/MinecraftServerGui.java
@@ -32,10 +32,12 @@
 import org.apache.logging.log4j.Logger;
 
 public class MinecraftServerGui extends JComponent {
    private static final Font MONOSPACED = new Font("Monospaced", 0, 12);
    private static final Logger LOGGER = LogManager.getLogger();
+   private static final String TITLE = "Minecraft server";
+   private static final String SHUTDOWN_TITLE = "Minecraft server - shutting down!";
    private final DedicatedServer server;
    private Thread logAppenderThread;
    private final Collection<Runnable> finalizers = Lists.newArrayList();
    private final AtomicBoolean isClosing = new AtomicBoolean();
 
