--- net/minecraft/server/gui/StatsComponent.java
+++ net/minecraft/server/gui/StatsComponent.java
@@ -34,11 +34,14 @@
    }
 
    private void tick() {
       long var1 = Runtime.getRuntime().totalMemory() - Runtime.getRuntime().freeMemory();
       this.msgs[0] = "Memory use: " + var1 / 1024L / 1024L + " mb (" + Runtime.getRuntime().freeMemory() * 100L / Runtime.getRuntime().maxMemory() + "% free)";
-      this.msgs[1] = "Avg tick: " + DECIMAL_FORMAT.format(this.getAverage(this.server.tickTimes) * 1.0E-6D) + " ms";
+      String[] var10000 = this.msgs;
+      DecimalFormat var10002 = DECIMAL_FORMAT;
+      double var10003 = this.getAverage(this.server.tickTimes);
+      var10000[1] = "Avg tick: " + var10002.format(var10003 * 1.0E-6D) + " ms";
       this.values[this.vp++ & 255] = (int)(var1 * 100L / Runtime.getRuntime().maxMemory());
       this.repaint();
    }
 
    private double getAverage(long[] var1) {
