--- net/minecraft/core/dispenser/BoatDispenseItemBehavior.java
+++ net/minecraft/core/dispenser/BoatDispenseItemBehavior.java
@@ -35,11 +35,11 @@
          var12 = 0.0D;
       }
 
       Boat var14 = new Boat(var4, var5, var7 + var12, var9);
       var14.setType(this.type);
-      var14.yRot = var3.toYRot();
+      var14.setYRot(var3.toYRot());
       var4.addFreshEntity(var14);
       var2.shrink(1);
       return var2;
    }
 
