--- net/minecraft/core/dispenser/ShearsDispenseItemBehavior.java
+++ net/minecraft/core/dispenser/ShearsDispenseItemBehavior.java
@@ -8,19 +8,21 @@
 import net.minecraft.server.level.ServerLevel;
 import net.minecraft.server.level.ServerPlayer;
 import net.minecraft.sounds.SoundEvents;
 import net.minecraft.sounds.SoundSource;
 import net.minecraft.tags.BlockTags;
+import net.minecraft.world.entity.Entity;
 import net.minecraft.world.entity.EntitySelector;
 import net.minecraft.world.entity.LivingEntity;
 import net.minecraft.world.entity.Shearable;
 import net.minecraft.world.entity.player.Player;
 import net.minecraft.world.item.ItemStack;
 import net.minecraft.world.level.block.BeehiveBlock;
 import net.minecraft.world.level.block.DispenserBlock;
 import net.minecraft.world.level.block.entity.BeehiveBlockEntity;
 import net.minecraft.world.level.block.state.BlockState;
+import net.minecraft.world.level.gameevent.GameEvent;
 import net.minecraft.world.phys.AABB;
 
 public class ShearsDispenseItemBehavior extends OptionalDispenseItemBehavior {
    public ShearsDispenseItemBehavior() {
    }
@@ -44,10 +46,11 @@
          int var3 = (Integer)var2.getValue(BeehiveBlock.HONEY_LEVEL);
          if (var3 >= 5) {
             var0.playSound((Player)null, var1, SoundEvents.BEEHIVE_SHEAR, SoundSource.BLOCKS, 1.0F, 1.0F);
             BeehiveBlock.dropHoneycomb(var0, var1);
             ((BeehiveBlock)var2.getBlock()).releaseBeesAndResetHoneyLevel(var0, var2, var1, (Player)null, BeehiveBlockEntity.BeeReleaseStatus.BEE_RELEASED);
+            var0.gameEvent((Entity)null, GameEvent.SHEAR, var1);
             return true;
          }
       }
 
       return false;
@@ -61,10 +64,11 @@
          LivingEntity var4 = (LivingEntity)var3.next();
          if (var4 instanceof Shearable) {
             Shearable var5 = (Shearable)var4;
             if (var5.readyForShearing()) {
                var5.shear(SoundSource.BLOCKS);
+               var0.gameEvent((Entity)null, GameEvent.SHEAR, var1);
                return true;
             }
          }
       }
 
