--- net/minecraft/world/item/enchantment/DamageEnchantment.java
+++ net/minecraft/world/item/enchantment/DamageEnchantment.java
@@ -52,11 +52,11 @@
    }
 
    public void doPostAttack(LivingEntity var1, Entity var2, int var3) {
       if (var2 instanceof LivingEntity) {
          LivingEntity var4 = (LivingEntity)var2;
-         if (this.type == 2 && var4.getMobType() == MobType.ARTHROPOD) {
+         if (this.type == 2 && var3 > 0 && var4.getMobType() == MobType.ARTHROPOD) {
             int var5 = 20 + var1.getRandom().nextInt(10 * var3);
             var4.addEffect(new MobEffectInstance(MobEffects.MOVEMENT_SLOWDOWN, var5, 3));
          }
       }
 
