--- net/minecraft/world/item/enchantment/ProtectionEnchantment.java
+++ net/minecraft/world/item/enchantment/ProtectionEnchantment.java
@@ -30,11 +30,11 @@
          return 0;
       } else if (this.type == ProtectionEnchantment.Type.ALL) {
          return var1;
       } else if (this.type == ProtectionEnchantment.Type.FIRE && var2.isFire()) {
          return var1 * 2;
-      } else if (this.type == ProtectionEnchantment.Type.FALL && var2 == DamageSource.FALL) {
+      } else if (this.type == ProtectionEnchantment.Type.FALL && var2.isFall()) {
          return var1 * 3;
       } else if (this.type == ProtectionEnchantment.Type.EXPLOSION && var2.isExplosion()) {
          return var1 * 2;
       } else {
          return this.type == ProtectionEnchantment.Type.PROJECTILE && var2.isProjectile() ? var1 * 2 : 0;
@@ -71,24 +71,22 @@
 
       return var1;
    }
 
    public static enum Type {
-      ALL("all", 1, 11),
-      FIRE("fire", 10, 8),
-      FALL("fall", 5, 6),
-      EXPLOSION("explosion", 5, 8),
-      PROJECTILE("projectile", 3, 6);
+      ALL(1, 11),
+      FIRE(10, 8),
+      FALL(5, 6),
+      EXPLOSION(5, 8),
+      PROJECTILE(3, 6);
 
-      private final String name;
       private final int minCost;
       private final int levelCost;
 
-      private Type(String var3, int var4, int var5) {
-         this.name = var3;
-         this.minCost = var4;
-         this.levelCost = var5;
+      private Type(int var3, int var4) {
+         this.minCost = var3;
+         this.levelCost = var4;
       }
 
       public int getMinCost() {
          return this.minCost;
       }
