--- net/minecraft/world/item/enchantment/ThornsEnchantment.java
+++ net/minecraft/world/item/enchantment/ThornsEnchantment.java
@@ -8,10 +8,12 @@
 import net.minecraft.world.entity.LivingEntity;
 import net.minecraft.world.item.ArmorItem;
 import net.minecraft.world.item.ItemStack;
 
 public class ThornsEnchantment extends Enchantment {
+   private static final float CHANCE_PER_LEVEL = 0.15F;
+
    public ThornsEnchantment(Enchantment.Rarity var1, EquipmentSlot... var2) {
       super(var1, EnchantmentCategory.ARMOR_CHEST, var2);
    }
 
    public int getMinCost(int var1) {
