--- net/minecraft/server/dedicated/Settings.java
+++ net/minecraft/server/dedicated/Settings.java
@@ -50,11 +50,11 @@
                }
             }
 
          }
       } catch (IOException var15) {
-         LOGGER.error("Failed to load properties from file: " + var0);
+         LOGGER.error("Failed to load properties from file: {}", var0);
       }
 
       return var1;
    }
 
@@ -81,11 +81,11 @@
                }
             }
 
          }
       } catch (IOException var15) {
-         LOGGER.error("Failed to store properties to file: " + var1);
+         LOGGER.error("Failed to store properties to file: {}", var1);
       }
 
    }
 
    private static <V extends Number> Function<String, V> wrapNumberDeserializer(Function<String, V> var0) {
