--- net/minecraft/core/Direction.java
+++ net/minecraft/core/Direction.java
@@ -113,10 +113,22 @@
 
    public Direction.AxisDirection getAxisDirection() {
       return this.axisDirection;
    }
 
+   public static Direction getFacingAxis(Entity var0, Direction.Axis var1) {
+      switch(var1) {
+      case X:
+         return EAST.isFacingAngle(var0.getViewYRot(1.0F)) ? EAST : WEST;
+      case Z:
+         return SOUTH.isFacingAngle(var0.getViewYRot(1.0F)) ? SOUTH : NORTH;
+      case Y:
+      default:
+         return var0.getViewXRot(1.0F) < 0.0F ? UP : DOWN;
+      }
+   }
+
    public Direction getOpposite() {
       return from3DDataValue(this.oppositeIndex);
    }
 
    public Direction getClockWise() {
@@ -188,15 +200,15 @@
 
    public static Direction fromAxisAndDirection(Direction.Axis var0, Direction.AxisDirection var1) {
       switch(var0) {
       case X:
          return var1 == Direction.AxisDirection.POSITIVE ? EAST : WEST;
-      case Y:
-         return var1 == Direction.AxisDirection.POSITIVE ? UP : DOWN;
       case Z:
       default:
          return var1 == Direction.AxisDirection.POSITIVE ? SOUTH : NORTH;
+      case Y:
+         return var1 == Direction.AxisDirection.POSITIVE ? UP : DOWN;
       }
    }
 
    public float toYRot() {
       return (float)((this.data2d & 3) * 90);
