--- net/minecraft/core/IdMapper.java
+++ net/minecraft/core/IdMapper.java
@@ -7,10 +7,11 @@
 import java.util.Iterator;
 import java.util.List;
 import javax.annotation.Nullable;
 
 public class IdMapper<T> implements IdMap<T> {
+   public static final int DEFAULT = -1;
    private int nextId;
    private final IdentityHashMap<T, Integer> tToId;
    private final List<T> idToT;
 
    public IdMapper() {
@@ -52,9 +53,13 @@
 
    public Iterator<T> iterator() {
       return Iterators.filter(this.idToT.iterator(), Predicates.notNull());
    }
 
+   public boolean contains(int var1) {
+      return this.byId(var1) != null;
+   }
+
    public int size() {
       return this.tToId.size();
    }
 }
