--- net/minecraft/core/MappedRegistry.java
+++ net/minecraft/core/MappedRegistry.java
@@ -180,10 +180,14 @@
       }
 
       return Util.getRandom(this.randomCache, var1);
    }
 
+   public boolean containsKey(ResourceLocation var1) {
+      return this.storage.containsKey(var1);
+   }
+
    public static <T> Codec<MappedRegistry<T>> networkCodec(ResourceKey<? extends Registry<T>> var0, Lifecycle var1, Codec<T> var2) {
       return withNameAndId(var0, var2.fieldOf("element")).codec().listOf().xmap((var2x) -> {
          MappedRegistry var3 = new MappedRegistry(var0, var1);
          Iterator var4 = var2x.iterator();
 
