--- net/minecraft/core/Vec3i.java
+++ net/minecraft/core/Vec3i.java
@@ -64,22 +64,29 @@
 
    public int getZ() {
       return this.z;
    }
 
-   protected void setX(int var1) {
+   protected Vec3i setX(int var1) {
       this.x = var1;
+      return this;
    }
 
-   protected void setY(int var1) {
+   protected Vec3i setY(int var1) {
       this.y = var1;
+      return this;
    }
 
-   protected void setZ(int var1) {
+   protected Vec3i setZ(int var1) {
       this.z = var1;
+      return this;
    }
 
+   public Vec3i offset(int var1, int var2, int var3) {
+      return var1 == 0 && var2 == 0 && var3 == 0 ? this : new Vec3i(this.getX() + var1, this.getY() + var2, this.getZ() + var3);
+   }
+
    public Vec3i above() {
       return this.above(1);
    }
 
    public Vec3i above(int var1) {
@@ -116,10 +123,14 @@
 
    public double distSqr(Position var1, boolean var2) {
       return this.distSqr(var1.x(), var1.y(), var1.z(), var2);
    }
 
+   public double distSqr(Vec3i var1, boolean var2) {
+      return this.distSqr((double)var1.x, (double)var1.y, (double)var1.z, var2);
+   }
+
    public double distSqr(double var1, double var3, double var5, boolean var7) {
       double var8 = var7 ? 0.5D : 0.0D;
       double var10 = (double)this.getX() + var8 - var1;
       double var12 = (double)this.getY() + var8 - var3;
       double var14 = (double)this.getZ() + var8 - var5;
@@ -141,11 +152,11 @@
       return MoreObjects.toStringHelper((Object)this).add("x", this.getX()).add("y", this.getY()).add("z", this.getZ()).toString();
    }
 
    static {
       CODEC = Codec.INT_STREAM.comapFlatMap((var0) -> {
-         return Util.fixedSize(var0, 3).map((var0x) -> {
+         return Util.fixedSize((IntStream)var0, 3).map((var0x) -> {
             return new Vec3i(var0x[0], var0x[1], var0x[2]);
          });
       }, (var0) -> {
          return IntStream.of(new int[]{var0.getX(), var0.getY(), var0.getZ()});
       });
