--- net/minecraft/data/models/ItemModelGenerators.java
+++ net/minecraft/data/models/ItemModelGenerators.java
@@ -30,10 +30,11 @@
       var3.create(ModelLocationUtils.getModelLocation(var1), TextureMapping.layer0(var2), this.output);
    }
 
    public void run() {
       this.generateFlatItem(Items.ACACIA_BOAT, ModelTemplates.FLAT_ITEM);
+      this.generateFlatItem(Items.AMETHYST_SHARD, ModelTemplates.FLAT_ITEM);
       this.generateFlatItem(Items.APPLE, ModelTemplates.FLAT_ITEM);
       this.generateFlatItem(Items.ARMOR_STAND, ModelTemplates.FLAT_ITEM);
       this.generateFlatItem(Items.ARROW, ModelTemplates.FLAT_ITEM);
       this.generateFlatItem(Items.BAKED_POTATO, ModelTemplates.FLAT_ITEM);
       this.generateFlatItem(Items.BAMBOO, ModelTemplates.FLAT_HANDHELD_ITEM);
@@ -85,10 +86,11 @@
       this.generateFlatItem(Items.COOKED_MUTTON, ModelTemplates.FLAT_ITEM);
       this.generateFlatItem(Items.COOKED_PORKCHOP, ModelTemplates.FLAT_ITEM);
       this.generateFlatItem(Items.COOKED_RABBIT, ModelTemplates.FLAT_ITEM);
       this.generateFlatItem(Items.COOKED_SALMON, ModelTemplates.FLAT_ITEM);
       this.generateFlatItem(Items.COOKIE, ModelTemplates.FLAT_ITEM);
+      this.generateFlatItem(Items.COPPER_INGOT, ModelTemplates.FLAT_ITEM);
       this.generateFlatItem(Items.CREEPER_BANNER_PATTERN, ModelTemplates.FLAT_ITEM);
       this.generateFlatItem(Items.CYAN_DYE, ModelTemplates.FLAT_ITEM);
       this.generateFlatItem(Items.DARK_OAK_BOAT, ModelTemplates.FLAT_ITEM);
       this.generateFlatItem(Items.DIAMOND, ModelTemplates.FLAT_ITEM);
       this.generateFlatItem(Items.DIAMOND_AXE, ModelTemplates.FLAT_HANDHELD_ITEM);
