--- net/minecraft/util/profiling/ActiveProfiler.java
+++ net/minecraft/util/profiling/ActiveProfiler.java
@@ -141,11 +141,11 @@
    }
 
    static class PathEntry implements ProfilerPathEntry {
       private long duration;
       private long count;
-      private Object2LongOpenHashMap<String> counters;
+      private final Object2LongOpenHashMap<String> counters;
 
       private PathEntry() {
          this.counters = new Object2LongOpenHashMap();
       }
 
