--- net/minecraft/util/profiling/InactiveProfiler.java
+++ net/minecraft/util/profiling/InactiveProfiler.java
@@ -1,8 +1,9 @@
 package net.minecraft.util.profiling;
 
 import java.util.function.Supplier;
+import javax.annotation.Nullable;
 
 public class InactiveProfiler implements ProfileCollector {
    public static final InactiveProfiler INSTANCE = new InactiveProfiler();
 
    private InactiveProfiler() {
@@ -24,15 +25,23 @@
    }
 
    public void popPush(String var1) {
    }
 
+   public void popPush(Supplier<String> var1) {
+   }
+
    public void incrementCounter(String var1) {
    }
 
    public void incrementCounter(Supplier<String> var1) {
    }
 
    public ProfileResults getResults() {
       return EmptyProfileResults.EMPTY;
    }
+
+   @Nullable
+   public ActiveProfiler.PathEntry getEntry(String var1) {
+      return null;
+   }
 }
