--- net/minecraft/util/profiling/ProfileResults.java
+++ net/minecraft/util/profiling/ProfileResults.java
@@ -1,12 +1,17 @@
 package net.minecraft.util.profiling;
 
-import java.io.File;
+import java.nio.file.Path;
+import java.util.List;
 
 public interface ProfileResults {
-   boolean saveResults(File var1);
+   char PATH_SEPARATOR = '\u001e';
 
+   List<ResultField> getTimes(String var1);
+
+   boolean saveResults(Path var1);
+
    long getStartTimeNano();
 
    int getStartTimeTicks();
 
    long getEndTimeNano();
@@ -19,9 +24,11 @@
 
    default int getTickDuration() {
       return this.getEndTimeTicks() - this.getStartTimeTicks();
    }
 
+   String getProfilerResults();
+
    static String demanglePath(String var0) {
       return var0.replace('\u001e', '.');
    }
 }
