--- net/minecraft/world/damagesource/IndirectEntityDamageSource.java
+++ net/minecraft/world/damagesource/IndirectEntityDamageSource.java
@@ -6,10 +6,11 @@
 import net.minecraft.world.entity.Entity;
 import net.minecraft.world.entity.LivingEntity;
 import net.minecraft.world.item.ItemStack;
 
 public class IndirectEntityDamageSource extends EntityDamageSource {
+   @Nullable
    private final Entity owner;
 
    public IndirectEntityDamageSource(String var1, Entity var2, @Nullable Entity var3) {
       super(var1, var2);
       this.owner = var3;
