--- net/minecraft/server/commands/data/EntityDataAccessor.java
+++ net/minecraft/server/commands/data/EntityDataAccessor.java
@@ -11,10 +11,11 @@
 import net.minecraft.commands.CommandSourceStack;
 import net.minecraft.commands.Commands;
 import net.minecraft.commands.arguments.EntityArgument;
 import net.minecraft.commands.arguments.NbtPathArgument;
 import net.minecraft.nbt.CompoundTag;
+import net.minecraft.nbt.NbtUtils;
 import net.minecraft.nbt.Tag;
 import net.minecraft.network.chat.Component;
 import net.minecraft.network.chat.TranslatableComponent;
 import net.minecraft.world.entity.Entity;
 import net.minecraft.world.entity.player.Player;
@@ -55,11 +56,11 @@
    public Component getModifiedSuccess() {
       return new TranslatableComponent("commands.data.entity.modified", new Object[]{this.entity.getDisplayName()});
    }
 
    public Component getPrintSuccess(Tag var1) {
-      return new TranslatableComponent("commands.data.entity.query", new Object[]{this.entity.getDisplayName(), var1.getPrettyDisplay()});
+      return new TranslatableComponent("commands.data.entity.query", new Object[]{this.entity.getDisplayName(), NbtUtils.toPrettyComponent(var1)});
    }
 
    public Component getPrintSuccess(NbtPathArgument.NbtPath var1, double var2, int var4) {
       return new TranslatableComponent("commands.data.entity.get", new Object[]{var1, this.entity.getDisplayName(), String.format(Locale.ROOT, "%.2f", var2), var4});
    }
