--- net/minecraft/server/commands/data/StorageDataAccessor.java
+++ net/minecraft/server/commands/data/StorageDataAccessor.java
@@ -9,10 +9,11 @@
 import net.minecraft.commands.Commands;
 import net.minecraft.commands.SharedSuggestionProvider;
 import net.minecraft.commands.arguments.NbtPathArgument;
 import net.minecraft.commands.arguments.ResourceLocationArgument;
 import net.minecraft.nbt.CompoundTag;
+import net.minecraft.nbt.NbtUtils;
 import net.minecraft.nbt.Tag;
 import net.minecraft.network.chat.Component;
 import net.minecraft.network.chat.TranslatableComponent;
 import net.minecraft.resources.ResourceLocation;
 import net.minecraft.world.level.storage.CommandStorage;
@@ -55,11 +56,11 @@
    public Component getModifiedSuccess() {
       return new TranslatableComponent("commands.data.storage.modified", new Object[]{this.id});
    }
 
    public Component getPrintSuccess(Tag var1) {
-      return new TranslatableComponent("commands.data.storage.query", new Object[]{this.id, var1.getPrettyDisplay()});
+      return new TranslatableComponent("commands.data.storage.query", new Object[]{this.id, NbtUtils.toPrettyComponent(var1)});
    }
 
    public Component getPrintSuccess(NbtPathArgument.NbtPath var1, double var2, int var4) {
       return new TranslatableComponent("commands.data.storage.get", new Object[]{var1, this.id, String.format(Locale.ROOT, "%.2f", var2), var4});
    }
