--- net/minecraft/client/renderer/block/model/multipart/AndCondition.java
+++ net/minecraft/client/renderer/block/model/multipart/AndCondition.java
@@ -7,10 +7,11 @@
 import net.minecraft.world.level.block.Block;
 import net.minecraft.world.level.block.state.BlockState;
 import net.minecraft.world.level.block.state.StateDefinition;
 
 public class AndCondition implements Condition {
+   public static final String TOKEN = "AND";
    private final Iterable<? extends Condition> conditions;
 
    public AndCondition(Iterable<? extends Condition> var1) {
       this.conditions = var1;
    }
