--- net/minecraft/client/renderer/block/model/multipart/MultiPart.java
+++ net/minecraft/client/renderer/block/model/multipart/MultiPart.java
@@ -116,11 +116,11 @@
          ArrayList var3 = Lists.newArrayList();
          Iterator var4 = var2.iterator();
 
          while(var4.hasNext()) {
             JsonElement var5 = (JsonElement)var4.next();
-            var3.add(var1.deserialize(var5, Selector.class));
+            var3.add((Selector)var1.deserialize(var5, Selector.class));
          }
 
          return var3;
       }
    }
