--- net/minecraft/client/renderer/block/model/multipart/package-info.java
+++ net/minecraft/client/renderer/block/model/multipart/package-info.java
@@ -1,0 +1,10 @@
+package net.minecraft.client.renderer.block.model.multipart;
+
+import javax.annotation.ParametersAreNonnullByDefault;
+import net.minecraft.MethodsReturnNonnullByDefault;
+
+// $FF: synthetic class
+@ParametersAreNonnullByDefault
+@MethodsReturnNonnullByDefault
+interface package-info {
+}
