--- net/minecraft/world/inventory/tooltip/package-info.java
+++ net/minecraft/world/inventory/tooltip/package-info.java
@@ -1,0 +1,12 @@
+package net.minecraft.world.inventory.tooltip;
+
+import javax.annotation.ParametersAreNonnullByDefault;
+import net.minecraft.FieldsAreNonnullByDefault;
+import net.minecraft.MethodsReturnNonnullByDefault;
+
+// $FF: synthetic class
+@ParametersAreNonnullByDefault
+@FieldsAreNonnullByDefault
+@MethodsReturnNonnullByDefault
+interface package-info {
+}
