--- net/minecraft/server/ChainedJsonException.java
+++ net/minecraft/server/ChainedJsonException.java
@@ -59,10 +59,15 @@
 
       private void addJsonKey(String var1) {
          this.jsonKeys.add(0, var1);
       }
 
+      @Nullable
+      public String getFilename() {
+         return this.filename;
+      }
+
       public String getJsonKeys() {
          return StringUtils.join(this.jsonKeys, "->");
       }
 
       public String toString() {
