--- net/minecraft/world/entity/vehicle/MinecartTNT.java
+++ net/minecraft/world/entity/vehicle/MinecartTNT.java
@@ -18,10 +18,11 @@
 import net.minecraft.world.level.block.Blocks;
 import net.minecraft.world.level.block.state.BlockState;
 import net.minecraft.world.level.material.FluidState;
 
 public class MinecartTNT extends AbstractMinecart {
+   private static final byte EVENT_PRIME = 10;
    private int fuse = -1;
 
    public MinecartTNT(EntityType<? extends MinecartTNT> var1, Level var2) {
       super(var1, var2);
    }
@@ -42,15 +43,15 @@
       super.tick();
       if (this.fuse > 0) {
          --this.fuse;
          this.level.addParticle(ParticleTypes.SMOKE, this.getX(), this.getY() + 0.5D, this.getZ(), 0.0D, 0.0D, 0.0D);
       } else if (this.fuse == 0) {
-         this.explode(getHorizontalDistanceSqr(this.getDeltaMovement()));
+         this.explode(this.getDeltaMovement().horizontalDistanceSqr());
       }
 
       if (this.horizontalCollision) {
-         double var1 = getHorizontalDistanceSqr(this.getDeltaMovement());
+         double var1 = this.getDeltaMovement().horizontalDistanceSqr();
          if (var1 >= 0.009999999776482582D) {
             this.explode(var1);
          }
       }
 
@@ -67,11 +68,11 @@
 
       return super.hurt(var1, var2);
    }
 
    public void destroy(DamageSource var1) {
-      double var2 = getHorizontalDistanceSqr(this.getDeltaMovement());
+      double var2 = this.getDeltaMovement().horizontalDistanceSqr();
       if (!var1.isFire() && !var1.isExplosion() && var2 < 0.009999999776482582D) {
          super.destroy(var1);
          if (!var1.isExplosion() && this.level.getGameRules().getBoolean(GameRules.RULE_DOENTITYDROPS)) {
             this.spawnAtLocation(Blocks.TNT);
          }
@@ -91,22 +92,22 @@
          if (var3 > 5.0D) {
             var3 = 5.0D;
          }
 
          this.level.explode(this, this.getX(), this.getY(), this.getZ(), (float)(4.0D + this.random.nextDouble() * 1.5D * var3), Explosion.BlockInteraction.BREAK);
-         this.remove();
+         this.discard();
       }
 
    }
 
-   public boolean causeFallDamage(float var1, float var2) {
+   public boolean causeFallDamage(float var1, float var2, DamageSource var3) {
       if (var1 >= 3.0F) {
-         float var3 = var1 / 10.0F;
-         this.explode((double)(var3 * var3));
+         float var4 = var1 / 10.0F;
+         this.explode((double)(var4 * var4));
       }
 
-      return super.causeFallDamage(var1, var2);
+      return super.causeFallDamage(var1, var2, var3);
    }
 
    public void activateMinecart(int var1, int var2, int var3, boolean var4) {
       if (var4 && this.fuse < 0) {
          this.primeFuse();
