--- net/minecraft/world/entity/vehicle/MinecartHopper.java
+++ net/minecraft/world/entity/vehicle/MinecartHopper.java
@@ -62,14 +62,10 @@
 
    public void setEnabled(boolean var1) {
       this.enabled = var1;
    }
 
-   public Level getLevel() {
-      return this.level;
-   }
-
    public double getLevelX() {
       return this.getX();
    }
 
    public double getLevelY() {
@@ -100,11 +96,11 @@
       }
 
    }
 
    public boolean suckInItems() {
-      if (HopperBlockEntity.suckInItems(this)) {
+      if (HopperBlockEntity.suckInItems(this.level, this)) {
          return true;
       } else {
          List var1 = this.level.getEntitiesOfClass(ItemEntity.class, this.getBoundingBox().inflate(0.25D, 0.0D, 0.25D), EntitySelector.ENTITY_STILL_ALIVE);
          if (!var1.isEmpty()) {
             HopperBlockEntity.addItem(this, (ItemEntity)var1.get(0));
