--- net/minecraft/world/entity/vehicle/MinecartTNT.java
+++ net/minecraft/world/entity/vehicle/MinecartTNT.java
@@ -91,22 +91,22 @@
          if (var3 > 5.0D) {
             var3 = 5.0D;
          }
 
          this.level.explode(this, this.getX(), this.getY(), this.getZ(), (float)(4.0D + this.random.nextDouble() * 1.5D * var3), Explosion.BlockInteraction.BREAK);
-         this.remove();
+         this.discard();
       }
 
    }
 
-   public boolean causeFallDamage(float var1, float var2) {
+   public boolean causeFallDamage(float var1, float var2, DamageSource var3) {
       if (var1 >= 3.0F) {
-         float var3 = var1 / 10.0F;
-         this.explode((double)(var3 * var3));
+         float var4 = var1 / 10.0F;
+         this.explode((double)(var4 * var4));
       }
 
-      return super.causeFallDamage(var1, var2);
+      return super.causeFallDamage(var1, var2, var3);
    }
 
    public void activateMinecart(int var1, int var2, int var3, boolean var4) {
       if (var4 && this.fuse < 0) {
          this.primeFuse();
