--- net/minecraft/world/entity/vehicle/MinecartHopper.java
+++ net/minecraft/world/entity/vehicle/MinecartHopper.java
@@ -16,10 +16,11 @@
 import net.minecraft.world.level.block.entity.Hopper;
 import net.minecraft.world.level.block.entity.HopperBlockEntity;
 import net.minecraft.world.level.block.state.BlockState;
 
 public class MinecartHopper extends AbstractMinecartContainer implements Hopper {
+   public static final int MOVE_ITEM_SPEED = 4;
    private boolean enabled = true;
    private int cooldownTime = -1;
    private final BlockPos lastPosition;
 
    public MinecartHopper(EntityType<? extends MinecartHopper> var1, Level var2) {
@@ -62,14 +63,10 @@
 
    public void setEnabled(boolean var1) {
       this.enabled = var1;
    }
 
-   public Level getLevel() {
-      return this.level;
-   }
-
    public double getLevelX() {
       return this.getX();
    }
 
    public double getLevelY() {
@@ -100,11 +97,11 @@
       }
 
    }
 
    public boolean suckInItems() {
-      if (HopperBlockEntity.suckInItems(this)) {
+      if (HopperBlockEntity.suckInItems(this.level, this)) {
          return true;
       } else {
          List var1 = this.level.getEntitiesOfClass(ItemEntity.class, this.getBoundingBox().inflate(0.25D, 0.0D, 0.25D), EntitySelector.ENTITY_STILL_ALIVE);
          if (!var1.isEmpty()) {
             HopperBlockEntity.addItem(this, (ItemEntity)var1.get(0));
