--- net/minecraft/world/food/Foods.java
+++ net/minecraft/world/food/Foods.java
@@ -40,10 +40,11 @@
    public static final FoodProperties ROTTEN_FLESH;
    public static final FoodProperties SALMON;
    public static final FoodProperties SPIDER_EYE;
    public static final FoodProperties SUSPICIOUS_STEW;
    public static final FoodProperties SWEET_BERRIES;
+   public static final FoodProperties GLOW_BERRIES;
    public static final FoodProperties TROPICAL_FISH;
 
    private static FoodProperties stew(int var0) {
       return (new FoodProperties.Builder()).nutrition(var0).saturationMod(0.6F).build();
    }
@@ -69,17 +70,18 @@
       MUSHROOM_STEW = stew(6);
       MUTTON = (new FoodProperties.Builder()).nutrition(2).saturationMod(0.3F).meat().build();
       POISONOUS_POTATO = (new FoodProperties.Builder()).nutrition(2).saturationMod(0.3F).effect(new MobEffectInstance(MobEffects.POISON, 100, 0), 0.6F).build();
       PORKCHOP = (new FoodProperties.Builder()).nutrition(3).saturationMod(0.3F).meat().build();
       POTATO = (new FoodProperties.Builder()).nutrition(1).saturationMod(0.3F).build();
-      PUFFERFISH = (new FoodProperties.Builder()).nutrition(1).saturationMod(0.1F).effect(new MobEffectInstance(MobEffects.POISON, 1200, 3), 1.0F).effect(new MobEffectInstance(MobEffects.HUNGER, 300, 2), 1.0F).effect(new MobEffectInstance(MobEffects.CONFUSION, 300, 0), 1.0F).build();
+      PUFFERFISH = (new FoodProperties.Builder()).nutrition(1).saturationMod(0.1F).effect(new MobEffectInstance(MobEffects.POISON, 1200, 1), 1.0F).effect(new MobEffectInstance(MobEffects.HUNGER, 300, 2), 1.0F).effect(new MobEffectInstance(MobEffects.CONFUSION, 300, 0), 1.0F).build();
       PUMPKIN_PIE = (new FoodProperties.Builder()).nutrition(8).saturationMod(0.3F).build();
       RABBIT = (new FoodProperties.Builder()).nutrition(3).saturationMod(0.3F).meat().build();
       RABBIT_STEW = stew(10);
       ROTTEN_FLESH = (new FoodProperties.Builder()).nutrition(4).saturationMod(0.1F).effect(new MobEffectInstance(MobEffects.HUNGER, 600, 0), 0.8F).meat().build();
       SALMON = (new FoodProperties.Builder()).nutrition(2).saturationMod(0.1F).build();
       SPIDER_EYE = (new FoodProperties.Builder()).nutrition(2).saturationMod(0.8F).effect(new MobEffectInstance(MobEffects.POISON, 100, 0), 1.0F).build();
       SUSPICIOUS_STEW = stew(6);
       SWEET_BERRIES = (new FoodProperties.Builder()).nutrition(2).saturationMod(0.1F).build();
+      GLOW_BERRIES = (new FoodProperties.Builder()).nutrition(2).saturationMod(0.1F).build();
       TROPICAL_FISH = (new FoodProperties.Builder()).nutrition(1).saturationMod(0.1F).build();
    }
 }
