--- net/minecraft/world/entity/npc/CatSpawner.java
+++ net/minecraft/world/entity/npc/CatSpawner.java
@@ -18,10 +18,11 @@
 import net.minecraft.world.level.NaturalSpawner;
 import net.minecraft.world.level.levelgen.feature.StructureFeature;
 import net.minecraft.world.phys.AABB;
 
 public class CatSpawner implements CustomSpawner {
+   private static final int TICK_DELAY = 1200;
    private int nextTick;
 
    public CatSpawner() {
    }
 
@@ -38,11 +39,12 @@
             } else {
                Random var5 = var1.random;
                int var6 = (8 + var5.nextInt(24)) * (var5.nextBoolean() ? -1 : 1);
                int var7 = (8 + var5.nextInt(24)) * (var5.nextBoolean() ? -1 : 1);
                BlockPos var8 = var4.blockPosition().offset(var6, 0, var7);
-               if (!var1.hasChunksAt(var8.getX() - 10, var8.getY() - 10, var8.getZ() - 10, var8.getX() + 10, var8.getY() + 10, var8.getZ() + 10)) {
+               boolean var9 = true;
+               if (!var1.hasChunksAt(var8.getX() - 10, var8.getZ() - 10, var8.getX() + 10, var8.getZ() + 10)) {
                   return 0;
                } else {
                   if (NaturalSpawner.isSpawnPositionOk(SpawnPlacements.Type.ON_GROUND, var1, var8, EntityType.CAT)) {
                      if (var1.isCloseToVillage(var8, 2)) {
                         return this.spawnInVillage(var1, var8);
