--- net/minecraft/world/entity/npc/VillagerType.java
+++ net/minecraft/world/entity/npc/VillagerType.java
@@ -78,9 +78,9 @@
       return (VillagerType)Registry.register(Registry.VILLAGER_TYPE, (ResourceLocation)(new ResourceLocation(var0)), new VillagerType(var0));
    }
 
    public static VillagerType byBiome(Optional<ResourceKey<Biome>> var0) {
       return (VillagerType)var0.flatMap((var0x) -> {
-         return Optional.ofNullable(BY_BIOME.get(var0x));
+         return Optional.ofNullable((VillagerType)BY_BIOME.get(var0x));
       }).orElse(PLAINS);
    }
 }
