--- net/minecraft/world/entity/npc/VillagerData.java
+++ net/minecraft/world/entity/npc/VillagerData.java
@@ -4,10 +4,12 @@
 import com.mojang.serialization.Codec;
 import com.mojang.serialization.codecs.RecordCodecBuilder;
 import net.minecraft.core.Registry;
 
 public class VillagerData {
+   public static final int MIN_VILLAGER_LEVEL = 1;
+   public static final int MAX_VILLAGER_LEVEL = 5;
    private static final int[] NEXT_LEVEL_XP_THRESHOLDS = new int[]{0, 10, 70, 150, 250};
    public static final Codec<VillagerData> CODEC = RecordCodecBuilder.create((var0) -> {
       return var0.group(Registry.VILLAGER_TYPE.fieldOf("type").orElseGet(() -> {
          return VillagerType.PLAINS;
       }).forGetter((var0x) -> {
@@ -52,10 +54,14 @@
 
    public VillagerData setLevel(int var1) {
       return new VillagerData(this.type, this.profession, var1);
    }
 
+   public static int getMinXpPerLevel(int var0) {
+      return canLevelUp(var0) ? NEXT_LEVEL_XP_THRESHOLDS[var0 - 1] : 0;
+   }
+
    public static int getMaxXpPerLevel(int var0) {
       return canLevelUp(var0) ? NEXT_LEVEL_XP_THRESHOLDS[var0] : 0;
    }
 
    public static boolean canLevelUp(int var0) {
