--- net/minecraft/util/thread/ProcessorMailbox.java
+++ net/minecraft/util/thread/ProcessorMailbox.java
@@ -3,11 +3,11 @@
 import it.unimi.dsi.fastutil.ints.Int2BooleanFunction;
 import java.util.concurrent.ConcurrentLinkedQueue;
 import java.util.concurrent.Executor;
 import java.util.concurrent.RejectedExecutionException;
 import java.util.concurrent.atomic.AtomicInteger;
-import net.minecraft.SharedConstants;
+import net.minecraft.Util;
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
 
 public class ProcessorMailbox<T> implements ProcessorHandle<T>, AutoCloseable, Runnable {
    private static final Logger LOGGER = LogManager.getLogger();
@@ -72,26 +72,11 @@
       } else {
          Runnable var1 = (Runnable)this.queue.pop();
          if (var1 == null) {
             return false;
          } else {
-            String var2;
-            Thread var3;
-            if (SharedConstants.IS_RUNNING_IN_IDE) {
-               var3 = Thread.currentThread();
-               var2 = var3.getName();
-               var3.setName(this.name);
-            } else {
-               var3 = null;
-               var2 = null;
-            }
-
-            var1.run();
-            if (var3 != null) {
-               var3.setName(var2);
-            }
-
+            Util.wrapThreadWithTaskName(this.name, var1).run();
             return true;
          }
       }
    }
 
