--- net/minecraft/client/Camera.java
+++ net/minecraft/client/Camera.java
@@ -1,18 +1,24 @@
 package net.minecraft.client;
 
 import com.mojang.math.Quaternion;
 import com.mojang.math.Vector3f;
+import java.util.Arrays;
+import java.util.Iterator;
+import java.util.List;
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.Direction;
+import net.minecraft.tags.FluidTags;
 import net.minecraft.util.Mth;
 import net.minecraft.world.entity.Entity;
 import net.minecraft.world.entity.LivingEntity;
 import net.minecraft.world.level.BlockGetter;
 import net.minecraft.world.level.ClipContext;
+import net.minecraft.world.level.block.Blocks;
+import net.minecraft.world.level.block.state.BlockState;
 import net.minecraft.world.level.material.FluidState;
-import net.minecraft.world.level.material.Fluids;
+import net.minecraft.world.level.material.FogType;
 import net.minecraft.world.phys.BlockHitResult;
 import net.minecraft.world.phys.HitResult;
 import net.minecraft.world.phys.Vec3;
 
 public class Camera {
@@ -26,13 +32,13 @@
    private final Vector3f left;
    private float xRot;
    private float yRot;
    private final Quaternion rotation;
    private boolean detached;
-   private boolean mirror;
    private float eyeHeight;
    private float eyeHeightOld;
+   public static final float FOG_DISTANCE_SCALE = 0.083333336F;
 
    public Camera() {
       this.position = Vec3.ZERO;
       this.blockPosition = new BlockPos.MutableBlockPos();
       this.forwards = new Vector3f(0.0F, 0.0F, 1.0F);
@@ -44,11 +50,10 @@
    public void setup(BlockGetter var1, Entity var2, boolean var3, boolean var4, float var5) {
       this.initialized = true;
       this.level = var1;
       this.entity = var2;
       this.detached = var3;
-      this.mirror = var4;
       this.setRotation(var2.getViewYRot(var5), var2.getViewXRot(var5));
       this.setPosition(Mth.lerp((double)var5, var2.xo, var2.getX()), Mth.lerp((double)var5, var2.yo, var2.getY()) + (double)Mth.lerp(var5, this.eyeHeightOld, this.eyeHeight), Mth.lerp((double)var5, var2.zo, var2.getZ()));
       if (var3) {
          if (var4) {
             this.setRotation(this.yRot + 180.0F, -this.xRot);
@@ -153,16 +158,52 @@
 
    public boolean isDetached() {
       return this.detached;
    }
 
-   public FluidState getFluidInCamera() {
+   public Camera.NearPlane getNearPlane() {
+      Minecraft var1 = Minecraft.getInstance();
+      double var2 = (double)var1.getWindow().getWidth() / (double)var1.getWindow().getHeight();
+      double var4 = Math.tan(var1.options.fov * 0.01745329238474369D / 2.0D) * 0.05000000074505806D;
+      double var6 = var4 * var2;
+      Vec3 var8 = (new Vec3(this.forwards)).scale(0.05000000074505806D);
+      Vec3 var9 = (new Vec3(this.left)).scale(var6);
+      Vec3 var10 = (new Vec3(this.up)).scale(var4);
+      return new Camera.NearPlane(var8, var9, var10);
+   }
+
+   public FogType getFluidInCamera() {
       if (!this.initialized) {
-         return Fluids.EMPTY.defaultFluidState();
+         return FogType.NONE;
       } else {
          FluidState var1 = this.level.getFluidState(this.blockPosition);
-         return !var1.isEmpty() && this.position.y >= (double)((float)this.blockPosition.getY() + var1.getHeight(this.level, this.blockPosition)) ? Fluids.EMPTY.defaultFluidState() : var1;
+         if (var1.is(FluidTags.WATER) && this.position.y < (double)((float)this.blockPosition.getY() + var1.getHeight(this.level, this.blockPosition))) {
+            return FogType.WATER;
+         } else {
+            Camera.NearPlane var2 = this.getNearPlane();
+            List var3 = Arrays.asList(var2.forward, var2.getTopLeft(), var2.getTopRight(), var2.getBottomLeft(), var2.getBottomRight());
+            Iterator var4 = var3.iterator();
+
+            while(var4.hasNext()) {
+               Vec3 var5 = (Vec3)var4.next();
+               Vec3 var6 = this.position.add(var5);
+               BlockPos var7 = new BlockPos(var6);
+               FluidState var8 = this.level.getFluidState(var7);
+               if (var8.is(FluidTags.LAVA)) {
+                  if (var6.y <= (double)(var8.getHeight(this.level, var7) + (float)var7.getY())) {
+                     return FogType.LAVA;
+                  }
+               } else {
+                  BlockState var9 = this.level.getBlockState(var7);
+                  if (var9.is(Blocks.POWDER_SNOW)) {
+                     return FogType.POWDER_SNOW;
+                  }
+               }
+            }
+
+            return FogType.NONE;
+         }
       }
    }
 
    public final Vector3f getLookVector() {
       return this.forwards;
@@ -170,11 +211,47 @@
 
    public final Vector3f getUpVector() {
       return this.up;
    }
 
+   public final Vector3f getLeftVector() {
+      return this.left;
+   }
+
    public void reset() {
       this.level = null;
       this.entity = null;
       this.initialized = false;
    }
+
+   public static class NearPlane {
+      final Vec3 forward;
+      private final Vec3 left;
+      private final Vec3 up;
+
+      NearPlane(Vec3 var1, Vec3 var2, Vec3 var3) {
+         this.forward = var1;
+         this.left = var2;
+         this.up = var3;
+      }
+
+      public Vec3 getTopLeft() {
+         return this.forward.add(this.up).add(this.left);
+      }
+
+      public Vec3 getTopRight() {
+         return this.forward.add(this.up).subtract(this.left);
+      }
+
+      public Vec3 getBottomLeft() {
+         return this.forward.subtract(this.up).add(this.left);
+      }
+
+      public Vec3 getBottomRight() {
+         return this.forward.subtract(this.up).subtract(this.left);
+      }
+
+      public Vec3 getPointOnPlane(float var1, float var2) {
+         return this.forward.add(this.up.scale((double)var2)).subtract(this.left.scale((double)var1));
+      }
+   }
 }
