--- net/minecraft/client/ClientRecipeBook.java
+++ net/minecraft/client/ClientRecipeBook.java
@@ -9,10 +9,11 @@
 import java.util.Collections;
 import java.util.HashMap;
 import java.util.Iterator;
 import java.util.List;
 import java.util.Map;
+import java.util.Objects;
 import java.util.stream.Stream;
 import net.minecraft.client.gui.screens.recipebook.RecipeCollection;
 import net.minecraft.core.Registry;
 import net.minecraft.stats.RecipeBook;
 import net.minecraft.world.item.BlockItem;
@@ -36,15 +37,15 @@
       Map var2 = categorizeAndGroupRecipes(var1);
       HashMap var3 = Maps.newHashMap();
       Builder var4 = ImmutableList.builder();
       var2.forEach((var2x, var3x) -> {
          Stream var10002 = var3x.stream().map(RecipeCollection::new);
-         var4.getClass();
-         List var10000 = (List)var3.put(var2x, var10002.peek(var4::add).collect(ImmutableList.toImmutableList()));
+         Objects.requireNonNull(var4);
+         var3.put(var2x, (List)var10002.peek(var4::add).collect(ImmutableList.toImmutableList()));
       });
       RecipeBookCategories.AGGREGATE_CATEGORIES.forEach((var1x, var2x) -> {
-         List var10000 = (List)var3.put(var1x, var2x.stream().flatMap((var1) -> {
+         var3.put(var1x, (List)var2x.stream().flatMap((var1) -> {
             return ((List)var3.getOrDefault(var1, ImmutableList.of())).stream();
          }).collect(ImmutableList.toImmutableList()));
       });
       this.collectionsByTab = ImmutableMap.copyOf(var3);
       this.allCollections = var4.build();
@@ -55,11 +56,11 @@
       HashBasedTable var2 = HashBasedTable.create();
       Iterator var3 = var0.iterator();
 
       while(var3.hasNext()) {
          Recipe var4 = (Recipe)var3.next();
-         if (!var4.isSpecial()) {
+         if (!var4.isSpecial() && !var4.isIncomplete()) {
             RecipeBookCategories var5 = getCategory(var4);
             String var6 = var4.getGroup();
             if (var6.isEmpty()) {
                ((List)var1.computeIfAbsent(var5, (var0x) -> {
                   return Lists.newArrayList();
@@ -109,13 +110,17 @@
       } else if (var1 == RecipeType.CAMPFIRE_COOKING) {
          return RecipeBookCategories.CAMPFIRE;
       } else if (var1 == RecipeType.SMITHING) {
          return RecipeBookCategories.SMITHING;
       } else {
-         LOGGER.warn("Unknown recipe category: {}/{}", new Supplier[]{() -> {
+         Logger var10000 = LOGGER;
+         Supplier[] var10002 = new Supplier[]{() -> {
             return Registry.RECIPE_TYPE.getKey(var0.getType());
-         }, var0::getId});
+         }, null};
+         Objects.requireNonNull(var0);
+         var10002[1] = var0::getId;
+         var10000.warn("Unknown recipe category: {}/{}", var10002);
          return RecipeBookCategories.UNKNOWN;
       }
    }
 
    public List<RecipeCollection> getCollections() {
