--- net/minecraft/client/ClientBrandRetriever.java
+++ net/minecraft/client/ClientBrandRetriever.java
@@ -1,10 +1,15 @@
 package net.minecraft.client;
 
+import net.minecraft.obfuscate.DontObfuscate;
+
 public class ClientBrandRetriever {
+   public static final String VANILLA_NAME = "vanilla";
+
    public ClientBrandRetriever() {
    }
 
+   @DontObfuscate
    public static String getClientModName() {
       return "vanilla";
    }
 }
