--- net/minecraft/client/ClientRecipeBook.java
+++ net/minecraft/client/ClientRecipeBook.java
@@ -55,11 +55,11 @@
       HashBasedTable var2 = HashBasedTable.create();
       Iterator var3 = var0.iterator();
 
       while(var3.hasNext()) {
          Recipe var4 = (Recipe)var3.next();
-         if (!var4.isSpecial()) {
+         if (!var4.isSpecial() && !var4.isIncomplete()) {
             RecipeBookCategories var5 = getCategory(var4);
             String var6 = var4.getGroup();
             if (var6.isEmpty()) {
                ((List)var1.computeIfAbsent(var5, (var0x) -> {
                   return Lists.newArrayList();
