--- net/minecraft/client/FullscreenResolutionProgressOption.java
+++ net/minecraft/client/FullscreenResolutionProgressOption.java
@@ -6,10 +6,12 @@
 import javax.annotation.Nullable;
 import net.minecraft.network.chat.TextComponent;
 import net.minecraft.network.chat.TranslatableComponent;
 
 public class FullscreenResolutionProgressOption extends ProgressOption {
+   private static final double CURRENT_MODE = -1.0D;
+
    public FullscreenResolutionProgressOption(Window var1) {
       this(var1, var1.findBestMonitor());
    }
 
    private FullscreenResolutionProgressOption(Window var1, @Nullable Monitor var2) {
