--- net/minecraft/client/Game.java
+++ net/minecraft/client/Game.java
@@ -71,11 +71,11 @@
 
    public void onLeaveGameSession() {
       this.listener.onLeaveGameSession(this.getCurrentSession());
    }
 
-   static class Metrics implements PerformanceMetrics {
+   private static class Metrics implements PerformanceMetrics {
       private final int min;
       private final int max;
       private final int average;
       private final int samples;
 
