--- net/minecraft/client/HotbarManager.java
+++ net/minecraft/client/HotbarManager.java
@@ -11,10 +11,11 @@
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
 
 public class HotbarManager {
    private static final Logger LOGGER = LogManager.getLogger();
+   public static final int NUM_HOTBAR_GROUPS = 9;
    private final File optionsFile;
    private final DataFixer fixerUpper;
    private final Hotbar[] hotbars = new Hotbar[9];
    private boolean loaded;
 
