--- net/minecraft/client/KeyMapping.java
+++ net/minecraft/client/KeyMapping.java
@@ -14,10 +14,17 @@
 
 public class KeyMapping implements Comparable<KeyMapping> {
    private static final Map<String, KeyMapping> ALL = Maps.newHashMap();
    private static final Map<InputConstants.Key, KeyMapping> MAP = Maps.newHashMap();
    private static final Set<String> CATEGORIES = Sets.newHashSet();
+   public static final String CATEGORY_MOVEMENT = "key.categories.movement";
+   public static final String CATEGORY_MISC = "key.categories.misc";
+   public static final String CATEGORY_MULTIPLAYER = "key.categories.multiplayer";
+   public static final String CATEGORY_GAMEPLAY = "key.categories.gameplay";
+   public static final String CATEGORY_INVENTORY = "key.categories.inventory";
+   public static final String CATEGORY_INTERFACE = "key.categories.ui";
+   public static final String CATEGORY_CREATIVE = "key.categories.creative";
    private static final Map<String, Integer> CATEGORY_SORT_ORDER = (Map)Util.make(Maps.newHashMap(), (var0) -> {
       var0.put("key.categories.movement", 1);
       var0.put("key.categories.gameplay", 2);
       var0.put("key.categories.inventory", 3);
       var0.put("key.categories.creative", 4);
