--- net/minecraft/client/Minecraft.java
+++ net/minecraft/client/Minecraft.java
@@ -1,61 +1,77 @@
 package net.minecraft.client;
 
+import com.google.common.collect.ImmutableList;
 import com.google.common.collect.ImmutableMap;
 import com.google.common.collect.Queues;
 import com.google.common.collect.UnmodifiableIterator;
+import com.google.common.hash.Hashing;
 import com.mojang.authlib.GameProfile;
 import com.mojang.authlib.GameProfileRepository;
 import com.mojang.authlib.exceptions.AuthenticationException;
 import com.mojang.authlib.minecraft.MinecraftSessionService;
 import com.mojang.authlib.minecraft.OfflineSocialInteractions;
 import com.mojang.authlib.minecraft.SocialInteractionsService;
 import com.mojang.authlib.properties.PropertyMap;
 import com.mojang.authlib.yggdrasil.YggdrasilAuthenticationService;
+import com.mojang.blaze3d.pipeline.MainTarget;
 import com.mojang.blaze3d.pipeline.RenderTarget;
+import com.mojang.blaze3d.pipeline.TextureTarget;
 import com.mojang.blaze3d.platform.DisplayData;
+import com.mojang.blaze3d.platform.GlDebug;
 import com.mojang.blaze3d.platform.GlUtil;
 import com.mojang.blaze3d.platform.Window;
 import com.mojang.blaze3d.platform.WindowEventHandler;
 import com.mojang.blaze3d.systems.RenderSystem;
 import com.mojang.blaze3d.vertex.BufferBuilder;
+import com.mojang.blaze3d.vertex.BufferUploader;
 import com.mojang.blaze3d.vertex.DefaultVertexFormat;
 import com.mojang.blaze3d.vertex.PoseStack;
 import com.mojang.blaze3d.vertex.Tesselator;
+import com.mojang.blaze3d.vertex.VertexFormat;
 import com.mojang.datafixers.DataFixer;
 import com.mojang.datafixers.util.Function4;
+import com.mojang.math.Matrix4f;
 import com.mojang.serialization.DataResult;
 import com.mojang.serialization.JsonOps;
 import com.mojang.serialization.Lifecycle;
 import java.io.File;
 import java.io.IOException;
 import java.io.InputStream;
+import java.io.UncheckedIOException;
 import java.net.Proxy;
 import java.net.SocketAddress;
+import java.nio.ByteBuffer;
 import java.nio.ByteOrder;
+import java.nio.file.Path;
+import java.nio.file.Paths;
 import java.text.DecimalFormat;
 import java.text.DecimalFormatSymbols;
 import java.text.SimpleDateFormat;
 import java.util.Collections;
 import java.util.Date;
 import java.util.Iterator;
 import java.util.List;
 import java.util.Locale;
+import java.util.Objects;
 import java.util.Queue;
 import java.util.UUID;
 import java.util.concurrent.CompletableFuture;
 import java.util.concurrent.ExecutionException;
 import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Consumer;
 import java.util.function.Function;
 import java.util.function.Supplier;
 import java.util.stream.Stream;
 import javax.annotation.Nullable;
 import net.minecraft.ChatFormatting;
 import net.minecraft.CrashReport;
 import net.minecraft.CrashReportCategory;
+import net.minecraft.FileUtil;
 import net.minecraft.ReportedException;
 import net.minecraft.SharedConstants;
+import net.minecraft.SystemReport;
 import net.minecraft.Util;
 import net.minecraft.client.color.block.BlockColors;
 import net.minecraft.client.color.item.ItemColors;
 import net.minecraft.client.gui.Font;
 import net.minecraft.client.gui.Gui;
@@ -83,22 +99,25 @@
 import net.minecraft.client.gui.screens.TitleScreen;
 import net.minecraft.client.gui.screens.WinScreen;
 import net.minecraft.client.gui.screens.advancements.AdvancementsScreen;
 import net.minecraft.client.gui.screens.inventory.CreativeModeInventoryScreen;
 import net.minecraft.client.gui.screens.inventory.InventoryScreen;
-import net.minecraft.client.gui.screens.multiplayer.JoinMultiplayerScreen;
 import net.minecraft.client.gui.screens.social.PlayerSocialManager;
 import net.minecraft.client.gui.screens.social.SocialInteractionsScreen;
 import net.minecraft.client.gui.screens.worldselection.EditWorldScreen;
 import net.minecraft.client.main.GameConfig;
+import net.minecraft.client.model.geom.EntityModelSet;
 import net.minecraft.client.multiplayer.ClientHandshakePacketListenerImpl;
 import net.minecraft.client.multiplayer.ClientLevel;
 import net.minecraft.client.multiplayer.ClientPacketListener;
 import net.minecraft.client.multiplayer.MultiPlayerGameMode;
 import net.minecraft.client.multiplayer.ServerData;
+import net.minecraft.client.multiplayer.resolver.ServerAddress;
 import net.minecraft.client.particle.ParticleEngine;
 import net.minecraft.client.player.LocalPlayer;
+import net.minecraft.client.profiling.ClientMetricsSamplersProvider;
+import net.minecraft.client.renderer.BlockEntityWithoutLevelRenderer;
 import net.minecraft.client.renderer.FogRenderer;
 import net.minecraft.client.renderer.GameRenderer;
 import net.minecraft.client.renderer.GpuWarnlistManager;
 import net.minecraft.client.renderer.ItemInHandRenderer;
 import net.minecraft.client.renderer.LevelRenderer;
@@ -107,10 +126,11 @@
 import net.minecraft.client.renderer.block.BlockModelShaper;
 import net.minecraft.client.renderer.block.BlockRenderDispatcher;
 import net.minecraft.client.renderer.blockentity.BlockEntityRenderDispatcher;
 import net.minecraft.client.renderer.debug.DebugRenderer;
 import net.minecraft.client.renderer.entity.EntityRenderDispatcher;
+import net.minecraft.client.renderer.entity.EntityRenderers;
 import net.minecraft.client.renderer.entity.ItemRenderer;
 import net.minecraft.client.renderer.texture.TextureAtlas;
 import net.minecraft.client.renderer.texture.TextureAtlasSprite;
 import net.minecraft.client.renderer.texture.TextureManager;
 import net.minecraft.client.resources.ClientPackSource;
@@ -144,13 +164,15 @@
 import net.minecraft.nbt.ListTag;
 import net.minecraft.nbt.NbtOps;
 import net.minecraft.nbt.StringTag;
 import net.minecraft.network.Connection;
 import net.minecraft.network.ConnectionProtocol;
+import net.minecraft.network.chat.ClickEvent;
 import net.minecraft.network.chat.CommonComponents;
 import net.minecraft.network.chat.Component;
 import net.minecraft.network.chat.KeybindComponent;
+import net.minecraft.network.chat.MutableComponent;
 import net.minecraft.network.chat.TextComponent;
 import net.minecraft.network.chat.TranslatableComponent;
 import net.minecraft.network.protocol.Packet;
 import net.minecraft.network.protocol.game.ServerboundPlayerActionPacket;
 import net.minecraft.network.protocol.handshake.ClientIntentionPacket;
@@ -177,43 +199,40 @@
 import net.minecraft.server.packs.resources.SimpleReloadableResourceManager;
 import net.minecraft.server.players.GameProfileCache;
 import net.minecraft.sounds.Music;
 import net.minecraft.sounds.Musics;
 import net.minecraft.tags.ItemTags;
+import net.minecraft.util.FileZipper;
 import net.minecraft.util.FrameTimer;
 import net.minecraft.util.Mth;
+import net.minecraft.util.TimeUtil;
 import net.minecraft.util.Unit;
 import net.minecraft.util.datafix.DataFixers;
 import net.minecraft.util.profiling.ContinuousProfiler;
 import net.minecraft.util.profiling.InactiveProfiler;
 import net.minecraft.util.profiling.ProfileResults;
 import net.minecraft.util.profiling.ProfilerFiller;
 import net.minecraft.util.profiling.ResultField;
 import net.minecraft.util.profiling.SingleTickProfiler;
+import net.minecraft.util.profiling.metrics.profiling.ActiveMetricsRecorder;
+import net.minecraft.util.profiling.metrics.profiling.InactiveMetricsRecorder;
+import net.minecraft.util.profiling.metrics.profiling.MetricsRecorder;
+import net.minecraft.util.profiling.metrics.storage.MetricsPersister;
 import net.minecraft.util.thread.ReentrantBlockableEventLoop;
 import net.minecraft.world.InteractionHand;
 import net.minecraft.world.InteractionResult;
 import net.minecraft.world.Snooper;
 import net.minecraft.world.SnooperPopulator;
 import net.minecraft.world.entity.Entity;
 import net.minecraft.world.entity.EntityType;
-import net.minecraft.world.entity.boss.enderdragon.EndCrystal;
-import net.minecraft.world.entity.decoration.ArmorStand;
-import net.minecraft.world.entity.decoration.ItemFrame;
-import net.minecraft.world.entity.decoration.LeashFenceKnotEntity;
-import net.minecraft.world.entity.decoration.Painting;
 import net.minecraft.world.entity.player.ChatVisiblity;
 import net.minecraft.world.entity.player.Inventory;
 import net.minecraft.world.entity.player.Player;
-import net.minecraft.world.entity.vehicle.AbstractMinecart;
-import net.minecraft.world.entity.vehicle.Boat;
 import net.minecraft.world.item.CreativeModeTab;
 import net.minecraft.world.item.Item;
 import net.minecraft.world.item.ItemStack;
-import net.minecraft.world.item.Items;
 import net.minecraft.world.item.PlayerHeadItem;
-import net.minecraft.world.item.SpawnEggItem;
 import net.minecraft.world.item.TooltipFlag;
 import net.minecraft.world.level.DataPackConfig;
 import net.minecraft.world.level.Level;
 import net.minecraft.world.level.LevelSettings;
 import net.minecraft.world.level.biome.Biome;
@@ -228,22 +247,27 @@
 import net.minecraft.world.level.storage.PrimaryLevelData;
 import net.minecraft.world.level.storage.WorldData;
 import net.minecraft.world.phys.BlockHitResult;
 import net.minecraft.world.phys.EntityHitResult;
 import net.minecraft.world.phys.HitResult;
+import org.apache.commons.io.Charsets;
+import org.apache.commons.io.FileUtils;
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
+import org.lwjgl.util.tinyfd.TinyFileDialogs;
 
 public class Minecraft extends ReentrantBlockableEventLoop<Runnable> implements SnooperPopulator, WindowEventHandler {
    private static Minecraft instance;
    private static final Logger LOGGER = LogManager.getLogger();
    public static final boolean ON_OSX;
+   private static final int MAX_TICKS_PER_UPDATE = 10;
    public static final ResourceLocation DEFAULT_FONT;
    public static final ResourceLocation UNIFORM_FONT;
    public static final ResourceLocation ALT_FONT;
    private static final CompletableFuture<Unit> RESOURCE_RELOAD_INITIAL_TASK;
    private static final Component SOCIAL_INTERACTIONS_NOT_AVAILABLE;
+   public static final String UPDATE_DRIVERS_ADVICE = "Please make sure you have up-to-date drivers (see aka.ms/mcdriver for instructions).";
    private final File resourcePackDirectory;
    private final PropertyMap profileProperties;
    private final TextureManager textureManager;
    private final DataFixer fixerUpper;
    private final VirtualScreen virtualScreen;
@@ -298,10 +322,12 @@
    private final MobEffectTextureManager mobEffectTextures;
    private final ToastComponent toast;
    private final Game game = new Game(this);
    private final Tutorial tutorial;
    private final PlayerSocialManager playerSocialManager;
+   private final EntityModelSet entityModels;
+   private final BlockEntityRenderDispatcher blockEntityRenderDispatcher;
    public static byte[] reserve;
    @Nullable
    public MultiPlayerGameMode gameMode;
    @Nullable
    public ClientLevel level;
@@ -329,18 +355,19 @@
    private int frames;
    public boolean noRender;
    @Nullable
    public Screen screen;
    @Nullable
-   public Overlay overlay;
+   private Overlay overlay;
    private boolean connectedToRealms;
    private Thread gameThread;
    private volatile boolean running = true;
    @Nullable
    private CrashReport delayedCrash;
    private static int fps;
    public String fpsString = "";
+   public boolean wireframe;
    public boolean chunkPath;
    public boolean chunkVisibility;
    public boolean smartCull = true;
    private boolean windowActive;
    private final Queue<Runnable> progressTasks = Queues.newConcurrentLinkedQueue();
@@ -351,18 +378,22 @@
    private ProfilerFiller profiler;
    private int fpsPieRenderTicks;
    private final ContinuousProfiler fpsPieProfiler;
    @Nullable
    private ProfileResults fpsPieResults;
+   private MetricsRecorder metricsRecorder;
+   private final ResourceLoadStateTracker reloadStateTracker;
    private String debugPath;
 
    public Minecraft(GameConfig var1) {
       super("Client");
       this.profiler = InactiveProfiler.INSTANCE;
       this.fpsPieProfiler = new ContinuousProfiler(Util.timeSource, () -> {
          return this.fpsPieRenderTicks;
       });
+      this.metricsRecorder = InactiveMetricsRecorder.INSTANCE;
+      this.reloadStateTracker = new ResourceLoadStateTracker();
       this.debugPath = "root";
       instance = this;
       this.gameDirectory = var1.location.gameDirectory;
       File var2 = var1.location.assetDirectory;
       this.resourcePackDirectory = var1.location.resourcePackDirectory;
@@ -394,13 +425,13 @@
       }
 
       KeybindComponent.setKeyResolver(KeyMapping::createNameSupplier);
       this.fixerUpper = DataFixers.getDataFixer();
       this.toast = new ToastComponent(this);
-      this.tutorial = new Tutorial(this);
       this.gameThread = Thread.currentThread();
       this.options = new Options(this, this.gameDirectory);
+      this.tutorial = new Tutorial(this, this.options);
       this.hotbarManager = new HotbarManager(this.gameDirectory, this.fixerUpper);
       LOGGER.info("Backend library: {}", RenderSystem.getBackendDescription());
       DisplayData var6;
       if (this.options.overrideHeight > 0 && this.options.overrideWidth > 0) {
          var6 = new DisplayData(this.options.overrideWidth, this.options.overrideHeight, var1.display.fullscreenWidth, var1.display.fullscreenHeight, var1.display.isFullscreen);
@@ -425,12 +456,13 @@
       this.mouseHandler = new MouseHandler(this);
       this.mouseHandler.setup(this.window.getWindow());
       this.keyboardHandler = new KeyboardHandler(this);
       this.keyboardHandler.setup(this.window.getWindow());
       RenderSystem.initRenderer(this.options.glDebugVerbosity, false);
-      this.mainRenderTarget = new RenderTarget(this.window.getWidth(), this.window.getHeight(), true, ON_OSX);
+      this.mainRenderTarget = new MainTarget(this.window.getWidth(), this.window.getHeight());
       this.mainRenderTarget.setClearColor(0.0F, 0.0F, 0.0F, 0.0F);
+      this.mainRenderTarget.clear(ON_OSX);
       this.resourceManager = new SimpleReloadableResourceManager(PackType.CLIENT_RESOURCES);
       this.resourcePackRepository.reload();
       this.options.loadSelectedResourcePacks(this.resourcePackRepository);
       this.languageManager = new LanguageManager(this.options.languageCode);
       this.resourceManager.registerReloadListener(this.languageManager);
@@ -454,19 +486,26 @@
       this.window.setErrorSection("Post startup");
       this.blockColors = BlockColors.createDefault();
       this.itemColors = ItemColors.createDefault(this.blockColors);
       this.modelManager = new ModelManager(this.textureManager, this.blockColors, this.options.mipmapLevels);
       this.resourceManager.registerReloadListener(this.modelManager);
-      this.itemRenderer = new ItemRenderer(this.textureManager, this.modelManager, this.itemColors);
-      this.entityRenderDispatcher = new EntityRenderDispatcher(this.textureManager, this.itemRenderer, this.resourceManager, this.font, this.options);
+      this.entityModels = new EntityModelSet();
+      this.resourceManager.registerReloadListener(this.entityModels);
+      this.blockEntityRenderDispatcher = new BlockEntityRenderDispatcher(this.font, this.entityModels, this::getBlockRenderer);
+      this.resourceManager.registerReloadListener(this.blockEntityRenderDispatcher);
+      BlockEntityWithoutLevelRenderer var10 = new BlockEntityWithoutLevelRenderer(this.blockEntityRenderDispatcher, this.entityModels);
+      this.resourceManager.registerReloadListener(var10);
+      this.itemRenderer = new ItemRenderer(this.textureManager, this.modelManager, this.itemColors, var10);
+      this.entityRenderDispatcher = new EntityRenderDispatcher(this.textureManager, this.itemRenderer, this.font, this.options, this.entityModels);
+      this.resourceManager.registerReloadListener(this.entityRenderDispatcher);
       this.itemInHandRenderer = new ItemInHandRenderer(this);
       this.resourceManager.registerReloadListener(this.itemRenderer);
       this.renderBuffers = new RenderBuffers();
       this.gameRenderer = new GameRenderer(this, this.resourceManager, this.renderBuffers);
       this.resourceManager.registerReloadListener(this.gameRenderer);
       this.playerSocialManager = new PlayerSocialManager(this, this.socialInteractionsService);
-      this.blockRenderer = new BlockRenderDispatcher(this.modelManager.getBlockModelShaper(), this.blockColors);
+      this.blockRenderer = new BlockRenderDispatcher(this.modelManager.getBlockModelShaper(), var10, this.blockColors);
       this.resourceManager.registerReloadListener(this.blockRenderer);
       this.levelRenderer = new LevelRenderer(this, this.renderBuffers);
       this.resourceManager.registerReloadListener(this.levelRenderer);
       this.createSearchTrees();
       this.resourceManager.registerReloadListener(this.searchRegistry);
@@ -479,33 +518,47 @@
       this.gpuWarnlistManager = new GpuWarnlistManager();
       this.resourceManager.registerReloadListener(this.gpuWarnlistManager);
       this.gui = new Gui(this);
       this.debugRenderer = new DebugRenderer(this);
       RenderSystem.setErrorCallback(this::onFullscreenError);
-      if (this.options.fullscreen && !this.window.isFullscreen()) {
-         this.window.toggleFullScreen();
-         this.options.fullscreen = this.window.isFullscreen();
+      if (this.mainRenderTarget.width == this.window.getWidth() && this.mainRenderTarget.height == this.window.getHeight()) {
+         if (this.options.fullscreen && !this.window.isFullscreen()) {
+            this.window.toggleFullScreen();
+            this.options.fullscreen = this.window.isFullscreen();
+         }
+      } else {
+         int var10002 = this.window.getWidth();
+         StringBuilder var11 = new StringBuilder("Recovering from unsupported resolution (" + var10002 + "x" + this.window.getHeight() + ").\nPlease make sure you have up-to-date drivers (see aka.ms/mcdriver for instructions).");
+         if (GlDebug.isDebugEnabled()) {
+            var11.append("\n\nReported GL debug messages:\n").append(String.join("\n", GlDebug.getLastOpenGlDebugMessages()));
+         }
+
+         this.window.setWindowed(this.mainRenderTarget.width, this.mainRenderTarget.height);
+         TinyFileDialogs.tinyfd_messageBox("Minecraft", var11.toString(), "ok", "error", false);
       }
 
       this.window.updateVsync(this.options.enableVsync);
       this.window.updateRawMouseInput(this.options.rawMouseInput);
       this.window.setDefaultErrorCallback();
       this.resizeDisplay();
       if (var4 != null) {
-         this.setScreen(new ConnectScreen(new TitleScreen(), this, var4, var5));
+         ConnectScreen.startConnecting(new TitleScreen(), this, new ServerAddress(var4, var5), (ServerData)null);
       } else {
          this.setScreen(new TitleScreen(true));
       }
 
+      this.gameRenderer.preloadUiShader(this.getClientPackSource().getVanillaPack());
       LoadingOverlay.registerTextures(this);
-      List var10 = this.resourcePackRepository.openAllSelected();
-      this.setOverlay(new LoadingOverlay(this, this.resourceManager.createFullReload(Util.backgroundExecutor(), this, RESOURCE_RELOAD_INITIAL_TASK, var10), (var1x) -> {
+      List var12 = this.resourcePackRepository.openAllSelected();
+      this.reloadStateTracker.startReload(ResourceLoadStateTracker.ReloadReason.INITIAL, var12);
+      this.setOverlay(new LoadingOverlay(this, this.resourceManager.createReload(Util.backgroundExecutor(), this, RESOURCE_RELOAD_INITIAL_TASK, var12), (var1x) -> {
          Util.ifElse(var1x, this::rollbackResourcePacks, () -> {
             if (SharedConstants.IS_RUNNING_IN_IDE) {
                this.selfTest();
             }
 
+            this.reloadStateTracker.finishReload();
          });
       }, false));
    }
 
    public void updateTitle() {
@@ -566,15 +619,16 @@
 
    }
 
    public void clearResourcePacksOnError(Throwable var1, @Nullable Component var2) {
       LOGGER.info("Caught error loading resourcepacks, removing all selected resourcepacks", var1);
+      this.reloadStateTracker.startRecovery(var1);
       this.resourcePackRepository.setSelected(Collections.emptyList());
       this.options.resourcePacks.clear();
       this.options.incompatibleResourcePacks.clear();
       this.options.save();
-      this.reloadResourcePacks().thenRun(() -> {
+      this.reloadResourcePacks(true).thenRun(() -> {
          ToastComponent var2x = this.getToasts();
          SystemToast.addOrUpdate(var2x, SystemToast.SystemToastIds.PACK_LOAD_FAILURE, new TranslatableComponent("resourcePack.load_fail"), var2);
       });
    }
 
@@ -591,13 +645,15 @@
             }
 
             try {
                SingleTickProfiler var7 = SingleTickProfiler.createTickProfiler("Renderer");
                boolean var3 = this.shouldRenderFpsPie();
-               this.startProfilers(var3, var7);
+               this.profiler = this.constructProfiler(var3, var7);
                this.profiler.startTick();
+               this.metricsRecorder.startTick();
                this.runTick(!var1);
+               this.metricsRecorder.endTick();
                this.profiler.endTick();
                this.finishProfilers(var3, var7);
             } catch (OutOfMemoryError var4) {
                if (var1) {
                   throw var4;
@@ -708,11 +764,13 @@
       this.delayedCrash = var1;
    }
 
    public static void crash(CrashReport var0) {
       File var1 = new File(getInstance().gameDirectory, "crash-reports");
-      File var2 = new File(var1, "crash-" + (new SimpleDateFormat("yyyy-MM-dd_HH.mm.ss")).format(new Date()) + "-client.txt");
+      SimpleDateFormat var10003 = new SimpleDateFormat("yyyy-MM-dd_HH.mm.ss");
+      Date var10004 = new Date();
+      File var2 = new File(var1, "crash-" + var10003.format(var10004) + "-client.txt");
       Bootstrap.realStdoutPrintln(var0.getFriendlyReport());
       if (var0.getSaveFile() != null) {
          Bootstrap.realStdoutPrintln("#@!@# Game crashed! Crash report saved to: #@!@# " + var0.getSaveFile());
          System.exit(-1);
       } else if (var0.saveToFile(var2)) {
@@ -728,27 +786,36 @@
    public boolean isEnforceUnicode() {
       return this.options.forceUnicodeFont;
    }
 
    public CompletableFuture<Void> reloadResourcePacks() {
+      return this.reloadResourcePacks(false);
+   }
+
+   private CompletableFuture<Void> reloadResourcePacks(boolean var1) {
       if (this.pendingReload != null) {
          return this.pendingReload;
       } else {
-         CompletableFuture var1 = new CompletableFuture();
-         if (this.overlay instanceof LoadingOverlay) {
-            this.pendingReload = var1;
-            return var1;
+         CompletableFuture var2 = new CompletableFuture();
+         if (!var1 && this.overlay instanceof LoadingOverlay) {
+            this.pendingReload = var2;
+            return var2;
          } else {
             this.resourcePackRepository.reload();
-            List var2 = this.resourcePackRepository.openAllSelected();
-            this.setOverlay(new LoadingOverlay(this, this.resourceManager.createFullReload(Util.backgroundExecutor(), this, RESOURCE_RELOAD_INITIAL_TASK, var2), (var2x) -> {
+            List var3 = this.resourcePackRepository.openAllSelected();
+            if (!var1) {
+               this.reloadStateTracker.startReload(ResourceLoadStateTracker.ReloadReason.MANUAL, var3);
+            }
+
+            this.setOverlay(new LoadingOverlay(this, this.resourceManager.createReload(Util.backgroundExecutor(), this, RESOURCE_RELOAD_INITIAL_TASK, var3), (var2x) -> {
                Util.ifElse(var2x, this::rollbackResourcePacks, () -> {
                   this.levelRenderer.allChanged();
-                  var1.complete((Object)null);
+                  this.reloadStateTracker.finishReload();
+                  var2.complete((Object)null);
                });
             }, true));
-            return var1;
+            return var2;
          }
       }
    }
 
    private void selfTest() {
@@ -808,31 +875,35 @@
             }
          }
       }
 
       var1 |= MenuScreens.selfTest();
+      var1 |= EntityRenderers.validateRegistrations();
       if (var1) {
          throw new IllegalStateException("Your game data is foobar, fix the errors above!");
       }
    }
 
    public LevelStorageSource getLevelSource() {
       return this.levelSource;
    }
 
    private void openChatScreen(String var1) {
-      if (!this.isLocalServer() && !this.allowsChat()) {
-         if (this.player != null) {
-            this.player.sendMessage((new TranslatableComponent("chat.cannotSend")).withStyle(ChatFormatting.RED), Util.NIL_UUID);
-         }
+      Minecraft.ChatStatus var2 = this.getChatStatus();
+      if (!var2.isChatAllowed(this.isLocalServer())) {
+         this.gui.setOverlayMessage(var2.getMessage(), false);
       } else {
          this.setScreen(new ChatScreen(var1));
       }
 
    }
 
    public void setScreen(@Nullable Screen var1) {
+      if (SharedConstants.IS_RUNNING_IN_IDE && Thread.currentThread() != this.gameThread) {
+         LOGGER.error("setScreen called from non-game thread");
+      }
+
       if (this.screen != null) {
          this.screen.removed();
       }
 
       if (var1 == null && this.level == null) {
@@ -843,22 +914,17 @@
          } else {
             this.player.respawn();
          }
       }
 
-      if (var1 instanceof TitleScreen || var1 instanceof JoinMultiplayerScreen) {
-         this.options.renderDebug = false;
-         this.gui.getChat().clearMessages(true);
-      }
-
       this.screen = (Screen)var1;
+      BufferUploader.reset();
       if (var1 != null) {
          this.mouseHandler.releaseMouse();
          KeyMapping.releaseAll();
          ((Screen)var1).init(this, this.window.getGuiScaledWidth(), this.window.getGuiScaledHeight());
          this.noRender = false;
-         NarratorChatListener.INSTANCE.sayNow(((Screen)var1).getNarrationMessage());
       } else {
          this.soundManager.resume();
          this.mouseHandler.grabMouse();
       }
 
@@ -939,24 +1005,24 @@
          this.reloadResourcePacks().thenRun(() -> {
             var4.complete((Object)null);
          });
       }
 
-      Runnable var9;
-      while((var9 = (Runnable)this.progressTasks.poll()) != null) {
-         var9.run();
+      Runnable var10;
+      while((var10 = (Runnable)this.progressTasks.poll()) != null) {
+         var10.run();
       }
 
-      int var5;
+      int var6;
       if (var1) {
-         var5 = this.timer.advanceTime(Util.getMillis());
+         int var5 = this.timer.advanceTime(Util.getMillis());
          this.profiler.push("scheduledExecutables");
          this.runAllTasks();
          this.profiler.pop();
          this.profiler.push("tick");
 
-         for(int var6 = 0; var6 < Math.min(10, var5); ++var6) {
+         for(var6 = 0; var6 < Math.min(10, var5); ++var6) {
             this.profiler.incrementCounter("clientTick");
             this.tick();
          }
 
          this.profiler.pop();
@@ -966,11 +1032,13 @@
       this.window.setErrorSection("Render");
       this.profiler.push("sound");
       this.soundManager.updateSource(this.gameRenderer.getMainCamera());
       this.profiler.pop();
       this.profiler.push("render");
-      RenderSystem.pushMatrix();
+      PoseStack var11 = RenderSystem.getModelViewStack();
+      var11.pushPose();
+      RenderSystem.applyModelViewMatrix();
       RenderSystem.clear(16640, ON_OSX);
       this.mainRenderTarget.bindWrite(true);
       FogRenderer.setupNoFog();
       this.profiler.push("display");
       RenderSystem.enableTexture();
@@ -990,40 +1058,42 @@
          this.profiler.pop();
       }
 
       this.profiler.push("blit");
       this.mainRenderTarget.unbindWrite();
-      RenderSystem.popMatrix();
-      RenderSystem.pushMatrix();
+      var11.popPose();
+      var11.pushPose();
+      RenderSystem.applyModelViewMatrix();
       this.mainRenderTarget.blitToScreen(this.window.getWidth(), this.window.getHeight());
-      RenderSystem.popMatrix();
+      var11.popPose();
+      RenderSystem.applyModelViewMatrix();
       this.profiler.popPush("updateDisplay");
       this.window.updateDisplay();
-      var5 = this.getFramerateLimit();
-      if ((double)var5 < Option.FRAMERATE_LIMIT.getMaxValue()) {
-         RenderSystem.limitDisplayFPS(var5);
+      var6 = this.getFramerateLimit();
+      if ((double)var6 < Option.FRAMERATE_LIMIT.getMaxValue()) {
+         RenderSystem.limitDisplayFPS(var6);
       }
 
       this.profiler.popPush("yield");
       Thread.yield();
       this.profiler.pop();
       this.window.setErrorSection("Post render");
       ++this.frames;
-      boolean var10 = this.hasSingleplayerServer() && (this.screen != null && this.screen.isPauseScreen() || this.overlay != null && this.overlay.isPauseScreen()) && !this.singleplayerServer.isPublished();
-      if (this.pause != var10) {
+      boolean var7 = this.hasSingleplayerServer() && (this.screen != null && this.screen.isPauseScreen() || this.overlay != null && this.overlay.isPauseScreen()) && !this.singleplayerServer.isPublished();
+      if (this.pause != var7) {
          if (this.pause) {
             this.pausePartialTick = this.timer.partialTick;
          } else {
             this.timer.partialTick = this.pausePartialTick;
          }
 
-         this.pause = var10;
+         this.pause = var7;
       }
 
-      long var7 = Util.getNanos();
-      this.frameTimer.logFrameDuration(var7 - this.lastNanoTime);
-      this.lastNanoTime = var7;
+      long var8 = Util.getNanos();
+      this.frameTimer.logFrameDuration(var8 - this.lastNanoTime);
+      this.lastNanoTime = var8;
       this.profiler.push("fpsUpdate");
 
       while(Util.getMillis() >= this.lastTime + 1000L) {
          fps = this.frames;
          this.fpsString = String.format("%d fps T: %s%s%s%s B: %d", fps, (double)this.options.framerateLimit == Option.FRAMERATE_LIMIT.getMaxValue() ? "inf" : this.options.framerateLimit, this.options.enableVsync ? " vsync" : "", this.options.graphicsMode.toString(), this.options.renderClouds == CloudStatus.OFF ? "" : (this.options.renderClouds == CloudStatus.FAST ? " fast-clouds" : " fancy-clouds"), this.options.biomeBlendRadius);
@@ -1040,23 +1110,29 @@
 
    private boolean shouldRenderFpsPie() {
       return this.options.renderDebug && this.options.renderDebugCharts && !this.options.hideGui;
    }
 
-   private void startProfilers(boolean var1, @Nullable SingleTickProfiler var2) {
-      if (var1) {
+   private ProfilerFiller constructProfiler(boolean var1, @Nullable SingleTickProfiler var2) {
+      if (!var1 && !this.metricsRecorder.isRecording()) {
+         return (ProfilerFiller)(var2 == null ? InactiveProfiler.INSTANCE : var2.startTick());
+      } else if (var1) {
          if (!this.fpsPieProfiler.isEnabled()) {
             this.fpsPieRenderTicks = 0;
             this.fpsPieProfiler.enable();
          }
 
          ++this.fpsPieRenderTicks;
+         ProfilerFiller var3 = this.metricsRecorder.isRecording() ? ProfilerFiller.tee(this.fpsPieProfiler.getFiller(), this.metricsRecorder.getProfiler()) : this.fpsPieProfiler.getFiller();
+         return SingleTickProfiler.decorateFiller(var3, var2);
       } else {
-         this.fpsPieProfiler.disable();
-      }
+         if (this.fpsPieProfiler.isEnabled()) {
+            this.fpsPieProfiler.disable();
+         }
 
-      this.profiler = SingleTickProfiler.decorateFiller(this.fpsPieProfiler.getFiller(), var2);
+         return SingleTickProfiler.decorateFiller(this.metricsRecorder.getProfiler(), var2);
+      }
    }
 
    private void finishProfilers(boolean var1, @Nullable SingleTickProfiler var2) {
       if (var2 != null) {
          var2.endTick();
@@ -1110,11 +1186,145 @@
       }
 
       System.gc();
    }
 
-   void debugFpsMeterKeyPress(int var1) {
+   public boolean debugClientMetricsStart(Consumer<TranslatableComponent> var1) {
+      if (this.metricsRecorder.isRecording()) {
+         this.debugClientMetricsStop();
+         return false;
+      } else {
+         Consumer var2 = (var2x) -> {
+            int var3 = var2x.getTickDuration();
+            double var4 = (double)var2x.getNanoDuration() / (double)TimeUtil.NANOSECONDS_PER_SECOND;
+            this.execute(() -> {
+               var1.accept(new TranslatableComponent("commands.debug.stopped", new Object[]{String.format(Locale.ROOT, "%.2f", var4), var3, String.format(Locale.ROOT, "%.2f", (double)var3 / var4)}));
+            });
+         };
+         Consumer var3 = (var2x) -> {
+            MutableComponent var3 = (new TextComponent(var2x.toString())).withStyle(ChatFormatting.UNDERLINE).withStyle((var1x) -> {
+               return var1x.withClickEvent(new ClickEvent(ClickEvent.Action.OPEN_FILE, var2x.toFile().getParent()));
+            });
+            this.execute(() -> {
+               var1.accept(new TranslatableComponent("debug.profiling.stop", new Object[]{var3}));
+            });
+         };
+         SystemReport var4 = fillSystemReport(new SystemReport(), this, this.languageManager, this.launchedVersion, this.options);
+         Consumer var5 = (var3x) -> {
+            Path var4x = this.archiveProfilingReport(var4, var3x);
+            var3.accept(var4x);
+         };
+         Consumer var6;
+         if (this.singleplayerServer == null) {
+            var6 = (var1x) -> {
+               var5.accept(ImmutableList.of(var1x));
+            };
+         } else {
+            this.singleplayerServer.fillSystemReport(var4);
+            CompletableFuture var7 = new CompletableFuture();
+            CompletableFuture var8 = new CompletableFuture();
+            CompletableFuture.allOf(var7, var8).thenRunAsync(() -> {
+               var5.accept(ImmutableList.of((Path)var7.join(), (Path)var8.join()));
+            }, Util.ioPool());
+            IntegratedServer var10000 = this.singleplayerServer;
+            Consumer var10001 = (var0) -> {
+            };
+            Objects.requireNonNull(var8);
+            var10000.startRecordingMetrics(var10001, var8::complete);
+            Objects.requireNonNull(var7);
+            var6 = var7::complete;
+         }
+
+         this.metricsRecorder = ActiveMetricsRecorder.createStarted(new ClientMetricsSamplersProvider(Util.timeSource, this.levelRenderer), Util.timeSource, Util.ioPool(), new MetricsPersister("client"), (var2x) -> {
+            this.metricsRecorder = InactiveMetricsRecorder.INSTANCE;
+            var2.accept(var2x);
+         }, var6);
+         return true;
+      }
+   }
+
+   private void debugClientMetricsStop() {
+      this.metricsRecorder.end();
+      if (this.singleplayerServer != null) {
+         this.singleplayerServer.finishRecordingMetrics();
+      }
+
+   }
+
+   private Path archiveProfilingReport(SystemReport var1, List<Path> var2) {
+      String var4;
+      if (this.isLocalServer()) {
+         var4 = this.getSingleplayerServer().getWorldData().getLevelName();
+      } else {
+         var4 = this.getCurrentServer().name;
+      }
+
+      Path var3;
+      try {
+         String var5 = String.format("%s-%s-%s", (new SimpleDateFormat("yyyy-MM-dd_HH.mm.ss")).format(new Date()), var4, SharedConstants.getCurrentVersion().getId());
+         String var6 = FileUtil.findAvailableName(MetricsPersister.PROFILING_RESULTS_DIR, var5, ".zip");
+         var3 = MetricsPersister.PROFILING_RESULTS_DIR.resolve(var6);
+      } catch (IOException var23) {
+         throw new UncheckedIOException(var23);
+      }
+
+      boolean var18 = false;
+
+      try {
+         var18 = true;
+         FileZipper var25 = new FileZipper(var3);
+
+         try {
+            var25.add(Paths.get("system.txt"), var1.toLineSeparatedString());
+            var25.add(Paths.get("client").resolve(this.options.getFile().getName()), this.options.dumpOptionsForReport());
+            Objects.requireNonNull(var25);
+            var2.forEach(var25::add);
+         } catch (Throwable var22) {
+            try {
+               var25.close();
+            } catch (Throwable var19) {
+               var22.addSuppressed(var19);
+            }
+
+            throw var22;
+         }
+
+         var25.close();
+         var18 = false;
+      } finally {
+         if (var18) {
+            Iterator var9 = var2.iterator();
+
+            while(var9.hasNext()) {
+               Path var10 = (Path)var9.next();
+
+               try {
+                  FileUtils.forceDelete(var10.toFile());
+               } catch (IOException var20) {
+                  LOGGER.warn("Failed to delete temporary profiling result {}", var10, var20);
+               }
+            }
+
+         }
+      }
+
+      Iterator var26 = var2.iterator();
+
+      while(var26.hasNext()) {
+         Path var27 = (Path)var26.next();
+
+         try {
+            FileUtils.forceDelete(var27.toFile());
+         } catch (IOException var21) {
+            LOGGER.warn("Failed to delete temporary profiling result {}", var27, var21);
+         }
+      }
+
+      return var3;
+   }
+
+   public void debugFpsMeterKeyPress(int var1) {
       if (this.fpsPieResults != null) {
          List var2 = this.fpsPieResults.getTimes(this.debugPath);
          if (!var2.isEmpty()) {
             ResultField var3 = (ResultField)var2.remove(0);
             if (var1 == 0) {
@@ -1126,14 +1336,15 @@
                }
             } else {
                --var1;
                if (var1 < var2.size() && !"unspecified".equals(((ResultField)var2.get(var1)).name)) {
                   if (!this.debugPath.isEmpty()) {
-                     this.debugPath = this.debugPath + '\u001e';
+                     this.debugPath = this.debugPath + "\u001e";
                   }
 
-                  this.debugPath = this.debugPath + ((ResultField)var2.get(var1)).name;
+                  String var10001 = this.debugPath;
+                  this.debugPath = var10001 + ((ResultField)var2.get(var1)).name;
                }
             }
 
          }
       }
@@ -1141,107 +1352,111 @@
 
    private void renderFpsMeter(PoseStack var1, ProfileResults var2) {
       List var3 = var2.getTimes(this.debugPath);
       ResultField var4 = (ResultField)var3.remove(0);
       RenderSystem.clear(256, ON_OSX);
-      RenderSystem.matrixMode(5889);
-      RenderSystem.loadIdentity();
-      RenderSystem.ortho(0.0D, (double)this.window.getWidth(), (double)this.window.getHeight(), 0.0D, 1000.0D, 3000.0D);
-      RenderSystem.matrixMode(5888);
-      RenderSystem.loadIdentity();
-      RenderSystem.translatef(0.0F, 0.0F, -2000.0F);
+      RenderSystem.setShader(GameRenderer::getPositionColorShader);
+      Matrix4f var5 = Matrix4f.orthographic(0.0F, (float)this.window.getWidth(), 0.0F, (float)this.window.getHeight(), 1000.0F, 3000.0F);
+      RenderSystem.setProjectionMatrix(var5);
+      PoseStack var6 = RenderSystem.getModelViewStack();
+      var6.setIdentity();
+      var6.translate(0.0D, 0.0D, -2000.0D);
+      RenderSystem.applyModelViewMatrix();
       RenderSystem.lineWidth(1.0F);
       RenderSystem.disableTexture();
-      Tesselator var5 = Tesselator.getInstance();
-      BufferBuilder var6 = var5.getBuilder();
-      boolean var7 = true;
-      int var8 = this.window.getWidth() - 160 - 10;
-      int var9 = this.window.getHeight() - 320;
+      Tesselator var7 = Tesselator.getInstance();
+      BufferBuilder var8 = var7.getBuilder();
+      boolean var9 = true;
+      int var10 = this.window.getWidth() - 160 - 10;
+      int var11 = this.window.getHeight() - 320;
       RenderSystem.enableBlend();
-      var6.begin(7, DefaultVertexFormat.POSITION_COLOR);
-      var6.vertex((double)((float)var8 - 176.0F), (double)((float)var9 - 96.0F - 16.0F), 0.0D).color(200, 0, 0, 0).endVertex();
-      var6.vertex((double)((float)var8 - 176.0F), (double)(var9 + 320), 0.0D).color(200, 0, 0, 0).endVertex();
-      var6.vertex((double)((float)var8 + 176.0F), (double)(var9 + 320), 0.0D).color(200, 0, 0, 0).endVertex();
-      var6.vertex((double)((float)var8 + 176.0F), (double)((float)var9 - 96.0F - 16.0F), 0.0D).color(200, 0, 0, 0).endVertex();
-      var5.end();
+      var8.begin(VertexFormat.Mode.QUADS, DefaultVertexFormat.POSITION_COLOR);
+      var8.vertex((double)((float)var10 - 176.0F), (double)((float)var11 - 96.0F - 16.0F), 0.0D).color(200, 0, 0, 0).endVertex();
+      var8.vertex((double)((float)var10 - 176.0F), (double)(var11 + 320), 0.0D).color(200, 0, 0, 0).endVertex();
+      var8.vertex((double)((float)var10 + 176.0F), (double)(var11 + 320), 0.0D).color(200, 0, 0, 0).endVertex();
+      var8.vertex((double)((float)var10 + 176.0F), (double)((float)var11 - 96.0F - 16.0F), 0.0D).color(200, 0, 0, 0).endVertex();
+      var7.end();
       RenderSystem.disableBlend();
-      double var10 = 0.0D;
+      double var12 = 0.0D;
 
-      ResultField var13;
-      int var15;
-      for(Iterator var12 = var3.iterator(); var12.hasNext(); var10 += var13.percentage) {
-         var13 = (ResultField)var12.next();
-         int var14 = Mth.floor(var13.percentage / 4.0D) + 1;
-         var6.begin(6, DefaultVertexFormat.POSITION_COLOR);
-         var15 = var13.getColor();
-         int var16 = var15 >> 16 & 255;
-         int var17 = var15 >> 8 & 255;
-         int var18 = var15 & 255;
-         var6.vertex((double)var8, (double)var9, 0.0D).color(var16, var17, var18, 255).endVertex();
+      ResultField var15;
+      int var17;
+      for(Iterator var14 = var3.iterator(); var14.hasNext(); var12 += var15.percentage) {
+         var15 = (ResultField)var14.next();
+         int var16 = Mth.floor(var15.percentage / 4.0D) + 1;
+         var8.begin(VertexFormat.Mode.TRIANGLE_FAN, DefaultVertexFormat.POSITION_COLOR);
+         var17 = var15.getColor();
+         int var18 = var17 >> 16 & 255;
+         int var19 = var17 >> 8 & 255;
+         int var20 = var17 & 255;
+         var8.vertex((double)var10, (double)var11, 0.0D).color(var18, var19, var20, 255).endVertex();
 
-         int var19;
-         float var20;
-         float var21;
+         int var21;
          float var22;
-         for(var19 = var14; var19 >= 0; --var19) {
-            var20 = (float)((var10 + var13.percentage * (double)var19 / (double)var14) * 6.2831854820251465D / 100.0D);
-            var21 = Mth.sin(var20) * 160.0F;
-            var22 = Mth.cos(var20) * 160.0F * 0.5F;
-            var6.vertex((double)((float)var8 + var21), (double)((float)var9 - var22), 0.0D).color(var16, var17, var18, 255).endVertex();
+         float var23;
+         float var24;
+         for(var21 = var16; var21 >= 0; --var21) {
+            var22 = (float)((var12 + var15.percentage * (double)var21 / (double)var16) * 6.2831854820251465D / 100.0D);
+            var23 = Mth.sin(var22) * 160.0F;
+            var24 = Mth.cos(var22) * 160.0F * 0.5F;
+            var8.vertex((double)((float)var10 + var23), (double)((float)var11 - var24), 0.0D).color(var18, var19, var20, 255).endVertex();
          }
 
-         var5.end();
-         var6.begin(5, DefaultVertexFormat.POSITION_COLOR);
+         var7.end();
+         var8.begin(VertexFormat.Mode.TRIANGLE_STRIP, DefaultVertexFormat.POSITION_COLOR);
 
-         for(var19 = var14; var19 >= 0; --var19) {
-            var20 = (float)((var10 + var13.percentage * (double)var19 / (double)var14) * 6.2831854820251465D / 100.0D);
-            var21 = Mth.sin(var20) * 160.0F;
-            var22 = Mth.cos(var20) * 160.0F * 0.5F;
-            if (var22 <= 0.0F) {
-               var6.vertex((double)((float)var8 + var21), (double)((float)var9 - var22), 0.0D).color(var16 >> 1, var17 >> 1, var18 >> 1, 255).endVertex();
-               var6.vertex((double)((float)var8 + var21), (double)((float)var9 - var22 + 10.0F), 0.0D).color(var16 >> 1, var17 >> 1, var18 >> 1, 255).endVertex();
+         for(var21 = var16; var21 >= 0; --var21) {
+            var22 = (float)((var12 + var15.percentage * (double)var21 / (double)var16) * 6.2831854820251465D / 100.0D);
+            var23 = Mth.sin(var22) * 160.0F;
+            var24 = Mth.cos(var22) * 160.0F * 0.5F;
+            if (var24 <= 0.0F) {
+               var8.vertex((double)((float)var10 + var23), (double)((float)var11 - var24), 0.0D).color(var18 >> 1, var19 >> 1, var20 >> 1, 255).endVertex();
+               var8.vertex((double)((float)var10 + var23), (double)((float)var11 - var24 + 10.0F), 0.0D).color(var18 >> 1, var19 >> 1, var20 >> 1, 255).endVertex();
             }
          }
 
-         var5.end();
+         var7.end();
       }
 
-      DecimalFormat var23 = new DecimalFormat("##0.00");
-      var23.setDecimalFormatSymbols(DecimalFormatSymbols.getInstance(Locale.ROOT));
+      DecimalFormat var25 = new DecimalFormat("##0.00");
+      var25.setDecimalFormatSymbols(DecimalFormatSymbols.getInstance(Locale.ROOT));
       RenderSystem.enableTexture();
-      String var24 = ProfileResults.demanglePath(var4.name);
-      String var26 = "";
-      if (!"unspecified".equals(var24)) {
-         var26 = var26 + "[0] ";
+      String var26 = ProfileResults.demanglePath(var4.name);
+      String var28 = "";
+      if (!"unspecified".equals(var26)) {
+         var28 = var28 + "[0] ";
       }
 
-      if (var24.isEmpty()) {
-         var26 = var26 + "ROOT ";
+      if (var26.isEmpty()) {
+         var28 = var28 + "ROOT ";
       } else {
-         var26 = var26 + var24 + ' ';
+         var28 = var28 + var26 + " ";
       }
 
-      var15 = 16777215;
-      this.font.drawShadow(var1, var26, (float)(var8 - 160), (float)(var9 - 80 - 16), 16777215);
-      var26 = var23.format(var4.globalPercentage) + "%";
-      this.font.drawShadow(var1, var26, (float)(var8 + 160 - this.font.width(var26)), (float)(var9 - 80 - 16), 16777215);
+      var17 = 16777215;
+      this.font.drawShadow(var1, var28, (float)(var10 - 160), (float)(var11 - 80 - 16), 16777215);
+      String var10000 = var25.format(var4.globalPercentage);
+      var28 = var10000 + "%";
+      this.font.drawShadow(var1, var28, (float)(var10 + 160 - this.font.width(var28)), (float)(var11 - 80 - 16), 16777215);
 
-      for(int var25 = 0; var25 < var3.size(); ++var25) {
-         ResultField var29 = (ResultField)var3.get(var25);
-         StringBuilder var27 = new StringBuilder();
-         if ("unspecified".equals(var29.name)) {
-            var27.append("[?] ");
+      for(int var27 = 0; var27 < var3.size(); ++var27) {
+         ResultField var31 = (ResultField)var3.get(var27);
+         StringBuilder var29 = new StringBuilder();
+         if ("unspecified".equals(var31.name)) {
+            var29.append("[?] ");
          } else {
-            var27.append("[").append(var25 + 1).append("] ");
+            var29.append("[").append(var27 + 1).append("] ");
          }
 
-         String var28 = var27.append(var29.name).toString();
-         this.font.drawShadow(var1, var28, (float)(var8 - 160), (float)(var9 + 80 + var25 * 8 + 20), var29.getColor());
-         var28 = var23.format(var29.percentage) + "%";
-         this.font.drawShadow(var1, var28, (float)(var8 + 160 - 50 - this.font.width(var28)), (float)(var9 + 80 + var25 * 8 + 20), var29.getColor());
-         var28 = var23.format(var29.globalPercentage) + "%";
-         this.font.drawShadow(var1, var28, (float)(var8 + 160 - this.font.width(var28)), (float)(var9 + 80 + var25 * 8 + 20), var29.getColor());
+         String var30 = var29.append(var31.name).toString();
+         this.font.drawShadow(var1, var30, (float)(var10 - 160), (float)(var11 + 80 + var27 * 8 + 20), var31.getColor());
+         var10000 = var25.format(var31.percentage);
+         var30 = var10000 + "%";
+         this.font.drawShadow(var1, var30, (float)(var10 + 160 - 50 - this.font.width(var30)), (float)(var11 + 80 + var27 * 8 + 20), var31.getColor());
+         var10000 = var25.format(var31.globalPercentage);
+         var30 = var10000 + "%";
+         this.font.drawShadow(var1, var30, (float)(var10 + 160 - this.font.width(var30)), (float)(var11 + 80 + var27 * 8 + 20), var31.getColor());
       }
 
    }
 
    public void stop() {
@@ -1508,11 +1723,11 @@
             }
          }
 
          this.profiler.popPush("animateTick");
          if (!this.pause && this.level != null) {
-            this.level.animateTick(Mth.floor(this.player.getX()), Mth.floor(this.player.getY()), Mth.floor(this.player.getZ()));
+            this.level.animateTick(this.player.getBlockX(), this.player.getBlockY(), this.player.getBlockZ());
          }
 
          this.profiler.popPush("particles");
          if (!this.pause) {
             this.particleEngine.tick();
@@ -1549,21 +1764,21 @@
          boolean var3 = this.options.keyLoadHotbarActivator.isDown();
          if (this.options.keyHotbarSlots[var4].consumeClick()) {
             if (this.player.isSpectator()) {
                this.gui.getSpectatorGui().onHotbarSelected(var4);
             } else if (!this.player.isCreative() || this.screen != null || !var3 && !var2) {
-               this.player.inventory.selected = var4;
+               this.player.getInventory().selected = var4;
             } else {
                CreativeModeInventoryScreen.handleHotbarLoadOrSave(this, var4, var3, var2);
             }
          }
       }
 
       while(this.options.keySocialInteractions.consumeClick()) {
          if (!this.isMultiplayerServer()) {
             this.player.displayClientMessage(SOCIAL_INTERACTIONS_NOT_AVAILABLE, true);
-            NarratorChatListener.INSTANCE.sayNow(SOCIAL_INTERACTIONS_NOT_AVAILABLE.getString());
+            NarratorChatListener.INSTANCE.sayNow(SOCIAL_INTERACTIONS_NOT_AVAILABLE);
          } else {
             if (this.socialInteractionsToast != null) {
                this.tutorial.removeTimedToast(this.socialInteractionsToast);
                this.socialInteractionsToast = null;
             }
@@ -1595,37 +1810,34 @@
          if (!this.player.isSpectator() && this.player.drop(Screen.hasControlDown())) {
             this.player.swing(InteractionHand.MAIN_HAND);
          }
       }
 
-      boolean var5 = this.options.chatVisibility != ChatVisiblity.HIDDEN;
-      if (var5) {
-         while(this.options.keyChat.consumeClick()) {
-            this.openChatScreen("");
-         }
+      while(this.options.keyChat.consumeClick()) {
+         this.openChatScreen("");
+      }
 
-         if (this.screen == null && this.overlay == null && this.options.keyCommand.consumeClick()) {
-            this.openChatScreen("/");
-         }
+      if (this.screen == null && this.overlay == null && this.options.keyCommand.consumeClick()) {
+         this.openChatScreen("/");
       }
 
       if (this.player.isUsingItem()) {
          if (!this.options.keyUse.isDown()) {
             this.gameMode.releaseUsingItem(this.player);
          }
 
-         label120:
+         label115:
          while(true) {
             if (!this.options.keyAttack.consumeClick()) {
                while(this.options.keyUse.consumeClick()) {
                }
 
                while(true) {
                   if (this.options.keyPickItem.consumeClick()) {
                      continue;
                   }
-                  break label120;
+                  break label115;
                }
             }
          }
       } else {
          while(this.options.keyAttack.consumeClick()) {
@@ -1657,11 +1869,11 @@
          return var1;
       }
    }
 
    public static WorldData loadWorldData(LevelStorageSource.LevelStorageAccess var0, RegistryAccess.RegistryHolder var1, ResourceManager var2, DataPackConfig var3) {
-      RegistryReadOps var4 = RegistryReadOps.create(NbtOps.INSTANCE, (ResourceManager)var2, var1);
+      RegistryReadOps var4 = RegistryReadOps.createAndLoad(NbtOps.INSTANCE, (ResourceManager)var2, var1);
       WorldData var5 = var0.getDataTag(var4, var3);
       if (var5 == null) {
          throw new IllegalStateException("Failed to load world");
       } else {
          return var5;
@@ -1675,16 +1887,16 @@
    public void createLevel(String var1, LevelSettings var2, RegistryAccess.RegistryHolder var3, WorldGenSettings var4) {
       this.doLoadLevel(var1, var3, (var1x) -> {
          return var2.getDataPackConfig();
       }, (var3x, var4x, var5, var6) -> {
          RegistryWriteOps var7 = RegistryWriteOps.create(JsonOps.INSTANCE, var3);
-         RegistryReadOps var8 = RegistryReadOps.create(JsonOps.INSTANCE, (ResourceManager)var5, var3);
+         RegistryReadOps var8 = RegistryReadOps.createAndLoad(JsonOps.INSTANCE, (ResourceManager)var5, var3);
          DataResult var9 = WorldGenSettings.CODEC.encodeStart(var7, var4).setLifecycle(Lifecycle.stable()).flatMap((var1) -> {
             return WorldGenSettings.CODEC.parse(var8, var1);
          });
          Logger var10002 = LOGGER;
-         var10002.getClass();
+         Objects.requireNonNull(var10002);
          WorldGenSettings var10 = (WorldGenSettings)var9.resultOrPartial(Util.prefix("Error reading worldgen settings after loading data packs: ", var10002::error)).orElse(var4);
          return new PrimaryLevelData(var2, var10, var9.lifecycle());
       }, false, Minecraft.ExperimentalDialogType.CREATE);
    }
 
@@ -1729,21 +1941,21 @@
             var8.serverResources().updateGlobals();
             YggdrasilAuthenticationService var12 = new YggdrasilAuthenticationService(this.proxy);
             MinecraftSessionService var23 = var12.createMinecraftSessionService();
             GameProfileRepository var25 = var12.createProfileRepository();
             GameProfileCache var15 = new GameProfileCache(var25, new File(this.gameDirectory, MinecraftServer.USERID_CACHE_FILE.getName()));
+            var15.setExecutor(this);
             SkullBlockEntity.setProfileCache(var15);
             SkullBlockEntity.setSessionService(var23);
             GameProfileCache.setUsesAuthentication(false);
             this.singleplayerServer = (IntegratedServer)MinecraftServer.spin((var8x) -> {
                return new IntegratedServer(var8x, this, var2, var7, var8.packRepository(), var8.serverResources(), var9, var23, var25, var15, (var1) -> {
                   StoringChunkProgressListener var2 = new StoringChunkProgressListener(var1 + 0);
-                  var2.start();
                   this.progressListener.set(var2);
-                  Queue var10003 = this.progressTasks;
-                  var10003.getClass();
-                  return new ProcessorChunkProgressListener(var2, var10003::add);
+                  Queue var10001 = this.progressTasks;
+                  Objects.requireNonNull(var10001);
+                  return ProcessorChunkProgressListener.createStarted(var2, var10001::add);
                });
             });
             this.isLocalServer = true;
          } catch (Throwable var19) {
             CrashReport var13 = CrashReport.forThrowable(var19, "Starting integrated server");
@@ -1825,49 +2037,47 @@
             } else {
                this.setScreen((Screen)null);
 
                try {
                   LevelStorageSource.LevelStorageAccess var4x = this.levelSource.createAccess(var2);
-                  Throwable var5 = null;
 
                   try {
                      var4x.deleteLevel();
-                  } catch (Throwable var15) {
-                     var5 = var15;
-                     throw var15;
-                  } finally {
+                  } catch (Throwable var8) {
                      if (var4x != null) {
-                        if (var5 != null) {
-                           try {
-                              var4x.close();
-                           } catch (Throwable var14) {
-                              var5.addSuppressed(var14);
-                           }
-                        } else {
+                        try {
                            var4x.close();
+                        } catch (Throwable var7) {
+                           var8.addSuppressed(var7);
                         }
                      }
 
+                     throw var8;
                   }
-               } catch (IOException var17) {
+
+                  if (var4x != null) {
+                     var4x.close();
+                  }
+               } catch (IOException var9) {
                   SystemToast.onWorldDeleteFailure(this, var2);
-                  LOGGER.error("Failed to delete world {}", var2, var17);
+                  LOGGER.error("Failed to delete world {}", var2, var9);
                }
             }
 
          }, new TranslatableComponent("selectWorld.backupQuestion.experimental"), new TranslatableComponent("selectWorld.backupWarning.experimental"), CommonComponents.GUI_PROCEED, CommonComponents.GUI_CANCEL));
       }
 
    }
 
    public Minecraft.ServerStem makeServerStem(RegistryAccess.RegistryHolder var1, Function<LevelStorageSource.LevelStorageAccess, DataPackConfig> var2, Function4<LevelStorageSource.LevelStorageAccess, RegistryAccess.RegistryHolder, ResourceManager, DataPackConfig, WorldData> var3, boolean var4, LevelStorageSource.LevelStorageAccess var5) throws InterruptedException, ExecutionException {
       DataPackConfig var6 = (DataPackConfig)var2.apply(var5);
-      PackRepository var7 = new PackRepository(new RepositorySource[]{new ServerPacksSource(), new FolderRepositorySource(var5.getLevelPath(LevelResource.DATAPACK_DIR).toFile(), PackSource.WORLD)});
+      PackRepository var7 = new PackRepository(PackType.SERVER_DATA, new RepositorySource[]{new ServerPacksSource(), new FolderRepositorySource(var5.getLevelPath(LevelResource.DATAPACK_DIR).toFile(), PackSource.WORLD)});
 
       try {
          DataPackConfig var8 = MinecraftServer.configurePackRepository(var7, var6, var4);
-         CompletableFuture var9 = ServerResources.loadResources(var7.openAllSelected(), Commands.CommandSelection.INTEGRATED, 2, Util.backgroundExecutor(), this);
+         CompletableFuture var9 = ServerResources.loadResources(var7.openAllSelected(), var1, Commands.CommandSelection.INTEGRATED, 2, Util.backgroundExecutor(), this);
+         Objects.requireNonNull(var9);
          this.managedBlock(var9::isDone);
          ServerResources var10 = (ServerResources)var9.get();
          WorldData var11 = (WorldData)var3.apply(var5, var1, var10.getResourceManager(), var8);
          return new Minecraft.ServerStem(var7, var10, var11);
       } catch (ExecutionException | InterruptedException var12) {
@@ -1875,29 +2085,30 @@
          throw var12;
       }
    }
 
    public void setLevel(ClientLevel var1) {
-      ProgressScreen var2 = new ProgressScreen();
+      ProgressScreen var2 = new ProgressScreen(true);
       var2.progressStartNoAbort(new TranslatableComponent("connect.joining"));
       this.updateScreenAndTick(var2);
       this.level = var1;
       this.updateLevelInEngines(var1);
       if (!this.isLocalServer) {
          YggdrasilAuthenticationService var3 = new YggdrasilAuthenticationService(this.proxy);
          MinecraftSessionService var4 = var3.createMinecraftSessionService();
          GameProfileRepository var5 = var3.createProfileRepository();
          GameProfileCache var6 = new GameProfileCache(var5, new File(this.gameDirectory, MinecraftServer.USERID_CACHE_FILE.getName()));
+         var6.setExecutor(this);
          SkullBlockEntity.setProfileCache(var6);
          SkullBlockEntity.setSessionService(var4);
          GameProfileCache.setUsesAuthentication(false);
       }
 
    }
 
    public void clearLevel() {
-      this.clearLevel(new ProgressScreen());
+      this.clearLevel(new ProgressScreen(true));
    }
 
    public void clearLevel(Screen var1) {
       ClientPacketListener var2 = this.getConnection();
       if (var2 != null) {
@@ -1952,28 +2163,38 @@
    }
 
    private void updateLevelInEngines(@Nullable ClientLevel var1) {
       this.levelRenderer.setLevel(var1);
       this.particleEngine.setLevel(var1);
-      BlockEntityRenderDispatcher.instance.setLevel(var1);
+      this.blockEntityRenderDispatcher.setLevel(var1);
       this.updateTitle();
    }
 
    public boolean allowsMultiplayer() {
       return this.allowsMultiplayer && this.socialInteractionsService.serversAllowed();
    }
 
+   public boolean allowsRealms() {
+      return this.socialInteractionsService.realmsAllowed();
+   }
+
    public boolean isBlocked(UUID var1) {
-      if (this.allowsChat()) {
+      if (this.getChatStatus().isChatAllowed(false)) {
          return this.playerSocialManager.shouldHideMessageFrom(var1);
       } else {
          return (this.player == null || !var1.equals(this.player.getUUID())) && !var1.equals(Util.NIL_UUID);
       }
    }
 
-   public boolean allowsChat() {
-      return this.allowsChat && this.socialInteractionsService.chatAllowed();
+   public Minecraft.ChatStatus getChatStatus() {
+      if (this.options.chatVisibility == ChatVisiblity.HIDDEN) {
+         return Minecraft.ChatStatus.DISABLED_BY_OPTIONS;
+      } else if (!this.allowsChat) {
+         return Minecraft.ChatStatus.DISABLED_BY_LAUNCHER;
+      } else {
+         return !this.socialInteractionsService.chatAllowed() ? Minecraft.ChatStatus.DISABLED_BY_PROFILE : Minecraft.ChatStatus.ENABLED;
+      }
    }
 
    public final boolean isDemo() {
       return this.demo;
    }
@@ -1990,94 +2211,48 @@
    public static boolean useFancyGraphics() {
       return instance.options.graphicsMode.getId() >= GraphicsStatus.FANCY.getId();
    }
 
    public static boolean useShaderTransparency() {
-      return instance.options.graphicsMode.getId() >= GraphicsStatus.FABULOUS.getId();
+      return !instance.gameRenderer.isPanoramicMode() && instance.options.graphicsMode.getId() >= GraphicsStatus.FABULOUS.getId();
    }
 
    public static boolean useAmbientOcclusion() {
       return instance.options.ambientOcclusion != AmbientOcclusionStatus.OFF;
    }
 
    private void pickBlock() {
       if (this.hitResult != null && this.hitResult.getType() != HitResult.Type.MISS) {
-         boolean var1 = this.player.abilities.instabuild;
+         boolean var1 = this.player.getAbilities().instabuild;
          BlockEntity var2 = null;
          HitResult.Type var4 = this.hitResult.getType();
          ItemStack var3;
          if (var4 == HitResult.Type.BLOCK) {
             BlockPos var8 = ((BlockHitResult)this.hitResult).getBlockPos();
-            BlockState var13 = this.level.getBlockState(var8);
-            Block var16 = var13.getBlock();
-            if (var13.isAir()) {
+            BlockState var6 = this.level.getBlockState(var8);
+            if (var6.isAir()) {
                return;
             }
 
-            var3 = var16.getCloneItemStack(this.level, var8, var13);
+            Block var7 = var6.getBlock();
+            var3 = var7.getCloneItemStack(this.level, var8, var6);
             if (var3.isEmpty()) {
                return;
             }
 
-            if (var1 && Screen.hasControlDown() && var16.isEntityBlock()) {
+            if (var1 && Screen.hasControlDown() && var6.hasBlockEntity()) {
                var2 = this.level.getBlockEntity(var8);
             }
          } else {
             if (var4 != HitResult.Type.ENTITY || !var1) {
                return;
             }
 
             Entity var5 = ((EntityHitResult)this.hitResult).getEntity();
-            if (var5 instanceof Painting) {
-               var3 = new ItemStack(Items.PAINTING);
-            } else if (var5 instanceof LeashFenceKnotEntity) {
-               var3 = new ItemStack(Items.LEAD);
-            } else if (var5 instanceof ItemFrame) {
-               ItemFrame var6 = (ItemFrame)var5;
-               ItemStack var7 = var6.getItem();
-               if (var7.isEmpty()) {
-                  var3 = new ItemStack(Items.ITEM_FRAME);
-               } else {
-                  var3 = var7.copy();
-               }
-            } else if (var5 instanceof AbstractMinecart) {
-               AbstractMinecart var11 = (AbstractMinecart)var5;
-               Item var15;
-               switch(var11.getMinecartType()) {
-               case FURNACE:
-                  var15 = Items.FURNACE_MINECART;
-                  break;
-               case CHEST:
-                  var15 = Items.CHEST_MINECART;
-                  break;
-               case TNT:
-                  var15 = Items.TNT_MINECART;
-                  break;
-               case HOPPER:
-                  var15 = Items.HOPPER_MINECART;
-                  break;
-               case COMMAND_BLOCK:
-                  var15 = Items.COMMAND_BLOCK_MINECART;
-                  break;
-               default:
-                  var15 = Items.MINECART;
-               }
-
-               var3 = new ItemStack(var15);
-            } else if (var5 instanceof Boat) {
-               var3 = new ItemStack(((Boat)var5).getDropItem());
-            } else if (var5 instanceof ArmorStand) {
-               var3 = new ItemStack(Items.ARMOR_STAND);
-            } else if (var5 instanceof EndCrystal) {
-               var3 = new ItemStack(Items.END_CRYSTAL);
-            } else {
-               SpawnEggItem var12 = SpawnEggItem.byId(var5.getType());
-               if (var12 == null) {
-                  return;
-               }
-
-               var3 = new ItemStack(var12);
+            var3 = var5.getPickResult();
+            if (var3 == null) {
+               return;
             }
          }
 
          if (var3.isEmpty()) {
             String var10 = "";
@@ -2087,24 +2262,24 @@
                var10 = Registry.ENTITY_TYPE.getKey(((EntityHitResult)this.hitResult).getEntity().getType()).toString();
             }
 
             LOGGER.warn("Picking on: [{}] {} gave null item", var4, var10);
          } else {
-            Inventory var9 = this.player.inventory;
+            Inventory var9 = this.player.getInventory();
             if (var2 != null) {
                this.addCustomNbtData(var3, var2);
             }
 
-            int var14 = var9.findSlotMatchingItem(var3);
+            int var11 = var9.findSlotMatchingItem(var3);
             if (var1) {
                var9.setPickedItem(var3);
                this.gameMode.handleCreativeModeItemAdd(this.player.getItemInHand(InteractionHand.MAIN_HAND), 36 + var9.selected);
-            } else if (var14 != -1) {
-               if (Inventory.isHotbarSlot(var14)) {
-                  var9.selected = var14;
+            } else if (var11 != -1) {
+               if (Inventory.isHotbarSlot(var11)) {
+                  var9.selected = var11;
                } else {
-                  this.gameMode.handlePickItem(var14);
+                  this.gameMode.handlePickItem(var11);
                }
             }
 
          }
       }
@@ -2127,74 +2302,91 @@
          return var1;
       }
    }
 
    public CrashReport fillReport(CrashReport var1) {
-      fillReport(this.languageManager, this.launchedVersion, this.options, var1);
+      SystemReport var2 = var1.getSystemReport();
+      fillSystemReport(var2, this, this.languageManager, this.launchedVersion, this.options);
       if (this.level != null) {
          this.level.fillReportDetails(var1);
       }
 
+      if (this.singleplayerServer != null) {
+         this.singleplayerServer.fillSystemReport(var2);
+      }
+
+      this.reloadStateTracker.fillCrashReport(var1);
       return var1;
    }
 
-   public static void fillReport(@Nullable LanguageManager var0, String var1, @Nullable Options var2, CrashReport var3) {
-      CrashReportCategory var4 = var3.getSystemDetails();
-      var4.setDetail("Launched Version", () -> {
-         return var1;
+   public static void fillReport(@Nullable Minecraft var0, @Nullable LanguageManager var1, String var2, @Nullable Options var3, CrashReport var4) {
+      SystemReport var5 = var4.getSystemReport();
+      fillSystemReport(var5, var0, var1, var2, var3);
+   }
+
+   private static SystemReport fillSystemReport(SystemReport var0, @Nullable Minecraft var1, @Nullable LanguageManager var2, String var3, Options var4) {
+      var0.setDetail("Launched Version", () -> {
+         return var3;
       });
-      var4.setDetail("Backend library", RenderSystem::getBackendDescription);
-      var4.setDetail("Backend API", RenderSystem::getApiDescription);
-      var4.setDetail("GL Caps", RenderSystem::getCapsString);
-      var4.setDetail("Using VBOs", () -> {
+      var0.setDetail("Backend library", RenderSystem::getBackendDescription);
+      var0.setDetail("Backend API", RenderSystem::getApiDescription);
+      var0.setDetail("Window size", () -> {
+         return var1 != null ? var1.window.getWidth() + "x" + var1.window.getHeight() : "<not initialized>";
+      });
+      var0.setDetail("GL Caps", RenderSystem::getCapsString);
+      var0.setDetail("GL debug messages", () -> {
+         return GlDebug.isDebugEnabled() ? String.join("\n", GlDebug.getLastOpenGlDebugMessages()) : "<disabled>";
+      });
+      var0.setDetail("Using VBOs", () -> {
          return "Yes";
       });
-      var4.setDetail("Is Modded", () -> {
+      var0.setDetail("Is Modded", () -> {
          String var0 = ClientBrandRetriever.getClientModName();
          if (!"vanilla".equals(var0)) {
             return "Definitely; Client brand changed to '" + var0 + "'";
          } else {
             return Minecraft.class.getSigners() == null ? "Very likely; Jar signature invalidated" : "Probably not. Jar signature remains and client brand is untouched.";
          }
       });
-      var4.setDetail("Type", (Object)"Client (map_client.txt)");
-      if (var2 != null) {
+      var0.setDetail("Type", "Client (map_client.txt)");
+      if (var4 != null) {
          if (instance != null) {
             String var5 = instance.getGpuWarnlistManager().getAllWarnings();
             if (var5 != null) {
-               var4.setDetail("GPU Warnings", (Object)var5);
+               var0.setDetail("GPU Warnings", var5);
             }
          }
 
-         var4.setDetail("Graphics mode", (Object)var2.graphicsMode);
-         var4.setDetail("Resource Packs", () -> {
+         var0.setDetail("Graphics mode", var4.graphicsMode.toString());
+         var0.setDetail("Resource Packs", () -> {
             StringBuilder var1 = new StringBuilder();
-            Iterator var2x = var2.resourcePacks.iterator();
+            Iterator var2 = var4.resourcePacks.iterator();
 
-            while(var2x.hasNext()) {
-               String var3 = (String)var2x.next();
+            while(var2.hasNext()) {
+               String var3 = (String)var2.next();
                if (var1.length() > 0) {
                   var1.append(", ");
                }
 
                var1.append(var3);
-               if (var2.incompatibleResourcePacks.contains(var3)) {
+               if (var4.incompatibleResourcePacks.contains(var3)) {
                   var1.append(" (incompatible)");
                }
             }
 
             return var1.toString();
          });
       }
 
-      if (var0 != null) {
-         var4.setDetail("Current Language", () -> {
-            return var0.getSelected().toString();
+      if (var2 != null) {
+         var0.setDetail("Current Language", () -> {
+            return var2.getSelected().toString();
          });
       }
 
-      var4.setDetail("CPU", GlUtil::getCpuInfo);
+      var0.setDetail("CPU", GlUtil::getCpuInfo);
+      return var0;
    }
 
    public static Minecraft getInstance() {
       return instance;
    }
@@ -2221,11 +2413,12 @@
       Iterator var4 = this.resourcePackRepository.getSelectedPacks().iterator();
 
       while(var4.hasNext()) {
          Pack var5 = (Pack)var4.next();
          if (!var5.isRequired() && !var5.isFixedPosition()) {
-            var1.setDynamicData("resource_pack[" + var3++ + "]", var5.getId());
+            int var10001 = var3++;
+            var1.setDynamicData("resource_pack[" + var10001 + "]", var5.getId());
          }
       }
 
       var1.setDynamicData("resource_packs", var3);
       if (this.singleplayerServer != null) {
@@ -2242,10 +2435,31 @@
       } else {
          return "out_of_game";
       }
    }
 
+   public void populateSnooperInitial(Snooper var1) {
+      var1.setFixedData("client_brand", ClientBrandRetriever.getClientModName());
+      var1.setFixedData("launched_version", this.launchedVersion);
+      populateSnooperWithOpenGL(var1);
+      var1.setFixedData("gl_max_texture_size", RenderSystem.maxSupportedTextureSize());
+      GameProfile var2 = this.user.getGameProfile();
+      if (var2.getId() != null) {
+         var1.setFixedData("uuid", Hashing.sha1().hashBytes(var2.getId().toString().getBytes(Charsets.ISO_8859_1)).toString());
+      }
+
+   }
+
+   private static void populateSnooperWithOpenGL(Snooper var0) {
+      Objects.requireNonNull(var0);
+      GlUtil.populateSnooperWithOpenGL(var0::setFixedData);
+   }
+
+   public boolean isSnooperEnabled() {
+      return this.options.snooperEnabled;
+   }
+
    public void setCurrentServer(@Nullable ServerData var1) {
       this.currentServer = var1;
    }
 
    @Nullable
@@ -2311,10 +2525,11 @@
       return this.languageManager;
    }
 
    public Function<ResourceLocation, TextureAtlasSprite> getTextureAtlas(ResourceLocation var1) {
       TextureAtlas var10000 = this.modelManager.getAtlas(var1);
+      Objects.requireNonNull(var10000);
       return var10000::getSprite;
    }
 
    public boolean is64Bit() {
       return this.is64bit;
@@ -2341,11 +2556,11 @@
          } else {
             Biome.BiomeCategory var1 = this.player.level.getBiome(this.player.blockPosition()).getBiomeCategory();
             if (this.musicManager.isPlayingMusic(Musics.UNDER_WATER) || this.player.isUnderWater() && (var1 == Biome.BiomeCategory.OCEAN || var1 == Biome.BiomeCategory.RIVER)) {
                return Musics.UNDER_WATER;
             } else {
-               return this.player.level.dimension() != Level.NETHER && this.player.abilities.instabuild && this.player.abilities.mayfly ? Musics.CREATIVE : (Music)this.level.getBiomeManager().getNoiseBiomeAtPosition(this.player.blockPosition()).getBackgroundMusic().orElse(Musics.GAME);
+               return this.player.level.dimension() != Level.NETHER && this.player.getAbilities().instabuild && this.player.getAbilities().mayfly ? Musics.CREATIVE : (Music)this.level.getBiomeManager().getNoiseBiomeAtPosition(this.player.blockPosition()).getBackgroundMusic().orElse(Musics.GAME);
             }
          }
       } else {
          return Musics.MENU;
       }
@@ -2368,11 +2583,11 @@
       this.cameraEntity = var1;
       this.gameRenderer.checkEntityPostEffect(var1);
    }
 
    public boolean shouldEntityAppearGlowing(Entity var1) {
-      return var1.isGlowing() || this.player != null && this.player.isSpectator() && this.options.keySpectatorOutlines.isDown() && var1.getType() == EntityType.PLAYER;
+      return var1.isCurrentlyGlowing() || this.player != null && this.player.isSpectator() && this.options.keySpectatorOutlines.isDown() && var1.getType() == EntityType.PLAYER;
    }
 
    protected Thread getRunningThread() {
       return this.gameThread;
    }
@@ -2391,10 +2606,14 @@
 
    public EntityRenderDispatcher getEntityRenderDispatcher() {
       return this.entityRenderDispatcher;
    }
 
+   public BlockEntityRenderDispatcher getBlockEntityRenderDispatcher() {
+      return this.blockEntityRenderDispatcher;
+   }
+
    public ItemRenderer getItemRenderer() {
       return this.itemRenderer;
    }
 
    public ItemInHandRenderer getItemInHandRenderer() {
@@ -2467,18 +2686,145 @@
 
    public void setWindowActive(boolean var1) {
       this.windowActive = var1;
    }
 
+   public Component grabPanoramixScreenshot(File var1, int var2, int var3) {
+      int var4 = this.window.getWidth();
+      int var5 = this.window.getHeight();
+      TextureTarget var6 = new TextureTarget(var2, var3, true, ON_OSX);
+      float var7 = this.player.getXRot();
+      float var8 = this.player.getYRot();
+      float var9 = this.player.xRotO;
+      float var10 = this.player.yRotO;
+      this.gameRenderer.setRenderBlockOutline(false);
+
+      TranslatableComponent var12;
+      try {
+         this.gameRenderer.setPanoramicMode(true);
+         this.levelRenderer.graphicsChanged();
+         this.window.setWidth(var2);
+         this.window.setHeight(var3);
+
+         for(int var11 = 0; var11 < 6; ++var11) {
+            switch(var11) {
+            case 0:
+               this.player.setYRot(var8);
+               this.player.setXRot(0.0F);
+               break;
+            case 1:
+               this.player.setYRot((var8 + 90.0F) % 360.0F);
+               this.player.setXRot(0.0F);
+               break;
+            case 2:
+               this.player.setYRot((var8 + 180.0F) % 360.0F);
+               this.player.setXRot(0.0F);
+               break;
+            case 3:
+               this.player.setYRot((var8 - 90.0F) % 360.0F);
+               this.player.setXRot(0.0F);
+               break;
+            case 4:
+               this.player.setYRot(var8);
+               this.player.setXRot(-90.0F);
+               break;
+            case 5:
+            default:
+               this.player.setYRot(var8);
+               this.player.setXRot(90.0F);
+            }
+
+            this.player.yRotO = this.player.getYRot();
+            this.player.xRotO = this.player.getXRot();
+            var6.bindWrite(true);
+            this.gameRenderer.renderLevel(1.0F, 0L, new PoseStack());
+
+            try {
+               Thread.sleep(10L);
+            } catch (InterruptedException var17) {
+            }
+
+            Screenshot.grab(var1, "panorama_" + var11 + ".png", var2, var3, var6, (var0) -> {
+            });
+         }
+
+         MutableComponent var20 = (new TextComponent(var1.getName())).withStyle(ChatFormatting.UNDERLINE).withStyle((var1x) -> {
+            return var1x.withClickEvent(new ClickEvent(ClickEvent.Action.OPEN_FILE, var1.getAbsolutePath()));
+         });
+         var12 = new TranslatableComponent("screenshot.success", new Object[]{var20});
+         return var12;
+      } catch (Exception var18) {
+         LOGGER.error("Couldn't save image", var18);
+         var12 = new TranslatableComponent("screenshot.failure", new Object[]{var18.getMessage()});
+      } finally {
+         this.player.setXRot(var7);
+         this.player.setYRot(var8);
+         this.player.xRotO = var9;
+         this.player.yRotO = var10;
+         this.gameRenderer.setRenderBlockOutline(true);
+         this.window.setWidth(var4);
+         this.window.setHeight(var5);
+         var6.destroyBuffers();
+         this.gameRenderer.setPanoramicMode(false);
+         this.levelRenderer.graphicsChanged();
+         this.getMainRenderTarget().bindWrite(true);
+      }
+
+      return var12;
+   }
+
+   private Component grabHugeScreenshot(File var1, int var2, int var3, int var4, int var5) {
+      try {
+         ByteBuffer var6 = GlUtil.allocateMemory(var2 * var3 * 3);
+         Screenshot var7 = new Screenshot(var1, var4, var5, var3);
+         float var8 = (float)var4 / (float)var2;
+         float var9 = (float)var5 / (float)var3;
+         float var10 = var8 > var9 ? var8 : var9;
+
+         for(int var11 = (var5 - 1) / var3 * var3; var11 >= 0; var11 -= var3) {
+            for(int var12 = 0; var12 < var4; var12 += var2) {
+               RenderSystem.setShaderTexture(0, TextureAtlas.LOCATION_BLOCKS);
+               float var13 = (float)(var4 - var2) / 2.0F * 2.0F - (float)(var12 * 2);
+               float var14 = (float)(var5 - var3) / 2.0F * 2.0F - (float)(var11 * 2);
+               var13 /= (float)var2;
+               var14 /= (float)var3;
+               this.gameRenderer.renderZoomed(var10, var13, var14);
+               var6.clear();
+               RenderSystem.pixelStore(3333, 1);
+               RenderSystem.pixelStore(3317, 1);
+               RenderSystem.readPixels(0, 0, var2, var3, 32992, 5121, var6);
+               var7.addRegion(var6, var12, var11, var2, var3);
+            }
+
+            var7.saveRow();
+         }
+
+         File var16 = var7.close();
+         GlUtil.freeMemory(var6);
+         MutableComponent var17 = (new TextComponent(var16.getName())).withStyle(ChatFormatting.UNDERLINE).withStyle((var1x) -> {
+            return var1x.withClickEvent(new ClickEvent(ClickEvent.Action.OPEN_FILE, var16.getAbsolutePath()));
+         });
+         return new TranslatableComponent("screenshot.success", new Object[]{var17});
+      } catch (Exception var15) {
+         LOGGER.warn("Couldn't save screenshot", var15);
+         return new TranslatableComponent("screenshot.failure", new Object[]{var15.getMessage()});
+      }
+   }
+
    public ProfilerFiller getProfiler() {
       return this.profiler;
    }
 
    public Game getGame() {
       return this.game;
    }
 
+   @Nullable
+   public StoringChunkProgressListener getProgressListener() {
+      return (StoringChunkProgressListener)this.progressListener.get();
+   }
+
    public SplashManager getSplashManager() {
       return this.splashManager;
    }
 
    @Nullable
@@ -2500,22 +2846,22 @@
 
    public RenderBuffers renderBuffers() {
       return this.renderBuffers;
    }
 
-   private static Pack createClientPackAdapter(String var0, boolean var1, Supplier<PackResources> var2, PackResources var3, PackMetadataSection var4, Pack.Position var5, PackSource var6) {
+   private static Pack createClientPackAdapter(String var0, Component var1, boolean var2, Supplier<PackResources> var3, PackMetadataSection var4, Pack.Position var5, PackSource var6) {
       int var7 = var4.getPackFormat();
-      Supplier var8 = var2;
+      Supplier var8 = var3;
       if (var7 <= 3) {
-         var8 = adaptV3(var2);
+         var8 = adaptV3(var3);
       }
 
       if (var7 <= 4) {
          var8 = adaptV4(var8);
       }
 
-      return new Pack(var0, var1, var8, var3, var4, var5, var6);
+      return new Pack(var0, var1, var2, var8, var4, PackType.CLIENT_RESOURCES, var5, var6);
    }
 
    private static Supplier<PackResources> adaptV3(Supplier<PackResources> var0) {
       return () -> {
          return new LegacyPackResourcesAdapter((PackResources)var0.get(), LegacyPackResourcesAdapter.V3);
@@ -2530,26 +2876,78 @@
 
    public void updateMaxMipLevel(int var1) {
       this.modelManager.updateMaxMipLevel(var1);
    }
 
+   public EntityModelSet getEntityModels() {
+      return this.entityModels;
+   }
+
+   public boolean isTextFilteringEnabled() {
+      return false;
+   }
+
    static {
       ON_OSX = Util.getPlatform() == Util.OS.OSX;
       DEFAULT_FONT = new ResourceLocation("default");
       UNIFORM_FONT = new ResourceLocation("uniform");
       ALT_FONT = new ResourceLocation("alt");
       RESOURCE_RELOAD_INITIAL_TASK = CompletableFuture.completedFuture(Unit.INSTANCE);
       SOCIAL_INTERACTIONS_NOT_AVAILABLE = new TranslatableComponent("multiplayer.socialInteractions.not_available");
       reserve = new byte[10485760];
    }
 
+   public static enum ChatStatus {
+      ENABLED(TextComponent.EMPTY) {
+         public boolean isChatAllowed(boolean var1) {
+            return true;
+         }
+      },
+      DISABLED_BY_OPTIONS((new TranslatableComponent("chat.disabled.options")).withStyle(ChatFormatting.RED)) {
+         public boolean isChatAllowed(boolean var1) {
+            return false;
+         }
+      },
+      DISABLED_BY_LAUNCHER((new TranslatableComponent("chat.disabled.launcher")).withStyle(ChatFormatting.RED)) {
+         public boolean isChatAllowed(boolean var1) {
+            return var1;
+         }
+      },
+      DISABLED_BY_PROFILE((new TranslatableComponent("chat.disabled.profile")).withStyle(ChatFormatting.RED)) {
+         public boolean isChatAllowed(boolean var1) {
+            return var1;
+         }
+      };
+
+      private final Component message;
+
+      ChatStatus(Component var3) {
+         this.message = var3;
+      }
+
+      public Component getMessage() {
+         return this.message;
+      }
+
+      public abstract boolean isChatAllowed(boolean var1);
+   }
+
+   static enum ExperimentalDialogType {
+      NONE,
+      CREATE,
+      BACKUP;
+
+      private ExperimentalDialogType() {
+      }
+   }
+
    public static final class ServerStem implements AutoCloseable {
       private final PackRepository packRepository;
       private final ServerResources serverResources;
       private final WorldData worldData;
 
-      private ServerStem(PackRepository var1, ServerResources var2, WorldData var3) {
+      ServerStem(PackRepository var1, ServerResources var2, WorldData var3) {
          this.packRepository = var1;
          this.serverResources = var2;
          this.worldData = var3;
       }
 
@@ -2568,15 +2966,6 @@
       public void close() {
          this.packRepository.close();
          this.serverResources.close();
       }
    }
-
-   static enum ExperimentalDialogType {
-      NONE,
-      CREATE,
-      BACKUP;
-
-      private ExperimentalDialogType() {
-      }
-   }
 }
