--- net/minecraft/client/User.java
+++ net/minecraft/client/User.java
@@ -46,10 +46,14 @@
       } catch (IllegalArgumentException var2) {
          return new GameProfile((UUID)null, this.getName());
       }
    }
 
+   public User.Type getType() {
+      return this.type;
+   }
+
    public static enum Type {
       LEGACY("legacy"),
       MOJANG("mojang");
 
       private static final Map<String, User.Type> BY_NAME = (Map)Arrays.stream(values()).collect(Collectors.toMap((var0) -> {
