--- net/minecraft/server/dedicated/DedicatedServerProperties.java
+++ net/minecraft/server/dedicated/DedicatedServerProperties.java
@@ -16,18 +16,19 @@
    public final boolean spawnAnimals = this.get("spawn-animals", true);
    public final boolean spawnNpcs = this.get("spawn-npcs", true);
    public final boolean pvp = this.get("pvp", true);
    public final boolean allowFlight = this.get("allow-flight", false);
    public final String resourcePack = this.get("resource-pack", "");
+   public final boolean requireResourcePack = this.get("require-resource-pack", false);
+   public final String resourcePackPrompt = this.get("resource-pack-prompt", "");
    public final String motd = this.get("motd", "A Minecraft Server");
    public final boolean forceGameMode = this.get("force-gamemode", false);
    public final boolean enforceWhitelist = this.get("enforce-whitelist", false);
    public final Difficulty difficulty;
    public final GameType gamemode;
    public final String levelName;
    public final int serverPort;
-   public final int maxBuildHeight;
    public final Boolean announcePlayerAchievements;
    public final boolean enableQuery;
    public final int queryPort;
    public final boolean enableRcon;
    public final int rconPort;
@@ -64,13 +65,10 @@
       super(var1);
       this.difficulty = (Difficulty)this.get("difficulty", dispatchNumberOrString(Difficulty::byId, Difficulty::byName), Difficulty::getKey, Difficulty.EASY);
       this.gamemode = (GameType)this.get("gamemode", dispatchNumberOrString(GameType::byId, GameType::byName), GameType::getName, GameType.SURVIVAL);
       this.levelName = this.get("level-name", "world");
       this.serverPort = this.get("server-port", 25565);
-      this.maxBuildHeight = this.get("max-build-height", (var0) -> {
-         return Mth.clamp((var0 + 8) / 16 * 16, 64, 256);
-      }, 256);
       this.announcePlayerAchievements = this.getLegacyBoolean("announce-player-achievements");
       this.enableQuery = this.get("enable-query", false);
       this.queryPort = this.get("query.port", 25565);
       this.enableRcon = this.get("enable-rcon", false);
       this.rconPort = this.get("rcon.port", 25575);
@@ -95,17 +93,17 @@
       this.maxPlayers = this.get("max-players", 20);
       this.networkCompressionThreshold = this.get("network-compression-threshold", 256);
       this.broadcastRconToOps = this.get("broadcast-rcon-to-ops", true);
       this.broadcastConsoleToOps = this.get("broadcast-console-to-ops", true);
       this.maxWorldSize = this.get("max-world-size", (var0) -> {
-         return Mth.clamp(var0, 1, 29999984);
+         return Mth.clamp((int)var0, (int)1, (int)29999984);
       }, 29999984);
       this.syncChunkWrites = this.get("sync-chunk-writes", true);
       this.enableJmxMonitoring = this.get("enable-jmx-monitoring", false);
       this.enableStatus = this.get("enable-status", true);
       this.entityBroadcastRangePercentage = this.get("entity-broadcast-range-percentage", (var0) -> {
-         return Mth.clamp(var0, 10, 1000);
+         return Mth.clamp((int)var0, (int)10, (int)1000);
       }, 100);
       this.textFilteringConfig = this.get("text-filtering-config", "");
       this.playerIdleTimeout = this.getMutable("player-idle-timeout", 0);
       this.whiteList = this.getMutable("white-list", false);
       this.worldGenSettings = WorldGenSettings.create(var2, var1);
