--- net/minecraft/server/dedicated/DedicatedServerProperties.java
+++ net/minecraft/server/dedicated/DedicatedServerProperties.java
@@ -16,18 +16,18 @@
    public final boolean spawnAnimals = this.get("spawn-animals", true);
    public final boolean spawnNpcs = this.get("spawn-npcs", true);
    public final boolean pvp = this.get("pvp", true);
    public final boolean allowFlight = this.get("allow-flight", false);
    public final String resourcePack = this.get("resource-pack", "");
+   public final boolean requireResourcePack = this.get("require-resource-pack", false);
    public final String motd = this.get("motd", "A Minecraft Server");
    public final boolean forceGameMode = this.get("force-gamemode", false);
    public final boolean enforceWhitelist = this.get("enforce-whitelist", false);
    public final Difficulty difficulty;
    public final GameType gamemode;
    public final String levelName;
    public final int serverPort;
-   public final int maxBuildHeight;
    public final Boolean announcePlayerAchievements;
    public final boolean enableQuery;
    public final int queryPort;
    public final boolean enableRcon;
    public final int rconPort;
@@ -64,13 +64,10 @@
       super(var1);
       this.difficulty = (Difficulty)this.get("difficulty", dispatchNumberOrString(Difficulty::byId, Difficulty::byName), Difficulty::getKey, Difficulty.EASY);
       this.gamemode = (GameType)this.get("gamemode", dispatchNumberOrString(GameType::byId, GameType::byName), GameType::getName, GameType.SURVIVAL);
       this.levelName = this.get("level-name", "world");
       this.serverPort = this.get("server-port", 25565);
-      this.maxBuildHeight = this.get("max-build-height", (var0) -> {
-         return Mth.clamp((var0 + 8) / 16 * 16, 64, 256);
-      }, 256);
       this.announcePlayerAchievements = this.getLegacyBoolean("announce-player-achievements");
       this.enableQuery = this.get("enable-query", false);
       this.queryPort = this.get("query.port", 25565);
       this.enableRcon = this.get("enable-rcon", false);
       this.rconPort = this.get("rcon.port", 25575);
