--- net/minecraft/client/gui/font/providers/LegacyUnicodeBitmapsProvider.java
+++ net/minecraft/client/gui/font/providers/LegacyUnicodeBitmapsProvider.java
@@ -19,10 +19,13 @@
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
 
 public class LegacyUnicodeBitmapsProvider implements GlyphProvider {
    private static final Logger LOGGER = LogManager.getLogger();
+   private static final int UNICODE_SHEETS = 256;
+   private static final int CHARS_PER_SHEET = 256;
+   private static final int TEXTURE_SIZE = 256;
    private final ResourceManager resourceManager;
    private final byte[] sizes;
    private final String texturePattern;
    private final Map<ResourceLocation, NativeImage> textures = Maps.newHashMap();
 
