--- com/mojang/blaze3d/vertex/DefaultedVertexConsumer.java
+++ com/mojang/blaze3d/vertex/DefaultedVertexConsumer.java
@@ -1,9 +1,9 @@
 package com.mojang.blaze3d.vertex;
 
 public abstract class DefaultedVertexConsumer implements VertexConsumer {
-   protected boolean defaultColorSet = false;
+   protected boolean defaultColorSet;
    protected int defaultR = 255;
    protected int defaultG = 255;
    protected int defaultB = 255;
    protected int defaultA = 255;
 
@@ -15,6 +15,10 @@
       this.defaultG = var2;
       this.defaultB = var3;
       this.defaultA = var4;
       this.defaultColorSet = true;
    }
+
+   public void unsetDefaultColor() {
+      this.defaultColorSet = false;
+   }
 }
