--- com/mojang/blaze3d/vertex/PoseStack.java
+++ com/mojang/blaze3d/vertex/PoseStack.java
@@ -20,11 +20,11 @@
    public PoseStack() {
    }
 
    public void translate(double var1, double var3, double var5) {
       PoseStack.Pose var7 = (PoseStack.Pose)this.poseStack.getLast();
-      var7.pose.multiply(Matrix4f.createTranslateMatrix((float)var1, (float)var3, (float)var5));
+      var7.pose.multiplyWithTranslation((float)var1, (float)var3, (float)var5);
    }
 
    public void scale(float var1, float var2, float var3) {
       PoseStack.Pose var4 = (PoseStack.Pose)this.poseStack.getLast();
       var4.pose.multiply(Matrix4f.createScaleMatrix(var1, var2, var3));
