--- net/minecraft/world/level/block/entity/BlastFurnaceBlockEntity.java
+++ net/minecraft/world/level/block/entity/BlastFurnaceBlockEntity.java
@@ -1,18 +1,20 @@
 package net.minecraft.world.level.block.entity;
 
+import net.minecraft.core.BlockPos;
 import net.minecraft.network.chat.Component;
 import net.minecraft.network.chat.TranslatableComponent;
 import net.minecraft.world.entity.player.Inventory;
 import net.minecraft.world.inventory.AbstractContainerMenu;
 import net.minecraft.world.inventory.BlastFurnaceMenu;
 import net.minecraft.world.item.ItemStack;
 import net.minecraft.world.item.crafting.RecipeType;
+import net.minecraft.world.level.block.state.BlockState;
 
 public class BlastFurnaceBlockEntity extends AbstractFurnaceBlockEntity {
-   public BlastFurnaceBlockEntity() {
-      super(BlockEntityType.BLAST_FURNACE, RecipeType.BLASTING);
+   public BlastFurnaceBlockEntity(BlockPos var1, BlockState var2) {
+      super(BlockEntityType.BLAST_FURNACE, var1, var2, RecipeType.BLASTING);
    }
 
    protected Component getDefaultName() {
       return new TranslatableComponent("container.blast_furnace");
    }
