--- net/minecraft/world/level/block/entity/JukeboxBlockEntity.java
+++ net/minecraft/world/level/block/entity/JukeboxBlockEntity.java
@@ -1,24 +1,25 @@
 package net.minecraft.world.level.block.entity;
 
+import net.minecraft.core.BlockPos;
 import net.minecraft.nbt.CompoundTag;
 import net.minecraft.world.Clearable;
 import net.minecraft.world.item.ItemStack;
 import net.minecraft.world.level.block.state.BlockState;
 
 public class JukeboxBlockEntity extends BlockEntity implements Clearable {
    private ItemStack record;
 
-   public JukeboxBlockEntity() {
-      super(BlockEntityType.JUKEBOX);
+   public JukeboxBlockEntity(BlockPos var1, BlockState var2) {
+      super(BlockEntityType.JUKEBOX, var1, var2);
       this.record = ItemStack.EMPTY;
    }
 
-   public void load(BlockState var1, CompoundTag var2) {
-      super.load(var1, var2);
-      if (var2.contains("RecordItem", 10)) {
-         this.setRecord(ItemStack.of(var2.getCompound("RecordItem")));
+   public void load(CompoundTag var1) {
+      super.load(var1);
+      if (var1.contains("RecordItem", 10)) {
+         this.setRecord(ItemStack.of(var1.getCompound("RecordItem")));
       }
 
    }
 
    public CompoundTag save(CompoundTag var1) {
