--- net/minecraft/world/level/block/state/predicate/BlockMaterialPredicate.java
+++ net/minecraft/world/level/block/state/predicate/BlockMaterialPredicate.java
@@ -7,11 +7,11 @@
 
 public class BlockMaterialPredicate implements Predicate<BlockState> {
    private static final BlockMaterialPredicate AIR;
    private final Material material;
 
-   private BlockMaterialPredicate(Material var1) {
+   BlockMaterialPredicate(Material var1) {
       this.material = var1;
    }
 
    public static BlockMaterialPredicate forMaterial(Material var0) {
       return var0 == Material.AIR ? AIR : new BlockMaterialPredicate(var0);
