--- net/minecraft/data/worldgen/BastionHousingUnitsPools.java
+++ net/minecraft/data/worldgen/BastionHousingUnitsPools.java
@@ -5,10 +5,13 @@
 import net.minecraft.resources.ResourceLocation;
 import net.minecraft.world.level.levelgen.feature.structures.StructurePoolElement;
 import net.minecraft.world.level.levelgen.feature.structures.StructureTemplatePool;
 
 public class BastionHousingUnitsPools {
+   public BastionHousingUnitsPools() {
+   }
+
    public static void bootstrap() {
    }
 
    static {
       Pools.register(new StructureTemplatePool(new ResourceLocation("bastion/units/center_pieces"), new ResourceLocation("empty"), ImmutableList.of(Pair.of(StructurePoolElement.single("bastion/units/center_pieces/center_0", ProcessorLists.HOUSING), 1), Pair.of(StructurePoolElement.single("bastion/units/center_pieces/center_1", ProcessorLists.HOUSING), 1), Pair.of(StructurePoolElement.single("bastion/units/center_pieces/center_2", ProcessorLists.HOUSING), 1)), StructureTemplatePool.Projection.RIGID));
