--- net/minecraft/data/worldgen/BiomeDefaultFeatures.java
+++ net/minecraft/data/worldgen/BiomeDefaultFeatures.java
@@ -5,10 +5,13 @@
 import net.minecraft.world.level.biome.BiomeGenerationSettings;
 import net.minecraft.world.level.biome.MobSpawnSettings;
 import net.minecraft.world.level.levelgen.GenerationStep;
 
 public class BiomeDefaultFeatures {
+   public BiomeDefaultFeatures() {
+   }
+
    public static void addDefaultOverworldLandMesaStructures(BiomeGenerationSettings.Builder var0) {
       var0.addStructureStart(StructureFeatures.MINESHAFT_MESA);
       var0.addStructureStart(StructureFeatures.STRONGHOLD);
    }
 
@@ -46,24 +49,43 @@
    public static void addDefaultMonsterRoom(BiomeGenerationSettings.Builder var0) {
       var0.addFeature(GenerationStep.Decoration.UNDERGROUND_STRUCTURES, Features.MONSTER_ROOM);
    }
 
    public static void addDefaultUndergroundVariety(BiomeGenerationSettings.Builder var0) {
+      addDefaultUndergroundVariety(var0, false);
+   }
+
+   public static void addDefaultUndergroundVariety(BiomeGenerationSettings.Builder var0, boolean var1) {
       var0.addFeature(GenerationStep.Decoration.UNDERGROUND_ORES, Features.ORE_DIRT);
       var0.addFeature(GenerationStep.Decoration.UNDERGROUND_ORES, Features.ORE_GRAVEL);
       var0.addFeature(GenerationStep.Decoration.UNDERGROUND_ORES, Features.ORE_GRANITE);
       var0.addFeature(GenerationStep.Decoration.UNDERGROUND_ORES, Features.ORE_DIORITE);
       var0.addFeature(GenerationStep.Decoration.UNDERGROUND_ORES, Features.ORE_ANDESITE);
+      if (!var1) {
+         var0.addFeature(GenerationStep.Decoration.VEGETAL_DECORATION, Features.GLOW_LICHEN);
+      }
+
+      var0.addFeature(GenerationStep.Decoration.UNDERGROUND_ORES, Features.ORE_TUFF);
+      var0.addFeature(GenerationStep.Decoration.UNDERGROUND_ORES, Features.ORE_DEEPSLATE);
+      var0.addFeature(GenerationStep.Decoration.UNDERGROUND_DECORATION, Features.RARE_DRIPSTONE_CLUSTER_FEATURE);
+      var0.addFeature(GenerationStep.Decoration.UNDERGROUND_DECORATION, Features.RARE_SMALL_DRIPSTONE_FEATURE);
    }
 
+   public static void addDripstone(BiomeGenerationSettings.Builder var0) {
+      var0.addFeature(GenerationStep.Decoration.LOCAL_MODIFICATIONS, Features.LARGE_DRIPSTONE_FEATURE);
+      var0.addFeature(GenerationStep.Decoration.UNDERGROUND_DECORATION, Features.DRIPSTONE_CLUSTER_FEATURE);
+      var0.addFeature(GenerationStep.Decoration.UNDERGROUND_DECORATION, Features.SMALL_DRIPSTONE_FEATURE);
+   }
+
    public static void addDefaultOres(BiomeGenerationSettings.Builder var0) {
       var0.addFeature(GenerationStep.Decoration.UNDERGROUND_ORES, Features.ORE_COAL);
       var0.addFeature(GenerationStep.Decoration.UNDERGROUND_ORES, Features.ORE_IRON);
       var0.addFeature(GenerationStep.Decoration.UNDERGROUND_ORES, Features.ORE_GOLD);
       var0.addFeature(GenerationStep.Decoration.UNDERGROUND_ORES, Features.ORE_REDSTONE);
       var0.addFeature(GenerationStep.Decoration.UNDERGROUND_ORES, Features.ORE_DIAMOND);
       var0.addFeature(GenerationStep.Decoration.UNDERGROUND_ORES, Features.ORE_LAPIS);
+      var0.addFeature(GenerationStep.Decoration.UNDERGROUND_ORES, Features.ORE_COPPER);
    }
 
    public static void addExtraGold(BiomeGenerationSettings.Builder var0) {
       var0.addFeature(GenerationStep.Decoration.UNDERGROUND_ORES, Features.ORE_GOLD_EXTRA);
    }
@@ -137,10 +159,24 @@
 
    public static void addShatteredSavannaTrees(BiomeGenerationSettings.Builder var0) {
       var0.addFeature(GenerationStep.Decoration.VEGETAL_DECORATION, Features.TREES_SHATTERED_SAVANNA);
    }
 
+   public static void addLushCavesVegetationFeatures(BiomeGenerationSettings.Builder var0) {
+      var0.addFeature(GenerationStep.Decoration.VEGETAL_DECORATION, Features.LUSH_CAVES_CEILING_VEGETATION);
+      var0.addFeature(GenerationStep.Decoration.VEGETAL_DECORATION, Features.CAVE_VINES);
+      var0.addFeature(GenerationStep.Decoration.VEGETAL_DECORATION, Features.LUSH_CAVES_CLAY);
+      var0.addFeature(GenerationStep.Decoration.VEGETAL_DECORATION, Features.LUSH_CAVES_VEGETATION);
+      var0.addFeature(GenerationStep.Decoration.VEGETAL_DECORATION, Features.ROOTED_AZALEA_TREES);
+      var0.addFeature(GenerationStep.Decoration.VEGETAL_DECORATION, Features.SPORE_BLOSSOM_FEATURE);
+      var0.addFeature(GenerationStep.Decoration.VEGETAL_DECORATION, Features.CLASSIC_VINES_CAVE_FEATURE);
+   }
+
+   public static void addLushCavesSpecialOres(BiomeGenerationSettings.Builder var0) {
+      var0.addFeature(GenerationStep.Decoration.UNDERGROUND_ORES, Features.ORE_CLAY);
+   }
+
    public static void addMountainTrees(BiomeGenerationSettings.Builder var0) {
       var0.addFeature(GenerationStep.Decoration.VEGETAL_DECORATION, Features.TREES_MOUNTAIN);
    }
 
    public static void addMountainEdgeTrees(BiomeGenerationSettings.Builder var0) {
@@ -154,15 +190,15 @@
    public static void addJungleEdgeTrees(BiomeGenerationSettings.Builder var0) {
       var0.addFeature(GenerationStep.Decoration.VEGETAL_DECORATION, Features.TREES_JUNGLE_EDGE);
    }
 
    public static void addBadlandsTrees(BiomeGenerationSettings.Builder var0) {
-      var0.addFeature(GenerationStep.Decoration.VEGETAL_DECORATION, Features.OAK_BADLANDS);
+      var0.addFeature(GenerationStep.Decoration.VEGETAL_DECORATION, Features.TREES_BADLANDS);
    }
 
    public static void addSnowyTrees(BiomeGenerationSettings.Builder var0) {
-      var0.addFeature(GenerationStep.Decoration.VEGETAL_DECORATION, Features.SPRUCE_SNOWY);
+      var0.addFeature(GenerationStep.Decoration.VEGETAL_DECORATION, Features.TREES_SNOWY);
    }
 
    public static void addJungleGrass(BiomeGenerationSettings.Builder var0) {
       var0.addFeature(GenerationStep.Decoration.VEGETAL_DECORATION, Features.PATCH_GRASS_JUNGLE);
    }
@@ -191,11 +227,11 @@
    public static void addForestGrass(BiomeGenerationSettings.Builder var0) {
       var0.addFeature(GenerationStep.Decoration.VEGETAL_DECORATION, Features.PATCH_GRASS_FOREST);
    }
 
    public static void addSwampVegetation(BiomeGenerationSettings.Builder var0) {
-      var0.addFeature(GenerationStep.Decoration.VEGETAL_DECORATION, Features.SWAMP_TREE);
+      var0.addFeature(GenerationStep.Decoration.VEGETAL_DECORATION, Features.TREES_SWAMP);
       var0.addFeature(GenerationStep.Decoration.VEGETAL_DECORATION, Features.FLOWER_SWAMP);
       var0.addFeature(GenerationStep.Decoration.VEGETAL_DECORATION, Features.PATCH_GRASS_NORMAL);
       var0.addFeature(GenerationStep.Decoration.VEGETAL_DECORATION, Features.PATCH_DEAD_BUSH);
       var0.addFeature(GenerationStep.Decoration.VEGETAL_DECORATION, Features.PATCH_WATERLILLY);
       var0.addFeature(GenerationStep.Decoration.VEGETAL_DECORATION, Features.BROWN_MUSHROOM_SWAMP);
@@ -328,26 +364,36 @@
    public static void addAncientDebris(BiomeGenerationSettings.Builder var0) {
       var0.addFeature(GenerationStep.Decoration.UNDERGROUND_DECORATION, Features.ORE_DEBRIS_LARGE);
       var0.addFeature(GenerationStep.Decoration.UNDERGROUND_DECORATION, Features.ORE_DEBRIS_SMALL);
    }
 
+   public static void addDefaultCrystalFormations(BiomeGenerationSettings.Builder var0) {
+      var0.addFeature(GenerationStep.Decoration.LOCAL_MODIFICATIONS, Features.AMETHYST_GEODE);
+   }
+
    public static void farmAnimals(MobSpawnSettings.Builder var0) {
       var0.addSpawn(MobCategory.CREATURE, new MobSpawnSettings.SpawnerData(EntityType.SHEEP, 12, 4, 4));
       var0.addSpawn(MobCategory.CREATURE, new MobSpawnSettings.SpawnerData(EntityType.PIG, 10, 4, 4));
       var0.addSpawn(MobCategory.CREATURE, new MobSpawnSettings.SpawnerData(EntityType.CHICKEN, 10, 4, 4));
       var0.addSpawn(MobCategory.CREATURE, new MobSpawnSettings.SpawnerData(EntityType.COW, 8, 4, 4));
    }
 
-   public static void ambientSpawns(MobSpawnSettings.Builder var0) {
+   public static void caveSpawns(MobSpawnSettings.Builder var0) {
       var0.addSpawn(MobCategory.AMBIENT, new MobSpawnSettings.SpawnerData(EntityType.BAT, 10, 8, 8));
+      caveWaterSpawns(var0);
    }
 
    public static void commonSpawns(MobSpawnSettings.Builder var0) {
-      ambientSpawns(var0);
+      caveSpawns(var0);
       monsters(var0, 95, 5, 100);
    }
 
+   public static void caveWaterSpawns(MobSpawnSettings.Builder var0) {
+      var0.addSpawn(MobCategory.UNDERGROUND_WATER_CREATURE, new MobSpawnSettings.SpawnerData(EntityType.GLOW_SQUID, 10, 4, 6));
+      var0.addSpawn(MobCategory.UNDERGROUND_WATER_CREATURE, new MobSpawnSettings.SpawnerData(EntityType.AXOLOTL, 10, 4, 6));
+   }
+
    public static void oceanSpawns(MobSpawnSettings.Builder var0, int var1, int var2, int var3) {
       var0.addSpawn(MobCategory.WATER_CREATURE, new MobSpawnSettings.SpawnerData(EntityType.SQUID, var1, 1, var2));
       var0.addSpawn(MobCategory.WATER_AMBIENT, new MobSpawnSettings.SpawnerData(EntityType.COD, var3, 3, 6));
       commonSpawns(var0);
       var0.addSpawn(MobCategory.MONSTER, new MobSpawnSettings.SpawnerData(EntityType.DROWNED, 5, 1, 1));
@@ -368,18 +414,18 @@
    }
 
    public static void snowySpawns(MobSpawnSettings.Builder var0) {
       var0.addSpawn(MobCategory.CREATURE, new MobSpawnSettings.SpawnerData(EntityType.RABBIT, 10, 2, 3));
       var0.addSpawn(MobCategory.CREATURE, new MobSpawnSettings.SpawnerData(EntityType.POLAR_BEAR, 1, 1, 2));
-      ambientSpawns(var0);
+      caveSpawns(var0);
       monsters(var0, 95, 5, 20);
       var0.addSpawn(MobCategory.MONSTER, new MobSpawnSettings.SpawnerData(EntityType.STRAY, 80, 4, 4));
    }
 
    public static void desertSpawns(MobSpawnSettings.Builder var0) {
       var0.addSpawn(MobCategory.CREATURE, new MobSpawnSettings.SpawnerData(EntityType.RABBIT, 4, 2, 3));
-      ambientSpawns(var0);
+      caveSpawns(var0);
       monsters(var0, 19, 1, 100);
       var0.addSpawn(MobCategory.MONSTER, new MobSpawnSettings.SpawnerData(EntityType.HUSK, 80, 4, 4));
    }
 
    public static void monsters(MobSpawnSettings.Builder var0, int var1, int var2, int var3) {
@@ -393,11 +439,11 @@
       var0.addSpawn(MobCategory.MONSTER, new MobSpawnSettings.SpawnerData(EntityType.WITCH, 5, 1, 1));
    }
 
    public static void mooshroomSpawns(MobSpawnSettings.Builder var0) {
       var0.addSpawn(MobCategory.CREATURE, new MobSpawnSettings.SpawnerData(EntityType.MOOSHROOM, 8, 4, 8));
-      ambientSpawns(var0);
+      caveSpawns(var0);
    }
 
    public static void baseJungleSpawns(MobSpawnSettings.Builder var0) {
       farmAnimals(var0);
       var0.addSpawn(MobCategory.CREATURE, new MobSpawnSettings.SpawnerData(EntityType.CHICKEN, 10, 4, 4));
