--- net/minecraft/data/worldgen/DesertVillagePools.java
+++ net/minecraft/data/worldgen/DesertVillagePools.java
@@ -7,10 +7,13 @@
 import net.minecraft.world.level.levelgen.feature.structures.StructureTemplatePool;
 
 public class DesertVillagePools {
    public static final StructureTemplatePool START;
 
+   public DesertVillagePools() {
+   }
+
    public static void bootstrap() {
    }
 
    static {
       START = Pools.register(new StructureTemplatePool(new ResourceLocation("village/desert/town_centers"), new ResourceLocation("empty"), ImmutableList.of(Pair.of(StructurePoolElement.legacy("village/desert/town_centers/desert_meeting_point_1"), 98), Pair.of(StructurePoolElement.legacy("village/desert/town_centers/desert_meeting_point_2"), 98), Pair.of(StructurePoolElement.legacy("village/desert/town_centers/desert_meeting_point_3"), 49), Pair.of(StructurePoolElement.legacy("village/desert/zombie/town_centers/desert_meeting_point_1", ProcessorLists.ZOMBIE_DESERT), 2), Pair.of(StructurePoolElement.legacy("village/desert/zombie/town_centers/desert_meeting_point_2", ProcessorLists.ZOMBIE_DESERT), 2), Pair.of(StructurePoolElement.legacy("village/desert/zombie/town_centers/desert_meeting_point_3", ProcessorLists.ZOMBIE_DESERT), 1)), StructureTemplatePool.Projection.RIGID));
