--- net/minecraft/nbt/DoubleTag.java
+++ net/minecraft/nbt/DoubleTag.java
@@ -1,13 +1,10 @@
 package net.minecraft.nbt;
 
 import java.io.DataInput;
 import java.io.DataOutput;
 import java.io.IOException;
-import net.minecraft.network.chat.Component;
-import net.minecraft.network.chat.MutableComponent;
-import net.minecraft.network.chat.TextComponent;
 import net.minecraft.util.Mth;
 
 public class DoubleTag extends NumericTag {
    public static final DoubleTag ZERO = new DoubleTag(0.0D);
    public static final TagType<DoubleTag> TYPE = new TagType<DoubleTag>() {
@@ -48,14 +45,10 @@
 
    public TagType<DoubleTag> getType() {
       return TYPE;
    }
 
-   public String toString() {
-      return this.data + "d";
-   }
-
    public DoubleTag copy() {
       return this;
    }
 
    public boolean equals(Object var1) {
@@ -69,13 +62,12 @@
    public int hashCode() {
       long var1 = Double.doubleToLongBits(this.data);
       return (int)(var1 ^ var1 >>> 32);
    }
 
-   public Component getPrettyDisplay(String var1, int var2) {
-      MutableComponent var3 = (new TextComponent("d")).withStyle(SYNTAX_HIGHLIGHTING_NUMBER_TYPE);
-      return (new TextComponent(String.valueOf(this.data))).append(var3).withStyle(SYNTAX_HIGHLIGHTING_NUMBER);
+   public void accept(TagVisitor var1) {
+      var1.visitDouble(this);
    }
 
    public long getAsLong() {
       return (long)Math.floor(this.data);
    }
