--- net/minecraft/nbt/EndTag.java
+++ net/minecraft/nbt/EndTag.java
@@ -1,12 +1,10 @@
 package net.minecraft.nbt;
 
 import java.io.DataInput;
 import java.io.DataOutput;
 import java.io.IOException;
-import net.minecraft.network.chat.Component;
-import net.minecraft.network.chat.TextComponent;
 
 public class EndTag implements Tag {
    public static final TagType<EndTag> TYPE = new TagType<EndTag>() {
       public EndTag load(DataInput var1, int var2, NbtAccounter var3) {
          var3.accountBits(64L);
@@ -40,16 +38,16 @@
    public TagType<EndTag> getType() {
       return TYPE;
    }
 
    public String toString() {
-      return "END";
+      return this.getAsString();
    }
 
    public EndTag copy() {
       return this;
    }
 
-   public Component getPrettyDisplay(String var1, int var2) {
-      return TextComponent.EMPTY;
+   public void accept(TagVisitor var1) {
+      var1.visitEnd(this);
    }
 }
