--- net/minecraft/nbt/FloatTag.java
+++ net/minecraft/nbt/FloatTag.java
@@ -1,16 +1,14 @@
 package net.minecraft.nbt;
 
 import java.io.DataInput;
 import java.io.DataOutput;
 import java.io.IOException;
-import net.minecraft.network.chat.Component;
-import net.minecraft.network.chat.MutableComponent;
-import net.minecraft.network.chat.TextComponent;
 import net.minecraft.util.Mth;
 
 public class FloatTag extends NumericTag {
+   private static final int SELF_SIZE_IN_BITS = 96;
    public static final FloatTag ZERO = new FloatTag(0.0F);
    public static final TagType<FloatTag> TYPE = new TagType<FloatTag>() {
       public FloatTag load(DataInput var1, int var2, NbtAccounter var3) throws IOException {
          var3.accountBits(96L);
          return FloatTag.valueOf(var1.readFloat());
@@ -48,14 +46,10 @@
 
    public TagType<FloatTag> getType() {
       return TYPE;
    }
 
-   public String toString() {
-      return this.data + "f";
-   }
-
    public FloatTag copy() {
       return this;
    }
 
    public boolean equals(Object var1) {
@@ -68,13 +62,12 @@
 
    public int hashCode() {
       return Float.floatToIntBits(this.data);
    }
 
-   public Component getPrettyDisplay(String var1, int var2) {
-      MutableComponent var3 = (new TextComponent("f")).withStyle(SYNTAX_HIGHLIGHTING_NUMBER_TYPE);
-      return (new TextComponent(String.valueOf(this.data))).append(var3).withStyle(SYNTAX_HIGHLIGHTING_NUMBER);
+   public void accept(TagVisitor var1) {
+      var1.visitFloat(this);
    }
 
    public long getAsLong() {
       return (long)this.data;
    }
