--- net/minecraft/nbt/IntArrayTag.java
+++ net/minecraft/nbt/IntArrayTag.java
@@ -3,16 +3,14 @@
 import java.io.DataInput;
 import java.io.DataOutput;
 import java.io.IOException;
 import java.util.Arrays;
 import java.util.List;
-import net.minecraft.network.chat.Component;
-import net.minecraft.network.chat.MutableComponent;
-import net.minecraft.network.chat.TextComponent;
 import org.apache.commons.lang3.ArrayUtils;
 
 public class IntArrayTag extends CollectionTag<IntTag> {
+   private static final int SELF_SIZE_IN_BITS = 192;
    public static final TagType<IntArrayTag> TYPE = new TagType<IntArrayTag>() {
       public IntArrayTag load(DataInput var1, int var2, NbtAccounter var3) throws IOException {
          var3.accountBits(192L);
          int var4 = var1.readInt();
          var3.accountBits(32L * (long)var4);
@@ -73,21 +71,11 @@
    public TagType<IntArrayTag> getType() {
       return TYPE;
    }
 
    public String toString() {
-      StringBuilder var1 = new StringBuilder("[I;");
-
-      for(int var2 = 0; var2 < this.data.length; ++var2) {
-         if (var2 != 0) {
-            var1.append(',');
-         }
-
-         var1.append(this.data[var2]);
-      }
-
-      return var1.append(']').toString();
+      return this.getAsString();
    }
 
    public IntArrayTag copy() {
       int[] var1 = new int[this.data.length];
       System.arraycopy(this.data, 0, var1, 0, this.data.length);
@@ -108,23 +96,12 @@
 
    public int[] getAsIntArray() {
       return this.data;
    }
 
-   public Component getPrettyDisplay(String var1, int var2) {
-      MutableComponent var3 = (new TextComponent("I")).withStyle(SYNTAX_HIGHLIGHTING_NUMBER_TYPE);
-      MutableComponent var4 = (new TextComponent("[")).append(var3).append(";");
-
-      for(int var5 = 0; var5 < this.data.length; ++var5) {
-         var4.append(" ").append((Component)(new TextComponent(String.valueOf(this.data[var5]))).withStyle(SYNTAX_HIGHLIGHTING_NUMBER));
-         if (var5 != this.data.length - 1) {
-            var4.append(",");
-         }
-      }
-
-      var4.append("]");
-      return var4;
+   public void accept(TagVisitor var1) {
+      var1.visitIntArray(this);
    }
 
    public int size() {
       return this.data.length;
    }
