--- net/minecraft/nbt/LongArrayTag.java
+++ net/minecraft/nbt/LongArrayTag.java
@@ -4,13 +4,10 @@
 import java.io.DataInput;
 import java.io.DataOutput;
 import java.io.IOException;
 import java.util.Arrays;
 import java.util.List;
-import net.minecraft.network.chat.Component;
-import net.minecraft.network.chat.MutableComponent;
-import net.minecraft.network.chat.TextComponent;
 import org.apache.commons.lang3.ArrayUtils;
 
 public class LongArrayTag extends CollectionTag<LongTag> {
    public static final TagType<LongArrayTag> TYPE = new TagType<LongArrayTag>() {
       public LongArrayTag load(DataInput var1, int var2, NbtAccounter var3) throws IOException {
@@ -78,21 +75,11 @@
    public TagType<LongArrayTag> getType() {
       return TYPE;
    }
 
    public String toString() {
-      StringBuilder var1 = new StringBuilder("[L;");
-
-      for(int var2 = 0; var2 < this.data.length; ++var2) {
-         if (var2 != 0) {
-            var1.append(',');
-         }
-
-         var1.append(this.data[var2]).append('L');
-      }
-
-      return var1.append(']').toString();
+      return this.getAsString();
    }
 
    public LongArrayTag copy() {
       long[] var1 = new long[this.data.length];
       System.arraycopy(this.data, 0, var1, 0, this.data.length);
@@ -109,24 +96,12 @@
 
    public int hashCode() {
       return Arrays.hashCode(this.data);
    }
 
-   public Component getPrettyDisplay(String var1, int var2) {
-      MutableComponent var3 = (new TextComponent("L")).withStyle(SYNTAX_HIGHLIGHTING_NUMBER_TYPE);
-      MutableComponent var4 = (new TextComponent("[")).append(var3).append(";");
-
-      for(int var5 = 0; var5 < this.data.length; ++var5) {
-         MutableComponent var6 = (new TextComponent(String.valueOf(this.data[var5]))).withStyle(SYNTAX_HIGHLIGHTING_NUMBER);
-         var4.append(" ").append((Component)var6).append((Component)var3);
-         if (var5 != this.data.length - 1) {
-            var4.append(",");
-         }
-      }
-
-      var4.append("]");
-      return var4;
+   public void accept(TagVisitor var1) {
+      var1.visitLongArray(this);
    }
 
    public long[] getAsLongArray() {
       return this.data;
    }
