--- net/minecraft/nbt/NbtIo.java
+++ net/minecraft/nbt/NbtIo.java
@@ -18,10 +18,13 @@
 import net.minecraft.CrashReport;
 import net.minecraft.CrashReportCategory;
 import net.minecraft.ReportedException;
 
 public class NbtIo {
+   public NbtIo() {
+   }
+
    public static CompoundTag readCompressed(File var0) throws IOException {
       FileInputStream var1 = new FileInputStream(var0);
       Throwable var2 = null;
 
       CompoundTag var3;
