--- net/minecraft/nbt/NbtOps.java
+++ net/minecraft/nbt/NbtOps.java
@@ -282,11 +282,11 @@
    }
 
    public DataResult<Consumer<Consumer<Tag>>> getList(Tag var1) {
       if (var1 instanceof CollectionTag) {
          CollectionTag var2 = (CollectionTag)var1;
-         var2.getClass();
+         Objects.requireNonNull(var2);
          return DataResult.success(var2::forEach);
       } else {
          return DataResult.error("Not a list: " + var1);
       }
    }
@@ -373,11 +373,11 @@
 
    public RecordBuilder<Tag> mapBuilder() {
       return new NbtOps.NbtRecordBuilder();
    }
 
-   class NbtRecordBuilder extends AbstractStringBuilder<Tag, CompoundTag> {
+   private class NbtRecordBuilder extends AbstractStringBuilder<Tag, CompoundTag> {
       protected NbtRecordBuilder() {
          super(NbtOps.this);
       }
 
       protected CompoundTag initBuilder() {
