--- net/minecraft/nbt/ShortTag.java
+++ net/minecraft/nbt/ShortTag.java
@@ -1,15 +1,13 @@
 package net.minecraft.nbt;
 
 import java.io.DataInput;
 import java.io.DataOutput;
 import java.io.IOException;
-import net.minecraft.network.chat.Component;
-import net.minecraft.network.chat.MutableComponent;
-import net.minecraft.network.chat.TextComponent;
 
 public class ShortTag extends NumericTag {
+   private static final int SELF_SIZE_IN_BITS = 80;
    public static final TagType<ShortTag> TYPE = new TagType<ShortTag>() {
       public ShortTag load(DataInput var1, int var2, NbtAccounter var3) throws IOException {
          var3.accountBits(80L);
          return ShortTag.valueOf(var1.readShort());
       }
@@ -26,16 +24,16 @@
          return true;
       }
    };
    private final short data;
 
-   private ShortTag(short var1) {
+   ShortTag(short var1) {
       this.data = var1;
    }
 
    public static ShortTag valueOf(short var0) {
-      return var0 >= -128 && var0 <= 1024 ? ShortTag.Cache.cache[var0 + 128] : new ShortTag(var0);
+      return var0 >= -128 && var0 <= 1024 ? ShortTag.Cache.cache[var0 - -128] : new ShortTag(var0);
    }
 
    public void write(DataOutput var1) throws IOException {
       var1.writeShort(this.data);
    }
@@ -46,14 +44,10 @@
 
    public TagType<ShortTag> getType() {
       return TYPE;
    }
 
-   public String toString() {
-      return this.data + "s";
-   }
-
    public ShortTag copy() {
       return this;
    }
 
    public boolean equals(Object var1) {
@@ -66,13 +60,12 @@
 
    public int hashCode() {
       return this.data;
    }
 
-   public Component getPrettyDisplay(String var1, int var2) {
-      MutableComponent var3 = (new TextComponent("s")).withStyle(SYNTAX_HIGHLIGHTING_NUMBER_TYPE);
-      return (new TextComponent(String.valueOf(this.data))).append(var3).withStyle(SYNTAX_HIGHLIGHTING_NUMBER);
+   public void accept(TagVisitor var1) {
+      var1.visitShort(this);
    }
 
    public long getAsLong() {
       return (long)this.data;
    }
@@ -99,13 +92,18 @@
 
    public Number getAsNumber() {
       return this.data;
    }
 
-   static class Cache {
+   private static class Cache {
+      private static final int HIGH = 1024;
+      private static final int LOW = -128;
       static final ShortTag[] cache = new ShortTag[1153];
 
+      private Cache() {
+      }
+
       static {
          for(int var0 = 0; var0 < cache.length; ++var0) {
             cache[var0] = new ShortTag((short)(-128 + var0));
          }
 
