--- net/minecraft/nbt/StringTag.java
+++ net/minecraft/nbt/StringTag.java
@@ -2,15 +2,13 @@
 
 import java.io.DataInput;
 import java.io.DataOutput;
 import java.io.IOException;
 import java.util.Objects;
-import net.minecraft.network.chat.Component;
-import net.minecraft.network.chat.MutableComponent;
-import net.minecraft.network.chat.TextComponent;
 
 public class StringTag implements Tag {
+   private static final int SELF_SIZE_IN_BITS = 288;
    public static final TagType<StringTag> TYPE = new TagType<StringTag>() {
       public StringTag load(DataInput var1, int var2, NbtAccounter var3) throws IOException {
          var3.accountBits(288L);
          String var4 = var1.readUTF();
          var3.accountBits((long)(16 * var4.length()));
@@ -28,10 +26,14 @@
       public boolean isValue() {
          return true;
       }
    };
    private static final StringTag EMPTY = new StringTag("");
+   private static final char DOUBLE_QUOTE = '"';
+   private static final char SINGLE_QUOTE = '\'';
+   private static final char ESCAPE = '\\';
+   private static final char NOT_SET = '\u0000';
    private final String data;
 
    private StringTag(String var1) {
       Objects.requireNonNull(var1, "Null string not allowed");
       this.data = var1;
@@ -52,11 +54,11 @@
    public TagType<StringTag> getType() {
       return TYPE;
    }
 
    public String toString() {
-      return quoteAndEscape(this.data);
+      return Tag.super.getAsString();
    }
 
    public StringTag copy() {
       return this;
    }
@@ -75,15 +77,12 @@
 
    public String getAsString() {
       return this.data;
    }
 
-   public Component getPrettyDisplay(String var1, int var2) {
-      String var3 = quoteAndEscape(this.data);
-      String var4 = var3.substring(0, 1);
-      MutableComponent var5 = (new TextComponent(var3.substring(1, var3.length() - 1))).withStyle(SYNTAX_HIGHLIGHTING_STRING);
-      return (new TextComponent(var4)).append(var5).append(var4);
+   public void accept(TagVisitor var1) {
+      var1.visitString(this);
    }
 
    public static String quoteAndEscape(String var0) {
       StringBuilder var1 = new StringBuilder(" ");
       int var2 = 0;
