--- net/minecraft/nbt/TagType.java
+++ net/minecraft/nbt/TagType.java
@@ -14,11 +14,11 @@
 
    String getPrettyName();
 
    static TagType<EndTag> createInvalid(final int var0) {
       return new TagType<EndTag>() {
-         public EndTag load(DataInput var1, int var2, NbtAccounter var3) throws IOException {
+         public EndTag load(DataInput var1, int var2, NbtAccounter var3) {
             throw new IllegalArgumentException("Invalid tag id: " + var0);
          }
 
          public String getName() {
             return "INVALID[" + var0 + "]";
