--- net/minecraft/util/profiling/metrics/MetricCategory.java
+++ net/minecraft/util/profiling/metrics/MetricCategory.java
@@ -1,0 +1,22 @@
+package net.minecraft.util.profiling.metrics;
+
+public enum MetricCategory {
+   PATH_FINDING("pathfinding"),
+   EVENT_LOOPS("event-loops"),
+   MAIL_BOXES("mailboxes"),
+   TICK_LOOP("ticking"),
+   JVM("jvm"),
+   CHUNK_RENDERING("chunk rendering"),
+   CHUNK_RENDERING_DISPATCHING("chunk rendering dispatching"),
+   CPU("cpu");
+
+   private final String description;
+
+   private MetricCategory(String var3) {
+      this.description = var3;
+   }
+
+   public String getDescription() {
+      return this.description;
+   }
+}
