--- net/minecraft/world/InteractionResult.java
+++ net/minecraft/world/InteractionResult.java
@@ -1,23 +1,28 @@
 package net.minecraft.world;
 
 public enum InteractionResult {
    SUCCESS,
    CONSUME,
+   CONSUME_PARTIAL,
    PASS,
    FAIL;
 
    private InteractionResult() {
    }
 
    public boolean consumesAction() {
-      return this == SUCCESS || this == CONSUME;
+      return this == SUCCESS || this == CONSUME || this == CONSUME_PARTIAL;
    }
 
    public boolean shouldSwing() {
       return this == SUCCESS;
    }
 
+   public boolean shouldAwardStats() {
+      return this == SUCCESS || this == CONSUME;
+   }
+
    public static InteractionResult sidedSuccess(boolean var0) {
       return var0 ? SUCCESS : CONSUME;
    }
 }
