--- net/minecraft/network/protocol/login/ClientboundGameProfilePacket.java
+++ net/minecraft/network/protocol/login/ClientboundGameProfilePacket.java
@@ -1,25 +1,21 @@
 package net.minecraft.network.protocol.login;
 
 import com.mojang.authlib.GameProfile;
-import java.io.IOException;
 import java.util.UUID;
 import net.minecraft.core.SerializableUUID;
 import net.minecraft.network.FriendlyByteBuf;
 import net.minecraft.network.protocol.Packet;
 
 public class ClientboundGameProfilePacket implements Packet<ClientLoginPacketListener> {
-   private GameProfile gameProfile;
+   private final GameProfile gameProfile;
 
-   public ClientboundGameProfilePacket() {
-   }
-
    public ClientboundGameProfilePacket(GameProfile var1) {
       this.gameProfile = var1;
    }
 
-   public void read(FriendlyByteBuf var1) throws IOException {
+   public ClientboundGameProfilePacket(FriendlyByteBuf var1) {
       int[] var2 = new int[4];
 
       for(int var3 = 0; var3 < var2.length; ++var3) {
          var2[var3] = var1.readInt();
       }
@@ -27,11 +23,11 @@
       UUID var5 = SerializableUUID.uuidFromIntArray(var2);
       String var4 = var1.readUtf(16);
       this.gameProfile = new GameProfile(var5, var4);
    }
 
-   public void write(FriendlyByteBuf var1) throws IOException {
+   public void write(FriendlyByteBuf var1) {
       int[] var2 = SerializableUUID.uuidToIntArray(this.gameProfile.getId());
       int var3 = var2.length;
 
       for(int var4 = 0; var4 < var3; ++var4) {
          int var5 = var2[var4];
@@ -42,6 +38,10 @@
    }
 
    public void handle(ClientLoginPacketListener var1) {
       var1.handleGameProfile(this);
    }
+
+   public GameProfile getGameProfile() {
+      return this.gameProfile;
+   }
 }
