--- net/minecraft/network/protocol/login/ClientboundLoginCompressionPacket.java
+++ net/minecraft/network/protocol/login/ClientboundLoginCompressionPacket.java
@@ -1,28 +1,28 @@
 package net.minecraft.network.protocol.login;
 
-import java.io.IOException;
 import net.minecraft.network.FriendlyByteBuf;
 import net.minecraft.network.protocol.Packet;
 
 public class ClientboundLoginCompressionPacket implements Packet<ClientLoginPacketListener> {
-   private int compressionThreshold;
+   private final int compressionThreshold;
 
-   public ClientboundLoginCompressionPacket() {
-   }
-
    public ClientboundLoginCompressionPacket(int var1) {
       this.compressionThreshold = var1;
    }
 
-   public void read(FriendlyByteBuf var1) throws IOException {
+   public ClientboundLoginCompressionPacket(FriendlyByteBuf var1) {
       this.compressionThreshold = var1.readVarInt();
    }
 
-   public void write(FriendlyByteBuf var1) throws IOException {
+   public void write(FriendlyByteBuf var1) {
       var1.writeVarInt(this.compressionThreshold);
    }
 
    public void handle(ClientLoginPacketListener var1) {
       var1.handleCompression(this);
    }
+
+   public int getCompressionThreshold() {
+      return this.compressionThreshold;
+   }
 }
