--- net/minecraft/network/protocol/login/package-info.java
+++ net/minecraft/network/protocol/login/package-info.java
@@ -1,0 +1,10 @@
+package net.minecraft.network.protocol.login;
+
+import javax.annotation.ParametersAreNonnullByDefault;
+import net.minecraft.MethodsReturnNonnullByDefault;
+
+// $FF: synthetic class
+@ParametersAreNonnullByDefault
+@MethodsReturnNonnullByDefault
+interface package-info {
+}
