--- net/minecraft/network/chat/BaseComponent.java
+++ net/minecraft/network/chat/BaseComponent.java
@@ -1,15 +1,19 @@
 package net.minecraft.network.chat;
 
 import com.google.common.collect.Lists;
 import java.util.List;
 import java.util.Objects;
+import javax.annotation.Nullable;
+import net.minecraft.locale.Language;
 import net.minecraft.util.FormattedCharSequence;
 
 public abstract class BaseComponent implements MutableComponent {
    protected final List<Component> siblings = Lists.newArrayList();
    private FormattedCharSequence visualOrderText;
+   @Nullable
+   private Language decomposedWith;
    private Style style;
 
    public BaseComponent() {
       this.visualOrderText = FormattedCharSequence.EMPTY;
       this.style = Style.EMPTY;
@@ -44,10 +48,20 @@
       var1.siblings.addAll(this.siblings);
       var1.setStyle(this.style);
       return var1;
    }
 
+   public FormattedCharSequence getVisualOrderText() {
+      Language var1 = Language.getInstance();
+      if (this.decomposedWith != var1) {
+         this.visualOrderText = var1.getVisualOrder((FormattedText)this);
+         this.decomposedWith = var1;
+      }
+
+      return this.visualOrderText;
+   }
+
    public boolean equals(Object var1) {
       if (this == var1) {
          return true;
       } else if (!(var1 instanceof BaseComponent)) {
          return false;
@@ -60,8 +74,8 @@
    public int hashCode() {
       return Objects.hash(new Object[]{this.getStyle(), this.siblings});
    }
 
    public String toString() {
-      return "BaseComponent{style=" + this.style + ", siblings=" + this.siblings + '}';
+      return "BaseComponent{style=" + this.style + ", siblings=" + this.siblings + "}";
    }
 }
