--- net/minecraft/network/chat/KeybindComponent.java
+++ net/minecraft/network/chat/KeybindComponent.java
@@ -15,10 +15,14 @@
 
    public KeybindComponent(String var1) {
       this.name = var1;
    }
 
+   public static void setKeyResolver(Function<String, Supplier<Component>> var0) {
+      keyResolver = var0;
+   }
+
    private Component getNestedComponent() {
       if (this.nameResolver == null) {
          this.nameResolver = (Supplier)keyResolver.apply(this.name);
       }
 
@@ -27,10 +31,14 @@
 
    public <T> Optional<T> visitSelf(FormattedText.ContentConsumer<T> var1) {
       return this.getNestedComponent().visit(var1);
    }
 
+   public <T> Optional<T> visitSelf(FormattedText.StyledContentConsumer<T> var1, Style var2) {
+      return this.getNestedComponent().visit(var1, var2);
+   }
+
    public KeybindComponent plainCopy() {
       return new KeybindComponent(this.name);
    }
 
    public boolean equals(Object var1) {
