--- net/minecraft/network/chat/NbtComponent.java
+++ net/minecraft/network/chat/NbtComponent.java
@@ -24,10 +24,11 @@
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
 
 public abstract class NbtComponent extends BaseComponent implements ContextAwareComponent {
    private static final Logger LOGGER = LogManager.getLogger();
+   private static final String SEPARATOR = ", ";
    protected final boolean interpreting;
    protected final String nbtPathPattern;
    @Nullable
    protected final NbtPathArgument.NbtPath compiledNbtPath;
 
@@ -72,11 +73,11 @@
          return (MutableComponent)(this.interpreting ? (MutableComponent)var4.flatMap((var3x) -> {
             try {
                MutableComponent var4 = Component.Serializer.fromJson(var3x);
                return Stream.of(ComponentUtils.updateForEntity(var1, var4, var2, var3));
             } catch (Exception var5) {
-               LOGGER.warn((String)("Failed to parse component: " + var3x), (Throwable)var5);
+               LOGGER.warn((String)"Failed to parse component: {}", (Object)var3x, (Object)var5);
                return Stream.of();
             }
          }).reduce((var0, var1x) -> {
             return var0.append(", ").append((Component)var1x);
          }).orElse(new TextComponent("")) : new TextComponent(Joiner.on(", ").join(var4.iterator())));
