--- net/minecraft/network/chat/ScoreComponent.java
+++ net/minecraft/network/chat/ScoreComponent.java
@@ -13,10 +13,11 @@
 import net.minecraft.world.entity.Entity;
 import net.minecraft.world.scores.Objective;
 import net.minecraft.world.scores.Score;
 
 public class ScoreComponent extends BaseComponent implements ContextAwareComponent {
+   private static final String SCORER_PLACEHOLDER = "*";
    private final String name;
    @Nullable
    private final EntitySelector selector;
    private final String objective;
 
@@ -41,10 +42,15 @@
 
    public String getName() {
       return this.name;
    }
 
+   @Nullable
+   public EntitySelector getSelector() {
+      return this.selector;
+   }
+
    public String getObjective() {
       return this.objective;
    }
 
    private String findTargetName(CommandSourceStack var1) throws CommandSyntaxException {
@@ -100,8 +106,9 @@
          return this.name.equals(var2.name) && this.objective.equals(var2.objective) && super.equals(var1);
       }
    }
 
    public String toString() {
-      return "ScoreComponent{name='" + this.name + '\'' + "objective='" + this.objective + '\'' + ", siblings=" + this.siblings + ", style=" + this.getStyle() + '}';
+      String var10000 = this.name;
+      return "ScoreComponent{name='" + var10000 + "'objective='" + this.objective + "', siblings=" + this.siblings + ", style=" + this.getStyle() + "}";
    }
 }
