--- net/minecraft/network/chat/SelectorComponent.java
+++ net/minecraft/network/chat/SelectorComponent.java
@@ -22,20 +22,25 @@
 
       try {
          EntitySelectorParser var3 = new EntitySelectorParser(new StringReader(var1));
          var2 = var3.parse();
       } catch (CommandSyntaxException var4) {
-         LOGGER.warn((String)"Invalid selector component: {}", (Object)var1, (Object)var4.getMessage());
+         LOGGER.warn((String)"Invalid selector component: {}: {}", (Object)var1, (Object)var4.getMessage());
       }
 
       this.selector = var2;
    }
 
    public String getPattern() {
       return this.pattern;
    }
 
+   @Nullable
+   public EntitySelector getSelector() {
+      return this.selector;
+   }
+
    public MutableComponent resolve(@Nullable CommandSourceStack var1, @Nullable Entity var2, int var3) throws CommandSyntaxException {
       return (MutableComponent)(var1 != null && this.selector != null ? EntitySelector.joinNames(this.selector.findEntities(var1)) : new TextComponent(""));
    }
 
    public String getContents() {
