--- net/minecraft/network/chat/Style.java
+++ net/minecraft/network/chat/Style.java
@@ -18,31 +18,31 @@
 
 public class Style {
    public static final Style EMPTY = new Style((TextColor)null, (Boolean)null, (Boolean)null, (Boolean)null, (Boolean)null, (Boolean)null, (ClickEvent)null, (HoverEvent)null, (String)null, (ResourceLocation)null);
    public static final ResourceLocation DEFAULT_FONT = new ResourceLocation("minecraft", "default");
    @Nullable
-   private final TextColor color;
+   final TextColor color;
    @Nullable
-   private final Boolean bold;
+   final Boolean bold;
    @Nullable
-   private final Boolean italic;
+   final Boolean italic;
    @Nullable
-   private final Boolean underlined;
+   final Boolean underlined;
    @Nullable
-   private final Boolean strikethrough;
+   final Boolean strikethrough;
    @Nullable
-   private final Boolean obfuscated;
+   final Boolean obfuscated;
    @Nullable
-   private final ClickEvent clickEvent;
+   final ClickEvent clickEvent;
    @Nullable
-   private final HoverEvent hoverEvent;
+   final HoverEvent hoverEvent;
    @Nullable
-   private final String insertion;
+   final String insertion;
    @Nullable
-   private final ResourceLocation font;
+   final ResourceLocation font;
 
-   private Style(@Nullable TextColor var1, @Nullable Boolean var2, @Nullable Boolean var3, @Nullable Boolean var4, @Nullable Boolean var5, @Nullable Boolean var6, @Nullable ClickEvent var7, @Nullable HoverEvent var8, @Nullable String var9, @Nullable ResourceLocation var10) {
+   Style(@Nullable TextColor var1, @Nullable Boolean var2, @Nullable Boolean var3, @Nullable Boolean var4, @Nullable Boolean var5, @Nullable Boolean var6, @Nullable ClickEvent var7, @Nullable HoverEvent var8, @Nullable String var9, @Nullable ResourceLocation var10) {
       this.color = var1;
       this.bold = var2;
       this.italic = var3;
       this.underlined = var4;
       this.strikethrough = var5;
@@ -107,18 +107,34 @@
 
    public Style withColor(@Nullable ChatFormatting var1) {
       return this.withColor(var1 != null ? TextColor.fromLegacyFormat(var1) : null);
    }
 
+   public Style withColor(int var1) {
+      return this.withColor(TextColor.fromRgb(var1));
+   }
+
    public Style withBold(@Nullable Boolean var1) {
       return new Style(this.color, var1, this.italic, this.underlined, this.strikethrough, this.obfuscated, this.clickEvent, this.hoverEvent, this.insertion, this.font);
    }
 
    public Style withItalic(@Nullable Boolean var1) {
       return new Style(this.color, this.bold, var1, this.underlined, this.strikethrough, this.obfuscated, this.clickEvent, this.hoverEvent, this.insertion, this.font);
    }
 
+   public Style withUnderlined(@Nullable Boolean var1) {
+      return new Style(this.color, this.bold, this.italic, var1, this.strikethrough, this.obfuscated, this.clickEvent, this.hoverEvent, this.insertion, this.font);
+   }
+
+   public Style withStrikethrough(@Nullable Boolean var1) {
+      return new Style(this.color, this.bold, this.italic, this.underlined, var1, this.obfuscated, this.clickEvent, this.hoverEvent, this.insertion, this.font);
+   }
+
+   public Style withObfuscated(@Nullable Boolean var1) {
+      return new Style(this.color, this.bold, this.italic, this.underlined, this.strikethrough, var1, this.clickEvent, this.hoverEvent, this.insertion, this.font);
+   }
+
    public Style withClickEvent(@Nullable ClickEvent var1) {
       return new Style(this.color, this.bold, this.italic, this.underlined, this.strikethrough, this.obfuscated, var1, this.hoverEvent, this.insertion, this.font);
    }
 
    public Style withHoverEvent(@Nullable HoverEvent var1) {
@@ -127,10 +143,14 @@
 
    public Style withInsertion(@Nullable String var1) {
       return new Style(this.color, this.bold, this.italic, this.underlined, this.strikethrough, this.obfuscated, this.clickEvent, this.hoverEvent, var1, this.font);
    }
 
+   public Style withFont(@Nullable ResourceLocation var1) {
+      return new Style(this.color, this.bold, this.italic, this.underlined, this.strikethrough, this.obfuscated, this.clickEvent, this.hoverEvent, this.insertion, var1);
+   }
+
    public Style applyFormat(ChatFormatting var1) {
       TextColor var2 = this.color;
       Boolean var3 = this.bold;
       Boolean var4 = this.italic;
       Boolean var5 = this.strikethrough;
@@ -159,10 +179,47 @@
       }
 
       return new Style(var2, var3, var4, var6, var5, var7, this.clickEvent, this.hoverEvent, this.insertion, this.font);
    }
 
+   public Style applyLegacyFormat(ChatFormatting var1) {
+      TextColor var2 = this.color;
+      Boolean var3 = this.bold;
+      Boolean var4 = this.italic;
+      Boolean var5 = this.strikethrough;
+      Boolean var6 = this.underlined;
+      Boolean var7 = this.obfuscated;
+      switch(var1) {
+      case OBFUSCATED:
+         var7 = true;
+         break;
+      case BOLD:
+         var3 = true;
+         break;
+      case STRIKETHROUGH:
+         var5 = true;
+         break;
+      case UNDERLINE:
+         var6 = true;
+         break;
+      case ITALIC:
+         var4 = true;
+         break;
+      case RESET:
+         return EMPTY;
+      default:
+         var7 = false;
+         var3 = false;
+         var5 = false;
+         var6 = false;
+         var4 = false;
+         var2 = TextColor.fromLegacyFormat(var1);
+      }
+
+      return new Style(var2, var3, var4, var6, var5, var7, this.clickEvent, this.hoverEvent, this.insertion, this.font);
+   }
+
    public Style applyFormats(ChatFormatting... var1) {
       TextColor var2 = this.color;
       Boolean var3 = this.bold;
       Boolean var4 = this.italic;
       Boolean var5 = this.strikethrough;
@@ -206,11 +263,12 @@
          return var1 == EMPTY ? this : new Style(this.color != null ? this.color : var1.color, this.bold != null ? this.bold : var1.bold, this.italic != null ? this.italic : var1.italic, this.underlined != null ? this.underlined : var1.underlined, this.strikethrough != null ? this.strikethrough : var1.strikethrough, this.obfuscated != null ? this.obfuscated : var1.obfuscated, this.clickEvent != null ? this.clickEvent : var1.clickEvent, this.hoverEvent != null ? this.hoverEvent : var1.hoverEvent, this.insertion != null ? this.insertion : var1.insertion, this.font != null ? this.font : var1.font);
       }
    }
 
    public String toString() {
-      return "Style{ color=" + this.color + ", bold=" + this.bold + ", italic=" + this.italic + ", underlined=" + this.underlined + ", strikethrough=" + this.strikethrough + ", obfuscated=" + this.obfuscated + ", clickEvent=" + this.getClickEvent() + ", hoverEvent=" + this.getHoverEvent() + ", insertion=" + this.getInsertion() + ", font=" + this.getFont() + '}';
+      TextColor var10000 = this.color;
+      return "Style{ color=" + var10000 + ", bold=" + this.bold + ", italic=" + this.italic + ", underlined=" + this.underlined + ", strikethrough=" + this.strikethrough + ", obfuscated=" + this.obfuscated + ", clickEvent=" + this.getClickEvent() + ", hoverEvent=" + this.getHoverEvent() + ", insertion=" + this.getInsertion() + ", font=" + this.getFont() + "}";
    }
 
    public boolean equals(Object var1) {
       if (this == var1) {
          return true;
