--- net/minecraft/network/chat/TranslatableComponent.java
+++ net/minecraft/network/chat/TranslatableComponent.java
@@ -118,10 +118,27 @@
 
    public TranslatableComponent plainCopy() {
       return new TranslatableComponent(this.key, this.args);
    }
 
+   public <T> Optional<T> visitSelf(FormattedText.StyledContentConsumer<T> var1, Style var2) {
+      this.decompose();
+      Iterator var3 = this.decomposedParts.iterator();
+
+      Optional var5;
+      do {
+         if (!var3.hasNext()) {
+            return Optional.empty();
+         }
+
+         FormattedText var4 = (FormattedText)var3.next();
+         var5 = var4.visit(var1, var2);
+      } while(!var5.isPresent());
+
+      return var5;
+   }
+
    public <T> Optional<T> visitSelf(FormattedText.ContentConsumer<T> var1) {
       this.decompose();
       Iterator var2 = this.decomposedParts.iterator();
 
       Optional var4;
