--- net/minecraft/world/entity/raid/Raid.java
+++ net/minecraft/world/entity/raid/Raid.java
@@ -7,10 +7,11 @@
 import java.util.HashSet;
 import java.util.Iterator;
 import java.util.List;
 import java.util.Locale;
 import java.util.Map;
+import java.util.Objects;
 import java.util.Optional;
 import java.util.Random;
 import java.util.Set;
 import java.util.UUID;
 import java.util.function.Predicate;
@@ -47,24 +48,44 @@
 import net.minecraft.world.entity.SpawnPlacements;
 import net.minecraft.world.entity.player.Player;
 import net.minecraft.world.item.DyeColor;
 import net.minecraft.world.item.ItemStack;
 import net.minecraft.world.item.Items;
-import net.minecraft.world.level.ChunkPos;
 import net.minecraft.world.level.Level;
 import net.minecraft.world.level.NaturalSpawner;
 import net.minecraft.world.level.block.Blocks;
 import net.minecraft.world.level.block.entity.BannerPattern;
 import net.minecraft.world.level.levelgen.Heightmap;
 import net.minecraft.world.phys.Vec3;
 
 public class Raid {
+   private static final int SECTION_RADIUS_FOR_FINDING_NEW_VILLAGE_CENTER = 2;
+   private static final int ATTEMPT_RAID_FARTHEST = 0;
+   private static final int ATTEMPT_RAID_CLOSE = 1;
+   private static final int ATTEMPT_RAID_INSIDE = 2;
+   private static final int VILLAGE_SEARCH_RADIUS = 32;
+   private static final int RAID_TIMEOUT_TICKS = 48000;
+   private static final int NUM_SPAWN_ATTEMPTS = 3;
+   private static final String OMINOUS_BANNER_PATTERN_NAME = "block.minecraft.ominous_banner";
+   private static final String RAIDERS_REMAINING = "event.minecraft.raid.raiders_remaining";
+   public static final int VILLAGE_RADIUS_BUFFER = 16;
+   private static final int POST_RAID_TICK_LIMIT = 40;
+   private static final int DEFAULT_PRE_RAID_TICKS = 300;
+   public static final int MAX_NO_ACTION_TIME = 2400;
+   public static final int MAX_CELEBRATION_TICKS = 600;
+   private static final int OUTSIDE_RAID_BOUNDS_TIMEOUT = 30;
+   public static final int TICKS_PER_DAY = 24000;
+   public static final int DEFAULT_MAX_BAD_OMEN_LEVEL = 5;
+   private static final int LOW_MOB_THRESHOLD = 2;
    private static final Component RAID_NAME_COMPONENT = new TranslatableComponent("event.minecraft.raid");
    private static final Component VICTORY = new TranslatableComponent("event.minecraft.raid.victory");
    private static final Component DEFEAT = new TranslatableComponent("event.minecraft.raid.defeat");
    private static final Component RAID_BAR_VICTORY_COMPONENT;
    private static final Component RAID_BAR_DEFEAT_COMPONENT;
+   private static final int HERO_OF_THE_VILLAGE_DURATION = 48000;
+   public static final int VALID_RAID_RADIUS_SQR = 9216;
+   public static final int RAID_REMOVAL_THRESHOLD_SQR = 12544;
    private final Map<Integer, Raider> groupToLeaderMap = Maps.newHashMap();
    private final Map<Integer, Set<Raider>> groupRaiderMap = Maps.newHashMap();
    private final Set<UUID> heroesOfTheVillage = Sets.newHashSet();
    private long ticksActive;
    private BlockPos center;
@@ -90,11 +111,11 @@
       this.waveSpawnPos = Optional.empty();
       this.id = var1;
       this.level = var2;
       this.active = true;
       this.raidCooldownTicks = 300;
-      this.raidEvent.setPercent(0.0F);
+      this.raidEvent.setProgress(0.0F);
       this.center = var3;
       this.numGroups = this.getNumGroups(var2.getDifficulty());
       this.status = Raid.RaidStatus.ONGOING;
    }
 
@@ -148,10 +169,26 @@
 
    public boolean isLoss() {
       return this.status == Raid.RaidStatus.LOSS;
    }
 
+   public float getTotalHealth() {
+      return this.totalHealth;
+   }
+
+   public Set<Raider> getAllRaiders() {
+      HashSet var1 = Sets.newHashSet();
+      Iterator var2 = this.groupRaiderMap.values().iterator();
+
+      while(var2.hasNext()) {
+         Set var3 = (Set)var2.next();
+         var1.addAll(var3);
+      }
+
+      return var1;
+   }
+
    public Level getLevel() {
       return this.level;
    }
 
    public boolean isStarted() {
@@ -199,14 +236,18 @@
 
    public int getBadOmenLevel() {
       return this.badOmenLevel;
    }
 
+   public void setBadOmenLevel(int var1) {
+      this.badOmenLevel = var1;
+   }
+
    public void absorbBadOmen(Player var1) {
       if (var1.hasEffect(MobEffects.BAD_OMEN)) {
          this.badOmenLevel += var1.getEffect(MobEffects.BAD_OMEN).getAmplifier() + 1;
-         this.badOmenLevel = Mth.clamp(this.badOmenLevel, 0, this.getMaxBadOmenLevel());
+         this.badOmenLevel = Mth.clamp((int)this.badOmenLevel, (int)0, (int)this.getMaxBadOmenLevel());
       }
 
       var1.removeEffect(MobEffects.BAD_OMEN);
    }
 
@@ -262,11 +303,11 @@
                      return;
                   }
                } else {
                   var3 = this.waveSpawnPos.isPresent();
                   boolean var4 = !var3 && this.raidCooldownTicks % 5 == 0;
-                  if (var3 && !this.level.getChunkSource().isEntityTickingChunk(new ChunkPos((BlockPos)this.waveSpawnPos.get()))) {
+                  if (var3 && !this.level.isPositionEntityTicking((BlockPos)this.waveSpawnPos.get())) {
                      var4 = true;
                   }
 
                   if (var4) {
                      byte var5 = 0;
@@ -282,11 +323,11 @@
                   if (this.raidCooldownTicks == 300 || this.raidCooldownTicks % 20 == 0) {
                      this.updatePlayers();
                   }
 
                   --this.raidCooldownTicks;
-                  this.raidEvent.setPercent(Mth.clamp((float)(300 - this.raidCooldownTicks) / 300.0F, 0.0F, 1.0F));
+                  this.raidEvent.setProgress(Mth.clamp((float)(300 - this.raidCooldownTicks) / 300.0F, 0.0F, 1.0F));
                }
             }
 
             if (this.ticksActive % 20L == 0L) {
                this.updatePlayers();
@@ -357,11 +398,11 @@
 
             if (this.celebrationTicks % 20 == 0) {
                this.updatePlayers();
                this.raidEvent.setVisible(true);
                if (this.isVictory()) {
-                  this.raidEvent.setPercent(0.0F);
+                  this.raidEvent.setProgress(0.0F);
                   this.raidEvent.setName(RAID_BAR_VICTORY_COMPONENT);
                } else {
                   this.raidEvent.setName(RAID_BAR_DEFEAT_COMPONENT);
                }
             }
@@ -371,11 +412,11 @@
    }
 
    private void moveRaidCenterToNearbyVillageSection() {
       Stream var1 = SectionPos.cube(SectionPos.of(this.center), 2);
       ServerLevel var10001 = this.level;
-      var10001.getClass();
+      Objects.requireNonNull(var10001);
       var1.filter(var10001::isVillage).map(SectionPos::center).min(Comparator.comparingDouble((var1x) -> {
          return var1x.distSqr(this.center);
       })).ifPresent(this::setCenter);
    }
 
@@ -429,11 +470,11 @@
                   continue label54;
                }
 
                Raider var5 = (Raider)var4.next();
                BlockPos var6 = var5.blockPosition();
-               if (!var5.removed && var5.level.dimension() == this.level.dimension() && this.center.distSqr(var6) < 12544.0D) {
+               if (!var5.isRemoved() && var5.level.dimension() == this.level.dimension() && this.center.distSqr(var6) < 12544.0D) {
                   if (var5.tickCount > 600) {
                      if (this.level.getEntity(var5.getUUID()) == null) {
                         var2.add(var5);
                      }
 
@@ -467,27 +508,27 @@
       Collection var4 = this.raidEvent.getPlayers();
       Iterator var5 = this.level.players().iterator();
 
       while(true) {
          ServerPlayer var6;
-         float var9;
-         double var10;
-         double var12;
+         double var9;
+         double var11;
+         double var13;
          do {
             if (!var5.hasNext()) {
                return;
             }
 
             var6 = (ServerPlayer)var5.next();
             Vec3 var7 = var6.position();
             Vec3 var8 = Vec3.atCenterOf(var1);
-            var9 = Mth.sqrt((var8.x - var7.x) * (var8.x - var7.x) + (var8.z - var7.z) * (var8.z - var7.z));
-            var10 = var7.x + (double)(13.0F / var9) * (var8.x - var7.x);
-            var12 = var7.z + (double)(13.0F / var9) * (var8.z - var7.z);
-         } while(var9 > 64.0F && !var4.contains(var6));
+            var9 = Math.sqrt((var8.x - var7.x) * (var8.x - var7.x) + (var8.z - var7.z) * (var8.z - var7.z));
+            var11 = var7.x + 13.0D / var9 * (var8.x - var7.x);
+            var13 = var7.z + 13.0D / var9 * (var8.z - var7.z);
+         } while(var9 > 64.0D && !var4.contains(var6));
 
-         var6.connection.send(new ClientboundSoundPacket(SoundEvents.RAID_HORN, SoundSource.NEUTRAL, var10, var6.getY(), var12, 64.0F, 1.0F));
+         var6.connection.send(new ClientboundSoundPacket(SoundEvents.RAID_HORN, SoundSource.NEUTRAL, var11, var6.getY(), var13, 64.0F, 1.0F));
       }
    }
 
    private void spawnGroup(BlockPos var1) {
       boolean var2 = false;
@@ -557,11 +598,11 @@
       }
 
    }
 
    public void updateBossbar() {
-      this.raidEvent.setPercent(Mth.clamp(this.getHealthOfLivingRaiders() / this.totalHealth, 0.0F, 1.0F));
+      this.raidEvent.setProgress(Mth.clamp(this.getHealthOfLivingRaiders() / this.totalHealth, 0.0F, 1.0F));
    }
 
    public float getHealthOfLivingRaiders() {
       float var1 = 0.0F;
       Iterator var2 = this.groupRaiderMap.values().iterator();
@@ -631,12 +672,15 @@
          float var9 = this.level.random.nextFloat() * 6.2831855F;
          int var4 = this.center.getX() + Mth.floor(Mth.cos(var9) * 32.0F * (float)var3) + this.level.random.nextInt(5);
          int var6 = this.center.getZ() + Mth.floor(Mth.sin(var9) * 32.0F * (float)var3) + this.level.random.nextInt(5);
          int var5 = this.level.getHeight(Heightmap.Types.WORLD_SURFACE, var4, var6);
          var7.set(var4, var5, var6);
-         if ((!this.level.isVillage((BlockPos)var7) || var1 >= 2) && this.level.hasChunksAt(var7.getX() - 10, var7.getY() - 10, var7.getZ() - 10, var7.getX() + 10, var7.getY() + 10, var7.getZ() + 10) && this.level.getChunkSource().isEntityTickingChunk(new ChunkPos(var7)) && (NaturalSpawner.isSpawnPositionOk(SpawnPlacements.Type.ON_GROUND, this.level, var7, EntityType.RAVAGER) || this.level.getBlockState(var7.below()).is(Blocks.SNOW) && this.level.getBlockState(var7).isAir())) {
-            return var7;
+         if (!this.level.isVillage((BlockPos)var7) || var1 >= 2) {
+            boolean var10 = true;
+            if (this.level.hasChunksAt(var7.getX() - 10, var7.getZ() - 10, var7.getX() + 10, var7.getZ() + 10) && this.level.isPositionEntityTicking((BlockPos)var7) && (NaturalSpawner.isSpawnPositionOk(SpawnPlacements.Type.ON_GROUND, this.level, var7, EntityType.RAVAGER) || this.level.getBlockState(var7.below()).is(Blocks.SNOW) && this.level.getBlockState(var7).isAir())) {
+               return var7;
+            }
          }
       }
 
       return null;
    }
@@ -799,39 +843,22 @@
    static {
       RAID_BAR_VICTORY_COMPONENT = RAID_NAME_COMPONENT.copy().append(" - ").append(VICTORY);
       RAID_BAR_DEFEAT_COMPONENT = RAID_NAME_COMPONENT.copy().append(" - ").append(DEFEAT);
    }
 
-   static enum RaiderType {
-      VINDICATOR(EntityType.VINDICATOR, new int[]{0, 0, 2, 0, 1, 4, 2, 5}),
-      EVOKER(EntityType.EVOKER, new int[]{0, 0, 0, 0, 0, 1, 1, 2}),
-      PILLAGER(EntityType.PILLAGER, new int[]{0, 4, 3, 3, 4, 4, 4, 2}),
-      WITCH(EntityType.WITCH, new int[]{0, 0, 0, 0, 3, 0, 0, 1}),
-      RAVAGER(EntityType.RAVAGER, new int[]{0, 0, 0, 1, 0, 1, 0, 2});
-
-      private static final Raid.RaiderType[] VALUES = values();
-      private final EntityType<? extends Raider> entityType;
-      private final int[] spawnsPerWaveBeforeBonus;
-
-      private RaiderType(EntityType<? extends Raider> var3, int[] var4) {
-         this.entityType = var3;
-         this.spawnsPerWaveBeforeBonus = var4;
-      }
-   }
-
-   static enum RaidStatus {
+   private static enum RaidStatus {
       ONGOING,
       VICTORY,
       LOSS,
       STOPPED;
 
       private static final Raid.RaidStatus[] VALUES = values();
 
       private RaidStatus() {
       }
 
-      private static Raid.RaidStatus getByName(String var0) {
+      static Raid.RaidStatus getByName(String var0) {
          Raid.RaidStatus[] var1 = VALUES;
          int var2 = var1.length;
 
          for(int var3 = 0; var3 < var2; ++var3) {
             Raid.RaidStatus var4 = var1[var3];
@@ -845,6 +872,23 @@
 
       public String getName() {
          return this.name().toLowerCase(Locale.ROOT);
       }
    }
+
+   private static enum RaiderType {
+      VINDICATOR(EntityType.VINDICATOR, new int[]{0, 0, 2, 0, 1, 4, 2, 5}),
+      EVOKER(EntityType.EVOKER, new int[]{0, 0, 0, 0, 0, 1, 1, 2}),
+      PILLAGER(EntityType.PILLAGER, new int[]{0, 4, 3, 3, 4, 4, 4, 2}),
+      WITCH(EntityType.WITCH, new int[]{0, 0, 0, 0, 3, 0, 0, 1}),
+      RAVAGER(EntityType.RAVAGER, new int[]{0, 0, 0, 1, 0, 1, 0, 2});
+
+      static final Raid.RaiderType[] VALUES = values();
+      final EntityType<? extends Raider> entityType;
+      final int[] spawnsPerWaveBeforeBonus;
+
+      private RaiderType(EntityType<? extends Raider> var3, int[] var4) {
+         this.entityType = var3;
+         this.spawnsPerWaveBeforeBonus = var4;
+      }
+   }
 }
