--- net/minecraft/world/entity/raid/Raid.java
+++ net/minecraft/world/entity/raid/Raid.java
@@ -90,11 +90,11 @@
       this.waveSpawnPos = Optional.empty();
       this.id = var1;
       this.level = var2;
       this.active = true;
       this.raidCooldownTicks = 300;
-      this.raidEvent.setPercent(0.0F);
+      this.raidEvent.setProgress(0.0F);
       this.center = var3;
       this.numGroups = this.getNumGroups(var2.getDifficulty());
       this.status = Raid.RaidStatus.ONGOING;
    }
 
@@ -282,11 +282,11 @@
                   if (this.raidCooldownTicks == 300 || this.raidCooldownTicks % 20 == 0) {
                      this.updatePlayers();
                   }
 
                   --this.raidCooldownTicks;
-                  this.raidEvent.setPercent(Mth.clamp((float)(300 - this.raidCooldownTicks) / 300.0F, 0.0F, 1.0F));
+                  this.raidEvent.setProgress(Mth.clamp((float)(300 - this.raidCooldownTicks) / 300.0F, 0.0F, 1.0F));
                }
             }
 
             if (this.ticksActive % 20L == 0L) {
                this.updatePlayers();
@@ -357,11 +357,11 @@
 
             if (this.celebrationTicks % 20 == 0) {
                this.updatePlayers();
                this.raidEvent.setVisible(true);
                if (this.isVictory()) {
-                  this.raidEvent.setPercent(0.0F);
+                  this.raidEvent.setProgress(0.0F);
                   this.raidEvent.setName(RAID_BAR_VICTORY_COMPONENT);
                } else {
                   this.raidEvent.setName(RAID_BAR_DEFEAT_COMPONENT);
                }
             }
@@ -429,11 +429,11 @@
                   continue label54;
                }
 
                Raider var5 = (Raider)var4.next();
                BlockPos var6 = var5.blockPosition();
-               if (!var5.removed && var5.level.dimension() == this.level.dimension() && this.center.distSqr(var6) < 12544.0D) {
+               if (!var5.isRemoved() && var5.level.dimension() == this.level.dimension() && this.center.distSqr(var6) < 12544.0D) {
                   if (var5.tickCount > 600) {
                      if (this.level.getEntity(var5.getUUID()) == null) {
                         var2.add(var5);
                      }
 
@@ -557,11 +557,11 @@
       }
 
    }
 
    public void updateBossbar() {
-      this.raidEvent.setPercent(Mth.clamp(this.getHealthOfLivingRaiders() / this.totalHealth, 0.0F, 1.0F));
+      this.raidEvent.setProgress(Mth.clamp(this.getHealthOfLivingRaiders() / this.totalHealth, 0.0F, 1.0F));
    }
 
    public float getHealthOfLivingRaiders() {
       float var1 = 0.0F;
       Iterator var2 = this.groupRaiderMap.values().iterator();
@@ -631,12 +631,15 @@
          float var9 = this.level.random.nextFloat() * 6.2831855F;
          int var4 = this.center.getX() + Mth.floor(Mth.cos(var9) * 32.0F * (float)var3) + this.level.random.nextInt(5);
          int var6 = this.center.getZ() + Mth.floor(Mth.sin(var9) * 32.0F * (float)var3) + this.level.random.nextInt(5);
          int var5 = this.level.getHeight(Heightmap.Types.WORLD_SURFACE, var4, var6);
          var7.set(var4, var5, var6);
-         if ((!this.level.isVillage((BlockPos)var7) || var1 >= 2) && this.level.hasChunksAt(var7.getX() - 10, var7.getY() - 10, var7.getZ() - 10, var7.getX() + 10, var7.getY() + 10, var7.getZ() + 10) && this.level.getChunkSource().isEntityTickingChunk(new ChunkPos(var7)) && (NaturalSpawner.isSpawnPositionOk(SpawnPlacements.Type.ON_GROUND, this.level, var7, EntityType.RAVAGER) || this.level.getBlockState(var7.below()).is(Blocks.SNOW) && this.level.getBlockState(var7).isAir())) {
-            return var7;
+         if (!this.level.isVillage((BlockPos)var7) || var1 >= 2) {
+            boolean var10 = true;
+            if (this.level.hasChunksAt(var7.getX() - 10, var7.getZ() - 10, var7.getX() + 10, var7.getZ() + 10) && this.level.getChunkSource().isEntityTickingChunk(new ChunkPos(var7)) && (NaturalSpawner.isSpawnPositionOk(SpawnPlacements.Type.ON_GROUND, this.level, var7, EntityType.RAVAGER) || this.level.getBlockState(var7.below()).is(Blocks.SNOW) && this.level.getBlockState(var7).isAir())) {
+               return var7;
+            }
          }
       }
 
       return null;
    }
