--- net/minecraft/client/gui/Font.java
+++ net/minecraft/client/gui/Font.java
@@ -241,11 +241,11 @@
       private final float a;
       private final Matrix4f pose;
       private final boolean seeThrough;
       private final int packedLightCoords;
       private float x;
-      private float y;
+      private final float y;
       @Nullable
       private List<BakedGlyph.Effect> effects;
 
       private void addEffect(BakedGlyph.Effect var1) {
          if (this.effects == null) {
