--- net/minecraft/data/models/blockstates/VariantProperty.java
+++ net/minecraft/data/models/blockstates/VariantProperty.java
@@ -26,10 +26,14 @@
 
       public Value(T var2) {
          this.value = var2;
       }
 
+      public VariantProperty<T> getKey() {
+         return VariantProperty.this;
+      }
+
       public void addToVariant(JsonObject var1) {
          var1.add(VariantProperty.this.key, (JsonElement)VariantProperty.this.serializer.apply(this.value));
       }
 
       public String toString() {
