--- net/minecraft/world/level/block/piston/PistonMath.java
+++ net/minecraft/world/level/block/piston/PistonMath.java
@@ -2,10 +2,13 @@
 
 import net.minecraft.core.Direction;
 import net.minecraft.world.phys.AABB;
 
 public class PistonMath {
+   public PistonMath() {
+   }
+
    public static AABB getMovementArea(AABB var0, Direction var1, double var2) {
       double var4 = var2 * (double)var1.getAxisDirection().getStep();
       double var6 = Math.min(var4, 0.0D);
       double var8 = Math.max(var4, 0.0D);
       switch(var1) {
