--- net/minecraft/world/level/block/piston/PistonHeadBlock.java
+++ net/minecraft/world/level/block/piston/PistonHeadBlock.java
@@ -92,11 +92,11 @@
       Block var3 = var1.getValue(TYPE) == PistonType.DEFAULT ? Blocks.PISTON : Blocks.STICKY_PISTON;
       return var2.is(var3) && (Boolean)var2.getValue(PistonBaseBlock.EXTENDED) && var2.getValue(FACING) == var1.getValue(FACING);
    }
 
    public void playerWillDestroy(Level var1, BlockPos var2, BlockState var3, Player var4) {
-      if (!var1.isClientSide && var4.abilities.instabuild) {
+      if (!var1.isClientSide && var4.getAbilities().instabuild) {
          BlockPos var5 = var2.relative(((Direction)var3.getValue(FACING)).getOpposite());
          if (this.isFittingBase(var3, var1.getBlockState(var5))) {
             var1.destroyBlock(var5, false);
          }
       }
