--- net/minecraft/world/scores/Objective.java
+++ net/minecraft/world/scores/Objective.java
@@ -5,10 +5,11 @@
 import net.minecraft.network.chat.HoverEvent;
 import net.minecraft.network.chat.TextComponent;
 import net.minecraft.world.scores.criteria.ObjectiveCriteria;
 
 public class Objective {
+   public static final int MAX_NAME_LENGTH = 16;
    private final Scoreboard scoreboard;
    private final String name;
    private final ObjectiveCriteria criteria;
    private Component displayName;
    private Component formattedDisplayName;
@@ -21,10 +22,14 @@
       this.displayName = var4;
       this.formattedDisplayName = this.createFormattedDisplayName();
       this.renderType = var5;
    }
 
+   public Scoreboard getScoreboard() {
+      return this.scoreboard;
+   }
+
    public String getName() {
       return this.name;
    }
 
    public ObjectiveCriteria getCriteria() {
