--- net/minecraft/world/scores/Scoreboard.java
+++ net/minecraft/world/scores/Scoreboard.java
@@ -17,10 +17,17 @@
 import net.minecraft.world.entity.Entity;
 import net.minecraft.world.entity.player.Player;
 import net.minecraft.world.scores.criteria.ObjectiveCriteria;
 
 public class Scoreboard {
+   public static final int DISPLAY_SLOT_LIST = 0;
+   public static final int DISPLAY_SLOT_SIDEBAR = 1;
+   public static final int DISPLAY_SLOT_BELOW_NAME = 2;
+   public static final int DISPLAY_SLOT_TEAMS_SIDEBAR_START = 3;
+   public static final int DISPLAY_SLOT_TEAMS_SIDEBAR_END = 18;
+   public static final int DISPLAY_SLOTS = 19;
+   public static final int MAX_NAME_LENGTH = 40;
    private final Map<String, Objective> objectivesByName = Maps.newHashMap();
    private final Map<ObjectiveCriteria, List<Objective>> objectivesByCriteria = Maps.newHashMap();
    private final Map<String, Map<Objective, Score>> playerScores = Maps.newHashMap();
    private final Objective[] displayObjectives = new Objective[19];
    private final Map<String, PlayerTeam> teamsByName = Maps.newHashMap();
@@ -28,10 +35,14 @@
    private static String[] displaySlotNames;
 
    public Scoreboard() {
    }
 
+   public boolean hasObjective(String var1) {
+      return this.objectivesByName.containsKey(var1);
+   }
+
    public Objective getOrCreateObjective(String var1) {
       return (Objective)this.objectivesByName.get(var1);
    }
 
    @Nullable
