--- net/minecraft/world/scores/Team.java
+++ net/minecraft/world/scores/Team.java
@@ -3,10 +3,11 @@
 import java.util.Arrays;
 import java.util.Collection;
 import java.util.Map;
 import java.util.stream.Collectors;
 import javax.annotation.Nullable;
+import net.minecraft.ChatFormatting;
 import net.minecraft.network.chat.Component;
 import net.minecraft.network.chat.MutableComponent;
 import net.minecraft.network.chat.TranslatableComponent;
 
 public abstract class Team {
@@ -23,12 +24,18 @@
 
    public abstract String getName();
 
    public abstract MutableComponent getFormattedName(Component var1);
 
+   public abstract boolean canSeeFriendlyInvisibles();
+
    public abstract boolean isAllowFriendlyFire();
 
+   public abstract Team.Visibility getNameTagVisibility();
+
+   public abstract ChatFormatting getColor();
+
    public abstract Collection<String> getPlayers();
 
    public abstract Team.Visibility getDeathMessageVisibility();
 
    public abstract Team.CollisionRule getCollisionRule();
@@ -74,10 +81,14 @@
          return var0;
       }));
       public final String name;
       public final int id;
 
+      public static String[] getAllNames() {
+         return (String[])BY_NAME.keySet().toArray(new String[BY_NAME.size()]);
+      }
+
       @Nullable
       public static Team.Visibility byName(String var0) {
          return (Team.Visibility)BY_NAME.get(var0);
       }
 
