--- net/minecraft/client/renderer/entity/ArmorStandRenderer.java
+++ net/minecraft/client/renderer/entity/ArmorStandRenderer.java
@@ -3,10 +3,11 @@
 import com.mojang.blaze3d.vertex.PoseStack;
 import com.mojang.math.Vector3f;
 import javax.annotation.Nullable;
 import net.minecraft.client.model.ArmorStandArmorModel;
 import net.minecraft.client.model.ArmorStandModel;
+import net.minecraft.client.model.geom.ModelLayers;
 import net.minecraft.client.renderer.RenderType;
 import net.minecraft.client.renderer.entity.layers.CustomHeadLayer;
 import net.minecraft.client.renderer.entity.layers.ElytraLayer;
 import net.minecraft.client.renderer.entity.layers.HumanoidArmorLayer;
 import net.minecraft.client.renderer.entity.layers.ItemInHandLayer;
@@ -15,16 +16,16 @@
 import net.minecraft.world.entity.decoration.ArmorStand;
 
 public class ArmorStandRenderer extends LivingEntityRenderer<ArmorStand, ArmorStandArmorModel> {
    public static final ResourceLocation DEFAULT_SKIN_LOCATION = new ResourceLocation("textures/entity/armorstand/wood.png");
 
-   public ArmorStandRenderer(EntityRenderDispatcher var1) {
-      super(var1, new ArmorStandModel(), 0.0F);
-      this.addLayer(new HumanoidArmorLayer(this, new ArmorStandArmorModel(0.5F), new ArmorStandArmorModel(1.0F)));
+   public ArmorStandRenderer(EntityRendererProvider.Context var1) {
+      super(var1, new ArmorStandModel(var1.getLayer(ModelLayers.ARMOR_STAND)), 0.0F);
+      this.addLayer(new HumanoidArmorLayer(this, new ArmorStandArmorModel(var1.getLayer(ModelLayers.ARMOR_STAND_INNER_ARMOR)), new ArmorStandArmorModel(var1.getLayer(ModelLayers.ARMOR_STAND_OUTER_ARMOR))));
       this.addLayer(new ItemInHandLayer(this));
-      this.addLayer(new ElytraLayer(this));
-      this.addLayer(new CustomHeadLayer(this));
+      this.addLayer(new ElytraLayer(this, var1.getModelSet()));
+      this.addLayer(new CustomHeadLayer(this, var1.getModelSet()));
    }
 
    public ResourceLocation getTextureLocation(ArmorStand var1) {
       return DEFAULT_SKIN_LOCATION;
    }
