--- net/minecraft/client/renderer/entity/ArrowRenderer.java
+++ net/minecraft/client/renderer/entity/ArrowRenderer.java
@@ -10,11 +10,11 @@
 import net.minecraft.client.renderer.texture.OverlayTexture;
 import net.minecraft.util.Mth;
 import net.minecraft.world.entity.projectile.AbstractArrow;
 
 public abstract class ArrowRenderer<T extends AbstractArrow> extends EntityRenderer<T> {
-   public ArrowRenderer(EntityRenderDispatcher var1) {
+   public ArrowRenderer(EntityRendererProvider.Context var1) {
       super(var1);
    }
 
    public void render(T var1, float var2, float var3, PoseStack var4, MultiBufferSource var5, int var6) {
       var4.pushPose();
