--- net/minecraft/client/renderer/entity/ArrowRenderer.java
+++ net/minecraft/client/renderer/entity/ArrowRenderer.java
@@ -10,18 +10,18 @@
 import net.minecraft.client.renderer.texture.OverlayTexture;
 import net.minecraft.util.Mth;
 import net.minecraft.world.entity.projectile.AbstractArrow;
 
 public abstract class ArrowRenderer<T extends AbstractArrow> extends EntityRenderer<T> {
-   public ArrowRenderer(EntityRenderDispatcher var1) {
+   public ArrowRenderer(EntityRendererProvider.Context var1) {
       super(var1);
    }
 
    public void render(T var1, float var2, float var3, PoseStack var4, MultiBufferSource var5, int var6) {
       var4.pushPose();
-      var4.mulPose(Vector3f.YP.rotationDegrees(Mth.lerp(var3, var1.yRotO, var1.yRot) - 90.0F));
-      var4.mulPose(Vector3f.ZP.rotationDegrees(Mth.lerp(var3, var1.xRotO, var1.xRot)));
+      var4.mulPose(Vector3f.YP.rotationDegrees(Mth.lerp(var3, var1.yRotO, var1.getYRot()) - 90.0F));
+      var4.mulPose(Vector3f.ZP.rotationDegrees(Mth.lerp(var3, var1.xRotO, var1.getXRot())));
       boolean var7 = false;
       float var8 = 0.0F;
       float var9 = 0.5F;
       float var10 = 0.0F;
       float var11 = 0.15625F;
