--- net/minecraft/client/renderer/entity/CowRenderer.java
+++ net/minecraft/client/renderer/entity/CowRenderer.java
@@ -1,16 +1,17 @@
 package net.minecraft.client.renderer.entity;
 
 import net.minecraft.client.model.CowModel;
+import net.minecraft.client.model.geom.ModelLayers;
 import net.minecraft.resources.ResourceLocation;
 import net.minecraft.world.entity.animal.Cow;
 
 public class CowRenderer extends MobRenderer<Cow, CowModel<Cow>> {
    private static final ResourceLocation COW_LOCATION = new ResourceLocation("textures/entity/cow/cow.png");
 
-   public CowRenderer(EntityRenderDispatcher var1) {
-      super(var1, new CowModel(), 0.7F);
+   public CowRenderer(EntityRendererProvider.Context var1) {
+      super(var1, new CowModel(var1.getLayer(ModelLayers.COW)), 0.7F);
    }
 
    public ResourceLocation getTextureLocation(Cow var1) {
       return COW_LOCATION;
    }
