--- net/minecraft/client/renderer/entity/CreeperRenderer.java
+++ net/minecraft/client/renderer/entity/CreeperRenderer.java
@@ -1,20 +1,21 @@
 package net.minecraft.client.renderer.entity;
 
 import com.mojang.blaze3d.vertex.PoseStack;
 import net.minecraft.client.model.CreeperModel;
+import net.minecraft.client.model.geom.ModelLayers;
 import net.minecraft.client.renderer.entity.layers.CreeperPowerLayer;
 import net.minecraft.resources.ResourceLocation;
 import net.minecraft.util.Mth;
 import net.minecraft.world.entity.monster.Creeper;
 
 public class CreeperRenderer extends MobRenderer<Creeper, CreeperModel<Creeper>> {
    private static final ResourceLocation CREEPER_LOCATION = new ResourceLocation("textures/entity/creeper/creeper.png");
 
-   public CreeperRenderer(EntityRenderDispatcher var1) {
-      super(var1, new CreeperModel(), 0.5F);
-      this.addLayer(new CreeperPowerLayer(this));
+   public CreeperRenderer(EntityRendererProvider.Context var1) {
+      super(var1, new CreeperModel(var1.bakeLayer(ModelLayers.CREEPER)), 0.5F);
+      this.addLayer(new CreeperPowerLayer(this, var1.getModelSet()));
    }
 
    protected void scale(Creeper var1, PoseStack var2, float var3) {
       float var4 = var1.getSwelling(var3);
       float var5 = 1.0F + Mth.sin(var4 * 100.0F) * var4 * 0.01F;
