--- net/minecraft/client/renderer/entity/DragonFireballRenderer.java
+++ net/minecraft/client/renderer/entity/DragonFireballRenderer.java
@@ -14,11 +14,11 @@
 
 public class DragonFireballRenderer extends EntityRenderer<DragonFireball> {
    private static final ResourceLocation TEXTURE_LOCATION = new ResourceLocation("textures/entity/enderdragon/dragon_fireball.png");
    private static final RenderType RENDER_TYPE;
 
-   public DragonFireballRenderer(EntityRenderDispatcher var1) {
+   public DragonFireballRenderer(EntityRendererProvider.Context var1) {
       super(var1);
    }
 
    protected int getBlockLightLevel(DragonFireball var1, BlockPos var2) {
       return 15;
