--- net/minecraft/client/renderer/entity/EntityRendererProvider.java
+++ net/minecraft/client/renderer/entity/EntityRendererProvider.java
@@ -1,0 +1,53 @@
+package net.minecraft.client.renderer.entity;
+
+import net.minecraft.client.gui.Font;
+import net.minecraft.client.model.geom.EntityModelSet;
+import net.minecraft.client.model.geom.ModelLayerLocation;
+import net.minecraft.client.model.geom.ModelPart;
+import net.minecraft.server.packs.resources.ResourceManager;
+import net.minecraft.world.entity.Entity;
+
+@FunctionalInterface
+public interface EntityRendererProvider<T extends Entity> {
+   EntityRenderer<T> create(EntityRendererProvider.Context var1);
+
+   public static class Context {
+      private final EntityRenderDispatcher entityRenderDispatcher;
+      private final ItemRenderer itemRenderer;
+      private final ResourceManager resourceManager;
+      private final EntityModelSet modelSet;
+      private final Font font;
+
+      public Context(EntityRenderDispatcher var1, ItemRenderer var2, ResourceManager var3, EntityModelSet var4, Font var5) {
+         this.entityRenderDispatcher = var1;
+         this.itemRenderer = var2;
+         this.resourceManager = var3;
+         this.modelSet = var4;
+         this.font = var5;
+      }
+
+      public EntityRenderDispatcher getEntityRenderDispatcher() {
+         return this.entityRenderDispatcher;
+      }
+
+      public ItemRenderer getItemRenderer() {
+         return this.itemRenderer;
+      }
+
+      public ResourceManager getResourceManager() {
+         return this.resourceManager;
+      }
+
+      public EntityModelSet getModelSet() {
+         return this.modelSet;
+      }
+
+      public ModelPart getLayer(ModelLayerLocation var1) {
+         return this.modelSet.getLayer(var1);
+      }
+
+      public Font getFont() {
+         return this.font;
+      }
+   }
+}
