--- net/minecraft/client/renderer/entity/EntityRenderer.java
+++ net/minecraft/client/renderer/entity/EntityRenderer.java
@@ -15,16 +15,19 @@
 import net.minecraft.world.level.LightLayer;
 import net.minecraft.world.phys.AABB;
 import net.minecraft.world.phys.Vec3;
 
 public abstract class EntityRenderer<T extends Entity> {
+   protected static final float NAMETAG_SCALE = 0.025F;
    protected final EntityRenderDispatcher entityRenderDispatcher;
+   private final Font font;
    protected float shadowRadius;
    protected float shadowStrength = 1.0F;
 
-   protected EntityRenderer(EntityRenderDispatcher var1) {
-      this.entityRenderDispatcher = var1;
+   protected EntityRenderer(EntityRendererProvider.Context var1) {
+      this.entityRenderDispatcher = var1.getEntityRenderDispatcher();
+      this.font = var1.getFont();
    }
 
    public final int getPackedLightCoords(T var1, float var2) {
       BlockPos var3 = new BlockPos(var1.getLightProbePosition(var2));
       return LightTexture.pack(this.getBlockLightLevel(var1, var3), this.getSkyLightLevel(var1, var3));
@@ -68,11 +71,11 @@
    }
 
    public abstract ResourceLocation getTextureLocation(T var1);
 
    public Font getFont() {
-      return this.entityRenderDispatcher.getFont();
+      return this.font;
    }
 
    protected void renderNameTag(T var1, Component var2, PoseStack var3, MultiBufferSource var4, int var5) {
       double var6 = this.entityRenderDispatcher.distanceToSqr(var1);
       if (var6 <= 4096.0D) {
@@ -94,10 +97,6 @@
          }
 
          var3.popPose();
       }
    }
-
-   public EntityRenderDispatcher getDispatcher() {
-      return this.entityRenderDispatcher;
-   }
 }
