--- net/minecraft/client/renderer/entity/ExperienceOrbRenderer.java
+++ net/minecraft/client/renderer/entity/ExperienceOrbRenderer.java
@@ -15,18 +15,18 @@
 
 public class ExperienceOrbRenderer extends EntityRenderer<ExperienceOrb> {
    private static final ResourceLocation EXPERIENCE_ORB_LOCATION = new ResourceLocation("textures/entity/experience_orb.png");
    private static final RenderType RENDER_TYPE;
 
-   public ExperienceOrbRenderer(EntityRenderDispatcher var1) {
+   public ExperienceOrbRenderer(EntityRendererProvider.Context var1) {
       super(var1);
       this.shadowRadius = 0.15F;
       this.shadowStrength = 0.75F;
    }
 
    protected int getBlockLightLevel(ExperienceOrb var1, BlockPos var2) {
-      return Mth.clamp(super.getBlockLightLevel(var1, var2) + 7, 0, 15);
+      return Mth.clamp((int)(super.getBlockLightLevel(var1, var2) + 7), (int)0, (int)15);
    }
 
    public void render(ExperienceOrb var1, float var2, float var3, PoseStack var4, MultiBufferSource var5, int var6) {
       var4.pushPose();
       int var7 = var1.getIcon();
