--- net/minecraft/client/renderer/entity/FallingBlockRenderer.java
+++ net/minecraft/client/renderer/entity/FallingBlockRenderer.java
@@ -14,11 +14,11 @@
 import net.minecraft.world.level.Level;
 import net.minecraft.world.level.block.RenderShape;
 import net.minecraft.world.level.block.state.BlockState;
 
 public class FallingBlockRenderer extends EntityRenderer<FallingBlockEntity> {
-   public FallingBlockRenderer(EntityRenderDispatcher var1) {
+   public FallingBlockRenderer(EntityRendererProvider.Context var1) {
       super(var1);
       this.shadowRadius = 0.5F;
    }
 
    public void render(FallingBlockEntity var1, float var2, float var3, PoseStack var4, MultiBufferSource var5, int var6) {
