--- net/minecraft/client/renderer/entity/FireworkEntityRenderer.java
+++ net/minecraft/client/renderer/entity/FireworkEntityRenderer.java
@@ -10,13 +10,13 @@
 import net.minecraft.world.entity.projectile.FireworkRocketEntity;
 
 public class FireworkEntityRenderer extends EntityRenderer<FireworkRocketEntity> {
    private final ItemRenderer itemRenderer;
 
-   public FireworkEntityRenderer(EntityRenderDispatcher var1, ItemRenderer var2) {
+   public FireworkEntityRenderer(EntityRendererProvider.Context var1) {
       super(var1);
-      this.itemRenderer = var2;
+      this.itemRenderer = var1.getItemRenderer();
    }
 
    public void render(FireworkRocketEntity var1, float var2, float var3, PoseStack var4, MultiBufferSource var5, int var6) {
       var4.pushPose();
       var4.mulPose(this.entityRenderDispatcher.cameraOrientation());
