--- net/minecraft/client/renderer/entity/FishingHookRenderer.java
+++ net/minecraft/client/renderer/entity/FishingHookRenderer.java
@@ -20,11 +20,11 @@
 
 public class FishingHookRenderer extends EntityRenderer<FishingHook> {
    private static final ResourceLocation TEXTURE_LOCATION = new ResourceLocation("textures/entity/fishing_hook.png");
    private static final RenderType RENDER_TYPE;
 
-   public FishingHookRenderer(EntityRenderDispatcher var1) {
+   public FishingHookRenderer(EntityRendererProvider.Context var1) {
       super(var1);
    }
 
    public void render(FishingHook var1, float var2, float var3, PoseStack var4, MultiBufferSource var5, int var6) {
       Player var7 = var1.getPlayerOwner();
@@ -43,11 +43,11 @@
          vertex(var11, var9, var10, var6, 1.0F, 1, 1, 0);
          vertex(var11, var9, var10, var6, 0.0F, 1, 0, 0);
          var4.popPose();
          int var12 = var7.getMainArm() == HumanoidArm.RIGHT ? 1 : -1;
          ItemStack var13 = var7.getMainHandItem();
-         if (var13.getItem() != Items.FISHING_ROD) {
+         if (!var13.is(Items.FISHING_ROD)) {
             var12 = -var12;
          }
 
          float var14 = var7.getAttackAnim(var3);
          float var15 = Mth.sin(Mth.sqrt(var14) * 3.1415927F);
