--- net/minecraft/client/renderer/entity/FoxRenderer.java
+++ net/minecraft/client/renderer/entity/FoxRenderer.java
@@ -1,10 +1,11 @@
 package net.minecraft.client.renderer.entity;
 
 import com.mojang.blaze3d.vertex.PoseStack;
 import com.mojang.math.Vector3f;
 import net.minecraft.client.model.FoxModel;
+import net.minecraft.client.model.geom.ModelLayers;
 import net.minecraft.client.renderer.entity.layers.FoxHeldItemLayer;
 import net.minecraft.resources.ResourceLocation;
 import net.minecraft.util.Mth;
 import net.minecraft.world.entity.animal.Fox;
 
@@ -12,19 +13,19 @@
    private static final ResourceLocation RED_FOX_TEXTURE = new ResourceLocation("textures/entity/fox/fox.png");
    private static final ResourceLocation RED_FOX_SLEEP_TEXTURE = new ResourceLocation("textures/entity/fox/fox_sleep.png");
    private static final ResourceLocation SNOW_FOX_TEXTURE = new ResourceLocation("textures/entity/fox/snow_fox.png");
    private static final ResourceLocation SNOW_FOX_SLEEP_TEXTURE = new ResourceLocation("textures/entity/fox/snow_fox_sleep.png");
 
-   public FoxRenderer(EntityRenderDispatcher var1) {
-      super(var1, new FoxModel(), 0.4F);
+   public FoxRenderer(EntityRendererProvider.Context var1) {
+      super(var1, new FoxModel(var1.bakeLayer(ModelLayers.FOX)), 0.4F);
       this.addLayer(new FoxHeldItemLayer(this));
    }
 
    protected void setupRotations(Fox var1, PoseStack var2, float var3, float var4, float var5) {
       super.setupRotations(var1, var2, var3, var4, var5);
       if (var1.isPouncing() || var1.isFaceplanted()) {
-         float var6 = -Mth.lerp(var5, var1.xRotO, var1.xRot);
+         float var6 = -Mth.lerp(var5, var1.xRotO, var1.getXRot());
          var2.mulPose(Vector3f.XP.rotationDegrees(var6));
       }
 
    }
 
