--- net/minecraft/client/renderer/entity/HoglinRenderer.java
+++ net/minecraft/client/renderer/entity/HoglinRenderer.java
@@ -1,16 +1,17 @@
 package net.minecraft.client.renderer.entity;
 
 import net.minecraft.client.model.HoglinModel;
+import net.minecraft.client.model.geom.ModelLayers;
 import net.minecraft.resources.ResourceLocation;
 import net.minecraft.world.entity.monster.hoglin.Hoglin;
 
 public class HoglinRenderer extends MobRenderer<Hoglin, HoglinModel<Hoglin>> {
    private static final ResourceLocation HOGLIN_LOCATION = new ResourceLocation("textures/entity/hoglin/hoglin.png");
 
-   public HoglinRenderer(EntityRenderDispatcher var1) {
-      super(var1, new HoglinModel(), 0.7F);
+   public HoglinRenderer(EntityRendererProvider.Context var1) {
+      super(var1, new HoglinModel(var1.getLayer(ModelLayers.HOGLIN)), 0.7F);
    }
 
    public ResourceLocation getTextureLocation(Hoglin var1) {
       return HOGLIN_LOCATION;
    }
