--- net/minecraft/client/renderer/entity/HorseRenderer.java
+++ net/minecraft/client/renderer/entity/HorseRenderer.java
@@ -2,10 +2,11 @@
 
 import com.google.common.collect.Maps;
 import java.util.Map;
 import net.minecraft.Util;
 import net.minecraft.client.model.HorseModel;
+import net.minecraft.client.model.geom.ModelLayers;
 import net.minecraft.client.renderer.entity.layers.HorseArmorLayer;
 import net.minecraft.client.renderer.entity.layers.HorseMarkingLayer;
 import net.minecraft.resources.ResourceLocation;
 import net.minecraft.world.entity.animal.horse.Horse;
 import net.minecraft.world.entity.animal.horse.Variant;
@@ -19,14 +20,14 @@
       var0.put(Variant.BLACK, new ResourceLocation("textures/entity/horse/horse_black.png"));
       var0.put(Variant.GRAY, new ResourceLocation("textures/entity/horse/horse_gray.png"));
       var0.put(Variant.DARKBROWN, new ResourceLocation("textures/entity/horse/horse_darkbrown.png"));
    });
 
-   public HorseRenderer(EntityRenderDispatcher var1) {
-      super(var1, new HorseModel(0.0F), 1.1F);
+   public HorseRenderer(EntityRendererProvider.Context var1) {
+      super(var1, new HorseModel(var1.bakeLayer(ModelLayers.HORSE)), 1.1F);
       this.addLayer(new HorseMarkingLayer(this));
-      this.addLayer(new HorseArmorLayer(this));
+      this.addLayer(new HorseArmorLayer(this, var1.getModelSet()));
    }
 
    public ResourceLocation getTextureLocation(Horse var1) {
       return (ResourceLocation)LOCATION_BY_VARIANT.get(var1.getVariant());
    }
