--- net/minecraft/client/renderer/entity/IllagerRenderer.java
+++ net/minecraft/client/renderer/entity/IllagerRenderer.java
@@ -4,13 +4,13 @@
 import net.minecraft.client.model.IllagerModel;
 import net.minecraft.client.renderer.entity.layers.CustomHeadLayer;
 import net.minecraft.world.entity.monster.AbstractIllager;
 
 public abstract class IllagerRenderer<T extends AbstractIllager> extends MobRenderer<T, IllagerModel<T>> {
-   protected IllagerRenderer(EntityRenderDispatcher var1, IllagerModel<T> var2, float var3) {
+   protected IllagerRenderer(EntityRendererProvider.Context var1, IllagerModel<T> var2, float var3) {
       super(var1, var2, var3);
-      this.addLayer(new CustomHeadLayer(this));
+      this.addLayer(new CustomHeadLayer(this, var1.getModelSet()));
    }
 
    protected void scale(T var1, PoseStack var2, float var3) {
       float var4 = 0.9375F;
       var2.scale(0.9375F, 0.9375F, 0.9375F);
