--- net/minecraft/client/renderer/entity/ItemEntityRenderer.java
+++ net/minecraft/client/renderer/entity/ItemEntityRenderer.java
@@ -17,13 +17,13 @@
 
 public class ItemEntityRenderer extends EntityRenderer<ItemEntity> {
    private final ItemRenderer itemRenderer;
    private final Random random = new Random();
 
-   public ItemEntityRenderer(EntityRenderDispatcher var1, ItemRenderer var2) {
+   public ItemEntityRenderer(EntityRendererProvider.Context var1) {
       super(var1);
-      this.itemRenderer = var2;
+      this.itemRenderer = var1.getItemRenderer();
       this.shadowRadius = 0.15F;
       this.shadowStrength = 0.75F;
    }
 
    private int getRenderAmount(ItemStack var1) {
@@ -44,11 +44,11 @@
    public void render(ItemEntity var1, float var2, float var3, PoseStack var4, MultiBufferSource var5, int var6) {
       var4.pushPose();
       ItemStack var7 = var1.getItem();
       int var8 = var7.isEmpty() ? 187 : Item.getId(var7.getItem()) + var7.getDamageValue();
       this.random.setSeed((long)var8);
-      BakedModel var9 = this.itemRenderer.getModel(var7, var1.level, (LivingEntity)null);
+      BakedModel var9 = this.itemRenderer.getModel(var7, var1.level, (LivingEntity)null, var1.getId());
       boolean var10 = var9.isGui3d();
       int var11 = this.getRenderAmount(var7);
       float var12 = 0.25F;
       float var13 = Mth.sin(((float)var1.getAge() + var3) / 10.0F + var1.bobOffs) * 0.1F + 0.1F;
       float var14 = var9.getTransforms().getTransform(ItemTransforms.TransformType.GROUND).scale.y();
