--- net/minecraft/client/renderer/entity/ItemEntityRenderer.java
+++ net/minecraft/client/renderer/entity/ItemEntityRenderer.java
@@ -14,16 +14,24 @@
 import net.minecraft.world.entity.item.ItemEntity;
 import net.minecraft.world.item.Item;
 import net.minecraft.world.item.ItemStack;
 
 public class ItemEntityRenderer extends EntityRenderer<ItemEntity> {
+   private static final float ITEM_BUNDLE_OFFSET_SCALE = 0.15F;
+   private static final int ITEM_COUNT_FOR_5_BUNDLE = 48;
+   private static final int ITEM_COUNT_FOR_4_BUNDLE = 32;
+   private static final int ITEM_COUNT_FOR_3_BUNDLE = 16;
+   private static final int ITEM_COUNT_FOR_2_BUNDLE = 1;
+   private static final float FLAT_ITEM_BUNDLE_OFFSET_X = 0.0F;
+   private static final float FLAT_ITEM_BUNDLE_OFFSET_Y = 0.0F;
+   private static final float FLAT_ITEM_BUNDLE_OFFSET_Z = 0.09375F;
    private final ItemRenderer itemRenderer;
    private final Random random = new Random();
 
-   public ItemEntityRenderer(EntityRenderDispatcher var1, ItemRenderer var2) {
+   public ItemEntityRenderer(EntityRendererProvider.Context var1) {
       super(var1);
-      this.itemRenderer = var2;
+      this.itemRenderer = var1.getItemRenderer();
       this.shadowRadius = 0.15F;
       this.shadowStrength = 0.75F;
    }
 
    private int getRenderAmount(ItemStack var1) {
@@ -44,11 +52,11 @@
    public void render(ItemEntity var1, float var2, float var3, PoseStack var4, MultiBufferSource var5, int var6) {
       var4.pushPose();
       ItemStack var7 = var1.getItem();
       int var8 = var7.isEmpty() ? 187 : Item.getId(var7.getItem()) + var7.getDamageValue();
       this.random.setSeed((long)var8);
-      BakedModel var9 = this.itemRenderer.getModel(var7, var1.level, (LivingEntity)null);
+      BakedModel var9 = this.itemRenderer.getModel(var7, var1.level, (LivingEntity)null, var1.getId());
       boolean var10 = var9.isGui3d();
       int var11 = this.getRenderAmount(var7);
       float var12 = 0.25F;
       float var13 = Mth.sin(((float)var1.getAge() + var3) / 10.0F + var1.bobOffs) * 0.1F + 0.1F;
       float var14 = var9.getTransforms().getTransform(ItemTransforms.TransformType.GROUND).scale.y();
