--- net/minecraft/client/renderer/entity/LightningBoltRenderer.java
+++ net/minecraft/client/renderer/entity/LightningBoltRenderer.java
@@ -9,11 +9,11 @@
 import net.minecraft.client.renderer.texture.TextureAtlas;
 import net.minecraft.resources.ResourceLocation;
 import net.minecraft.world.entity.LightningBolt;
 
 public class LightningBoltRenderer extends EntityRenderer<LightningBolt> {
-   public LightningBoltRenderer(EntityRenderDispatcher var1) {
+   public LightningBoltRenderer(EntityRendererProvider.Context var1) {
       super(var1);
    }
 
    public void render(LightningBolt var1, float var2, float var3, PoseStack var4, MultiBufferSource var5, int var6) {
       float[] var7 = new float[8];
