--- net/minecraft/client/renderer/entity/LivingEntityRenderer.java
+++ net/minecraft/client/renderer/entity/LivingEntityRenderer.java
@@ -29,11 +29,11 @@
 public abstract class LivingEntityRenderer<T extends LivingEntity, M extends EntityModel<T>> extends EntityRenderer<T> implements RenderLayerParent<T, M> {
    private static final Logger LOGGER = LogManager.getLogger();
    protected M model;
    protected final List<RenderLayer<T, M>> layers = Lists.newArrayList();
 
-   public LivingEntityRenderer(EntityRenderDispatcher var1, M var2, float var3) {
+   public LivingEntityRenderer(EntityRendererProvider.Context var1, M var2, float var3) {
       super(var1);
       this.model = var2;
       this.shadowRadius = var3;
    }
 
