--- net/minecraft/client/renderer/entity/LivingEntityRenderer.java
+++ net/minecraft/client/renderer/entity/LivingEntityRenderer.java
@@ -26,14 +26,15 @@
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
 
 public abstract class LivingEntityRenderer<T extends LivingEntity, M extends EntityModel<T>> extends EntityRenderer<T> implements RenderLayerParent<T, M> {
    private static final Logger LOGGER = LogManager.getLogger();
+   private static final float EYE_BED_OFFSET = 0.1F;
    protected M model;
    protected final List<RenderLayer<T, M>> layers = Lists.newArrayList();
 
-   public LivingEntityRenderer(EntityRenderDispatcher var1, M var2, float var3) {
+   public LivingEntityRenderer(EntityRendererProvider.Context var1, M var2, float var3) {
       super(var1);
       this.model = var2;
       this.shadowRadius = var3;
    }
 
@@ -164,11 +165,11 @@
          return 0.0F;
       }
    }
 
    protected boolean isShaking(T var1) {
-      return false;
+      return var1.isFullyFrozen();
    }
 
    protected void setupRotations(T var1, PoseStack var2, float var3, float var4, float var5) {
       if (this.isShaking(var1)) {
          var4 += (float)(Math.cos((double)var1.tickCount * 3.25D) * 3.141592653589793D * 0.4000000059604645D);
