--- net/minecraft/client/renderer/entity/PillagerRenderer.java
+++ net/minecraft/client/renderer/entity/PillagerRenderer.java
@@ -1,17 +1,18 @@
 package net.minecraft.client.renderer.entity;
 
 import net.minecraft.client.model.IllagerModel;
+import net.minecraft.client.model.geom.ModelLayers;
 import net.minecraft.client.renderer.entity.layers.ItemInHandLayer;
 import net.minecraft.resources.ResourceLocation;
 import net.minecraft.world.entity.monster.Pillager;
 
 public class PillagerRenderer extends IllagerRenderer<Pillager> {
    private static final ResourceLocation PILLAGER = new ResourceLocation("textures/entity/illager/pillager.png");
 
-   public PillagerRenderer(EntityRenderDispatcher var1) {
-      super(var1, new IllagerModel(0.0F, 0.0F, 64, 64), 0.5F);
+   public PillagerRenderer(EntityRendererProvider.Context var1) {
+      super(var1, new IllagerModel(var1.bakeLayer(ModelLayers.PILLAGER)), 0.5F);
       this.addLayer(new ItemInHandLayer(this));
    }
 
    public ResourceLocation getTextureLocation(Pillager var1) {
       return PILLAGER;
