--- net/minecraft/client/renderer/entity/RabbitRenderer.java
+++ net/minecraft/client/renderer/entity/RabbitRenderer.java
@@ -1,9 +1,10 @@
 package net.minecraft.client.renderer.entity;
 
 import net.minecraft.ChatFormatting;
 import net.minecraft.client.model.RabbitModel;
+import net.minecraft.client.model.geom.ModelLayers;
 import net.minecraft.resources.ResourceLocation;
 import net.minecraft.world.entity.animal.Rabbit;
 
 public class RabbitRenderer extends MobRenderer<Rabbit, RabbitModel<Rabbit>> {
    private static final ResourceLocation RABBIT_BROWN_LOCATION = new ResourceLocation("textures/entity/rabbit/brown.png");
@@ -13,12 +14,12 @@
    private static final ResourceLocation RABBIT_SALT_LOCATION = new ResourceLocation("textures/entity/rabbit/salt.png");
    private static final ResourceLocation RABBIT_WHITE_SPLOTCHED_LOCATION = new ResourceLocation("textures/entity/rabbit/white_splotched.png");
    private static final ResourceLocation RABBIT_TOAST_LOCATION = new ResourceLocation("textures/entity/rabbit/toast.png");
    private static final ResourceLocation RABBIT_EVIL_LOCATION = new ResourceLocation("textures/entity/rabbit/caerbannog.png");
 
-   public RabbitRenderer(EntityRenderDispatcher var1) {
-      super(var1, new RabbitModel(), 0.3F);
+   public RabbitRenderer(EntityRendererProvider.Context var1) {
+      super(var1, new RabbitModel(var1.bakeLayer(ModelLayers.RABBIT)), 0.3F);
    }
 
    public ResourceLocation getTextureLocation(Rabbit var1) {
       String var2 = ChatFormatting.stripFormatting(var1.getName().getString());
       if (var2 != null && "Toast".equals(var2)) {
