--- net/minecraft/client/renderer/entity/ShulkerRenderer.java
+++ net/minecraft/client/renderer/entity/ShulkerRenderer.java
@@ -1,9 +1,10 @@
 package net.minecraft.client.renderer.entity;
 
 import com.mojang.blaze3d.vertex.PoseStack;
 import net.minecraft.client.model.ShulkerModel;
+import net.minecraft.client.model.geom.ModelLayers;
 import net.minecraft.client.renderer.Sheets;
 import net.minecraft.client.renderer.culling.Frustum;
 import net.minecraft.client.renderer.entity.layers.ShulkerHeadLayer;
 import net.minecraft.core.BlockPos;
 import net.minecraft.resources.ResourceLocation;
@@ -14,12 +15,12 @@
 
 public class ShulkerRenderer extends MobRenderer<Shulker, ShulkerModel<Shulker>> {
    public static final ResourceLocation DEFAULT_TEXTURE_LOCATION;
    public static final ResourceLocation[] TEXTURE_LOCATION;
 
-   public ShulkerRenderer(EntityRenderDispatcher var1) {
-      super(var1, new ShulkerModel(), 0.0F);
+   public ShulkerRenderer(EntityRendererProvider.Context var1) {
+      super(var1, new ShulkerModel(var1.getLayer(ModelLayers.SHULKER)), 0.0F);
       this.addLayer(new ShulkerHeadLayer(this));
    }
 
    public Vec3 getRenderOffset(Shulker var1, float var2) {
       int var3 = var1.getClientSideTeleportInterpolation();
