--- net/minecraft/client/renderer/entity/ShulkerBulletRenderer.java
+++ net/minecraft/client/renderer/entity/ShulkerBulletRenderer.java
@@ -2,10 +2,11 @@
 
 import com.mojang.blaze3d.vertex.PoseStack;
 import com.mojang.blaze3d.vertex.VertexConsumer;
 import com.mojang.math.Vector3f;
 import net.minecraft.client.model.ShulkerBulletModel;
+import net.minecraft.client.model.geom.ModelLayers;
 import net.minecraft.client.renderer.MultiBufferSource;
 import net.minecraft.client.renderer.RenderType;
 import net.minecraft.client.renderer.texture.OverlayTexture;
 import net.minecraft.core.BlockPos;
 import net.minecraft.resources.ResourceLocation;
@@ -13,24 +14,25 @@
 import net.minecraft.world.entity.projectile.ShulkerBullet;
 
 public class ShulkerBulletRenderer extends EntityRenderer<ShulkerBullet> {
    private static final ResourceLocation TEXTURE_LOCATION = new ResourceLocation("textures/entity/shulker/spark.png");
    private static final RenderType RENDER_TYPE;
-   private final ShulkerBulletModel<ShulkerBullet> model = new ShulkerBulletModel();
+   private final ShulkerBulletModel<ShulkerBullet> model;
 
-   public ShulkerBulletRenderer(EntityRenderDispatcher var1) {
+   public ShulkerBulletRenderer(EntityRendererProvider.Context var1) {
       super(var1);
+      this.model = new ShulkerBulletModel(var1.bakeLayer(ModelLayers.SHULKER_BULLET));
    }
 
    protected int getBlockLightLevel(ShulkerBullet var1, BlockPos var2) {
       return 15;
    }
 
    public void render(ShulkerBullet var1, float var2, float var3, PoseStack var4, MultiBufferSource var5, int var6) {
       var4.pushPose();
-      float var7 = Mth.rotlerp(var1.yRotO, var1.yRot, var3);
-      float var8 = Mth.lerp(var3, var1.xRotO, var1.xRot);
+      float var7 = Mth.rotlerp(var1.yRotO, var1.getYRot(), var3);
+      float var8 = Mth.lerp(var3, var1.xRotO, var1.getXRot());
       float var9 = (float)var1.tickCount + var3;
       var4.translate(0.0D, 0.15000000596046448D, 0.0D);
       var4.mulPose(Vector3f.YP.rotationDegrees(Mth.sin(var9 * 0.1F) * 180.0F));
       var4.mulPose(Vector3f.XP.rotationDegrees(Mth.cos(var9 * 0.1F) * 180.0F));
       var4.mulPose(Vector3f.ZP.rotationDegrees(Mth.sin(var9 * 0.15F) * 360.0F));
