--- net/minecraft/client/renderer/entity/SpectralArrowRenderer.java
+++ net/minecraft/client/renderer/entity/SpectralArrowRenderer.java
@@ -4,11 +4,11 @@
 import net.minecraft.world.entity.projectile.SpectralArrow;
 
 public class SpectralArrowRenderer extends ArrowRenderer<SpectralArrow> {
    public static final ResourceLocation SPECTRAL_ARROW_LOCATION = new ResourceLocation("textures/entity/projectiles/spectral_arrow.png");
 
-   public SpectralArrowRenderer(EntityRenderDispatcher var1) {
+   public SpectralArrowRenderer(EntityRendererProvider.Context var1) {
       super(var1);
    }
 
    public ResourceLocation getTextureLocation(SpectralArrow var1) {
       return SPECTRAL_ARROW_LOCATION;
