--- net/minecraft/client/renderer/entity/ThrownItemRenderer.java
+++ net/minecraft/client/renderer/entity/ThrownItemRenderer.java
@@ -10,23 +10,24 @@
 import net.minecraft.resources.ResourceLocation;
 import net.minecraft.world.entity.Entity;
 import net.minecraft.world.entity.projectile.ItemSupplier;
 
 public class ThrownItemRenderer<T extends Entity & ItemSupplier> extends EntityRenderer<T> {
+   private static final float MIN_CAMERA_DISTANCE_SQUARED = 12.25F;
    private final ItemRenderer itemRenderer;
    private final float scale;
    private final boolean fullBright;
 
-   public ThrownItemRenderer(EntityRenderDispatcher var1, ItemRenderer var2, float var3, boolean var4) {
+   public ThrownItemRenderer(EntityRendererProvider.Context var1, float var2, boolean var3) {
       super(var1);
-      this.itemRenderer = var2;
-      this.scale = var3;
-      this.fullBright = var4;
+      this.itemRenderer = var1.getItemRenderer();
+      this.scale = var2;
+      this.fullBright = var3;
    }
 
-   public ThrownItemRenderer(EntityRenderDispatcher var1, ItemRenderer var2) {
-      this(var1, var2, 1.0F, false);
+   public ThrownItemRenderer(EntityRendererProvider.Context var1) {
+      this(var1, 1.0F, false);
    }
 
    protected int getBlockLightLevel(T var1, BlockPos var2) {
       return this.fullBright ? 15 : super.getBlockLightLevel(var1, var2);
    }
@@ -35,11 +36,11 @@
       if (var1.tickCount >= 2 || this.entityRenderDispatcher.camera.getEntity().distanceToSqr(var1) >= 12.25D) {
          var4.pushPose();
          var4.scale(this.scale, this.scale, this.scale);
          var4.mulPose(this.entityRenderDispatcher.cameraOrientation());
          var4.mulPose(Vector3f.YP.rotationDegrees(180.0F));
-         this.itemRenderer.renderStatic(((ItemSupplier)var1).getItem(), ItemTransforms.TransformType.GROUND, var6, OverlayTexture.NO_OVERLAY, var4, var5);
+         this.itemRenderer.renderStatic(((ItemSupplier)var1).getItem(), ItemTransforms.TransformType.GROUND, var6, OverlayTexture.NO_OVERLAY, var4, var5, var1.getId());
          var4.popPose();
          super.render(var1, var2, var3, var4, var5, var6);
       }
    }
 
