--- net/minecraft/client/renderer/entity/TippableArrowRenderer.java
+++ net/minecraft/client/renderer/entity/TippableArrowRenderer.java
@@ -5,11 +5,11 @@
 
 public class TippableArrowRenderer extends ArrowRenderer<Arrow> {
    public static final ResourceLocation NORMAL_ARROW_LOCATION = new ResourceLocation("textures/entity/projectiles/arrow.png");
    public static final ResourceLocation TIPPED_ARROW_LOCATION = new ResourceLocation("textures/entity/projectiles/tipped_arrow.png");
 
-   public TippableArrowRenderer(EntityRenderDispatcher var1) {
+   public TippableArrowRenderer(EntityRendererProvider.Context var1) {
       super(var1);
    }
 
    public ResourceLocation getTextureLocation(Arrow var1) {
       return var1.getColor() > 0 ? TIPPED_ARROW_LOCATION : NORMAL_ARROW_LOCATION;
