--- net/minecraft/client/renderer/entity/TntMinecartRenderer.java
+++ net/minecraft/client/renderer/entity/TntMinecartRenderer.java
@@ -1,18 +1,19 @@
 package net.minecraft.client.renderer.entity;
 
 import com.mojang.blaze3d.vertex.PoseStack;
 import net.minecraft.client.Minecraft;
+import net.minecraft.client.model.geom.ModelLayers;
 import net.minecraft.client.renderer.MultiBufferSource;
 import net.minecraft.client.renderer.texture.OverlayTexture;
 import net.minecraft.util.Mth;
 import net.minecraft.world.entity.vehicle.MinecartTNT;
 import net.minecraft.world.level.block.state.BlockState;
 
 public class TntMinecartRenderer extends MinecartRenderer<MinecartTNT> {
-   public TntMinecartRenderer(EntityRenderDispatcher var1) {
-      super(var1);
+   public TntMinecartRenderer(EntityRendererProvider.Context var1) {
+      super(var1, ModelLayers.TNT_MINECART);
    }
 
    protected void renderMinecartContents(MinecartTNT var1, float var2, BlockState var3, PoseStack var4, MultiBufferSource var5, int var6) {
       int var7 = var1.getFuse();
       if (var7 > -1 && (float)var7 - var2 + 1.0F < 10.0F) {
