--- net/minecraft/client/renderer/entity/TntRenderer.java
+++ net/minecraft/client/renderer/entity/TntRenderer.java
@@ -8,11 +8,11 @@
 import net.minecraft.util.Mth;
 import net.minecraft.world.entity.item.PrimedTnt;
 import net.minecraft.world.level.block.Blocks;
 
 public class TntRenderer extends EntityRenderer<PrimedTnt> {
-   public TntRenderer(EntityRenderDispatcher var1) {
+   public TntRenderer(EntityRendererProvider.Context var1) {
       super(var1);
       this.shadowRadius = 0.5F;
    }
 
    public void render(PrimedTnt var1, float var2, float var3, PoseStack var4, MultiBufferSource var5, int var6) {
