--- net/minecraft/client/gui/components/events/ContainerEventHandler.java
+++ net/minecraft/client/gui/components/events/ContainerEventHandler.java
@@ -1,10 +1,11 @@
 package net.minecraft.client.gui.components.events;
 
 import java.util.Iterator;
 import java.util.List;
 import java.util.ListIterator;
+import java.util.Objects;
 import java.util.Optional;
 import java.util.function.BooleanSupplier;
 import java.util.function.Supplier;
 import javax.annotation.Nullable;
 
@@ -109,13 +110,31 @@
          } else {
             var5 = var4.size();
          }
 
          ListIterator var7 = var4.listIterator(var5);
-         BooleanSupplier var8 = var1 ? var7::hasNext : var7::hasPrevious;
-         Supplier var9 = var1 ? var7::next : var7::previous;
+         BooleanSupplier var10000;
+         if (var1) {
+            Objects.requireNonNull(var7);
+            var10000 = var7::hasNext;
+         } else {
+            Objects.requireNonNull(var7);
+            var10000 = var7::hasPrevious;
+         }
 
+         BooleanSupplier var8 = var10000;
+         Supplier var11;
+         if (var1) {
+            Objects.requireNonNull(var7);
+            var11 = var7::next;
+         } else {
+            Objects.requireNonNull(var7);
+            var11 = var7::previous;
+         }
+
+         Supplier var9 = var11;
+
          GuiEventListener var10;
          do {
             if (!var8.getAsBoolean()) {
                this.setFocused((GuiEventListener)null);
                return false;
