--- net/minecraft/network/protocol/handshake/ClientIntentionPacket.java
+++ net/minecraft/network/protocol/handshake/ClientIntentionPacket.java
@@ -1,29 +1,25 @@
 package net.minecraft.network.protocol.handshake;
 
-import java.io.IOException;
 import net.minecraft.network.ConnectionProtocol;
 import net.minecraft.network.FriendlyByteBuf;
 import net.minecraft.network.protocol.Packet;
 
 public class ClientIntentionPacket implements Packet<ServerHandshakePacketListener> {
-   private int protocolVersion;
-   private String hostName;
-   private int port;
-   private ConnectionProtocol intention;
+   private final int protocolVersion;
+   private final String hostName;
+   private final int port;
+   private final ConnectionProtocol intention;
 
-   public ClientIntentionPacket() {
-   }
-
-   public void read(FriendlyByteBuf var1) throws IOException {
+   public ClientIntentionPacket(FriendlyByteBuf var1) {
       this.protocolVersion = var1.readVarInt();
       this.hostName = var1.readUtf(255);
       this.port = var1.readUnsignedShort();
       this.intention = ConnectionProtocol.getById(var1.readVarInt());
    }
 
-   public void write(FriendlyByteBuf var1) throws IOException {
+   public void write(FriendlyByteBuf var1) {
       var1.writeVarInt(this.protocolVersion);
       var1.writeUtf(this.hostName);
       var1.writeShort(this.port);
       var1.writeVarInt(this.intention.getId());
    }
