--- net/minecraft/client/server/IntegratedServer.java
+++ net/minecraft/client/server/IntegratedServer.java
@@ -47,11 +47,11 @@
       this.setPlayerList(new IntegratedPlayerList(this, this.registryHolder, this.playerDataStorage));
       this.minecraft = var2;
    }
 
    public boolean initServer() {
-      LOGGER.info("Starting integrated minecraft server version " + SharedConstants.getCurrentVersion().getName());
+      LOGGER.info("Starting integrated minecraft server version {}", SharedConstants.getCurrentVersion().getName());
       this.setUsesAuthentication(true);
       this.setPvpAllowed(true);
       this.setFlightAllowed(true);
       this.initializeKeyPair();
       this.loadLevel();
