--- net/minecraft/client/server/LanServerDetection.java
+++ net/minecraft/client/server/LanServerDetection.java
@@ -14,13 +14,16 @@
 import net.minecraft.DefaultUncaughtExceptionHandler;
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
 
 public class LanServerDetection {
-   private static final AtomicInteger UNIQUE_THREAD_ID = new AtomicInteger(0);
-   private static final Logger LOGGER = LogManager.getLogger();
+   static final AtomicInteger UNIQUE_THREAD_ID = new AtomicInteger(0);
+   static final Logger LOGGER = LogManager.getLogger();
 
+   public LanServerDetection() {
+   }
+
    public static class LanServerDetector extends Thread {
       private final LanServerDetection.LanServerList serverList;
       private final InetAddress pingGroup;
       private final MulticastSocket socket;
 
@@ -85,11 +88,12 @@
 
       public synchronized void addServer(String var1, InetAddress var2) {
          String var3 = LanServerPinger.parseMotd(var1);
          String var4 = LanServerPinger.parseAddress(var1);
          if (var4 != null) {
-            var4 = var2.getHostAddress() + ":" + var4;
+            String var10000 = var2.getHostAddress();
+            var4 = var10000 + ":" + var4;
             boolean var5 = false;
             Iterator var6 = this.servers.iterator();
 
             while(var6.hasNext()) {
                LanServer var7 = (LanServer)var6.next();
