--- net/minecraft/client/server/LanServerDetection.java
+++ net/minecraft/client/server/LanServerDetection.java
@@ -17,10 +17,13 @@
 
 public class LanServerDetection {
    private static final AtomicInteger UNIQUE_THREAD_ID = new AtomicInteger(0);
    private static final Logger LOGGER = LogManager.getLogger();
 
+   public LanServerDetection() {
+   }
+
    public static class LanServerDetector extends Thread {
       private final LanServerDetection.LanServerList serverList;
       private final InetAddress pingGroup;
       private final MulticastSocket socket;
 
