--- net/minecraft/client/gui/components/AbstractWidget.java
+++ net/minecraft/client/gui/components/AbstractWidget.java
@@ -90,13 +90,17 @@
       }
 
    }
 
    protected MutableComponent createNarrationMessage() {
-      return new TranslatableComponent("gui.narrate.button", new Object[]{this.getMessage()});
+      return wrapDefaultNarrationMessage(this.getMessage());
    }
 
+   public static MutableComponent wrapDefaultNarrationMessage(Component var0) {
+      return new TranslatableComponent("gui.narrate.button", new Object[]{var0});
+   }
+
    public void renderButton(PoseStack var1, int var2, int var3, float var4) {
       Minecraft var5 = Minecraft.getInstance();
       Font var6 = var5.font;
       var5.getTextureManager().bind(WIDGETS_LOCATION);
       RenderSystem.color4f(1.0F, 1.0F, 1.0F, this.alpha);
