--- net/minecraft/client/gui/components/AbstractSliderButton.java
+++ net/minecraft/client/gui/components/AbstractSliderButton.java
@@ -24,12 +24,12 @@
    protected MutableComponent createNarrationMessage() {
       return new TranslatableComponent("gui.narrate.slider", new Object[]{this.getMessage()});
    }
 
    protected void renderBg(PoseStack var1, Minecraft var2, int var3, int var4) {
-      var2.getTextureManager().bind(WIDGETS_LOCATION);
-      RenderSystem.color4f(1.0F, 1.0F, 1.0F, 1.0F);
+      RenderSystem.setShaderTexture(0, WIDGETS_LOCATION);
+      RenderSystem.setShaderColor(1.0F, 1.0F, 1.0F, 1.0F);
       int var5 = (this.isHovered() ? 2 : 1) * 20;
       this.blit(var1, this.x + (int)(this.value * (double)(this.width - 8)), this.y, 0, 46 + var5, 4, 20);
       this.blit(var1, this.x + (int)(this.value * (double)(this.width - 8)) + 4, this.y, 196, 46 + var5, 4, 20);
    }
 
