--- net/minecraft/client/gui/components/AbstractWidget.java
+++ net/minecraft/client/gui/components/AbstractWidget.java
@@ -7,10 +7,11 @@
 import net.minecraft.client.Minecraft;
 import net.minecraft.client.gui.Font;
 import net.minecraft.client.gui.GuiComponent;
 import net.minecraft.client.gui.chat.NarratorChatListener;
 import net.minecraft.client.gui.components.events.GuiEventListener;
+import net.minecraft.client.renderer.GameRenderer;
 import net.minecraft.client.resources.sounds.SimpleSoundInstance;
 import net.minecraft.client.sounds.SoundManager;
 import net.minecraft.network.chat.Component;
 import net.minecraft.network.chat.MutableComponent;
 import net.minecraft.network.chat.TranslatableComponent;
@@ -90,18 +91,23 @@
       }
 
    }
 
    protected MutableComponent createNarrationMessage() {
-      return new TranslatableComponent("gui.narrate.button", new Object[]{this.getMessage()});
+      return wrapDefaultNarrationMessage(this.getMessage());
    }
 
+   public static MutableComponent wrapDefaultNarrationMessage(Component var0) {
+      return new TranslatableComponent("gui.narrate.button", new Object[]{var0});
+   }
+
    public void renderButton(PoseStack var1, int var2, int var3, float var4) {
       Minecraft var5 = Minecraft.getInstance();
       Font var6 = var5.font;
-      var5.getTextureManager().bind(WIDGETS_LOCATION);
-      RenderSystem.color4f(1.0F, 1.0F, 1.0F, this.alpha);
+      RenderSystem.setShader(GameRenderer::getPositionTexShader);
+      RenderSystem.setShaderTexture(0, WIDGETS_LOCATION);
+      RenderSystem.setShaderColor(1.0F, 1.0F, 1.0F, this.alpha);
       int var7 = this.getYImage(this.isHovered());
       RenderSystem.enableBlend();
       RenderSystem.defaultBlendFunc();
       RenderSystem.enableDepthTest();
       this.blit(var1, this.x, this.y, 0, 46 + var7 * 20, this.width / 2, this.height);
