--- net/minecraft/client/gui/components/AbstractSelectionList.java
+++ net/minecraft/client/gui/components/AbstractSelectionList.java
@@ -5,10 +5,11 @@
 import com.mojang.blaze3d.systems.RenderSystem;
 import com.mojang.blaze3d.vertex.BufferBuilder;
 import com.mojang.blaze3d.vertex.DefaultVertexFormat;
 import com.mojang.blaze3d.vertex.PoseStack;
 import com.mojang.blaze3d.vertex.Tesselator;
+import com.mojang.blaze3d.vertex.VertexFormat;
 import java.util.AbstractList;
 import java.util.Collection;
 import java.util.List;
 import java.util.Objects;
 import java.util.function.Predicate;
@@ -168,11 +169,11 @@
       BufferBuilder var8 = var7.getBuilder();
       if (this.renderBackground) {
          this.minecraft.getTextureManager().bind(GuiComponent.BACKGROUND_LOCATION);
          RenderSystem.color4f(1.0F, 1.0F, 1.0F, 1.0F);
          float var9 = 32.0F;
-         var8.begin(7, DefaultVertexFormat.POSITION_TEX_COLOR);
+         var8.begin(VertexFormat.Mode.QUADS, DefaultVertexFormat.POSITION_TEX_COLOR);
          var8.vertex((double)this.x0, (double)this.y1, 0.0D).uv((float)this.x0 / 32.0F, (float)(this.y1 + (int)this.getScrollAmount()) / 32.0F).color(32, 32, 32, 255).endVertex();
          var8.vertex((double)this.x1, (double)this.y1, 0.0D).uv((float)this.x1 / 32.0F, (float)(this.y1 + (int)this.getScrollAmount()) / 32.0F).color(32, 32, 32, 255).endVertex();
          var8.vertex((double)this.x1, (double)this.y0, 0.0D).uv((float)this.x1 / 32.0F, (float)(this.y0 + (int)this.getScrollAmount()) / 32.0F).color(32, 32, 32, 255).endVertex();
          var8.vertex((double)this.x0, (double)this.y0, 0.0D).uv((float)this.x0 / 32.0F, (float)(this.y0 + (int)this.getScrollAmount()) / 32.0F).color(32, 32, 32, 255).endVertex();
          var7.end();
@@ -189,11 +190,11 @@
          this.minecraft.getTextureManager().bind(GuiComponent.BACKGROUND_LOCATION);
          RenderSystem.enableDepthTest();
          RenderSystem.depthFunc(519);
          float var11 = 32.0F;
          boolean var12 = true;
-         var8.begin(7, DefaultVertexFormat.POSITION_TEX_COLOR);
+         var8.begin(VertexFormat.Mode.QUADS, DefaultVertexFormat.POSITION_TEX_COLOR);
          var8.vertex((double)this.x0, (double)this.y0, -100.0D).uv(0.0F, (float)this.y0 / 32.0F).color(64, 64, 64, 255).endVertex();
          var8.vertex((double)(this.x0 + this.width), (double)this.y0, -100.0D).uv((float)this.width / 32.0F, (float)this.y0 / 32.0F).color(64, 64, 64, 255).endVertex();
          var8.vertex((double)(this.x0 + this.width), 0.0D, -100.0D).uv((float)this.width / 32.0F, 0.0F).color(64, 64, 64, 255).endVertex();
          var8.vertex((double)this.x0, 0.0D, -100.0D).uv(0.0F, 0.0F).color(64, 64, 64, 255).endVertex();
          var8.vertex((double)this.x0, (double)this.height, -100.0D).uv(0.0F, (float)this.height / 32.0F).color(64, 64, 64, 255).endVertex();
@@ -207,11 +208,11 @@
          RenderSystem.blendFuncSeparate(GlStateManager.SourceFactor.SRC_ALPHA, GlStateManager.DestFactor.ONE_MINUS_SRC_ALPHA, GlStateManager.SourceFactor.ZERO, GlStateManager.DestFactor.ONE);
          RenderSystem.disableAlphaTest();
          RenderSystem.shadeModel(7425);
          RenderSystem.disableTexture();
          boolean var13 = true;
-         var8.begin(7, DefaultVertexFormat.POSITION_TEX_COLOR);
+         var8.begin(VertexFormat.Mode.QUADS, DefaultVertexFormat.POSITION_TEX_COLOR);
          var8.vertex((double)this.x0, (double)(this.y0 + 4), 0.0D).uv(0.0F, 1.0F).color(0, 0, 0, 0).endVertex();
          var8.vertex((double)this.x1, (double)(this.y0 + 4), 0.0D).uv(1.0F, 1.0F).color(0, 0, 0, 0).endVertex();
          var8.vertex((double)this.x1, (double)this.y0, 0.0D).uv(1.0F, 0.0F).color(0, 0, 0, 255).endVertex();
          var8.vertex((double)this.x0, (double)this.y0, 0.0D).uv(0.0F, 0.0F).color(0, 0, 0, 255).endVertex();
          var8.vertex((double)this.x0, (double)this.y1, 0.0D).uv(0.0F, 1.0F).color(0, 0, 0, 255).endVertex();
@@ -229,11 +230,11 @@
          int var17 = (int)this.getScrollAmount() * (this.y1 - this.y0 - var16) / var15 + this.y0;
          if (var17 < this.y0) {
             var17 = this.y0;
          }
 
-         var8.begin(7, DefaultVertexFormat.POSITION_TEX_COLOR);
+         var8.begin(VertexFormat.Mode.QUADS, DefaultVertexFormat.POSITION_TEX_COLOR);
          var8.vertex((double)var5, (double)this.y1, 0.0D).uv(0.0F, 1.0F).color(0, 0, 0, 255).endVertex();
          var8.vertex((double)var6, (double)this.y1, 0.0D).uv(1.0F, 1.0F).color(0, 0, 0, 255).endVertex();
          var8.vertex((double)var6, (double)this.y0, 0.0D).uv(1.0F, 0.0F).color(0, 0, 0, 255).endVertex();
          var8.vertex((double)var5, (double)this.y0, 0.0D).uv(0.0F, 0.0F).color(0, 0, 0, 255).endVertex();
          var8.vertex((double)var5, (double)(var17 + var16), 0.0D).uv(0.0F, 1.0F).color(128, 128, 128, 255).endVertex();
@@ -427,18 +428,18 @@
                var17 = this.x0 + this.width / 2 - var16 / 2;
                int var18 = this.x0 + this.width / 2 + var16 / 2;
                RenderSystem.disableTexture();
                float var19 = this.isFocused() ? 1.0F : 0.5F;
                RenderSystem.color4f(var19, var19, var19, 1.0F);
-               var9.begin(7, DefaultVertexFormat.POSITION);
+               var9.begin(VertexFormat.Mode.QUADS, DefaultVertexFormat.POSITION);
                var9.vertex((double)var17, (double)(var13 + var14 + 2), 0.0D).endVertex();
                var9.vertex((double)var18, (double)(var13 + var14 + 2), 0.0D).endVertex();
                var9.vertex((double)var18, (double)(var13 - 2), 0.0D).endVertex();
                var9.vertex((double)var17, (double)(var13 - 2), 0.0D).endVertex();
                var8.end();
                RenderSystem.color4f(0.0F, 0.0F, 0.0F, 1.0F);
-               var9.begin(7, DefaultVertexFormat.POSITION);
+               var9.begin(VertexFormat.Mode.QUADS, DefaultVertexFormat.POSITION);
                var9.vertex((double)(var17 + 1), (double)(var13 + var14 + 1), 0.0D).endVertex();
                var9.vertex((double)(var18 - 1), (double)(var13 + var14 + 1), 0.0D).endVertex();
                var9.vertex((double)(var18 - 1), (double)(var13 - 1), 0.0D).endVertex();
                var9.vertex((double)(var17 + 1), (double)(var13 - 1), 0.0D).endVertex();
                var8.end();
