--- net/minecraft/client/gui/components/AbstractSelectionList.java
+++ net/minecraft/client/gui/components/AbstractSelectionList.java
@@ -5,23 +5,30 @@
 import com.mojang.blaze3d.systems.RenderSystem;
 import com.mojang.blaze3d.vertex.BufferBuilder;
 import com.mojang.blaze3d.vertex.DefaultVertexFormat;
 import com.mojang.blaze3d.vertex.PoseStack;
 import com.mojang.blaze3d.vertex.Tesselator;
+import com.mojang.blaze3d.vertex.VertexFormat;
 import java.util.AbstractList;
 import java.util.Collection;
 import java.util.List;
 import java.util.Objects;
 import java.util.function.Predicate;
 import javax.annotation.Nullable;
 import net.minecraft.client.Minecraft;
 import net.minecraft.client.gui.GuiComponent;
 import net.minecraft.client.gui.components.events.AbstractContainerEventHandler;
 import net.minecraft.client.gui.components.events.GuiEventListener;
+import net.minecraft.client.gui.narration.NarratableEntry;
+import net.minecraft.client.gui.narration.NarratedElementType;
+import net.minecraft.client.gui.narration.NarrationElementOutput;
+import net.minecraft.client.renderer.GameRenderer;
+import net.minecraft.network.chat.Component;
+import net.minecraft.network.chat.TranslatableComponent;
 import net.minecraft.util.Mth;
 
-public abstract class AbstractSelectionList<E extends AbstractSelectionList.Entry<E>> extends AbstractContainerEventHandler implements Widget {
+public abstract class AbstractSelectionList<E extends AbstractSelectionList.Entry<E>> extends AbstractContainerEventHandler implements Widget, NarratableEntry {
    protected final Minecraft minecraft;
    protected final int itemHeight;
    private final List<E> children = new AbstractSelectionList.TrackedList();
    protected int width;
    protected int height;
@@ -33,13 +40,16 @@
    private double scrollAmount;
    private boolean renderSelection = true;
    private boolean renderHeader;
    protected int headerHeight;
    private boolean scrolling;
+   @Nullable
    private E selected;
    private boolean renderBackground = true;
    private boolean renderTopAndBottom = true;
+   @Nullable
+   private E hovered;
 
    public AbstractSelectionList(Minecraft var1, int var2, int var3, int var4, int var5, int var6) {
       this.minecraft = var1;
       this.width = var2;
       this.height = var3;
@@ -164,15 +174,17 @@
       this.renderBackground(var1);
       int var5 = this.getScrollbarPosition();
       int var6 = var5 + 6;
       Tesselator var7 = Tesselator.getInstance();
       BufferBuilder var8 = var7.getBuilder();
+      RenderSystem.setShader(GameRenderer::getPositionTexColorShader);
+      this.hovered = this.isMouseOver((double)var2, (double)var3) ? this.getEntryAtPosition((double)var2, (double)var3) : null;
       if (this.renderBackground) {
-         this.minecraft.getTextureManager().bind(GuiComponent.BACKGROUND_LOCATION);
-         RenderSystem.color4f(1.0F, 1.0F, 1.0F, 1.0F);
+         RenderSystem.setShaderTexture(0, GuiComponent.BACKGROUND_LOCATION);
+         RenderSystem.setShaderColor(1.0F, 1.0F, 1.0F, 1.0F);
          float var9 = 32.0F;
-         var8.begin(7, DefaultVertexFormat.POSITION_TEX_COLOR);
+         var8.begin(VertexFormat.Mode.QUADS, DefaultVertexFormat.POSITION_TEX_COLOR);
          var8.vertex((double)this.x0, (double)this.y1, 0.0D).uv((float)this.x0 / 32.0F, (float)(this.y1 + (int)this.getScrollAmount()) / 32.0F).color(32, 32, 32, 255).endVertex();
          var8.vertex((double)this.x1, (double)this.y1, 0.0D).uv((float)this.x1 / 32.0F, (float)(this.y1 + (int)this.getScrollAmount()) / 32.0F).color(32, 32, 32, 255).endVertex();
          var8.vertex((double)this.x1, (double)this.y0, 0.0D).uv((float)this.x1 / 32.0F, (float)(this.y0 + (int)this.getScrollAmount()) / 32.0F).color(32, 32, 32, 255).endVertex();
          var8.vertex((double)this.x0, (double)this.y0, 0.0D).uv((float)this.x0 / 32.0F, (float)(this.y0 + (int)this.getScrollAmount()) / 32.0F).color(32, 32, 32, 255).endVertex();
          var7.end();
@@ -184,16 +196,17 @@
          this.renderHeader(var1, var14, var10, var7);
       }
 
       this.renderList(var1, var14, var10, var2, var3, var4);
       if (this.renderTopAndBottom) {
-         this.minecraft.getTextureManager().bind(GuiComponent.BACKGROUND_LOCATION);
+         RenderSystem.setShader(GameRenderer::getPositionTexColorShader);
+         RenderSystem.setShaderTexture(0, GuiComponent.BACKGROUND_LOCATION);
          RenderSystem.enableDepthTest();
          RenderSystem.depthFunc(519);
          float var11 = 32.0F;
          boolean var12 = true;
-         var8.begin(7, DefaultVertexFormat.POSITION_TEX_COLOR);
+         var8.begin(VertexFormat.Mode.QUADS, DefaultVertexFormat.POSITION_TEX_COLOR);
          var8.vertex((double)this.x0, (double)this.y0, -100.0D).uv(0.0F, (float)this.y0 / 32.0F).color(64, 64, 64, 255).endVertex();
          var8.vertex((double)(this.x0 + this.width), (double)this.y0, -100.0D).uv((float)this.width / 32.0F, (float)this.y0 / 32.0F).color(64, 64, 64, 255).endVertex();
          var8.vertex((double)(this.x0 + this.width), 0.0D, -100.0D).uv((float)this.width / 32.0F, 0.0F).color(64, 64, 64, 255).endVertex();
          var8.vertex((double)this.x0, 0.0D, -100.0D).uv(0.0F, 0.0F).color(64, 64, 64, 255).endVertex();
          var8.vertex((double)this.x0, (double)this.height, -100.0D).uv(0.0F, (float)this.height / 32.0F).color(64, 64, 64, 255).endVertex();
@@ -203,56 +216,54 @@
          var7.end();
          RenderSystem.depthFunc(515);
          RenderSystem.disableDepthTest();
          RenderSystem.enableBlend();
          RenderSystem.blendFuncSeparate(GlStateManager.SourceFactor.SRC_ALPHA, GlStateManager.DestFactor.ONE_MINUS_SRC_ALPHA, GlStateManager.SourceFactor.ZERO, GlStateManager.DestFactor.ONE);
-         RenderSystem.disableAlphaTest();
-         RenderSystem.shadeModel(7425);
          RenderSystem.disableTexture();
+         RenderSystem.setShader(GameRenderer::getPositionColorShader);
          boolean var13 = true;
-         var8.begin(7, DefaultVertexFormat.POSITION_TEX_COLOR);
-         var8.vertex((double)this.x0, (double)(this.y0 + 4), 0.0D).uv(0.0F, 1.0F).color(0, 0, 0, 0).endVertex();
-         var8.vertex((double)this.x1, (double)(this.y0 + 4), 0.0D).uv(1.0F, 1.0F).color(0, 0, 0, 0).endVertex();
-         var8.vertex((double)this.x1, (double)this.y0, 0.0D).uv(1.0F, 0.0F).color(0, 0, 0, 255).endVertex();
-         var8.vertex((double)this.x0, (double)this.y0, 0.0D).uv(0.0F, 0.0F).color(0, 0, 0, 255).endVertex();
-         var8.vertex((double)this.x0, (double)this.y1, 0.0D).uv(0.0F, 1.0F).color(0, 0, 0, 255).endVertex();
-         var8.vertex((double)this.x1, (double)this.y1, 0.0D).uv(1.0F, 1.0F).color(0, 0, 0, 255).endVertex();
-         var8.vertex((double)this.x1, (double)(this.y1 - 4), 0.0D).uv(1.0F, 0.0F).color(0, 0, 0, 0).endVertex();
-         var8.vertex((double)this.x0, (double)(this.y1 - 4), 0.0D).uv(0.0F, 0.0F).color(0, 0, 0, 0).endVertex();
+         var8.begin(VertexFormat.Mode.QUADS, DefaultVertexFormat.POSITION_COLOR);
+         var8.vertex((double)this.x0, (double)(this.y0 + 4), 0.0D).color(0, 0, 0, 0).endVertex();
+         var8.vertex((double)this.x1, (double)(this.y0 + 4), 0.0D).color(0, 0, 0, 0).endVertex();
+         var8.vertex((double)this.x1, (double)this.y0, 0.0D).color(0, 0, 0, 255).endVertex();
+         var8.vertex((double)this.x0, (double)this.y0, 0.0D).color(0, 0, 0, 255).endVertex();
+         var8.vertex((double)this.x0, (double)this.y1, 0.0D).color(0, 0, 0, 255).endVertex();
+         var8.vertex((double)this.x1, (double)this.y1, 0.0D).color(0, 0, 0, 255).endVertex();
+         var8.vertex((double)this.x1, (double)(this.y1 - 4), 0.0D).color(0, 0, 0, 0).endVertex();
+         var8.vertex((double)this.x0, (double)(this.y1 - 4), 0.0D).color(0, 0, 0, 0).endVertex();
          var7.end();
       }
 
       int var15 = this.getMaxScroll();
       if (var15 > 0) {
          RenderSystem.disableTexture();
+         RenderSystem.setShader(GameRenderer::getPositionColorShader);
          int var16 = (int)((float)((this.y1 - this.y0) * (this.y1 - this.y0)) / (float)this.getMaxPosition());
-         var16 = Mth.clamp(var16, 32, this.y1 - this.y0 - 8);
+         var16 = Mth.clamp((int)var16, (int)32, (int)(this.y1 - this.y0 - 8));
          int var17 = (int)this.getScrollAmount() * (this.y1 - this.y0 - var16) / var15 + this.y0;
          if (var17 < this.y0) {
             var17 = this.y0;
          }
 
-         var8.begin(7, DefaultVertexFormat.POSITION_TEX_COLOR);
-         var8.vertex((double)var5, (double)this.y1, 0.0D).uv(0.0F, 1.0F).color(0, 0, 0, 255).endVertex();
-         var8.vertex((double)var6, (double)this.y1, 0.0D).uv(1.0F, 1.0F).color(0, 0, 0, 255).endVertex();
-         var8.vertex((double)var6, (double)this.y0, 0.0D).uv(1.0F, 0.0F).color(0, 0, 0, 255).endVertex();
-         var8.vertex((double)var5, (double)this.y0, 0.0D).uv(0.0F, 0.0F).color(0, 0, 0, 255).endVertex();
-         var8.vertex((double)var5, (double)(var17 + var16), 0.0D).uv(0.0F, 1.0F).color(128, 128, 128, 255).endVertex();
-         var8.vertex((double)var6, (double)(var17 + var16), 0.0D).uv(1.0F, 1.0F).color(128, 128, 128, 255).endVertex();
-         var8.vertex((double)var6, (double)var17, 0.0D).uv(1.0F, 0.0F).color(128, 128, 128, 255).endVertex();
-         var8.vertex((double)var5, (double)var17, 0.0D).uv(0.0F, 0.0F).color(128, 128, 128, 255).endVertex();
-         var8.vertex((double)var5, (double)(var17 + var16 - 1), 0.0D).uv(0.0F, 1.0F).color(192, 192, 192, 255).endVertex();
-         var8.vertex((double)(var6 - 1), (double)(var17 + var16 - 1), 0.0D).uv(1.0F, 1.0F).color(192, 192, 192, 255).endVertex();
-         var8.vertex((double)(var6 - 1), (double)var17, 0.0D).uv(1.0F, 0.0F).color(192, 192, 192, 255).endVertex();
-         var8.vertex((double)var5, (double)var17, 0.0D).uv(0.0F, 0.0F).color(192, 192, 192, 255).endVertex();
+         var8.begin(VertexFormat.Mode.QUADS, DefaultVertexFormat.POSITION_COLOR);
+         var8.vertex((double)var5, (double)this.y1, 0.0D).color(0, 0, 0, 255).endVertex();
+         var8.vertex((double)var6, (double)this.y1, 0.0D).color(0, 0, 0, 255).endVertex();
+         var8.vertex((double)var6, (double)this.y0, 0.0D).color(0, 0, 0, 255).endVertex();
+         var8.vertex((double)var5, (double)this.y0, 0.0D).color(0, 0, 0, 255).endVertex();
+         var8.vertex((double)var5, (double)(var17 + var16), 0.0D).color(128, 128, 128, 255).endVertex();
+         var8.vertex((double)var6, (double)(var17 + var16), 0.0D).color(128, 128, 128, 255).endVertex();
+         var8.vertex((double)var6, (double)var17, 0.0D).color(128, 128, 128, 255).endVertex();
+         var8.vertex((double)var5, (double)var17, 0.0D).color(128, 128, 128, 255).endVertex();
+         var8.vertex((double)var5, (double)(var17 + var16 - 1), 0.0D).color(192, 192, 192, 255).endVertex();
+         var8.vertex((double)(var6 - 1), (double)(var17 + var16 - 1), 0.0D).color(192, 192, 192, 255).endVertex();
+         var8.vertex((double)(var6 - 1), (double)var17, 0.0D).color(192, 192, 192, 255).endVertex();
+         var8.vertex((double)var5, (double)var17, 0.0D).color(192, 192, 192, 255).endVertex();
          var7.end();
       }
 
       this.renderDecorations(var1, var2, var3);
       RenderSystem.enableTexture();
-      RenderSystem.shadeModel(7424);
-      RenderSystem.enableAlphaTest();
       RenderSystem.disableBlend();
    }
 
    protected void centerScrollOn(E var1) {
       this.setScrollAmount((double)(this.children().indexOf(var1) * this.itemHeight + this.itemHeight / 2 - (this.y1 - this.y0) / 2));
@@ -286,10 +297,14 @@
 
    public int getMaxScroll() {
       return Math.max(0, this.getMaxPosition() - (this.y1 - this.y0 - 4));
    }
 
+   public int getScrollBottom() {
+      return (int)this.getScrollAmount() - this.height - this.headerHeight;
+   }
+
    protected void updateScrollingState(double var1, double var3, int var5) {
       this.scrolling = var5 == 0 && var1 >= (double)this.getScrollbarPosition() && var1 < (double)(this.getScrollbarPosition() + 6);
    }
 
    protected int getScrollbarPosition() {
@@ -334,11 +349,11 @@
          } else if (var3 > (double)this.y1) {
             this.setScrollAmount((double)this.getMaxScroll());
          } else {
             double var10 = (double)Math.max(1, this.getMaxScroll());
             int var12 = this.y1 - this.y0;
-            int var13 = Mth.clamp((int)((float)(var12 * var12) / (float)this.getMaxPosition()), 32, var12 - 8);
+            int var13 = Mth.clamp((int)((int)((float)(var12 * var12) / (float)this.getMaxPosition())), (int)32, (int)(var12 - 8));
             double var14 = Math.max(1.0D, var10 / (double)(var12 - var13));
             this.setScrollAmount(this.getScrollAmount() + var8 * var14);
          }
 
          return true;
@@ -385,11 +400,11 @@
       int var3 = var1 == AbstractSelectionList.SelectionDirection.UP ? -1 : 1;
       if (!this.children().isEmpty()) {
          int var4 = this.children().indexOf(this.getSelected());
 
          while(true) {
-            int var5 = Mth.clamp(var4 + var3, 0, this.getItemCount() - 1);
+            int var5 = Mth.clamp((int)(var4 + var3), (int)0, (int)(this.getItemCount() - 1));
             if (var4 == var5) {
                break;
             }
 
             AbstractSelectionList.Entry var6 = (AbstractSelectionList.Entry)this.children().get(var5);
@@ -425,30 +440,31 @@
             int var17;
             if (this.renderSelection && this.isSelectedItem(var10)) {
                var17 = this.x0 + this.width / 2 - var16 / 2;
                int var18 = this.x0 + this.width / 2 + var16 / 2;
                RenderSystem.disableTexture();
+               RenderSystem.setShader(GameRenderer::getPositionShader);
                float var19 = this.isFocused() ? 1.0F : 0.5F;
-               RenderSystem.color4f(var19, var19, var19, 1.0F);
-               var9.begin(7, DefaultVertexFormat.POSITION);
+               RenderSystem.setShaderColor(var19, var19, var19, 1.0F);
+               var9.begin(VertexFormat.Mode.QUADS, DefaultVertexFormat.POSITION);
                var9.vertex((double)var17, (double)(var13 + var14 + 2), 0.0D).endVertex();
                var9.vertex((double)var18, (double)(var13 + var14 + 2), 0.0D).endVertex();
                var9.vertex((double)var18, (double)(var13 - 2), 0.0D).endVertex();
                var9.vertex((double)var17, (double)(var13 - 2), 0.0D).endVertex();
                var8.end();
-               RenderSystem.color4f(0.0F, 0.0F, 0.0F, 1.0F);
-               var9.begin(7, DefaultVertexFormat.POSITION);
+               RenderSystem.setShaderColor(0.0F, 0.0F, 0.0F, 1.0F);
+               var9.begin(VertexFormat.Mode.QUADS, DefaultVertexFormat.POSITION);
                var9.vertex((double)(var17 + 1), (double)(var13 + var14 + 1), 0.0D).endVertex();
                var9.vertex((double)(var18 - 1), (double)(var13 + var14 + 1), 0.0D).endVertex();
                var9.vertex((double)(var18 - 1), (double)(var13 - 1), 0.0D).endVertex();
                var9.vertex((double)(var17 + 1), (double)(var13 - 1), 0.0D).endVertex();
                var8.end();
                RenderSystem.enableTexture();
             }
 
             var17 = this.getRowLeft();
-            var15.render(var1, var10, var11, var17, var16, var14, var4, var5, this.isMouseOver((double)var4, (double)var5) && Objects.equals(this.getEntryAtPosition((double)var4, (double)var5), var15), var6);
+            var15.render(var1, var10, var11, var17, var16, var14, var4, var5, Objects.equals(this.hovered, var15), var6);
          }
       }
 
    }
 
@@ -470,10 +486,19 @@
 
    protected boolean isFocused() {
       return false;
    }
 
+   public NarratableEntry.NarrationPriority narrationPriority() {
+      if (this.isFocused()) {
+         return NarratableEntry.NarrationPriority.FOCUSED;
+      } else {
+         return this.hovered != null ? NarratableEntry.NarrationPriority.HOVERED : NarratableEntry.NarrationPriority.NONE;
+      }
+   }
+
+   @Nullable
    protected E remove(int var1) {
       AbstractSelectionList.Entry var2 = (AbstractSelectionList.Entry)this.children.get(var1);
       return this.removeEntry((AbstractSelectionList.Entry)this.children.get(var1)) ? var2 : null;
    }
 
@@ -484,19 +509,34 @@
       }
 
       return var2;
    }
 
-   private void bindEntryToSelf(AbstractSelectionList.Entry<E> var1) {
+   @Nullable
+   protected E getHovered() {
+      return this.hovered;
+   }
+
+   void bindEntryToSelf(AbstractSelectionList.Entry<E> var1) {
       var1.list = this;
    }
 
+   protected void narrateListElementPosition(NarrationElementOutput var1, E var2) {
+      List var3 = this.children();
+      if (var3.size() > 1) {
+         int var4 = var3.indexOf(var2);
+         if (var4 != -1) {
+            var1.add(NarratedElementType.POSITION, (Component)(new TranslatableComponent("narrator.position.list", new Object[]{var4 + 1, var3.size()})));
+         }
+      }
+
+   }
+
    class TrackedList extends AbstractList<E> {
-      private final List<E> delegate;
+      private final List<E> delegate = Lists.newArrayList();
 
-      private TrackedList() {
-         this.delegate = Lists.newArrayList();
+      TrackedList() {
       }
 
       public E get(int var1) {
          return (AbstractSelectionList.Entry)this.delegate.get(var1);
       }
@@ -521,11 +561,11 @@
       }
    }
 
    public abstract static class Entry<E extends AbstractSelectionList.Entry<E>> implements GuiEventListener {
       @Deprecated
-      private AbstractSelectionList<E> list;
+      AbstractSelectionList<E> list;
 
       public Entry() {
       }
 
       public abstract void render(PoseStack var1, int var2, int var3, int var4, int var5, int var6, int var7, int var8, boolean var9, float var10);
@@ -533,11 +573,11 @@
       public boolean isMouseOver(double var1, double var3) {
          return Objects.equals(this.list.getEntryAtPosition(var1, var3), this);
       }
    }
 
-   public static enum SelectionDirection {
+   protected static enum SelectionDirection {
       UP,
       DOWN;
 
       private SelectionDirection() {
       }
