--- net/minecraft/client/gui/components/BossHealthOverlay.java
+++ net/minecraft/client/gui/components/BossHealthOverlay.java
@@ -54,11 +54,11 @@
       this.blit(var1, var2, var3, 0, var4.getColor().ordinal() * 5 * 2, 182, 5);
       if (var4.getOverlay() != BossEvent.BossBarOverlay.PROGRESS) {
          this.blit(var1, var2, var3, 0, 80 + (var4.getOverlay().ordinal() - 1) * 5 * 2, 182, 5);
       }
 
-      int var5 = (int)(var4.getPercent() * 183.0F);
+      int var5 = (int)(var4.getProgress() * 183.0F);
       if (var5 > 0) {
          this.blit(var1, var2, var3, 0, var4.getColor().ordinal() * 5 * 2 + 5, var5, 5);
          if (var4.getOverlay() != BossEvent.BossBarOverlay.PROGRESS) {
             this.blit(var1, var2, var3, 0, 80 + (var4.getOverlay().ordinal() - 1) * 5 * 2 + 5, var5, 5);
          }
