--- net/minecraft/client/gui/components/Button.java
+++ net/minecraft/client/gui/components/Button.java
@@ -1,8 +1,11 @@
 package net.minecraft.client.gui.components;
 
 import com.mojang.blaze3d.vertex.PoseStack;
+import java.util.function.Consumer;
+import net.minecraft.client.gui.narration.NarratedElementType;
+import net.minecraft.client.gui.narration.NarrationElementOutput;
 import net.minecraft.network.chat.Component;
 
 public class Button extends AbstractButton {
    public static final Button.OnTooltip NO_TOOLTIP = (var0, var1, var2, var3) -> {
    };
@@ -33,12 +36,22 @@
 
    public void renderToolTip(PoseStack var1, int var2, int var3) {
       this.onTooltip.onTooltip(this, var1, var2, var3);
    }
 
+   public void updateNarration(NarrationElementOutput var1) {
+      this.defaultButtonNarrationText(var1);
+      this.onTooltip.narrateTooltip((var1x) -> {
+         var1.add(NarratedElementType.HINT, var1x);
+      });
+   }
+
    public interface OnTooltip {
       void onTooltip(Button var1, PoseStack var2, int var3, int var4);
+
+      default void narrateTooltip(Consumer<Component> var1) {
+      }
    }
 
    public interface OnPress {
       void onPress(Button var1);
    }
