--- net/minecraft/client/gui/components/ChatComponent.java
+++ net/minecraft/client/gui/components/ChatComponent.java
@@ -28,11 +28,11 @@
    private final List<GuiMessage<Component>> allMessages = Lists.newArrayList();
    private final List<GuiMessage<FormattedCharSequence>> trimmedMessages = Lists.newArrayList();
    private final Deque<Component> chatQueue = Queues.newArrayDeque();
    private int chatScrollbarPos;
    private boolean newMessageSinceScroll;
-   private long lastMessage = 0L;
+   private long lastMessage;
 
    public ChatComponent(Minecraft var1) {
       this.minecraft = var1;
    }
 
