--- net/minecraft/client/gui/components/ChatComponent.java
+++ net/minecraft/client/gui/components/ChatComponent.java
@@ -28,11 +28,11 @@
    private final List<GuiMessage<Component>> allMessages = Lists.newArrayList();
    private final List<GuiMessage<FormattedCharSequence>> trimmedMessages = Lists.newArrayList();
    private final Deque<Component> chatQueue = Queues.newArrayDeque();
    private int chatScrollbarPos;
    private boolean newMessageSinceScroll;
-   private long lastMessage = 0L;
+   private long lastMessage;
 
    public ChatComponent(Minecraft var1) {
       this.minecraft = var1;
    }
 
@@ -45,83 +45,80 @@
             boolean var5 = false;
             if (this.isChatFocused()) {
                var5 = true;
             }
 
-            double var6 = this.getScale();
-            int var8 = Mth.ceil((double)this.getWidth() / var6);
-            RenderSystem.pushMatrix();
-            RenderSystem.translatef(2.0F, 8.0F, 0.0F);
-            RenderSystem.scaled(var6, var6, 1.0D);
-            double var9 = this.minecraft.options.chatOpacity * 0.8999999761581421D + 0.10000000149011612D;
-            double var11 = this.minecraft.options.textBackgroundOpacity;
-            double var13 = 9.0D * (this.minecraft.options.chatLineSpacing + 1.0D);
-            double var15 = -8.0D * (this.minecraft.options.chatLineSpacing + 1.0D) + 4.0D * this.minecraft.options.chatLineSpacing;
-            int var17 = 0;
+            float var6 = (float)this.getScale();
+            int var7 = Mth.ceil((float)this.getWidth() / var6);
+            var1.pushPose();
+            var1.translate(2.0D, 8.0D, 0.0D);
+            var1.scale(var6, var6, 1.0F);
+            double var8 = this.minecraft.options.chatOpacity * 0.8999999761581421D + 0.10000000149011612D;
+            double var10 = this.minecraft.options.textBackgroundOpacity;
+            double var12 = 9.0D * (this.minecraft.options.chatLineSpacing + 1.0D);
+            double var14 = -8.0D * (this.minecraft.options.chatLineSpacing + 1.0D) + 4.0D * this.minecraft.options.chatLineSpacing;
+            int var16 = 0;
 
-            int var18;
-            int var20;
+            int var17;
+            int var19;
+            int var22;
             int var23;
-            int var24;
-            for(var18 = 0; var18 + this.chatScrollbarPos < this.trimmedMessages.size() && var18 < var3; ++var18) {
-               GuiMessage var19 = (GuiMessage)this.trimmedMessages.get(var18 + this.chatScrollbarPos);
-               if (var19 != null) {
-                  var20 = var2 - var19.getAddedTime();
-                  if (var20 < 200 || var5) {
-                     double var21 = var5 ? 1.0D : getTimeFactor(var20);
-                     var23 = (int)(255.0D * var21 * var9);
-                     var24 = (int)(255.0D * var21 * var11);
-                     ++var17;
-                     if (var23 > 3) {
-                        boolean var25 = false;
-                        double var26 = (double)(-var18) * var13;
+            for(var17 = 0; var17 + this.chatScrollbarPos < this.trimmedMessages.size() && var17 < var3; ++var17) {
+               GuiMessage var18 = (GuiMessage)this.trimmedMessages.get(var17 + this.chatScrollbarPos);
+               if (var18 != null) {
+                  var19 = var2 - var18.getAddedTime();
+                  if (var19 < 200 || var5) {
+                     double var20 = var5 ? 1.0D : getTimeFactor(var19);
+                     var22 = (int)(255.0D * var20 * var8);
+                     var23 = (int)(255.0D * var20 * var10);
+                     ++var16;
+                     if (var22 > 3) {
+                        boolean var24 = false;
+                        double var25 = (double)(-var17) * var12;
                         var1.pushPose();
                         var1.translate(0.0D, 0.0D, 50.0D);
-                        fill(var1, -2, (int)(var26 - var13), 0 + var8 + 4, (int)var26, var24 << 24);
+                        fill(var1, -2, (int)(var25 - var12), 0 + var7 + 4, (int)var25, var23 << 24);
                         RenderSystem.enableBlend();
                         var1.translate(0.0D, 0.0D, 50.0D);
-                        this.minecraft.font.drawShadow(var1, (FormattedCharSequence)var19.getMessage(), 0.0F, (float)((int)(var26 + var15)), 16777215 + (var23 << 24));
-                        RenderSystem.disableAlphaTest();
+                        this.minecraft.font.drawShadow(var1, (FormattedCharSequence)var18.getMessage(), 0.0F, (float)((int)(var25 + var14)), 16777215 + (var22 << 24));
                         RenderSystem.disableBlend();
                         var1.popPose();
                      }
                   }
                }
             }
 
-            int var29;
+            int var28;
             if (!this.chatQueue.isEmpty()) {
-               var18 = (int)(128.0D * var9);
-               var29 = (int)(255.0D * var11);
+               var17 = (int)(128.0D * var8);
+               var28 = (int)(255.0D * var10);
                var1.pushPose();
                var1.translate(0.0D, 0.0D, 50.0D);
-               fill(var1, -2, 0, var8 + 4, 9, var29 << 24);
+               fill(var1, -2, 0, var7 + 4, 9, var28 << 24);
                RenderSystem.enableBlend();
                var1.translate(0.0D, 0.0D, 50.0D);
-               this.minecraft.font.drawShadow(var1, (Component)(new TranslatableComponent("chat.queue", new Object[]{this.chatQueue.size()})), 0.0F, 1.0F, 16777215 + (var18 << 24));
+               this.minecraft.font.drawShadow(var1, (Component)(new TranslatableComponent("chat.queue", new Object[]{this.chatQueue.size()})), 0.0F, 1.0F, 16777215 + (var17 << 24));
                var1.popPose();
-               RenderSystem.disableAlphaTest();
                RenderSystem.disableBlend();
             }
 
             if (var5) {
                this.minecraft.font.getClass();
-               byte var28 = 9;
-               RenderSystem.translatef(-3.0F, 0.0F, 0.0F);
-               var29 = var4 * var28 + var4;
-               var20 = var17 * var28 + var17;
-               int var30 = this.chatScrollbarPos * var20 / var4;
-               int var22 = var20 * var20 / var29;
-               if (var29 != var20) {
-                  var23 = var30 > 0 ? 170 : 96;
-                  var24 = this.newMessageSinceScroll ? 13382451 : 3355562;
-                  fill(var1, 0, -var30, 2, -var30 - var22, var24 + (var23 << 24));
-                  fill(var1, 2, -var30, 1, -var30 - var22, 13421772 + (var23 << 24));
+               byte var27 = 9;
+               var28 = var4 * var27 + var4;
+               var19 = var16 * var27 + var16;
+               int var29 = this.chatScrollbarPos * var19 / var4;
+               int var21 = var19 * var19 / var28;
+               if (var28 != var19) {
+                  var22 = var29 > 0 ? 170 : 96;
+                  var23 = this.newMessageSinceScroll ? 13382451 : 3355562;
+                  fill(var1, 0, -var29, 2, -var29 - var21, var23 + (var22 << 24));
+                  fill(var1, 2, -var29, 1, -var29 - var21, 13421772 + (var22 << 24));
                }
             }
 
-            RenderSystem.popMatrix();
+            var1.popPose();
          }
       }
    }
 
    private boolean isChatHidden() {
