--- net/minecraft/client/gui/components/Checkbox.java
+++ net/minecraft/client/gui/components/Checkbox.java
@@ -32,14 +32,14 @@
       return this.selected;
    }
 
    public void renderButton(PoseStack var1, int var2, int var3, float var4) {
       Minecraft var5 = Minecraft.getInstance();
-      var5.getTextureManager().bind(TEXTURE);
+      RenderSystem.setShaderTexture(0, TEXTURE);
       RenderSystem.enableDepthTest();
       Font var6 = var5.font;
-      RenderSystem.color4f(1.0F, 1.0F, 1.0F, this.alpha);
+      RenderSystem.setShaderColor(1.0F, 1.0F, 1.0F, this.alpha);
       RenderSystem.enableBlend();
       RenderSystem.defaultBlendFunc();
       RenderSystem.blendFunc(GlStateManager.SourceFactor.SRC_ALPHA, GlStateManager.DestFactor.ONE_MINUS_SRC_ALPHA);
       blit(var1, this.x, this.y, this.isFocused() ? 20.0F : 0.0F, this.selected ? 20.0F : 0.0F, 20, this.height, 64, 64);
       this.renderBg(var1, var5, var2, var3);
