--- net/minecraft/client/gui/components/CommandSuggestions.java
+++ net/minecraft/client/gui/components/CommandSuggestions.java
@@ -120,11 +120,11 @@
             Suggestion var5;
             for(Iterator var4 = var2.getList().iterator(); var4.hasNext(); var3 = Math.max(var3, this.font.width(var5.getText()))) {
                var5 = (Suggestion)var4.next();
             }
 
-            int var6 = Mth.clamp(this.input.getScreenX(var2.getRange().getStart()), 0, this.input.getScreenX(0) + this.input.getInnerWidth() - var3);
+            int var6 = Mth.clamp((int)this.input.getScreenX(var2.getRange().getStart()), (int)0, (int)(this.input.getScreenX(0) + this.input.getInnerWidth() - var3));
             int var7 = this.anchorToBottom ? this.screen.height - 12 : 72;
             this.suggestions = new CommandSuggestions.SuggestionsList(var6, var7, var3, this.sortSuggestions(var2), var1);
          }
       }
 
@@ -271,11 +271,11 @@
          }
       }
 
       if (!var5.isEmpty()) {
          this.commandUsage.addAll(var5);
-         this.commandUsagePosition = Mth.clamp(this.input.getScreenX(var3.startPos), 0, this.input.getScreenX(0) + this.input.getInnerWidth() - var6);
+         this.commandUsagePosition = Mth.clamp((int)this.input.getScreenX(var3.startPos), (int)0, (int)(this.input.getScreenX(0) + this.input.getInnerWidth() - var6));
          this.commandUsageWidth = var6;
       }
 
    }
 
@@ -324,11 +324,11 @@
             var4 = var12;
          }
       }
 
       var3.add(FormattedCharSequence.forward(var1.substring(var4), LITERAL_STYLE));
-      return FormattedCharSequence.composite(var3);
+      return FormattedCharSequence.composite((List)var3);
    }
 
    public void render(PoseStack var1, int var2, int var3) {
       if (this.suggestions != null) {
          this.suggestions.render(var1, var2, var3);
@@ -452,11 +452,11 @@
 
       public boolean mouseScrolled(double var1) {
          int var3 = (int)(CommandSuggestions.this.minecraft.mouseHandler.xpos() * (double)CommandSuggestions.this.minecraft.getWindow().getGuiScaledWidth() / (double)CommandSuggestions.this.minecraft.getWindow().getScreenWidth());
          int var4 = (int)(CommandSuggestions.this.minecraft.mouseHandler.ypos() * (double)CommandSuggestions.this.minecraft.getWindow().getGuiScaledHeight() / (double)CommandSuggestions.this.minecraft.getWindow().getScreenHeight());
          if (this.rect.contains(var3, var4)) {
-            this.offset = Mth.clamp((int)((double)this.offset - var1), 0, Math.max(this.suggestionList.size() - CommandSuggestions.this.suggestionLineLimit, 0));
+            this.offset = Mth.clamp((int)((int)((double)this.offset - var1)), (int)0, (int)Math.max(this.suggestionList.size() - CommandSuggestions.this.suggestionLineLimit, 0));
             return true;
          } else {
             return false;
          }
       }
@@ -488,13 +488,13 @@
       public void cycle(int var1) {
          this.select(this.current + var1);
          int var2 = this.offset;
          int var3 = this.offset + CommandSuggestions.this.suggestionLineLimit - 1;
          if (this.current < var2) {
-            this.offset = Mth.clamp(this.current, 0, Math.max(this.suggestionList.size() - CommandSuggestions.this.suggestionLineLimit, 0));
+            this.offset = Mth.clamp((int)this.current, (int)0, (int)Math.max(this.suggestionList.size() - CommandSuggestions.this.suggestionLineLimit, 0));
          } else if (this.current > var3) {
-            this.offset = Mth.clamp(this.current + CommandSuggestions.this.lineStartOffset - CommandSuggestions.this.suggestionLineLimit, 0, Math.max(this.suggestionList.size() - CommandSuggestions.this.suggestionLineLimit, 0));
+            this.offset = Mth.clamp((int)(this.current + CommandSuggestions.this.lineStartOffset - CommandSuggestions.this.suggestionLineLimit), (int)0, (int)Math.max(this.suggestionList.size() - CommandSuggestions.this.suggestionLineLimit, 0));
          }
 
       }
 
       public void select(int var1) {
