--- net/minecraft/client/gui/components/ComponentRenderUtils.java
+++ net/minecraft/client/gui/components/ComponentRenderUtils.java
@@ -14,10 +14,13 @@
 import net.minecraft.util.FormattedCharSequence;
 
 public class ComponentRenderUtils {
    private static final FormattedCharSequence INDENT;
 
+   public ComponentRenderUtils() {
+   }
+
    private static String stripColor(String var0) {
       return Minecraft.getInstance().options.chatColors ? var0 : ChatFormatting.stripFormatting(var0);
    }
 
    public static List<FormattedCharSequence> wrapComponents(FormattedText var0, int var1, Font var2) {
