--- net/minecraft/client/gui/components/EditBox.java
+++ net/minecraft/client/gui/components/EditBox.java
@@ -4,10 +4,11 @@
 import com.mojang.blaze3d.systems.RenderSystem;
 import com.mojang.blaze3d.vertex.BufferBuilder;
 import com.mojang.blaze3d.vertex.DefaultVertexFormat;
 import com.mojang.blaze3d.vertex.PoseStack;
 import com.mojang.blaze3d.vertex.Tesselator;
+import com.mojang.blaze3d.vertex.VertexFormat;
 import java.util.Objects;
 import java.util.function.BiFunction;
 import java.util.function.Consumer;
 import java.util.function.Predicate;
 import javax.annotation.Nullable;
@@ -16,10 +17,11 @@
 import net.minecraft.client.Minecraft;
 import net.minecraft.client.gui.Font;
 import net.minecraft.client.gui.GuiComponent;
 import net.minecraft.client.gui.components.events.GuiEventListener;
 import net.minecraft.client.gui.screens.Screen;
+import net.minecraft.client.renderer.GameRenderer;
 import net.minecraft.network.chat.Component;
 import net.minecraft.network.chat.MutableComponent;
 import net.minecraft.network.chat.Style;
 import net.minecraft.network.chat.TranslatableComponent;
 import net.minecraft.util.FormattedCharSequence;
@@ -462,15 +464,16 @@
          var1 = this.x + this.width;
       }
 
       Tesselator var7 = Tesselator.getInstance();
       BufferBuilder var6 = var7.getBuilder();
-      RenderSystem.color4f(0.0F, 0.0F, 255.0F, 255.0F);
+      RenderSystem.setShader(GameRenderer::getPositionShader);
+      RenderSystem.setShaderColor(0.0F, 0.0F, 255.0F, 255.0F);
       RenderSystem.disableTexture();
       RenderSystem.enableColorLogicOp();
       RenderSystem.logicOp(GlStateManager.LogicOp.OR_REVERSE);
-      var6.begin(7, DefaultVertexFormat.POSITION);
+      var6.begin(VertexFormat.Mode.QUADS, DefaultVertexFormat.POSITION);
       var6.vertex((double)var1, (double)var4, 0.0D).endVertex();
       var6.vertex((double)var3, (double)var4, 0.0D).endVertex();
       var6.vertex((double)var3, (double)var2, 0.0D).endVertex();
       var6.vertex((double)var1, (double)var2, 0.0D).endVertex();
       var7.end();
