--- net/minecraft/client/gui/components/ImageButton.java
+++ net/minecraft/client/gui/components/ImageButton.java
@@ -1,10 +1,11 @@
 package net.minecraft.client.gui.components;
 
 import com.mojang.blaze3d.systems.RenderSystem;
 import com.mojang.blaze3d.vertex.PoseStack;
 import net.minecraft.client.Minecraft;
+import net.minecraft.client.renderer.GameRenderer;
 import net.minecraft.network.chat.Component;
 import net.minecraft.network.chat.TextComponent;
 import net.minecraft.resources.ResourceLocation;
 
 public class ImageButton extends Button {
@@ -42,11 +43,12 @@
       this.y = var2;
    }
 
    public void renderButton(PoseStack var1, int var2, int var3, float var4) {
       Minecraft var5 = Minecraft.getInstance();
-      var5.getTextureManager().bind(this.resourceLocation);
+      RenderSystem.setShader(GameRenderer::getPositionTexShader);
+      RenderSystem.setShaderTexture(0, this.resourceLocation);
       int var6 = this.yTexStart;
       if (this.isHovered()) {
          var6 += this.yDiffTex;
       }
 
