--- net/minecraft/client/gui/components/ImageButton.java
+++ net/minecraft/client/gui/components/ImageButton.java
@@ -1,10 +1,11 @@
 package net.minecraft.client.gui.components;
 
 import com.mojang.blaze3d.systems.RenderSystem;
 import com.mojang.blaze3d.vertex.PoseStack;
 import net.minecraft.client.Minecraft;
+import net.minecraft.client.renderer.GameRenderer;
 import net.minecraft.network.chat.Component;
 import net.minecraft.network.chat.TextComponent;
 import net.minecraft.resources.ResourceLocation;
 
 public class ImageButton extends Button {
@@ -13,10 +14,14 @@
    private final int yTexStart;
    private final int yDiffTex;
    private final int textureWidth;
    private final int textureHeight;
 
+   public ImageButton(int var1, int var2, int var3, int var4, int var5, int var6, ResourceLocation var7, Button.OnPress var8) {
+      this(var1, var2, var3, var4, var5, var6, var4, var7, 256, 256, var8);
+   }
+
    public ImageButton(int var1, int var2, int var3, int var4, int var5, int var6, int var7, ResourceLocation var8, Button.OnPress var9) {
       this(var1, var2, var3, var4, var5, var6, var7, var8, 256, 256, var9);
    }
 
    public ImageButton(int var1, int var2, int var3, int var4, int var5, int var6, int var7, ResourceLocation var8, int var9, int var10, Button.OnPress var11) {
@@ -42,11 +47,12 @@
       this.y = var2;
    }
 
    public void renderButton(PoseStack var1, int var2, int var3, float var4) {
       Minecraft var5 = Minecraft.getInstance();
-      var5.getTextureManager().bind(this.resourceLocation);
+      RenderSystem.setShader(GameRenderer::getPositionTexShader);
+      RenderSystem.setShaderTexture(0, this.resourceLocation);
       int var6 = this.yTexStart;
       if (this.isHovered()) {
          var6 += this.yDiffTex;
       }
 
