--- net/minecraft/client/gui/components/MultiLineLabel.java
+++ net/minecraft/client/gui/components/MultiLineLabel.java
@@ -49,10 +49,16 @@
       return createFixed(var0, (List)Arrays.stream(var1).map(Component::getVisualOrderText).map((var1x) -> {
          return new MultiLineLabel.TextWithWidth(var1x, var0.width(var1x));
       }).collect(ImmutableList.toImmutableList()));
    }
 
+   static MultiLineLabel create(Font var0, List<Component> var1) {
+      return createFixed(var0, (List)var1.stream().map(Component::getVisualOrderText).map((var1x) -> {
+         return new MultiLineLabel.TextWithWidth(var1x, var0.width(var1x));
+      }).collect(ImmutableList.toImmutableList()));
+   }
+
    static MultiLineLabel createFixed(final Font var0, final List<MultiLineLabel.TextWithWidth> var1) {
       return var1.isEmpty() ? EMPTY : new MultiLineLabel() {
          public int renderCentered(PoseStack var1x, int var2, int var3) {
             var0.getClass();
             return this.renderCentered(var1x, var2, var3, 9, 16777215);
