--- net/minecraft/client/gui/components/MultiLineLabel.java
+++ net/minecraft/client/gui/components/MultiLineLabel.java
@@ -3,10 +3,11 @@
 import com.google.common.collect.ImmutableList;
 import com.mojang.blaze3d.vertex.PoseStack;
 import java.util.Arrays;
 import java.util.Iterator;
 import java.util.List;
+import java.util.Objects;
 import net.minecraft.client.gui.Font;
 import net.minecraft.network.chat.Component;
 import net.minecraft.network.chat.FormattedText;
 import net.minecraft.util.FormattedCharSequence;
 
@@ -49,14 +50,20 @@
       return createFixed(var0, (List)Arrays.stream(var1).map(Component::getVisualOrderText).map((var1x) -> {
          return new MultiLineLabel.TextWithWidth(var1x, var0.width(var1x));
       }).collect(ImmutableList.toImmutableList()));
    }
 
+   static MultiLineLabel create(Font var0, List<Component> var1) {
+      return createFixed(var0, (List)var1.stream().map(Component::getVisualOrderText).map((var1x) -> {
+         return new MultiLineLabel.TextWithWidth(var1x, var0.width(var1x));
+      }).collect(ImmutableList.toImmutableList()));
+   }
+
    static MultiLineLabel createFixed(final Font var0, final List<MultiLineLabel.TextWithWidth> var1) {
       return var1.isEmpty() ? EMPTY : new MultiLineLabel() {
          public int renderCentered(PoseStack var1x, int var2, int var3) {
-            var0.getClass();
+            Objects.requireNonNull(var0);
             return this.renderCentered(var1x, var2, var3, 9, 16777215);
          }
 
          public int renderCentered(PoseStack var1x, int var2, int var3, int var4, int var5) {
             int var6 = var3;
@@ -106,14 +113,14 @@
    int renderLeftAlignedNoShadow(PoseStack var1, int var2, int var3, int var4, int var5);
 
    int getLineCount();
 
    public static class TextWithWidth {
-      private final FormattedCharSequence text;
-      private final int width;
+      final FormattedCharSequence text;
+      final int width;
 
-      private TextWithWidth(FormattedCharSequence var1, int var2) {
+      TextWithWidth(FormattedCharSequence var1, int var2) {
          this.text = var1;
          this.width = var2;
       }
    }
 }
