--- net/minecraft/client/gui/components/PlayerTabOverlay.java
+++ net/minecraft/client/gui/components/PlayerTabOverlay.java
@@ -154,21 +154,20 @@
          var22 = var39 / var35;
          var23 = var39 % var35;
          int var24 = var15 + var22 * var14 + var22 * 5;
          int var25 = var16 + var23 * 9;
          fill(var1, var24, var25, var24 + var14, var25 + 8, var38);
-         RenderSystem.color4f(1.0F, 1.0F, 1.0F, 1.0F);
-         RenderSystem.enableAlphaTest();
+         RenderSystem.setShaderColor(1.0F, 1.0F, 1.0F, 1.0F);
          RenderSystem.enableBlend();
          RenderSystem.defaultBlendFunc();
          if (var39 < var6.size()) {
             PlayerInfo var26 = (PlayerInfo)var6.get(var39);
             GameProfile var27 = var26.getProfile();
             if (var12) {
                Player var28 = this.minecraft.level.getPlayerByUUID(var27.getId());
                boolean var29 = var28 != null && var28.isModelPartShown(PlayerModelPart.CAPE) && ("Dinnerbone".equals(var27.getName()) || "Grumm".equals(var27.getName()));
-               this.minecraft.getTextureManager().bind(var26.getSkinLocation());
+               RenderSystem.setShaderTexture(0, var26.getSkinLocation());
                int var30 = 8 + (var29 ? 8 : 0);
                int var31 = 8 * (var29 ? -1 : 1);
                GuiComponent.blit(var1, var24, var25, 8, 8, 8.0F, (float)var30, 8, var31, 64, 64);
                if (var28 != null && var28.isModelPartShown(PlayerModelPart.HAT)) {
                   int var32 = 8 + (var29 ? 8 : 0);
@@ -210,12 +209,12 @@
       }
 
    }
 
    protected void renderPingIcon(PoseStack var1, int var2, int var3, int var4, PlayerInfo var5) {
-      RenderSystem.color4f(1.0F, 1.0F, 1.0F, 1.0F);
-      this.minecraft.getTextureManager().bind(GUI_ICONS_LOCATION);
+      RenderSystem.setShaderColor(1.0F, 1.0F, 1.0F, 1.0F);
+      RenderSystem.setShaderTexture(0, GUI_ICONS_LOCATION);
       boolean var6 = false;
       byte var7;
       if (var5.getLatency() < 0) {
          var7 = 5;
       } else if (var5.getLatency() < 150) {
@@ -236,11 +235,11 @@
    }
 
    private void renderTablistScore(Objective var1, int var2, String var3, int var4, int var5, PlayerInfo var6, PoseStack var7) {
       int var8 = var1.getScoreboard().getOrCreatePlayerScore(var3, var1).getScore();
       if (var1.getRenderType() == ObjectiveCriteria.RenderType.HEARTS) {
-         this.minecraft.getTextureManager().bind(GUI_ICONS_LOCATION);
+         RenderSystem.setShaderTexture(0, GUI_ICONS_LOCATION);
          long var9 = Util.getMillis();
          if (this.visibilityId == var6.getRenderVisibilityId()) {
             if (var8 < var6.getLastHealth()) {
                var6.setLastHealthTime(var9);
                var6.setHealthBlinkTime((long)(this.gui.getGuiTicks() + 20));
