--- net/minecraft/client/gui/components/StateSwitchingButton.java
+++ net/minecraft/client/gui/components/StateSwitchingButton.java
@@ -1,10 +1,11 @@
 package net.minecraft.client.gui.components;
 
 import com.mojang.blaze3d.systems.RenderSystem;
 import com.mojang.blaze3d.vertex.PoseStack;
-import net.minecraft.client.Minecraft;
+import net.minecraft.client.gui.narration.NarrationElementOutput;
+import net.minecraft.client.renderer.GameRenderer;
 import net.minecraft.network.chat.TextComponent;
 import net.minecraft.resources.ResourceLocation;
 
 public class StateSwitchingButton extends AbstractWidget {
    protected ResourceLocation resourceLocation;
@@ -38,23 +39,27 @@
    public void setPosition(int var1, int var2) {
       this.x = var1;
       this.y = var2;
    }
 
+   public void updateNarration(NarrationElementOutput var1) {
+      this.defaultButtonNarrationText(var1);
+   }
+
    public void renderButton(PoseStack var1, int var2, int var3, float var4) {
-      Minecraft var5 = Minecraft.getInstance();
-      var5.getTextureManager().bind(this.resourceLocation);
+      RenderSystem.setShader(GameRenderer::getPositionTexShader);
+      RenderSystem.setShaderTexture(0, this.resourceLocation);
       RenderSystem.disableDepthTest();
-      int var6 = this.xTexStart;
-      int var7 = this.yTexStart;
+      int var5 = this.xTexStart;
+      int var6 = this.yTexStart;
       if (this.isStateTriggered) {
-         var6 += this.xDiffTex;
+         var5 += this.xDiffTex;
       }
 
       if (this.isHovered()) {
-         var7 += this.yDiffTex;
+         var6 += this.yDiffTex;
       }
 
-      this.blit(var1, this.x, this.y, var6, var7, this.width, this.height);
+      this.blit(var1, this.x, this.y, var5, var6, this.width, this.height);
       RenderSystem.enableDepthTest();
    }
 }
