--- net/minecraft/client/gui/components/SubtitleOverlay.java
+++ net/minecraft/client/gui/components/SubtitleOverlay.java
@@ -33,11 +33,10 @@
          this.minecraft.getSoundManager().removeListener(this);
          this.isListening = false;
       }
 
       if (this.isListening && !this.subtitles.isEmpty()) {
-         RenderSystem.pushMatrix();
          RenderSystem.enableBlend();
          RenderSystem.defaultBlendFunc();
          Vec3 var2 = new Vec3(this.minecraft.player.getX(), this.minecraft.player.getEyeY(), this.minecraft.player.getZ());
          Vec3 var3 = (new Vec3(0.0D, 0.0D, -1.0D)).xRot(-this.minecraft.player.xRot * 0.017453292F).yRot(-this.minecraft.player.yRot * 0.017453292F);
          Vec3 var4 = (new Vec3(0.0D, 1.0D, 0.0D)).xRot(-this.minecraft.player.xRot * 0.017453292F).yRot(-this.minecraft.player.yRot * 0.017453292F);
@@ -72,13 +71,13 @@
             int var20 = var19 / 2;
             float var21 = 1.0F;
             int var22 = this.minecraft.font.width((FormattedText)var11);
             int var23 = Mth.floor(Mth.clampedLerp(255.0D, 75.0D, (double)((float)(Util.getMillis() - var9.getTime()) / 3000.0F)));
             int var24 = var23 << 16 | var23 << 8 | var23;
-            RenderSystem.pushMatrix();
-            RenderSystem.translatef((float)this.minecraft.getWindow().getGuiScaledWidth() - (float)var18 * 1.0F - 2.0F, (float)(this.minecraft.getWindow().getGuiScaledHeight() - 30) - (float)(var6 * (var19 + 1)) * 1.0F, 0.0F);
-            RenderSystem.scalef(1.0F, 1.0F, 1.0F);
+            var1.pushPose();
+            var1.translate((double)((float)this.minecraft.getWindow().getGuiScaledWidth() - (float)var18 * 1.0F - 2.0F), (double)((float)(this.minecraft.getWindow().getGuiScaledHeight() - 30) - (float)(var6 * (var19 + 1)) * 1.0F), 0.0D);
+            var1.scale(1.0F, 1.0F, 1.0F);
             fill(var1, -var18 - 1, -var20 - 1, var18 + 1, var20 + 1, this.minecraft.options.getBackgroundColor(0.8F));
             RenderSystem.enableBlend();
             if (!var17) {
                if (var13 > 0.0D) {
                   this.minecraft.font.draw(var1, ">", (float)(var18 - this.minecraft.font.width(">")), (float)(-var20), var24 + -16777216);
@@ -86,15 +85,14 @@
                   this.minecraft.font.draw(var1, "<", (float)(-var18), (float)(-var20), var24 + -16777216);
                }
             }
 
             this.minecraft.font.draw(var1, var11, (float)(-var22 / 2), (float)(-var20), var24 + -16777216);
-            RenderSystem.popMatrix();
+            var1.popPose();
          }
 
          RenderSystem.disableBlend();
-         RenderSystem.popMatrix();
       }
    }
 
    public void onPlaySound(SoundInstance var1, WeighedSoundEvents var2) {
       if (var2.getSubtitle() != null) {
