--- net/minecraft/client/gui/components/TooltipAccessor.java
+++ net/minecraft/client/gui/components/TooltipAccessor.java
@@ -1,9 +1,8 @@
 package net.minecraft.client.gui.components;
 
 import java.util.List;
-import java.util.Optional;
 import net.minecraft.util.FormattedCharSequence;
 
 public interface TooltipAccessor {
-   Optional<List<FormattedCharSequence>> getTooltip();
+   List<FormattedCharSequence> getTooltip();
 }
