--- net/minecraft/world/phys/shapes/CubeVoxelShape.java
+++ net/minecraft/world/phys/shapes/CubeVoxelShape.java
@@ -13,8 +13,8 @@
       return new CubePointRange(this.shape.getSize(var1));
    }
 
    protected int findIndex(Direction.Axis var1, double var2) {
       int var4 = this.shape.getSize(var1);
-      return Mth.clamp(Mth.floor(var2 * (double)var4), -1, var4);
+      return Mth.floor(Mth.clamp(var2 * (double)var4, -1.0D, (double)var4));
    }
 }
