--- net/minecraft/world/phys/shapes/CubePointRange.java
+++ net/minecraft/world/phys/shapes/CubePointRange.java
@@ -4,11 +4,15 @@
 
 public class CubePointRange extends AbstractDoubleList {
    private final int parts;
 
    CubePointRange(int var1) {
-      this.parts = var1;
+      if (var1 <= 0) {
+         throw new IllegalArgumentException("Need at least 1 part");
+      } else {
+         this.parts = var1;
+      }
    }
 
    public double getDouble(int var1) {
       return (double)var1 / (double)this.parts;
    }
