--- net/minecraft/client/resources/language/ClientLanguage.java
+++ net/minecraft/client/resources/language/ClientLanguage.java
@@ -7,10 +7,11 @@
 import java.io.InputStream;
 import java.util.HashMap;
 import java.util.Iterator;
 import java.util.List;
 import java.util.Map;
+import java.util.Objects;
 import net.minecraft.locale.Language;
 import net.minecraft.network.chat.FormattedText;
 import net.minecraft.resources.ResourceLocation;
 import net.minecraft.server.packs.resources.Resource;
 import net.minecraft.server.packs.resources.ResourceManager;
@@ -61,33 +62,31 @@
       while(var2.hasNext()) {
          Resource var3 = (Resource)var2.next();
 
          try {
             InputStream var4 = var3.getInputStream();
-            Throwable var5 = null;
 
             try {
+               Objects.requireNonNull(var1);
                Language.loadFromJson(var4, var1::put);
-            } catch (Throwable var15) {
-               var5 = var15;
-               throw var15;
-            } finally {
+            } catch (Throwable var8) {
                if (var4 != null) {
-                  if (var5 != null) {
-                     try {
-                        var4.close();
-                     } catch (Throwable var14) {
-                        var5.addSuppressed(var14);
-                     }
-                  } else {
+                  try {
                      var4.close();
+                  } catch (Throwable var7) {
+                     var8.addSuppressed(var7);
                   }
                }
 
+               throw var8;
             }
-         } catch (IOException var17) {
-            LOGGER.warn("Failed to load translations from {}", var3, var17);
+
+            if (var4 != null) {
+               var4.close();
+            }
+         } catch (IOException var9) {
+            LOGGER.warn("Failed to load translations from {}", var3, var9);
          }
       }
 
    }
 
