--- net/minecraft/client/resources/language/I18n.java
+++ net/minecraft/client/resources/language/I18n.java
@@ -4,10 +4,13 @@
 import net.minecraft.locale.Language;
 
 public class I18n {
    private static volatile Language language = Language.getInstance();
 
+   private I18n() {
+   }
+
    static void setLanguage(Language var0) {
       language = var0;
    }
 
    public static String get(String var0, Object... var1) {
