--- net/minecraft/client/resources/language/LanguageManager.java
+++ net/minecraft/client/resources/language/LanguageManager.java
@@ -19,10 +19,11 @@
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
 
 public class LanguageManager implements ResourceManagerReloadListener {
    private static final Logger LOGGER = LogManager.getLogger();
+   public static final String DEFAULT_LANGUAGE_CODE = "en_us";
    private static final LanguageInfo DEFAULT_LANGUAGE = new LanguageInfo("en_us", "US", "English", false);
    private Map<String, LanguageInfo> languages;
    private String currentCode;
    private LanguageInfo currentLanguage;
 
