--- net/minecraft/world/inventory/HopperMenu.java
+++ net/minecraft/world/inventory/HopperMenu.java
@@ -5,10 +5,11 @@
 import net.minecraft.world.entity.player.Inventory;
 import net.minecraft.world.entity.player.Player;
 import net.minecraft.world.item.ItemStack;
 
 public class HopperMenu extends AbstractContainerMenu {
+   public static final int CONTAINER_SIZE = 5;
    private final Container hopper;
 
    public HopperMenu(int var1, Inventory var2) {
       this(var1, var2, new SimpleContainer(5));
    }
