--- net/minecraft/world/inventory/HorseInventoryMenu.java
+++ net/minecraft/world/inventory/HorseInventoryMenu.java
@@ -19,11 +19,11 @@
       boolean var5 = true;
       var3.startOpen(var2.player);
       boolean var6 = true;
       this.addSlot(new Slot(var3, 0, 8, 18) {
          public boolean mayPlace(ItemStack var1) {
-            return var1.getItem() == Items.SADDLE && !this.hasItem() && var4.isSaddleable();
+            return var1.is(Items.SADDLE) && !this.hasItem() && var4.isSaddleable();
          }
       });
       this.addSlot(new Slot(var3, 1, 8, 36) {
          public boolean mayPlace(ItemStack var1) {
             return var4.isArmor(var1);
