--- net/minecraft/world/inventory/InventoryMenu.java
+++ net/minecraft/world/inventory/InventoryMenu.java
@@ -90,11 +90,11 @@
 
    public void removed(Player var1) {
       super.removed(var1);
       this.resultSlots.clearContent();
       if (!var1.level.isClientSide) {
-         this.clearContainer(var1, var1.level, this.craftSlots);
+         this.clearContainer(var1, this.craftSlots);
       }
    }
 
    public boolean stillValid(Player var1) {
       return true;
@@ -175,14 +175,22 @@
 
    public int getGridHeight() {
       return this.craftSlots.getHeight();
    }
 
+   public int getSize() {
+      return 5;
+   }
+
    public CraftingContainer getCraftSlots() {
       return this.craftSlots;
    }
 
+   public boolean shouldMoveToInventory(int var1) {
+      return var1 != this.getResultSlotIndex();
+   }
+
    static {
       TEXTURE_EMPTY_SLOTS = new ResourceLocation[]{EMPTY_ARMOR_SLOT_BOOTS, EMPTY_ARMOR_SLOT_LEGGINGS, EMPTY_ARMOR_SLOT_CHESTPLATE, EMPTY_ARMOR_SLOT_HELMET};
       SLOT_IDS = new EquipmentSlot[]{EquipmentSlot.HEAD, EquipmentSlot.CHEST, EquipmentSlot.LEGS, EquipmentSlot.FEET};
    }
 }
