--- net/minecraft/world/inventory/ItemCombinerMenu.java
+++ net/minecraft/world/inventory/ItemCombinerMenu.java
@@ -69,11 +69,11 @@
    }
 
    public void removed(Player var1) {
       super.removed(var1);
       this.access.execute((var2, var3) -> {
-         this.clearContainer(var1, var2, this.inputSlots);
+         this.clearContainer(var1, this.inputSlots);
       });
    }
 
    public boolean stillValid(Player var1) {
       return (Boolean)this.access.evaluate((var2, var3) -> {
