--- net/minecraft/world/inventory/ItemCombinerMenu.java
+++ net/minecraft/world/inventory/ItemCombinerMenu.java
@@ -7,10 +7,17 @@
 import net.minecraft.world.entity.player.Player;
 import net.minecraft.world.item.ItemStack;
 import net.minecraft.world.level.block.state.BlockState;
 
 public abstract class ItemCombinerMenu extends AbstractContainerMenu {
+   public static final int INPUT_SLOT = 0;
+   public static final int ADDITIONAL_SLOT = 1;
+   public static final int RESULT_SLOT = 2;
+   private static final int INV_SLOT_START = 3;
+   private static final int INV_SLOT_END = 30;
+   private static final int USE_ROW_SLOT_START = 30;
+   private static final int USE_ROW_SLOT_END = 39;
    protected final ResultContainer resultSlots = new ResultContainer();
    protected final Container inputSlots = new SimpleContainer(2) {
       public void setChanged() {
          super.setChanged();
          ItemCombinerMenu.this.slotsChanged(this);
@@ -19,11 +26,11 @@
    protected final ContainerLevelAccess access;
    protected final Player player;
 
    protected abstract boolean mayPickup(Player var1, boolean var2);
 
-   protected abstract ItemStack onTake(Player var1, ItemStack var2);
+   protected abstract void onTake(Player var1, ItemStack var2);
 
    protected abstract boolean isValidBlock(BlockState var1);
 
    public ItemCombinerMenu(@Nullable MenuType<?> var1, int var2, Inventory var3, ContainerLevelAccess var4) {
       super(var1, var2);
@@ -38,12 +45,12 @@
 
          public boolean mayPickup(Player var1) {
             return ItemCombinerMenu.this.mayPickup(var1, this.hasItem());
          }
 
-         public ItemStack onTake(Player var1, ItemStack var2) {
-            return ItemCombinerMenu.this.onTake(var1, var2);
+         public void onTake(Player var1, ItemStack var2) {
+            ItemCombinerMenu.this.onTake(var1, var2);
          }
       });
 
       int var5;
       for(var5 = 0; var5 < 3; ++var5) {
@@ -69,11 +76,11 @@
    }
 
    public void removed(Player var1) {
       super.removed(var1);
       this.access.execute((var2, var3) -> {
-         this.clearContainer(var1, var2, this.inputSlots);
+         this.clearContainer(var1, this.inputSlots);
       });
    }
 
    public boolean stillValid(Player var1) {
       return (Boolean)this.access.evaluate((var2, var3) -> {
