--- net/minecraft/world/inventory/LoomMenu.java
+++ net/minecraft/world/inventory/LoomMenu.java
@@ -196,11 +196,11 @@
    }
 
    public void removed(Player var1) {
       super.removed(var1);
       this.access.execute((var2, var3) -> {
-         this.clearContainer(var1, var1.level, this.inputContainer);
+         this.clearContainer(var1, this.inputContainer);
       });
    }
 
    private void setupResultSlot() {
       if (this.selectedBannerPatternIndex.get() > 0) {
