--- net/minecraft/world/inventory/LoomMenu.java
+++ net/minecraft/world/inventory/LoomMenu.java
@@ -70,11 +70,11 @@
       this.resultSlot = this.addSlot(new Slot(this.outputContainer, 0, 143, 58) {
          public boolean mayPlace(ItemStack var1) {
             return false;
          }
 
-         public ItemStack onTake(Player var1, ItemStack var2) {
+         public void onTake(Player var1, ItemStack var2) {
             LoomMenu.this.bannerSlot.remove(1);
             LoomMenu.this.dyeSlot.remove(1);
             if (!LoomMenu.this.bannerSlot.hasItem() || !LoomMenu.this.dyeSlot.hasItem()) {
                LoomMenu.this.selectedBannerPatternIndex.set(0);
             }
@@ -85,11 +85,11 @@
                   var1x.playSound((Player)null, (BlockPos)var2x, SoundEvents.UI_LOOM_TAKE_RESULT, SoundSource.BLOCKS, 1.0F, 1.0F);
                   LoomMenu.this.lastSoundTime = var3x;
                }
 
             });
-            return super.onTake(var1, var2);
+            super.onTake(var1, var2);
          }
       });
 
       int var4;
       for(var4 = 0; var4 < 3; ++var4) {
@@ -196,11 +196,11 @@
    }
 
    public void removed(Player var1) {
       super.removed(var1);
       this.access.execute((var2, var3) -> {
-         this.clearContainer(var1, var1.level, this.inputContainer);
+         this.clearContainer(var1, this.inputContainer);
       });
    }
 
    private void setupResultSlot() {
       if (this.selectedBannerPatternIndex.get() > 0) {
