--- net/minecraft/world/inventory/MerchantResultSlot.java
+++ net/minecraft/world/inventory/MerchantResultSlot.java
@@ -39,11 +39,11 @@
    protected void checkTakeAchievements(ItemStack var1) {
       var1.onCraftedBy(this.player.level, this.player, this.removeCount);
       this.removeCount = 0;
    }
 
-   public ItemStack onTake(Player var1, ItemStack var2) {
+   public void onTake(Player var1, ItemStack var2) {
       this.checkTakeAchievements(var2);
       MerchantOffer var3 = this.slots.getActiveOffer();
       if (var3 != null) {
          ItemStack var4 = this.slots.getItem(0);
          ItemStack var5 = this.slots.getItem(1);
@@ -55,8 +55,7 @@
          }
 
          this.merchant.overrideXp(this.merchant.getVillagerXp() + var3.getXp());
       }
 
-      return var2;
    }
 }
