--- net/minecraft/world/inventory/PlayerEnderChestContainer.java
+++ net/minecraft/world/inventory/PlayerEnderChestContainer.java
@@ -1,25 +1,31 @@
 package net.minecraft.world.inventory;
 
+import javax.annotation.Nullable;
 import net.minecraft.nbt.CompoundTag;
 import net.minecraft.nbt.ListTag;
 import net.minecraft.world.SimpleContainer;
 import net.minecraft.world.entity.player.Player;
 import net.minecraft.world.item.ItemStack;
 import net.minecraft.world.level.block.entity.EnderChestBlockEntity;
 
 public class PlayerEnderChestContainer extends SimpleContainer {
+   @Nullable
    private EnderChestBlockEntity activeChest;
 
    public PlayerEnderChestContainer() {
       super(27);
    }
 
    public void setActiveChest(EnderChestBlockEntity var1) {
       this.activeChest = var1;
    }
 
+   public boolean isActiveChest(EnderChestBlockEntity var1) {
+      return this.activeChest == var1;
+   }
+
    public void fromTag(ListTag var1) {
       int var2;
       for(var2 = 0; var2 < this.getContainerSize(); ++var2) {
          this.setItem(var2, ItemStack.EMPTY);
       }
@@ -54,19 +60,19 @@
       return this.activeChest != null && !this.activeChest.stillValid(var1) ? false : super.stillValid(var1);
    }
 
    public void startOpen(Player var1) {
       if (this.activeChest != null) {
-         this.activeChest.startOpen();
+         this.activeChest.startOpen(var1);
       }
 
       super.startOpen(var1);
    }
 
    public void stopOpen(Player var1) {
       if (this.activeChest != null) {
-         this.activeChest.stopOpen();
+         this.activeChest.stopOpen(var1);
       }
 
       super.stopOpen(var1);
       this.activeChest = null;
    }
