--- net/minecraft/world/inventory/RecipeBookMenu.java
+++ net/minecraft/world/inventory/RecipeBookMenu.java
@@ -24,6 +24,12 @@
    public abstract int getResultSlotIndex();
 
    public abstract int getGridWidth();
 
    public abstract int getGridHeight();
+
+   public abstract int getSize();
+
+   public abstract RecipeBookType getRecipeBookType();
+
+   public abstract boolean shouldMoveToInventory(int var1);
 }
