--- net/minecraft/world/inventory/Slot.java
+++ net/minecraft/world/inventory/Slot.java
@@ -1,7 +1,8 @@
 package net.minecraft.world.inventory;
 
+import java.util.Optional;
 import net.minecraft.world.Container;
 import net.minecraft.world.entity.player.Player;
 import net.minecraft.world.item.ItemStack;
 
 public class Slot {
@@ -33,13 +34,12 @@
    }
 
    protected void checkTakeAchievements(ItemStack var1) {
    }
 
-   public ItemStack onTake(Player var1, ItemStack var2) {
+   public void onTake(Player var1, ItemStack var2) {
       this.setChanged();
-      return var2;
    }
 
    public boolean mayPlace(ItemStack var1) {
       return true;
    }
@@ -64,16 +64,74 @@
    public int getMaxStackSize() {
       return this.container.getMaxStackSize();
    }
 
    public int getMaxStackSize(ItemStack var1) {
-      return this.getMaxStackSize();
+      return Math.min(this.getMaxStackSize(), var1.getMaxStackSize());
    }
 
    public ItemStack remove(int var1) {
       return this.container.removeItem(this.slot, var1);
    }
 
    public boolean mayPickup(Player var1) {
       return true;
    }
+
+   public Optional<ItemStack> tryRemove(int var1, int var2, Player var3) {
+      if (!this.mayPickup(var3)) {
+         return Optional.empty();
+      } else if (!this.allowModification(var3) && var2 < this.getItem().getCount()) {
+         return Optional.empty();
+      } else {
+         if (!this.allowModification(var3)) {
+            var1 = this.getItem().getCount();
+         }
+
+         var1 = Math.min(var1, var2);
+         ItemStack var4 = this.remove(var1);
+         if (this.getItem().isEmpty()) {
+            this.set(ItemStack.EMPTY);
+         }
+
+         return Optional.of(var4);
+      }
+   }
+
+   public ItemStack safeTake(int var1, int var2, Player var3) {
+      Optional var4 = this.tryRemove(var1, var2, var3);
+      var4.ifPresent((var2x) -> {
+         this.onTake(var3, var2x);
+      });
+      return (ItemStack)var4.orElse(ItemStack.EMPTY);
+   }
+
+   public ItemStack safeInsert(ItemStack var1) {
+      return this.safeInsert(var1, var1.getCount());
+   }
+
+   public ItemStack safeInsert(ItemStack var1, int var2) {
+      if (!var1.isEmpty() && this.mayPlace(var1)) {
+         ItemStack var3 = this.getItem();
+         int var4 = Math.min(Math.min(var2, var1.getCount()), this.getMaxStackSize(var1) - var3.getCount());
+         if (var3.isEmpty()) {
+            this.set(var1.split(var4));
+         } else if (ItemStack.isSameItemSameTags(var3, var1)) {
+            var1.shrink(var4);
+            var3.grow(var4);
+            this.set(var3);
+         }
+
+         return var1;
+      } else {
+         return var1;
+      }
+   }
+
+   public boolean allowModification(Player var1) {
+      return this.mayPickup(var1) && this.mayPlace(this.getItem());
+   }
+
+   public int getContainerSlot() {
+      return this.slot;
+   }
 }
