--- net/minecraft/world/inventory/StonecutterMenu.java
+++ net/minecraft/world/inventory/StonecutterMenu.java
@@ -107,11 +107,11 @@
       return var1 >= 0 && var1 < this.recipes.size();
    }
 
    public void slotsChanged(Container var1) {
       ItemStack var2 = this.inputSlot.getItem();
-      if (var2.getItem() != this.input.getItem()) {
+      if (!var2.is(this.input.getItem())) {
          this.input = var2.copy();
          this.setupRecipeList(var1, var2);
       }
 
    }
@@ -194,9 +194,9 @@
 
    public void removed(Player var1) {
       super.removed(var1);
       this.resultContainer.removeItemNoUpdate(1);
       this.access.execute((var2, var3) -> {
-         this.clearContainer(var1, var1.level, this.container);
+         this.clearContainer(var1, this.container);
       });
    }
 }
