--- net/minecraft/world/level/newbiome/layer/Layer.java
+++ net/minecraft/world/level/newbiome/layer/Layer.java
@@ -28,11 +28,11 @@
          Biome var6 = (Biome)var1.get(var5);
          if (var6 == null) {
             if (SharedConstants.IS_RUNNING_IN_IDE) {
                throw (IllegalStateException)Util.pauseInIde(new IllegalStateException("Unknown biome id: " + var4));
             } else {
-               LOGGER.warn("Unknown biome id: ", var4);
+               LOGGER.warn("Unknown biome id: {}", var4);
                return (Biome)var1.get(Biomes.byId(0));
             }
          } else {
             return var6;
          }
