--- net/minecraft/world/level/newbiome/layer/Layer.java
+++ net/minecraft/world/level/newbiome/layer/Layer.java
@@ -1,8 +1,7 @@
 package net.minecraft.world.level.newbiome.layer;
 
-import net.minecraft.SharedConstants;
 import net.minecraft.Util;
 import net.minecraft.core.Registry;
 import net.minecraft.data.worldgen.biome.Biomes;
 import net.minecraft.resources.ResourceKey;
 import net.minecraft.world.level.biome.Biome;
@@ -25,16 +24,12 @@
       if (var5 == null) {
          throw new IllegalStateException("Unknown biome id emitted by layers: " + var4);
       } else {
          Biome var6 = (Biome)var1.get(var5);
          if (var6 == null) {
-            if (SharedConstants.IS_RUNNING_IN_IDE) {
-               throw (IllegalStateException)Util.pauseInIde(new IllegalStateException("Unknown biome id: " + var4));
-            } else {
-               LOGGER.warn("Unknown biome id: ", var4);
-               return (Biome)var1.get(Biomes.byId(0));
-            }
+            Util.logAndPauseIfInIde("Unknown biome id: " + var4);
+            return (Biome)var1.get(Biomes.byId(0));
          } else {
             return var6;
          }
       }
    }
