--- net/minecraft/world/level/newbiome/layer/OceanLayer.java
+++ net/minecraft/world/level/newbiome/layer/OceanLayer.java
@@ -10,11 +10,11 @@
    private OceanLayer() {
    }
 
    public int applyPixel(Context var1, int var2, int var3) {
       ImprovedNoise var4 = var1.getBiomeNoise();
-      double var5 = var4.noise((double)var2 / 8.0D, (double)var3 / 8.0D, 0.0D, 0.0D, 0.0D);
+      double var5 = var4.noise((double)var2 / 8.0D, (double)var3 / 8.0D, 0.0D);
       if (var5 > 0.4D) {
          return 44;
       } else if (var5 > 0.2D) {
          return 45;
       } else if (var5 < -0.4D) {
