--- net/minecraft/world/level/newbiome/layer/ZoomLayer.java
+++ net/minecraft/world/level/newbiome/layer/ZoomLayer.java
@@ -10,11 +10,14 @@
       protected int modeOrRandom(BigContext<?> var1, int var2, int var3, int var4, int var5) {
          return var1.random(var2, var3, var4, var5);
       }
    };
 
-   private ZoomLayer() {
+   private static final int ZOOM_BITS = 1;
+   private static final int ZOOM_MASK = 1;
+
+   ZoomLayer() {
    }
 
    public int getParentX(int var1) {
       return var1 >> 1;
    }
