--- net/minecraft/commands/Commands.java
+++ net/minecraft/commands/Commands.java
@@ -55,10 +55,11 @@
 import net.minecraft.server.commands.FunctionCommand;
 import net.minecraft.server.commands.GameModeCommand;
 import net.minecraft.server.commands.GameRuleCommand;
 import net.minecraft.server.commands.GiveCommand;
 import net.minecraft.server.commands.HelpCommand;
+import net.minecraft.server.commands.ItemCommands;
 import net.minecraft.server.commands.KickCommand;
 import net.minecraft.server.commands.KillCommand;
 import net.minecraft.server.commands.ListPlayersCommand;
 import net.minecraft.server.commands.LocateBiomeCommand;
 import net.minecraft.server.commands.LocateCommand;
@@ -70,11 +71,10 @@
 import net.minecraft.server.commands.ParticleCommand;
 import net.minecraft.server.commands.PlaySoundCommand;
 import net.minecraft.server.commands.PublishCommand;
 import net.minecraft.server.commands.RecipeCommand;
 import net.minecraft.server.commands.ReloadCommand;
-import net.minecraft.server.commands.ReplaceItemCommand;
 import net.minecraft.server.commands.SaveAllCommand;
 import net.minecraft.server.commands.SaveOffCommand;
 import net.minecraft.server.commands.SaveOnCommand;
 import net.minecraft.server.commands.SayCommand;
 import net.minecraft.server.commands.ScheduleCommand;
@@ -130,10 +130,11 @@
       FunctionCommand.register(this.dispatcher);
       GameModeCommand.register(this.dispatcher);
       GameRuleCommand.register(this.dispatcher);
       GiveCommand.register(this.dispatcher);
       HelpCommand.register(this.dispatcher);
+      ItemCommands.register(this.dispatcher);
       KickCommand.register(this.dispatcher);
       KillCommand.register(this.dispatcher);
       ListPlayersCommand.register(this.dispatcher);
       LocateCommand.register(this.dispatcher);
       LocateBiomeCommand.register(this.dispatcher);
@@ -141,11 +142,10 @@
       MsgCommand.register(this.dispatcher);
       ParticleCommand.register(this.dispatcher);
       PlaySoundCommand.register(this.dispatcher);
       ReloadCommand.register(this.dispatcher);
       RecipeCommand.register(this.dispatcher);
-      ReplaceItemCommand.register(this.dispatcher);
       SayCommand.register(this.dispatcher);
       ScheduleCommand.register(this.dispatcher);
       ScoreboardCommand.register(this.dispatcher);
       SeedCommand.register(this.dispatcher, var1 != Commands.CommandSelection.INTEGRATED);
       SetBlockCommand.register(this.dispatcher);
@@ -250,11 +250,11 @@
             var1.sendFailure((new TranslatableComponent("command.failed")).withStyle((var1x) -> {
                return var1x.withHoverEvent(new HoverEvent(HoverEvent.Action.SHOW_TEXT, var5));
             }));
             if (SharedConstants.IS_RUNNING_IN_IDE) {
                var1.sendFailure(new TextComponent(Util.describeError(var15)));
-               LOGGER.error("'" + var2 + "' threw an exception", var15);
+               LOGGER.error("'{}' threw an exception", var2, var15);
             }
 
             byte var19 = 0;
             return var19;
          }
