--- net/minecraft/commands/CommandSourceStack.java
+++ net/minecraft/commands/CommandSourceStack.java
@@ -118,15 +118,15 @@
          Vec3 var4 = new Vec3(this.worldPosition.x * var2, this.worldPosition.y, this.worldPosition.z * var2);
          return new CommandSourceStack(this.source, var4, this.rotation, var1, this.permissionLevel, this.textName, this.displayName, this.server, this.entity, this.silent, this.consumer, this.anchor);
       }
    }
 
-   public CommandSourceStack facing(Entity var1, EntityAnchorArgument.Anchor var2) throws CommandSyntaxException {
+   public CommandSourceStack facing(Entity var1, EntityAnchorArgument.Anchor var2) {
       return this.facing(var2.apply(var1));
    }
 
-   public CommandSourceStack facing(Vec3 var1) throws CommandSyntaxException {
+   public CommandSourceStack facing(Vec3 var1) {
       Vec3 var2 = this.anchor.apply(this);
       double var3 = var1.x - var2.x;
       double var5 = var1.y - var2.y;
       double var7 = var1.z - var2.z;
       double var9 = (double)Mth.sqrt(var3 * var3 + var7 * var7);
