--- net/minecraft/commands/arguments/blocks/BlockInput.java
+++ net/minecraft/commands/arguments/blocks/BlockInput.java
@@ -67,11 +67,11 @@
             if (var5 != null) {
                CompoundTag var6 = this.tag.copy();
                var6.putInt("x", var2.getX());
                var6.putInt("y", var2.getY());
                var6.putInt("z", var2.getZ());
-               var5.load(var4, var6);
+               var5.load(var6);
             }
          }
 
          return true;
       }
