--- net/minecraft/commands/arguments/blocks/package-info.java
+++ net/minecraft/commands/arguments/blocks/package-info.java
@@ -1,0 +1,10 @@
+package net.minecraft.commands.arguments.blocks;
+
+import javax.annotation.ParametersAreNonnullByDefault;
+import net.minecraft.MethodsReturnNonnullByDefault;
+
+// $FF: synthetic class
+@ParametersAreNonnullByDefault
+@MethodsReturnNonnullByDefault
+interface package-info {
+}
