--- net/minecraft/world/level/block/grower/AbstractTreeGrower.java
+++ net/minecraft/world/level/block/grower/AbstractTreeGrower.java
@@ -24,11 +24,10 @@
       ConfiguredFeature var6 = this.getConfiguredFeature(var5, this.hasFlowers(var1, var3));
       if (var6 == null) {
          return false;
       } else {
          var1.setBlock(var3, Blocks.AIR.defaultBlockState(), 4);
-         ((TreeConfiguration)var6.config).setFromSapling();
          if (var6.place(var1, var2, var5, var3)) {
             return true;
          } else {
             var1.setBlock(var3, var4, 4);
             return false;
