--- net/minecraft/core/Rotations.java
+++ net/minecraft/core/Rotations.java
@@ -1,9 +1,10 @@
 package net.minecraft.core;
 
 import net.minecraft.nbt.FloatTag;
 import net.minecraft.nbt.ListTag;
+import net.minecraft.util.Mth;
 
 public class Rotations {
    protected final float x;
    protected final float y;
    protected final float z;
@@ -44,6 +45,18 @@
    }
 
    public float getZ() {
       return this.z;
    }
+
+   public float getWrappedX() {
+      return Mth.wrapDegrees(this.x);
+   }
+
+   public float getWrappedY() {
+      return Mth.wrapDegrees(this.y);
+   }
+
+   public float getWrappedZ() {
+      return Mth.wrapDegrees(this.z);
+   }
 }
