--- net/minecraft/world/entity/NeutralMob.java
+++ net/minecraft/world/entity/NeutralMob.java
@@ -8,10 +8,13 @@
 import net.minecraft.world.entity.player.Player;
 import net.minecraft.world.level.GameRules;
 import net.minecraft.world.level.Level;
 
 public interface NeutralMob {
+   String TAG_ANGER_TIME = "AngerTime";
+   String TAG_ANGRY_AT = "AngryAt";
+
    int getRemainingPersistentAngerTime();
 
    void setRemainingPersistentAngerTime(int var1);
 
    @Nullable
@@ -27,27 +30,29 @@
          var1.putUUID("AngryAt", this.getPersistentAngerTarget());
       }
 
    }
 
-   default void readPersistentAngerSaveData(ServerLevel var1, CompoundTag var2) {
+   default void readPersistentAngerSaveData(Level var1, CompoundTag var2) {
       this.setRemainingPersistentAngerTime(var2.getInt("AngerTime"));
-      if (!var2.hasUUID("AngryAt")) {
-         this.setPersistentAngerTarget((UUID)null);
-      } else {
-         UUID var3 = var2.getUUID("AngryAt");
-         this.setPersistentAngerTarget(var3);
-         Entity var4 = var1.getEntity(var3);
-         if (var4 != null) {
-            if (var4 instanceof Mob) {
-               this.setLastHurtByMob((Mob)var4);
-            }
+      if (var1 instanceof ServerLevel) {
+         if (!var2.hasUUID("AngryAt")) {
+            this.setPersistentAngerTarget((UUID)null);
+         } else {
+            UUID var3 = var2.getUUID("AngryAt");
+            this.setPersistentAngerTarget(var3);
+            Entity var4 = ((ServerLevel)var1).getEntity(var3);
+            if (var4 != null) {
+               if (var4 instanceof Mob) {
+                  this.setLastHurtByMob((Mob)var4);
+               }
 
-            if (var4.getType() == EntityType.PLAYER) {
-               this.setLastHurtByPlayer((Player)var4);
-            }
+               if (var4.getType() == EntityType.PLAYER) {
+                  this.setLastHurtByPlayer((Player)var4);
+               }
 
+            }
          }
       }
    }
 
    default void updatePersistentAnger(ServerLevel var1, boolean var2) {
@@ -105,10 +110,13 @@
       this.setPersistentAngerTarget((UUID)null);
       this.setTarget((LivingEntity)null);
       this.setRemainingPersistentAngerTime(0);
    }
 
+   @Nullable
+   LivingEntity getLastHurtByMob();
+
    void setLastHurtByMob(@Nullable LivingEntity var1);
 
    void setLastHurtByPlayer(@Nullable Player var1);
 
    void setTarget(@Nullable LivingEntity var1);
