--- net/minecraft/world/entity/TamableAnimal.java
+++ net/minecraft/world/entity/TamableAnimal.java
@@ -18,11 +18,11 @@
 import net.minecraft.world.entity.player.Player;
 import net.minecraft.world.level.GameRules;
 import net.minecraft.world.level.Level;
 import net.minecraft.world.scores.Team;
 
-public abstract class TamableAnimal extends Animal {
+public abstract class TamableAnimal extends Animal implements OwnableEntity {
    protected static final EntityDataAccessor<Byte> DATA_FLAGS_ID;
    protected static final EntityDataAccessor<Optional<UUID>> DATA_OWNERUUID_ID;
    private boolean orderedToSit;
 
    protected TamableAnimal(EntityType<? extends TamableAnimal> var1, Level var2) {
