--- net/minecraft/world/level/pathfinder/TurtleNodeEvaluator.java
+++ net/minecraft/world/level/pathfinder/TurtleNodeEvaluator.java
@@ -168,11 +168,11 @@
                   return var7;
                }
 
                int var17 = 0;
 
-               while(var2 > 0 && var11 == BlockPathTypes.OPEN) {
+               while(var2 > this.level.getMinBuildHeight() && var11 == BlockPathTypes.OPEN) {
                   --var2;
                   if (var17++ >= this.mob.getMaxFallDistance()) {
                      return null;
                   }
 
