--- net/minecraft/network/chat/BaseComponent.java
+++ net/minecraft/network/chat/BaseComponent.java
@@ -74,8 +74,8 @@
    public int hashCode() {
       return Objects.hash(new Object[]{this.getStyle(), this.siblings});
    }
 
    public String toString() {
-      return "BaseComponent{style=" + this.style + ", siblings=" + this.siblings + '}';
+      return "BaseComponent{style=" + this.style + ", siblings=" + this.siblings + "}";
    }
 }
