--- net/minecraft/network/chat/CommonComponents.java
+++ net/minecraft/network/chat/CommonComponents.java
@@ -9,13 +9,13 @@
    public static final Component GUI_NO = new TranslatableComponent("gui.no");
    public static final Component GUI_PROCEED = new TranslatableComponent("gui.proceed");
    public static final Component GUI_BACK = new TranslatableComponent("gui.back");
    public static final Component CONNECT_FAILED = new TranslatableComponent("connect.failed");
 
-   public static Component optionStatus(boolean var0) {
-      return var0 ? OPTION_ON : OPTION_OFF;
-   }
-
    public static MutableComponent optionStatus(Component var0, boolean var1) {
       return new TranslatableComponent(var1 ? "options.on.composed" : "options.off.composed", new Object[]{var0});
    }
+
+   public static MutableComponent optionNameValue(Component var0, Component var1) {
+      return new TranslatableComponent("options.generic_value", new Object[]{var0, var1});
+   }
 }
