--- net/minecraft/network/chat/ComponentUtils.java
+++ net/minecraft/network/chat/ComponentUtils.java
@@ -12,10 +12,13 @@
 import net.minecraft.ChatFormatting;
 import net.minecraft.commands.CommandSourceStack;
 import net.minecraft.world.entity.Entity;
 
 public class ComponentUtils {
+   public ComponentUtils() {
+   }
+
    public static MutableComponent mergeStyles(MutableComponent var0, Style var1) {
       if (var1.isEmpty()) {
          return var0;
       } else {
          Style var2 = var0.getStyle();
