--- net/minecraft/network/chat/Component.java
+++ net/minecraft/network/chat/Component.java
@@ -178,15 +178,15 @@
             } else {
                String var6;
                if (var4.has("translate")) {
                   var6 = GsonHelper.getAsString(var4, "translate");
                   if (var4.has("with")) {
-                     JsonArray var7 = GsonHelper.getAsJsonArray(var4, "with");
-                     Object[] var8 = new Object[var7.size()];
+                     JsonArray var15 = GsonHelper.getAsJsonArray(var4, "with");
+                     Object[] var8 = new Object[var15.size()];
 
                      for(int var9 = 0; var9 < var8.length; ++var9) {
-                        var8[var9] = this.deserialize(var7.get(var9), var2, var3);
+                        var8[var9] = this.deserialize(var15.get(var9), var2, var3);
                         if (var8[var9] instanceof TextComponent) {
                            TextComponent var10 = (TextComponent)var8[var9];
                            if (var10.getStyle().isEmpty() && var10.getSiblings().isEmpty()) {
                               var8[var9] = var10.getText();
                            }
@@ -212,21 +212,21 @@
                   if (!var4.has("nbt")) {
                      throw new JsonParseException("Don't know how to turn " + var1 + " into a Component");
                   }
 
                   var6 = GsonHelper.getAsString(var4, "nbt");
-                  boolean var15 = GsonHelper.getAsBoolean(var4, "interpret", false);
+                  boolean var7 = GsonHelper.getAsBoolean(var4, "interpret", false);
                   if (var4.has("block")) {
-                     var5 = new NbtComponent.BlockNbtComponent(var6, var15, GsonHelper.getAsString(var4, "block"));
+                     var5 = new NbtComponent.BlockNbtComponent(var6, var7, GsonHelper.getAsString(var4, "block"));
                   } else if (var4.has("entity")) {
-                     var5 = new NbtComponent.EntityNbtComponent(var6, var15, GsonHelper.getAsString(var4, "entity"));
+                     var5 = new NbtComponent.EntityNbtComponent(var6, var7, GsonHelper.getAsString(var4, "entity"));
                   } else {
                      if (!var4.has("storage")) {
                         throw new JsonParseException("Don't know how to turn " + var1 + " into a Component");
                      }
 
-                     var5 = new NbtComponent.StorageNbtComponent(var6, var15, new ResourceLocation(GsonHelper.getAsString(var4, "storage")));
+                     var5 = new NbtComponent.StorageNbtComponent(var6, var7, new ResourceLocation(GsonHelper.getAsString(var4, "storage")));
                   }
                }
             }
 
             if (var4.has("extra")) {
