--- net/minecraft/network/chat/FormattedText.java
+++ net/minecraft/network/chat/FormattedText.java
@@ -48,11 +48,11 @@
 
    static FormattedText composite(FormattedText... var0) {
       return composite((List)ImmutableList.copyOf(var0));
    }
 
-   static FormattedText composite(final List<FormattedText> var0) {
+   static FormattedText composite(final List<? extends FormattedText> var0) {
       return new FormattedText() {
          public <T> Optional<T> visit(FormattedText.ContentConsumer<T> var1) {
             Iterator var2 = var0.iterator();
 
             Optional var4;
