--- net/minecraft/network/chat/HoverEvent.java
+++ net/minecraft/network/chat/HoverEvent.java
@@ -100,11 +100,11 @@
          return ((HoverEvent.ItemStackInfo)var0).serialize();
       }, (var0) -> {
          return HoverEvent.ItemStackInfo.create(var0);
       });
       public static final HoverEvent.Action<HoverEvent.EntityTooltipInfo> SHOW_ENTITY = new HoverEvent.Action("show_entity", true, HoverEvent.EntityTooltipInfo::create, HoverEvent.EntityTooltipInfo::serialize, HoverEvent.EntityTooltipInfo::create);
-      private static final Map<String, HoverEvent.Action> LOOKUP;
+      private static final Map<String, HoverEvent.Action<?>> LOOKUP;
       private final String name;
       private final boolean allowFromServer;
       private final Function<JsonElement, T> argDeserializer;
       private final Function<T, JsonElement> argSerializer;
       private final Function<Component, T> legacyArgDeserializer;
@@ -124,11 +124,11 @@
       public String getName() {
          return this.name;
       }
 
       @Nullable
-      public static HoverEvent.Action getByName(String var0) {
+      public static HoverEvent.Action<?> getByName(String var0) {
          return (HoverEvent.Action)LOOKUP.get(var0);
       }
 
       private T cast(Object var1) {
          return var1;
