--- net/minecraft/network/chat/KeybindComponent.java
+++ net/minecraft/network/chat/KeybindComponent.java
@@ -51,11 +51,12 @@
          return this.name.equals(var2.name) && super.equals(var1);
       }
    }
 
    public String toString() {
-      return "KeybindComponent{keybind='" + this.name + '\'' + ", siblings=" + this.siblings + ", style=" + this.getStyle() + '}';
+      String var10000 = this.name;
+      return "KeybindComponent{keybind='" + var10000 + "', siblings=" + this.siblings + ", style=" + this.getStyle() + "}";
    }
 
    public String getName() {
       return this.name;
    }
