--- net/minecraft/network/chat/NbtComponent.java
+++ net/minecraft/network/chat/NbtComponent.java
@@ -72,11 +72,11 @@
          return (MutableComponent)(this.interpreting ? (MutableComponent)var4.flatMap((var3x) -> {
             try {
                MutableComponent var4 = Component.Serializer.fromJson(var3x);
                return Stream.of(ComponentUtils.updateForEntity(var1, var4, var2, var3));
             } catch (Exception var5) {
-               LOGGER.warn("Failed to parse component: " + var3x, var5);
+               LOGGER.warn("Failed to parse component: {}", var3x, var5);
                return Stream.of();
             }
          }).reduce((var0, var1x) -> {
             return var0.append(", ").append((Component)var1x);
          }).orElse(new TextComponent("")) : new TextComponent(Joiner.on(", ").join(var4.iterator())));
