--- net/minecraft/network/chat/SubStringSource.java
+++ net/minecraft/network/chat/SubStringSource.java
@@ -51,10 +51,18 @@
 
          return (List)(var3 ? Lists.reverse(var4) : var4);
       }
    }
 
+   public static SubStringSource create(FormattedText var0) {
+      return create(var0, (var0x) -> {
+         return var0x;
+      }, (var0x) -> {
+         return var0x;
+      });
+   }
+
    public static SubStringSource create(FormattedText var0, Int2IntFunction var1, UnaryOperator<String> var2) {
       StringBuilder var3 = new StringBuilder();
       ArrayList var4 = Lists.newArrayList();
       var0.visit((var2x, var3x) -> {
          StringDecomposer.iterateFormatted(var3x, var2x, (var2, var3xx, var4x) -> {
