--- net/minecraft/network/chat/TextComponent.java
+++ net/minecraft/network/chat/TextComponent.java
@@ -30,8 +30,9 @@
          return this.text.equals(var2.getText()) && super.equals(var1);
       }
    }
 
    public String toString() {
-      return "TextComponent{text='" + this.text + '\'' + ", siblings=" + this.siblings + ", style=" + this.getStyle() + '}';
+      String var10000 = this.text;
+      return "TextComponent{text='" + var10000 + "', siblings=" + this.siblings + ", style=" + this.getStyle() + "}";
    }
 }
