--- net/minecraft/network/chat/TextColor.java
+++ net/minecraft/network/chat/TextColor.java
@@ -7,10 +7,11 @@
 import java.util.stream.Stream;
 import javax.annotation.Nullable;
 import net.minecraft.ChatFormatting;
 
 public final class TextColor {
+   private static final String CUSTOM_COLOR_PREFIX = "#";
    private static final Map<ChatFormatting, TextColor> LEGACY_FORMAT_TO_COLOR = (Map)Stream.of(ChatFormatting.values()).filter(ChatFormatting::isColor).collect(ImmutableMap.toImmutableMap(Function.identity(), (var0) -> {
       return new TextColor(var0.getColor(), var0.getName());
    }));
    private static final Map<String, TextColor> NAMED_COLORS;
    private final int value;
