--- net/minecraft/network/chat/TranslatableComponent.java
+++ net/minecraft/network/chat/TranslatableComponent.java
@@ -186,11 +186,12 @@
       var1 = 31 * var1 + Arrays.hashCode(this.args);
       return var1;
    }
 
    public String toString() {
-      return "TranslatableComponent{key='" + this.key + '\'' + ", args=" + Arrays.toString(this.args) + ", siblings=" + this.siblings + ", style=" + this.getStyle() + '}';
+      String var10000 = this.key;
+      return "TranslatableComponent{key='" + var10000 + "', args=" + Arrays.toString(this.args) + ", siblings=" + this.siblings + ", style=" + this.getStyle() + "}";
    }
 
    public String getKey() {
       return this.key;
    }
