--- net/minecraft/client/gui/spectator/PlayerMenuItem.java
+++ net/minecraft/client/gui/spectator/PlayerMenuItem.java
@@ -17,11 +17,11 @@
 import net.minecraft.world.entity.player.Player;
 
 public class PlayerMenuItem implements SpectatorMenuItem {
    private final GameProfile profile;
    private final ResourceLocation location;
-   private final TextComponent name;
+   private final Component name;
 
    public PlayerMenuItem(GameProfile var1) {
       this.profile = var1;
       Minecraft var2 = Minecraft.getInstance();
       Map var3 = var2.getSkinManager().getInsecureSkinInformation(var1);
