--- net/minecraft/client/gui/spectator/PlayerMenuItem.java
+++ net/minecraft/client/gui/spectator/PlayerMenuItem.java
@@ -17,11 +17,11 @@
 import net.minecraft.world.entity.player.Player;
 
 public class PlayerMenuItem implements SpectatorMenuItem {
    private final GameProfile profile;
    private final ResourceLocation location;
-   private final TextComponent name;
+   private final Component name;
 
    public PlayerMenuItem(GameProfile var1) {
       this.profile = var1;
       Minecraft var2 = Minecraft.getInstance();
       Map var3 = var2.getSkinManager().getInsecureSkinInformation(var1);
@@ -41,12 +41,12 @@
    public Component getName() {
       return this.name;
    }
 
    public void renderIcon(PoseStack var1, float var2, int var3) {
-      Minecraft.getInstance().getTextureManager().bind(this.location);
-      RenderSystem.color4f(1.0F, 1.0F, 1.0F, (float)var3 / 255.0F);
+      RenderSystem.setShaderTexture(0, this.location);
+      RenderSystem.setShaderColor(1.0F, 1.0F, 1.0F, (float)var3 / 255.0F);
       GuiComponent.blit(var1, 2, 2, 12, 12, 8.0F, 8.0F, 8, 8, 64, 64);
       GuiComponent.blit(var1, 2, 2, 12, 12, 40.0F, 8.0F, 8, 8, 64, 64);
    }
 
    public boolean isEnabled() {
