--- net/minecraft/client/gui/spectator/SpectatorMenu.java
+++ net/minecraft/client/gui/spectator/SpectatorMenu.java
@@ -101,11 +101,11 @@
       this.selectedSlot = -1;
       this.page = 0;
    }
 
    public SpectatorPage getCurrentPage() {
-      return new SpectatorPage(this.category, this.getItems(), this.selectedSlot);
+      return new SpectatorPage(this.getItems(), this.selectedSlot);
    }
 
    static class ScrollMenuItem implements SpectatorMenuItem {
       private final int direction;
       private final boolean enabled;
