--- net/minecraft/commands/arguments/item/ItemPredicateArgument.java
+++ net/minecraft/commands/arguments/item/ItemPredicateArgument.java
@@ -85,11 +85,11 @@
          this.tag = var1;
          this.nbt = var2;
       }
 
       public boolean test(ItemStack var1) {
-         return this.tag.contains(var1.getItem()) && NbtUtils.compareNbt(this.nbt, var1.getTag(), true);
+         return var1.is(this.tag) && NbtUtils.compareNbt(this.nbt, var1.getTag(), true);
       }
    }
 
    static class ItemPredicate implements Predicate<ItemStack> {
       private final Item item;
@@ -100,11 +100,11 @@
          this.item = var1;
          this.nbt = var2;
       }
 
       public boolean test(ItemStack var1) {
-         return var1.getItem() == this.item && NbtUtils.compareNbt(this.nbt, var1.getTag(), true);
+         return var1.is(this.item) && NbtUtils.compareNbt(this.nbt, var1.getTag(), true);
       }
    }
 
    public interface Result {
       Predicate<ItemStack> create(CommandContext<CommandSourceStack> var1) throws CommandSyntaxException;
