--- net/minecraft/commands/arguments/item/ItemInput.java
+++ net/minecraft/commands/arguments/item/ItemInput.java
@@ -27,11 +27,11 @@
    public Item getItem() {
       return this.item;
    }
 
    public boolean test(ItemStack var1) {
-      return var1.getItem() == this.item && NbtUtils.compareNbt(this.tag, var1.getTag(), true);
+      return var1.is(this.item) && NbtUtils.compareNbt(this.tag, var1.getTag(), true);
    }
 
    public ItemStack createItemStack(int var1, boolean var2) throws CommandSyntaxException {
       ItemStack var3 = new ItemStack(this.item, var1);
       if (this.tag != null) {
