--- net/minecraft/resources/RegistryWriteOps.java
+++ net/minecraft/resources/RegistryWriteOps.java
@@ -7,23 +7,23 @@
 import net.minecraft.core.Registry;
 import net.minecraft.core.RegistryAccess;
 import net.minecraft.core.WritableRegistry;
 
 public class RegistryWriteOps<T> extends DelegatingOps<T> {
-   private final RegistryAccess registryHolder;
+   private final RegistryAccess registryAccess;
 
    public static <T> RegistryWriteOps<T> create(DynamicOps<T> var0, RegistryAccess var1) {
       return new RegistryWriteOps(var0, var1);
    }
 
    private RegistryWriteOps(DynamicOps<T> var1, RegistryAccess var2) {
       super(var1);
-      this.registryHolder = var2;
+      this.registryAccess = var2;
    }
 
    protected <E> DataResult<T> encode(E var1, T var2, ResourceKey<? extends Registry<E>> var3, Codec<E> var4) {
-      Optional var5 = this.registryHolder.registry(var3);
+      Optional var5 = this.registryAccess.ownedRegistry(var3);
       if (var5.isPresent()) {
          WritableRegistry var6 = (WritableRegistry)var5.get();
          Optional var7 = var6.getResourceKey(var1);
          if (var7.isPresent()) {
             ResourceKey var8 = (ResourceKey)var7.get();
