--- net/minecraft/resources/ResourceKey.java
+++ net/minecraft/resources/ResourceKey.java
@@ -30,11 +30,11 @@
       this.registryName = var1;
       this.location = var2;
    }
 
    public String toString() {
-      return "ResourceKey[" + this.registryName + " / " + this.location + ']';
+      return "ResourceKey[" + this.registryName + " / " + this.location + "]";
    }
 
    public boolean isFor(ResourceKey<? extends Registry<?>> var1) {
       return this.registryName.equals(var1.location());
    }
