--- net/minecraft/data/Main.java
+++ net/minecraft/data/Main.java
@@ -21,17 +21,20 @@
 import net.minecraft.data.structures.SnbtToNbt;
 import net.minecraft.data.structures.StructureUpdater;
 import net.minecraft.data.tags.BlockTagsProvider;
 import net.minecraft.data.tags.EntityTypeTagsProvider;
 import net.minecraft.data.tags.FluidTagsProvider;
+import net.minecraft.data.tags.GameEventTagsProvider;
 import net.minecraft.data.tags.ItemTagsProvider;
 import net.minecraft.data.worldgen.biome.BiomeReport;
+import net.minecraft.obfuscate.DontObfuscate;
 
 public class Main {
    public Main() {
    }
 
+   @DontObfuscate
    public static void main(String[] var0) throws IOException {
       OptionParser var1 = new OptionParser();
       AbstractOptionSpec var2 = var1.accepts("help", "Show the help menu").forHelp();
       OptionSpecBuilder var3 = var1.accepts("server", "Include server generators");
       OptionSpecBuilder var4 = var1.accepts("client", "Include client generators");
@@ -76,10 +79,11 @@
          var7.addProvider(new ItemTagsProvider(var7, var8));
          var7.addProvider(new EntityTypeTagsProvider(var7));
          var7.addProvider(new RecipeProvider(var7));
          var7.addProvider(new AdvancementProvider(var7));
          var7.addProvider(new LootTableProvider(var7));
+         var7.addProvider(new GameEventTagsProvider(var7));
       }
 
       if (var4) {
          var7.addProvider(new NbtToSnbt(var7));
       }
