--- net/minecraft/world/level/biome/AmbientAdditionsSettings.java
+++ net/minecraft/world/level/biome/AmbientAdditionsSettings.java
@@ -11,12 +11,12 @@
          return var0x.soundEvent;
       }), Codec.DOUBLE.fieldOf("tick_chance").forGetter((var0x) -> {
          return var0x.tickChance;
       })).apply(var0, (BiFunction)(AmbientAdditionsSettings::new));
    });
-   private SoundEvent soundEvent;
-   private double tickChance;
+   private final SoundEvent soundEvent;
+   private final double tickChance;
 
    public AmbientAdditionsSettings(SoundEvent var1, double var2) {
       this.soundEvent = var1;
       this.tickChance = var2;
    }
