--- net/minecraft/world/level/biome/BiomeManager.java
+++ net/minecraft/world/level/biome/BiomeManager.java
@@ -1,11 +1,13 @@
 package net.minecraft.world.level.biome;
 
 import com.google.common.hash.Hashing;
 import net.minecraft.core.BlockPos;
+import net.minecraft.core.QuartPos;
 
 public class BiomeManager {
+   private static final int CHUNK_CENTER_QUART = QuartPos.fromBlock(8);
    private final BiomeManager.NoiseBiomeSource noiseBiomeSource;
    private final long biomeZoomSeed;
    private final BiomeZoomer zoomer;
 
    public BiomeManager(BiomeManager.NoiseBiomeSource var1, long var2, BiomeZoomer var4) {
@@ -24,9 +26,17 @@
 
    public Biome getBiome(BlockPos var1) {
       return this.zoomer.getBiome(this.biomeZoomSeed, var1.getX(), var1.getY(), var1.getZ(), this.noiseBiomeSource);
    }
 
+   public Biome getPrimaryBiomeAtChunk(int var1, int var2) {
+      return this.noiseBiomeSource.getPrimaryBiome(var1, var2);
+   }
+
    public interface NoiseBiomeSource {
       Biome getNoiseBiome(int var1, int var2, int var3);
+
+      default Biome getPrimaryBiome(int var1, int var2) {
+         return this.getNoiseBiome(QuartPos.fromSection(var1) + BiomeManager.CHUNK_CENTER_QUART, 0, QuartPos.fromSection(var2) + BiomeManager.CHUNK_CENTER_QUART);
+      }
    }
 }
