--- net/minecraft/world/level/biome/Biomes.java
+++ net/minecraft/world/level/biome/Biomes.java
@@ -82,10 +82,11 @@
    public static final ResourceKey<Biome> BAMBOO_JUNGLE_HILLS = register("bamboo_jungle_hills");
    public static final ResourceKey<Biome> SOUL_SAND_VALLEY = register("soul_sand_valley");
    public static final ResourceKey<Biome> CRIMSON_FOREST = register("crimson_forest");
    public static final ResourceKey<Biome> WARPED_FOREST = register("warped_forest");
    public static final ResourceKey<Biome> BASALT_DELTAS = register("basalt_deltas");
+   public static final ResourceKey<Biome> DRIPSTONE_CAVES = register("dripstone_caves");
 
    private static ResourceKey<Biome> register(String var0) {
       return ResourceKey.create(Registry.BIOME_REGISTRY, new ResourceLocation(var0));
    }
 }
