--- net/minecraft/world/level/biome/CheckerboardColumnBiomeSource.java
+++ net/minecraft/world/level/biome/CheckerboardColumnBiomeSource.java
@@ -27,9 +27,13 @@
 
    protected Codec<? extends BiomeSource> codec() {
       return CODEC;
    }
 
+   public BiomeSource withSeed(long var1) {
+      return this;
+   }
+
    public Biome getNoiseBiome(int var1, int var2, int var3) {
       return (Biome)((Supplier)this.allowedBiomes.get(Math.floorMod((var1 >> this.bitShift) + (var3 >> this.bitShift), this.allowedBiomes.size()))).get();
    }
 }
