--- net/minecraft/world/level/biome/FixedBiomeSource.java
+++ net/minecraft/world/level/biome/FixedBiomeSource.java
@@ -20,25 +20,29 @@
          return var1;
       });
    }
 
    public FixedBiomeSource(Supplier<Biome> var1) {
-      super((List)ImmutableList.of(var1.get()));
+      super((List)ImmutableList.of((Biome)var1.get()));
       this.biome = var1;
    }
 
    protected Codec<? extends BiomeSource> codec() {
       return CODEC;
    }
 
+   public BiomeSource withSeed(long var1) {
+      return this;
+   }
+
    public Biome getNoiseBiome(int var1, int var2, int var3) {
       return (Biome)this.biome.get();
    }
 
    @Nullable
    public BlockPos findBiomeHorizontal(int var1, int var2, int var3, int var4, int var5, Predicate<Biome> var6, Random var7, boolean var8) {
-      if (var6.test(this.biome.get())) {
+      if (var6.test((Biome)this.biome.get())) {
          return var8 ? new BlockPos(var1, var2, var3) : new BlockPos(var1 - var4 + var7.nextInt(var4 * 2 + 1), var2, var3 - var4 + var7.nextInt(var4 * 2 + 1));
       } else {
          return null;
       }
    }
