--- net/minecraft/world/level/biome/FixedBiomeSource.java
+++ net/minecraft/world/level/biome/FixedBiomeSource.java
@@ -28,10 +28,14 @@
 
    protected Codec<? extends BiomeSource> codec() {
       return CODEC;
    }
 
+   public BiomeSource withSeed(long var1) {
+      return this;
+   }
+
    public Biome getNoiseBiome(int var1, int var2, int var3) {
       return (Biome)this.biome.get();
    }
 
    @Nullable
