--- net/minecraft/world/level/biome/MobSpawnSettings.java
+++ net/minecraft/world/level/biome/MobSpawnSettings.java
@@ -30,11 +30,11 @@
       return var0;
    }, (var0) -> {
       return ImmutableList.of();
    })), ImmutableMap.of(), false);
    public static final MapCodec<MobSpawnSettings> CODEC = RecordCodecBuilder.mapCodec((var0) -> {
-      RecordCodecBuilder var10001 = Codec.FLOAT.optionalFieldOf("creature_spawn_probability", 0.1F).forGetter((var0x) -> {
+      RecordCodecBuilder var10001 = Codec.floatRange(0.0F, 0.9999999F).optionalFieldOf("creature_spawn_probability", 0.1F).forGetter((var0x) -> {
          return var0x.creatureGenerationProbability;
       });
       Codec var10002 = MobCategory.CODEC;
       Codec var10003 = MobSpawnSettings.SpawnerData.CODEC.listOf();
       Logger var10005 = LOGGER;
