--- net/minecraft/world/level/biome/NearestNeighborBiomeZoomer.java
+++ net/minecraft/world/level/biome/NearestNeighborBiomeZoomer.java
@@ -1,12 +1,14 @@
 package net.minecraft.world.level.biome;
 
+import net.minecraft.core.QuartPos;
+
 public enum NearestNeighborBiomeZoomer implements BiomeZoomer {
    INSTANCE;
 
    private NearestNeighborBiomeZoomer() {
    }
 
    public Biome getBiome(long var1, int var3, int var4, int var5, BiomeManager.NoiseBiomeSource var6) {
-      return var6.getNoiseBiome(var3 >> 2, var4 >> 2, var5 >> 2);
+      return var6.getNoiseBiome(QuartPos.fromBlock(var3), QuartPos.fromBlock(var4), QuartPos.fromBlock(var5));
    }
 }
