--- net/minecraft/world/level/biome/OverworldBiomeSource.java
+++ net/minecraft/world/level/biome/OverworldBiomeSource.java
@@ -46,10 +46,14 @@
 
    protected Codec<? extends BiomeSource> codec() {
       return CODEC;
    }
 
+   public BiomeSource withSeed(long var1) {
+      return new OverworldBiomeSource(var1, this.legacyBiomeInitLayer, this.largeBiomes, this.biomes);
+   }
+
    public Biome getNoiseBiome(int var1, int var2, int var3) {
       return this.noiseBiomeLayer.get(this.biomes, var1, var3);
    }
 
    static {
