--- com/mojang/blaze3d/shaders/AbstractUniform.java
+++ com/mojang/blaze3d/shaders/AbstractUniform.java
@@ -1,8 +1,9 @@
 package com.mojang.blaze3d.shaders;
 
 import com.mojang.math.Matrix4f;
+import com.mojang.math.Vector3f;
 
 public class AbstractUniform {
    public AbstractUniform() {
    }
 
@@ -25,8 +26,11 @@
    }
 
    public void set(float[] var1) {
    }
 
+   public void set(Vector3f var1) {
+   }
+
    public void set(Matrix4f var1) {
    }
 }
