--- com/mojang/blaze3d/shaders/AbstractUniform.java
+++ com/mojang/blaze3d/shaders/AbstractUniform.java
@@ -1,8 +1,10 @@
 package com.mojang.blaze3d.shaders;
 
 import com.mojang.math.Matrix4f;
+import com.mojang.math.Vector3f;
+import com.mojang.math.Vector4f;
 
 public class AbstractUniform {
    public AbstractUniform() {
    }
 
@@ -22,11 +24,56 @@
    }
 
    public void setSafe(int var1, int var2, int var3, int var4) {
    }
 
+   public void set(int var1) {
+   }
+
+   public void set(int var1, int var2) {
+   }
+
+   public void set(int var1, int var2, int var3) {
+   }
+
+   public void set(int var1, int var2, int var3, int var4) {
+   }
+
    public void set(float[] var1) {
    }
 
+   public void set(Vector3f var1) {
+   }
+
+   public void set(Vector4f var1) {
+   }
+
+   public void setMat2x2(float var1, float var2, float var3, float var4) {
+   }
+
+   public void setMat2x3(float var1, float var2, float var3, float var4, float var5, float var6) {
+   }
+
+   public void setMat2x4(float var1, float var2, float var3, float var4, float var5, float var6, float var7, float var8) {
+   }
+
+   public void setMat3x2(float var1, float var2, float var3, float var4, float var5, float var6) {
+   }
+
+   public void setMat3x3(float var1, float var2, float var3, float var4, float var5, float var6, float var7, float var8, float var9) {
+   }
+
+   public void setMat3x4(float var1, float var2, float var3, float var4, float var5, float var6, float var7, float var8, float var9, float var10, float var11, float var12) {
+   }
+
+   public void setMat4x2(float var1, float var2, float var3, float var4, float var5, float var6, float var7, float var8) {
+   }
+
+   public void setMat4x3(float var1, float var2, float var3, float var4, float var5, float var6, float var7, float var8, float var9, float var10, float var11, float var12) {
+   }
+
+   public void setMat4x4(float var1, float var2, float var3, float var4, float var5, float var6, float var7, float var8, float var9, float var10, float var11, float var12, float var13, float var14, float var15, float var16) {
+   }
+
    public void set(Matrix4f var1) {
    }
 }
