--- com/mojang/blaze3d/shaders/ProgramManager.java
+++ com/mojang/blaze3d/shaders/ProgramManager.java
@@ -29,11 +29,11 @@
       } else {
          return var0;
       }
    }
 
-   public static void linkProgram(Effect var0) throws IOException {
+   public static void linkProgram(Effect var0) {
       RenderSystem.assertThread(RenderSystem::isOnRenderThread);
       var0.getFragmentProgram().attachToEffect(var0);
       var0.getVertexProgram().attachToEffect(var0);
       GlStateManager.glLinkProgram(var0.getId());
       int var1 = GlStateManager.glGetProgrami(var0.getId(), 35714);
