--- com/mojang/blaze3d/shaders/Uniform.java
+++ com/mojang/blaze3d/shaders/Uniform.java
@@ -204,11 +204,11 @@
       }
 
    }
 
    private void uploadAsInteger() {
-      this.floatValues.clear();
+      this.intValues.rewind();
       switch(this.type) {
       case 0:
          RenderSystem.glUniform1(this.location, this.intValues);
          break;
       case 1:
@@ -225,11 +225,11 @@
       }
 
    }
 
    private void uploadAsFloat() {
-      this.floatValues.clear();
+      this.floatValues.rewind();
       switch(this.type) {
       case 4:
          RenderSystem.glUniform1(this.location, this.floatValues);
          break;
       case 5:
