--- com/mojang/blaze3d/shaders/Uniform.java
+++ com/mojang/blaze3d/shaders/Uniform.java
@@ -1,28 +1,42 @@
 package com.mojang.blaze3d.shaders;
 
 import com.mojang.blaze3d.platform.GlStateManager;
 import com.mojang.blaze3d.systems.RenderSystem;
 import com.mojang.math.Matrix4f;
+import com.mojang.math.Vector3f;
+import com.mojang.math.Vector4f;
 import java.nio.FloatBuffer;
 import java.nio.IntBuffer;
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
 import org.lwjgl.system.MemoryUtil;
 
 public class Uniform extends AbstractUniform implements AutoCloseable {
    private static final Logger LOGGER = LogManager.getLogger();
+   public static final int UT_INT1 = 0;
+   public static final int UT_INT2 = 1;
+   public static final int UT_INT3 = 2;
+   public static final int UT_INT4 = 3;
+   public static final int UT_FLOAT1 = 4;
+   public static final int UT_FLOAT2 = 5;
+   public static final int UT_FLOAT3 = 6;
+   public static final int UT_FLOAT4 = 7;
+   public static final int UT_MAT2 = 8;
+   public static final int UT_MAT3 = 9;
+   public static final int UT_MAT4 = 10;
+   private static final boolean TRANSPOSE_MATRICIES = false;
    private int location;
    private final int count;
    private final int type;
    private final IntBuffer intValues;
    private final FloatBuffer floatValues;
    private final String name;
    private boolean dirty;
-   private final Effect parent;
+   private final Shader parent;
 
-   public Uniform(String var1, int var2, int var3, Effect var4) {
+   public Uniform(String var1, int var2, int var3, Shader var4) {
       this.name = var1;
       this.count = var3;
       this.type = var2;
       this.parent = var4;
       if (var2 <= 3) {
@@ -47,10 +61,14 @@
 
    public static int glGetAttribLocation(int var0, CharSequence var1) {
       return GlStateManager._glGetAttribLocation(var0, var1);
    }
 
+   public static void glBindAttribLocation(int var0, int var1, CharSequence var2) {
+      GlStateManager._glBindAttribLocation(var0, var1, var2);
+   }
+
    public void close() {
       if (this.intValues != null) {
          MemoryUtil.memFree(this.intValues);
       }
 
@@ -93,43 +111,66 @@
 
    public String getName() {
       return this.name;
    }
 
-   public void set(float var1) {
+   public final void set(float var1) {
       this.floatValues.position(0);
       this.floatValues.put(0, var1);
       this.markDirty();
    }
 
-   public void set(float var1, float var2) {
+   public final void set(float var1, float var2) {
       this.floatValues.position(0);
       this.floatValues.put(0, var1);
       this.floatValues.put(1, var2);
       this.markDirty();
    }
 
-   public void set(float var1, float var2, float var3) {
+   public final void set(int var1, float var2) {
       this.floatValues.position(0);
+      this.floatValues.put(var1, var2);
+      this.markDirty();
+   }
+
+   public final void set(float var1, float var2, float var3) {
+      this.floatValues.position(0);
       this.floatValues.put(0, var1);
       this.floatValues.put(1, var2);
       this.floatValues.put(2, var3);
       this.markDirty();
    }
 
-   public void set(float var1, float var2, float var3, float var4) {
+   public final void set(Vector3f var1) {
       this.floatValues.position(0);
+      this.floatValues.put(0, var1.x());
+      this.floatValues.put(1, var1.y());
+      this.floatValues.put(2, var1.z());
+      this.markDirty();
+   }
+
+   public final void set(float var1, float var2, float var3, float var4) {
+      this.floatValues.position(0);
       this.floatValues.put(var1);
       this.floatValues.put(var2);
       this.floatValues.put(var3);
       this.floatValues.put(var4);
       this.floatValues.flip();
       this.markDirty();
    }
 
-   public void setSafe(float var1, float var2, float var3, float var4) {
+   public final void set(Vector4f var1) {
       this.floatValues.position(0);
+      this.floatValues.put(0, var1.x());
+      this.floatValues.put(1, var1.y());
+      this.floatValues.put(2, var1.z());
+      this.floatValues.put(3, var1.w());
+      this.markDirty();
+   }
+
+   public final void setSafe(float var1, float var2, float var3, float var4) {
+      this.floatValues.position(0);
       if (this.type >= 4) {
          this.floatValues.put(0, var1);
       }
 
       if (this.type >= 5) {
@@ -145,11 +186,11 @@
       }
 
       this.markDirty();
    }
 
-   public void setSafe(int var1, int var2, int var3, int var4) {
+   public final void setSafe(int var1, int var2, int var3, int var4) {
       this.intValues.position(0);
       if (this.type >= 0) {
          this.intValues.put(0, var1);
       }
 
@@ -166,23 +207,179 @@
       }
 
       this.markDirty();
    }
 
-   public void set(float[] var1) {
+   public final void set(int var1) {
+      this.intValues.position(0);
+      this.intValues.put(0, var1);
+      this.markDirty();
+   }
+
+   public final void set(int var1, int var2) {
+      this.intValues.position(0);
+      this.intValues.put(0, var1);
+      this.intValues.put(1, var2);
+      this.markDirty();
+   }
+
+   public final void set(int var1, int var2, int var3) {
+      this.intValues.position(0);
+      this.intValues.put(0, var1);
+      this.intValues.put(1, var2);
+      this.intValues.put(2, var3);
+      this.markDirty();
+   }
+
+   public final void set(int var1, int var2, int var3, int var4) {
+      this.intValues.position(0);
+      this.intValues.put(0, var1);
+      this.intValues.put(1, var2);
+      this.intValues.put(2, var3);
+      this.intValues.put(3, var4);
+      this.markDirty();
+   }
+
+   public final void set(float[] var1) {
       if (var1.length < this.count) {
          LOGGER.warn("Uniform.set called with a too-small value array (expected {}, got {}). Ignoring.", this.count, var1.length);
       } else {
          this.floatValues.position(0);
          this.floatValues.put(var1);
          this.floatValues.position(0);
          this.markDirty();
       }
    }
 
-   public void set(Matrix4f var1) {
+   public final void setMat2x2(float var1, float var2, float var3, float var4) {
       this.floatValues.position(0);
+      this.floatValues.put(0, var1);
+      this.floatValues.put(1, var2);
+      this.floatValues.put(2, var3);
+      this.floatValues.put(3, var4);
+      this.markDirty();
+   }
+
+   public final void setMat2x3(float var1, float var2, float var3, float var4, float var5, float var6) {
+      this.floatValues.position(0);
+      this.floatValues.put(0, var1);
+      this.floatValues.put(1, var2);
+      this.floatValues.put(2, var3);
+      this.floatValues.put(3, var4);
+      this.floatValues.put(4, var5);
+      this.floatValues.put(5, var6);
+      this.markDirty();
+   }
+
+   public final void setMat2x4(float var1, float var2, float var3, float var4, float var5, float var6, float var7, float var8) {
+      this.floatValues.position(0);
+      this.floatValues.put(0, var1);
+      this.floatValues.put(1, var2);
+      this.floatValues.put(2, var3);
+      this.floatValues.put(3, var4);
+      this.floatValues.put(4, var5);
+      this.floatValues.put(5, var6);
+      this.floatValues.put(6, var7);
+      this.floatValues.put(7, var8);
+      this.markDirty();
+   }
+
+   public final void setMat3x2(float var1, float var2, float var3, float var4, float var5, float var6) {
+      this.floatValues.position(0);
+      this.floatValues.put(0, var1);
+      this.floatValues.put(1, var2);
+      this.floatValues.put(2, var3);
+      this.floatValues.put(3, var4);
+      this.floatValues.put(4, var5);
+      this.floatValues.put(5, var6);
+      this.markDirty();
+   }
+
+   public final void setMat3x3(float var1, float var2, float var3, float var4, float var5, float var6, float var7, float var8, float var9) {
+      this.floatValues.position(0);
+      this.floatValues.put(0, var1);
+      this.floatValues.put(1, var2);
+      this.floatValues.put(2, var3);
+      this.floatValues.put(3, var4);
+      this.floatValues.put(4, var5);
+      this.floatValues.put(5, var6);
+      this.floatValues.put(6, var7);
+      this.floatValues.put(7, var8);
+      this.floatValues.put(8, var9);
+      this.markDirty();
+   }
+
+   public final void setMat3x4(float var1, float var2, float var3, float var4, float var5, float var6, float var7, float var8, float var9, float var10, float var11, float var12) {
+      this.floatValues.position(0);
+      this.floatValues.put(0, var1);
+      this.floatValues.put(1, var2);
+      this.floatValues.put(2, var3);
+      this.floatValues.put(3, var4);
+      this.floatValues.put(4, var5);
+      this.floatValues.put(5, var6);
+      this.floatValues.put(6, var7);
+      this.floatValues.put(7, var8);
+      this.floatValues.put(8, var9);
+      this.floatValues.put(9, var10);
+      this.floatValues.put(10, var11);
+      this.floatValues.put(11, var12);
+      this.markDirty();
+   }
+
+   public final void setMat4x2(float var1, float var2, float var3, float var4, float var5, float var6, float var7, float var8) {
+      this.floatValues.position(0);
+      this.floatValues.put(0, var1);
+      this.floatValues.put(1, var2);
+      this.floatValues.put(2, var3);
+      this.floatValues.put(3, var4);
+      this.floatValues.put(4, var5);
+      this.floatValues.put(5, var6);
+      this.floatValues.put(6, var7);
+      this.floatValues.put(7, var8);
+      this.markDirty();
+   }
+
+   public final void setMat4x3(float var1, float var2, float var3, float var4, float var5, float var6, float var7, float var8, float var9, float var10, float var11, float var12) {
+      this.floatValues.position(0);
+      this.floatValues.put(0, var1);
+      this.floatValues.put(1, var2);
+      this.floatValues.put(2, var3);
+      this.floatValues.put(3, var4);
+      this.floatValues.put(4, var5);
+      this.floatValues.put(5, var6);
+      this.floatValues.put(6, var7);
+      this.floatValues.put(7, var8);
+      this.floatValues.put(8, var9);
+      this.floatValues.put(9, var10);
+      this.floatValues.put(10, var11);
+      this.floatValues.put(11, var12);
+      this.markDirty();
+   }
+
+   public final void setMat4x4(float var1, float var2, float var3, float var4, float var5, float var6, float var7, float var8, float var9, float var10, float var11, float var12, float var13, float var14, float var15, float var16) {
+      this.floatValues.position(0);
+      this.floatValues.put(0, var1);
+      this.floatValues.put(1, var2);
+      this.floatValues.put(2, var3);
+      this.floatValues.put(3, var4);
+      this.floatValues.put(4, var5);
+      this.floatValues.put(5, var6);
+      this.floatValues.put(6, var7);
+      this.floatValues.put(7, var8);
+      this.floatValues.put(8, var9);
+      this.floatValues.put(9, var10);
+      this.floatValues.put(10, var11);
+      this.floatValues.put(11, var12);
+      this.floatValues.put(12, var13);
+      this.floatValues.put(13, var14);
+      this.floatValues.put(14, var15);
+      this.floatValues.put(15, var16);
+      this.markDirty();
+   }
+
+   public final void set(Matrix4f var1) {
+      this.floatValues.position(0);
       var1.store(this.floatValues);
       this.markDirty();
    }
 
    public void upload() {
@@ -204,11 +401,11 @@
       }
 
    }
 
    private void uploadAsInteger() {
-      this.floatValues.clear();
+      this.intValues.rewind();
       switch(this.type) {
       case 0:
          RenderSystem.glUniform1(this.location, this.intValues);
          break;
       case 1:
@@ -225,11 +422,11 @@
       }
 
    }
 
    private void uploadAsFloat() {
-      this.floatValues.clear();
+      this.floatValues.rewind();
       switch(this.type) {
       case 4:
          RenderSystem.glUniform1(this.location, this.floatValues);
          break;
       case 5:
@@ -259,6 +456,26 @@
       case 10:
          RenderSystem.glUniformMatrix4(this.location, false, this.floatValues);
       }
 
    }
+
+   public int getLocation() {
+      return this.location;
+   }
+
+   public int getCount() {
+      return this.count;
+   }
+
+   public int getType() {
+      return this.type;
+   }
+
+   public IntBuffer getIntBuffer() {
+      return this.intValues;
+   }
+
+   public FloatBuffer getFloatBuffer() {
+      return this.floatValues;
+   }
 }
