--- net/minecraft/network/protocol/login/ClientboundHelloPacket.java
+++ net/minecraft/network/protocol/login/ClientboundHelloPacket.java
@@ -1,35 +1,31 @@
 package net.minecraft.network.protocol.login;
 
-import java.io.IOException;
 import java.security.PublicKey;
 import net.minecraft.network.FriendlyByteBuf;
 import net.minecraft.network.protocol.Packet;
 import net.minecraft.util.Crypt;
 import net.minecraft.util.CryptException;
 
 public class ClientboundHelloPacket implements Packet<ClientLoginPacketListener> {
-   private String serverId;
-   private byte[] publicKey;
-   private byte[] nonce;
+   private final String serverId;
+   private final byte[] publicKey;
+   private final byte[] nonce;
 
-   public ClientboundHelloPacket() {
-   }
-
    public ClientboundHelloPacket(String var1, byte[] var2, byte[] var3) {
       this.serverId = var1;
       this.publicKey = var2;
       this.nonce = var3;
    }
 
-   public void read(FriendlyByteBuf var1) throws IOException {
+   public ClientboundHelloPacket(FriendlyByteBuf var1) {
       this.serverId = var1.readUtf(20);
       this.publicKey = var1.readByteArray();
       this.nonce = var1.readByteArray();
    }
 
-   public void write(FriendlyByteBuf var1) throws IOException {
+   public void write(FriendlyByteBuf var1) {
       var1.writeUtf(this.serverId);
       var1.writeByteArray(this.publicKey);
       var1.writeByteArray(this.nonce);
    }
 
