--- net/minecraft/advancements/AdvancementList.java
+++ net/minecraft/advancements/AdvancementList.java
@@ -1,12 +1,11 @@
 package net.minecraft.advancements;
 
-import com.google.common.base.Function;
-import com.google.common.base.Functions;
 import com.google.common.collect.Maps;
 import com.google.common.collect.Sets;
 import java.util.Collection;
+import java.util.HashMap;
 import java.util.Iterator;
 import java.util.Map;
 import java.util.Set;
 import java.util.Map.Entry;
 import javax.annotation.Nullable;
@@ -23,23 +22,25 @@
 
    public AdvancementList() {
    }
 
    public void add(Map<ResourceLocation, Advancement.Builder> var1) {
-      Function var2 = Functions.forMap(this.advancements, (Object)null);
+      HashMap var2 = Maps.newHashMap(var1);
 
       label42:
-      while(!var1.isEmpty()) {
+      while(!var2.isEmpty()) {
          boolean var3 = false;
-         Iterator var4 = var1.entrySet().iterator();
+         Iterator var4 = var2.entrySet().iterator();
 
          Entry var5;
          while(var4.hasNext()) {
             var5 = (Entry)var4.next();
             ResourceLocation var6 = (ResourceLocation)var5.getKey();
             Advancement.Builder var7 = (Advancement.Builder)var5.getValue();
-            if (var7.canBuild(var2)) {
+            Map var10001 = this.advancements;
+            var10001.getClass();
+            if (var7.canBuild(var10001::get)) {
                Advancement var8 = var7.build(var6);
                this.advancements.put(var6, var8);
                var3 = true;
                var4.remove();
                if (var8.getParent() == null) {
@@ -55,11 +56,11 @@
                }
             }
          }
 
          if (!var3) {
-            var4 = var1.entrySet().iterator();
+            var4 = var2.entrySet().iterator();
 
             while(true) {
                if (!var4.hasNext()) {
                   break label42;
                }
