--- net/minecraft/advancements/Advancement.java
+++ net/minecraft/advancements/Advancement.java
@@ -93,10 +93,14 @@
 
    public Map<String, Criterion> getCriteria() {
       return this.criteria;
    }
 
+   public int getMaxCriteraRequired() {
+      return this.requirements.length;
+   }
+
    public void addChild(Advancement var1) {
       this.children.add(var1);
    }
 
    public ResourceLocation getId() {
@@ -204,10 +208,15 @@
       public Advancement.Builder requirements(RequirementsStrategy var1) {
          this.requirementsStrategy = var1;
          return this;
       }
 
+      public Advancement.Builder requirements(String[][] var1) {
+         this.requirements = var1;
+         return this;
+      }
+
       public boolean canBuild(Function<ResourceLocation, Advancement> var1) {
          if (this.parentId == null) {
             return true;
          } else {
             if (this.parent == null) {
@@ -412,11 +421,11 @@
 
          for(int var5 = 0; var5 < var4.length; ++var5) {
             var4[var5] = new String[var0.readVarInt()];
 
             for(int var6 = 0; var6 < var4[var5].length; ++var6) {
-               var4[var5][var6] = var0.readUtf(32767);
+               var4[var5][var6] = var0.readUtf();
             }
          }
 
          return new Advancement.Builder(var1, var2, AdvancementRewards.EMPTY, var3, var4);
       }
