--- net/minecraft/advancements/CriteriaTriggers.java
+++ net/minecraft/advancements/CriteriaTriggers.java
@@ -86,10 +86,13 @@
    public static final ItemUsedOnBlockTrigger ITEM_USED_ON_BLOCK = (ItemUsedOnBlockTrigger)register(new ItemUsedOnBlockTrigger());
    public static final LootTableTrigger GENERATE_LOOT = (LootTableTrigger)register(new LootTableTrigger());
    public static final ItemPickedUpByEntityTrigger ITEM_PICKED_UP_BY_ENTITY = (ItemPickedUpByEntityTrigger)register(new ItemPickedUpByEntityTrigger());
    public static final PlayerInteractTrigger PLAYER_INTERACTED_WITH_ENTITY = (PlayerInteractTrigger)register(new PlayerInteractTrigger());
 
+   public CriteriaTriggers() {
+   }
+
    private static <T extends CriterionTrigger<?>> T register(T var0) {
       if (CRITERIA.containsKey(var0.getId())) {
          throw new IllegalArgumentException("Duplicate criterion id " + var0.getId());
       } else {
          CRITERIA.put(var0.getId(), var0);
