--- net/minecraft/advancements/DisplayInfo.java
+++ net/minecraft/advancements/DisplayInfo.java
@@ -51,14 +51,35 @@
 
    public Component getDescription() {
       return this.description;
    }
 
+   public ItemStack getIcon() {
+      return this.icon;
+   }
+
+   @Nullable
+   public ResourceLocation getBackground() {
+      return this.background;
+   }
+
    public FrameType getFrame() {
       return this.frame;
    }
 
+   public float getX() {
+      return this.x;
+   }
+
+   public float getY() {
+      return this.y;
+   }
+
+   public boolean shouldShowToast() {
+      return this.showToast;
+   }
+
    public boolean shouldAnnounceChat() {
       return this.announceChat;
    }
 
    public boolean isHidden() {
