--- net/minecraft/advancements/FrameType.java
+++ net/minecraft/advancements/FrameType.java
@@ -23,10 +23,14 @@
 
    public String getName() {
       return this.name;
    }
 
+   public int getTexture() {
+      return this.texture;
+   }
+
    public static FrameType byName(String var0) {
       FrameType[] var1 = values();
       int var2 = var1.length;
 
       for(int var3 = 0; var3 < var2; ++var3) {
@@ -40,6 +44,10 @@
    }
 
    public ChatFormatting getChatColor() {
       return this.chatColor;
    }
+
+   public Component getDisplayName() {
+      return this.displayName;
+   }
 }
