--- net/minecraft/gametest/framework/GameTestSequence.java
+++ net/minecraft/gametest/framework/GameTestSequence.java
@@ -9,19 +9,19 @@
    private long lastTick;
 
    public void tickAndContinue(long var1) {
       try {
          this.tick(var1);
-      } catch (Exception var4) {
+      } catch (GameTestAssertException var4) {
       }
 
    }
 
    public void tickAndFailIfNotComplete(long var1) {
       try {
          this.tick(var1);
-      } catch (Exception var4) {
+      } catch (GameTestAssertException var4) {
          this.parent.fail(var4);
       }
 
    }
 
