--- net/minecraft/gametest/framework/GameTestBatch.java
+++ net/minecraft/gametest/framework/GameTestBatch.java
@@ -8,18 +8,21 @@
 public class GameTestBatch {
    private final String name;
    private final Collection<TestFunction> testFunctions;
    @Nullable
    private final Consumer<ServerLevel> beforeBatchFunction;
+   @Nullable
+   private final Consumer<ServerLevel> afterBatchFunction;
 
-   public GameTestBatch(String var1, Collection<TestFunction> var2, @Nullable Consumer<ServerLevel> var3) {
+   public GameTestBatch(String var1, Collection<TestFunction> var2, @Nullable Consumer<ServerLevel> var3, @Nullable Consumer<ServerLevel> var4) {
       if (var2.isEmpty()) {
          throw new IllegalArgumentException("A GameTestBatch must include at least one TestFunction!");
       } else {
          this.name = var1;
          this.testFunctions = var2;
          this.beforeBatchFunction = var3;
+         this.afterBatchFunction = var4;
       }
    }
 
    public String getName() {
       return this.name;
@@ -33,6 +36,13 @@
       if (this.beforeBatchFunction != null) {
          this.beforeBatchFunction.accept(var1);
       }
 
    }
+
+   public void runAfterBatchFunction(ServerLevel var1) {
+      if (this.afterBatchFunction != null) {
+         this.afterBatchFunction.accept(var1);
+      }
+
+   }
 }
