--- net/minecraft/gametest/framework/LogTestReporter.java
+++ net/minecraft/gametest/framework/LogTestReporter.java
@@ -10,12 +10,15 @@
    public LogTestReporter() {
    }
 
    public void onTestFailed(GameTestInfo var1) {
       if (var1.isRequired()) {
-         LOGGER.error(var1.getTestName() + " failed! " + Util.describeError(var1.getError()));
+         LOGGER.error("{} failed! {}", var1.getTestName(), Util.describeError(var1.getError()));
       } else {
-         LOGGER.warn("(optional) " + var1.getTestName() + " failed. " + Util.describeError(var1.getError()));
+         LOGGER.warn("(optional) {} failed. {}", var1.getTestName(), Util.describeError(var1.getError()));
       }
 
    }
+
+   public void onTestSuccess(GameTestInfo var1) {
+   }
 }
