--- net/minecraft/gametest/framework/TestFunction.java
+++ net/minecraft/gametest/framework/TestFunction.java
@@ -6,10 +6,12 @@
 public class TestFunction {
    private final String batchName;
    private final String testName;
    private final String structureName;
    private final boolean required;
+   private final int maxAttempts;
+   private final int requiredSuccesses;
    private final Consumer<GameTestHelper> function;
    private final int maxTicks;
    private final long setupTicks;
    private final Rotation rotation;
 
@@ -46,6 +48,18 @@
    }
 
    public Rotation getRotation() {
       return this.rotation;
    }
+
+   public boolean isFlaky() {
+      return this.maxAttempts > 1;
+   }
+
+   public int getMaxAttempts() {
+      return this.maxAttempts;
+   }
+
+   public int getRequiredSuccesses() {
+      return this.requiredSuccesses;
+   }
 }
