--- net/minecraft/world/level/levelgen/structure/templatesystem/package-info.java
+++ net/minecraft/world/level/levelgen/structure/templatesystem/package-info.java
@@ -1,0 +1,10 @@
+package net.minecraft.world.level.levelgen.structure.templatesystem;
+
+import javax.annotation.ParametersAreNonnullByDefault;
+import net.minecraft.MethodsReturnNonnullByDefault;
+
+// $FF: synthetic class
+@ParametersAreNonnullByDefault
+@MethodsReturnNonnullByDefault
+interface package-info {
+}
