--- net/minecraft/util/datafix/DataFixTypes.java
+++ net/minecraft/util/datafix/DataFixTypes.java
@@ -12,11 +12,12 @@
    STRUCTURE(References.STRUCTURE),
    STATS(References.STATS),
    SAVED_DATA(References.SAVED_DATA),
    ADVANCEMENTS(References.ADVANCEMENTS),
    POI_CHUNK(References.POI_CHUNK),
-   WORLD_GEN_SETTINGS(References.WORLD_GEN_SETTINGS);
+   WORLD_GEN_SETTINGS(References.WORLD_GEN_SETTINGS),
+   ENTITY_CHUNK(References.ENTITY_CHUNK);
 
    private final DSL.TypeReference type;
 
    private DataFixTypes(DSL.TypeReference var3) {
       this.type = var3;
