--- net/minecraft/network/protocol/PacketUtils.java
+++ net/minecraft/network/protocol/PacketUtils.java
@@ -18,11 +18,11 @@
       if (!var2.isSameThread()) {
          var2.execute(() -> {
             if (var1.getConnection().isConnected()) {
                var0.handle(var1);
             } else {
-               LOGGER.debug("Ignoring packet due to disconnection: " + var0);
+               LOGGER.debug((String)"Ignoring packet due to disconnection: {}", (Object)var0);
             }
 
          });
          throw RunningOnDifferentThreadException.RUNNING_ON_DIFFERENT_THREAD;
       }
