--- net/minecraft/world/entity/npc/VillagerProfession.java
+++ net/minecraft/world/entity/npc/VillagerProfession.java
@@ -41,10 +41,14 @@
       this.requestedItems = var3;
       this.secondaryPoi = var4;
       this.workSound = var5;
    }
 
+   public String getName() {
+      return this.name;
+   }
+
    public PoiType getJobPoiType() {
       return this.jobPoiType;
    }
 
    public ImmutableSet<Item> getRequestedItems() {
