--- net/minecraft/world/entity/npc/WanderingTraderSpawner.java
+++ net/minecraft/world/entity/npc/WanderingTraderSpawner.java
@@ -25,10 +25,17 @@
 import net.minecraft.world.level.biome.Biomes;
 import net.minecraft.world.level.levelgen.Heightmap;
 import net.minecraft.world.level.storage.ServerLevelData;
 
 public class WanderingTraderSpawner implements CustomSpawner {
+   private static final int DEFAULT_TICK_DELAY = 1200;
+   public static final int DEFAULT_SPAWN_DELAY = 24000;
+   private static final int MIN_SPAWN_CHANCE = 25;
+   private static final int MAX_SPAWN_CHANCE = 75;
+   private static final int SPAWN_CHANCE_INCREASE = 25;
+   private static final int SPAWN_ONE_IN_X_CHANCE = 10;
+   private static final int NUMBER_OF_SPAWN_ATTEMPTS = 10;
    private final Random random = new Random();
    private final ServerLevelData serverLevelData;
    private int tickDelay;
    private int spawnDelay;
    private int spawnChance;
@@ -62,11 +69,11 @@
             this.spawnDelay = 24000;
             if (!var1.getGameRules().getBoolean(GameRules.RULE_DOMOBSPAWNING)) {
                return 0;
             } else {
                int var4 = this.spawnChance;
-               this.spawnChance = Mth.clamp(this.spawnChance + 25, 25, 75);
+               this.spawnChance = Mth.clamp((int)(this.spawnChance + 25), (int)25, (int)75);
                this.serverLevelData.setWanderingTraderSpawnChance(this.spawnChance);
                if (this.random.nextInt(100) > var4) {
                   return 0;
                } else if (this.spawn(var1)) {
                   this.spawnChance = 25;
