--- com/mojang/realmsclient/util/RealmsUtil.java
+++ com/mojang/realmsclient/util/RealmsUtil.java
@@ -15,13 +15,19 @@
 import java.util.concurrent.TimeUnit;
 import net.minecraft.client.Minecraft;
 
 public class RealmsUtil {
    private static final YggdrasilAuthenticationService AUTHENTICATION_SERVICE = new YggdrasilAuthenticationService(Minecraft.getInstance().getProxy());
-   private static final MinecraftSessionService SESSION_SERVICE;
+   static final MinecraftSessionService SESSION_SERVICE;
    public static LoadingCache<String, GameProfile> gameProfileCache;
+   private static final int MINUTES = 60;
+   private static final int HOURS = 3600;
+   private static final int DAYS = 86400;
 
+   public RealmsUtil() {
+   }
+
    public static String uuidToName(String var0) throws Exception {
       GameProfile var1 = (GameProfile)gameProfileCache.get(var0);
       return var1.getName();
    }
 
@@ -37,23 +43,28 @@
    public static String convertToAgePresentation(long var0) {
       if (var0 < 0L) {
          return "right now";
       } else {
          long var2 = var0 / 1000L;
+         String var10000;
          if (var2 < 60L) {
-            return (var2 == 1L ? "1 second" : var2 + " seconds") + " ago";
+            var10000 = var2 == 1L ? "1 second" : var2 + " seconds";
+            return var10000 + " ago";
          } else {
             long var4;
             if (var2 < 3600L) {
                var4 = var2 / 60L;
-               return (var4 == 1L ? "1 minute" : var4 + " minutes") + " ago";
+               var10000 = var4 == 1L ? "1 minute" : var4 + " minutes";
+               return var10000 + " ago";
             } else if (var2 < 86400L) {
                var4 = var2 / 3600L;
-               return (var4 == 1L ? "1 hour" : var4 + " hours") + " ago";
+               var10000 = var4 == 1L ? "1 hour" : var4 + " hours";
+               return var10000 + " ago";
             } else {
                var4 = var2 / 86400L;
-               return (var4 == 1L ? "1 day" : var4 + " days") + " ago";
+               var10000 = var4 == 1L ? "1 day" : var4 + " days";
+               return var10000 + " ago";
             }
          }
       }
    }
 
