--- com/mojang/realmsclient/util/RealmsUtil.java
+++ com/mojang/realmsclient/util/RealmsUtil.java
@@ -17,11 +17,17 @@
 
 public class RealmsUtil {
    private static final YggdrasilAuthenticationService AUTHENTICATION_SERVICE = new YggdrasilAuthenticationService(Minecraft.getInstance().getProxy());
    private static final MinecraftSessionService SESSION_SERVICE;
    public static LoadingCache<String, GameProfile> gameProfileCache;
+   private static final int MINUTES = 60;
+   private static final int HOURS = 3600;
+   private static final int DAYS = 86400;
 
+   public RealmsUtil() {
+   }
+
    public static String uuidToName(String var0) throws Exception {
       GameProfile var1 = (GameProfile)gameProfileCache.get(var0);
       return var1.getName();
    }
 
