--- com/mojang/realmsclient/util/RealmsPersistence.java
+++ com/mojang/realmsclient/util/RealmsPersistence.java
@@ -8,18 +8,32 @@
 import java.nio.charset.StandardCharsets;
 import net.minecraft.client.Minecraft;
 import org.apache.commons.io.FileUtils;
 
 public class RealmsPersistence {
+   private static final String FILE_NAME = "realms_persistence.json";
    private static final GuardedSerializer GSON = new GuardedSerializer();
 
+   public RealmsPersistence() {
+   }
+
+   public RealmsPersistence.RealmsPersistenceData read() {
+      return readFile();
+   }
+
+   public void save(RealmsPersistence.RealmsPersistenceData var1) {
+      writeFile(var1);
+   }
+
    public static RealmsPersistence.RealmsPersistenceData readFile() {
       File var0 = getPathToData();
 
       try {
-         return (RealmsPersistence.RealmsPersistenceData)GSON.fromJson(FileUtils.readFileToString(var0, StandardCharsets.UTF_8), RealmsPersistence.RealmsPersistenceData.class);
-      } catch (IOException var2) {
+         String var1 = FileUtils.readFileToString(var0, StandardCharsets.UTF_8);
+         RealmsPersistence.RealmsPersistenceData var2 = (RealmsPersistence.RealmsPersistenceData)GSON.fromJson(var1, RealmsPersistence.RealmsPersistenceData.class);
+         return var2 != null ? var2 : new RealmsPersistence.RealmsPersistenceData();
+      } catch (IOException var3) {
          return new RealmsPersistence.RealmsPersistenceData();
       }
    }
 
    public static void writeFile(RealmsPersistence.RealmsPersistenceData var0) {
