--- com/mojang/realmsclient/util/TextRenderingUtils.java
+++ com/mojang/realmsclient/util/TextRenderingUtils.java
@@ -7,10 +7,13 @@
 import java.util.Iterator;
 import java.util.List;
 import java.util.Objects;
 
 public class TextRenderingUtils {
+   private TextRenderingUtils() {
+   }
+
    @VisibleForTesting
    protected static List<String> lineBreak(String var0) {
       return Arrays.asList(var0.split("\\n"));
    }
 
@@ -35,11 +38,11 @@
          Iterator var8 = var7.iterator();
 
          while(var8.hasNext()) {
             String var9 = (String)var8.next();
             if ("%link".equals(var9)) {
-               var6.add(var1.get(var2++));
+               var6.add((TextRenderingUtils.LineSegment)var1.get(var2++));
             } else {
                var6.add(TextRenderingUtils.LineSegment.text(var9));
             }
          }
 
@@ -104,11 +107,11 @@
       public int hashCode() {
          return Objects.hash(new Object[]{this.fullText, this.linkTitle, this.linkUrl});
       }
 
       public String toString() {
-         return "Segment{fullText='" + this.fullText + '\'' + ", linkTitle='" + this.linkTitle + '\'' + ", linkUrl='" + this.linkUrl + '\'' + '}';
+         return "Segment{fullText='" + this.fullText + "', linkTitle='" + this.linkTitle + "', linkUrl='" + this.linkUrl + "'}";
       }
 
       public String renderedText() {
          return this.isLink() ? this.linkTitle : this.fullText;
       }
@@ -136,16 +139,20 @@
    }
 
    public static class Line {
       public final List<TextRenderingUtils.LineSegment> segments;
 
+      Line(TextRenderingUtils.LineSegment... var1) {
+         this(Arrays.asList(var1));
+      }
+
       Line(List<TextRenderingUtils.LineSegment> var1) {
          this.segments = var1;
       }
 
       public String toString() {
-         return "Line{segments=" + this.segments + '}';
+         return "Line{segments=" + this.segments + "}";
       }
 
       public boolean equals(Object var1) {
          if (this == var1) {
             return true;
