--- com/mojang/realmsclient/util/TextRenderingUtils.java
+++ com/mojang/realmsclient/util/TextRenderingUtils.java
@@ -7,10 +7,13 @@
 import java.util.Iterator;
 import java.util.List;
 import java.util.Objects;
 
 public class TextRenderingUtils {
+   private TextRenderingUtils() {
+   }
+
    @VisibleForTesting
    protected static List<String> lineBreak(String var0) {
       return Arrays.asList(var0.split("\\n"));
    }
 
@@ -136,10 +139,14 @@
    }
 
    public static class Line {
       public final List<TextRenderingUtils.LineSegment> segments;
 
+      Line(TextRenderingUtils.LineSegment... var1) {
+         this(Arrays.asList(var1));
+      }
+
       Line(List<TextRenderingUtils.LineSegment> var1) {
          this.segments = var1;
       }
 
       public String toString() {
