--- com/mojang/realmsclient/util/UploadTokenCache.java
+++ com/mojang/realmsclient/util/UploadTokenCache.java
@@ -4,10 +4,13 @@
 import it.unimi.dsi.fastutil.longs.Long2ObjectOpenHashMap;
 
 public class UploadTokenCache {
    private static final Long2ObjectMap<String> TOKEN_CACHE = new Long2ObjectOpenHashMap();
 
+   public UploadTokenCache() {
+   }
+
    public static String get(long var0) {
       return (String)TOKEN_CACHE.get(var0);
    }
 
    public static void invalidate(long var0) {
