--- net/minecraft/commands/arguments/item/FunctionArgument.java
+++ net/minecraft/commands/arguments/item/FunctionArgument.java
@@ -59,17 +59,17 @@
             }
          };
       }
    }
 
-   private static CommandFunction getFunction(CommandContext<CommandSourceStack> var0, ResourceLocation var1) throws CommandSyntaxException {
+   static CommandFunction getFunction(CommandContext<CommandSourceStack> var0, ResourceLocation var1) throws CommandSyntaxException {
       return (CommandFunction)((CommandSourceStack)var0.getSource()).getServer().getFunctions().get(var1).orElseThrow(() -> {
          return ERROR_UNKNOWN_FUNCTION.create(var1.toString());
       });
    }
 
-   private static Tag<CommandFunction> getFunctionTag(CommandContext<CommandSourceStack> var0, ResourceLocation var1) throws CommandSyntaxException {
+   static Tag<CommandFunction> getFunctionTag(CommandContext<CommandSourceStack> var0, ResourceLocation var1) throws CommandSyntaxException {
       Tag var2 = ((CommandSourceStack)var0.getSource()).getServer().getFunctions().getTag(var1);
       if (var2 == null) {
          throw ERROR_UNKNOWN_TAG.create(var1.toString());
       } else {
          return var2;
