--- net/minecraft/world/level/block/piston/PistonHeadBlock.java
+++ net/minecraft/world/level/block/piston/PistonHeadBlock.java
@@ -27,16 +27,20 @@
 import net.minecraft.world.phys.shapes.VoxelShape;
 
 public class PistonHeadBlock extends DirectionalBlock {
    public static final EnumProperty<PistonType> TYPE;
    public static final BooleanProperty SHORT;
+   public static final float PLATFORM = 4.0F;
    protected static final VoxelShape EAST_AABB;
    protected static final VoxelShape WEST_AABB;
    protected static final VoxelShape SOUTH_AABB;
    protected static final VoxelShape NORTH_AABB;
    protected static final VoxelShape UP_AABB;
    protected static final VoxelShape DOWN_AABB;
+   protected static final float AABB_OFFSET = 2.0F;
+   protected static final float EDGE_MIN = 6.0F;
+   protected static final float EDGE_MAX = 10.0F;
    protected static final VoxelShape UP_ARM_AABB;
    protected static final VoxelShape DOWN_ARM_AABB;
    protected static final VoxelShape SOUTH_ARM_AABB;
    protected static final VoxelShape NORTH_ARM_AABB;
    protected static final VoxelShape EAST_ARM_AABB;
@@ -93,11 +97,11 @@
       Block var3 = var1.getValue(TYPE) == PistonType.DEFAULT ? Blocks.PISTON : Blocks.STICKY_PISTON;
       return var2.is(var3) && (Boolean)var2.getValue(PistonBaseBlock.EXTENDED) && var2.getValue(FACING) == var1.getValue(FACING);
    }
 
    public void playerWillDestroy(Level var1, BlockPos var2, BlockState var3, Player var4) {
-      if (!var1.isClientSide && var4.abilities.instabuild) {
+      if (!var1.isClientSide && var4.getAbilities().instabuild) {
          BlockPos var5 = var2.relative(((Direction)var3.getValue(FACING)).getOpposite());
          if (this.isFittingBase(var3, var1.getBlockState(var5))) {
             var1.destroyBlock(var5, false);
          }
       }
