--- com/mojang/blaze3d/platform/ClipboardManager.java
+++ com/mojang/blaze3d/platform/ClipboardManager.java
@@ -8,10 +8,11 @@
 import org.lwjgl.glfw.GLFWErrorCallback;
 import org.lwjgl.glfw.GLFWErrorCallbackI;
 import org.lwjgl.system.MemoryUtil;
 
 public class ClipboardManager {
+   public static final int FORMAT_UNAVAILABLE = 65545;
    private final ByteBuffer clipboardScratchBuffer = BufferUtils.createByteBuffer(8192);
 
    public ClipboardManager() {
    }
 
