--- com/mojang/blaze3d/platform/DebugMemoryUntracker.java
+++ com/mojang/blaze3d/platform/DebugMemoryUntracker.java
@@ -23,10 +23,13 @@
       } catch (NoSuchMethodException | NoSuchFieldException | IllegalAccessException | ClassNotFoundException var5) {
          throw new RuntimeException(var5);
       }
    });
 
+   public DebugMemoryUntracker() {
+   }
+
    public static void untrack(long var0) {
       if (UNTRACK != null) {
          try {
             UNTRACK.invoke(var0);
          } catch (Throwable var3) {
