--- com/mojang/blaze3d/platform/GlDebug.java
+++ com/mojang/blaze3d/platform/GlDebug.java
@@ -35,11 +35,11 @@
 
    private static String printUnknownToken(int var0) {
       return "Unknown (0x" + Integer.toHexString(var0).toUpperCase() + ")";
    }
 
-   private static String sourceToString(int var0) {
+   public static String sourceToString(int var0) {
       switch(var0) {
       case 33350:
          return "API";
       case 33351:
          return "WINDOW SYSTEM";
@@ -54,11 +54,11 @@
       default:
          return printUnknownToken(var0);
       }
    }
 
-   private static String typeToString(int var0) {
+   public static String typeToString(int var0) {
       switch(var0) {
       case 33356:
          return "ERROR";
       case 33357:
          return "DEPRECATED BEHAVIOR";
@@ -75,11 +75,11 @@
       default:
          return printUnknownToken(var0);
       }
    }
 
-   private static String severityToString(int var0) {
+   public static String severityToString(int var0) {
       switch(var0) {
       case 33387:
          return "NOTIFICATION";
       case 37190:
          return "HIGH";
@@ -162,13 +162,10 @@
       setup(2, "GL11.GL_LINE_LOOP");
       setup(3, "GL11.GL_LINE_STRIP");
       setup(4, "GL11.GL_TRIANGLES");
       setup(5, "GL11.GL_TRIANGLE_STRIP");
       setup(6, "GL11.GL_TRIANGLE_FAN");
-      setup(7, "GL11.GL_QUADS");
-      setup(8, "GL11.GL_QUAD_STRIP");
-      setup(9, "GL11.GL_POLYGON");
       setup(0, "GL11.GL_ZERO");
       setup(1, "GL11.GL_ONE");
       setup(768, "GL11.GL_SRC_COLOR");
       setup(769, "GL11.GL_ONE_MINUS_SRC_COLOR");
       setup(770, "GL11.GL_SRC_ALPHA");
