--- com/mojang/blaze3d/platform/GlUtil.java
+++ com/mojang/blaze3d/platform/GlUtil.java
@@ -1,8 +1,26 @@
 package com.mojang.blaze3d.platform;
 
+import java.nio.Buffer;
+import java.nio.ByteBuffer;
+import org.lwjgl.system.MemoryUtil;
+
 public class GlUtil {
+   public GlUtil() {
+   }
+
+   public static void populateSnooperWithOpenGL(SnooperAccess var0) {
+   }
+
+   public static ByteBuffer allocateMemory(int var0) {
+      return MemoryUtil.memAlloc(var0);
+   }
+
+   public static void freeMemory(Buffer var0) {
+      MemoryUtil.memFree(var0);
+   }
+
    public static String getVendor() {
       return GlStateManager._getString(7936);
    }
 
    public static String getCpuInfo() {
