--- com/mojang/blaze3d/platform/InputConstants.java
+++ com/mojang/blaze3d/platform/InputConstants.java
@@ -25,11 +25,11 @@
 import org.lwjgl.glfw.GLFWMouseButtonCallbackI;
 import org.lwjgl.glfw.GLFWScrollCallbackI;
 
 public class InputConstants {
    @Nullable
-   private static final MethodHandle glfwRawMouseMotionSupported;
+   private static final MethodHandle GLFW_RAW_MOUSE_MOTION_SUPPORTED;
    private static final int GLFW_RAW_MOUSE_MOTION;
    public static final InputConstants.Key UNKNOWN;
 
    public static InputConstants.Key getKey(int var0, int var1) {
       return var0 == -1 ? InputConstants.Type.SCANCODE.getOrCreate(var1) : InputConstants.Type.KEYSYM.getOrCreate(var0);
@@ -75,11 +75,11 @@
       GLFW.glfwSetInputMode(var0, 208897, var2);
    }
 
    public static boolean isRawMouseInputSupported() {
       try {
-         return glfwRawMouseMotionSupported != null && glfwRawMouseMotionSupported.invokeExact();
+         return GLFW_RAW_MOUSE_MOTION_SUPPORTED != null && GLFW_RAW_MOUSE_MOTION_SUPPORTED.invokeExact();
       } catch (Throwable var1) {
          throw new RuntimeException(var1);
       }
    }
 
@@ -103,11 +103,11 @@
       } catch (NoSuchFieldException | NoSuchMethodException var5) {
       } catch (Throwable var6) {
          throw new RuntimeException(var6);
       }
 
-      glfwRawMouseMotionSupported = var2;
+      GLFW_RAW_MOUSE_MOTION_SUPPORTED = var2;
       GLFW_RAW_MOUSE_MOTION = var3;
       UNKNOWN = InputConstants.Type.KEYSYM.getOrCreate(-1);
    }
 
    public static final class Key {
