--- com/mojang/blaze3d/platform/Lighting.java
+++ com/mojang/blaze3d/platform/Lighting.java
@@ -9,21 +9,10 @@
    private static final Vector3f DIFFUSE_LIGHT_0 = (Vector3f)Util.make(new Vector3f(0.2F, 1.0F, -0.7F), Vector3f::normalize);
    private static final Vector3f DIFFUSE_LIGHT_1 = (Vector3f)Util.make(new Vector3f(-0.2F, 1.0F, 0.7F), Vector3f::normalize);
    private static final Vector3f NETHER_DIFFUSE_LIGHT_0 = (Vector3f)Util.make(new Vector3f(0.2F, 1.0F, -0.7F), Vector3f::normalize);
    private static final Vector3f NETHER_DIFFUSE_LIGHT_1 = (Vector3f)Util.make(new Vector3f(-0.2F, -1.0F, 0.7F), Vector3f::normalize);
 
-   public static void turnBackOn() {
-      RenderSystem.enableLighting();
-      RenderSystem.enableColorMaterial();
-      RenderSystem.colorMaterial(1032, 5634);
-   }
-
-   public static void turnOff() {
-      RenderSystem.disableLighting();
-      RenderSystem.disableColorMaterial();
-   }
-
    public static void setupNetherLevel(Matrix4f var0) {
       RenderSystem.setupLevelDiffuseLighting(NETHER_DIFFUSE_LIGHT_0, NETHER_DIFFUSE_LIGHT_1, var0);
    }
 
    public static void setupLevel(Matrix4f var0) {
