--- com/mojang/blaze3d/platform/NativeImage.java
+++ com/mojang/blaze3d/platform/NativeImage.java
@@ -135,22 +135,10 @@
 
          return var8;
       }
    }
 
-   private static void setClamp(boolean var0) {
-      RenderSystem.assertThread(RenderSystem::isOnRenderThreadOrInit);
-      if (var0) {
-         GlStateManager._texParameter(3553, 10242, 10496);
-         GlStateManager._texParameter(3553, 10243, 10496);
-      } else {
-         GlStateManager._texParameter(3553, 10242, 10497);
-         GlStateManager._texParameter(3553, 10243, 10497);
-      }
-
-   }
-
    private static void setFilter(boolean var0, boolean var1) {
       RenderSystem.assertThread(RenderSystem::isOnRenderThreadOrInit);
       if (var0) {
          GlStateManager._texParameter(3553, 10241, var1 ? 9987 : 9729);
          GlStateManager._texParameter(3553, 10240, 9729);
@@ -271,11 +259,10 @@
 
    private void _upload(int var1, int var2, int var3, int var4, int var5, int var6, int var7, boolean var8, boolean var9, boolean var10, boolean var11) {
       RenderSystem.assertThread(RenderSystem::isOnRenderThreadOrInit);
       this.checkAllocated();
       setFilter(var8, var10);
-      setClamp(var9);
       if (var6 == this.getWidth()) {
          GlStateManager._pixelStore(3314, 0);
       } else {
          GlStateManager._pixelStore(3314, this.getWidth());
       }
