--- com/mojang/blaze3d/platform/PngInfo.java
+++ com/mojang/blaze3d/platform/PngInfo.java
@@ -149,10 +149,11 @@
    private static PngInfo.StbReader createCallbacks(InputStream var0) {
       return (PngInfo.StbReader)(var0 instanceof FileInputStream ? new PngInfo.StbReaderSeekableByteChannel(((FileInputStream)var0).getChannel()) : new PngInfo.StbReaderBufferedChannel(Channels.newChannel(var0)));
    }
 
    static class StbReaderBufferedChannel extends PngInfo.StbReader {
+      private static final int START_BUFFER_SIZE = 128;
       private final ReadableByteChannel channel;
       private long readBufferAddress;
       private int bufferSize;
       private int read;
       private int consumed;
