--- com/mojang/blaze3d/platform/TextureUtil.java
+++ com/mojang/blaze3d/platform/TextureUtil.java
@@ -8,10 +8,11 @@
 import java.nio.IntBuffer;
 import java.nio.channels.Channels;
 import java.nio.channels.FileChannel;
 import java.nio.channels.ReadableByteChannel;
 import java.util.concurrent.ThreadLocalRandom;
+import javax.annotation.Nullable;
 import net.minecraft.SharedConstants;
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
 import org.lwjgl.opengl.GL11;
 import org.lwjgl.system.MemoryUtil;
@@ -94,10 +95,11 @@
       }
 
       return var1;
    }
 
+   @Nullable
    public static String readResourceAsString(InputStream var0) {
       RenderSystem.assertThread(RenderSystem::isOnRenderThread);
       ByteBuffer var1 = null;
 
       try {
@@ -124,11 +126,9 @@
       GL11.glPixelStorei(3314, 0);
       GL11.glPixelStorei(3315, 0);
       GL11.glPixelStorei(3316, 0);
       GL11.glPixelStorei(3317, 4);
       GL11.glTexImage2D(3553, 0, 6408, var1, var2, 0, 32993, 33639, var0);
-      GL11.glTexParameteri(3553, 10242, 10497);
-      GL11.glTexParameteri(3553, 10243, 10497);
       GL11.glTexParameteri(3553, 10240, 9728);
       GL11.glTexParameteri(3553, 10241, 9729);
    }
 }
