--- com/mojang/blaze3d/platform/Window.java
+++ com/mojang/blaze3d/platform/Window.java
@@ -71,13 +71,14 @@
       this.windowedWidth = this.width = var3.width > 0 ? var3.width : 1;
       this.windowedHeight = this.height = var3.height > 0 ? var3.height : 1;
       GLFW.glfwDefaultWindowHints();
       GLFW.glfwWindowHint(139265, 196609);
       GLFW.glfwWindowHint(139275, 221185);
-      GLFW.glfwWindowHint(139266, 2);
-      GLFW.glfwWindowHint(139267, 0);
-      GLFW.glfwWindowHint(139272, 0);
+      GLFW.glfwWindowHint(139266, 3);
+      GLFW.glfwWindowHint(139267, 2);
+      GLFW.glfwWindowHint(139272, 204801);
+      GLFW.glfwWindowHint(139270, 1);
       this.window = GLFW.glfwCreateWindow(this.width, this.height, var5, this.fullscreen && var7 != null ? var7.getMonitor() : 0L, 0L);
       if (var7 != null) {
          VideoMode var8 = var7.getPreferredVidMode(this.fullscreen ? this.preferredFullscreenVideoMode : Optional.empty());
          this.windowedX = this.x = var7.getX() + var8.getWidth() / 2 - this.width / 2;
          this.windowedY = this.y = var7.getY() + var8.getHeight() / 2 - this.height / 2;
@@ -449,10 +450,18 @@
 
    public int getHeight() {
       return this.framebufferHeight;
    }
 
+   public void setWidth(int var1) {
+      this.framebufferWidth = var1;
+   }
+
+   public void setHeight(int var1) {
+      this.framebufferHeight = var1;
+   }
+
    public int getScreenWidth() {
       return this.width;
    }
 
    public int getScreenHeight() {
