--- net/minecraft/world/level/levelgen/structure/BoundingBox.java
+++ net/minecraft/world/level/levelgen/structure/BoundingBox.java
@@ -54,19 +54,10 @@
 
    public static BoundingBox createProper(int var0, int var1, int var2, int var3, int var4, int var5) {
       return new BoundingBox(Math.min(var0, var3), Math.min(var1, var4), Math.min(var2, var5), Math.max(var0, var3), Math.max(var1, var4), Math.max(var2, var5));
    }
 
-   public BoundingBox(BoundingBox var1) {
-      this.x0 = var1.x0;
-      this.y0 = var1.y0;
-      this.z0 = var1.z0;
-      this.x1 = var1.x1;
-      this.y1 = var1.y1;
-      this.z1 = var1.z1;
-   }
-
    public BoundingBox(int var1, int var2, int var3, int var4, int var5, int var6) {
       this.x0 = var1;
       this.y0 = var2;
       this.z0 = var3;
       this.x1 = var4;
@@ -81,19 +72,10 @@
       this.x1 = Math.max(var1.getX(), var2.getX());
       this.y1 = Math.max(var1.getY(), var2.getY());
       this.z1 = Math.max(var1.getZ(), var2.getZ());
    }
 
-   public BoundingBox(int var1, int var2, int var3, int var4) {
-      this.x0 = var1;
-      this.z0 = var2;
-      this.x1 = var3;
-      this.z1 = var4;
-      this.y0 = 1;
-      this.y1 = 512;
-   }
-
    public boolean intersects(BoundingBox var1) {
       return this.x1 >= var1.x0 && this.x0 <= var1.x1 && this.z1 >= var1.z0 && this.z0 <= var1.z1 && this.y1 >= var1.y0 && this.y0 <= var1.y1;
    }
 
    public boolean intersects(int var1, int var2, int var3, int var4) {
