--- net/minecraft/world/level/levelgen/structure/EndCityPieces.java
+++ net/minecraft/world/level/levelgen/structure/EndCityPieces.java
@@ -280,11 +280,10 @@
                RandomizableContainerBlockEntity.setLootTable(var3, var4, var6, BuiltInLootTables.END_CITY_TREASURE);
             }
          } else if (var1.startsWith("Sentry")) {
             Shulker var7 = (Shulker)EntityType.SHULKER.create(var3.getLevel());
             var7.setPos((double)var2.getX() + 0.5D, (double)var2.getY() + 0.5D, (double)var2.getZ() + 0.5D);
-            var7.setAttachPosition(var2);
             var3.addFreshEntity(var7);
          } else if (var1.startsWith("Elytra")) {
             ItemFrame var8 = new ItemFrame(var3.getLevel(), var2, this.rotation.rotate(Direction.SOUTH));
             var8.setItem(new ItemStack(Items.ELYTRA), false);
             var3.addFreshEntity(var8);
