--- net/minecraft/world/level/levelgen/structure/StructureFeatureIndexSavedData.java
+++ net/minecraft/world/level/levelgen/structure/StructureFeatureIndexSavedData.java
@@ -4,22 +4,28 @@
 import it.unimi.dsi.fastutil.longs.LongSet;
 import net.minecraft.nbt.CompoundTag;
 import net.minecraft.world.level.saveddata.SavedData;
 
 public class StructureFeatureIndexSavedData extends SavedData {
-   private LongSet all = new LongOpenHashSet();
-   private LongSet remaining = new LongOpenHashSet();
+   private static final String TAG_REMAINING_INDEXES = "Remaining";
+   private static final String TAG_All_INDEXES = "All";
+   private final LongSet all;
+   private final LongSet remaining;
 
-   public StructureFeatureIndexSavedData(String var1) {
-      super(var1);
+   private StructureFeatureIndexSavedData(LongSet var1, LongSet var2) {
+      this.all = var1;
+      this.remaining = var2;
    }
 
-   public void load(CompoundTag var1) {
-      this.all = new LongOpenHashSet(var1.getLongArray("All"));
-      this.remaining = new LongOpenHashSet(var1.getLongArray("Remaining"));
+   public StructureFeatureIndexSavedData() {
+      this(new LongOpenHashSet(), new LongOpenHashSet());
    }
 
+   public static StructureFeatureIndexSavedData load(CompoundTag var0) {
+      return new StructureFeatureIndexSavedData(new LongOpenHashSet(var0.getLongArray("All")), new LongOpenHashSet(var0.getLongArray("Remaining")));
+   }
+
    public CompoundTag save(CompoundTag var1) {
       var1.putLongArray("All", this.all.toLongArray());
       var1.putLongArray("Remaining", this.remaining.toLongArray());
       return var1;
    }
