--- net/minecraft/world/level/levelgen/structure/StructurePieceAccessor.java
+++ net/minecraft/world/level/levelgen/structure/StructurePieceAccessor.java
@@ -1,0 +1,10 @@
+package net.minecraft.world.level.levelgen.structure;
+
+import javax.annotation.Nullable;
+
+public interface StructurePieceAccessor {
+   void addPiece(StructurePiece var1);
+
+   @Nullable
+   StructurePiece findCollisionPiece(BoundingBox var1);
+}
