--- net/minecraft/world/entity/monster/hoglin/HoglinBase.java
+++ net/minecraft/world/entity/monster/hoglin/HoglinBase.java
@@ -4,10 +4,12 @@
 import net.minecraft.world.entity.LivingEntity;
 import net.minecraft.world.entity.ai.attributes.Attributes;
 import net.minecraft.world.phys.Vec3;
 
 public interface HoglinBase {
+   int ATTACK_ANIMATION_DURATION = 10;
+
    int getAttackAnimationRemainingTicks();
 
    static boolean hurtAndThrowTarget(LivingEntity var0, LivingEntity var1) {
       float var3 = (float)var0.getAttributeValue(Attributes.ATTACK_DAMAGE);
       float var2;
