--- net/minecraft/world/entity/ai/memory/ExpirableValue.java
+++ net/minecraft/world/entity/ai/memory/ExpirableValue.java
@@ -35,11 +35,11 @@
    public boolean hasExpired() {
       return this.timeToLive <= 0L;
    }
 
    public String toString() {
-      return this.value.toString() + (this.canExpire() ? " (ttl: " + this.timeToLive + ")" : "");
+      return this.value + (this.canExpire() ? " (ttl: " + this.timeToLive + ")" : "");
    }
 
    public boolean canExpire() {
       return this.timeToLive != Long.MAX_VALUE;
    }
