--- net/minecraft/world/entity/ai/memory/WalkTarget.java
+++ net/minecraft/world/entity/ai/memory/WalkTarget.java
@@ -1,9 +1,11 @@
 package net.minecraft.world.entity.ai.memory;
 
 import net.minecraft.core.BlockPos;
+import net.minecraft.world.entity.Entity;
 import net.minecraft.world.entity.ai.behavior.BlockPosTracker;
+import net.minecraft.world.entity.ai.behavior.EntityTracker;
 import net.minecraft.world.entity.ai.behavior.PositionTracker;
 import net.minecraft.world.phys.Vec3;
 
 public class WalkTarget {
    private final PositionTracker target;
@@ -16,10 +18,14 @@
 
    public WalkTarget(Vec3 var1, float var2, int var3) {
       this((PositionTracker)(new BlockPosTracker(new BlockPos(var1))), var2, var3);
    }
 
+   public WalkTarget(Entity var1, float var2, int var3) {
+      this((PositionTracker)(new EntityTracker(var1, false)), var2, var3);
+   }
+
    public WalkTarget(PositionTracker var1, float var2, int var3) {
       this.target = var1;
       this.speedModifier = var2;
       this.closeEnoughDist = var3;
    }
