--- net/minecraft/data/recipes/ShapelessRecipeBuilder.java
+++ net/minecraft/data/recipes/ShapelessRecipeBuilder.java
@@ -20,11 +20,11 @@
 import net.minecraft.world.item.crafting.RecipeSerializer;
 import net.minecraft.world.level.ItemLike;
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
 
-public class ShapelessRecipeBuilder {
+public class ShapelessRecipeBuilder implements RecipeBuilder {
    private static final Logger LOGGER = LogManager.getLogger();
    private final Item result;
    private final int count;
    private final List<Ingredient> ingredients = Lists.newArrayList();
    private final Advancement.Builder advancement = Advancement.Builder.advancement();
