--- net/minecraft/data/recipes/SimpleCookingRecipeBuilder.java
+++ net/minecraft/data/recipes/SimpleCookingRecipeBuilder.java
@@ -36,23 +36,36 @@
 
    public static SimpleCookingRecipeBuilder cooking(Ingredient var0, ItemLike var1, float var2, int var3, SimpleCookingSerializer<?> var4) {
       return new SimpleCookingRecipeBuilder(var1, var0, var2, var3, var4);
    }
 
+   public static SimpleCookingRecipeBuilder campfireCooking(Ingredient var0, ItemLike var1, float var2, int var3) {
+      return cooking(var0, var1, var2, var3, RecipeSerializer.CAMPFIRE_COOKING_RECIPE);
+   }
+
    public static SimpleCookingRecipeBuilder blasting(Ingredient var0, ItemLike var1, float var2, int var3) {
       return cooking(var0, var1, var2, var3, RecipeSerializer.BLASTING_RECIPE);
    }
 
    public static SimpleCookingRecipeBuilder smelting(Ingredient var0, ItemLike var1, float var2, int var3) {
       return cooking(var0, var1, var2, var3, RecipeSerializer.SMELTING_RECIPE);
    }
 
+   public static SimpleCookingRecipeBuilder smoking(Ingredient var0, ItemLike var1, float var2, int var3) {
+      return cooking(var0, var1, var2, var3, RecipeSerializer.SMOKING_RECIPE);
+   }
+
    public SimpleCookingRecipeBuilder unlockedBy(String var1, CriterionTriggerInstance var2) {
       this.advancement.addCriterion(var1, var2);
       return this;
    }
 
+   public SimpleCookingRecipeBuilder group(String var1) {
+      this.group = var1;
+      return this;
+   }
+
    public void save(Consumer<FinishedRecipe> var1) {
       this.save(var1, Registry.ITEM.getKey(this.result));
    }
 
    public void save(Consumer<FinishedRecipe> var1, String var2) {
