--- net/minecraft/data/recipes/SpecialRecipeBuilder.java
+++ net/minecraft/data/recipes/SpecialRecipeBuilder.java
@@ -6,11 +6,11 @@
 import net.minecraft.resources.ResourceLocation;
 import net.minecraft.world.item.crafting.RecipeSerializer;
 import net.minecraft.world.item.crafting.SimpleRecipeSerializer;
 
 public class SpecialRecipeBuilder {
-   private final SimpleRecipeSerializer<?> serializer;
+   final SimpleRecipeSerializer<?> serializer;
 
    public SpecialRecipeBuilder(SimpleRecipeSerializer<?> var1) {
       this.serializer = var1;
    }
 
