--- net/minecraft/network/protocol/status/ServerboundPingRequestPacket.java
+++ net/minecraft/network/protocol/status/ServerboundPingRequestPacket.java
@@ -1,22 +1,18 @@
 package net.minecraft.network.protocol.status;
 
-import java.io.IOException;
 import net.minecraft.network.FriendlyByteBuf;
 import net.minecraft.network.protocol.Packet;
 
 public class ServerboundPingRequestPacket implements Packet<ServerStatusPacketListener> {
-   private long time;
+   private final long time;
 
-   public ServerboundPingRequestPacket() {
-   }
-
-   public void read(FriendlyByteBuf var1) throws IOException {
+   public ServerboundPingRequestPacket(FriendlyByteBuf var1) {
       this.time = var1.readLong();
    }
 
-   public void write(FriendlyByteBuf var1) throws IOException {
+   public void write(FriendlyByteBuf var1) {
       var1.writeLong(this.time);
    }
 
    public void handle(ServerStatusPacketListener var1) {
       var1.handlePingRequest(this);
