--- net/minecraft/network/protocol/status/ServerboundStatusRequestPacket.java
+++ net/minecraft/network/protocol/status/ServerboundStatusRequestPacket.java
@@ -1,19 +1,18 @@
 package net.minecraft.network.protocol.status;
 
-import java.io.IOException;
 import net.minecraft.network.FriendlyByteBuf;
 import net.minecraft.network.protocol.Packet;
 
 public class ServerboundStatusRequestPacket implements Packet<ServerStatusPacketListener> {
    public ServerboundStatusRequestPacket() {
    }
 
-   public void read(FriendlyByteBuf var1) throws IOException {
+   public ServerboundStatusRequestPacket(FriendlyByteBuf var1) {
    }
 
-   public void write(FriendlyByteBuf var1) throws IOException {
+   public void write(FriendlyByteBuf var1) {
    }
 
    public void handle(ServerStatusPacketListener var1) {
       var1.handleStatusRequest(this);
    }
