--- net/minecraft/world/level/newbiome/layer/traits/DimensionTransformer.java
+++ net/minecraft/world/level/newbiome/layer/traits/DimensionTransformer.java
@@ -1,7 +1,9 @@
 package net.minecraft.world.level.newbiome.layer.traits;
 
-public interface DimensionTransformer {
+import net.minecraft.world.level.newbiome.layer.LayerBiomes;
+
+public interface DimensionTransformer extends LayerBiomes {
    int getParentX(int var1);
 
    int getParentY(int var1);
 }
