--- net/minecraft/data/advancements/AdvancementProvider.java
+++ net/minecraft/data/advancements/AdvancementProvider.java
@@ -25,11 +25,11 @@
 
    public AdvancementProvider(DataGenerator var1) {
       this.generator = var1;
    }
 
-   public void run(HashCache var1) throws IOException {
+   public void run(HashCache var1) {
       Path var2 = this.generator.getOutputFolder();
       HashSet var3 = Sets.newHashSet();
       Consumer var4 = (var3x) -> {
          if (!var3.add(var3x.getId())) {
             throw new IllegalStateException("Duplicate advancement " + var3x.getId());
@@ -52,11 +52,12 @@
       }
 
    }
 
    private static Path createPath(Path var0, Advancement var1) {
-      return var0.resolve("data/" + var1.getId().getNamespace() + "/advancements/" + var1.getId().getPath() + ".json");
+      String var10001 = var1.getId().getNamespace();
+      return var0.resolve("data/" + var10001 + "/advancements/" + var1.getId().getPath() + ".json");
    }
 
    public String getName() {
       return "Advancements";
    }
