--- net/minecraft/world/level/levelgen/feature/BastionFeature.java
+++ net/minecraft/world/level/levelgen/feature/BastionFeature.java
@@ -1,9 +1,10 @@
 package net.minecraft.world.level.levelgen.feature;
 
 import com.mojang.serialization.Codec;
 import net.minecraft.world.level.ChunkPos;
+import net.minecraft.world.level.LevelHeightAccessor;
 import net.minecraft.world.level.biome.Biome;
 import net.minecraft.world.level.biome.BiomeSource;
 import net.minecraft.world.level.chunk.ChunkGenerator;
 import net.minecraft.world.level.levelgen.WorldgenRandom;
 import net.minecraft.world.level.levelgen.feature.configurations.JigsawConfiguration;
@@ -11,9 +12,9 @@
 public class BastionFeature extends JigsawFeature {
    public BastionFeature(Codec<JigsawConfiguration> var1) {
       super(var1, 33, false, false);
    }
 
-   protected boolean isFeatureChunk(ChunkGenerator var1, BiomeSource var2, long var3, WorldgenRandom var5, int var6, int var7, Biome var8, ChunkPos var9, JigsawConfiguration var10) {
+   protected boolean isFeatureChunk(ChunkGenerator var1, BiomeSource var2, long var3, WorldgenRandom var5, ChunkPos var6, Biome var7, ChunkPos var8, JigsawConfiguration var9, LevelHeightAccessor var10) {
       return var5.nextInt(5) >= 2;
    }
 }
