--- net/minecraft/world/level/levelgen/feature/DiskReplaceFeature.java
+++ net/minecraft/world/level/levelgen/feature/DiskReplaceFeature.java
@@ -1,19 +1,15 @@
 package net.minecraft.world.level.levelgen.feature;
 
 import com.mojang.serialization.Codec;
-import java.util.Random;
-import net.minecraft.core.BlockPos;
 import net.minecraft.tags.FluidTags;
-import net.minecraft.world.level.WorldGenLevel;
-import net.minecraft.world.level.chunk.ChunkGenerator;
 import net.minecraft.world.level.levelgen.feature.configurations.DiskConfiguration;
 
 public class DiskReplaceFeature extends BaseDiskFeature {
    public DiskReplaceFeature(Codec<DiskConfiguration> var1) {
       super(var1);
    }
 
-   public boolean place(WorldGenLevel var1, ChunkGenerator var2, Random var3, BlockPos var4, DiskConfiguration var5) {
-      return !var1.getFluidState(var4).is(FluidTags.WATER) ? false : super.place(var1, var2, var3, var4, var5);
+   public boolean place(FeaturePlaceContext<DiskConfiguration> var1) {
+      return !var1.level().getFluidState(var1.origin()).is(FluidTags.WATER) ? false : super.place(var1);
    }
 }
