--- net/minecraft/world/level/levelgen/feature/IcePatchFeature.java
+++ net/minecraft/world/level/levelgen/feature/IcePatchFeature.java
@@ -12,11 +12,11 @@
    public IcePatchFeature(Codec<DiskConfiguration> var1) {
       super(var1);
    }
 
    public boolean place(WorldGenLevel var1, ChunkGenerator var2, Random var3, BlockPos var4, DiskConfiguration var5) {
-      while(var1.isEmptyBlock(var4) && var4.getY() > 2) {
+      while(var1.isEmptyBlock(var4) && var4.getY() > var1.getMinBuildHeight() + 2) {
          var4 = var4.below();
       }
 
       if (!var1.getBlockState(var4).is(Blocks.SNOW_BLOCK)) {
          return false;
