--- net/minecraft/world/level/levelgen/feature/JunglePyramidFeature.java
+++ net/minecraft/world/level/levelgen/feature/JunglePyramidFeature.java
@@ -1,9 +1,11 @@
 package net.minecraft.world.level.levelgen.feature;
 
 import com.mojang.serialization.Codec;
 import net.minecraft.core.RegistryAccess;
+import net.minecraft.core.SectionPos;
+import net.minecraft.world.level.LevelHeightAccessor;
 import net.minecraft.world.level.biome.Biome;
 import net.minecraft.world.level.chunk.ChunkGenerator;
 import net.minecraft.world.level.levelgen.feature.configurations.NoneFeatureConfiguration;
 import net.minecraft.world.level.levelgen.structure.BoundingBox;
 import net.minecraft.world.level.levelgen.structure.JunglePyramidPiece;
@@ -22,12 +24,12 @@
    public static class FeatureStart extends StructureStart<NoneFeatureConfiguration> {
       public FeatureStart(StructureFeature<NoneFeatureConfiguration> var1, int var2, int var3, BoundingBox var4, int var5, long var6) {
          super(var1, var2, var3, var4, var5, var6);
       }
 
-      public void generatePieces(RegistryAccess var1, ChunkGenerator var2, StructureManager var3, int var4, int var5, Biome var6, NoneFeatureConfiguration var7) {
-         JunglePyramidPiece var8 = new JunglePyramidPiece(this.random, var4 * 16, var5 * 16);
-         this.pieces.add(var8);
+      public void generatePieces(RegistryAccess var1, ChunkGenerator var2, StructureManager var3, int var4, int var5, Biome var6, NoneFeatureConfiguration var7, LevelHeightAccessor var8) {
+         JunglePyramidPiece var9 = new JunglePyramidPiece(this.random, SectionPos.sectionToBlockCoord(var4), SectionPos.sectionToBlockCoord(var5));
+         this.pieces.add(var9);
          this.calculateBoundingBox();
       }
    }
 }
