--- net/minecraft/world/level/levelgen/feature/SimpleRandomSelectorFeature.java
+++ net/minecraft/world/level/levelgen/feature/SimpleRandomSelectorFeature.java
@@ -11,11 +11,16 @@
 public class SimpleRandomSelectorFeature extends Feature<SimpleRandomFeatureConfiguration> {
    public SimpleRandomSelectorFeature(Codec<SimpleRandomFeatureConfiguration> var1) {
       super(var1);
    }
 
-   public boolean place(WorldGenLevel var1, ChunkGenerator var2, Random var3, BlockPos var4, SimpleRandomFeatureConfiguration var5) {
-      int var6 = var3.nextInt(var5.features.size());
-      ConfiguredFeature var7 = (ConfiguredFeature)((Supplier)var5.features.get(var6)).get();
-      return var7.place(var1, var2, var3, var4);
+   public boolean place(FeaturePlaceContext<SimpleRandomFeatureConfiguration> var1) {
+      Random var2 = var1.random();
+      SimpleRandomFeatureConfiguration var3 = (SimpleRandomFeatureConfiguration)var1.config();
+      WorldGenLevel var4 = var1.level();
+      BlockPos var5 = var1.origin();
+      ChunkGenerator var6 = var1.chunkGenerator();
+      int var7 = var2.nextInt(var3.features.size());
+      ConfiguredFeature var8 = (ConfiguredFeature)((Supplier)var3.features.get(var7)).get();
+      return var8.place(var4, var6, var2, var5);
    }
 }
