--- net/minecraft/world/level/levelgen/feature/StrongholdFeature.java
+++ net/minecraft/world/level/levelgen/feature/StrongholdFeature.java
@@ -1,10 +1,11 @@
 package net.minecraft.world.level.levelgen.feature;
 
 import com.mojang.serialization.Codec;
 import java.util.List;
 import net.minecraft.core.RegistryAccess;
+import net.minecraft.core.SectionPos;
 import net.minecraft.world.level.ChunkPos;
 import net.minecraft.world.level.biome.Biome;
 import net.minecraft.world.level.biome.BiomeSource;
 import net.minecraft.world.level.chunk.ChunkGenerator;
 import net.minecraft.world.level.levelgen.WorldgenRandom;
@@ -43,11 +44,11 @@
          do {
             this.pieces.clear();
             this.boundingBox = BoundingBox.getUnknownBox();
             this.random.setLargeFeatureSeed(this.seed + (long)(var8++), var4, var5);
             StrongholdPieces.resetPieces();
-            var9 = new StrongholdPieces.StartPiece(this.random, (var4 << 4) + 2, (var5 << 4) + 2);
+            var9 = new StrongholdPieces.StartPiece(this.random, SectionPos.sectionToBlockCoord(var4, 2), SectionPos.sectionToBlockCoord(var5, 2));
             this.pieces.add(var9);
             var9.addChildren(var9, this.pieces, this.random);
             List var10 = var9.pendingChildren;
 
             while(!var10.isEmpty()) {
