--- net/minecraft/world/level/levelgen/feature/StructureFeature.java
+++ net/minecraft/world/level/levelgen/feature/StructureFeature.java
@@ -15,11 +15,14 @@
 import net.minecraft.core.RegistryAccess;
 import net.minecraft.core.SectionPos;
 import net.minecraft.nbt.CompoundTag;
 import net.minecraft.nbt.ListTag;
 import net.minecraft.resources.ResourceLocation;
+import net.minecraft.server.level.ServerLevel;
+import net.minecraft.util.random.WeightedRandomList;
 import net.minecraft.world.level.ChunkPos;
+import net.minecraft.world.level.LevelHeightAccessor;
 import net.minecraft.world.level.LevelReader;
 import net.minecraft.world.level.StructureFeatureManager;
 import net.minecraft.world.level.biome.Biome;
 import net.minecraft.world.level.biome.BiomeSource;
 import net.minecraft.world.level.biome.MobSpawnSettings;
@@ -35,11 +38,10 @@
 import net.minecraft.world.level.levelgen.feature.configurations.OceanRuinConfiguration;
 import net.minecraft.world.level.levelgen.feature.configurations.ProbabilityFeatureConfiguration;
 import net.minecraft.world.level.levelgen.feature.configurations.RuinedPortalConfiguration;
 import net.minecraft.world.level.levelgen.feature.configurations.ShipwreckConfiguration;
 import net.minecraft.world.level.levelgen.feature.configurations.StructureFeatureConfiguration;
-import net.minecraft.world.level.levelgen.structure.BoundingBox;
 import net.minecraft.world.level.levelgen.structure.NetherFossilFeature;
 import net.minecraft.world.level.levelgen.structure.OceanRuinFeature;
 import net.minecraft.world.level.levelgen.structure.StructurePiece;
 import net.minecraft.world.level.levelgen.structure.StructureStart;
 import net.minecraft.world.level.levelgen.structure.templatesystem.StructureManager;
@@ -69,10 +71,11 @@
    public static final StructureFeature<NoneFeatureConfiguration> NETHER_FOSSIL;
    public static final StructureFeature<JigsawConfiguration> BASTION_REMNANT;
    public static final List<StructureFeature<?>> NOISE_AFFECTING_FEATURES;
    private static final ResourceLocation JIGSAW_RENAME;
    private static final Map<ResourceLocation, ResourceLocation> RENAMES;
+   public static final int MAX_STRUCTURE_RANGE = 8;
    private final Codec<ConfiguredStructureFeature<C, StructureFeature<C>>> configuredStructureCodec;
 
    private static <F extends StructureFeature<?>> F register(String var0, F var1, GenerationStep.Decoration var2) {
       STRUCTURES_REGISTRY.put(var0.toLowerCase(Locale.ROOT), var1);
       STEP.put(var1, var2);
@@ -93,50 +96,48 @@
 
    public static void bootstrap() {
    }
 
    @Nullable
-   public static StructureStart<?> loadStaticStart(StructureManager var0, CompoundTag var1, long var2) {
+   public static StructureStart<?> loadStaticStart(ServerLevel var0, CompoundTag var1, long var2) {
       String var4 = var1.getString("id");
       if ("INVALID".equals(var4)) {
          return StructureStart.INVALID_START;
       } else {
          StructureFeature var5 = (StructureFeature)Registry.STRUCTURE_FEATURE.get(new ResourceLocation(var4.toLowerCase(Locale.ROOT)));
          if (var5 == null) {
             LOGGER.error("Unknown feature id: {}", var4);
             return null;
          } else {
-            int var6 = var1.getInt("ChunkX");
-            int var7 = var1.getInt("ChunkZ");
-            int var8 = var1.getInt("references");
-            BoundingBox var9 = var1.contains("BB") ? new BoundingBox(var1.getIntArray("BB")) : BoundingBox.getUnknownBox();
-            ListTag var10 = var1.getList("Children", 10);
+            ChunkPos var6 = new ChunkPos(var1.getInt("ChunkX"), var1.getInt("ChunkZ"));
+            int var7 = var1.getInt("references");
+            ListTag var8 = var1.getList("Children", 10);
 
             try {
-               StructureStart var11 = var5.createStart(var6, var7, var9, var8, var2);
+               StructureStart var9 = var5.createStart(var6, var7, var2);
 
-               for(int var12 = 0; var12 < var10.size(); ++var12) {
-                  CompoundTag var13 = var10.getCompound(var12);
-                  String var14 = var13.getString("id").toLowerCase(Locale.ROOT);
-                  ResourceLocation var15 = new ResourceLocation(var14);
-                  ResourceLocation var16 = (ResourceLocation)RENAMES.getOrDefault(var15, var15);
-                  StructurePieceType var17 = (StructurePieceType)Registry.STRUCTURE_PIECE.get(var16);
-                  if (var17 == null) {
-                     LOGGER.error("Unknown structure piece id: {}", var16);
+               for(int var10 = 0; var10 < var8.size(); ++var10) {
+                  CompoundTag var11 = var8.getCompound(var10);
+                  String var12 = var11.getString("id").toLowerCase(Locale.ROOT);
+                  ResourceLocation var13 = new ResourceLocation(var12);
+                  ResourceLocation var14 = (ResourceLocation)RENAMES.getOrDefault(var13, var13);
+                  StructurePieceType var15 = (StructurePieceType)Registry.STRUCTURE_PIECE.get(var14);
+                  if (var15 == null) {
+                     LOGGER.error("Unknown structure piece id: {}", var14);
                   } else {
                      try {
-                        StructurePiece var18 = var17.load(var0, var13);
-                        var11.getPieces().add(var18);
-                     } catch (Exception var19) {
-                        LOGGER.error("Exception loading structure piece with id {}", var16, var19);
+                        StructurePiece var16 = var15.load(var0, var11);
+                        var9.addPiece(var16);
+                     } catch (Exception var17) {
+                        LOGGER.error("Exception loading structure piece with id {}", var14, var17);
                      }
                   }
                }
 
-               return var11;
-            } catch (Exception var20) {
-               LOGGER.error("Failed Start with id {}", var4, var20);
+               return var9;
+            } catch (Exception var18) {
+               LOGGER.error("Failed Start with id {}", var4, var18);
                return null;
             }
          }
       }
    }
@@ -150,12 +151,12 @@
    }
 
    @Nullable
    public BlockPos getNearestGeneratedFeature(LevelReader var1, StructureFeatureManager var2, BlockPos var3, int var4, boolean var5, long var6, StructureFeatureConfiguration var8) {
       int var9 = var8.spacing();
-      int var10 = var3.getX() >> 4;
-      int var11 = var3.getZ() >> 4;
+      int var10 = SectionPos.blockToSectionCoord(var3.getX());
+      int var11 = SectionPos.blockToSectionCoord(var3.getZ());
       int var12 = 0;
 
       for(WorldgenRandom var13 = new WorldgenRandom(); var12 <= var4; ++var12) {
          for(int var14 = -var12; var14 <= var12; ++var14) {
             boolean var15 = var14 == -var12 || var14 == var12;
@@ -164,20 +165,23 @@
                boolean var17 = var16 == -var12 || var16 == var12;
                if (var15 || var17) {
                   int var18 = var10 + var9 * var14;
                   int var19 = var11 + var9 * var16;
                   ChunkPos var20 = this.getPotentialFeatureChunk(var8, var6, var13, var18, var19);
-                  ChunkAccess var21 = var1.getChunk(var20.x, var20.z, ChunkStatus.STRUCTURE_STARTS);
-                  StructureStart var22 = var2.getStartForFeature(SectionPos.of(var21.getPos(), 0), this, var21);
-                  if (var22 != null && var22.isValid()) {
-                     if (var5 && var22.canBeReferenced()) {
-                        var22.addReference();
-                        return var22.getLocatePos();
-                     }
+                  boolean var21 = var1.getBiomeManager().getPrimaryBiomeAtChunk(var20).getGenerationSettings().isValidStart(this);
+                  if (var21) {
+                     ChunkAccess var22 = var1.getChunk(var20.x, var20.z, ChunkStatus.STRUCTURE_STARTS);
+                     StructureStart var23 = var2.getStartForFeature(SectionPos.bottomOf(var22), this, var22);
+                     if (var23 != null && var23.isValid()) {
+                        if (var5 && var23.canBeReferenced()) {
+                           var23.addReference();
+                           return var23.getLocatePos();
+                        }
 
-                     if (!var5) {
-                        return var22.getLocatePos();
+                        if (!var5) {
+                           return var23.getLocatePos();
+                        }
                      }
                   }
 
                   if (var12 == 0) {
                      break;
@@ -215,25 +219,25 @@
       }
 
       return new ChunkPos(var9 * var7 + var11, var10 * var7 + var12);
    }
 
-   protected boolean isFeatureChunk(ChunkGenerator var1, BiomeSource var2, long var3, WorldgenRandom var5, int var6, int var7, Biome var8, ChunkPos var9, C var10) {
+   protected boolean isFeatureChunk(ChunkGenerator var1, BiomeSource var2, long var3, WorldgenRandom var5, ChunkPos var6, Biome var7, ChunkPos var8, C var9, LevelHeightAccessor var10) {
       return true;
    }
 
-   private StructureStart<C> createStart(int var1, int var2, BoundingBox var3, int var4, long var5) {
-      return this.getStartFactory().create(this, var1, var2, var3, var4, var5);
+   private StructureStart<C> createStart(ChunkPos var1, int var2, long var3) {
+      return this.getStartFactory().create(this, var1, var2, var3);
    }
 
-   public StructureStart<?> generate(RegistryAccess var1, ChunkGenerator var2, BiomeSource var3, StructureManager var4, long var5, ChunkPos var7, Biome var8, int var9, WorldgenRandom var10, StructureFeatureConfiguration var11, C var12) {
-      ChunkPos var13 = this.getPotentialFeatureChunk(var11, var5, var10, var7.x, var7.z);
-      if (var7.x == var13.x && var7.z == var13.z && this.isFeatureChunk(var2, var3, var5, var10, var7.x, var7.z, var8, var13, var12)) {
-         StructureStart var14 = this.createStart(var7.x, var7.z, BoundingBox.getUnknownBox(), var9, var5);
-         var14.generatePieces(var1, var2, var4, var7.x, var7.z, var8, var12);
-         if (var14.isValid()) {
-            return var14;
+   public StructureStart<?> generate(RegistryAccess var1, ChunkGenerator var2, BiomeSource var3, StructureManager var4, long var5, ChunkPos var7, Biome var8, int var9, WorldgenRandom var10, StructureFeatureConfiguration var11, C var12, LevelHeightAccessor var13) {
+      ChunkPos var14 = this.getPotentialFeatureChunk(var11, var5, var10, var7.x, var7.z);
+      if (var7.x == var14.x && var7.z == var14.z && this.isFeatureChunk(var2, var3, var5, var10, var7, var8, var14, var12, var13)) {
+         StructureStart var15 = this.createStart(var7, var9, var5);
+         var15.generatePieces(var1, var2, var4, var7, var8, var12, var13);
+         if (var15.isValid()) {
+            return var15;
          }
       }
 
       return StructureStart.INVALID_START;
    }
@@ -242,16 +246,16 @@
 
    public String getFeatureName() {
       return (String)STRUCTURES_REGISTRY.inverse().get(this);
    }
 
-   public List<MobSpawnSettings.SpawnerData> getSpecialEnemies() {
-      return ImmutableList.of();
+   public WeightedRandomList<MobSpawnSettings.SpawnerData> getSpecialEnemies() {
+      return MobSpawnSettings.EMPTY_MOB_LIST;
    }
 
-   public List<MobSpawnSettings.SpawnerData> getSpecialAnimals() {
-      return ImmutableList.of();
+   public WeightedRandomList<MobSpawnSettings.SpawnerData> getSpecialAnimals() {
+      return MobSpawnSettings.EMPTY_MOB_LIST;
    }
 
    static {
       PILLAGER_OUTPOST = register("Pillager_Outpost", new PillagerOutpostFeature(JigsawConfiguration.CODEC), GenerationStep.Decoration.SURFACE_STRUCTURES);
       MINESHAFT = register("Mineshaft", new MineshaftFeature(MineshaftConfiguration.CODEC), GenerationStep.Decoration.UNDERGROUND_STRUCTURES);
@@ -269,14 +273,14 @@
       END_CITY = register("EndCity", new EndCityFeature(NoneFeatureConfiguration.CODEC), GenerationStep.Decoration.SURFACE_STRUCTURES);
       BURIED_TREASURE = register("Buried_Treasure", new BuriedTreasureFeature(ProbabilityFeatureConfiguration.CODEC), GenerationStep.Decoration.UNDERGROUND_STRUCTURES);
       VILLAGE = register("Village", new VillageFeature(JigsawConfiguration.CODEC), GenerationStep.Decoration.SURFACE_STRUCTURES);
       NETHER_FOSSIL = register("Nether_Fossil", new NetherFossilFeature(NoneFeatureConfiguration.CODEC), GenerationStep.Decoration.UNDERGROUND_DECORATION);
       BASTION_REMNANT = register("Bastion_Remnant", new BastionFeature(JigsawConfiguration.CODEC), GenerationStep.Decoration.SURFACE_STRUCTURES);
-      NOISE_AFFECTING_FEATURES = ImmutableList.of(PILLAGER_OUTPOST, VILLAGE, NETHER_FOSSIL);
+      NOISE_AFFECTING_FEATURES = ImmutableList.of(PILLAGER_OUTPOST, VILLAGE, NETHER_FOSSIL, STRONGHOLD);
       JIGSAW_RENAME = new ResourceLocation("jigsaw");
       RENAMES = ImmutableMap.builder().put(new ResourceLocation("nvi"), JIGSAW_RENAME).put(new ResourceLocation("pcp"), JIGSAW_RENAME).put(new ResourceLocation("bastionremnant"), JIGSAW_RENAME).put(new ResourceLocation("runtime"), JIGSAW_RENAME).build();
    }
 
    public interface StructureStartFactory<C extends FeatureConfiguration> {
-      StructureStart<C> create(StructureFeature<C> var1, int var2, int var3, BoundingBox var4, int var5, long var6);
+      StructureStart<C> create(StructureFeature<C> var1, ChunkPos var2, int var3, long var4);
    }
 }
