--- net/minecraft/client/resources/sounds/AbstractSoundInstance.java
+++ net/minecraft/client/resources/sounds/AbstractSoundInstance.java
@@ -16,11 +16,10 @@
    protected double y;
    protected double z;
    protected boolean looping;
    protected int delay;
    protected SoundInstance.Attenuation attenuation;
-   protected boolean priority;
    protected boolean relative;
 
    protected AbstractSoundInstance(SoundEvent var1, SoundSource var2) {
       this(var1.getLocation(), var2);
    }
