--- net/minecraft/client/resources/sounds/BeeSoundInstance.java
+++ net/minecraft/client/resources/sounds/BeeSoundInstance.java
@@ -6,10 +6,13 @@
 import net.minecraft.util.Mth;
 import net.minecraft.world.entity.Entity;
 import net.minecraft.world.entity.animal.Bee;
 
 public abstract class BeeSoundInstance extends AbstractTickableSoundInstance {
+   private static final float VOLUME_MIN = 0.0F;
+   private static final float VOLUME_MAX = 1.2F;
+   private static final float PITCH_MIN = 0.0F;
    protected final Bee bee;
    private boolean hasSwitched;
 
    public BeeSoundInstance(Bee var1, SoundEvent var2, SoundSource var3) {
       super(var2, var3);
@@ -27,11 +30,11 @@
       if (var1 && !this.isStopped()) {
          Minecraft.getInstance().getSoundManager().queueTickingSound(this.getAlternativeSoundInstance());
          this.hasSwitched = true;
       }
 
-      if (!this.bee.removed && !this.hasSwitched) {
+      if (!this.bee.isRemoved() && !this.hasSwitched) {
          this.x = (double)((float)this.bee.getX());
          this.y = (double)((float)this.bee.getY());
          this.z = (double)((float)this.bee.getZ());
          float var2 = Mth.sqrt(Entity.getHorizontalDistanceSqr(this.bee.getDeltaMovement()));
          if ((double)var2 >= 0.01D) {
