--- net/minecraft/client/resources/sounds/BeeSoundInstance.java
+++ net/minecraft/client/resources/sounds/BeeSoundInstance.java
@@ -27,11 +27,11 @@
       if (var1 && !this.isStopped()) {
          Minecraft.getInstance().getSoundManager().queueTickingSound(this.getAlternativeSoundInstance());
          this.hasSwitched = true;
       }
 
-      if (!this.bee.removed && !this.hasSwitched) {
+      if (!this.bee.isRemoved() && !this.hasSwitched) {
          this.x = (double)((float)this.bee.getX());
          this.y = (double)((float)this.bee.getY());
          this.z = (double)((float)this.bee.getZ());
          float var2 = Mth.sqrt(Entity.getHorizontalDistanceSqr(this.bee.getDeltaMovement()));
          if ((double)var2 >= 0.01D) {
