--- net/minecraft/client/resources/sounds/BiomeAmbientSoundsHandler.java
+++ net/minecraft/client/resources/sounds/BiomeAmbientSoundsHandler.java
@@ -15,15 +15,17 @@
 import net.minecraft.world.level.biome.AmbientMoodSettings;
 import net.minecraft.world.level.biome.Biome;
 import net.minecraft.world.level.biome.BiomeManager;
 
 public class BiomeAmbientSoundsHandler implements AmbientSoundHandler {
+   private static final int LOOP_SOUND_CROSS_FADE_TIME = 40;
+   private static final float SKY_MOOD_RECOVERY_RATE = 0.001F;
    private final LocalPlayer player;
    private final SoundManager soundManager;
    private final BiomeManager biomeManager;
    private final Random random;
-   private Object2ObjectArrayMap<Biome, BiomeAmbientSoundsHandler.LoopSoundInstance> loopSounds = new Object2ObjectArrayMap();
+   private final Object2ObjectArrayMap<Biome, BiomeAmbientSoundsHandler.LoopSoundInstance> loopSounds = new Object2ObjectArrayMap();
    private Optional<AmbientMoodSettings> moodSettings = Optional.empty();
    private Optional<AmbientAdditionsSettings> additionsSettings = Optional.empty();
    private float moodiness;
    private Biome previousBiome;
 
