--- net/minecraft/client/resources/sounds/BubbleColumnAmbientSoundHandler.java
+++ net/minecraft/client/resources/sounds/BubbleColumnAmbientSoundHandler.java
@@ -16,11 +16,11 @@
       this.player = var1;
    }
 
    public void tick() {
       Level var1 = this.player.level;
-      BlockState var2 = (BlockState)var1.getBlockStatesIfLoaded(this.player.getBoundingBox().inflate(0.0D, -0.4000000059604645D, 0.0D).deflate(0.001D)).filter((var0) -> {
+      BlockState var2 = (BlockState)var1.getBlockStatesIfLoaded(this.player.getBoundingBox().inflate(0.0D, -0.4000000059604645D, 0.0D).deflate(1.0E-6D)).filter((var0) -> {
          return var0.is(Blocks.BUBBLE_COLUMN);
       }).findFirst().orElse((Object)null);
       if (var2 != null) {
          if (!this.wasInBubbleColumn && !this.firstTick && var2.is(Blocks.BUBBLE_COLUMN) && !this.player.isSpectator()) {
             boolean var3 = (Boolean)var2.getValue(BubbleColumnBlock.DRAG_DOWN);
