--- net/minecraft/client/resources/sounds/ElytraOnPlayerSoundInstance.java
+++ net/minecraft/client/resources/sounds/ElytraOnPlayerSoundInstance.java
@@ -17,11 +17,11 @@
       this.volume = 0.1F;
    }
 
    public void tick() {
       ++this.time;
-      if (!this.player.removed && (this.time <= 20 || this.player.isFallFlying())) {
+      if (!this.player.isRemoved() && (this.time <= 20 || this.player.isFallFlying())) {
          this.x = (double)((float)this.player.getX());
          this.y = (double)((float)this.player.getY());
          this.z = (double)((float)this.player.getZ());
          float var1 = (float)this.player.getDeltaMovement().lengthSqr();
          if ((double)var1 >= 1.0E-7D) {
