--- net/minecraft/client/resources/sounds/EntityBoundSoundInstance.java
+++ net/minecraft/client/resources/sounds/EntityBoundSoundInstance.java
@@ -5,14 +5,10 @@
 import net.minecraft.world.entity.Entity;
 
 public class EntityBoundSoundInstance extends AbstractTickableSoundInstance {
    private final Entity entity;
 
-   public EntityBoundSoundInstance(SoundEvent var1, SoundSource var2, Entity var3) {
-      this(var1, var2, 1.0F, 1.0F, var3);
-   }
-
    public EntityBoundSoundInstance(SoundEvent var1, SoundSource var2, float var3, float var4, Entity var5) {
       super(var1, var2);
       this.volume = var3;
       this.pitch = var4;
       this.entity = var5;
@@ -24,11 +20,11 @@
    public boolean canPlaySound() {
       return !this.entity.isSilent();
    }
 
    public void tick() {
-      if (this.entity.removed) {
+      if (this.entity.isRemoved()) {
          this.stop();
       } else {
          this.x = (double)((float)this.entity.getX());
          this.y = (double)((float)this.entity.getY());
          this.z = (double)((float)this.entity.getZ());
