--- net/minecraft/client/resources/sounds/EntityBoundSoundInstance.java
+++ net/minecraft/client/resources/sounds/EntityBoundSoundInstance.java
@@ -24,11 +24,11 @@
    public boolean canPlaySound() {
       return !this.entity.isSilent();
    }
 
    public void tick() {
-      if (this.entity.removed) {
+      if (this.entity.isRemoved()) {
          this.stop();
       } else {
          this.x = (double)((float)this.entity.getX());
          this.y = (double)((float)this.entity.getY());
          this.z = (double)((float)this.entity.getZ());
