--- net/minecraft/client/resources/sounds/GuardianAttackSoundInstance.java
+++ net/minecraft/client/resources/sounds/GuardianAttackSoundInstance.java
@@ -3,10 +3,14 @@
 import net.minecraft.sounds.SoundEvents;
 import net.minecraft.sounds.SoundSource;
 import net.minecraft.world.entity.monster.Guardian;
 
 public class GuardianAttackSoundInstance extends AbstractTickableSoundInstance {
+   private static final float VOLUME_MIN = 0.0F;
+   private static final float VOLUME_SCALE = 1.0F;
+   private static final float PITCH_MIN = 0.7F;
+   private static final float PITCH_SCALE = 0.5F;
    private final Guardian guardian;
 
    public GuardianAttackSoundInstance(Guardian var1) {
       super(SoundEvents.GUARDIAN_ATTACK, SoundSource.HOSTILE);
       this.guardian = var1;
@@ -18,11 +22,11 @@
    public boolean canPlaySound() {
       return !this.guardian.isSilent();
    }
 
    public void tick() {
-      if (!this.guardian.removed && this.guardian.getTarget() == null) {
+      if (!this.guardian.isRemoved() && this.guardian.getTarget() == null) {
          this.x = (double)((float)this.guardian.getX());
          this.y = (double)((float)this.guardian.getY());
          this.z = (double)((float)this.guardian.getZ());
          float var1 = this.guardian.getAttackAnimationScale(0.0F);
          this.volume = 0.0F + 1.0F * var1 * var1;
