--- net/minecraft/client/resources/sounds/GuardianAttackSoundInstance.java
+++ net/minecraft/client/resources/sounds/GuardianAttackSoundInstance.java
@@ -18,11 +18,11 @@
    public boolean canPlaySound() {
       return !this.guardian.isSilent();
    }
 
    public void tick() {
-      if (!this.guardian.removed && this.guardian.getTarget() == null) {
+      if (!this.guardian.isRemoved() && this.guardian.getTarget() == null) {
          this.x = (double)((float)this.guardian.getX());
          this.y = (double)((float)this.guardian.getY());
          this.z = (double)((float)this.guardian.getZ());
          float var1 = this.guardian.getAttackAnimationScale(0.0F);
          this.volume = 0.0F + 1.0F * var1 * var1;
