--- net/minecraft/client/resources/sounds/MinecartSoundInstance.java
+++ net/minecraft/client/resources/sounds/MinecartSoundInstance.java
@@ -28,11 +28,11 @@
    public boolean canStartSilent() {
       return true;
    }
 
    public void tick() {
-      if (this.minecart.removed) {
+      if (this.minecart.isRemoved()) {
          this.stop();
       } else {
          this.x = (double)((float)this.minecart.getX());
          this.y = (double)((float)this.minecart.getY());
          this.z = (double)((float)this.minecart.getZ());
