--- net/minecraft/client/resources/sounds/UnderwaterAmbientSoundInstances.java
+++ net/minecraft/client/resources/sounds/UnderwaterAmbientSoundInstances.java
@@ -19,11 +19,11 @@
          this.priority = true;
          this.relative = true;
       }
 
       public void tick() {
-         if (!this.player.removed && this.fade >= 0) {
+         if (!this.player.isRemoved() && this.fade >= 0) {
             if (this.player.isUnderWater()) {
                ++this.fade;
             } else {
                this.fade -= 2;
             }
@@ -48,11 +48,11 @@
          this.priority = true;
          this.relative = true;
       }
 
       public void tick() {
-         if (this.player.removed || !this.player.isUnderWater()) {
+         if (this.player.isRemoved() || !this.player.isUnderWater()) {
             this.stop();
          }
 
       }
    }
