--- net/minecraft/world/level/levelgen/feature/structures/FeaturePoolElement.java
+++ net/minecraft/world/level/levelgen/feature/structures/FeaturePoolElement.java
@@ -9,10 +9,11 @@
 import java.util.function.Supplier;
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.Direction;
 import net.minecraft.core.FrontAndTop;
 import net.minecraft.core.Registry;
+import net.minecraft.core.Vec3i;
 import net.minecraft.nbt.CompoundTag;
 import net.minecraft.world.level.StructureFeatureManager;
 import net.minecraft.world.level.WorldGenLevel;
 import net.minecraft.world.level.block.Blocks;
 import net.minecraft.world.level.block.JigsawBlock;
@@ -48,22 +49,22 @@
       var1.putString("target", "minecraft:empty");
       var1.putString("joint", JigsawBlockEntity.JointType.ROLLABLE.getSerializedName());
       return var1;
    }
 
-   public BlockPos getSize(StructureManager var1, Rotation var2) {
-      return BlockPos.ZERO;
+   public Vec3i getSize(StructureManager var1, Rotation var2) {
+      return Vec3i.ZERO;
    }
 
    public List<StructureTemplate.StructureBlockInfo> getShuffledJigsawBlocks(StructureManager var1, BlockPos var2, Rotation var3, Random var4) {
       ArrayList var5 = Lists.newArrayList();
       var5.add(new StructureTemplate.StructureBlockInfo(var2, (BlockState)Blocks.JIGSAW.defaultBlockState().setValue(JigsawBlock.ORIENTATION, FrontAndTop.fromFrontAndTop(Direction.DOWN, Direction.SOUTH)), this.defaultJigsawNBT));
       return var5;
    }
 
    public BoundingBox getBoundingBox(StructureManager var1, BlockPos var2, Rotation var3) {
-      BlockPos var4 = this.getSize(var1, var3);
+      Vec3i var4 = this.getSize(var1, var3);
       return new BoundingBox(var2.getX(), var2.getY(), var2.getZ(), var2.getX() + var4.getX(), var2.getY() + var4.getY(), var2.getZ() + var4.getZ());
    }
 
    public boolean place(StructureManager var1, WorldGenLevel var2, StructureFeatureManager var3, ChunkGenerator var4, BlockPos var5, BlockPos var6, Rotation var7, BoundingBox var8, Random var9, boolean var10) {
       return ((ConfiguredFeature)this.feature.get()).place(var2, var4, var9, var5);
