--- net/minecraft/world/level/entity/ChunkEntities.java
+++ net/minecraft/world/level/entity/ChunkEntities.java
@@ -1,0 +1,27 @@
+package net.minecraft.world.level.entity;
+
+import java.util.List;
+import java.util.stream.Stream;
+import net.minecraft.world.level.ChunkPos;
+
+public class ChunkEntities<T> {
+   private final ChunkPos pos;
+   private final List<T> entities;
+
+   public ChunkEntities(ChunkPos var1, List<T> var2) {
+      this.pos = var1;
+      this.entities = var2;
+   }
+
+   public ChunkPos getPos() {
+      return this.pos;
+   }
+
+   public Stream<T> getEntities() {
+      return this.entities.stream();
+   }
+
+   public boolean isEmpty() {
+      return this.entities.isEmpty();
+   }
+}
