--- net/minecraft/client/tutorial/CompletedTutorialStepInstance.java
+++ net/minecraft/client/tutorial/CompletedTutorialStepInstance.java
@@ -1,9 +1,6 @@
 package net.minecraft.client.tutorial;
 
 public class CompletedTutorialStepInstance implements TutorialStepInstance {
-   private final Tutorial tutorial;
-
    public CompletedTutorialStepInstance(Tutorial var1) {
-      this.tutorial = var1;
    }
 }
