--- net/minecraft/client/tutorial/CraftPlanksTutorialStep.java
+++ net/minecraft/client/tutorial/CraftPlanksTutorialStep.java
@@ -8,13 +8,13 @@
 import net.minecraft.stats.Stats;
 import net.minecraft.tags.ItemTags;
 import net.minecraft.tags.Tag;
 import net.minecraft.world.item.Item;
 import net.minecraft.world.item.ItemStack;
-import net.minecraft.world.level.GameType;
 
 public class CraftPlanksTutorialStep implements TutorialStepInstance {
+   private static final int HINT_DELAY = 1200;
    private static final Component CRAFT_TITLE = new TranslatableComponent("tutorial.craft_planks.title");
    private static final Component CRAFT_DESCRIPTION = new TranslatableComponent("tutorial.craft_planks.description");
    private final Tutorial tutorial;
    private TutorialToast toast;
    private int timeWaiting;
@@ -23,17 +23,17 @@
       this.tutorial = var1;
    }
 
    public void tick() {
       ++this.timeWaiting;
-      if (this.tutorial.getGameMode() != GameType.SURVIVAL) {
+      if (!this.tutorial.isSurvival()) {
          this.tutorial.setStep(TutorialSteps.NONE);
       } else {
          if (this.timeWaiting == 1) {
             LocalPlayer var1 = this.tutorial.getMinecraft().player;
             if (var1 != null) {
-               if (var1.inventory.contains((Tag)ItemTags.PLANKS)) {
+               if (var1.getInventory().contains((Tag)ItemTags.PLANKS)) {
                   this.tutorial.setStep(TutorialSteps.NONE);
                   return;
                }
 
                if (hasCraftedPlanksPreviously(var1, ItemTags.PLANKS)) {
@@ -58,12 +58,11 @@
       }
 
    }
 
    public void onGetItem(ItemStack var1) {
-      Item var2 = var1.getItem();
-      if (ItemTags.PLANKS.contains(var2)) {
+      if (var1.is((Tag)ItemTags.PLANKS)) {
          this.tutorial.setStep(TutorialSteps.NONE);
       }
 
    }
 
