--- net/minecraft/client/tutorial/CraftPlanksTutorialStep.java
+++ net/minecraft/client/tutorial/CraftPlanksTutorialStep.java
@@ -29,11 +29,11 @@
          this.tutorial.setStep(TutorialSteps.NONE);
       } else {
          if (this.timeWaiting == 1) {
             LocalPlayer var1 = this.tutorial.getMinecraft().player;
             if (var1 != null) {
-               if (var1.inventory.contains((Tag)ItemTags.PLANKS)) {
+               if (var1.getInventory().contains((Tag)ItemTags.PLANKS)) {
                   this.tutorial.setStep(TutorialSteps.NONE);
                   return;
                }
 
                if (hasCraftedPlanksPreviously(var1, ItemTags.PLANKS)) {
@@ -58,12 +58,11 @@
       }
 
    }
 
    public void onGetItem(ItemStack var1) {
-      Item var2 = var1.getItem();
-      if (ItemTags.PLANKS.contains(var2)) {
+      if (var1.is((Tag)ItemTags.PLANKS)) {
          this.tutorial.setStep(TutorialSteps.NONE);
       }
 
    }
 
