--- net/minecraft/client/tutorial/FindTreeTutorialStepInstance.java
+++ net/minecraft/client/tutorial/FindTreeTutorialStepInstance.java
@@ -8,11 +8,10 @@
 import net.minecraft.client.player.LocalPlayer;
 import net.minecraft.network.chat.Component;
 import net.minecraft.network.chat.TranslatableComponent;
 import net.minecraft.stats.Stats;
 import net.minecraft.world.item.ItemStack;
-import net.minecraft.world.level.GameType;
 import net.minecraft.world.level.block.Block;
 import net.minecraft.world.level.block.Blocks;
 import net.minecraft.world.level.block.state.BlockState;
 import net.minecraft.world.phys.BlockHitResult;
 import net.minecraft.world.phys.HitResult;
@@ -29,21 +28,21 @@
       this.tutorial = var1;
    }
 
    public void tick() {
       ++this.timeWaiting;
-      if (this.tutorial.getGameMode() != GameType.SURVIVAL) {
+      if (!this.tutorial.isSurvival()) {
          this.tutorial.setStep(TutorialSteps.NONE);
       } else {
          if (this.timeWaiting == 1) {
             LocalPlayer var1 = this.tutorial.getMinecraft().player;
             if (var1 != null) {
                Iterator var2 = TREE_BLOCKS.iterator();
 
                while(var2.hasNext()) {
                   Block var3 = (Block)var2.next();
-                  if (var1.inventory.contains(new ItemStack(var3))) {
+                  if (var1.getInventory().contains(new ItemStack(var3))) {
                      this.tutorial.setStep(TutorialSteps.CRAFT_PLANKS);
                      return;
                   }
                }
 
@@ -88,11 +87,11 @@
          if (!var2.hasNext()) {
             return;
          }
 
          var3 = (Block)var2.next();
-      } while(var1.getItem() != var3.asItem());
+      } while(!var1.is(var3.asItem()));
 
       this.tutorial.setStep(TutorialSteps.CRAFT_PLANKS);
    }
 
    public static boolean hasPunchedTreesPreviously(LocalPlayer var0) {
