--- net/minecraft/client/tutorial/FindTreeTutorialStepInstance.java
+++ net/minecraft/client/tutorial/FindTreeTutorialStepInstance.java
@@ -39,11 +39,11 @@
             if (var1 != null) {
                Iterator var2 = TREE_BLOCKS.iterator();
 
                while(var2.hasNext()) {
                   Block var3 = (Block)var2.next();
-                  if (var1.inventory.contains(new ItemStack(var3))) {
+                  if (var1.getInventory().contains(new ItemStack(var3))) {
                      this.tutorial.setStep(TutorialSteps.CRAFT_PLANKS);
                      return;
                   }
                }
 
@@ -88,11 +88,11 @@
          if (!var2.hasNext()) {
             return;
          }
 
          var3 = (Block)var2.next();
-      } while(var1.getItem() != var3.asItem());
+      } while(!var1.is(var3.asItem()));
 
       this.tutorial.setStep(TutorialSteps.CRAFT_PLANKS);
    }
 
    public static boolean hasPunchedTreesPreviously(LocalPlayer var0) {
