--- net/minecraft/client/tutorial/MovementTutorialStepInstance.java
+++ net/minecraft/client/tutorial/MovementTutorialStepInstance.java
@@ -2,13 +2,17 @@
 
 import net.minecraft.client.gui.components.toasts.TutorialToast;
 import net.minecraft.client.player.Input;
 import net.minecraft.network.chat.Component;
 import net.minecraft.network.chat.TranslatableComponent;
-import net.minecraft.world.level.GameType;
 
 public class MovementTutorialStepInstance implements TutorialStepInstance {
+   private static final int MINIMUM_TIME_MOVED = 40;
+   private static final int MINIMUM_TIME_LOOKED = 40;
+   private static final int MOVE_HINT_DELAY = 100;
+   private static final int LOOK_HINT_DELAY = 20;
+   private static final int INCOMPLETE = -1;
    private static final Component MOVE_TITLE = new TranslatableComponent("tutorial.move.title", new Object[]{Tutorial.key("forward"), Tutorial.key("left"), Tutorial.key("back"), Tutorial.key("right")});
    private static final Component MOVE_DESCRIPTION = new TranslatableComponent("tutorial.move.description", new Object[]{Tutorial.key("jump")});
    private static final Component LOOK_TITLE = new TranslatableComponent("tutorial.look.title");
    private static final Component LOOK_DESCRIPTION = new TranslatableComponent("tutorial.look.description");
    private final Tutorial tutorial;
@@ -55,11 +59,11 @@
 
          this.lookCompleted = this.timeWaiting;
       }
 
       if (this.moveCompleted != -1 && this.lookCompleted != -1) {
-         if (this.tutorial.getGameMode() == GameType.SURVIVAL) {
+         if (this.tutorial.isSurvival()) {
             this.tutorial.setStep(TutorialSteps.FIND_TREE);
          } else {
             this.tutorial.setStep(TutorialSteps.NONE);
          }
       }
