--- net/minecraft/client/tutorial/PunchTreeTutorialStepInstance.java
+++ net/minecraft/client/tutorial/PunchTreeTutorialStepInstance.java
@@ -31,11 +31,11 @@
          this.tutorial.setStep(TutorialSteps.NONE);
       } else {
          if (this.timeWaiting == 1) {
             LocalPlayer var1 = this.tutorial.getMinecraft().player;
             if (var1 != null) {
-               if (var1.inventory.contains((Tag)ItemTags.LOGS)) {
+               if (var1.getInventory().contains((Tag)ItemTags.LOGS)) {
                   this.tutorial.setStep(TutorialSteps.CRAFT_PLANKS);
                   return;
                }
 
                if (FindTreeTutorialStepInstance.hasPunchedTreesPreviously(var1)) {
@@ -78,10 +78,10 @@
       }
 
    }
 
    public void onGetItem(ItemStack var1) {
-      if (ItemTags.LOGS.contains(var1.getItem())) {
+      if (var1.is((Tag)ItemTags.LOGS)) {
          this.tutorial.setStep(TutorialSteps.CRAFT_PLANKS);
       }
    }
 }
