--- net/minecraft/client/tutorial/PunchTreeTutorialStepInstance.java
+++ net/minecraft/client/tutorial/PunchTreeTutorialStepInstance.java
@@ -8,11 +8,10 @@
 import net.minecraft.network.chat.TranslatableComponent;
 import net.minecraft.tags.BlockTags;
 import net.minecraft.tags.ItemTags;
 import net.minecraft.tags.Tag;
 import net.minecraft.world.item.ItemStack;
-import net.minecraft.world.level.GameType;
 import net.minecraft.world.level.block.state.BlockState;
 
 public class PunchTreeTutorialStepInstance implements TutorialStepInstance {
    private static final Component TITLE = new TranslatableComponent("tutorial.punch_tree.title");
    private static final Component DESCRIPTION = new TranslatableComponent("tutorial.punch_tree.description", new Object[]{Tutorial.key("attack")});
@@ -25,17 +24,17 @@
       this.tutorial = var1;
    }
 
    public void tick() {
       ++this.timeWaiting;
-      if (this.tutorial.getGameMode() != GameType.SURVIVAL) {
+      if (!this.tutorial.isSurvival()) {
          this.tutorial.setStep(TutorialSteps.NONE);
       } else {
          if (this.timeWaiting == 1) {
             LocalPlayer var1 = this.tutorial.getMinecraft().player;
             if (var1 != null) {
-               if (var1.inventory.contains((Tag)ItemTags.LOGS)) {
+               if (var1.getInventory().contains((Tag)ItemTags.LOGS)) {
                   this.tutorial.setStep(TutorialSteps.CRAFT_PLANKS);
                   return;
                }
 
                if (FindTreeTutorialStepInstance.hasPunchedTreesPreviously(var1)) {
@@ -78,10 +77,10 @@
       }
 
    }
 
    public void onGetItem(ItemStack var1) {
-      if (ItemTags.LOGS.contains(var1.getItem())) {
+      if (var1.is((Tag)ItemTags.LOGS)) {
          this.tutorial.setStep(TutorialSteps.CRAFT_PLANKS);
       }
    }
 }
