--- net/minecraft/client/tutorial/Tutorial.java
+++ net/minecraft/client/tutorial/Tutorial.java
@@ -123,12 +123,16 @@
 
    public Minecraft getMinecraft() {
       return this.minecraft;
    }
 
-   public GameType getGameMode() {
-      return this.minecraft.gameMode == null ? GameType.NOT_SET : this.minecraft.gameMode.getPlayerMode();
+   public boolean isSurvival() {
+      if (this.minecraft.gameMode == null) {
+         return false;
+      } else {
+         return this.minecraft.gameMode.getPlayerMode() == GameType.SURVIVAL;
+      }
    }
 
    public static Component key(String var0) {
       return (new KeybindComponent("key." + var0)).withStyle(ChatFormatting.BOLD);
    }
