--- net/minecraft/world/level/block/state/properties/DirectionProperty.java
+++ net/minecraft/world/level/block/state/properties/DirectionProperty.java
@@ -1,7 +1,8 @@
 package net.minecraft.world.level.block.state.properties;
 
+import com.google.common.base.Predicates;
 import com.google.common.collect.Lists;
 import java.util.Arrays;
 import java.util.Collection;
 import java.util.function.Predicate;
 import java.util.stream.Collectors;
@@ -10,10 +11,14 @@
 public class DirectionProperty extends EnumProperty<Direction> {
    protected DirectionProperty(String var1, Collection<Direction> var2) {
       super(var1, Direction.class, var2);
    }
 
+   public static DirectionProperty create(String var0) {
+      return create(var0, (Predicate)Predicates.alwaysTrue());
+   }
+
    public static DirectionProperty create(String var0, Predicate<Direction> var1) {
       return create(var0, (Collection)Arrays.stream(Direction.values()).filter(var1).collect(Collectors.toList()));
    }
 
    public static DirectionProperty create(String var0, Direction... var1) {
