--- net/minecraft/world/level/block/state/properties/WoodType.java
+++ net/minecraft/world/level/block/state/properties/WoodType.java
@@ -1,9 +1,10 @@
 package net.minecraft.world.level.block.state.properties;
 
 import it.unimi.dsi.fastutil.objects.ObjectArraySet;
 import java.util.Set;
+import java.util.stream.Stream;
 
 public class WoodType {
    private static final Set<WoodType> VALUES = new ObjectArraySet();
    public static final WoodType OAK = register(new WoodType("oak"));
    public static final WoodType SPRUCE = register(new WoodType("spruce"));
@@ -21,6 +22,14 @@
 
    private static WoodType register(WoodType var0) {
       VALUES.add(var0);
       return var0;
    }
+
+   public static Stream<WoodType> values() {
+      return VALUES.stream();
+   }
+
+   public String name() {
+      return this.name;
+   }
 }
