--- net/minecraft/server/Bootstrap.java
+++ net/minecraft/server/Bootstrap.java
@@ -7,10 +7,11 @@
 import net.minecraft.SharedConstants;
 import net.minecraft.commands.Commands;
 import net.minecraft.commands.arguments.selector.options.EntitySelectorOptions;
 import net.minecraft.commands.synchronization.ArgumentTypes;
 import net.minecraft.core.Registry;
+import net.minecraft.core.cauldron.CauldronInteraction;
 import net.minecraft.core.dispenser.DispenseItemBehavior;
 import net.minecraft.locale.Language;
 import net.minecraft.tags.StaticTags;
 import net.minecraft.world.effect.MobEffect;
 import net.minecraft.world.entity.EntityType;
@@ -43,10 +44,11 @@
                throw new IllegalStateException("Failed loading EntityTypes");
             } else {
                PotionBrewing.bootStrap();
                EntitySelectorOptions.bootStrap();
                DispenseItemBehavior.bootStrap();
+               CauldronInteraction.bootStrap();
                ArgumentTypes.bootStrap();
                StaticTags.bootStrap();
                wrapStreams();
             }
          }
@@ -95,11 +97,11 @@
       if (!isBootstrapped) {
          throw new IllegalArgumentException("Not bootstrapped");
       } else {
          if (SharedConstants.IS_RUNNING_IN_IDE) {
             getMissingTranslations().forEach((var0) -> {
-               LOGGER.error("Missing translations: " + var0);
+               LOGGER.error((String)"Missing translations: {}", (Object)var0);
             });
             Commands.validate();
          }
 
          DefaultAttributes.validate();
