--- net/minecraft/server/Eula.java
+++ net/minecraft/server/Eula.java
@@ -20,37 +20,34 @@
    }
 
    private boolean readFile() {
       try {
          InputStream var1 = Files.newInputStream(this.file);
-         Throwable var2 = null;
 
-         boolean var4;
+         boolean var3;
          try {
-            Properties var3 = new Properties();
-            var3.load(var1);
-            var4 = Boolean.parseBoolean(var3.getProperty("eula", "false"));
-         } catch (Throwable var14) {
-            var2 = var14;
-            throw var14;
-         } finally {
+            Properties var2 = new Properties();
+            var2.load(var1);
+            var3 = Boolean.parseBoolean(var2.getProperty("eula", "false"));
+         } catch (Throwable var5) {
             if (var1 != null) {
-               if (var2 != null) {
-                  try {
-                     var1.close();
-                  } catch (Throwable var13) {
-                     var2.addSuppressed(var13);
-                  }
-               } else {
+               try {
                   var1.close();
+               } catch (Throwable var4) {
+                  var5.addSuppressed(var4);
                }
             }
 
+            throw var5;
          }
 
-         return var4;
-      } catch (Exception var16) {
+         if (var1 != null) {
+            var1.close();
+         }
+
+         return var3;
+      } catch (Exception var6) {
          LOGGER.warn((String)"Failed to load {}", (Object)this.file);
          this.saveDefaults();
          return false;
       }
    }
@@ -61,35 +58,32 @@
 
    private void saveDefaults() {
       if (!SharedConstants.IS_RUNNING_IN_IDE) {
          try {
             OutputStream var1 = Files.newOutputStream(this.file);
-            Throwable var2 = null;
 
             try {
-               Properties var3 = new Properties();
-               var3.setProperty("eula", "false");
-               var3.store(var1, "By changing the setting below to TRUE you are indicating your agreement to our EULA (https://account.mojang.com/documents/minecraft_eula).");
-            } catch (Throwable var12) {
-               var2 = var12;
-               throw var12;
-            } finally {
+               Properties var2 = new Properties();
+               var2.setProperty("eula", "false");
+               var2.store(var1, "By changing the setting below to TRUE you are indicating your agreement to our EULA (https://account.mojang.com/documents/minecraft_eula).");
+            } catch (Throwable var5) {
                if (var1 != null) {
-                  if (var2 != null) {
-                     try {
-                        var1.close();
-                     } catch (Throwable var11) {
-                        var2.addSuppressed(var11);
-                     }
-                  } else {
+                  try {
                      var1.close();
+                  } catch (Throwable var4) {
+                     var5.addSuppressed(var4);
                   }
                }
 
+               throw var5;
             }
-         } catch (Exception var14) {
-            LOGGER.warn((String)"Failed to save {}", (Object)this.file, (Object)var14);
+
+            if (var1 != null) {
+               var1.close();
+            }
+         } catch (Exception var6) {
+            LOGGER.warn((String)"Failed to save {}", (Object)this.file, (Object)var6);
          }
 
       }
    }
 }
