--- net/minecraft/server/ServerScoreboard.java
+++ net/minecraft/server/ServerScoreboard.java
@@ -1,30 +1,31 @@
 package net.minecraft.server;
 
 import com.google.common.collect.Lists;
 import com.google.common.collect.Sets;
 import java.util.ArrayList;
-import java.util.Arrays;
 import java.util.Iterator;
 import java.util.List;
 import java.util.Set;
 import javax.annotation.Nullable;
+import net.minecraft.nbt.CompoundTag;
 import net.minecraft.network.protocol.Packet;
 import net.minecraft.network.protocol.game.ClientboundSetDisplayObjectivePacket;
 import net.minecraft.network.protocol.game.ClientboundSetObjectivePacket;
 import net.minecraft.network.protocol.game.ClientboundSetPlayerTeamPacket;
 import net.minecraft.network.protocol.game.ClientboundSetScorePacket;
 import net.minecraft.server.level.ServerPlayer;
 import net.minecraft.world.scores.Objective;
 import net.minecraft.world.scores.PlayerTeam;
 import net.minecraft.world.scores.Score;
 import net.minecraft.world.scores.Scoreboard;
+import net.minecraft.world.scores.ScoreboardSaveData;
 
 public class ServerScoreboard extends Scoreboard {
    private final MinecraftServer server;
    private final Set<Objective> trackedObjectives = Sets.newHashSet();
-   private Runnable[] dirtyListeners = new Runnable[0];
+   private final List<Runnable> dirtyListeners = Lists.newArrayList();
 
    public ServerScoreboard(MinecraftServer var1) {
       this.server = var1;
    }
 
@@ -74,21 +75,21 @@
       this.setDirty();
    }
 
    public boolean addPlayerToTeam(String var1, PlayerTeam var2) {
       if (super.addPlayerToTeam(var1, var2)) {
-         this.server.getPlayerList().broadcastAll(new ClientboundSetPlayerTeamPacket(var2, Arrays.asList(var1), 3));
+         this.server.getPlayerList().broadcastAll(ClientboundSetPlayerTeamPacket.createPlayerPacket(var2, var1, ClientboundSetPlayerTeamPacket.Action.ADD));
          this.setDirty();
          return true;
       } else {
          return false;
       }
    }
 
    public void removePlayerFromTeam(String var1, PlayerTeam var2) {
       super.removePlayerFromTeam(var1, var2);
-      this.server.getPlayerList().broadcastAll(new ClientboundSetPlayerTeamPacket(var2, Arrays.asList(var1), 4));
+      this.server.getPlayerList().broadcastAll(ClientboundSetPlayerTeamPacket.createPlayerPacket(var2, var1, ClientboundSetPlayerTeamPacket.Action.REMOVE));
       this.setDirty();
    }
 
    public void onObjectiveAdded(Objective var1) {
       super.onObjectiveAdded(var1);
@@ -113,38 +114,36 @@
       this.setDirty();
    }
 
    public void onTeamAdded(PlayerTeam var1) {
       super.onTeamAdded(var1);
-      this.server.getPlayerList().broadcastAll(new ClientboundSetPlayerTeamPacket(var1, 0));
+      this.server.getPlayerList().broadcastAll(ClientboundSetPlayerTeamPacket.createAddOrModifyPacket(var1, true));
       this.setDirty();
    }
 
    public void onTeamChanged(PlayerTeam var1) {
       super.onTeamChanged(var1);
-      this.server.getPlayerList().broadcastAll(new ClientboundSetPlayerTeamPacket(var1, 2));
+      this.server.getPlayerList().broadcastAll(ClientboundSetPlayerTeamPacket.createAddOrModifyPacket(var1, false));
       this.setDirty();
    }
 
    public void onTeamRemoved(PlayerTeam var1) {
       super.onTeamRemoved(var1);
-      this.server.getPlayerList().broadcastAll(new ClientboundSetPlayerTeamPacket(var1, 1));
+      this.server.getPlayerList().broadcastAll(ClientboundSetPlayerTeamPacket.createRemovePacket(var1));
       this.setDirty();
    }
 
    public void addDirtyListener(Runnable var1) {
-      this.dirtyListeners = (Runnable[])Arrays.copyOf(this.dirtyListeners, this.dirtyListeners.length + 1);
-      this.dirtyListeners[this.dirtyListeners.length - 1] = var1;
+      this.dirtyListeners.add(var1);
    }
 
    protected void setDirty() {
-      Runnable[] var1 = this.dirtyListeners;
-      int var2 = var1.length;
+      Iterator var1 = this.dirtyListeners.iterator();
 
-      for(int var3 = 0; var3 < var2; ++var3) {
-         Runnable var4 = var1[var3];
-         var4.run();
+      while(var1.hasNext()) {
+         Runnable var2 = (Runnable)var1.next();
+         var2.run();
       }
 
    }
 
    public List<Packet<?>> getStartTrackingPackets(Objective var1) {
@@ -224,10 +223,20 @@
       }
 
       return var2;
    }
 
+   public ScoreboardSaveData createData() {
+      ScoreboardSaveData var1 = new ScoreboardSaveData(this);
+      this.addDirtyListener(var1::setDirty);
+      return var1;
+   }
+
+   public ScoreboardSaveData createData(CompoundTag var1) {
+      return this.createData().load(var1);
+   }
+
    public static enum Method {
       CHANGE,
       REMOVE;
 
       private Method() {
