--- net/minecraft/world/level/block/state/pattern/package-info.java
+++ net/minecraft/world/level/block/state/pattern/package-info.java
@@ -1,0 +1,10 @@
+package net.minecraft.world.level.block.state.pattern;
+
+import javax.annotation.ParametersAreNonnullByDefault;
+import net.minecraft.MethodsReturnNonnullByDefault;
+
+// $FF: synthetic class
+@ParametersAreNonnullByDefault
+@MethodsReturnNonnullByDefault
+interface package-info {
+}
