--- com/mojang/math/SymmetricGroup3.java
+++ com/mojang/math/SymmetricGroup3.java
@@ -11,10 +11,11 @@
    P312(2, 0, 1),
    P321(2, 1, 0);
 
    private final int[] permutation;
    private final Matrix3f transformation;
+   private static final int ORDER = 3;
    private static final SymmetricGroup3[][] cayleyTable = (SymmetricGroup3[][])Util.make(new SymmetricGroup3[values().length][values().length], (var0) -> {
       SymmetricGroup3[] var1 = values();
       int var2 = var1.length;
 
       for(int var3 = 0; var3 < var2; ++var3) {
