--- net/minecraft/client/profiling/ClientMetricsLogger.java
+++ net/minecraft/client/profiling/ClientMetricsLogger.java
@@ -1,0 +1,15 @@
+package net.minecraft.client.profiling;
+
+import net.minecraft.util.profiling.ProfilerFiller;
+
+public interface ClientMetricsLogger {
+   void end();
+
+   void startTick();
+
+   boolean isRecording();
+
+   ProfilerFiller getProfiler();
+
+   void endTick();
+}
