--- net/minecraft/client/profiling/InactiveClientMetricsLogger.java
+++ net/minecraft/client/profiling/InactiveClientMetricsLogger.java
@@ -1,0 +1,28 @@
+package net.minecraft.client.profiling;
+
+import net.minecraft.util.profiling.InactiveProfiler;
+import net.minecraft.util.profiling.ProfilerFiller;
+
+public class InactiveClientMetricsLogger implements ClientMetricsLogger {
+   public static final ClientMetricsLogger INSTANCE = new InactiveClientMetricsLogger();
+
+   public InactiveClientMetricsLogger() {
+   }
+
+   public void end() {
+   }
+
+   public void startTick() {
+   }
+
+   public boolean isRecording() {
+      return false;
+   }
+
+   public ProfilerFiller getProfiler() {
+      return InactiveProfiler.INSTANCE;
+   }
+
+   public void endTick() {
+   }
+}
