--- net/minecraft/world/level/block/AbstractBannerBlock.java
+++ net/minecraft/world/level/block/AbstractBannerBlock.java
@@ -3,11 +3,10 @@
 import javax.annotation.Nullable;
 import net.minecraft.core.BlockPos;
 import net.minecraft.world.entity.LivingEntity;
 import net.minecraft.world.item.DyeColor;
 import net.minecraft.world.item.ItemStack;
-import net.minecraft.world.level.BlockGetter;
 import net.minecraft.world.level.Level;
 import net.minecraft.world.level.block.entity.BannerBlockEntity;
 import net.minecraft.world.level.block.entity.BlockEntity;
 import net.minecraft.world.level.block.state.BlockBehaviour;
 import net.minecraft.world.level.block.state.BlockState;
@@ -22,12 +21,12 @@
 
    public boolean isPossibleToRespawnInThis() {
       return true;
    }
 
-   public BlockEntity newBlockEntity(BlockGetter var1) {
-      return new BannerBlockEntity(this.color);
+   public BlockEntity newBlockEntity(BlockPos var1, BlockState var2) {
+      return new BannerBlockEntity(var1, var2, this.color);
    }
 
    public void setPlacedBy(Level var1, BlockPos var2, BlockState var3, @Nullable LivingEntity var4, ItemStack var5) {
       if (var5.hasCustomHoverName()) {
          BlockEntity var6 = var1.getBlockEntity(var2);
