--- net/minecraft/world/level/block/AbstractBannerBlock.java
+++ net/minecraft/world/level/block/AbstractBannerBlock.java
@@ -22,12 +22,12 @@
 
    public boolean isPossibleToRespawnInThis() {
       return true;
    }
 
-   public BlockEntity newBlockEntity(BlockGetter var1) {
-      return new BannerBlockEntity(this.color);
+   public BlockEntity newBlockEntity(BlockPos var1, BlockState var2) {
+      return new BannerBlockEntity(var1, var2, this.color);
    }
 
    public void setPlacedBy(Level var1, BlockPos var2, BlockState var3, @Nullable LivingEntity var4, ItemStack var5) {
       if (var5.hasCustomHoverName()) {
          BlockEntity var6 = var1.getBlockEntity(var2);
@@ -36,9 +36,14 @@
          }
       }
 
    }
 
+   public ItemStack getCloneItemStack(BlockGetter var1, BlockPos var2, BlockState var3) {
+      BlockEntity var4 = var1.getBlockEntity(var2);
+      return var4 instanceof BannerBlockEntity ? ((BannerBlockEntity)var4).getItem() : super.getCloneItemStack(var1, var2, var3);
+   }
+
    public DyeColor getColor() {
       return this.color;
    }
 }
