--- net/minecraft/world/level/block/AbstractGlassBlock.java
+++ net/minecraft/world/level/block/AbstractGlassBlock.java
@@ -15,9 +15,13 @@
 
    public VoxelShape getVisualShape(BlockState var1, BlockGetter var2, BlockPos var3, CollisionContext var4) {
       return Shapes.empty();
    }
 
+   public float getShadeBrightness(BlockState var1, BlockGetter var2, BlockPos var3) {
+      return 1.0F;
+   }
+
    public boolean propagatesSkylightDown(BlockState var1, BlockGetter var2, BlockPos var3) {
       return true;
    }
 }
