--- net/minecraft/world/level/block/AbstractCandleBlock.java
+++ net/minecraft/world/level/block/AbstractCandleBlock.java
@@ -1,0 +1,47 @@
+package net.minecraft.world.level.block;
+
+import javax.annotation.Nullable;
+import net.minecraft.core.BlockPos;
+import net.minecraft.core.particles.ParticleTypes;
+import net.minecraft.sounds.SoundEvents;
+import net.minecraft.sounds.SoundSource;
+import net.minecraft.world.entity.player.Player;
+import net.minecraft.world.entity.projectile.Projectile;
+import net.minecraft.world.level.Level;
+import net.minecraft.world.level.LevelAccessor;
+import net.minecraft.world.level.block.state.BlockBehaviour;
+import net.minecraft.world.level.block.state.BlockState;
+import net.minecraft.world.level.block.state.properties.BlockStateProperties;
+import net.minecraft.world.level.block.state.properties.BooleanProperty;
+import net.minecraft.world.level.gameevent.GameEvent;
+import net.minecraft.world.phys.BlockHitResult;
+
+public abstract class AbstractCandleBlock extends Block {
+   public static final BooleanProperty LIT;
+
+   protected AbstractCandleBlock(BlockBehaviour.Properties var1) {
+      super(var1);
+   }
+
+   public void onProjectileHit(Level var1, BlockState var2, BlockHitResult var3, Projectile var4) {
+      if (!var1.isClientSide && var4.isOnFire() && !(Boolean)var2.getValue(LIT)) {
+         setLit(var1, var2, var3.getBlockPos(), true);
+      }
+
+   }
+
+   protected static void extinguish(@Nullable Player var0, BlockState var1, LevelAccessor var2, BlockPos var3) {
+      setLit(var2, var1, var3, false);
+      var2.addParticle(ParticleTypes.SMOKE, (double)var3.getX(), (double)var3.getY(), (double)var3.getZ(), 0.0D, 0.10000000149011612D, 0.0D);
+      var2.playSound((Player)null, var3, SoundEvents.CANDLE_EXTINGUISH, SoundSource.BLOCKS, 1.0F, 1.0F);
+      var2.gameEvent(var0, GameEvent.BLOCK_CHANGE, (BlockPos)var3);
+   }
+
+   private static void setLit(LevelAccessor var0, BlockState var1, BlockPos var2, boolean var3) {
+      var0.setBlock(var2, (BlockState)var1.setValue(LIT, var3), 11);
+   }
+
+   static {
+      LIT = BlockStateProperties.LIT;
+   }
+}
