--- net/minecraft/world/level/block/BambooBlock.java
+++ net/minecraft/world/level/block/BambooBlock.java
@@ -25,16 +25,24 @@
 import net.minecraft.world.phys.Vec3;
 import net.minecraft.world.phys.shapes.CollisionContext;
 import net.minecraft.world.phys.shapes.VoxelShape;
 
 public class BambooBlock extends Block implements BonemealableBlock {
+   protected static final float SMALL_LEAVES_AABB_OFFSET = 3.0F;
+   protected static final float LARGE_LEAVES_AABB_OFFSET = 5.0F;
+   protected static final float COLLISION_AABB_OFFSET = 1.5F;
    protected static final VoxelShape SMALL_SHAPE = Block.box(5.0D, 0.0D, 5.0D, 11.0D, 16.0D, 11.0D);
    protected static final VoxelShape LARGE_SHAPE = Block.box(3.0D, 0.0D, 3.0D, 13.0D, 16.0D, 13.0D);
    protected static final VoxelShape COLLISION_SHAPE = Block.box(6.5D, 0.0D, 6.5D, 9.5D, 16.0D, 9.5D);
    public static final IntegerProperty AGE;
    public static final EnumProperty<BambooLeaves> LEAVES;
    public static final IntegerProperty STAGE;
+   public static final int MAX_HEIGHT = 16;
+   public static final int STAGE_GROWING = 0;
+   public static final int STAGE_DONE_GROWING = 1;
+   public static final int AGE_THIN_BAMBOO = 0;
+   public static final int AGE_THICK_BAMBOO = 1;
 
    public BambooBlock(BlockBehaviour.Properties var1) {
       super(var1);
       this.registerDefaultState((BlockState)((BlockState)((BlockState)((BlockState)this.stateDefinition.any()).setValue(AGE, 0)).setValue(LEAVES, BambooLeaves.NONE)).setValue(STAGE, 0));
    }
@@ -79,11 +87,11 @@
             } else if (var3.is(Blocks.BAMBOO)) {
                int var5 = (Integer)var3.getValue(AGE) > 0 ? 1 : 0;
                return (BlockState)this.defaultBlockState().setValue(AGE, var5);
             } else {
                BlockState var4 = var1.getLevel().getBlockState(var1.getClickedPos().above());
-               return !var4.is(Blocks.BAMBOO) && !var4.is(Blocks.BAMBOO_SAPLING) ? Blocks.BAMBOO_SAPLING.defaultBlockState() : (BlockState)this.defaultBlockState().setValue(AGE, var4.getValue(AGE));
+               return var4.is(Blocks.BAMBOO) ? (BlockState)this.defaultBlockState().setValue(AGE, var4.getValue(AGE)) : Blocks.BAMBOO_SAPLING.defaultBlockState();
             }
          } else {
             return null;
          }
       }
