--- net/minecraft/world/level/block/BannerBlock.java
+++ net/minecraft/world/level/block/BannerBlock.java
@@ -55,10 +55,14 @@
 
    protected void createBlockStateDefinition(StateDefinition.Builder<Block, BlockState> var1) {
       var1.add(ROTATION);
    }
 
+   public static Block byColor(DyeColor var0) {
+      return (Block)BY_COLOR.getOrDefault(var0, Blocks.WHITE_BANNER);
+   }
+
    static {
       ROTATION = BlockStateProperties.ROTATION_16;
       BY_COLOR = Maps.newHashMap();
       SHAPE = Block.box(4.0D, 0.0D, 4.0D, 12.0D, 16.0D, 12.0D);
    }
