--- net/minecraft/world/level/block/BaseEntityBlock.java
+++ net/minecraft/world/level/block/BaseEntityBlock.java
@@ -3,10 +3,12 @@
 import javax.annotation.Nullable;
 import net.minecraft.core.BlockPos;
 import net.minecraft.world.MenuProvider;
 import net.minecraft.world.level.Level;
 import net.minecraft.world.level.block.entity.BlockEntity;
+import net.minecraft.world.level.block.entity.BlockEntityTicker;
+import net.minecraft.world.level.block.entity.BlockEntityType;
 import net.minecraft.world.level.block.state.BlockBehaviour;
 import net.minecraft.world.level.block.state.BlockState;
 
 public abstract class BaseEntityBlock extends Block implements EntityBlock {
    protected BaseEntityBlock(BlockBehaviour.Properties var1) {
@@ -26,6 +28,11 @@
    @Nullable
    public MenuProvider getMenuProvider(BlockState var1, Level var2, BlockPos var3) {
       BlockEntity var4 = var2.getBlockEntity(var3);
       return var4 instanceof MenuProvider ? (MenuProvider)var4 : null;
    }
+
+   @Nullable
+   protected static <E extends BlockEntity, A extends BlockEntity> BlockEntityTicker<A> createTickerHelper(BlockEntityType<A> var0, BlockEntityType<E> var1, BlockEntityTicker<? super E> var2) {
+      return var1 == var0 ? var2 : null;
+   }
 }
