--- net/minecraft/world/level/block/BeehiveBlock.java
+++ net/minecraft/world/level/block/BeehiveBlock.java
@@ -27,22 +27,24 @@
 import net.minecraft.world.item.ItemStack;
 import net.minecraft.world.item.Items;
 import net.minecraft.world.item.context.BlockPlaceContext;
 import net.minecraft.world.item.enchantment.EnchantmentHelper;
 import net.minecraft.world.item.enchantment.Enchantments;
-import net.minecraft.world.level.BlockGetter;
 import net.minecraft.world.level.GameRules;
 import net.minecraft.world.level.Level;
 import net.minecraft.world.level.LevelAccessor;
 import net.minecraft.world.level.block.entity.BeehiveBlockEntity;
 import net.minecraft.world.level.block.entity.BlockEntity;
+import net.minecraft.world.level.block.entity.BlockEntityTicker;
+import net.minecraft.world.level.block.entity.BlockEntityType;
 import net.minecraft.world.level.block.state.BlockBehaviour;
 import net.minecraft.world.level.block.state.BlockState;
 import net.minecraft.world.level.block.state.StateDefinition;
 import net.minecraft.world.level.block.state.properties.BlockStateProperties;
 import net.minecraft.world.level.block.state.properties.DirectionProperty;
 import net.minecraft.world.level.block.state.properties.IntegerProperty;
+import net.minecraft.world.level.gameevent.GameEvent;
 import net.minecraft.world.level.storage.loot.LootContext;
 import net.minecraft.world.level.storage.loot.parameters.LootContextParams;
 import net.minecraft.world.phys.AABB;
 import net.minecraft.world.phys.BlockHitResult;
 
@@ -72,11 +74,11 @@
             var7.emptyAllLivingFromHive(var2, var4, BeehiveBlockEntity.BeeReleaseStatus.EMERGENCY);
             var1.updateNeighbourForOutputSignal(var3, this);
             this.angerNearbyBees(var1, var3);
          }
 
-         CriteriaTriggers.BEE_NEST_DESTROYED.trigger((ServerPlayer)var2, var4.getBlock(), var6, var7.getOccupantCount());
+         CriteriaTriggers.BEE_NEST_DESTROYED.trigger((ServerPlayer)var2, var4, var6, var7.getOccupantCount());
       }
 
    }
 
    private void angerNearbyBees(Level var1, BlockPos var2) {
@@ -103,27 +105,29 @@
    public InteractionResult use(BlockState var1, Level var2, BlockPos var3, Player var4, InteractionHand var5, BlockHitResult var6) {
       ItemStack var7 = var4.getItemInHand(var5);
       int var8 = (Integer)var1.getValue(HONEY_LEVEL);
       boolean var9 = false;
       if (var8 >= 5) {
-         if (var7.getItem() == Items.SHEARS) {
+         if (var7.is(Items.SHEARS)) {
             var2.playSound(var4, var4.getX(), var4.getY(), var4.getZ(), SoundEvents.BEEHIVE_SHEAR, SoundSource.NEUTRAL, 1.0F, 1.0F);
             dropHoneycomb(var2, var3);
             var7.hurtAndBreak(1, var4, (var1x) -> {
                var1x.broadcastBreakEvent(var5);
             });
             var9 = true;
-         } else if (var7.getItem() == Items.GLASS_BOTTLE) {
+            var2.gameEvent(var4, GameEvent.SHEAR, var3);
+         } else if (var7.is(Items.GLASS_BOTTLE)) {
             var7.shrink(1);
             var2.playSound(var4, var4.getX(), var4.getY(), var4.getZ(), SoundEvents.BOTTLE_FILL, SoundSource.NEUTRAL, 1.0F, 1.0F);
             if (var7.isEmpty()) {
                var4.setItemInHand(var5, new ItemStack(Items.HONEY_BOTTLE));
-            } else if (!var4.inventory.add(new ItemStack(Items.HONEY_BOTTLE))) {
+            } else if (!var4.getInventory().add(new ItemStack(Items.HONEY_BOTTLE))) {
                var4.drop(new ItemStack(Items.HONEY_BOTTLE), false);
             }
 
             var9 = true;
+            var2.gameEvent(var4, GameEvent.FLUID_PICKUP, var3);
          }
       }
 
       if (var9) {
          if (!CampfireBlock.isSmokeyPos(var2, var3)) {
@@ -177,39 +181,42 @@
    public RenderShape getRenderShape(BlockState var1) {
       return RenderShape.MODEL;
    }
 
    @Nullable
-   public BlockEntity newBlockEntity(BlockGetter var1) {
-      return new BeehiveBlockEntity();
+   public BlockEntity newBlockEntity(BlockPos var1, BlockState var2) {
+      return new BeehiveBlockEntity(var1, var2);
    }
 
+   @Nullable
+   public <T extends BlockEntity> BlockEntityTicker<T> getTicker(Level var1, BlockState var2, BlockEntityType<T> var3) {
+      return var1.isClientSide ? null : createTickerHelper(var3, BlockEntityType.BEEHIVE, BeehiveBlockEntity::serverTick);
+   }
+
    public void playerWillDestroy(Level var1, BlockPos var2, BlockState var3, Player var4) {
       if (!var1.isClientSide && var4.isCreative() && var1.getGameRules().getBoolean(GameRules.RULE_DOBLOCKDROPS)) {
          BlockEntity var5 = var1.getBlockEntity(var2);
          if (var5 instanceof BeehiveBlockEntity) {
             BeehiveBlockEntity var6 = (BeehiveBlockEntity)var5;
             ItemStack var7 = new ItemStack(this);
             int var8 = (Integer)var3.getValue(HONEY_LEVEL);
             boolean var9 = !var6.isEmpty();
-            if (!var9 && var8 == 0) {
-               return;
-            }
+            if (var9 || var8 > 0) {
+               CompoundTag var10;
+               if (var9) {
+                  var10 = new CompoundTag();
+                  var10.put("Bees", var6.writeBees());
+                  var7.addTagElement("BlockEntityTag", var10);
+               }
 
-            CompoundTag var10;
-            if (var9) {
                var10 = new CompoundTag();
-               var10.put("Bees", var6.writeBees());
-               var7.addTagElement("BlockEntityTag", var10);
+               var10.putInt("honey_level", var8);
+               var7.addTagElement("BlockStateTag", var10);
+               ItemEntity var11 = new ItemEntity(var1, (double)var2.getX(), (double)var2.getY(), (double)var2.getZ(), var7);
+               var11.setDefaultPickUpDelay();
+               var1.addFreshEntity(var11);
             }
-
-            var10 = new CompoundTag();
-            var10.putInt("honey_level", var8);
-            var7.addTagElement("BlockStateTag", var10);
-            ItemEntity var11 = new ItemEntity(var1, (double)var2.getX(), (double)var2.getY(), (double)var2.getZ(), var7);
-            var11.setDefaultPickUpDelay();
-            var1.addFreshEntity(var11);
          }
       }
 
       super.playerWillDestroy(var1, var2, var3, var4);
    }
