--- net/minecraft/world/level/block/BeetrootBlock.java
+++ net/minecraft/world/level/block/BeetrootBlock.java
@@ -1,21 +1,24 @@
 package net.minecraft.world.level.block;
 
 import java.util.Random;
 import net.minecraft.core.BlockPos;
 import net.minecraft.server.level.ServerLevel;
+import net.minecraft.world.item.Items;
 import net.minecraft.world.level.BlockGetter;
+import net.minecraft.world.level.ItemLike;
 import net.minecraft.world.level.Level;
 import net.minecraft.world.level.block.state.BlockBehaviour;
 import net.minecraft.world.level.block.state.BlockState;
 import net.minecraft.world.level.block.state.StateDefinition;
 import net.minecraft.world.level.block.state.properties.BlockStateProperties;
 import net.minecraft.world.level.block.state.properties.IntegerProperty;
 import net.minecraft.world.phys.shapes.CollisionContext;
 import net.minecraft.world.phys.shapes.VoxelShape;
 
 public class BeetrootBlock extends CropBlock {
+   public static final int MAX_AGE = 3;
    public static final IntegerProperty AGE;
    private static final VoxelShape[] SHAPE_BY_AGE;
 
    public BeetrootBlock(BlockBehaviour.Properties var1) {
       super(var1);
@@ -27,10 +30,14 @@
 
    public int getMaxAge() {
       return 3;
    }
 
+   protected ItemLike getBaseSeedId() {
+      return Items.BEETROOT_SEEDS;
+   }
+
    public void randomTick(BlockState var1, ServerLevel var2, BlockPos var3, Random var4) {
       if (var4.nextInt(3) != 0) {
          super.randomTick(var1, var2, var3, var4);
       }
 
