--- net/minecraft/world/level/block/BubbleColumnBlock.java
+++ net/minecraft/world/level/block/BubbleColumnBlock.java
@@ -1,23 +1,26 @@
 package net.minecraft.world.level.block;
 
+import java.util.Optional;
 import java.util.Random;
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.Direction;
 import net.minecraft.core.particles.ParticleTypes;
 import net.minecraft.server.level.ServerLevel;
+import net.minecraft.sounds.SoundEvent;
 import net.minecraft.world.entity.Entity;
+import net.minecraft.world.item.ItemStack;
+import net.minecraft.world.item.Items;
 import net.minecraft.world.level.BlockGetter;
 import net.minecraft.world.level.Level;
 import net.minecraft.world.level.LevelAccessor;
 import net.minecraft.world.level.LevelReader;
 import net.minecraft.world.level.block.state.BlockBehaviour;
 import net.minecraft.world.level.block.state.BlockState;
 import net.minecraft.world.level.block.state.StateDefinition;
 import net.minecraft.world.level.block.state.properties.BlockStateProperties;
 import net.minecraft.world.level.block.state.properties.BooleanProperty;
-import net.minecraft.world.level.material.Fluid;
 import net.minecraft.world.level.material.FluidState;
 import net.minecraft.world.level.material.Fluids;
 import net.minecraft.world.phys.shapes.CollisionContext;
 import net.minecraft.world.phys.shapes.Shapes;
 import net.minecraft.world.phys.shapes.VoxelShape;
@@ -46,25 +49,24 @@
          var4.onInsideBubbleColumn((Boolean)var1.getValue(DRAG_DOWN));
       }
 
    }
 
-   public void onPlace(BlockState var1, Level var2, BlockPos var3, BlockState var4, boolean var5) {
-      growColumn(var2, var3.above(), getDrag(var2, var3.below()));
-   }
-
    public void tick(BlockState var1, ServerLevel var2, BlockPos var3, Random var4) {
       growColumn(var2, var3.above(), getDrag(var2, var3));
    }
 
    public FluidState getFluidState(BlockState var1) {
       return Fluids.WATER.getSource(false);
    }
 
    public static void growColumn(LevelAccessor var0, BlockPos var1, boolean var2) {
-      if (canExistIn(var0, var1)) {
-         var0.setBlock(var1, (BlockState)Blocks.BUBBLE_COLUMN.defaultBlockState().setValue(DRAG_DOWN, var2), 2);
+      BlockPos.MutableBlockPos var3 = var1.mutable();
+
+      while(canExistIn(var0, var3)) {
+         var0.setBlock(var3, (BlockState)Blocks.BUBBLE_COLUMN.defaultBlockState().setValue(DRAG_DOWN, var2), 2);
+         var3.move(Direction.UP);
       }
 
    }
 
    public static boolean canExistIn(LevelAccessor var0, BlockPos var1) {
@@ -111,14 +113,18 @@
 
    protected void createBlockStateDefinition(StateDefinition.Builder<Block, BlockState> var1) {
       var1.add(DRAG_DOWN);
    }
 
-   public Fluid takeLiquid(LevelAccessor var1, BlockPos var2, BlockState var3) {
+   public ItemStack pickupBlock(LevelAccessor var1, BlockPos var2, BlockState var3) {
       var1.setBlock(var2, Blocks.AIR.defaultBlockState(), 11);
-      return Fluids.WATER;
+      return new ItemStack(Items.WATER_BUCKET);
    }
 
+   public Optional<SoundEvent> getPickupSound() {
+      return Fluids.WATER.getPickupSound();
+   }
+
    static {
       DRAG_DOWN = BlockStateProperties.DRAG;
    }
 }
