--- net/minecraft/world/level/block/CactusBlock.java
+++ net/minecraft/world/level/block/CactusBlock.java
@@ -22,10 +22,12 @@
 import net.minecraft.world.phys.shapes.CollisionContext;
 import net.minecraft.world.phys.shapes.VoxelShape;
 
 public class CactusBlock extends Block {
    public static final IntegerProperty AGE;
+   public static final int MAX_AGE = 15;
+   protected static final int AABB_OFFSET = 1;
    protected static final VoxelShape COLLISION_SHAPE;
    protected static final VoxelShape OUTLINE_SHAPE;
 
    protected CactusBlock(BlockBehaviour.Properties var1) {
       super(var1);
