--- net/minecraft/world/level/block/DiodeBlock.java
+++ net/minecraft/world/level/block/DiodeBlock.java
@@ -65,11 +65,11 @@
 
    public void neighborChanged(BlockState var1, Level var2, BlockPos var3, Block var4, BlockPos var5, boolean var6) {
       if (var1.canSurvive(var2, var3)) {
          this.checkTickOnNeighbor(var2, var3, var1);
       } else {
-         BlockEntity var7 = this.isEntityBlock() ? var2.getBlockEntity(var3) : null;
+         BlockEntity var7 = var1.hasBlockEntity() ? var2.getBlockEntity(var3) : null;
          dropResources(var1, var2, var3, var7);
          var2.removeBlock(var3, false);
          Direction[] var8 = Direction.values();
          int var9 = var8.length;
 
