--- net/minecraft/world/level/block/EndPortalBlock.java
+++ net/minecraft/world/level/block/EndPortalBlock.java
@@ -21,12 +21,12 @@
 
    protected EndPortalBlock(BlockBehaviour.Properties var1) {
       super(var1);
    }
 
-   public BlockEntity newBlockEntity(BlockGetter var1) {
-      return new TheEndPortalBlockEntity();
+   public BlockEntity newBlockEntity(BlockPos var1, BlockState var2) {
+      return new TheEndPortalBlockEntity(var1, var2);
    }
 
    public VoxelShape getShape(BlockState var1, BlockGetter var2, BlockPos var3, CollisionContext var4) {
       return SHAPE;
    }
