--- net/minecraft/world/level/block/FarmBlock.java
+++ net/minecraft/world/level/block/FarmBlock.java
@@ -26,10 +26,11 @@
 import net.minecraft.world.phys.shapes.VoxelShape;
 
 public class FarmBlock extends Block {
    public static final IntegerProperty MOISTURE;
    protected static final VoxelShape SHAPE;
+   public static final int MAX_MOISTURE = 7;
 
    protected FarmBlock(BlockBehaviour.Properties var1) {
       super(var1);
       this.registerDefaultState((BlockState)((BlockState)this.stateDefinition.any()).setValue(MOISTURE, 0));
    }
@@ -78,16 +79,16 @@
          var2.setBlock(var3, (BlockState)var1.setValue(MOISTURE, 7), 2);
       }
 
    }
 
-   public void fallOn(Level var1, BlockPos var2, Entity var3, float var4) {
-      if (!var1.isClientSide && var1.random.nextFloat() < var4 - 0.5F && var3 instanceof LivingEntity && (var3 instanceof Player || var1.getGameRules().getBoolean(GameRules.RULE_MOBGRIEFING)) && var3.getBbWidth() * var3.getBbWidth() * var3.getBbHeight() > 0.512F) {
-         turnToDirt(var1.getBlockState(var2), var1, var2);
+   public void fallOn(Level var1, BlockState var2, BlockPos var3, Entity var4, float var5) {
+      if (!var1.isClientSide && var1.random.nextFloat() < var5 - 0.5F && var4 instanceof LivingEntity && (var4 instanceof Player || var1.getGameRules().getBoolean(GameRules.RULE_MOBGRIEFING)) && var4.getBbWidth() * var4.getBbWidth() * var4.getBbHeight() > 0.512F) {
+         turnToDirt(var2, var1, var3);
       }
 
-      super.fallOn(var1, var2, var3, var4);
+      super.fallOn(var1, var2, var3, var4, var5);
    }
 
    public static void turnToDirt(BlockState var0, Level var1, BlockPos var2) {
       var1.setBlockAndUpdate(var2, pushEntitiesUp(var0, Blocks.DIRT.defaultBlockState(), var1, var2));
    }
