--- net/minecraft/world/level/block/HoneyBlock.java
+++ net/minecraft/world/level/block/HoneyBlock.java
@@ -2,10 +2,11 @@
 
 import net.minecraft.advancements.CriteriaTriggers;
 import net.minecraft.core.BlockPos;
 import net.minecraft.server.level.ServerPlayer;
 import net.minecraft.sounds.SoundEvents;
+import net.minecraft.world.damagesource.DamageSource;
 import net.minecraft.world.entity.Entity;
 import net.minecraft.world.entity.LivingEntity;
 import net.minecraft.world.entity.item.PrimedTnt;
 import net.minecraft.world.entity.vehicle.AbstractMinecart;
 import net.minecraft.world.entity.vehicle.Boat;
@@ -36,11 +37,11 @@
       var3.playSound(SoundEvents.HONEY_BLOCK_SLIDE, 1.0F, 1.0F);
       if (!var1.isClientSide) {
          var1.broadcastEntityEvent(var3, (byte)54);
       }
 
-      if (var3.causeFallDamage(var4, 0.2F)) {
+      if (var3.causeFallDamage(var4, 0.2F, DamageSource.FALL)) {
          var3.playSound(this.soundType.getFallSound(), this.soundType.getVolume() * 0.5F, this.soundType.getPitch() * 0.75F);
       }
 
    }
 
