--- net/minecraft/world/level/block/HugeMushroomBlock.java
+++ net/minecraft/world/level/block/HugeMushroomBlock.java
@@ -27,23 +27,23 @@
    }
 
    public BlockState getStateForPlacement(BlockPlaceContext var1) {
       Level var2 = var1.getLevel();
       BlockPos var3 = var1.getClickedPos();
-      return (BlockState)((BlockState)((BlockState)((BlockState)((BlockState)((BlockState)this.defaultBlockState().setValue(DOWN, this != var2.getBlockState(var3.below()).getBlock())).setValue(UP, this != var2.getBlockState(var3.above()).getBlock())).setValue(NORTH, this != var2.getBlockState(var3.north()).getBlock())).setValue(EAST, this != var2.getBlockState(var3.east()).getBlock())).setValue(SOUTH, this != var2.getBlockState(var3.south()).getBlock())).setValue(WEST, this != var2.getBlockState(var3.west()).getBlock());
+      return (BlockState)((BlockState)((BlockState)((BlockState)((BlockState)((BlockState)this.defaultBlockState().setValue(DOWN, !var2.getBlockState(var3.below()).is(this))).setValue(UP, !var2.getBlockState(var3.above()).is(this))).setValue(NORTH, !var2.getBlockState(var3.north()).is(this))).setValue(EAST, !var2.getBlockState(var3.east()).is(this))).setValue(SOUTH, !var2.getBlockState(var3.south()).is(this))).setValue(WEST, !var2.getBlockState(var3.west()).is(this));
    }
 
    public BlockState updateShape(BlockState var1, Direction var2, BlockState var3, LevelAccessor var4, BlockPos var5, BlockPos var6) {
       return var3.is(this) ? (BlockState)var1.setValue((Property)PROPERTY_BY_DIRECTION.get(var2), false) : super.updateShape(var1, var2, var3, var4, var5, var6);
    }
 
    public BlockState rotate(BlockState var1, Rotation var2) {
-      return (BlockState)((BlockState)((BlockState)((BlockState)((BlockState)((BlockState)var1.setValue((Property)PROPERTY_BY_DIRECTION.get(var2.rotate(Direction.NORTH)), var1.getValue(NORTH))).setValue((Property)PROPERTY_BY_DIRECTION.get(var2.rotate(Direction.SOUTH)), var1.getValue(SOUTH))).setValue((Property)PROPERTY_BY_DIRECTION.get(var2.rotate(Direction.EAST)), var1.getValue(EAST))).setValue((Property)PROPERTY_BY_DIRECTION.get(var2.rotate(Direction.WEST)), var1.getValue(WEST))).setValue((Property)PROPERTY_BY_DIRECTION.get(var2.rotate(Direction.UP)), var1.getValue(UP))).setValue((Property)PROPERTY_BY_DIRECTION.get(var2.rotate(Direction.DOWN)), var1.getValue(DOWN));
+      return (BlockState)((BlockState)((BlockState)((BlockState)((BlockState)((BlockState)var1.setValue((Property)PROPERTY_BY_DIRECTION.get(var2.rotate(Direction.NORTH)), (Boolean)var1.getValue(NORTH))).setValue((Property)PROPERTY_BY_DIRECTION.get(var2.rotate(Direction.SOUTH)), (Boolean)var1.getValue(SOUTH))).setValue((Property)PROPERTY_BY_DIRECTION.get(var2.rotate(Direction.EAST)), (Boolean)var1.getValue(EAST))).setValue((Property)PROPERTY_BY_DIRECTION.get(var2.rotate(Direction.WEST)), (Boolean)var1.getValue(WEST))).setValue((Property)PROPERTY_BY_DIRECTION.get(var2.rotate(Direction.UP)), (Boolean)var1.getValue(UP))).setValue((Property)PROPERTY_BY_DIRECTION.get(var2.rotate(Direction.DOWN)), (Boolean)var1.getValue(DOWN));
    }
 
    public BlockState mirror(BlockState var1, Mirror var2) {
-      return (BlockState)((BlockState)((BlockState)((BlockState)((BlockState)((BlockState)var1.setValue((Property)PROPERTY_BY_DIRECTION.get(var2.mirror(Direction.NORTH)), var1.getValue(NORTH))).setValue((Property)PROPERTY_BY_DIRECTION.get(var2.mirror(Direction.SOUTH)), var1.getValue(SOUTH))).setValue((Property)PROPERTY_BY_DIRECTION.get(var2.mirror(Direction.EAST)), var1.getValue(EAST))).setValue((Property)PROPERTY_BY_DIRECTION.get(var2.mirror(Direction.WEST)), var1.getValue(WEST))).setValue((Property)PROPERTY_BY_DIRECTION.get(var2.mirror(Direction.UP)), var1.getValue(UP))).setValue((Property)PROPERTY_BY_DIRECTION.get(var2.mirror(Direction.DOWN)), var1.getValue(DOWN));
+      return (BlockState)((BlockState)((BlockState)((BlockState)((BlockState)((BlockState)var1.setValue((Property)PROPERTY_BY_DIRECTION.get(var2.mirror(Direction.NORTH)), (Boolean)var1.getValue(NORTH))).setValue((Property)PROPERTY_BY_DIRECTION.get(var2.mirror(Direction.SOUTH)), (Boolean)var1.getValue(SOUTH))).setValue((Property)PROPERTY_BY_DIRECTION.get(var2.mirror(Direction.EAST)), (Boolean)var1.getValue(EAST))).setValue((Property)PROPERTY_BY_DIRECTION.get(var2.mirror(Direction.WEST)), (Boolean)var1.getValue(WEST))).setValue((Property)PROPERTY_BY_DIRECTION.get(var2.mirror(Direction.UP)), (Boolean)var1.getValue(UP))).setValue((Property)PROPERTY_BY_DIRECTION.get(var2.mirror(Direction.DOWN)), (Boolean)var1.getValue(DOWN));
    }
 
    protected void createBlockStateDefinition(StateDefinition.Builder<Block, BlockState> var1) {
       var1.add(UP, DOWN, NORTH, EAST, SOUTH, WEST);
    }
