--- net/minecraft/world/level/block/IronBarsBlock.java
+++ net/minecraft/world/level/block/IronBarsBlock.java
@@ -1,11 +1,10 @@
 package net.minecraft.world.level.block;
 
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.Direction;
 import net.minecraft.tags.BlockTags;
-import net.minecraft.tags.Tag;
 import net.minecraft.world.item.context.BlockPlaceContext;
 import net.minecraft.world.level.BlockGetter;
 import net.minecraft.world.level.Level;
 import net.minecraft.world.level.LevelAccessor;
 import net.minecraft.world.level.block.state.BlockBehaviour;
@@ -49,13 +48,26 @@
 
    public VoxelShape getVisualShape(BlockState var1, BlockGetter var2, BlockPos var3, CollisionContext var4) {
       return Shapes.empty();
    }
 
+   public boolean skipRendering(BlockState var1, BlockState var2, Direction var3) {
+      if (var2.is(this)) {
+         if (!var3.getAxis().isHorizontal()) {
+            return true;
+         }
+
+         if ((Boolean)var1.getValue((Property)PROPERTY_BY_DIRECTION.get(var3)) && (Boolean)var2.getValue((Property)PROPERTY_BY_DIRECTION.get(var3.getOpposite()))) {
+            return true;
+         }
+      }
+
+      return super.skipRendering(var1, var2, var3);
+   }
+
    public final boolean attachsTo(BlockState var1, boolean var2) {
-      Block var3 = var1.getBlock();
-      return !isExceptionForConnection(var3) && var2 || var3 instanceof IronBarsBlock || var3.is((Tag)BlockTags.WALLS);
+      return !isExceptionForConnection(var1) && var2 || var1.getBlock() instanceof IronBarsBlock || var1.is(BlockTags.WALLS);
    }
 
    protected void createBlockStateDefinition(StateDefinition.Builder<Block, BlockState> var1) {
       var1.add(NORTH, EAST, WEST, SOUTH, WATERLOGGED);
    }
