--- net/minecraft/world/level/block/KelpBlock.java
+++ net/minecraft/world/level/block/KelpBlock.java
@@ -15,10 +15,11 @@
 import net.minecraft.world.level.material.Fluids;
 import net.minecraft.world.phys.shapes.VoxelShape;
 
 public class KelpBlock extends GrowingPlantHeadBlock implements LiquidBlockContainer {
    protected static final VoxelShape SHAPE = Block.box(0.0D, 0.0D, 0.0D, 16.0D, 9.0D, 16.0D);
+   private static final double GROW_PER_TICK_PROBABILITY = 0.14D;
 
    protected KelpBlock(BlockBehaviour.Properties var1) {
       super(var1, Direction.UP, SHAPE, true, 0.14D);
    }
 
@@ -28,12 +29,12 @@
 
    protected Block getBodyBlock() {
       return Blocks.KELP_PLANT;
    }
 
-   protected boolean canAttachToBlock(Block var1) {
-      return var1 != Blocks.MAGMA_BLOCK;
+   protected boolean canAttachTo(BlockState var1) {
+      return !var1.is(Blocks.MAGMA_BLOCK);
    }
 
    public boolean canPlaceLiquid(BlockGetter var1, BlockPos var2, BlockState var3, Fluid var4) {
       return false;
    }
