--- net/minecraft/world/level/block/KelpPlantBlock.java
+++ net/minecraft/world/level/block/KelpPlantBlock.java
@@ -22,10 +22,14 @@
 
    public FluidState getFluidState(BlockState var1) {
       return Fluids.WATER.getSource(false);
    }
 
+   protected boolean canAttachTo(BlockState var1) {
+      return this.getHeadBlock().canAttachTo(var1);
+   }
+
    public boolean canPlaceLiquid(BlockGetter var1, BlockPos var2, BlockState var3, Fluid var4) {
       return false;
    }
 
    public boolean placeLiquid(LevelAccessor var1, BlockPos var2, BlockState var3, FluidState var4) {
