--- net/minecraft/world/level/block/LeavesBlock.java
+++ net/minecraft/world/level/block/LeavesBlock.java
@@ -76,11 +76,11 @@
 
       return (BlockState)var0.setValue(DISTANCE, var3);
    }
 
    private static int getDistanceAt(BlockState var0) {
-      if (BlockTags.LOGS.contains(var0.getBlock())) {
+      if (var0.is(BlockTags.LOGS)) {
          return 0;
       } else {
          return var0.getBlock() instanceof LeavesBlock ? (Integer)var0.getValue(DISTANCE) : 7;
       }
    }
