--- net/minecraft/world/level/block/MagmaBlock.java
+++ net/minecraft/world/level/block/MagmaBlock.java
@@ -17,24 +17,26 @@
 import net.minecraft.world.level.LevelAccessor;
 import net.minecraft.world.level.block.state.BlockBehaviour;
 import net.minecraft.world.level.block.state.BlockState;
 
 public class MagmaBlock extends Block {
+   private static final int BUBBLE_COLUMN_CHECK_DELAY = 20;
+
    public MagmaBlock(BlockBehaviour.Properties var1) {
       super(var1);
    }
 
-   public void stepOn(Level var1, BlockPos var2, Entity var3) {
-      if (!var3.fireImmune() && var3 instanceof LivingEntity && !EnchantmentHelper.hasFrostWalker((LivingEntity)var3)) {
-         var3.hurt(DamageSource.HOT_FLOOR, 1.0F);
+   public void stepOn(Level var1, BlockPos var2, BlockState var3, Entity var4) {
+      if (!var4.fireImmune() && var4 instanceof LivingEntity && !EnchantmentHelper.hasFrostWalker((LivingEntity)var4)) {
+         var4.hurt(DamageSource.HOT_FLOOR, 1.0F);
       }
 
-      super.stepOn(var1, var2, var3);
+      super.stepOn(var1, var2, var3, var4);
    }
 
    public void tick(BlockState var1, ServerLevel var2, BlockPos var3, Random var4) {
-      BubbleColumnBlock.growColumn(var2, var3.above(), true);
+      BubbleColumnBlock.updateColumn(var2, var3.above(), var1);
    }
 
    public BlockState updateShape(BlockState var1, Direction var2, BlockState var3, LevelAccessor var4, BlockPos var5, BlockPos var6) {
       if (var2 == Direction.UP && var3.is(Blocks.WATER)) {
          var4.getBlockTicks().scheduleTick(var5, this, 20);
