--- net/minecraft/world/level/block/Mirror.java
+++ net/minecraft/world/level/block/Mirror.java
@@ -1,19 +1,23 @@
 package net.minecraft.world.level.block;
 
 import com.mojang.math.OctahedralGroup;
 import net.minecraft.core.Direction;
+import net.minecraft.network.chat.Component;
+import net.minecraft.network.chat.TranslatableComponent;
 
 public enum Mirror {
-   NONE(OctahedralGroup.IDENTITY),
-   LEFT_RIGHT(OctahedralGroup.INVERT_Z),
-   FRONT_BACK(OctahedralGroup.INVERT_X);
+   NONE(new TranslatableComponent("mirror.none"), OctahedralGroup.IDENTITY),
+   LEFT_RIGHT(new TranslatableComponent("mirror.left_right"), OctahedralGroup.INVERT_Z),
+   FRONT_BACK(new TranslatableComponent("mirror.front_back"), OctahedralGroup.INVERT_X);
 
+   private final Component symbol;
    private final OctahedralGroup rotation;
 
-   private Mirror(OctahedralGroup var3) {
-      this.rotation = var3;
+   private Mirror(Component var3, OctahedralGroup var4) {
+      this.symbol = var3;
+      this.rotation = var4;
    }
 
    public int mirror(int var1, int var2) {
       int var3 = var2 / 2;
       int var4 = var1 > var3 ? var1 - var2 : var1;
