--- net/minecraft/world/level/block/MossBlock.java
+++ net/minecraft/world/level/block/MossBlock.java
@@ -1,0 +1,30 @@
+package net.minecraft.world.level.block;
+
+import java.util.Random;
+import net.minecraft.core.BlockPos;
+import net.minecraft.data.worldgen.Features;
+import net.minecraft.server.level.ServerLevel;
+import net.minecraft.world.level.BlockGetter;
+import net.minecraft.world.level.Level;
+import net.minecraft.world.level.block.state.BlockBehaviour;
+import net.minecraft.world.level.block.state.BlockState;
+import net.minecraft.world.level.levelgen.feature.Feature;
+import net.minecraft.world.level.levelgen.feature.FeaturePlaceContext;
+
+public class MossBlock extends Block implements BonemealableBlock {
+   public MossBlock(BlockBehaviour.Properties var1) {
+      super(var1);
+   }
+
+   public boolean isValidBonemealTarget(BlockGetter var1, BlockPos var2, BlockState var3, boolean var4) {
+      return var1.getBlockState(var2.above()).isAir();
+   }
+
+   public boolean isBonemealSuccess(Level var1, Random var2, BlockPos var3, BlockState var4) {
+      return var1.getBlockState(var3.above()).isAir();
+   }
+
+   public void performBonemeal(ServerLevel var1, Random var2, BlockPos var3, BlockState var4) {
+      Feature.VEGETATION_PATCH.place(new FeaturePlaceContext(var1, var1.getChunkSource().getGenerator(), var2, var3.above(), Features.MOSS_PATCH_BONEMEAL.config()));
+   }
+}
