--- net/minecraft/world/level/block/NetherVines.java
+++ net/minecraft/world/level/block/NetherVines.java
@@ -2,10 +2,16 @@
 
 import java.util.Random;
 import net.minecraft.world.level.block.state.BlockState;
 
 public class NetherVines {
+   private static final double BONEMEAL_GROW_PROBABILITY_DECREASE_RATE = 0.826D;
+   public static final double GROW_PER_TICK_PROBABILITY = 0.1D;
+
+   public NetherVines() {
+   }
+
    public static boolean isValidGrowthState(BlockState var0) {
       return var0.isAir();
    }
 
    public static int getBlocksToGrowWhenBonemealed(Random var0) {
