--- net/minecraft/world/level/block/PipeBlock.java
+++ net/minecraft/world/level/block/PipeBlock.java
@@ -1,7 +1,8 @@
 package net.minecraft.world.level.block;
 
+import com.google.common.collect.ImmutableMap;
 import com.google.common.collect.Maps;
 import java.util.Map;
 import net.minecraft.Util;
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.Direction;
@@ -84,15 +85,15 @@
       EAST = BlockStateProperties.EAST;
       SOUTH = BlockStateProperties.SOUTH;
       WEST = BlockStateProperties.WEST;
       UP = BlockStateProperties.UP;
       DOWN = BlockStateProperties.DOWN;
-      PROPERTY_BY_DIRECTION = (Map)Util.make(Maps.newEnumMap(Direction.class), (var0) -> {
+      PROPERTY_BY_DIRECTION = ImmutableMap.copyOf((Map)Util.make(Maps.newEnumMap(Direction.class), (var0) -> {
          var0.put(Direction.NORTH, NORTH);
          var0.put(Direction.EAST, EAST);
          var0.put(Direction.SOUTH, SOUTH);
          var0.put(Direction.WEST, WEST);
          var0.put(Direction.UP, UP);
          var0.put(Direction.DOWN, DOWN);
-      });
+      }));
    }
 }
