--- net/minecraft/world/level/block/PotatoBlock.java
+++ net/minecraft/world/level/block/PotatoBlock.java
@@ -1,9 +1,11 @@
 package net.minecraft.world.level.block;
 
 import net.minecraft.core.BlockPos;
+import net.minecraft.world.item.Items;
 import net.minecraft.world.level.BlockGetter;
+import net.minecraft.world.level.ItemLike;
 import net.minecraft.world.level.block.state.BlockBehaviour;
 import net.minecraft.world.level.block.state.BlockState;
 import net.minecraft.world.phys.shapes.CollisionContext;
 import net.minecraft.world.phys.shapes.VoxelShape;
 
@@ -12,9 +14,13 @@
 
    public PotatoBlock(BlockBehaviour.Properties var1) {
       super(var1);
    }
 
+   protected ItemLike getBaseSeedId() {
+      return Items.POTATO;
+   }
+
    public VoxelShape getShape(BlockState var1, BlockGetter var2, BlockPos var3, CollisionContext var4) {
       return SHAPE_BY_AGE[(Integer)var1.getValue(this.getAgeProperty())];
    }
 }
