--- net/minecraft/world/level/block/PoweredRailBlock.java
+++ net/minecraft/world/level/block/PoweredRailBlock.java
@@ -15,11 +15,11 @@
    public static final EnumProperty<RailShape> SHAPE;
    public static final BooleanProperty POWERED;
 
    protected PoweredRailBlock(BlockBehaviour.Properties var1) {
       super(true, var1);
-      this.registerDefaultState((BlockState)((BlockState)((BlockState)this.stateDefinition.any()).setValue(SHAPE, RailShape.NORTH_SOUTH)).setValue(POWERED, false));
+      this.registerDefaultState((BlockState)((BlockState)((BlockState)((BlockState)this.stateDefinition.any()).setValue(SHAPE, RailShape.NORTH_SOUTH)).setValue(POWERED, false)).setValue(WATERLOGGED, false));
    }
 
    protected boolean findPoweredRailSignal(Level var1, BlockPos var2, BlockState var3, boolean var4, int var5) {
       if (var5 >= 8) {
          return false;
@@ -247,11 +247,11 @@
 
       return super.mirror(var1, var2);
    }
 
    protected void createBlockStateDefinition(StateDefinition.Builder<Block, BlockState> var1) {
-      var1.add(SHAPE, POWERED);
+      var1.add(SHAPE, POWERED, WATERLOGGED);
    }
 
    static {
       SHAPE = BlockStateProperties.RAIL_SHAPE_STRAIGHT;
       POWERED = BlockStateProperties.POWERED;
