--- net/minecraft/world/level/block/RailBlock.java
+++ net/minecraft/world/level/block/RailBlock.java
@@ -13,11 +13,11 @@
 public class RailBlock extends BaseRailBlock {
    public static final EnumProperty<RailShape> SHAPE;
 
    protected RailBlock(BlockBehaviour.Properties var1) {
       super(false, var1);
-      this.registerDefaultState((BlockState)((BlockState)this.stateDefinition.any()).setValue(SHAPE, RailShape.NORTH_SOUTH));
+      this.registerDefaultState((BlockState)((BlockState)((BlockState)this.stateDefinition.any()).setValue(SHAPE, RailShape.NORTH_SOUTH)).setValue(WATERLOGGED, false));
    }
 
    protected void updateState(BlockState var1, Level var2, BlockPos var3, Block var4) {
       if (var4.defaultBlockState().isSignalSource() && (new RailState(var2, var3, var1)).countPotentialConnections() == 3) {
          this.updateDir(var2, var3, var1, false);
@@ -144,11 +144,11 @@
 
       return super.mirror(var1, var2);
    }
 
    protected void createBlockStateDefinition(StateDefinition.Builder<Block, BlockState> var1) {
-      var1.add(SHAPE);
+      var1.add(SHAPE, WATERLOGGED);
    }
 
    static {
       SHAPE = BlockStateProperties.RAIL_SHAPE;
    }
