--- net/minecraft/world/level/block/RedStoneOreBlock.java
+++ net/minecraft/world/level/block/RedStoneOreBlock.java
@@ -76,10 +76,17 @@
          this.popExperience(var2, var3, var5);
       }
 
    }
 
+   public void animateTick(BlockState var1, Level var2, BlockPos var3, Random var4) {
+      if ((Boolean)var1.getValue(LIT)) {
+         spawnParticles(var2, var3);
+      }
+
+   }
+
    private static void spawnParticles(Level var0, BlockPos var1) {
       double var2 = 0.5625D;
       Random var4 = var0.random;
       Direction[] var5 = Direction.values();
       int var6 = var5.length;
