--- net/minecraft/world/level/block/RedstoneWallTorchBlock.java
+++ net/minecraft/world/level/block/RedstoneWallTorchBlock.java
@@ -1,7 +1,8 @@
 package net.minecraft.world.level.block;
 
+import java.util.Random;
 import javax.annotation.Nullable;
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.Direction;
 import net.minecraft.world.item.context.BlockPlaceContext;
 import net.minecraft.world.level.BlockGetter;
@@ -42,13 +43,24 @@
    }
 
    @Nullable
    public BlockState getStateForPlacement(BlockPlaceContext var1) {
       BlockState var2 = Blocks.WALL_TORCH.getStateForPlacement(var1);
-      return var2 == null ? null : (BlockState)this.defaultBlockState().setValue(FACING, var2.getValue(FACING));
+      return var2 == null ? null : (BlockState)this.defaultBlockState().setValue(FACING, (Direction)var2.getValue(FACING));
    }
 
+   public void animateTick(BlockState var1, Level var2, BlockPos var3, Random var4) {
+      if ((Boolean)var1.getValue(LIT)) {
+         Direction var5 = ((Direction)var1.getValue(FACING)).getOpposite();
+         double var6 = 0.27D;
+         double var8 = (double)var3.getX() + 0.5D + (var4.nextDouble() - 0.5D) * 0.2D + 0.27D * (double)var5.getStepX();
+         double var10 = (double)var3.getY() + 0.7D + (var4.nextDouble() - 0.5D) * 0.2D + 0.22D;
+         double var12 = (double)var3.getZ() + 0.5D + (var4.nextDouble() - 0.5D) * 0.2D + 0.27D * (double)var5.getStepZ();
+         var2.addParticle(this.flameParticle, var8, var10, var12, 0.0D, 0.0D, 0.0D);
+      }
+   }
+
    protected boolean hasNeighborSignal(Level var1, BlockPos var2, BlockState var3) {
       Direction var4 = ((Direction)var3.getValue(FACING)).getOpposite();
       return var1.hasSignal(var2.relative(var4), var4);
    }
 
