--- net/minecraft/world/level/block/RepeaterBlock.java
+++ net/minecraft/world/level/block/RepeaterBlock.java
@@ -25,11 +25,11 @@
       super(var1);
       this.registerDefaultState((BlockState)((BlockState)((BlockState)((BlockState)((BlockState)this.stateDefinition.any()).setValue(FACING, Direction.NORTH)).setValue(DELAY, 1)).setValue(LOCKED, false)).setValue(POWERED, false));
    }
 
    public InteractionResult use(BlockState var1, Level var2, BlockPos var3, Player var4, InteractionHand var5, BlockHitResult var6) {
-      if (!var4.abilities.mayBuild) {
+      if (!var4.getAbilities().mayBuild) {
          return InteractionResult.PASS;
       } else {
          var2.setBlock(var3, (BlockState)var1.cycle(DELAY), 3);
          return InteractionResult.sidedSuccess(var2.isClientSide);
       }
