--- net/minecraft/world/level/block/SaplingBlock.java
+++ net/minecraft/world/level/block/SaplingBlock.java
@@ -14,10 +14,11 @@
 import net.minecraft.world.phys.shapes.CollisionContext;
 import net.minecraft.world.phys.shapes.VoxelShape;
 
 public class SaplingBlock extends BushBlock implements BonemealableBlock {
    public static final IntegerProperty STAGE;
+   protected static final float AABB_OFFSET = 6.0F;
    protected static final VoxelShape SHAPE;
    private final AbstractTreeGrower treeGrower;
 
    protected SaplingBlock(AbstractTreeGrower var1, BlockBehaviour.Properties var2) {
       super(var2);
