--- net/minecraft/world/level/block/ScaffoldingBlock.java
+++ net/minecraft/world/level/block/ScaffoldingBlock.java
@@ -52,11 +52,11 @@
    public VoxelShape getInteractionShape(BlockState var1, BlockGetter var2, BlockPos var3) {
       return Shapes.block();
    }
 
    public boolean canBeReplaced(BlockState var1, BlockPlaceContext var2) {
-      return var2.getItemInHand().getItem() == this.asItem();
+      return var2.getItemInHand().is(this.asItem());
    }
 
    public BlockState getStateForPlacement(BlockPlaceContext var1) {
       BlockPos var2 = var1.getClickedPos();
       Level var3 = var1.getLevel();
