--- net/minecraft/world/level/block/ShulkerBoxBlock.java
+++ net/minecraft/world/level/block/ShulkerBoxBlock.java
@@ -1,37 +1,50 @@
 package net.minecraft.world.level.block;
 
+import java.util.Iterator;
 import java.util.List;
 import javax.annotation.Nullable;
+import net.minecraft.ChatFormatting;
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.Direction;
+import net.minecraft.core.NonNullList;
 import net.minecraft.nbt.CompoundTag;
+import net.minecraft.network.chat.Component;
+import net.minecraft.network.chat.MutableComponent;
+import net.minecraft.network.chat.TextComponent;
+import net.minecraft.network.chat.TranslatableComponent;
 import net.minecraft.resources.ResourceLocation;
 import net.minecraft.stats.Stats;
 import net.minecraft.world.Container;
+import net.minecraft.world.ContainerHelper;
 import net.minecraft.world.InteractionHand;
 import net.minecraft.world.InteractionResult;
-import net.minecraft.world.ShulkerSharedHelper;
 import net.minecraft.world.entity.LivingEntity;
 import net.minecraft.world.entity.item.ItemEntity;
+import net.minecraft.world.entity.monster.Shulker;
 import net.minecraft.world.entity.monster.piglin.PiglinAi;
 import net.minecraft.world.entity.player.Player;
 import net.minecraft.world.inventory.AbstractContainerMenu;
 import net.minecraft.world.item.DyeColor;
+import net.minecraft.world.item.Item;
 import net.minecraft.world.item.ItemStack;
+import net.minecraft.world.item.TooltipFlag;
 import net.minecraft.world.item.context.BlockPlaceContext;
 import net.minecraft.world.level.BlockGetter;
 import net.minecraft.world.level.Level;
 import net.minecraft.world.level.block.entity.BlockEntity;
+import net.minecraft.world.level.block.entity.BlockEntityTicker;
+import net.minecraft.world.level.block.entity.BlockEntityType;
 import net.minecraft.world.level.block.entity.ShulkerBoxBlockEntity;
 import net.minecraft.world.level.block.state.BlockBehaviour;
 import net.minecraft.world.level.block.state.BlockState;
 import net.minecraft.world.level.block.state.StateDefinition;
 import net.minecraft.world.level.block.state.properties.EnumProperty;
 import net.minecraft.world.level.material.PushReaction;
 import net.minecraft.world.level.storage.loot.LootContext;
 import net.minecraft.world.level.storage.loot.parameters.LootContextParams;
+import net.minecraft.world.phys.AABB;
 import net.minecraft.world.phys.BlockHitResult;
 import net.minecraft.world.phys.shapes.CollisionContext;
 import net.minecraft.world.phys.shapes.Shapes;
 import net.minecraft.world.phys.shapes.VoxelShape;
 
@@ -45,14 +58,19 @@
       super(var2);
       this.color = var1;
       this.registerDefaultState((BlockState)((BlockState)this.stateDefinition.any()).setValue(FACING, Direction.UP));
    }
 
-   public BlockEntity newBlockEntity(BlockGetter var1) {
-      return new ShulkerBoxBlockEntity(this.color);
+   public BlockEntity newBlockEntity(BlockPos var1, BlockState var2) {
+      return new ShulkerBoxBlockEntity(this.color, var1, var2);
    }
 
+   @Nullable
+   public <T extends BlockEntity> BlockEntityTicker<T> getTicker(Level var1, BlockState var2, BlockEntityType<T> var3) {
+      return createTickerHelper(var3, BlockEntityType.SHULKER_BOX, ShulkerBoxBlockEntity::tick);
+   }
+
    public RenderShape getRenderShape(BlockState var1) {
       return RenderShape.ENTITYBLOCK_ANIMATED;
    }
 
    public InteractionResult use(BlockState var1, Level var2, BlockPos var3, Player var4, InteractionHand var5, BlockHitResult var6) {
@@ -62,19 +80,11 @@
          return InteractionResult.CONSUME;
       } else {
          BlockEntity var7 = var2.getBlockEntity(var3);
          if (var7 instanceof ShulkerBoxBlockEntity) {
             ShulkerBoxBlockEntity var8 = (ShulkerBoxBlockEntity)var7;
-            boolean var9;
-            if (var8.getAnimationStatus() == ShulkerBoxBlockEntity.AnimationStatus.CLOSED) {
-               Direction var10 = (Direction)var1.getValue(FACING);
-               var9 = var2.noCollision(ShulkerSharedHelper.openBoundingBox(var3, var10));
-            } else {
-               var9 = true;
-            }
-
-            if (var9) {
+            if (canOpen(var1, var2, var3, var8)) {
                var4.openMenu(var8);
                var4.awardStat(Stats.OPEN_SHULKER_BOX);
                PiglinAi.angerNearbyPiglins(var4, true);
             }
 
@@ -83,10 +93,19 @@
             return InteractionResult.PASS;
          }
       }
    }
 
+   private static boolean canOpen(BlockState var0, Level var1, BlockPos var2, ShulkerBoxBlockEntity var3) {
+      if (var3.getAnimationStatus() != ShulkerBoxBlockEntity.AnimationStatus.CLOSED) {
+         return true;
+      } else {
+         AABB var4 = Shulker.getProgressDeltaAabb((Direction)var0.getValue(FACING), 0.0F, 0.5F).move(var2).deflate(1.0E-6D);
+         return var1.noCollision(var4);
+      }
+   }
+
    public BlockState getStateForPlacement(BlockPlaceContext var1) {
       return (BlockState)this.defaultBlockState().setValue(FACING, var1.getClickedFace());
    }
 
    protected void createBlockStateDefinition(StateDefinition.Builder<Block, BlockState> var1) {
@@ -153,10 +172,46 @@
 
          super.onRemove(var1, var2, var3, var4, var5);
       }
    }
 
+   public void appendHoverText(ItemStack var1, @Nullable BlockGetter var2, List<Component> var3, TooltipFlag var4) {
+      super.appendHoverText(var1, var2, var3, var4);
+      CompoundTag var5 = var1.getTagElement("BlockEntityTag");
+      if (var5 != null) {
+         if (var5.contains("LootTable", 8)) {
+            var3.add(new TextComponent("???????"));
+         }
+
+         if (var5.contains("Items", 9)) {
+            NonNullList var6 = NonNullList.withSize(27, ItemStack.EMPTY);
+            ContainerHelper.loadAllItems(var5, var6);
+            int var7 = 0;
+            int var8 = 0;
+            Iterator var9 = var6.iterator();
+
+            while(var9.hasNext()) {
+               ItemStack var10 = (ItemStack)var9.next();
+               if (!var10.isEmpty()) {
+                  ++var8;
+                  if (var7 <= 4) {
+                     ++var7;
+                     MutableComponent var11 = var10.getHoverName().copy();
+                     var11.append(" x").append(String.valueOf(var10.getCount()));
+                     var3.add(var11);
+                  }
+               }
+            }
+
+            if (var8 - var7 > 0) {
+               var3.add((new TranslatableComponent("container.shulkerBox.more", new Object[]{var8 - var7})).withStyle(ChatFormatting.ITALIC));
+            }
+         }
+      }
+
+   }
+
    public PushReaction getPistonPushReaction(BlockState var1) {
       return PushReaction.DESTROY;
    }
 
    public VoxelShape getShape(BlockState var1, BlockGetter var2, BlockPos var3, CollisionContext var4) {
@@ -170,10 +225,31 @@
 
    public int getAnalogOutputSignal(BlockState var1, Level var2, BlockPos var3) {
       return AbstractContainerMenu.getRedstoneSignalFromContainer((Container)var2.getBlockEntity(var3));
    }
 
+   public ItemStack getCloneItemStack(BlockGetter var1, BlockPos var2, BlockState var3) {
+      ItemStack var4 = super.getCloneItemStack(var1, var2, var3);
+      ShulkerBoxBlockEntity var5 = (ShulkerBoxBlockEntity)var1.getBlockEntity(var2);
+      CompoundTag var6 = var5.saveToTag(new CompoundTag());
+      if (!var6.isEmpty()) {
+         var4.addTagElement("BlockEntityTag", var6);
+      }
+
+      return var4;
+   }
+
+   @Nullable
+   public static DyeColor getColorFromItem(Item var0) {
+      return getColorFromBlock(Block.byItem(var0));
+   }
+
+   @Nullable
+   public static DyeColor getColorFromBlock(Block var0) {
+      return var0 instanceof ShulkerBoxBlock ? ((ShulkerBoxBlock)var0).getColor() : null;
+   }
+
    public static Block getBlockByColor(@Nullable DyeColor var0) {
       if (var0 == null) {
          return Blocks.SHULKER_BOX;
       } else {
          switch(var0) {
