--- net/minecraft/world/level/block/SimpleWaterloggedBlock.java
+++ net/minecraft/world/level/block/SimpleWaterloggedBlock.java
@@ -1,8 +1,12 @@
 package net.minecraft.world.level.block;
 
+import java.util.Optional;
 import net.minecraft.core.BlockPos;
+import net.minecraft.sounds.SoundEvent;
+import net.minecraft.world.item.ItemStack;
+import net.minecraft.world.item.Items;
 import net.minecraft.world.level.BlockGetter;
 import net.minecraft.world.level.LevelAccessor;
 import net.minecraft.world.level.block.state.BlockState;
 import net.minecraft.world.level.block.state.properties.BlockStateProperties;
 import net.minecraft.world.level.material.Fluid;
@@ -25,14 +29,22 @@
       } else {
          return false;
       }
    }
 
-   default Fluid takeLiquid(LevelAccessor var1, BlockPos var2, BlockState var3) {
+   default ItemStack pickupBlock(LevelAccessor var1, BlockPos var2, BlockState var3) {
       if ((Boolean)var3.getValue(BlockStateProperties.WATERLOGGED)) {
          var1.setBlock(var2, (BlockState)var3.setValue(BlockStateProperties.WATERLOGGED, false), 3);
-         return Fluids.WATER;
+         if (!var3.canSurvive(var1, var2)) {
+            var1.destroyBlock(var2, true);
+         }
+
+         return new ItemStack(Items.WATER_BUCKET);
       } else {
-         return Fluids.EMPTY;
+         return ItemStack.EMPTY;
       }
    }
+
+   default Optional<SoundEvent> getPickupSound() {
+      return Fluids.WATER.getPickupSound();
+   }
 }
