--- net/minecraft/world/level/block/SlimeBlock.java
+++ net/minecraft/world/level/block/SlimeBlock.java
@@ -1,8 +1,9 @@
 package net.minecraft.world.level.block;
 
 import net.minecraft.core.BlockPos;
+import net.minecraft.world.damagesource.DamageSource;
 import net.minecraft.world.entity.Entity;
 import net.minecraft.world.entity.LivingEntity;
 import net.minecraft.world.level.BlockGetter;
 import net.minecraft.world.level.Level;
 import net.minecraft.world.level.block.state.BlockBehaviour;
@@ -15,11 +16,11 @@
 
    public void fallOn(Level var1, BlockPos var2, Entity var3, float var4) {
       if (var3.isSuppressingBounce()) {
          super.fallOn(var1, var2, var3, var4);
       } else {
-         var3.causeFallDamage(var4, 0.0F);
+         var3.causeFallDamage(var4, 0.0F, DamageSource.FALL);
       }
 
    }
 
    public void updateEntityAfterFallOn(BlockGetter var1, Entity var2) {
