--- net/minecraft/world/level/block/SoulSandBlock.java
+++ net/minecraft/world/level/block/SoulSandBlock.java
@@ -32,11 +32,11 @@
    public VoxelShape getVisualShape(BlockState var1, BlockGetter var2, BlockPos var3, CollisionContext var4) {
       return Shapes.block();
    }
 
    public void tick(BlockState var1, ServerLevel var2, BlockPos var3, Random var4) {
-      BubbleColumnBlock.growColumn(var2, var3.above(), false);
+      BubbleColumnBlock.updateColumn(var2, var3.above(), var1);
    }
 
    public BlockState updateShape(BlockState var1, Direction var2, BlockState var3, LevelAccessor var4, BlockPos var5, BlockPos var6) {
       if (var2 == Direction.UP && var3.is(Blocks.WATER)) {
          var4.getBlockTicks().scheduleTick(var5, this, 20);
