--- net/minecraft/world/level/block/SpongeBlock.java
+++ net/minecraft/world/level/block/SpongeBlock.java
@@ -9,14 +9,16 @@
 import net.minecraft.world.level.Level;
 import net.minecraft.world.level.block.entity.BlockEntity;
 import net.minecraft.world.level.block.state.BlockBehaviour;
 import net.minecraft.world.level.block.state.BlockState;
 import net.minecraft.world.level.material.FluidState;
-import net.minecraft.world.level.material.Fluids;
 import net.minecraft.world.level.material.Material;
 
 public class SpongeBlock extends Block {
+   public static final int MAX_DEPTH = 6;
+   public static final int MAX_COUNT = 64;
+
    protected SpongeBlock(BlockBehaviour.Properties var1) {
       super(var1);
    }
 
    public void onPlace(BlockState var1, Level var2, BlockPos var3, BlockState var4, boolean var5) {
@@ -55,11 +57,11 @@
             BlockPos var12 = var6.relative(var11);
             BlockState var13 = var1.getBlockState(var12);
             FluidState var14 = var1.getFluidState(var12);
             Material var15 = var13.getMaterial();
             if (var14.is(FluidTags.WATER)) {
-               if (var13.getBlock() instanceof BucketPickup && ((BucketPickup)var13.getBlock()).takeLiquid(var1, var12, var13) != Fluids.EMPTY) {
+               if (var13.getBlock() instanceof BucketPickup && !((BucketPickup)var13.getBlock()).pickupBlock(var1, var12, var13).isEmpty()) {
                   ++var4;
                   if (var7 < 6) {
                      var3.add(new Tuple(var12, var7 + 1));
                   }
                } else if (var13.getBlock() instanceof LiquidBlock) {
@@ -67,11 +69,11 @@
                   ++var4;
                   if (var7 < 6) {
                      var3.add(new Tuple(var12, var7 + 1));
                   }
                } else if (var15 == Material.WATER_PLANT || var15 == Material.REPLACEABLE_WATER_PLANT) {
-                  BlockEntity var16 = var13.getBlock().isEntityBlock() ? var1.getBlockEntity(var12) : null;
+                  BlockEntity var16 = var13.hasBlockEntity() ? var1.getBlockEntity(var12) : null;
                   dropResources(var13, var1, var12, var16);
                   var1.setBlock(var12, Blocks.AIR.defaultBlockState(), 3);
                   ++var4;
                   if (var7 < 6) {
                      var3.add(new Tuple(var12, var7 + 1));
