--- net/minecraft/world/level/block/StructureVoidBlock.java
+++ net/minecraft/world/level/block/StructureVoidBlock.java
@@ -7,10 +7,11 @@
 import net.minecraft.world.level.material.PushReaction;
 import net.minecraft.world.phys.shapes.CollisionContext;
 import net.minecraft.world.phys.shapes.VoxelShape;
 
 public class StructureVoidBlock extends Block {
+   private static final double SIZE = 5.0D;
    private static final VoxelShape SHAPE = Block.box(5.0D, 5.0D, 5.0D, 11.0D, 11.0D, 11.0D);
 
    protected StructureVoidBlock(BlockBehaviour.Properties var1) {
       super(var1);
    }
@@ -21,9 +22,13 @@
 
    public VoxelShape getShape(BlockState var1, BlockGetter var2, BlockPos var3, CollisionContext var4) {
       return SHAPE;
    }
 
+   public float getShadeBrightness(BlockState var1, BlockGetter var2, BlockPos var3) {
+      return 1.0F;
+   }
+
    public PushReaction getPistonPushReaction(BlockState var1) {
       return PushReaction.DESTROY;
    }
 }
