--- net/minecraft/world/level/block/SweetBerryBushBlock.java
+++ net/minecraft/world/level/block/SweetBerryBushBlock.java
@@ -25,19 +25,25 @@
 import net.minecraft.world.phys.Vec3;
 import net.minecraft.world.phys.shapes.CollisionContext;
 import net.minecraft.world.phys.shapes.VoxelShape;
 
 public class SweetBerryBushBlock extends BushBlock implements BonemealableBlock {
+   private static final float HURT_SPEED_THRESHOLD = 0.003F;
+   public static final int MAX_AGE = 3;
    public static final IntegerProperty AGE;
    private static final VoxelShape SAPLING_SHAPE;
    private static final VoxelShape MID_GROWTH_SHAPE;
 
    public SweetBerryBushBlock(BlockBehaviour.Properties var1) {
       super(var1);
       this.registerDefaultState((BlockState)((BlockState)this.stateDefinition.any()).setValue(AGE, 0));
    }
 
+   public ItemStack getCloneItemStack(BlockGetter var1, BlockPos var2, BlockState var3) {
+      return new ItemStack(Items.SWEET_BERRIES);
+   }
+
    public VoxelShape getShape(BlockState var1, BlockGetter var2, BlockPos var3, CollisionContext var4) {
       if ((Integer)var1.getValue(AGE) == 0) {
          return SAPLING_SHAPE;
       } else {
          return (Integer)var1.getValue(AGE) < 3 ? MID_GROWTH_SHAPE : super.getShape(var1, var2, var3, var4);
@@ -71,11 +77,11 @@
    }
 
    public InteractionResult use(BlockState var1, Level var2, BlockPos var3, Player var4, InteractionHand var5, BlockHitResult var6) {
       int var7 = (Integer)var1.getValue(AGE);
       boolean var8 = var7 == 3;
-      if (!var8 && var4.getItemInHand(var5).getItem() == Items.BONE_MEAL) {
+      if (!var8 && var4.getItemInHand(var5).is(Items.BONE_MEAL)) {
          return InteractionResult.PASS;
       } else if (var7 > 1) {
          int var9 = 1 + var2.random.nextInt(2);
          popResource(var2, var3, new ItemStack(Items.SWEET_BERRIES, var9 + (var8 ? 1 : 0)));
          var2.playSound((Player)null, (BlockPos)var3, SoundEvents.SWEET_BERRY_BUSH_PICK_BERRIES, SoundSource.BLOCKS, 1.0F, 0.8F + var2.random.nextFloat() * 0.4F);
