--- net/minecraft/world/level/block/TargetBlock.java
+++ net/minecraft/world/level/block/TargetBlock.java
@@ -22,10 +22,12 @@
 import net.minecraft.world.phys.BlockHitResult;
 import net.minecraft.world.phys.Vec3;
 
 public class TargetBlock extends Block {
    private static final IntegerProperty OUTPUT_POWER;
+   private static final int ACTIVATION_TICKS_ARROWS = 20;
+   private static final int ACTIVATION_TICKS_OTHER = 8;
 
    public TargetBlock(BlockBehaviour.Properties var1) {
       super(var1);
       this.registerDefaultState((BlockState)((BlockState)this.stateDefinition.any()).setValue(OUTPUT_POWER, 0));
    }
