--- net/minecraft/world/level/block/TorchBlock.java
+++ net/minecraft/world/level/block/TorchBlock.java
@@ -1,19 +1,23 @@
 package net.minecraft.world.level.block;
 
+import java.util.Random;
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.Direction;
 import net.minecraft.core.particles.ParticleOptions;
+import net.minecraft.core.particles.ParticleTypes;
 import net.minecraft.world.level.BlockGetter;
+import net.minecraft.world.level.Level;
 import net.minecraft.world.level.LevelAccessor;
 import net.minecraft.world.level.LevelReader;
 import net.minecraft.world.level.block.state.BlockBehaviour;
 import net.minecraft.world.level.block.state.BlockState;
 import net.minecraft.world.phys.shapes.CollisionContext;
 import net.minecraft.world.phys.shapes.VoxelShape;
 
 public class TorchBlock extends Block {
+   protected static final int AABB_STANDING_OFFSET = 2;
    protected static final VoxelShape AABB = Block.box(6.0D, 0.0D, 6.0D, 10.0D, 10.0D, 10.0D);
    protected final ParticleOptions flameParticle;
 
    protected TorchBlock(BlockBehaviour.Properties var1, ParticleOptions var2) {
       super(var1);
@@ -29,6 +33,14 @@
    }
 
    public boolean canSurvive(BlockState var1, LevelReader var2, BlockPos var3) {
       return canSupportCenter(var2, var3.below(), Direction.UP);
    }
+
+   public void animateTick(BlockState var1, Level var2, BlockPos var3, Random var4) {
+      double var5 = (double)var3.getX() + 0.5D;
+      double var7 = (double)var3.getY() + 0.7D;
+      double var9 = (double)var3.getZ() + 0.5D;
+      var2.addParticle(ParticleTypes.SMOKE, var5, var7, var9, 0.0D, 0.0D, 0.0D);
+      var2.addParticle(this.flameParticle, var5, var7, var9, 0.0D, 0.0D, 0.0D);
+   }
 }
