--- net/minecraft/world/level/block/TrappedChestBlock.java
+++ net/minecraft/world/level/block/TrappedChestBlock.java
@@ -19,12 +19,12 @@
       super(var1, () -> {
          return BlockEntityType.TRAPPED_CHEST;
       });
    }
 
-   public BlockEntity newBlockEntity(BlockGetter var1) {
-      return new TrappedChestBlockEntity();
+   public BlockEntity newBlockEntity(BlockPos var1, BlockState var2) {
+      return new TrappedChestBlockEntity(var1, var2);
    }
 
    protected Stat<ResourceLocation> getOpenChestStat() {
       return Stats.CUSTOM.get(Stats.TRIGGER_TRAPPED_CHEST);
    }
@@ -32,11 +32,11 @@
    public boolean isSignalSource(BlockState var1) {
       return true;
    }
 
    public int getSignal(BlockState var1, BlockGetter var2, BlockPos var3, Direction var4) {
-      return Mth.clamp(ChestBlockEntity.getOpenCount(var2, var3), 0, 15);
+      return Mth.clamp((int)ChestBlockEntity.getOpenCount(var2, var3), (int)0, (int)15);
    }
 
    public int getDirectSignal(BlockState var1, BlockGetter var2, BlockPos var3, Direction var4) {
       return var4 == Direction.UP ? var1.getSignal(var2, var3, var4) : 0;
    }
