--- net/minecraft/world/level/block/TripWireBlock.java
+++ net/minecraft/world/level/block/TripWireBlock.java
@@ -67,11 +67,11 @@
          this.updateSource(var2, var3, (BlockState)var1.setValue(POWERED, true));
       }
    }
 
    public void playerWillDestroy(Level var1, BlockPos var2, BlockState var3, Player var4) {
-      if (!var1.isClientSide && !var4.getMainHandItem().isEmpty() && var4.getMainHandItem().getItem() == Items.SHEARS) {
+      if (!var1.isClientSide && !var4.getMainHandItem().isEmpty() && var4.getMainHandItem().is(Items.SHEARS)) {
          var1.setBlock(var2, (BlockState)var3.setValue(DISARMED, true), 4);
       }
 
       super.playerWillDestroy(var1, var2, var3, var4);
    }
@@ -143,15 +143,14 @@
       }
 
    }
 
    public boolean shouldConnectTo(BlockState var1, Direction var2) {
-      Block var3 = var1.getBlock();
-      if (var3 == this.hook) {
+      if (var1.is(this.hook)) {
          return var1.getValue(TripWireHookBlock.FACING) == var2.getOpposite();
       } else {
-         return var3 == this;
+         return var1.is(this);
       }
    }
 
    public BlockState rotate(BlockState var1, Rotation var2) {
       switch(var2) {
