--- net/minecraft/world/level/block/WallSignBlock.java
+++ net/minecraft/world/level/block/WallSignBlock.java
@@ -25,11 +25,11 @@
    public static final DirectionProperty FACING;
    private static final Map<Direction, VoxelShape> AABBS;
 
    public WallSignBlock(BlockBehaviour.Properties var1, WoodType var2) {
       super(var1, var2);
-      this.registerDefaultState((BlockState)((BlockState)((BlockState)this.stateDefinition.any()).setValue(FACING, Direction.NORTH)).setValue(WATERLOGGED, false));
+      this.registerDefaultState((BlockState)((BlockState)((BlockState)((BlockState)this.stateDefinition.any()).setValue(FACING, Direction.NORTH)).setValue(WATERLOGGED, false)).setValue(LIT, false));
    }
 
    public String getDescriptionId() {
       return this.asItem().getDescriptionId();
    }
@@ -77,11 +77,11 @@
    public BlockState mirror(BlockState var1, Mirror var2) {
       return var1.rotate(var2.getRotation((Direction)var1.getValue(FACING)));
    }
 
    protected void createBlockStateDefinition(StateDefinition.Builder<Block, BlockState> var1) {
-      var1.add(FACING, WATERLOGGED);
+      var1.add(FACING, WATERLOGGED, LIT);
    }
 
    static {
       FACING = HorizontalDirectionalBlock.FACING;
       AABBS = Maps.newEnumMap((Map)ImmutableMap.of(Direction.NORTH, Block.box(0.0D, 4.5D, 14.0D, 16.0D, 12.5D, 16.0D), Direction.SOUTH, Block.box(0.0D, 4.5D, 0.0D, 16.0D, 12.5D, 2.0D), Direction.EAST, Block.box(0.0D, 4.5D, 0.0D, 2.0D, 12.5D, 16.0D), Direction.WEST, Block.box(14.0D, 4.5D, 0.0D, 16.0D, 12.5D, 16.0D)));
