--- net/minecraft/ChatFormatting.java
+++ net/minecraft/ChatFormatting.java
@@ -32,10 +32,11 @@
    STRIKETHROUGH("STRIKETHROUGH", 'm', true),
    UNDERLINE("UNDERLINE", 'n', true),
    ITALIC("ITALIC", 'o', true),
    RESET("RESET", 'r', -1, (Integer)null);
 
+   public static final char PREFIX_CODE = '\u00a7';
    private static final Map<String, ChatFormatting> FORMATTING_BY_NAME = (Map)Arrays.stream(values()).collect(Collectors.toMap((var0) -> {
       return cleanName(var0.name);
    }, (var0) -> {
       return var0;
    }));
@@ -67,10 +68,14 @@
       this.id = var6;
       this.color = var7;
       this.toString = "\u00a7" + var4;
    }
 
+   public char getChar() {
+      return this.code;
+   }
+
    public int getId() {
       return this.id;
    }
 
    public boolean isFormat() {
