--- net/minecraft/world/entity/ai/targeting/TargetingConditions.java
+++ net/minecraft/world/entity/ai/targeting/TargetingConditions.java
@@ -5,10 +5,11 @@
 import net.minecraft.world.entity.LivingEntity;
 import net.minecraft.world.entity.Mob;
 
 public class TargetingConditions {
    public static final TargetingConditions DEFAULT = new TargetingConditions();
+   private static final double MIN_VISIBILITY_DISTANCE_FOR_INVISIBLE_TARGET = 2.0D;
    private double range = -1.0D;
    private boolean allowInvulnerable;
    private boolean allowSameTeam;
    private boolean allowUnseeable;
    private boolean allowNonAttackable;
@@ -16,10 +17,22 @@
    private Predicate<LivingEntity> selector;
 
    public TargetingConditions() {
    }
 
+   public TargetingConditions copy() {
+      TargetingConditions var1 = new TargetingConditions();
+      var1.range = this.range;
+      var1.allowInvulnerable = this.allowInvulnerable;
+      var1.allowSameTeam = this.allowSameTeam;
+      var1.allowUnseeable = this.allowUnseeable;
+      var1.allowNonAttackable = this.allowNonAttackable;
+      var1.testInvisible = this.testInvisible;
+      var1.selector = this.selector;
+      return var1;
+   }
+
    public TargetingConditions range(double var1) {
       this.range = var1;
       return this;
    }
 
