--- net/minecraft/world/level/levelgen/feature/blockplacers/DoublePlantPlacer.java
+++ net/minecraft/world/level/levelgen/feature/blockplacers/DoublePlantPlacer.java
@@ -19,8 +19,8 @@
    protected BlockPlacerType<?> type() {
       return BlockPlacerType.DOUBLE_PLANT_PLACER;
    }
 
    public void place(LevelAccessor var1, BlockPos var2, BlockState var3, Random var4) {
-      ((DoublePlantBlock)var3.getBlock()).placeAt(var1, var2, 2);
+      ((DoublePlantBlock)var3.getBlock()).placeAt(var1, var3, var2, 2);
    }
 }
