--- net/minecraft/advancements/critereon/ChanneledLightningTrigger.java
+++ net/minecraft/advancements/critereon/ChanneledLightningTrigger.java
@@ -10,11 +10,11 @@
 import net.minecraft.server.level.ServerPlayer;
 import net.minecraft.world.entity.Entity;
 import net.minecraft.world.level.storage.loot.LootContext;
 
 public class ChanneledLightningTrigger extends SimpleCriterionTrigger<ChanneledLightningTrigger.TriggerInstance> {
-   private static final ResourceLocation ID = new ResourceLocation("channeled_lightning");
+   static final ResourceLocation ID = new ResourceLocation("channeled_lightning");
 
    public ChanneledLightningTrigger() {
    }
 
    public ResourceLocation getId() {
