--- net/minecraft/advancements/critereon/ChangeDimensionTrigger.java
+++ net/minecraft/advancements/critereon/ChangeDimensionTrigger.java
@@ -41,14 +41,26 @@
          super(ChangeDimensionTrigger.ID, var1);
          this.from = var2;
          this.to = var3;
       }
 
+      public static ChangeDimensionTrigger.TriggerInstance changedDimension() {
+         return new ChangeDimensionTrigger.TriggerInstance(EntityPredicate.Composite.ANY, (ResourceKey)null, (ResourceKey)null);
+      }
+
+      public static ChangeDimensionTrigger.TriggerInstance changedDimension(ResourceKey<Level> var0, ResourceKey<Level> var1) {
+         return new ChangeDimensionTrigger.TriggerInstance(EntityPredicate.Composite.ANY, var0, var1);
+      }
+
       public static ChangeDimensionTrigger.TriggerInstance changedDimensionTo(ResourceKey<Level> var0) {
          return new ChangeDimensionTrigger.TriggerInstance(EntityPredicate.Composite.ANY, (ResourceKey)null, var0);
       }
 
+      public static ChangeDimensionTrigger.TriggerInstance changedDimensionFrom(ResourceKey<Level> var0) {
+         return new ChangeDimensionTrigger.TriggerInstance(EntityPredicate.Composite.ANY, var0, (ResourceKey)null);
+      }
+
       public boolean matches(ResourceKey<Level> var1, ResourceKey<Level> var2) {
          if (this.from != null && this.from != var1) {
             return false;
          } else {
             return this.to == null || this.to == var2;
