--- net/minecraft/advancements/critereon/DamagePredicate.java
+++ net/minecraft/advancements/critereon/DamagePredicate.java
@@ -95,15 +95,35 @@
 
       public static DamagePredicate.Builder damageInstance() {
          return new DamagePredicate.Builder();
       }
 
+      public DamagePredicate.Builder dealtDamage(MinMaxBounds.Floats var1) {
+         this.dealtDamage = var1;
+         return this;
+      }
+
+      public DamagePredicate.Builder takenDamage(MinMaxBounds.Floats var1) {
+         this.takenDamage = var1;
+         return this;
+      }
+
+      public DamagePredicate.Builder sourceEntity(EntityPredicate var1) {
+         this.sourceEntity = var1;
+         return this;
+      }
+
       public DamagePredicate.Builder blocked(Boolean var1) {
          this.blocked = var1;
          return this;
       }
 
+      public DamagePredicate.Builder type(DamageSourcePredicate var1) {
+         this.type = var1;
+         return this;
+      }
+
       public DamagePredicate.Builder type(DamageSourcePredicate.Builder var1) {
          this.type = var1.build();
          return this;
       }
 
