--- net/minecraft/advancements/critereon/EnchantedItemTrigger.java
+++ net/minecraft/advancements/critereon/EnchantedItemTrigger.java
@@ -4,11 +4,11 @@
 import net.minecraft.resources.ResourceLocation;
 import net.minecraft.server.level.ServerPlayer;
 import net.minecraft.world.item.ItemStack;
 
 public class EnchantedItemTrigger extends SimpleCriterionTrigger<EnchantedItemTrigger.TriggerInstance> {
-   private static final ResourceLocation ID = new ResourceLocation("enchanted_item");
+   static final ResourceLocation ID = new ResourceLocation("enchanted_item");
 
    public EnchantedItemTrigger() {
    }
 
    public ResourceLocation getId() {
