--- net/minecraft/advancements/critereon/EntityEquipmentPredicate.java
+++ net/minecraft/advancements/critereon/EntityEquipmentPredicate.java
@@ -128,10 +128,20 @@
       public EntityEquipmentPredicate.Builder feet(ItemPredicate var1) {
          this.feet = var1;
          return this;
       }
 
+      public EntityEquipmentPredicate.Builder mainhand(ItemPredicate var1) {
+         this.mainhand = var1;
+         return this;
+      }
+
+      public EntityEquipmentPredicate.Builder offhand(ItemPredicate var1) {
+         this.offhand = var1;
+         return this;
+      }
+
       public EntityEquipmentPredicate build() {
          return new EntityEquipmentPredicate(this.head, this.chest, this.legs, this.feet, this.mainhand, this.offhand);
       }
    }
 }
