--- net/minecraft/advancements/critereon/EntityPredicate.java
+++ net/minecraft/advancements/critereon/EntityPredicate.java
@@ -215,11 +215,11 @@
          }
       }
 
       private static EntityPredicate.Composite fromElement(String var0, DeserializationContext var1, @Nullable JsonElement var2) {
          if (var2 != null && var2.isJsonArray()) {
-            LootItemCondition[] var4 = var1.deserializeConditions(var2.getAsJsonArray(), var1.getAdvancementId().toString() + "/" + var0, LootContextParamSets.ADVANCEMENT_ENTITY);
+            LootItemCondition[] var4 = var1.deserializeConditions(var2.getAsJsonArray(), var1.getAdvancementId() + "/" + var0, LootContextParamSets.ADVANCEMENT_ENTITY);
             return new EntityPredicate.Composite(var4);
          } else {
             EntityPredicate var3 = EntityPredicate.fromJson(var2);
             return wrap(var3);
          }
