--- net/minecraft/advancements/critereon/EntityTypePredicate.java
+++ net/minecraft/advancements/critereon/EntityTypePredicate.java
@@ -65,11 +65,11 @@
       public TagPredicate(Tag<EntityType<?>> var1) {
          this.tag = var1;
       }
 
       public boolean matches(EntityType<?> var1) {
-         return this.tag.contains(var1);
+         return var1.is(this.tag);
       }
 
       public JsonElement serializeToJson() {
          return new JsonPrimitive("#" + SerializationTags.getInstance().getEntityTypes().getIdOrThrow(this.tag));
       }
