--- net/minecraft/advancements/critereon/FishingRodHookedTrigger.java
+++ net/minecraft/advancements/critereon/FishingRodHookedTrigger.java
@@ -11,11 +11,11 @@
 import net.minecraft.world.item.ItemStack;
 import net.minecraft.world.level.storage.loot.LootContext;
 import net.minecraft.world.level.storage.loot.parameters.LootContextParams;
 
 public class FishingRodHookedTrigger extends SimpleCriterionTrigger<FishingRodHookedTrigger.TriggerInstance> {
-   private static final ResourceLocation ID = new ResourceLocation("fishing_rod_hooked");
+   static final ResourceLocation ID = new ResourceLocation("fishing_rod_hooked");
 
    public FishingRodHookedTrigger() {
    }
 
    public ResourceLocation getId() {
