--- net/minecraft/advancements/critereon/FluidPredicate.java
+++ net/minecraft/advancements/critereon/FluidPredicate.java
@@ -35,11 +35,11 @@
       } else if (!var1.isLoaded(var2)) {
          return false;
       } else {
          FluidState var3 = var1.getFluidState(var2);
          Fluid var4 = var3.getType();
-         if (this.tag != null && !this.tag.contains(var4)) {
+         if (this.tag != null && !var4.is(this.tag)) {
             return false;
          } else if (this.fluid != null && var4 != this.fluid) {
             return false;
          } else {
             return this.properties.matches(var3);
