--- net/minecraft/advancements/critereon/ItemPredicate.java
+++ net/minecraft/advancements/critereon/ItemPredicate.java
@@ -61,13 +61,13 @@
    }
 
    public boolean matches(ItemStack var1) {
       if (this == ANY) {
          return true;
-      } else if (this.tag != null && !this.tag.contains(var1.getItem())) {
+      } else if (this.tag != null && !var1.is(this.tag)) {
          return false;
-      } else if (this.item != null && var1.getItem() != this.item) {
+      } else if (this.item != null && !var1.is(this.item)) {
          return false;
       } else if (!this.count.matches(var1.getCount())) {
          return false;
       } else if (!this.durability.isAny() && !var1.isDamageableItem()) {
          return false;
