--- net/minecraft/advancements/critereon/ItemDurabilityTrigger.java
+++ net/minecraft/advancements/critereon/ItemDurabilityTrigger.java
@@ -38,10 +38,14 @@
          this.item = var2;
          this.durability = var3;
          this.delta = var4;
       }
 
+      public static ItemDurabilityTrigger.TriggerInstance changedDurability(ItemPredicate var0, MinMaxBounds.Ints var1) {
+         return changedDurability(EntityPredicate.Composite.ANY, var0, var1);
+      }
+
       public static ItemDurabilityTrigger.TriggerInstance changedDurability(EntityPredicate.Composite var0, ItemPredicate var1, MinMaxBounds.Ints var2) {
          return new ItemDurabilityTrigger.TriggerInstance(var0, var1, var2, MinMaxBounds.Ints.ANY);
       }
 
       public boolean matches(ItemStack var1, int var2) {
