--- net/minecraft/advancements/critereon/LevitationTrigger.java
+++ net/minecraft/advancements/critereon/LevitationTrigger.java
@@ -4,11 +4,11 @@
 import net.minecraft.resources.ResourceLocation;
 import net.minecraft.server.level.ServerPlayer;
 import net.minecraft.world.phys.Vec3;
 
 public class LevitationTrigger extends SimpleCriterionTrigger<LevitationTrigger.TriggerInstance> {
-   private static final ResourceLocation ID = new ResourceLocation("levitation");
+   static final ResourceLocation ID = new ResourceLocation("levitation");
 
    public LevitationTrigger() {
    }
 
    public ResourceLocation getId() {
