--- net/minecraft/advancements/critereon/LightPredicate.java
+++ net/minecraft/advancements/critereon/LightPredicate.java
@@ -10,11 +10,11 @@
 
 public class LightPredicate {
    public static final LightPredicate ANY;
    private final MinMaxBounds.Ints composite;
 
-   private LightPredicate(MinMaxBounds.Ints var1) {
+   LightPredicate(MinMaxBounds.Ints var1) {
       this.composite = var1;
    }
 
    public boolean matches(ServerLevel var1, BlockPos var2) {
       if (this == ANY) {
@@ -47,6 +47,27 @@
    }
 
    static {
       ANY = new LightPredicate(MinMaxBounds.Ints.ANY);
    }
+
+   public static class Builder {
+      private MinMaxBounds.Ints composite;
+
+      public Builder() {
+         this.composite = MinMaxBounds.Ints.ANY;
+      }
+
+      public static LightPredicate.Builder light() {
+         return new LightPredicate.Builder();
+      }
+
+      public LightPredicate.Builder setComposite(MinMaxBounds.Ints var1) {
+         this.composite = var1;
+         return this;
+      }
+
+      public LightPredicate build() {
+         return new LightPredicate(this.composite);
+      }
+   }
 }
