--- net/minecraft/advancements/critereon/LocationPredicate.java
+++ net/minecraft/advancements/critereon/LocationPredicate.java
@@ -84,20 +84,18 @@
             return false;
          } else if (this.biome != null && (!var6 || this.biome != var7.get())) {
             return false;
          } else if (this.feature != null && (!var6 || !var1.structureFeatureManager().getStructureAt(var5, true, this.feature).isValid())) {
             return false;
-         } else if (this.smokey == null || var6 && this.smokey == CampfireBlock.isSmokeyPos(var1, var5)) {
-            if (!this.light.matches(var1, var5)) {
-               return false;
-            } else if (!this.block.matches(var1, var5)) {
-               return false;
-            } else {
-               return this.fluid.matches(var1, var5);
-            }
-         } else {
+         } else if (this.smokey != null && (!var6 || this.smokey != CampfireBlock.isSmokeyPos(var1, var5))) {
             return false;
+         } else if (!this.light.matches(var1, var5)) {
+            return false;
+         } else if (!this.block.matches(var1, var5)) {
+            return false;
+         } else {
+            return this.fluid.matches(var1, var5);
          }
       }
    }
 
    public JsonElement serializeToJson() {
