--- net/minecraft/advancements/critereon/SummonedEntityTrigger.java
+++ net/minecraft/advancements/critereon/SummonedEntityTrigger.java
@@ -5,11 +5,11 @@
 import net.minecraft.server.level.ServerPlayer;
 import net.minecraft.world.entity.Entity;
 import net.minecraft.world.level.storage.loot.LootContext;
 
 public class SummonedEntityTrigger extends SimpleCriterionTrigger<SummonedEntityTrigger.TriggerInstance> {
-   private static final ResourceLocation ID = new ResourceLocation("summoned_entity");
+   static final ResourceLocation ID = new ResourceLocation("summoned_entity");
 
    public SummonedEntityTrigger() {
    }
 
    public ResourceLocation getId() {
