--- net/minecraft/client/gui/screens/multiplayer/SafetyScreen.java
+++ net/minecraft/client/gui/screens/multiplayer/SafetyScreen.java
@@ -1,8 +1,9 @@
 package net.minecraft.client.gui.screens.multiplayer;
 
 import com.mojang.blaze3d.vertex.PoseStack;
+import java.util.Objects;
 import net.minecraft.ChatFormatting;
 import net.minecraft.client.gui.chat.NarratorChatListener;
 import net.minecraft.client.gui.components.Button;
 import net.minecraft.client.gui.components.Checkbox;
 import net.minecraft.client.gui.components.MultiLineLabel;
@@ -28,36 +29,36 @@
 
    protected void init() {
       super.init();
       this.message = MultiLineLabel.create(this.font, CONTENT, this.width - 50);
       int var10000 = this.message.getLineCount() + 1;
-      this.font.getClass();
+      Objects.requireNonNull(this.font);
       int var1 = var10000 * 9 * 2;
-      this.addButton(new Button(this.width / 2 - 155, 100 + var1, 150, 20, CommonComponents.GUI_PROCEED, (var1x) -> {
+      this.addRenderableWidget(new Button(this.width / 2 - 155, 100 + var1, 150, 20, CommonComponents.GUI_PROCEED, (var1x) -> {
          if (this.stopShowing.selected()) {
             this.minecraft.options.skipMultiplayerWarning = true;
             this.minecraft.options.save();
          }
 
          this.minecraft.setScreen(new JoinMultiplayerScreen(this.previous));
       }));
-      this.addButton(new Button(this.width / 2 - 155 + 160, 100 + var1, 150, 20, CommonComponents.GUI_BACK, (var1x) -> {
+      this.addRenderableWidget(new Button(this.width / 2 - 155 + 160, 100 + var1, 150, 20, CommonComponents.GUI_BACK, (var1x) -> {
          this.minecraft.setScreen(this.previous);
       }));
       this.stopShowing = new Checkbox(this.width / 2 - 155 + 80, 76 + var1, 150, 20, CHECK, false);
-      this.addButton(this.stopShowing);
+      this.addRenderableWidget(this.stopShowing);
    }
 
-   public String getNarrationMessage() {
-      return NARRATION.getString();
+   public Component getNarrationMessage() {
+      return NARRATION;
    }
 
    public void render(PoseStack var1, int var2, int var3, float var4) {
       this.renderDirtBackground(0);
       drawString(var1, this.font, TITLE, 25, 30, 16777215);
       MultiLineLabel var10000 = this.message;
-      this.font.getClass();
+      Objects.requireNonNull(this.font);
       var10000.renderLeftAligned(var1, 25, 70, 9 * 2, 16777215);
       super.render(var1, var2, var3, var4);
    }
 
    static {
