--- net/minecraft/world/item/context/BlockPlaceContext.java
+++ net/minecraft/world/item/context/BlockPlaceContext.java
@@ -47,10 +47,14 @@
 
    public Direction getNearestLookingDirection() {
       return Direction.orderedByNearest(this.getPlayer())[0];
    }
 
+   public Direction getNearestLookingVerticalDirection() {
+      return Direction.getFacingAxis(this.getPlayer(), Direction.Axis.Y);
+   }
+
    public Direction[] getNearestLookingDirections() {
       Direction[] var1 = Direction.orderedByNearest(this.getPlayer());
       if (this.replaceClicked) {
          return var1;
       } else {
