--- net/minecraft/util/worldupdate/WorldUpgrader.java
+++ net/minecraft/util/worldupdate/WorldUpgrader.java
@@ -216,39 +216,32 @@
                int var11 = Integer.parseInt(var10.group(1)) << 5;
                int var12 = Integer.parseInt(var10.group(2)) << 5;
 
                try {
                   RegionFile var13 = new RegionFile(var9, var3, true);
-                  Throwable var14 = null;
 
                   try {
-                     for(int var15 = 0; var15 < 32; ++var15) {
-                        for(int var16 = 0; var16 < 32; ++var16) {
-                           ChunkPos var17 = new ChunkPos(var15 + var11, var16 + var12);
-                           if (var13.doesChunkExist(var17)) {
-                              var5.add(var17);
+                     for(int var14 = 0; var14 < 32; ++var14) {
+                        for(int var15 = 0; var15 < 32; ++var15) {
+                           ChunkPos var16 = new ChunkPos(var14 + var11, var15 + var12);
+                           if (var13.doesChunkExist(var16)) {
+                              var5.add(var16);
                            }
                         }
                      }
-                  } catch (Throwable var26) {
-                     var14 = var26;
-                     throw var26;
-                  } finally {
-                     if (var13 != null) {
-                        if (var14 != null) {
-                           try {
-                              var13.close();
-                           } catch (Throwable var25) {
-                              var14.addSuppressed(var25);
-                           }
-                        } else {
-                           var13.close();
-                        }
+                  } catch (Throwable var18) {
+                     try {
+                        var13.close();
+                     } catch (Throwable var17) {
+                        var18.addSuppressed(var17);
                      }
 
+                     throw var18;
                   }
-               } catch (Throwable var28) {
+
+                  var13.close();
+               } catch (Throwable var19) {
                }
             }
          }
 
          return var5;
@@ -257,10 +250,22 @@
 
    public boolean isFinished() {
       return this.finished;
    }
 
+   public ImmutableSet<ResourceKey<Level>> levels() {
+      return this.levels;
+   }
+
+   public float dimensionProgress(ResourceKey<Level> var1) {
+      return this.progressMap.getFloat(var1);
+   }
+
+   public float getProgress() {
+      return this.progress;
+   }
+
    public int getTotalChunks() {
       return this.totalChunks;
    }
 
    public int getConverted() {
