--- net/minecraft/util/worldupdate/WorldUpgrader.java
+++ net/minecraft/util/worldupdate/WorldUpgrader.java
@@ -257,10 +257,22 @@
 
    public boolean isFinished() {
       return this.finished;
    }
 
+   public ImmutableSet<ResourceKey<Level>> levels() {
+      return this.levels;
+   }
+
+   public float dimensionProgress(ResourceKey<Level> var1) {
+      return this.progressMap.getFloat(var1);
+   }
+
+   public float getProgress() {
+      return this.progress;
+   }
+
    public int getTotalChunks() {
       return this.totalChunks;
    }
 
    public int getConverted() {
