--- net/minecraft/client/color/item/ItemColors.java
+++ net/minecraft/client/color/item/ItemColors.java
@@ -19,10 +19,11 @@
 import net.minecraft.world.level.ItemLike;
 import net.minecraft.world.level.block.Blocks;
 import net.minecraft.world.level.block.state.BlockState;
 
 public class ItemColors {
+   private static final int DEFAULT = -1;
    private final IdMapper<ItemColor> itemColors = new IdMapper(32);
 
    public ItemColors() {
    }
 
