--- net/minecraft/server/level/BlockDestructionProgress.java
+++ net/minecraft/server/level/BlockDestructionProgress.java
@@ -11,10 +11,14 @@
    public BlockDestructionProgress(int var1, BlockPos var2) {
       this.id = var1;
       this.pos = var2;
    }
 
+   public int getId() {
+      return this.id;
+   }
+
    public BlockPos getPos() {
       return this.pos;
    }
 
    public void setProgress(int var1) {
