--- net/minecraft/server/level/ServerEntity.java
+++ net/minecraft/server/level/ServerEntity.java
@@ -201,12 +201,12 @@
       this.entity.startSeenByPlayer(var1);
       var1.cancelRemoveEntity(this.entity);
    }
 
    public void sendPairingData(Consumer<Packet<?>> var1) {
-      if (this.entity.removed) {
-         LOGGER.warn("Fetching packet for removed entity " + this.entity);
+      if (this.entity.isRemoved()) {
+         LOGGER.warn("Fetching packet for removed entity {}", this.entity);
       }
 
       Packet var2 = this.entity.getAddEntityPacket();
       this.yHeadRotp = Mth.floor(this.entity.getYHeadRot() * 256.0F / 360.0F);
       var1.accept(var2);
