--- net/minecraft/server/rcon/thread/GenericThread.java
+++ net/minecraft/server/rcon/thread/GenericThread.java
@@ -7,10 +7,11 @@
 import org.apache.logging.log4j.Logger;
 
 public abstract class GenericThread implements Runnable {
    private static final Logger LOGGER = LogManager.getLogger();
    private static final AtomicInteger UNIQUE_THREAD_ID = new AtomicInteger(0);
+   private static final int MAX_STOP_WAIT = 5;
    protected volatile boolean running;
    protected final String name;
    @Nullable
    protected Thread thread;
 
@@ -21,11 +22,12 @@
    public synchronized boolean start() {
       if (this.running) {
          return true;
       } else {
          this.running = true;
-         this.thread = new Thread(this, this.name + " #" + UNIQUE_THREAD_ID.incrementAndGet());
+         String var10004 = this.name;
+         this.thread = new Thread(this, var10004 + " #" + UNIQUE_THREAD_ID.incrementAndGet());
          this.thread.setUncaughtExceptionHandler(new DefaultUncaughtExceptionHandlerWithName(LOGGER));
          this.thread.start();
          LOGGER.info((String)"Thread {} started", (Object)this.name);
          return true;
       }
