--- net/minecraft/server/rcon/thread/RconClient.java
+++ net/minecraft/server/rcon/thread/RconClient.java
@@ -11,10 +11,15 @@
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
 
 public class RconClient extends GenericThread {
    private static final Logger LOGGER = LogManager.getLogger();
+   private static final int SERVERDATA_AUTH = 3;
+   private static final int SERVERDATA_EXECCOMMAND = 2;
+   private static final int SERVERDATA_RESPONSE_VALUE = 0;
+   private static final int SERVERDATA_AUTH_RESPONSE = 2;
+   private static final int SERVERDATA_AUTH_FAILURE = -1;
    private boolean authed;
    private final Socket client;
    private final byte[] buf = new byte[1460];
    private final String rconPassword;
    private final ServerInterface serverInterface;
