--- net/minecraft/server/rcon/thread/RconClient.java
+++ net/minecraft/server/rcon/thread/RconClient.java
@@ -37,17 +37,17 @@
       while(true) {
          try {
             if (this.running) {
                BufferedInputStream var1 = new BufferedInputStream(this.client.getInputStream());
                int var2 = var1.read(this.buf, 0, 1460);
-               if (10 <= var2) {
-                  byte var3 = 0;
-                  int var4 = PktUtils.intFromByteArray(this.buf, 0, var2);
-                  if (var4 != var2 - 4) {
-                     return;
-                  }
+               if (10 > var2) {
+                  return;
+               }
 
+               byte var3 = 0;
+               int var4 = PktUtils.intFromByteArray(this.buf, 0, var2);
+               if (var4 == var2 - 4) {
                   int var19 = var3 + 4;
                   int var5 = PktUtils.intFromByteArray(this.buf, var19, var2);
                   var19 += 4;
                   int var6 = PktUtils.intFromByteArray(this.buf, var19);
                   var19 += 4;
