--- net/minecraft/world/item/enchantment/DiggingEnchantment.java
+++ net/minecraft/world/item/enchantment/DiggingEnchantment.java
@@ -20,8 +20,8 @@
    public int getMaxLevel() {
       return 5;
    }
 
    public boolean canEnchant(ItemStack var1) {
-      return var1.getItem() == Items.SHEARS ? true : super.canEnchant(var1);
+      return var1.is(Items.SHEARS) ? true : super.canEnchant(var1);
    }
 }
