--- net/minecraft/world/item/enchantment/EnchantmentCategory.java
+++ net/minecraft/world/item/enchantment/EnchantmentCategory.java
@@ -83,10 +83,10 @@
       public boolean canEnchant(Item var1) {
          return var1 instanceof Vanishable || Block.byItem(var1) instanceof Vanishable || BREAKABLE.canEnchant(var1);
       }
    };
 
-   private EnchantmentCategory() {
+   EnchantmentCategory() {
    }
 
    public abstract boolean canEnchant(Item var1);
 }
