--- net/minecraft/world/item/enchantment/EnchantmentInstance.java
+++ net/minecraft/world/item/enchantment/EnchantmentInstance.java
@@ -1,10 +1,10 @@
 package net.minecraft.world.item.enchantment;
 
-import net.minecraft.util.WeighedRandom;
+import net.minecraft.util.random.WeightedEntry;
 
-public class EnchantmentInstance extends WeighedRandom.WeighedRandomItem {
+public class EnchantmentInstance extends WeightedEntry.IntrusiveBase {
    public final Enchantment enchantment;
    public final int level;
 
    public EnchantmentInstance(Enchantment var1, int var2) {
       super(var1.getRarity().getWeight());
