--- net/minecraft/world/item/enchantment/Enchantments.java
+++ net/minecraft/world/item/enchantment/Enchantments.java
@@ -42,10 +42,13 @@
    public static final Enchantment QUICK_CHARGE;
    public static final Enchantment PIERCING;
    public static final Enchantment MENDING;
    public static final Enchantment VANISHING_CURSE;
 
+   public Enchantments() {
+   }
+
    private static Enchantment register(String var0, Enchantment var1) {
       return (Enchantment)Registry.register(Registry.ENCHANTMENT, (String)var0, var1);
    }
 
    static {
