--- net/minecraft/world/item/enchantment/Enchantment.java
+++ net/minecraft/world/item/enchantment/Enchantment.java
@@ -21,10 +21,15 @@
    private final Enchantment.Rarity rarity;
    public final EnchantmentCategory category;
    @Nullable
    protected String descriptionId;
 
+   @Nullable
+   public static Enchantment byId(int var0) {
+      return (Enchantment)Registry.ENCHANTMENT.byId(var0);
+   }
+
    protected Enchantment(Enchantment.Rarity var1, EnchantmentCategory var2, EquipmentSlot[] var3) {
       this.rarity = var1;
       this.category = var2;
       this.slots = var3;
    }
