--- net/minecraft/world/level/storage/loot/predicates/ConditionReference.java
+++ net/minecraft/world/level/storage/loot/predicates/ConditionReference.java
@@ -52,10 +52,16 @@
          LOGGER.warn("Detected infinite loop in loot tables");
          return false;
       }
    }
 
+   public static LootItemCondition.Builder conditionReference(ResourceLocation var0) {
+      return () -> {
+         return new ConditionReference(var0);
+      };
+   }
+
    public static class Serializer implements net.minecraft.world.level.storage.loot.Serializer<ConditionReference> {
       public Serializer() {
       }
 
       public void serialize(JsonObject var1, ConditionReference var2, JsonSerializationContext var3) {
