--- net/minecraft/world/level/storage/loot/predicates/InvertedLootItemCondition.java
+++ net/minecraft/world/level/storage/loot/predicates/InvertedLootItemCondition.java
@@ -8,13 +8,13 @@
 import net.minecraft.world.level.storage.loot.LootContext;
 import net.minecraft.world.level.storage.loot.ValidationContext;
 import net.minecraft.world.level.storage.loot.parameters.LootContextParam;
 
 public class InvertedLootItemCondition implements LootItemCondition {
-   private final LootItemCondition term;
+   final LootItemCondition term;
 
-   private InvertedLootItemCondition(LootItemCondition var1) {
+   InvertedLootItemCondition(LootItemCondition var1) {
       this.term = var1;
    }
 
    public LootItemConditionType getType() {
       return LootItemConditions.INVERTED;
