--- net/minecraft/world/level/storage/loot/predicates/LootItemRandomChanceWithLootingCondition.java
+++ net/minecraft/world/level/storage/loot/predicates/LootItemRandomChanceWithLootingCondition.java
@@ -12,14 +12,14 @@
 import net.minecraft.world.level.storage.loot.LootContext;
 import net.minecraft.world.level.storage.loot.parameters.LootContextParam;
 import net.minecraft.world.level.storage.loot.parameters.LootContextParams;
 
 public class LootItemRandomChanceWithLootingCondition implements LootItemCondition {
-   private final float percent;
-   private final float lootingMultiplier;
+   final float percent;
+   final float lootingMultiplier;
 
-   private LootItemRandomChanceWithLootingCondition(float var1, float var2) {
+   LootItemRandomChanceWithLootingCondition(float var1, float var2) {
       this.percent = var1;
       this.lootingMultiplier = var2;
    }
 
    public LootItemConditionType getType() {
