--- net/minecraft/world/level/storage/loot/predicates/LootItemEntityPropertyCondition.java
+++ net/minecraft/world/level/storage/loot/predicates/LootItemEntityPropertyCondition.java
@@ -12,14 +12,14 @@
 import net.minecraft.world.level.storage.loot.parameters.LootContextParam;
 import net.minecraft.world.level.storage.loot.parameters.LootContextParams;
 import net.minecraft.world.phys.Vec3;
 
 public class LootItemEntityPropertyCondition implements LootItemCondition {
-   private final EntityPredicate predicate;
-   private final LootContext.EntityTarget entityTarget;
+   final EntityPredicate predicate;
+   final LootContext.EntityTarget entityTarget;
 
-   private LootItemEntityPropertyCondition(EntityPredicate var1, LootContext.EntityTarget var2) {
+   LootItemEntityPropertyCondition(EntityPredicate var1, LootContext.EntityTarget var2) {
       this.predicate = var1;
       this.entityTarget = var2;
    }
 
    public LootItemConditionType getType() {
