--- net/minecraft/world/level/storage/loot/predicates/LootItemBlockStatePropertyCondition.java
+++ net/minecraft/world/level/storage/loot/predicates/LootItemBlockStatePropertyCondition.java
@@ -15,14 +15,14 @@
 import net.minecraft.world.level.storage.loot.LootContext;
 import net.minecraft.world.level.storage.loot.parameters.LootContextParam;
 import net.minecraft.world.level.storage.loot.parameters.LootContextParams;
 
 public class LootItemBlockStatePropertyCondition implements LootItemCondition {
-   private final Block block;
-   private final StatePropertiesPredicate properties;
+   final Block block;
+   final StatePropertiesPredicate properties;
 
-   private LootItemBlockStatePropertyCondition(Block var1, StatePropertiesPredicate var2) {
+   LootItemBlockStatePropertyCondition(Block var1, StatePropertiesPredicate var2) {
       this.block = var1;
       this.properties = var2;
    }
 
    public LootItemConditionType getType() {
@@ -33,17 +33,36 @@
       return ImmutableSet.of(LootContextParams.BLOCK_STATE);
    }
 
    public boolean test(LootContext var1) {
       BlockState var2 = (BlockState)var1.getParamOrNull(LootContextParams.BLOCK_STATE);
-      return var2 != null && this.block == var2.getBlock() && this.properties.matches(var2);
+      return var2 != null && var2.is(this.block) && this.properties.matches(var2);
    }
 
    public static LootItemBlockStatePropertyCondition.Builder hasBlockStateProperties(Block var0) {
       return new LootItemBlockStatePropertyCondition.Builder(var0);
    }
 
+   public static class Builder implements LootItemCondition.Builder {
+      private final Block block;
+      private StatePropertiesPredicate properties;
+
+      public Builder(Block var1) {
+         this.properties = StatePropertiesPredicate.ANY;
+         this.block = var1;
+      }
+
+      public LootItemBlockStatePropertyCondition.Builder setProperties(StatePropertiesPredicate.Builder var1) {
+         this.properties = var1.build();
+         return this;
+      }
+
+      public LootItemCondition build() {
+         return new LootItemBlockStatePropertyCondition(this.block, this.properties);
+      }
+   }
+
    public static class Serializer implements net.minecraft.world.level.storage.loot.Serializer<LootItemBlockStatePropertyCondition> {
       public Serializer() {
       }
 
       public void serialize(JsonObject var1, LootItemBlockStatePropertyCondition var2, JsonSerializationContext var3) {
@@ -61,25 +80,6 @@
             throw new JsonSyntaxException("Block " + var4 + " has no property " + var1x);
          });
          return new LootItemBlockStatePropertyCondition(var4, var5);
       }
    }
-
-   public static class Builder implements LootItemCondition.Builder {
-      private final Block block;
-      private StatePropertiesPredicate properties;
-
-      public Builder(Block var1) {
-         this.properties = StatePropertiesPredicate.ANY;
-         this.block = var1;
-      }
-
-      public LootItemBlockStatePropertyCondition.Builder setProperties(StatePropertiesPredicate.Builder var1) {
-         this.properties = var1.build();
-         return this;
-      }
-
-      public LootItemCondition build() {
-         return new LootItemBlockStatePropertyCondition(this.block, this.properties);
-      }
-   }
 }
