--- net/minecraft/world/level/storage/loot/predicates/WeatherCheck.java
+++ net/minecraft/world/level/storage/loot/predicates/WeatherCheck.java
@@ -30,10 +30,14 @@
       } else {
          return this.isThundering == null || this.isThundering == var2.isThundering();
       }
    }
 
+   public static WeatherCheck.Builder weather() {
+      return new WeatherCheck.Builder();
+   }
+
    public static class Serializer implements net.minecraft.world.level.storage.loot.Serializer<WeatherCheck> {
       public Serializer() {
       }
 
       public void serialize(JsonObject var1, WeatherCheck var2, JsonSerializationContext var3) {
@@ -45,6 +49,30 @@
          Boolean var3 = var1.has("raining") ? GsonHelper.getAsBoolean(var1, "raining") : null;
          Boolean var4 = var1.has("thundering") ? GsonHelper.getAsBoolean(var1, "thundering") : null;
          return new WeatherCheck(var3, var4);
       }
    }
+
+   public static class Builder implements LootItemCondition.Builder {
+      @Nullable
+      private Boolean isRaining;
+      @Nullable
+      private Boolean isThundering;
+
+      public Builder() {
+      }
+
+      public WeatherCheck.Builder setRaining(@Nullable Boolean var1) {
+         this.isRaining = var1;
+         return this;
+      }
+
+      public WeatherCheck.Builder setThundering(@Nullable Boolean var1) {
+         this.isThundering = var1;
+         return this;
+      }
+
+      public WeatherCheck build() {
+         return new WeatherCheck(this.isRaining, this.isThundering);
+      }
+   }
 }
