--- net/minecraft/client/renderer/item/ClampedItemPropertyFunction.java
+++ net/minecraft/client/renderer/item/ClampedItemPropertyFunction.java
@@ -1,0 +1,16 @@
+package net.minecraft.client.renderer.item;
+
+import javax.annotation.Nullable;
+import net.minecraft.client.multiplayer.ClientLevel;
+import net.minecraft.util.Mth;
+import net.minecraft.world.entity.LivingEntity;
+import net.minecraft.world.item.ItemStack;
+
+public interface ClampedItemPropertyFunction extends ItemPropertyFunction {
+   @Deprecated
+   default float call(ItemStack var1, @Nullable ClientLevel var2, @Nullable LivingEntity var3, int var4) {
+      return Mth.clamp(this.unclampedCall(var1, var2, var3, var4), 0.0F, 1.0F);
+   }
+
+   float unclampedCall(ItemStack var1, @Nullable ClientLevel var2, @Nullable LivingEntity var3, int var4);
+}
