--- net/minecraft/client/renderer/item/ItemPropertyFunction.java
+++ net/minecraft/client/renderer/item/ItemPropertyFunction.java
@@ -3,8 +3,9 @@
 import javax.annotation.Nullable;
 import net.minecraft.client.multiplayer.ClientLevel;
 import net.minecraft.world.entity.LivingEntity;
 import net.minecraft.world.item.ItemStack;
 
+@Deprecated
 public interface ItemPropertyFunction {
-   float call(ItemStack var1, @Nullable ClientLevel var2, @Nullable LivingEntity var3);
+   float call(ItemStack var1, @Nullable ClientLevel var2, @Nullable LivingEntity var3, int var4);
 }
