--- net/minecraft/client/renderer/item/ItemPropertyFunction.java
+++ net/minecraft/client/renderer/item/ItemPropertyFunction.java
@@ -4,7 +4,7 @@
 import net.minecraft.client.multiplayer.ClientLevel;
 import net.minecraft.world.entity.LivingEntity;
 import net.minecraft.world.item.ItemStack;
 
 public interface ItemPropertyFunction {
-   float call(ItemStack var1, @Nullable ClientLevel var2, @Nullable LivingEntity var3);
+   float call(ItemStack var1, @Nullable ClientLevel var2, @Nullable LivingEntity var3, int var4);
 }
