--- net/minecraft/world/entity/animal/Cod.java
+++ net/minecraft/world/entity/animal/Cod.java
@@ -11,11 +11,11 @@
 public class Cod extends AbstractSchoolingFish {
    public Cod(EntityType<? extends Cod> var1, Level var2) {
       super(var1, var2);
    }
 
-   protected ItemStack getBucketItemStack() {
+   public ItemStack getBucketItemStack() {
       return new ItemStack(Items.COD_BUCKET);
    }
 
    protected SoundEvent getAmbientSound() {
       return SoundEvents.COD_AMBIENT;
