--- net/minecraft/world/entity/animal/Fox.java
+++ net/minecraft/world/entity/animal/Fox.java
@@ -28,12 +28,13 @@
 import net.minecraft.sounds.SoundEvents;
 import net.minecraft.stats.Stats;
 import net.minecraft.tags.FluidTags;
 import net.minecraft.util.Mth;
 import net.minecraft.world.DifficultyInstance;
+import net.minecraft.world.InteractionHand;
 import net.minecraft.world.damagesource.DamageSource;
-import net.minecraft.world.entity.AgableMob;
+import net.minecraft.world.entity.AgeableMob;
 import net.minecraft.world.entity.Entity;
 import net.minecraft.world.entity.EntityDimensions;
 import net.minecraft.world.entity.EntitySelector;
 import net.minecraft.world.entity.EntityType;
 import net.minecraft.world.entity.EquipmentSlot;
@@ -232,11 +233,11 @@
 
    public static AttributeSupplier.Builder createAttributes() {
       return Mob.createMobAttributes().add(Attributes.MOVEMENT_SPEED, 0.30000001192092896D).add(Attributes.MAX_HEALTH, 10.0D).add(Attributes.FOLLOW_RANGE, 32.0D).add(Attributes.ATTACK_DAMAGE, 2.0D);
    }
 
-   public Fox getBreedOffspring(ServerLevel var1, AgableMob var2) {
+   public Fox getBreedOffspring(ServerLevel var1, AgeableMob var2) {
       Fox var3 = (Fox)EntityType.FOX.create(var1);
       var3.setFoxType(this.random.nextBoolean() ? this.getFoxType() : ((Fox)var2).getFoxType());
       return var3;
    }
 
@@ -278,16 +279,16 @@
          this.targetSelector.addGoal(6, this.turtleEggTargetGoal);
       }
 
    }
 
-   protected void usePlayerItem(Player var1, ItemStack var2) {
-      if (this.isFood(var2)) {
-         this.playSound(this.getEatingSound(var2), 1.0F, 1.0F);
+   protected void usePlayerItem(Player var1, InteractionHand var2, ItemStack var3) {
+      if (this.isFood(var3)) {
+         this.playSound(this.getEatingSound(var3), 1.0F, 1.0F);
       }
 
-      super.usePlayerItem(var1, var2);
+      super.usePlayerItem(var1, var2, var3);
    }
 
    protected float getStandingEyeHeight(Pose var1, EntityDimensions var2) {
       return this.isBaby() ? var2.height * 0.85F : 0.4F;
    }
@@ -440,11 +441,11 @@
          this.spitOutItem(this.getItemBySlot(EquipmentSlot.MAINHAND));
          this.onItemPickup(var1);
          this.setItemSlot(EquipmentSlot.MAINHAND, var2.split(1));
          this.handDropChances[EquipmentSlot.MAINHAND.getIndex()] = 2.0F;
          this.take(var1, var2.getCount());
-         var1.remove();
+         var1.discard();
          this.ticksSinceEaten = 0;
       }
 
    }
 
@@ -485,11 +486,11 @@
       }
 
    }
 
    public boolean isFood(ItemStack var1) {
-      return var1.getItem() == Items.SWEET_BERRIES;
+      return var1.is(Items.SWEET_BERRIES);
    }
 
    protected void onOffspringSpawnedFromEgg(Player var1, Mob var2) {
       ((Fox)var2).addTrustedUUID(var1.getUUID());
    }
@@ -861,11 +862,11 @@
       public void tick() {
          --this.countdown;
       }
    }
 
-   public static class FoxGroupData extends AgableMob.AgableMobGroupData {
+   public static class FoxGroupData extends AgeableMob.AgeableMobGroupData {
       public final Fox.Type type;
 
       public FoxGroupData(Fox.Type var1) {
          super(false);
          this.type = var1;
