--- net/minecraft/world/entity/animal/Pufferfish.java
+++ net/minecraft/world/entity/animal/Pufferfish.java
@@ -29,13 +29,17 @@
 public class Pufferfish extends AbstractFish {
    private static final EntityDataAccessor<Integer> PUFF_STATE;
    private int inflateCounter;
    private int deflateTimer;
    private static final Predicate<LivingEntity> NO_SPECTATORS_AND_NO_WATER_MOB;
+   public static final int STATE_SMALL = 0;
+   public static final int STATE_MID = 1;
+   public static final int STATE_FULL = 2;
 
    public Pufferfish(EntityType<? extends Pufferfish> var1, Level var2) {
       super(var1, var2);
+      this.refreshDimensions();
    }
 
    protected void defineSynchedData() {
       super.defineSynchedData();
       this.entityData.define(PUFF_STATE, 0);
@@ -65,11 +69,11 @@
    public void readAdditionalSaveData(CompoundTag var1) {
       super.readAdditionalSaveData(var1);
       this.setPuffState(var1.getInt("PuffState"));
    }
 
-   protected ItemStack getBucketItemStack() {
+   public ItemStack getBucketItemStack() {
       return new ItemStack(Items.PUFFERFISH_BUCKET);
    }
 
    protected void registerGoals() {
       super.registerGoals();
