--- net/minecraft/world/entity/animal/Salmon.java
+++ net/minecraft/world/entity/animal/Salmon.java
@@ -15,11 +15,11 @@
 
    public int getMaxSchoolSize() {
       return 5;
    }
 
-   protected ItemStack getBucketItemStack() {
+   public ItemStack getBucketItemStack() {
       return new ItemStack(Items.SALMON_BUCKET);
    }
 
    protected SoundEvent getAmbientSound() {
       return SoundEvents.SALMON_AMBIENT;
