--- net/minecraft/world/entity/animal/ShoulderRidingEntity.java
+++ net/minecraft/world/entity/animal/ShoulderRidingEntity.java
@@ -16,11 +16,11 @@
    public boolean setEntityOnShoulder(ServerPlayer var1) {
       CompoundTag var2 = new CompoundTag();
       var2.putString("id", this.getEncodeId());
       this.saveWithoutId(var2);
       if (var1.setEntityOnShoulder(var2)) {
-         this.remove();
+         this.discard();
          return true;
       } else {
          return false;
       }
    }
