--- net/minecraft/world/entity/animal/SnowGolem.java
+++ net/minecraft/world/entity/animal/SnowGolem.java
@@ -35,13 +35,17 @@
 import net.minecraft.world.item.Items;
 import net.minecraft.world.level.GameRules;
 import net.minecraft.world.level.Level;
 import net.minecraft.world.level.block.Blocks;
 import net.minecraft.world.level.block.state.BlockState;
+import net.minecraft.world.level.gameevent.GameEvent;
+import net.minecraft.world.phys.Vec3;
 
 public class SnowGolem extends AbstractGolem implements Shearable, RangedAttackMob {
    private static final EntityDataAccessor<Byte> DATA_PUMPKIN_ID;
+   private static final byte PUMPKIN_FLAG = 16;
+   private static final float EYE_HEIGHT = 1.7F;
 
    public SnowGolem(EntityType<? extends SnowGolem> var1, Level var2) {
       super(var1, var2);
    }
 
@@ -126,12 +130,13 @@
       return 1.7F;
    }
 
    protected InteractionResult mobInteract(Player var1, InteractionHand var2) {
       ItemStack var3 = var1.getItemInHand(var2);
-      if (var3.getItem() == Items.SHEARS && this.readyForShearing()) {
+      if (var3.is(Items.SHEARS) && this.readyForShearing()) {
          this.shear(SoundSource.PLAYERS);
+         this.gameEvent(GameEvent.SHEAR, var1);
          if (!this.level.isClientSide) {
             var3.hurtAndBreak(1, var1, (var1x) -> {
                var1x.broadcastBreakEvent(var2);
             });
          }
@@ -182,9 +187,13 @@
    @Nullable
    protected SoundEvent getDeathSound() {
       return SoundEvents.SNOW_GOLEM_DEATH;
    }
 
+   public Vec3 getLeashOffset() {
+      return new Vec3(0.0D, (double)(0.75F * this.getEyeHeight()), (double)(this.getBbWidth() * 0.4F));
+   }
+
    static {
       DATA_PUMPKIN_ID = SynchedEntityData.defineId(SnowGolem.class, EntityDataSerializers.BYTE);
    }
 }
