--- net/minecraft/world/level/levelgen/feature/featuresize/FeatureSize.java
+++ net/minecraft/world/level/levelgen/feature/featuresize/FeatureSize.java
@@ -6,10 +6,11 @@
 import java.util.OptionalInt;
 import net.minecraft.core.Registry;
 
 public abstract class FeatureSize {
    public static final Codec<FeatureSize> CODEC;
+   protected static final int MAX_WIDTH = 16;
    protected final OptionalInt minClippedHeight;
 
    protected static <S extends FeatureSize> RecordCodecBuilder<S, OptionalInt> minClippedHeightCodec() {
       return Codec.intRange(0, 80).optionalFieldOf("min_clipped_height").xmap((var0) -> {
          return (OptionalInt)var0.map(OptionalInt::of).orElse(OptionalInt.empty());
