--- net/minecraft/client/particle/AttackSweepParticle.java
+++ net/minecraft/client/particle/AttackSweepParticle.java
@@ -4,11 +4,11 @@
 import net.minecraft.core.particles.SimpleParticleType;
 
 public class AttackSweepParticle extends TextureSheetParticle {
    private final SpriteSet sprites;
 
-   private AttackSweepParticle(ClientLevel var1, double var2, double var4, double var6, double var8, SpriteSet var10) {
+   AttackSweepParticle(ClientLevel var1, double var2, double var4, double var6, double var8, SpriteSet var10) {
       super(var1, var2, var4, var6, 0.0D, 0.0D, 0.0D);
       this.sprites = var10;
       this.lifetime = 4;
       float var11 = this.random.nextFloat() * 0.6F + 0.4F;
       this.rCol = var11;
