--- net/minecraft/client/particle/BreakingItemParticle.java
+++ net/minecraft/client/particle/BreakingItemParticle.java
@@ -26,11 +26,11 @@
       return ParticleRenderType.TERRAIN_SHEET;
    }
 
    protected BreakingItemParticle(ClientLevel var1, double var2, double var4, double var6, ItemStack var8) {
       super(var1, var2, var4, var6, 0.0D, 0.0D, 0.0D);
-      this.setSprite(Minecraft.getInstance().getItemRenderer().getModel(var8, var1, (LivingEntity)null).getParticleIcon());
+      this.setSprite(Minecraft.getInstance().getItemRenderer().getModel(var8, var1, (LivingEntity)null, 0).getParticleIcon());
       this.gravity = 1.0F;
       this.quadSize /= 2.0F;
       this.uo = this.random.nextFloat() * 3.0F;
       this.vo = this.random.nextFloat() * 3.0F;
    }
