--- net/minecraft/client/particle/BreakingItemParticle.java
+++ net/minecraft/client/particle/BreakingItemParticle.java
@@ -10,11 +10,11 @@
 
 public class BreakingItemParticle extends TextureSheetParticle {
    private final float uo;
    private final float vo;
 
-   private BreakingItemParticle(ClientLevel var1, double var2, double var4, double var6, double var8, double var10, double var12, ItemStack var14) {
+   BreakingItemParticle(ClientLevel var1, double var2, double var4, double var6, double var8, double var10, double var12, ItemStack var14) {
       this(var1, var2, var4, var6, var14);
       this.xd *= 0.10000000149011612D;
       this.yd *= 0.10000000149011612D;
       this.zd *= 0.10000000149011612D;
       this.xd += var8;
@@ -26,11 +26,11 @@
       return ParticleRenderType.TERRAIN_SHEET;
    }
 
    protected BreakingItemParticle(ClientLevel var1, double var2, double var4, double var6, ItemStack var8) {
       super(var1, var2, var4, var6, 0.0D, 0.0D, 0.0D);
-      this.setSprite(Minecraft.getInstance().getItemRenderer().getModel(var8, var1, (LivingEntity)null).getParticleIcon());
+      this.setSprite(Minecraft.getInstance().getItemRenderer().getModel(var8, var1, (LivingEntity)null, 0).getParticleIcon());
       this.gravity = 1.0F;
       this.quadSize /= 2.0F;
       this.uo = this.random.nextFloat() * 3.0F;
       this.vo = this.random.nextFloat() * 3.0F;
    }
