--- net/minecraft/client/particle/BubblePopParticle.java
+++ net/minecraft/client/particle/BubblePopParticle.java
@@ -4,11 +4,11 @@
 import net.minecraft.core.particles.SimpleParticleType;
 
 public class BubblePopParticle extends TextureSheetParticle {
    private final SpriteSet sprites;
 
-   private BubblePopParticle(ClientLevel var1, double var2, double var4, double var6, double var8, double var10, double var12, SpriteSet var14) {
+   BubblePopParticle(ClientLevel var1, double var2, double var4, double var6, double var8, double var10, double var12, SpriteSet var14) {
       super(var1, var2, var4, var6);
       this.sprites = var14;
       this.lifetime = 4;
       this.gravity = 0.008F;
       this.xd = var8;
