--- net/minecraft/client/particle/BubbleColumnUpParticle.java
+++ net/minecraft/client/particle/BubbleColumnUpParticle.java
@@ -6,35 +6,26 @@
 import net.minecraft.tags.FluidTags;
 
 public class BubbleColumnUpParticle extends TextureSheetParticle {
    private BubbleColumnUpParticle(ClientLevel var1, double var2, double var4, double var6, double var8, double var10, double var12) {
       super(var1, var2, var4, var6);
+      this.gravity = -0.125F;
+      this.friction = 0.85F;
       this.setSize(0.02F, 0.02F);
       this.quadSize *= this.random.nextFloat() * 0.6F + 0.2F;
       this.xd = var8 * 0.20000000298023224D + (Math.random() * 2.0D - 1.0D) * 0.019999999552965164D;
       this.yd = var10 * 0.20000000298023224D + (Math.random() * 2.0D - 1.0D) * 0.019999999552965164D;
       this.zd = var12 * 0.20000000298023224D + (Math.random() * 2.0D - 1.0D) * 0.019999999552965164D;
       this.lifetime = (int)(40.0D / (Math.random() * 0.8D + 0.2D));
    }
 
    public void tick() {
-      this.xo = this.x;
-      this.yo = this.y;
-      this.zo = this.z;
-      this.yd += 0.005D;
-      if (this.lifetime-- <= 0) {
+      super.tick();
+      if (!this.removed && !this.level.getFluidState(new BlockPos(this.x, this.y, this.z)).is(FluidTags.WATER)) {
          this.remove();
-      } else {
-         this.move(this.xd, this.yd, this.zd);
-         this.xd *= 0.8500000238418579D;
-         this.yd *= 0.8500000238418579D;
-         this.zd *= 0.8500000238418579D;
-         if (!this.level.getFluidState(new BlockPos(this.x, this.y, this.z)).is(FluidTags.WATER)) {
-            this.remove();
-         }
-
       }
+
    }
 
    public ParticleRenderType getRenderType() {
       return ParticleRenderType.PARTICLE_SHEET_OPAQUE;
    }
