--- net/minecraft/client/particle/CampfireSmokeParticle.java
+++ net/minecraft/client/particle/CampfireSmokeParticle.java
@@ -2,11 +2,11 @@
 
 import net.minecraft.client.multiplayer.ClientLevel;
 import net.minecraft.core.particles.SimpleParticleType;
 
 public class CampfireSmokeParticle extends TextureSheetParticle {
-   private CampfireSmokeParticle(ClientLevel var1, double var2, double var4, double var6, double var8, double var10, double var12, boolean var14) {
+   CampfireSmokeParticle(ClientLevel var1, double var2, double var4, double var6, double var8, double var10, double var12, boolean var14) {
       super(var1, var2, var4, var6);
       this.scale(3.0F);
       this.setSize(0.25F, 0.25F);
       if (var14) {
          this.lifetime = this.random.nextInt(50) + 280;
