--- net/minecraft/client/particle/DragonBreathParticle.java
+++ net/minecraft/client/particle/DragonBreathParticle.java
@@ -8,10 +8,11 @@
    private boolean hasHitGround;
    private final SpriteSet sprites;
 
    private DragonBreathParticle(ClientLevel var1, double var2, double var4, double var6, double var8, double var10, double var12, SpriteSet var14) {
       super(var1, var2, var4, var6);
+      this.friction = 0.96F;
       this.xd = var8;
       this.yd = var10;
       this.zd = var12;
       this.rCol = Mth.nextFloat(this.random, 0.7176471F, 0.8745098F);
       this.gCol = Mth.nextFloat(this.random, 0.0F, 0.0F);
@@ -45,14 +46,14 @@
          if (this.y == this.yo) {
             this.xd *= 1.1D;
             this.zd *= 1.1D;
          }
 
-         this.xd *= 0.9599999785423279D;
-         this.zd *= 0.9599999785423279D;
+         this.xd *= (double)this.friction;
+         this.zd *= (double)this.friction;
          if (this.hasHitGround) {
-            this.yd *= 0.9599999785423279D;
+            this.yd *= (double)this.friction;
          }
 
       }
    }
 
