--- net/minecraft/client/particle/DragonBreathParticle.java
+++ net/minecraft/client/particle/DragonBreathParticle.java
@@ -3,15 +3,24 @@
 import net.minecraft.client.multiplayer.ClientLevel;
 import net.minecraft.core.particles.SimpleParticleType;
 import net.minecraft.util.Mth;
 
 public class DragonBreathParticle extends TextureSheetParticle {
+   private static final int COLOR_MIN = 11993298;
+   private static final int COLOR_MAX = 14614777;
+   private static final float COLOR_MIN_RED = 0.7176471F;
+   private static final float COLOR_MIN_GREEN = 0.0F;
+   private static final float COLOR_MIN_BLUE = 0.8235294F;
+   private static final float COLOR_MAX_RED = 0.8745098F;
+   private static final float COLOR_MAX_GREEN = 0.0F;
+   private static final float COLOR_MAX_BLUE = 0.9764706F;
    private boolean hasHitGround;
    private final SpriteSet sprites;
 
-   private DragonBreathParticle(ClientLevel var1, double var2, double var4, double var6, double var8, double var10, double var12, SpriteSet var14) {
+   DragonBreathParticle(ClientLevel var1, double var2, double var4, double var6, double var8, double var10, double var12, SpriteSet var14) {
       super(var1, var2, var4, var6);
+      this.friction = 0.96F;
       this.xd = var8;
       this.yd = var10;
       this.zd = var12;
       this.rCol = Mth.nextFloat(this.random, 0.7176471F, 0.8745098F);
       this.gCol = Mth.nextFloat(this.random, 0.0F, 0.0F);
@@ -45,14 +54,14 @@
          if (this.y == this.yo) {
             this.xd *= 1.1D;
             this.zd *= 1.1D;
          }
 
-         this.xd *= 0.9599999785423279D;
-         this.zd *= 0.9599999785423279D;
+         this.xd *= (double)this.friction;
+         this.zd *= (double)this.friction;
          if (this.hasHitGround) {
-            this.yd *= 0.9599999785423279D;
+            this.yd *= (double)this.friction;
          }
 
       }
    }
 
