--- net/minecraft/client/particle/EndRodParticle.java
+++ net/minecraft/client/particle/EndRodParticle.java
@@ -2,12 +2,12 @@
 
 import net.minecraft.client.multiplayer.ClientLevel;
 import net.minecraft.core.particles.SimpleParticleType;
 
 public class EndRodParticle extends SimpleAnimatedParticle {
-   private EndRodParticle(ClientLevel var1, double var2, double var4, double var6, double var8, double var10, double var12, SpriteSet var14) {
-      super(var1, var2, var4, var6, var14, -5.0E-4F);
+   EndRodParticle(ClientLevel var1, double var2, double var4, double var6, double var8, double var10, double var12, SpriteSet var14) {
+      super(var1, var2, var4, var6, var14, 0.0125F);
       this.xd = var8;
       this.yd = var10;
       this.zd = var12;
       this.quadSize *= 0.75F;
       this.lifetime = 60 + this.random.nextInt(12);
