--- net/minecraft/client/particle/ExplodeParticle.java
+++ net/minecraft/client/particle/ExplodeParticle.java
@@ -6,10 +6,12 @@
 public class ExplodeParticle extends TextureSheetParticle {
    private final SpriteSet sprites;
 
    protected ExplodeParticle(ClientLevel var1, double var2, double var4, double var6, double var8, double var10, double var12, SpriteSet var14) {
       super(var1, var2, var4, var6);
+      this.gravity = -0.1F;
+      this.friction = 0.9F;
       this.sprites = var14;
       this.xd = var8 + (Math.random() * 2.0D - 1.0D) * 0.05000000074505806D;
       this.yd = var10 + (Math.random() * 2.0D - 1.0D) * 0.05000000074505806D;
       this.zd = var12 + (Math.random() * 2.0D - 1.0D) * 0.05000000074505806D;
       float var15 = this.random.nextFloat() * 0.3F + 0.7F;
@@ -24,28 +26,12 @@
    public ParticleRenderType getRenderType() {
       return ParticleRenderType.PARTICLE_SHEET_OPAQUE;
    }
 
    public void tick() {
-      this.xo = this.x;
-      this.yo = this.y;
-      this.zo = this.z;
-      if (this.age++ >= this.lifetime) {
-         this.remove();
-      } else {
-         this.setSpriteFromAge(this.sprites);
-         this.yd += 0.004D;
-         this.move(this.xd, this.yd, this.zd);
-         this.xd *= 0.8999999761581421D;
-         this.yd *= 0.8999999761581421D;
-         this.zd *= 0.8999999761581421D;
-         if (this.onGround) {
-            this.xd *= 0.699999988079071D;
-            this.zd *= 0.699999988079071D;
-         }
-
-      }
+      super.tick();
+      this.setSpriteFromAge(this.sprites);
    }
 
    public static class Provider implements ParticleProvider<SimpleParticleType> {
       private final SpriteSet sprites;
 
