--- net/minecraft/client/particle/FallingDustParticle.java
+++ net/minecraft/client/particle/FallingDustParticle.java
@@ -12,11 +12,11 @@
 
 public class FallingDustParticle extends TextureSheetParticle {
    private final float rotSpeed;
    private final SpriteSet sprites;
 
-   private FallingDustParticle(ClientLevel var1, double var2, double var4, double var6, float var8, float var9, float var10, SpriteSet var11) {
+   FallingDustParticle(ClientLevel var1, double var2, double var4, double var6, float var8, float var9, float var10, SpriteSet var11) {
       super(var1, var2, var4, var6);
       this.sprites = var11;
       this.rCol = var8;
       this.gCol = var9;
       this.bCol = var10;
