--- net/minecraft/client/particle/FireworkParticles.java
+++ net/minecraft/client/particle/FireworkParticles.java
@@ -16,10 +16,13 @@
 import net.minecraft.util.Mth;
 import net.minecraft.world.item.DyeColor;
 import net.minecraft.world.item.FireworkRocketItem;
 
 public class FireworkParticles {
+   public FireworkParticles() {
+   }
+
    public static class SparkProvider implements ParticleProvider<SimpleParticleType> {
       private final SpriteSet sprites;
 
       public SparkProvider(SpriteSet var1) {
          this.sprites = var1;
@@ -74,11 +77,11 @@
       private float fadeG;
       private float fadeB;
       private boolean hasFade;
 
       private SparkParticle(ClientLevel var1, double var2, double var4, double var6, double var8, double var10, double var12, ParticleEngine var14, SpriteSet var15) {
-         super(var1, var2, var4, var6, var15, -0.004F);
+         super(var1, var2, var4, var6, var15, 0.1F);
          this.xd = var8;
          this.yd = var10;
          this.zd = var12;
          this.engine = var14;
          this.quadSize *= 0.75F;
