--- net/minecraft/client/particle/FireworkParticles.java
+++ net/minecraft/client/particle/FireworkParticles.java
@@ -16,10 +16,13 @@
 import net.minecraft.util.Mth;
 import net.minecraft.world.item.DyeColor;
 import net.minecraft.world.item.FireworkRocketItem;
 
 public class FireworkParticles {
+   public FireworkParticles() {
+   }
+
    public static class SparkProvider implements ParticleProvider<SimpleParticleType> {
       private final SpriteSet sprites;
 
       public SparkProvider(SpriteSet var1) {
          this.sprites = var1;
@@ -45,11 +48,11 @@
          return var15;
       }
    }
 
    public static class OverlayParticle extends TextureSheetParticle {
-      private OverlayParticle(ClientLevel var1, double var2, double var4, double var6) {
+      OverlayParticle(ClientLevel var1, double var2, double var4, double var6) {
          super(var1, var2, var4, var6);
          this.lifetime = 4;
       }
 
       public ParticleRenderType getRenderType() {
@@ -73,12 +76,12 @@
       private float fadeR;
       private float fadeG;
       private float fadeB;
       private boolean hasFade;
 
-      private SparkParticle(ClientLevel var1, double var2, double var4, double var6, double var8, double var10, double var12, ParticleEngine var14, SpriteSet var15) {
-         super(var1, var2, var4, var6, var15, -0.004F);
+      SparkParticle(ClientLevel var1, double var2, double var4, double var6, double var8, double var10, double var12, ParticleEngine var14, SpriteSet var15) {
+         super(var1, var2, var4, var6, var15, 0.1F);
          this.xd = var8;
          this.yd = var10;
          this.zd = var12;
          this.engine = var14;
          this.quadSize *= 0.75F;
@@ -260,11 +263,11 @@
             for(int var15 = -var3; var15 <= var3; ++var15) {
                for(int var16 = -var3; var16 <= var3; ++var16) {
                   double var17 = (double)var15 + (this.random.nextDouble() - this.random.nextDouble()) * 0.5D;
                   double var19 = (double)var14 + (this.random.nextDouble() - this.random.nextDouble()) * 0.5D;
                   double var21 = (double)var16 + (this.random.nextDouble() - this.random.nextDouble()) * 0.5D;
-                  double var23 = (double)Mth.sqrt(var17 * var17 + var19 * var19 + var21 * var21) / var1 + this.random.nextGaussian() * 0.05D;
+                  double var23 = Math.sqrt(var17 * var17 + var19 * var19 + var21 * var21) / var1 + this.random.nextGaussian() * 0.05D;
                   this.createParticle(var8, var10, var12, var17 / var23, var19 / var23, var21 / var23, var4, var5, var6, var7);
                   if (var14 != -var3 && var14 != var3 && var15 != -var3 && var15 != var3) {
                      var16 += var3 * 2 - 1;
                   }
                }
