--- net/minecraft/client/particle/ItemPickupParticle.java
+++ net/minecraft/client/particle/ItemPickupParticle.java
@@ -11,10 +11,11 @@
 import net.minecraft.world.entity.Entity;
 import net.minecraft.world.entity.item.ItemEntity;
 import net.minecraft.world.phys.Vec3;
 
 public class ItemPickupParticle extends Particle {
+   private static final int LIFE_TIME = 3;
    private final RenderBuffers renderBuffers;
    private final Entity itemEntity;
    private final Entity target;
    private int life;
    private final EntityRenderDispatcher entityRenderDispatcher;
@@ -48,11 +49,11 @@
       double var11 = Mth.lerp((double)var4, this.itemEntity.getX(), var5);
       double var13 = Mth.lerp((double)var4, this.itemEntity.getY(), var7);
       double var15 = Mth.lerp((double)var4, this.itemEntity.getZ(), var9);
       MultiBufferSource.BufferSource var17 = this.renderBuffers.bufferSource();
       Vec3 var18 = var2.getPosition();
-      this.entityRenderDispatcher.render(this.itemEntity, var11 - var18.x(), var13 - var18.y(), var15 - var18.z(), this.itemEntity.yRot, var3, new PoseStack(), var17, this.entityRenderDispatcher.getPackedLightCoords(this.itemEntity, var3));
+      this.entityRenderDispatcher.render(this.itemEntity, var11 - var18.x(), var13 - var18.y(), var15 - var18.z(), this.itemEntity.getYRot(), var3, new PoseStack(), var17, this.entityRenderDispatcher.getPackedLightCoords(this.itemEntity, var3));
       var17.endBatch();
    }
 
    public void tick() {
       ++this.life;
