--- net/minecraft/client/particle/ItemPickupParticle.java
+++ net/minecraft/client/particle/ItemPickupParticle.java
@@ -11,10 +11,11 @@
 import net.minecraft.world.entity.Entity;
 import net.minecraft.world.entity.item.ItemEntity;
 import net.minecraft.world.phys.Vec3;
 
 public class ItemPickupParticle extends Particle {
+   private static final int LIFE_TIME = 3;
    private final RenderBuffers renderBuffers;
    private final Entity itemEntity;
    private final Entity target;
    private int life;
    private final EntityRenderDispatcher entityRenderDispatcher;
