--- net/minecraft/client/particle/LavaParticle.java
+++ net/minecraft/client/particle/LavaParticle.java
@@ -5,10 +5,12 @@
 import net.minecraft.core.particles.SimpleParticleType;
 
 public class LavaParticle extends TextureSheetParticle {
    private LavaParticle(ClientLevel var1, double var2, double var4, double var6) {
       super(var1, var2, var4, var6, 0.0D, 0.0D, 0.0D);
+      this.gravity = 0.75F;
+      this.friction = 0.999F;
       this.xd *= 0.800000011920929D;
       this.yd *= 0.800000011920929D;
       this.zd *= 0.800000011920929D;
       this.yd = (double)(this.random.nextFloat() * 0.4F + 0.05F);
       this.quadSize *= this.random.nextFloat() * 2.0F + 0.2F;
@@ -30,32 +32,18 @@
       float var2 = ((float)this.age + var1) / (float)this.lifetime;
       return this.quadSize * (1.0F - var2 * var2);
    }
 
    public void tick() {
-      this.xo = this.x;
-      this.yo = this.y;
-      this.zo = this.z;
-      float var1 = (float)this.age / (float)this.lifetime;
-      if (this.random.nextFloat() > var1) {
-         this.level.addParticle(ParticleTypes.SMOKE, this.x, this.y, this.z, this.xd, this.yd, this.zd);
-      }
-
-      if (this.age++ >= this.lifetime) {
-         this.remove();
-      } else {
-         this.yd -= 0.03D;
-         this.move(this.xd, this.yd, this.zd);
-         this.xd *= 0.9990000128746033D;
-         this.yd *= 0.9990000128746033D;
-         this.zd *= 0.9990000128746033D;
-         if (this.onGround) {
-            this.xd *= 0.699999988079071D;
-            this.zd *= 0.699999988079071D;
+      super.tick();
+      if (!this.removed) {
+         float var1 = (float)this.age / (float)this.lifetime;
+         if (this.random.nextFloat() > var1) {
+            this.level.addParticle(ParticleTypes.SMOKE, this.x, this.y, this.z, this.xd, this.yd, this.zd);
          }
-
       }
+
    }
 
    public static class Provider implements ParticleProvider<SimpleParticleType> {
       private final SpriteSet sprite;
 
