--- net/minecraft/client/particle/MobAppearanceParticle.java
+++ net/minecraft/client/particle/MobAppearanceParticle.java
@@ -5,10 +5,11 @@
 import com.mojang.math.Vector3f;
 import net.minecraft.client.Camera;
 import net.minecraft.client.Minecraft;
 import net.minecraft.client.model.GuardianModel;
 import net.minecraft.client.model.Model;
+import net.minecraft.client.model.geom.ModelLayers;
 import net.minecraft.client.multiplayer.ClientLevel;
 import net.minecraft.client.renderer.MultiBufferSource;
 import net.minecraft.client.renderer.RenderType;
 import net.minecraft.client.renderer.entity.ElderGuardianRenderer;
 import net.minecraft.client.renderer.texture.OverlayTexture;
@@ -19,12 +20,12 @@
    private final Model model;
    private final RenderType renderType;
 
    private MobAppearanceParticle(ClientLevel var1, double var2, double var4, double var6) {
       super(var1, var2, var4, var6);
-      this.model = new GuardianModel();
       this.renderType = RenderType.entityTranslucent(ElderGuardianRenderer.GUARDIAN_ELDER_LOCATION);
+      this.model = new GuardianModel(Minecraft.getInstance().getEntityModels().bakeLayer(ModelLayers.ELDER_GUARDIAN));
       this.gravity = 0.0F;
       this.lifetime = 30;
    }
 
    public ParticleRenderType getRenderType() {
