--- net/minecraft/client/particle/NoteParticle.java
+++ net/minecraft/client/particle/NoteParticle.java
@@ -3,12 +3,14 @@
 import net.minecraft.client.multiplayer.ClientLevel;
 import net.minecraft.core.particles.SimpleParticleType;
 import net.minecraft.util.Mth;
 
 public class NoteParticle extends TextureSheetParticle {
-   private NoteParticle(ClientLevel var1, double var2, double var4, double var6, double var8) {
+   NoteParticle(ClientLevel var1, double var2, double var4, double var6, double var8) {
       super(var1, var2, var4, var6, 0.0D, 0.0D, 0.0D);
+      this.friction = 0.66F;
+      this.speedUpWhenYMotionIsBlocked = true;
       this.xd *= 0.009999999776482582D;
       this.yd *= 0.009999999776482582D;
       this.zd *= 0.009999999776482582D;
       this.yd += 0.2D;
       this.rCol = Math.max(0.0F, Mth.sin(((float)var8 + 0.0F) * 6.2831855F) * 0.65F + 0.35F);
@@ -24,34 +26,10 @@
 
    public float getQuadSize(float var1) {
       return this.quadSize * Mth.clamp(((float)this.age + var1) / (float)this.lifetime * 32.0F, 0.0F, 1.0F);
    }
 
-   public void tick() {
-      this.xo = this.x;
-      this.yo = this.y;
-      this.zo = this.z;
-      if (this.age++ >= this.lifetime) {
-         this.remove();
-      } else {
-         this.move(this.xd, this.yd, this.zd);
-         if (this.y == this.yo) {
-            this.xd *= 1.1D;
-            this.zd *= 1.1D;
-         }
-
-         this.xd *= 0.6600000262260437D;
-         this.yd *= 0.6600000262260437D;
-         this.zd *= 0.6600000262260437D;
-         if (this.onGround) {
-            this.xd *= 0.699999988079071D;
-            this.zd *= 0.699999988079071D;
-         }
-
-      }
-   }
-
    public static class Provider implements ParticleProvider<SimpleParticleType> {
       private final SpriteSet sprite;
 
       public Provider(SpriteSet var1) {
          this.sprite = var1;
