--- net/minecraft/client/particle/ParticleRenderType.java
+++ net/minecraft/client/particle/ParticleRenderType.java
@@ -3,21 +3,22 @@
 import com.mojang.blaze3d.platform.GlStateManager;
 import com.mojang.blaze3d.systems.RenderSystem;
 import com.mojang.blaze3d.vertex.BufferBuilder;
 import com.mojang.blaze3d.vertex.DefaultVertexFormat;
 import com.mojang.blaze3d.vertex.Tesselator;
+import com.mojang.blaze3d.vertex.VertexFormat;
 import net.minecraft.client.renderer.texture.TextureAtlas;
 import net.minecraft.client.renderer.texture.TextureManager;
 
 public interface ParticleRenderType {
    ParticleRenderType TERRAIN_SHEET = new ParticleRenderType() {
       public void begin(BufferBuilder var1, TextureManager var2) {
          RenderSystem.enableBlend();
          RenderSystem.defaultBlendFunc();
          RenderSystem.depthMask(true);
          var2.bind(TextureAtlas.LOCATION_BLOCKS);
-         var1.begin(7, DefaultVertexFormat.PARTICLE);
+         var1.begin(VertexFormat.Mode.QUADS, DefaultVertexFormat.PARTICLE);
       }
 
       public void end(Tesselator var1) {
          var1.end();
       }
@@ -29,11 +30,11 @@
    ParticleRenderType PARTICLE_SHEET_OPAQUE = new ParticleRenderType() {
       public void begin(BufferBuilder var1, TextureManager var2) {
          RenderSystem.disableBlend();
          RenderSystem.depthMask(true);
          var2.bind(TextureAtlas.LOCATION_PARTICLES);
-         var1.begin(7, DefaultVertexFormat.PARTICLE);
+         var1.begin(VertexFormat.Mode.QUADS, DefaultVertexFormat.PARTICLE);
       }
 
       public void end(Tesselator var1) {
          var1.end();
       }
@@ -45,13 +46,13 @@
    ParticleRenderType PARTICLE_SHEET_TRANSLUCENT = new ParticleRenderType() {
       public void begin(BufferBuilder var1, TextureManager var2) {
          RenderSystem.depthMask(true);
          var2.bind(TextureAtlas.LOCATION_PARTICLES);
          RenderSystem.enableBlend();
-         RenderSystem.blendFuncSeparate(GlStateManager.SourceFactor.SRC_ALPHA, GlStateManager.DestFactor.ONE_MINUS_SRC_ALPHA, GlStateManager.SourceFactor.ONE, GlStateManager.DestFactor.ONE_MINUS_SRC_ALPHA);
+         RenderSystem.blendFunc(GlStateManager.SourceFactor.SRC_ALPHA, GlStateManager.DestFactor.ONE_MINUS_SRC_ALPHA);
          RenderSystem.alphaFunc(516, 0.003921569F);
-         var1.begin(7, DefaultVertexFormat.PARTICLE);
+         var1.begin(VertexFormat.Mode.QUADS, DefaultVertexFormat.PARTICLE);
       }
 
       public void end(Tesselator var1) {
          var1.end();
       }
@@ -63,11 +64,11 @@
    ParticleRenderType PARTICLE_SHEET_LIT = new ParticleRenderType() {
       public void begin(BufferBuilder var1, TextureManager var2) {
          RenderSystem.disableBlend();
          RenderSystem.depthMask(true);
          var2.bind(TextureAtlas.LOCATION_PARTICLES);
-         var1.begin(7, DefaultVertexFormat.PARTICLE);
+         var1.begin(VertexFormat.Mode.QUADS, DefaultVertexFormat.PARTICLE);
       }
 
       public void end(Tesselator var1) {
          var1.end();
       }
