--- net/minecraft/client/particle/PlayerCloudParticle.java
+++ net/minecraft/client/particle/PlayerCloudParticle.java
@@ -8,10 +8,11 @@
 public class PlayerCloudParticle extends TextureSheetParticle {
    private final SpriteSet sprites;
 
    private PlayerCloudParticle(ClientLevel var1, double var2, double var4, double var6, double var8, double var10, double var12, SpriteSet var14) {
       super(var1, var2, var4, var6, 0.0D, 0.0D, 0.0D);
+      this.friction = 0.96F;
       this.sprites = var14;
       float var15 = 2.5F;
       this.xd *= 0.10000000149011612D;
       this.yd *= 0.10000000149011612D;
       this.zd *= 0.10000000149011612D;
@@ -36,37 +37,24 @@
    public float getQuadSize(float var1) {
       return this.quadSize * Mth.clamp(((float)this.age + var1) / (float)this.lifetime * 32.0F, 0.0F, 1.0F);
    }
 
    public void tick() {
-      this.xo = this.x;
-      this.yo = this.y;
-      this.zo = this.z;
-      if (this.age++ >= this.lifetime) {
-         this.remove();
-      } else {
+      super.tick();
+      if (!this.removed) {
          this.setSpriteFromAge(this.sprites);
-         this.move(this.xd, this.yd, this.zd);
-         this.xd *= 0.9599999785423279D;
-         this.yd *= 0.9599999785423279D;
-         this.zd *= 0.9599999785423279D;
          Player var1 = this.level.getNearestPlayer(this.x, this.y, this.z, 2.0D, false);
          if (var1 != null) {
             double var2 = var1.getY();
             if (this.y > var2) {
                this.y += (var2 - this.y) * 0.2D;
                this.yd += (var1.getDeltaMovement().y - this.yd) * 0.2D;
                this.setPos(this.x, this.y, this.z);
             }
          }
-
-         if (this.onGround) {
-            this.xd *= 0.699999988079071D;
-            this.zd *= 0.699999988079071D;
-         }
-
       }
+
    }
 
    public static class SneezeProvider implements ParticleProvider<SimpleParticleType> {
       private final SpriteSet sprites;
 
