--- net/minecraft/client/particle/ReversePortalParticle.java
+++ net/minecraft/client/particle/ReversePortalParticle.java
@@ -2,11 +2,11 @@
 
 import net.minecraft.client.multiplayer.ClientLevel;
 import net.minecraft.core.particles.SimpleParticleType;
 
 public class ReversePortalParticle extends PortalParticle {
-   private ReversePortalParticle(ClientLevel var1, double var2, double var4, double var6, double var8, double var10, double var12) {
+   ReversePortalParticle(ClientLevel var1, double var2, double var4, double var6, double var8, double var10, double var12) {
       super(var1, var2, var4, var6, var8, var10, var12);
       this.quadSize = (float)((double)this.quadSize * 1.5D);
       this.lifetime = (int)(Math.random() * 2.0D) + 60;
    }
 
