--- net/minecraft/client/particle/SoulParticle.java
+++ net/minecraft/client/particle/SoulParticle.java
@@ -17,14 +17,11 @@
       return ParticleRenderType.PARTICLE_SHEET_TRANSLUCENT;
    }
 
    public void tick() {
       super.tick();
-      if (!this.removed) {
-         this.setSpriteFromAge(this.sprites);
-      }
-
+      this.setSpriteFromAge(this.sprites);
    }
 
    public static class Provider implements ParticleProvider<SimpleParticleType> {
       private final SpriteSet sprite;
 
