--- net/minecraft/client/particle/SoulParticle.java
+++ net/minecraft/client/particle/SoulParticle.java
@@ -4,11 +4,11 @@
 import net.minecraft.core.particles.SimpleParticleType;
 
 public class SoulParticle extends RisingParticle {
    private final SpriteSet sprites;
 
-   private SoulParticle(ClientLevel var1, double var2, double var4, double var6, double var8, double var10, double var12, SpriteSet var14) {
+   SoulParticle(ClientLevel var1, double var2, double var4, double var6, double var8, double var10, double var12, SpriteSet var14) {
       super(var1, var2, var4, var6, var8, var10, var12);
       this.sprites = var14;
       this.scale(1.5F);
       this.setSpriteFromAge(var14);
    }
@@ -17,14 +17,11 @@
       return ParticleRenderType.PARTICLE_SHEET_TRANSLUCENT;
    }
 
    public void tick() {
       super.tick();
-      if (!this.removed) {
-         this.setSpriteFromAge(this.sprites);
-      }
-
+      this.setSpriteFromAge(this.sprites);
    }
 
    public static class Provider implements ParticleProvider<SimpleParticleType> {
       private final SpriteSet sprite;
 
