--- net/minecraft/client/particle/SpitParticle.java
+++ net/minecraft/client/particle/SpitParticle.java
@@ -7,15 +7,10 @@
    private SpitParticle(ClientLevel var1, double var2, double var4, double var6, double var8, double var10, double var12, SpriteSet var14) {
       super(var1, var2, var4, var6, var8, var10, var12, var14);
       this.gravity = 0.5F;
    }
 
-   public void tick() {
-      super.tick();
-      this.yd -= 0.004D + 0.04D * (double)this.gravity;
-   }
-
    public static class Provider implements ParticleProvider<SimpleParticleType> {
       private final SpriteSet sprites;
 
       public Provider(SpriteSet var1) {
          this.sprites = var1;
