--- net/minecraft/client/particle/SplashParticle.java
+++ net/minecraft/client/particle/SplashParticle.java
@@ -2,11 +2,11 @@
 
 import net.minecraft.client.multiplayer.ClientLevel;
 import net.minecraft.core.particles.SimpleParticleType;
 
 public class SplashParticle extends WaterDropParticle {
-   private SplashParticle(ClientLevel var1, double var2, double var4, double var6, double var8, double var10, double var12) {
+   SplashParticle(ClientLevel var1, double var2, double var4, double var6, double var8, double var10, double var12) {
       super(var1, var2, var4, var6);
       this.gravity = 0.04F;
       if (var10 == 0.0D && (var8 != 0.0D || var12 != 0.0D)) {
          this.xd = var8;
          this.yd = 0.1D;
