--- net/minecraft/client/particle/StationaryItemParticle.java
+++ net/minecraft/client/particle/StationaryItemParticle.java
@@ -1,0 +1,44 @@
+package net.minecraft.client.particle;
+
+import net.minecraft.client.Minecraft;
+import net.minecraft.client.multiplayer.ClientLevel;
+import net.minecraft.core.particles.SimpleParticleType;
+import net.minecraft.world.item.Items;
+import net.minecraft.world.level.ItemLike;
+import net.minecraft.world.level.block.Blocks;
+
+public class StationaryItemParticle extends TextureSheetParticle {
+   private StationaryItemParticle(ClientLevel var1, double var2, double var4, double var6, ItemLike var8) {
+      super(var1, var2, var4, var6);
+      this.setSprite(Minecraft.getInstance().getItemRenderer().getItemModelShaper().getParticleIcon(var8));
+      this.gravity = 0.0F;
+      this.lifetime = 80;
+      this.hasPhysics = false;
+   }
+
+   public ParticleRenderType getRenderType() {
+      return ParticleRenderType.TERRAIN_SHEET;
+   }
+
+   public float getQuadSize(float var1) {
+      return 0.5F;
+   }
+
+   public static class LightProvider implements ParticleProvider<SimpleParticleType> {
+      public LightProvider() {
+      }
+
+      public Particle createParticle(SimpleParticleType var1, ClientLevel var2, double var3, double var5, double var7, double var9, double var11, double var13) {
+         return new StationaryItemParticle(var2, var3, var5, var7, Items.LIGHT);
+      }
+   }
+
+   public static class BarrierProvider implements ParticleProvider<SimpleParticleType> {
+      public BarrierProvider() {
+      }
+
+      public Particle createParticle(SimpleParticleType var1, ClientLevel var2, double var3, double var5, double var7, double var9, double var11, double var13) {
+         return new StationaryItemParticle(var2, var3, var5, var7, Blocks.BARRIER.asItem());
+      }
+   }
+}
