--- net/minecraft/client/particle/SuspendedTownParticle.java
+++ net/minecraft/client/particle/SuspendedTownParticle.java
@@ -2,11 +2,11 @@
 
 import net.minecraft.client.multiplayer.ClientLevel;
 import net.minecraft.core.particles.SimpleParticleType;
 
 public class SuspendedTownParticle extends TextureSheetParticle {
-   private SuspendedTownParticle(ClientLevel var1, double var2, double var4, double var6, double var8, double var10, double var12) {
+   SuspendedTownParticle(ClientLevel var1, double var2, double var4, double var6, double var8, double var10, double var12) {
       super(var1, var2, var4, var6, var8, var10, var12);
       float var14 = this.random.nextFloat() * 0.1F + 0.2F;
       this.rCol = var14;
       this.gCol = var14;
       this.bCol = var14;
