--- net/minecraft/client/particle/TextureSheetParticle.java
+++ net/minecraft/client/particle/TextureSheetParticle.java
@@ -37,8 +37,11 @@
    public void pickSprite(SpriteSet var1) {
       this.setSprite(var1.get(this.random));
    }
 
    public void setSpriteFromAge(SpriteSet var1) {
-      this.setSprite(var1.get(this.age, this.lifetime));
+      if (!this.removed) {
+         this.setSprite(var1.get(this.age, this.lifetime));
+      }
+
    }
 }
