--- net/minecraft/client/particle/TotemParticle.java
+++ net/minecraft/client/particle/TotemParticle.java
@@ -3,11 +3,12 @@
 import net.minecraft.client.multiplayer.ClientLevel;
 import net.minecraft.core.particles.SimpleParticleType;
 
 public class TotemParticle extends SimpleAnimatedParticle {
    private TotemParticle(ClientLevel var1, double var2, double var4, double var6, double var8, double var10, double var12, SpriteSet var14) {
-      super(var1, var2, var4, var6, var14, -0.05F);
+      super(var1, var2, var4, var6, var14, 1.25F);
+      this.friction = 0.6F;
       this.xd = var8;
       this.yd = var10;
       this.zd = var12;
       this.quadSize *= 0.75F;
       this.lifetime = 60 + this.random.nextInt(12);
@@ -16,11 +17,10 @@
          this.setColor(0.6F + this.random.nextFloat() * 0.2F, 0.6F + this.random.nextFloat() * 0.3F, this.random.nextFloat() * 0.2F);
       } else {
          this.setColor(0.1F + this.random.nextFloat() * 0.2F, 0.4F + this.random.nextFloat() * 0.3F, this.random.nextFloat() * 0.2F);
       }
 
-      this.setBaseAirFriction(0.6F);
    }
 
    public static class Provider implements ParticleProvider<SimpleParticleType> {
       private final SpriteSet sprites;
 
