--- net/minecraft/client/particle/WakeParticle.java
+++ net/minecraft/client/particle/WakeParticle.java
@@ -4,11 +4,11 @@
 import net.minecraft.core.particles.SimpleParticleType;
 
 public class WakeParticle extends TextureSheetParticle {
    private final SpriteSet sprites;
 
-   private WakeParticle(ClientLevel var1, double var2, double var4, double var6, double var8, double var10, double var12, SpriteSet var14) {
+   WakeParticle(ClientLevel var1, double var2, double var4, double var6, double var8, double var10, double var12, SpriteSet var14) {
       super(var1, var2, var4, var6, 0.0D, 0.0D, 0.0D);
       this.sprites = var14;
       this.xd *= 0.30000001192092896D;
       this.yd = Math.random() * 0.20000000298023224D + 0.10000000149011612D;
       this.zd *= 0.30000001192092896D;
