--- net/minecraft/client/particle/WhiteAshParticle.java
+++ net/minecraft/client/particle/WhiteAshParticle.java
@@ -3,12 +3,14 @@
 import java.util.Random;
 import net.minecraft.client.multiplayer.ClientLevel;
 import net.minecraft.core.particles.SimpleParticleType;
 
 public class WhiteAshParticle extends BaseAshSmokeParticle {
+   private static final int COLOR_RGB24 = 12235202;
+
    protected WhiteAshParticle(ClientLevel var1, double var2, double var4, double var6, double var8, double var10, double var12, float var14, SpriteSet var15) {
-      super(var1, var2, var4, var6, 0.1F, -0.1F, 0.1F, var8, var10, var12, var14, var15, 0.0F, 20, -5.0E-4D, false);
+      super(var1, var2, var4, var6, 0.1F, -0.1F, 0.1F, var8, var10, var12, var14, var15, 0.0F, 20, 0.0125F, false);
       this.rCol = 0.7294118F;
       this.gCol = 0.69411767F;
       this.bCol = 0.7607843F;
    }
 
