--- net/minecraft/world/entity/monster/piglin/PiglinBrute.java
+++ net/minecraft/world/entity/monster/piglin/PiglinBrute.java
@@ -27,10 +27,13 @@
 import net.minecraft.world.level.Level;
 import net.minecraft.world.level.ServerLevelAccessor;
 import net.minecraft.world.level.block.state.BlockState;
 
 public class PiglinBrute extends AbstractPiglin {
+   private static final int MAX_HEALTH = 50;
+   private static final float MOVEMENT_SPEED_WHEN_FIGHTING = 0.35F;
+   private static final int ATTACK_DAMAGE = 7;
    protected static final ImmutableList<SensorType<? extends Sensor<? super PiglinBrute>>> SENSOR_TYPES;
    protected static final ImmutableList<MemoryModuleType<?>> MEMORY_TYPES;
 
    public PiglinBrute(EntityType<? extends PiglinBrute> var1, Level var2) {
       super(var1, var2);
@@ -67,11 +70,11 @@
    public boolean canHunt() {
       return false;
    }
 
    public boolean wantsToPickUp(ItemStack var1) {
-      return var1.getItem() == Items.GOLDEN_AXE ? super.wantsToPickUp(var1) : false;
+      return var1.is(Items.GOLDEN_AXE) ? super.wantsToPickUp(var1) : false;
    }
 
    protected void customServerAiStep() {
       this.level.getProfiler().push("piglinBruteBrain");
       this.getBrain().tick((ServerLevel)this.level, this);
@@ -79,10 +82,14 @@
       PiglinBruteAi.updateActivity(this);
       PiglinBruteAi.maybePlayActivitySound(this);
       super.customServerAiStep();
    }
 
+   public PiglinArmPose getArmPose() {
+      return this.isAggressive() && this.isHoldingMeleeWeapon() ? PiglinArmPose.ATTACKING_WITH_MELEE_WEAPON : PiglinArmPose.DEFAULT;
+   }
+
    public boolean hurt(DamageSource var1, float var2) {
       boolean var3 = super.hurt(var1, var2);
       if (this.level.isClientSide) {
          return false;
       } else {
