--- net/minecraft/world/entity/monster/piglin/StopHoldingItemIfNoLongerAdmiring.java
+++ net/minecraft/world/entity/monster/piglin/StopHoldingItemIfNoLongerAdmiring.java
@@ -11,11 +11,11 @@
    public StopHoldingItemIfNoLongerAdmiring() {
       super(ImmutableMap.of(MemoryModuleType.ADMIRING_ITEM, MemoryStatus.VALUE_ABSENT));
    }
 
    protected boolean checkExtraStartConditions(ServerLevel var1, E var2) {
-      return !var2.getOffhandItem().isEmpty() && var2.getOffhandItem().getItem() != Items.SHIELD;
+      return !var2.getOffhandItem().isEmpty() && !var2.getOffhandItem().is(Items.SHIELD);
    }
 
    protected void start(ServerLevel var1, E var2, long var3) {
       PiglinAi.stopHoldingOffHandItem(var2, true);
    }
