--- net/minecraft/world/level/lighting/BlockLightSectionStorage.java
+++ net/minecraft/world/level/lighting/BlockLightSectionStorage.java
@@ -16,11 +16,11 @@
       long var3 = SectionPos.blockToSection(var1);
       DataLayer var5 = this.getDataLayer(var3, false);
       return var5 == null ? 0 : var5.get(SectionPos.sectionRelative(BlockPos.getX(var1)), SectionPos.sectionRelative(BlockPos.getY(var1)), SectionPos.sectionRelative(BlockPos.getZ(var1)));
    }
 
-   public static final class BlockDataLayerStorageMap extends DataLayerStorageMap<BlockLightSectionStorage.BlockDataLayerStorageMap> {
+   protected static final class BlockDataLayerStorageMap extends DataLayerStorageMap<BlockLightSectionStorage.BlockDataLayerStorageMap> {
       public BlockDataLayerStorageMap(Long2ObjectOpenHashMap<DataLayer> var1) {
          super(var1);
       }
 
       public BlockLightSectionStorage.BlockDataLayerStorageMap copy() {
