--- net/minecraft/world/level/lighting/FlatDataLayer.java
+++ net/minecraft/world/level/lighting/FlatDataLayer.java
@@ -1,10 +1,12 @@
 package net.minecraft.world.level.lighting;
 
 import net.minecraft.world.level.chunk.DataLayer;
 
 public class FlatDataLayer extends DataLayer {
+   public static final int SIZE = 128;
+
    public FlatDataLayer() {
       super(128);
    }
 
    public FlatDataLayer(DataLayer var1, int var2) {
