--- net/minecraft/client/renderer/block/model/BakedQuad.java
+++ net/minecraft/client/renderer/block/model/BakedQuad.java
@@ -16,10 +16,14 @@
       this.direction = var3;
       this.sprite = var4;
       this.shade = var5;
    }
 
+   public TextureAtlasSprite getSprite() {
+      return this.sprite;
+   }
+
    public int[] getVertices() {
       return this.vertices;
    }
 
    public boolean isTinted() {
