--- net/minecraft/client/renderer/block/model/BlockFaceUV.java
+++ net/minecraft/client/renderer/block/model/BlockFaceUV.java
@@ -50,11 +50,13 @@
          this.uvs = var1;
       }
 
    }
 
-   public static class Deserializer implements JsonDeserializer<BlockFaceUV> {
+   protected static class Deserializer implements JsonDeserializer<BlockFaceUV> {
+      private static final int DEFAULT_ROTATION = 0;
+
       protected Deserializer() {
       }
 
       public BlockFaceUV deserialize(JsonElement var1, Type var2, JsonDeserializationContext var3) throws JsonParseException {
          JsonObject var4 = var1.getAsJsonObject();
