--- net/minecraft/client/renderer/block/model/FaceBakery.java
+++ net/minecraft/client/renderer/block/model/FaceBakery.java
@@ -135,26 +135,26 @@
       if (var2 != null) {
          Vector3f var3;
          Vector3f var4;
          switch(var2.axis) {
          case X:
-            var3 = new Vector3f(1.0F, 0.0F, 0.0F);
+            var3 = Vector3f.XP;
             var4 = new Vector3f(0.0F, 1.0F, 1.0F);
             break;
          case Y:
-            var3 = new Vector3f(0.0F, 1.0F, 0.0F);
+            var3 = Vector3f.YP;
             var4 = new Vector3f(1.0F, 0.0F, 1.0F);
             break;
          case Z:
-            var3 = new Vector3f(0.0F, 0.0F, 1.0F);
+            var3 = Vector3f.ZP;
             var4 = new Vector3f(1.0F, 1.0F, 0.0F);
             break;
          default:
             throw new IllegalArgumentException("There are only 3 axes");
          }
 
-         Quaternion var5 = new Quaternion(var3, var2.angle, true);
+         Quaternion var5 = var3.rotationDegrees(var2.angle);
          if (var2.rescale) {
             if (Math.abs(var2.angle) == 22.5F) {
                var4.mul(RESCALE_22_5);
             } else {
                var4.mul(RESCALE_45);
