--- net/minecraft/client/renderer/block/model/ItemOverrides.java
+++ net/minecraft/client/renderer/block/model/ItemOverrides.java
@@ -37,21 +37,21 @@
       }
 
    }
 
    @Nullable
-   public BakedModel resolve(BakedModel var1, ItemStack var2, @Nullable ClientLevel var3, @Nullable LivingEntity var4) {
+   public BakedModel resolve(BakedModel var1, ItemStack var2, @Nullable ClientLevel var3, @Nullable LivingEntity var4, int var5) {
       if (!this.overrides.isEmpty()) {
-         for(int var5 = 0; var5 < this.overrides.size(); ++var5) {
-            ItemOverride var6 = (ItemOverride)this.overrides.get(var5);
-            if (var6.test(var2, var3, var4)) {
-               BakedModel var7 = (BakedModel)this.overrideModels.get(var5);
-               if (var7 == null) {
+         for(int var6 = 0; var6 < this.overrides.size(); ++var6) {
+            ItemOverride var7 = (ItemOverride)this.overrides.get(var6);
+            if (var7.test(var2, var3, var4, var5)) {
+               BakedModel var8 = (BakedModel)this.overrideModels.get(var6);
+               if (var8 == null) {
                   return var1;
                }
 
-               return var7;
+               return var8;
             }
          }
       }
 
       return var1;
