--- net/minecraft/client/renderer/block/model/ItemOverride.java
+++ net/minecraft/client/renderer/block/model/ItemOverride.java
@@ -32,24 +32,24 @@
 
    public ResourceLocation getModel() {
       return this.model;
    }
 
-   boolean test(ItemStack var1, @Nullable ClientLevel var2, @Nullable LivingEntity var3) {
-      Item var4 = var1.getItem();
-      Iterator var5 = this.predicates.entrySet().iterator();
+   boolean test(ItemStack var1, @Nullable ClientLevel var2, @Nullable LivingEntity var3, int var4) {
+      Item var5 = var1.getItem();
+      Iterator var6 = this.predicates.entrySet().iterator();
 
-      Entry var6;
-      ItemPropertyFunction var7;
+      Entry var7;
+      ItemPropertyFunction var8;
       do {
-         if (!var5.hasNext()) {
+         if (!var6.hasNext()) {
             return true;
          }
 
-         var6 = (Entry)var5.next();
-         var7 = ItemProperties.getProperty(var4, (ResourceLocation)var6.getKey());
-      } while(var7 != null && var7.call(var1, var2, var3) >= (Float)var6.getValue());
+         var7 = (Entry)var6.next();
+         var8 = ItemProperties.getProperty(var5, (ResourceLocation)var7.getKey());
+      } while(var8 != null && var8.call(var1, var2, var3, var4) >= (Float)var7.getValue());
 
       return false;
    }
 
    public static class Deserializer implements JsonDeserializer<ItemOverride> {
