--- net/minecraft/client/renderer/block/model/ItemTransforms.java
+++ net/minecraft/client/renderer/block/model/ItemTransforms.java
@@ -69,11 +69,30 @@
 
    public boolean hasTransform(ItemTransforms.TransformType var1) {
       return this.getTransform(var1) != ItemTransform.NO_TRANSFORM;
    }
 
-   public static class Deserializer implements JsonDeserializer<ItemTransforms> {
+   public static enum TransformType {
+      NONE,
+      THIRD_PERSON_LEFT_HAND,
+      THIRD_PERSON_RIGHT_HAND,
+      FIRST_PERSON_LEFT_HAND,
+      FIRST_PERSON_RIGHT_HAND,
+      HEAD,
+      GUI,
+      GROUND,
+      FIXED;
+
+      private TransformType() {
+      }
+
+      public boolean firstPerson() {
+         return this == FIRST_PERSON_LEFT_HAND || this == FIRST_PERSON_RIGHT_HAND;
+      }
+   }
+
+   protected static class Deserializer implements JsonDeserializer<ItemTransforms> {
       protected Deserializer() {
       }
 
       public ItemTransforms deserialize(JsonElement var1, Type var2, JsonDeserializationContext var3) throws JsonParseException {
          JsonObject var4 = var1.getAsJsonObject();
@@ -98,25 +117,6 @@
 
       private ItemTransform getTransform(JsonDeserializationContext var1, JsonObject var2, String var3) {
          return var2.has(var3) ? (ItemTransform)var1.deserialize(var2.get(var3), ItemTransform.class) : ItemTransform.NO_TRANSFORM;
       }
    }
-
-   public static enum TransformType {
-      NONE,
-      THIRD_PERSON_LEFT_HAND,
-      THIRD_PERSON_RIGHT_HAND,
-      FIRST_PERSON_LEFT_HAND,
-      FIRST_PERSON_RIGHT_HAND,
-      HEAD,
-      GUI,
-      GROUND,
-      FIXED;
-
-      private TransformType() {
-      }
-
-      public boolean firstPerson() {
-         return this == FIRST_PERSON_LEFT_HAND || this == FIRST_PERSON_RIGHT_HAND;
-      }
-   }
 }
