--- net/minecraft/client/renderer/block/model/ItemTransform.java
+++ net/minecraft/client/renderer/block/model/ItemTransform.java
@@ -57,14 +57,16 @@
       var1 = 31 * var1 + this.translation.hashCode();
       var1 = 31 * var1 + this.scale.hashCode();
       return var1;
    }
 
-   public static class Deserializer implements JsonDeserializer<ItemTransform> {
+   protected static class Deserializer implements JsonDeserializer<ItemTransform> {
       private static final Vector3f DEFAULT_ROTATION = new Vector3f(0.0F, 0.0F, 0.0F);
       private static final Vector3f DEFAULT_TRANSLATION = new Vector3f(0.0F, 0.0F, 0.0F);
       private static final Vector3f DEFAULT_SCALE = new Vector3f(1.0F, 1.0F, 1.0F);
+      public static final float MAX_TRANSLATION = 5.0F;
+      public static final float MAX_SCALE = 4.0F;
 
       protected Deserializer() {
       }
 
       public ItemTransform deserialize(JsonElement var1, Type var2, JsonDeserializationContext var3) throws JsonParseException {
