--- net/minecraft/client/renderer/block/model/MultiVariant.java
+++ net/minecraft/client/renderer/block/model/MultiVariant.java
@@ -93,14 +93,14 @@
 
             Iterator var6 = var5.iterator();
 
             while(var6.hasNext()) {
                JsonElement var7 = (JsonElement)var6.next();
-               var4.add(var3.deserialize(var7, Variant.class));
+               var4.add((Variant)var3.deserialize(var7, Variant.class));
             }
          } else {
-            var4.add(var3.deserialize(var1, Variant.class));
+            var4.add((Variant)var3.deserialize(var1, Variant.class));
          }
 
          return new MultiVariant(var4);
       }
    }
