--- net/minecraft/client/renderer/block/model/Variant.java
+++ net/minecraft/client/renderer/block/model/Variant.java
@@ -1,7 +1,8 @@
 package net.minecraft.client.renderer.block.model;
 
+import com.google.common.annotations.VisibleForTesting;
 import com.google.gson.JsonDeserializationContext;
 import com.google.gson.JsonDeserializer;
 import com.google.gson.JsonElement;
 import com.google.gson.JsonObject;
 import com.google.gson.JsonParseException;
@@ -41,11 +42,11 @@
    public int getWeight() {
       return this.weight;
    }
 
    public String toString() {
-      return "Variant{modelLocation=" + this.modelLocation + ", rotation=" + this.rotation + ", uvLock=" + this.uvLock + ", weight=" + this.weight + '}';
+      return "Variant{modelLocation=" + this.modelLocation + ", rotation=" + this.rotation + ", uvLock=" + this.uvLock + ", weight=" + this.weight + "}";
    }
 
    public boolean equals(Object var1) {
       if (this == var1) {
          return true;
@@ -64,10 +65,19 @@
       var1 = 31 * var1 + this.weight;
       return var1;
    }
 
    public static class Deserializer implements JsonDeserializer<Variant> {
+      @VisibleForTesting
+      static final boolean DEFAULT_UVLOCK = false;
+      @VisibleForTesting
+      static final int DEFAULT_WEIGHT = 1;
+      @VisibleForTesting
+      static final int DEFAULT_X_ROTATION = 0;
+      @VisibleForTesting
+      static final int DEFAULT_Y_ROTATION = 0;
+
       public Deserializer() {
       }
 
       public Variant deserialize(JsonElement var1, Type var2, JsonDeserializationContext var3) throws JsonParseException {
          JsonObject var4 = var1.getAsJsonObject();
