--- net/minecraft/client/gui/font/FontTexture.java
+++ net/minecraft/client/gui/font/FontTexture.java
@@ -9,10 +9,11 @@
 import net.minecraft.client.renderer.texture.AbstractTexture;
 import net.minecraft.resources.ResourceLocation;
 import net.minecraft.server.packs.resources.ResourceManager;
 
 public class FontTexture extends AbstractTexture {
+   private static final int SIZE = 256;
    private final ResourceLocation name;
    private final RenderType normalType;
    private final RenderType seeThroughType;
    private final boolean colored;
    private final FontTexture.Node root;
