--- net/minecraft/client/gui/font/FontManager.java
+++ net/minecraft/client/gui/font/FontManager.java
@@ -33,10 +33,11 @@
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
 
 public class FontManager implements AutoCloseable {
    private static final Logger LOGGER = LogManager.getLogger();
+   private static final String FONTS_PATH = "fonts.json";
    public static final ResourceLocation MISSING_FONT = new ResourceLocation("minecraft", "missing");
    private final FontSet missingFontSet;
    private final Map<ResourceLocation, FontSet> fontSets = Maps.newHashMap();
    private final TextureManager textureManager;
    private Map<ResourceLocation, ResourceLocation> renames = ImmutableMap.of();
@@ -88,11 +89,11 @@
                                     var9.add(var21);
                                  }
 
                                  var2.pop();
                               } catch (RuntimeException var49) {
-                                 FontManager.LOGGER.warn("Unable to read definition '{}' in fonts.json in resourcepack: '{}': {}", var8, var11.getSourceName(), var49.getMessage());
+                                 FontManager.LOGGER.warn("Unable to read definition '{}' in {} in resourcepack: '{}': {}", var8, "fonts.json", var11.getSourceName(), var49.getMessage());
                               }
                            }
 
                            var2.pop();
                         } catch (Throwable var50) {
@@ -128,15 +129,15 @@
                            }
                         }
 
                      }
                   } catch (RuntimeException var54) {
-                     FontManager.LOGGER.warn("Unable to load font '{}' in fonts.json in resourcepack: '{}': {}", var8, var11.getSourceName(), var54.getMessage());
+                     FontManager.LOGGER.warn("Unable to load font '{}' in {} in resourcepack: '{}': {}", var8, "fonts.json", var11.getSourceName(), var54.getMessage());
                   }
                }
             } catch (IOException var55) {
-               FontManager.LOGGER.warn("Unable to load font '{}' in fonts.json: {}", var8, var55.getMessage());
+               FontManager.LOGGER.warn("Unable to load font '{}' in {}: {}", var8, "fonts.json", var55.getMessage());
             }
 
             var2.push("caching");
             IntOpenHashSet var56 = new IntOpenHashSet();
             Iterator var57 = var9.iterator();
