--- net/minecraft/client/gui/font/FontManager.java
+++ net/minecraft/client/gui/font/FontManager.java
@@ -88,11 +88,11 @@
                                     var9.add(var21);
                                  }
 
                                  var2.pop();
                               } catch (RuntimeException var49) {
-                                 FontManager.LOGGER.warn("Unable to read definition '{}' in fonts.json in resourcepack: '{}': {}", var8, var11.getSourceName(), var49.getMessage());
+                                 FontManager.LOGGER.warn("Unable to read definition '{}' in {} in resourcepack: '{}': {}", var8, "fonts.json", var11.getSourceName(), var49.getMessage());
                               }
                            }
 
                            var2.pop();
                         } catch (Throwable var50) {
@@ -128,15 +128,15 @@
                            }
                         }
 
                      }
                   } catch (RuntimeException var54) {
-                     FontManager.LOGGER.warn("Unable to load font '{}' in fonts.json in resourcepack: '{}': {}", var8, var11.getSourceName(), var54.getMessage());
+                     FontManager.LOGGER.warn("Unable to load font '{}' in {} in resourcepack: '{}': {}", var8, "fonts.json", var11.getSourceName(), var54.getMessage());
                   }
                }
             } catch (IOException var55) {
-               FontManager.LOGGER.warn("Unable to load font '{}' in fonts.json: {}", var8, var55.getMessage());
+               FontManager.LOGGER.warn("Unable to load font '{}' in {}: {}", var8, "fonts.json", var55.getMessage());
             }
 
             var2.push("caching");
             IntOpenHashSet var56 = new IntOpenHashSet();
             Iterator var57 = var9.iterator();
