--- net/minecraft/world/level/saveddata/SaveDataDirtyRunnable.java
+++ net/minecraft/world/level/saveddata/SaveDataDirtyRunnable.java
@@ -1,13 +1,0 @@
-package net.minecraft.world.level.saveddata;
-
-public class SaveDataDirtyRunnable implements Runnable {
-   private final SavedData savedData;
-
-   public SaveDataDirtyRunnable(SavedData var1) {
-      this.savedData = var1;
-   }
-
-   public void run() {
-      this.savedData.setDirty();
-   }
-}
