--- net/minecraft/commands/synchronization/ArgumentTypes.java
+++ net/minecraft/commands/synchronization/ArgumentTypes.java
@@ -65,10 +65,13 @@
 public class ArgumentTypes {
    private static final Logger LOGGER = LogManager.getLogger();
    private static final Map<Class<?>, ArgumentTypes.Entry<?>> BY_CLASS = Maps.newHashMap();
    private static final Map<ResourceLocation, ArgumentTypes.Entry<?>> BY_NAME = Maps.newHashMap();
 
+   public ArgumentTypes() {
+   }
+
    public static <T extends ArgumentType<?>> void register(String var0, Class<T> var1, ArgumentSerializer<T> var2) {
       ResourceLocation var3 = new ResourceLocation(var0);
       if (BY_CLASS.containsKey(var1)) {
          throw new IllegalArgumentException("Class " + var1.getName() + " already has a serializer!");
       } else if (BY_NAME.containsKey(var3)) {
