--- net/minecraft/server/commands/AttributeCommand.java
+++ net/minecraft/server/commands/AttributeCommand.java
@@ -41,10 +41,13 @@
    });
    private static final Dynamic3CommandExceptionType ERROR_MODIFIER_ALREADY_PRESENT = new Dynamic3CommandExceptionType((var0, var1, var2) -> {
       return new TranslatableComponent("commands.attribute.failed.modifier_already_present", new Object[]{var2, var1, var0});
    });
 
+   public AttributeCommand() {
+   }
+
    public static void register(CommandDispatcher<CommandSourceStack> var0) {
       var0.register((LiteralArgumentBuilder)((LiteralArgumentBuilder)Commands.literal("attribute").requires((var0x) -> {
          return var0x.hasPermission(2);
       })).then(Commands.argument("target", EntityArgument.entity()).then(((RequiredArgumentBuilder)((RequiredArgumentBuilder)Commands.argument("attribute", ResourceLocationArgument.id()).suggests(AVAILABLE_ATTRIBUTES).then(((LiteralArgumentBuilder)Commands.literal("get").executes((var0x) -> {
          return getAttributeValue((CommandSourceStack)var0x.getSource(), EntityArgument.getEntity(var0x, "target"), ResourceLocationArgument.getAttribute(var0x, "attribute"), 1.0D);
