--- net/minecraft/server/commands/BanListCommands.java
+++ net/minecraft/server/commands/BanListCommands.java
@@ -11,10 +11,13 @@
 import net.minecraft.network.chat.TranslatableComponent;
 import net.minecraft.server.players.BanListEntry;
 import net.minecraft.server.players.PlayerList;
 
 public class BanListCommands {
+   public BanListCommands() {
+   }
+
    public static void register(CommandDispatcher<CommandSourceStack> var0) {
       var0.register((LiteralArgumentBuilder)((LiteralArgumentBuilder)((LiteralArgumentBuilder)((LiteralArgumentBuilder)Commands.literal("banlist").requires((var0x) -> {
          return var0x.hasPermission(3);
       })).executes((var0x) -> {
          PlayerList var1 = ((CommandSourceStack)var0x.getSource()).getServer().getPlayerList();
