--- net/minecraft/server/commands/BanPlayerCommands.java
+++ net/minecraft/server/commands/BanPlayerCommands.java
@@ -22,10 +22,13 @@
 import net.minecraft.server.players.UserBanListEntry;
 
 public class BanPlayerCommands {
    private static final SimpleCommandExceptionType ERROR_ALREADY_BANNED = new SimpleCommandExceptionType(new TranslatableComponent("commands.ban.failed"));
 
+   public BanPlayerCommands() {
+   }
+
    public static void register(CommandDispatcher<CommandSourceStack> var0) {
       var0.register((LiteralArgumentBuilder)((LiteralArgumentBuilder)Commands.literal("ban").requires((var0x) -> {
          return var0x.hasPermission(3);
       })).then(((RequiredArgumentBuilder)Commands.argument("targets", GameProfileArgument.gameProfile()).executes((var0x) -> {
          return banPlayers((CommandSourceStack)var0x.getSource(), GameProfileArgument.getGameProfiles(var0x, "targets"), (Component)null);
