--- net/minecraft/server/commands/BossBarCommands.java
+++ net/minecraft/server/commands/BossBarCommands.java
@@ -47,10 +47,13 @@
    private static final SimpleCommandExceptionType ERROR_ALREADY_VISIBLE = new SimpleCommandExceptionType(new TranslatableComponent("commands.bossbar.set.visibility.unchanged.visible"));
    public static final SuggestionProvider<CommandSourceStack> SUGGEST_BOSS_BAR = (var0, var1) -> {
       return SharedSuggestionProvider.suggestResource((Iterable)((CommandSourceStack)var0.getSource()).getServer().getCustomBossEvents().getIds(), var1);
    };
 
+   public BossBarCommands() {
+   }
+
    public static void register(CommandDispatcher<CommandSourceStack> var0) {
       var0.register((LiteralArgumentBuilder)((LiteralArgumentBuilder)((LiteralArgumentBuilder)((LiteralArgumentBuilder)((LiteralArgumentBuilder)((LiteralArgumentBuilder)Commands.literal("bossbar").requires((var0x) -> {
          return var0x.hasPermission(2);
       })).then(Commands.literal("add").then(Commands.argument("id", ResourceLocationArgument.id()).then(Commands.argument("name", ComponentArgument.textComponent()).executes((var0x) -> {
          return createBar((CommandSourceStack)var0x.getSource(), ResourceLocationArgument.getId(var0x, "id"), ComponentArgument.getComponent(var0x, "name"));
