--- net/minecraft/server/commands/CloneCommands.java
+++ net/minecraft/server/commands/CloneCommands.java
@@ -79,13 +79,13 @@
          return clone((CommandSourceStack)var0x.getSource(), BlockPosArgument.getLoadedBlockPos(var0x, "begin"), BlockPosArgument.getLoadedBlockPos(var0x, "end"), BlockPosArgument.getLoadedBlockPos(var0x, "destination"), BlockPredicateArgument.getBlockPredicate(var0x, "filter"), CloneCommands.Mode.NORMAL);
       }))))))));
    }
 
    private static int clone(CommandSourceStack var0, BlockPos var1, BlockPos var2, BlockPos var3, Predicate<BlockInWorld> var4, CloneCommands.Mode var5) throws CommandSyntaxException {
-      BoundingBox var6 = new BoundingBox(var1, var2);
+      BoundingBox var6 = BoundingBox.createProper(var1, var2);
       BlockPos var7 = var3.offset(var6.getLength());
-      BoundingBox var8 = new BoundingBox(var3, var7);
+      BoundingBox var8 = BoundingBox.createProper(var3, var7);
       if (!var5.canOverlap() && var8.intersects(var6)) {
          throw ERROR_OVERLAP.create();
       } else {
          int var9 = var6.getXSpan() * var6.getYSpan() * var6.getZSpan();
          if (var9 > 32768) {
@@ -174,11 +174,11 @@
                   BlockEntity var35 = var10.getBlockEntity(var34.pos);
                   if (var34.tag != null && var35 != null) {
                      var34.tag.putInt("x", var34.pos.getX());
                      var34.tag.putInt("y", var34.pos.getY());
                      var34.tag.putInt("z", var34.pos.getZ());
-                     var35.load(var34.state, var34.tag);
+                     var35.load(var34.tag);
                      var35.setChanged();
                   }
                }
 
                var32 = var28.iterator();
