--- net/minecraft/server/commands/DeOpCommands.java
+++ net/minecraft/server/commands/DeOpCommands.java
@@ -15,10 +15,13 @@
 import net.minecraft.server.players.PlayerList;
 
 public class DeOpCommands {
    private static final SimpleCommandExceptionType ERROR_NOT_OP = new SimpleCommandExceptionType(new TranslatableComponent("commands.deop.failed"));
 
+   public DeOpCommands() {
+   }
+
    public static void register(CommandDispatcher<CommandSourceStack> var0) {
       var0.register((LiteralArgumentBuilder)((LiteralArgumentBuilder)Commands.literal("deop").requires((var0x) -> {
          return var0x.hasPermission(3);
       })).then(Commands.argument("targets", GameProfileArgument.gameProfile()).suggests((var0x, var1) -> {
          return SharedSuggestionProvider.suggest(((CommandSourceStack)var0x.getSource()).getServer().getPlayerList().getOpNames(), var1);
