--- net/minecraft/server/commands/EnchantCommand.java
+++ net/minecraft/server/commands/EnchantCommand.java
@@ -34,10 +34,13 @@
    private static final Dynamic2CommandExceptionType ERROR_LEVEL_TOO_HIGH = new Dynamic2CommandExceptionType((var0, var1) -> {
       return new TranslatableComponent("commands.enchant.failed.level", new Object[]{var0, var1});
    });
    private static final SimpleCommandExceptionType ERROR_NOTHING_HAPPENED = new SimpleCommandExceptionType(new TranslatableComponent("commands.enchant.failed"));
 
+   public EnchantCommand() {
+   }
+
    public static void register(CommandDispatcher<CommandSourceStack> var0) {
       var0.register((LiteralArgumentBuilder)((LiteralArgumentBuilder)Commands.literal("enchant").requires((var0x) -> {
          return var0x.hasPermission(2);
       })).then(Commands.argument("targets", EntityArgument.entities()).then(((RequiredArgumentBuilder)Commands.argument("enchantment", ItemEnchantmentArgument.enchantment()).executes((var0x) -> {
          return enchant((CommandSourceStack)var0x.getSource(), EntityArgument.getEntities(var0x, "targets"), ItemEnchantmentArgument.getEnchantment(var0x, "enchantment"), 1);
