--- net/minecraft/server/commands/ExperienceCommand.java
+++ net/minecraft/server/commands/ExperienceCommand.java
@@ -21,10 +21,13 @@
 import net.minecraft.world.entity.player.Player;
 
 public class ExperienceCommand {
    private static final SimpleCommandExceptionType ERROR_SET_POINTS_INVALID = new SimpleCommandExceptionType(new TranslatableComponent("commands.experience.set.points.invalid"));
 
+   public ExperienceCommand() {
+   }
+
    public static void register(CommandDispatcher<CommandSourceStack> var0) {
       LiteralCommandNode var1 = var0.register((LiteralArgumentBuilder)((LiteralArgumentBuilder)((LiteralArgumentBuilder)((LiteralArgumentBuilder)Commands.literal("experience").requires((var0x) -> {
          return var0x.hasPermission(2);
       })).then(Commands.literal("add").then(Commands.argument("targets", EntityArgument.players()).then(((RequiredArgumentBuilder)((RequiredArgumentBuilder)Commands.argument("amount", IntegerArgumentType.integer()).executes((var0x) -> {
          return addExperience((CommandSourceStack)var0x.getSource(), EntityArgument.getPlayers(var0x, "targets"), IntegerArgumentType.getInteger(var0x, "amount"), ExperienceCommand.Type.POINTS);
