--- net/minecraft/server/commands/GameModeCommand.java
+++ net/minecraft/server/commands/GameModeCommand.java
@@ -23,17 +23,15 @@
       GameType[] var2 = GameType.values();
       int var3 = var2.length;
 
       for(int var4 = 0; var4 < var3; ++var4) {
          GameType var5 = var2[var4];
-         if (var5 != GameType.NOT_SET) {
-            var1.then(((LiteralArgumentBuilder)Commands.literal(var5.getName()).executes((var1x) -> {
-               return setMode(var1x, Collections.singleton(((CommandSourceStack)var1x.getSource()).getPlayerOrException()), var5);
-            })).then(Commands.argument("target", EntityArgument.players()).executes((var1x) -> {
-               return setMode(var1x, EntityArgument.getPlayers(var1x, "target"), var5);
-            })));
-         }
+         var1.then(((LiteralArgumentBuilder)Commands.literal(var5.getName()).executes((var1x) -> {
+            return setMode(var1x, Collections.singleton(((CommandSourceStack)var1x.getSource()).getPlayerOrException()), var5);
+         })).then(Commands.argument("target", EntityArgument.players()).executes((var1x) -> {
+            return setMode(var1x, EntityArgument.getPlayers(var1x, "target"), var5);
+         })));
       }
 
       var0.register(var1);
    }
 
@@ -55,12 +53,11 @@
       int var3 = 0;
       Iterator var4 = var1.iterator();
 
       while(var4.hasNext()) {
          ServerPlayer var5 = (ServerPlayer)var4.next();
-         if (var5.gameMode.getGameModeForPlayer() != var2) {
-            var5.setGameMode(var2);
+         if (var5.setGameMode(var2)) {
             logGamemodeChange((CommandSourceStack)var0.getSource(), var5, var2);
             ++var3;
          }
       }
 
