--- net/minecraft/server/commands/HelpCommand.java
+++ net/minecraft/server/commands/HelpCommand.java
@@ -15,10 +15,13 @@
 import net.minecraft.network.chat.TranslatableComponent;
 
 public class HelpCommand {
    private static final SimpleCommandExceptionType ERROR_FAILED = new SimpleCommandExceptionType(new TranslatableComponent("commands.help.failed"));
 
+   public HelpCommand() {
+   }
+
    public static void register(CommandDispatcher<CommandSourceStack> var0) {
       var0.register((LiteralArgumentBuilder)((LiteralArgumentBuilder)Commands.literal("help").executes((var1) -> {
          Map var2 = var0.getSmartUsage(var0.getRoot(), var1.getSource());
          Iterator var3 = var2.values().iterator();
 
