--- net/minecraft/server/commands/HelpCommand.java
+++ net/minecraft/server/commands/HelpCommand.java
@@ -15,32 +15,37 @@
 import net.minecraft.network.chat.TranslatableComponent;
 
 public class HelpCommand {
    private static final SimpleCommandExceptionType ERROR_FAILED = new SimpleCommandExceptionType(new TranslatableComponent("commands.help.failed"));
 
+   public HelpCommand() {
+   }
+
    public static void register(CommandDispatcher<CommandSourceStack> var0) {
       var0.register((LiteralArgumentBuilder)((LiteralArgumentBuilder)Commands.literal("help").executes((var1) -> {
-         Map var2 = var0.getSmartUsage(var0.getRoot(), var1.getSource());
+         Map var2 = var0.getSmartUsage(var0.getRoot(), (CommandSourceStack)var1.getSource());
          Iterator var3 = var2.values().iterator();
 
          while(var3.hasNext()) {
             String var4 = (String)var3.next();
             ((CommandSourceStack)var1.getSource()).sendSuccess(new TextComponent("/" + var4), false);
          }
 
          return var2.size();
       })).then(Commands.argument("command", StringArgumentType.greedyString()).executes((var1) -> {
-         ParseResults var2 = var0.parse(StringArgumentType.getString(var1, "command"), var1.getSource());
+         ParseResults var2 = var0.parse((String)StringArgumentType.getString(var1, "command"), (CommandSourceStack)var1.getSource());
          if (var2.getContext().getNodes().isEmpty()) {
             throw ERROR_FAILED.create();
          } else {
-            Map var3 = var0.getSmartUsage(((ParsedCommandNode)Iterables.getLast(var2.getContext().getNodes())).getNode(), var1.getSource());
+            Map var3 = var0.getSmartUsage(((ParsedCommandNode)Iterables.getLast(var2.getContext().getNodes())).getNode(), (CommandSourceStack)var1.getSource());
             Iterator var4 = var3.values().iterator();
 
             while(var4.hasNext()) {
                String var5 = (String)var4.next();
-               ((CommandSourceStack)var1.getSource()).sendSuccess(new TextComponent("/" + var2.getReader().getString() + " " + var5), false);
+               CommandSourceStack var10000 = (CommandSourceStack)var1.getSource();
+               String var10003 = var2.getReader().getString();
+               var10000.sendSuccess(new TextComponent("/" + var10003 + " " + var5), false);
             }
 
             return var3.size();
          }
       })));
