--- net/minecraft/server/commands/ListPlayersCommand.java
+++ net/minecraft/server/commands/ListPlayersCommand.java
@@ -13,10 +13,13 @@
 import net.minecraft.server.level.ServerPlayer;
 import net.minecraft.server.players.PlayerList;
 import net.minecraft.world.entity.player.Player;
 
 public class ListPlayersCommand {
+   public ListPlayersCommand() {
+   }
+
    public static void register(CommandDispatcher<CommandSourceStack> var0) {
       var0.register((LiteralArgumentBuilder)((LiteralArgumentBuilder)Commands.literal("list").executes((var0x) -> {
          return listPlayers((CommandSourceStack)var0x.getSource());
       })).then(Commands.literal("uuids").executes((var0x) -> {
          return listPlayersWithUuids((CommandSourceStack)var0x.getSource());
