--- net/minecraft/server/commands/LocateBiomeCommand.java
+++ net/minecraft/server/commands/LocateBiomeCommand.java
@@ -19,11 +19,16 @@
       return new TranslatableComponent("commands.locatebiome.invalid", new Object[]{var0});
    });
    private static final DynamicCommandExceptionType ERROR_BIOME_NOT_FOUND = new DynamicCommandExceptionType((var0) -> {
       return new TranslatableComponent("commands.locatebiome.notFound", new Object[]{var0});
    });
+   private static final int MAX_SEARCH_RADIUS = 6400;
+   private static final int SEARCH_STEP = 8;
 
+   public LocateBiomeCommand() {
+   }
+
    public static void register(CommandDispatcher<CommandSourceStack> var0) {
       var0.register((LiteralArgumentBuilder)((LiteralArgumentBuilder)Commands.literal("locatebiome").requires((var0x) -> {
          return var0x.hasPermission(2);
       })).then(Commands.argument("biome", ResourceLocationArgument.id()).suggests(SuggestionProviders.AVAILABLE_BIOMES).executes((var0x) -> {
          return locateBiome((CommandSourceStack)var0x.getSource(), (ResourceLocation)var0x.getArgument("biome", ResourceLocation.class));
