--- net/minecraft/server/commands/LocateCommand.java
+++ net/minecraft/server/commands/LocateCommand.java
@@ -19,10 +19,13 @@
 import net.minecraft.world.level.levelgen.feature.StructureFeature;
 
 public class LocateCommand {
    private static final SimpleCommandExceptionType ERROR_FAILED = new SimpleCommandExceptionType(new TranslatableComponent("commands.locate.failed"));
 
+   public LocateCommand() {
+   }
+
    public static void register(CommandDispatcher<CommandSourceStack> var0) {
       LiteralArgumentBuilder var1 = (LiteralArgumentBuilder)Commands.literal("locate").requires((var0x) -> {
          return var0x.hasPermission(2);
       });
 
