--- net/minecraft/server/commands/LootCommand.java
+++ net/minecraft/server/commands/LootCommand.java
@@ -176,11 +176,11 @@
          throw ReplaceItemCommand.ERROR_INAPPLICABLE_SLOT.create(var2);
       }
    }
 
    private static boolean canMergeItems(ItemStack var0, ItemStack var1) {
-      return var0.getItem() == var1.getItem() && var0.getDamageValue() == var1.getDamageValue() && var0.getCount() <= var0.getMaxStackSize() && Objects.equals(var0.getTag(), var1.getTag());
+      return var0.is(var1.getItem()) && var0.getDamageValue() == var1.getDamageValue() && var0.getCount() <= var0.getMaxStackSize() && Objects.equals(var0.getTag(), var1.getTag());
    }
 
    private static int playerGive(Collection<ServerPlayer> var0, List<ItemStack> var1, LootCommand.Callback var2) throws CommandSyntaxException {
       ArrayList var3 = Lists.newArrayListWithCapacity(var1.size());
       Iterator var4 = var1.iterator();
@@ -189,11 +189,11 @@
          ItemStack var5 = (ItemStack)var4.next();
          Iterator var6 = var0.iterator();
 
          while(var6.hasNext()) {
             ServerPlayer var7 = (ServerPlayer)var6.next();
-            if (var7.inventory.add(var5.copy())) {
+            if (var7.getInventory().add(var5.copy())) {
                var3.add(var5);
             }
          }
       }
 
