--- net/minecraft/server/commands/PardonCommand.java
+++ net/minecraft/server/commands/PardonCommand.java
@@ -16,10 +16,13 @@
 import net.minecraft.server.players.UserBanList;
 
 public class PardonCommand {
    private static final SimpleCommandExceptionType ERROR_NOT_BANNED = new SimpleCommandExceptionType(new TranslatableComponent("commands.pardon.failed"));
 
+   public PardonCommand() {
+   }
+
    public static void register(CommandDispatcher<CommandSourceStack> var0) {
       var0.register((LiteralArgumentBuilder)((LiteralArgumentBuilder)Commands.literal("pardon").requires((var0x) -> {
          return var0x.hasPermission(3);
       })).then(Commands.argument("targets", GameProfileArgument.gameProfile()).suggests((var0x, var1) -> {
          return SharedSuggestionProvider.suggest(((CommandSourceStack)var0x.getSource()).getServer().getPlayerList().getBans().getUserList(), var1);
