--- net/minecraft/server/commands/PlaySoundCommand.java
+++ net/minecraft/server/commands/PlaySoundCommand.java
@@ -23,10 +23,13 @@
 import net.minecraft.world.phys.Vec3;
 
 public class PlaySoundCommand {
    private static final SimpleCommandExceptionType ERROR_TOO_FAR = new SimpleCommandExceptionType(new TranslatableComponent("commands.playsound.failed"));
 
+   public PlaySoundCommand() {
+   }
+
    public static void register(CommandDispatcher<CommandSourceStack> var0) {
       RequiredArgumentBuilder var1 = Commands.argument("sound", ResourceLocationArgument.id()).suggests(SuggestionProviders.AVAILABLE_SOUNDS);
       SoundSource[] var2 = SoundSource.values();
       int var3 = var2.length;
 
