--- net/minecraft/server/commands/RecipeCommand.java
+++ net/minecraft/server/commands/RecipeCommand.java
@@ -19,10 +19,13 @@
 
 public class RecipeCommand {
    private static final SimpleCommandExceptionType ERROR_GIVE_FAILED = new SimpleCommandExceptionType(new TranslatableComponent("commands.recipe.give.failed"));
    private static final SimpleCommandExceptionType ERROR_TAKE_FAILED = new SimpleCommandExceptionType(new TranslatableComponent("commands.recipe.take.failed"));
 
+   public RecipeCommand() {
+   }
+
    public static void register(CommandDispatcher<CommandSourceStack> var0) {
       var0.register((LiteralArgumentBuilder)((LiteralArgumentBuilder)((LiteralArgumentBuilder)Commands.literal("recipe").requires((var0x) -> {
          return var0x.hasPermission(2);
       })).then(Commands.literal("give").then(((RequiredArgumentBuilder)Commands.argument("targets", EntityArgument.players()).then(Commands.argument("recipe", ResourceLocationArgument.id()).suggests(SuggestionProviders.ALL_RECIPES).executes((var0x) -> {
          return giveRecipes((CommandSourceStack)var0x.getSource(), EntityArgument.getPlayers(var0x, "targets"), Collections.singleton(ResourceLocationArgument.getRecipe(var0x, "recipe")));
