--- net/minecraft/server/commands/SaveOnCommand.java
+++ net/minecraft/server/commands/SaveOnCommand.java
@@ -10,10 +10,13 @@
 import net.minecraft.server.level.ServerLevel;
 
 public class SaveOnCommand {
    private static final SimpleCommandExceptionType ERROR_ALREADY_ON = new SimpleCommandExceptionType(new TranslatableComponent("commands.save.alreadyOn"));
 
+   public SaveOnCommand() {
+   }
+
    public static void register(CommandDispatcher<CommandSourceStack> var0) {
       var0.register((LiteralArgumentBuilder)((LiteralArgumentBuilder)Commands.literal("save-on").requires((var0x) -> {
          return var0x.hasPermission(4);
       })).executes((var0x) -> {
          CommandSourceStack var1 = (CommandSourceStack)var0x.getSource();
