--- net/minecraft/server/commands/SeedCommand.java
+++ net/minecraft/server/commands/SeedCommand.java
@@ -11,10 +11,13 @@
 import net.minecraft.network.chat.MutableComponent;
 import net.minecraft.network.chat.TextComponent;
 import net.minecraft.network.chat.TranslatableComponent;
 
 public class SeedCommand {
+   public SeedCommand() {
+   }
+
    public static void register(CommandDispatcher<CommandSourceStack> var0, boolean var1) {
       var0.register((LiteralArgumentBuilder)((LiteralArgumentBuilder)Commands.literal("seed").requires((var1x) -> {
          return !var1 || var1x.hasPermission(2);
       })).executes((var0x) -> {
          long var1 = ((CommandSourceStack)var0x.getSource()).getLevel().getSeed();
