--- net/minecraft/server/commands/SetBlockCommand.java
+++ net/minecraft/server/commands/SetBlockCommand.java
@@ -21,10 +21,13 @@
 import net.minecraft.world.level.levelgen.structure.BoundingBox;
 
 public class SetBlockCommand {
    private static final SimpleCommandExceptionType ERROR_FAILED = new SimpleCommandExceptionType(new TranslatableComponent("commands.setblock.failed"));
 
+   public SetBlockCommand() {
+   }
+
    public static void register(CommandDispatcher<CommandSourceStack> var0) {
       var0.register((LiteralArgumentBuilder)((LiteralArgumentBuilder)Commands.literal("setblock").requires((var0x) -> {
          return var0x.hasPermission(2);
       })).then(Commands.argument("pos", BlockPosArgument.blockPos()).then(((RequiredArgumentBuilder)((RequiredArgumentBuilder)((RequiredArgumentBuilder)Commands.argument("block", BlockStateArgument.block()).executes((var0x) -> {
          return setBlock((CommandSourceStack)var0x.getSource(), BlockPosArgument.getLoadedBlockPos(var0x, "pos"), BlockStateArgument.getBlock(var0x, "block"), SetBlockCommand.Mode.REPLACE, (Predicate)null);
@@ -62,18 +65,18 @@
             return 1;
          }
       }
    }
 
-   public interface Filter {
-      @Nullable
-      BlockInput filter(BoundingBox var1, BlockPos var2, BlockInput var3, ServerLevel var4);
-   }
-
    public static enum Mode {
       REPLACE,
       DESTROY;
 
       private Mode() {
       }
    }
+
+   public interface Filter {
+      @Nullable
+      BlockInput filter(BoundingBox var1, BlockPos var2, BlockInput var3, ServerLevel var4);
+   }
 }
