--- net/minecraft/server/commands/SetSpawnCommand.java
+++ net/minecraft/server/commands/SetSpawnCommand.java
@@ -15,21 +15,24 @@
 import net.minecraft.network.chat.TranslatableComponent;
 import net.minecraft.resources.ResourceKey;
 import net.minecraft.server.level.ServerPlayer;
 
 public class SetSpawnCommand {
+   public SetSpawnCommand() {
+   }
+
    public static void register(CommandDispatcher<CommandSourceStack> var0) {
       var0.register((LiteralArgumentBuilder)((LiteralArgumentBuilder)((LiteralArgumentBuilder)Commands.literal("spawnpoint").requires((var0x) -> {
          return var0x.hasPermission(2);
       })).executes((var0x) -> {
          return setSpawn((CommandSourceStack)var0x.getSource(), Collections.singleton(((CommandSourceStack)var0x.getSource()).getPlayerOrException()), new BlockPos(((CommandSourceStack)var0x.getSource()).getPosition()), 0.0F);
       })).then(((RequiredArgumentBuilder)Commands.argument("targets", EntityArgument.players()).executes((var0x) -> {
          return setSpawn((CommandSourceStack)var0x.getSource(), EntityArgument.getPlayers(var0x, "targets"), new BlockPos(((CommandSourceStack)var0x.getSource()).getPosition()), 0.0F);
       })).then(((RequiredArgumentBuilder)Commands.argument("pos", BlockPosArgument.blockPos()).executes((var0x) -> {
-         return setSpawn((CommandSourceStack)var0x.getSource(), EntityArgument.getPlayers(var0x, "targets"), BlockPosArgument.getOrLoadBlockPos(var0x, "pos"), 0.0F);
+         return setSpawn((CommandSourceStack)var0x.getSource(), EntityArgument.getPlayers(var0x, "targets"), BlockPosArgument.getSpawnablePos(var0x, "pos"), 0.0F);
       })).then(Commands.argument("angle", AngleArgument.angle()).executes((var0x) -> {
-         return setSpawn((CommandSourceStack)var0x.getSource(), EntityArgument.getPlayers(var0x, "targets"), BlockPosArgument.getOrLoadBlockPos(var0x, "pos"), AngleArgument.getAngle(var0x, "angle"));
+         return setSpawn((CommandSourceStack)var0x.getSource(), EntityArgument.getPlayers(var0x, "targets"), BlockPosArgument.getSpawnablePos(var0x, "pos"), AngleArgument.getAngle(var0x, "angle"));
       })))));
    }
 
    private static int setSpawn(CommandSourceStack var0, Collection<ServerPlayer> var1, BlockPos var2, float var3) {
       ResourceKey var4 = var0.getLevel().dimension();
