--- net/minecraft/server/commands/StopCommand.java
+++ net/minecraft/server/commands/StopCommand.java
@@ -5,10 +5,13 @@
 import net.minecraft.commands.CommandSourceStack;
 import net.minecraft.commands.Commands;
 import net.minecraft.network.chat.TranslatableComponent;
 
 public class StopCommand {
+   public StopCommand() {
+   }
+
    public static void register(CommandDispatcher<CommandSourceStack> var0) {
       var0.register((LiteralArgumentBuilder)((LiteralArgumentBuilder)Commands.literal("stop").requires((var0x) -> {
          return var0x.hasPermission(4);
       })).executes((var0x) -> {
          ((CommandSourceStack)var0x.getSource()).sendSuccess(new TranslatableComponent("commands.stop.stopping"), true);
