--- net/minecraft/server/commands/StopSoundCommand.java
+++ net/minecraft/server/commands/StopSoundCommand.java
@@ -16,10 +16,13 @@
 import net.minecraft.resources.ResourceLocation;
 import net.minecraft.server.level.ServerPlayer;
 import net.minecraft.sounds.SoundSource;
 
 public class StopSoundCommand {
+   public StopSoundCommand() {
+   }
+
    public static void register(CommandDispatcher<CommandSourceStack> var0) {
       RequiredArgumentBuilder var1 = (RequiredArgumentBuilder)((RequiredArgumentBuilder)Commands.argument("targets", EntityArgument.players()).executes((var0x) -> {
          return stopSound((CommandSourceStack)var0x.getSource(), EntityArgument.getPlayers(var0x, "targets"), (SoundSource)null, (ResourceLocation)null);
       })).then(Commands.literal("*").then(Commands.argument("sound", ResourceLocationArgument.id()).suggests(SuggestionProviders.AVAILABLE_SOUNDS).executes((var0x) -> {
          return stopSound((CommandSourceStack)var0x.getSource(), EntityArgument.getPlayers(var0x, "targets"), (SoundSource)null, ResourceLocationArgument.getId(var0x, "sound"));
