--- net/minecraft/server/commands/TagCommand.java
+++ net/minecraft/server/commands/TagCommand.java
@@ -20,10 +20,13 @@
 
 public class TagCommand {
    private static final SimpleCommandExceptionType ERROR_ADD_FAILED = new SimpleCommandExceptionType(new TranslatableComponent("commands.tag.add.failed"));
    private static final SimpleCommandExceptionType ERROR_REMOVE_FAILED = new SimpleCommandExceptionType(new TranslatableComponent("commands.tag.remove.failed"));
 
+   public TagCommand() {
+   }
+
    public static void register(CommandDispatcher<CommandSourceStack> var0) {
       var0.register((LiteralArgumentBuilder)((LiteralArgumentBuilder)Commands.literal("tag").requires((var0x) -> {
          return var0x.hasPermission(2);
       })).then(((RequiredArgumentBuilder)((RequiredArgumentBuilder)Commands.argument("targets", EntityArgument.entities()).then(Commands.literal("add").then(Commands.argument("name", StringArgumentType.word()).executes((var0x) -> {
          return addTag((CommandSourceStack)var0x.getSource(), EntityArgument.getEntities(var0x, "targets"), StringArgumentType.getString(var0x, "name"));
