--- net/minecraft/server/commands/TimeCommand.java
+++ net/minecraft/server/commands/TimeCommand.java
@@ -9,10 +9,13 @@
 import net.minecraft.commands.arguments.TimeArgument;
 import net.minecraft.network.chat.TranslatableComponent;
 import net.minecraft.server.level.ServerLevel;
 
 public class TimeCommand {
+   public TimeCommand() {
+   }
+
    public static void register(CommandDispatcher<CommandSourceStack> var0) {
       var0.register((LiteralArgumentBuilder)((LiteralArgumentBuilder)((LiteralArgumentBuilder)((LiteralArgumentBuilder)Commands.literal("time").requires((var0x) -> {
          return var0x.hasPermission(2);
       })).then(((LiteralArgumentBuilder)((LiteralArgumentBuilder)((LiteralArgumentBuilder)((LiteralArgumentBuilder)Commands.literal("set").then(Commands.literal("day").executes((var0x) -> {
          return setTime((CommandSourceStack)var0x.getSource(), 1000);
