--- net/minecraft/server/commands/TriggerCommand.java
+++ net/minecraft/server/commands/TriggerCommand.java
@@ -27,10 +27,13 @@
 
 public class TriggerCommand {
    private static final SimpleCommandExceptionType ERROR_NOT_PRIMED = new SimpleCommandExceptionType(new TranslatableComponent("commands.trigger.failed.unprimed"));
    private static final SimpleCommandExceptionType ERROR_INVALID_OBJECTIVE = new SimpleCommandExceptionType(new TranslatableComponent("commands.trigger.failed.invalid"));
 
+   public TriggerCommand() {
+   }
+
    public static void register(CommandDispatcher<CommandSourceStack> var0) {
       var0.register((LiteralArgumentBuilder)Commands.literal("trigger").then(((RequiredArgumentBuilder)((RequiredArgumentBuilder)Commands.argument("objective", ObjectiveArgument.objective()).suggests((var0x, var1) -> {
          return suggestObjectives((CommandSourceStack)var0x.getSource(), var1);
       }).executes((var0x) -> {
          return simpleTrigger((CommandSourceStack)var0x.getSource(), getScore(((CommandSourceStack)var0x.getSource()).getPlayerOrException(), ObjectiveArgument.getObjective(var0x, "objective")));
