--- net/minecraft/server/commands/WeatherCommand.java
+++ net/minecraft/server/commands/WeatherCommand.java
@@ -6,10 +6,15 @@
 import net.minecraft.commands.CommandSourceStack;
 import net.minecraft.commands.Commands;
 import net.minecraft.network.chat.TranslatableComponent;
 
 public class WeatherCommand {
+   private static final int DEFAULT_TIME = 6000;
+
+   public WeatherCommand() {
+   }
+
    public static void register(CommandDispatcher<CommandSourceStack> var0) {
       var0.register((LiteralArgumentBuilder)((LiteralArgumentBuilder)((LiteralArgumentBuilder)((LiteralArgumentBuilder)Commands.literal("weather").requires((var0x) -> {
          return var0x.hasPermission(2);
       })).then(((LiteralArgumentBuilder)Commands.literal("clear").executes((var0x) -> {
          return setClear((CommandSourceStack)var0x.getSource(), 6000);
