--- net/minecraft/commands/synchronization/brigadier/BrigadierArgumentSerializers.java
+++ net/minecraft/commands/synchronization/brigadier/BrigadierArgumentSerializers.java
@@ -8,10 +8,16 @@
 import com.mojang.brigadier.arguments.StringArgumentType;
 import net.minecraft.commands.synchronization.ArgumentTypes;
 import net.minecraft.commands.synchronization.EmptyArgumentSerializer;
 
 public class BrigadierArgumentSerializers {
+   private static final byte NUMBER_FLAG_MIN = 1;
+   private static final byte NUMBER_FLAG_MAX = 2;
+
+   public BrigadierArgumentSerializers() {
+   }
+
    public static void bootstrap() {
       ArgumentTypes.register("brigadier:bool", BoolArgumentType.class, new EmptyArgumentSerializer(BoolArgumentType::bool));
       ArgumentTypes.register("brigadier:float", FloatArgumentType.class, new FloatArgumentSerializer());
       ArgumentTypes.register("brigadier:double", DoubleArgumentType.class, new DoubleArgumentSerializer());
       ArgumentTypes.register("brigadier:integer", IntegerArgumentType.class, new IntegerArgumentSerializer());
