--- net/minecraft/world/level/block/AnvilBlock.java
+++ net/minecraft/world/level/block/AnvilBlock.java
@@ -8,10 +8,11 @@
 import net.minecraft.stats.Stats;
 import net.minecraft.world.InteractionHand;
 import net.minecraft.world.InteractionResult;
 import net.minecraft.world.MenuProvider;
 import net.minecraft.world.SimpleMenuProvider;
+import net.minecraft.world.damagesource.DamageSource;
 import net.minecraft.world.entity.item.FallingBlockEntity;
 import net.minecraft.world.entity.player.Player;
 import net.minecraft.world.inventory.AnvilMenu;
 import net.minecraft.world.inventory.ContainerLevelAccess;
 import net.minecraft.world.item.context.BlockPlaceContext;
@@ -70,27 +71,31 @@
       Direction var5 = (Direction)var1.getValue(FACING);
       return var5.getAxis() == Direction.Axis.X ? X_AXIS_AABB : Z_AXIS_AABB;
    }
 
    protected void falling(FallingBlockEntity var1) {
-      var1.setHurtsEntities(true);
+      var1.setHurtsEntities(2.0F, 40);
    }
 
    public void onLand(Level var1, BlockPos var2, BlockState var3, BlockState var4, FallingBlockEntity var5) {
       if (!var5.isSilent()) {
          var1.levelEvent(1031, var2, 0);
       }
 
    }
 
-   public void onBroken(Level var1, BlockPos var2, FallingBlockEntity var3) {
+   public void onBrokenAfterFall(Level var1, BlockPos var2, FallingBlockEntity var3) {
       if (!var3.isSilent()) {
          var1.levelEvent(1029, var2, 0);
       }
 
    }
 
+   public DamageSource getFallDamageSource() {
+      return DamageSource.ANVIL;
+   }
+
    @Nullable
    public static BlockState damage(BlockState var0) {
       if (var0.is(Blocks.ANVIL)) {
          return (BlockState)Blocks.CHIPPED_ANVIL.defaultBlockState().setValue(FACING, var0.getValue(FACING));
       } else {
