--- net/minecraft/world/level/block/BambooBlock.java
+++ net/minecraft/world/level/block/BambooBlock.java
@@ -79,11 +79,11 @@
             } else if (var3.is(Blocks.BAMBOO)) {
                int var5 = (Integer)var3.getValue(AGE) > 0 ? 1 : 0;
                return (BlockState)this.defaultBlockState().setValue(AGE, var5);
             } else {
                BlockState var4 = var1.getLevel().getBlockState(var1.getClickedPos().above());
-               return !var4.is(Blocks.BAMBOO) && !var4.is(Blocks.BAMBOO_SAPLING) ? Blocks.BAMBOO_SAPLING.defaultBlockState() : (BlockState)this.defaultBlockState().setValue(AGE, var4.getValue(AGE));
+               return var4.is(Blocks.BAMBOO) ? (BlockState)this.defaultBlockState().setValue(AGE, var4.getValue(AGE)) : Blocks.BAMBOO_SAPLING.defaultBlockState();
             }
          } else {
             return null;
          }
       }
