--- net/minecraft/world/level/block/BambooSaplingBlock.java
+++ net/minecraft/world/level/block/BambooSaplingBlock.java
@@ -19,10 +19,11 @@
 import net.minecraft.world.phys.Vec3;
 import net.minecraft.world.phys.shapes.CollisionContext;
 import net.minecraft.world.phys.shapes.VoxelShape;
 
 public class BambooSaplingBlock extends Block implements BonemealableBlock {
+   protected static final float SAPLING_AABB_OFFSET = 4.0F;
    protected static final VoxelShape SAPLING_SHAPE = Block.box(4.0D, 0.0D, 4.0D, 12.0D, 12.0D, 12.0D);
 
    public BambooSaplingBlock(BlockBehaviour.Properties var1) {
       super(var1);
    }
