--- net/minecraft/world/level/block/BaseCoralPlantBlock.java
+++ net/minecraft/world/level/block/BaseCoralPlantBlock.java
@@ -6,10 +6,11 @@
 import net.minecraft.world.level.block.state.BlockState;
 import net.minecraft.world.phys.shapes.CollisionContext;
 import net.minecraft.world.phys.shapes.VoxelShape;
 
 public class BaseCoralPlantBlock extends BaseCoralPlantTypeBlock {
+   protected static final float AABB_OFFSET = 6.0F;
    protected static final VoxelShape SHAPE = Block.box(2.0D, 0.0D, 2.0D, 14.0D, 15.0D, 14.0D);
 
    protected BaseCoralPlantBlock(BlockBehaviour.Properties var1) {
       super(var1);
    }
