--- net/minecraft/world/level/block/BedBlock.java
+++ net/minecraft/world/level/block/BedBlock.java
@@ -45,11 +45,13 @@
 import org.apache.commons.lang3.ArrayUtils;
 
 public class BedBlock extends HorizontalDirectionalBlock implements EntityBlock {
    public static final EnumProperty<BedPart> PART;
    public static final BooleanProperty OCCUPIED;
+   protected static final int HEIGHT = 9;
    protected static final VoxelShape BASE;
+   private static final int LEG_WIDTH = 3;
    protected static final VoxelShape LEG_NORTH_WEST;
    protected static final VoxelShape LEG_SOUTH_WEST;
    protected static final VoxelShape LEG_NORTH_EAST;
    protected static final VoxelShape LEG_SOUTH_EAST;
    protected static final VoxelShape NORTH_SHAPE;
@@ -161,11 +163,11 @@
       if (!var1.isClientSide && var4.isCreative()) {
          BedPart var5 = (BedPart)var3.getValue(PART);
          if (var5 == BedPart.FOOT) {
             BlockPos var6 = var2.relative(getNeighbourDirection(var5, (Direction)var3.getValue(FACING)));
             BlockState var7 = var1.getBlockState(var6);
-            if (var7.getBlock() == this && var7.getValue(PART) == BedPart.HEAD) {
+            if (var7.is(this) && var7.getValue(PART) == BedPart.HEAD) {
                var1.setBlock(var6, Blocks.AIR.defaultBlockState(), 35);
                var1.levelEvent(var4, 2001, var6, Block.getId(var7));
             }
          }
       }
@@ -277,12 +279,12 @@
 
    protected void createBlockStateDefinition(StateDefinition.Builder<Block, BlockState> var1) {
       var1.add(FACING, PART, OCCUPIED);
    }
 
-   public BlockEntity newBlockEntity(BlockGetter var1) {
-      return new BedBlockEntity(this.color);
+   public BlockEntity newBlockEntity(BlockPos var1, BlockState var2) {
+      return new BedBlockEntity(var1, var2, this.color);
    }
 
    public void setPlacedBy(Level var1, BlockPos var2, BlockState var3, @Nullable LivingEntity var4, ItemStack var5) {
       super.setPlacedBy(var1, var2, var3, var4, var5);
       if (!var1.isClientSide) {
