--- net/minecraft/world/level/block/BeetrootBlock.java
+++ net/minecraft/world/level/block/BeetrootBlock.java
@@ -14,10 +14,11 @@
 import net.minecraft.world.level.block.state.properties.IntegerProperty;
 import net.minecraft.world.phys.shapes.CollisionContext;
 import net.minecraft.world.phys.shapes.VoxelShape;
 
 public class BeetrootBlock extends CropBlock {
+   public static final int MAX_AGE = 3;
    public static final IntegerProperty AGE;
    private static final VoxelShape[] SHAPE_BY_AGE;
 
    public BeetrootBlock(BlockBehaviour.Properties var1) {
       super(var1);
