--- net/minecraft/world/level/block/ButtonBlock.java
+++ net/minecraft/world/level/block/ButtonBlock.java
@@ -20,16 +20,21 @@
 import net.minecraft.world.level.block.state.BlockState;
 import net.minecraft.world.level.block.state.StateDefinition;
 import net.minecraft.world.level.block.state.properties.AttachFace;
 import net.minecraft.world.level.block.state.properties.BlockStateProperties;
 import net.minecraft.world.level.block.state.properties.BooleanProperty;
+import net.minecraft.world.level.gameevent.GameEvent;
 import net.minecraft.world.phys.BlockHitResult;
 import net.minecraft.world.phys.shapes.CollisionContext;
 import net.minecraft.world.phys.shapes.VoxelShape;
 
 public abstract class ButtonBlock extends FaceAttachedHorizontalDirectionalBlock {
    public static final BooleanProperty POWERED;
+   private static final int PRESSED_DEPTH = 1;
+   private static final int UNPRESSED_DEPTH = 2;
+   protected static final int HALF_AABB_HEIGHT = 2;
+   protected static final int HALF_AABB_WIDTH = 3;
    protected static final VoxelShape CEILING_AABB_X;
    protected static final VoxelShape CEILING_AABB_Z;
    protected static final VoxelShape FLOOR_AABB_X;
    protected static final VoxelShape FLOOR_AABB_Z;
    protected static final VoxelShape NORTH_AABB;
@@ -92,10 +97,11 @@
       if ((Boolean)var1.getValue(POWERED)) {
          return InteractionResult.CONSUME;
       } else {
          this.press(var1, var2, var3);
          this.playSound(var4, var2, var3, true);
+         var2.gameEvent(var4, GameEvent.BLOCK_PRESS, var3);
          return InteractionResult.sidedSuccess(var2.isClientSide);
       }
    }
 
    public void press(BlockState var1, Level var2, BlockPos var3) {
@@ -138,10 +144,11 @@
             this.checkPressed(var1, var2, var3);
          } else {
             var2.setBlock(var3, (BlockState)var1.setValue(POWERED, false), 3);
             this.updateNeighbours(var1, var2, var3);
             this.playSound((Player)null, var2, var3, false);
+            var2.gameEvent(GameEvent.BLOCK_UNPRESS, var3);
          }
 
       }
    }
 
@@ -157,10 +164,11 @@
       boolean var6 = (Boolean)var1.getValue(POWERED);
       if (var5 != var6) {
          var2.setBlock(var3, (BlockState)var1.setValue(POWERED, var5), 3);
          this.updateNeighbours(var1, var2, var3);
          this.playSound((Player)null, var2, var3, var5);
+         var2.gameEvent((Entity)var4.stream().findFirst().orElse((Object)null), var5 ? GameEvent.BLOCK_PRESS : GameEvent.BLOCK_UNPRESS, var3);
       }
 
       if (var5) {
          var2.getBlockTicks().scheduleTick(new BlockPos(var3), this, this.getPressDuration());
       }
