--- net/minecraft/world/level/block/ChainBlock.java
+++ net/minecraft/world/level/block/ChainBlock.java
@@ -17,10 +17,12 @@
 import net.minecraft.world.phys.shapes.CollisionContext;
 import net.minecraft.world.phys.shapes.VoxelShape;
 
 public class ChainBlock extends RotatedPillarBlock implements SimpleWaterloggedBlock {
    public static final BooleanProperty WATERLOGGED;
+   protected static final float AABB_MIN = 6.5F;
+   protected static final float AABB_MAX = 9.5F;
    protected static final VoxelShape Y_AXIS_AABB;
    protected static final VoxelShape Z_AXIS_AABB;
    protected static final VoxelShape X_AXIS_AABB;
 
    public ChainBlock(BlockBehaviour.Properties var1) {
