--- net/minecraft/world/level/block/ComposterBlock.java
+++ net/minecraft/world/level/block/ComposterBlock.java
@@ -35,12 +35,16 @@
 import net.minecraft.world.phys.shapes.CollisionContext;
 import net.minecraft.world.phys.shapes.Shapes;
 import net.minecraft.world.phys.shapes.VoxelShape;
 
 public class ComposterBlock extends Block implements WorldlyContainerHolder {
+   public static final int READY = 8;
+   public static final int MIN_LEVEL = 0;
+   public static final int MAX_LEVEL = 7;
    public static final IntegerProperty LEVEL;
    public static final Object2FloatMap<ItemLike> COMPOSTABLES;
+   private static final int AABB_SIDE_THICKNESS = 2;
    private static final VoxelShape OUTER_SHAPE;
    private static final VoxelShape[] SHAPES;
 
    public static void bootStrap() {
       COMPOSTABLES.defaultReturnValue(-1.0F);
@@ -53,10 +57,11 @@
       add(0.3F, Items.OAK_LEAVES);
       add(0.3F, Items.SPRUCE_LEAVES);
       add(0.3F, Items.DARK_OAK_LEAVES);
       add(0.3F, Items.ACACIA_LEAVES);
       add(0.3F, Items.BIRCH_LEAVES);
+      add(0.3F, Items.AZALEA_LEAVES);
       add(0.3F, Items.OAK_SAPLING);
       add(0.3F, Items.SPRUCE_SAPLING);
       add(0.3F, Items.BIRCH_SAPLING);
       add(0.3F, Items.JUNGLE_SAPLING);
       add(0.3F, Items.ACACIA_SAPLING);
@@ -67,20 +72,26 @@
       add(0.3F, Items.KELP);
       add(0.3F, Items.MELON_SEEDS);
       add(0.3F, Items.PUMPKIN_SEEDS);
       add(0.3F, Items.SEAGRASS);
       add(0.3F, Items.SWEET_BERRIES);
+      add(0.3F, Items.GLOW_BERRIES);
       add(0.3F, Items.WHEAT_SEEDS);
+      add(0.3F, Items.MOSS_CARPET);
+      add(0.3F, Items.SMALL_DRIPLEAF);
+      add(0.3F, Items.HANGING_ROOTS);
       add(0.5F, Items.DRIED_KELP_BLOCK);
       add(0.5F, Items.TALL_GRASS);
+      add(0.5F, Items.AZALEA_LEAVES_FLOWERS);
       add(0.5F, Items.CACTUS);
       add(0.5F, Items.SUGAR_CANE);
       add(0.5F, Items.VINE);
       add(0.5F, Items.NETHER_SPROUTS);
       add(0.5F, Items.WEEPING_VINES);
       add(0.5F, Items.TWISTING_VINES);
       add(0.5F, Items.MELON_SLICE);
+      add(0.5F, Items.GLOW_LICHEN);
       add(0.65F, Items.SEA_PICKLE);
       add(0.65F, Items.LILY_PAD);
       add(0.65F, Items.PUMPKIN);
       add(0.65F, Items.CARVED_PUMPKIN);
       add(0.65F, Items.MELON);
@@ -116,15 +127,20 @@
       add(0.65F, Items.SUNFLOWER);
       add(0.65F, Items.LILAC);
       add(0.65F, Items.ROSE_BUSH);
       add(0.65F, Items.PEONY);
       add(0.65F, Items.LARGE_FERN);
+      add(0.65F, Items.SPORE_BLOSSOM);
+      add(0.65F, Items.AZALEA);
+      add(0.65F, Items.MOSS_BLOCK);
+      add(0.65F, Items.BIG_DRIPLEAF);
       add(0.85F, Items.HAY_BLOCK);
       add(0.85F, Items.BROWN_MUSHROOM_BLOCK);
       add(0.85F, Items.RED_MUSHROOM_BLOCK);
       add(0.85F, Items.NETHER_WART_BLOCK);
       add(0.85F, Items.WARPED_WART_BLOCK);
+      add(0.85F, Items.FLOWERING_AZALEA);
       add(0.85F, Items.BREAD);
       add(0.85F, Items.BAKED_POTATO);
       add(0.85F, Items.COOKIE);
       add(1.0F, Items.CAKE);
       add(1.0F, Items.PUMPKIN_PIE);
@@ -180,11 +196,11 @@
       ItemStack var8 = var4.getItemInHand(var5);
       if (var7 < 8 && COMPOSTABLES.containsKey(var8.getItem())) {
          if (var7 < 7 && !var2.isClientSide) {
             BlockState var9 = addItem(var1, var2, var3, var8);
             var2.levelEvent(1500, var3, var1 != var9 ? 1 : 0);
-            if (!var4.abilities.instabuild) {
+            if (!var4.getAbilities().instabuild) {
                var8.shrink(1);
             }
          }
 
          return InteractionResult.sidedSuccess(var2.isClientSide);
@@ -357,11 +373,11 @@
       public boolean canPlaceItemThroughFace(int var1, ItemStack var2, @Nullable Direction var3) {
          return false;
       }
 
       public boolean canTakeItemThroughFace(int var1, ItemStack var2, Direction var3) {
-         return !this.changed && var3 == Direction.DOWN && var2.getItem() == Items.BONE_MEAL;
+         return !this.changed && var3 == Direction.DOWN && var2.is(Items.BONE_MEAL);
       }
 
       public void setChanged() {
          ComposterBlock.empty(this.state, this.level, this.pos);
          this.changed = true;
