--- net/minecraft/world/level/block/ComposterBlock.java
+++ net/minecraft/world/level/block/ComposterBlock.java
@@ -180,11 +180,11 @@
       ItemStack var8 = var4.getItemInHand(var5);
       if (var7 < 8 && COMPOSTABLES.containsKey(var8.getItem())) {
          if (var7 < 7 && !var2.isClientSide) {
             BlockState var9 = addItem(var1, var2, var3, var8);
             var2.levelEvent(1500, var3, var1 != var9 ? 1 : 0);
-            if (!var4.abilities.instabuild) {
+            if (!var4.getAbilities().instabuild) {
                var8.shrink(1);
             }
          }
 
          return InteractionResult.sidedSuccess(var2.isClientSide);
@@ -357,11 +357,11 @@
       public boolean canPlaceItemThroughFace(int var1, ItemStack var2, @Nullable Direction var3) {
          return false;
       }
 
       public boolean canTakeItemThroughFace(int var1, ItemStack var2, Direction var3) {
-         return !this.changed && var3 == Direction.DOWN && var2.getItem() == Items.BONE_MEAL;
+         return !this.changed && var3 == Direction.DOWN && var2.is(Items.BONE_MEAL);
       }
 
       public void setChanged() {
          ComposterBlock.empty(this.state, this.level, this.pos);
          this.changed = true;
