--- net/minecraft/world/level/block/CoralPlantBlock.java
+++ net/minecraft/world/level/block/CoralPlantBlock.java
@@ -13,10 +13,11 @@
 import net.minecraft.world.phys.shapes.CollisionContext;
 import net.minecraft.world.phys.shapes.VoxelShape;
 
 public class CoralPlantBlock extends BaseCoralPlantTypeBlock {
    private final Block deadBlock;
+   protected static final float AABB_OFFSET = 6.0F;
    protected static final VoxelShape SHAPE = Block.box(2.0D, 0.0D, 2.0D, 14.0D, 15.0D, 14.0D);
 
    protected CoralPlantBlock(Block var1, BlockBehaviour.Properties var2) {
       super(var2);
       this.deadBlock = var1;
