--- net/minecraft/world/level/block/DoublePlantBlock.java
+++ net/minecraft/world/level/block/DoublePlantBlock.java
@@ -37,11 +37,12 @@
    }
 
    @Nullable
    public BlockState getStateForPlacement(BlockPlaceContext var1) {
       BlockPos var2 = var1.getClickedPos();
-      return var2.getY() < 255 && var1.getLevel().getBlockState(var2.above()).canBeReplaced(var1) ? super.getStateForPlacement(var1) : null;
+      Level var3 = var1.getLevel();
+      return var2.getY() < var3.getMaxBuildHeight() - 1 && var3.getBlockState(var2.above()).canBeReplaced(var1) ? super.getStateForPlacement(var1) : null;
    }
 
    public void setPlacedBy(Level var1, BlockPos var2, BlockState var3, LivingEntity var4, ItemStack var5) {
       var1.setBlock(var2.above(), (BlockState)this.defaultBlockState().setValue(HALF, DoubleBlockHalf.UPPER), 3);
    }
@@ -53,13 +54,13 @@
          BlockState var4 = var2.getBlockState(var3.below());
          return var4.is(this) && var4.getValue(HALF) == DoubleBlockHalf.LOWER;
       }
    }
 
-   public void placeAt(LevelAccessor var1, BlockPos var2, int var3) {
-      var1.setBlock(var2, (BlockState)this.defaultBlockState().setValue(HALF, DoubleBlockHalf.LOWER), var3);
-      var1.setBlock(var2.above(), (BlockState)this.defaultBlockState().setValue(HALF, DoubleBlockHalf.UPPER), var3);
+   public void placeAt(LevelAccessor var1, BlockState var2, BlockPos var3, int var4) {
+      var1.setBlock(var3, (BlockState)var2.setValue(HALF, DoubleBlockHalf.LOWER), var4);
+      var1.setBlock(var3.above(), (BlockState)var2.setValue(HALF, DoubleBlockHalf.UPPER), var4);
    }
 
    public void playerWillDestroy(Level var1, BlockPos var2, BlockState var3, Player var4) {
       if (!var1.isClientSide) {
          if (var4.isCreative()) {
@@ -79,11 +80,11 @@
    protected static void preventCreativeDropFromBottomPart(Level var0, BlockPos var1, BlockState var2, Player var3) {
       DoubleBlockHalf var4 = (DoubleBlockHalf)var2.getValue(HALF);
       if (var4 == DoubleBlockHalf.UPPER) {
          BlockPos var5 = var1.below();
          BlockState var6 = var0.getBlockState(var5);
-         if (var6.getBlock() == var2.getBlock() && var6.getValue(HALF) == DoubleBlockHalf.LOWER) {
+         if (var6.is(var2.getBlock()) && var6.getValue(HALF) == DoubleBlockHalf.LOWER) {
             var0.setBlock(var5, Blocks.AIR.defaultBlockState(), 35);
             var0.levelEvent(var3, 2001, var5, Block.getId(var6));
          }
       }
 
