--- net/minecraft/world/level/block/FarmBlock.java
+++ net/minecraft/world/level/block/FarmBlock.java
@@ -26,10 +26,11 @@
 import net.minecraft.world.phys.shapes.VoxelShape;
 
 public class FarmBlock extends Block {
    public static final IntegerProperty MOISTURE;
    protected static final VoxelShape SHAPE;
+   public static final int MAX_MOISTURE = 7;
 
    protected FarmBlock(BlockBehaviour.Properties var1) {
       super(var1);
       this.registerDefaultState((BlockState)((BlockState)this.stateDefinition.any()).setValue(MOISTURE, 0));
    }
