--- net/minecraft/world/level/block/FireBlock.java
+++ net/minecraft/world/level/block/FireBlock.java
@@ -26,10 +26,11 @@
 import net.minecraft.world.phys.shapes.CollisionContext;
 import net.minecraft.world.phys.shapes.Shapes;
 import net.minecraft.world.phys.shapes.VoxelShape;
 
 public class FireBlock extends BaseFireBlock {
+   public static final int MAX_AGE = 15;
    public static final IntegerProperty AGE;
    public static final BooleanProperty NORTH;
    public static final BooleanProperty EAST;
    public static final BooleanProperty SOUTH;
    public static final BooleanProperty WEST;
@@ -39,10 +40,18 @@
    private static final VoxelShape WEST_AABB;
    private static final VoxelShape EAST_AABB;
    private static final VoxelShape NORTH_AABB;
    private static final VoxelShape SOUTH_AABB;
    private final Map<BlockState, VoxelShape> shapesCache;
+   private static final int FLAME_INSTANT = 60;
+   private static final int FLAME_EASY = 30;
+   private static final int FLAME_MEDIUM = 15;
+   private static final int FLAME_HARD = 5;
+   private static final int BURN_INSTANT = 100;
+   private static final int BURN_EASY = 60;
+   private static final int BURN_MEDIUM = 20;
+   private static final int BURN_HARD = 5;
    private final Object2IntMap<Block> flameOdds = new Object2IntOpenHashMap();
    private final Object2IntMap<Block> burnOdds = new Object2IntOpenHashMap();
 
    public FireBlock(BlockBehaviour.Properties var1) {
       super(var1, 1.0F);
@@ -411,10 +420,22 @@
       var0.setFlammable(Blocks.LECTERN, 30, 20);
       var0.setFlammable(Blocks.COMPOSTER, 5, 20);
       var0.setFlammable(Blocks.SWEET_BERRY_BUSH, 60, 100);
       var0.setFlammable(Blocks.BEEHIVE, 5, 20);
       var0.setFlammable(Blocks.BEE_NEST, 30, 20);
+      var0.setFlammable(Blocks.AZALEA_LEAVES, 30, 60);
+      var0.setFlammable(Blocks.AZALEA_LEAVES_FLOWERS, 30, 60);
+      var0.setFlammable(Blocks.CAVE_VINES, 15, 60);
+      var0.setFlammable(Blocks.CAVE_VINES_PLANT, 15, 60);
+      var0.setFlammable(Blocks.SPORE_BLOSSOM, 60, 100);
+      var0.setFlammable(Blocks.AZALEA, 30, 60);
+      var0.setFlammable(Blocks.FLOWERING_AZALEA, 30, 60);
+      var0.setFlammable(Blocks.BIG_DRIPLEAF, 60, 100);
+      var0.setFlammable(Blocks.BIG_DRIPLEAF_STEM, 60, 100);
+      var0.setFlammable(Blocks.SMALL_DRIPLEAF, 60, 100);
+      var0.setFlammable(Blocks.HANGING_ROOTS, 30, 60);
+      var0.setFlammable(Blocks.GLOW_LICHEN, 15, 100);
    }
 
    static {
       AGE = BlockStateProperties.AGE_15;
       NORTH = PipeBlock.NORTH;
