--- net/minecraft/world/level/block/FrostedIceBlock.java
+++ net/minecraft/world/level/block/FrostedIceBlock.java
@@ -13,11 +13,14 @@
 import net.minecraft.world.level.block.state.StateDefinition;
 import net.minecraft.world.level.block.state.properties.BlockStateProperties;
 import net.minecraft.world.level.block.state.properties.IntegerProperty;
 
 public class FrostedIceBlock extends IceBlock {
+   public static final int MAX_AGE = 3;
    public static final IntegerProperty AGE;
+   private static final int NEIGHBORS_TO_AGE = 4;
+   private static final int NEIGHBORS_TO_MELT = 2;
 
    public FrostedIceBlock(BlockBehaviour.Properties var1) {
       super(var1);
       this.registerDefaultState((BlockState)((BlockState)this.stateDefinition.any()).setValue(AGE, 0));
    }
@@ -32,11 +35,11 @@
          Direction[] var6 = Direction.values();
          int var7 = var6.length;
 
          for(int var8 = 0; var8 < var7; ++var8) {
             Direction var9 = var6[var8];
-            var5.setWithOffset(var3, var9);
+            var5.setWithOffset(var3, (Direction)var9);
             BlockState var10 = var2.getBlockState(var5);
             if (var10.is(this) && !this.slightlyMelt(var10, var2, var5)) {
                var2.getBlockTicks().scheduleTick(var5, this, Mth.nextInt(var4, 20, 40));
             }
          }
@@ -56,11 +59,11 @@
          return true;
       }
    }
 
    public void neighborChanged(BlockState var1, Level var2, BlockPos var3, Block var4, BlockPos var5, boolean var6) {
-      if (var4 == this && this.fewerNeigboursThan(var2, var3, 2)) {
+      if (var4.defaultBlockState().is(this) && this.fewerNeigboursThan(var2, var3, 2)) {
          this.melt(var1, var2, var3);
       }
 
       super.neighborChanged(var1, var2, var3, var4, var5, var6);
    }
@@ -71,11 +74,11 @@
       Direction[] var6 = Direction.values();
       int var7 = var6.length;
 
       for(int var8 = 0; var8 < var7; ++var8) {
          Direction var9 = var6[var8];
-         var5.setWithOffset(var2, var9);
+         var5.setWithOffset(var2, (Direction)var9);
          if (var1.getBlockState(var5).is(this)) {
             ++var4;
             if (var4 >= var3) {
                return false;
             }
