--- net/minecraft/world/level/block/GrowingPlantBlock.java
+++ net/minecraft/world/level/block/GrowingPlantBlock.java
@@ -37,26 +37,25 @@
    }
 
    public boolean canSurvive(BlockState var1, LevelReader var2, BlockPos var3) {
       BlockPos var4 = var3.relative(this.growthDirection.getOpposite());
       BlockState var5 = var2.getBlockState(var4);
-      Block var6 = var5.getBlock();
-      if (!this.canAttachToBlock(var6)) {
+      if (!this.canAttachTo(var5)) {
          return false;
       } else {
-         return var6 == this.getHeadBlock() || var6 == this.getBodyBlock() || var5.isFaceSturdy(var2, var4, this.growthDirection);
+         return var5.is(this.getHeadBlock()) || var5.is(this.getBodyBlock()) || var5.isFaceSturdy(var2, var4, this.growthDirection);
       }
    }
 
    public void tick(BlockState var1, ServerLevel var2, BlockPos var3, Random var4) {
       if (!var1.canSurvive(var2, var3)) {
          var2.destroyBlock(var3, true);
       }
 
    }
 
-   protected boolean canAttachToBlock(Block var1) {
+   protected boolean canAttachTo(BlockState var1) {
       return true;
    }
 
    public VoxelShape getShape(BlockState var1, BlockGetter var2, BlockPos var3, CollisionContext var4) {
       return this.shape;
