--- net/minecraft/world/level/block/HayBlock.java
+++ net/minecraft/world/level/block/HayBlock.java
@@ -1,9 +1,10 @@
 package net.minecraft.world.level.block;
 
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.Direction;
+import net.minecraft.world.damagesource.DamageSource;
 import net.minecraft.world.entity.Entity;
 import net.minecraft.world.level.Level;
 import net.minecraft.world.level.block.state.BlockBehaviour;
 import net.minecraft.world.level.block.state.BlockState;
 
@@ -11,9 +12,9 @@
    public HayBlock(BlockBehaviour.Properties var1) {
       super(var1);
       this.registerDefaultState((BlockState)((BlockState)this.stateDefinition.any()).setValue(AXIS, Direction.Axis.Y));
    }
 
-   public void fallOn(Level var1, BlockPos var2, Entity var3, float var4) {
-      var3.causeFallDamage(var4, 0.2F);
+   public void fallOn(Level var1, BlockState var2, BlockPos var3, Entity var4, float var5) {
+      var4.causeFallDamage(var5, 0.2F, DamageSource.FALL);
    }
 }
