--- net/minecraft/world/level/block/HugeMushroomBlock.java
+++ net/minecraft/world/level/block/HugeMushroomBlock.java
@@ -27,11 +27,11 @@
    }
 
    public BlockState getStateForPlacement(BlockPlaceContext var1) {
       Level var2 = var1.getLevel();
       BlockPos var3 = var1.getClickedPos();
-      return (BlockState)((BlockState)((BlockState)((BlockState)((BlockState)((BlockState)this.defaultBlockState().setValue(DOWN, this != var2.getBlockState(var3.below()).getBlock())).setValue(UP, this != var2.getBlockState(var3.above()).getBlock())).setValue(NORTH, this != var2.getBlockState(var3.north()).getBlock())).setValue(EAST, this != var2.getBlockState(var3.east()).getBlock())).setValue(SOUTH, this != var2.getBlockState(var3.south()).getBlock())).setValue(WEST, this != var2.getBlockState(var3.west()).getBlock());
+      return (BlockState)((BlockState)((BlockState)((BlockState)((BlockState)((BlockState)this.defaultBlockState().setValue(DOWN, !var2.getBlockState(var3.below()).is(this))).setValue(UP, !var2.getBlockState(var3.above()).is(this))).setValue(NORTH, !var2.getBlockState(var3.north()).is(this))).setValue(EAST, !var2.getBlockState(var3.east()).is(this))).setValue(SOUTH, !var2.getBlockState(var3.south()).is(this))).setValue(WEST, !var2.getBlockState(var3.west()).is(this));
    }
 
    public BlockState updateShape(BlockState var1, Direction var2, BlockState var3, LevelAccessor var4, BlockPos var5, BlockPos var6) {
       return var3.is(this) ? (BlockState)var1.setValue((Property)PROPERTY_BY_DIRECTION.get(var2), false) : super.updateShape(var1, var2, var3, var4, var5, var6);
    }
