--- net/minecraft/world/level/block/JukeboxBlock.java
+++ net/minecraft/world/level/block/JukeboxBlock.java
@@ -9,11 +9,10 @@
 import net.minecraft.world.entity.item.ItemEntity;
 import net.minecraft.world.entity.player.Player;
 import net.minecraft.world.item.Item;
 import net.minecraft.world.item.ItemStack;
 import net.minecraft.world.item.RecordItem;
-import net.minecraft.world.level.BlockGetter;
 import net.minecraft.world.level.Level;
 import net.minecraft.world.level.LevelAccessor;
 import net.minecraft.world.level.block.entity.BlockEntity;
 import net.minecraft.world.level.block.entity.JukeboxBlockEntity;
 import net.minecraft.world.level.block.state.BlockBehaviour;
@@ -89,12 +88,12 @@
          this.dropRecording(var2, var3);
          super.onRemove(var1, var2, var3, var4, var5);
       }
    }
 
-   public BlockEntity newBlockEntity(BlockGetter var1) {
-      return new JukeboxBlockEntity();
+   public BlockEntity newBlockEntity(BlockPos var1, BlockState var2) {
+      return new JukeboxBlockEntity(var1, var2);
    }
 
    public boolean hasAnalogOutputSignal(BlockState var1) {
       return true;
    }
