--- net/minecraft/world/level/block/KelpBlock.java
+++ net/minecraft/world/level/block/KelpBlock.java
@@ -28,12 +28,12 @@
 
    protected Block getBodyBlock() {
       return Blocks.KELP_PLANT;
    }
 
-   protected boolean canAttachToBlock(Block var1) {
-      return var1 != Blocks.MAGMA_BLOCK;
+   protected boolean canAttachTo(BlockState var1) {
+      return !var1.is(Blocks.MAGMA_BLOCK);
    }
 
    public boolean canPlaceLiquid(BlockGetter var1, BlockPos var2, BlockState var3, Fluid var4) {
       return false;
    }
