--- net/minecraft/world/level/block/LadderBlock.java
+++ net/minecraft/world/level/block/LadderBlock.java
@@ -20,10 +20,11 @@
 import net.minecraft.world.phys.shapes.VoxelShape;
 
 public class LadderBlock extends Block implements SimpleWaterloggedBlock {
    public static final DirectionProperty FACING;
    public static final BooleanProperty WATERLOGGED;
+   protected static final float AABB_OFFSET = 3.0F;
    protected static final VoxelShape EAST_AABB;
    protected static final VoxelShape WEST_AABB;
    protected static final VoxelShape SOUTH_AABB;
    protected static final VoxelShape NORTH_AABB;
 
