--- net/minecraft/world/level/block/LeavesBlock.java
+++ net/minecraft/world/level/block/LeavesBlock.java
@@ -18,12 +18,14 @@
 import net.minecraft.world.level.block.state.properties.IntegerProperty;
 import net.minecraft.world.phys.shapes.Shapes;
 import net.minecraft.world.phys.shapes.VoxelShape;
 
 public class LeavesBlock extends Block {
+   public static final int DECAY_DISTANCE = 7;
    public static final IntegerProperty DISTANCE;
    public static final BooleanProperty PERSISTENT;
+   private static final int TICK_DELAY = 1;
 
    public LeavesBlock(BlockBehaviour.Properties var1) {
       super(var1);
       this.registerDefaultState((BlockState)((BlockState)((BlockState)this.stateDefinition.any()).setValue(DISTANCE, 7)).setValue(PERSISTENT, false));
    }
@@ -67,22 +69,22 @@
       Direction[] var5 = Direction.values();
       int var6 = var5.length;
 
       for(int var7 = 0; var7 < var6; ++var7) {
          Direction var8 = var5[var7];
-         var4.setWithOffset(var2, var8);
+         var4.setWithOffset(var2, (Direction)var8);
          var3 = Math.min(var3, getDistanceAt(var1.getBlockState(var4)) + 1);
          if (var3 == 1) {
             break;
          }
       }
 
       return (BlockState)var0.setValue(DISTANCE, var3);
    }
 
    private static int getDistanceAt(BlockState var0) {
-      if (BlockTags.LOGS.contains(var0.getBlock())) {
+      if (var0.is(BlockTags.LOGS)) {
          return 0;
       } else {
          return var0.getBlock() instanceof LeavesBlock ? (Integer)var0.getValue(DISTANCE) : 7;
       }
    }
