--- net/minecraft/world/level/block/NetherWartBlock.java
+++ net/minecraft/world/level/block/NetherWartBlock.java
@@ -13,10 +13,11 @@
 import net.minecraft.world.level.block.state.properties.IntegerProperty;
 import net.minecraft.world.phys.shapes.CollisionContext;
 import net.minecraft.world.phys.shapes.VoxelShape;
 
 public class NetherWartBlock extends BushBlock {
+   public static final int MAX_AGE = 3;
    public static final IntegerProperty AGE;
    private static final VoxelShape[] SHAPE_BY_AGE;
 
    protected NetherWartBlock(BlockBehaviour.Properties var1) {
       super(var1);
