--- net/minecraft/world/level/block/NetherPortalBlock.java
+++ net/minecraft/world/level/block/NetherPortalBlock.java
@@ -27,10 +27,11 @@
 import net.minecraft.world.phys.shapes.CollisionContext;
 import net.minecraft.world.phys.shapes.VoxelShape;
 
 public class NetherPortalBlock extends Block {
    public static final EnumProperty<Direction.Axis> AXIS;
+   protected static final int AABB_OFFSET = 2;
    protected static final VoxelShape X_AXIS_AABB;
    protected static final VoxelShape Z_AXIS_AABB;
 
    public NetherPortalBlock(BlockBehaviour.Properties var1) {
       super(var1);
