--- net/minecraft/world/level/block/RedstoneTorchBlock.java
+++ net/minecraft/world/level/block/RedstoneTorchBlock.java
@@ -18,10 +18,14 @@
 import net.minecraft.world.level.block.state.properties.BooleanProperty;
 
 public class RedstoneTorchBlock extends TorchBlock {
    public static final BooleanProperty LIT;
    private static final Map<BlockGetter, List<RedstoneTorchBlock.Toggle>> RECENT_TOGGLES;
+   public static final int RECENT_TOGGLE_TIMER = 60;
+   public static final int MAX_RECENT_TOGGLES = 8;
+   public static final int RESTART_DELAY = 160;
+   private static final int TOGGLE_DELAY = 2;
 
    protected RedstoneTorchBlock(BlockBehaviour.Properties var1) {
       super(var1, DustParticleOptions.REDSTONE);
       this.registerDefaultState((BlockState)((BlockState)this.stateDefinition.any()).setValue(LIT, true));
    }
