--- net/minecraft/world/level/block/RedstoneWallTorchBlock.java
+++ net/minecraft/world/level/block/RedstoneWallTorchBlock.java
@@ -43,11 +43,11 @@
    }
 
    @Nullable
    public BlockState getStateForPlacement(BlockPlaceContext var1) {
       BlockState var2 = Blocks.WALL_TORCH.getStateForPlacement(var1);
-      return var2 == null ? null : (BlockState)this.defaultBlockState().setValue(FACING, var2.getValue(FACING));
+      return var2 == null ? null : (BlockState)this.defaultBlockState().setValue(FACING, (Direction)var2.getValue(FACING));
    }
 
    public void animateTick(BlockState var1, Level var2, BlockPos var3, Random var4) {
       if ((Boolean)var1.getValue(LIT)) {
          Direction var5 = ((Direction)var1.getValue(FACING)).getOpposite();
