--- net/minecraft/world/level/block/RedStoneOreBlock.java
+++ net/minecraft/world/level/block/RedStoneOreBlock.java
@@ -32,13 +32,13 @@
    public void attack(BlockState var1, Level var2, BlockPos var3, Player var4) {
       interact(var1, var2, var3);
       super.attack(var1, var2, var3, var4);
    }
 
-   public void stepOn(Level var1, BlockPos var2, Entity var3) {
-      interact(var1.getBlockState(var2), var1, var2);
-      super.stepOn(var1, var2, var3);
+   public void stepOn(Level var1, BlockPos var2, BlockState var3, Entity var4) {
+      interact(var3, var1, var2);
+      super.stepOn(var1, var2, var3, var4);
    }
 
    public InteractionResult use(BlockState var1, Level var2, BlockPos var3, Player var4, InteractionHand var5, BlockHitResult var6) {
       if (var2.isClientSide) {
          spawnParticles(var2, var3);
