--- net/minecraft/world/level/block/RespawnAnchorBlock.java
+++ net/minecraft/world/level/block/RespawnAnchorBlock.java
@@ -53,11 +53,11 @@
       ItemStack var7 = var4.getItemInHand(var5);
       if (var5 == InteractionHand.MAIN_HAND && !isRespawnFuel(var7) && isRespawnFuel(var4.getItemInHand(InteractionHand.OFF_HAND))) {
          return InteractionResult.PASS;
       } else if (isRespawnFuel(var7) && canBeCharged(var1)) {
          charge(var2, var3, var1);
-         if (!var4.abilities.instabuild) {
+         if (!var4.getAbilities().instabuild) {
             var7.shrink(1);
          }
 
          return InteractionResult.sidedSuccess(var2.isClientSide);
       } else if ((Integer)var1.getValue(CHARGE) == 0) {
@@ -81,11 +81,11 @@
          return InteractionResult.CONSUME;
       }
    }
 
    private static boolean isRespawnFuel(ItemStack var0) {
-      return var0.getItem() == Items.GLOWSTONE;
+      return var0.is(Items.GLOWSTONE);
    }
 
    private static boolean canBeCharged(BlockState var0) {
       return (Integer)var0.getValue(CHARGE) < 4;
    }
