--- net/minecraft/world/level/block/RespawnAnchorBlock.java
+++ net/minecraft/world/level/block/RespawnAnchorBlock.java
@@ -1,10 +1,11 @@
 package net.minecraft.world.level.block;
 
 import com.google.common.collect.ImmutableList;
 import com.google.common.collect.UnmodifiableIterator;
 import com.google.common.collect.ImmutableList.Builder;
+import java.util.Objects;
 import java.util.Optional;
 import java.util.Random;
 import java.util.stream.Stream;
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.Direction;
@@ -38,10 +39,12 @@
 import net.minecraft.world.level.pathfinder.PathComputationType;
 import net.minecraft.world.phys.BlockHitResult;
 import net.minecraft.world.phys.Vec3;
 
 public class RespawnAnchorBlock extends Block {
+   public static final int MIN_CHARGES = 0;
+   public static final int MAX_CHARGES = 4;
    public static final IntegerProperty CHARGE;
    private static final ImmutableList<Vec3i> RESPAWN_HORIZONTAL_OFFSETS;
    private static final ImmutableList<Vec3i> RESPAWN_OFFSETS;
 
    public RespawnAnchorBlock(BlockBehaviour.Properties var1) {
@@ -53,11 +56,11 @@
       ItemStack var7 = var4.getItemInHand(var5);
       if (var5 == InteractionHand.MAIN_HAND && !isRespawnFuel(var7) && isRespawnFuel(var4.getItemInHand(InteractionHand.OFF_HAND))) {
          return InteractionResult.PASS;
       } else if (isRespawnFuel(var7) && canBeCharged(var1)) {
          charge(var2, var3, var1);
-         if (!var4.abilities.instabuild) {
+         if (!var4.getAbilities().instabuild) {
             var7.shrink(1);
          }
 
          return InteractionResult.sidedSuccess(var2.isClientSide);
       } else if ((Integer)var1.getValue(CHARGE) == 0) {
@@ -69,11 +72,11 @@
 
          return InteractionResult.sidedSuccess(var2.isClientSide);
       } else {
          if (!var2.isClientSide) {
             ServerPlayer var8 = (ServerPlayer)var4;
-            if (var8.getRespawnDimension() != var2.dimension() || !var8.getRespawnPosition().equals(var3)) {
+            if (var8.getRespawnDimension() != var2.dimension() || !var3.equals(var8.getRespawnPosition())) {
                var8.setRespawnPosition(var2.dimension(), var3, 0.0F, false, true);
                var2.playSound((Player)null, (double)var3.getX() + 0.5D, (double)var3.getY() + 0.5D, (double)var3.getZ() + 0.5D, SoundEvents.RESPAWN_ANCHOR_SET_SPAWN, SoundSource.BLOCKS, 1.0F, 1.0F);
                return InteractionResult.SUCCESS;
             }
          }
@@ -81,11 +84,11 @@
          return InteractionResult.CONSUME;
       }
    }
 
    private static boolean isRespawnFuel(ItemStack var0) {
-      return var0.getItem() == Items.GLOWSTONE;
+      return var0.is(Items.GLOWSTONE);
    }
 
    private static boolean canBeCharged(BlockState var0) {
       return (Integer)var0.getValue(CHARGE) < 4;
    }
@@ -108,11 +111,11 @@
    }
 
    private void explode(BlockState var1, Level var2, final BlockPos var3) {
       var2.removeBlock(var3, false);
       Stream var10000 = Direction.Plane.HORIZONTAL.stream();
-      var3.getClass();
+      Objects.requireNonNull(var3);
       boolean var4 = var10000.map(var3::relative).anyMatch((var1x) -> {
          return isWaterThatWouldFlow(var1x, var2);
       });
       final boolean var5 = var4 || var2.getFluidState(var3.above()).is(FluidTags.WATER);
       ExplosionDamageCalculator var6 = new ExplosionDamageCalculator() {
