--- net/minecraft/world/level/block/RotatedPillarBlock.java
+++ net/minecraft/world/level/block/RotatedPillarBlock.java
@@ -15,23 +15,27 @@
       super(var1);
       this.registerDefaultState((BlockState)this.defaultBlockState().setValue(AXIS, Direction.Axis.Y));
    }
 
    public BlockState rotate(BlockState var1, Rotation var2) {
-      switch(var2) {
+      return rotatePillar(var1, var2);
+   }
+
+   public static BlockState rotatePillar(BlockState var0, Rotation var1) {
+      switch(var1) {
       case COUNTERCLOCKWISE_90:
       case CLOCKWISE_90:
-         switch((Direction.Axis)var1.getValue(AXIS)) {
+         switch((Direction.Axis)var0.getValue(AXIS)) {
          case X:
-            return (BlockState)var1.setValue(AXIS, Direction.Axis.Z);
+            return (BlockState)var0.setValue(AXIS, Direction.Axis.Z);
          case Z:
-            return (BlockState)var1.setValue(AXIS, Direction.Axis.X);
+            return (BlockState)var0.setValue(AXIS, Direction.Axis.X);
          default:
-            return var1;
+            return var0;
          }
       default:
-         return var1;
+         return var0;
       }
    }
 
    protected void createBlockStateDefinition(StateDefinition.Builder<Block, BlockState> var1) {
       var1.add(AXIS);
