--- net/minecraft/world/level/block/SkullBlock.java
+++ net/minecraft/world/level/block/SkullBlock.java
@@ -12,10 +12,12 @@
 import net.minecraft.world.phys.shapes.CollisionContext;
 import net.minecraft.world.phys.shapes.Shapes;
 import net.minecraft.world.phys.shapes.VoxelShape;
 
 public class SkullBlock extends AbstractSkullBlock {
+   public static final int MAX = 15;
+   private static final int ROTATIONS = 16;
    public static final IntegerProperty ROTATION;
    protected static final VoxelShape SHAPE;
 
    protected SkullBlock(SkullBlock.Type var1, BlockBehaviour.Properties var2) {
       super(var1, var2);
