--- net/minecraft/world/level/block/SlimeBlock.java
+++ net/minecraft/world/level/block/SlimeBlock.java
@@ -1,25 +1,27 @@
 package net.minecraft.world.level.block;
 
 import net.minecraft.core.BlockPos;
+import net.minecraft.world.damagesource.DamageSource;
 import net.minecraft.world.entity.Entity;
 import net.minecraft.world.entity.LivingEntity;
 import net.minecraft.world.level.BlockGetter;
 import net.minecraft.world.level.Level;
 import net.minecraft.world.level.block.state.BlockBehaviour;
+import net.minecraft.world.level.block.state.BlockState;
 import net.minecraft.world.phys.Vec3;
 
 public class SlimeBlock extends HalfTransparentBlock {
    public SlimeBlock(BlockBehaviour.Properties var1) {
       super(var1);
    }
 
-   public void fallOn(Level var1, BlockPos var2, Entity var3, float var4) {
-      if (var3.isSuppressingBounce()) {
-         super.fallOn(var1, var2, var3, var4);
+   public void fallOn(Level var1, BlockState var2, BlockPos var3, Entity var4, float var5) {
+      if (var4.isSuppressingBounce()) {
+         super.fallOn(var1, var2, var3, var4, var5);
       } else {
-         var3.causeFallDamage(var4, 0.0F);
+         var4.causeFallDamage(var5, 0.0F, DamageSource.FALL);
       }
 
    }
 
    public void updateEntityAfterFallOn(BlockGetter var1, Entity var2) {
@@ -38,15 +40,15 @@
          var1.setDeltaMovement(var2.x, -var2.y * var3, var2.z);
       }
 
    }
 
-   public void stepOn(Level var1, BlockPos var2, Entity var3) {
-      double var4 = Math.abs(var3.getDeltaMovement().y);
-      if (var4 < 0.1D && !var3.isSteppingCarefully()) {
-         double var6 = 0.4D + var4 * 0.2D;
-         var3.setDeltaMovement(var3.getDeltaMovement().multiply(var6, 1.0D, var6));
+   public void stepOn(Level var1, BlockPos var2, BlockState var3, Entity var4) {
+      double var5 = Math.abs(var4.getDeltaMovement().y);
+      if (var5 < 0.1D && !var4.isSteppingCarefully()) {
+         double var7 = 0.4D + var5 * 0.2D;
+         var4.setDeltaMovement(var4.getDeltaMovement().multiply(var7, 1.0D, var7));
       }
 
-      super.stepOn(var1, var2, var3);
+      super.stepOn(var1, var2, var3, var4);
    }
 }
