--- net/minecraft/world/level/block/SmokerBlock.java
+++ net/minecraft/world/level/block/SmokerBlock.java
@@ -1,31 +1,38 @@
 package net.minecraft.world.level.block;
 
 import java.util.Random;
+import javax.annotation.Nullable;
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.particles.ParticleTypes;
 import net.minecraft.sounds.SoundEvents;
 import net.minecraft.sounds.SoundSource;
 import net.minecraft.stats.Stats;
 import net.minecraft.world.MenuProvider;
 import net.minecraft.world.entity.player.Player;
-import net.minecraft.world.level.BlockGetter;
 import net.minecraft.world.level.Level;
 import net.minecraft.world.level.block.entity.BlockEntity;
+import net.minecraft.world.level.block.entity.BlockEntityTicker;
+import net.minecraft.world.level.block.entity.BlockEntityType;
 import net.minecraft.world.level.block.entity.SmokerBlockEntity;
 import net.minecraft.world.level.block.state.BlockBehaviour;
 import net.minecraft.world.level.block.state.BlockState;
 
 public class SmokerBlock extends AbstractFurnaceBlock {
    protected SmokerBlock(BlockBehaviour.Properties var1) {
       super(var1);
    }
 
-   public BlockEntity newBlockEntity(BlockGetter var1) {
-      return new SmokerBlockEntity();
+   public BlockEntity newBlockEntity(BlockPos var1, BlockState var2) {
+      return new SmokerBlockEntity(var1, var2);
    }
 
+   @Nullable
+   public <T extends BlockEntity> BlockEntityTicker<T> getTicker(Level var1, BlockState var2, BlockEntityType<T> var3) {
+      return createFurnaceTicker(var1, var3, BlockEntityType.SMOKER);
+   }
+
    protected void openContainer(Level var1, BlockPos var2, Player var3) {
       BlockEntity var4 = var1.getBlockEntity(var2);
       if (var4 instanceof SmokerBlockEntity) {
          var3.openMenu((MenuProvider)var4);
          var3.awardStat(Stats.INTERACT_WITH_SMOKER);
