--- net/minecraft/world/level/block/SoulFireBlock.java
+++ net/minecraft/world/level/block/SoulFireBlock.java
@@ -1,11 +1,10 @@
 package net.minecraft.world.level.block;
 
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.Direction;
 import net.minecraft.tags.BlockTags;
-import net.minecraft.tags.Tag;
 import net.minecraft.world.level.LevelAccessor;
 import net.minecraft.world.level.LevelReader;
 import net.minecraft.world.level.block.state.BlockBehaviour;
 import net.minecraft.world.level.block.state.BlockState;
 
@@ -17,15 +16,15 @@
    public BlockState updateShape(BlockState var1, Direction var2, BlockState var3, LevelAccessor var4, BlockPos var5, BlockPos var6) {
       return this.canSurvive(var1, var4, var5) ? this.defaultBlockState() : Blocks.AIR.defaultBlockState();
    }
 
    public boolean canSurvive(BlockState var1, LevelReader var2, BlockPos var3) {
-      return canSurviveOnBlock(var2.getBlockState(var3.below()).getBlock());
+      return canSurviveOnBlock(var2.getBlockState(var3.below()));
    }
 
-   public static boolean canSurviveOnBlock(Block var0) {
-      return var0.is((Tag)BlockTags.SOUL_FIRE_BASE_BLOCKS);
+   public static boolean canSurviveOnBlock(BlockState var0) {
+      return var0.is(BlockTags.SOUL_FIRE_BASE_BLOCKS);
    }
 
    protected boolean canBurn(BlockState var1) {
       return true;
    }
