--- net/minecraft/world/level/block/StairBlock.java
+++ net/minecraft/world/level/block/StairBlock.java
@@ -127,12 +127,12 @@
       if (!var1.is(var4.getBlock())) {
          this.baseState.onRemove(var2, var3, var4, var5);
       }
    }
 
-   public void stepOn(Level var1, BlockPos var2, Entity var3) {
-      this.base.stepOn(var1, var2, var3);
+   public void stepOn(Level var1, BlockPos var2, BlockState var3, Entity var4) {
+      this.base.stepOn(var1, var2, var3, var4);
    }
 
    public boolean isRandomlyTicking(BlockState var1) {
       return this.base.isRandomlyTicking(var1);
    }
