--- net/minecraft/world/level/block/StandingSignBlock.java
+++ net/minecraft/world/level/block/StandingSignBlock.java
@@ -18,11 +18,11 @@
 public class StandingSignBlock extends SignBlock {
    public static final IntegerProperty ROTATION;
 
    public StandingSignBlock(BlockBehaviour.Properties var1, WoodType var2) {
       super(var1, var2);
-      this.registerDefaultState((BlockState)((BlockState)((BlockState)this.stateDefinition.any()).setValue(ROTATION, 0)).setValue(WATERLOGGED, false));
+      this.registerDefaultState((BlockState)((BlockState)((BlockState)((BlockState)this.stateDefinition.any()).setValue(ROTATION, 0)).setValue(WATERLOGGED, false)).setValue(LIT, false));
    }
 
    public boolean canSurvive(BlockState var1, LevelReader var2, BlockPos var3) {
       return var2.getBlockState(var3.below()).getMaterial().isSolid();
    }
@@ -43,11 +43,11 @@
    public BlockState mirror(BlockState var1, Mirror var2) {
       return (BlockState)var1.setValue(ROTATION, var2.mirror((Integer)var1.getValue(ROTATION), 16));
    }
 
    protected void createBlockStateDefinition(StateDefinition.Builder<Block, BlockState> var1) {
-      var1.add(ROTATION, WATERLOGGED);
+      var1.add(ROTATION, WATERLOGGED, LIT);
    }
 
    static {
       ROTATION = BlockStateProperties.ROTATION_16;
    }
