--- net/minecraft/world/level/block/SupportType.java
+++ net/minecraft/world/level/block/SupportType.java
@@ -33,10 +33,10 @@
       public boolean isSupporting(BlockState var1, BlockGetter var2, BlockPos var3, Direction var4) {
          return !Shapes.joinIsNotEmpty(var1.getBlockSupportShape(var2, var3).getFaceShape(var4), this.RIGID_SUPPORT_SHAPE, BooleanOp.ONLY_SECOND);
       }
    };
 
-   private SupportType() {
+   SupportType() {
    }
 
    public abstract boolean isSupporting(BlockState var1, BlockGetter var2, BlockPos var3, Direction var4);
 }
