--- net/minecraft/world/level/block/SweetBerryBushBlock.java
+++ net/minecraft/world/level/block/SweetBerryBushBlock.java
@@ -75,11 +75,11 @@
    }
 
    public InteractionResult use(BlockState var1, Level var2, BlockPos var3, Player var4, InteractionHand var5, BlockHitResult var6) {
       int var7 = (Integer)var1.getValue(AGE);
       boolean var8 = var7 == 3;
-      if (!var8 && var4.getItemInHand(var5).getItem() == Items.BONE_MEAL) {
+      if (!var8 && var4.getItemInHand(var5).is(Items.BONE_MEAL)) {
          return InteractionResult.PASS;
       } else if (var7 > 1) {
          int var9 = 1 + var2.random.nextInt(2);
          popResource(var2, var3, new ItemStack(Items.SWEET_BERRIES, var9 + (var8 ? 1 : 0)));
          var2.playSound((Player)null, (BlockPos)var3, SoundEvents.SWEET_BERRY_BUSH_PICK_BERRIES, SoundSource.BLOCKS, 1.0F, 0.8F + var2.random.nextFloat() * 0.4F);
