--- net/minecraft/world/level/block/TorchBlock.java
+++ net/minecraft/world/level/block/TorchBlock.java
@@ -13,10 +13,11 @@
 import net.minecraft.world.level.block.state.BlockState;
 import net.minecraft.world.phys.shapes.CollisionContext;
 import net.minecraft.world.phys.shapes.VoxelShape;
 
 public class TorchBlock extends Block {
+   protected static final int AABB_STANDING_OFFSET = 2;
    protected static final VoxelShape AABB = Block.box(6.0D, 0.0D, 6.0D, 10.0D, 10.0D, 10.0D);
    protected final ParticleOptions flameParticle;
 
    protected TorchBlock(BlockBehaviour.Properties var1, ParticleOptions var2) {
       super(var1);
