--- net/minecraft/world/level/block/TrappedChestBlock.java
+++ net/minecraft/world/level/block/TrappedChestBlock.java
@@ -19,12 +19,12 @@
       super(var1, () -> {
          return BlockEntityType.TRAPPED_CHEST;
       });
    }
 
-   public BlockEntity newBlockEntity(BlockGetter var1) {
-      return new TrappedChestBlockEntity();
+   public BlockEntity newBlockEntity(BlockPos var1, BlockState var2) {
+      return new TrappedChestBlockEntity(var1, var2);
    }
 
    protected Stat<ResourceLocation> getOpenChestStat() {
       return Stats.CUSTOM.get(Stats.TRIGGER_TRAPPED_CHEST);
    }
