--- net/minecraft/world/level/block/WallTorchBlock.java
+++ net/minecraft/world/level/block/WallTorchBlock.java
@@ -21,10 +21,11 @@
 import net.minecraft.world.phys.shapes.CollisionContext;
 import net.minecraft.world.phys.shapes.VoxelShape;
 
 public class WallTorchBlock extends TorchBlock {
    public static final DirectionProperty FACING;
+   protected static final float AABB_OFFSET = 2.5F;
    private static final Map<Direction, VoxelShape> AABBS;
 
    protected WallTorchBlock(BlockBehaviour.Properties var1, ParticleOptions var2) {
       super(var1, var2);
       this.registerDefaultState((BlockState)((BlockState)this.stateDefinition.any()).setValue(FACING, Direction.NORTH));
