--- net/minecraft/world/level/block/WallSignBlock.java
+++ net/minecraft/world/level/block/WallSignBlock.java
@@ -21,10 +21,13 @@
 import net.minecraft.world.phys.shapes.CollisionContext;
 import net.minecraft.world.phys.shapes.VoxelShape;
 
 public class WallSignBlock extends SignBlock {
    public static final DirectionProperty FACING;
+   protected static final float AABB_THICKNESS = 2.0F;
+   protected static final float AABB_BOTTOM = 4.5F;
+   protected static final float AABB_TOP = 12.5F;
    private static final Map<Direction, VoxelShape> AABBS;
 
    public WallSignBlock(BlockBehaviour.Properties var1, WoodType var2) {
       super(var1, var2);
       this.registerDefaultState((BlockState)((BlockState)((BlockState)this.stateDefinition.any()).setValue(FACING, Direction.NORTH)).setValue(WATERLOGGED, false));
