--- net/minecraft/world/level/block/WoolCarpetBlock.java
+++ net/minecraft/world/level/block/WoolCarpetBlock.java
@@ -1,38 +1,17 @@
 package net.minecraft.world.level.block;
 
-import net.minecraft.core.BlockPos;
-import net.minecraft.core.Direction;
 import net.minecraft.world.item.DyeColor;
-import net.minecraft.world.level.BlockGetter;
-import net.minecraft.world.level.LevelAccessor;
-import net.minecraft.world.level.LevelReader;
 import net.minecraft.world.level.block.state.BlockBehaviour;
-import net.minecraft.world.level.block.state.BlockState;
-import net.minecraft.world.phys.shapes.CollisionContext;
-import net.minecraft.world.phys.shapes.VoxelShape;
 
-public class WoolCarpetBlock extends Block {
-   protected static final VoxelShape SHAPE = Block.box(0.0D, 0.0D, 0.0D, 16.0D, 1.0D, 16.0D);
+public class WoolCarpetBlock extends CarpetBlock {
    private final DyeColor color;
 
    protected WoolCarpetBlock(DyeColor var1, BlockBehaviour.Properties var2) {
       super(var2);
       this.color = var1;
    }
 
    public DyeColor getColor() {
       return this.color;
    }
-
-   public VoxelShape getShape(BlockState var1, BlockGetter var2, BlockPos var3, CollisionContext var4) {
-      return SHAPE;
-   }
-
-   public BlockState updateShape(BlockState var1, Direction var2, BlockState var3, LevelAccessor var4, BlockPos var5, BlockPos var6) {
-      return !var1.canSurvive(var4, var5) ? Blocks.AIR.defaultBlockState() : super.updateShape(var1, var2, var3, var4, var5, var6);
-   }
-
-   public boolean canSurvive(BlockState var1, LevelReader var2, BlockPos var3) {
-      return !var2.isEmptyBlock(var3.below());
-   }
 }
