--- net/minecraft/world/scores/criteria/package-info.java
+++ net/minecraft/world/scores/criteria/package-info.java
@@ -1,0 +1,10 @@
+package net.minecraft.world.scores.criteria;
+
+import javax.annotation.ParametersAreNonnullByDefault;
+import net.minecraft.MethodsReturnNonnullByDefault;
+
+// $FF: synthetic class
+@ParametersAreNonnullByDefault
+@MethodsReturnNonnullByDefault
+interface package-info {
+}
