--- com/mojang/math/OctahedralGroup.java
+++ com/mojang/math/OctahedralGroup.java
@@ -128,10 +128,18 @@
 
    public OctahedralGroup compose(OctahedralGroup var1) {
       return cayleyTable[this.ordinal()][var1.ordinal()];
    }
 
+   public OctahedralGroup inverse() {
+      return inverseTable[this.ordinal()];
+   }
+
+   public Matrix3f transformation() {
+      return this.transformation.copy();
+   }
+
    public String toString() {
       return this.name;
    }
 
    public String getSerializedName() {
