--- com/mojang/math/Vector4f.java
+++ com/mojang/math/Vector4f.java
@@ -63,10 +63,17 @@
 
    public float w() {
       return this.w;
    }
 
+   public void mul(float var1) {
+      this.x *= var1;
+      this.y *= var1;
+      this.z *= var1;
+      this.w *= var1;
+   }
+
    public void mul(Vector3f var1) {
       this.x *= var1.x();
       this.y *= var1.y();
       this.z *= var1.z();
    }
@@ -76,10 +83,17 @@
       this.y = var2;
       this.z = var3;
       this.w = var4;
    }
 
+   public void add(float var1, float var2, float var3, float var4) {
+      this.x += var1;
+      this.y += var2;
+      this.z += var3;
+      this.w += var4;
+   }
+
    public float dot(Vector4f var1) {
       return this.x * var1.x + this.y * var1.y + this.z * var1.z + this.w * var1.w;
    }
 
    public boolean normalize() {
@@ -121,9 +135,17 @@
       this.y /= this.w;
       this.z /= this.w;
       this.w = 1.0F;
    }
 
+   public void lerp(Vector4f var1, float var2) {
+      float var3 = 1.0F - var2;
+      this.x = this.x * var3 + var1.x * var2;
+      this.y = this.y * var3 + var1.y * var2;
+      this.z = this.z * var3 + var1.z * var2;
+      this.w = this.w * var3 + var1.w * var2;
+   }
+
    public String toString() {
       return "[" + this.x + ", " + this.y + ", " + this.z + ", " + this.w + "]";
    }
 }
