--- com/mojang/math/Vector3f.java
+++ com/mojang/math/Vector3f.java
@@ -1,18 +1,24 @@
 package com.mojang.math;
 
+import com.google.common.collect.ImmutableList;
+import com.mojang.serialization.Codec;
 import it.unimi.dsi.fastutil.floats.Float2FloatFunction;
+import java.util.List;
+import net.minecraft.Util;
 import net.minecraft.util.Mth;
 import net.minecraft.world.phys.Vec3;
 
 public final class Vector3f {
-   public static Vector3f XN = new Vector3f(-1.0F, 0.0F, 0.0F);
-   public static Vector3f XP = new Vector3f(1.0F, 0.0F, 0.0F);
-   public static Vector3f YN = new Vector3f(0.0F, -1.0F, 0.0F);
-   public static Vector3f YP = new Vector3f(0.0F, 1.0F, 0.0F);
-   public static Vector3f ZN = new Vector3f(0.0F, 0.0F, -1.0F);
-   public static Vector3f ZP = new Vector3f(0.0F, 0.0F, 1.0F);
+   public static final Codec<Vector3f> CODEC;
+   public static Vector3f XN;
+   public static Vector3f XP;
+   public static Vector3f YN;
+   public static Vector3f YP;
+   public static Vector3f ZN;
+   public static Vector3f ZP;
+   public static Vector3f ZERO;
    private float x;
    private float y;
    private float z;
 
    public Vector3f() {
@@ -22,10 +28,14 @@
       this.x = var1;
       this.y = var2;
       this.z = var3;
    }
 
+   public Vector3f(Vector4f var1) {
+      this(var1.x(), var1.y(), var1.z());
+   }
+
    public Vector3f(Vec3 var1) {
       this((float)var1.x, (float)var1.y, (float)var1.z);
    }
 
    public boolean equals(Object var1) {
@@ -74,10 +84,16 @@
       this.x *= var1;
       this.y *= var2;
       this.z *= var3;
    }
 
+   public void clamp(Vector3f var1, Vector3f var2) {
+      this.x = Mth.clamp(this.x, var1.x(), var2.x());
+      this.y = Mth.clamp(this.y, var1.x(), var2.y());
+      this.z = Mth.clamp(this.z, var1.z(), var2.z());
+   }
+
    public void clamp(float var1, float var2) {
       this.x = Mth.clamp(this.x, var1, var2);
       this.y = Mth.clamp(this.y, var1, var2);
       this.z = Mth.clamp(this.z, var1, var2);
    }
@@ -86,10 +102,16 @@
       this.x = var1;
       this.y = var2;
       this.z = var3;
    }
 
+   public void load(Vector3f var1) {
+      this.x = var1.x;
+      this.y = var1.y;
+      this.z = var1.z;
+   }
+
    public void add(float var1, float var2, float var3) {
       this.x += var1;
       this.y += var2;
       this.z += var3;
    }
@@ -179,6 +201,23 @@
    }
 
    public String toString() {
       return "[" + this.x + ", " + this.y + ", " + this.z + "]";
    }
+
+   static {
+      CODEC = Codec.FLOAT.listOf().comapFlatMap((var0) -> {
+         return Util.fixedSize((List)var0, 3).map((var0x) -> {
+            return new Vector3f((Float)var0x.get(0), (Float)var0x.get(1), (Float)var0x.get(2));
+         });
+      }, (var0) -> {
+         return ImmutableList.of(var0.x, var0.y, var0.z);
+      });
+      XN = new Vector3f(-1.0F, 0.0F, 0.0F);
+      XP = new Vector3f(1.0F, 0.0F, 0.0F);
+      YN = new Vector3f(0.0F, -1.0F, 0.0F);
+      YP = new Vector3f(0.0F, 1.0F, 0.0F);
+      ZN = new Vector3f(0.0F, 0.0F, -1.0F);
+      ZP = new Vector3f(0.0F, 0.0F, 1.0F);
+      ZERO = new Vector3f(0.0F, 0.0F, 0.0F);
+   }
 }
