--- net/minecraft/world/level/storage/loot/parameters/LootContextParams.java
+++ net/minecraft/world/level/storage/loot/parameters/LootContextParams.java
@@ -19,9 +19,12 @@
    public static final LootContextParam<BlockState> BLOCK_STATE = create("block_state");
    public static final LootContextParam<BlockEntity> BLOCK_ENTITY = create("block_entity");
    public static final LootContextParam<ItemStack> TOOL = create("tool");
    public static final LootContextParam<Float> EXPLOSION_RADIUS = create("explosion_radius");
 
+   public LootContextParams() {
+   }
+
    private static <T> LootContextParam<T> create(String var0) {
       return new LootContextParam(new ResourceLocation(var0));
    }
 }
