--- net/minecraft/world/level/storage/loot/entries/DynamicLoot.java
+++ net/minecraft/world/level/storage/loot/entries/DynamicLoot.java
@@ -10,13 +10,13 @@
 import net.minecraft.world.level.storage.loot.LootContext;
 import net.minecraft.world.level.storage.loot.functions.LootItemFunction;
 import net.minecraft.world.level.storage.loot.predicates.LootItemCondition;
 
 public class DynamicLoot extends LootPoolSingletonContainer {
-   private final ResourceLocation name;
+   final ResourceLocation name;
 
-   private DynamicLoot(ResourceLocation var1, int var2, int var3, LootItemCondition[] var4, LootItemFunction[] var5) {
+   DynamicLoot(ResourceLocation var1, int var2, int var3, LootItemCondition[] var4, LootItemFunction[] var5) {
       super(var2, var3, var4, var5);
       this.name = var1;
    }
 
    public LootPoolEntryType getType() {
