--- net/minecraft/world/level/storage/loot/entries/LootPoolSingletonContainer.java
+++ net/minecraft/world/level/storage/loot/entries/LootPoolSingletonContainer.java
@@ -17,10 +17,12 @@
 import net.minecraft.world.level.storage.loot.functions.LootItemFunctions;
 import net.minecraft.world.level.storage.loot.predicates.LootItemCondition;
 import org.apache.commons.lang3.ArrayUtils;
 
 public abstract class LootPoolSingletonContainer extends LootPoolEntryContainer {
+   public static final int DEFAULT_WEIGHT = 1;
+   public static final int DEFAULT_QUALITY = 0;
    protected final int weight;
    protected final int quality;
    protected final LootItemFunction[] functions;
    private final BiFunction<ItemStack, LootContext, ItemStack> compositeFunction;
    private final LootPoolEntry entry = new LootPoolSingletonContainer.EntryBase() {
