--- net/minecraft/world/level/storage/loot/entries/LootPoolEntryContainer.java
+++ net/minecraft/world/level/storage/loot/entries/LootPoolEntryContainer.java
@@ -81,8 +81,16 @@
 
       public AlternativesEntry.Builder otherwise(LootPoolEntryContainer.Builder<?> var1) {
          return new AlternativesEntry.Builder(new LootPoolEntryContainer.Builder[]{this, var1});
       }
 
+      public EntryGroup.Builder append(LootPoolEntryContainer.Builder<?> var1) {
+         return new EntryGroup.Builder(new LootPoolEntryContainer.Builder[]{this, var1});
+      }
+
+      public SequentialEntry.Builder then(LootPoolEntryContainer.Builder<?> var1) {
+         return new SequentialEntry.Builder(new LootPoolEntryContainer.Builder[]{this, var1});
+      }
+
       public abstract LootPoolEntryContainer build();
    }
 }
