--- net/minecraft/world/level/storage/loot/entries/LootItem.java
+++ net/minecraft/world/level/storage/loot/entries/LootItem.java
@@ -13,13 +13,13 @@
 import net.minecraft.world.level.storage.loot.LootContext;
 import net.minecraft.world.level.storage.loot.functions.LootItemFunction;
 import net.minecraft.world.level.storage.loot.predicates.LootItemCondition;
 
 public class LootItem extends LootPoolSingletonContainer {
-   private final Item item;
+   final Item item;
 
-   private LootItem(Item var1, int var2, int var3, LootItemCondition[] var4, LootItemFunction[] var5) {
+   LootItem(Item var1, int var2, int var3, LootItemCondition[] var4, LootItemFunction[] var5) {
       super(var2, var3, var4, var5);
       this.item = var1;
    }
 
    public LootPoolEntryType getType() {
