--- net/minecraft/advancements/CriterionProgress.java
+++ net/minecraft/advancements/CriterionProgress.java
@@ -31,11 +31,12 @@
    public Date getObtained() {
       return this.obtained;
    }
 
    public String toString() {
-      return "CriterionProgress{obtained=" + (this.obtained == null ? "false" : this.obtained) + '}';
+      Object var10000 = this.obtained == null ? "false" : this.obtained;
+      return "CriterionProgress{obtained=" + var10000 + "}";
    }
 
    public void serializeToNetwork(FriendlyByteBuf var1) {
       var1.writeBoolean(this.obtained != null);
       if (this.obtained != null) {
