--- net/minecraft/server/network/LegacyQueryHandler.java
+++ net/minecraft/server/network/LegacyQueryHandler.java
@@ -11,17 +11,18 @@
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
 
 public class LegacyQueryHandler extends ChannelInboundHandlerAdapter {
    private static final Logger LOGGER = LogManager.getLogger();
+   public static final int FAKE_PROTOCOL_VERSION = 127;
    private final ServerConnectionListener serverConnectionListener;
 
    public LegacyQueryHandler(ServerConnectionListener var1) {
       this.serverConnectionListener = var1;
    }
 
-   public void channelRead(ChannelHandlerContext var1, Object var2) throws Exception {
+   public void channelRead(ChannelHandlerContext var1, Object var2) {
       ByteBuf var3 = (ByteBuf)var2;
       var3.markReaderIndex();
       boolean var4 = true;
 
       try {
