--- net/minecraft/server/network/ServerConnectionListener.java
+++ net/minecraft/server/network/ServerConnectionListener.java
@@ -69,11 +69,11 @@
             var5 = SERVER_EVENT_GROUP;
             LOGGER.info("Using default channel type");
          }
 
          this.channels.add(((ServerBootstrap)((ServerBootstrap)(new ServerBootstrap()).channel(var4)).childHandler(new ChannelInitializer<Channel>() {
-            protected void initChannel(Channel var1) throws Exception {
+            protected void initChannel(Channel var1) {
                try {
                   var1.config().setOption(ChannelOption.TCP_NODELAY, true);
                } catch (ChannelException var4) {
                }
 
