--- net/minecraft/world/phys/AABB.java
+++ net/minecraft/world/phys/AABB.java
@@ -314,9 +314,9 @@
 
    public Vec3 getCenter() {
       return new Vec3(Mth.lerp(0.5D, this.minX, this.maxX), Mth.lerp(0.5D, this.minY, this.maxY), Mth.lerp(0.5D, this.minZ, this.maxZ));
    }
 
-   public static AABB ofSize(double var0, double var2, double var4) {
-      return new AABB(-var0 / 2.0D, -var2 / 2.0D, -var4 / 2.0D, var0 / 2.0D, var2 / 2.0D, var4 / 2.0D);
+   public static AABB ofSize(Vec3 var0, double var1, double var3, double var5) {
+      return new AABB(var0.x - var1 / 2.0D, var0.y - var3 / 2.0D, var0.z - var5 / 2.0D, var0.x + var1 / 2.0D, var0.y + var3 / 2.0D, var0.z + var5 / 2.0D);
    }
 }
