--- net/minecraft/world/level/block/entity/CommandBlockEntity.java
+++ net/minecraft/world/level/block/entity/CommandBlockEntity.java
@@ -39,12 +39,12 @@
       public CommandSourceStack createCommandSourceStack() {
          return new CommandSourceStack(this, Vec3.atCenterOf(CommandBlockEntity.this.worldPosition), Vec2.ZERO, this.getLevel(), 2, this.getName().getString(), this.getName(), this.getLevel().getServer(), (Entity)null);
       }
    };
 
-   public CommandBlockEntity() {
-      super(BlockEntityType.COMMAND_BLOCK);
+   public CommandBlockEntity(BlockPos var1, BlockState var2) {
+      super(BlockEntityType.COMMAND_BLOCK, var1, var2);
    }
 
    public CompoundTag save(CompoundTag var1) {
       super.save(var1);
       this.commandBlock.save(var1);
@@ -52,16 +52,16 @@
       var1.putBoolean("conditionMet", this.wasConditionMet());
       var1.putBoolean("auto", this.isAutomatic());
       return var1;
    }
 
-   public void load(BlockState var1, CompoundTag var2) {
-      super.load(var1, var2);
-      this.commandBlock.load(var2);
-      this.powered = var2.getBoolean("powered");
-      this.conditionMet = var2.getBoolean("conditionMet");
-      this.setAutomatic(var2.getBoolean("auto"));
+   public void load(CompoundTag var1) {
+      super.load(var1);
+      this.commandBlock.load(var1);
+      this.powered = var1.getBoolean("powered");
+      this.conditionMet = var1.getBoolean("conditionMet");
+      this.setAutomatic(var1.getBoolean("auto"));
    }
 
    @Nullable
    public ClientboundBlockEntityDataPacket getUpdatePacket() {
       if (this.isSendToClient()) {
@@ -160,15 +160,10 @@
    public boolean isConditional() {
       BlockState var1 = this.level.getBlockState(this.getBlockPos());
       return var1.getBlock() instanceof CommandBlock ? (Boolean)var1.getValue(CommandBlock.CONDITIONAL) : false;
    }
 
-   public void clearRemoved() {
-      this.clearCache();
-      super.clearRemoved();
-   }
-
    public static enum Mode {
       SEQUENCE,
       AUTO,
       REDSTONE;
 
