--- net/minecraft/world/level/block/entity/Hopper.java
+++ net/minecraft/world/level/block/entity/Hopper.java
@@ -1,10 +1,8 @@
 package net.minecraft.world.level.block.entity;
 
-import javax.annotation.Nullable;
 import net.minecraft.world.Container;
-import net.minecraft.world.level.Level;
 import net.minecraft.world.level.block.Block;
 import net.minecraft.world.phys.shapes.Shapes;
 import net.minecraft.world.phys.shapes.VoxelShape;
 
 public interface Hopper extends Container {
@@ -14,13 +12,10 @@
 
    default VoxelShape getSuckShape() {
       return SUCK;
    }
 
-   @Nullable
-   Level getLevel();
-
    double getLevelX();
 
    double getLevelY();
 
    double getLevelZ();
