--- net/minecraft/world/level/block/entity/JigsawBlockEntity.java
+++ net/minecraft/world/level/block/entity/JigsawBlockEntity.java
@@ -6,11 +6,14 @@
 import java.util.Iterator;
 import java.util.Optional;
 import java.util.Random;
 import javax.annotation.Nullable;
 import net.minecraft.core.BlockPos;
+import net.minecraft.core.Vec3i;
 import net.minecraft.nbt.CompoundTag;
+import net.minecraft.network.chat.Component;
+import net.minecraft.network.chat.TranslatableComponent;
 import net.minecraft.network.protocol.game.ClientboundBlockEntityDataPacket;
 import net.minecraft.resources.ResourceLocation;
 import net.minecraft.server.level.ServerLevel;
 import net.minecraft.util.StringRepresentable;
 import net.minecraft.world.level.StructureFeatureManager;
@@ -25,29 +28,47 @@
 import net.minecraft.world.level.levelgen.structure.PoolElementStructurePiece;
 import net.minecraft.world.level.levelgen.structure.templatesystem.StructureManager;
 import net.minecraft.world.level.levelgen.structure.templatesystem.StructureTemplate;
 
 public class JigsawBlockEntity extends BlockEntity {
-   private ResourceLocation name;
-   private ResourceLocation target;
-   private ResourceLocation pool;
+   public static final String TARGET = "target";
+   public static final String POOL = "pool";
+   public static final String JOINT = "joint";
+   public static final String NAME = "name";
+   public static final String FINAL_STATE = "final_state";
+   private ResourceLocation name = new ResourceLocation("empty");
+   private ResourceLocation target = new ResourceLocation("empty");
+   private ResourceLocation pool = new ResourceLocation("empty");
    private JigsawBlockEntity.JointType joint;
    private String finalState;
 
-   public JigsawBlockEntity(BlockEntityType<?> var1) {
-      super(var1);
-      this.name = new ResourceLocation("empty");
-      this.target = new ResourceLocation("empty");
-      this.pool = new ResourceLocation("empty");
+   public JigsawBlockEntity(BlockPos var1, BlockState var2) {
+      super(BlockEntityType.JIGSAW, var1, var2);
       this.joint = JigsawBlockEntity.JointType.ROLLABLE;
       this.finalState = "minecraft:air";
    }
 
-   public JigsawBlockEntity() {
-      this(BlockEntityType.JIGSAW);
+   public ResourceLocation getName() {
+      return this.name;
    }
 
+   public ResourceLocation getTarget() {
+      return this.target;
+   }
+
+   public ResourceLocation getPool() {
+      return this.pool;
+   }
+
+   public String getFinalState() {
+      return this.finalState;
+   }
+
+   public JigsawBlockEntity.JointType getJoint() {
+      return this.joint;
+   }
+
    public void setName(ResourceLocation var1) {
       this.name = var1;
    }
 
    public void setTarget(ResourceLocation var1) {
@@ -74,18 +95,18 @@
       var1.putString("final_state", this.finalState);
       var1.putString("joint", this.joint.getSerializedName());
       return var1;
    }
 
-   public void load(BlockState var1, CompoundTag var2) {
-      super.load(var1, var2);
-      this.name = new ResourceLocation(var2.getString("name"));
-      this.target = new ResourceLocation(var2.getString("target"));
-      this.pool = new ResourceLocation(var2.getString("pool"));
-      this.finalState = var2.getString("final_state");
-      this.joint = (JigsawBlockEntity.JointType)JigsawBlockEntity.JointType.byName(var2.getString("joint")).orElseGet(() -> {
-         return JigsawBlock.getFrontFacing(var1).getAxis().isHorizontal() ? JigsawBlockEntity.JointType.ALIGNED : JigsawBlockEntity.JointType.ROLLABLE;
+   public void load(CompoundTag var1) {
+      super.load(var1);
+      this.name = new ResourceLocation(var1.getString("name"));
+      this.target = new ResourceLocation(var1.getString("target"));
+      this.pool = new ResourceLocation(var1.getString("pool"));
+      this.finalState = var1.getString("final_state");
+      this.joint = (JigsawBlockEntity.JointType)JigsawBlockEntity.JointType.byName(var1.getString("joint")).orElseGet(() -> {
+         return JigsawBlock.getFrontFacing(this.getBlockState()).getAxis().isHorizontal() ? JigsawBlockEntity.JointType.ALIGNED : JigsawBlockEntity.JointType.ROLLABLE;
       });
    }
 
    @Nullable
    public ClientboundBlockEntityDataPacket getUpdatePacket() {
@@ -102,14 +123,14 @@
       StructureFeatureManager var6 = var1.structureFeatureManager();
       Random var7 = var1.getRandom();
       BlockPos var8 = this.getBlockPos();
       ArrayList var9 = Lists.newArrayList();
       StructureTemplate var10 = new StructureTemplate();
-      var10.fillFromWorld(var1, var8, new BlockPos(1, 1, 1), false, (Block)null);
+      var10.fillFromWorld(var1, var8, new Vec3i(1, 1, 1), false, (Block)null);
       SinglePoolElement var11 = new SinglePoolElement(var10);
-      PoolElementStructurePiece var12 = new PoolElementStructurePiece(var5, var11, var8, 1, Rotation.NONE, new BoundingBox(var8, var8));
-      JigsawPlacement.addPieces(var1.registryAccess(), var12, var2, PoolElementStructurePiece::new, var4, var5, var9, var7);
+      PoolElementStructurePiece var12 = new PoolElementStructurePiece(var5, var11, var8, 1, Rotation.NONE, new BoundingBox(var8));
+      JigsawPlacement.addPieces(var1.registryAccess(), var12, var2, PoolElementStructurePiece::new, var4, var5, var9, var7, var1);
       Iterator var13 = var9.iterator();
 
       while(var13.hasNext()) {
          PoolElementStructurePiece var14 = (PoolElementStructurePiece)var13.next();
          var14.place(var1, var6, var4, var7, BoundingBox.infinite(), var8, var3);
@@ -134,7 +155,11 @@
       public static Optional<JigsawBlockEntity.JointType> byName(String var0) {
          return Arrays.stream(values()).filter((var1) -> {
             return var1.getSerializedName().equals(var0);
          }).findFirst();
       }
+
+      public Component getTranslatedName() {
+         return new TranslatableComponent("jigsaw_block.joint." + this.name);
+      }
    }
 }
