--- net/minecraft/world/level/block/entity/package-info.java
+++ net/minecraft/world/level/block/entity/package-info.java
@@ -1,0 +1,12 @@
+package net.minecraft.world.level.block.entity;
+
+import com.mojang.math.FieldsAreNonnullByDefault;
+import javax.annotation.ParametersAreNonnullByDefault;
+import net.minecraft.MethodsReturnNonnullByDefault;
+
+// $FF: synthetic class
+@ParametersAreNonnullByDefault
+@MethodsReturnNonnullByDefault
+@FieldsAreNonnullByDefault
+interface package-info {
+}
