--- net/minecraft/world/level/block/entity/SmokerBlockEntity.java
+++ net/minecraft/world/level/block/entity/SmokerBlockEntity.java
@@ -1,18 +1,20 @@
 package net.minecraft.world.level.block.entity;
 
+import net.minecraft.core.BlockPos;
 import net.minecraft.network.chat.Component;
 import net.minecraft.network.chat.TranslatableComponent;
 import net.minecraft.world.entity.player.Inventory;
 import net.minecraft.world.inventory.AbstractContainerMenu;
 import net.minecraft.world.inventory.SmokerMenu;
 import net.minecraft.world.item.ItemStack;
 import net.minecraft.world.item.crafting.RecipeType;
+import net.minecraft.world.level.block.state.BlockState;
 
 public class SmokerBlockEntity extends AbstractFurnaceBlockEntity {
-   public SmokerBlockEntity() {
-      super(BlockEntityType.SMOKER, RecipeType.SMOKING);
+   public SmokerBlockEntity(BlockPos var1, BlockState var2) {
+      super(BlockEntityType.SMOKER, var1, var2, RecipeType.SMOKING);
    }
 
    protected Component getDefaultName() {
       return new TranslatableComponent("container.smoker");
    }
