--- net/minecraft/data/models/model/TextureMapping.java
+++ net/minecraft/data/models/model/TextureMapping.java
@@ -7,10 +7,11 @@
 import java.util.stream.Stream;
 import net.minecraft.core.Registry;
 import net.minecraft.resources.ResourceLocation;
 import net.minecraft.world.item.Item;
 import net.minecraft.world.level.block.Block;
+import net.minecraft.world.level.block.Blocks;
 
 public class TextureMapping {
    private final Map<TextureSlot, ResourceLocation> slots = Maps.newHashMap();
    private final Set<TextureSlot> forcedSlots = Sets.newHashSet();
 
@@ -20,14 +21,25 @@
    public TextureMapping put(TextureSlot var1, ResourceLocation var2) {
       this.slots.put(var1, var2);
       return this;
    }
 
+   public TextureMapping putForced(TextureSlot var1, ResourceLocation var2) {
+      this.slots.put(var1, var2);
+      this.forcedSlots.add(var1);
+      return this;
+   }
+
    public Stream<TextureSlot> getForced() {
       return this.forcedSlots.stream();
    }
 
+   public TextureMapping copySlot(TextureSlot var1, TextureSlot var2) {
+      this.slots.put(var2, this.slots.get(var1));
+      return this;
+   }
+
    public TextureMapping copyForced(TextureSlot var1, TextureSlot var2) {
       this.slots.put(var2, this.slots.get(var1));
       this.forcedSlots.add(var2);
       return this;
    }
@@ -95,10 +107,14 @@
 
    public static TextureMapping wool(Block var0) {
       return singleSlot(TextureSlot.WOOL, getBlockTexture(var0));
    }
 
+   public static TextureMapping wool(ResourceLocation var0) {
+      return singleSlot(TextureSlot.WOOL, var0);
+   }
+
    public static TextureMapping stem(Block var0) {
       return singleSlot(TextureSlot.STEM, getBlockTexture(var0));
    }
 
    public static TextureMapping attachedStem(Block var0, Block var1) {
@@ -153,10 +169,14 @@
    public static TextureMapping columnWithWall(Block var0) {
       ResourceLocation var1 = getBlockTexture(var0);
       return (new TextureMapping()).put(TextureSlot.WALL, var1).put(TextureSlot.SIDE, var1).put(TextureSlot.END, getBlockTexture(var0, "_top"));
    }
 
+   public static TextureMapping door(ResourceLocation var0, ResourceLocation var1) {
+      return (new TextureMapping()).put(TextureSlot.TOP, var0).put(TextureSlot.BOTTOM, var1);
+   }
+
    public static TextureMapping door(Block var0) {
       return (new TextureMapping()).put(TextureSlot.TOP, getBlockTexture(var0, "_top")).put(TextureSlot.BOTTOM, getBlockTexture(var0, "_bottom"));
    }
 
    public static TextureMapping particle(Block var0) {
@@ -221,10 +241,18 @@
 
    public static TextureMapping campfire(Block var0) {
       return (new TextureMapping()).put(TextureSlot.LIT_LOG, getBlockTexture(var0, "_log_lit")).put(TextureSlot.FIRE, getBlockTexture(var0, "_fire"));
    }
 
+   public static TextureMapping candleCake(Block var0) {
+      return (new TextureMapping()).put(TextureSlot.PARTICLE, getBlockTexture(Blocks.CAKE, "_side")).put(TextureSlot.BOTTOM, getBlockTexture(Blocks.CAKE, "_bottom")).put(TextureSlot.TOP, getBlockTexture(Blocks.CAKE, "_top")).put(TextureSlot.SIDE, getBlockTexture(Blocks.CAKE, "_side")).put(TextureSlot.CANDLE, getBlockTexture(var0));
+   }
+
+   public static TextureMapping cauldron(ResourceLocation var0) {
+      return (new TextureMapping()).put(TextureSlot.PARTICLE, getBlockTexture(Blocks.CAULDRON, "_side")).put(TextureSlot.SIDE, getBlockTexture(Blocks.CAULDRON, "_side")).put(TextureSlot.TOP, getBlockTexture(Blocks.CAULDRON, "_top")).put(TextureSlot.BOTTOM, getBlockTexture(Blocks.CAULDRON, "_bottom")).put(TextureSlot.INSIDE, getBlockTexture(Blocks.CAULDRON, "_inner")).put(TextureSlot.CONTENT, var0);
+   }
+
    public static TextureMapping layer0(Item var0) {
       return (new TextureMapping()).put(TextureSlot.LAYER0, getItemTexture(var0));
    }
 
    public static TextureMapping layer0(Block var0) {
