--- net/minecraft/advancements/critereon/AbstractCriterionTriggerInstance.java
+++ net/minecraft/advancements/critereon/AbstractCriterionTriggerInstance.java
@@ -26,8 +26,8 @@
       var2.add("player", this.player.toJson(var1));
       return var2;
    }
 
    public String toString() {
-      return "AbstractCriterionInstance{criterion=" + this.criterion + '}';
+      return "AbstractCriterionInstance{criterion=" + this.criterion + "}";
    }
 }
