--- net/minecraft/advancements/critereon/BrewedPotionTrigger.java
+++ net/minecraft/advancements/critereon/BrewedPotionTrigger.java
@@ -8,11 +8,11 @@
 import net.minecraft.server.level.ServerPlayer;
 import net.minecraft.util.GsonHelper;
 import net.minecraft.world.item.alchemy.Potion;
 
 public class BrewedPotionTrigger extends SimpleCriterionTrigger<BrewedPotionTrigger.TriggerInstance> {
-   private static final ResourceLocation ID = new ResourceLocation("brewed_potion");
+   static final ResourceLocation ID = new ResourceLocation("brewed_potion");
 
    public BrewedPotionTrigger() {
    }
 
    public ResourceLocation getId() {
