--- net/minecraft/advancements/critereon/ConsumeItemTrigger.java
+++ net/minecraft/advancements/critereon/ConsumeItemTrigger.java
@@ -38,10 +38,14 @@
 
       public static ConsumeItemTrigger.TriggerInstance usedItem() {
          return new ConsumeItemTrigger.TriggerInstance(EntityPredicate.Composite.ANY, ItemPredicate.ANY);
       }
 
+      public static ConsumeItemTrigger.TriggerInstance usedItem(ItemPredicate var0) {
+         return new ConsumeItemTrigger.TriggerInstance(EntityPredicate.Composite.ANY, var0);
+      }
+
       public static ConsumeItemTrigger.TriggerInstance usedItem(ItemLike var0) {
          return new ConsumeItemTrigger.TriggerInstance(EntityPredicate.Composite.ANY, new ItemPredicate((Tag)null, var0.asItem(), MinMaxBounds.Ints.ANY, MinMaxBounds.Ints.ANY, EnchantmentPredicate.NONE, EnchantmentPredicate.NONE, (Potion)null, NbtPredicate.ANY));
       }
 
       public boolean matches(ItemStack var1) {
