--- net/minecraft/advancements/critereon/CuredZombieVillagerTrigger.java
+++ net/minecraft/advancements/critereon/CuredZombieVillagerTrigger.java
@@ -6,11 +6,11 @@
 import net.minecraft.world.entity.monster.Zombie;
 import net.minecraft.world.entity.npc.Villager;
 import net.minecraft.world.level.storage.loot.LootContext;
 
 public class CuredZombieVillagerTrigger extends SimpleCriterionTrigger<CuredZombieVillagerTrigger.TriggerInstance> {
-   private static final ResourceLocation ID = new ResourceLocation("cured_zombie_villager");
+   static final ResourceLocation ID = new ResourceLocation("cured_zombie_villager");
 
    public CuredZombieVillagerTrigger() {
    }
 
    public ResourceLocation getId() {
