--- net/minecraft/advancements/critereon/DeserializationContext.java
+++ net/minecraft/advancements/critereon/DeserializationContext.java
@@ -1,10 +1,11 @@
 package net.minecraft.advancements.critereon;
 
 import com.google.gson.Gson;
 import com.google.gson.JsonArray;
 import com.google.gson.JsonElement;
+import java.util.Objects;
 import net.minecraft.resources.ResourceLocation;
 import net.minecraft.world.level.storage.loot.Deserializers;
 import net.minecraft.world.level.storage.loot.PredicateManager;
 import net.minecraft.world.level.storage.loot.ValidationContext;
 import net.minecraft.world.level.storage.loot.parameters.LootContextParamSet;
@@ -23,11 +24,13 @@
       this.predicateManager = var2;
    }
 
    public final LootItemCondition[] deserializeConditions(JsonArray var1, String var2, LootContextParamSet var3) {
       LootItemCondition[] var4 = (LootItemCondition[])this.predicateGson.fromJson((JsonElement)var1, (Class)LootItemCondition[].class);
-      ValidationContext var5 = new ValidationContext(var3, this.predicateManager::get, (var0) -> {
+      PredicateManager var10003 = this.predicateManager;
+      Objects.requireNonNull(var10003);
+      ValidationContext var5 = new ValidationContext(var3, var10003::get, (var0) -> {
          return null;
       });
       LootItemCondition[] var6 = var4;
       int var7 = var4.length;
 
