--- net/minecraft/advancements/critereon/EnterBlockTrigger.java
+++ net/minecraft/advancements/critereon/EnterBlockTrigger.java
@@ -9,11 +9,11 @@
 import net.minecraft.util.GsonHelper;
 import net.minecraft.world.level.block.Block;
 import net.minecraft.world.level.block.state.BlockState;
 
 public class EnterBlockTrigger extends SimpleCriterionTrigger<EnterBlockTrigger.TriggerInstance> {
-   private static final ResourceLocation ID = new ResourceLocation("enter_block");
+   static final ResourceLocation ID = new ResourceLocation("enter_block");
 
    public EnterBlockTrigger() {
    }
 
    public ResourceLocation getId() {
