--- net/minecraft/advancements/critereon/EntityHurtPlayerTrigger.java
+++ net/minecraft/advancements/critereon/EntityHurtPlayerTrigger.java
@@ -32,10 +32,18 @@
       public TriggerInstance(EntityPredicate.Composite var1, DamagePredicate var2) {
          super(EntityHurtPlayerTrigger.ID, var1);
          this.damage = var2;
       }
 
+      public static EntityHurtPlayerTrigger.TriggerInstance entityHurtPlayer() {
+         return new EntityHurtPlayerTrigger.TriggerInstance(EntityPredicate.Composite.ANY, DamagePredicate.ANY);
+      }
+
+      public static EntityHurtPlayerTrigger.TriggerInstance entityHurtPlayer(DamagePredicate var0) {
+         return new EntityHurtPlayerTrigger.TriggerInstance(EntityPredicate.Composite.ANY, var0);
+      }
+
       public static EntityHurtPlayerTrigger.TriggerInstance entityHurtPlayer(DamagePredicate.Builder var0) {
          return new EntityHurtPlayerTrigger.TriggerInstance(EntityPredicate.Composite.ANY, var0.build());
       }
 
       public boolean matches(ServerPlayer var1, DamageSource var2, float var3, float var4, boolean var5) {
