--- net/minecraft/advancements/critereon/EntityTypePredicate.java
+++ net/minecraft/advancements/critereon/EntityTypePredicate.java
@@ -5,13 +5,15 @@
 import com.google.gson.JsonNull;
 import com.google.gson.JsonPrimitive;
 import com.google.gson.JsonSyntaxException;
 import javax.annotation.Nullable;
 import net.minecraft.core.Registry;
+import net.minecraft.resources.ResourceKey;
 import net.minecraft.resources.ResourceLocation;
 import net.minecraft.tags.SerializationTags;
 import net.minecraft.tags.Tag;
+import net.minecraft.tags.TagContainer;
 import net.minecraft.util.GsonHelper;
 import net.minecraft.world.entity.EntityType;
 
 public abstract class EntityTypePredicate {
    public static final EntityTypePredicate ANY = new EntityTypePredicate() {
@@ -36,11 +38,13 @@
       if (var0 != null && !var0.isJsonNull()) {
          String var1 = GsonHelper.convertToString(var0, "type");
          ResourceLocation var2;
          if (var1.startsWith("#")) {
             var2 = new ResourceLocation(var1.substring(1));
-            return new EntityTypePredicate.TagPredicate(SerializationTags.getInstance().getEntityTypes().getTagOrEmpty(var2));
+            return new EntityTypePredicate.TagPredicate(SerializationTags.getInstance().getTagOrThrow(Registry.ENTITY_TYPE_REGISTRY, var2, (var0x) -> {
+               return new JsonSyntaxException("Unknown entity tag '" + var0x + "'");
+            }));
          } else {
             var2 = new ResourceLocation(var1);
             EntityType var3 = (EntityType)Registry.ENTITY_TYPE.getOptional(var2).orElseThrow(() -> {
                return new JsonSyntaxException("Unknown entity type '" + var2 + "', valid types are: " + COMMA_JOINER.join((Iterable)Registry.ENTITY_TYPE.keySet()));
             });
@@ -57,23 +61,28 @@
 
    public static EntityTypePredicate of(Tag<EntityType<?>> var0) {
       return new EntityTypePredicate.TagPredicate(var0);
    }
 
-   static class TagPredicate extends EntityTypePredicate {
+   private static class TagPredicate extends EntityTypePredicate {
       private final Tag<EntityType<?>> tag;
 
       public TagPredicate(Tag<EntityType<?>> var1) {
          this.tag = var1;
       }
 
       public boolean matches(EntityType<?> var1) {
-         return this.tag.contains(var1);
+         return var1.is(this.tag);
       }
 
       public JsonElement serializeToJson() {
-         return new JsonPrimitive("#" + SerializationTags.getInstance().getEntityTypes().getIdOrThrow(this.tag));
+         TagContainer var10002 = SerializationTags.getInstance();
+         ResourceKey var10003 = Registry.ENTITY_TYPE_REGISTRY;
+         Tag var10004 = this.tag;
+         return new JsonPrimitive("#" + var10002.getIdOrThrow(var10003, var10004, () -> {
+            return new IllegalStateException("Unknown entity type tag");
+         }));
       }
    }
 
    static class TypePredicate extends EntityTypePredicate {
       private final EntityType<?> type;
