--- net/minecraft/advancements/critereon/EntityFlagsPredicate.java
+++ net/minecraft/advancements/critereon/EntityFlagsPredicate.java
@@ -105,10 +105,25 @@
       public EntityFlagsPredicate.Builder setOnFire(@Nullable Boolean var1) {
          this.isOnFire = var1;
          return this;
       }
 
+      public EntityFlagsPredicate.Builder setCrouching(@Nullable Boolean var1) {
+         this.isCrouching = var1;
+         return this;
+      }
+
+      public EntityFlagsPredicate.Builder setSprinting(@Nullable Boolean var1) {
+         this.isSprinting = var1;
+         return this;
+      }
+
+      public EntityFlagsPredicate.Builder setSwimming(@Nullable Boolean var1) {
+         this.isSwimming = var1;
+         return this;
+      }
+
       public EntityFlagsPredicate.Builder setIsBaby(@Nullable Boolean var1) {
          this.isBaby = var1;
          return this;
       }
 
