--- net/minecraft/advancements/critereon/FishingHookPredicate.java
+++ net/minecraft/advancements/critereon/FishingHookPredicate.java
@@ -9,11 +9,11 @@
 import net.minecraft.world.entity.Entity;
 import net.minecraft.world.entity.projectile.FishingHook;
 
 public class FishingHookPredicate {
    public static final FishingHookPredicate ANY = new FishingHookPredicate(false);
-   private boolean inOpenWater;
+   private final boolean inOpenWater;
 
    private FishingHookPredicate(boolean var1) {
       this.inOpenWater = var1;
    }
 
