--- net/minecraft/advancements/critereon/ImpossibleTrigger.java
+++ net/minecraft/advancements/critereon/ImpossibleTrigger.java
@@ -5,11 +5,11 @@
 import net.minecraft.advancements.CriterionTriggerInstance;
 import net.minecraft.resources.ResourceLocation;
 import net.minecraft.server.PlayerAdvancements;
 
 public class ImpossibleTrigger implements CriterionTrigger<ImpossibleTrigger.TriggerInstance> {
-   private static final ResourceLocation ID = new ResourceLocation("impossible");
+   static final ResourceLocation ID = new ResourceLocation("impossible");
 
    public ImpossibleTrigger() {
    }
 
    public ResourceLocation getId() {
