--- net/minecraft/advancements/critereon/ItemUsedOnBlockTrigger.java
+++ net/minecraft/advancements/critereon/ItemUsedOnBlockTrigger.java
@@ -7,11 +7,11 @@
 import net.minecraft.server.level.ServerPlayer;
 import net.minecraft.world.item.ItemStack;
 import net.minecraft.world.level.block.state.BlockState;
 
 public class ItemUsedOnBlockTrigger extends SimpleCriterionTrigger<ItemUsedOnBlockTrigger.TriggerInstance> {
-   private static final ResourceLocation ID = new ResourceLocation("item_used_on_block");
+   static final ResourceLocation ID = new ResourceLocation("item_used_on_block");
 
    public ItemUsedOnBlockTrigger() {
    }
 
    public ResourceLocation getId() {
