--- net/minecraft/advancements/critereon/KilledTrigger.java
+++ net/minecraft/advancements/critereon/KilledTrigger.java
@@ -38,26 +38,66 @@
          super(var1, var2);
          this.entityPredicate = var3;
          this.killingBlow = var4;
       }
 
+      public static KilledTrigger.TriggerInstance playerKilledEntity(EntityPredicate var0) {
+         return new KilledTrigger.TriggerInstance(CriteriaTriggers.PLAYER_KILLED_ENTITY.id, EntityPredicate.Composite.ANY, EntityPredicate.Composite.wrap(var0), DamageSourcePredicate.ANY);
+      }
+
       public static KilledTrigger.TriggerInstance playerKilledEntity(EntityPredicate.Builder var0) {
          return new KilledTrigger.TriggerInstance(CriteriaTriggers.PLAYER_KILLED_ENTITY.id, EntityPredicate.Composite.ANY, EntityPredicate.Composite.wrap(var0.build()), DamageSourcePredicate.ANY);
       }
 
       public static KilledTrigger.TriggerInstance playerKilledEntity() {
          return new KilledTrigger.TriggerInstance(CriteriaTriggers.PLAYER_KILLED_ENTITY.id, EntityPredicate.Composite.ANY, EntityPredicate.Composite.ANY, DamageSourcePredicate.ANY);
       }
 
+      public static KilledTrigger.TriggerInstance playerKilledEntity(EntityPredicate var0, DamageSourcePredicate var1) {
+         return new KilledTrigger.TriggerInstance(CriteriaTriggers.PLAYER_KILLED_ENTITY.id, EntityPredicate.Composite.ANY, EntityPredicate.Composite.wrap(var0), var1);
+      }
+
+      public static KilledTrigger.TriggerInstance playerKilledEntity(EntityPredicate.Builder var0, DamageSourcePredicate var1) {
+         return new KilledTrigger.TriggerInstance(CriteriaTriggers.PLAYER_KILLED_ENTITY.id, EntityPredicate.Composite.ANY, EntityPredicate.Composite.wrap(var0.build()), var1);
+      }
+
+      public static KilledTrigger.TriggerInstance playerKilledEntity(EntityPredicate var0, DamageSourcePredicate.Builder var1) {
+         return new KilledTrigger.TriggerInstance(CriteriaTriggers.PLAYER_KILLED_ENTITY.id, EntityPredicate.Composite.ANY, EntityPredicate.Composite.wrap(var0), var1.build());
+      }
+
       public static KilledTrigger.TriggerInstance playerKilledEntity(EntityPredicate.Builder var0, DamageSourcePredicate.Builder var1) {
          return new KilledTrigger.TriggerInstance(CriteriaTriggers.PLAYER_KILLED_ENTITY.id, EntityPredicate.Composite.ANY, EntityPredicate.Composite.wrap(var0.build()), var1.build());
       }
 
+      public static KilledTrigger.TriggerInstance entityKilledPlayer(EntityPredicate var0) {
+         return new KilledTrigger.TriggerInstance(CriteriaTriggers.ENTITY_KILLED_PLAYER.id, EntityPredicate.Composite.ANY, EntityPredicate.Composite.wrap(var0), DamageSourcePredicate.ANY);
+      }
+
+      public static KilledTrigger.TriggerInstance entityKilledPlayer(EntityPredicate.Builder var0) {
+         return new KilledTrigger.TriggerInstance(CriteriaTriggers.ENTITY_KILLED_PLAYER.id, EntityPredicate.Composite.ANY, EntityPredicate.Composite.wrap(var0.build()), DamageSourcePredicate.ANY);
+      }
+
       public static KilledTrigger.TriggerInstance entityKilledPlayer() {
          return new KilledTrigger.TriggerInstance(CriteriaTriggers.ENTITY_KILLED_PLAYER.id, EntityPredicate.Composite.ANY, EntityPredicate.Composite.ANY, DamageSourcePredicate.ANY);
       }
 
+      public static KilledTrigger.TriggerInstance entityKilledPlayer(EntityPredicate var0, DamageSourcePredicate var1) {
+         return new KilledTrigger.TriggerInstance(CriteriaTriggers.ENTITY_KILLED_PLAYER.id, EntityPredicate.Composite.ANY, EntityPredicate.Composite.wrap(var0), var1);
+      }
+
+      public static KilledTrigger.TriggerInstance entityKilledPlayer(EntityPredicate.Builder var0, DamageSourcePredicate var1) {
+         return new KilledTrigger.TriggerInstance(CriteriaTriggers.ENTITY_KILLED_PLAYER.id, EntityPredicate.Composite.ANY, EntityPredicate.Composite.wrap(var0.build()), var1);
+      }
+
+      public static KilledTrigger.TriggerInstance entityKilledPlayer(EntityPredicate var0, DamageSourcePredicate.Builder var1) {
+         return new KilledTrigger.TriggerInstance(CriteriaTriggers.ENTITY_KILLED_PLAYER.id, EntityPredicate.Composite.ANY, EntityPredicate.Composite.wrap(var0), var1.build());
+      }
+
+      public static KilledTrigger.TriggerInstance entityKilledPlayer(EntityPredicate.Builder var0, DamageSourcePredicate.Builder var1) {
+         return new KilledTrigger.TriggerInstance(CriteriaTriggers.ENTITY_KILLED_PLAYER.id, EntityPredicate.Composite.ANY, EntityPredicate.Composite.wrap(var0.build()), var1.build());
+      }
+
       public boolean matches(ServerPlayer var1, LootContext var2, DamageSource var3) {
          return !this.killingBlow.matches(var1, var3) ? false : this.entityPredicate.matches(var2);
       }
 
       public JsonObject serializeToJson(SerializationContext var1) {
