--- net/minecraft/advancements/critereon/KilledByCrossbowTrigger.java
+++ net/minecraft/advancements/critereon/KilledByCrossbowTrigger.java
@@ -11,11 +11,11 @@
 import net.minecraft.server.level.ServerPlayer;
 import net.minecraft.world.entity.Entity;
 import net.minecraft.world.level.storage.loot.LootContext;
 
 public class KilledByCrossbowTrigger extends SimpleCriterionTrigger<KilledByCrossbowTrigger.TriggerInstance> {
-   private static final ResourceLocation ID = new ResourceLocation("killed_by_crossbow");
+   static final ResourceLocation ID = new ResourceLocation("killed_by_crossbow");
 
    public KilledByCrossbowTrigger() {
    }
 
    public ResourceLocation getId() {
