--- net/minecraft/advancements/critereon/LocationPredicate.java
+++ net/minecraft/advancements/critereon/LocationPredicate.java
@@ -84,20 +84,18 @@
             return false;
          } else if (this.biome != null && (!var6 || this.biome != var7.get())) {
             return false;
          } else if (this.feature != null && (!var6 || !var1.structureFeatureManager().getStructureAt(var5, true, this.feature).isValid())) {
             return false;
-         } else if (this.smokey == null || var6 && this.smokey == CampfireBlock.isSmokeyPos(var1, var5)) {
-            if (!this.light.matches(var1, var5)) {
-               return false;
-            } else if (!this.block.matches(var1, var5)) {
-               return false;
-            } else {
-               return this.fluid.matches(var1, var5);
-            }
-         } else {
+         } else if (this.smokey != null && (!var6 || this.smokey != CampfireBlock.isSmokeyPos(var1, var5))) {
             return false;
+         } else if (!this.light.matches(var1, var5)) {
+            return false;
+         } else if (!this.block.matches(var1, var5)) {
+            return false;
+         } else {
+            return this.fluid.matches(var1, var5);
          }
       }
    }
 
    public JsonElement serializeToJson() {
@@ -209,20 +207,55 @@
 
       public static LocationPredicate.Builder location() {
          return new LocationPredicate.Builder();
       }
 
+      public LocationPredicate.Builder setX(MinMaxBounds.Floats var1) {
+         this.x = var1;
+         return this;
+      }
+
+      public LocationPredicate.Builder setY(MinMaxBounds.Floats var1) {
+         this.y = var1;
+         return this;
+      }
+
+      public LocationPredicate.Builder setZ(MinMaxBounds.Floats var1) {
+         this.z = var1;
+         return this;
+      }
+
       public LocationPredicate.Builder setBiome(@Nullable ResourceKey<Biome> var1) {
          this.biome = var1;
          return this;
       }
 
+      public LocationPredicate.Builder setFeature(@Nullable StructureFeature<?> var1) {
+         this.feature = var1;
+         return this;
+      }
+
+      public LocationPredicate.Builder setDimension(@Nullable ResourceKey<Level> var1) {
+         this.dimension = var1;
+         return this;
+      }
+
+      public LocationPredicate.Builder setLight(LightPredicate var1) {
+         this.light = var1;
+         return this;
+      }
+
       public LocationPredicate.Builder setBlock(BlockPredicate var1) {
          this.block = var1;
          return this;
       }
 
+      public LocationPredicate.Builder setFluid(FluidPredicate var1) {
+         this.fluid = var1;
+         return this;
+      }
+
       public LocationPredicate.Builder setSmokey(Boolean var1) {
          this.smokey = var1;
          return this;
       }
 
