--- net/minecraft/advancements/critereon/LootTableTrigger.java
+++ net/minecraft/advancements/critereon/LootTableTrigger.java
@@ -4,11 +4,11 @@
 import net.minecraft.resources.ResourceLocation;
 import net.minecraft.server.level.ServerPlayer;
 import net.minecraft.util.GsonHelper;
 
 public class LootTableTrigger extends SimpleCriterionTrigger<LootTableTrigger.TriggerInstance> {
-   private static final ResourceLocation ID = new ResourceLocation("player_generates_container_loot");
+   static final ResourceLocation ID = new ResourceLocation("player_generates_container_loot");
 
    public LootTableTrigger() {
    }
 
    public ResourceLocation getId() {
