--- net/minecraft/advancements/critereon/MobEffectsPredicate.java
+++ net/minecraft/advancements/critereon/MobEffectsPredicate.java
@@ -34,10 +34,15 @@
    public MobEffectsPredicate and(MobEffect var1) {
       this.effects.put(var1, new MobEffectsPredicate.MobEffectInstancePredicate());
       return this;
    }
 
+   public MobEffectsPredicate and(MobEffect var1, MobEffectsPredicate.MobEffectInstancePredicate var2) {
+      this.effects.put(var1, var2);
+      return this;
+   }
+
    public boolean matches(Entity var1) {
       if (this == ANY) {
          return true;
       } else {
          return var1 instanceof LivingEntity ? this.matches(((LivingEntity)var1).getActiveEffectsMap()) : false;
