--- net/minecraft/advancements/critereon/NetherTravelTrigger.java
+++ net/minecraft/advancements/critereon/NetherTravelTrigger.java
@@ -5,11 +5,11 @@
 import net.minecraft.server.level.ServerLevel;
 import net.minecraft.server.level.ServerPlayer;
 import net.minecraft.world.phys.Vec3;
 
 public class NetherTravelTrigger extends SimpleCriterionTrigger<NetherTravelTrigger.TriggerInstance> {
-   private static final ResourceLocation ID = new ResourceLocation("nether_travel");
+   static final ResourceLocation ID = new ResourceLocation("nether_travel");
 
    public NetherTravelTrigger() {
    }
 
    public ResourceLocation getId() {
