--- net/minecraft/advancements/critereon/PlacedBlockTrigger.java
+++ net/minecraft/advancements/critereon/PlacedBlockTrigger.java
@@ -12,11 +12,11 @@
 import net.minecraft.world.item.ItemStack;
 import net.minecraft.world.level.block.Block;
 import net.minecraft.world.level.block.state.BlockState;
 
 public class PlacedBlockTrigger extends SimpleCriterionTrigger<PlacedBlockTrigger.TriggerInstance> {
-   private static final ResourceLocation ID = new ResourceLocation("placed_block");
+   static final ResourceLocation ID = new ResourceLocation("placed_block");
 
    public PlacedBlockTrigger() {
    }
 
    public ResourceLocation getId() {
@@ -77,11 +77,11 @@
       public boolean matches(BlockState var1, BlockPos var2, ServerLevel var3, ItemStack var4) {
          if (this.block != null && !var1.is(this.block)) {
             return false;
          } else if (!this.state.matches(var1)) {
             return false;
-         } else if (!this.location.matches(var3, (float)var2.getX(), (float)var2.getY(), (float)var2.getZ())) {
+         } else if (!this.location.matches(var3, (double)var2.getX(), (double)var2.getY(), (double)var2.getZ())) {
             return false;
          } else {
             return this.item.matches(var4);
          }
       }
