--- net/minecraft/advancements/critereon/RecipeUnlockedTrigger.java
+++ net/minecraft/advancements/critereon/RecipeUnlockedTrigger.java
@@ -5,11 +5,11 @@
 import net.minecraft.server.level.ServerPlayer;
 import net.minecraft.util.GsonHelper;
 import net.minecraft.world.item.crafting.Recipe;
 
 public class RecipeUnlockedTrigger extends SimpleCriterionTrigger<RecipeUnlockedTrigger.TriggerInstance> {
-   private static final ResourceLocation ID = new ResourceLocation("recipe_unlocked");
+   static final ResourceLocation ID = new ResourceLocation("recipe_unlocked");
 
    public RecipeUnlockedTrigger() {
    }
 
    public ResourceLocation getId() {
