--- net/minecraft/advancements/critereon/ShotCrossbowTrigger.java
+++ net/minecraft/advancements/critereon/ShotCrossbowTrigger.java
@@ -5,11 +5,11 @@
 import net.minecraft.server.level.ServerPlayer;
 import net.minecraft.world.item.ItemStack;
 import net.minecraft.world.level.ItemLike;
 
 public class ShotCrossbowTrigger extends SimpleCriterionTrigger<ShotCrossbowTrigger.TriggerInstance> {
-   private static final ResourceLocation ID = new ResourceLocation("shot_crossbow");
+   static final ResourceLocation ID = new ResourceLocation("shot_crossbow");
 
    public ShotCrossbowTrigger() {
    }
 
    public ResourceLocation getId() {
@@ -33,10 +33,14 @@
       public TriggerInstance(EntityPredicate.Composite var1, ItemPredicate var2) {
          super(ShotCrossbowTrigger.ID, var1);
          this.item = var2;
       }
 
+      public static ShotCrossbowTrigger.TriggerInstance shotCrossbow(ItemPredicate var0) {
+         return new ShotCrossbowTrigger.TriggerInstance(EntityPredicate.Composite.ANY, var0);
+      }
+
       public static ShotCrossbowTrigger.TriggerInstance shotCrossbow(ItemLike var0) {
          return new ShotCrossbowTrigger.TriggerInstance(EntityPredicate.Composite.ANY, ItemPredicate.Builder.item().of(var0).build());
       }
 
       public boolean matches(ItemStack var1) {
