--- net/minecraft/advancements/critereon/SimpleCriterionTrigger.java
+++ net/minecraft/advancements/critereon/SimpleCriterionTrigger.java
@@ -58,11 +58,11 @@
 
          CriterionTrigger.Listener var8;
          while(var7.hasNext()) {
             var8 = (CriterionTrigger.Listener)var7.next();
             AbstractCriterionTriggerInstance var9 = (AbstractCriterionTriggerInstance)var8.getTriggerInstance();
-            if (var9.getPlayerPredicate().matches(var5) && var2.test(var9)) {
+            if (var2.test(var9) && var9.getPlayerPredicate().matches(var5)) {
                if (var6 == null) {
                   var6 = Lists.newArrayList();
                }
 
                var6.add(var8);
