--- net/minecraft/advancements/critereon/TameAnimalTrigger.java
+++ net/minecraft/advancements/critereon/TameAnimalTrigger.java
@@ -5,11 +5,11 @@
 import net.minecraft.server.level.ServerPlayer;
 import net.minecraft.world.entity.animal.Animal;
 import net.minecraft.world.level.storage.loot.LootContext;
 
 public class TameAnimalTrigger extends SimpleCriterionTrigger<TameAnimalTrigger.TriggerInstance> {
-   private static final ResourceLocation ID = new ResourceLocation("tame_animal");
+   static final ResourceLocation ID = new ResourceLocation("tame_animal");
 
    public TameAnimalTrigger() {
    }
 
    public ResourceLocation getId() {
