--- net/minecraft/advancements/critereon/UsedTotemTrigger.java
+++ net/minecraft/advancements/critereon/UsedTotemTrigger.java
@@ -33,10 +33,14 @@
       public TriggerInstance(EntityPredicate.Composite var1, ItemPredicate var2) {
          super(UsedTotemTrigger.ID, var1);
          this.item = var2;
       }
 
+      public static UsedTotemTrigger.TriggerInstance usedTotem(ItemPredicate var0) {
+         return new UsedTotemTrigger.TriggerInstance(EntityPredicate.Composite.ANY, var0);
+      }
+
       public static UsedTotemTrigger.TriggerInstance usedTotem(ItemLike var0) {
          return new UsedTotemTrigger.TriggerInstance(EntityPredicate.Composite.ANY, ItemPredicate.Builder.item().of(var0).build());
       }
 
       public boolean matches(ItemStack var1) {
