--- net/minecraft/commands/arguments/coordinates/LocalCoordinates.java
+++ net/minecraft/commands/arguments/coordinates/LocalCoordinates.java
@@ -7,10 +7,11 @@
 import net.minecraft.util.Mth;
 import net.minecraft.world.phys.Vec2;
 import net.minecraft.world.phys.Vec3;
 
 public class LocalCoordinates implements Coordinates {
+   public static final char PREFIX_LOCAL_COORDINATE = '^';
    private final double left;
    private final double up;
    private final double forwards;
 
    public LocalCoordinates(double var1, double var3, double var5) {
