--- net/minecraft/commands/arguments/coordinates/WorldCoordinates.java
+++ net/minecraft/commands/arguments/coordinates/WorldCoordinates.java
@@ -92,10 +92,18 @@
          var0.setCursor(var2);
          throw Vec3Argument.ERROR_NOT_COMPLETE.createWithContext(var0);
       }
    }
 
+   public static WorldCoordinates absolute(double var0, double var2, double var4) {
+      return new WorldCoordinates(new WorldCoordinate(false, var0), new WorldCoordinate(false, var2), new WorldCoordinate(false, var4));
+   }
+
+   public static WorldCoordinates absolute(Vec2 var0) {
+      return new WorldCoordinates(new WorldCoordinate(false, (double)var0.x), new WorldCoordinate(false, (double)var0.y), new WorldCoordinate(true, 0.0D));
+   }
+
    public static WorldCoordinates current() {
       return new WorldCoordinates(new WorldCoordinate(true, 0.0D), new WorldCoordinate(true, 0.0D), new WorldCoordinate(true, 0.0D));
    }
 
    public int hashCode() {
