--- net/minecraft/client/gui/components/toasts/AdvancementToast.java
+++ net/minecraft/client/gui/components/toasts/AdvancementToast.java
@@ -2,13 +2,15 @@
 
 import com.mojang.blaze3d.systems.RenderSystem;
 import com.mojang.blaze3d.vertex.PoseStack;
 import java.util.Iterator;
 import java.util.List;
+import java.util.Objects;
 import net.minecraft.advancements.Advancement;
 import net.minecraft.advancements.DisplayInfo;
 import net.minecraft.advancements.FrameType;
+import net.minecraft.client.renderer.GameRenderer;
 import net.minecraft.client.resources.sounds.SimpleSoundInstance;
 import net.minecraft.sounds.SoundEvents;
 import net.minecraft.util.FormattedCharSequence;
 import net.minecraft.util.Mth;
 
@@ -19,12 +21,13 @@
    public AdvancementToast(Advancement var1) {
       this.advancement = var1;
    }
 
    public Toast.Visibility render(PoseStack var1, ToastComponent var2, long var3) {
-      var2.getMinecraft().getTextureManager().bind(TEXTURE);
-      RenderSystem.color3f(1.0F, 1.0F, 1.0F);
+      RenderSystem.setShader(GameRenderer::getPositionTexShader);
+      RenderSystem.setShaderTexture(0, TEXTURE);
+      RenderSystem.setShaderColor(1.0F, 1.0F, 1.0F, 1.0F);
       DisplayInfo var5 = this.advancement.getDisplay();
       var2.blit(var1, 0, 0, 0, 0, this.width(), this.height());
       if (var5 != null) {
          List var6 = var2.getMinecraft().font.split(var5.getTitle(), 125);
          int var7 = var5.getFrame() == FrameType.CHALLENGE ? 16746751 : 16776960;
@@ -40,17 +43,17 @@
                var2.getMinecraft().font.draw(var1, var5.getFrame().getDisplayName(), 30.0F, 11.0F, var7 | var10);
             } else {
                var10 = Mth.floor(Mth.clamp((float)(var3 - 1500L) / 300.0F, 0.0F, 1.0F) * 252.0F) << 24 | 67108864;
                int var10000 = this.height() / 2;
                int var10001 = var6.size();
-               var2.getMinecraft().font.getClass();
+               Objects.requireNonNull(var2.getMinecraft().font);
                int var11 = var10000 - var10001 * 9 / 2;
 
                for(Iterator var12 = var6.iterator(); var12.hasNext(); var11 += 9) {
                   FormattedCharSequence var13 = (FormattedCharSequence)var12.next();
                   var2.getMinecraft().font.draw(var1, var13, 30.0F, (float)var11, 16777215 | var10);
-                  var2.getMinecraft().font.getClass();
+                  Objects.requireNonNull(var2.getMinecraft().font);
                }
             }
          }
 
          if (!this.playedSound && var3 > 0L) {
