--- net/minecraft/client/gui/components/toasts/AdvancementToast.java
+++ net/minecraft/client/gui/components/toasts/AdvancementToast.java
@@ -5,10 +5,11 @@
 import java.util.Iterator;
 import java.util.List;
 import net.minecraft.advancements.Advancement;
 import net.minecraft.advancements.DisplayInfo;
 import net.minecraft.advancements.FrameType;
+import net.minecraft.client.renderer.GameRenderer;
 import net.minecraft.client.resources.sounds.SimpleSoundInstance;
 import net.minecraft.sounds.SoundEvents;
 import net.minecraft.util.FormattedCharSequence;
 import net.minecraft.util.Mth;
 
@@ -19,12 +20,13 @@
    public AdvancementToast(Advancement var1) {
       this.advancement = var1;
    }
 
    public Toast.Visibility render(PoseStack var1, ToastComponent var2, long var3) {
-      var2.getMinecraft().getTextureManager().bind(TEXTURE);
-      RenderSystem.color3f(1.0F, 1.0F, 1.0F);
+      RenderSystem.setShader(GameRenderer::getPositionTexShader);
+      RenderSystem.setShaderTexture(0, TEXTURE);
+      RenderSystem.setShaderColor(1.0F, 1.0F, 1.0F);
       DisplayInfo var5 = this.advancement.getDisplay();
       var2.blit(var1, 0, 0, 0, 0, this.width(), this.height());
       if (var5 != null) {
          List var6 = var2.getMinecraft().font.split(var5.getTitle(), 125);
          int var7 = var5.getFrame() == FrameType.CHALLENGE ? 16746751 : 16776960;
