--- net/minecraft/client/gui/components/toasts/SystemToast.java
+++ net/minecraft/client/gui/components/toasts/SystemToast.java
@@ -53,12 +53,12 @@
       if (this.changed) {
          this.lastChanged = var3;
          this.changed = false;
       }
 
-      var2.getMinecraft().getTextureManager().bind(TEXTURE);
-      RenderSystem.color3f(1.0F, 1.0F, 1.0F);
+      RenderSystem.setShaderTexture(0, TEXTURE);
+      RenderSystem.setShaderColor(1.0F, 1.0F, 1.0F);
       int var5 = this.width();
       boolean var6 = true;
       int var7;
       if (var5 == 160 && this.messageLines.size() <= 1) {
          var2.blit(var1, 0, 0, 0, 64, var5, this.height());
