--- net/minecraft/client/gui/components/toasts/SystemToast.java
+++ net/minecraft/client/gui/components/toasts/SystemToast.java
@@ -2,20 +2,23 @@
 
 import com.google.common.collect.ImmutableList;
 import com.mojang.blaze3d.systems.RenderSystem;
 import com.mojang.blaze3d.vertex.PoseStack;
 import java.util.List;
+import java.util.Objects;
 import java.util.stream.Stream;
 import javax.annotation.Nullable;
 import net.minecraft.client.Minecraft;
 import net.minecraft.client.gui.Font;
 import net.minecraft.network.chat.Component;
 import net.minecraft.network.chat.TextComponent;
 import net.minecraft.network.chat.TranslatableComponent;
 import net.minecraft.util.FormattedCharSequence;
 
 public class SystemToast implements Toast {
+   private static final long DISPLAY_TIME = 5000L;
+   private static final int MAX_LINE_SIZE = 200;
    private final SystemToast.SystemToastIds id;
    private Component title;
    private List<FormattedCharSequence> messageLines;
    private long lastChanged;
    private boolean changed;
@@ -27,11 +30,11 @@
 
    public static SystemToast multiline(Minecraft var0, SystemToast.SystemToastIds var1, Component var2, Component var3) {
       Font var4 = var0.font;
       List var5 = var4.split(var3, 200);
       Stream var10001 = var5.stream();
-      var4.getClass();
+      Objects.requireNonNull(var4);
       int var6 = Math.max(200, var10001.mapToInt(var4::width).max().orElse(200));
       return new SystemToast(var1, var2, var5, var6 + 30);
    }
 
    private SystemToast(SystemToast.SystemToastIds var1, Component var2, List<FormattedCharSequence> var3, int var4) {
@@ -53,12 +56,12 @@
       if (this.changed) {
          this.lastChanged = var3;
          this.changed = false;
       }
 
-      var2.getMinecraft().getTextureManager().bind(TEXTURE);
-      RenderSystem.color3f(1.0F, 1.0F, 1.0F);
+      RenderSystem.setShaderTexture(0, TEXTURE);
+      RenderSystem.setShaderColor(1.0F, 1.0F, 1.0F, 1.0F);
       int var5 = this.width();
       boolean var6 = true;
       int var7;
       if (var5 == 160 && this.messageLines.size() <= 1) {
          var2.blit(var1, 0, 0, 0, 64, var5, this.height());
