--- net/minecraft/client/gui/components/toasts/SystemToast.java
+++ net/minecraft/client/gui/components/toasts/SystemToast.java
@@ -12,10 +12,12 @@
 import net.minecraft.network.chat.TextComponent;
 import net.minecraft.network.chat.TranslatableComponent;
 import net.minecraft.util.FormattedCharSequence;
 
 public class SystemToast implements Toast {
+   private static final long DISPLAY_TIME = 5000L;
+   private static final int MAX_LINE_SIZE = 200;
    private final SystemToast.SystemToastIds id;
    private Component title;
    private List<FormattedCharSequence> messageLines;
    private long lastChanged;
    private boolean changed;
@@ -53,12 +55,12 @@
       if (this.changed) {
          this.lastChanged = var3;
          this.changed = false;
       }
 
-      var2.getMinecraft().getTextureManager().bind(TEXTURE);
-      RenderSystem.color3f(1.0F, 1.0F, 1.0F);
+      RenderSystem.setShaderTexture(0, TEXTURE);
+      RenderSystem.setShaderColor(1.0F, 1.0F, 1.0F, 1.0F);
       int var5 = this.width();
       boolean var6 = true;
       int var7;
       if (var5 == 160 && this.messageLines.size() <= 1) {
          var2.blit(var1, 0, 0, 0, 64, var5, this.height());
