--- net/minecraft/client/gui/components/toasts/ToastComponent.java
+++ net/minecraft/client/gui/components/toasts/ToastComponent.java
@@ -108,17 +108,20 @@
 
          if (this.visibility == Toast.Visibility.SHOW && var4 - this.animationTime <= 600L) {
             this.visibleTime = var4;
          }
 
-         RenderSystem.pushMatrix();
-         RenderSystem.translatef((float)var1 - (float)this.toast.width() * this.getVisibility(var4), (float)(var2 * this.toast.height()), (float)(800 + var2));
-         Toast.Visibility var6 = this.toast.render(var3, ToastComponent.this, var4 - this.visibleTime);
-         RenderSystem.popMatrix();
-         if (var6 != this.visibility) {
+         PoseStack var6 = RenderSystem.getModelViewStack();
+         var6.pushPose();
+         var6.translate((double)((float)var1 - (float)this.toast.width() * this.getVisibility(var4)), (double)(var2 * this.toast.height()), (double)(800 + var2));
+         RenderSystem.applyModelViewMatrix();
+         Toast.Visibility var7 = this.toast.render(var3, ToastComponent.this, var4 - this.visibleTime);
+         var6.popPose();
+         RenderSystem.applyModelViewMatrix();
+         if (var7 != this.visibility) {
             this.animationTime = var4 - (long)((int)((1.0F - this.getVisibility(var4)) * 600.0F));
-            this.visibility = var6;
+            this.visibility = var7;
             this.visibility.playSound(ToastComponent.this.minecraft.getSoundManager());
          }
 
          return this.visibility == Toast.Visibility.HIDE && var4 - this.animationTime > 600L;
       }
