--- net/minecraft/client/gui/components/toasts/TutorialToast.java
+++ net/minecraft/client/gui/components/toasts/TutorialToast.java
@@ -6,10 +6,14 @@
 import net.minecraft.client.gui.GuiComponent;
 import net.minecraft.network.chat.Component;
 import net.minecraft.util.Mth;
 
 public class TutorialToast implements Toast {
+   public static final int PROGRESS_BAR_WIDTH = 154;
+   public static final int PROGRESS_BAR_HEIGHT = 1;
+   public static final int PROGRESS_BAR_X = 3;
+   public static final int PROGRESS_BAR_Y = 28;
    private final TutorialToast.Icons icon;
    private final Component title;
    private final Component message;
    private Toast.Visibility visibility;
    private long lastProgressTime;
@@ -24,12 +28,12 @@
       this.message = var3;
       this.progressable = var4;
    }
 
    public Toast.Visibility render(PoseStack var1, ToastComponent var2, long var3) {
-      var2.getMinecraft().getTextureManager().bind(TEXTURE);
-      RenderSystem.color3f(1.0F, 1.0F, 1.0F);
+      RenderSystem.setShaderTexture(0, TEXTURE);
+      RenderSystem.setShaderColor(1.0F, 1.0F, 1.0F, 1.0F);
       var2.blit(var1, 0, 0, 0, 96, this.width(), this.height());
       this.icon.render(var1, var2, 6, 6);
       if (this.message == null) {
          var2.getMinecraft().font.draw(var1, this.title, 30.0F, 12.0F, -11534256);
       } else {
@@ -67,11 +71,12 @@
       MOVEMENT_KEYS(0, 0),
       MOUSE(1, 0),
       TREE(2, 0),
       RECIPE_BOOK(0, 1),
       WOODEN_PLANKS(1, 1),
-      SOCIAL_INTERACTIONS(2, 1);
+      SOCIAL_INTERACTIONS(2, 1),
+      RIGHT_CLICK(3, 1);
 
       private final int x;
       private final int y;
 
       private Icons(int var3, int var4) {
