--- net/minecraft/client/gui/components/toasts/TutorialToast.java
+++ net/minecraft/client/gui/components/toasts/TutorialToast.java
@@ -67,11 +67,12 @@
       MOVEMENT_KEYS(0, 0),
       MOUSE(1, 0),
       TREE(2, 0),
       RECIPE_BOOK(0, 1),
       WOODEN_PLANKS(1, 1),
-      SOCIAL_INTERACTIONS(2, 1);
+      SOCIAL_INTERACTIONS(2, 1),
+      RIGHT_CLICK(3, 1);
 
       private final int x;
       private final int y;
 
       private Icons(int var3, int var4) {
