--- net/minecraft/client/renderer/BiomeColors.java
+++ net/minecraft/client/renderer/BiomeColors.java
@@ -12,10 +12,13 @@
    };
    public static final ColorResolver WATER_COLOR_RESOLVER = (var0, var1, var3) -> {
       return var0.getWaterColor();
    };
 
+   public BiomeColors() {
+   }
+
    private static int getAverageColor(BlockAndTintGetter var0, BlockPos var1, ColorResolver var2) {
       return var0.getBlockTint(var1, var2);
    }
 
    public static int getAverageGrassColor(BlockAndTintGetter var0, BlockPos var1) {
