--- net/minecraft/client/renderer/DimensionSpecialEffects.java
+++ net/minecraft/client/renderer/DimensionSpecialEffects.java
@@ -77,30 +77,22 @@
 
    public boolean constantAmbientLight() {
       return this.constantAmbientLight;
    }
 
-   public static class EndEffects extends DimensionSpecialEffects {
-      public EndEffects() {
-         super(Float.NaN, false, DimensionSpecialEffects.SkyType.END, true, false);
-      }
+   public static enum SkyType {
+      NONE,
+      NORMAL,
+      END;
 
-      public Vec3 getBrightnessDependentFogColor(Vec3 var1, float var2) {
-         return var1.scale(0.15000000596046448D);
+      private SkyType() {
       }
-
-      public boolean isFoggyAt(int var1, int var2) {
-         return false;
-      }
-
-      @Nullable
-      public float[] getSunriseColor(float var1, float var2) {
-         return null;
-      }
    }
 
    public static class OverworldEffects extends DimensionSpecialEffects {
+      public static final int CLOUD_LEVEL = 128;
+
       public OverworldEffects() {
          super(128.0F, true, DimensionSpecialEffects.SkyType.NORMAL, false, false);
       }
 
       public Vec3 getBrightnessDependentFogColor(Vec3 var1, float var2) {
@@ -124,14 +116,24 @@
       public boolean isFoggyAt(int var1, int var2) {
          return true;
       }
    }
 
-   public static enum SkyType {
-      NONE,
-      NORMAL,
-      END;
+   public static class EndEffects extends DimensionSpecialEffects {
+      public EndEffects() {
+         super(Float.NaN, false, DimensionSpecialEffects.SkyType.END, true, false);
+      }
 
-      private SkyType() {
+      public Vec3 getBrightnessDependentFogColor(Vec3 var1, float var2) {
+         return var1.scale(0.15000000596046448D);
       }
+
+      public boolean isFoggyAt(int var1, int var2) {
+         return false;
+      }
+
+      @Nullable
+      public float[] getSunriseColor(float var1, float var2) {
+         return null;
+      }
    }
 }
