--- net/minecraft/client/renderer/DimensionSpecialEffects.java
+++ net/minecraft/client/renderer/DimensionSpecialEffects.java
@@ -98,11 +98,11 @@
       }
    }
 
    public static class OverworldEffects extends DimensionSpecialEffects {
       public OverworldEffects() {
-         super(128.0F, true, DimensionSpecialEffects.SkyType.NORMAL, false, false);
+         super(192.0F, true, DimensionSpecialEffects.SkyType.NORMAL, false, false);
       }
 
       public Vec3 getBrightnessDependentFogColor(Vec3 var1, float var2) {
          return var1.multiply((double)(var2 * 0.94F + 0.06F), (double)(var2 * 0.94F + 0.06F), (double)(var2 * 0.91F + 0.09F));
       }
