--- net/minecraft/client/renderer/FaceInfo.java
+++ net/minecraft/client/renderer/FaceInfo.java
@@ -36,11 +36,11 @@
    public static class VertexInfo {
       public final int xFace;
       public final int yFace;
       public final int zFace;
 
-      private VertexInfo(int var1, int var2, int var3) {
+      VertexInfo(int var1, int var2, int var3) {
          this.xFace = var1;
          this.yFace = var2;
          this.zFace = var3;
       }
    }
@@ -51,10 +51,13 @@
       public static final int MAX_X;
       public static final int MIN_Z;
       public static final int MIN_Y;
       public static final int MIN_X;
 
+      public Constants() {
+      }
+
       static {
          MAX_Z = Direction.SOUTH.get3DDataValue();
          MAX_Y = Direction.UP.get3DDataValue();
          MAX_X = Direction.EAST.get3DDataValue();
          MIN_Z = Direction.NORTH.get3DDataValue();
