--- net/minecraft/client/renderer/FaceInfo.java
+++ net/minecraft/client/renderer/FaceInfo.java
@@ -51,10 +51,13 @@
       public static final int MAX_X;
       public static final int MIN_Z;
       public static final int MIN_Y;
       public static final int MIN_X;
 
+      public Constants() {
+      }
+
       static {
          MAX_Z = Direction.SOUTH.get3DDataValue();
          MAX_Y = Direction.UP.get3DDataValue();
          MAX_X = Direction.EAST.get3DDataValue();
          MIN_Z = Direction.NORTH.get3DDataValue();
