--- net/minecraft/client/renderer/ItemBlockRenderTypes.java
+++ net/minecraft/client/renderer/ItemBlockRenderTypes.java
@@ -108,10 +108,11 @@
       var0.put(Blocks.ATTACHED_PUMPKIN_STEM, var3);
       var0.put(Blocks.ATTACHED_MELON_STEM, var3);
       var0.put(Blocks.PUMPKIN_STEM, var3);
       var0.put(Blocks.MELON_STEM, var3);
       var0.put(Blocks.VINE, var3);
+      var0.put(Blocks.GLOW_LICHEN, var3);
       var0.put(Blocks.LILY_PAD, var3);
       var0.put(Blocks.NETHER_WART, var3);
       var0.put(Blocks.BREWING_STAND, var3);
       var0.put(Blocks.COCOA, var3);
       var0.put(Blocks.BEACON, var3);
@@ -218,10 +219,17 @@
       var0.put(Blocks.POTTED_WARPED_FUNGUS, var3);
       var0.put(Blocks.POTTED_CRIMSON_ROOTS, var3);
       var0.put(Blocks.POTTED_WARPED_ROOTS, var3);
       var0.put(Blocks.CRIMSON_DOOR, var3);
       var0.put(Blocks.WARPED_DOOR, var3);
+      var0.put(Blocks.POINTED_DRIPSTONE, var3);
+      var0.put(Blocks.SMALL_AMETHYST_BUD, var3);
+      var0.put(Blocks.MEDIUM_AMETHYST_BUD, var3);
+      var0.put(Blocks.LARGE_AMETHYST_BUD, var3);
+      var0.put(Blocks.AMETHYST_CLUSTER, var3);
+      var0.put(Blocks.LIGHTNING_ROD, var3);
+      var0.put(Blocks.SCULK_SENSOR, var3);
       RenderType var4 = RenderType.translucent();
       var0.put(Blocks.ICE, var4);
       var0.put(Blocks.NETHER_PORTAL, var4);
       var0.put(Blocks.WHITE_STAINED_GLASS, var4);
       var0.put(Blocks.ORANGE_STAINED_GLASS, var4);
@@ -257,10 +265,11 @@
       var0.put(Blocks.BLACK_STAINED_GLASS_PANE, var4);
       var0.put(Blocks.SLIME_BLOCK, var4);
       var0.put(Blocks.HONEY_BLOCK, var4);
       var0.put(Blocks.FROSTED_ICE, var4);
       var0.put(Blocks.BUBBLE_COLUMN, var4);
+      var0.put(Blocks.TINTED_GLASS, var4);
    });
    private static final Map<Fluid, RenderType> TYPE_BY_FLUID = (Map)Util.make(Maps.newHashMap(), (var0) -> {
       RenderType var1 = RenderType.translucent();
       var0.put(Fluids.FLOWING_WATER, var1);
       var0.put(Fluids.WATER, var1);
