--- net/minecraft/client/renderer/LightTexture.java
+++ net/minecraft/client/renderer/LightTexture.java
@@ -44,31 +44,19 @@
       this.blockLightRedFlicker = (float)((double)this.blockLightRedFlicker * 0.9D);
       this.updateLightTexture = true;
    }
 
    public void turnOffLightLayer() {
-      RenderSystem.activeTexture(33986);
-      RenderSystem.disableTexture();
-      RenderSystem.activeTexture(33984);
+      RenderSystem.setShaderTexture(2, 0);
    }
 
    public void turnOnLightLayer() {
-      RenderSystem.activeTexture(33986);
-      RenderSystem.matrixMode(5890);
-      RenderSystem.loadIdentity();
-      float var1 = 0.00390625F;
-      RenderSystem.scalef(0.00390625F, 0.00390625F, 0.00390625F);
-      RenderSystem.translatef(8.0F, 8.0F, 8.0F);
-      RenderSystem.matrixMode(5888);
-      this.minecraft.getTextureManager().bind(this.lightTextureLocation);
+      RenderSystem.setShaderTexture(2, this.lightTextureLocation);
+      this.minecraft.getTextureManager().bindForSetup(this.lightTextureLocation);
       RenderSystem.texParameter(3553, 10241, 9729);
       RenderSystem.texParameter(3553, 10240, 9729);
-      RenderSystem.texParameter(3553, 10242, 10496);
-      RenderSystem.texParameter(3553, 10243, 10496);
-      RenderSystem.color4f(1.0F, 1.0F, 1.0F, 1.0F);
-      RenderSystem.enableTexture();
-      RenderSystem.activeTexture(33984);
+      RenderSystem.setShaderColor(1.0F, 1.0F, 1.0F, 1.0F);
    }
 
    public void updateLightTexture(float var1) {
       if (this.updateLightTexture) {
          this.updateLightTexture = false;
