--- net/minecraft/client/renderer/MultiBufferSource.java
+++ net/minecraft/client/renderer/MultiBufferSource.java
@@ -55,10 +55,22 @@
 
       private BufferBuilder getBuilderRaw(RenderType var1) {
          return (BufferBuilder)this.fixedBuffers.getOrDefault(var1, this.builder);
       }
 
+      public void endLastBatch() {
+         if (this.lastState.isPresent()) {
+            RenderType var1 = (RenderType)this.lastState.get();
+            if (!this.fixedBuffers.containsKey(var1)) {
+               this.endBatch(var1);
+            }
+
+            this.lastState = Optional.empty();
+         }
+
+      }
+
       public void endBatch() {
          this.lastState.ifPresent((var1x) -> {
             VertexConsumer var2 = this.getBuffer(var1x);
             if (var2 == this.builder) {
                this.endBatch(var1x);
