--- net/minecraft/client/renderer/OutlineBufferSource.java
+++ net/minecraft/client/renderer/OutlineBufferSource.java
@@ -61,10 +61,13 @@
       }
 
       public void defaultColor(int var1, int var2, int var3, int var4) {
       }
 
+      public void unsetDefaultColor() {
+      }
+
       public VertexConsumer vertex(double var1, double var3, double var5) {
          this.x = var1;
          this.y = var3;
          this.z = var5;
          return this;
