--- net/minecraft/client/renderer/OutlineBufferSource.java
+++ net/minecraft/client/renderer/OutlineBufferSource.java
@@ -53,18 +53,21 @@
       private double y;
       private double z;
       private float u;
       private float v;
 
-      private EntityOutlineGenerator(VertexConsumer var1, int var2, int var3, int var4, int var5) {
+      EntityOutlineGenerator(VertexConsumer var1, int var2, int var3, int var4, int var5) {
          this.delegate = var1;
          super.defaultColor(var2, var3, var4, var5);
       }
 
       public void defaultColor(int var1, int var2, int var3, int var4) {
       }
 
+      public void unsetDefaultColor() {
+      }
+
       public VertexConsumer vertex(double var1, double var3, double var5) {
          this.x = var1;
          this.y = var3;
          this.z = var5;
          return this;
