--- net/minecraft/client/renderer/PostChain.java
+++ net/minecraft/client/renderer/PostChain.java
@@ -180,11 +180,12 @@
                         throw new ChainedJsonException("Render target or texture '" + var18 + "' does not exist");
                      } finally {
                         IOUtils.closeQuietly(var21);
                      }
 
-                     var1.bind(var20);
+                     RenderSystem.setShaderTexture(0, var20);
+                     var1.bindForSetup(var20);
                      AbstractTexture var22 = var1.getTexture(var20);
                      int var23 = GsonHelper.getAsInt(var14, "width");
                      int var24 = GsonHelper.getAsInt(var14, "height");
                      boolean var25 = GsonHelper.getAsBoolean(var14, "bilinear");
                      if (var25) {
@@ -309,11 +310,11 @@
       this.passes.add(this.passes.size(), var4);
       return var4;
    }
 
    private void updateOrthoMatrix() {
-      this.shaderOrthoMatrix = Matrix4f.orthographic((float)this.screenTarget.width, (float)this.screenTarget.height, 0.1F, 1000.0F);
+      this.shaderOrthoMatrix = Matrix4f.orthographic(0.0F, (float)this.screenTarget.width, (float)this.screenTarget.height, 0.0F, 0.1F, 1000.0F);
    }
 
    public void resize(int var1, int var2) {
       this.screenWidth = this.screenTarget.width;
       this.screenHeight = this.screenTarget.height;
