--- net/minecraft/client/renderer/PostChain.java
+++ net/minecraft/client/renderer/PostChain.java
@@ -27,10 +27,11 @@
 import net.minecraft.server.packs.resources.ResourceManager;
 import net.minecraft.util.GsonHelper;
 import org.apache.commons.io.IOUtils;
 
 public class PostChain implements AutoCloseable {
+   private static final String MAIN_RENDER_TARGET = "minecraft:main";
    private final RenderTarget screenTarget;
    private final ResourceManager resourceManager;
    private final String name;
    private final List<PostPass> passes = Lists.newArrayList();
    private final Map<String, RenderTarget> customRenderTargets = Maps.newHashMap();
@@ -180,11 +181,12 @@
                         throw new ChainedJsonException("Render target or texture '" + var18 + "' does not exist");
                      } finally {
                         IOUtils.closeQuietly(var21);
                      }
 
-                     var1.bind(var20);
+                     RenderSystem.setShaderTexture(0, var20);
+                     var1.bindForSetup(var20);
                      AbstractTexture var22 = var1.getTexture(var20);
                      int var23 = GsonHelper.getAsInt(var14, "width");
                      int var24 = GsonHelper.getAsInt(var14, "height");
                      boolean var25 = GsonHelper.getAsBoolean(var14, "bilinear");
                      if (var25) {
@@ -309,11 +311,11 @@
       this.passes.add(this.passes.size(), var4);
       return var4;
    }
 
    private void updateOrthoMatrix() {
-      this.shaderOrthoMatrix = Matrix4f.orthographic((float)this.screenTarget.width, (float)this.screenTarget.height, 0.1F, 1000.0F);
+      this.shaderOrthoMatrix = Matrix4f.orthographic(0.0F, (float)this.screenTarget.width, (float)this.screenTarget.height, 0.0F, 0.1F, 1000.0F);
    }
 
    public void resize(int var1, int var2) {
       this.screenWidth = this.screenTarget.width;
       this.screenHeight = this.screenTarget.height;
