--- net/minecraft/client/renderer/SpriteCoordinateExpander.java
+++ net/minecraft/client/renderer/SpriteCoordinateExpander.java
@@ -38,9 +38,17 @@
 
    public void endVertex() {
       this.delegate.endVertex();
    }
 
+   public void defaultColor(int var1, int var2, int var3, int var4) {
+      this.delegate.defaultColor(var1, var2, var3, var4);
+   }
+
+   public void unsetDefaultColor() {
+      this.delegate.unsetDefaultColor();
+   }
+
    public void vertex(float var1, float var2, float var3, float var4, float var5, float var6, float var7, float var8, float var9, int var10, int var11, float var12, float var13, float var14) {
       this.delegate.vertex(var1, var2, var3, var4, var5, var6, var7, this.sprite.getU((double)(var8 * 16.0F)), this.sprite.getV((double)(var9 * 16.0F)), var10, var11, var12, var13, var14);
    }
 }
