--- com/mojang/authlib/yggdrasil/response/ErrorResponse.java
+++ com/mojang/authlib/yggdrasil/response/ErrorResponse.java
@@ -1,0 +1,38 @@
+package com.mojang.authlib.yggdrasil.response;
+
+import java.util.Map;
+import java.util.StringJoiner;
+
+public class ErrorResponse {
+   private final String path;
+   private final String error;
+   private final String errorMessage;
+   private final Map<String, Object> details;
+
+   public ErrorResponse(String var1, String var2, String var3, Map<String, Object> var4) {
+      this.path = var1;
+      this.error = var2;
+      this.errorMessage = var3;
+      this.details = var4;
+   }
+
+   public String getPath() {
+      return this.path;
+   }
+
+   public String getError() {
+      return this.error;
+   }
+
+   public Map<String, Object> getDetails() {
+      return this.details;
+   }
+
+   public String getErrorMessage() {
+      return this.errorMessage;
+   }
+
+   public String toString() {
+      return (new StringJoiner(", ", ErrorResponse.class.getSimpleName() + "[", "]")).add("path='" + this.path + "'").add("error='" + this.error + "'").add("details=" + this.details).add("errorMessage='" + this.errorMessage + "'").toString();
+   }
+}
