--- net/minecraft/client/model/geom/ModelLayerLocation.java
+++ net/minecraft/client/model/geom/ModelLayerLocation.java
@@ -1,0 +1,34 @@
+package net.minecraft.client.model.geom;
+
+import net.minecraft.resources.ResourceLocation;
+
+public final class ModelLayerLocation {
+   private final ResourceLocation model;
+   private final String layer;
+
+   public ModelLayerLocation(ResourceLocation var1, String var2) {
+      this.model = var1;
+      this.layer = var2;
+   }
+
+   public boolean equals(Object var1) {
+      if (this == var1) {
+         return true;
+      } else if (!(var1 instanceof ModelLayerLocation)) {
+         return false;
+      } else {
+         ModelLayerLocation var2 = (ModelLayerLocation)var1;
+         return this.model.equals(var2.model) && this.layer.equals(var2.layer);
+      }
+   }
+
+   public int hashCode() {
+      int var1 = this.model.hashCode();
+      var1 = 31 * var1 + this.layer.hashCode();
+      return var1;
+   }
+
+   public String toString() {
+      return this.model + "#" + this.layer;
+   }
+}
