--- net/minecraft/client/model/geom/package-info.java
+++ net/minecraft/client/model/geom/package-info.java
@@ -1,0 +1,12 @@
+package net.minecraft.client.model.geom;
+
+import com.mojang.math.FieldsAreNonnullByDefault;
+import javax.annotation.ParametersAreNonnullByDefault;
+import net.minecraft.MethodsReturnNonnullByDefault;
+
+// $FF: synthetic class
+@ParametersAreNonnullByDefault
+@MethodsReturnNonnullByDefault
+@FieldsAreNonnullByDefault
+interface package-info {
+}
