--- net/minecraft/realms/NarrationHelper.java
+++ net/minecraft/realms/NarrationHelper.java
@@ -8,10 +8,13 @@
 import net.minecraft.network.chat.TextComponent;
 
 public class NarrationHelper {
    private static final RepeatedNarrator REPEATED_NARRATOR = new RepeatedNarrator(Duration.ofSeconds(5L));
 
+   public NarrationHelper() {
+   }
+
    public static void now(String var0) {
       NarratorChatListener var1 = NarratorChatListener.INSTANCE;
       var1.clear();
       var1.handle(ChatType.SYSTEM, new TextComponent(fixNarrationNewlines(var0)), Util.NIL_UUID);
    }
