--- net/minecraft/realms/RealmsObjectSelectionList.java
+++ net/minecraft/realms/RealmsObjectSelectionList.java
@@ -12,11 +12,11 @@
 
    public void setSelectedItem(int var1) {
       if (var1 == -1) {
          this.setSelected((AbstractSelectionList.Entry)null);
       } else if (super.getItemCount() != 0) {
-         this.setSelected(this.getEntry(var1));
+         this.setSelected((ObjectSelectionList.Entry)this.getEntry(var1));
       }
 
    }
 
    public void selectItem(int var1) {
