--- net/minecraft/realms/RealmsLabel.java
+++ net/minecraft/realms/RealmsLabel.java
@@ -1,14 +1,14 @@
 package net.minecraft.realms;
 
 import com.mojang.blaze3d.vertex.PoseStack;
 import net.minecraft.client.Minecraft;
-import net.minecraft.client.gui.components.events.GuiEventListener;
-import net.minecraft.client.gui.screens.Screen;
+import net.minecraft.client.gui.GuiComponent;
+import net.minecraft.client.gui.components.Widget;
 import net.minecraft.network.chat.Component;
 
-public class RealmsLabel implements GuiEventListener {
+public class RealmsLabel implements Widget {
    private final Component text;
    private final int x;
    private final int y;
    private final int color;
 
@@ -17,13 +17,13 @@
       this.x = var2;
       this.y = var3;
       this.color = var4;
    }
 
-   public void render(Screen var1, PoseStack var2) {
-      Screen.drawCenteredString(var2, Minecraft.getInstance().font, this.text, this.x, this.y, this.color);
+   public void render(PoseStack var1, int var2, int var3, float var4) {
+      GuiComponent.drawCenteredString(var1, Minecraft.getInstance().font, this.text, this.x, this.y, this.color);
    }
 
-   public String getText() {
-      return this.text.getString();
+   public Component getText() {
+      return this.text;
    }
 }
