--- net/minecraft/core/particles/ParticleTypes.java
+++ net/minecraft/core/particles/ParticleTypes.java
@@ -18,10 +18,11 @@
    public static final SimpleParticleType FALLING_LAVA;
    public static final SimpleParticleType LANDING_LAVA;
    public static final SimpleParticleType DRIPPING_WATER;
    public static final SimpleParticleType FALLING_WATER;
    public static final ParticleType<DustParticleOptions> DUST;
+   public static final ParticleType<DustColorTransitionOptions> DUST_COLOR_TRANSITION;
    public static final SimpleParticleType EFFECT;
    public static final SimpleParticleType ELDER_GUARDIAN;
    public static final SimpleParticleType ENCHANTED_HIT;
    public static final SimpleParticleType ENCHANT;
    public static final SimpleParticleType END_ROD;
@@ -38,10 +39,11 @@
    public static final SimpleParticleType HAPPY_VILLAGER;
    public static final SimpleParticleType COMPOSTER;
    public static final SimpleParticleType HEART;
    public static final SimpleParticleType INSTANT_EFFECT;
    public static final ParticleType<ItemParticleOption> ITEM;
+   public static final ParticleType<VibrationParticleOption> VIBRATION;
    public static final SimpleParticleType ITEM_SLIME;
    public static final SimpleParticleType ITEM_SNOWBALL;
    public static final SimpleParticleType LARGE_SMOKE;
    public static final SimpleParticleType LAVA;
    public static final SimpleParticleType MYCELIUM;
@@ -67,18 +69,28 @@
    public static final SimpleParticleType CAMPFIRE_SIGNAL_SMOKE;
    public static final SimpleParticleType DRIPPING_HONEY;
    public static final SimpleParticleType FALLING_HONEY;
    public static final SimpleParticleType LANDING_HONEY;
    public static final SimpleParticleType FALLING_NECTAR;
+   public static final SimpleParticleType FALLING_SPORE_BLOSSOM;
    public static final SimpleParticleType ASH;
    public static final SimpleParticleType CRIMSON_SPORE;
    public static final SimpleParticleType WARPED_SPORE;
+   public static final SimpleParticleType SPORE_BLOSSOM_AIR;
    public static final SimpleParticleType DRIPPING_OBSIDIAN_TEAR;
    public static final SimpleParticleType FALLING_OBSIDIAN_TEAR;
    public static final SimpleParticleType LANDING_OBSIDIAN_TEAR;
    public static final SimpleParticleType REVERSE_PORTAL;
    public static final SimpleParticleType WHITE_ASH;
+   public static final SimpleParticleType SMALL_FLAME;
+   public static final SimpleParticleType SNOWFLAKE;
+   public static final SimpleParticleType DRIPPING_DRIPSTONE_LAVA;
+   public static final SimpleParticleType FALLING_DRIPSTONE_LAVA;
+   public static final SimpleParticleType DRIPPING_DRIPSTONE_WATER;
+   public static final SimpleParticleType FALLING_DRIPSTONE_WATER;
+   public static final SimpleParticleType GLOW_SQUID_INK;
+   public static final SimpleParticleType GLOW;
    public static final Codec<ParticleOptions> CODEC;
 
    private static SimpleParticleType register(String var0, boolean var1) {
       return (SimpleParticleType)Registry.register(Registry.PARTICLE_TYPE, (String)var0, new SimpleParticleType(var1));
    }
@@ -104,10 +116,13 @@
       DRIPPING_WATER = register("dripping_water", false);
       FALLING_WATER = register("falling_water", false);
       DUST = register("dust", DustParticleOptions.DESERIALIZER, (var0) -> {
          return DustParticleOptions.CODEC;
       });
+      DUST_COLOR_TRANSITION = register("dust_color_transition", DustColorTransitionOptions.DESERIALIZER, (var0) -> {
+         return DustColorTransitionOptions.CODEC;
+      });
       EFFECT = register("effect", false);
       ELDER_GUARDIAN = register("elder_guardian", true);
       ENCHANTED_HIT = register("enchanted_hit", false);
       ENCHANT = register("enchant", false);
       END_ROD = register("end_rod", false);
@@ -124,10 +139,13 @@
       HAPPY_VILLAGER = register("happy_villager", false);
       COMPOSTER = register("composter", false);
       HEART = register("heart", false);
       INSTANT_EFFECT = register("instant_effect", false);
       ITEM = register("item", ItemParticleOption.DESERIALIZER, ItemParticleOption::codec);
+      VIBRATION = register("vibration", VibrationParticleOption.DESERIALIZER, (var0) -> {
+         return VibrationParticleOption.CODEC;
+      });
       ITEM_SLIME = register("item_slime", false);
       ITEM_SNOWBALL = register("item_snowball", false);
       LARGE_SMOKE = register("large_smoke", false);
       LAVA = register("lava", false);
       MYCELIUM = register("mycelium", false);
@@ -153,16 +171,26 @@
       CAMPFIRE_SIGNAL_SMOKE = register("campfire_signal_smoke", true);
       DRIPPING_HONEY = register("dripping_honey", false);
       FALLING_HONEY = register("falling_honey", false);
       LANDING_HONEY = register("landing_honey", false);
       FALLING_NECTAR = register("falling_nectar", false);
+      FALLING_SPORE_BLOSSOM = register("falling_spore_blossom", false);
       ASH = register("ash", false);
       CRIMSON_SPORE = register("crimson_spore", false);
       WARPED_SPORE = register("warped_spore", false);
+      SPORE_BLOSSOM_AIR = register("spore_blossom_air", false);
       DRIPPING_OBSIDIAN_TEAR = register("dripping_obsidian_tear", false);
       FALLING_OBSIDIAN_TEAR = register("falling_obsidian_tear", false);
       LANDING_OBSIDIAN_TEAR = register("landing_obsidian_tear", false);
       REVERSE_PORTAL = register("reverse_portal", false);
       WHITE_ASH = register("white_ash", false);
+      SMALL_FLAME = register("small_flame", false);
+      SNOWFLAKE = register("snowflake", false);
+      DRIPPING_DRIPSTONE_LAVA = register("dripping_dripstone_lava", false);
+      FALLING_DRIPSTONE_LAVA = register("falling_dripstone_lava", false);
+      DRIPPING_DRIPSTONE_WATER = register("dripping_dripstone_water", false);
+      FALLING_DRIPSTONE_WATER = register("falling_dripstone_water", false);
+      GLOW_SQUID_INK = register("glow_squid_ink", true);
+      GLOW = register("glow", true);
       CODEC = Registry.PARTICLE_TYPE.dispatch("type", ParticleOptions::getType, ParticleType::codec);
    }
 }
