--- net/minecraft/core/particles/SimpleParticleType.java
+++ net/minecraft/core/particles/SimpleParticleType.java
@@ -1,16 +1,15 @@
 package net.minecraft.core.particles;
 
 import com.mojang.brigadier.StringReader;
-import com.mojang.brigadier.exceptions.CommandSyntaxException;
 import com.mojang.serialization.Codec;
 import net.minecraft.core.Registry;
 import net.minecraft.network.FriendlyByteBuf;
 
 public class SimpleParticleType extends ParticleType<SimpleParticleType> implements ParticleOptions {
    private static final ParticleOptions.Deserializer<SimpleParticleType> DESERIALIZER = new ParticleOptions.Deserializer<SimpleParticleType>() {
-      public SimpleParticleType fromCommand(ParticleType<SimpleParticleType> var1, StringReader var2) throws CommandSyntaxException {
+      public SimpleParticleType fromCommand(ParticleType<SimpleParticleType> var1, StringReader var2) {
          return (SimpleParticleType)var1;
       }
 
       public SimpleParticleType fromNetwork(ParticleType<SimpleParticleType> var1, FriendlyByteBuf var2) {
          return (SimpleParticleType)var1;
