--- net/minecraft/world/entity/player/StackedContents.java
+++ net/minecraft/world/entity/player/StackedContents.java
@@ -18,10 +18,11 @@
 import net.minecraft.world.item.ItemStack;
 import net.minecraft.world.item.crafting.Ingredient;
 import net.minecraft.world.item.crafting.Recipe;
 
 public class StackedContents {
+   private static final int EMPTY = 0;
    public final Int2IntMap contents = new Int2IntOpenHashMap();
 
    public StackedContents() {
    }
 
@@ -47,25 +48,25 @@
 
    public static int getStackingIndex(ItemStack var0) {
       return Registry.ITEM.getId(var0.getItem());
    }
 
-   private boolean has(int var1) {
+   boolean has(int var1) {
       return this.contents.get(var1) > 0;
    }
 
-   private int take(int var1, int var2) {
+   int take(int var1, int var2) {
       int var3 = this.contents.get(var1);
       if (var3 >= var2) {
          this.contents.put(var1, var3 - var2);
          return var1;
       } else {
          return 0;
       }
    }
 
-   private void put(int var1, int var2) {
+   void put(int var1, int var2) {
       this.contents.put(var1, this.contents.get(var1) + var2);
    }
 
    public boolean canCraft(Recipe<?> var1, @Nullable IntList var2) {
       return this.canCraft(var1, var2, 1);
@@ -89,11 +90,11 @@
 
    public void clear() {
       this.contents.clear();
    }
 
-   class RecipePicker {
+   private class RecipePicker {
       private final Recipe<?> recipe;
       private final List<Ingredient> ingredients = Lists.newArrayList();
       private final int ingredientCount;
       private final int[] items;
       private final int itemCount;
