--- com/mojang/realmsclient/exception/package-info.java
+++ com/mojang/realmsclient/exception/package-info.java
@@ -1,0 +1,10 @@
+package com.mojang.realmsclient.exception;
+
+import javax.annotation.ParametersAreNonnullByDefault;
+import net.minecraft.MethodsReturnNonnullByDefault;
+
+// $FF: synthetic class
+@ParametersAreNonnullByDefault
+@MethodsReturnNonnullByDefault
+interface package-info {
+}
