--- com/mojang/realmsclient/exception/RealmsServiceException.java
+++ com/mojang/realmsclient/exception/RealmsServiceException.java
@@ -29,9 +29,10 @@
       if (this.errorCode == -1) {
          return "Realms (" + this.httpResultCode + ") " + this.httpResponseContent;
       } else {
          String var1 = "mco.errorMessage." + this.errorCode;
          String var2 = I18n.get(var1);
-         return (var2.equals(var1) ? this.errorMsg : var2) + " - " + this.errorCode;
+         String var10000 = var2.equals(var1) ? this.errorMsg : var2;
+         return var10000 + " - " + this.errorCode;
       }
    }
 }
