--- net/minecraft/world/level/material/FluidState.java
+++ net/minecraft/world/level/material/FluidState.java
@@ -31,10 +31,14 @@
 
    public boolean isSource() {
       return this.getType().isSource(this);
    }
 
+   public boolean isSourceOfType(Fluid var1) {
+      return this.owner == var1 && ((Fluid)this.owner).isSource(this);
+   }
+
    public boolean isEmpty() {
       return this.getType().isEmpty();
    }
 
    public float getHeight(BlockGetter var1, BlockPos var2) {
