--- net/minecraft/world/level/material/Fluids.java
+++ net/minecraft/world/level/material/Fluids.java
@@ -9,10 +9,13 @@
    public static final FlowingFluid FLOWING_WATER = (FlowingFluid)register("flowing_water", new WaterFluid.Flowing());
    public static final FlowingFluid WATER = (FlowingFluid)register("water", new WaterFluid.Source());
    public static final FlowingFluid FLOWING_LAVA = (FlowingFluid)register("flowing_lava", new LavaFluid.Flowing());
    public static final FlowingFluid LAVA = (FlowingFluid)register("lava", new LavaFluid.Source());
 
+   public Fluids() {
+   }
+
    private static <T extends Fluid> T register(String var0, T var1) {
       return (Fluid)Registry.register(Registry.FLUID, (String)var0, var1);
    }
 
    static {
