--- net/minecraft/world/level/material/FluidState.java
+++ net/minecraft/world/level/material/FluidState.java
@@ -18,10 +18,12 @@
 import net.minecraft.world.phys.Vec3;
 import net.minecraft.world.phys.shapes.VoxelShape;
 
 public final class FluidState extends StateHolder<Fluid, FluidState> {
    public static final Codec<FluidState> CODEC;
+   public static final int AMOUNT_MAX = 9;
+   public static final int AMOUNT_FULL = 8;
 
    public FluidState(Fluid var1, ImmutableMap<Property<?>, Comparable<?>> var2, MapCodec<FluidState> var3) {
       super(var1, var2, var3);
    }
 
@@ -31,10 +33,14 @@
 
    public boolean isSource() {
       return this.getType().isSource(this);
    }
 
+   public boolean isSourceOfType(Fluid var1) {
+      return this.owner == var1 && ((Fluid)this.owner).isSource(this);
+   }
+
    public boolean isEmpty() {
       return this.getType().isEmpty();
    }
 
    public float getHeight(BlockGetter var1, BlockPos var2) {
