--- net/minecraft/world/level/material/MaterialColor.java
+++ net/minecraft/world/level/material/MaterialColor.java
@@ -59,10 +59,13 @@
    public static final MaterialColor CRIMSON_HYPHAE = new MaterialColor(54, 6035741);
    public static final MaterialColor WARPED_NYLIUM = new MaterialColor(55, 1474182);
    public static final MaterialColor WARPED_STEM = new MaterialColor(56, 3837580);
    public static final MaterialColor WARPED_HYPHAE = new MaterialColor(57, 5647422);
    public static final MaterialColor WARPED_WART_BLOCK = new MaterialColor(58, 1356933);
+   public static final MaterialColor DEEPSLATE = new MaterialColor(59, 6579300);
+   public static final MaterialColor RAW_IRON = new MaterialColor(60, 14200723);
+   public static final MaterialColor GLOW_LICHEN = new MaterialColor(61, 8365974);
    public final int col;
    public final int id;
 
    private MaterialColor(int var1, int var2) {
       if (var1 >= 0 && var1 <= 63) {
