--- net/minecraft/world/level/material/WaterFluid.java
+++ net/minecraft/world/level/material/WaterFluid.java
@@ -57,11 +57,11 @@
    protected boolean canConvertToSource() {
       return true;
    }
 
    protected void beforeDestroyingBlock(LevelAccessor var1, BlockPos var2, BlockState var3) {
-      BlockEntity var4 = var3.getBlock().isEntityBlock() ? var1.getBlockEntity(var2) : null;
+      BlockEntity var4 = var3.hasBlockEntity() ? var1.getBlockEntity(var2) : null;
       Block.dropResources(var3, var1, var2, var4);
    }
 
    public int getSlopeFindDistance(LevelReader var1) {
       return 4;
