--- net/minecraft/world/level/newbiome/layer/AddEdgeLayer.java
+++ net/minecraft/world/level/newbiome/layer/AddEdgeLayer.java
@@ -3,10 +3,13 @@
 import net.minecraft.world.level.newbiome.context.Context;
 import net.minecraft.world.level.newbiome.layer.traits.C0Transformer;
 import net.minecraft.world.level.newbiome.layer.traits.CastleTransformer;
 
 public class AddEdgeLayer {
+   public AddEdgeLayer() {
+   }
+
    public static enum IntroduceSpecial implements C0Transformer {
       INSTANCE;
 
       private IntroduceSpecial() {
       }
