--- net/minecraft/client/gui/spectator/categories/TeleportToTeamMenuCategory.java
+++ net/minecraft/client/gui/spectator/categories/TeleportToTeamMenuCategory.java
@@ -52,11 +52,11 @@
    public Component getName() {
       return TELEPORT_TEXT;
    }
 
    public void renderIcon(PoseStack var1, float var2, int var3) {
-      Minecraft.getInstance().getTextureManager().bind(SpectatorGui.SPECTATOR_LOCATION);
+      RenderSystem.setShaderTexture(0, SpectatorGui.SPECTATOR_LOCATION);
       GuiComponent.blit(var1, 0, 0, 16.0F, 0.0F, 16, 16, 256, 256);
    }
 
    public boolean isEnabled() {
       Iterator var1 = this.items.iterator();
@@ -116,12 +116,12 @@
             float var6 = (float)(var4 >> 8 & 255) / 255.0F;
             float var7 = (float)(var4 & 255) / 255.0F;
             GuiComponent.fill(var1, 1, 1, 15, 15, Mth.color(var5 * var2, var6 * var2, var7 * var2) | var3 << 24);
          }
 
-         Minecraft.getInstance().getTextureManager().bind(this.location);
-         RenderSystem.color4f(var2, var2, var2, (float)var3 / 255.0F);
+         RenderSystem.setShaderTexture(0, this.location);
+         RenderSystem.setShaderColor(var2, var2, var2, (float)var3 / 255.0F);
          GuiComponent.blit(var1, 2, 2, 12, 12, 8.0F, 8.0F, 8, 8, 64, 64);
          GuiComponent.blit(var1, 2, 2, 12, 12, 40.0F, 8.0F, 8, 8, 64, 64);
       }
 
       public boolean isEnabled() {
