--- net/minecraft/world/level/chunk/ChunkGenerator.java
+++ net/minecraft/world/level/chunk/ChunkGenerator.java
@@ -13,21 +13,22 @@
 import javax.annotation.Nullable;
 import net.minecraft.CrashReport;
 import net.minecraft.CrashReportCategory;
 import net.minecraft.ReportedException;
 import net.minecraft.core.BlockPos;
+import net.minecraft.core.QuartPos;
 import net.minecraft.core.Registry;
 import net.minecraft.core.RegistryAccess;
 import net.minecraft.core.SectionPos;
 import net.minecraft.data.worldgen.StructureFeatures;
 import net.minecraft.network.protocol.game.DebugPackets;
 import net.minecraft.server.level.ServerLevel;
 import net.minecraft.server.level.WorldGenRegion;
 import net.minecraft.world.entity.MobCategory;
-import net.minecraft.world.level.BlockGetter;
 import net.minecraft.world.level.ChunkPos;
 import net.minecraft.world.level.LevelAccessor;
+import net.minecraft.world.level.NoiseColumn;
 import net.minecraft.world.level.StructureFeatureManager;
 import net.minecraft.world.level.WorldGenLevel;
 import net.minecraft.world.level.biome.Biome;
 import net.minecraft.world.level.biome.BiomeGenerationSettings;
 import net.minecraft.world.level.biome.BiomeManager;
@@ -93,14 +94,14 @@
 
             for(int var11 = 0; var11 < var18; ++var11) {
                double var12 = (double)(4 * var17 + var17 * var10 * 6) + (var6.nextDouble() - 0.5D) * (double)var17 * 2.5D;
                int var14 = (int)Math.round(Math.cos(var7) * var12);
                int var15 = (int)Math.round(Math.sin(var7) * var12);
-               BlockPos var16 = this.biomeSource.findBiomeHorizontal((var14 << 4) + 8, 0, (var15 << 4) + 8, 112, var2::contains, var6);
+               BlockPos var16 = this.biomeSource.findBiomeHorizontal(SectionPos.sectionToBlockCoord(var14, 8), 0, SectionPos.sectionToBlockCoord(var15, 8), 112, var2::contains, var6);
                if (var16 != null) {
-                  var14 = var16.getX() >> 4;
-                  var15 = var16.getZ() >> 4;
+                  var14 = SectionPos.blockToSectionCoord(var16.getX());
+                  var15 = SectionPos.blockToSectionCoord(var16.getZ());
                }
 
                this.strongholdPositions.add(new ChunkPos(var14, var15));
                var7 += 6.283185307179586D / (double)var5;
                ++var9;
@@ -121,21 +122,21 @@
 
    public abstract ChunkGenerator withSeed(long var1);
 
    public void createBiomes(Registry<Biome> var1, ChunkAccess var2) {
       ChunkPos var3 = var2.getPos();
-      ((ProtoChunk)var2).setBiomes(new ChunkBiomeContainer(var1, var3, this.runtimeBiomeSource));
+      ((ProtoChunk)var2).setBiomes(new ChunkBiomeContainer(var1, var2, var3, this.runtimeBiomeSource));
    }
 
    public void applyCarvers(long var1, BiomeManager var3, ChunkAccess var4, GenerationStep.Carving var5) {
       BiomeManager var6 = var3.withDifferentSource(this.biomeSource);
       WorldgenRandom var7 = new WorldgenRandom();
       boolean var8 = true;
       ChunkPos var9 = var4.getPos();
       int var10 = var9.x;
       int var11 = var9.z;
-      BiomeGenerationSettings var12 = this.biomeSource.getNoiseBiome(var9.x << 2, 0, var9.z << 2).getGenerationSettings();
+      BiomeGenerationSettings var12 = this.biomeSource.getNoiseBiome(QuartPos.fromBlock(var9.getMinBlockX()), 0, QuartPos.fromBlock(var9.getMinBlockZ())).getGenerationSettings();
       BitSet var13 = ((ProtoChunk)var4).getOrCreateCarvingMask(var5);
 
       for(int var14 = var10 - 8; var14 <= var10 + 8; ++var14) {
          for(int var15 = var11 - 8; var15 <= var11 + 8; ++var15) {
             List var16 = var12.getCarvers(var5);
@@ -165,11 +166,11 @@
          BlockPos.MutableBlockPos var9 = new BlockPos.MutableBlockPos();
          Iterator var10 = this.strongholdPositions.iterator();
 
          while(var10.hasNext()) {
             ChunkPos var11 = (ChunkPos)var10.next();
-            var9.set((var11.x << 4) + 8, 32, (var11.z << 4) + 8);
+            var9.set(SectionPos.sectionToBlockCoord(var11.x, 8), 32, SectionPos.sectionToBlockCoord(var11.z, 8));
             double var12 = var9.distSqr(var3);
             if (var14 == null) {
                var14 = new BlockPos(var9);
                var7 = var12;
             } else if (var12 < var7) {
@@ -186,14 +187,14 @@
    }
 
    public void applyBiomeDecoration(WorldGenRegion var1, StructureFeatureManager var2) {
       int var3 = var1.getCenterX();
       int var4 = var1.getCenterZ();
-      int var5 = var3 * 16;
-      int var6 = var4 * 16;
+      int var5 = SectionPos.sectionToBlockCoord(var3);
+      int var6 = SectionPos.sectionToBlockCoord(var4);
       BlockPos var7 = new BlockPos(var5, 0, var6);
-      Biome var8 = this.biomeSource.getNoiseBiome((var3 << 2) + 2, 2, (var4 << 2) + 2);
+      Biome var8 = this.biomeSource.getPrimaryBiome(var3, var4);
       WorldgenRandom var9 = new WorldgenRandom();
       long var10 = var9.setDecorationSeed(var1.getSeed(), var5, var6);
 
       try {
          var8.generate(var2, this, var1, var10, var9, var7);
@@ -229,11 +230,11 @@
       return var1.getMobSettings().getMobs(var3);
    }
 
    public void createStructures(RegistryAccess var1, StructureFeatureManager var2, ChunkAccess var3, StructureManager var4, long var5) {
       ChunkPos var7 = var3.getPos();
-      Biome var8 = this.biomeSource.getNoiseBiome((var7.x << 2) + 2, 0, (var7.z << 2) + 2);
+      Biome var8 = this.biomeSource.getPrimaryBiome(var7.x, var7.z);
       this.createStructure(StructureFeatures.STRONGHOLD, var1, var2, var3, var4, var5, var7, var8);
       Iterator var9 = var8.getGenerationSettings().structures().iterator();
 
       while(var9.hasNext()) {
          Supplier var10 = (Supplier)var9.next();
@@ -255,12 +256,12 @@
 
    public void createReferences(WorldGenLevel var1, StructureFeatureManager var2, ChunkAccess var3) {
       boolean var4 = true;
       int var5 = var3.getPos().x;
       int var6 = var3.getPos().z;
-      int var7 = var5 << 4;
-      int var8 = var6 << 4;
+      int var7 = SectionPos.sectionToBlockCoord(var5);
+      int var8 = SectionPos.sectionToBlockCoord(var6);
       SectionPos var9 = SectionPos.of(var3.getPos(), 0);
 
       for(int var10 = var5 - 8; var10 <= var5 + 8; ++var10) {
          for(int var11 = var6 - 8; var11 <= var6 + 8; ++var11) {
             long var12 = ChunkPos.asLong(var10, var11);
@@ -300,11 +301,11 @@
       return 63;
    }
 
    public abstract int getBaseHeight(int var1, int var2, Heightmap.Types var3);
 
-   public abstract BlockGetter getBaseColumn(int var1, int var2);
+   public abstract NoiseColumn getBaseColumn(int var1, int var2);
 
    public int getFirstFreeHeight(int var1, int var2, Heightmap.Types var3) {
       return this.getBaseHeight(var1, var2, var3);
    }
 
