--- net/minecraft/world/level/chunk/ChunkBiomeContainer.java
+++ net/minecraft/world/level/chunk/ChunkBiomeContainer.java
@@ -34,11 +34,11 @@
 
       for(int var3 = 0; var3 < this.biomes.length; ++var3) {
          int var4 = var2[var3];
          Biome var5 = (Biome)var1.byId(var4);
          if (var5 == null) {
-            LOGGER.warn("Received invalid biome id: " + var4);
+            LOGGER.warn("Received invalid biome id: {}", var4);
             this.biomes[var3] = (Biome)var1.byId(0);
          } else {
             this.biomes[var3] = var5;
          }
       }
