--- net/minecraft/world/level/chunk/DataLayer.java
+++ net/minecraft/world/level/chunk/DataLayer.java
@@ -1,11 +1,15 @@
 package net.minecraft.world.level.chunk;
 
 import javax.annotation.Nullable;
 import net.minecraft.Util;
+import net.minecraft.util.VisibleForDebug;
 
 public class DataLayer {
+   public static final int SIZE = 2048;
+   public static final int LAYER_SIZE = 128;
+   private static final int NIBBLE_SIZE = 4;
    @Nullable
    protected byte[] data;
 
    public DataLayer() {
    }
@@ -91,9 +95,23 @@
       }
 
       return var1.toString();
    }
 
+   @VisibleForDebug
+   public String layerToString(int var1) {
+      StringBuilder var2 = new StringBuilder();
+
+      for(int var3 = 0; var3 < 256; ++var3) {
+         var2.append(Integer.toHexString(this.get(var3)));
+         if ((var3 & 15) == 15) {
+            var2.append("\n");
+         }
+      }
+
+      return var2.toString();
+   }
+
    public boolean isEmpty() {
       return this.data == null;
    }
 }
