--- net/minecraft/world/level/chunk/LevelChunkSection.java
+++ net/minecraft/world/level/chunk/LevelChunkSection.java
@@ -20,17 +20,21 @@
    public LevelChunkSection(int var1) {
       this(var1, (short)0, (short)0, (short)0);
    }
 
    public LevelChunkSection(int var1, short var2, short var3, short var4) {
-      this.bottomBlockY = var1;
+      this.bottomBlockY = getBottomBlockY(var1);
       this.nonEmptyBlockCount = var2;
       this.tickingBlockCount = var3;
       this.tickingFluidCount = var4;
       this.states = new PalettedContainer(GLOBAL_BLOCKSTATE_PALETTE, Block.BLOCK_STATE_REGISTRY, NbtUtils::readBlockState, NbtUtils::writeBlockState, Blocks.AIR.defaultBlockState());
    }
 
+   public static int getBottomBlockY(int var0) {
+      return var0 << 4;
+   }
+
    public BlockState getBlockState(int var1, int var2, int var3) {
       return (BlockState)this.states.get(var1, var2, var3);
    }
 
    public FluidState getFluidState(int var1, int var2, int var3) {
