--- net/minecraft/world/level/chunk/LinearPalette.java
+++ net/minecraft/world/level/chunk/LinearPalette.java
@@ -16,11 +16,11 @@
    private final int bits;
    private int size;
 
    public LinearPalette(IdMapper<T> var1, int var2, PaletteResize<T> var3, Function<CompoundTag, T> var4) {
       this.registry = var1;
-      this.values = (Object[])(new Object[1 << var2]);
+      this.values = new Object[1 << var2];
       this.bits = var2;
       this.resizeHandler = var3;
       this.reader = var4;
    }
 
