--- net/minecraft/world/entity/ai/behavior/BehaviorUtils.java
+++ net/minecraft/world/entity/ai/behavior/BehaviorUtils.java
@@ -137,11 +137,13 @@
    }
 
    public static Optional<LivingEntity> getLivingEntityFromUUIDMemory(LivingEntity var0, MemoryModuleType<UUID> var1) {
       Optional var2 = var0.getBrain().getMemory(var1);
       return var2.map((var1x) -> {
-         return (LivingEntity)((ServerLevel)var0.level).getEntity(var1x);
+         return ((ServerLevel)var0.level).getEntity(var1x);
+      }).map((var0x) -> {
+         return var0x instanceof LivingEntity ? (LivingEntity)var0x : null;
       });
    }
 
    public static Stream<Villager> getNearbyVillagersWithCondition(Villager var0, Predicate<Villager> var1) {
       return (Stream)var0.getBrain().getMemory(MemoryModuleType.LIVING_ENTITIES).map((var2) -> {
