--- net/minecraft/world/entity/ai/behavior/EntityTracker.java
+++ net/minecraft/world/entity/ai/behavior/EntityTracker.java
@@ -32,9 +32,13 @@
          Optional var2 = var1.getBrain().getMemory(MemoryModuleType.VISIBLE_LIVING_ENTITIES);
          return this.entity.isAlive() && var2.isPresent() && ((List)var2.get()).contains(this.entity);
       }
    }
 
+   public Entity getEntity() {
+      return this.entity;
+   }
+
    public String toString() {
       return "EntityTracker for " + this.entity;
    }
 }
