--- net/minecraft/world/entity/ai/behavior/GiveGiftToHero.java
+++ net/minecraft/world/entity/ai/behavior/GiveGiftToHero.java
@@ -26,11 +26,15 @@
 import net.minecraft.world.level.storage.loot.LootTable;
 import net.minecraft.world.level.storage.loot.parameters.LootContextParamSets;
 import net.minecraft.world.level.storage.loot.parameters.LootContextParams;
 
 public class GiveGiftToHero extends Behavior<Villager> {
-   private static final Map<VillagerProfession, ResourceLocation> gifts = (Map)Util.make(Maps.newHashMap(), (var0) -> {
+   private static final int THROW_GIFT_AT_DISTANCE = 5;
+   private static final int MIN_TIME_BETWEEN_GIFTS = 600;
+   private static final int MAX_TIME_BETWEEN_GIFTS = 6600;
+   private static final int TIME_TO_DELAY_FOR_HEAD_TO_FINISH_TURNING = 20;
+   private static final Map<VillagerProfession, ResourceLocation> GIFTS = (Map)Util.make(Maps.newHashMap(), (var0) -> {
       var0.put(VillagerProfession.ARMORER, BuiltInLootTables.ARMORER_GIFT);
       var0.put(VillagerProfession.BUTCHER, BuiltInLootTables.BUTCHER_GIFT);
       var0.put(VillagerProfession.CARTOGRAPHER, BuiltInLootTables.CARTOGRAPHER_GIFT);
       var0.put(VillagerProfession.CLERIC, BuiltInLootTables.CLERIC_GIFT);
       var0.put(VillagerProfession.FARMER, BuiltInLootTables.FARMER_GIFT);
@@ -41,10 +45,11 @@
       var0.put(VillagerProfession.MASON, BuiltInLootTables.MASON_GIFT);
       var0.put(VillagerProfession.SHEPHERD, BuiltInLootTables.SHEPHERD_GIFT);
       var0.put(VillagerProfession.TOOLSMITH, BuiltInLootTables.TOOLSMITH_GIFT);
       var0.put(VillagerProfession.WEAPONSMITH, BuiltInLootTables.WEAPONSMITH_GIFT);
    });
+   private static final float SPEED_MODIFIER = 0.5F;
    private int timeUntilNextGift = 600;
    private boolean giftGivenDuringThisRun;
    private long timeSinceStart;
 
    public GiveGiftToHero(int var1) {
@@ -109,12 +114,12 @@
    private List<ItemStack> getItemToThrow(Villager var1) {
       if (var1.isBaby()) {
          return ImmutableList.of(new ItemStack(Items.POPPY));
       } else {
          VillagerProfession var2 = var1.getVillagerData().getProfession();
-         if (gifts.containsKey(var2)) {
-            LootTable var3 = var1.level.getServer().getLootTables().get((ResourceLocation)gifts.get(var2));
+         if (GIFTS.containsKey(var2)) {
+            LootTable var3 = var1.level.getServer().getLootTables().get((ResourceLocation)GIFTS.get(var2));
             LootContext.Builder var4 = (new LootContext.Builder((ServerLevel)var1.level)).withParameter(LootContextParams.ORIGIN, var1.position()).withParameter(LootContextParams.THIS_ENTITY, var1).withRandom(var1.getRandom());
             return var3.getRandomItems(var4.create(LootContextParamSets.GIFT));
          } else {
             return ImmutableList.of(new ItemStack(Items.WHEAT_SEEDS));
          }
