--- net/minecraft/world/entity/ai/behavior/GoToPotentialJobSite.java
+++ net/minecraft/world/entity/ai/behavior/GoToPotentialJobSite.java
@@ -11,10 +11,11 @@
 import net.minecraft.world.entity.ai.village.poi.PoiManager;
 import net.minecraft.world.entity.npc.Villager;
 import net.minecraft.world.entity.schedule.Activity;
 
 public class GoToPotentialJobSite extends Behavior<Villager> {
+   private static final int TICKS_UNTIL_TIMEOUT = 1200;
    final float speedModifier;
 
    public GoToPotentialJobSite(float var1) {
       super(ImmutableMap.of(MemoryModuleType.POTENTIAL_JOB_SITE, MemoryStatus.VALUE_PRESENT), 1200);
       this.speedModifier = var1;
