--- net/minecraft/world/entity/ai/behavior/InteractWithDoor.java
+++ net/minecraft/world/entity/ai/behavior/InteractWithDoor.java
@@ -55,22 +55,22 @@
       BlockPos var8 = var6.asBlockPos();
       BlockState var9 = var1.getBlockState(var8);
       if (var9.is(BlockTags.WOODEN_DOORS)) {
          DoorBlock var10 = (DoorBlock)var9.getBlock();
          if (!var10.isOpen(var9)) {
-            var10.setOpen(var1, var9, var8, true);
+            var10.setOpen(var2, var1, var9, var8, true);
          }
 
          this.rememberDoorToClose(var1, var2, var8);
       }
 
       BlockPos var13 = var7.asBlockPos();
       BlockState var11 = var1.getBlockState(var13);
       if (var11.is(BlockTags.WOODEN_DOORS)) {
          DoorBlock var12 = (DoorBlock)var11.getBlock();
          if (!var12.isOpen(var11)) {
-            var12.setOpen(var1, var11, var13, true);
+            var12.setOpen(var2, var1, var11, var13, true);
             this.rememberDoorToClose(var1, var2, var13);
          }
       }
 
       closeDoorsThatIHaveOpenedOrPassedThrough(var1, var2, var6, var7);
@@ -106,11 +106,11 @@
                   if (!var9.isOpen(var8)) {
                      var5.remove();
                   } else if (areOtherMobsComingThroughDoor(var0, var1, var7)) {
                      var5.remove();
                   } else {
-                     var9.setOpen(var0, var8, var7, false);
+                     var9.setOpen(var1, var0, var8, var7, false);
                      var5.remove();
                   }
                }
             }
          }
