--- net/minecraft/world/entity/ai/behavior/MeleeAttack.java
+++ net/minecraft/world/entity/ai/behavior/MeleeAttack.java
@@ -5,10 +5,11 @@
 import net.minecraft.world.InteractionHand;
 import net.minecraft.world.entity.LivingEntity;
 import net.minecraft.world.entity.Mob;
 import net.minecraft.world.entity.ai.memory.MemoryModuleType;
 import net.minecraft.world.entity.ai.memory.MemoryStatus;
+import net.minecraft.world.item.Item;
 import net.minecraft.world.item.ProjectileWeaponItem;
 
 public class MeleeAttack extends Behavior<Mob> {
    private final int cooldownBetweenAttacks;
 
@@ -22,11 +23,12 @@
       return !this.isHoldingUsableProjectileWeapon(var2) && BehaviorUtils.canSee(var2, var3) && BehaviorUtils.isWithinMeleeAttackRange(var2, var3);
    }
 
    private boolean isHoldingUsableProjectileWeapon(Mob var1) {
       return var1.isHolding((var1x) -> {
-         return var1x instanceof ProjectileWeaponItem && var1.canFireProjectileWeapon((ProjectileWeaponItem)var1x);
+         Item var2 = var1x.getItem();
+         return var2 instanceof ProjectileWeaponItem && var1.canFireProjectileWeapon((ProjectileWeaponItem)var2);
       });
    }
 
    protected void start(ServerLevel var1, Mob var2, long var3) {
       LivingEntity var5 = this.getAttackTarget(var2);
