--- net/minecraft/world/entity/ai/behavior/PoiCompetitorScan.java
+++ net/minecraft/world/entity/ai/behavior/PoiCompetitorScan.java
@@ -11,11 +11,11 @@
 
 public class PoiCompetitorScan extends Behavior<Villager> {
    final VillagerProfession profession;
 
    public PoiCompetitorScan(VillagerProfession var1) {
-      super(ImmutableMap.of(MemoryModuleType.JOB_SITE, MemoryStatus.VALUE_PRESENT, MemoryModuleType.LIVING_ENTITIES, MemoryStatus.VALUE_PRESENT));
+      super(ImmutableMap.of(MemoryModuleType.JOB_SITE, MemoryStatus.VALUE_PRESENT, MemoryModuleType.NEAREST_LIVING_ENTITIES, MemoryStatus.VALUE_PRESENT));
       this.profession = var1;
    }
 
    protected void start(ServerLevel var1, Villager var2, long var3) {
       GlobalPos var5 = (GlobalPos)var2.getBrain().getMemory(MemoryModuleType.JOB_SITE).get();
