--- net/minecraft/world/entity/ai/behavior/RingBell.java
+++ net/minecraft/world/entity/ai/behavior/RingBell.java
@@ -12,10 +12,13 @@
 import net.minecraft.world.level.block.BellBlock;
 import net.minecraft.world.level.block.Blocks;
 import net.minecraft.world.level.block.state.BlockState;
 
 public class RingBell extends Behavior<LivingEntity> {
+   private static final float BELL_RING_CHANCE = 0.95F;
+   public static final int RING_BELL_FROM_DISTANCE = 3;
+
    public RingBell() {
       super(ImmutableMap.of(MemoryModuleType.MEETING_POINT, MemoryStatus.VALUE_PRESENT));
    }
 
    protected boolean checkExtraStartConditions(ServerLevel var1, LivingEntity var2) {
@@ -27,11 +30,11 @@
       BlockPos var6 = ((GlobalPos)var5.getMemory(MemoryModuleType.MEETING_POINT).get()).pos();
       if (var6.closerThan(var2.blockPosition(), 3.0D)) {
          BlockState var7 = var1.getBlockState(var6);
          if (var7.is(Blocks.BELL)) {
             BellBlock var8 = (BellBlock)var7.getBlock();
-            var8.attemptToRing(var1, var6, (Direction)null);
+            var8.attemptToRing(var2, var1, var6, (Direction)null);
          }
       }
 
    }
 }
