--- net/minecraft/world/entity/ai/behavior/RingBell.java
+++ net/minecraft/world/entity/ai/behavior/RingBell.java
@@ -27,11 +27,11 @@
       BlockPos var6 = ((GlobalPos)var5.getMemory(MemoryModuleType.MEETING_POINT).get()).pos();
       if (var6.closerThan(var2.blockPosition(), 3.0D)) {
          BlockState var7 = var1.getBlockState(var6);
          if (var7.is(Blocks.BELL)) {
             BellBlock var8 = (BellBlock)var7.getBlock();
-            var8.attemptToRing(var1, var6, (Direction)null);
+            var8.attemptToRing(var2, var1, var6, (Direction)null);
          }
       }
 
    }
 }
