--- net/minecraft/world/entity/ai/behavior/RunIf.java
+++ net/minecraft/world/entity/ai/behavior/RunIf.java
@@ -27,14 +27,24 @@
       var2.putAll(var0);
       var2.putAll(var1);
       return var2;
    }
 
+   public RunIf(Predicate<E> var1, Behavior<? super E> var2, boolean var3) {
+      this(ImmutableMap.of(), var1, var2, var3);
+   }
+
    public RunIf(Predicate<E> var1, Behavior<? super E> var2) {
       this(ImmutableMap.of(), var1, var2, false);
    }
 
+   public RunIf(Map<MemoryModuleType<?>, MemoryStatus> var1, Behavior<? super E> var2) {
+      this(var1, (var0) -> {
+         return true;
+      }, var2, false);
+   }
+
    protected boolean checkExtraStartConditions(ServerLevel var1, E var2) {
       return this.predicate.test(var2) && this.wrappedBehavior.checkExtraStartConditions(var1, var2);
    }
 
    protected boolean canStillUse(ServerLevel var1, E var2, long var3) {
