--- net/minecraft/world/entity/ai/behavior/SetEntityLookTarget.java
+++ net/minecraft/world/entity/ai/behavior/SetEntityLookTarget.java
@@ -2,10 +2,11 @@
 
 import com.google.common.collect.ImmutableMap;
 import java.util.List;
 import java.util.function.Predicate;
 import net.minecraft.server.level.ServerLevel;
+import net.minecraft.tags.Tag;
 import net.minecraft.world.entity.EntityType;
 import net.minecraft.world.entity.LivingEntity;
 import net.minecraft.world.entity.MobCategory;
 import net.minecraft.world.entity.ai.Brain;
 import net.minecraft.world.entity.ai.memory.MemoryModuleType;
@@ -13,10 +14,16 @@
 
 public class SetEntityLookTarget extends Behavior<LivingEntity> {
    private final Predicate<LivingEntity> predicate;
    private final float maxDistSqr;
 
+   public SetEntityLookTarget(Tag<EntityType<?>> var1, float var2) {
+      this((var1x) -> {
+         return var1x.getType().is(var1);
+      }, var2);
+   }
+
    public SetEntityLookTarget(MobCategory var1, float var2) {
       this((var1x) -> {
          return var1.equals(var1x.getType().getCategory());
       }, var2);
    }
