--- net/minecraft/world/entity/ai/behavior/SetWalkTargetFromBlockMemory.java
+++ net/minecraft/world/entity/ai/behavior/SetWalkTargetFromBlockMemory.java
@@ -7,11 +7,11 @@
 import net.minecraft.server.level.ServerLevel;
 import net.minecraft.world.entity.ai.Brain;
 import net.minecraft.world.entity.ai.memory.MemoryModuleType;
 import net.minecraft.world.entity.ai.memory.MemoryStatus;
 import net.minecraft.world.entity.ai.memory.WalkTarget;
-import net.minecraft.world.entity.ai.util.RandomPos;
+import net.minecraft.world.entity.ai.util.DefaultRandomPos;
 import net.minecraft.world.entity.npc.Villager;
 import net.minecraft.world.phys.Vec3;
 
 public class SetWalkTargetFromBlockMemory extends Behavior<Villager> {
    private final MemoryModuleType<GlobalPos> memoryType;
@@ -43,11 +43,11 @@
             if (this.tooFar(var2, var6)) {
                Vec3 var7 = null;
                int var8 = 0;
 
                for(boolean var9 = true; var8 < 1000 && (var7 == null || this.tooFar(var2, GlobalPos.of(var1.dimension(), new BlockPos(var7)))); ++var8) {
-                  var7 = RandomPos.getPosTowards(var2, 15, 7, Vec3.atBottomCenterOf(var6.pos()));
+                  var7 = DefaultRandomPos.getPosTowards(var2, 15, 7, Vec3.atBottomCenterOf(var6.pos()), 1.5707963705062866D);
                }
 
                if (var8 == 1000) {
                   this.dropPOI(var2, var3);
                   return;
