--- net/minecraft/world/entity/ai/behavior/ShowTradesToPlayer.java
+++ net/minecraft/world/entity/ai/behavior/ShowTradesToPlayer.java
@@ -15,10 +15,12 @@
 import net.minecraft.world.entity.npc.Villager;
 import net.minecraft.world.item.ItemStack;
 import net.minecraft.world.item.trading.MerchantOffer;
 
 public class ShowTradesToPlayer extends Behavior<Villager> {
+   private static final int MAX_LOOK_TIME = 900;
+   private static final int STARTING_LOOK_TIME = 40;
    @Nullable
    private ItemStack playerItemStack;
    private final List<ItemStack> displayItems = Lists.newArrayList();
    private int cycleCounter;
    private int displayIndex;
