--- net/minecraft/world/entity/ai/behavior/SocializeAtBell.java
+++ net/minecraft/world/entity/ai/behavior/SocializeAtBell.java
@@ -11,10 +11,12 @@
 import net.minecraft.world.entity.ai.memory.MemoryModuleType;
 import net.minecraft.world.entity.ai.memory.MemoryStatus;
 import net.minecraft.world.entity.ai.memory.WalkTarget;
 
 public class SocializeAtBell extends Behavior<LivingEntity> {
+   private static final float SPEED_MODIFIER = 0.3F;
+
    public SocializeAtBell() {
       super(ImmutableMap.of(MemoryModuleType.WALK_TARGET, MemoryStatus.REGISTERED, MemoryModuleType.LOOK_TARGET, MemoryStatus.REGISTERED, MemoryModuleType.MEETING_POINT, MemoryStatus.VALUE_PRESENT, MemoryModuleType.VISIBLE_LIVING_ENTITIES, MemoryStatus.VALUE_PRESENT, MemoryModuleType.INTERACTION_TARGET, MemoryStatus.VALUE_ABSENT));
    }
 
    protected boolean checkExtraStartConditions(ServerLevel var1, LivingEntity var2) {
