--- net/minecraft/world/entity/ai/behavior/UseBonemeal.java
+++ net/minecraft/world/entity/ai/behavior/UseBonemeal.java
@@ -16,10 +16,11 @@
 import net.minecraft.world.level.block.Block;
 import net.minecraft.world.level.block.CropBlock;
 import net.minecraft.world.level.block.state.BlockState;
 
 public class UseBonemeal extends Behavior<Villager> {
+   private static final int BONEMEALING_DURATION = 80;
    private long nextWorkCycleTime;
    private long lastBonemealingSession;
    private int timeWorkedSoFar;
    private Optional<BlockPos> cropPos = Optional.empty();
 
@@ -99,11 +100,11 @@
          SimpleContainer var7 = var2.getInventory();
          int var8 = var7.getContainerSize();
 
          for(int var9 = 0; var9 < var8; ++var9) {
             ItemStack var10 = var7.getItem(var9);
-            if (var10.getItem() == Items.BONE_MEAL) {
+            if (var10.is(Items.BONE_MEAL)) {
                var6 = var10;
                break;
             }
          }
 
