--- net/minecraft/world/entity/ai/behavior/UseBonemeal.java
+++ net/minecraft/world/entity/ai/behavior/UseBonemeal.java
@@ -99,11 +99,11 @@
          SimpleContainer var7 = var2.getInventory();
          int var8 = var7.getContainerSize();
 
          for(int var9 = 0; var9 < var8; ++var9) {
             ItemStack var10 = var7.getItem(var9);
-            if (var10.getItem() == Items.BONE_MEAL) {
+            if (var10.is(Items.BONE_MEAL)) {
                var6 = var10;
                break;
             }
          }
 
