--- net/minecraft/client/renderer/chunk/VisGraph.java
+++ net/minecraft/client/renderer/chunk/VisGraph.java
@@ -7,13 +7,21 @@
 import net.minecraft.Util;
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.Direction;
 
 public class VisGraph {
+   private static final int SIZE_IN_BITS = 4;
+   private static final int LEN = 16;
+   private static final int MASK = 15;
+   private static final int SIZE = 4096;
+   private static final int X_SHIFT = 0;
+   private static final int Z_SHIFT = 4;
+   private static final int Y_SHIFT = 8;
    private static final int DX = (int)Math.pow(16.0D, 0.0D);
    private static final int DZ = (int)Math.pow(16.0D, 1.0D);
    private static final int DY = (int)Math.pow(16.0D, 2.0D);
+   private static final int INVALID_INDEX = -1;
    private static final Direction[] DIRECTIONS = Direction.values();
    private final BitSet bitSet = new BitSet(4096);
    private static final int[] INDEX_OF_EDGES = (int[])Util.make(new int[1352], (var0) -> {
       boolean var1 = false;
       boolean var2 = true;
