--- net/minecraft/world/level/storage/loot/BuiltInLootTables.java
+++ net/minecraft/world/level/storage/loot/BuiltInLootTables.java
@@ -84,10 +84,13 @@
    public static final ResourceLocation SHEPHERD_GIFT;
    public static final ResourceLocation TOOLSMITH_GIFT;
    public static final ResourceLocation WEAPONSMITH_GIFT;
    public static final ResourceLocation PIGLIN_BARTERING;
 
+   public BuiltInLootTables() {
+   }
+
    private static ResourceLocation register(String var0) {
       return register(new ResourceLocation(var0));
    }
 
    private static ResourceLocation register(ResourceLocation var0) {
