--- net/minecraft/world/level/storage/loot/LootContext.java
+++ net/minecraft/world/level/storage/loot/LootContext.java
@@ -6,10 +6,11 @@
 import com.google.gson.TypeAdapter;
 import com.google.gson.stream.JsonReader;
 import com.google.gson.stream.JsonWriter;
 import java.io.IOException;
 import java.util.Map;
+import java.util.NoSuchElementException;
 import java.util.Random;
 import java.util.Set;
 import java.util.function.Consumer;
 import java.util.function.Function;
 import javax.annotation.Nullable;
@@ -48,10 +49,19 @@
 
    public boolean hasParam(LootContextParam<?> var1) {
       return this.params.containsKey(var1);
    }
 
+   public <T> T getParam(LootContextParam<T> var1) {
+      Object var2 = this.params.get(var1);
+      if (var2 == null) {
+         throw new NoSuchElementException(var1.getName().toString());
+      } else {
+         return var2;
+      }
+   }
+
    public void addDynamicDrops(ResourceLocation var1, Consumer<ItemStack> var2) {
       LootContext.DynamicDrop var3 = (LootContext.DynamicDrop)this.dynamicDrops.get(var1);
       if (var3 != null) {
          var3.add(this, var2);
       }
