--- net/minecraft/world/level/storage/loot/ValidationContext.java
+++ net/minecraft/world/level/storage/loot/ValidationContext.java
@@ -50,25 +50,28 @@
       this.problems.put(this.getContext(), var1);
    }
 
    public ValidationContext forChild(String var1) {
       return new ValidationContext(this.problems, () -> {
-         return this.getContext() + var1;
+         String var10000 = this.getContext();
+         return var10000 + var1;
       }, this.params, this.conditionResolver, this.visitedConditions, this.tableResolver, this.visitedTables);
    }
 
    public ValidationContext enterTable(String var1, ResourceLocation var2) {
       ImmutableSet var3 = ImmutableSet.builder().addAll((Iterable)this.visitedTables).add((Object)var2).build();
       return new ValidationContext(this.problems, () -> {
-         return this.getContext() + var1;
+         String var10000 = this.getContext();
+         return var10000 + var1;
       }, this.params, this.conditionResolver, this.visitedConditions, this.tableResolver, var3);
    }
 
    public ValidationContext enterCondition(String var1, ResourceLocation var2) {
       ImmutableSet var3 = ImmutableSet.builder().addAll((Iterable)this.visitedConditions).add((Object)var2).build();
       return new ValidationContext(this.problems, () -> {
-         return this.getContext() + var1;
+         String var10000 = this.getContext();
+         return var10000 + var1;
       }, this.params, this.conditionResolver, var3, this.tableResolver, this.visitedTables);
    }
 
    public boolean hasVisitedTable(ResourceLocation var1) {
       return this.visitedTables.contains(var1);
