--- net/minecraft/client/renderer/texture/AbstractTexture.java
+++ net/minecraft/client/renderer/texture/AbstractTexture.java
@@ -7,10 +7,11 @@
 import java.util.concurrent.Executor;
 import net.minecraft.resources.ResourceLocation;
 import net.minecraft.server.packs.resources.ResourceManager;
 
 public abstract class AbstractTexture implements AutoCloseable {
+   public static final int NOT_ASSIGNED = -1;
    protected int id = -1;
    protected boolean blur;
    protected boolean mipmap;
 
    public AbstractTexture() {
@@ -28,10 +29,11 @@
       } else {
          var3 = var2 ? 9986 : 9728;
          var4 = 9728;
       }
 
+      this.bind();
       GlStateManager._texParameter(3553, 10241, var3);
       GlStateManager._texParameter(3553, 10240, var4);
    }
 
    public int getId() {
