--- net/minecraft/client/renderer/texture/MipmapGenerator.java
+++ net/minecraft/client/renderer/texture/MipmapGenerator.java
@@ -2,17 +2,21 @@
 
 import com.mojang.blaze3d.platform.NativeImage;
 import net.minecraft.Util;
 
 public class MipmapGenerator {
+   private static final int ALPHA_CUTOUT_CUTOFF = 96;
    private static final float[] POW22 = (float[])Util.make(new float[256], (var0) -> {
       for(int var1 = 0; var1 < var0.length; ++var1) {
          var0[var1] = (float)Math.pow((double)((float)var1 / 255.0F), 2.2D);
       }
 
    });
 
+   private MipmapGenerator() {
+   }
+
    public static NativeImage[] generateMipLevels(NativeImage var0, int var1) {
       NativeImage[] var2 = new NativeImage[var1 + 1];
       var2[0] = var0;
       if (var1 > 0) {
          boolean var3 = false;
