--- net/minecraft/client/renderer/texture/OverlayTexture.java
+++ net/minecraft/client/renderer/texture/OverlayTexture.java
@@ -1,11 +1,16 @@
 package net.minecraft.client.renderer.texture;
 
 import com.mojang.blaze3d.platform.NativeImage;
 import com.mojang.blaze3d.systems.RenderSystem;
+import java.util.Objects;
 
 public class OverlayTexture implements AutoCloseable {
+   private static final int SIZE = 16;
+   public static final int NO_WHITE_U = 0;
+   public static final int RED_OVERLAY_V = 3;
+   public static final int WHITE_OVERLAY_V = 10;
    public static final int NO_OVERLAY = pack(0, 10);
    private final DynamicTexture texture = new DynamicTexture(16, 16, false);
 
    public OverlayTexture() {
       NativeImage var1 = this.texture.getPixels();
@@ -21,26 +26,22 @@
          }
       }
 
       RenderSystem.activeTexture(33985);
       this.texture.bind();
-      RenderSystem.matrixMode(5890);
-      RenderSystem.loadIdentity();
-      float var5 = 0.06666667F;
-      RenderSystem.scalef(0.06666667F, 0.06666667F, 0.06666667F);
-      RenderSystem.matrixMode(5888);
-      this.texture.bind();
       var1.upload(0, 0, 0, 0, 0, var1.getWidth(), var1.getHeight(), false, true, false, false);
       RenderSystem.activeTexture(33984);
    }
 
    public void close() {
       this.texture.close();
    }
 
    public void setupOverlayColor() {
-      RenderSystem.setupOverlayColor(this.texture::getId, 16);
+      DynamicTexture var10000 = this.texture;
+      Objects.requireNonNull(var10000);
+      RenderSystem.setupOverlayColor(var10000::getId, 16);
    }
 
    public static int u(float var0) {
       return (int)(var0 * 15.0F);
    }
