--- net/minecraft/client/renderer/texture/OverlayTexture.java
+++ net/minecraft/client/renderer/texture/OverlayTexture.java
@@ -2,10 +2,14 @@
 
 import com.mojang.blaze3d.platform.NativeImage;
 import com.mojang.blaze3d.systems.RenderSystem;
 
 public class OverlayTexture implements AutoCloseable {
+   private static final int SIZE = 16;
+   public static final int NO_WHITE_U = 0;
+   public static final int RED_OVERLAY_V = 3;
+   public static final int WHITE_OVERLAY_V = 10;
    public static final int NO_OVERLAY = pack(0, 10);
    private final DynamicTexture texture = new DynamicTexture(16, 16, false);
 
    public OverlayTexture() {
       NativeImage var1 = this.texture.getPixels();
@@ -21,16 +25,10 @@
          }
       }
 
       RenderSystem.activeTexture(33985);
       this.texture.bind();
-      RenderSystem.matrixMode(5890);
-      RenderSystem.loadIdentity();
-      float var5 = 0.06666667F;
-      RenderSystem.scalef(0.06666667F, 0.06666667F, 0.06666667F);
-      RenderSystem.matrixMode(5888);
-      this.texture.bind();
       var1.upload(0, 0, 0, 0, 0, var1.getWidth(), var1.getHeight(), false, true, false, false);
       RenderSystem.activeTexture(33984);
    }
 
    public void close() {
