--- net/minecraft/client/renderer/texture/OverlayTexture.java
+++ net/minecraft/client/renderer/texture/OverlayTexture.java
@@ -21,16 +21,10 @@
          }
       }
 
       RenderSystem.activeTexture(33985);
       this.texture.bind();
-      RenderSystem.matrixMode(5890);
-      RenderSystem.loadIdentity();
-      float var5 = 0.06666667F;
-      RenderSystem.scalef(0.06666667F, 0.06666667F, 0.06666667F);
-      RenderSystem.matrixMode(5888);
-      this.texture.bind();
       var1.upload(0, 0, 0, 0, 0, var1.getWidth(), var1.getHeight(), false, true, false, false);
       RenderSystem.activeTexture(33984);
    }
 
    public void close() {
