--- net/minecraft/client/renderer/texture/PreloadedTexture.java
+++ net/minecraft/client/renderer/texture/PreloadedTexture.java
@@ -1,8 +1,9 @@
 package net.minecraft.client.renderer.texture;
 
 import com.mojang.blaze3d.systems.RenderSystem;
+import java.util.Objects;
 import java.util.concurrent.CompletableFuture;
 import java.util.concurrent.Executor;
 import javax.annotation.Nullable;
 import net.minecraft.Util;
 import net.minecraft.resources.ResourceLocation;
@@ -45,10 +46,11 @@
    }
 
    private static Executor executor(Executor var0) {
       return (var1) -> {
          var0.execute(() -> {
+            Objects.requireNonNull(var1);
             RenderSystem.recordRenderCall(var1::run);
          });
       };
    }
 }
