--- net/minecraft/client/renderer/texture/TextureAtlas.java
+++ net/minecraft/client/renderer/texture/TextureAtlas.java
@@ -38,22 +38,23 @@
    private static final Logger LOGGER = LogManager.getLogger();
    @Deprecated
    public static final ResourceLocation LOCATION_BLOCKS;
    @Deprecated
    public static final ResourceLocation LOCATION_PARTICLES;
-   private final List<TextureAtlasSprite> animatedTextures = Lists.newArrayList();
+   private static final String FILE_EXTENSION = ".png";
+   private final List<Tickable> animatedTextures = Lists.newArrayList();
    private final Set<ResourceLocation> sprites = Sets.newHashSet();
    private final Map<ResourceLocation, TextureAtlasSprite> texturesByName = Maps.newHashMap();
    private final ResourceLocation location;
    private final int maxSupportedTextureSize;
 
    public TextureAtlas(ResourceLocation var1) {
       this.location = var1;
       this.maxSupportedTextureSize = RenderSystem.maxSupportedTextureSize();
    }
 
-   public void load(ResourceManager var1) throws IOException {
+   public void load(ResourceManager var1) {
    }
 
    public void reload(TextureAtlas.Preparations var1) {
       this.sprites.clear();
       this.sprites.addAll(var1.sprites);
@@ -74,12 +75,13 @@
             var6.setDetail("Atlas path", (Object)this.location);
             var6.setDetail("Sprite", (Object)var3);
             throw new ReportedException(var5);
          }
 
-         if (var3.isAnimation()) {
-            this.animatedTextures.add(var3);
+         Tickable var4 = var3.getAnimationTicker();
+         if (var4 != null) {
+            this.animatedTextures.add(var4);
          }
       }
 
    }
 
@@ -267,12 +269,12 @@
    public void cycleAnimationFrames() {
       this.bind();
       Iterator var1 = this.animatedTextures.iterator();
 
       while(var1.hasNext()) {
-         TextureAtlasSprite var2 = (TextureAtlasSprite)var1.next();
-         var2.cycleFrames();
+         Tickable var2 = (Tickable)var1.next();
+         var2.tick();
       }
 
    }
 
    public void tick() {
