--- net/minecraft/client/renderer/texture/TextureAtlas.java
+++ net/minecraft/client/renderer/texture/TextureAtlas.java
@@ -49,11 +49,11 @@
    public TextureAtlas(ResourceLocation var1) {
       this.location = var1;
       this.maxSupportedTextureSize = RenderSystem.maxSupportedTextureSize();
    }
 
-   public void load(ResourceManager var1) throws IOException {
+   public void load(ResourceManager var1) {
    }
 
    public void reload(TextureAtlas.Preparations var1) {
       this.sprites.clear();
       this.sprites.addAll(var1.sprites);
