--- net/minecraft/data/worldgen/BastionPieces.java
+++ net/minecraft/data/worldgen/BastionPieces.java
@@ -7,10 +7,13 @@
 import net.minecraft.world.level.levelgen.feature.structures.StructureTemplatePool;
 
 public class BastionPieces {
    public static final StructureTemplatePool START;
 
+   public BastionPieces() {
+   }
+
    public static void bootstrap() {
       BastionHousingUnitsPools.bootstrap();
       BastionHoglinStablePools.bootstrap();
       BastionTreasureRoomPools.bootstrap();
       BastionBridgePools.bootstrap();
