--- net/minecraft/data/worldgen/BastionTreasureRoomPools.java
+++ net/minecraft/data/worldgen/BastionTreasureRoomPools.java
@@ -5,10 +5,13 @@
 import net.minecraft.resources.ResourceLocation;
 import net.minecraft.world.level.levelgen.feature.structures.StructurePoolElement;
 import net.minecraft.world.level.levelgen.feature.structures.StructureTemplatePool;
 
 public class BastionTreasureRoomPools {
+   public BastionTreasureRoomPools() {
+   }
+
    public static void bootstrap() {
    }
 
    static {
       Pools.register(new StructureTemplatePool(new ResourceLocation("bastion/treasure/bases"), new ResourceLocation("empty"), ImmutableList.of(Pair.of(StructurePoolElement.single("bastion/treasure/bases/lava_basin", ProcessorLists.TREASURE_ROOMS), 1)), StructureTemplatePool.Projection.RIGID));
