--- net/minecraft/data/worldgen/BiomeDefaultFeatures.java
+++ net/minecraft/data/worldgen/BiomeDefaultFeatures.java
@@ -23,17 +23,16 @@
    }
 
    public static void addDefaultCarvers(BiomeGenerationSettings.Builder var0) {
       var0.addCarver(GenerationStep.Carving.AIR, Carvers.CAVE);
       var0.addCarver(GenerationStep.Carving.AIR, Carvers.CANYON);
+      var0.addCarver(GenerationStep.Carving.AIR, Carvers.CRACK);
    }
 
    public static void addOceanCarvers(BiomeGenerationSettings.Builder var0) {
       var0.addCarver(GenerationStep.Carving.AIR, Carvers.OCEAN_CAVE);
       var0.addCarver(GenerationStep.Carving.AIR, Carvers.CANYON);
-      var0.addCarver(GenerationStep.Carving.LIQUID, Carvers.UNDERWATER_CANYON);
-      var0.addCarver(GenerationStep.Carving.LIQUID, Carvers.UNDERWATER_CAVE);
    }
 
    public static void addDefaultLakes(BiomeGenerationSettings.Builder var0) {
       var0.addFeature(GenerationStep.Decoration.LAKES, Features.LAKE_WATER);
       var0.addFeature(GenerationStep.Decoration.LAKES, Features.LAKE_LAVA);
@@ -51,19 +50,34 @@
       var0.addFeature(GenerationStep.Decoration.UNDERGROUND_ORES, Features.ORE_DIRT);
       var0.addFeature(GenerationStep.Decoration.UNDERGROUND_ORES, Features.ORE_GRAVEL);
       var0.addFeature(GenerationStep.Decoration.UNDERGROUND_ORES, Features.ORE_GRANITE);
       var0.addFeature(GenerationStep.Decoration.UNDERGROUND_ORES, Features.ORE_DIORITE);
       var0.addFeature(GenerationStep.Decoration.UNDERGROUND_ORES, Features.ORE_ANDESITE);
+      var0.addFeature(GenerationStep.Decoration.UNDERGROUND_ORES, Features.ORE_TUFF);
+      var0.addFeature(GenerationStep.Decoration.VEGETAL_DECORATION, Features.GLOW_LICHEN);
    }
 
+   public static void addDripstone(BiomeGenerationSettings.Builder var0) {
+      var0.addFeature(GenerationStep.Decoration.LOCAL_MODIFICATIONS, Features.LARGE_DRIPSTONE_FEATURE);
+      var0.addFeature(GenerationStep.Decoration.UNDERGROUND_DECORATION, Features.DRIPSTONE_CLUSTER_FEATURE);
+      var0.addFeature(GenerationStep.Decoration.UNDERGROUND_DECORATION, Features.SMALL_DRIPSTONE_FEATURE);
+   }
+
    public static void addDefaultOres(BiomeGenerationSettings.Builder var0) {
-      var0.addFeature(GenerationStep.Decoration.UNDERGROUND_ORES, Features.ORE_COAL);
-      var0.addFeature(GenerationStep.Decoration.UNDERGROUND_ORES, Features.ORE_IRON);
+      var0.addFeature(GenerationStep.Decoration.UNDERGROUND_ORES, Features.ORE_COAL_UPPER);
+      var0.addFeature(GenerationStep.Decoration.UNDERGROUND_ORES, Features.ORE_COAL_LOWER);
+      var0.addFeature(GenerationStep.Decoration.UNDERGROUND_ORES, Features.ORE_IRON_UPPER);
+      var0.addFeature(GenerationStep.Decoration.UNDERGROUND_ORES, Features.ORE_IRON_MIDDLE);
+      var0.addFeature(GenerationStep.Decoration.UNDERGROUND_ORES, Features.ORE_IRON_LOWER);
       var0.addFeature(GenerationStep.Decoration.UNDERGROUND_ORES, Features.ORE_GOLD);
       var0.addFeature(GenerationStep.Decoration.UNDERGROUND_ORES, Features.ORE_REDSTONE);
+      var0.addFeature(GenerationStep.Decoration.UNDERGROUND_ORES, Features.ORE_REDSTONE_LOWER);
       var0.addFeature(GenerationStep.Decoration.UNDERGROUND_ORES, Features.ORE_DIAMOND);
       var0.addFeature(GenerationStep.Decoration.UNDERGROUND_ORES, Features.ORE_LAPIS);
+      var0.addFeature(GenerationStep.Decoration.UNDERGROUND_ORES, Features.ORE_LAPIS_BURIED);
+      var0.addFeature(GenerationStep.Decoration.UNDERGROUND_ORES, Features.ORE_COPPER);
+      var0.addFeature(GenerationStep.Decoration.UNDERGROUND_ORES, Features.UNDERWATER_MAGMA);
    }
 
    public static void addExtraGold(BiomeGenerationSettings.Builder var0) {
       var0.addFeature(GenerationStep.Decoration.UNDERGROUND_ORES, Features.ORE_GOLD_EXTRA);
    }
@@ -328,10 +342,14 @@
    public static void addAncientDebris(BiomeGenerationSettings.Builder var0) {
       var0.addFeature(GenerationStep.Decoration.UNDERGROUND_DECORATION, Features.ORE_DEBRIS_LARGE);
       var0.addFeature(GenerationStep.Decoration.UNDERGROUND_DECORATION, Features.ORE_DEBRIS_SMALL);
    }
 
+   public static void addDefaultCrystalFormations(BiomeGenerationSettings.Builder var0) {
+      var0.addFeature(GenerationStep.Decoration.UNDERGROUND_STRUCTURES, Features.AMETHYST_GEODE);
+   }
+
    public static void farmAnimals(MobSpawnSettings.Builder var0) {
       var0.addSpawn(MobCategory.CREATURE, new MobSpawnSettings.SpawnerData(EntityType.SHEEP, 12, 4, 4));
       var0.addSpawn(MobCategory.CREATURE, new MobSpawnSettings.SpawnerData(EntityType.PIG, 10, 4, 4));
       var0.addSpawn(MobCategory.CREATURE, new MobSpawnSettings.SpawnerData(EntityType.CHICKEN, 10, 4, 4));
       var0.addSpawn(MobCategory.CREATURE, new MobSpawnSettings.SpawnerData(EntityType.COW, 8, 4, 4));
