--- net/minecraft/data/worldgen/PillagerOutpostPools.java
+++ net/minecraft/data/worldgen/PillagerOutpostPools.java
@@ -7,10 +7,13 @@
 import net.minecraft.world.level.levelgen.feature.structures.StructureTemplatePool;
 
 public class PillagerOutpostPools {
    public static final StructureTemplatePool START;
 
+   public PillagerOutpostPools() {
+   }
+
    public static void bootstrap() {
    }
 
    static {
       START = Pools.register(new StructureTemplatePool(new ResourceLocation("pillager_outpost/base_plates"), new ResourceLocation("empty"), ImmutableList.of(Pair.of(StructurePoolElement.legacy("pillager_outpost/base_plate"), 1)), StructureTemplatePool.Projection.RIGID));
