--- net/minecraft/data/worldgen/Pools.java
+++ net/minecraft/data/worldgen/Pools.java
@@ -9,10 +9,13 @@
 
 public class Pools {
    public static final ResourceKey<StructureTemplatePool> EMPTY;
    private static final StructureTemplatePool BUILTIN_EMPTY;
 
+   public Pools() {
+   }
+
    public static StructureTemplatePool register(StructureTemplatePool var0) {
       return (StructureTemplatePool)BuiltinRegistries.register(BuiltinRegistries.TEMPLATE_POOL, (ResourceLocation)var0.getName(), var0);
    }
 
    public static StructureTemplatePool bootstrap() {
