--- net/minecraft/data/worldgen/StructureFeatures.java
+++ net/minecraft/data/worldgen/StructureFeatures.java
@@ -46,10 +46,13 @@
    public static final ConfiguredStructureFeature<RuinedPortalConfiguration, ? extends StructureFeature<RuinedPortalConfiguration>> RUINED_PORTAL_SWAMP;
    public static final ConfiguredStructureFeature<RuinedPortalConfiguration, ? extends StructureFeature<RuinedPortalConfiguration>> RUINED_PORTAL_MOUNTAIN;
    public static final ConfiguredStructureFeature<RuinedPortalConfiguration, ? extends StructureFeature<RuinedPortalConfiguration>> RUINED_PORTAL_OCEAN;
    public static final ConfiguredStructureFeature<RuinedPortalConfiguration, ? extends StructureFeature<RuinedPortalConfiguration>> RUINED_PORTAL_NETHER;
 
+   public StructureFeatures() {
+   }
+
    private static <FC extends FeatureConfiguration, F extends StructureFeature<FC>> ConfiguredStructureFeature<FC, F> register(String var0, ConfiguredStructureFeature<FC, F> var1) {
       return (ConfiguredStructureFeature)BuiltinRegistries.register(BuiltinRegistries.CONFIGURED_STRUCTURE_FEATURE, (String)var0, var1);
    }
 
    static {
