--- net/minecraft/data/worldgen/SurfaceBuilders.java
+++ net/minecraft/data/worldgen/SurfaceBuilders.java
@@ -30,10 +30,13 @@
    public static final ConfiguredSurfaceBuilder<SurfaceBuilderBaseConfiguration> STONE;
    public static final ConfiguredSurfaceBuilder<SurfaceBuilderBaseConfiguration> SWAMP;
    public static final ConfiguredSurfaceBuilder<SurfaceBuilderBaseConfiguration> WARPED_FOREST;
    public static final ConfiguredSurfaceBuilder<SurfaceBuilderBaseConfiguration> WOODED_BADLANDS;
 
+   public SurfaceBuilders() {
+   }
+
    private static <SC extends SurfaceBuilderConfiguration> ConfiguredSurfaceBuilder<SC> register(String var0, ConfiguredSurfaceBuilder<SC> var1) {
       return (ConfiguredSurfaceBuilder)BuiltinRegistries.register(BuiltinRegistries.CONFIGURED_SURFACE_BUILDER, (String)var0, var1);
    }
 
    static {
