--- net/minecraft/data/worldgen/TaigaVillagePools.java
+++ net/minecraft/data/worldgen/TaigaVillagePools.java
@@ -7,10 +7,13 @@
 import net.minecraft.world.level.levelgen.feature.structures.StructureTemplatePool;
 
 public class TaigaVillagePools {
    public static final StructureTemplatePool START;
 
+   public TaigaVillagePools() {
+   }
+
    public static void bootstrap() {
    }
 
    static {
       START = Pools.register(new StructureTemplatePool(new ResourceLocation("village/taiga/town_centers"), new ResourceLocation("empty"), ImmutableList.of(Pair.of(StructurePoolElement.legacy("village/taiga/town_centers/taiga_meeting_point_1", ProcessorLists.MOSSIFY_10_PERCENT), 49), Pair.of(StructurePoolElement.legacy("village/taiga/town_centers/taiga_meeting_point_2", ProcessorLists.MOSSIFY_10_PERCENT), 49), Pair.of(StructurePoolElement.legacy("village/taiga/zombie/town_centers/taiga_meeting_point_1", ProcessorLists.ZOMBIE_TAIGA), 1), Pair.of(StructurePoolElement.legacy("village/taiga/zombie/town_centers/taiga_meeting_point_2", ProcessorLists.ZOMBIE_TAIGA), 1)), StructureTemplatePool.Projection.RIGID));
