--- net/minecraft/data/worldgen/VillagePools.java
+++ net/minecraft/data/worldgen/VillagePools.java
@@ -1,8 +1,11 @@
 package net.minecraft.data.worldgen;
 
 public class VillagePools {
+   public VillagePools() {
+   }
+
    public static void bootstrap() {
       PlainVillagePools.bootstrap();
       SnowyVillagePools.bootstrap();
       SavannaVillagePools.bootstrap();
       DesertVillagePools.bootstrap();
