--- net/minecraft/client/resources/AssetIndex.java
+++ net/minecraft/client/resources/AssetIndex.java
@@ -46,11 +46,12 @@
                Entry var9 = (Entry)var8.next();
                JsonObject var10 = (JsonObject)var9.getValue();
                String var11 = (String)var9.getKey();
                String[] var12 = var11.split("/", 2);
                String var13 = GsonHelper.getAsString(var10, "hash");
-               File var14 = new File(var3, var13.substring(0, 2) + "/" + var13);
+               String var10003 = var13.substring(0, 2);
+               File var14 = new File(var3, var10003 + "/" + var13);
                if (var12.length == 1) {
                   this.rootFiles.put(var12[0], var14);
                } else {
                   this.namespacedFiles.put(new ResourceLocation(var12[0], var12[1]), var14);
                }
