--- net/minecraft/client/resources/DefaultPlayerSkin.java
+++ net/minecraft/client/resources/DefaultPlayerSkin.java
@@ -4,11 +4,16 @@
 import net.minecraft.resources.ResourceLocation;
 
 public class DefaultPlayerSkin {
    private static final ResourceLocation STEVE_SKIN_LOCATION = new ResourceLocation("textures/entity/steve.png");
    private static final ResourceLocation ALEX_SKIN_LOCATION = new ResourceLocation("textures/entity/alex.png");
+   private static final String STEVE_MODEL = "default";
+   private static final String ALEX_MODEL = "slim";
 
+   public DefaultPlayerSkin() {
+   }
+
    public static ResourceLocation getDefaultSkin() {
       return STEVE_SKIN_LOCATION;
    }
 
    public static ResourceLocation getDefaultSkin(UUID var0) {
