--- net/minecraft/client/resources/LegacyStuffWrapper.java
+++ net/minecraft/client/resources/LegacyStuffWrapper.java
@@ -5,56 +5,52 @@
 import net.minecraft.resources.ResourceLocation;
 import net.minecraft.server.packs.resources.Resource;
 import net.minecraft.server.packs.resources.ResourceManager;
 
 public class LegacyStuffWrapper {
+   public LegacyStuffWrapper() {
+   }
+
    @Deprecated
    public static int[] getPixels(ResourceManager var0, ResourceLocation var1) throws IOException {
       Resource var2 = var0.getResource(var1);
-      Throwable var3 = null;
 
-      Object var6;
+      int[] var4;
       try {
-         NativeImage var4 = NativeImage.read(var2.getInputStream());
-         Throwable var5 = null;
+         NativeImage var3 = NativeImage.read(var2.getInputStream());
 
          try {
-            var6 = var4.makePixelArray();
-         } catch (Throwable var29) {
-            var6 = var29;
-            var5 = var29;
-            throw var29;
-         } finally {
-            if (var4 != null) {
-               if (var5 != null) {
-                  try {
-                     var4.close();
-                  } catch (Throwable var28) {
-                     var5.addSuppressed(var28);
-                  }
-               } else {
-                  var4.close();
+            var4 = var3.makePixelArray();
+         } catch (Throwable var8) {
+            if (var3 != null) {
+               try {
+                  var3.close();
+               } catch (Throwable var7) {
+                  var8.addSuppressed(var7);
                }
             }
 
+            throw var8;
          }
-      } catch (Throwable var31) {
-         var3 = var31;
-         throw var31;
-      } finally {
+
+         if (var3 != null) {
+            var3.close();
+         }
+      } catch (Throwable var9) {
          if (var2 != null) {
-            if (var3 != null) {
-               try {
-                  var2.close();
-               } catch (Throwable var27) {
-                  var3.addSuppressed(var27);
-               }
-            } else {
+            try {
                var2.close();
+            } catch (Throwable var6) {
+               var9.addSuppressed(var6);
             }
          }
 
+         throw var9;
       }
 
-      return (int[])var6;
+      if (var2 != null) {
+         var2.close();
+      }
+
+      return var4;
    }
 }
