--- net/minecraft/client/resources/LegacyStuffWrapper.java
+++ net/minecraft/client/resources/LegacyStuffWrapper.java
@@ -5,10 +5,13 @@
 import net.minecraft.resources.ResourceLocation;
 import net.minecraft.server.packs.resources.Resource;
 import net.minecraft.server.packs.resources.ResourceManager;
 
 public class LegacyStuffWrapper {
+   public LegacyStuffWrapper() {
+   }
+
    @Deprecated
    public static int[] getPixels(ResourceManager var0, ResourceLocation var1) throws IOException {
       Resource var2 = var0.getResource(var1);
       Throwable var3 = null;
 
