--- net/minecraft/client/resources/PackResourcesAdapterV4.java
+++ net/minecraft/client/resources/PackResourcesAdapterV4.java
@@ -37,11 +37,15 @@
    private static final List<String> PATTERNS = Lists.newArrayList(new String[]{"base", "border", "bricks", "circle", "creeper", "cross", "curly_border", "diagonal_left", "diagonal_right", "diagonal_up_left", "diagonal_up_right", "flower", "globe", "gradient", "gradient_up", "half_horizontal", "half_horizontal_bottom", "half_vertical", "half_vertical_right", "mojang", "rhombus", "skull", "small_stripes", "square_bottom_left", "square_bottom_right", "square_top_left", "square_top_right", "straight_cross", "stripe_bottom", "stripe_center", "stripe_downleft", "stripe_downright", "stripe_left", "stripe_middle", "stripe_right", "stripe_top", "triangle_bottom", "triangle_top", "triangles_bottom", "triangles_top"});
    private static final Set<String> SHIELDS;
    private static final Set<String> BANNERS;
    public static final ResourceLocation SHIELD_BASE;
    public static final ResourceLocation BANNER_BASE;
+   public static final int DEFAULT_CHEST_SIZE = 64;
+   public static final int DEFAULT_SHIELD_SIZE = 64;
+   public static final int DEFAULT_BANNER_SIZE = 64;
    public static final ResourceLocation OLD_IRON_GOLEM_LOCATION;
+   public static final String NEW_IRON_GOLEM_PATH = "textures/entity/iron_golem/iron_golem.png";
    private final PackResources pack;
 
    public PackResourcesAdapterV4(PackResources var1) {
       this.pack = var1;
    }
