--- net/minecraft/client/resources/TextureAtlasHolder.java
+++ net/minecraft/client/resources/TextureAtlasHolder.java
@@ -25,11 +25,13 @@
    protected TextureAtlasSprite getSprite(ResourceLocation var1) {
       return this.textureAtlas.getSprite(this.resolveLocation(var1));
    }
 
    private ResourceLocation resolveLocation(ResourceLocation var1) {
-      return new ResourceLocation(var1.getNamespace(), this.prefix + "/" + var1.getPath());
+      String var10002 = var1.getNamespace();
+      String var10003 = this.prefix;
+      return new ResourceLocation(var10002, var10003 + "/" + var1.getPath());
    }
 
    protected TextureAtlas.Preparations prepare(ResourceManager var1, ProfilerFiller var2) {
       var2.startTick();
       var2.push("stitching");
