--- net/minecraft/client/resources/metadata/language/LanguageMetadataSection.java
+++ net/minecraft/client/resources/metadata/language/LanguageMetadataSection.java
@@ -3,10 +3,11 @@
 import java.util.Collection;
 import net.minecraft.client.resources.language.LanguageInfo;
 
 public class LanguageMetadataSection {
    public static final LanguageMetadataSectionSerializer SERIALIZER = new LanguageMetadataSectionSerializer();
+   public static final boolean DEFAULT_BIDIRECTIONAL = false;
    private final Collection<LanguageInfo> languages;
 
    public LanguageMetadataSection(Collection<LanguageInfo> var1) {
       this.languages = var1;
    }
