--- net/minecraft/client/resources/metadata/language/LanguageMetadataSectionSerializer.java
+++ net/minecraft/client/resources/metadata/language/LanguageMetadataSectionSerializer.java
@@ -10,10 +10,12 @@
 import net.minecraft.client.resources.language.LanguageInfo;
 import net.minecraft.server.packs.metadata.MetadataSectionSerializer;
 import net.minecraft.util.GsonHelper;
 
 public class LanguageMetadataSectionSerializer implements MetadataSectionSerializer<LanguageMetadataSection> {
+   private static final int MAX_LANGUAGE_LENGTH = 16;
+
    public LanguageMetadataSectionSerializer() {
    }
 
    public LanguageMetadataSection fromJson(JsonObject var1) {
       HashSet var2 = Sets.newHashSet();
@@ -29,11 +31,11 @@
          }
 
          Entry var4 = (Entry)var3.next();
          var5 = (String)var4.getKey();
          if (var5.length() > 16) {
-            throw new JsonParseException("Invalid language->'" + var5 + "': language code must not be more than " + 16 + " characters long");
+            throw new JsonParseException("Invalid language->'" + var5 + "': language code must not be more than 16 characters long");
          }
 
          JsonObject var6 = GsonHelper.convertToJsonObject((JsonElement)var4.getValue(), "language");
          var7 = GsonHelper.getAsString(var6, "region");
          var8 = GsonHelper.getAsString(var6, "name");
