--- net/minecraft/world/entity/animal/AbstractGolem.java
+++ net/minecraft/world/entity/animal/AbstractGolem.java
@@ -10,11 +10,11 @@
 public abstract class AbstractGolem extends PathfinderMob {
    protected AbstractGolem(EntityType<? extends AbstractGolem> var1, Level var2) {
       super(var1, var2);
    }
 
-   public boolean causeFallDamage(float var1, float var2) {
+   public boolean causeFallDamage(float var1, float var2, DamageSource var3) {
       return false;
    }
 
    @Nullable
    protected SoundEvent getAmbientSound() {
