--- net/minecraft/world/entity/animal/AbstractSchoolingFish.java
+++ net/minecraft/world/entity/animal/AbstractSchoolingFish.java
@@ -89,11 +89,11 @@
          this.getNavigation().moveTo((Entity)this.leader, 1.0D);
       }
 
    }
 
-   public void addFollowers(Stream<AbstractSchoolingFish> var1) {
+   public void addFollowers(Stream<? extends AbstractSchoolingFish> var1) {
       var1.limit((long)(this.getMaxSchoolSize() - this.schoolSize)).filter((var1x) -> {
          return var1x != this;
       }).forEach((var1x) -> {
          var1x.startFollowing(this);
       });
