--- net/minecraft/world/entity/animal/Chicken.java
+++ net/minecraft/world/entity/animal/Chicken.java
@@ -5,11 +5,11 @@
 import net.minecraft.server.level.ServerLevel;
 import net.minecraft.sounds.SoundEvent;
 import net.minecraft.sounds.SoundEvents;
 import net.minecraft.util.Mth;
 import net.minecraft.world.damagesource.DamageSource;
-import net.minecraft.world.entity.AgableMob;
+import net.minecraft.world.entity.AgeableMob;
 import net.minecraft.world.entity.Entity;
 import net.minecraft.world.entity.EntityDimensions;
 import net.minecraft.world.entity.EntityType;
 import net.minecraft.world.entity.LivingEntity;
 import net.minecraft.world.entity.Mob;
@@ -38,10 +38,11 @@
    public float flap;
    public float flapSpeed;
    public float oFlapSpeed;
    public float oFlap;
    public float flapping = 1.0F;
+   private float nextFlap = 1.0F;
    public int eggTime;
    public boolean isChickenJockey;
 
    public Chicken(EntityType<? extends Chicken> var1, Level var2) {
       super(var1, var2);
@@ -51,11 +52,11 @@
 
    protected void registerGoals() {
       this.goalSelector.addGoal(0, new FloatGoal(this));
       this.goalSelector.addGoal(1, new PanicGoal(this, 1.4D));
       this.goalSelector.addGoal(2, new BreedGoal(this, 1.0D));
-      this.goalSelector.addGoal(3, new TemptGoal(this, 1.0D, false, FOOD_ITEMS));
+      this.goalSelector.addGoal(3, new TemptGoal(this, 1.0D, FOOD_ITEMS, false));
       this.goalSelector.addGoal(4, new FollowParentGoal(this, 1.1D));
       this.goalSelector.addGoal(5, new WaterAvoidingRandomStrollGoal(this, 1.0D));
       this.goalSelector.addGoal(6, new LookAtPlayerGoal(this, Player.class, 6.0F));
       this.goalSelector.addGoal(7, new RandomLookAroundGoal(this));
    }
@@ -91,11 +92,19 @@
          this.eggTime = this.random.nextInt(6000) + 6000;
       }
 
    }
 
-   public boolean causeFallDamage(float var1, float var2) {
+   protected boolean isFlapping() {
+      return this.flyDist > this.nextFlap;
+   }
+
+   protected void onFlap() {
+      this.nextFlap = this.flyDist + this.flapSpeed / 2.0F;
+   }
+
+   public boolean causeFallDamage(float var1, float var2, DamageSource var3) {
       return false;
    }
 
    protected SoundEvent getAmbientSound() {
       return SoundEvents.CHICKEN_AMBIENT;
@@ -111,11 +120,11 @@
 
    protected void playStepSound(BlockPos var1, BlockState var2) {
       this.playSound(SoundEvents.CHICKEN_STEP, 0.15F, 1.0F);
    }
 
-   public Chicken getBreedOffspring(ServerLevel var1, AgableMob var2) {
+   public Chicken getBreedOffspring(ServerLevel var1, AgeableMob var2) {
       return (Chicken)EntityType.CHICKEN.create(var1);
    }
 
    public boolean isFood(ItemStack var1) {
       return FOOD_ITEMS.test(var1);
