--- net/minecraft/world/entity/animal/Panda.java
+++ net/minecraft/world/entity/animal/Panda.java
@@ -21,11 +21,11 @@
 import net.minecraft.util.Mth;
 import net.minecraft.world.DifficultyInstance;
 import net.minecraft.world.InteractionHand;
 import net.minecraft.world.InteractionResult;
 import net.minecraft.world.damagesource.DamageSource;
-import net.minecraft.world.entity.AgableMob;
+import net.minecraft.world.entity.AgeableMob;
 import net.minecraft.world.entity.Entity;
 import net.minecraft.world.entity.EntitySelector;
 import net.minecraft.world.entity.EntityType;
 import net.minecraft.world.entity.EquipmentSlot;
 import net.minecraft.world.entity.LivingEntity;
@@ -49,11 +49,10 @@
 import net.minecraft.world.entity.ai.goal.target.HurtByTargetGoal;
 import net.minecraft.world.entity.ai.targeting.TargetingConditions;
 import net.minecraft.world.entity.item.ItemEntity;
 import net.minecraft.world.entity.monster.Monster;
 import net.minecraft.world.entity.player.Player;
-import net.minecraft.world.item.Item;
 import net.minecraft.world.item.ItemStack;
 import net.minecraft.world.item.Items;
 import net.minecraft.world.item.crafting.Ingredient;
 import net.minecraft.world.level.GameRules;
 import net.minecraft.world.level.Level;
@@ -228,11 +227,11 @@
       this.setMainGene(Panda.Gene.byName(var1.getString("MainGene")));
       this.setHiddenGene(Panda.Gene.byName(var1.getString("HiddenGene")));
    }
 
    @Nullable
-   public AgableMob getBreedOffspring(ServerLevel var1, AgableMob var2) {
+   public AgeableMob getBreedOffspring(ServerLevel var1, AgeableMob var2) {
       Panda var3 = (Panda)EntityType.PANDA.create(var1);
       if (var2 instanceof Panda) {
          var3.setGeneFromParents(this, (Panda)var2);
       }
 
@@ -493,11 +492,11 @@
          this.onItemPickup(var1);
          ItemStack var2 = var1.getItem();
          this.setItemSlot(EquipmentSlot.MAINHAND, var2);
          this.handDropChances[EquipmentSlot.MAINHAND.getIndex()] = 2.0F;
          this.take(var1, var2.getCount());
-         var1.remove();
+         var1.discard();
       }
 
    }
 
    public boolean hurt(DamageSource var1, float var2) {
@@ -509,11 +508,11 @@
    public SpawnGroupData finalizeSpawn(ServerLevelAccessor var1, DifficultyInstance var2, MobSpawnType var3, @Nullable SpawnGroupData var4, @Nullable CompoundTag var5) {
       this.setMainGene(Panda.Gene.getRandom(this.random));
       this.setHiddenGene(Panda.Gene.getRandom(this.random));
       this.setAttributes();
       if (var4 == null) {
-         var4 = new AgableMob.AgableMobGroupData(0.2F);
+         var4 = new AgeableMob.AgeableMobGroupData(0.2F);
       }
 
       return super.finalizeSpawn(var1, var2, var3, (SpawnGroupData)var4, var5);
    }
 
@@ -592,11 +591,11 @@
             }
 
             this.tryToSit();
             this.eat(true);
             ItemStack var4 = this.getItemBySlot(EquipmentSlot.MAINHAND);
-            if (!var4.isEmpty() && !var1.abilities.instabuild) {
+            if (!var4.isEmpty() && !var1.getAbilities().instabuild) {
                this.spawnAtLocation(var4);
             }
 
             this.setItemSlot(EquipmentSlot.MAINHAND, new ItemStack(var3.getItem(), 1));
             this.usePlayerItem(var1, var3);
@@ -620,15 +619,15 @@
    protected void playStepSound(BlockPos var1, BlockState var2) {
       this.playSound(SoundEvents.PANDA_STEP, 0.15F, 1.0F);
    }
 
    public boolean isFood(ItemStack var1) {
-      return var1.getItem() == Blocks.BAMBOO.asItem();
+      return var1.is(Blocks.BAMBOO.asItem());
    }
 
    private boolean isFoodOrCake(ItemStack var1) {
-      return this.isFood(var1) || var1.getItem() == Blocks.CAKE.asItem();
+      return this.isFood(var1) || var1.is(Blocks.CAKE.asItem());
    }
 
    @Nullable
    protected SoundEvent getDeathSound() {
       return SoundEvents.PANDA_DEATH;
@@ -650,12 +649,12 @@
       MAIN_GENE_ID = SynchedEntityData.defineId(Panda.class, EntityDataSerializers.BYTE);
       HIDDEN_GENE_ID = SynchedEntityData.defineId(Panda.class, EntityDataSerializers.BYTE);
       DATA_ID_FLAGS = SynchedEntityData.defineId(Panda.class, EntityDataSerializers.BYTE);
       BREED_TARGETING = (new TargetingConditions()).range(8.0D).allowSameTeam().allowInvulnerable();
       PANDA_ITEMS = (var0) -> {
-         Item var1 = var0.getItem().getItem();
-         return (var1 == Blocks.BAMBOO.asItem() || var1 == Blocks.CAKE.asItem()) && var0.isAlive() && !var0.hasPickUpDelay();
+         ItemStack var1 = var0.getItem();
+         return (var1.is(Blocks.BAMBOO.asItem()) || var1.is(Blocks.CAKE.asItem())) && var0.isAlive() && !var0.hasPickUpDelay();
       };
    }
 
    static class PandaPanicGoal extends PanicGoal {
       private final Panda panda;
@@ -969,11 +968,13 @@
          } else {
             if (this.lookAt == null) {
                if (this.lookAtType == Player.class) {
                   this.lookAt = this.mob.level.getNearestPlayer(this.lookAtContext, this.mob, this.mob.getX(), this.mob.getEyeY(), this.mob.getZ());
                } else {
-                  this.lookAt = this.mob.level.getNearestLoadedEntity(this.lookAtType, this.lookAtContext, this.mob, this.mob.getX(), this.mob.getEyeY(), this.mob.getZ(), this.mob.getBoundingBox().inflate((double)this.lookDistance, 3.0D, (double)this.lookDistance));
+                  this.lookAt = this.mob.level.getNearestEntity(this.mob.level.getEntitiesOfClass(this.lookAtType, this.mob.getBoundingBox().inflate((double)this.lookDistance, 3.0D, (double)this.lookDistance), (var0) -> {
+                     return true;
+                  }), this.lookAtContext, this.mob, this.mob.getX(), this.mob.getEyeY(), this.mob.getZ());
                }
             }
 
             return this.panda.canPerformAction() && this.lookAt != null;
          }
