--- net/minecraft/world/entity/animal/Parrot.java
+++ net/minecraft/world/entity/animal/Parrot.java
@@ -27,11 +27,11 @@
 import net.minecraft.world.InteractionHand;
 import net.minecraft.world.InteractionResult;
 import net.minecraft.world.damagesource.DamageSource;
 import net.minecraft.world.effect.MobEffectInstance;
 import net.minecraft.world.effect.MobEffects;
-import net.minecraft.world.entity.AgableMob;
+import net.minecraft.world.entity.AgeableMob;
 import net.minecraft.world.entity.Entity;
 import net.minecraft.world.entity.EntityDimensions;
 import net.minecraft.world.entity.EntityType;
 import net.minecraft.world.entity.Mob;
 import net.minecraft.world.entity.MobSpawnType;
@@ -86,11 +86,11 @@
 
    @Nullable
    public SpawnGroupData finalizeSpawn(ServerLevelAccessor var1, DifficultyInstance var2, MobSpawnType var3, @Nullable SpawnGroupData var4, @Nullable CompoundTag var5) {
       this.setVariant(this.random.nextInt(5));
       if (var4 == null) {
-         var4 = new AgableMob.AgableMobGroupData(false);
+         var4 = new AgeableMob.AgeableMobGroupData(false);
       }
 
       return super.finalizeSpawn(var1, var2, var3, (SpawnGroupData)var4, var5);
    }
 
@@ -185,11 +185,11 @@
    }
 
    public InteractionResult mobInteract(Player var1, InteractionHand var2) {
       ItemStack var3 = var1.getItemInHand(var2);
       if (!this.isTame() && TAME_FOOD.contains(var3.getItem())) {
-         if (!var1.abilities.instabuild) {
+         if (!var1.getAbilities().instabuild) {
             var3.shrink(1);
          }
 
          if (!this.isSilent()) {
             this.level.playSound((Player)null, this.getX(), this.getY(), this.getZ(), SoundEvents.PARROT_EAT, this.getSoundSource(), 1.0F, 1.0F + (this.random.nextFloat() - this.random.nextFloat()) * 0.2F);
@@ -203,12 +203,12 @@
                this.level.broadcastEntityEvent(this, (byte)6);
             }
          }
 
          return InteractionResult.sidedSuccess(this.level.isClientSide);
-      } else if (var3.getItem() == POISONOUS_FOOD) {
-         if (!var1.abilities.instabuild) {
+      } else if (var3.is(POISONOUS_FOOD)) {
+         if (!var1.getAbilities().instabuild) {
             var3.shrink(1);
          }
 
          this.addEffect(new MobEffectInstance(MobEffects.POISON, 900));
          if (var1.isCreative() || !this.isInvulnerable()) {
@@ -234,11 +234,11 @@
    public static boolean checkParrotSpawnRules(EntityType<Parrot> var0, LevelAccessor var1, MobSpawnType var2, BlockPos var3, Random var4) {
       BlockState var5 = var1.getBlockState(var3.below());
       return (var5.is(BlockTags.LEAVES) || var5.is(Blocks.GRASS_BLOCK) || var5.is(BlockTags.LOGS) || var5.is(Blocks.AIR)) && var1.getRawBrightness(var3, 0) > 8;
    }
 
-   public boolean causeFallDamage(float var1, float var2) {
+   public boolean causeFallDamage(float var1, float var2, DamageSource var3) {
       return false;
    }
 
    protected void checkFallDamage(double var1, boolean var3, BlockState var4, BlockPos var5) {
    }
@@ -246,11 +246,11 @@
    public boolean canMate(Animal var1) {
       return false;
    }
 
    @Nullable
-   public AgableMob getBreedOffspring(ServerLevel var1, AgableMob var2) {
+   public AgeableMob getBreedOffspring(ServerLevel var1, AgeableMob var2) {
       return null;
    }
 
    public boolean doHurtTarget(Entity var1) {
       return var1.hurt(DamageSource.mobAttack(this), 3.0F);
