--- net/minecraft/world/entity/animal/Pufferfish.java
+++ net/minecraft/world/entity/animal/Pufferfish.java
@@ -32,10 +32,11 @@
    private int deflateTimer;
    private static final Predicate<LivingEntity> NO_SPECTATORS_AND_NO_WATER_MOB;
 
    public Pufferfish(EntityType<? extends Pufferfish> var1, Level var2) {
       super(var1, var2);
+      this.refreshDimensions();
    }
 
    protected void defineSynchedData() {
       super.defineSynchedData();
       this.entityData.define(PUFF_STATE, 0);
