--- net/minecraft/server/players/GameProfileCache.java
+++ net/minecraft/server/players/GameProfileCache.java
@@ -39,10 +39,12 @@
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
 
 public class GameProfileCache {
    private static final Logger LOGGER = LogManager.getLogger();
+   private static final int GAMEPROFILES_MRU_LIMIT = 1000;
+   private static final int GAMEPROFILES_EXPIRATION_MONTHS = 1;
    private static boolean usesAuthentication;
    private final Map<String, GameProfileCache.GameProfileInfo> profilesByName = Maps.newConcurrentMap();
    private final Map<UUID, GameProfileCache.GameProfileInfo> profilesByUUID = Maps.newConcurrentMap();
    private final GameProfileRepository profileRepository;
    private final Gson gson = (new GsonBuilder()).create();
