--- net/minecraft/world/entity/ai/sensing/GolemSensor.java
+++ net/minecraft/world/entity/ai/sensing/GolemSensor.java
@@ -8,10 +8,13 @@
 import net.minecraft.world.entity.EntityType;
 import net.minecraft.world.entity.LivingEntity;
 import net.minecraft.world.entity.ai.memory.MemoryModuleType;
 
 public class GolemSensor extends Sensor<LivingEntity> {
+   private static final int GOLEM_SCAN_RATE = 200;
+   private static final int MEMORY_TIME_TO_LIVE = 600;
+
    public GolemSensor() {
       this(200);
    }
 
    public GolemSensor(int var1) {
@@ -21,15 +24,15 @@
    protected void doTick(ServerLevel var1, LivingEntity var2) {
       checkForNearbyGolem(var2);
    }
 
    public Set<MemoryModuleType<?>> requires() {
-      return ImmutableSet.of(MemoryModuleType.LIVING_ENTITIES);
+      return ImmutableSet.of(MemoryModuleType.NEAREST_LIVING_ENTITIES);
    }
 
    public static void checkForNearbyGolem(LivingEntity var0) {
-      Optional var1 = var0.getBrain().getMemory(MemoryModuleType.LIVING_ENTITIES);
+      Optional var1 = var0.getBrain().getMemory(MemoryModuleType.NEAREST_LIVING_ENTITIES);
       if (var1.isPresent()) {
          boolean var2 = ((List)var1.get()).stream().anyMatch((var0x) -> {
             return var0x.getType().equals(EntityType.IRON_GOLEM);
          });
          if (var2) {
