--- net/minecraft/world/entity/ai/sensing/NearestItemSensor.java
+++ net/minecraft/world/entity/ai/sensing/NearestItemSensor.java
@@ -11,10 +11,14 @@
 import net.minecraft.world.entity.ai.Brain;
 import net.minecraft.world.entity.ai.memory.MemoryModuleType;
 import net.minecraft.world.entity.item.ItemEntity;
 
 public class NearestItemSensor extends Sensor<Mob> {
+   private static final long XZ_RANGE = 8L;
+   private static final long Y_RANGE = 4L;
+   public static final int MAX_DISTANCE_TO_WANTED_ITEM = 9;
+
    public NearestItemSensor() {
    }
 
    public Set<MemoryModuleType<?>> requires() {
       return ImmutableSet.of(MemoryModuleType.NEAREST_VISIBLE_WANTED_ITEM);
