--- net/minecraft/world/entity/ai/sensing/Sensor.java
+++ net/minecraft/world/entity/ai/sensing/Sensor.java
@@ -7,10 +7,12 @@
 import net.minecraft.world.entity.ai.memory.MemoryModuleType;
 import net.minecraft.world.entity.ai.targeting.TargetingConditions;
 
 public abstract class Sensor<E extends LivingEntity> {
    private static final Random RANDOM = new Random();
+   private static final int DEFAULT_SCAN_RATE = 20;
+   protected static final int TARGETING_RANGE = 16;
    private static final TargetingConditions TARGET_CONDITIONS = (new TargetingConditions()).range(16.0D).allowSameTeam().allowNonAttackable();
    private static final TargetingConditions TARGET_CONDITIONS_IGNORE_INVISIBILITY_TESTING = (new TargetingConditions()).range(16.0D).allowSameTeam().allowNonAttackable().ignoreInvisibilityTesting();
    private final int scanRate;
    private long timeToTick;
 
