--- net/minecraft/world/entity/ai/sensing/Sensing.java
+++ net/minecraft/world/entity/ai/sensing/Sensing.java
@@ -17,18 +17,18 @@
    public void tick() {
       this.seen.clear();
       this.unseen.clear();
    }
 
-   public boolean canSee(Entity var1) {
+   public boolean hasLineOfSight(Entity var1) {
       if (this.seen.contains(var1)) {
          return true;
       } else if (this.unseen.contains(var1)) {
          return false;
       } else {
-         this.mob.level.getProfiler().push("canSee");
-         boolean var2 = this.mob.canSee(var1);
+         this.mob.level.getProfiler().push("hasLineOfSight");
+         boolean var2 = this.mob.hasLineOfSight(var1);
          this.mob.level.getProfiler().pop();
          if (var2) {
             this.seen.add(var1);
          } else {
             this.unseen.add(var1);
