--- net/minecraft/world/entity/ai/control/JumpControl.java
+++ net/minecraft/world/entity/ai/control/JumpControl.java
@@ -1,10 +1,10 @@
 package net.minecraft.world.entity.ai.control;
 
 import net.minecraft.world.entity.Mob;
 
-public class JumpControl {
+public class JumpControl implements Control {
    private final Mob mob;
    protected boolean jump;
 
    public JumpControl(Mob var1) {
       this.mob = var1;
