--- net/minecraft/world/entity/ai/control/LookControl.java
+++ net/minecraft/world/entity/ai/control/LookControl.java
@@ -4,11 +4,11 @@
 import net.minecraft.world.entity.Entity;
 import net.minecraft.world.entity.LivingEntity;
 import net.minecraft.world.entity.Mob;
 import net.minecraft.world.phys.Vec3;
 
-public class LookControl {
+public class LookControl implements Control {
    protected final Mob mob;
    protected float yMaxRotSpeed;
    protected float xMaxRotAngle;
    protected boolean hasWanted;
    protected double wantedX;
@@ -21,10 +21,14 @@
 
    public void setLookAt(Vec3 var1) {
       this.setLookAt(var1.x, var1.y, var1.z);
    }
 
+   public void setLookAt(Entity var1) {
+      this.setLookAt(var1.getX(), getWantedY(var1), var1.getZ());
+   }
+
    public void setLookAt(Entity var1, float var2, float var3) {
       this.setLookAt(var1.getX(), getWantedY(var1), var1.getZ(), var2, var3);
    }
 
    public void setLookAt(double var1, double var3, double var5) {
