--- net/minecraft/world/item/crafting/AbstractCookingRecipe.java
+++ net/minecraft/world/item/crafting/AbstractCookingRecipe.java
@@ -31,10 +31,14 @@
 
    public ItemStack assemble(Container var1) {
       return this.result.copy();
    }
 
+   public boolean canCraftInDimensions(int var1, int var2) {
+      return true;
+   }
+
    public NonNullList<Ingredient> getIngredients() {
       NonNullList var1 = NonNullList.create();
       var1.add(this.ingredient);
       return var1;
    }
@@ -45,10 +49,14 @@
 
    public ItemStack getResultItem() {
       return this.result;
    }
 
+   public String getGroup() {
+      return this.group;
+   }
+
    public int getCookingTime() {
       return this.cookingTime;
    }
 
    public ResourceLocation getId() {
