--- net/minecraft/world/item/crafting/ArmorDyeRecipe.java
+++ net/minecraft/world/item/crafting/ArmorDyeRecipe.java
@@ -70,9 +70,13 @@
       } else {
          return ItemStack.EMPTY;
       }
    }
 
+   public boolean canCraftInDimensions(int var1, int var2) {
+      return var1 * var2 >= 2;
+   }
+
    public RecipeSerializer<?> getSerializer() {
       return RecipeSerializer.ARMOR_DYE;
    }
 }
