--- net/minecraft/world/item/crafting/FireworkRocketRecipe.java
+++ net/minecraft/world/item/crafting/FireworkRocketRecipe.java
@@ -70,10 +70,14 @@
       }
 
       return var2;
    }
 
+   public boolean canCraftInDimensions(int var1, int var2) {
+      return var1 * var2 >= 2;
+   }
+
    public ItemStack getResultItem() {
       return new ItemStack(Items.FIREWORK_ROCKET);
    }
 
    public RecipeSerializer<?> getSerializer() {
