--- net/minecraft/world/item/crafting/FireworkStarRecipe.java
+++ net/minecraft/world/item/crafting/FireworkStarRecipe.java
@@ -98,10 +98,14 @@
       var3.putIntArray("Colors", (List)var5);
       var3.putByte("Type", (byte)var4.getId());
       return var2;
    }
 
+   public boolean canCraftInDimensions(int var1, int var2) {
+      return var1 * var2 >= 2;
+   }
+
    public ItemStack getResultItem() {
       return new ItemStack(Items.FIREWORK_STAR);
    }
 
    public RecipeSerializer<?> getSerializer() {
