--- net/minecraft/world/item/crafting/Ingredient.java
+++ net/minecraft/world/item/crafting/Ingredient.java
@@ -63,11 +63,11 @@
             ItemStack[] var2 = this.itemStacks;
             int var3 = var2.length;
 
             for(int var4 = 0; var4 < var3; ++var4) {
                ItemStack var5 = var2[var4];
-               if (var5.getItem() == var1.getItem()) {
+               if (var5.is(var1.getItem())) {
                   return true;
                }
             }
 
             return false;
