--- net/minecraft/world/item/crafting/MapExtendingRecipe.java
+++ net/minecraft/world/item/crafting/MapExtendingRecipe.java
@@ -22,11 +22,11 @@
       } else {
          ItemStack var3 = ItemStack.EMPTY;
 
          for(int var4 = 0; var4 < var1.getContainerSize() && var3.isEmpty(); ++var4) {
             ItemStack var5 = var1.getItem(var4);
-            if (var5.getItem() == Items.FILLED_MAP) {
+            if (var5.is(Items.FILLED_MAP)) {
                var3 = var5;
             }
          }
 
          if (var3.isEmpty()) {
@@ -62,11 +62,11 @@
    public ItemStack assemble(CraftingContainer var1) {
       ItemStack var2 = ItemStack.EMPTY;
 
       for(int var3 = 0; var3 < var1.getContainerSize() && var2.isEmpty(); ++var3) {
          ItemStack var4 = var1.getItem(var3);
-         if (var4.getItem() == Items.FILLED_MAP) {
+         if (var4.is(Items.FILLED_MAP)) {
             var2 = var4;
          }
       }
 
       var2 = var2.copy();
