--- net/minecraft/world/item/crafting/ShapedRecipe.java
+++ net/minecraft/world/item/crafting/ShapedRecipe.java
@@ -258,11 +258,11 @@
       }
 
       public ShapedRecipe fromNetwork(ResourceLocation var1, FriendlyByteBuf var2) {
          int var3 = var2.readVarInt();
          int var4 = var2.readVarInt();
-         String var5 = var2.readUtf(32767);
+         String var5 = var2.readUtf();
          NonNullList var6 = NonNullList.withSize(var3 * var4, Ingredient.EMPTY);
 
          for(int var7 = 0; var7 < var6.size(); ++var7) {
             var6.set(var7, Ingredient.fromNetwork(var2));
          }
