--- net/minecraft/world/item/crafting/ShapelessRecipe.java
+++ net/minecraft/world/item/crafting/ShapelessRecipe.java
@@ -91,11 +91,11 @@
 
          return var1;
       }
 
       public ShapelessRecipe fromNetwork(ResourceLocation var1, FriendlyByteBuf var2) {
-         String var3 = var2.readUtf(32767);
+         String var3 = var2.readUtf();
          int var4 = var2.readVarInt();
          NonNullList var5 = NonNullList.withSize(var4, Ingredient.EMPTY);
 
          for(int var6 = 0; var6 < var5.size(); ++var6) {
             var5.set(var6, Ingredient.fromNetwork(var2));
