--- net/minecraft/world/item/crafting/ShieldDecorationRecipe.java
+++ net/minecraft/world/item/crafting/ShieldDecorationRecipe.java
@@ -25,11 +25,11 @@
                   return false;
                }
 
                var4 = var6;
             } else {
-               if (var6.getItem() != Items.SHIELD) {
+               if (!var6.is(Items.SHIELD)) {
                   return false;
                }
 
                if (!var3.isEmpty()) {
                   return false;
@@ -58,11 +58,11 @@
       for(int var4 = 0; var4 < var1.getContainerSize(); ++var4) {
          ItemStack var5 = var1.getItem(var4);
          if (!var5.isEmpty()) {
             if (var5.getItem() instanceof BannerItem) {
                var2 = var5;
-            } else if (var5.getItem() == Items.SHIELD) {
+            } else if (var5.is(Items.SHIELD)) {
                var3 = var5.copy();
             }
          }
       }
 
@@ -75,9 +75,13 @@
          var3.addTagElement("BlockEntityTag", var7);
          return var3;
       }
    }
 
+   public boolean canCraftInDimensions(int var1, int var2) {
+      return var1 * var2 >= 2;
+   }
+
    public RecipeSerializer<?> getSerializer() {
       return RecipeSerializer.SHIELD_DECORATION;
    }
 }
