--- net/minecraft/world/entity/schedule/Activity.java
+++ net/minecraft/world/entity/schedule/Activity.java
@@ -16,10 +16,12 @@
    public static final Activity FIGHT = register("fight");
    public static final Activity CELEBRATE = register("celebrate");
    public static final Activity ADMIRE_ITEM = register("admire_item");
    public static final Activity AVOID = register("avoid");
    public static final Activity RIDE = register("ride");
+   public static final Activity PLAY_DEAD = register("play_dead");
+   public static final Activity LONG_JUMP = register("long_jump");
    private final String name;
    private final int hashCode;
 
    private Activity(String var1) {
       this.name = var1;
