--- net/minecraft/commands/arguments/coordinates/WorldCoordinate.java
+++ net/minecraft/commands/arguments/coordinates/WorldCoordinate.java
@@ -4,10 +4,11 @@
 import com.mojang.brigadier.exceptions.CommandSyntaxException;
 import com.mojang.brigadier.exceptions.SimpleCommandExceptionType;
 import net.minecraft.network.chat.TranslatableComponent;
 
 public class WorldCoordinate {
+   private static final char PREFIX_RELATIVE = '~';
    public static final SimpleCommandExceptionType ERROR_EXPECTED_DOUBLE = new SimpleCommandExceptionType(new TranslatableComponent("argument.pos.missing.double"));
    public static final SimpleCommandExceptionType ERROR_EXPECTED_INT = new SimpleCommandExceptionType(new TranslatableComponent("argument.pos.missing.int"));
    private final boolean relative;
    private final double value;
 
