--- net/minecraft/server/bossevents/CustomBossEvent.java
+++ net/minecraft/server/bossevents/CustomBossEvent.java
@@ -26,11 +26,11 @@
    private int max = 100;
 
    public CustomBossEvent(ResourceLocation var1, Component var2) {
       super(var2, BossEvent.BossBarColor.WHITE, BossEvent.BossBarOverlay.PROGRESS);
       this.id = var1;
-      this.setPercent(0.0F);
+      this.setProgress(0.0F);
    }
 
    public ResourceLocation getTextId() {
       return this.id;
    }
@@ -62,16 +62,16 @@
       return this.max;
    }
 
    public void setValue(int var1) {
       this.value = var1;
-      this.setPercent(Mth.clamp((float)var1 / (float)this.max, 0.0F, 1.0F));
+      this.setProgress(Mth.clamp((float)var1 / (float)this.max, 0.0F, 1.0F));
    }
 
    public void setMax(int var1) {
       this.max = var1;
-      this.setPercent(Mth.clamp((float)this.value / (float)var1, 0.0F, 1.0F));
+      this.setProgress(Mth.clamp((float)this.value / (float)var1, 0.0F, 1.0F));
    }
 
    public final Component getDisplayName() {
       return ComponentUtils.wrapInSquareBrackets(this.getName()).withStyle((var1) -> {
          return var1.withColor(this.getColor().getFormatting()).withHoverEvent(new HoverEvent(HoverEvent.Action.SHOW_TEXT, new TextComponent(this.getTextId().toString()))).withInsertion(this.getTextId().toString());
