--- net/minecraft/util/ClassInstanceMultiMap.java
+++ net/minecraft/util/ClassInstanceMultiMap.java
@@ -8,10 +8,11 @@
 import java.util.Collection;
 import java.util.Collections;
 import java.util.Iterator;
 import java.util.List;
 import java.util.Map;
+import java.util.Objects;
 import java.util.Map.Entry;
 import java.util.stream.Collectors;
 import java.util.stream.Stream;
 
 public class ClassInstanceMultiMap<T> extends AbstractCollection<T> {
@@ -61,11 +62,11 @@
       if (!this.baseClass.isAssignableFrom(var1)) {
          throw new IllegalArgumentException("Don't know how to search for " + var1);
       } else {
          List var2 = (List)this.byClass.computeIfAbsent(var1, (var1x) -> {
             Stream var10000 = this.allInstances.stream();
-            var1x.getClass();
+            Objects.requireNonNull(var1x);
             return (List)var10000.filter(var1x::isInstance).collect(Collectors.toList());
          });
          return Collections.unmodifiableCollection(var2);
       }
    }
