--- net/minecraft/util/Crypt.java
+++ net/minecraft/util/Crypt.java
@@ -13,10 +13,20 @@
 import javax.crypto.SecretKey;
 import javax.crypto.spec.IvParameterSpec;
 import javax.crypto.spec.SecretKeySpec;
 
 public class Crypt {
+   private static final String SYMMETRIC_ALGORITHM = "AES";
+   private static final int SYMMETRIC_BITS = 128;
+   private static final String ASYMMETRIC_ALGORITHM = "RSA";
+   private static final int ASYMMETRIC_BITS = 1024;
+   private static final String BYTE_ENCODING = "ISO_8859_1";
+   private static final String HASH_ALGORITHM = "SHA-1";
+
+   public Crypt() {
+   }
+
    public static SecretKey generateSecretKey() throws CryptException {
       try {
          KeyGenerator var0 = KeyGenerator.getInstance("AES");
          var0.init(128);
          return var0.generateKey();
