--- net/minecraft/util/CsvOutput.java
+++ net/minecraft/util/CsvOutput.java
@@ -8,14 +8,16 @@
 import java.util.stream.Stream;
 import javax.annotation.Nullable;
 import org.apache.commons.lang3.StringEscapeUtils;
 
 public class CsvOutput {
+   private static final String LINE_SEPARATOR = "\r\n";
+   private static final String FIELD_SEPARATOR = ",";
    private final Writer output;
    private final int columnCount;
 
-   private CsvOutput(Writer var1, List<String> var2) throws IOException {
+   CsvOutput(Writer var1, List<String> var2) throws IOException {
       this.output = var1;
       this.columnCount = var2.size();
       this.writeLine(var2.stream());
    }
 
@@ -30,11 +32,13 @@
          this.writeLine(Stream.of(var1));
       }
    }
 
    private void writeLine(Stream<?> var1) throws IOException {
-      this.output.write((String)var1.map(CsvOutput::getStringValue).collect(Collectors.joining(",")) + "\r\n");
+      Writer var10000 = this.output;
+      Stream var10001 = var1.map(CsvOutput::getStringValue);
+      var10000.write((String)var10001.collect(Collectors.joining(",")) + "\r\n");
    }
 
    private static String getStringValue(@Nullable Object var0) {
       return StringEscapeUtils.escapeCsv(var0 != null ? var0.toString() : "[null]");
    }
