--- net/minecraft/util/CubicSampler.java
+++ net/minecraft/util/CubicSampler.java
@@ -1,14 +1,18 @@
 package net.minecraft.util;
 
-import javax.annotation.Nonnull;
+import net.minecraft.obfuscate.DontObfuscate;
 import net.minecraft.world.phys.Vec3;
 
 public class CubicSampler {
+   private static final int GAUSSIAN_SAMPLE_RADIUS = 2;
+   private static final int GAUSSIAN_SAMPLE_BREADTH = 6;
    private static final double[] GAUSSIAN_SAMPLE_KERNEL = new double[]{0.0D, 1.0D, 4.0D, 6.0D, 4.0D, 1.0D, 0.0D};
 
-   @Nonnull
+   private CubicSampler() {
+   }
+
    public static Vec3 gaussianSampleVec3(Vec3 var0, CubicSampler.Vec3Fetcher var1) {
       int var2 = Mth.floor(var0.x());
       int var3 = Mth.floor(var0.y());
       int var4 = Mth.floor(var0.z());
       double var5 = var0.x() - (double)var2;
@@ -37,9 +41,10 @@
 
       var13 = var13.scale(1.0D / var11);
       return var13;
    }
 
+   @DontObfuscate
    public interface Vec3Fetcher {
       Vec3 fetch(int var1, int var2, int var3);
    }
 }
