--- net/minecraft/util/DirectoryLock.java
+++ net/minecraft/util/DirectoryLock.java
@@ -11,10 +11,11 @@
 import java.nio.file.NoSuchFileException;
 import java.nio.file.Path;
 import java.nio.file.StandardOpenOption;
 
 public class DirectoryLock implements AutoCloseable {
+   public static final String LOCK_FILE = "session.lock";
    private final FileChannel lockFile;
    private final FileLock lock;
    private static final ByteBuffer DUMMY;
 
    public static DirectoryLock create(Path var0) throws IOException {
