--- net/minecraft/util/LazyLoadedValue.java
+++ net/minecraft/util/LazyLoadedValue.java
@@ -1,22 +1,19 @@
 package net.minecraft.util;
 
+import com.google.common.base.Suppliers;
+import java.util.Objects;
 import java.util.function.Supplier;
 
+@Deprecated
 public class LazyLoadedValue<T> {
-   private Supplier<T> factory;
-   private T value;
+   private final Supplier<T> factory;
 
    public LazyLoadedValue(Supplier<T> var1) {
-      this.factory = var1;
+      Objects.requireNonNull(var1);
+      this.factory = Suppliers.memoize(var1::get);
    }
 
    public T get() {
-      Supplier var1 = this.factory;
-      if (var1 != null) {
-         this.value = var1.get();
-         this.factory = null;
-      }
-
-      return this.value;
+      return this.factory.get();
    }
 }
