--- net/minecraft/util/LowerCaseEnumTypeAdapterFactory.java
+++ net/minecraft/util/LowerCaseEnumTypeAdapterFactory.java
@@ -53,9 +53,9 @@
             }
          };
       }
    }
 
-   private String toLowercase(Object var1) {
+   String toLowercase(Object var1) {
       return var1 instanceof Enum ? ((Enum)var1).name().toLowerCase(Locale.ROOT) : var1.toString().toLowerCase(Locale.ROOT);
    }
 }
