--- net/minecraft/util/StringDecomposer.java
+++ net/minecraft/util/StringDecomposer.java
@@ -4,12 +4,16 @@
 import net.minecraft.ChatFormatting;
 import net.minecraft.network.chat.FormattedText;
 import net.minecraft.network.chat.Style;
 
 public class StringDecomposer {
+   private static final char REPLACEMENT_CHAR = '\ufffd';
    private static final Optional<Object> STOP_ITERATION;
 
+   public StringDecomposer() {
+   }
+
    private static boolean feedChar(Style var0, FormattedCharSink var1, int var2, char var3) {
       return Character.isSurrogate(var3) ? var1.accept(var2, var0, 65533) : var1.accept(var2, var0, var3);
    }
 
    public static boolean iterate(String var0, Style var1, FormattedCharSink var2) {
