--- net/minecraft/util/TimeUtil.java
+++ net/minecraft/util/TimeUtil.java
@@ -1,7 +1,12 @@
 package net.minecraft.util;
 
+import net.minecraft.util.valueproviders.UniformInt;
+
 public class TimeUtil {
-   public static IntRange rangeOfSeconds(int var0, int var1) {
-      return new IntRange(var0 * 20, var1 * 20);
+   public TimeUtil() {
    }
+
+   public static UniformInt rangeOfSeconds(int var0, int var1) {
+      return UniformInt.of(var0 * 20, var1 * 20);
+   }
 }
