--- net/minecraft/world/entity/boss/enderdragon/phases/DragonChargePlayerPhase.java
+++ net/minecraft/world/entity/boss/enderdragon/phases/DragonChargePlayerPhase.java
@@ -6,10 +6,11 @@
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
 
 public class DragonChargePlayerPhase extends AbstractDragonPhaseInstance {
    private static final Logger LOGGER = LogManager.getLogger();
+   private static final int CHARGE_RECOVERY_TIME = 10;
    private Vec3 targetLocation;
    private int timeSinceCharge;
 
    public DragonChargePlayerPhase(EnderDragon var1) {
       super(var1);
