--- net/minecraft/world/entity/boss/enderdragon/phases/DragonSittingScanningPhase.java
+++ net/minecraft/world/entity/boss/enderdragon/phases/DragonSittingScanningPhase.java
@@ -5,10 +5,14 @@
 import net.minecraft.world.entity.boss.enderdragon.EnderDragon;
 import net.minecraft.world.entity.player.Player;
 import net.minecraft.world.phys.Vec3;
 
 public class DragonSittingScanningPhase extends AbstractDragonSittingPhase {
+   private static final int SITTING_SCANNING_IDLE_TICKS = 100;
+   private static final int SITTING_ATTACK_Y_VIEW_RANGE = 10;
+   private static final int SITTING_ATTACK_VIEW_RANGE = 20;
+   private static final int SITTING_CHARGE_VIEW_RANGE = 150;
    private static final TargetingConditions CHARGE_TARGETING = (new TargetingConditions()).range(150.0D);
    private final TargetingConditions scanTargeting;
    private int scanningTime;
 
    public DragonSittingScanningPhase(EnderDragon var1) {
