--- net/minecraft/world/entity/boss/enderdragon/phases/DragonSittingFlamingPhase.java
+++ net/minecraft/world/entity/boss/enderdragon/phases/DragonSittingFlamingPhase.java
@@ -8,10 +8,13 @@
 import net.minecraft.world.entity.AreaEffectCloud;
 import net.minecraft.world.entity.boss.enderdragon.EnderDragon;
 import net.minecraft.world.phys.Vec3;
 
 public class DragonSittingFlamingPhase extends AbstractDragonSittingPhase {
+   private static final int FLAME_DURATION = 200;
+   private static final int SITTING_FLAME_ATTACKS_COUNT = 4;
+   private static final int WARMUP_TIME = 10;
    private int flameTicks;
    private int flameCount;
    private AreaEffectCloud flame;
 
    public DragonSittingFlamingPhase(EnderDragon var1) {
@@ -86,11 +89,11 @@
       ++this.flameCount;
    }
 
    public void end() {
       if (this.flame != null) {
-         this.flame.remove();
+         this.flame.discard();
          this.flame = null;
       }
 
    }
 
