--- net/minecraft/world/entity/boss/enderdragon/phases/DragonSittingAttackingPhase.java
+++ net/minecraft/world/entity/boss/enderdragon/phases/DragonSittingAttackingPhase.java
@@ -2,10 +2,11 @@
 
 import net.minecraft.sounds.SoundEvents;
 import net.minecraft.world.entity.boss.enderdragon.EnderDragon;
 
 public class DragonSittingAttackingPhase extends AbstractDragonSittingPhase {
+   private static final int ROAR_DURATION = 40;
    private int attackingTicks;
 
    public DragonSittingAttackingPhase(EnderDragon var1) {
       super(var1);
    }
