--- net/minecraft/world/item/AirItem.java
+++ net/minecraft/world/item/AirItem.java
@@ -1,7 +1,11 @@
 package net.minecraft.world.item;
 
+import java.util.List;
+import javax.annotation.Nullable;
+import net.minecraft.network.chat.Component;
+import net.minecraft.world.level.Level;
 import net.minecraft.world.level.block.Block;
 
 public class AirItem extends Item {
    private final Block block;
 
@@ -11,6 +15,11 @@
    }
 
    public String getDescriptionId() {
       return this.block.getDescriptionId();
    }
+
+   public void appendHoverText(ItemStack var1, @Nullable Level var2, List<Component> var3, TooltipFlag var4) {
+      super.appendHoverText(var1, var2, var3, var4);
+      this.block.appendHoverText(var1, var2, var3, var4);
+   }
 }
