--- net/minecraft/world/item/ArmorMaterials.java
+++ net/minecraft/world/item/ArmorMaterials.java
@@ -69,10 +69,14 @@
 
    public Ingredient getRepairIngredient() {
       return (Ingredient)this.repairIngredient.get();
    }
 
+   public String getName() {
+      return this.name;
+   }
+
    public float getToughness() {
       return this.toughness;
    }
 
    public float getKnockbackResistance() {
