--- net/minecraft/world/item/BoatItem.java
+++ net/minecraft/world/item/BoatItem.java
@@ -55,11 +55,11 @@
             if (!var1.noCollision(var14, var14.getBoundingBox().inflate(-0.1D))) {
                return InteractionResultHolder.fail(var4);
             } else {
                if (!var1.isClientSide) {
                   var1.addFreshEntity(var14);
-                  if (!var2.abilities.instabuild) {
+                  if (!var2.getAbilities().instabuild) {
                      var4.shrink(1);
                   }
                }
 
                var2.awardStat(Stats.ITEM_USED.get(this));
