--- net/minecraft/world/item/CompassItem.java
+++ net/minecraft/world/item/CompassItem.java
@@ -20,10 +20,13 @@
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
 
 public class CompassItem extends Item implements Vanishable {
    private static final Logger LOGGER = LogManager.getLogger();
+   public static final String TAG_LODESTONE_POS = "LodestonePos";
+   public static final String TAG_LODESTONE_DIMENSION = "LodestoneDimension";
+   public static final String TAG_LODESTONE_TRACKED = "LodestoneTracked";
 
    public CompassItem(Item.Properties var1) {
       super(var1);
    }
 
@@ -47,12 +50,15 @@
             if (var6.contains("LodestoneTracked") && !var6.getBoolean("LodestoneTracked")) {
                return;
             }
 
             Optional var7 = getLodestoneDimension(var6);
-            if (var7.isPresent() && var7.get() == var2.dimension() && var6.contains("LodestonePos") && !((ServerLevel)var2).getPoiManager().existsAtPosition(PoiType.LODESTONE, NbtUtils.readBlockPos(var6.getCompound("LodestonePos")))) {
-               var6.remove("LodestonePos");
+            if (var7.isPresent() && var7.get() == var2.dimension() && var6.contains("LodestonePos")) {
+               BlockPos var8 = NbtUtils.readBlockPos(var6.getCompound("LodestonePos"));
+               if (!var2.isInWorldBounds(var8) || !((ServerLevel)var2).getPoiManager().existsAtPosition(PoiType.LODESTONE, var8)) {
+                  var6.remove("LodestonePos");
+               }
             }
          }
 
       }
    }
@@ -64,23 +70,23 @@
          return super.useOn(var1);
       } else {
          var3.playSound((Player)null, (BlockPos)var2, SoundEvents.LODESTONE_COMPASS_LOCK, SoundSource.PLAYERS, 1.0F, 1.0F);
          Player var4 = var1.getPlayer();
          ItemStack var5 = var1.getItemInHand();
-         boolean var6 = !var4.abilities.instabuild && var5.getCount() == 1;
+         boolean var6 = !var4.getAbilities().instabuild && var5.getCount() == 1;
          if (var6) {
             this.addLodestoneTags(var3.dimension(), var2, var5.getOrCreateTag());
          } else {
             ItemStack var7 = new ItemStack(Items.COMPASS, 1);
             CompoundTag var8 = var5.hasTag() ? var5.getTag().copy() : new CompoundTag();
             var7.setTag(var8);
-            if (!var4.abilities.instabuild) {
+            if (!var4.getAbilities().instabuild) {
                var5.shrink(1);
             }
 
             this.addLodestoneTags(var3.dimension(), var2, var8);
-            if (!var4.inventory.add(var7)) {
+            if (!var4.getInventory().add(var7)) {
                var4.drop(var7, false);
             }
          }
 
          return InteractionResult.sidedSuccess(var3.isClientSide);
