--- net/minecraft/world/item/DyeableLeatherItem.java
+++ net/minecraft/world/item/DyeableLeatherItem.java
@@ -3,10 +3,14 @@
 import java.util.Iterator;
 import java.util.List;
 import net.minecraft.nbt.CompoundTag;
 
 public interface DyeableLeatherItem {
+   String TAG_COLOR = "color";
+   String TAG_DISPLAY = "display";
+   int DEFAULT_LEATHER_COLOR = 10511680;
+
    default boolean hasCustomColor(ItemStack var1) {
       CompoundTag var2 = var1.getTagElement("display");
       return var2 != null && var2.contains("color", 99);
    }
 
