--- net/minecraft/world/item/DyeColor.java
+++ net/minecraft/world/item/DyeColor.java
@@ -3,10 +3,11 @@
 import it.unimi.dsi.fastutil.ints.Int2ObjectOpenHashMap;
 import java.util.Arrays;
 import java.util.Comparator;
 import java.util.Map;
 import java.util.stream.Collectors;
+import javax.annotation.Nullable;
 import net.minecraft.util.StringRepresentable;
 import net.minecraft.world.level.material.MaterialColor;
 
 public enum DyeColor implements StringRepresentable {
    WHITE(0, "white", 16383998, MaterialColor.SNOW, 15790320, 16777215),
@@ -35,26 +36,22 @@
       return var0;
    })));
    private final int id;
    private final String name;
    private final MaterialColor color;
-   private final int textureDiffuseColor;
-   private final int textureDiffuseColorBGR;
    private final float[] textureDiffuseColors;
    private final int fireworkColor;
    private final int textColor;
 
    private DyeColor(int var3, String var4, int var5, MaterialColor var6, int var7, int var8) {
       this.id = var3;
       this.name = var4;
-      this.textureDiffuseColor = var5;
       this.color = var6;
       this.textColor = var8;
       int var9 = (var5 & 16711680) >> 16;
       int var10 = (var5 & '\uff00') >> 8;
       int var11 = (var5 & 255) >> 0;
-      this.textureDiffuseColorBGR = var11 << 16 | var10 << 8 | var9 << 0;
       this.textureDiffuseColors = new float[]{(float)var9 / 255.0F, (float)var10 / 255.0F, (float)var11 / 255.0F};
       this.fireworkColor = var7;
    }
 
    public int getId() {
@@ -75,10 +72,14 @@
 
    public int getFireworkColor() {
       return this.fireworkColor;
    }
 
+   public int getTextColor() {
+      return this.textColor;
+   }
+
    public static DyeColor byId(int var0) {
       if (var0 < 0 || var0 >= BY_ID.length) {
          var0 = 0;
       }
 
@@ -97,10 +98,15 @@
       }
 
       return var1;
    }
 
+   @Nullable
+   public static DyeColor byFireworkColor(int var0) {
+      return (DyeColor)BY_FIREWORK_COLOR.get(var0);
+   }
+
    public String toString() {
       return this.name;
    }
 
    public String getSerializedName() {
