--- net/minecraft/world/item/DyeItem.java
+++ net/minecraft/world/item/DyeItem.java
@@ -1,11 +1,14 @@
 package net.minecraft.world.item;
 
 import com.google.common.collect.Maps;
 import java.util.Map;
+import net.minecraft.sounds.SoundEvents;
+import net.minecraft.sounds.SoundSource;
 import net.minecraft.world.InteractionHand;
 import net.minecraft.world.InteractionResult;
+import net.minecraft.world.entity.Entity;
 import net.minecraft.world.entity.LivingEntity;
 import net.minecraft.world.entity.animal.Sheep;
 import net.minecraft.world.entity.player.Player;
 
 public class DyeItem extends Item {
@@ -20,10 +23,11 @@
 
    public InteractionResult interactLivingEntity(ItemStack var1, Player var2, LivingEntity var3, InteractionHand var4) {
       if (var3 instanceof Sheep) {
          Sheep var5 = (Sheep)var3;
          if (var5.isAlive() && !var5.isSheared() && var5.getColor() != this.dyeColor) {
+            var5.level.playSound(var2, (Entity)var5, SoundEvents.DYE_USE, SoundSource.PLAYERS, 1.0F, 1.0F);
             if (!var2.level.isClientSide) {
                var5.setColor(this.dyeColor);
                var1.shrink(1);
             }
 
