--- net/minecraft/world/item/EnchantedBookItem.java
+++ net/minecraft/world/item/EnchantedBookItem.java
@@ -1,17 +1,23 @@
 package net.minecraft.world.item;
 
 import java.util.Iterator;
+import java.util.List;
+import javax.annotation.Nullable;
 import net.minecraft.core.NonNullList;
 import net.minecraft.core.Registry;
 import net.minecraft.nbt.CompoundTag;
 import net.minecraft.nbt.ListTag;
+import net.minecraft.network.chat.Component;
 import net.minecraft.resources.ResourceLocation;
 import net.minecraft.world.item.enchantment.Enchantment;
 import net.minecraft.world.item.enchantment.EnchantmentInstance;
+import net.minecraft.world.level.Level;
 
 public class EnchantedBookItem extends Item {
+   public static final String TAG_STORED_ENCHANTMENTS = "StoredEnchantments";
+
    public EnchantedBookItem(Item.Properties var1) {
       super(var1);
    }
 
    public boolean isFoil(ItemStack var1) {
@@ -25,10 +31,15 @@
    public static ListTag getEnchantments(ItemStack var0) {
       CompoundTag var1 = var0.getTag();
       return var1 != null ? var1.getList("StoredEnchantments", 10) : new ListTag();
    }
 
+   public void appendHoverText(ItemStack var1, @Nullable Level var2, List<Component> var3, TooltipFlag var4) {
+      super.appendHoverText(var1, var2, var3, var4);
+      ItemStack.appendEnchantmentNames(var3, getEnchantments(var1));
+   }
+
    public static void addEnchantment(ItemStack var0, EnchantmentInstance var1) {
       ListTag var2 = getEnchantments(var0);
       boolean var3 = true;
       ResourceLocation var4 = Registry.ENCHANTMENT.getKey(var1.enchantment);
 
