--- net/minecraft/world/item/EndCrystalItem.java
+++ net/minecraft/world/item/EndCrystalItem.java
@@ -9,10 +9,11 @@
 import net.minecraft.world.item.context.UseOnContext;
 import net.minecraft.world.level.Level;
 import net.minecraft.world.level.block.Blocks;
 import net.minecraft.world.level.block.state.BlockState;
 import net.minecraft.world.level.dimension.end.EndDragonFight;
+import net.minecraft.world.level.gameevent.GameEvent;
 import net.minecraft.world.phys.AABB;
 
 public class EndCrystalItem extends Item {
    public EndCrystalItem(Item.Properties var1) {
       super(var1);
@@ -38,10 +39,11 @@
             } else {
                if (var2 instanceof ServerLevel) {
                   EndCrystal var13 = new EndCrystal(var2, var6 + 0.5D, var8, var10 + 0.5D);
                   var13.setShowBottom(false);
                   var2.addFreshEntity(var13);
+                  var2.gameEvent(var1.getPlayer(), GameEvent.BLOCK_PLACE, var5);
                   EndDragonFight var14 = ((ServerLevel)var2).dragonFight();
                   if (var14 != null) {
                      var14.tryRespawn();
                   }
                }
