--- net/minecraft/world/item/FireworkRocketItem.java
+++ net/minecraft/world/item/FireworkRocketItem.java
@@ -34,21 +34,27 @@
    public InteractionResultHolder<ItemStack> use(Level var1, Player var2, InteractionHand var3) {
       if (var2.isFallFlying()) {
          ItemStack var4 = var2.getItemInHand(var3);
          if (!var1.isClientSide) {
             var1.addFreshEntity(new FireworkRocketEntity(var1, var4, var2));
-            if (!var2.abilities.instabuild) {
+            if (!var2.getAbilities().instabuild) {
                var4.shrink(1);
             }
          }
 
          return InteractionResultHolder.sidedSuccess(var2.getItemInHand(var3), var1.isClientSide());
       } else {
          return InteractionResultHolder.pass(var2.getItemInHand(var3));
       }
    }
 
+   public ItemStack getDefaultInstance() {
+      ItemStack var1 = new ItemStack(this);
+      var1.getOrCreateTag().putByte("Flight", (byte)1);
+      return var1;
+   }
+
    public static enum Shape {
       SMALL_BALL(0, "small_ball"),
       LARGE_BALL(1, "large_ball"),
       STAR(2, "star"),
       CREEPER(3, "creeper"),
