--- net/minecraft/world/item/ItemCooldowns.java
+++ net/minecraft/world/item/ItemCooldowns.java
@@ -47,10 +47,15 @@
    public void addCooldown(Item var1, int var2) {
       this.cooldowns.put(var1, new ItemCooldowns.CooldownInstance(this.tickCount, this.tickCount + var2));
       this.onCooldownStarted(var1, var2);
    }
 
+   public void removeCooldown(Item var1) {
+      this.cooldowns.remove(var1);
+      this.onCooldownEnded(var1);
+   }
+
    protected void onCooldownStarted(Item var1, int var2) {
    }
 
    protected void onCooldownEnded(Item var1) {
    }
