--- net/minecraft/world/item/ItemCooldowns.java
+++ net/minecraft/world/item/ItemCooldowns.java
@@ -47,21 +47,26 @@
    public void addCooldown(Item var1, int var2) {
       this.cooldowns.put(var1, new ItemCooldowns.CooldownInstance(this.tickCount, this.tickCount + var2));
       this.onCooldownStarted(var1, var2);
    }
 
+   public void removeCooldown(Item var1) {
+      this.cooldowns.remove(var1);
+      this.onCooldownEnded(var1);
+   }
+
    protected void onCooldownStarted(Item var1, int var2) {
    }
 
    protected void onCooldownEnded(Item var1) {
    }
 
-   class CooldownInstance {
-      private final int startTime;
-      private final int endTime;
+   private class CooldownInstance {
+      final int startTime;
+      final int endTime;
 
-      private CooldownInstance(int var2, int var3) {
+      CooldownInstance(int var2, int var3) {
          this.startTime = var2;
          this.endTime = var3;
       }
    }
 }
