--- net/minecraft/world/item/PlayerHeadItem.java
+++ net/minecraft/world/item/PlayerHeadItem.java
@@ -9,16 +9,18 @@
 import net.minecraft.world.level.block.Block;
 import net.minecraft.world.level.block.entity.SkullBlockEntity;
 import org.apache.commons.lang3.StringUtils;
 
 public class PlayerHeadItem extends StandingAndWallBlockItem {
+   public static final String TAG_SKULL_OWNER = "SkullOwner";
+
    public PlayerHeadItem(Block var1, Block var2, Item.Properties var3) {
       super(var1, var2, var3);
    }
 
    public Component getName(ItemStack var1) {
-      if (var1.getItem() == Items.PLAYER_HEAD && var1.hasTag()) {
+      if (var1.is(Items.PLAYER_HEAD) && var1.hasTag()) {
          String var2 = null;
          CompoundTag var3 = var1.getTag();
          if (var3.contains("SkullOwner", 8)) {
             var2 = var3.getString("SkullOwner");
          } else if (var3.contains("SkullOwner", 10)) {
@@ -34,17 +36,17 @@
       }
 
       return super.getName(var1);
    }
 
-   public boolean verifyTagAfterLoad(CompoundTag var1) {
+   public void verifyTagAfterLoad(CompoundTag var1) {
       super.verifyTagAfterLoad(var1);
       if (var1.contains("SkullOwner", 8) && !StringUtils.isBlank(var1.getString("SkullOwner"))) {
          GameProfile var2 = new GameProfile((UUID)null, var1.getString("SkullOwner"));
-         var2 = SkullBlockEntity.updateGameprofile(var2);
-         var1.put("SkullOwner", NbtUtils.writeGameProfile(new CompoundTag(), var2));
-         return true;
-      } else {
-         return false;
+         var1.remove("SkullOwner");
+         SkullBlockEntity.updateGameprofile(var2, (var1x) -> {
+            var1.put("SkullOwner", NbtUtils.writeGameProfile(new CompoundTag(), var1x));
+         });
       }
+
    }
 }
