--- net/minecraft/world/item/PotionItem.java
+++ net/minecraft/world/item/PotionItem.java
@@ -1,12 +1,14 @@
 package net.minecraft.world.item;
 
 import java.util.Iterator;
 import java.util.List;
+import javax.annotation.Nullable;
 import net.minecraft.advancements.CriteriaTriggers;
 import net.minecraft.core.NonNullList;
 import net.minecraft.core.Registry;
+import net.minecraft.network.chat.Component;
 import net.minecraft.server.level.ServerPlayer;
 import net.minecraft.stats.Stats;
 import net.minecraft.world.InteractionHand;
 import net.minecraft.world.InteractionResultHolder;
 import net.minecraft.world.effect.MobEffectInstance;
@@ -14,12 +16,15 @@
 import net.minecraft.world.entity.player.Player;
 import net.minecraft.world.item.alchemy.Potion;
 import net.minecraft.world.item.alchemy.PotionUtils;
 import net.minecraft.world.item.alchemy.Potions;
 import net.minecraft.world.level.Level;
+import net.minecraft.world.level.gameevent.GameEvent;
 
 public class PotionItem extends Item {
+   private static final int DRINK_DURATION = 32;
+
    public PotionItem(Item.Properties var1) {
       super(var1);
    }
 
    public ItemStack getDefaultInstance() {
@@ -46,25 +51,26 @@
          }
       }
 
       if (var4 != null) {
          var4.awardStat(Stats.ITEM_USED.get(this));
-         if (!var4.abilities.instabuild) {
+         if (!var4.getAbilities().instabuild) {
             var1.shrink(1);
          }
       }
 
-      if (var4 == null || !var4.abilities.instabuild) {
+      if (var4 == null || !var4.getAbilities().instabuild) {
          if (var1.isEmpty()) {
             return new ItemStack(Items.GLASS_BOTTLE);
          }
 
          if (var4 != null) {
-            var4.inventory.add(new ItemStack(Items.GLASS_BOTTLE));
+            var4.getInventory().add(new ItemStack(Items.GLASS_BOTTLE));
          }
       }
 
+      var2.gameEvent(var3, GameEvent.DRINKING_FINISH, var3.eyeBlockPosition());
       return var1;
    }
 
    public int getUseDuration(ItemStack var1) {
       return 32;
@@ -73,17 +79,21 @@
    public UseAnim getUseAnimation(ItemStack var1) {
       return UseAnim.DRINK;
    }
 
    public InteractionResultHolder<ItemStack> use(Level var1, Player var2, InteractionHand var3) {
-      return ItemUtils.useDrink(var1, var2, var3);
+      return ItemUtils.startUsingInstantly(var1, var2, var3);
    }
 
    public String getDescriptionId(ItemStack var1) {
       return PotionUtils.getPotion(var1).getName(this.getDescriptionId() + ".effect.");
    }
 
+   public void appendHoverText(ItemStack var1, @Nullable Level var2, List<Component> var3, TooltipFlag var4) {
+      PotionUtils.addPotionTooltip(var1, var3, 1.0F);
+   }
+
    public boolean isFoil(ItemStack var1) {
       return super.isFoil(var1) || !PotionUtils.getMobEffects(var1).isEmpty();
    }
 
    public void fillItemCategory(CreativeModeTab var1, NonNullList<ItemStack> var2) {
