--- net/minecraft/world/item/ScaffoldingBlockItem.java
+++ net/minecraft/world/item/ScaffoldingBlockItem.java
@@ -5,11 +5,10 @@
 import net.minecraft.Util;
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.Direction;
 import net.minecraft.network.chat.ChatType;
 import net.minecraft.network.chat.TranslatableComponent;
-import net.minecraft.network.protocol.game.ClientboundChatPacket;
 import net.minecraft.server.level.ServerPlayer;
 import net.minecraft.world.entity.player.Player;
 import net.minecraft.world.item.context.BlockPlaceContext;
 import net.minecraft.world.level.Level;
 import net.minecraft.world.level.block.Block;
@@ -39,16 +38,15 @@
 
          int var7 = 0;
          BlockPos.MutableBlockPos var8 = var2.mutable().move(var6);
 
          while(var7 < 7) {
-            if (!var3.isClientSide && !Level.isInWorldBounds(var8)) {
+            if (!var3.isClientSide && !var3.isInWorldBounds(var8)) {
                Player var9 = var1.getPlayer();
                int var10 = var3.getMaxBuildHeight();
                if (var9 instanceof ServerPlayer && var8.getY() >= var10) {
-                  ClientboundChatPacket var11 = new ClientboundChatPacket((new TranslatableComponent("build.tooHigh", new Object[]{var10})).withStyle(ChatFormatting.RED), ChatType.GAME_INFO, Util.NIL_UUID);
-                  ((ServerPlayer)var9).connection.send(var11);
+                  ((ServerPlayer)var9).sendMessage((new TranslatableComponent("build.tooHigh", new Object[]{var10})).withStyle(ChatFormatting.RED), ChatType.GAME_INFO, Util.NIL_UUID);
                }
                break;
             }
 
             var4 = var3.getBlockState(var8);
