--- net/minecraft/world/item/SwordItem.java
+++ net/minecraft/world/item/SwordItem.java
@@ -39,11 +39,11 @@
    public float getDestroySpeed(ItemStack var1, BlockState var2) {
       if (var2.is(Blocks.COBWEB)) {
          return 15.0F;
       } else {
          Material var3 = var2.getMaterial();
-         return var3 != Material.PLANT && var3 != Material.REPLACEABLE_PLANT && var3 != Material.CORAL && !var2.is(BlockTags.LEAVES) && var3 != Material.VEGETABLE ? 1.0F : 1.5F;
+         return var3 != Material.PLANT && var3 != Material.REPLACEABLE_PLANT && !var2.is(BlockTags.LEAVES) && var3 != Material.VEGETABLE ? 1.0F : 1.5F;
       }
    }
 
    public boolean hurtEnemy(ItemStack var1, LivingEntity var2, LivingEntity var3) {
       var1.hurtAndBreak(1, var3, (var0) -> {
