--- net/minecraft/world/item/TippedArrowItem.java
+++ net/minecraft/world/item/TippedArrowItem.java
@@ -1,13 +1,17 @@
 package net.minecraft.world.item;
 
 import java.util.Iterator;
+import java.util.List;
+import javax.annotation.Nullable;
 import net.minecraft.core.NonNullList;
 import net.minecraft.core.Registry;
+import net.minecraft.network.chat.Component;
 import net.minecraft.world.item.alchemy.Potion;
 import net.minecraft.world.item.alchemy.PotionUtils;
 import net.minecraft.world.item.alchemy.Potions;
+import net.minecraft.world.level.Level;
 
 public class TippedArrowItem extends ArrowItem {
    public TippedArrowItem(Item.Properties var1) {
       super(var1);
    }
@@ -28,9 +32,13 @@
          }
       }
 
    }
 
+   public void appendHoverText(ItemStack var1, @Nullable Level var2, List<Component> var3, TooltipFlag var4) {
+      PotionUtils.addPotionTooltip(var1, var3, 0.125F);
+   }
+
    public String getDescriptionId(ItemStack var1) {
       return PotionUtils.getPotion(var1).getName(this.getDescriptionId() + ".effect.");
    }
 }
