--- net/minecraft/network/protocol/game/ClientboundSelectAdvancementsTabPacket.java
+++ net/minecraft/network/protocol/game/ClientboundSelectAdvancementsTabPacket.java
@@ -1,36 +1,34 @@
 package net.minecraft.network.protocol.game;
 
-import java.io.IOException;
 import javax.annotation.Nullable;
 import net.minecraft.network.FriendlyByteBuf;
 import net.minecraft.network.protocol.Packet;
 import net.minecraft.resources.ResourceLocation;
 
 public class ClientboundSelectAdvancementsTabPacket implements Packet<ClientGamePacketListener> {
    @Nullable
-   private ResourceLocation tab;
+   private final ResourceLocation tab;
 
-   public ClientboundSelectAdvancementsTabPacket() {
-   }
-
    public ClientboundSelectAdvancementsTabPacket(@Nullable ResourceLocation var1) {
       this.tab = var1;
    }
 
    public void handle(ClientGamePacketListener var1) {
       var1.handleSelectAdvancementsTab(this);
    }
 
-   public void read(FriendlyByteBuf var1) throws IOException {
+   public ClientboundSelectAdvancementsTabPacket(FriendlyByteBuf var1) {
       if (var1.readBoolean()) {
          this.tab = var1.readResourceLocation();
+      } else {
+         this.tab = null;
       }
 
    }
 
-   public void write(FriendlyByteBuf var1) throws IOException {
+   public void write(FriendlyByteBuf var1) {
       var1.writeBoolean(this.tab != null);
       if (this.tab != null) {
          var1.writeResourceLocation(this.tab);
       }
 
