--- net/minecraft/network/protocol/game/ClientboundCustomPayloadPacket.java
+++ net/minecraft/network/protocol/game/ClientboundCustomPayloadPacket.java
@@ -21,10 +21,12 @@
    public static final ResourceLocation DEBUG_BEE = new ResourceLocation("debug/bee");
    public static final ResourceLocation DEBUG_HIVE = new ResourceLocation("debug/hive");
    public static final ResourceLocation DEBUG_GAME_TEST_ADD_MARKER = new ResourceLocation("debug/game_test_add_marker");
    public static final ResourceLocation DEBUG_GAME_TEST_CLEAR = new ResourceLocation("debug/game_test_clear");
    public static final ResourceLocation DEBUG_RAIDS = new ResourceLocation("debug/raids");
+   public static final ResourceLocation DEBUG_GAME_EVENT = new ResourceLocation("debug/game_event");
+   public static final ResourceLocation DEBUG_GAME_EVENT_LISTENER = new ResourceLocation("debug/game_event_listeners");
    private ResourceLocation identifier;
    private FriendlyByteBuf data;
 
    public ClientboundCustomPayloadPacket() {
    }
