--- net/minecraft/network/protocol/game/ClientboundOpenScreenPacket.java
+++ net/minecraft/network/protocol/game/ClientboundOpenScreenPacket.java
@@ -1,35 +1,31 @@
 package net.minecraft.network.protocol.game;
 
-import java.io.IOException;
 import net.minecraft.core.Registry;
 import net.minecraft.network.FriendlyByteBuf;
 import net.minecraft.network.chat.Component;
 import net.minecraft.network.protocol.Packet;
 import net.minecraft.world.inventory.MenuType;
 
 public class ClientboundOpenScreenPacket implements Packet<ClientGamePacketListener> {
-   private int containerId;
-   private int type;
-   private Component title;
+   private final int containerId;
+   private final int type;
+   private final Component title;
 
-   public ClientboundOpenScreenPacket() {
-   }
-
    public ClientboundOpenScreenPacket(int var1, MenuType<?> var2, Component var3) {
       this.containerId = var1;
       this.type = Registry.MENU.getId(var2);
       this.title = var3;
    }
 
-   public void read(FriendlyByteBuf var1) throws IOException {
+   public ClientboundOpenScreenPacket(FriendlyByteBuf var1) {
       this.containerId = var1.readVarInt();
       this.type = var1.readVarInt();
       this.title = var1.readComponent();
    }
 
-   public void write(FriendlyByteBuf var1) throws IOException {
+   public void write(FriendlyByteBuf var1) {
       var1.writeVarInt(this.containerId);
       var1.writeVarInt(this.type);
       var1.writeComponent(this.title);
    }
 
