--- net/minecraft/network/protocol/game/ClientboundBossEventPacket.java
+++ net/minecraft/network/protocol/game/ClientboundBossEventPacket.java
@@ -23,11 +23,11 @@
 
    public ClientboundBossEventPacket(ClientboundBossEventPacket.Operation var1, BossEvent var2) {
       this.operation = var1;
       this.id = var2.getId();
       this.name = var2.getName();
-      this.pct = var2.getPercent();
+      this.pct = var2.getProgress();
       this.color = var2.getColor();
       this.overlay = var2.getOverlay();
       this.darkenScreen = var2.shouldDarkenScreen();
       this.playMusic = var2.shouldPlayBossMusic();
       this.createWorldFog = var2.shouldCreateWorldFog();
@@ -44,11 +44,11 @@
          this.overlay = (BossEvent.BossBarOverlay)var1.readEnum(BossEvent.BossBarOverlay.class);
          this.decodeProperties(var1.readUnsignedByte());
       case REMOVE:
       default:
          break;
-      case UPDATE_PCT:
+      case UPDATE_PROGRESS:
          this.pct = var1.readFloat();
          break;
       case UPDATE_NAME:
          this.name = var1.readComponent();
          break;
@@ -79,11 +79,11 @@
          var1.writeEnum(this.overlay);
          var1.writeByte(this.encodeProperties());
       case REMOVE:
       default:
          break;
-      case UPDATE_PCT:
+      case UPDATE_PROGRESS:
          var1.writeFloat(this.pct);
          break;
       case UPDATE_NAME:
          var1.writeComponent(this.name);
          break;
@@ -119,11 +119,11 @@
    }
 
    public static enum Operation {
       ADD,
       REMOVE,
-      UPDATE_PCT,
+      UPDATE_PROGRESS,
       UPDATE_NAME,
       UPDATE_STYLE,
       UPDATE_PROPERTIES;
 
       private Operation() {
