--- net/minecraft/network/protocol/game/ClientboundSetChunkCacheCenterPacket.java
+++ net/minecraft/network/protocol/game/ClientboundSetChunkCacheCenterPacket.java
@@ -1,32 +1,36 @@
 package net.minecraft.network.protocol.game;
 
-import java.io.IOException;
 import net.minecraft.network.FriendlyByteBuf;
 import net.minecraft.network.protocol.Packet;
 
 public class ClientboundSetChunkCacheCenterPacket implements Packet<ClientGamePacketListener> {
-   private int x;
-   private int z;
+   private final int x;
+   private final int z;
 
-   public ClientboundSetChunkCacheCenterPacket() {
-   }
-
    public ClientboundSetChunkCacheCenterPacket(int var1, int var2) {
       this.x = var1;
       this.z = var2;
    }
 
-   public void read(FriendlyByteBuf var1) throws IOException {
+   public ClientboundSetChunkCacheCenterPacket(FriendlyByteBuf var1) {
       this.x = var1.readVarInt();
       this.z = var1.readVarInt();
    }
 
-   public void write(FriendlyByteBuf var1) throws IOException {
+   public void write(FriendlyByteBuf var1) {
       var1.writeVarInt(this.x);
       var1.writeVarInt(this.z);
    }
 
    public void handle(ClientGamePacketListener var1) {
       var1.handleSetChunkCacheCenter(this);
    }
+
+   public int getX() {
+      return this.x;
+   }
+
+   public int getZ() {
+      return this.z;
+   }
 }
