--- net/minecraft/network/protocol/game/ClientboundSetSubtitleTextPacket.java
+++ net/minecraft/network/protocol/game/ClientboundSetSubtitleTextPacket.java
@@ -1,0 +1,25 @@
+package net.minecraft.network.protocol.game;
+
+import net.minecraft.network.FriendlyByteBuf;
+import net.minecraft.network.chat.Component;
+import net.minecraft.network.protocol.Packet;
+
+public class ClientboundSetSubtitleTextPacket implements Packet<ClientGamePacketListener> {
+   private final Component text;
+
+   public ClientboundSetSubtitleTextPacket(Component var1) {
+      this.text = var1;
+   }
+
+   public ClientboundSetSubtitleTextPacket(FriendlyByteBuf var1) {
+      this.text = var1.readComponent();
+   }
+
+   public void write(FriendlyByteBuf var1) {
+      var1.writeComponent(this.text);
+   }
+
+   public void handle(ClientGamePacketListener var1) {
+      var1.setSubtitleText(this);
+   }
+}
