--- net/minecraft/network/protocol/game/ClientboundAddMobPacket.java
+++ net/minecraft/network/protocol/game/ClientboundAddMobPacket.java
@@ -1,33 +1,29 @@
 package net.minecraft.network.protocol.game;
 
-import java.io.IOException;
 import java.util.UUID;
 import net.minecraft.core.Registry;
 import net.minecraft.network.FriendlyByteBuf;
 import net.minecraft.network.protocol.Packet;
 import net.minecraft.util.Mth;
 import net.minecraft.world.entity.LivingEntity;
 import net.minecraft.world.phys.Vec3;
 
 public class ClientboundAddMobPacket implements Packet<ClientGamePacketListener> {
-   private int id;
-   private UUID uuid;
-   private int type;
-   private double x;
-   private double y;
-   private double z;
-   private int xd;
-   private int yd;
-   private int zd;
-   private byte yRot;
-   private byte xRot;
-   private byte yHeadRot;
+   private final int id;
+   private final UUID uuid;
+   private final int type;
+   private final double x;
+   private final double y;
+   private final double z;
+   private final int xd;
+   private final int yd;
+   private final int zd;
+   private final byte yRot;
+   private final byte xRot;
+   private final byte yHeadRot;
 
-   public ClientboundAddMobPacket() {
-   }
-
    public ClientboundAddMobPacket(LivingEntity var1) {
       this.id = var1.getId();
       this.uuid = var1.getUUID();
       this.type = Registry.ENTITY_TYPE.getId(var1.getType());
       this.x = var1.getX();
@@ -44,11 +40,11 @@
       this.xd = (int)(var5 * 8000.0D);
       this.yd = (int)(var7 * 8000.0D);
       this.zd = (int)(var9 * 8000.0D);
    }
 
-   public void read(FriendlyByteBuf var1) throws IOException {
+   public ClientboundAddMobPacket(FriendlyByteBuf var1) {
       this.id = var1.readVarInt();
       this.uuid = var1.readUUID();
       this.type = var1.readVarInt();
       this.x = var1.readDouble();
       this.y = var1.readDouble();
@@ -59,11 +55,11 @@
       this.xd = var1.readShort();
       this.yd = var1.readShort();
       this.zd = var1.readShort();
    }
 
-   public void write(FriendlyByteBuf var1) throws IOException {
+   public void write(FriendlyByteBuf var1) {
       var1.writeVarInt(this.id);
       var1.writeUUID(this.uuid);
       var1.writeVarInt(this.type);
       var1.writeDouble(this.x);
       var1.writeDouble(this.y);
@@ -77,6 +73,54 @@
    }
 
    public void handle(ClientGamePacketListener var1) {
       var1.handleAddMob(this);
    }
+
+   public int getId() {
+      return this.id;
+   }
+
+   public UUID getUUID() {
+      return this.uuid;
+   }
+
+   public int getType() {
+      return this.type;
+   }
+
+   public double getX() {
+      return this.x;
+   }
+
+   public double getY() {
+      return this.y;
+   }
+
+   public double getZ() {
+      return this.z;
+   }
+
+   public int getXd() {
+      return this.xd;
+   }
+
+   public int getYd() {
+      return this.yd;
+   }
+
+   public int getZd() {
+      return this.zd;
+   }
+
+   public byte getyRot() {
+      return this.yRot;
+   }
+
+   public byte getxRot() {
+      return this.xRot;
+   }
+
+   public byte getyHeadRot() {
+      return this.yHeadRot;
+   }
 }
