--- net/minecraft/network/protocol/game/ClientboundMerchantOffersPacket.java
+++ net/minecraft/network/protocol/game/ClientboundMerchantOffersPacket.java
@@ -1,42 +1,38 @@
 package net.minecraft.network.protocol.game;
 
-import java.io.IOException;
 import net.minecraft.network.FriendlyByteBuf;
 import net.minecraft.network.protocol.Packet;
 import net.minecraft.world.item.trading.MerchantOffers;
 
 public class ClientboundMerchantOffersPacket implements Packet<ClientGamePacketListener> {
-   private int containerId;
-   private MerchantOffers offers;
-   private int villagerLevel;
-   private int villagerXp;
-   private boolean showProgress;
-   private boolean canRestock;
+   private final int containerId;
+   private final MerchantOffers offers;
+   private final int villagerLevel;
+   private final int villagerXp;
+   private final boolean showProgress;
+   private final boolean canRestock;
 
-   public ClientboundMerchantOffersPacket() {
-   }
-
    public ClientboundMerchantOffersPacket(int var1, MerchantOffers var2, int var3, int var4, boolean var5, boolean var6) {
       this.containerId = var1;
       this.offers = var2;
       this.villagerLevel = var3;
       this.villagerXp = var4;
       this.showProgress = var5;
       this.canRestock = var6;
    }
 
-   public void read(FriendlyByteBuf var1) throws IOException {
+   public ClientboundMerchantOffersPacket(FriendlyByteBuf var1) {
       this.containerId = var1.readVarInt();
       this.offers = MerchantOffers.createFromStream(var1);
       this.villagerLevel = var1.readVarInt();
       this.villagerXp = var1.readVarInt();
       this.showProgress = var1.readBoolean();
       this.canRestock = var1.readBoolean();
    }
 
-   public void write(FriendlyByteBuf var1) throws IOException {
+   public void write(FriendlyByteBuf var1) {
       var1.writeVarInt(this.containerId);
       this.offers.writeToStream(var1);
       var1.writeVarInt(this.villagerLevel);
       var1.writeVarInt(this.villagerXp);
       var1.writeBoolean(this.showProgress);
