--- net/minecraft/network/protocol/game/ClientboundRotateHeadPacket.java
+++ net/minecraft/network/protocol/game/ClientboundRotateHeadPacket.java
@@ -1,30 +1,26 @@
 package net.minecraft.network.protocol.game;
 
-import java.io.IOException;
 import net.minecraft.network.FriendlyByteBuf;
 import net.minecraft.network.protocol.Packet;
 import net.minecraft.world.entity.Entity;
 
 public class ClientboundRotateHeadPacket implements Packet<ClientGamePacketListener> {
-   private int entityId;
-   private byte yHeadRot;
+   private final int entityId;
+   private final byte yHeadRot;
 
-   public ClientboundRotateHeadPacket() {
-   }
-
    public ClientboundRotateHeadPacket(Entity var1, byte var2) {
       this.entityId = var1.getId();
       this.yHeadRot = var2;
    }
 
-   public void read(FriendlyByteBuf var1) throws IOException {
+   public ClientboundRotateHeadPacket(FriendlyByteBuf var1) {
       this.entityId = var1.readVarInt();
       this.yHeadRot = var1.readByte();
    }
 
-   public void write(FriendlyByteBuf var1) throws IOException {
+   public void write(FriendlyByteBuf var1) {
       var1.writeVarInt(this.entityId);
       var1.writeByte(this.yHeadRot);
    }
 
    public void handle(ClientGamePacketListener var1) {
