--- net/minecraft/network/protocol/game/ClientboundOpenBookPacket.java
+++ net/minecraft/network/protocol/game/ClientboundOpenBookPacket.java
@@ -1,27 +1,23 @@
 package net.minecraft.network.protocol.game;
 
-import java.io.IOException;
 import net.minecraft.network.FriendlyByteBuf;
 import net.minecraft.network.protocol.Packet;
 import net.minecraft.world.InteractionHand;
 
 public class ClientboundOpenBookPacket implements Packet<ClientGamePacketListener> {
-   private InteractionHand hand;
+   private final InteractionHand hand;
 
-   public ClientboundOpenBookPacket() {
-   }
-
    public ClientboundOpenBookPacket(InteractionHand var1) {
       this.hand = var1;
    }
 
-   public void read(FriendlyByteBuf var1) throws IOException {
+   public ClientboundOpenBookPacket(FriendlyByteBuf var1) {
       this.hand = (InteractionHand)var1.readEnum(InteractionHand.class);
    }
 
-   public void write(FriendlyByteBuf var1) throws IOException {
+   public void write(FriendlyByteBuf var1) {
       var1.writeEnum(this.hand);
    }
 
    public void handle(ClientGamePacketListener var1) {
       var1.handleOpenBook(this);
