--- net/minecraft/network/protocol/game/ClientboundSetEntityDataPacket.java
+++ net/minecraft/network/protocol/game/ClientboundSetEntityDataPacket.java
@@ -1,20 +1,18 @@
 package net.minecraft.network.protocol.game;
 
-import java.io.IOException;
 import java.util.List;
+import javax.annotation.Nullable;
 import net.minecraft.network.FriendlyByteBuf;
 import net.minecraft.network.protocol.Packet;
 import net.minecraft.network.syncher.SynchedEntityData;
 
 public class ClientboundSetEntityDataPacket implements Packet<ClientGamePacketListener> {
-   private int id;
-   private List<SynchedEntityData.DataItem<?>> packedItems;
+   private final int id;
+   @Nullable
+   private final List<SynchedEntityData.DataItem<?>> packedItems;
 
-   public ClientboundSetEntityDataPacket() {
-   }
-
    public ClientboundSetEntityDataPacket(int var1, SynchedEntityData var2, boolean var3) {
       this.id = var1;
       if (var3) {
          this.packedItems = var2.getAll();
          var2.clearDirty();
@@ -22,16 +20,16 @@
          this.packedItems = var2.packDirty();
       }
 
    }
 
-   public void read(FriendlyByteBuf var1) throws IOException {
+   public ClientboundSetEntityDataPacket(FriendlyByteBuf var1) {
       this.id = var1.readVarInt();
       this.packedItems = SynchedEntityData.unpack(var1);
    }
 
-   public void write(FriendlyByteBuf var1) throws IOException {
+   public void write(FriendlyByteBuf var1) {
       var1.writeVarInt(this.id);
       SynchedEntityData.pack(this.packedItems, var1);
    }
 
    public void handle(ClientGamePacketListener var1) {
