--- net/minecraft/network/protocol/game/ClientboundStopSoundPacket.java
+++ net/minecraft/network/protocol/game/ClientboundStopSoundPacket.java
@@ -1,39 +1,43 @@
 package net.minecraft.network.protocol.game;
 
-import java.io.IOException;
 import javax.annotation.Nullable;
 import net.minecraft.network.FriendlyByteBuf;
 import net.minecraft.network.protocol.Packet;
 import net.minecraft.resources.ResourceLocation;
 import net.minecraft.sounds.SoundSource;
 
 public class ClientboundStopSoundPacket implements Packet<ClientGamePacketListener> {
-   private ResourceLocation name;
-   private SoundSource source;
+   private static final int HAS_SOURCE = 1;
+   private static final int HAS_SOUND = 2;
+   @Nullable
+   private final ResourceLocation name;
+   @Nullable
+   private final SoundSource source;
 
-   public ClientboundStopSoundPacket() {
-   }
-
    public ClientboundStopSoundPacket(@Nullable ResourceLocation var1, @Nullable SoundSource var2) {
       this.name = var1;
       this.source = var2;
    }
 
-   public void read(FriendlyByteBuf var1) throws IOException {
+   public ClientboundStopSoundPacket(FriendlyByteBuf var1) {
       byte var2 = var1.readByte();
       if ((var2 & 1) > 0) {
          this.source = (SoundSource)var1.readEnum(SoundSource.class);
+      } else {
+         this.source = null;
       }
 
       if ((var2 & 2) > 0) {
          this.name = var1.readResourceLocation();
+      } else {
+         this.name = null;
       }
 
    }
 
-   public void write(FriendlyByteBuf var1) throws IOException {
+   public void write(FriendlyByteBuf var1) {
       if (this.source != null) {
          if (this.name != null) {
             var1.writeByte(3);
             var1.writeEnum(this.source);
             var1.writeResourceLocation(this.name);
@@ -48,9 +52,19 @@
          var1.writeByte(0);
       }
 
    }
 
+   @Nullable
+   public ResourceLocation getName() {
+      return this.name;
+   }
+
+   @Nullable
+   public SoundSource getSource() {
+      return this.source;
+   }
+
    public void handle(ClientGamePacketListener var1) {
       var1.handleStopSoundEvent(this);
    }
 }
