--- net/minecraft/network/protocol/game/ClientboundPlaceGhostRecipePacket.java
+++ net/minecraft/network/protocol/game/ClientboundPlaceGhostRecipePacket.java
@@ -1,34 +1,38 @@
 package net.minecraft.network.protocol.game;
 
-import java.io.IOException;
 import net.minecraft.network.FriendlyByteBuf;
 import net.minecraft.network.protocol.Packet;
 import net.minecraft.resources.ResourceLocation;
 import net.minecraft.world.item.crafting.Recipe;
 
 public class ClientboundPlaceGhostRecipePacket implements Packet<ClientGamePacketListener> {
-   private int containerId;
-   private ResourceLocation recipe;
+   private final int containerId;
+   private final ResourceLocation recipe;
 
-   public ClientboundPlaceGhostRecipePacket() {
-   }
-
    public ClientboundPlaceGhostRecipePacket(int var1, Recipe<?> var2) {
       this.containerId = var1;
       this.recipe = var2.getId();
    }
 
-   public void read(FriendlyByteBuf var1) throws IOException {
+   public ClientboundPlaceGhostRecipePacket(FriendlyByteBuf var1) {
       this.containerId = var1.readByte();
       this.recipe = var1.readResourceLocation();
    }
 
-   public void write(FriendlyByteBuf var1) throws IOException {
+   public void write(FriendlyByteBuf var1) {
       var1.writeByte(this.containerId);
       var1.writeResourceLocation(this.recipe);
    }
 
    public void handle(ClientGamePacketListener var1) {
       var1.handlePlaceRecipe(this);
    }
+
+   public ResourceLocation getRecipe() {
+      return this.recipe;
+   }
+
+   public int getContainerId() {
+      return this.containerId;
+   }
 }
