--- net/minecraft/network/protocol/game/DebugPackets.java
+++ net/minecraft/network/protocol/game/DebugPackets.java
@@ -17,10 +17,13 @@
 import net.minecraft.world.entity.raid.Raid;
 import net.minecraft.world.level.ChunkPos;
 import net.minecraft.world.level.Level;
 import net.minecraft.world.level.WorldGenLevel;
 import net.minecraft.world.level.block.entity.BeehiveBlockEntity;
+import net.minecraft.world.level.block.state.BlockState;
+import net.minecraft.world.level.gameevent.GameEvent;
+import net.minecraft.world.level.gameevent.GameEventListener;
 import net.minecraft.world.level.levelgen.structure.StructureStart;
 import net.minecraft.world.level.pathfinder.Path;
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
 
@@ -81,13 +84,19 @@
    }
 
    public static void sendBeeInfo(Bee var0) {
    }
 
-   public static void sendHiveInfo(BeehiveBlockEntity var0) {
+   public static void sendGameEventInfo(Level var0, GameEvent var1, BlockPos var2) {
    }
 
+   public static void sendGameEventListenerInfo(Level var0, GameEventListener var1) {
+   }
+
+   public static void sendHiveInfo(Level var0, BlockPos var1, BlockState var2, BeehiveBlockEntity var3) {
+   }
+
    private static void sendPacketToAllPlayers(ServerLevel var0, FriendlyByteBuf var1, ResourceLocation var2) {
       ClientboundCustomPayloadPacket var3 = new ClientboundCustomPayloadPacket(var2, var1);
       Iterator var4 = var0.getLevel().players().iterator();
 
       while(var4.hasNext()) {
