--- net/minecraft/network/protocol/game/package-info.java
+++ net/minecraft/network/protocol/game/package-info.java
@@ -1,0 +1,12 @@
+package net.minecraft.network.protocol.game;
+
+import javax.annotation.ParametersAreNonnullByDefault;
+import net.minecraft.FieldsAreNonnullByDefault;
+import net.minecraft.MethodsReturnNonnullByDefault;
+
+// $FF: synthetic class
+@ParametersAreNonnullByDefault
+@MethodsReturnNonnullByDefault
+@FieldsAreNonnullByDefault
+interface package-info {
+}
