--- net/minecraft/network/protocol/game/ServerboundEditBookPacket.java
+++ net/minecraft/network/protocol/game/ServerboundEditBookPacket.java
@@ -1,27 +1,23 @@
 package net.minecraft.network.protocol.game;
 
-import java.io.IOException;
 import net.minecraft.network.FriendlyByteBuf;
 import net.minecraft.network.protocol.Packet;
 import net.minecraft.world.item.ItemStack;
 
 public class ServerboundEditBookPacket implements Packet<ServerGamePacketListener> {
-   private ItemStack book;
-   private boolean signing;
-   private int slot;
+   private final ItemStack book;
+   private final boolean signing;
+   private final int slot;
 
-   public ServerboundEditBookPacket() {
-   }
-
-   public void read(FriendlyByteBuf var1) throws IOException {
+   public ServerboundEditBookPacket(FriendlyByteBuf var1) {
       this.book = var1.readItem();
       this.signing = var1.readBoolean();
       this.slot = var1.readVarInt();
    }
 
-   public void write(FriendlyByteBuf var1) throws IOException {
+   public void write(FriendlyByteBuf var1) {
       var1.writeItem(this.book);
       var1.writeBoolean(this.signing);
       var1.writeVarInt(this.slot);
    }
 
