--- net/minecraft/network/protocol/game/ServerboundContainerClickPacket.java
+++ net/minecraft/network/protocol/game/ServerboundContainerClickPacket.java
@@ -1,47 +1,47 @@
 package net.minecraft.network.protocol.game;
 
-import java.io.IOException;
+import it.unimi.dsi.fastutil.ints.Int2ObjectMap;
+import it.unimi.dsi.fastutil.ints.Int2ObjectMaps;
+import it.unimi.dsi.fastutil.ints.Int2ObjectOpenHashMap;
 import net.minecraft.network.FriendlyByteBuf;
 import net.minecraft.network.protocol.Packet;
 import net.minecraft.world.inventory.ClickType;
 import net.minecraft.world.item.ItemStack;
 
 public class ServerboundContainerClickPacket implements Packet<ServerGamePacketListener> {
-   private int containerId;
-   private int slotNum;
-   private int buttonNum;
-   private short uid;
-   private ItemStack itemStack;
-   private ClickType clickType;
+   private final int containerId;
+   private final int slotNum;
+   private final int buttonNum;
+   private final ClickType clickType;
+   private final ItemStack carriedItem;
+   private final Int2ObjectMap<ItemStack> changedSlots;
 
-   public ServerboundContainerClickPacket() {
-      this.itemStack = ItemStack.EMPTY;
-   }
-
-   public void handle(ServerGamePacketListener var1) {
-      var1.handleContainerClick(this);
-   }
-
-   public void read(FriendlyByteBuf var1) throws IOException {
+   public ServerboundContainerClickPacket(FriendlyByteBuf var1) {
       this.containerId = var1.readByte();
       this.slotNum = var1.readShort();
       this.buttonNum = var1.readByte();
-      this.uid = var1.readShort();
       this.clickType = (ClickType)var1.readEnum(ClickType.class);
-      this.itemStack = var1.readItem();
+      this.changedSlots = Int2ObjectMaps.unmodifiable((Int2ObjectMap)var1.readMap(Int2ObjectOpenHashMap::new, (var0) -> {
+         return Integer.valueOf(var0.readShort());
+      }, FriendlyByteBuf::readItem));
+      this.carriedItem = var1.readItem();
    }
 
-   public void write(FriendlyByteBuf var1) throws IOException {
+   public void write(FriendlyByteBuf var1) {
       var1.writeByte(this.containerId);
       var1.writeShort(this.slotNum);
       var1.writeByte(this.buttonNum);
-      var1.writeShort(this.uid);
       var1.writeEnum(this.clickType);
-      var1.writeItem(this.itemStack);
+      var1.writeMap(this.changedSlots, FriendlyByteBuf::writeShort, FriendlyByteBuf::writeItem);
+      var1.writeItem(this.carriedItem);
    }
 
+   public void handle(ServerGamePacketListener var1) {
+      var1.handleContainerClick(this);
+   }
+
    public int getContainerId() {
       return this.containerId;
    }
 
    public int getSlotNum() {
@@ -50,16 +50,16 @@
 
    public int getButtonNum() {
       return this.buttonNum;
    }
 
-   public short getUid() {
-      return this.uid;
+   public ItemStack getCarriedItem() {
+      return this.carriedItem;
    }
 
-   public ItemStack getItem() {
-      return this.itemStack;
+   public Int2ObjectMap<ItemStack> getChangedSlots() {
+      return this.changedSlots;
    }
 
    public ClickType getClickType() {
       return this.clickType;
    }
