--- net/minecraft/network/protocol/game/ServerGamePacketListener.java
+++ net/minecraft/network/protocol/game/ServerGamePacketListener.java
@@ -9,12 +9,10 @@
 
    void handleClientCommand(ServerboundClientCommandPacket var1);
 
    void handleClientInformation(ServerboundClientInformationPacket var1);
 
-   void handleContainerAck(ServerboundContainerAckPacket var1);
-
    void handleContainerButtonClick(ServerboundContainerButtonClickPacket var1);
 
    void handleContainerClick(ServerboundContainerClickPacket var1);
 
    void handlePlaceRecipe(ServerboundPlaceRecipePacket var1);
@@ -27,10 +25,12 @@
 
    void handleKeepAlive(ServerboundKeepAlivePacket var1);
 
    void handleMovePlayer(ServerboundMovePlayerPacket var1);
 
+   void handlePong(ServerboundPongPacket var1);
+
    void handlePlayerAbilities(ServerboundPlayerAbilitiesPacket var1);
 
    void handlePlayerAction(ServerboundPlayerActionPacket var1);
 
    void handlePlayerCommand(ServerboundPlayerCommandPacket var1);
