--- net/minecraft/network/protocol/game/ServerboundSeenAdvancementsPacket.java
+++ net/minecraft/network/protocol/game/ServerboundSeenAdvancementsPacket.java
@@ -1,28 +1,42 @@
 package net.minecraft.network.protocol.game;
 
-import java.io.IOException;
+import javax.annotation.Nullable;
+import net.minecraft.advancements.Advancement;
 import net.minecraft.network.FriendlyByteBuf;
 import net.minecraft.network.protocol.Packet;
 import net.minecraft.resources.ResourceLocation;
 
 public class ServerboundSeenAdvancementsPacket implements Packet<ServerGamePacketListener> {
-   private ServerboundSeenAdvancementsPacket.Action action;
-   private ResourceLocation tab;
+   private final ServerboundSeenAdvancementsPacket.Action action;
+   @Nullable
+   private final ResourceLocation tab;
 
-   public ServerboundSeenAdvancementsPacket() {
+   public ServerboundSeenAdvancementsPacket(ServerboundSeenAdvancementsPacket.Action var1, @Nullable ResourceLocation var2) {
+      this.action = var1;
+      this.tab = var2;
    }
 
-   public void read(FriendlyByteBuf var1) throws IOException {
+   public static ServerboundSeenAdvancementsPacket openedTab(Advancement var0) {
+      return new ServerboundSeenAdvancementsPacket(ServerboundSeenAdvancementsPacket.Action.OPENED_TAB, var0.getId());
+   }
+
+   public static ServerboundSeenAdvancementsPacket closedScreen() {
+      return new ServerboundSeenAdvancementsPacket(ServerboundSeenAdvancementsPacket.Action.CLOSED_SCREEN, (ResourceLocation)null);
+   }
+
+   public ServerboundSeenAdvancementsPacket(FriendlyByteBuf var1) {
       this.action = (ServerboundSeenAdvancementsPacket.Action)var1.readEnum(ServerboundSeenAdvancementsPacket.Action.class);
       if (this.action == ServerboundSeenAdvancementsPacket.Action.OPENED_TAB) {
          this.tab = var1.readResourceLocation();
+      } else {
+         this.tab = null;
       }
 
    }
 
-   public void write(FriendlyByteBuf var1) throws IOException {
+   public void write(FriendlyByteBuf var1) {
       var1.writeEnum(this.action);
       if (this.action == ServerboundSeenAdvancementsPacket.Action.OPENED_TAB) {
          var1.writeResourceLocation(this.tab);
       }
 
@@ -34,10 +48,11 @@
 
    public ServerboundSeenAdvancementsPacket.Action getAction() {
       return this.action;
    }
 
+   @Nullable
    public ResourceLocation getTab() {
       return this.tab;
    }
 
    public static enum Action {
