--- net/minecraft/network/protocol/game/ServerboundInteractPacket.java
+++ net/minecraft/network/protocol/game/ServerboundInteractPacket.java
@@ -2,13 +2,13 @@
 
 import java.io.IOException;
 import javax.annotation.Nullable;
 import net.minecraft.network.FriendlyByteBuf;
 import net.minecraft.network.protocol.Packet;
+import net.minecraft.server.level.ServerLevel;
 import net.minecraft.world.InteractionHand;
 import net.minecraft.world.entity.Entity;
-import net.minecraft.world.level.Level;
 import net.minecraft.world.phys.Vec3;
 
 public class ServerboundInteractPacket implements Packet<ServerGamePacketListener> {
    private int entityId;
    private ServerboundInteractPacket.Action action;
@@ -52,12 +52,12 @@
    public void handle(ServerGamePacketListener var1) {
       var1.handleInteract(this);
    }
 
    @Nullable
-   public Entity getTarget(Level var1) {
-      return var1.getEntity(this.entityId);
+   public Entity getTarget(ServerLevel var1) {
+      return var1.getEntityOrPart(this.entityId);
    }
 
    public ServerboundInteractPacket.Action getAction() {
       return this.action;
    }
