--- net/minecraft/network/protocol/game/ServerboundSetCreativeModeSlotPacket.java
+++ net/minecraft/network/protocol/game/ServerboundSetCreativeModeSlotPacket.java
@@ -1,30 +1,25 @@
 package net.minecraft.network.protocol.game;
 
-import java.io.IOException;
 import net.minecraft.network.FriendlyByteBuf;
 import net.minecraft.network.protocol.Packet;
 import net.minecraft.world.item.ItemStack;
 
 public class ServerboundSetCreativeModeSlotPacket implements Packet<ServerGamePacketListener> {
-   private int slotNum;
-   private ItemStack itemStack;
+   private final int slotNum;
+   private final ItemStack itemStack;
 
-   public ServerboundSetCreativeModeSlotPacket() {
-      this.itemStack = ItemStack.EMPTY;
-   }
-
    public void handle(ServerGamePacketListener var1) {
       var1.handleSetCreativeModeSlot(this);
    }
 
-   public void read(FriendlyByteBuf var1) throws IOException {
+   public ServerboundSetCreativeModeSlotPacket(FriendlyByteBuf var1) {
       this.slotNum = var1.readShort();
       this.itemStack = var1.readItem();
    }
 
-   public void write(FriendlyByteBuf var1) throws IOException {
+   public void write(FriendlyByteBuf var1) {
       var1.writeShort(this.slotNum);
       var1.writeItem(this.itemStack);
    }
 
    public int getSlotNum() {
