--- net/minecraft/network/protocol/game/ServerboundSetBeaconPacket.java
+++ net/minecraft/network/protocol/game/ServerboundSetBeaconPacket.java
@@ -1,24 +1,25 @@
 package net.minecraft.network.protocol.game;
 
-import java.io.IOException;
 import net.minecraft.network.FriendlyByteBuf;
 import net.minecraft.network.protocol.Packet;
 
 public class ServerboundSetBeaconPacket implements Packet<ServerGamePacketListener> {
-   private int primary;
-   private int secondary;
+   private final int primary;
+   private final int secondary;
 
-   public ServerboundSetBeaconPacket() {
+   public ServerboundSetBeaconPacket(int var1, int var2) {
+      this.primary = var1;
+      this.secondary = var2;
    }
 
-   public void read(FriendlyByteBuf var1) throws IOException {
+   public ServerboundSetBeaconPacket(FriendlyByteBuf var1) {
       this.primary = var1.readVarInt();
       this.secondary = var1.readVarInt();
    }
 
-   public void write(FriendlyByteBuf var1) throws IOException {
+   public void write(FriendlyByteBuf var1) {
       var1.writeVarInt(this.primary);
       var1.writeVarInt(this.secondary);
    }
 
    public void handle(ServerGamePacketListener var1) {
