--- net/minecraft/network/protocol/game/ServerboundPlayerCommandPacket.java
+++ net/minecraft/network/protocol/game/ServerboundPlayerCommandPacket.java
@@ -1,35 +1,46 @@
 package net.minecraft.network.protocol.game;
 
-import java.io.IOException;
 import net.minecraft.network.FriendlyByteBuf;
 import net.minecraft.network.protocol.Packet;
+import net.minecraft.world.entity.Entity;
 
 public class ServerboundPlayerCommandPacket implements Packet<ServerGamePacketListener> {
-   private int id;
-   private ServerboundPlayerCommandPacket.Action action;
-   private int data;
+   private final int id;
+   private final ServerboundPlayerCommandPacket.Action action;
+   private final int data;
 
-   public ServerboundPlayerCommandPacket() {
+   public ServerboundPlayerCommandPacket(Entity var1, ServerboundPlayerCommandPacket.Action var2) {
+      this(var1, var2, 0);
    }
 
-   public void read(FriendlyByteBuf var1) throws IOException {
+   public ServerboundPlayerCommandPacket(Entity var1, ServerboundPlayerCommandPacket.Action var2, int var3) {
+      this.id = var1.getId();
+      this.action = var2;
+      this.data = var3;
+   }
+
+   public ServerboundPlayerCommandPacket(FriendlyByteBuf var1) {
       this.id = var1.readVarInt();
       this.action = (ServerboundPlayerCommandPacket.Action)var1.readEnum(ServerboundPlayerCommandPacket.Action.class);
       this.data = var1.readVarInt();
    }
 
-   public void write(FriendlyByteBuf var1) throws IOException {
+   public void write(FriendlyByteBuf var1) {
       var1.writeVarInt(this.id);
       var1.writeEnum(this.action);
       var1.writeVarInt(this.data);
    }
 
    public void handle(ServerGamePacketListener var1) {
       var1.handlePlayerCommand(this);
    }
 
+   public int getId() {
+      return this.id;
+   }
+
    public ServerboundPlayerCommandPacket.Action getAction() {
       return this.action;
    }
 
    public int getData() {
