--- net/minecraft/network/protocol/game/ServerboundRecipeBookChangeSettingsPacket.java
+++ net/minecraft/network/protocol/game/ServerboundRecipeBookChangeSettingsPacket.java
@@ -1,27 +1,23 @@
 package net.minecraft.network.protocol.game;
 
-import java.io.IOException;
 import net.minecraft.network.FriendlyByteBuf;
 import net.minecraft.network.protocol.Packet;
 import net.minecraft.world.inventory.RecipeBookType;
 
 public class ServerboundRecipeBookChangeSettingsPacket implements Packet<ServerGamePacketListener> {
-   private RecipeBookType bookType;
-   private boolean isOpen;
-   private boolean isFiltering;
+   private final RecipeBookType bookType;
+   private final boolean isOpen;
+   private final boolean isFiltering;
 
-   public ServerboundRecipeBookChangeSettingsPacket() {
-   }
-
-   public void read(FriendlyByteBuf var1) throws IOException {
+   public ServerboundRecipeBookChangeSettingsPacket(FriendlyByteBuf var1) {
       this.bookType = (RecipeBookType)var1.readEnum(RecipeBookType.class);
       this.isOpen = var1.readBoolean();
       this.isFiltering = var1.readBoolean();
    }
 
-   public void write(FriendlyByteBuf var1) throws IOException {
+   public void write(FriendlyByteBuf var1) {
       var1.writeEnum(this.bookType);
       var1.writeBoolean(this.isOpen);
       var1.writeBoolean(this.isFiltering);
    }
 
