--- net/minecraft/network/protocol/game/ServerboundPlaceRecipePacket.java
+++ net/minecraft/network/protocol/game/ServerboundPlaceRecipePacket.java
@@ -1,27 +1,23 @@
 package net.minecraft.network.protocol.game;
 
-import java.io.IOException;
 import net.minecraft.network.FriendlyByteBuf;
 import net.minecraft.network.protocol.Packet;
 import net.minecraft.resources.ResourceLocation;
 
 public class ServerboundPlaceRecipePacket implements Packet<ServerGamePacketListener> {
-   private int containerId;
-   private ResourceLocation recipe;
-   private boolean shiftDown;
+   private final int containerId;
+   private final ResourceLocation recipe;
+   private final boolean shiftDown;
 
-   public ServerboundPlaceRecipePacket() {
-   }
-
-   public void read(FriendlyByteBuf var1) throws IOException {
+   public ServerboundPlaceRecipePacket(FriendlyByteBuf var1) {
       this.containerId = var1.readByte();
       this.recipe = var1.readResourceLocation();
       this.shiftDown = var1.readBoolean();
    }
 
-   public void write(FriendlyByteBuf var1) throws IOException {
+   public void write(FriendlyByteBuf var1) {
       var1.writeByte(this.containerId);
       var1.writeResourceLocation(this.recipe);
       var1.writeBoolean(this.shiftDown);
    }
 
