--- net/minecraft/network/protocol/game/ServerboundPlayerAbilitiesPacket.java
+++ net/minecraft/network/protocol/game/ServerboundPlayerAbilitiesPacket.java
@@ -1,28 +1,25 @@
 package net.minecraft.network.protocol.game;
 
-import java.io.IOException;
 import net.minecraft.network.FriendlyByteBuf;
 import net.minecraft.network.protocol.Packet;
 import net.minecraft.world.entity.player.Abilities;
 
 public class ServerboundPlayerAbilitiesPacket implements Packet<ServerGamePacketListener> {
-   private boolean isFlying;
+   private static final int FLAG_FLYING = 2;
+   private final boolean isFlying;
 
-   public ServerboundPlayerAbilitiesPacket() {
-   }
-
    public ServerboundPlayerAbilitiesPacket(Abilities var1) {
       this.isFlying = var1.flying;
    }
 
-   public void read(FriendlyByteBuf var1) throws IOException {
+   public ServerboundPlayerAbilitiesPacket(FriendlyByteBuf var1) {
       byte var2 = var1.readByte();
       this.isFlying = (var2 & 2) != 0;
    }
 
-   public void write(FriendlyByteBuf var1) throws IOException {
+   public void write(FriendlyByteBuf var1) {
       byte var2 = 0;
       if (this.isFlying) {
          var2 = (byte)(var2 | 2);
       }
 
