--- net/minecraft/util/profiling/registry/MeasurementRegistry.java
+++ net/minecraft/util/profiling/registry/MeasurementRegistry.java
@@ -1,0 +1,25 @@
+package net.minecraft.util.profiling.registry;
+
+import java.util.EnumMap;
+import java.util.List;
+import java.util.Map;
+import java.util.WeakHashMap;
+import java.util.stream.Collectors;
+
+public class MeasurementRegistry {
+   public static final MeasurementRegistry INSTANCE = new MeasurementRegistry();
+   private final WeakHashMap<ProfilerMeasured, Void> measuredInstances = new WeakHashMap();
+
+   private MeasurementRegistry() {
+   }
+
+   public void add(ProfilerMeasured var1) {
+      this.measuredInstances.put(var1, (Object)null);
+   }
+
+   public Map<MeasurementCategory, List<MeasuredMetric>> getMetricsByCategories() {
+      return (Map)this.measuredInstances.keySet().stream().flatMap((var0) -> {
+         return var0.metrics().stream();
+      }).collect(Collectors.collectingAndThen(Collectors.groupingBy(MeasuredMetric::getGetCategory), EnumMap::new));
+   }
+}
