--- net/minecraft/world/entity/ai/goal/target/ResetUniversalAngerTargetGoal.java
+++ net/minecraft/world/entity/ai/goal/target/ResetUniversalAngerTargetGoal.java
@@ -1,17 +1,19 @@
 package net.minecraft.world.entity.ai.goal.target;
 
 import java.util.List;
+import net.minecraft.world.entity.EntitySelector;
 import net.minecraft.world.entity.EntityType;
 import net.minecraft.world.entity.Mob;
 import net.minecraft.world.entity.NeutralMob;
 import net.minecraft.world.entity.ai.attributes.Attributes;
 import net.minecraft.world.entity.ai.goal.Goal;
 import net.minecraft.world.level.GameRules;
 import net.minecraft.world.phys.AABB;
 
 public class ResetUniversalAngerTargetGoal<T extends Mob & NeutralMob> extends Goal {
+   private static final int ALERT_RANGE_Y = 10;
    private final T mob;
    private final boolean alertOthersOfSameType;
    private int lastHurtByPlayerTimestamp;
 
    public ResetUniversalAngerTargetGoal(T var1, boolean var2) {
@@ -39,11 +41,11 @@
       }
 
       super.start();
    }
 
-   private List<Mob> getNearbyMobsOfSameType() {
+   private List<? extends Mob> getNearbyMobsOfSameType() {
       double var1 = this.mob.getAttributeValue(Attributes.FOLLOW_RANGE);
       AABB var3 = AABB.unitCubeFromLowerCorner(this.mob.position()).inflate(var1, 10.0D, var1);
-      return this.mob.level.getLoadedEntitiesOfClass(this.mob.getClass(), var3);
+      return this.mob.level.getEntitiesOfClass(this.mob.getClass(), var3, EntitySelector.NO_SPECTATORS);
    }
 }
