--- net/minecraft/world/level/storage/loot/predicates/BonusLevelTableCondition.java
+++ net/minecraft/world/level/storage/loot/predicates/BonusLevelTableCondition.java
@@ -15,14 +15,14 @@
 import net.minecraft.world.level.storage.loot.LootContext;
 import net.minecraft.world.level.storage.loot.parameters.LootContextParam;
 import net.minecraft.world.level.storage.loot.parameters.LootContextParams;
 
 public class BonusLevelTableCondition implements LootItemCondition {
-   private final Enchantment enchantment;
-   private final float[] values;
+   final Enchantment enchantment;
+   final float[] values;
 
-   private BonusLevelTableCondition(Enchantment var1, float[] var2) {
+   BonusLevelTableCondition(Enchantment var1, float[] var2) {
       this.enchantment = var1;
       this.values = var2;
    }
 
    public LootItemConditionType getType() {
