--- net/minecraft/world/level/storage/loot/predicates/DamageSourceCondition.java
+++ net/minecraft/world/level/storage/loot/predicates/DamageSourceCondition.java
@@ -11,13 +11,13 @@
 import net.minecraft.world.level.storage.loot.parameters.LootContextParam;
 import net.minecraft.world.level.storage.loot.parameters.LootContextParams;
 import net.minecraft.world.phys.Vec3;
 
 public class DamageSourceCondition implements LootItemCondition {
-   private final DamageSourcePredicate predicate;
+   final DamageSourcePredicate predicate;
 
-   private DamageSourceCondition(DamageSourcePredicate var1) {
+   DamageSourceCondition(DamageSourcePredicate var1) {
       this.predicate = var1;
    }
 
    public LootItemConditionType getType() {
       return LootItemConditions.DAMAGE_SOURCE_PROPERTIES;
