--- net/minecraft/world/level/storage/loot/predicates/ExplosionCondition.java
+++ net/minecraft/world/level/storage/loot/predicates/ExplosionCondition.java
@@ -9,11 +9,11 @@
 import net.minecraft.world.level.storage.loot.LootContext;
 import net.minecraft.world.level.storage.loot.parameters.LootContextParam;
 import net.minecraft.world.level.storage.loot.parameters.LootContextParams;
 
 public class ExplosionCondition implements LootItemCondition {
-   private static final ExplosionCondition INSTANCE = new ExplosionCondition();
+   static final ExplosionCondition INSTANCE = new ExplosionCondition();
 
    private ExplosionCondition() {
    }
 
    public LootItemConditionType getType() {
