--- net/minecraft/world/level/storage/loot/predicates/LocationCheck.java
+++ net/minecraft/world/level/storage/loot/predicates/LocationCheck.java
@@ -9,14 +9,14 @@
 import net.minecraft.world.level.storage.loot.LootContext;
 import net.minecraft.world.level.storage.loot.parameters.LootContextParams;
 import net.minecraft.world.phys.Vec3;
 
 public class LocationCheck implements LootItemCondition {
-   private final LocationPredicate predicate;
-   private final BlockPos offset;
+   final LocationPredicate predicate;
+   final BlockPos offset;
 
-   private LocationCheck(LocationPredicate var1, BlockPos var2) {
+   LocationCheck(LocationPredicate var1, BlockPos var2) {
       this.predicate = var1;
       this.offset = var2;
    }
 
    public LootItemConditionType getType() {
