--- net/minecraft/world/level/storage/loot/predicates/LootItemConditions.java
+++ net/minecraft/world/level/storage/loot/predicates/LootItemConditions.java
@@ -21,10 +21,11 @@
    public static final LootItemConditionType DAMAGE_SOURCE_PROPERTIES = register("damage_source_properties", new DamageSourceCondition.Serializer());
    public static final LootItemConditionType LOCATION_CHECK = register("location_check", new LocationCheck.Serializer());
    public static final LootItemConditionType WEATHER_CHECK = register("weather_check", new WeatherCheck.Serializer());
    public static final LootItemConditionType REFERENCE = register("reference", new ConditionReference.Serializer());
    public static final LootItemConditionType TIME_CHECK = register("time_check", new TimeCheck.Serializer());
+   public static final LootItemConditionType VALUE_CHECK = register("value_check", new ValueCheckCondition.Serializer());
 
    private static LootItemConditionType register(String var0, Serializer<? extends LootItemCondition> var1) {
       return (LootItemConditionType)Registry.register(Registry.LOOT_CONDITION_TYPE, (ResourceLocation)(new ResourceLocation(var0)), new LootItemConditionType(var1));
    }
 
