--- net/minecraft/world/level/storage/loot/predicates/MatchTool.java
+++ net/minecraft/world/level/storage/loot/predicates/MatchTool.java
@@ -10,11 +10,11 @@
 import net.minecraft.world.level.storage.loot.LootContext;
 import net.minecraft.world.level.storage.loot.parameters.LootContextParam;
 import net.minecraft.world.level.storage.loot.parameters.LootContextParams;
 
 public class MatchTool implements LootItemCondition {
-   private final ItemPredicate predicate;
+   final ItemPredicate predicate;
 
    public MatchTool(ItemPredicate var1) {
       this.predicate = var1;
    }
 
