--- com/mojang/authlib/exceptions/MinecraftClientException.java
+++ com/mojang/authlib/exceptions/MinecraftClientException.java
@@ -1,0 +1,32 @@
+package com.mojang.authlib.exceptions;
+
+public class MinecraftClientException extends RuntimeException {
+   protected final MinecraftClientException.ErrorType type;
+
+   protected MinecraftClientException(MinecraftClientException.ErrorType var1, String var2) {
+      super(var2);
+      this.type = var1;
+   }
+
+   public MinecraftClientException(MinecraftClientException.ErrorType var1, String var2, Throwable var3) {
+      super(var2, var3);
+      this.type = var1;
+   }
+
+   public MinecraftClientException.ErrorType getType() {
+      return this.type;
+   }
+
+   public AuthenticationException toAuthenticationException() {
+      return new AuthenticationException(this);
+   }
+
+   public static enum ErrorType {
+      SERVICE_UNAVAILABLE,
+      HTTP_ERROR,
+      JSON_ERROR;
+
+      private ErrorType() {
+      }
+   }
+}
