--- net/minecraft/world/entity/decoration/Painting.java
+++ net/minecraft/world/entity/decoration/Painting.java
@@ -86,11 +86,11 @@
    public void dropItem(@Nullable Entity var1) {
       if (this.level.getGameRules().getBoolean(GameRules.RULE_DOENTITYDROPS)) {
          this.playSound(SoundEvents.PAINTING_BREAK, 1.0F, 1.0F);
          if (var1 instanceof Player) {
             Player var2 = (Player)var1;
-            if (var2.abilities.instabuild) {
+            if (var2.getAbilities().instabuild) {
                return;
             }
          }
 
          this.spawnAtLocation(Items.PAINTING);
