--- net/minecraft/world/level/block/grower/AbstractMegaTreeGrower.java
+++ net/minecraft/world/level/block/grower/AbstractMegaTreeGrower.java
@@ -34,11 +34,10 @@
    public boolean placeMega(ServerLevel var1, ChunkGenerator var2, BlockPos var3, BlockState var4, Random var5, int var6, int var7) {
       ConfiguredFeature var8 = this.getConfiguredMegaFeature(var5);
       if (var8 == null) {
          return false;
       } else {
-         ((TreeConfiguration)var8.config).setFromSapling();
          BlockState var9 = Blocks.AIR.defaultBlockState();
          var1.setBlock(var3.offset(var6, 0, var7), var9, 4);
          var1.setBlock(var3.offset(var6 + 1, 0, var7), var9, 4);
          var1.setBlock(var3.offset(var6, 0, var7 + 1), var9, 4);
          var1.setBlock(var3.offset(var6 + 1, 0, var7 + 1), var9, 4);
@@ -54,8 +53,8 @@
       }
    }
 
    public static boolean isTwoByTwoSapling(BlockState var0, BlockGetter var1, BlockPos var2, int var3, int var4) {
       Block var5 = var0.getBlock();
-      return var5 == var1.getBlockState(var2.offset(var3, 0, var4)).getBlock() && var5 == var1.getBlockState(var2.offset(var3 + 1, 0, var4)).getBlock() && var5 == var1.getBlockState(var2.offset(var3, 0, var4 + 1)).getBlock() && var5 == var1.getBlockState(var2.offset(var3 + 1, 0, var4 + 1)).getBlock();
+      return var1.getBlockState(var2.offset(var3, 0, var4)).is(var5) && var1.getBlockState(var2.offset(var3 + 1, 0, var4)).is(var5) && var1.getBlockState(var2.offset(var3, 0, var4 + 1)).is(var5) && var1.getBlockState(var2.offset(var3 + 1, 0, var4 + 1)).is(var5);
    }
 }
