--- net/minecraft/util/datafix/schemas/package-info.java
+++ net/minecraft/util/datafix/schemas/package-info.java
@@ -1,0 +1,10 @@
+package net.minecraft.util.datafix.schemas;
+
+import javax.annotation.ParametersAreNonnullByDefault;
+import net.minecraft.MethodsReturnNonnullByDefault;
+
+// $FF: synthetic class
+@ParametersAreNonnullByDefault
+@MethodsReturnNonnullByDefault
+interface package-info {
+}
