--- net/minecraft/util/datafix/schemas/V1460.java
+++ net/minecraft/util/datafix/schemas/V1460.java
@@ -258,7 +258,10 @@
       var1.registerType(false, References.ENTITY_NAME, () -> {
          return DSL.constType(namespacedString());
       });
       var1.registerType(false, References.POI_CHUNK, DSL::remainder);
       var1.registerType(true, References.WORLD_GEN_SETTINGS, DSL::remainder);
+      var1.registerType(false, References.ENTITY_CHUNK, () -> {
+         return DSL.optionalFields("Entities", DSL.list(References.ENTITY_TREE.in(var1)));
+      });
    }
 }
