--- net/minecraft/util/datafix/schemas/V1486.java
+++ net/minecraft/util/datafix/schemas/V1486.java
@@ -10,10 +10,10 @@
       super(var1, var2);
    }
 
    public Map<String, Supplier<TypeTemplate>> registerEntities(Schema var1) {
       Map var2 = super.registerEntities(var1);
-      var2.put("minecraft:cod", var2.remove("minecraft:cod_mob"));
-      var2.put("minecraft:salmon", var2.remove("minecraft:salmon_mob"));
+      var2.put("minecraft:cod", (Supplier)var2.remove("minecraft:cod_mob"));
+      var2.put("minecraft:salmon", (Supplier)var2.remove("minecraft:salmon_mob"));
       return var2;
    }
 }
