--- net/minecraft/util/datafix/schemas/V2100.java
+++ net/minecraft/util/datafix/schemas/V2100.java
@@ -26,10 +26,10 @@
    }
 
    public Map<String, Supplier<TypeTemplate>> registerBlockEntities(Schema var1) {
       Map var2 = super.registerBlockEntities(var1);
       var1.register(var2, "minecraft:beehive", () -> {
-         return DSL.optionalFields("Items", DSL.list(References.ITEM_STACK.in(var1)), "Bees", DSL.list(DSL.optionalFields("EntityData", References.ENTITY_TREE.in(var1))));
+         return DSL.optionalFields("Bees", DSL.list(DSL.optionalFields("EntityData", References.ENTITY_TREE.in(var1))));
       });
       return var2;
    }
 }
