--- net/minecraft/util/datafix/schemas/V99.java
+++ net/minecraft/util/datafix/schemas/V99.java
@@ -266,10 +266,13 @@
       var1.registerType(false, References.OBJECTIVE, DSL::remainder);
       var1.registerType(false, References.TEAM, DSL::remainder);
       var1.registerType(true, References.UNTAGGED_SPAWNER, DSL::remainder);
       var1.registerType(false, References.POI_CHUNK, DSL::remainder);
       var1.registerType(true, References.WORLD_GEN_SETTINGS, DSL::remainder);
+      var1.registerType(false, References.ENTITY_CHUNK, () -> {
+         return DSL.optionalFields("Entities", DSL.list(References.ENTITY_TREE.in(var1)));
+      });
    }
 
    protected static <T> T addNames(Dynamic<T> var0, Map<String, String> var1, String var2) {
       return var0.update("tag", (var3) -> {
          return var3.update("BlockEntityTag", (var2x) -> {
