--- net/minecraft/network/protocol/game/ClientGamePacketListener.java
+++ net/minecraft/network/protocol/game/ClientGamePacketListener.java
@@ -5,10 +5,12 @@
 public interface ClientGamePacketListener extends PacketListener {
    void handleAddEntity(ClientboundAddEntityPacket var1);
 
    void handleAddExperienceOrb(ClientboundAddExperienceOrbPacket var1);
 
+   void handleAddVibrationSignal(ClientboundAddVibrationSignalPacket var1);
+
    void handleAddMob(ClientboundAddMobPacket var1);
 
    void handleAddObjective(ClientboundSetObjectivePacket var1);
 
    void handleAddPainting(ClientboundAddPaintingPacket var1);
