--- net/minecraft/network/protocol/game/ClientboundBlockEntityDataPacket.java
+++ net/minecraft/network/protocol/game/ClientboundBlockEntityDataPacket.java
@@ -1,34 +1,43 @@
 package net.minecraft.network.protocol.game;
 
-import java.io.IOException;
 import net.minecraft.core.BlockPos;
 import net.minecraft.nbt.CompoundTag;
 import net.minecraft.network.FriendlyByteBuf;
 import net.minecraft.network.protocol.Packet;
 
 public class ClientboundBlockEntityDataPacket implements Packet<ClientGamePacketListener> {
-   private BlockPos pos;
-   private int type;
-   private CompoundTag tag;
+   public static final int TYPE_MOB_SPAWNER = 1;
+   public static final int TYPE_ADV_COMMAND = 2;
+   public static final int TYPE_BEACON = 3;
+   public static final int TYPE_SKULL = 4;
+   public static final int TYPE_CONDUIT = 5;
+   public static final int TYPE_BANNER = 6;
+   public static final int TYPE_STRUCT_COMMAND = 7;
+   public static final int TYPE_END_GATEWAY = 8;
+   public static final int TYPE_SIGN = 9;
+   public static final int TYPE_BED = 11;
+   public static final int TYPE_JIGSAW = 12;
+   public static final int TYPE_CAMPFIRE = 13;
+   public static final int TYPE_BEEHIVE = 14;
+   private final BlockPos pos;
+   private final int type;
+   private final CompoundTag tag;
 
-   public ClientboundBlockEntityDataPacket() {
-   }
-
    public ClientboundBlockEntityDataPacket(BlockPos var1, int var2, CompoundTag var3) {
       this.pos = var1;
       this.type = var2;
       this.tag = var3;
    }
 
-   public void read(FriendlyByteBuf var1) throws IOException {
+   public ClientboundBlockEntityDataPacket(FriendlyByteBuf var1) {
       this.pos = var1.readBlockPos();
       this.type = var1.readUnsignedByte();
       this.tag = var1.readNbt();
    }
 
-   public void write(FriendlyByteBuf var1) throws IOException {
+   public void write(FriendlyByteBuf var1) {
       var1.writeBlockPos(this.pos);
       var1.writeByte((byte)this.type);
       var1.writeNbt(this.tag);
    }
 
