--- net/minecraft/network/protocol/game/ClientboundSetDisplayObjectivePacket.java
+++ net/minecraft/network/protocol/game/ClientboundSetDisplayObjectivePacket.java
@@ -1,37 +1,33 @@
 package net.minecraft.network.protocol.game;
 
-import java.io.IOException;
 import java.util.Objects;
 import javax.annotation.Nullable;
 import net.minecraft.network.FriendlyByteBuf;
 import net.minecraft.network.protocol.Packet;
 import net.minecraft.world.scores.Objective;
 
 public class ClientboundSetDisplayObjectivePacket implements Packet<ClientGamePacketListener> {
-   private int slot;
-   private String objectiveName;
+   private final int slot;
+   private final String objectiveName;
 
-   public ClientboundSetDisplayObjectivePacket() {
-   }
-
    public ClientboundSetDisplayObjectivePacket(int var1, @Nullable Objective var2) {
       this.slot = var1;
       if (var2 == null) {
          this.objectiveName = "";
       } else {
          this.objectiveName = var2.getName();
       }
 
    }
 
-   public void read(FriendlyByteBuf var1) throws IOException {
+   public ClientboundSetDisplayObjectivePacket(FriendlyByteBuf var1) {
       this.slot = var1.readByte();
       this.objectiveName = var1.readUtf(16);
    }
 
-   public void write(FriendlyByteBuf var1) throws IOException {
+   public void write(FriendlyByteBuf var1) {
       var1.writeByte(this.slot);
       var1.writeUtf(this.objectiveName);
    }
 
    public void handle(ClientGamePacketListener var1) {
