--- net/minecraft/network/protocol/game/ClientboundSetTitleTextPacket.java
+++ net/minecraft/network/protocol/game/ClientboundSetTitleTextPacket.java
@@ -1,0 +1,29 @@
+package net.minecraft.network.protocol.game;
+
+import net.minecraft.network.FriendlyByteBuf;
+import net.minecraft.network.chat.Component;
+import net.minecraft.network.protocol.Packet;
+
+public class ClientboundSetTitleTextPacket implements Packet<ClientGamePacketListener> {
+   private final Component text;
+
+   public ClientboundSetTitleTextPacket(Component var1) {
+      this.text = var1;
+   }
+
+   public ClientboundSetTitleTextPacket(FriendlyByteBuf var1) {
+      this.text = var1.readComponent();
+   }
+
+   public void write(FriendlyByteBuf var1) {
+      var1.writeComponent(this.text);
+   }
+
+   public void handle(ClientGamePacketListener var1) {
+      var1.setTitleText(this);
+   }
+
+   public Component getText() {
+      return this.text;
+   }
+}
