--- net/minecraft/network/protocol/game/ClientboundAddMobPacket.java
+++ net/minecraft/network/protocol/game/ClientboundAddMobPacket.java
@@ -1,42 +1,38 @@
 package net.minecraft.network.protocol.game;
 
-import java.io.IOException;
 import java.util.UUID;
 import net.minecraft.core.Registry;
 import net.minecraft.network.FriendlyByteBuf;
 import net.minecraft.network.protocol.Packet;
 import net.minecraft.util.Mth;
 import net.minecraft.world.entity.LivingEntity;
 import net.minecraft.world.phys.Vec3;
 
 public class ClientboundAddMobPacket implements Packet<ClientGamePacketListener> {
-   private int id;
-   private UUID uuid;
-   private int type;
-   private double x;
-   private double y;
-   private double z;
-   private int xd;
-   private int yd;
-   private int zd;
-   private byte yRot;
-   private byte xRot;
-   private byte yHeadRot;
+   private final int id;
+   private final UUID uuid;
+   private final int type;
+   private final double x;
+   private final double y;
+   private final double z;
+   private final int xd;
+   private final int yd;
+   private final int zd;
+   private final byte yRot;
+   private final byte xRot;
+   private final byte yHeadRot;
 
-   public ClientboundAddMobPacket() {
-   }
-
    public ClientboundAddMobPacket(LivingEntity var1) {
       this.id = var1.getId();
       this.uuid = var1.getUUID();
       this.type = Registry.ENTITY_TYPE.getId(var1.getType());
       this.x = var1.getX();
       this.y = var1.getY();
       this.z = var1.getZ();
-      this.yRot = (byte)((int)(var1.yRot * 256.0F / 360.0F));
-      this.xRot = (byte)((int)(var1.xRot * 256.0F / 360.0F));
+      this.yRot = (byte)((int)(var1.getYRot() * 256.0F / 360.0F));
+      this.xRot = (byte)((int)(var1.getXRot() * 256.0F / 360.0F));
       this.yHeadRot = (byte)((int)(var1.yHeadRot * 256.0F / 360.0F));
       double var2 = 3.9D;
       Vec3 var4 = var1.getDeltaMovement();
       double var5 = Mth.clamp(var4.x, -3.9D, 3.9D);
       double var7 = Mth.clamp(var4.y, -3.9D, 3.9D);
@@ -44,11 +40,11 @@
       this.xd = (int)(var5 * 8000.0D);
       this.yd = (int)(var7 * 8000.0D);
       this.zd = (int)(var9 * 8000.0D);
    }
 
-   public void read(FriendlyByteBuf var1) throws IOException {
+   public ClientboundAddMobPacket(FriendlyByteBuf var1) {
       this.id = var1.readVarInt();
       this.uuid = var1.readUUID();
       this.type = var1.readVarInt();
       this.x = var1.readDouble();
       this.y = var1.readDouble();
@@ -59,11 +55,11 @@
       this.xd = var1.readShort();
       this.yd = var1.readShort();
       this.zd = var1.readShort();
    }
 
-   public void write(FriendlyByteBuf var1) throws IOException {
+   public void write(FriendlyByteBuf var1) {
       var1.writeVarInt(this.id);
       var1.writeUUID(this.uuid);
       var1.writeVarInt(this.type);
       var1.writeDouble(this.x);
       var1.writeDouble(this.y);
