--- net/minecraft/network/protocol/game/ClientboundAddPaintingPacket.java
+++ net/minecraft/network/protocol/game/ClientboundAddPaintingPacket.java
@@ -1,44 +1,40 @@
 package net.minecraft.network.protocol.game;
 
-import java.io.IOException;
 import java.util.UUID;
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.Direction;
 import net.minecraft.core.Registry;
 import net.minecraft.network.FriendlyByteBuf;
 import net.minecraft.network.protocol.Packet;
 import net.minecraft.world.entity.decoration.Motive;
 import net.minecraft.world.entity.decoration.Painting;
 
 public class ClientboundAddPaintingPacket implements Packet<ClientGamePacketListener> {
-   private int id;
-   private UUID uuid;
-   private BlockPos pos;
-   private Direction direction;
-   private int motive;
+   private final int id;
+   private final UUID uuid;
+   private final BlockPos pos;
+   private final Direction direction;
+   private final int motive;
 
-   public ClientboundAddPaintingPacket() {
-   }
-
    public ClientboundAddPaintingPacket(Painting var1) {
       this.id = var1.getId();
       this.uuid = var1.getUUID();
       this.pos = var1.getPos();
       this.direction = var1.getDirection();
       this.motive = Registry.MOTIVE.getId(var1.motive);
    }
 
-   public void read(FriendlyByteBuf var1) throws IOException {
+   public ClientboundAddPaintingPacket(FriendlyByteBuf var1) {
       this.id = var1.readVarInt();
       this.uuid = var1.readUUID();
       this.motive = var1.readVarInt();
       this.pos = var1.readBlockPos();
       this.direction = Direction.from2DDataValue(var1.readUnsignedByte());
    }
 
-   public void write(FriendlyByteBuf var1) throws IOException {
+   public void write(FriendlyByteBuf var1) {
       var1.writeVarInt(this.id);
       var1.writeUUID(this.uuid);
       var1.writeVarInt(this.motive);
       var1.writeBlockPos(this.pos);
       var1.writeByte(this.direction.get2DDataValue());
