--- net/minecraft/network/protocol/game/ClientboundEntityEventPacket.java
+++ net/minecraft/network/protocol/game/ClientboundEntityEventPacket.java
@@ -1,8 +1,9 @@
 package net.minecraft.network.protocol.game;
 
 import java.io.IOException;
+import javax.annotation.Nullable;
 import net.minecraft.network.FriendlyByteBuf;
 import net.minecraft.network.protocol.Packet;
 import net.minecraft.world.entity.Entity;
 import net.minecraft.world.level.Level;
 
@@ -30,10 +31,11 @@
 
    public void handle(ClientGamePacketListener var1) {
       var1.handleEntityEvent(this);
    }
 
+   @Nullable
    public Entity getEntity(Level var1) {
       return var1.getEntity(this.entityId);
    }
 
    public byte getEventId() {
