--- net/minecraft/network/protocol/game/ClientboundSetEntityMotionPacket.java
+++ net/minecraft/network/protocol/game/ClientboundSetEntityMotionPacket.java
@@ -1,23 +1,19 @@
 package net.minecraft.network.protocol.game;
 
-import java.io.IOException;
 import net.minecraft.network.FriendlyByteBuf;
 import net.minecraft.network.protocol.Packet;
 import net.minecraft.util.Mth;
 import net.minecraft.world.entity.Entity;
 import net.minecraft.world.phys.Vec3;
 
 public class ClientboundSetEntityMotionPacket implements Packet<ClientGamePacketListener> {
-   private int id;
-   private int xa;
-   private int ya;
-   private int za;
+   private final int id;
+   private final int xa;
+   private final int ya;
+   private final int za;
 
-   public ClientboundSetEntityMotionPacket() {
-   }
-
    public ClientboundSetEntityMotionPacket(Entity var1) {
       this(var1.getId(), var1.getDeltaMovement());
    }
 
    public ClientboundSetEntityMotionPacket(int var1, Vec3 var2) {
@@ -29,18 +25,18 @@
       this.xa = (int)(var5 * 8000.0D);
       this.ya = (int)(var7 * 8000.0D);
       this.za = (int)(var9 * 8000.0D);
    }
 
-   public void read(FriendlyByteBuf var1) throws IOException {
+   public ClientboundSetEntityMotionPacket(FriendlyByteBuf var1) {
       this.id = var1.readVarInt();
       this.xa = var1.readShort();
       this.ya = var1.readShort();
       this.za = var1.readShort();
    }
 
-   public void write(FriendlyByteBuf var1) throws IOException {
+   public void write(FriendlyByteBuf var1) {
       var1.writeVarInt(this.id);
       var1.writeShort(this.xa);
       var1.writeShort(this.ya);
       var1.writeShort(this.za);
    }
