--- net/minecraft/network/protocol/game/ClientboundRemoveMobEffectPacket.java
+++ net/minecraft/network/protocol/game/ClientboundRemoveMobEffectPacket.java
@@ -1,33 +1,29 @@
 package net.minecraft.network.protocol.game;
 
-import java.io.IOException;
 import javax.annotation.Nullable;
 import net.minecraft.network.FriendlyByteBuf;
 import net.minecraft.network.protocol.Packet;
 import net.minecraft.world.effect.MobEffect;
 import net.minecraft.world.entity.Entity;
 import net.minecraft.world.level.Level;
 
 public class ClientboundRemoveMobEffectPacket implements Packet<ClientGamePacketListener> {
-   private int entityId;
-   private MobEffect effect;
+   private final int entityId;
+   private final MobEffect effect;
 
-   public ClientboundRemoveMobEffectPacket() {
-   }
-
    public ClientboundRemoveMobEffectPacket(int var1, MobEffect var2) {
       this.entityId = var1;
       this.effect = var2;
    }
 
-   public void read(FriendlyByteBuf var1) throws IOException {
+   public ClientboundRemoveMobEffectPacket(FriendlyByteBuf var1) {
       this.entityId = var1.readVarInt();
       this.effect = MobEffect.byId(var1.readUnsignedByte());
    }
 
-   public void write(FriendlyByteBuf var1) throws IOException {
+   public void write(FriendlyByteBuf var1) {
       var1.writeVarInt(this.entityId);
       var1.writeByte(MobEffect.getId(this.effect));
    }
 
    public void handle(ClientGamePacketListener var1) {
