--- net/minecraft/network/protocol/game/ClientboundTakeItemEntityPacket.java
+++ net/minecraft/network/protocol/game/ClientboundTakeItemEntityPacket.java
@@ -1,32 +1,28 @@
 package net.minecraft.network.protocol.game;
 
-import java.io.IOException;
 import net.minecraft.network.FriendlyByteBuf;
 import net.minecraft.network.protocol.Packet;
 
 public class ClientboundTakeItemEntityPacket implements Packet<ClientGamePacketListener> {
-   private int itemId;
-   private int playerId;
-   private int amount;
+   private final int itemId;
+   private final int playerId;
+   private final int amount;
 
-   public ClientboundTakeItemEntityPacket() {
-   }
-
    public ClientboundTakeItemEntityPacket(int var1, int var2, int var3) {
       this.itemId = var1;
       this.playerId = var2;
       this.amount = var3;
    }
 
-   public void read(FriendlyByteBuf var1) throws IOException {
+   public ClientboundTakeItemEntityPacket(FriendlyByteBuf var1) {
       this.itemId = var1.readVarInt();
       this.playerId = var1.readVarInt();
       this.amount = var1.readVarInt();
    }
 
-   public void write(FriendlyByteBuf var1) throws IOException {
+   public void write(FriendlyByteBuf var1) {
       var1.writeVarInt(this.itemId);
       var1.writeVarInt(this.playerId);
       var1.writeVarInt(this.amount);
    }
 
