--- net/minecraft/network/protocol/game/ClientboundMoveVehiclePacket.java
+++ net/minecraft/network/protocol/game/ClientboundMoveVehiclePacket.java
@@ -1,39 +1,35 @@
 package net.minecraft.network.protocol.game;
 
-import java.io.IOException;
 import net.minecraft.network.FriendlyByteBuf;
 import net.minecraft.network.protocol.Packet;
 import net.minecraft.world.entity.Entity;
 
 public class ClientboundMoveVehiclePacket implements Packet<ClientGamePacketListener> {
-   private double x;
-   private double y;
-   private double z;
-   private float yRot;
-   private float xRot;
+   private final double x;
+   private final double y;
+   private final double z;
+   private final float yRot;
+   private final float xRot;
 
-   public ClientboundMoveVehiclePacket() {
-   }
-
    public ClientboundMoveVehiclePacket(Entity var1) {
       this.x = var1.getX();
       this.y = var1.getY();
       this.z = var1.getZ();
-      this.yRot = var1.yRot;
-      this.xRot = var1.xRot;
+      this.yRot = var1.getYRot();
+      this.xRot = var1.getXRot();
    }
 
-   public void read(FriendlyByteBuf var1) throws IOException {
+   public ClientboundMoveVehiclePacket(FriendlyByteBuf var1) {
       this.x = var1.readDouble();
       this.y = var1.readDouble();
       this.z = var1.readDouble();
       this.yRot = var1.readFloat();
       this.xRot = var1.readFloat();
    }
 
-   public void write(FriendlyByteBuf var1) throws IOException {
+   public void write(FriendlyByteBuf var1) {
       var1.writeDouble(this.x);
       var1.writeDouble(this.y);
       var1.writeDouble(this.z);
       var1.writeFloat(this.yRot);
       var1.writeFloat(this.xRot);
