--- net/minecraft/network/protocol/game/ClientGamePacketListener.java
+++ net/minecraft/network/protocol/game/ClientGamePacketListener.java
@@ -5,10 +5,12 @@
 public interface ClientGamePacketListener extends PacketListener {
    void handleAddEntity(ClientboundAddEntityPacket var1);
 
    void handleAddExperienceOrb(ClientboundAddExperienceOrbPacket var1);
 
+   void handleAddVibrationSignal(ClientboundAddVibrationSignalPacket var1);
+
    void handleAddMob(ClientboundAddMobPacket var1);
 
    void handleAddObjective(ClientboundSetObjectivePacket var1);
 
    void handleAddPainting(ClientboundAddPaintingPacket var1);
@@ -127,20 +129,32 @@
 
    void handleUpdateMobEffect(ClientboundUpdateMobEffectPacket var1);
 
    void handleUpdateTags(ClientboundUpdateTagsPacket var1);
 
-   void handlePlayerCombat(ClientboundPlayerCombatPacket var1);
+   void handlePlayerCombatEnd(ClientboundPlayerCombatEndPacket var1);
 
+   void handlePlayerCombatEnter(ClientboundPlayerCombatEnterPacket var1);
+
+   void handlePlayerCombatKill(ClientboundPlayerCombatKillPacket var1);
+
    void handleChangeDifficulty(ClientboundChangeDifficultyPacket var1);
 
    void handleSetCamera(ClientboundSetCameraPacket var1);
 
-   void handleSetBorder(ClientboundSetBorderPacket var1);
+   void handleInitializeBorder(ClientboundInitializeBorderPacket var1);
 
-   void handleSetTitles(ClientboundSetTitlesPacket var1);
+   void handleSetBorderLerpSize(ClientboundSetBorderLerpSizePacket var1);
 
+   void handleSetBorderSize(ClientboundSetBorderSizePacket var1);
+
+   void handleSetBorderWarningDelay(ClientboundSetBorderWarningDelayPacket var1);
+
+   void handleSetBorderWarningDistance(ClientboundSetBorderWarningDistancePacket var1);
+
+   void handleSetBorderCenter(ClientboundSetBorderCenterPacket var1);
+
    void handleTabListCustomisation(ClientboundTabListPacket var1);
 
    void handleResourcePack(ClientboundResourcePackPacket var1);
 
    void handleBossUpdate(ClientboundBossEventPacket var1);
@@ -178,6 +192,16 @@
    void handleSetChunkCacheRadius(ClientboundSetChunkCacheRadiusPacket var1);
 
    void handleSetChunkCacheCenter(ClientboundSetChunkCacheCenterPacket var1);
 
    void handleBlockBreakAck(ClientboundBlockBreakAckPacket var1);
+
+   void setActionBarText(ClientboundSetActionBarTextPacket var1);
+
+   void setSubtitleText(ClientboundSetSubtitleTextPacket var1);
+
+   void setTitleText(ClientboundSetTitleTextPacket var1);
+
+   void setTitlesAnimation(ClientboundSetTitlesAnimationPacket var1);
+
+   void handleTitlesClear(ClientboundClearTitlesPacket var1);
 }
