--- net/minecraft/network/protocol/game/ClientboundContainerClosePacket.java
+++ net/minecraft/network/protocol/game/ClientboundContainerClosePacket.java
@@ -1,28 +1,24 @@
 package net.minecraft.network.protocol.game;
 
-import java.io.IOException;
 import net.minecraft.network.FriendlyByteBuf;
 import net.minecraft.network.protocol.Packet;
 
 public class ClientboundContainerClosePacket implements Packet<ClientGamePacketListener> {
-   private int containerId;
+   private final int containerId;
 
-   public ClientboundContainerClosePacket() {
-   }
-
    public ClientboundContainerClosePacket(int var1) {
       this.containerId = var1;
    }
 
-   public void handle(ClientGamePacketListener var1) {
-      var1.handleContainerClose(this);
-   }
-
-   public void read(FriendlyByteBuf var1) throws IOException {
+   public ClientboundContainerClosePacket(FriendlyByteBuf var1) {
       this.containerId = var1.readUnsignedByte();
    }
 
-   public void write(FriendlyByteBuf var1) throws IOException {
+   public void write(FriendlyByteBuf var1) {
       var1.writeByte(this.containerId);
    }
+
+   public void handle(ClientGamePacketListener var1) {
+      var1.handleContainerClose(this);
+   }
 }
