--- net/minecraft/network/protocol/game/ServerboundSetCommandMinecartPacket.java
+++ net/minecraft/network/protocol/game/ServerboundSetCommandMinecartPacket.java
@@ -1,37 +1,33 @@
 package net.minecraft.network.protocol.game;
 
-import java.io.IOException;
 import javax.annotation.Nullable;
 import net.minecraft.network.FriendlyByteBuf;
 import net.minecraft.network.protocol.Packet;
 import net.minecraft.world.entity.Entity;
 import net.minecraft.world.entity.vehicle.MinecartCommandBlock;
 import net.minecraft.world.level.BaseCommandBlock;
 import net.minecraft.world.level.Level;
 
 public class ServerboundSetCommandMinecartPacket implements Packet<ServerGamePacketListener> {
-   private int entity;
-   private String command;
-   private boolean trackOutput;
+   private final int entity;
+   private final String command;
+   private final boolean trackOutput;
 
-   public ServerboundSetCommandMinecartPacket() {
-   }
-
    public ServerboundSetCommandMinecartPacket(int var1, String var2, boolean var3) {
       this.entity = var1;
       this.command = var2;
       this.trackOutput = var3;
    }
 
-   public void read(FriendlyByteBuf var1) throws IOException {
+   public ServerboundSetCommandMinecartPacket(FriendlyByteBuf var1) {
       this.entity = var1.readVarInt();
-      this.command = var1.readUtf(32767);
+      this.command = var1.readUtf();
       this.trackOutput = var1.readBoolean();
    }
 
-   public void write(FriendlyByteBuf var1) throws IOException {
+   public void write(FriendlyByteBuf var1) {
       var1.writeVarInt(this.entity);
       var1.writeUtf(this.command);
       var1.writeBoolean(this.trackOutput);
    }
 
