--- net/minecraft/network/protocol/game/ServerboundLockDifficultyPacket.java
+++ net/minecraft/network/protocol/game/ServerboundLockDifficultyPacket.java
@@ -1,30 +1,26 @@
 package net.minecraft.network.protocol.game;
 
-import java.io.IOException;
 import net.minecraft.network.FriendlyByteBuf;
 import net.minecraft.network.protocol.Packet;
 
 public class ServerboundLockDifficultyPacket implements Packet<ServerGamePacketListener> {
-   private boolean locked;
+   private final boolean locked;
 
-   public ServerboundLockDifficultyPacket() {
-   }
-
    public ServerboundLockDifficultyPacket(boolean var1) {
       this.locked = var1;
    }
 
    public void handle(ServerGamePacketListener var1) {
       var1.handleLockDifficulty(this);
    }
 
-   public void read(FriendlyByteBuf var1) throws IOException {
+   public ServerboundLockDifficultyPacket(FriendlyByteBuf var1) {
       this.locked = var1.readBoolean();
    }
 
-   public void write(FriendlyByteBuf var1) throws IOException {
+   public void write(FriendlyByteBuf var1) {
       var1.writeBoolean(this.locked);
    }
 
    public boolean isLocked() {
       return this.locked;
