--- net/minecraft/network/protocol/game/ServerboundBlockEntityTagQuery.java
+++ net/minecraft/network/protocol/game/ServerboundBlockEntityTagQuery.java
@@ -1,30 +1,26 @@
 package net.minecraft.network.protocol.game;
 
-import java.io.IOException;
 import net.minecraft.core.BlockPos;
 import net.minecraft.network.FriendlyByteBuf;
 import net.minecraft.network.protocol.Packet;
 
 public class ServerboundBlockEntityTagQuery implements Packet<ServerGamePacketListener> {
-   private int transactionId;
-   private BlockPos pos;
+   private final int transactionId;
+   private final BlockPos pos;
 
-   public ServerboundBlockEntityTagQuery() {
-   }
-
    public ServerboundBlockEntityTagQuery(int var1, BlockPos var2) {
       this.transactionId = var1;
       this.pos = var2;
    }
 
-   public void read(FriendlyByteBuf var1) throws IOException {
+   public ServerboundBlockEntityTagQuery(FriendlyByteBuf var1) {
       this.transactionId = var1.readVarInt();
       this.pos = var1.readBlockPos();
    }
 
-   public void write(FriendlyByteBuf var1) throws IOException {
+   public void write(FriendlyByteBuf var1) {
       var1.writeVarInt(this.transactionId);
       var1.writeBlockPos(this.pos);
    }
 
    public void handle(ServerGamePacketListener var1) {
