--- net/minecraft/network/protocol/game/ServerGamePacketListener.java
+++ net/minecraft/network/protocol/game/ServerGamePacketListener.java
@@ -9,12 +9,10 @@
 
    void handleClientCommand(ServerboundClientCommandPacket var1);
 
    void handleClientInformation(ServerboundClientInformationPacket var1);
 
-   void handleContainerAck(ServerboundContainerAckPacket var1);
-
    void handleContainerButtonClick(ServerboundContainerButtonClickPacket var1);
 
    void handleContainerClick(ServerboundContainerClickPacket var1);
 
    void handlePlaceRecipe(ServerboundPlaceRecipePacket var1);
