--- net/minecraft/network/protocol/game/ServerboundJigsawGeneratePacket.java
+++ net/minecraft/network/protocol/game/ServerboundJigsawGeneratePacket.java
@@ -1,33 +1,29 @@
 package net.minecraft.network.protocol.game;
 
-import java.io.IOException;
 import net.minecraft.core.BlockPos;
 import net.minecraft.network.FriendlyByteBuf;
 import net.minecraft.network.protocol.Packet;
 
 public class ServerboundJigsawGeneratePacket implements Packet<ServerGamePacketListener> {
-   private BlockPos pos;
-   private int levels;
-   private boolean keepJigsaws;
+   private final BlockPos pos;
+   private final int levels;
+   private final boolean keepJigsaws;
 
-   public ServerboundJigsawGeneratePacket() {
-   }
-
    public ServerboundJigsawGeneratePacket(BlockPos var1, int var2, boolean var3) {
       this.pos = var1;
       this.levels = var2;
       this.keepJigsaws = var3;
    }
 
-   public void read(FriendlyByteBuf var1) throws IOException {
+   public ServerboundJigsawGeneratePacket(FriendlyByteBuf var1) {
       this.pos = var1.readBlockPos();
       this.levels = var1.readVarInt();
       this.keepJigsaws = var1.readBoolean();
    }
 
-   public void write(FriendlyByteBuf var1) throws IOException {
+   public void write(FriendlyByteBuf var1) {
       var1.writeBlockPos(this.pos);
       var1.writeVarInt(this.levels);
       var1.writeBoolean(this.keepJigsaws);
    }
 
