--- net/minecraft/world/level/timers/FunctionCallback.java
+++ net/minecraft/world/level/timers/FunctionCallback.java
@@ -4,11 +4,11 @@
 import net.minecraft.resources.ResourceLocation;
 import net.minecraft.server.MinecraftServer;
 import net.minecraft.server.ServerFunctionManager;
 
 public class FunctionCallback implements TimerCallback<MinecraftServer> {
-   private final ResourceLocation functionId;
+   final ResourceLocation functionId;
 
    public FunctionCallback(ResourceLocation var1) {
       this.functionId = var1;
    }
 
