--- net/minecraft/world/level/timers/TimerCallbacks.java
+++ net/minecraft/world/level/timers/TimerCallbacks.java
@@ -41,17 +41,17 @@
    @Nullable
    public TimerCallback<C> deserialize(CompoundTag var1) {
       ResourceLocation var2 = ResourceLocation.tryParse(var1.getString("Type"));
       TimerCallback.Serializer var3 = (TimerCallback.Serializer)this.idToSerializer.get(var2);
       if (var3 == null) {
-         LOGGER.error("Failed to deserialize timer callback: " + var1);
+         LOGGER.error((String)"Failed to deserialize timer callback: {}", (Object)var1);
          return null;
       } else {
          try {
             return var3.deserialize(var1);
          } catch (Exception var5) {
-            LOGGER.error((String)("Failed to deserialize timer callback: " + var1), (Throwable)var5);
+            LOGGER.error((String)"Failed to deserialize timer callback: {}", (Object)var1, (Object)var5);
             return null;
          }
       }
    }
 }
