--- net/minecraft/world/entity/projectile/AbstractHurtingProjectile.java
+++ net/minecraft/world/entity/projectile/AbstractHurtingProjectile.java
@@ -46,11 +46,11 @@
    protected void defineSynchedData() {
    }
 
    public void tick() {
       Entity var1 = this.getOwner();
-      if (this.level.isClientSide || (var1 == null || !var1.removed) && this.level.hasChunkAt(this.blockPosition())) {
+      if (this.level.isClientSide || (var1 == null || !var1.isRemoved()) && this.level.hasChunkAt(this.blockPosition())) {
          super.tick();
          if (this.shouldBurn()) {
             this.setSecondsOnFire(1);
          }
 
@@ -77,11 +77,11 @@
 
          this.setDeltaMovement(var3.add(this.xPower, this.yPower, this.zPower).scale((double)var10));
          this.level.addParticle(this.getTrailParticle(), var4, var6 + 0.5D, var8, 0.0D, 0.0D, 0.0D);
          this.setPos(var4, var6, var8);
       } else {
-         this.remove();
+         this.discard();
       }
    }
 
    protected boolean canHitEntity(Entity var1) {
       return super.canHitEntity(var1) && !var1.noPhysics;
