--- net/minecraft/world/entity/projectile/DragonFireball.java
+++ net/minecraft/world/entity/projectile/DragonFireball.java
@@ -53,11 +53,11 @@
                }
             }
 
             this.level.levelEvent(2006, this.blockPosition(), this.isSilent() ? -1 : 1);
             this.level.addFreshEntity(var4);
-            this.remove();
+            this.discard();
          }
 
       }
    }
 
