--- net/minecraft/world/entity/projectile/LargeFireball.java
+++ net/minecraft/world/entity/projectile/LargeFireball.java
@@ -25,11 +25,11 @@
    protected void onHit(HitResult var1) {
       super.onHit(var1);
       if (!this.level.isClientSide) {
          boolean var2 = this.level.getGameRules().getBoolean(GameRules.RULE_MOBGRIEFING);
          this.level.explode((Entity)null, this.getX(), this.getY(), this.getZ(), (float)this.explosionPower, var2, var2 ? Explosion.BlockInteraction.DESTROY : Explosion.BlockInteraction.NONE);
-         this.remove();
+         this.discard();
       }
 
    }
 
    protected void onHitEntity(EntityHitResult var1) {
