--- net/minecraft/world/entity/projectile/LlamaSpit.java
+++ net/minecraft/world/entity/projectile/LlamaSpit.java
@@ -1,9 +1,7 @@
 package net.minecraft.world.entity.projectile;
 
-import net.minecraft.network.protocol.Packet;
-import net.minecraft.network.protocol.game.ClientboundAddEntityPacket;
 import net.minecraft.util.Mth;
 import net.minecraft.world.damagesource.DamageSource;
 import net.minecraft.world.entity.Entity;
 import net.minecraft.world.entity.EntityType;
 import net.minecraft.world.entity.LivingEntity;
@@ -39,13 +37,13 @@
       double var7 = this.getZ() + var1.z;
       this.updateRotation();
       float var9 = 0.99F;
       float var10 = 0.06F;
       if (this.level.getBlockStates(this.getBoundingBox()).noneMatch(BlockBehaviour.BlockStateBase::isAir)) {
-         this.remove();
+         this.discard();
       } else if (this.isInWaterOrBubble()) {
-         this.remove();
+         this.discard();
       } else {
          this.setDeltaMovement(var1.scale(0.9900000095367432D));
          if (!this.isNoGravity()) {
             this.setDeltaMovement(this.getDeltaMovement().add(0.0D, -0.05999999865889549D, 0.0D));
          }
@@ -64,17 +62,13 @@
    }
 
    protected void onHitBlock(BlockHitResult var1) {
       super.onHitBlock(var1);
       if (!this.level.isClientSide) {
-         this.remove();
+         this.discard();
       }
 
    }
 
    protected void defineSynchedData() {
    }
-
-   public Packet<?> getAddEntityPacket() {
-      return new ClientboundAddEntityPacket(this);
-   }
 }
