--- net/minecraft/world/entity/projectile/ThrowableProjectile.java
+++ net/minecraft/world/entity/projectile/ThrowableProjectile.java
@@ -1,11 +1,9 @@
 package net.minecraft.world.entity.projectile;
 
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.particles.ParticleTypes;
-import net.minecraft.network.protocol.Packet;
-import net.minecraft.network.protocol.game.ClientboundAddEntityPacket;
 import net.minecraft.world.entity.EntityType;
 import net.minecraft.world.entity.LivingEntity;
 import net.minecraft.world.level.Level;
 import net.minecraft.world.level.block.Blocks;
 import net.minecraft.world.level.block.entity.BlockEntity;
@@ -28,10 +26,20 @@
    protected ThrowableProjectile(EntityType<? extends ThrowableProjectile> var1, LivingEntity var2, Level var3) {
       this(var1, var2.getX(), var2.getEyeY() - 0.10000000149011612D, var2.getZ(), var3);
       this.setOwner(var2);
    }
 
+   public boolean shouldRenderAtSqrDistance(double var1) {
+      double var3 = this.getBoundingBox().getSize() * 4.0D;
+      if (Double.isNaN(var3)) {
+         var3 = 4.0D;
+      }
+
+      var3 *= 64.0D;
+      return var1 < var3 * var3;
+   }
+
    public void tick() {
       super.tick();
       HitResult var1 = ProjectileUtil.getHitResult(this, this::canHitEntity);
       boolean var2 = false;
       if (var1.getType() == HitResult.Type.BLOCK) {
@@ -41,11 +49,11 @@
             this.handleInsidePortal(var3);
             var2 = true;
          } else if (var4.is(Blocks.END_GATEWAY)) {
             BlockEntity var5 = this.level.getBlockEntity(var3);
             if (var5 instanceof TheEndGatewayBlockEntity && TheEndGatewayBlockEntity.canEntityTeleport(this)) {
-               ((TheEndGatewayBlockEntity)var5).teleportEntity(this);
+               TheEndGatewayBlockEntity.teleportEntity(this.level, var3, var4, this, (TheEndGatewayBlockEntity)var5);
             }
 
             var2 = true;
          }
       }
@@ -82,10 +90,6 @@
    }
 
    protected float getGravity() {
       return 0.03F;
    }
-
-   public Packet<?> getAddEntityPacket() {
-      return new ClientboundAddEntityPacket(this);
-   }
 }
