--- net/minecraft/world/entity/projectile/WitherSkull.java
+++ net/minecraft/world/entity/projectile/WitherSkull.java
@@ -83,11 +83,11 @@
    protected void onHit(HitResult var1) {
       super.onHit(var1);
       if (!this.level.isClientSide) {
          Explosion.BlockInteraction var2 = this.level.getGameRules().getBoolean(GameRules.RULE_MOBGRIEFING) ? Explosion.BlockInteraction.DESTROY : Explosion.BlockInteraction.NONE;
          this.level.explode(this, this.getX(), this.getY(), this.getZ(), 1.0F, false, var2);
-         this.remove();
+         this.discard();
       }
 
    }
 
    public boolean isPickable() {
