--- net/minecraft/world/level/saveddata/SavedData.java
+++ net/minecraft/world/level/saveddata/SavedData.java
@@ -8,19 +8,15 @@
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
 
 public abstract class SavedData {
    private static final Logger LOGGER = LogManager.getLogger();
-   private final String id;
    private boolean dirty;
 
-   public SavedData(String var1) {
-      this.id = var1;
+   public SavedData() {
    }
 
-   public abstract void load(CompoundTag var1);
-
    public abstract CompoundTag save(CompoundTag var1);
 
    public void setDirty() {
       this.setDirty(true);
    }
@@ -31,14 +27,10 @@
 
    public boolean isDirty() {
       return this.dirty;
    }
 
-   public String getId() {
-      return this.id;
-   }
-
    public void save(File var1) {
       if (this.isDirty()) {
          CompoundTag var2 = new CompoundTag();
          var2.put("data", this.save(new CompoundTag()));
          var2.putInt("DataVersion", SharedConstants.getCurrentVersion().getWorldVersion());
