--- net/minecraft/server/packs/resources/FallbackResourceManager.java
+++ net/minecraft/server/packs/resources/FallbackResourceManager.java
@@ -21,11 +21,11 @@
 import net.minecraft.server.packs.PackType;
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
 
 public class FallbackResourceManager implements ResourceManager {
-   private static final Logger LOGGER = LogManager.getLogger();
+   static final Logger LOGGER = LogManager.getLogger();
    protected final List<PackResources> fallbacks = Lists.newArrayList();
    private final PackType type;
    private final String namespace;
 
    public FallbackResourceManager(PackType var1, String var2) {
