--- net/minecraft/server/packs/resources/ReloadableResourceManager.java
+++ net/minecraft/server/packs/resources/ReloadableResourceManager.java
@@ -6,14 +6,14 @@
 import net.minecraft.server.packs.PackResources;
 import net.minecraft.util.Unit;
 
 public interface ReloadableResourceManager extends ResourceManager, AutoCloseable {
    default CompletableFuture<Unit> reload(Executor var1, Executor var2, List<PackResources> var3, CompletableFuture<Unit> var4) {
-      return this.createFullReload(var1, var2, var4, var3).done();
+      return this.createReload(var1, var2, var4, var3).done();
    }
 
-   ReloadInstance createFullReload(Executor var1, Executor var2, CompletableFuture<Unit> var3, List<PackResources> var4);
+   ReloadInstance createReload(Executor var1, Executor var2, CompletableFuture<Unit> var3, List<PackResources> var4);
 
    void registerReloadListener(PreparableReloadListener var1);
 
    void close();
 }
