--- net/minecraft/server/packs/resources/SimpleJsonResourceReloadListener.java
+++ net/minecraft/server/packs/resources/SimpleJsonResourceReloadListener.java
@@ -19,10 +19,11 @@
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
 
 public abstract class SimpleJsonResourceReloadListener extends SimplePreparableReloadListener<Map<ResourceLocation, JsonElement>> {
    private static final Logger LOGGER = LogManager.getLogger();
+   private static final String PATH_SUFFIX = ".json";
    private static final int PATH_SUFFIX_LENGTH = ".json".length();
    private final Gson gson;
    private final String directory;
 
    public SimpleJsonResourceReloadListener(Gson var1, String var2) {
