--- net/minecraft/server/packs/resources/SimplePreparableReloadListener.java
+++ net/minecraft/server/packs/resources/SimplePreparableReloadListener.java
@@ -1,7 +1,8 @@
 package net.minecraft.server.packs.resources;
 
+import java.util.Objects;
 import java.util.concurrent.CompletableFuture;
 import java.util.concurrent.Executor;
 import net.minecraft.util.profiling.ProfilerFiller;
 
 public abstract class SimplePreparableReloadListener<T> implements PreparableReloadListener {
@@ -10,11 +11,11 @@
 
    public final CompletableFuture<Void> reload(PreparableReloadListener.PreparationBarrier var1, ResourceManager var2, ProfilerFiller var3, ProfilerFiller var4, Executor var5, Executor var6) {
       CompletableFuture var10000 = CompletableFuture.supplyAsync(() -> {
          return this.prepare(var2, var3);
       }, var5);
-      var1.getClass();
+      Objects.requireNonNull(var1);
       return var10000.thenCompose(var1::wait).thenAcceptAsync((var3x) -> {
          this.apply(var3x, var2, var4);
       }, var6);
    }
 
