--- net/minecraft/server/packs/resources/SimpleReloadInstance.java
+++ net/minecraft/server/packs/resources/SimpleReloadInstance.java
@@ -70,11 +70,11 @@
             });
          });
          var7.add(var12);
       }
 
-      this.allDone = Util.sequence(var7);
+      this.allDone = Util.sequenceFailFast(var7);
    }
 
    public CompletableFuture<Unit> done() {
       return this.allDone.thenApply((var0) -> {
          return Unit.INSTANCE;
