--- net/minecraft/world/level/levelgen/placement/DecorationContext.java
+++ net/minecraft/world/level/levelgen/placement/DecorationContext.java
@@ -1,18 +1,19 @@
 package net.minecraft.world.level.levelgen.placement;
 
 import java.util.BitSet;
 import net.minecraft.core.BlockPos;
 import net.minecraft.world.level.ChunkPos;
+import net.minecraft.world.level.LevelHeightAccessor;
 import net.minecraft.world.level.WorldGenLevel;
 import net.minecraft.world.level.block.state.BlockState;
 import net.minecraft.world.level.chunk.ChunkGenerator;
 import net.minecraft.world.level.chunk.ProtoChunk;
 import net.minecraft.world.level.levelgen.GenerationStep;
 import net.minecraft.world.level.levelgen.Heightmap;
 
-public class DecorationContext {
+public class DecorationContext implements LevelHeightAccessor {
    private final WorldGenLevel level;
    private final ChunkGenerator generator;
 
    public DecorationContext(WorldGenLevel var1, ChunkGenerator var2) {
       this.level = var1;
@@ -36,6 +37,14 @@
    }
 
    public BlockState getBlockState(BlockPos var1) {
       return this.level.getBlockState(var1);
    }
+
+   public int getMinBuildHeight() {
+      return this.level.getMinBuildHeight();
+   }
+
+   public int getHeight() {
+      return this.level.getHeight();
+   }
 }
