--- net/minecraft/world/level/levelgen/placement/EndGatewayPlacementDecorator.java
+++ net/minecraft/world/level/levelgen/placement/EndGatewayPlacementDecorator.java
@@ -15,11 +15,11 @@
    public Stream<BlockPos> getPositions(DecorationContext var1, Random var2, NoneDecoratorConfiguration var3, BlockPos var4) {
       if (var2.nextInt(700) == 0) {
          int var5 = var2.nextInt(16) + var4.getX();
          int var6 = var2.nextInt(16) + var4.getZ();
          int var7 = var1.getHeight(Heightmap.Types.MOTION_BLOCKING, var5, var6);
-         if (var7 > 0) {
+         if (var7 > var1.getMinBuildHeight()) {
             int var8 = var7 + 3 + var2.nextInt(7);
             return Stream.of(new BlockPos(var5, var8, var6));
          }
       }
 
