--- net/minecraft/world/level/levelgen/placement/Spread32Decorator.java
+++ net/minecraft/world/level/levelgen/placement/Spread32Decorator.java
@@ -10,9 +10,9 @@
    public Spread32Decorator(Codec<NoneDecoratorConfiguration> var1) {
       super(var1);
    }
 
    public Stream<BlockPos> getPositions(DecorationContext var1, Random var2, NoneDecoratorConfiguration var3, BlockPos var4) {
-      int var5 = var2.nextInt(var4.getY() + 32);
+      int var5 = var2.nextInt(Math.max(var4.getY(), 0) + 32);
       return Stream.of(new BlockPos(var4.getX(), var5, var4.getZ()));
    }
 }
