--- net/minecraft/world/level/chunk/ChunkSource.java
+++ net/minecraft/world/level/chunk/ChunkSource.java
@@ -1,11 +1,10 @@
 package net.minecraft.world.level.chunk;
 
 import java.io.IOException;
 import javax.annotation.Nullable;
 import net.minecraft.core.BlockPos;
-import net.minecraft.world.entity.Entity;
 import net.minecraft.world.level.BlockGetter;
 import net.minecraft.world.level.ChunkPos;
 import net.minecraft.world.level.lighting.LevelLightEngine;
 
 public abstract class ChunkSource implements LightChunkGetter, AutoCloseable {
@@ -45,14 +44,10 @@
    }
 
    public void updateChunkForced(ChunkPos var1, boolean var2) {
    }
 
-   public boolean isEntityTickingChunk(Entity var1) {
-      return true;
-   }
-
    public boolean isEntityTickingChunk(ChunkPos var1) {
       return true;
    }
 
    public boolean isTickingChunk(BlockPos var1) {
