--- net/minecraft/world/level/chunk/ChunkStatus.java
+++ net/minecraft/world/level/chunk/ChunkStatus.java
@@ -87,11 +87,11 @@
 
    private static boolean isLighted(ChunkStatus var0, ChunkAccess var1) {
       return var1.getStatus().isOrAfter(var0) && var1.isLightCorrect();
    }
 
-   public static ChunkStatus getStatus(int var0) {
+   public static ChunkStatus getStatusAroundFullChunk(int var0) {
       if (var0 >= STATUS_BY_RANGE.size()) {
          return EMPTY;
       } else {
          return var0 < 0 ? FULL : (ChunkStatus)STATUS_BY_RANGE.get(var0);
       }
