--- net/minecraft/world/level/chunk/GlobalPalette.java
+++ net/minecraft/world/level/chunk/GlobalPalette.java
@@ -26,15 +26,22 @@
    public T valueFor(int var1) {
       Object var2 = this.registry.byId(var1);
       return var2 == null ? this.defaultValue : var2;
    }
 
+   public void read(FriendlyByteBuf var1) {
+   }
+
    public void write(FriendlyByteBuf var1) {
    }
 
    public int getSerializedSize() {
       return FriendlyByteBuf.getVarIntSize(0);
    }
 
+   public int getSize() {
+      return this.registry.size();
+   }
+
    public void read(ListTag var1) {
    }
 }
