--- net/minecraft/world/level/chunk/HashMapPalette.java
+++ net/minecraft/world/level/chunk/HashMapPalette.java
@@ -4,10 +4,11 @@
 import java.util.function.Predicate;
 import javax.annotation.Nullable;
 import net.minecraft.core.IdMapper;
 import net.minecraft.nbt.CompoundTag;
 import net.minecraft.nbt.ListTag;
+import net.minecraft.nbt.Tag;
 import net.minecraft.network.FriendlyByteBuf;
 import net.minecraft.util.CrudeIncrementalIntIdentityHashBiMap;
 
 public class HashMapPalette<T> implements Palette<T> {
    private final IdMapper<T> registry;
@@ -51,10 +52,20 @@
    @Nullable
    public T valueFor(int var1) {
       return this.values.byId(var1);
    }
 
+   public void read(FriendlyByteBuf var1) {
+      this.values.clear();
+      int var2 = var1.readVarInt();
+
+      for(int var3 = 0; var3 < var2; ++var3) {
+         this.values.add(this.registry.byId(var1.readVarInt()));
+      }
+
+   }
+
    public void write(FriendlyByteBuf var1) {
       int var2 = this.getSize();
       var1.writeVarInt(var2);
 
       for(int var3 = 0; var3 < var2; ++var3) {
@@ -86,10 +97,10 @@
 
    }
 
    public void write(ListTag var1) {
       for(int var2 = 0; var2 < this.getSize(); ++var2) {
-         var1.add(this.writer.apply(this.values.byId(var2)));
+         var1.add((Tag)this.writer.apply(this.values.byId(var2)));
       }
 
    }
 }
