--- net/minecraft/world/level/chunk/LinearPalette.java
+++ net/minecraft/world/level/chunk/LinearPalette.java
@@ -16,11 +16,11 @@
    private final int bits;
    private int size;
 
    public LinearPalette(IdMapper<T> var1, int var2, PaletteResize<T> var3, Function<CompoundTag, T> var4) {
       this.registry = var1;
-      this.values = (Object[])(new Object[1 << var2]);
+      this.values = new Object[1 << var2];
       this.bits = var2;
       this.resizeHandler = var3;
       this.reader = var4;
    }
 
@@ -55,10 +55,19 @@
    @Nullable
    public T valueFor(int var1) {
       return var1 >= 0 && var1 < this.size ? this.values[var1] : null;
    }
 
+   public void read(FriendlyByteBuf var1) {
+      this.size = var1.readVarInt();
+
+      for(int var2 = 0; var2 < this.size; ++var2) {
+         this.values[var2] = this.registry.byId(var1.readVarInt());
+      }
+
+   }
+
    public void write(FriendlyByteBuf var1) {
       var1.writeVarInt(this.size);
 
       for(int var2 = 0; var2 < this.size; ++var2) {
          var1.writeVarInt(this.registry.getId(this.values[var2]));
