--- net/minecraft/world/level/chunk/LinearPalette.java
+++ net/minecraft/world/level/chunk/LinearPalette.java
@@ -55,10 +55,19 @@
    @Nullable
    public T valueFor(int var1) {
       return var1 >= 0 && var1 < this.size ? this.values[var1] : null;
    }
 
+   public void read(FriendlyByteBuf var1) {
+      this.size = var1.readVarInt();
+
+      for(int var2 = 0; var2 < this.size; ++var2) {
+         this.values[var2] = this.registry.byId(var1.readVarInt());
+      }
+
+   }
+
    public void write(FriendlyByteBuf var1) {
       var1.writeVarInt(this.size);
 
       for(int var2 = 0; var2 < this.size; ++var2) {
          var1.writeVarInt(this.registry.getId(this.values[var2]));
