--- net/minecraft/client/player/AbstractClientPlayer.java
+++ net/minecraft/client/player/AbstractClientPlayer.java
@@ -7,21 +7,34 @@
 import net.minecraft.client.Minecraft;
 import net.minecraft.client.multiplayer.ClientLevel;
 import net.minecraft.client.multiplayer.PlayerInfo;
 import net.minecraft.client.renderer.texture.AbstractTexture;
 import net.minecraft.client.renderer.texture.HttpTexture;
+import net.minecraft.client.renderer.texture.MissingTextureAtlasSprite;
 import net.minecraft.client.renderer.texture.TextureManager;
 import net.minecraft.client.resources.DefaultPlayerSkin;
 import net.minecraft.resources.ResourceLocation;
 import net.minecraft.util.Mth;
 import net.minecraft.util.StringUtil;
 import net.minecraft.world.entity.ai.attributes.Attributes;
 import net.minecraft.world.entity.player.Player;
+import net.minecraft.world.item.ItemStack;
 import net.minecraft.world.item.Items;
 import net.minecraft.world.level.GameType;
 
 public abstract class AbstractClientPlayer extends Player {
+   private static final String SKIN_URL_TEMPLATE = "http://skins.minecraft.net/MinecraftSkins/%s.png";
+   public static final int SKIN_HEAD_U = 8;
+   public static final int SKIN_HEAD_V = 8;
+   public static final int SKIN_HEAD_WIDTH = 8;
+   public static final int SKIN_HEAD_HEIGHT = 8;
+   public static final int SKIN_HAT_U = 40;
+   public static final int SKIN_HAT_V = 8;
+   public static final int SKIN_HAT_WIDTH = 8;
+   public static final int SKIN_HAT_HEIGHT = 8;
+   public static final int SKIN_TEX_WIDTH = 64;
+   public static final int SKIN_TEX_HEIGHT = 64;
    private PlayerInfo playerInfo;
    public float elytraRotX;
    public float elytraRotY;
    public float elytraRotZ;
    public final ClientLevel clientLevel;
@@ -78,19 +91,18 @@
    public ResourceLocation getElytraTextureLocation() {
       PlayerInfo var1 = this.getPlayerInfo();
       return var1 == null ? null : var1.getElytraLocation();
    }
 
-   public static HttpTexture registerSkinTexture(ResourceLocation var0, String var1) {
+   public static void registerSkinTexture(ResourceLocation var0, String var1) {
       TextureManager var2 = Minecraft.getInstance().getTextureManager();
-      Object var3 = var2.getTexture(var0);
-      if (var3 == null) {
-         var3 = new HttpTexture((File)null, String.format("http://skins.minecraft.net/MinecraftSkins/%s.png", StringUtil.stripColor(var1)), DefaultPlayerSkin.getDefaultSkin(createPlayerUUID(var1)), true, (Runnable)null);
-         var2.register((ResourceLocation)var0, (AbstractTexture)var3);
+      AbstractTexture var3 = var2.getTexture(var0, MissingTextureAtlasSprite.getTexture());
+      if (var3 == MissingTextureAtlasSprite.getTexture()) {
+         HttpTexture var4 = new HttpTexture((File)null, String.format("http://skins.minecraft.net/MinecraftSkins/%s.png", StringUtil.stripColor(var1)), DefaultPlayerSkin.getDefaultSkin(createPlayerUUID(var1)), true, (Runnable)null);
+         var2.register((ResourceLocation)var0, (AbstractTexture)var4);
       }
 
-      return (HttpTexture)var3;
    }
 
    public static ResourceLocation getSkinLocation(String var0) {
       return new ResourceLocation("skins/" + Hashing.sha1().hashUnencodedChars(StringUtil.stripColor(var0)));
    }
@@ -100,29 +112,34 @@
       return var1 == null ? DefaultPlayerSkin.getSkinModelName(this.getUUID()) : var1.getModelName();
    }
 
    public float getFieldOfViewModifier() {
       float var1 = 1.0F;
-      if (this.abilities.flying) {
+      if (this.getAbilities().flying) {
          var1 *= 1.1F;
       }
 
-      var1 = (float)((double)var1 * ((this.getAttributeValue(Attributes.MOVEMENT_SPEED) / (double)this.abilities.getWalkingSpeed() + 1.0D) / 2.0D));
-      if (this.abilities.getWalkingSpeed() == 0.0F || Float.isNaN(var1) || Float.isInfinite(var1)) {
+      var1 = (float)((double)var1 * ((this.getAttributeValue(Attributes.MOVEMENT_SPEED) / (double)this.getAbilities().getWalkingSpeed() + 1.0D) / 2.0D));
+      if (this.getAbilities().getWalkingSpeed() == 0.0F || Float.isNaN(var1) || Float.isInfinite(var1)) {
          var1 = 1.0F;
       }
 
-      if (this.isUsingItem() && this.getUseItem().getItem() == Items.BOW) {
-         int var2 = this.getTicksUsingItem();
-         float var3 = (float)var2 / 20.0F;
-         if (var3 > 1.0F) {
-            var3 = 1.0F;
-         } else {
-            var3 *= var3;
-         }
+      ItemStack var2 = this.getUseItem();
+      if (this.isUsingItem()) {
+         if (var2.is(Items.BOW)) {
+            int var3 = this.getTicksUsingItem();
+            float var4 = (float)var3 / 20.0F;
+            if (var4 > 1.0F) {
+               var4 = 1.0F;
+            } else {
+               var4 *= var4;
+            }
 
-         var1 *= 1.0F - var3 * 0.15F;
+            var1 *= 1.0F - var4 * 0.15F;
+         } else if (Minecraft.getInstance().options.getCameraType().isFirstPerson() && this.isScoping()) {
+            return 0.1F;
+         }
       }
 
       return Mth.lerp(Minecraft.getInstance().options.fovEffectScale, 1.0F, var1);
    }
 }
