--- net/minecraft/client/player/LocalPlayer.java
+++ net/minecraft/client/player/LocalPlayer.java
@@ -59,12 +59,12 @@
 import net.minecraft.world.entity.MoverType;
 import net.minecraft.world.entity.PlayerRideableJumping;
 import net.minecraft.world.entity.Pose;
 import net.minecraft.world.entity.vehicle.AbstractMinecart;
 import net.minecraft.world.entity.vehicle.Boat;
+import net.minecraft.world.inventory.ClickAction;
 import net.minecraft.world.item.ElytraItem;
-import net.minecraft.world.item.Item;
 import net.minecraft.world.item.ItemStack;
 import net.minecraft.world.item.Items;
 import net.minecraft.world.item.crafting.Recipe;
 import net.minecraft.world.level.BaseCommandBlock;
 import net.minecraft.world.level.block.entity.CommandBlockEntity;
@@ -140,11 +140,12 @@
    public boolean startRiding(Entity var1, boolean var2) {
       if (!super.startRiding(var1, var2)) {
          return false;
       } else {
          if (var1 instanceof AbstractMinecart) {
-            this.minecraft.getSoundManager().play(new RidingMinecartSoundInstance(this, (AbstractMinecart)var1));
+            this.minecraft.getSoundManager().play(new RidingMinecartSoundInstance(this, (AbstractMinecart)var1, true));
+            this.minecraft.getSoundManager().play(new RidingMinecartSoundInstance(this, (AbstractMinecart)var1, false));
          }
 
          if (var1 instanceof Boat) {
             this.yRotO = var1.yRot;
             this.yRot = var1.yRot;
@@ -167,11 +168,11 @@
    public float getViewYRot(float var1) {
       return this.isPassenger() ? super.getViewYRot(var1) : this.yRot;
    }
 
    public void tick() {
-      if (this.level.hasChunkAt(new BlockPos(this.getX(), 0.0D, this.getZ()))) {
+      if (this.level.hasChunkAt(this.getBlockX(), this.getBlockZ())) {
          super.tick();
          if (this.isPassenger()) {
             this.connection.send((Packet)(new ServerboundMovePlayerPacket.Rot(this.yRot, this.xRot, this.onGround)));
             this.connection.send((Packet)(new ServerboundPlayerInputPacket(this.xxa, this.zza, this.input.jumping, this.input.shiftKeyDown)));
             Entity var1 = this.getRootVehicle();
@@ -264,11 +265,11 @@
    }
 
    public boolean drop(boolean var1) {
       ServerboundPlayerActionPacket.Action var2 = var1 ? ServerboundPlayerActionPacket.Action.DROP_ALL_ITEMS : ServerboundPlayerActionPacket.Action.DROP_ITEM;
       this.connection.send((Packet)(new ServerboundPlayerActionPacket(var2, BlockPos.ZERO, Direction.DOWN)));
-      return this.inventory.removeItem(this.inventory.selected, var1 && !this.inventory.getSelected().isEmpty() ? this.inventory.getSelected().getCount() : 1) != ItemStack.EMPTY;
+      return this.getInventory().removeItem(this.getInventory().selected, var1 && !this.getInventory().getSelected().isEmpty() ? this.getInventory().getSelected().getCount() : 1) != ItemStack.EMPTY;
    }
 
    public void chat(String var1) {
       this.connection.send((Packet)(new ServerboundChatPacket(var1)));
    }
@@ -292,11 +293,11 @@
       this.connection.send((Packet)(new ServerboundContainerClosePacket(this.containerMenu.containerId)));
       this.clientSideCloseContainer();
    }
 
    public void clientSideCloseContainer() {
-      this.inventory.setCarried(ItemStack.EMPTY);
+      this.getInventory().setCarried(ItemStack.EMPTY);
       super.closeContainer();
       this.minecraft.setScreen((Screen)null);
    }
 
    public void hurtTo(float var1) {
@@ -321,27 +322,27 @@
       }
 
    }
 
    public void onUpdateAbilities() {
-      this.connection.send((Packet)(new ServerboundPlayerAbilitiesPacket(this.abilities)));
+      this.connection.send((Packet)(new ServerboundPlayerAbilitiesPacket(this.getAbilities())));
    }
 
    public boolean isLocalPlayer() {
       return true;
    }
 
    public boolean isSuppressingSlidingDownLadder() {
-      return !this.abilities.flying && super.isSuppressingSlidingDownLadder();
+      return !this.getAbilities().flying && super.isSuppressingSlidingDownLadder();
    }
 
    public boolean canSpawnSprintParticle() {
-      return !this.abilities.flying && super.canSpawnSprintParticle();
+      return !this.getAbilities().flying && super.canSpawnSprintParticle();
    }
 
    public boolean canSpawnSoulSpeedParticle() {
-      return !this.abilities.flying && super.canSpawnSoulSpeedParticle();
+      return !this.getAbilities().flying && super.canSpawnSoulSpeedParticle();
    }
 
    protected void sendRidingJump() {
       this.connection.send((Packet)(new ServerboundPlayerCommandPacket(this, ServerboundPlayerCommandPacket.Action.START_RIDING_JUMP, Mth.floor(this.getJumpRidingScale() * 100.0F))));
    }
@@ -425,11 +426,11 @@
    }
 
    private boolean suffocatesAt(BlockPos var1) {
       AABB var2 = this.getBoundingBox();
       AABB var3 = (new AABB((double)var1.getX(), var2.minY, (double)var1.getZ(), (double)var1.getX() + 1.0D, var2.maxY, (double)var1.getZ() + 1.0D)).deflate(1.0E-7D);
-      return !this.level.noBlockCollision(this, var3, (var1x, var2x) -> {
+      return this.level.hasBlockCollision(this, var3, (var1x, var2x) -> {
          return var1x.isSuffocating(this.level, var2x);
       });
    }
 
    public void setSprinting(boolean var1) {
@@ -544,12 +545,11 @@
    public void openJigsawBlock(JigsawBlockEntity var1) {
       this.minecraft.setScreen(new JigsawBlockEditScreen(var1));
    }
 
    public void openItemGui(ItemStack var1, InteractionHand var2) {
-      Item var3 = var1.getItem();
-      if (var3 == Items.WRITABLE_BOOK) {
+      if (var1.is(Items.WRITABLE_BOOK)) {
          this.minecraft.setScreen(new BookEditScreen(this, var1, var2));
       }
 
    }
 
@@ -589,21 +589,39 @@
 
    protected boolean isControlledCamera() {
       return this.minecraft.getCameraEntity() == this;
    }
 
+   public void resetPos() {
+      this.setPose(Pose.STANDING);
+      if (this.level != null) {
+         for(double var1 = this.getY(); var1 > (double)this.level.getMinBuildHeight() && var1 < (double)this.level.getMaxBuildHeight(); ++var1) {
+            this.setPos(this.getX(), var1, this.getZ());
+            if (this.level.noCollision(this)) {
+               break;
+            }
+         }
+
+         this.setDeltaMovement(Vec3.ZERO);
+         this.xRot = 0.0F;
+      }
+
+      this.setHealth(this.getMaxHealth());
+      this.deathTime = 0;
+   }
+
    public void aiStep() {
       ++this.sprintTime;
       if (this.sprintTriggerTime > 0) {
          --this.sprintTriggerTime;
       }
 
       this.handleNetherPortalClient();
       boolean var1 = this.input.jumping;
       boolean var2 = this.input.shiftKeyDown;
       boolean var3 = this.hasEnoughImpulseToStartSprinting();
-      this.crouching = !this.abilities.flying && !this.isSwimming() && this.canEnterPose(Pose.CROUCHING) && (this.isShiftKeyDown() || !this.isSleeping() && !this.canEnterPose(Pose.STANDING));
+      this.crouching = !this.getAbilities().flying && !this.isSwimming() && this.canEnterPose(Pose.CROUCHING) && (this.isShiftKeyDown() || !this.isSleeping() && !this.canEnterPose(Pose.STANDING));
       this.input.tick(this.isMovingSlowly());
       this.minecraft.getTutorial().onInput(this.input);
       if (this.isUsingItem() && !this.isPassenger()) {
          Input var10000 = this.input;
          var10000.leftImpulse *= 0.2F;
@@ -628,11 +646,11 @@
 
       if (var2) {
          this.sprintTriggerTime = 0;
       }
 
-      boolean var5 = (float)this.getFoodData().getFoodLevel() > 6.0F || this.abilities.mayfly;
+      boolean var5 = (float)this.getFoodData().getFoodLevel() > 6.0F || this.getAbilities().mayfly;
       if ((this.onGround || this.isUnderWater()) && !var2 && !var3 && this.hasEnoughImpulseToStartSprinting() && !this.isSprinting() && var5 && !this.isUsingItem() && !this.hasEffect(MobEffects.BLINDNESS)) {
          if (this.sprintTriggerTime <= 0 && !this.minecraft.options.keySprint.isDown()) {
             this.sprintTriggerTime = 7;
          } else {
             this.setSprinting(true);
@@ -655,32 +673,32 @@
             this.setSprinting(false);
          }
       }
 
       var6 = false;
-      if (this.abilities.mayfly) {
+      if (this.getAbilities().mayfly) {
          if (this.minecraft.gameMode.isAlwaysFlying()) {
-            if (!this.abilities.flying) {
-               this.abilities.flying = true;
+            if (!this.getAbilities().flying) {
+               this.getAbilities().flying = true;
                var6 = true;
                this.onUpdateAbilities();
             }
          } else if (!var1 && this.input.jumping && !var4) {
             if (this.jumpTriggerTime == 0) {
                this.jumpTriggerTime = 7;
             } else if (!this.isSwimming()) {
-               this.abilities.flying = !this.abilities.flying;
+               this.getAbilities().flying = !this.getAbilities().flying;
                var6 = true;
                this.onUpdateAbilities();
                this.jumpTriggerTime = 0;
             }
          }
       }
 
-      if (this.input.jumping && !var6 && !var1 && !this.abilities.flying && !this.isPassenger() && !this.onClimbable()) {
+      if (this.input.jumping && !var6 && !var1 && !this.getAbilities().flying && !this.isPassenger() && !this.onClimbable()) {
          ItemStack var8 = this.getItemBySlot(EquipmentSlot.CHEST);
-         if (var8.getItem() == Items.ELYTRA && ElytraItem.isFlyEnabled(var8) && this.tryToStartFallFlying()) {
+         if (var8.is(Items.ELYTRA) && ElytraItem.isFlyEnabled(var8) && this.tryToStartFallFlying()) {
             this.connection.send((Packet)(new ServerboundPlayerCommandPacket(this, ServerboundPlayerCommandPacket.Action.START_FALL_FLYING)));
          }
       }
 
       this.wasFallFlying = this.isFallFlying();
@@ -695,22 +713,22 @@
       } else if (this.waterVisionTime > 0) {
          this.isEyeInFluid(FluidTags.WATER);
          this.waterVisionTime = Mth.clamp(this.waterVisionTime - 10, 0, 600);
       }
 
-      if (this.abilities.flying && this.isControlledCamera()) {
+      if (this.getAbilities().flying && this.isControlledCamera()) {
          var9 = 0;
          if (this.input.shiftKeyDown) {
             --var9;
          }
 
          if (this.input.jumping) {
             ++var9;
          }
 
          if (var9 != 0) {
-            this.setDeltaMovement(this.getDeltaMovement().add(0.0D, (double)((float)var9 * this.abilities.getFlyingSpeed() * 3.0F), 0.0D));
+            this.setDeltaMovement(this.getDeltaMovement().add(0.0D, (double)((float)var9 * this.getAbilities().getFlyingSpeed() * 3.0F), 0.0D));
          }
       }
 
       if (this.isRidingJumpable()) {
          PlayerRideableJumping var10 = (PlayerRideableJumping)this.getVehicle();
@@ -739,12 +757,12 @@
       } else {
          this.jumpRidingScale = 0.0F;
       }
 
       super.aiStep();
-      if (this.onGround && this.abilities.flying && !this.minecraft.gameMode.isAlwaysFlying()) {
-         this.abilities.flying = false;
+      if (this.onGround && this.getAbilities().flying && !this.minecraft.gameMode.isAlwaysFlying()) {
+         this.getAbilities().flying = false;
          this.onUpdateAbilities();
       }
 
    }
 
@@ -993,6 +1011,10 @@
          return var6.xRot(-var3).yRot(-var2).add(this.getEyePosition(var1));
       } else {
          return super.getRopeHoldPosition(var1);
       }
    }
+
+   public void updateTutorialInventoryAction(ItemStack var1, ItemStack var2, ClickAction var3) {
+      this.minecraft.getTutorial().onInventoryAction(var1, var2, var3);
+   }
 }
