--- net/minecraft/client/player/RemotePlayer.java
+++ net/minecraft/client/player/RemotePlayer.java
@@ -37,15 +37,15 @@
    public void aiStep() {
       if (this.lerpSteps > 0) {
          double var1 = this.getX() + (this.lerpX - this.getX()) / (double)this.lerpSteps;
          double var3 = this.getY() + (this.lerpY - this.getY()) / (double)this.lerpSteps;
          double var5 = this.getZ() + (this.lerpZ - this.getZ()) / (double)this.lerpSteps;
-         this.yRot = (float)((double)this.yRot + Mth.wrapDegrees(this.lerpYRot - (double)this.yRot) / (double)this.lerpSteps);
-         this.xRot = (float)((double)this.xRot + (this.lerpXRot - (double)this.xRot) / (double)this.lerpSteps);
+         this.setYRot(this.getYRot() + (float)Mth.wrapDegrees(this.lerpYRot - (double)this.getYRot()) / (float)this.lerpSteps);
+         this.setXRot(this.getXRot() + (float)(this.lerpXRot - (double)this.getXRot()) / (float)this.lerpSteps);
          --this.lerpSteps;
          this.setPos(var1, var3, var5);
-         this.setRot(this.yRot, this.xRot);
+         this.setRot(this.getYRot(), this.getXRot());
       }
 
       if (this.lerpHeadSteps > 0) {
          this.yHeadRot = (float)((double)this.yHeadRot + Mth.wrapDegrees(this.lyHeadRot - (double)this.yHeadRot) / (double)this.lerpHeadSteps);
          --this.lerpHeadSteps;
@@ -53,22 +53,15 @@
 
       this.oBob = this.bob;
       this.updateSwingTime();
       float var7;
       if (this.onGround && !this.isDeadOrDying()) {
-         var7 = Math.min(0.1F, Mth.sqrt(getHorizontalDistanceSqr(this.getDeltaMovement())));
+         var7 = (float)Math.min(0.1D, this.getDeltaMovement().horizontalDistance());
       } else {
          var7 = 0.0F;
       }
 
-      float var2;
-      if (!this.onGround && !this.isDeadOrDying()) {
-         var2 = (float)Math.atan(-this.getDeltaMovement().y * 0.20000000298023224D) * 15.0F;
-      } else {
-         var2 = 0.0F;
-      }
-
       this.bob += (var7 - this.bob) * 0.4F;
       this.level.getProfiler().push("push");
       this.pushEntities();
       this.level.getProfiler().pop();
    }
