--- net/minecraft/world/level/storage/loot/entries/LootTableReference.java
+++ net/minecraft/world/level/storage/loot/entries/LootTableReference.java
@@ -12,13 +12,13 @@
 import net.minecraft.world.level.storage.loot.ValidationContext;
 import net.minecraft.world.level.storage.loot.functions.LootItemFunction;
 import net.minecraft.world.level.storage.loot.predicates.LootItemCondition;
 
 public class LootTableReference extends LootPoolSingletonContainer {
-   private final ResourceLocation name;
+   final ResourceLocation name;
 
-   private LootTableReference(ResourceLocation var1, int var2, int var3, LootItemCondition[] var4, LootItemFunction[] var5) {
+   LootTableReference(ResourceLocation var1, int var2, int var3, LootItemCondition[] var4, LootItemFunction[] var5) {
       super(var2, var3, var4, var5);
       this.name = var1;
    }
 
    public LootPoolEntryType getType() {
