--- net/minecraft/world/level/storage/loot/entries/LootPoolEntries.java
+++ net/minecraft/world/level/storage/loot/entries/LootPoolEntries.java
@@ -13,10 +13,13 @@
    public static final LootPoolEntryType TAG = register("tag", new TagEntry.Serializer());
    public static final LootPoolEntryType ALTERNATIVES = register("alternatives", CompositeEntryBase.createSerializer(AlternativesEntry::new));
    public static final LootPoolEntryType SEQUENCE = register("sequence", CompositeEntryBase.createSerializer(SequentialEntry::new));
    public static final LootPoolEntryType GROUP = register("group", CompositeEntryBase.createSerializer(EntryGroup::new));
 
+   public LootPoolEntries() {
+   }
+
    private static LootPoolEntryType register(String var0, Serializer<? extends LootPoolEntryContainer> var1) {
       return (LootPoolEntryType)Registry.register(Registry.LOOT_POOL_ENTRY_TYPE, (ResourceLocation)(new ResourceLocation(var0)), new LootPoolEntryType(var1));
    }
 
    public static Object createGsonAdapter() {
