--- net/minecraft/world/phys/shapes/ArrayVoxelShape.java
+++ net/minecraft/world/phys/shapes/ArrayVoxelShape.java
@@ -4,11 +4,11 @@
 import it.unimi.dsi.fastutil.doubles.DoubleList;
 import java.util.Arrays;
 import net.minecraft.Util;
 import net.minecraft.core.Direction;
 
-public final class ArrayVoxelShape extends VoxelShape {
+public class ArrayVoxelShape extends VoxelShape {
    private final DoubleList xs;
    private final DoubleList ys;
    private final DoubleList zs;
 
    protected ArrayVoxelShape(DiscreteVoxelShape var1, double[] var2, double[] var3, double[] var4) {
