--- net/minecraft/world/phys/shapes/VoxelShape.java
+++ net/minecraft/world/phys/shapes/VoxelShape.java
@@ -3,10 +3,11 @@
 import com.google.common.collect.Lists;
 import com.google.common.math.DoubleMath;
 import it.unimi.dsi.fastutil.doubles.DoubleList;
 import java.util.ArrayList;
 import java.util.List;
+import java.util.Optional;
 import javax.annotation.Nullable;
 import net.minecraft.Util;
 import net.minecraft.core.AxisCycle;
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.Direction;
@@ -81,39 +82,47 @@
       return var1;
    }
 
    protected int findIndex(Direction.Axis var1, double var2) {
       return Mth.binarySearch(0, this.shape.getSize(var1) + 1, (var4) -> {
-         if (var4 < 0) {
-            return false;
-         } else if (var4 > this.shape.getSize(var1)) {
-            return true;
-         } else {
-            return var2 < this.get(var1, var4);
-         }
+         return var2 < this.get(var1, var4);
       }) - 1;
    }
 
-   protected boolean isFullWide(double var1, double var3, double var5) {
-      return this.shape.isFullWide(this.findIndex(Direction.Axis.X, var1), this.findIndex(Direction.Axis.Y, var3), this.findIndex(Direction.Axis.Z, var5));
-   }
-
    @Nullable
    public BlockHitResult clip(Vec3 var1, Vec3 var2, BlockPos var3) {
       if (this.isEmpty()) {
          return null;
       } else {
          Vec3 var4 = var2.subtract(var1);
          if (var4.lengthSqr() < 1.0E-7D) {
             return null;
          } else {
             Vec3 var5 = var1.add(var4.scale(0.001D));
-            return this.isFullWide(var5.x - (double)var3.getX(), var5.y - (double)var3.getY(), var5.z - (double)var3.getZ()) ? new BlockHitResult(var5, Direction.getNearest(var4.x, var4.y, var4.z).getOpposite(), var3, true) : AABB.clip(this.toAabbs(), var1, var2, var3);
+            return this.shape.isFullWide(this.findIndex(Direction.Axis.X, var5.x - (double)var3.getX()), this.findIndex(Direction.Axis.Y, var5.y - (double)var3.getY()), this.findIndex(Direction.Axis.Z, var5.z - (double)var3.getZ())) ? new BlockHitResult(var5, Direction.getNearest(var4.x, var4.y, var4.z).getOpposite(), var3, true) : AABB.clip(this.toAabbs(), var1, var2, var3);
          }
       }
    }
 
+   public Optional<Vec3> closestPointTo(Vec3 var1) {
+      if (this.isEmpty()) {
+         return Optional.empty();
+      } else {
+         Vec3[] var2 = new Vec3[1];
+         this.forAllBoxes((var2x, var4, var6, var8, var10, var12) -> {
+            double var14 = Mth.clamp(var1.x(), var2x, var8);
+            double var16 = Mth.clamp(var1.y(), var4, var10);
+            double var18 = Mth.clamp(var1.z(), var6, var12);
+            if (var2[0] == null || var1.distanceToSqr(var14, var16, var18) < var1.distanceToSqr(var2[0])) {
+               var2[0] = new Vec3(var14, var16, var18);
+            }
+
+         });
+         return Optional.of(var2[0]);
+      }
+   }
+
    public VoxelShape getFaceShape(Direction var1) {
       if (!this.isEmpty() && this != Shapes.block()) {
          VoxelShape var2;
          if (this.faces != null) {
             var2 = this.faces[var1.ordinal()];
@@ -132,16 +141,16 @@
       }
    }
 
    private VoxelShape calculateFace(Direction var1) {
       Direction.Axis var2 = var1.getAxis();
-      Direction.AxisDirection var3 = var1.getAxisDirection();
-      DoubleList var4 = this.getCoords(var2);
-      if (var4.size() == 2 && DoubleMath.fuzzyEquals(var4.getDouble(0), 0.0D, 1.0E-7D) && DoubleMath.fuzzyEquals(var4.getDouble(1), 1.0D, 1.0E-7D)) {
+      DoubleList var3 = this.getCoords(var2);
+      if (var3.size() == 2 && DoubleMath.fuzzyEquals(var3.getDouble(0), 0.0D, 1.0E-7D) && DoubleMath.fuzzyEquals(var3.getDouble(1), 1.0D, 1.0E-7D)) {
          return this;
       } else {
-         int var5 = this.findIndex(var2, var3 == Direction.AxisDirection.POSITIVE ? 0.9999999D : 1.0E-7D);
+         Direction.AxisDirection var4 = var1.getAxisDirection();
+         int var5 = this.findIndex(var2, var4 == Direction.AxisDirection.POSITIVE ? 0.9999999D : 1.0E-7D);
          return new SliceShape(this, var2, var5);
       }
    }
 
    public double collide(Direction.Axis var1, AABB var2, double var3) {
