--- net/minecraft/stats/ServerStatsCounter.java
+++ net/minecraft/stats/ServerStatsCounter.java
@@ -35,10 +35,11 @@
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
 
 public class ServerStatsCounter extends StatsCounter {
    private static final Logger LOGGER = LogManager.getLogger();
+   private static final int STAT_REQUEST_COOLDOWN = 300;
    private final MinecraftServer server;
    private final File file;
    private final Set<Stat<?>> dirty = Sets.newHashSet();
    private int lastStatRequest = -300;
 
