--- net/minecraft/stats/Stat.java
+++ net/minecraft/stats/Stat.java
@@ -17,11 +17,12 @@
       this.formatter = var3;
       this.value = var2;
    }
 
    public static <T> String buildName(StatType<T> var0, T var1) {
-      return locationToKey(Registry.STAT_TYPE.getKey(var0)) + ":" + locationToKey(var0.getRegistry().getKey(var1));
+      String var10000 = locationToKey(Registry.STAT_TYPE.getKey(var0));
+      return var10000 + ":" + locationToKey(var0.getRegistry().getKey(var1));
    }
 
    private static <T> String locationToKey(@Nullable ResourceLocation var0) {
       return var0.toString().replace(':', '.');
    }
@@ -45,8 +46,9 @@
    public int hashCode() {
       return this.getName().hashCode();
    }
 
    public String toString() {
-      return "Stat{name=" + this.getName() + ", formatter=" + this.formatter + '}';
+      String var10000 = this.getName();
+      return "Stat{name=" + var10000 + ", formatter=" + this.formatter + "}";
    }
 }
