--- net/minecraft/world/level/levelgen/feature/trunkplacers/DarkOakTrunkPlacer.java
+++ net/minecraft/world/level/levelgen/feature/trunkplacers/DarkOakTrunkPlacer.java
@@ -29,14 +29,14 @@
    }
 
    public List<FoliagePlacer.FoliageAttachment> placeTrunk(LevelSimulatedRW var1, Random var2, int var3, BlockPos var4, Set<BlockPos> var5, BoundingBox var6, TreeConfiguration var7) {
       ArrayList var8 = Lists.newArrayList();
       BlockPos var9 = var4.below();
-      setDirtAt(var1, var9);
-      setDirtAt(var1, var9.east());
-      setDirtAt(var1, var9.south());
-      setDirtAt(var1, var9.south().east());
+      setDirtAt(var1, var2, var9, var7);
+      setDirtAt(var1, var2, var9.east(), var7);
+      setDirtAt(var1, var2, var9.south(), var7);
+      setDirtAt(var1, var2, var9.south().east(), var7);
       Direction var10 = Direction.Plane.HORIZONTAL.getRandomDirection(var2);
       int var11 = var3 - var2.nextInt(4);
       int var12 = 2 - var2.nextInt(3);
       int var13 = var4.getX();
       int var14 = var4.getY();
