--- net/minecraft/world/level/levelgen/feature/trunkplacers/StraightTrunkPlacer.java
+++ net/minecraft/world/level/levelgen/feature/trunkplacers/StraightTrunkPlacer.java
@@ -24,11 +24,11 @@
    protected TrunkPlacerType<?> type() {
       return TrunkPlacerType.STRAIGHT_TRUNK_PLACER;
    }
 
    public List<FoliagePlacer.FoliageAttachment> placeTrunk(LevelSimulatedRW var1, Random var2, int var3, BlockPos var4, Set<BlockPos> var5, BoundingBox var6, TreeConfiguration var7) {
-      setDirtAt(var1, var4.below());
+      setDirtAt(var1, var2, var4.below(), var7);
 
       for(int var8 = 0; var8 < var3; ++var8) {
          placeLog(var1, var2, var4.above(var8), var5, var6, var7);
       }
 
