--- net/minecraft/server/commands/data/BlockDataAccessor.java
+++ net/minecraft/server/commands/data/BlockDataAccessor.java
@@ -10,10 +10,11 @@
 import net.minecraft.commands.Commands;
 import net.minecraft.commands.arguments.NbtPathArgument;
 import net.minecraft.commands.arguments.coordinates.BlockPosArgument;
 import net.minecraft.core.BlockPos;
 import net.minecraft.nbt.CompoundTag;
+import net.minecraft.nbt.NbtUtils;
 import net.minecraft.nbt.Tag;
 import net.minecraft.network.chat.Component;
 import net.minecraft.network.chat.TranslatableComponent;
 import net.minecraft.world.level.block.entity.BlockEntity;
 import net.minecraft.world.level.block.state.BlockState;
@@ -48,11 +49,11 @@
    public void setData(CompoundTag var1) {
       var1.putInt("x", this.pos.getX());
       var1.putInt("y", this.pos.getY());
       var1.putInt("z", this.pos.getZ());
       BlockState var2 = this.entity.getLevel().getBlockState(this.pos);
-      this.entity.load(var2, var1);
+      this.entity.load(var1);
       this.entity.setChanged();
       this.entity.getLevel().sendBlockUpdated(this.pos, var2, var2, 3);
    }
 
    public CompoundTag getData() {
@@ -62,11 +63,11 @@
    public Component getModifiedSuccess() {
       return new TranslatableComponent("commands.data.block.modified", new Object[]{this.pos.getX(), this.pos.getY(), this.pos.getZ()});
    }
 
    public Component getPrintSuccess(Tag var1) {
-      return new TranslatableComponent("commands.data.block.query", new Object[]{this.pos.getX(), this.pos.getY(), this.pos.getZ(), var1.getPrettyDisplay()});
+      return new TranslatableComponent("commands.data.block.query", new Object[]{this.pos.getX(), this.pos.getY(), this.pos.getZ(), NbtUtils.toPrettyComponent(var1)});
    }
 
    public Component getPrintSuccess(NbtPathArgument.NbtPath var1, double var2, int var4) {
       return new TranslatableComponent("commands.data.block.get", new Object[]{var1, this.pos.getX(), this.pos.getY(), this.pos.getZ(), String.format(Locale.ROOT, "%.2f", var2), var4});
    }
