--- net/minecraft/world/level/block/state/pattern/BlockPattern.java
+++ net/minecraft/world/level/block/state/pattern/BlockPattern.java
@@ -155,10 +155,10 @@
       public BlockCacheLoader(LevelReader var1, boolean var2) {
          this.level = var1;
          this.loadChunks = var2;
       }
 
-      public BlockInWorld load(BlockPos var1) throws Exception {
+      public BlockInWorld load(BlockPos var1) {
          return new BlockInWorld(this.level, var1, this.loadChunks);
       }
    }
 }
