--- net/minecraft/world/item/ArmorStandItem.java
+++ net/minecraft/world/item/ArmorStandItem.java
@@ -16,10 +16,11 @@
 import net.minecraft.world.entity.decoration.ArmorStand;
 import net.minecraft.world.entity.player.Player;
 import net.minecraft.world.item.context.BlockPlaceContext;
 import net.minecraft.world.item.context.UseOnContext;
 import net.minecraft.world.level.Level;
+import net.minecraft.world.level.gameevent.GameEvent;
 import net.minecraft.world.phys.AABB;
 import net.minecraft.world.phys.Vec3;
 
 public class ArmorStandItem extends Item {
    public ArmorStandItem(Item.Properties var1) {
@@ -45,16 +46,16 @@
                ArmorStand var10 = (ArmorStand)EntityType.ARMOR_STAND.create(var9, var6.getTag(), (Component)null, var1.getPlayer(), var5, MobSpawnType.SPAWN_EGG, true, true);
                if (var10 == null) {
                   return InteractionResult.FAIL;
                }
 
-               var9.addFreshEntityWithPassengers(var10);
                float var11 = (float)Mth.floor((Mth.wrapDegrees(var1.getRotation() - 180.0F) + 22.5F) / 45.0F) * 45.0F;
                var10.moveTo(var10.getX(), var10.getY(), var10.getZ(), var11, 0.0F);
                this.randomizePose(var10, var3.random);
-               var3.addFreshEntity(var10);
+               var9.addFreshEntityWithPassengers(var10);
                var3.playSound((Player)null, var10.getX(), var10.getY(), var10.getZ(), SoundEvents.ARMOR_STAND_PLACE, SoundSource.BLOCKS, 0.75F, 0.8F);
+               var3.gameEvent(var1.getPlayer(), GameEvent.BLOCK_PLACE, var10);
             }
 
             var6.shrink(1);
             return InteractionResult.sidedSuccess(var3.isClientSide);
          } else {
