--- net/minecraft/world/item/BottleItem.java
+++ net/minecraft/world/item/BottleItem.java
@@ -13,10 +13,11 @@
 import net.minecraft.world.entity.player.Player;
 import net.minecraft.world.item.alchemy.PotionUtils;
 import net.minecraft.world.item.alchemy.Potions;
 import net.minecraft.world.level.ClipContext;
 import net.minecraft.world.level.Level;
+import net.minecraft.world.level.gameevent.GameEvent;
 import net.minecraft.world.phys.BlockHitResult;
 import net.minecraft.world.phys.HitResult;
 
 public class BottleItem extends Item {
    public BottleItem(Item.Properties var1) {
@@ -30,10 +31,11 @@
       ItemStack var5 = var2.getItemInHand(var3);
       if (!var4.isEmpty()) {
          AreaEffectCloud var8 = (AreaEffectCloud)var4.get(0);
          var8.setRadius(var8.getRadius() - 0.5F);
          var1.playSound((Player)null, var2.getX(), var2.getY(), var2.getZ(), SoundEvents.BOTTLE_FILL_DRAGONBREATH, SoundSource.NEUTRAL, 1.0F, 1.0F);
+         var1.gameEvent(var2, GameEvent.FLUID_PICKUP, var2.blockPosition());
          return InteractionResultHolder.sidedSuccess(this.turnBottleIntoItem(var5, var2, new ItemStack(Items.DRAGON_BREATH)), var1.isClientSide());
       } else {
          BlockHitResult var6 = getPlayerPOVHitResult(var1, var2, ClipContext.Fluid.SOURCE_ONLY);
          if (var6.getType() == HitResult.Type.MISS) {
             return InteractionResultHolder.pass(var5);
@@ -44,10 +46,11 @@
                   return InteractionResultHolder.pass(var5);
                }
 
                if (var1.getFluidState(var7).is(FluidTags.WATER)) {
                   var1.playSound(var2, var2.getX(), var2.getY(), var2.getZ(), SoundEvents.BOTTLE_FILL, SoundSource.NEUTRAL, 1.0F, 1.0F);
+                  var1.gameEvent(var2, GameEvent.FLUID_PICKUP, var7);
                   return InteractionResultHolder.sidedSuccess(this.turnBottleIntoItem(var5, var2, PotionUtils.setPotion(new ItemStack(Items.POTION), Potions.WATER)), var1.isClientSide());
                }
             }
 
             return InteractionResultHolder.pass(var5);
