--- net/minecraft/world/item/BowlFoodItem.java
+++ net/minecraft/world/item/BowlFoodItem.java
@@ -9,8 +9,8 @@
       super(var1);
    }
 
    public ItemStack finishUsingItem(ItemStack var1, Level var2, LivingEntity var3) {
       ItemStack var4 = super.finishUsingItem(var1, var2, var3);
-      return var3 instanceof Player && ((Player)var3).abilities.instabuild ? var4 : new ItemStack(Items.BOWL);
+      return var3 instanceof Player && ((Player)var3).getAbilities().instabuild ? var4 : new ItemStack(Items.BOWL);
    }
 }
