--- net/minecraft/world/item/DyeColor.java
+++ net/minecraft/world/item/DyeColor.java
@@ -36,26 +36,22 @@
       return var0;
    })));
    private final int id;
    private final String name;
    private final MaterialColor color;
-   private final int textureDiffuseColor;
-   private final int textureDiffuseColorBGR;
    private final float[] textureDiffuseColors;
    private final int fireworkColor;
    private final int textColor;
 
    private DyeColor(int var3, String var4, int var5, MaterialColor var6, int var7, int var8) {
       this.id = var3;
       this.name = var4;
-      this.textureDiffuseColor = var5;
       this.color = var6;
       this.textColor = var8;
       int var9 = (var5 & 16711680) >> 16;
       int var10 = (var5 & '\uff00') >> 8;
       int var11 = (var5 & 255) >> 0;
-      this.textureDiffuseColorBGR = var11 << 16 | var10 << 8 | var9 << 0;
       this.textureDiffuseColors = new float[]{(float)var9 / 255.0F, (float)var10 / 255.0F, (float)var11 / 255.0F};
       this.fireworkColor = var7;
    }
 
    public int getId() {
