--- net/minecraft/world/item/EnchantedBookItem.java
+++ net/minecraft/world/item/EnchantedBookItem.java
@@ -12,10 +12,12 @@
 import net.minecraft.world.item.enchantment.Enchantment;
 import net.minecraft.world.item.enchantment.EnchantmentInstance;
 import net.minecraft.world.level.Level;
 
 public class EnchantedBookItem extends Item {
+   public static final String TAG_STORED_ENCHANTMENTS = "StoredEnchantments";
+
    public EnchantedBookItem(Item.Properties var1) {
       super(var1);
    }
 
    public boolean isFoil(ItemStack var1) {
