--- net/minecraft/world/item/HoneyBottleItem.java
+++ net/minecraft/world/item/HoneyBottleItem.java
@@ -11,10 +11,12 @@
 import net.minecraft.world.entity.LivingEntity;
 import net.minecraft.world.entity.player.Player;
 import net.minecraft.world.level.Level;
 
 public class HoneyBottleItem extends Item {
+   private static final int DRINK_DURATION = 40;
+
    public HoneyBottleItem(Item.Properties var1) {
       super(var1);
    }
 
    public ItemStack finishUsingItem(ItemStack var1, Level var2, LivingEntity var3) {
@@ -30,14 +32,14 @@
       }
 
       if (var1.isEmpty()) {
          return new ItemStack(Items.GLASS_BOTTLE);
       } else {
-         if (var3 instanceof Player && !((Player)var3).abilities.instabuild) {
+         if (var3 instanceof Player && !((Player)var3).getAbilities().instabuild) {
             ItemStack var6 = new ItemStack(Items.GLASS_BOTTLE);
             Player var5 = (Player)var3;
-            if (!var5.inventory.add(var6)) {
+            if (!var5.getInventory().add(var6)) {
                var5.drop(var6, false);
             }
          }
 
          return var1;
@@ -59,8 +61,8 @@
    public SoundEvent getEatingSound() {
       return SoundEvents.HONEY_DRINK;
    }
 
    public InteractionResultHolder<ItemStack> use(Level var1, Player var2, InteractionHand var3) {
-      return ItemUtils.useDrink(var1, var2, var3);
+      return ItemUtils.startUsingInstantly(var1, var2, var3);
    }
 }
