--- net/minecraft/world/item/HorseArmorItem.java
+++ net/minecraft/world/item/HorseArmorItem.java
@@ -1,10 +1,11 @@
 package net.minecraft.world.item;
 
 import net.minecraft.resources.ResourceLocation;
 
 public class HorseArmorItem extends Item {
+   private static final String TEX_FOLDER = "textures/entity/horse/";
    private final int protection;
    private final String texture;
 
    public HorseArmorItem(int var1, String var2, Item.Properties var3) {
       super(var3);
