--- net/minecraft/world/item/ItemCooldowns.java
+++ net/minecraft/world/item/ItemCooldowns.java
@@ -58,15 +58,15 @@
    }
 
    protected void onCooldownEnded(Item var1) {
    }
 
-   class CooldownInstance {
-      private final int startTime;
-      private final int endTime;
+   private class CooldownInstance {
+      final int startTime;
+      final int endTime;
 
-      private CooldownInstance(int var2, int var3) {
+      CooldownInstance(int var2, int var3) {
          this.startTime = var2;
          this.endTime = var3;
       }
    }
 }
