--- net/minecraft/world/item/ItemUtils.java
+++ net/minecraft/world/item/ItemUtils.java
@@ -4,20 +4,20 @@
 import net.minecraft.world.InteractionResultHolder;
 import net.minecraft.world.entity.player.Player;
 import net.minecraft.world.level.Level;
 
 public class ItemUtils {
-   public static InteractionResultHolder<ItemStack> useDrink(Level var0, Player var1, InteractionHand var2) {
+   public static InteractionResultHolder<ItemStack> startUsingInstantly(Level var0, Player var1, InteractionHand var2) {
       var1.startUsingItem(var2);
       return InteractionResultHolder.consume(var1.getItemInHand(var2));
    }
 
    public static ItemStack createFilledResult(ItemStack var0, Player var1, ItemStack var2, boolean var3) {
-      boolean var4 = var1.abilities.instabuild;
+      boolean var4 = var1.getAbilities().instabuild;
       if (var3 && var4) {
-         if (!var1.inventory.contains(var2)) {
-            var1.inventory.add(var2);
+         if (!var1.getInventory().contains(var2)) {
+            var1.getInventory().add(var2);
          }
 
          return var0;
       } else {
          if (!var4) {
@@ -25,11 +25,11 @@
          }
 
          if (var0.isEmpty()) {
             return var2;
          } else {
-            if (!var1.inventory.add(var2)) {
+            if (!var1.getInventory().add(var2)) {
                var1.drop(var2, false);
             }
 
             return var0;
          }
