--- net/minecraft/world/item/KnowledgeBookItem.java
+++ net/minecraft/world/item/KnowledgeBookItem.java
@@ -14,20 +14,21 @@
 import net.minecraft.world.level.Level;
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
 
 public class KnowledgeBookItem extends Item {
+   private static final String RECIPE_TAG = "Recipes";
    private static final Logger LOGGER = LogManager.getLogger();
 
    public KnowledgeBookItem(Item.Properties var1) {
       super(var1);
    }
 
    public InteractionResultHolder<ItemStack> use(Level var1, Player var2, InteractionHand var3) {
       ItemStack var4 = var2.getItemInHand(var3);
       CompoundTag var5 = var4.getTag();
-      if (!var2.abilities.instabuild) {
+      if (!var2.getAbilities().instabuild) {
          var2.setItemInHand(var3, ItemStack.EMPTY);
       }
 
       if (var5 != null && var5.contains("Recipes", 9)) {
          if (!var1.isClientSide) {
