--- net/minecraft/world/item/MilkBucketItem.java
+++ net/minecraft/world/item/MilkBucketItem.java
@@ -19,11 +19,11 @@
          ServerPlayer var4 = (ServerPlayer)var3;
          CriteriaTriggers.CONSUME_ITEM.trigger(var4, var1);
          var4.awardStat(Stats.ITEM_USED.get(this));
       }
 
-      if (var3 instanceof Player && !((Player)var3).abilities.instabuild) {
+      if (var3 instanceof Player && !((Player)var3).getAbilities().instabuild) {
          var1.shrink(1);
       }
 
       if (!var2.isClientSide) {
          var3.removeAllEffects();
@@ -39,8 +39,8 @@
    public UseAnim getUseAnimation(ItemStack var1) {
       return UseAnim.DRINK;
    }
 
    public InteractionResultHolder<ItemStack> use(Level var1, Player var2, InteractionHand var3) {
-      return ItemUtils.useDrink(var1, var2, var3);
+      return ItemUtils.startUsingInstantly(var1, var2, var3);
    }
 }
