--- net/minecraft/world/item/MinecartItem.java
+++ net/minecraft/world/item/MinecartItem.java
@@ -13,10 +13,11 @@
 import net.minecraft.world.level.Level;
 import net.minecraft.world.level.block.BaseRailBlock;
 import net.minecraft.world.level.block.DispenserBlock;
 import net.minecraft.world.level.block.state.BlockState;
 import net.minecraft.world.level.block.state.properties.RailShape;
+import net.minecraft.world.level.gameevent.GameEvent;
 
 public class MinecartItem extends Item {
    private static final DispenseItemBehavior DISPENSE_ITEM_BEHAVIOR = new DefaultDispenseItemBehavior() {
       private final DefaultDispenseItemBehavior defaultDispenseItemBehavior = new DefaultDispenseItemBehavior();
 
@@ -62,11 +63,11 @@
 
       protected void playSound(BlockSource var1) {
          var1.getLevel().levelEvent(1000, var1.getPos(), 0);
       }
    };
-   private final AbstractMinecart.Type type;
+   final AbstractMinecart.Type type;
 
    public MinecartItem(AbstractMinecart.Type var1, Item.Properties var2) {
       super(var2);
       this.type = var1;
       DispenserBlock.registerBehavior(this, DISPENSE_ITEM_BEHAVIOR);
@@ -91,10 +92,11 @@
             if (var5.hasCustomHoverName()) {
                var9.setCustomName(var5.getHoverName());
             }
 
             var2.addFreshEntity(var9);
+            var2.gameEvent(var1.getPlayer(), GameEvent.ENTITY_PLACE, var3);
          }
 
          var5.shrink(1);
          return InteractionResult.sidedSuccess(var2.isClientSide);
       }
