--- net/minecraft/world/item/PotionItem.java
+++ net/minecraft/world/item/PotionItem.java
@@ -16,12 +16,15 @@
 import net.minecraft.world.entity.player.Player;
 import net.minecraft.world.item.alchemy.Potion;
 import net.minecraft.world.item.alchemy.PotionUtils;
 import net.minecraft.world.item.alchemy.Potions;
 import net.minecraft.world.level.Level;
+import net.minecraft.world.level.gameevent.GameEvent;
 
 public class PotionItem extends Item {
+   private static final int DRINK_DURATION = 32;
+
    public PotionItem(Item.Properties var1) {
       super(var1);
    }
 
    public ItemStack getDefaultInstance() {
@@ -48,25 +51,26 @@
          }
       }
 
       if (var4 != null) {
          var4.awardStat(Stats.ITEM_USED.get(this));
-         if (!var4.abilities.instabuild) {
+         if (!var4.getAbilities().instabuild) {
             var1.shrink(1);
          }
       }
 
-      if (var4 == null || !var4.abilities.instabuild) {
+      if (var4 == null || !var4.getAbilities().instabuild) {
          if (var1.isEmpty()) {
             return new ItemStack(Items.GLASS_BOTTLE);
          }
 
          if (var4 != null) {
-            var4.inventory.add(new ItemStack(Items.GLASS_BOTTLE));
+            var4.getInventory().add(new ItemStack(Items.GLASS_BOTTLE));
          }
       }
 
+      var2.gameEvent(var3, GameEvent.DRINKING_FINISH, var3.eyeBlockPosition());
       return var1;
    }
 
    public int getUseDuration(ItemStack var1) {
       return 32;
@@ -75,11 +79,11 @@
    public UseAnim getUseAnimation(ItemStack var1) {
       return UseAnim.DRINK;
    }
 
    public InteractionResultHolder<ItemStack> use(Level var1, Player var2, InteractionHand var3) {
-      return ItemUtils.useDrink(var1, var2, var3);
+      return ItemUtils.startUsingInstantly(var1, var2, var3);
    }
 
    public String getDescriptionId(ItemStack var1) {
       return PotionUtils.getPotion(var1).getName(this.getDescriptionId() + ".effect.");
    }
