--- net/minecraft/world/item/SnowballItem.java
+++ net/minecraft/world/item/SnowballItem.java
@@ -14,20 +14,20 @@
       super(var1);
    }
 
    public InteractionResultHolder<ItemStack> use(Level var1, Player var2, InteractionHand var3) {
       ItemStack var4 = var2.getItemInHand(var3);
-      var1.playSound((Player)null, var2.getX(), var2.getY(), var2.getZ(), SoundEvents.SNOWBALL_THROW, SoundSource.NEUTRAL, 0.5F, 0.4F / (random.nextFloat() * 0.4F + 0.8F));
+      var1.playSound((Player)null, var2.getX(), var2.getY(), var2.getZ(), SoundEvents.SNOWBALL_THROW, SoundSource.NEUTRAL, 0.5F, 0.4F / (var1.getRandom().nextFloat() * 0.4F + 0.8F));
       if (!var1.isClientSide) {
          Snowball var5 = new Snowball(var1, var2);
          var5.setItem(var4);
-         var5.shootFromRotation(var2, var2.xRot, var2.yRot, 0.0F, 1.5F, 1.0F);
+         var5.shootFromRotation(var2, var2.getXRot(), var2.getYRot(), 0.0F, 1.5F, 1.0F);
          var1.addFreshEntity(var5);
       }
 
       var2.awardStat(Stats.ITEM_USED.get(this));
-      if (!var2.abilities.instabuild) {
+      if (!var2.getAbilities().instabuild) {
          var4.shrink(1);
       }
 
       return InteractionResultHolder.sidedSuccess(var4, var1.isClientSide());
    }
