--- net/minecraft/world/item/SplashPotionItem.java
+++ net/minecraft/world/item/SplashPotionItem.java
@@ -11,9 +11,9 @@
    public SplashPotionItem(Item.Properties var1) {
       super(var1);
    }
 
    public InteractionResultHolder<ItemStack> use(Level var1, Player var2, InteractionHand var3) {
-      var1.playSound((Player)null, var2.getX(), var2.getY(), var2.getZ(), SoundEvents.SPLASH_POTION_THROW, SoundSource.PLAYERS, 0.5F, 0.4F / (random.nextFloat() * 0.4F + 0.8F));
+      var1.playSound((Player)null, var2.getX(), var2.getY(), var2.getZ(), SoundEvents.SPLASH_POTION_THROW, SoundSource.PLAYERS, 0.5F, 0.4F / (var1.getRandom().nextFloat() * 0.4F + 0.8F));
       return super.use(var1, var2, var3);
    }
 }
