--- net/minecraft/world/level/levelgen/DebugLevelSource.java
+++ net/minecraft/world/level/levelgen/DebugLevelSource.java
@@ -4,14 +4,14 @@
 import java.util.List;
 import java.util.stream.Collectors;
 import java.util.stream.StreamSupport;
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.Registry;
+import net.minecraft.core.SectionPos;
 import net.minecraft.resources.RegistryLookupCodec;
 import net.minecraft.server.level.WorldGenRegion;
 import net.minecraft.util.Mth;
-import net.minecraft.world.level.BlockGetter;
 import net.minecraft.world.level.LevelAccessor;
 import net.minecraft.world.level.NoiseColumn;
 import net.minecraft.world.level.StructureFeatureManager;
 import net.minecraft.world.level.biome.Biome;
 import net.minecraft.world.level.biome.BiomeManager;
@@ -55,12 +55,12 @@
       int var4 = var1.getCenterX();
       int var5 = var1.getCenterZ();
 
       for(int var6 = 0; var6 < 16; ++var6) {
          for(int var7 = 0; var7 < 16; ++var7) {
-            int var8 = (var4 << 4) + var6;
-            int var9 = (var5 << 4) + var7;
+            int var8 = SectionPos.sectionToBlockCoord(var4, var6);
+            int var9 = SectionPos.sectionToBlockCoord(var5, var7);
             var1.setBlock(var3.set(var8, 60, var9), BARRIER, 2);
             BlockState var10 = getBlockStateFor(var8, var9);
             if (var10 != null) {
                var1.setBlock(var3.set(var8, 70, var9), var10, 2);
             }
@@ -74,12 +74,12 @@
 
    public int getBaseHeight(int var1, int var2, Heightmap.Types var3) {
       return 0;
    }
 
-   public BlockGetter getBaseColumn(int var1, int var2) {
-      return new NoiseColumn(new BlockState[0]);
+   public NoiseColumn getBaseColumn(int var1, int var2) {
+      return new NoiseColumn(0, new BlockState[0]);
    }
 
    public static BlockState getBlockStateFor(int var0, int var1) {
       BlockState var2 = AIR;
       if (var0 > 0 && var1 > 0 && var0 % 2 != 0 && var1 % 2 != 0) {
