--- net/minecraft/world/level/levelgen/Decoratable.java
+++ net/minecraft/world/level/levelgen/Decoratable.java
@@ -9,11 +9,11 @@
 import net.minecraft.world.level.levelgen.placement.FeatureDecorator;
 
 public interface Decoratable<R> {
    R decorated(ConfiguredDecorator<?> var1);
 
-   default R chance(int var1) {
+   default R rarity(int var1) {
       return this.decorated(FeatureDecorator.CHANCE.configured(new ChanceDecoratorConfiguration(var1)));
    }
 
    default R count(UniformInt var1) {
       return this.decorated(FeatureDecorator.COUNT.configured(new CountConfiguration(var1)));
