--- net/minecraft/world/level/levelgen/Heightmap.java
+++ net/minecraft/world/level/levelgen/Heightmap.java
@@ -23,17 +23,19 @@
       return !var0.isAir();
    };
    private static final Predicate<BlockState> MATERIAL_MOTION_BLOCKING = (var0) -> {
       return var0.getMaterial().blocksMotion();
    };
-   private final BitStorage data = new BitStorage(9, 256);
+   private final BitStorage data;
    private final Predicate<BlockState> isOpaque;
    private final ChunkAccess chunk;
 
    public Heightmap(ChunkAccess var1, Heightmap.Types var2) {
       this.isOpaque = var2.isOpaque();
       this.chunk = var1;
+      int var3 = (int)Math.ceil(Math.log((double)(var1.getHeight() + 1)) / Math.log(2.0D));
+      this.data = new BitStorage(var3, 256);
    }
 
    public static void primeHeightmaps(ChunkAccess var0, Set<Heightmap.Types> var1) {
       int var2 = var1.size();
       ObjectArrayList var3 = new ObjectArrayList(var2);
@@ -48,11 +50,11 @@
             while(var9.hasNext()) {
                Heightmap.Types var10 = (Heightmap.Types)var9.next();
                var3.add(var0.getOrCreateHeightmapUnprimed(var10));
             }
 
-            for(int var12 = var5 - 1; var12 >= 0; --var12) {
+            for(int var12 = var5 - 1; var12 >= var0.getMinBuildHeight(); --var12) {
                var6.set(var7, var12, var8);
                BlockState var13 = var0.getBlockState(var6);
                if (!var13.is(Blocks.AIR)) {
                   while(var4.hasNext()) {
                      Heightmap var11 = (Heightmap)var4.next();
@@ -85,19 +87,19 @@
                return true;
             }
          } else if (var5 - 1 == var2) {
             BlockPos.MutableBlockPos var6 = new BlockPos.MutableBlockPos();
 
-            for(int var7 = var2 - 1; var7 >= 0; --var7) {
+            for(int var7 = var2 - 1; var7 >= this.chunk.getMinBuildHeight(); --var7) {
                var6.set(var1, var7, var3);
                if (this.isOpaque.test(this.chunk.getBlockState(var6))) {
                   this.setHeight(var1, var3, var7 + 1);
                   return true;
                }
             }
 
-            this.setHeight(var1, var3, 0);
+            this.setHeight(var1, var3, this.chunk.getMinBuildHeight());
             return true;
          }
 
          return false;
       }
@@ -106,15 +108,15 @@
    public int getFirstAvailable(int var1, int var2) {
       return this.getFirstAvailable(getIndex(var1, var2));
    }
 
    private int getFirstAvailable(int var1) {
-      return this.data.get(var1);
+      return this.data.get(var1) + this.chunk.getMinBuildHeight();
    }
 
    private void setHeight(int var1, int var2, int var3) {
-      this.data.set(getIndex(var1, var2), var3);
+      this.data.set(getIndex(var1, var2), var3 - this.chunk.getMinBuildHeight());
    }
 
    public void setRawData(long[] var1) {
       System.arraycopy(var1, 0, this.data.getRaw(), 0, var1.length);
    }
