--- net/minecraft/world/level/levelgen/PatrolSpawner.java
+++ net/minecraft/world/level/levelgen/PatrolSpawner.java
@@ -49,25 +49,26 @@
                         return 0;
                      } else {
                         int var9 = (24 + var4.nextInt(24)) * (var4.nextBoolean() ? -1 : 1);
                         int var10 = (24 + var4.nextInt(24)) * (var4.nextBoolean() ? -1 : 1);
                         BlockPos.MutableBlockPos var11 = var8.blockPosition().mutable().move(var9, 0, var10);
-                        if (!var1.hasChunksAt(var11.getX() - 10, var11.getY() - 10, var11.getZ() - 10, var11.getX() + 10, var11.getY() + 10, var11.getZ() + 10)) {
+                        boolean var12 = true;
+                        if (!var1.hasChunksAt(var11.getX() - 10, var11.getZ() - 10, var11.getX() + 10, var11.getZ() + 10)) {
                            return 0;
                         } else {
-                           Biome var12 = var1.getBiome(var11);
-                           Biome.BiomeCategory var13 = var12.getBiomeCategory();
-                           if (var13 == Biome.BiomeCategory.MUSHROOM) {
+                           Biome var13 = var1.getBiome(var11);
+                           Biome.BiomeCategory var14 = var13.getBiomeCategory();
+                           if (var14 == Biome.BiomeCategory.MUSHROOM) {
                               return 0;
                            } else {
-                              int var14 = 0;
-                              int var15 = (int)Math.ceil((double)var1.getCurrentDifficultyAt(var11).getEffectiveDifficulty()) + 1;
+                              int var15 = 0;
+                              int var16 = (int)Math.ceil((double)var1.getCurrentDifficultyAt(var11).getEffectiveDifficulty()) + 1;
 
-                              for(int var16 = 0; var16 < var15; ++var16) {
-                                 ++var14;
+                              for(int var17 = 0; var17 < var16; ++var17) {
+                                 ++var15;
                                  var11.setY(var1.getHeightmapPos(Heightmap.Types.MOTION_BLOCKING_NO_LEAVES, var11).getY());
-                                 if (var16 == 0) {
+                                 if (var17 == 0) {
                                     if (!this.spawnPatrolMember(var1, var11, var4, true)) {
                                        break;
                                     }
                                  } else {
                                     this.spawnPatrolMember(var1, var11, var4, false);
@@ -75,11 +76,11 @@
 
                                  var11.setX(var11.getX() + var4.nextInt(5) - var4.nextInt(5));
                                  var11.setZ(var11.getZ() + var4.nextInt(5) - var4.nextInt(5));
                               }
 
-                              return var14;
+                              return var15;
                            }
                         }
                      }
                   }
                }
