--- net/minecraft/world/level/levelgen/PhantomSpawner.java
+++ net/minecraft/world/level/levelgen/PhantomSpawner.java
@@ -70,11 +70,11 @@
 
                            var9 = var1.getCurrentDifficultyAt(var8);
                         } while(!var9.isHarderThan(var4.nextFloat() * 3.0F));
 
                         ServerStatsCounter var10 = ((ServerPlayer)var7).getStats();
-                        var11 = Mth.clamp(var10.getValue(Stats.CUSTOM.get(Stats.TIME_SINCE_REST)), 1, Integer.MAX_VALUE);
+                        var11 = Mth.clamp((int)var10.getValue(Stats.CUSTOM.get(Stats.TIME_SINCE_REST)), (int)1, (int)Integer.MAX_VALUE);
                         boolean var12 = true;
                      } while(var4.nextInt(var11) < 72000);
 
                      var13 = var8.above(20 + var4.nextInt(15)).east(-10 + var4.nextInt(21)).south(-10 + var4.nextInt(21));
                      var14 = var1.getBlockState(var13);
