--- net/minecraft/world/level/levelgen/WorldgenRandom.java
+++ net/minecraft/world/level/levelgen/WorldgenRandom.java
@@ -1,12 +1,10 @@
 package net.minecraft.world.level.levelgen;
 
 import java.util.Random;
 
 public class WorldgenRandom extends Random {
-   private int count;
-
    public WorldgenRandom() {
    }
 
    public WorldgenRandom(long var1) {
       super(var1);
@@ -17,15 +15,10 @@
          this.next(1);
       }
 
    }
 
-   protected int next(int var1) {
-      ++this.count;
-      return super.next(var1);
-   }
-
    public long setBaseChunkSeed(int var1, int var2) {
       long var3 = (long)var1 * 341873128712L + (long)var2 * 132897987541L;
       this.setSeed(var3);
       return var3;
    }
