--- net/minecraft/world/entity/ai/targeting/TargetingConditions.java
+++ net/minecraft/world/entity/ai/targeting/TargetingConditions.java
@@ -16,10 +16,22 @@
    private Predicate<LivingEntity> selector;
 
    public TargetingConditions() {
    }
 
+   public TargetingConditions copy() {
+      TargetingConditions var1 = new TargetingConditions();
+      var1.range = this.range;
+      var1.allowInvulnerable = this.allowInvulnerable;
+      var1.allowSameTeam = this.allowSameTeam;
+      var1.allowUnseeable = this.allowUnseeable;
+      var1.allowNonAttackable = this.allowNonAttackable;
+      var1.testInvisible = this.testInvisible;
+      var1.selector = this.selector;
+      return var1;
+   }
+
    public TargetingConditions range(double var1) {
       this.range = var1;
       return this;
    }
 
