--- net/minecraft/client/renderer/entity/layers/Deadmau5EarsLayer.java
+++ net/minecraft/client/renderer/entity/layers/Deadmau5EarsLayer.java
@@ -20,12 +20,12 @@
       if ("deadmau5".equals(var4.getName().getString()) && var4.isSkinLoaded() && !var4.isInvisible()) {
          VertexConsumer var11 = var2.getBuffer(RenderType.entitySolid(var4.getSkinTextureLocation()));
          int var12 = LivingEntityRenderer.getOverlayCoords(var4, 0.0F);
 
          for(int var13 = 0; var13 < 2; ++var13) {
-            float var14 = Mth.lerp(var7, var4.yRotO, var4.yRot) - Mth.lerp(var7, var4.yBodyRotO, var4.yBodyRot);
-            float var15 = Mth.lerp(var7, var4.xRotO, var4.xRot);
+            float var14 = Mth.lerp(var7, var4.yRotO, var4.getYRot()) - Mth.lerp(var7, var4.yBodyRotO, var4.yBodyRot);
+            float var15 = Mth.lerp(var7, var4.xRotO, var4.getXRot());
             var1.pushPose();
             var1.mulPose(Vector3f.YP.rotationDegrees(var14));
             var1.mulPose(Vector3f.XP.rotationDegrees(var15));
             var1.translate((double)(0.375F * (float)(var13 * 2 - 1)), 0.0D, 0.0D);
             var1.translate(0.0D, -0.375D, 0.0D);
