--- net/minecraft/client/renderer/entity/layers/DolphinCarryingItemLayer.java
+++ net/minecraft/client/renderer/entity/layers/DolphinCarryingItemLayer.java
@@ -19,12 +19,12 @@
    public void render(PoseStack var1, MultiBufferSource var2, int var3, Dolphin var4, float var5, float var6, float var7, float var8, float var9, float var10) {
       boolean var11 = var4.getMainArm() == HumanoidArm.RIGHT;
       var1.pushPose();
       float var12 = 1.0F;
       float var13 = -1.0F;
-      float var14 = Mth.abs(var4.xRot) / 60.0F;
-      if (var4.xRot < 0.0F) {
+      float var14 = Mth.abs(var4.getXRot()) / 60.0F;
+      if (var4.getXRot() < 0.0F) {
          var1.translate(0.0D, (double)(1.0F - var14 * 0.5F), (double)(-1.0F + var14 * 0.5F));
       } else {
          var1.translate(0.0D, (double)(1.0F + var14 * 0.8F), (double)(-1.0F + var14 * 0.2F));
       }
 
