--- net/minecraft/client/renderer/entity/layers/EnergySwirlLayer.java
+++ net/minecraft/client/renderer/entity/layers/EnergySwirlLayer.java
@@ -20,11 +20,11 @@
       if (((PowerableMob)var4).isPowered()) {
          float var11 = (float)var4.tickCount + var7;
          EntityModel var12 = this.model();
          var12.prepareMobModel(var4, var5, var6, var7);
          this.getParentModel().copyPropertiesTo(var12);
-         VertexConsumer var13 = var2.getBuffer(RenderType.energySwirl(this.getTextureLocation(), this.xOffset(var11), var11 * 0.01F));
+         VertexConsumer var13 = var2.getBuffer(RenderType.energySwirl(this.getTextureLocation(), this.xOffset(var11) % 1.0F, var11 * 0.01F % 1.0F));
          var12.setupAnim(var4, var5, var6, var8, var9, var10);
          var12.renderToBuffer(var1, var13, var3, OverlayTexture.NO_OVERLAY, 0.5F, 0.5F, 0.5F, 1.0F);
       }
    }
 
