--- net/minecraft/client/renderer/entity/layers/HumanoidArmorLayer.java
+++ net/minecraft/client/renderer/entity/layers/HumanoidArmorLayer.java
@@ -96,9 +96,10 @@
    private boolean usesInnerModel(EquipmentSlot var1) {
       return var1 == EquipmentSlot.LEGS;
    }
 
    private ResourceLocation getArmorLocation(ArmorItem var1, boolean var2, @Nullable String var3) {
-      String var4 = "textures/models/armor/" + var1.getMaterial().getName() + "_layer_" + (var2 ? 2 : 1) + (var3 == null ? "" : "_" + var3) + ".png";
+      String var10000 = var1.getMaterial().getName();
+      String var4 = "textures/models/armor/" + var10000 + "_layer_" + (var2 ? 2 : 1) + (var3 == null ? "" : "_" + var3) + ".png";
       return (ResourceLocation)ARMOR_LOCATION_CACHE.computeIfAbsent(var4, ResourceLocation::new);
    }
 }
