--- net/minecraft/client/renderer/entity/layers/ItemInHandLayer.java
+++ net/minecraft/client/renderer/entity/layers/ItemInHandLayer.java
@@ -33,11 +33,11 @@
          this.renderArmWithItem(var4, var12, ItemTransforms.TransformType.THIRD_PERSON_LEFT_HAND, HumanoidArm.LEFT, var1, var2, var3);
          var1.popPose();
       }
    }
 
-   private void renderArmWithItem(LivingEntity var1, ItemStack var2, ItemTransforms.TransformType var3, HumanoidArm var4, PoseStack var5, MultiBufferSource var6, int var7) {
+   protected void renderArmWithItem(LivingEntity var1, ItemStack var2, ItemTransforms.TransformType var3, HumanoidArm var4, PoseStack var5, MultiBufferSource var6, int var7) {
       if (!var2.isEmpty()) {
          var5.pushPose();
          ((ArmedModel)this.getParentModel()).translateToHand(var4, var5);
          var5.mulPose(Vector3f.XP.rotationDegrees(-90.0F));
          var5.mulPose(Vector3f.YP.rotationDegrees(180.0F));
