--- net/minecraft/client/renderer/entity/layers/SnowGolemHeadLayer.java
+++ net/minecraft/client/renderer/entity/layers/SnowGolemHeadLayer.java
@@ -24,10 +24,10 @@
          float var11 = 0.625F;
          var1.translate(0.0D, -0.34375D, 0.0D);
          var1.mulPose(Vector3f.YP.rotationDegrees(180.0F));
          var1.scale(0.625F, -0.625F, -0.625F);
          ItemStack var12 = new ItemStack(Blocks.CARVED_PUMPKIN);
-         Minecraft.getInstance().getItemRenderer().renderStatic(var4, var12, ItemTransforms.TransformType.HEAD, false, var1, var2, var4.level, var3, LivingEntityRenderer.getOverlayCoords(var4, 0.0F));
+         Minecraft.getInstance().getItemRenderer().renderStatic(var4, var12, ItemTransforms.TransformType.HEAD, false, var1, var2, var4.level, var3, LivingEntityRenderer.getOverlayCoords(var4, 0.0F), var4.getId());
          var1.popPose();
       }
    }
 }
