--- net/minecraft/client/renderer/entity/layers/TropicalFishPatternLayer.java
+++ net/minecraft/client/renderer/entity/layers/TropicalFishPatternLayer.java
@@ -1,21 +1,26 @@
 package net.minecraft.client.renderer.entity.layers;
 
 import com.mojang.blaze3d.vertex.PoseStack;
+import net.minecraft.client.model.ColorableHierarchicalModel;
 import net.minecraft.client.model.EntityModel;
 import net.minecraft.client.model.TropicalFishModelA;
 import net.minecraft.client.model.TropicalFishModelB;
+import net.minecraft.client.model.geom.EntityModelSet;
+import net.minecraft.client.model.geom.ModelLayers;
 import net.minecraft.client.renderer.MultiBufferSource;
 import net.minecraft.client.renderer.entity.RenderLayerParent;
 import net.minecraft.world.entity.animal.TropicalFish;
 
-public class TropicalFishPatternLayer extends RenderLayer<TropicalFish, EntityModel<TropicalFish>> {
-   private final TropicalFishModelA<TropicalFish> modelA = new TropicalFishModelA(0.008F);
-   private final TropicalFishModelB<TropicalFish> modelB = new TropicalFishModelB(0.008F);
+public class TropicalFishPatternLayer extends RenderLayer<TropicalFish, ColorableHierarchicalModel<TropicalFish>> {
+   private final TropicalFishModelA<TropicalFish> modelA;
+   private final TropicalFishModelB<TropicalFish> modelB;
 
-   public TropicalFishPatternLayer(RenderLayerParent<TropicalFish, EntityModel<TropicalFish>> var1) {
+   public TropicalFishPatternLayer(RenderLayerParent<TropicalFish, ColorableHierarchicalModel<TropicalFish>> var1, EntityModelSet var2) {
       super(var1);
+      this.modelA = new TropicalFishModelA(var2.getLayer(ModelLayers.TROPICAL_FISH_SMALL_PATTERN));
+      this.modelB = new TropicalFishModelB(var2.getLayer(ModelLayers.TROPICAL_FISH_LARGE_PATTERN));
    }
 
    public void render(PoseStack var1, MultiBufferSource var2, int var3, TropicalFish var4, float var5, float var6, float var7, float var8, float var9, float var10) {
       Object var11 = var4.getBaseVariant() == 0 ? this.modelA : this.modelB;
       float[] var12 = var4.getPatternColor();
