--- net/minecraft/resources/RegistryWriteOps.java
+++ net/minecraft/resources/RegistryWriteOps.java
@@ -19,11 +19,11 @@
       super(var1);
       this.registryHolder = var2;
    }
 
    protected <E> DataResult<T> encode(E var1, T var2, ResourceKey<? extends Registry<E>> var3, Codec<E> var4) {
-      Optional var5 = this.registryHolder.registry(var3);
+      Optional var5 = this.registryHolder.ownedRegistry(var3);
       if (var5.isPresent()) {
          WritableRegistry var6 = (WritableRegistry)var5.get();
          Optional var7 = var6.getResourceKey(var1);
          if (var7.isPresent()) {
             ResourceKey var8 = (ResourceKey)var7.get();
