--- net/minecraft/gametest/framework/GameTestAssertPosException.java
+++ net/minecraft/gametest/framework/GameTestAssertPosException.java
@@ -6,10 +6,17 @@
 public class GameTestAssertPosException extends GameTestAssertException {
    private final BlockPos absolutePos;
    private final BlockPos relativePos;
    private final long tick;
 
+   public GameTestAssertPosException(String var1, BlockPos var2, BlockPos var3, long var4) {
+      super(var1);
+      this.absolutePos = var2;
+      this.relativePos = var3;
+      this.tick = var4;
+   }
+
    public String getMessage() {
       String var1 = "" + this.absolutePos.getX() + "," + this.absolutePos.getY() + "," + this.absolutePos.getZ() + " (relative: " + this.relativePos.getX() + "," + this.relativePos.getY() + "," + this.relativePos.getZ() + ")";
       return super.getMessage() + " at " + var1 + " (t=" + this.tick + ")";
    }
 
@@ -17,9 +24,14 @@
    public String getMessageToShowAtBlock() {
       return super.getMessage() + " here";
    }
 
    @Nullable
+   public BlockPos getRelativePos() {
+      return this.relativePos;
+   }
+
+   @Nullable
    public BlockPos getAbsolutePos() {
       return this.absolutePos;
    }
 }
