--- net/minecraft/gametest/framework/LogTestReporter.java
+++ net/minecraft/gametest/framework/LogTestReporter.java
@@ -10,12 +10,12 @@
    public LogTestReporter() {
    }
 
    public void onTestFailed(GameTestInfo var1) {
       if (var1.isRequired()) {
-         LOGGER.error(var1.getTestName() + " failed! " + Util.describeError(var1.getError()));
+         LOGGER.error((String)"{} failed! {}", (Object)var1.getTestName(), (Object)Util.describeError(var1.getError()));
       } else {
-         LOGGER.warn("(optional) " + var1.getTestName() + " failed. " + Util.describeError(var1.getError()));
+         LOGGER.warn((String)"(optional) {} failed. {}", (Object)var1.getTestName(), (Object)Util.describeError(var1.getError()));
       }
 
    }
 }
