--- net/minecraft/gametest/framework/MultipleTestTracker.java
+++ net/minecraft/gametest/framework/MultipleTestTracker.java
@@ -6,11 +6,11 @@
 import javax.annotation.Nullable;
 
 public class MultipleTestTracker {
    private final Collection<GameTestInfo> tests = Lists.newArrayList();
    @Nullable
-   private Collection<GameTestListener> listeners = Lists.newArrayList();
+   private final Collection<GameTestListener> listeners = Lists.newArrayList();
 
    public MultipleTestTracker() {
    }
 
    public MultipleTestTracker(Collection<GameTestInfo> var1) {
