--- net/minecraft/gametest/framework/MultipleTestTracker.java
+++ net/minecraft/gametest/framework/MultipleTestTracker.java
@@ -1,27 +1,36 @@
 package net.minecraft.gametest.framework;
 
 import com.google.common.collect.Lists;
 import java.util.Collection;
+import java.util.Objects;
 import java.util.function.Consumer;
+import java.util.stream.Collectors;
 import javax.annotation.Nullable;
 
 public class MultipleTestTracker {
+   private static final char NOT_STARTED_TEST_CHAR = ' ';
+   private static final char ONGOING_TEST_CHAR = '_';
+   private static final char SUCCESSFUL_TEST_CHAR = '+';
+   private static final char FAILED_OPTIONAL_TEST_CHAR = 'x';
+   private static final char FAILED_REQUIRED_TEST_CHAR = 'X';
    private final Collection<GameTestInfo> tests = Lists.newArrayList();
    @Nullable
-   private Collection<GameTestListener> listeners = Lists.newArrayList();
+   private final Collection<GameTestListener> listeners = Lists.newArrayList();
 
    public MultipleTestTracker() {
    }
 
    public MultipleTestTracker(Collection<GameTestInfo> var1) {
       this.tests.addAll(var1);
    }
 
    public void addTestToTrack(GameTestInfo var1) {
       this.tests.add(var1);
-      this.listeners.forEach(var1::addListener);
+      Collection var10000 = this.listeners;
+      Objects.requireNonNull(var1);
+      var10000.forEach(var1::addListener);
    }
 
    public void addListener(GameTestListener var1) {
       this.listeners.add(var1);
       this.tests.forEach((var1x) -> {
@@ -32,10 +41,13 @@
    public void addFailureListener(final Consumer<GameTestInfo> var1) {
       this.addListener(new GameTestListener() {
          public void testStructureLoaded(GameTestInfo var1x) {
          }
 
+         public void testPassed(GameTestInfo var1x) {
+         }
+
          public void testFailed(GameTestInfo var1x) {
             var1.accept(var1x);
          }
       });
    }
@@ -58,10 +70,18 @@
 
    public boolean hasFailedOptional() {
       return this.getFailedOptionalCount() > 0;
    }
 
+   public Collection<GameTestInfo> getFailedRequired() {
+      return (Collection)this.tests.stream().filter(GameTestInfo::hasFailed).filter(GameTestInfo::isRequired).collect(Collectors.toList());
+   }
+
+   public Collection<GameTestInfo> getFailedOptional() {
+      return (Collection)this.tests.stream().filter(GameTestInfo::hasFailed).filter(GameTestInfo::isOptional).collect(Collectors.toList());
+   }
+
    public int getTotalCount() {
       return this.tests.size();
    }
 
    public boolean isDone() {
