--- net/minecraft/world/entity/monster/piglin/AbstractPiglin.java
+++ net/minecraft/world/entity/monster/piglin/AbstractPiglin.java
@@ -20,11 +20,11 @@
 import net.minecraft.world.level.Level;
 import net.minecraft.world.level.pathfinder.BlockPathTypes;
 
 public abstract class AbstractPiglin extends Monster {
    protected static final EntityDataAccessor<Boolean> DATA_IMMUNE_TO_ZOMBIFICATION;
-   protected int timeInOverworld = 0;
+   protected int timeInOverworld;
 
    public AbstractPiglin(EntityType<? extends AbstractPiglin> var1, Level var2) {
       super(var1, var2);
       this.setCanPickUpLoot(true);
       this.applyOpenDoorsAbility();
