--- net/minecraft/world/effect/MobEffect.java
+++ net/minecraft/world/effect/MobEffect.java
@@ -133,20 +133,28 @@
 
    public Component getDisplayName() {
       return new TranslatableComponent(this.getDescriptionId());
    }
 
+   public MobEffectCategory getCategory() {
+      return this.category;
+   }
+
    public int getColor() {
       return this.color;
    }
 
    public MobEffect addAttributeModifier(Attribute var1, String var2, double var3, AttributeModifier.Operation var5) {
       AttributeModifier var6 = new AttributeModifier(UUID.fromString(var2), this::getDescriptionId, var3, var5);
       this.attributeModifiers.put(var1, var6);
       return this;
    }
 
+   public Map<Attribute, AttributeModifier> getAttributeModifiers() {
+      return this.attributeModifiers;
+   }
+
    public void removeAttributeModifiers(LivingEntity var1, AttributeMap var2, int var3) {
       Iterator var4 = this.attributeModifiers.entrySet().iterator();
 
       while(var4.hasNext()) {
          Entry var5 = (Entry)var4.next();
@@ -174,6 +182,10 @@
    }
 
    public double getAttributeModifierValue(int var1, AttributeModifier var2) {
       return var2.getAmount() * (double)(var1 + 1);
    }
+
+   public boolean isBeneficial() {
+      return this.category == MobEffectCategory.BENEFICIAL;
+   }
 }
