--- net/minecraft/world/food/FoodData.java
+++ net/minecraft/world/food/FoodData.java
@@ -93,18 +93,26 @@
 
    public int getFoodLevel() {
       return this.foodLevel;
    }
 
+   public int getLastFoodLevel() {
+      return this.lastFoodLevel;
+   }
+
    public boolean needsFood() {
       return this.foodLevel < 20;
    }
 
    public void addExhaustion(float var1) {
       this.exhaustionLevel = Math.min(this.exhaustionLevel + var1, 40.0F);
    }
 
+   public float getExhaustionLevel() {
+      return this.exhaustionLevel;
+   }
+
    public float getSaturationLevel() {
       return this.saturationLevel;
    }
 
    public void setFoodLevel(int var1) {
@@ -112,6 +120,10 @@
    }
 
    public void setSaturation(float var1) {
       this.saturationLevel = var1;
    }
+
+   public void setExhaustion(float var1) {
+      this.exhaustionLevel = var1;
+   }
 }
