--- net/minecraft/world/food/Foods.java
+++ net/minecraft/world/food/Foods.java
@@ -69,11 +69,11 @@
       MUSHROOM_STEW = stew(6);
       MUTTON = (new FoodProperties.Builder()).nutrition(2).saturationMod(0.3F).meat().build();
       POISONOUS_POTATO = (new FoodProperties.Builder()).nutrition(2).saturationMod(0.3F).effect(new MobEffectInstance(MobEffects.POISON, 100, 0), 0.6F).build();
       PORKCHOP = (new FoodProperties.Builder()).nutrition(3).saturationMod(0.3F).meat().build();
       POTATO = (new FoodProperties.Builder()).nutrition(1).saturationMod(0.3F).build();
-      PUFFERFISH = (new FoodProperties.Builder()).nutrition(1).saturationMod(0.1F).effect(new MobEffectInstance(MobEffects.POISON, 1200, 3), 1.0F).effect(new MobEffectInstance(MobEffects.HUNGER, 300, 2), 1.0F).effect(new MobEffectInstance(MobEffects.CONFUSION, 300, 0), 1.0F).build();
+      PUFFERFISH = (new FoodProperties.Builder()).nutrition(1).saturationMod(0.1F).effect(new MobEffectInstance(MobEffects.POISON, 1200, 1), 1.0F).effect(new MobEffectInstance(MobEffects.HUNGER, 300, 2), 1.0F).effect(new MobEffectInstance(MobEffects.CONFUSION, 300, 0), 1.0F).build();
       PUMPKIN_PIE = (new FoodProperties.Builder()).nutrition(8).saturationMod(0.3F).build();
       RABBIT = (new FoodProperties.Builder()).nutrition(3).saturationMod(0.3F).meat().build();
       RABBIT_STEW = stew(10);
       ROTTEN_FLESH = (new FoodProperties.Builder()).nutrition(4).saturationMod(0.1F).effect(new MobEffectInstance(MobEffects.HUNGER, 600, 0), 0.8F).meat().build();
       SALMON = (new FoodProperties.Builder()).nutrition(2).saturationMod(0.1F).build();
