--- net/minecraft/world/food/FoodProperties.java
+++ net/minecraft/world/food/FoodProperties.java
@@ -11,11 +11,11 @@
    private final boolean isMeat;
    private final boolean canAlwaysEat;
    private final boolean fastFood;
    private final List<Pair<MobEffectInstance, Float>> effects;
 
-   private FoodProperties(int var1, float var2, boolean var3, boolean var4, boolean var5, List<Pair<MobEffectInstance, Float>> var6) {
+   FoodProperties(int var1, float var2, boolean var3, boolean var4, boolean var5, List<Pair<MobEffectInstance, Float>> var6) {
       this.nutrition = var1;
       this.saturationModifier = var2;
       this.isMeat = var3;
       this.canAlwaysEat = var4;
       this.fastFood = var5;
