--- net/minecraft/util/datafix/fixes/BitStorageAlignFix.java
+++ net/minecraft/util/datafix/fixes/BitStorageAlignFix.java
@@ -14,10 +14,17 @@
 import java.util.List;
 import java.util.stream.LongStream;
 import net.minecraft.util.Mth;
 
 public class BitStorageAlignFix extends DataFix {
+   private static final int BIT_TO_LONG_SHIFT = 6;
+   private static final int SECTION_WIDTH = 16;
+   private static final int SECTION_HEIGHT = 16;
+   private static final int SECTION_SIZE = 4096;
+   private static final int HEIGHTMAP_BITS = 9;
+   private static final int HEIGHTMAP_SIZE = 256;
+
    public BitStorageAlignFix(Schema var1) {
       super(var1, false);
    }
 
    protected TypeRewriteRule makeRule() {
