--- net/minecraft/util/datafix/fixes/ChunkPalettedStorageFix.java
+++ net/minecraft/util/datafix/fixes/ChunkPalettedStorageFix.java
@@ -31,10 +31,18 @@
 import net.minecraft.util.datafix.PackedBitStorage;
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
 
 public class ChunkPalettedStorageFix extends DataFix {
+   private static final int NORTH_WEST_MASK = 128;
+   private static final int WEST_MASK = 64;
+   private static final int SOUTH_WEST_MASK = 32;
+   private static final int SOUTH_MASK = 16;
+   private static final int SOUTH_EAST_MASK = 8;
+   private static final int EAST_MASK = 4;
+   private static final int NORTH_EAST_MASK = 2;
+   private static final int NORTH_MASK = 1;
    private static final Logger LOGGER = LogManager.getLogger();
    private static final BitSet VIRTUAL = new BitSet(256);
    private static final BitSet FIX = new BitSet(256);
    private static final Dynamic<?> PUMPKIN = BlockStateData.parse("{Name:'minecraft:pumpkin'}");
    private static final Dynamic<?> SNOWY_PODZOL = BlockStateData.parse("{Name:'minecraft:podzol',Properties:{snowy:'true'}}");
@@ -133,10 +141,11 @@
          }
       }
 
    });
    private static final Dynamic<?> AIR;
+   private static final int SIZE = 4096;
 
    public ChunkPalettedStorageFix(Schema var1, boolean var2) {
       super(var1, var2);
    }
 
@@ -412,10 +421,12 @@
          }
       }
    }
 
    static class DataLayer {
+      private static final int SIZE = 2048;
+      private static final int NIBBLE_SIZE = 4;
       private final byte[] data;
 
       public DataLayer() {
          this.data = new byte[2048];
       }
