--- net/minecraft/util/datafix/fixes/ChunkToProtochunkFix.java
+++ net/minecraft/util/datafix/fixes/ChunkToProtochunkFix.java
@@ -19,10 +19,12 @@
 import java.util.stream.Collectors;
 import java.util.stream.IntStream;
 import java.util.stream.Stream;
 
 public class ChunkToProtochunkFix extends DataFix {
+   private static final int NUM_SECTIONS = 16;
+
    public ChunkToProtochunkFix(Schema var1, boolean var2) {
       super(var1, var2);
    }
 
    public TypeRewriteRule makeRule() {
