--- net/minecraft/util/datafix/fixes/EntityEquipmentToArmorAndHandFix.java
+++ net/minecraft/util/datafix/fixes/EntityEquipmentToArmorAndHandFix.java
@@ -11,10 +11,11 @@
 import com.mojang.datafixers.util.Pair;
 import com.mojang.serialization.Dynamic;
 import java.util.ArrayList;
 import java.util.Iterator;
 import java.util.List;
+import java.util.Objects;
 import java.util.Optional;
 import java.util.stream.Stream;
 
 public class EntityEquipmentToArmorAndHandFix extends DataFix {
    public EntityEquipmentToArmorAndHandFix(Schema var1, boolean var2) {
@@ -63,18 +64,18 @@
             float var16 = ((Dynamic)var15.next()).asFloat(0.0F);
             Dynamic var13;
             Stream var10001;
             if (!var7.get("HandDropChances").result().isPresent()) {
                var10001 = Stream.of(var16, 0.0F);
-               var7.getClass();
+               Objects.requireNonNull(var7);
                var13 = var7.createList(var10001.map(var7::createFloat));
                var7 = var7.set("HandDropChances", var13);
             }
 
             if (!var7.get("ArmorDropChances").result().isPresent()) {
                var10001 = Stream.of(((Dynamic)var15.next()).asFloat(0.0F), ((Dynamic)var15.next()).asFloat(0.0F), ((Dynamic)var15.next()).asFloat(0.0F), ((Dynamic)var15.next()).asFloat(0.0F));
-               var7.getClass();
+               Objects.requireNonNull(var7);
                var13 = var7.createList(var10001.map(var7::createFloat));
                var7 = var7.set("ArmorDropChances", var13);
             }
 
             var7 = var7.remove("DropChances");
