--- net/minecraft/util/datafix/fixes/IglooMetadataRemovalFix.java
+++ net/minecraft/util/datafix/fixes/IglooMetadataRemovalFix.java
@@ -4,10 +4,11 @@
 import com.mojang.datafixers.TypeRewriteRule;
 import com.mojang.datafixers.schemas.Schema;
 import com.mojang.datafixers.types.Type;
 import com.mojang.serialization.DataResult;
 import com.mojang.serialization.Dynamic;
+import java.util.Objects;
 
 public class IglooMetadataRemovalFix extends DataFix {
    public IglooMetadataRemovalFix(Schema var1, boolean var2) {
       super(var1, var2);
    }
@@ -29,11 +30,11 @@
       DataResult var10000 = var0.asStreamOpt().map((var0x) -> {
          return var0x.filter((var0) -> {
             return !isIglooPiece(var0);
          });
       });
-      var0.getClass();
+      Objects.requireNonNull(var0);
       return (Dynamic)var10000.map(var0::createList).result().orElse(var0);
    }
 
    private static boolean isIglooPiece(Dynamic<?> var0) {
       return var0.get("id").asString("").equals("Iglu");
