--- net/minecraft/util/datafix/fixes/ItemWrittenBookPagesStrictJsonFix.java
+++ net/minecraft/util/datafix/fixes/ItemWrittenBookPagesStrictJsonFix.java
@@ -8,10 +8,11 @@
 import com.mojang.datafixers.TypeRewriteRule;
 import com.mojang.datafixers.schemas.Schema;
 import com.mojang.datafixers.types.Type;
 import com.mojang.serialization.DataResult;
 import com.mojang.serialization.Dynamic;
+import java.util.Objects;
 import net.minecraft.network.chat.Component;
 import net.minecraft.network.chat.TextComponent;
 import net.minecraft.util.GsonHelper;
 import org.apache.commons.lang3.StringUtils;
 
@@ -65,11 +66,11 @@
 
                   return var0x.createString(Component.Serializer.toJson((Component)var2));
                }
             });
          });
-         var1.getClass();
+         Objects.requireNonNull(var1);
          return (Dynamic)DataFixUtils.orElse(var10000.map(var1::createList).result(), var1.emptyList());
       });
    }
 
    public TypeRewriteRule makeRule() {
