--- net/minecraft/util/datafix/fixes/ItemStackSpawnEggFix.java
+++ net/minecraft/util/datafix/fixes/ItemStackSpawnEggFix.java
@@ -79,11 +79,11 @@
          if (var5x.isPresent() && Objects.equals(((Pair)var5x.get()).getSecond(), "minecraft:spawn_egg")) {
             Typed var6 = var4x.getOrCreateTyped(var4);
             Typed var7 = var6.getOrCreateTyped(var5);
             Optional var8 = var7.getOptional(var3);
             if (var8.isPresent()) {
-               return var4x.set(var2, Pair.of(References.ITEM_NAME.typeName(), MAP.getOrDefault(var8.get(), "minecraft:pig_spawn_egg")));
+               return var4x.set(var2, Pair.of(References.ITEM_NAME.typeName(), (String)MAP.getOrDefault(var8.get(), "minecraft:pig_spawn_egg")));
             }
          }
 
          return var4x;
       });
