--- net/minecraft/util/datafix/fixes/ItemStackTheFlatteningFix.java
+++ net/minecraft/util/datafix/fixes/ItemStackTheFlatteningFix.java
@@ -380,11 +380,11 @@
    }
 
    @Nullable
    public static String updateItem(@Nullable String var0, int var1) {
       if (IDS.contains(var0)) {
-         String var2 = (String)MAP.get(var0 + '.' + var1);
+         String var2 = (String)MAP.get(var0 + "." + var1);
          return var2 == null ? (String)MAP.get(var0 + ".0") : var2;
       } else {
          return null;
       }
    }
