--- net/minecraft/util/datafix/fixes/ItemPotionFix.java
+++ net/minecraft/util/datafix/fixes/ItemPotionFix.java
@@ -13,10 +13,11 @@
 import java.util.Objects;
 import java.util.Optional;
 import net.minecraft.util.datafix.schemas.NamespacedSchema;
 
 public class ItemPotionFix extends DataFix {
+   private static final int SPLASH = 16384;
    private static final String[] POTIONS = (String[])DataFixUtils.make(new String[128], (var0) -> {
       var0[0] = "minecraft:water";
       var0[1] = "minecraft:regeneration";
       var0[2] = "minecraft:swiftness";
       var0[3] = "minecraft:fire_resistance";
@@ -143,10 +144,11 @@
       var0[124] = "minecraft:strong_harming";
       var0[125] = "minecraft:long_water_breathing";
       var0[126] = "minecraft:long_invisibility";
       var0[127] = null;
    });
+   public static final String DEFAULT = "minecraft:water";
 
    public ItemPotionFix(Schema var1, boolean var2) {
       super(var1, var2);
    }
 
