--- net/minecraft/util/datafix/fixes/LevelDataGeneratorOptionsFix.java
+++ net/minecraft/util/datafix/fixes/LevelDataGeneratorOptionsFix.java
@@ -100,10 +100,11 @@
       var0.put("164", "minecraft:shattered_savanna_plateau");
       var0.put("165", "minecraft:eroded_badlands");
       var0.put("166", "minecraft:modified_wooded_badlands_plateau");
       var0.put("167", "minecraft:modified_badlands_plateau");
    });
+   public static final String GENERATOR_OPTIONS = "generatorOptions";
 
    public LevelDataGeneratorOptionsFix(Schema var1, boolean var2) {
       super(var1, var2);
    }
 
