--- net/minecraft/util/datafix/fixes/PoiTypeRename.java
+++ net/minecraft/util/datafix/fixes/PoiTypeRename.java
@@ -44,11 +44,11 @@
    private <T> Optional<Dynamic<T>> renameRecords(Dynamic<T> var1) {
       return var1.asStreamOpt().map((var2) -> {
          return var1.createList(var2.map((var1x) -> {
             return var1x.update("type", (var1) -> {
                DataResult var10000 = var1.asString().map(this::rename);
-               var1.getClass();
+               Objects.requireNonNull(var1);
                return (Dynamic)DataFixUtils.orElse(var10000.map(var1::createString).result(), var1);
             });
          }));
       }).result();
    }
