--- net/minecraft/util/datafix/fixes/References.java
+++ net/minecraft/util/datafix/fixes/References.java
@@ -31,10 +31,13 @@
       return "advancements";
    };
    public static final TypeReference POI_CHUNK = () -> {
       return "poi_chunk";
    };
+   public static final TypeReference ENTITY_CHUNK = () -> {
+      return "entity_chunk";
+   };
    public static final TypeReference BLOCK_ENTITY = () -> {
       return "block_entity";
    };
    public static final TypeReference ITEM_STACK = () -> {
       return "item_stack";
