--- net/minecraft/util/datafix/fixes/TeamDisplayNameFix.java
+++ net/minecraft/util/datafix/fixes/TeamDisplayNameFix.java
@@ -27,11 +27,11 @@
                return var0x.mapSecond((var0) -> {
                   return var0.update("DisplayName", (var1) -> {
                      DataResult var10000 = var1.asString().map((var0x) -> {
                         return Component.Serializer.toJson(new TextComponent(var0x));
                      });
-                     var0.getClass();
+                     Objects.requireNonNull(var0);
                      return (Dynamic)DataFixUtils.orElse(var10000.map(var0::createString).result(), var1);
                   });
                });
             };
          });
