--- net/minecraft/util/datafix/fixes/VillagerFollowRangeFix.java
+++ net/minecraft/util/datafix/fixes/VillagerFollowRangeFix.java
@@ -4,10 +4,13 @@
 import com.mojang.datafixers.Typed;
 import com.mojang.datafixers.schemas.Schema;
 import com.mojang.serialization.Dynamic;
 
 public class VillagerFollowRangeFix extends NamedEntityFix {
+   private static final double ORIGINAL_VALUE = 16.0D;
+   private static final double NEW_BASE_VALUE = 48.0D;
+
    public VillagerFollowRangeFix(Schema var1) {
       super(var1, false, "Villager Follow Range Fix", References.ENTITY, "minecraft:villager");
    }
 
    protected Typed<?> fix(Typed<?> var1) {
