--- net/minecraft/server/level/DemoMode.java
+++ net/minecraft/server/level/DemoMode.java
@@ -16,11 +16,11 @@
    private boolean displayedIntro;
    private boolean demoHasEnded;
    private int demoEndedReminder;
    private int gameModeTicks;
 
-   public DemoMode(ServerLevel var1) {
+   public DemoMode(ServerPlayer var1) {
       super(var1);
    }
 
    public void tick() {
       super.tick();
