--- net/minecraft/server/level/DistanceManager.java
+++ net/minecraft/server/level/DistanceManager.java
@@ -91,12 +91,12 @@
       boolean var3 = var2 != 0;
       if (var3) {
       }
 
       if (!this.chunksToUpdateFutures.isEmpty()) {
-         this.chunksToUpdateFutures.forEach((var1x) -> {
-            var1x.updateFutures(var1);
+         this.chunksToUpdateFutures.forEach((var2x) -> {
+            var2x.updateFutures(var1, this.mainThreadExecutor);
          });
          this.chunksToUpdateFutures.clear();
          return true;
       } else {
          if (!this.ticketsToRelease.isEmpty()) {
