--- net/minecraft/server/level/ServerBossEvent.java
+++ net/minecraft/server/level/ServerBossEvent.java
@@ -21,14 +21,14 @@
       super(Mth.createInsecureUUID(), var1, var2, var3);
       this.unmodifiablePlayers = Collections.unmodifiableSet(this.players);
       this.visible = true;
    }
 
-   public void setPercent(float var1) {
-      if (var1 != this.percent) {
-         super.setPercent(var1);
-         this.broadcast(ClientboundBossEventPacket.Operation.UPDATE_PCT);
+   public void setProgress(float var1) {
+      if (var1 != this.progress) {
+         super.setProgress(var1);
+         this.broadcast(ClientboundBossEventPacket.Operation.UPDATE_PROGRESS);
       }
 
    }
 
    public void setColor(BossEvent.BossBarColor var1) {
