--- net/minecraft/server/level/ServerEntity.java
+++ net/minecraft/server/level/ServerEntity.java
@@ -5,15 +5,17 @@
 import java.util.ArrayList;
 import java.util.Collection;
 import java.util.Collections;
 import java.util.Iterator;
 import java.util.List;
+import java.util.Objects;
 import java.util.Set;
 import java.util.function.Consumer;
 import net.minecraft.network.protocol.Packet;
 import net.minecraft.network.protocol.game.ClientboundAddMobPacket;
 import net.minecraft.network.protocol.game.ClientboundMoveEntityPacket;
+import net.minecraft.network.protocol.game.ClientboundRemoveEntityPacket;
 import net.minecraft.network.protocol.game.ClientboundRotateHeadPacket;
 import net.minecraft.network.protocol.game.ClientboundSetEntityDataPacket;
 import net.minecraft.network.protocol.game.ClientboundSetEntityLinkPacket;
 import net.minecraft.network.protocol.game.ClientboundSetEntityMotionPacket;
 import net.minecraft.network.protocol.game.ClientboundSetEquipmentPacket;
@@ -38,10 +40,11 @@
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
 
 public class ServerEntity {
    private static final Logger LOGGER = LogManager.getLogger();
+   private static final int TOLERANCE_LEVEL_ROTATION = 1;
    private final ServerLevel level;
    private final Entity entity;
    private final int updateInterval;
    private final boolean trackDelta;
    private final Consumer<Packet<?>> broadcast;
@@ -65,12 +68,12 @@
       this.broadcast = var5;
       this.entity = var2;
       this.updateInterval = var3;
       this.trackDelta = var4;
       this.updateSentPos();
-      this.yRotp = Mth.floor(var2.yRot * 256.0F / 360.0F);
-      this.xRotp = Mth.floor(var2.xRot * 256.0F / 360.0F);
+      this.yRotp = Mth.floor(var2.getYRot() * 256.0F / 360.0F);
+      this.xRotp = Mth.floor(var2.getXRot() * 256.0F / 360.0F);
       this.yHeadRotp = Mth.floor(var2.getYHeadRot() * 256.0F / 360.0F);
       this.wasOnGround = var2.isOnGround();
    }
 
    public void sendChanges() {
@@ -82,19 +85,22 @@
 
       if (this.entity instanceof ItemFrame && this.tickCount % 10 == 0) {
          ItemFrame var2 = (ItemFrame)this.entity;
          ItemStack var3 = var2.getItem();
          if (var3.getItem() instanceof MapItem) {
-            MapItemSavedData var4 = MapItem.getOrCreateSavedData(var3, this.level);
-            Iterator var5 = this.level.players().iterator();
+            Integer var4 = MapItem.getMapId(var3);
+            MapItemSavedData var5 = MapItem.getSavedData((Integer)var4, this.level);
+            if (var5 != null) {
+               Iterator var6 = this.level.players().iterator();
 
-            while(var5.hasNext()) {
-               ServerPlayer var6 = (ServerPlayer)var5.next();
-               var4.tickCarriedBy(var6, var3);
-               Packet var7 = ((MapItem)var3.getItem()).getUpdatePacket(var3, this.level, var6);
-               if (var7 != null) {
-                  var6.connection.send(var7);
+               while(var6.hasNext()) {
+                  ServerPlayer var7 = (ServerPlayer)var6.next();
+                  var5.tickCarriedBy(var7, var3);
+                  Packet var8 = var5.getUpdatePacket(var4, var7);
+                  if (var8 != null) {
+                     var7.connection.send(var8);
+                  }
                }
             }
          }
 
          this.sendDirtyEntityData();
@@ -102,12 +108,12 @@
 
       if (this.tickCount % this.updateInterval == 0 || this.entity.hasImpulse || this.entity.getEntityData().isDirty()) {
          int var16;
          int var17;
          if (this.entity.isPassenger()) {
-            var16 = Mth.floor(this.entity.yRot * 256.0F / 360.0F);
-            var17 = Mth.floor(this.entity.xRot * 256.0F / 360.0F);
+            var16 = Mth.floor(this.entity.getYRot() * 256.0F / 360.0F);
+            var17 = Mth.floor(this.entity.getXRot() * 256.0F / 360.0F);
             boolean var19 = Math.abs(var16 - this.yRotp) >= 1 || Math.abs(var17 - this.xRotp) >= 1;
             if (var19) {
                this.broadcast.accept(new ClientboundMoveEntityPacket.Rot(this.entity.getId(), (byte)var16, (byte)var17, this.entity.isOnGround()));
                this.yRotp = var16;
                this.xRotp = var17;
@@ -116,27 +122,27 @@
             this.updateSentPos();
             this.sendDirtyEntityData();
             this.wasRiding = true;
          } else {
             ++this.teleportDelay;
-            var16 = Mth.floor(this.entity.yRot * 256.0F / 360.0F);
-            var17 = Mth.floor(this.entity.xRot * 256.0F / 360.0F);
+            var16 = Mth.floor(this.entity.getYRot() * 256.0F / 360.0F);
+            var17 = Mth.floor(this.entity.getXRot() * 256.0F / 360.0F);
             Vec3 var18 = this.entity.position().subtract(ClientboundMoveEntityPacket.packetToEntity(this.xp, this.yp, this.zp));
             boolean var20 = var18.lengthSqr() >= 7.62939453125E-6D;
             Object var21 = null;
             boolean var22 = var20 || this.tickCount % 60 == 0;
-            boolean var8 = Math.abs(var16 - this.yRotp) >= 1 || Math.abs(var17 - this.xRotp) >= 1;
+            boolean var23 = Math.abs(var16 - this.yRotp) >= 1 || Math.abs(var17 - this.xRotp) >= 1;
             if (this.tickCount > 0 || this.entity instanceof AbstractArrow) {
                long var9 = ClientboundMoveEntityPacket.entityToPacket(var18.x);
                long var11 = ClientboundMoveEntityPacket.entityToPacket(var18.y);
                long var13 = ClientboundMoveEntityPacket.entityToPacket(var18.z);
                boolean var15 = var9 < -32768L || var9 > 32767L || var11 < -32768L || var11 > 32767L || var13 < -32768L || var13 > 32767L;
                if (!var15 && this.teleportDelay <= 400 && !this.wasRiding && this.wasOnGround == this.entity.isOnGround()) {
-                  if ((!var22 || !var8) && !(this.entity instanceof AbstractArrow)) {
+                  if ((!var22 || !var23) && !(this.entity instanceof AbstractArrow)) {
                      if (var22) {
                         var21 = new ClientboundMoveEntityPacket.Pos(this.entity.getId(), (short)((int)var9), (short)((int)var11), (short)((int)var13), this.entity.isOnGround());
-                     } else if (var8) {
+                     } else if (var23) {
                         var21 = new ClientboundMoveEntityPacket.Rot(this.entity.getId(), (byte)var16, (byte)var17, this.entity.isOnGround());
                      }
                   } else {
                      var21 = new ClientboundMoveEntityPacket.PosRot(this.entity.getId(), (short)((int)var9), (short)((int)var11), (short)((int)var13), (byte)var16, (byte)var17, this.entity.isOnGround());
                   }
@@ -146,14 +152,14 @@
                   var21 = new ClientboundTeleportEntityPacket(this.entity);
                }
             }
 
             if ((this.trackDelta || this.entity.hasImpulse || this.entity instanceof LivingEntity && ((LivingEntity)this.entity).isFallFlying()) && this.tickCount > 0) {
-               Vec3 var23 = this.entity.getDeltaMovement();
-               double var10 = var23.distanceToSqr(this.ap);
-               if (var10 > 1.0E-7D || var10 > 0.0D && var23.lengthSqr() == 0.0D) {
-                  this.ap = var23;
+               Vec3 var24 = this.entity.getDeltaMovement();
+               double var10 = var24.distanceToSqr(this.ap);
+               if (var10 > 1.0E-7D || var10 > 0.0D && var24.lengthSqr() == 0.0D) {
+                  this.ap = var24;
                   this.broadcast.accept(new ClientboundSetEntityMotionPacket(this.entity.getId(), this.ap));
                }
             }
 
             if (var21 != null) {
@@ -163,11 +169,11 @@
             this.sendDirtyEntityData();
             if (var22) {
                this.updateSentPos();
             }
 
-            if (var8) {
+            if (var23) {
                this.yRotp = var16;
                this.xRotp = var17;
             }
 
             this.wasRiding = false;
@@ -190,23 +196,23 @@
 
    }
 
    public void removePairing(ServerPlayer var1) {
       this.entity.stopSeenByPlayer(var1);
-      var1.sendRemoveEntity(this.entity);
+      var1.connection.send(new ClientboundRemoveEntityPacket(this.entity.getId()));
    }
 
    public void addPairing(ServerPlayer var1) {
       ServerGamePacketListenerImpl var10001 = var1.connection;
+      Objects.requireNonNull(var10001);
       this.sendPairingData(var10001::send);
       this.entity.startSeenByPlayer(var1);
-      var1.cancelRemoveEntity(this.entity);
    }
 
    public void sendPairingData(Consumer<Packet<?>> var1) {
-      if (this.entity.removed) {
-         LOGGER.warn("Fetching packet for removed entity " + this.entity);
+      if (this.entity.isRemoved()) {
+         LOGGER.warn((String)"Fetching packet for removed entity {}", (Object)this.entity);
       }
 
       Packet var2 = this.entity.getAddEntityPacket();
       this.yHeadRotp = Mth.floor(this.entity.getYHeadRot() * 256.0F / 360.0F);
       var1.accept(var2);
