--- net/minecraft/world/level/levelgen/synth/ImprovedNoise.java
+++ net/minecraft/world/level/levelgen/synth/ImprovedNoise.java
@@ -42,11 +42,11 @@
       double var26 = Mth.smoothstep(var20);
       double var28 = Mth.smoothstep(var22);
       double var30 = Mth.smoothstep(var24);
       double var32;
       if (var7 != 0.0D) {
-         double var34 = Math.min(var9, var22);
+         double var34 = var9 < 0.0D ? 0.0D : Math.min(var9, var22);
          var32 = (double)Mth.floor(var34 / var7) * var7;
       } else {
          var32 = 0.0D;
       }
 
