--- net/minecraft/world/level/levelgen/synth/NormalNoise.java
+++ net/minecraft/world/level/levelgen/synth/NormalNoise.java
@@ -1,21 +1,26 @@
 package net.minecraft.world.level.levelgen.synth;
 
+import it.unimi.dsi.fastutil.doubles.DoubleArrayList;
 import it.unimi.dsi.fastutil.doubles.DoubleList;
 import it.unimi.dsi.fastutil.doubles.DoubleListIterator;
-import net.minecraft.world.level.levelgen.WorldgenRandom;
+import net.minecraft.world.level.levelgen.RandomSource;
 
 public class NormalNoise {
    private final double valueFactor;
    private final PerlinNoise first;
    private final PerlinNoise second;
 
-   public static NormalNoise create(WorldgenRandom var0, int var1, DoubleList var2) {
+   public static NormalNoise create(RandomSource var0, int var1, double... var2) {
+      return new NormalNoise(var0, var1, new DoubleArrayList(var2));
+   }
+
+   public static NormalNoise create(RandomSource var0, int var1, DoubleList var2) {
       return new NormalNoise(var0, var1, var2);
    }
 
-   private NormalNoise(WorldgenRandom var1, int var2, DoubleList var3) {
+   private NormalNoise(RandomSource var1, int var2, DoubleList var3) {
       this.first = PerlinNoise.create(var1, var2, var3);
       this.second = PerlinNoise.create(var1, var2, var3);
       int var4 = Integer.MAX_VALUE;
       int var5 = Integer.MIN_VALUE;
       DoubleListIterator var6 = var3.iterator();
