--- net/minecraft/world/level/levelgen/synth/PerlinSimplexNoise.java
+++ net/minecraft/world/level/levelgen/synth/PerlinSimplexNoise.java
@@ -3,26 +3,27 @@
 import com.google.common.collect.ImmutableList;
 import it.unimi.dsi.fastutil.ints.IntRBTreeSet;
 import it.unimi.dsi.fastutil.ints.IntSortedSet;
 import java.util.List;
 import java.util.stream.IntStream;
+import net.minecraft.world.level.levelgen.RandomSource;
 import net.minecraft.world.level.levelgen.WorldgenRandom;
 
 public class PerlinSimplexNoise implements SurfaceNoise {
    private final SimplexNoise[] noiseLevels;
    private final double highestFreqValueFactor;
    private final double highestFreqInputFactor;
 
-   public PerlinSimplexNoise(WorldgenRandom var1, IntStream var2) {
+   public PerlinSimplexNoise(RandomSource var1, IntStream var2) {
       this(var1, (List)var2.boxed().collect(ImmutableList.toImmutableList()));
    }
 
-   public PerlinSimplexNoise(WorldgenRandom var1, List<Integer> var2) {
+   public PerlinSimplexNoise(RandomSource var1, List<Integer> var2) {
       this(var1, (IntSortedSet)(new IntRBTreeSet(var2)));
    }
 
-   private PerlinSimplexNoise(WorldgenRandom var1, IntSortedSet var2) {
+   private PerlinSimplexNoise(RandomSource var1, IntSortedSet var2) {
       if (var2.isEmpty()) {
          throw new IllegalArgumentException("Need some octaves!");
       } else {
          int var3 = -var2.firstInt();
          int var4 = var2.lastInt();
