--- net/minecraft/world/level/levelgen/synth/SimplexNoise.java
+++ net/minecraft/world/level/levelgen/synth/SimplexNoise.java
@@ -1,9 +1,9 @@
 package net.minecraft.world.level.levelgen.synth;
 
-import java.util.Random;
 import net.minecraft.util.Mth;
+import net.minecraft.world.level.levelgen.RandomSource;
 
 public class SimplexNoise {
    protected static final int[][] GRADIENT = new int[][]{{1, 1, 0}, {-1, 1, 0}, {1, -1, 0}, {-1, -1, 0}, {1, 0, 1}, {-1, 0, 1}, {1, 0, -1}, {-1, 0, -1}, {0, 1, 1}, {0, -1, 1}, {0, 1, -1}, {0, -1, -1}, {1, 1, 0}, {0, -1, 1}, {-1, 1, 0}, {0, -1, -1}};
    private static final double SQRT_3 = Math.sqrt(3.0D);
    private static final double F2;
@@ -11,11 +11,11 @@
    private final int[] p = new int[512];
    public final double xo;
    public final double yo;
    public final double zo;
 
-   public SimplexNoise(Random var1) {
+   public SimplexNoise(RandomSource var1) {
       this.xo = var1.nextDouble() * 256.0D;
       this.yo = var1.nextDouble() * 256.0D;
       this.zo = var1.nextDouble() * 256.0D;
 
       int var2;
