--- net/minecraft/client/gui/screens/social/PlayerEntry.java
+++ net/minecraft/client/gui/screens/social/PlayerEntry.java
@@ -22,10 +22,12 @@
 import net.minecraft.resources.ResourceLocation;
 import net.minecraft.util.FastColor;
 import net.minecraft.util.FormattedCharSequence;
 
 public class PlayerEntry extends ContainerObjectSelectionList.Entry<PlayerEntry> {
+   private static final int TOOLTIP_DELAY = 10;
+   private static final int TOOLTIP_MAX_WIDTH = 150;
    private final Minecraft minecraft;
    private final List<GuiEventListener> children;
    private final UUID id;
    private final String playerName;
    private final Supplier<ResourceLocation> skinGetter;
@@ -40,10 +42,15 @@
    private static final Component HIDDEN;
    private static final Component BLOCKED;
    private static final Component OFFLINE;
    private static final Component HIDDEN_OFFLINE;
    private static final Component BLOCKED_OFFLINE;
+   private static final int SKIN_SIZE = 24;
+   private static final int PADDING = 4;
+   private static final int CHAT_TOGGLE_ICON_SIZE = 20;
+   private static final int CHAT_TOGGLE_ICON_X = 0;
+   private static final int CHAT_TOGGLE_ICON_Y = 38;
    public static final int SKIN_SHADE;
    public static final int BG_FILL;
    public static final int BG_FILL_REMOVED;
    public static final int PLAYERNAME_COLOR;
    public static final int PLAYER_STATUS_COLOR;
@@ -114,11 +121,11 @@
          this.minecraft.font.getClass();
          var14 = var3 + (var6 - (9 + 9)) / 2;
          this.minecraft.font.draw(var1, var15, (float)var13, (float)(var14 + 12), PLAYER_STATUS_COLOR);
       }
 
-      this.minecraft.getTextureManager().bind((ResourceLocation)this.skinGetter.get());
+      RenderSystem.setShaderTexture(0, (ResourceLocation)this.skinGetter.get());
       GuiComponent.blit(var1, var11, var12, 24, 24, 8.0F, 8.0F, 8, 8, 64, 64);
       RenderSystem.enableBlend();
       GuiComponent.blit(var1, var11, var12, 24, 24, 40.0F, 8.0F, 8, 8, 64, 64);
       RenderSystem.disableBlend();
       this.minecraft.font.draw(var1, this.playerName, (float)var13, (float)var14, PLAYERNAME_COLOR);
