--- net/minecraft/client/gui/screens/social/PlayerEntry.java
+++ net/minecraft/client/gui/screens/social/PlayerEntry.java
@@ -114,11 +114,11 @@
          this.minecraft.font.getClass();
          var14 = var3 + (var6 - (9 + 9)) / 2;
          this.minecraft.font.draw(var1, var15, (float)var13, (float)(var14 + 12), PLAYER_STATUS_COLOR);
       }
 
-      this.minecraft.getTextureManager().bind((ResourceLocation)this.skinGetter.get());
+      RenderSystem.setShaderTexture(0, (ResourceLocation)this.skinGetter.get());
       GuiComponent.blit(var1, var11, var12, 24, 24, 8.0F, 8.0F, 8, 8, 64, 64);
       RenderSystem.enableBlend();
       GuiComponent.blit(var1, var11, var12, 24, 24, 40.0F, 8.0F, 8, 8, 64, 64);
       RenderSystem.disableBlend();
       this.minecraft.font.draw(var1, this.playerName, (float)var13, (float)var14, PLAYERNAME_COLOR);
