--- net/minecraft/client/gui/screens/social/SocialInteractionsPlayerList.java
+++ net/minecraft/client/gui/screens/social/SocialInteractionsPlayerList.java
@@ -6,27 +6,26 @@
 import com.mojang.blaze3d.vertex.PoseStack;
 import java.util.Collection;
 import java.util.Iterator;
 import java.util.List;
 import java.util.Locale;
+import java.util.Objects;
 import java.util.UUID;
 import javax.annotation.Nullable;
 import net.minecraft.client.Minecraft;
 import net.minecraft.client.gui.components.ContainerObjectSelectionList;
 import net.minecraft.client.multiplayer.PlayerInfo;
 
 public class SocialInteractionsPlayerList extends ContainerObjectSelectionList<PlayerEntry> {
    private final SocialInteractionsScreen socialInteractionsScreen;
-   private final Minecraft minecraft;
    private final List<PlayerEntry> players = Lists.newArrayList();
    @Nullable
    private String filter;
 
    public SocialInteractionsPlayerList(SocialInteractionsScreen var1, Minecraft var2, int var3, int var4, int var5, int var6, int var7) {
       super(var2, var3, var4, var5, var6, var7);
       this.socialInteractionsScreen = var1;
-      this.minecraft = var2;
       this.setRenderBackground(false);
       this.setRenderTopAndBottom(false);
    }
 
    public void render(PoseStack var1, int var2, int var3, float var4) {
@@ -47,11 +46,11 @@
             List var10000 = this.players;
             Minecraft var10003 = this.minecraft;
             SocialInteractionsScreen var10004 = this.socialInteractionsScreen;
             UUID var10005 = var6.getProfile().getId();
             String var10006 = var6.getProfile().getName();
-            var6.getClass();
+            Objects.requireNonNull(var6);
             var10000.add(new PlayerEntry(var10003, var10004, var10005, var10006, var6::getSkinLocation));
          }
       }
 
       this.updateFilteredPlayers();
@@ -95,11 +94,11 @@
       if ((var2 == SocialInteractionsScreen.Page.ALL || this.minecraft.getPlayerSocialManager().shouldHideMessageFrom(var3)) && (Strings.isNullOrEmpty(this.filter) || var1.getProfile().getName().toLowerCase(Locale.ROOT).contains(this.filter))) {
          Minecraft var10002 = this.minecraft;
          SocialInteractionsScreen var10003 = this.socialInteractionsScreen;
          UUID var10004 = var1.getProfile().getId();
          String var10005 = var1.getProfile().getName();
-         var1.getClass();
+         Objects.requireNonNull(var1);
          PlayerEntry var6 = new PlayerEntry(var10002, var10003, var10004, var10005, var1::getSkinLocation);
          this.addEntry(var6);
          this.players.add(var6);
       }
 
