--- net/minecraft/client/gui/screens/social/SocialInteractionsScreen.java
+++ net/minecraft/client/gui/screens/social/SocialInteractionsScreen.java
@@ -1,8 +1,9 @@
 package net.minecraft.client.gui.screens.social;
 
 import com.google.common.collect.ImmutableList;
+import com.mojang.blaze3d.systems.RenderSystem;
 import com.mojang.blaze3d.vertex.PoseStack;
 import java.util.Collection;
 import java.util.Locale;
 import java.util.UUID;
 import java.util.stream.Collectors;
@@ -178,11 +179,11 @@
    }
 
    public void renderBackground(PoseStack var1) {
       int var2 = this.marginX() + 3;
       super.renderBackground(var1);
-      this.minecraft.getTextureManager().bind(SOCIAL_INTERACTIONS_LOCATION);
+      RenderSystem.setShaderTexture(0, SOCIAL_INTERACTIONS_LOCATION);
       this.blit(var1, var2, 64, 1, 1, 236, 8);
       int var3 = this.backgroundUnits();
 
       for(int var4 = 0; var4 < var3; ++var4) {
          this.blit(var1, var2, 72 + 16 * var4, 1, 10, 236, 16);
