--- net/minecraft/world/entity/boss/enderdragon/EndCrystal.java
+++ net/minecraft/world/entity/boss/enderdragon/EndCrystal.java
@@ -82,12 +82,12 @@
       if (this.isInvulnerableTo(var1)) {
          return false;
       } else if (var1.getEntity() instanceof EnderDragon) {
          return false;
       } else {
-         if (!this.removed && !this.level.isClientSide) {
-            this.remove();
+         if (!this.isRemoved() && !this.level.isClientSide) {
+            this.remove(Entity.RemovalReason.KILLED);
             if (!var1.isExplosion()) {
                this.level.explode((Entity)null, this.getX(), this.getY(), this.getZ(), 6.0F, Explosion.BlockInteraction.DESTROY);
             }
 
             this.onDestroyedBy(var1);
