--- net/minecraft/world/entity/boss/enderdragon/EndCrystal.java
+++ net/minecraft/world/entity/boss/enderdragon/EndCrystal.java
@@ -12,10 +12,12 @@
 import net.minecraft.network.syncher.SynchedEntityData;
 import net.minecraft.server.level.ServerLevel;
 import net.minecraft.world.damagesource.DamageSource;
 import net.minecraft.world.entity.Entity;
 import net.minecraft.world.entity.EntityType;
+import net.minecraft.world.item.ItemStack;
+import net.minecraft.world.item.Items;
 import net.minecraft.world.level.Explosion;
 import net.minecraft.world.level.Level;
 import net.minecraft.world.level.block.BaseFireBlock;
 import net.minecraft.world.level.dimension.end.EndDragonFight;
 
@@ -33,12 +35,12 @@
    public EndCrystal(Level var1, double var2, double var4, double var6) {
       this(EntityType.END_CRYSTAL, var1);
       this.setPos(var2, var4, var6);
    }
 
-   protected boolean isMovementNoisy() {
-      return false;
+   protected Entity.MovementEmission getMovementEmission() {
+      return Entity.MovementEmission.NONE;
    }
 
    protected void defineSynchedData() {
       this.getEntityData().define(DATA_BEAM_TARGET, Optional.empty());
       this.getEntityData().define(DATA_SHOW_BOTTOM, true);
@@ -82,12 +84,12 @@
       if (this.isInvulnerableTo(var1)) {
          return false;
       } else if (var1.getEntity() instanceof EnderDragon) {
          return false;
       } else {
-         if (!this.removed && !this.level.isClientSide) {
-            this.remove();
+         if (!this.isRemoved() && !this.level.isClientSide) {
+            this.remove(Entity.RemovalReason.KILLED);
             if (!var1.isExplosion()) {
                this.level.explode((Entity)null, this.getX(), this.getY(), this.getZ(), 6.0F, Explosion.BlockInteraction.DESTROY);
             }
 
             this.onDestroyedBy(var1);
@@ -127,10 +129,18 @@
 
    public boolean showsBottom() {
       return (Boolean)this.getEntityData().get(DATA_SHOW_BOTTOM);
    }
 
+   public boolean shouldRenderAtSqrDistance(double var1) {
+      return super.shouldRenderAtSqrDistance(var1) || this.getBeamTarget() != null;
+   }
+
+   public ItemStack getPickResult() {
+      return new ItemStack(Items.END_CRYSTAL);
+   }
+
    public Packet<?> getAddEntityPacket() {
       return new ClientboundAddEntityPacket(this);
    }
 
    static {
