--- net/minecraft/client/sounds/ChannelAccess.java
+++ net/minecraft/client/sounds/ChannelAccess.java
@@ -12,12 +12,12 @@
 import java.util.stream.Stream;
 import javax.annotation.Nullable;
 
 public class ChannelAccess {
    private final Set<ChannelAccess.ChannelHandle> channels = Sets.newIdentityHashSet();
-   private final Library library;
-   private final Executor executor;
+   final Library library;
+   final Executor executor;
 
    public ChannelAccess(Library var1, Executor var2) {
       this.library = var1;
       this.executor = var2;
    }
@@ -67,11 +67,11 @@
       this.channels.clear();
    }
 
    public class ChannelHandle {
       @Nullable
-      private Channel channel;
+      Channel channel;
       private boolean stopped;
 
       public boolean isStopped() {
          return this.stopped;
       }
