--- net/minecraft/client/sounds/LoopingAudioStream.java
+++ net/minecraft/client/sounds/LoopingAudioStream.java
@@ -44,11 +44,11 @@
    public interface AudioStreamProvider {
       AudioStream create(InputStream var1) throws IOException;
    }
 
    static class NoCloseBuffer extends FilterInputStream {
-      private NoCloseBuffer(InputStream var1) {
+      NoCloseBuffer(InputStream var1) {
          super(var1);
       }
 
       public void close() {
       }
