--- net/minecraft/client/sounds/MusicManager.java
+++ net/minecraft/client/sounds/MusicManager.java
@@ -7,10 +7,11 @@
 import net.minecraft.client.resources.sounds.SoundInstance;
 import net.minecraft.sounds.Music;
 import net.minecraft.util.Mth;
 
 public class MusicManager {
+   private static final int STARTING_DELAY = 100;
    private final Random random = new Random();
    private final Minecraft minecraft;
    @Nullable
    private SoundInstance currentMusic;
    private int nextSongDelay = 100;
