--- net/minecraft/client/sounds/SoundManager.java
+++ net/minecraft/client/sounds/SoundManager.java
@@ -116,11 +116,11 @@
                         }
                      }
 
                   }
                } catch (RuntimeException var45) {
-                  LOGGER.warn("Invalid sounds.json in resourcepack: '{}'", var8.getSourceName(), var45);
+                  LOGGER.warn("Invalid {} in resourcepack: '{}'", "sounds.json", var8.getSourceName(), var45);
                }
             }
          } catch (IOException var46) {
          }
       }
