--- net/minecraft/client/sounds/SoundManager.java
+++ net/minecraft/client/sounds/SoundManager.java
@@ -37,10 +37,11 @@
 import org.apache.logging.log4j.Logger;
 
 public class SoundManager extends SimplePreparableReloadListener<SoundManager.Preparations> {
    public static final Sound EMPTY_SOUND;
    private static final Logger LOGGER;
+   private static final String SOUNDS_PATH = "sounds.json";
    private static final Gson GSON;
    private static final TypeToken<Map<String, SoundEventRegistration>> SOUND_EVENT_REGISTRATION_TYPE;
    private final Map<ResourceLocation, WeighedSoundEvents> registry = Maps.newHashMap();
    private final SoundEngine soundEngine;
 
@@ -116,11 +117,11 @@
                         }
                      }
 
                   }
                } catch (RuntimeException var45) {
-                  LOGGER.warn("Invalid sounds.json in resourcepack: '{}'", var8.getSourceName(), var45);
+                  LOGGER.warn("Invalid {} in resourcepack: '{}'", "sounds.json", var8.getSourceName(), var45);
                }
             }
          } catch (IOException var46) {
          }
       }
