--- net/minecraft/client/sounds/SoundEngine.java
+++ net/minecraft/client/sounds/SoundEngine.java
@@ -34,11 +34,17 @@
 import org.apache.logging.log4j.MarkerManager;
 
 public class SoundEngine {
    private static final Marker MARKER = MarkerManager.getMarker("SOUNDS");
    private static final Logger LOGGER = LogManager.getLogger();
+   private static final float PITCH_MIN = 0.5F;
+   private static final float PITCH_MAX = 2.0F;
+   private static final float VOLUME_MIN = 0.0F;
+   private static final float VOLUME_MAX = 1.0F;
+   private static final int MIN_SOURCE_LIFETIME = 20;
    private static final Set<ResourceLocation> ONLY_WARN_ONCE = Sets.newHashSet();
+   public static final String MISSING_SOUND = "FOR THE DEBUG!";
    private final SoundManager soundManager;
    private final Options options;
    private boolean loaded;
    private final Library library = new Library();
    private final Listener listener;
