--- net/minecraft/world/entity/ai/control/BodyRotationControl.java
+++ net/minecraft/world/entity/ai/control/BodyRotationControl.java
@@ -49,11 +49,11 @@
       float var3 = (float)this.mob.getMaxHeadYRot() * (1.0F - var2);
       this.mob.yBodyRot = Mth.rotateIfNecessary(this.mob.yBodyRot, this.mob.yHeadRot, var3);
    }
 
    private boolean notCarryingMobPassengers() {
-      return this.mob.getPassengers().isEmpty() || !(this.mob.getPassengers().get(0) instanceof Mob);
+      return !(this.mob.getFirstPassenger() instanceof Mob);
    }
 
    private boolean isMoving() {
       double var1 = this.mob.getX() - this.mob.xo;
       double var3 = this.mob.getZ() - this.mob.zo;
