--- com/mojang/bridge/game/GameVersion.java
+++ com/mojang/bridge/game/GameVersion.java
@@ -11,11 +11,19 @@
 
    int getWorldVersion();
 
    int getProtocolVersion();
 
-   int getPackVersion();
+   /** @deprecated */
+   @Deprecated
+   default int getPackVersion() {
+      return this.getPackVersion(PackType.RESOURCE);
+   }
 
+   default int getPackVersion(PackType var1) {
+      return this.getPackVersion();
+   }
+
    Date getBuildTime();
 
    boolean isStable();
 }
