--- net/minecraft/data/structures/NbtToSnbt.java
+++ net/minecraft/data/structures/NbtToSnbt.java
@@ -7,13 +7,12 @@
 import java.util.Iterator;
 import javax.annotation.Nullable;
 import net.minecraft.data.DataGenerator;
 import net.minecraft.data.DataProvider;
 import net.minecraft.data.HashCache;
-import net.minecraft.nbt.CompoundTag;
 import net.minecraft.nbt.NbtIo;
-import net.minecraft.network.chat.Component;
+import net.minecraft.nbt.NbtUtils;
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
 
 public class NbtToSnbt implements DataProvider {
    private static final Logger LOGGER = LogManager.getLogger();
@@ -48,41 +47,39 @@
    }
 
    @Nullable
    public static Path convertStructure(Path var0, String var1, Path var2) {
       try {
-         CompoundTag var3 = NbtIo.readCompressed(Files.newInputStream(var0));
-         Component var4 = var3.getPrettyDisplay("    ", 0);
-         String var5 = var4.getString() + "\n";
-         Path var6 = var2.resolve(var1 + ".snbt");
-         Files.createDirectories(var6.getParent());
-         BufferedWriter var7 = Files.newBufferedWriter(var6);
-         Throwable var8 = null;
+         writeSnbt(var2.resolve(var1 + ".snbt"), NbtUtils.structureToSnbt(NbtIo.readCompressed(Files.newInputStream(var0))));
+         LOGGER.info("Converted {} from NBT to SNBT", var1);
+         return var2.resolve(var1 + ".snbt");
+      } catch (IOException var4) {
+         LOGGER.error("Couldn't convert {} from NBT to SNBT at {}", var1, var0, var4);
+         return null;
+      }
+   }
 
-         try {
-            var7.write(var5);
-         } catch (Throwable var18) {
-            var8 = var18;
-            throw var18;
-         } finally {
-            if (var7 != null) {
-               if (var8 != null) {
-                  try {
-                     var7.close();
-                  } catch (Throwable var17) {
-                     var8.addSuppressed(var17);
-                  }
-               } else {
-                  var7.close();
-               }
-            }
+   public static void writeSnbt(Path var0, String var1) throws IOException {
+      Files.createDirectories(var0.getParent());
+      BufferedWriter var2 = Files.newBufferedWriter(var0);
 
+      try {
+         var2.write(var1);
+         var2.write(10);
+      } catch (Throwable var6) {
+         if (var2 != null) {
+            try {
+               var2.close();
+            } catch (Throwable var5) {
+               var6.addSuppressed(var5);
+            }
          }
 
-         LOGGER.info("Converted {} from NBT to SNBT", var1);
-         return var6;
-      } catch (IOException var20) {
-         LOGGER.error("Couldn't convert {} from NBT to SNBT at {}", var1, var0, var20);
-         return null;
+         throw var6;
       }
+
+      if (var2 != null) {
+         var2.close();
+      }
+
    }
 }
