--- net/minecraft/data/structures/StructureUpdater.java
+++ net/minecraft/data/structures/StructureUpdater.java
@@ -13,13 +13,17 @@
 
    public StructureUpdater() {
    }
 
    public CompoundTag apply(String var1, CompoundTag var2) {
-      return var1.startsWith("data/minecraft/structures/") ? updateStructure(var1, patchVersion(var2)) : var2;
+      return var1.startsWith("data/minecraft/structures/") ? update(var1, var2) : var2;
    }
 
+   public static CompoundTag update(String var0, CompoundTag var1) {
+      return updateStructure(var0, patchVersion(var1));
+   }
+
    private static CompoundTag patchVersion(CompoundTag var0) {
       if (!var0.contains("DataVersion", 99)) {
          var0.putInt("DataVersion", 500);
       }
 
@@ -28,12 +32,12 @@
 
    private static CompoundTag updateStructure(String var0, CompoundTag var1) {
       StructureTemplate var2 = new StructureTemplate();
       int var3 = var1.getInt("DataVersion");
       boolean var4 = true;
-      if (var3 < 2532) {
-         LOGGER.warn("SNBT Too old, do not forget to update: " + var3 + " < " + 2532 + ": " + var0);
+      if (var3 < 2678) {
+         LOGGER.warn("SNBT Too old, do not forget to update: {} < {}: {}", var3, 2678, var0);
       }
 
       CompoundTag var5 = NbtUtils.update(DataFixers.getDataFixer(), DataFixTypes.STRUCTURE, var1, var3);
       var2.load(var5);
       return var2.save(new CompoundTag());
