--- net/minecraft/world/level/levelgen/structure/templatesystem/BlackstoneReplaceProcessor.java
+++ net/minecraft/world/level/levelgen/structure/templatesystem/BlackstoneReplaceProcessor.java
@@ -3,16 +3,19 @@
 import com.google.common.collect.Maps;
 import com.mojang.serialization.Codec;
 import java.util.Map;
 import net.minecraft.Util;
 import net.minecraft.core.BlockPos;
+import net.minecraft.core.Direction;
 import net.minecraft.world.level.LevelReader;
 import net.minecraft.world.level.block.Block;
 import net.minecraft.world.level.block.Blocks;
 import net.minecraft.world.level.block.SlabBlock;
 import net.minecraft.world.level.block.StairBlock;
 import net.minecraft.world.level.block.state.BlockState;
+import net.minecraft.world.level.block.state.properties.Half;
+import net.minecraft.world.level.block.state.properties.SlabType;
 
 public class BlackstoneReplaceProcessor extends StructureProcessor {
    public static final Codec<BlackstoneReplaceProcessor> CODEC = Codec.unit(() -> {
       return INSTANCE;
    });
@@ -52,19 +55,19 @@
          return var5;
       } else {
          BlockState var8 = var5.state;
          BlockState var9 = var7.defaultBlockState();
          if (var8.hasProperty(StairBlock.FACING)) {
-            var9 = (BlockState)var9.setValue(StairBlock.FACING, var8.getValue(StairBlock.FACING));
+            var9 = (BlockState)var9.setValue(StairBlock.FACING, (Direction)var8.getValue(StairBlock.FACING));
          }
 
          if (var8.hasProperty(StairBlock.HALF)) {
-            var9 = (BlockState)var9.setValue(StairBlock.HALF, var8.getValue(StairBlock.HALF));
+            var9 = (BlockState)var9.setValue(StairBlock.HALF, (Half)var8.getValue(StairBlock.HALF));
          }
 
          if (var8.hasProperty(SlabBlock.TYPE)) {
-            var9 = (BlockState)var9.setValue(SlabBlock.TYPE, var8.getValue(SlabBlock.TYPE));
+            var9 = (BlockState)var9.setValue(SlabBlock.TYPE, (SlabType)var8.getValue(SlabBlock.TYPE));
          }
 
          return new StructureTemplate.StructureBlockInfo(var5.pos, var9, var5.nbt);
       }
    }
