--- net/minecraft/world/level/levelgen/structure/templatesystem/StructureProcessorType.java
+++ net/minecraft/world/level/levelgen/structure/templatesystem/StructureProcessorType.java
@@ -14,10 +14,11 @@
    StructureProcessorType<RuleProcessor> RULE = register("rule", RuleProcessor.CODEC);
    StructureProcessorType<NopProcessor> NOP = register("nop", NopProcessor.CODEC);
    StructureProcessorType<BlockAgeProcessor> BLOCK_AGE = register("block_age", BlockAgeProcessor.CODEC);
    StructureProcessorType<BlackstoneReplaceProcessor> BLACKSTONE_REPLACE = register("blackstone_replace", BlackstoneReplaceProcessor.CODEC);
    StructureProcessorType<LavaSubmergedBlockProcessor> LAVA_SUBMERGED_BLOCK = register("lava_submerged_block", LavaSubmergedBlockProcessor.CODEC);
+   StructureProcessorType<ProtectedBlockProcessor> PROTECTED_BLOCKS = register("protected_blocks", ProtectedBlockProcessor.CODEC);
    Codec<StructureProcessor> SINGLE_CODEC = Registry.STRUCTURE_PROCESSOR.dispatch("processor_type", StructureProcessor::getType, StructureProcessorType::codec);
    Codec<StructureProcessorList> LIST_OBJECT_CODEC = SINGLE_CODEC.listOf().xmap(StructureProcessorList::new, StructureProcessorList::list);
    Codec<StructureProcessorList> DIRECT_CODEC = Codec.either(LIST_OBJECT_CODEC.fieldOf("processors").codec(), LIST_OBJECT_CODEC).xmap((var0) -> {
       return (StructureProcessorList)var0.map((var0x) -> {
          return var0x;
