--- net/minecraft/util/profiling/registry/MeasuredMetric.java
+++ net/minecraft/util/profiling/registry/MeasuredMetric.java
@@ -1,0 +1,27 @@
+package net.minecraft.util.profiling.registry;
+
+import java.util.function.DoubleSupplier;
+
+public class MeasuredMetric {
+   private final Metric metric;
+   private final DoubleSupplier currentValue;
+   private final MeasurementCategory getCategory;
+
+   public MeasuredMetric(Metric var1, DoubleSupplier var2, MeasurementCategory var3) {
+      this.metric = var1;
+      this.currentValue = var2;
+      this.getCategory = var3;
+   }
+
+   public Metric getMetric() {
+      return this.metric;
+   }
+
+   public DoubleSupplier getCurrentValue() {
+      return this.currentValue;
+   }
+
+   public MeasurementCategory getGetCategory() {
+      return this.getCategory;
+   }
+}
