--- net/minecraft/util/profiling/ActiveProfiler.java
+++ net/minecraft/util/profiling/ActiveProfiler.java
@@ -143,17 +143,21 @@
 
    public ProfileResults getResults() {
       return new FilledProfileResults(this.entries, this.startTimeNano, this.startTimeTicks, this.getRealTime.getAsLong(), this.getTickTime.getAsInt());
    }
 
-   static class PathEntry implements ProfilerPathEntry {
+   @Nullable
+   public ActiveProfiler.PathEntry getEntry(String var1) {
+      return (ActiveProfiler.PathEntry)this.entries.get(var1);
+   }
+
+   public static class PathEntry implements ProfilerPathEntry {
       private long duration;
       private long count;
-      private Object2LongOpenHashMap<String> counters;
+      private final Object2LongOpenHashMap<String> counters = new Object2LongOpenHashMap();
 
-      private PathEntry() {
-         this.counters = new Object2LongOpenHashMap();
+      public PathEntry() {
       }
 
       public long getDuration() {
          return this.duration;
       }
