--- net/minecraft/util/profiling/ProfilerFiller.java
+++ net/minecraft/util/profiling/ProfilerFiller.java
@@ -1,10 +1,13 @@
 package net.minecraft.util.profiling;
 
 import java.util.function.Supplier;
+import net.minecraft.util.profiling.metrics.MetricCategory;
 
 public interface ProfilerFiller {
+   String ROOT = "root";
+
    void startTick();
 
    void endTick();
 
    void push(String var1);
@@ -15,10 +18,12 @@
 
    void popPush(String var1);
 
    void popPush(Supplier<String> var1);
 
+   void markForCharting(MetricCategory var1);
+
    void incrementCounter(String var1);
 
    void incrementCounter(Supplier<String> var1);
 
    static ProfilerFiller tee(final ProfilerFiller var0, final ProfilerFiller var1) {
@@ -44,10 +49,15 @@
             public void push(Supplier<String> var1x) {
                var0.push(var1x);
                var1.push(var1x);
             }
 
+            public void markForCharting(MetricCategory var1x) {
+               var0.markForCharting(var1x);
+               var1.markForCharting(var1x);
+            }
+
             public void pop() {
                var0.pop();
                var1.pop();
             }
 
