--- net/minecraft/data/worldgen/biome/Biomes.java
+++ net/minecraft/data/worldgen/biome/Biomes.java
@@ -10,10 +10,13 @@
 public abstract class Biomes {
    private static final Int2ObjectMap<ResourceKey<Biome>> TO_NAME = new Int2ObjectArrayMap();
    public static final Biome PLAINS;
    public static final Biome THE_VOID;
 
+   public Biomes() {
+   }
+
    private static Biome register(int var0, ResourceKey<Biome> var1, Biome var2) {
       TO_NAME.put(var0, var1);
       return (Biome)BuiltinRegistries.registerMapping(BuiltinRegistries.BIOME, var0, var1, var2);
    }
 
@@ -99,7 +102,9 @@
       register(169, net.minecraft.world.level.biome.Biomes.BAMBOO_JUNGLE_HILLS, VanillaBiomes.bambooJungleHillsBiome());
       register(170, net.minecraft.world.level.biome.Biomes.SOUL_SAND_VALLEY, VanillaBiomes.soulSandValleyBiome());
       register(171, net.minecraft.world.level.biome.Biomes.CRIMSON_FOREST, VanillaBiomes.crimsonForestBiome());
       register(172, net.minecraft.world.level.biome.Biomes.WARPED_FOREST, VanillaBiomes.warpedForestBiome());
       register(173, net.minecraft.world.level.biome.Biomes.BASALT_DELTAS, VanillaBiomes.basaltDeltasBiome());
+      register(174, net.minecraft.world.level.biome.Biomes.DRIPSTONE_CAVES, VanillaBiomes.dripstoneCaves());
+      register(175, net.minecraft.world.level.biome.Biomes.LUSH_CAVES, VanillaBiomes.lushCaves());
    }
 }
