--- net/minecraft/data/worldgen/biome/VanillaBiomes.java
+++ net/minecraft/data/worldgen/biome/VanillaBiomes.java
@@ -46,10 +46,11 @@
       BiomeGenerationSettings.Builder var5 = (new BiomeGenerationSettings.Builder()).surfaceBuilder(SurfaceBuilders.GIANT_TREE_TAIGA);
       BiomeDefaultFeatures.addDefaultOverworldLandStructures(var5);
       var5.addStructureStart(StructureFeatures.RUINED_PORTAL_STANDARD);
       BiomeDefaultFeatures.addDefaultCarvers(var5);
       BiomeDefaultFeatures.addDefaultLakes(var5);
+      BiomeDefaultFeatures.addDefaultCrystalFormations(var5);
       BiomeDefaultFeatures.addDefaultMonsterRoom(var5);
       BiomeDefaultFeatures.addMossyStoneBlock(var5);
       BiomeDefaultFeatures.addFerns(var5);
       BiomeDefaultFeatures.addDefaultUndergroundVariety(var5);
       BiomeDefaultFeatures.addDefaultOres(var5);
@@ -72,10 +73,11 @@
       BiomeGenerationSettings.Builder var4 = (new BiomeGenerationSettings.Builder()).surfaceBuilder(SurfaceBuilders.GRASS);
       BiomeDefaultFeatures.addDefaultOverworldLandStructures(var4);
       var4.addStructureStart(StructureFeatures.RUINED_PORTAL_STANDARD);
       BiomeDefaultFeatures.addDefaultCarvers(var4);
       BiomeDefaultFeatures.addDefaultLakes(var4);
+      BiomeDefaultFeatures.addDefaultCrystalFormations(var4);
       BiomeDefaultFeatures.addDefaultMonsterRoom(var4);
       BiomeDefaultFeatures.addForestFlowers(var4);
       BiomeDefaultFeatures.addDefaultUndergroundVariety(var4);
       BiomeDefaultFeatures.addDefaultOres(var4);
       BiomeDefaultFeatures.addDefaultSoftDisks(var4);
@@ -152,10 +154,11 @@
 
       BiomeDefaultFeatures.addDefaultOverworldLandStructures(var7);
       var7.addStructureStart(StructureFeatures.RUINED_PORTAL_JUNGLE);
       BiomeDefaultFeatures.addDefaultCarvers(var7);
       BiomeDefaultFeatures.addDefaultLakes(var7);
+      BiomeDefaultFeatures.addDefaultCrystalFormations(var7);
       BiomeDefaultFeatures.addDefaultMonsterRoom(var7);
       BiomeDefaultFeatures.addDefaultUndergroundVariety(var7);
       BiomeDefaultFeatures.addDefaultOres(var7);
       BiomeDefaultFeatures.addDefaultSoftDisks(var7);
       if (var3) {
@@ -190,10 +193,11 @@
       BiomeGenerationSettings.Builder var5 = (new BiomeGenerationSettings.Builder()).surfaceBuilder(var2);
       BiomeDefaultFeatures.addDefaultOverworldLandStructures(var5);
       var5.addStructureStart(StructureFeatures.RUINED_PORTAL_MOUNTAIN);
       BiomeDefaultFeatures.addDefaultCarvers(var5);
       BiomeDefaultFeatures.addDefaultLakes(var5);
+      BiomeDefaultFeatures.addDefaultCrystalFormations(var5);
       BiomeDefaultFeatures.addDefaultMonsterRoom(var5);
       BiomeDefaultFeatures.addDefaultUndergroundVariety(var5);
       BiomeDefaultFeatures.addDefaultOres(var5);
       BiomeDefaultFeatures.addDefaultSoftDisks(var5);
       if (var3) {
@@ -232,10 +236,11 @@
 
       BiomeDefaultFeatures.addDefaultOverworldLandStructures(var6);
       var6.addStructureStart(StructureFeatures.RUINED_PORTAL_DESERT);
       BiomeDefaultFeatures.addDefaultCarvers(var6);
       BiomeDefaultFeatures.addDesertLakes(var6);
+      BiomeDefaultFeatures.addDefaultCrystalFormations(var6);
       BiomeDefaultFeatures.addDefaultMonsterRoom(var6);
       BiomeDefaultFeatures.addDefaultUndergroundVariety(var6);
       BiomeDefaultFeatures.addDefaultOres(var6);
       BiomeDefaultFeatures.addDefaultSoftDisks(var6);
       BiomeDefaultFeatures.addDefaultFlowers(var6);
@@ -263,10 +268,11 @@
 
       BiomeDefaultFeatures.addDefaultOverworldLandStructures(var2);
       var2.addStructureStart(StructureFeatures.RUINED_PORTAL_STANDARD);
       BiomeDefaultFeatures.addDefaultCarvers(var2);
       BiomeDefaultFeatures.addDefaultLakes(var2);
+      BiomeDefaultFeatures.addDefaultCrystalFormations(var2);
       BiomeDefaultFeatures.addDefaultMonsterRoom(var2);
       BiomeDefaultFeatures.addPlainGrass(var2);
       if (var0) {
          var2.addFeature(GenerationStep.Decoration.VEGETAL_DECORATION, Features.PATCH_SUNFLOWER);
       }
@@ -328,10 +334,11 @@
       BiomeGenerationSettings.Builder var3 = (new BiomeGenerationSettings.Builder()).surfaceBuilder(SurfaceBuilders.MYCELIUM);
       BiomeDefaultFeatures.addDefaultOverworldLandStructures(var3);
       var3.addStructureStart(StructureFeatures.RUINED_PORTAL_STANDARD);
       BiomeDefaultFeatures.addDefaultCarvers(var3);
       BiomeDefaultFeatures.addDefaultLakes(var3);
+      BiomeDefaultFeatures.addDefaultCrystalFormations(var3);
       BiomeDefaultFeatures.addDefaultMonsterRoom(var3);
       BiomeDefaultFeatures.addDefaultUndergroundVariety(var3);
       BiomeDefaultFeatures.addDefaultOres(var3);
       BiomeDefaultFeatures.addDefaultSoftDisks(var3);
       BiomeDefaultFeatures.addMushroomFieldVegetation(var3);
@@ -350,10 +357,11 @@
 
       BiomeDefaultFeatures.addDefaultOverworldLandStructures(var6);
       var6.addStructureStart(var3 ? StructureFeatures.RUINED_PORTAL_MOUNTAIN : StructureFeatures.RUINED_PORTAL_STANDARD);
       BiomeDefaultFeatures.addDefaultCarvers(var6);
       BiomeDefaultFeatures.addDefaultLakes(var6);
+      BiomeDefaultFeatures.addDefaultCrystalFormations(var6);
       BiomeDefaultFeatures.addDefaultMonsterRoom(var6);
       if (!var4) {
          BiomeDefaultFeatures.addSavannaGrass(var6);
       }
 
@@ -402,10 +410,11 @@
       BiomeGenerationSettings.Builder var6 = (new BiomeGenerationSettings.Builder()).surfaceBuilder(var0);
       BiomeDefaultFeatures.addDefaultOverworldLandMesaStructures(var6);
       var6.addStructureStart(var3 ? StructureFeatures.RUINED_PORTAL_MOUNTAIN : StructureFeatures.RUINED_PORTAL_STANDARD);
       BiomeDefaultFeatures.addDefaultCarvers(var6);
       BiomeDefaultFeatures.addDefaultLakes(var6);
+      BiomeDefaultFeatures.addDefaultCrystalFormations(var6);
       BiomeDefaultFeatures.addDefaultMonsterRoom(var6);
       BiomeDefaultFeatures.addDefaultUndergroundVariety(var6);
       BiomeDefaultFeatures.addDefaultOres(var6);
       BiomeDefaultFeatures.addExtraGold(var6);
       BiomeDefaultFeatures.addDefaultSoftDisks(var6);
@@ -457,10 +466,11 @@
       }
 
       var4.addStructureStart(StructureFeatures.RUINED_PORTAL_OCEAN);
       BiomeDefaultFeatures.addOceanCarvers(var4);
       BiomeDefaultFeatures.addDefaultLakes(var4);
+      BiomeDefaultFeatures.addDefaultCrystalFormations(var4);
       BiomeDefaultFeatures.addDefaultMonsterRoom(var4);
       BiomeDefaultFeatures.addDefaultUndergroundVariety(var4);
       BiomeDefaultFeatures.addDefaultOres(var4);
       BiomeDefaultFeatures.addDefaultSoftDisks(var4);
       BiomeDefaultFeatures.addWaterTrees(var4);
@@ -549,10 +559,11 @@
       BiomeDefaultFeatures.addDefaultOverworldOceanStructures(var3);
       var3.addStructureStart(StructureFeatures.RUINED_PORTAL_OCEAN);
       BiomeDefaultFeatures.addOceanCarvers(var3);
       BiomeDefaultFeatures.addDefaultLakes(var3);
       BiomeDefaultFeatures.addIcebergs(var3);
+      BiomeDefaultFeatures.addDefaultCrystalFormations(var3);
       BiomeDefaultFeatures.addDefaultMonsterRoom(var3);
       BiomeDefaultFeatures.addBlueIce(var3);
       BiomeDefaultFeatures.addDefaultUndergroundVariety(var3);
       BiomeDefaultFeatures.addDefaultOres(var3);
       BiomeDefaultFeatures.addDefaultSoftDisks(var3);
@@ -570,10 +581,11 @@
       BiomeGenerationSettings.Builder var4 = (new BiomeGenerationSettings.Builder()).surfaceBuilder(SurfaceBuilders.GRASS);
       BiomeDefaultFeatures.addDefaultOverworldLandStructures(var4);
       var4.addStructureStart(StructureFeatures.RUINED_PORTAL_STANDARD);
       BiomeDefaultFeatures.addDefaultCarvers(var4);
       BiomeDefaultFeatures.addDefaultLakes(var4);
+      BiomeDefaultFeatures.addDefaultCrystalFormations(var4);
       BiomeDefaultFeatures.addDefaultMonsterRoom(var4);
       if (var2) {
          var4.addFeature(GenerationStep.Decoration.VEGETAL_DECORATION, Features.FOREST_FLOWER_VEGETATION_COMMON);
       } else {
          BiomeDefaultFeatures.addForestFlowers(var4);
@@ -638,10 +650,11 @@
 
       BiomeDefaultFeatures.addDefaultOverworldLandStructures(var8);
       var8.addStructureStart(var3 ? StructureFeatures.RUINED_PORTAL_MOUNTAIN : StructureFeatures.RUINED_PORTAL_STANDARD);
       BiomeDefaultFeatures.addDefaultCarvers(var8);
       BiomeDefaultFeatures.addDefaultLakes(var8);
+      BiomeDefaultFeatures.addDefaultCrystalFormations(var8);
       BiomeDefaultFeatures.addDefaultMonsterRoom(var8);
       BiomeDefaultFeatures.addFerns(var8);
       BiomeDefaultFeatures.addDefaultUndergroundVariety(var8);
       BiomeDefaultFeatures.addDefaultOres(var8);
       BiomeDefaultFeatures.addDefaultSoftDisks(var8);
@@ -669,10 +682,11 @@
       var4.addStructureStart(StructureFeatures.WOODLAND_MANSION);
       BiomeDefaultFeatures.addDefaultOverworldLandStructures(var4);
       var4.addStructureStart(StructureFeatures.RUINED_PORTAL_STANDARD);
       BiomeDefaultFeatures.addDefaultCarvers(var4);
       BiomeDefaultFeatures.addDefaultLakes(var4);
+      BiomeDefaultFeatures.addDefaultCrystalFormations(var4);
       BiomeDefaultFeatures.addDefaultMonsterRoom(var4);
       var4.addFeature(GenerationStep.Decoration.VEGETAL_DECORATION, var2 ? Features.DARK_FOREST_VEGETATION_RED : Features.DARK_FOREST_VEGETATION_BROWN);
       BiomeDefaultFeatures.addForestFlowers(var4);
       BiomeDefaultFeatures.addDefaultUndergroundVariety(var4);
       BiomeDefaultFeatures.addDefaultOres(var4);
@@ -702,10 +716,11 @@
       if (!var2) {
          BiomeDefaultFeatures.addFossilDecoration(var4);
       }
 
       BiomeDefaultFeatures.addDefaultLakes(var4);
+      BiomeDefaultFeatures.addDefaultCrystalFormations(var4);
       BiomeDefaultFeatures.addDefaultMonsterRoom(var4);
       BiomeDefaultFeatures.addDefaultUndergroundVariety(var4);
       BiomeDefaultFeatures.addDefaultOres(var4);
       BiomeDefaultFeatures.addSwampClayDisk(var4);
       BiomeDefaultFeatures.addSwampVegetation(var4);
@@ -736,10 +751,11 @@
       }
 
       var5.addStructureStart(var3 ? StructureFeatures.RUINED_PORTAL_MOUNTAIN : StructureFeatures.RUINED_PORTAL_STANDARD);
       BiomeDefaultFeatures.addDefaultCarvers(var5);
       BiomeDefaultFeatures.addDefaultLakes(var5);
+      BiomeDefaultFeatures.addDefaultCrystalFormations(var5);
       BiomeDefaultFeatures.addDefaultMonsterRoom(var5);
       if (var2) {
          var5.addFeature(GenerationStep.Decoration.SURFACE_STRUCTURES, Features.ICE_SPIKE);
          var5.addFeature(GenerationStep.Decoration.SURFACE_STRUCTURES, Features.ICE_PATCH);
       }
@@ -764,10 +780,11 @@
       BiomeGenerationSettings.Builder var6 = (new BiomeGenerationSettings.Builder()).surfaceBuilder(SurfaceBuilders.GRASS);
       var6.addStructureStart(StructureFeatures.MINESHAFT);
       var6.addStructureStart(StructureFeatures.RUINED_PORTAL_STANDARD);
       BiomeDefaultFeatures.addDefaultCarvers(var6);
       BiomeDefaultFeatures.addDefaultLakes(var6);
+      BiomeDefaultFeatures.addDefaultCrystalFormations(var6);
       BiomeDefaultFeatures.addDefaultMonsterRoom(var6);
       BiomeDefaultFeatures.addDefaultUndergroundVariety(var6);
       BiomeDefaultFeatures.addDefaultOres(var6);
       BiomeDefaultFeatures.addDefaultSoftDisks(var6);
       BiomeDefaultFeatures.addWaterTrees(var6);
@@ -801,10 +818,11 @@
       }
 
       var8.addStructureStart(var6 ? StructureFeatures.RUINED_PORTAL_MOUNTAIN : StructureFeatures.RUINED_PORTAL_STANDARD);
       BiomeDefaultFeatures.addDefaultCarvers(var8);
       BiomeDefaultFeatures.addDefaultLakes(var8);
+      BiomeDefaultFeatures.addDefaultCrystalFormations(var8);
       BiomeDefaultFeatures.addDefaultMonsterRoom(var8);
       BiomeDefaultFeatures.addDefaultUndergroundVariety(var8);
       BiomeDefaultFeatures.addDefaultOres(var8);
       BiomeDefaultFeatures.addDefaultSoftDisks(var8);
       BiomeDefaultFeatures.addDefaultFlowers(var8);
