--- net/minecraft/world/level/levelgen/placement/CarvingMaskDecorator.java
+++ net/minecraft/world/level/levelgen/placement/CarvingMaskDecorator.java
@@ -1,9 +1,10 @@
 package net.minecraft.world.level.levelgen.placement;
 
 import com.mojang.serialization.Codec;
 import java.util.BitSet;
+import java.util.Objects;
 import java.util.Random;
 import java.util.stream.IntStream;
 import java.util.stream.Stream;
 import net.minecraft.core.BlockPos;
 import net.minecraft.world.level.ChunkPos;
@@ -14,13 +15,13 @@
    }
 
    public Stream<BlockPos> getPositions(DecorationContext var1, Random var2, CarvingMaskDecoratorConfiguration var3, BlockPos var4) {
       ChunkPos var5 = new ChunkPos(var4);
       BitSet var6 = var1.getCarvingMask(var5, var3.step);
-      return IntStream.range(0, var6.length()).filter((var3x) -> {
-         return var6.get(var3x) && var2.nextFloat() < var3.probability;
-      }).mapToObj((var1x) -> {
+      IntStream var10000 = IntStream.range(0, var6.length());
+      Objects.requireNonNull(var6);
+      return var10000.filter(var6::get).mapToObj((var1x) -> {
          int var2 = var1x & 15;
          int var3 = var1x >> 4 & 15;
          int var4 = var1x >> 8;
          return new BlockPos(var5.getMinBlockX() + var2, var4, var5.getMinBlockZ() + var3);
       });
