--- net/minecraft/world/level/levelgen/placement/SquareDecorator.java
+++ net/minecraft/world/level/levelgen/placement/SquareDecorator.java
@@ -4,17 +4,16 @@
 import java.util.Random;
 import java.util.stream.Stream;
 import net.minecraft.core.BlockPos;
 import net.minecraft.world.level.levelgen.feature.configurations.NoneDecoratorConfiguration;
 
-public class SquareDecorator extends SimpleFeatureDecorator<NoneDecoratorConfiguration> {
+public class SquareDecorator extends FeatureDecorator<NoneDecoratorConfiguration> {
    public SquareDecorator(Codec<NoneDecoratorConfiguration> var1) {
       super(var1);
    }
 
-   public Stream<BlockPos> place(Random var1, NoneDecoratorConfiguration var2, BlockPos var3) {
-      int var4 = var1.nextInt(16) + var3.getX();
-      int var5 = var1.nextInt(16) + var3.getZ();
-      int var6 = var3.getY();
-      return Stream.of(new BlockPos(var4, var6, var5));
+   public Stream<BlockPos> getPositions(DecorationContext var1, Random var2, NoneDecoratorConfiguration var3, BlockPos var4) {
+      int var5 = var2.nextInt(16) + var4.getX();
+      int var6 = var2.nextInt(16) + var4.getZ();
+      return Stream.of(new BlockPos(var5, var4.getY(), var6));
    }
 }
