--- net/minecraft/world/level/material/Material.java
+++ net/minecraft/world/level/material/Material.java
@@ -15,10 +15,11 @@
    public static final Material LAVA;
    public static final Material TOP_SNOW;
    public static final Material FIRE;
    public static final Material DECORATION;
    public static final Material WEB;
+   public static final Material SCULK;
    public static final Material BUILDABLE_GLASS;
    public static final Material CLAY;
    public static final Material DIRT;
    public static final Material GRASS;
    public static final Material ICE_SOLID;
@@ -39,14 +40,16 @@
    public static final Material METAL;
    public static final Material SNOW;
    public static final Material HEAVY_METAL;
    public static final Material BARRIER;
    public static final Material PISTON;
-   public static final Material CORAL;
+   public static final Material MOSS;
    public static final Material VEGETABLE;
    public static final Material EGG;
    public static final Material CAKE;
+   public static final Material AMETHYST;
+   public static final Material POWDER_SNOW;
    private final MaterialColor color;
    private final PushReaction pushReaction;
    private final boolean blocksMotion;
    private final boolean flammable;
    private final boolean liquid;
@@ -112,10 +115,11 @@
       LAVA = (new Material.Builder(MaterialColor.FIRE)).noCollider().notSolidBlocking().nonSolid().destroyOnPush().replaceable().liquid().build();
       TOP_SNOW = (new Material.Builder(MaterialColor.SNOW)).noCollider().notSolidBlocking().nonSolid().destroyOnPush().replaceable().build();
       FIRE = (new Material.Builder(MaterialColor.NONE)).noCollider().notSolidBlocking().nonSolid().destroyOnPush().replaceable().build();
       DECORATION = (new Material.Builder(MaterialColor.NONE)).noCollider().notSolidBlocking().nonSolid().destroyOnPush().build();
       WEB = (new Material.Builder(MaterialColor.WOOL)).noCollider().notSolidBlocking().destroyOnPush().build();
+      SCULK = (new Material.Builder(MaterialColor.COLOR_BLACK)).build();
       BUILDABLE_GLASS = (new Material.Builder(MaterialColor.NONE)).build();
       CLAY = (new Material.Builder(MaterialColor.CLAY)).build();
       DIRT = (new Material.Builder(MaterialColor.DIRT)).build();
       GRASS = (new Material.Builder(MaterialColor.GRASS)).build();
       ICE_SOLID = (new Material.Builder(MaterialColor.ICE)).build();
@@ -136,14 +140,16 @@
       METAL = (new Material.Builder(MaterialColor.METAL)).build();
       SNOW = (new Material.Builder(MaterialColor.SNOW)).build();
       HEAVY_METAL = (new Material.Builder(MaterialColor.METAL)).notPushable().build();
       BARRIER = (new Material.Builder(MaterialColor.NONE)).notPushable().build();
       PISTON = (new Material.Builder(MaterialColor.STONE)).notPushable().build();
-      CORAL = (new Material.Builder(MaterialColor.PLANT)).destroyOnPush().build();
+      MOSS = (new Material.Builder(MaterialColor.PLANT)).destroyOnPush().build();
       VEGETABLE = (new Material.Builder(MaterialColor.PLANT)).destroyOnPush().build();
       EGG = (new Material.Builder(MaterialColor.PLANT)).destroyOnPush().build();
       CAKE = (new Material.Builder(MaterialColor.NONE)).destroyOnPush().build();
+      AMETHYST = (new Material.Builder(MaterialColor.COLOR_PURPLE)).notSolidBlocking().build();
+      POWDER_SNOW = (new Material.Builder(MaterialColor.SNOW)).nonSolid().noCollider().build();
    }
 
    public static class Builder {
       private PushReaction pushReaction;
       private boolean blocksMotion;
