--- net/minecraft/world/level/material/MaterialColor.java
+++ net/minecraft/world/level/material/MaterialColor.java
@@ -71,6 +71,30 @@
          MATERIAL_COLORS[var1] = this;
       } else {
          throw new IndexOutOfBoundsException("Map colour ID must be between 0 and 63 (inclusive)");
       }
    }
+
+   public int calculateRGBColor(int var1) {
+      short var2 = 220;
+      if (var1 == 3) {
+         var2 = 135;
+      }
+
+      if (var1 == 2) {
+         var2 = 255;
+      }
+
+      if (var1 == 1) {
+         var2 = 220;
+      }
+
+      if (var1 == 0) {
+         var2 = 180;
+      }
+
+      int var3 = (this.col >> 16 & 255) * var2 / 255;
+      int var4 = (this.col >> 8 & 255) * var2 / 255;
+      int var5 = (this.col & 255) * var2 / 255;
+      return -16777216 | var5 << 16 | var4 << 8 | var3;
+   }
 }
