--- net/minecraft/world/level/material/Material.java
+++ net/minecraft/world/level/material/Material.java
@@ -43,10 +43,11 @@
    public static final Material PISTON;
    public static final Material CORAL;
    public static final Material VEGETABLE;
    public static final Material EGG;
    public static final Material CAKE;
+   public static final Material AMETHYST;
    private final MaterialColor color;
    private final PushReaction pushReaction;
    private final boolean blocksMotion;
    private final boolean flammable;
    private final boolean liquid;
@@ -140,10 +141,11 @@
       PISTON = (new Material.Builder(MaterialColor.STONE)).notPushable().build();
       CORAL = (new Material.Builder(MaterialColor.PLANT)).destroyOnPush().build();
       VEGETABLE = (new Material.Builder(MaterialColor.PLANT)).destroyOnPush().build();
       EGG = (new Material.Builder(MaterialColor.PLANT)).destroyOnPush().build();
       CAKE = (new Material.Builder(MaterialColor.NONE)).destroyOnPush().build();
+      AMETHYST = (new Material.Builder(MaterialColor.COLOR_PURPLE)).notSolidBlocking().destroyOnPush().build();
    }
 
    public static class Builder {
       private PushReaction pushReaction;
       private boolean blocksMotion;
