--- net/minecraft/world/level/material/WaterFluid.java
+++ net/minecraft/world/level/material/WaterFluid.java
@@ -1,9 +1,12 @@
 package net.minecraft.world.level.material;
 
+import java.util.Optional;
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.Direction;
+import net.minecraft.sounds.SoundEvent;
+import net.minecraft.sounds.SoundEvents;
 import net.minecraft.tags.FluidTags;
 import net.minecraft.world.item.Item;
 import net.minecraft.world.item.Items;
 import net.minecraft.world.level.BlockGetter;
 import net.minecraft.world.level.LevelAccessor;
@@ -34,11 +37,11 @@
    protected boolean canConvertToSource() {
       return true;
    }
 
    protected void beforeDestroyingBlock(LevelAccessor var1, BlockPos var2, BlockState var3) {
-      BlockEntity var4 = var3.getBlock().isEntityBlock() ? var1.getBlockEntity(var2) : null;
+      BlockEntity var4 = var3.hasBlockEntity() ? var1.getBlockEntity(var2) : null;
       Block.dropResources(var3, var1, var2, var4);
    }
 
    public int getSlopeFindDistance(LevelReader var1) {
       return 4;
@@ -66,10 +69,14 @@
 
    protected float getExplosionResistance() {
       return 100.0F;
    }
 
+   public Optional<SoundEvent> getPickupSound() {
+      return Optional.of(SoundEvents.BUCKET_FILL);
+   }
+
    public static class Flowing extends WaterFluid {
       public Flowing() {
       }
 
       protected void createFluidStateDefinition(StateDefinition.Builder<Fluid, FluidState> var1) {
