--- net/minecraft/tags/BlockTags.java
+++ net/minecraft/tags/BlockTags.java
@@ -47,10 +47,11 @@
    public static final Tag.Named<Block> TALL_FLOWERS = bind("tall_flowers");
    public static final Tag.Named<Block> FLOWERS = bind("flowers");
    public static final Tag.Named<Block> PIGLIN_REPELLENTS = bind("piglin_repellents");
    public static final Tag.Named<Block> GOLD_ORES = bind("gold_ores");
    public static final Tag.Named<Block> NON_FLAMMABLE_WOOD = bind("non_flammable_wood");
+   public static final Tag.Named<Block> CANDLES = bind("candles");
    public static final Tag.Named<Block> FLOWER_POTS = bind("flower_pots");
    public static final Tag.Named<Block> ENDERMAN_HOLDABLE = bind("enderman_holdable");
    public static final Tag.Named<Block> ICE = bind("ice");
    public static final Tag.Named<Block> VALID_SPAWN = bind("valid_spawn");
    public static final Tag.Named<Block> IMPERMEABLE = bind("impermeable");
@@ -90,10 +91,13 @@
    public static final Tag.Named<Block> INFINIBURN_OVERWORLD = bind("infiniburn_overworld");
    public static final Tag.Named<Block> INFINIBURN_NETHER = bind("infiniburn_nether");
    public static final Tag.Named<Block> INFINIBURN_END = bind("infiniburn_end");
    public static final Tag.Named<Block> BASE_STONE_OVERWORLD = bind("base_stone_overworld");
    public static final Tag.Named<Block> BASE_STONE_NETHER = bind("base_stone_nether");
+   public static final Tag.Named<Block> CANDLE_CAKES = bind("candle_cakes");
+   public static final Tag.Named<Block> CAULDRONS = bind("cauldrons");
+   public static final Tag.Named<Block> CRYSTAL_SOUND_BLOCKS = bind("crystal_sound_blocks");
 
    private static Tag.Named<Block> bind(String var0) {
       return HELPER.bind(var0);
    }
 
