--- net/minecraft/tags/EntityTypeTags.java
+++ net/minecraft/tags/EntityTypeTags.java
@@ -1,26 +1,36 @@
 package net.minecraft.tags;
 
-import java.util.List;
-import net.minecraft.resources.ResourceLocation;
+import net.minecraft.core.Registry;
 import net.minecraft.world.entity.EntityType;
 
 public final class EntityTypeTags {
-   protected static final StaticTagHelper<EntityType<?>> HELPER = StaticTags.create(new ResourceLocation("entity_type"), TagContainer::getEntityTypes);
-   public static final Tag.Named<EntityType<?>> SKELETONS = bind("skeletons");
-   public static final Tag.Named<EntityType<?>> RAIDERS = bind("raiders");
-   public static final Tag.Named<EntityType<?>> BEEHIVE_INHABITORS = bind("beehive_inhabitors");
-   public static final Tag.Named<EntityType<?>> ARROWS = bind("arrows");
-   public static final Tag.Named<EntityType<?>> IMPACT_PROJECTILES = bind("impact_projectiles");
+   protected static final StaticTagHelper<EntityType<?>> HELPER;
+   public static final Tag.Named<EntityType<?>> SKELETONS;
+   public static final Tag.Named<EntityType<?>> RAIDERS;
+   public static final Tag.Named<EntityType<?>> BEEHIVE_INHABITORS;
+   public static final Tag.Named<EntityType<?>> ARROWS;
+   public static final Tag.Named<EntityType<?>> IMPACT_PROJECTILES;
+   public static final Tag.Named<EntityType<?>> POWDER_SNOW_WALKABLE_MOBS;
+   public static final Tag.Named<EntityType<?>> AXOLOTL_ALWAYS_HOSTILES;
+   public static final Tag.Named<EntityType<?>> AXOLOTL_TEMPTED_HOSTILES;
 
    private static Tag.Named<EntityType<?>> bind(String var0) {
       return HELPER.bind(var0);
    }
 
    public static TagCollection<EntityType<?>> getAllTags() {
       return HELPER.getAllTags();
    }
 
-   public static List<? extends Tag.Named<EntityType<?>>> getWrappers() {
-      return HELPER.getWrappers();
+   static {
+      HELPER = StaticTags.create(Registry.ENTITY_TYPE_REGISTRY, "tags/entity_types");
+      SKELETONS = bind("skeletons");
+      RAIDERS = bind("raiders");
+      BEEHIVE_INHABITORS = bind("beehive_inhabitors");
+      ARROWS = bind("arrows");
+      IMPACT_PROJECTILES = bind("impact_projectiles");
+      POWDER_SNOW_WALKABLE_MOBS = bind("powder_snow_walkable_mobs");
+      AXOLOTL_ALWAYS_HOSTILES = bind("axolotl_always_hostiles");
+      AXOLOTL_TEMPTED_HOSTILES = bind("axolotl_tempted_hostiles");
    }
 }
