--- net/minecraft/tags/Tag.java
+++ net/minecraft/tags/Tag.java
@@ -126,11 +126,11 @@
          var2.addProperty("required", false);
          var1.add((JsonElement)var2);
       }
 
       public String toString() {
-         return this.id.toString() + "?";
+         return this.id + "?";
       }
    }
 
    public static class ElementEntry implements Tag.Entry {
       private final ResourceLocation id;
@@ -290,9 +290,9 @@
       public Tag.Entry getEntry() {
          return this.entry;
       }
 
       public String toString() {
-         return this.entry.toString() + " (from " + this.source + ")";
+         return this.entry + " (from " + this.source + ")";
       }
    }
 }
