--- net/minecraft/client/resources/metadata/animation/AnimationFrame.java
+++ net/minecraft/client/resources/metadata/animation/AnimationFrame.java
@@ -1,8 +1,9 @@
 package net.minecraft.client.resources.metadata.animation;
 
 public class AnimationFrame {
+   public static final int UNKNOWN_FRAME_TIME = -1;
    private final int index;
    private final int time;
 
    public AnimationFrame(int var1) {
       this(var1, -1);
@@ -11,17 +12,13 @@
    public AnimationFrame(int var1, int var2) {
       this.index = var1;
       this.time = var2;
    }
 
-   public boolean isTimeUnknown() {
-      return this.time == -1;
+   public int getTime(int var1) {
+      return this.time == -1 ? var1 : this.time;
    }
 
-   public int getTime() {
-      return this.time;
-   }
-
    public int getIndex() {
       return this.index;
    }
 }
