--- net/minecraft/client/resources/metadata/animation/AnimationFrame.java
+++ net/minecraft/client/resources/metadata/animation/AnimationFrame.java
@@ -11,17 +11,13 @@
    public AnimationFrame(int var1, int var2) {
       this.index = var1;
       this.time = var2;
    }
 
-   public boolean isTimeUnknown() {
-      return this.time == -1;
+   public int getTime(int var1) {
+      return this.time == -1 ? var1 : this.time;
    }
 
-   public int getTime() {
-      return this.time;
-   }
-
    public int getIndex() {
       return this.index;
    }
 }
