--- net/minecraft/client/resources/metadata/animation/VillagerMetaDataSection.java
+++ net/minecraft/client/resources/metadata/animation/VillagerMetaDataSection.java
@@ -4,10 +4,11 @@
 import java.util.Map;
 import java.util.stream.Collectors;
 
 public class VillagerMetaDataSection {
    public static final VillagerMetadataSectionSerializer SERIALIZER = new VillagerMetadataSectionSerializer();
+   public static final String SECTION_NAME = "villager";
    private final VillagerMetaDataSection.Hat hat;
 
    public VillagerMetaDataSection(VillagerMetaDataSection.Hat var1) {
       this.hat = var1;
    }
