--- net/minecraft/world/item/alchemy/PotionBrewing.java
+++ net/minecraft/world/item/alchemy/PotionBrewing.java
@@ -11,10 +11,11 @@
 import net.minecraft.world.item.PotionItem;
 import net.minecraft.world.item.crafting.Ingredient;
 import net.minecraft.world.level.ItemLike;
 
 public class PotionBrewing {
+   public static final int BREWING_TIME_SECONDS = 20;
    private static final List<PotionBrewing.Mix<Potion>> POTION_MIXES = Lists.newArrayList();
    private static final List<PotionBrewing.Mix<Item>> CONTAINER_MIXES = Lists.newArrayList();
    private static final List<Ingredient> ALLOWED_CONTAINERS = Lists.newArrayList();
    private static final Predicate<ItemStack> ALLOWED_CONTAINER = (var0) -> {
       Iterator var1 = ALLOWED_CONTAINERS.iterator();
@@ -29,10 +30,13 @@
       } while(!var2.test(var0));
 
       return true;
    };
 
+   public PotionBrewing() {
+   }
+
    public static boolean isIngredient(ItemStack var0) {
       return isContainerIngredient(var0) || isPotionIngredient(var0);
    }
 
    protected static boolean isContainerIngredient(ItemStack var0) {
@@ -218,14 +222,14 @@
 
    private static void addMix(Potion var0, Item var1, Potion var2) {
       POTION_MIXES.add(new PotionBrewing.Mix(var0, Ingredient.of(var1), var2));
    }
 
-   static class Mix<T> {
-      private final T from;
-      private final Ingredient ingredient;
-      private final T to;
+   private static class Mix<T> {
+      final T from;
+      final Ingredient ingredient;
+      final T to;
 
       public Mix(T var1, Ingredient var2, T var3) {
          this.from = var1;
          this.ingredient = var2;
          this.to = var3;
