--- net/minecraft/world/item/alchemy/Potions.java
+++ net/minecraft/world/item/alchemy/Potions.java
@@ -47,10 +47,13 @@
    public static final Potion LONG_WEAKNESS;
    public static final Potion LUCK;
    public static final Potion SLOW_FALLING;
    public static final Potion LONG_SLOW_FALLING;
 
+   public Potions() {
+   }
+
    private static Potion register(String var0, Potion var1) {
       return (Potion)Registry.register(Registry.POTION, (String)var0, var1);
    }
 
    static {
