--- net/minecraft/world/entity/ai/goal/target/DefendVillageTargetGoal.java
+++ net/minecraft/world/entity/ai/goal/target/DefendVillageTargetGoal.java
@@ -12,11 +12,11 @@
 import net.minecraft.world.phys.AABB;
 
 public class DefendVillageTargetGoal extends TargetGoal {
    private final IronGolem golem;
    private LivingEntity potentialTarget;
-   private final TargetingConditions attackTargeting = (new TargetingConditions()).range(64.0D);
+   private final TargetingConditions attackTargeting = TargetingConditions.forCombat().range(64.0D);
 
    public DefendVillageTargetGoal(IronGolem var1) {
       super(var1, false, true);
       this.golem = var1;
       this.setFlags(EnumSet.of(Goal.Flag.TARGET));
