--- net/minecraft/world/entity/ai/goal/target/NearestAttackableTargetGoal.java
+++ net/minecraft/world/entity/ai/goal/target/NearestAttackableTargetGoal.java
@@ -46,11 +46,13 @@
       return this.mob.getBoundingBox().inflate(var1, 4.0D, var1);
    }
 
    protected void findTarget() {
       if (this.targetType != Player.class && this.targetType != ServerPlayer.class) {
-         this.target = this.mob.level.getNearestLoadedEntity(this.targetType, this.targetConditions, this.mob, this.mob.getX(), this.mob.getEyeY(), this.mob.getZ(), this.getTargetSearchArea(this.getFollowDistance()));
+         this.target = this.mob.level.getNearestEntity(this.mob.level.getEntitiesOfClass(this.targetType, this.getTargetSearchArea(this.getFollowDistance()), (var0) -> {
+            return true;
+         }), this.targetConditions, this.mob, this.mob.getX(), this.mob.getEyeY(), this.mob.getZ());
       } else {
          this.target = this.mob.level.getNearestPlayer(this.targetConditions, this.mob, this.mob.getX(), this.mob.getEyeY(), this.mob.getZ());
       }
 
    }
