--- net/minecraft/world/level/block/state/StateHolder.java
+++ net/minecraft/world/level/block/state/StateHolder.java
@@ -19,10 +19,12 @@
 import java.util.stream.Collectors;
 import javax.annotation.Nullable;
 import net.minecraft.world.level.block.state.properties.Property;
 
 public abstract class StateHolder<O, S> {
+   public static final String NAME_TAG = "Name";
+   public static final String PROPERTIES_TAG = "Properties";
    private static final Function<Entry<Property<?>, Comparable<?>>, String> PROPERTY_ENTRY_TO_STRING_FUNCTION = new Function<Entry<Property<?>, Comparable<?>>, String>() {
       public String apply(@Nullable Entry<Property<?>, Comparable<?>> var1) {
          if (var1 == null) {
             return "<NULL>";
          } else {
