--- net/minecraft/client/gui/components/spectator/package-info.java
+++ net/minecraft/client/gui/components/spectator/package-info.java
@@ -1,0 +1,10 @@
+package net.minecraft.client.gui.components.spectator;
+
+import javax.annotation.ParametersAreNonnullByDefault;
+import net.minecraft.MethodsReturnNonnullByDefault;
+
+// $FF: synthetic class
+@ParametersAreNonnullByDefault
+@MethodsReturnNonnullByDefault
+interface package-info {
+}
