--- net/minecraft/client/gui/components/spectator/SpectatorGui.java
+++ net/minecraft/client/gui/components/spectator/SpectatorGui.java
@@ -56,38 +56,36 @@
          }
       }
    }
 
    protected void renderPage(PoseStack var1, float var2, int var3, int var4, SpectatorPage var5) {
-      RenderSystem.enableRescaleNormal();
       RenderSystem.enableBlend();
       RenderSystem.defaultBlendFunc();
-      RenderSystem.color4f(1.0F, 1.0F, 1.0F, var2);
-      this.minecraft.getTextureManager().bind(WIDGETS_LOCATION);
+      RenderSystem.setShaderColor(1.0F, 1.0F, 1.0F, var2);
+      RenderSystem.setShaderTexture(0, WIDGETS_LOCATION);
       this.blit(var1, var3 - 91, var4, 0, 0, 182, 22);
       if (var5.getSelectedSlot() >= 0) {
          this.blit(var1, var3 - 91 - 1 + var5.getSelectedSlot() * 20, var4 - 1, 0, 22, 24, 22);
       }
 
       for(int var6 = 0; var6 < 9; ++var6) {
          this.renderSlot(var1, var6, this.minecraft.getWindow().getGuiScaledWidth() / 2 - 90 + var6 * 20 + 2, (float)(var4 + 3), var2, var5.getItem(var6));
       }
 
-      RenderSystem.disableRescaleNormal();
       RenderSystem.disableBlend();
    }
 
    private void renderSlot(PoseStack var1, int var2, int var3, float var4, float var5, SpectatorMenuItem var6) {
-      this.minecraft.getTextureManager().bind(SPECTATOR_LOCATION);
+      RenderSystem.setShaderTexture(0, SPECTATOR_LOCATION);
       if (var6 != SpectatorMenu.EMPTY_SLOT) {
          int var7 = (int)(var5 * 255.0F);
-         RenderSystem.pushMatrix();
-         RenderSystem.translatef((float)var3, var4, 0.0F);
+         var1.pushPose();
+         var1.translate((double)var3, (double)var4, 0.0D);
          float var8 = var6.isEnabled() ? 1.0F : 0.25F;
-         RenderSystem.color4f(var8, var8, var8, var5);
+         RenderSystem.setShaderColor(var8, var8, var8, var5);
          var6.renderIcon(var1, var8, var7);
-         RenderSystem.popMatrix();
+         var1.popPose();
          if (var7 > 3 && var6.isEnabled()) {
             Component var9 = this.minecraft.options.keyHotbarSlots[var2].getTranslatedKeyMessage();
             this.minecraft.font.drawShadow(var1, var9, (float)(var3 + 19 - 2 - this.minecraft.font.width((FormattedText)var9)), var4 + 6.0F + 3.0F, 16777215 + (var7 << 24));
          }
       }
@@ -100,16 +98,14 @@
          SpectatorMenuItem var3 = this.menu.getSelectedItem();
          Component var4 = var3 == SpectatorMenu.EMPTY_SLOT ? this.menu.getSelectedCategory().getPrompt() : var3.getName();
          if (var4 != null) {
             int var5 = (this.minecraft.getWindow().getGuiScaledWidth() - this.minecraft.font.width((FormattedText)var4)) / 2;
             int var6 = this.minecraft.getWindow().getGuiScaledHeight() - 35;
-            RenderSystem.pushMatrix();
             RenderSystem.enableBlend();
             RenderSystem.defaultBlendFunc();
             this.minecraft.font.drawShadow(var1, var4, (float)var5, (float)var6, 16777215 + (var2 << 24));
             RenderSystem.disableBlend();
-            RenderSystem.popMatrix();
          }
       }
 
    }
 
