--- net/minecraft/world/level/storage/loot/parameters/LootContextParamSet.java
+++ net/minecraft/world/level/storage/loot/parameters/LootContextParamSet.java
@@ -15,10 +15,14 @@
    private LootContextParamSet(Set<LootContextParam<?>> var1, Set<LootContextParam<?>> var2) {
       this.required = ImmutableSet.copyOf(var1);
       this.all = ImmutableSet.copyOf(Sets.union(var1, var2));
    }
 
+   public boolean isAllowed(LootContextParam<?> var1) {
+      return this.all.contains(var1);
+   }
+
    public Set<LootContextParam<?>> getRequired() {
       return this.required;
    }
 
    public Set<LootContextParam<?>> getAllowed() {
@@ -38,10 +42,14 @@
          var1.reportProblem("Parameters " + var4 + " are not provided in this context");
       }
 
    }
 
+   public static LootContextParamSet.Builder builder() {
+      return new LootContextParamSet.Builder();
+   }
+
    public static class Builder {
       private final Set<LootContextParam<?>> required = Sets.newIdentityHashSet();
       private final Set<LootContextParam<?>> optional = Sets.newIdentityHashSet();
 
       public Builder() {
