--- net/minecraft/world/level/storage/loot/parameters/LootContextParamSets.java
+++ net/minecraft/world/level/storage/loot/parameters/LootContextParamSets.java
@@ -42,10 +42,13 @@
    });
    public static final LootContextParamSet BLOCK = register("block", (var0) -> {
       var0.required(LootContextParams.BLOCK_STATE).required(LootContextParams.ORIGIN).required(LootContextParams.TOOL).optional(LootContextParams.THIS_ENTITY).optional(LootContextParams.BLOCK_ENTITY).optional(LootContextParams.EXPLOSION_RADIUS);
    });
 
+   public LootContextParamSets() {
+   }
+
    private static LootContextParamSet register(String var0, Consumer<LootContextParamSet.Builder> var1) {
       LootContextParamSet.Builder var2 = new LootContextParamSet.Builder();
       var1.accept(var2);
       LootContextParamSet var3 = var2.build();
       ResourceLocation var4 = new ResourceLocation(var0);
