--- net/minecraft/data/loot/LootTableProvider.java
+++ net/minecraft/data/loot/LootTableProvider.java
@@ -68,11 +68,11 @@
          LootTables.validate(var4, var1x, var2x);
       });
       Multimap var8 = var4.getProblems();
       if (!var8.isEmpty()) {
          var8.forEach((var0, var1x) -> {
-            LOGGER.warn("Found validation problem in " + var0 + ": " + var1x);
+            LOGGER.warn((String)"Found validation problem in {}: {}", (Object)var0, (Object)var1x);
          });
          throw new IllegalStateException("Failed to validate loot tables, see logs");
       } else {
          var3.forEach((var2x, var3x) -> {
             Path var4 = createPath(var2, var2x);
