--- net/minecraft/world/entity/ai/attributes/AttributeModifier.java
+++ net/minecraft/world/entity/ai/attributes/AttributeModifier.java
@@ -66,11 +66,12 @@
    public int hashCode() {
       return this.id.hashCode();
    }
 
    public String toString() {
-      return "AttributeModifier{amount=" + this.amount + ", operation=" + this.operation + ", name='" + (String)this.nameGetter.get() + '\'' + ", id=" + this.id + '}';
+      double var10000 = this.amount;
+      return "AttributeModifier{amount=" + var10000 + ", operation=" + this.operation + ", name='" + (String)this.nameGetter.get() + "', id=" + this.id + "}";
    }
 
    public CompoundTag save() {
       CompoundTag var1 = new CompoundTag();
       var1.putString("Name", this.getName());
