--- net/minecraft/world/entity/ai/village/poi/package-info.java
+++ net/minecraft/world/entity/ai/village/poi/package-info.java
@@ -1,0 +1,12 @@
+package net.minecraft.world.entity.ai.village.poi;
+
+import javax.annotation.ParametersAreNonnullByDefault;
+import net.minecraft.FieldsAreNonnullByDefault;
+import net.minecraft.MethodsReturnNonnullByDefault;
+
+// $FF: synthetic class
+@ParametersAreNonnullByDefault
+@MethodsReturnNonnullByDefault
+@FieldsAreNonnullByDefault
+interface package-info {
+}
