--- net/minecraft/world/entity/ai/village/poi/PoiSection.java
+++ net/minecraft/world/entity/ai/village/poi/PoiSection.java
@@ -98,11 +98,11 @@
    }
 
    public void remove(BlockPos var1) {
       PoiRecord var2 = (PoiRecord)this.records.remove(SectionPos.sectionRelativePos(var1));
       if (var2 == null) {
-         LOGGER.error("POI data mismatch: never registered at " + var1);
+         LOGGER.error((String)"POI data mismatch: never registered at {}", (Object)var1);
       } else {
          ((Set)this.byType.get(var2.getPoiType())).remove(var2);
          LOGGER.debug("Removed POI of type {} @ {}", var2::getPoiType, var2::getPos);
          this.setDirty.run();
       }
