--- net/minecraft/world/entity/ai/village/poi/PoiType.java
+++ net/minecraft/world/entity/ai/village/poi/PoiType.java
@@ -53,10 +53,11 @@
    public static final PoiType MEETING;
    public static final PoiType BEEHIVE;
    public static final PoiType BEE_NEST;
    public static final PoiType NETHER_PORTAL;
    public static final PoiType LODESTONE;
+   public static final PoiType LIGHTNING_ROD;
    protected static final Set<BlockState> ALL_STATES;
    private final String name;
    private final Set<BlockState> matchingStates;
    private final int maxTickets;
    private final Predicate<PoiType> predicate;
@@ -148,8 +149,9 @@
       MEETING = register("meeting", getBlockStates(Blocks.BELL), 32, 6);
       BEEHIVE = register("beehive", getBlockStates(Blocks.BEEHIVE), 0, 1);
       BEE_NEST = register("bee_nest", getBlockStates(Blocks.BEE_NEST), 0, 1);
       NETHER_PORTAL = register("nether_portal", getBlockStates(Blocks.NETHER_PORTAL), 0, 1);
       LODESTONE = register("lodestone", getBlockStates(Blocks.LODESTONE), 0, 1);
+      LIGHTNING_ROD = register("lightning_rod", getBlockStates(Blocks.LIGHTNING_ROD), 0, 1);
       ALL_STATES = new ObjectOpenHashSet(TYPE_BY_STATE.keySet());
    }
 }
