--- net/minecraft/world/entity/ai/goal/BegGoal.java
+++ net/minecraft/world/entity/ai/goal/BegGoal.java
@@ -60,11 +60,11 @@
       int var3 = var2.length;
 
       for(int var4 = 0; var4 < var3; ++var4) {
          InteractionHand var5 = var2[var4];
          ItemStack var6 = var1.getItemInHand(var5);
-         if (this.wolf.isTame() && var6.getItem() == Items.BONE) {
+         if (this.wolf.isTame() && var6.is(Items.BONE)) {
             return true;
          }
 
          if (this.wolf.isFood(var6)) {
             return true;
