--- net/minecraft/world/entity/ai/goal/BreathAirGoal.java
+++ net/minecraft/world/entity/ai/goal/BreathAirGoal.java
@@ -35,11 +35,11 @@
    public void start() {
       this.findAirPosition();
    }
 
    private void findAirPosition() {
-      Iterable var1 = BlockPos.betweenClosed(Mth.floor(this.mob.getX() - 1.0D), Mth.floor(this.mob.getY()), Mth.floor(this.mob.getZ() - 1.0D), Mth.floor(this.mob.getX() + 1.0D), Mth.floor(this.mob.getY() + 8.0D), Mth.floor(this.mob.getZ() + 1.0D));
+      Iterable var1 = BlockPos.betweenClosed(Mth.floor(this.mob.getX() - 1.0D), this.mob.getBlockY(), Mth.floor(this.mob.getZ() - 1.0D), Mth.floor(this.mob.getX() + 1.0D), Mth.floor(this.mob.getY() + 8.0D), Mth.floor(this.mob.getZ() + 1.0D));
       BlockPos var2 = null;
       Iterator var3 = var1.iterator();
 
       while(var3.hasNext()) {
          BlockPos var4 = (BlockPos)var3.next();
