--- net/minecraft/world/entity/ai/goal/BreakDoorGoal.java
+++ net/minecraft/world/entity/ai/goal/BreakDoorGoal.java
@@ -5,10 +5,11 @@
 import net.minecraft.world.entity.Mob;
 import net.minecraft.world.level.GameRules;
 import net.minecraft.world.level.block.Block;
 
 public class BreakDoorGoal extends DoorInteractGoal {
+   private static final int DEFAULT_DOOR_BREAK_TIME = 240;
    private final Predicate<Difficulty> validDifficulties;
    protected int breakTime;
    protected int lastBreakProgress;
    protected int doorBreakTime;
 
