--- net/minecraft/world/entity/ai/goal/DoorInteractGoal.java
+++ net/minecraft/world/entity/ai/goal/DoorInteractGoal.java
@@ -41,11 +41,11 @@
 
    protected void setOpen(boolean var1) {
       if (this.hasDoor) {
          BlockState var2 = this.mob.level.getBlockState(this.doorPos);
          if (var2.getBlock() instanceof DoorBlock) {
-            ((DoorBlock)var2.getBlock()).setOpen(this.mob.level, var2, this.doorPos, var1);
+            ((DoorBlock)var2.getBlock()).setOpen(this.mob, this.mob.level, var2, this.doorPos, var1);
          }
       }
 
    }
 
