--- net/minecraft/world/entity/ai/goal/FollowOwnerGoal.java
+++ net/minecraft/world/entity/ai/goal/FollowOwnerGoal.java
@@ -13,10 +13,14 @@
 import net.minecraft.world.level.block.state.BlockState;
 import net.minecraft.world.level.pathfinder.BlockPathTypes;
 import net.minecraft.world.level.pathfinder.WalkNodeEvaluator;
 
 public class FollowOwnerGoal extends Goal {
+   public static final int TELEPORT_WHEN_DISTANCE_IS = 12;
+   private static final int MIN_HORIZONTAL_DISTANCE_FROM_PLAYER_WHEN_TELEPORTING = 2;
+   private static final int MAX_HORIZONTAL_DISTANCE_FROM_PLAYER_WHEN_TELEPORTING = 3;
+   private static final int MAX_VERTICAL_DISTANCE_FROM_PLAYER_WHEN_TELEPORTING = 1;
    private final TamableAnimal tamable;
    private LivingEntity owner;
    private final LevelReader level;
    private final double speedModifier;
    private final PathNavigation navigation;
@@ -112,11 +116,11 @@
       if (Math.abs((double)var1 - this.owner.getX()) < 2.0D && Math.abs((double)var3 - this.owner.getZ()) < 2.0D) {
          return false;
       } else if (!this.canTeleportTo(new BlockPos(var1, var2, var3))) {
          return false;
       } else {
-         this.tamable.moveTo((double)var1 + 0.5D, (double)var2, (double)var3 + 0.5D, this.tamable.yRot, this.tamable.xRot);
+         this.tamable.moveTo((double)var1 + 0.5D, (double)var2, (double)var3 + 0.5D, this.tamable.getYRot(), this.tamable.getXRot());
          this.navigation.stop();
          return true;
       }
    }
 
