--- net/minecraft/world/entity/ai/goal/GolemRandomStrollInVillageGoal.java
+++ net/minecraft/world/entity/ai/goal/GolemRandomStrollInVillageGoal.java
@@ -7,11 +7,11 @@
 import net.minecraft.core.SectionPos;
 import net.minecraft.server.level.ServerLevel;
 import net.minecraft.world.entity.Entity;
 import net.minecraft.world.entity.EntityType;
 import net.minecraft.world.entity.PathfinderMob;
-import net.minecraft.world.entity.ai.util.RandomPos;
+import net.minecraft.world.entity.ai.util.LandRandomPos;
 import net.minecraft.world.entity.ai.village.poi.PoiManager;
 import net.minecraft.world.entity.ai.village.poi.PoiRecord;
 import net.minecraft.world.entity.npc.Villager;
 import net.minecraft.world.phys.Vec3;
 
@@ -43,11 +43,11 @@
       }
    }
 
    @Nullable
    private Vec3 getPositionTowardsAnywhere() {
-      return RandomPos.getLandPos(this.mob, 10, 7);
+      return LandRandomPos.getPos(this.mob, 10, 7);
    }
 
    @Nullable
    private Vec3 getPositionTowardsVillagerWhoWantsGolem() {
       ServerLevel var1 = (ServerLevel)this.mob.level;
@@ -55,22 +55,22 @@
       if (var2.isEmpty()) {
          return null;
       } else {
          Villager var3 = (Villager)var2.get(this.mob.level.random.nextInt(var2.size()));
          Vec3 var4 = var3.position();
-         return RandomPos.getLandPosTowards(this.mob, 10, 7, var4);
+         return LandRandomPos.getPosTowards(this.mob, 10, 7, var4);
       }
    }
 
    @Nullable
    private Vec3 getPositionTowardsPoi() {
       SectionPos var1 = this.getRandomVillageSection();
       if (var1 == null) {
          return null;
       } else {
          BlockPos var2 = this.getRandomPoiWithinSection(var1);
-         return var2 == null ? null : RandomPos.getLandPosTowards(this.mob, 10, 7, Vec3.atBottomCenterOf(var2));
+         return var2 == null ? null : LandRandomPos.getPosTowards(this.mob, 10, 7, Vec3.atBottomCenterOf(var2));
       }
    }
 
    @Nullable
    private SectionPos getRandomVillageSection() {
