--- net/minecraft/world/entity/ai/goal/MoveToBlockGoal.java
+++ net/minecraft/world/entity/ai/goal/MoveToBlockGoal.java
@@ -4,10 +4,13 @@
 import net.minecraft.core.BlockPos;
 import net.minecraft.world.entity.PathfinderMob;
 import net.minecraft.world.level.LevelReader;
 
 public abstract class MoveToBlockGoal extends Goal {
+   private static final int GIVE_UP_TICKS = 1200;
+   private static final int STAY_TICKS = 1200;
+   private static final int INTERVAL_TICKS = 200;
    protected final PathfinderMob mob;
    public final double speedModifier;
    protected int nextStartTick;
    protected int tryTicks;
    private int maxStayTicks;
