--- net/minecraft/world/entity/ai/goal/OfferFlowerGoal.java
+++ net/minecraft/world/entity/ai/goal/OfferFlowerGoal.java
@@ -4,11 +4,12 @@
 import net.minecraft.world.entity.ai.targeting.TargetingConditions;
 import net.minecraft.world.entity.animal.IronGolem;
 import net.minecraft.world.entity.npc.Villager;
 
 public class OfferFlowerGoal extends Goal {
-   private static final TargetingConditions OFFER_TARGER_CONTEXT = (new TargetingConditions()).range(6.0D).allowSameTeam().allowInvulnerable();
+   private static final TargetingConditions OFFER_TARGER_CONTEXT = TargetingConditions.forNonCombat().range(6.0D);
+   public static final int OFFER_TICKS = 400;
    private final IronGolem golem;
    private Villager villager;
    private int tick;
 
    public OfferFlowerGoal(IronGolem var1) {
