--- net/minecraft/world/entity/ai/goal/RemoveBlockGoal.java
+++ net/minecraft/world/entity/ai/goal/RemoveBlockGoal.java
@@ -1,10 +1,11 @@
 package net.minecraft.world.entity.ai.goal;
 
 import java.util.Random;
 import javax.annotation.Nullable;
 import net.minecraft.core.BlockPos;
+import net.minecraft.core.SectionPos;
 import net.minecraft.core.particles.ItemParticleOption;
 import net.minecraft.core.particles.ParticleTypes;
 import net.minecraft.server.level.ServerLevel;
 import net.minecraft.world.entity.Mob;
 import net.minecraft.world.entity.PathfinderMob;
@@ -22,10 +23,11 @@
 
 public class RemoveBlockGoal extends MoveToBlockGoal {
    private final Block blockToRemove;
    private final Mob removerMob;
    private int ticksSinceReachedGoal;
+   private static final int WAIT_AFTER_BLOCK_FOUND = 20;
 
    public RemoveBlockGoal(Block var1, PathfinderMob var2, double var3, int var5) {
       super(var2, var3, 24, var5);
       this.blockToRemove = var1;
       this.removerMob = var2;
@@ -130,11 +132,11 @@
          return null;
       }
    }
 
    protected boolean isValidTarget(LevelReader var1, BlockPos var2) {
-      ChunkAccess var3 = var1.getChunk(var2.getX() >> 4, var2.getZ() >> 4, ChunkStatus.FULL, false);
+      ChunkAccess var3 = var1.getChunk(SectionPos.blockToSectionCoord(var2.getX()), SectionPos.blockToSectionCoord(var2.getZ()), ChunkStatus.FULL, false);
       if (var3 == null) {
          return false;
       } else {
          return var3.getBlockState(var2).is(this.blockToRemove) && var3.getBlockState(var2.above()).isAir() && var3.getBlockState(var2.above(2)).isAir();
       }
