--- net/minecraft/world/entity/ai/goal/StrollThroughVillageGoal.java
+++ net/minecraft/world/entity/ai/goal/StrollThroughVillageGoal.java
@@ -6,15 +6,16 @@
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.SectionPos;
 import net.minecraft.server.level.ServerLevel;
 import net.minecraft.world.entity.PathfinderMob;
 import net.minecraft.world.entity.ai.navigation.PathNavigation;
-import net.minecraft.world.entity.ai.util.RandomPos;
+import net.minecraft.world.entity.ai.util.LandRandomPos;
 import net.minecraft.world.level.levelgen.Heightmap;
 import net.minecraft.world.phys.Vec3;
 
 public class StrollThroughVillageGoal extends Goal {
+   private static final int DISTANCE_THRESHOLD = 10;
    private final PathfinderMob mob;
    private final int interval;
    @Nullable
    private BlockPos wantedPos;
 
@@ -35,11 +36,11 @@
          ServerLevel var1 = (ServerLevel)this.mob.level;
          BlockPos var2 = this.mob.blockPosition();
          if (!var1.isCloseToVillage(var2, 6)) {
             return false;
          } else {
-            Vec3 var3 = RandomPos.getLandPos(this.mob, 15, 7, (var1x) -> {
+            Vec3 var3 = LandRandomPos.getPos(this.mob, 15, 7, (var1x) -> {
                return (double)(-var1.sectionsToVillage(SectionPos.of(var1x)));
             });
             this.wantedPos = var3 == null ? null : new BlockPos(var3);
             return this.wantedPos != null;
          }
