--- net/minecraft/world/entity/ai/goal/TryFindWaterGoal.java
+++ net/minecraft/world/entity/ai/goal/TryFindWaterGoal.java
@@ -17,11 +17,11 @@
       return this.mob.isOnGround() && !this.mob.level.getFluidState(this.mob.blockPosition()).is(FluidTags.WATER);
    }
 
    public void start() {
       BlockPos var1 = null;
-      Iterable var2 = BlockPos.betweenClosed(Mth.floor(this.mob.getX() - 2.0D), Mth.floor(this.mob.getY() - 2.0D), Mth.floor(this.mob.getZ() - 2.0D), Mth.floor(this.mob.getX() + 2.0D), Mth.floor(this.mob.getY()), Mth.floor(this.mob.getZ() + 2.0D));
+      Iterable var2 = BlockPos.betweenClosed(Mth.floor(this.mob.getX() - 2.0D), Mth.floor(this.mob.getY() - 2.0D), Mth.floor(this.mob.getZ() - 2.0D), Mth.floor(this.mob.getX() + 2.0D), this.mob.getBlockY(), Mth.floor(this.mob.getZ() + 2.0D));
       Iterator var3 = var2.iterator();
 
       while(var3.hasNext()) {
          BlockPos var4 = (BlockPos)var3.next();
          if (this.mob.level.getFluidState(var4).is(FluidTags.WATER)) {
