--- net/minecraft/world/level/block/state/properties/EnumProperty.java
+++ net/minecraft/world/level/block/state/properties/EnumProperty.java
@@ -37,11 +37,11 @@
    public Collection<T> getPossibleValues() {
       return this.values;
    }
 
    public Optional<T> getValue(String var1) {
-      return Optional.ofNullable(this.names.get(var1));
+      return Optional.ofNullable((Enum)this.names.get(var1));
    }
 
    public String getName(T var1) {
       return ((StringRepresentable)var1).getSerializedName();
    }
@@ -67,11 +67,11 @@
    public static <T extends Enum<T> & StringRepresentable> EnumProperty<T> create(String var0, Class<T> var1) {
       return create(var0, var1, (Predicate)Predicates.alwaysTrue());
    }
 
    public static <T extends Enum<T> & StringRepresentable> EnumProperty<T> create(String var0, Class<T> var1, Predicate<T> var2) {
-      return create(var0, var1, (Collection)Arrays.stream(var1.getEnumConstants()).filter(var2).collect(Collectors.toList()));
+      return create(var0, var1, (Collection)Arrays.stream((Enum[])var1.getEnumConstants()).filter(var2).collect(Collectors.toList()));
    }
 
    public static <T extends Enum<T> & StringRepresentable> EnumProperty<T> create(String var0, Class<T> var1, T... var2) {
       return create(var0, var1, (Collection)Lists.newArrayList(var2));
    }
