--- net/minecraft/world/level/levelgen/feature/AbstractFlowerFeature.java
+++ net/minecraft/world/level/levelgen/feature/AbstractFlowerFeature.java
@@ -18,11 +18,11 @@
       BlockState var6 = this.getRandomFlower(var3, var4, var5);
       int var7 = 0;
 
       for(int var8 = 0; var8 < this.getCount(var5); ++var8) {
          BlockPos var9 = this.getPos(var3, var4, var5);
-         if (var1.isEmptyBlock(var9) && var9.getY() < 255 && var6.canSurvive(var1, var9) && this.isValid(var1, var9, var5)) {
+         if (var1.isEmptyBlock(var9) && var6.canSurvive(var1, var9) && this.isValid(var1, var9, var5)) {
             var1.setBlock(var9, var6, 2);
             ++var7;
          }
       }
 
