--- net/minecraft/world/level/levelgen/feature/BambooFeature.java
+++ net/minecraft/world/level/levelgen/feature/BambooFeature.java
@@ -38,11 +38,11 @@
                   for(int var12 = var4.getZ() - var10; var12 <= var4.getZ() + var10; ++var12) {
                      int var13 = var11 - var4.getX();
                      int var14 = var12 - var4.getZ();
                      if (var13 * var13 + var14 * var14 <= var10 * var10) {
                         var8.set(var11, var1.getHeight(Heightmap.Types.WORLD_SURFACE, var11, var12) - 1, var12);
-                        if (isDirt(var1.getBlockState(var8).getBlock())) {
+                        if (isDirt(var1.getBlockState(var8))) {
                            var1.setBlock(var8, Blocks.PODZOL.defaultBlockState(), 2);
                         }
                      }
                   }
                }
