--- net/minecraft/world/level/levelgen/feature/BasaltPillarFeature.java
+++ net/minecraft/world/level/levelgen/feature/BasaltPillarFeature.java
@@ -3,11 +3,10 @@
 import com.mojang.serialization.Codec;
 import java.util.Random;
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.Direction;
 import net.minecraft.util.Mth;
-import net.minecraft.world.level.Level;
 import net.minecraft.world.level.LevelAccessor;
 import net.minecraft.world.level.WorldGenLevel;
 import net.minecraft.world.level.block.Blocks;
 import net.minecraft.world.level.chunk.ChunkGenerator;
 import net.minecraft.world.level.levelgen.feature.configurations.NoneFeatureConfiguration;
@@ -25,11 +24,11 @@
          boolean var9 = true;
          boolean var10 = true;
          boolean var11 = true;
 
          while(var1.isEmptyBlock(var6)) {
-            if (Level.isOutsideBuildHeight(var6)) {
+            if (var1.isOutsideBuildHeight(var6)) {
                return true;
             }
 
             var1.setBlock(var6, Blocks.BASALT.defaultBlockState(), 2);
             var8 = var8 && this.placeHangOff(var1, var3, var7.setWithOffset(var6, Direction.NORTH));
