--- net/minecraft/world/level/levelgen/feature/BlockPileFeature.java
+++ net/minecraft/world/level/levelgen/feature/BlockPileFeature.java
@@ -16,11 +16,11 @@
    public BlockPileFeature(Codec<BlockPileConfiguration> var1) {
       super(var1);
    }
 
    public boolean place(WorldGenLevel var1, ChunkGenerator var2, Random var3, BlockPos var4, BlockPileConfiguration var5) {
-      if (var4.getY() < 5) {
+      if (var4.getY() < var1.getMinBuildHeight() + 5) {
          return false;
       } else {
          int var6 = 2 + var3.nextInt(2);
          int var7 = 2 + var3.nextInt(2);
          Iterator var8 = BlockPos.betweenClosed(var4.offset(-var6, 0, -var7), var4.offset(var6, 1, var7)).iterator();
@@ -41,11 +41,11 @@
    }
 
    private boolean mayPlaceOn(LevelAccessor var1, BlockPos var2, Random var3) {
       BlockPos var4 = var2.below();
       BlockState var5 = var1.getBlockState(var4);
-      return var5.is(Blocks.GRASS_PATH) ? var3.nextBoolean() : var5.isFaceSturdy(var1, var4, Direction.UP);
+      return var5.is(Blocks.DIRT_PATH) ? var3.nextBoolean() : var5.isFaceSturdy(var1, var4, Direction.UP);
    }
 
    private void tryPlaceBlock(LevelAccessor var1, BlockPos var2, Random var3, BlockPileConfiguration var4) {
       if (var1.isEmptyBlock(var2) && this.mayPlaceOn(var1, var2, var3)) {
          var1.setBlock(var2, var4.stateProvider.getState(var3, var2), 4);
