--- net/minecraft/world/level/levelgen/feature/ChorusPlantFeature.java
+++ net/minecraft/world/level/levelgen/feature/ChorusPlantFeature.java
@@ -4,21 +4,23 @@
 import java.util.Random;
 import net.minecraft.core.BlockPos;
 import net.minecraft.world.level.WorldGenLevel;
 import net.minecraft.world.level.block.Blocks;
 import net.minecraft.world.level.block.ChorusFlowerBlock;
-import net.minecraft.world.level.chunk.ChunkGenerator;
 import net.minecraft.world.level.levelgen.feature.configurations.NoneFeatureConfiguration;
 
 public class ChorusPlantFeature extends Feature<NoneFeatureConfiguration> {
    public ChorusPlantFeature(Codec<NoneFeatureConfiguration> var1) {
       super(var1);
    }
 
-   public boolean place(WorldGenLevel var1, ChunkGenerator var2, Random var3, BlockPos var4, NoneFeatureConfiguration var5) {
-      if (var1.isEmptyBlock(var4) && var1.getBlockState(var4.below()).is(Blocks.END_STONE)) {
-         ChorusFlowerBlock.generatePlant(var1, var4, var3, 8);
+   public boolean place(FeaturePlaceContext<NoneFeatureConfiguration> var1) {
+      WorldGenLevel var2 = var1.level();
+      BlockPos var3 = var1.origin();
+      Random var4 = var1.random();
+      if (var2.isEmptyBlock(var3) && var2.getBlockState(var3.below()).is(Blocks.END_STONE)) {
+         ChorusFlowerBlock.generatePlant(var2, var3, var4, 8);
          return true;
       } else {
          return false;
       }
    }
