--- net/minecraft/world/level/levelgen/feature/ConfiguredFeature.java
+++ net/minecraft/world/level/levelgen/feature/ConfiguredFeature.java
@@ -47,11 +47,11 @@
    public WeightedConfiguredFeature weighted(float var1) {
       return new WeightedConfiguredFeature(this, var1);
    }
 
    public boolean place(WorldGenLevel var1, ChunkGenerator var2, Random var3, BlockPos var4) {
-      return this.feature.place(var1, var2, var3, var4, this.config);
+      return this.feature.place(new FeaturePlaceContext(var1, var2, var3, var4, this.config));
    }
 
    public Stream<ConfiguredFeature<?, ?>> getFeatures() {
       return Stream.concat(Stream.of(this), this.config.getFeatures());
    }
