--- net/minecraft/world/level/levelgen/feature/CoralFeature.java
+++ net/minecraft/world/level/levelgen/feature/CoralFeature.java
@@ -11,21 +11,23 @@
 import net.minecraft.world.level.block.BaseCoralWallFanBlock;
 import net.minecraft.world.level.block.Block;
 import net.minecraft.world.level.block.Blocks;
 import net.minecraft.world.level.block.SeaPickleBlock;
 import net.minecraft.world.level.block.state.BlockState;
-import net.minecraft.world.level.chunk.ChunkGenerator;
 import net.minecraft.world.level.levelgen.feature.configurations.NoneFeatureConfiguration;
 
 public abstract class CoralFeature extends Feature<NoneFeatureConfiguration> {
    public CoralFeature(Codec<NoneFeatureConfiguration> var1) {
       super(var1);
    }
 
-   public boolean place(WorldGenLevel var1, ChunkGenerator var2, Random var3, BlockPos var4, NoneFeatureConfiguration var5) {
-      BlockState var6 = ((Block)BlockTags.CORAL_BLOCKS.getRandomElement(var3)).defaultBlockState();
-      return this.placeFeature(var1, var3, var4, var6);
+   public boolean place(FeaturePlaceContext<NoneFeatureConfiguration> var1) {
+      Random var2 = var1.random();
+      WorldGenLevel var3 = var1.level();
+      BlockPos var4 = var1.origin();
+      BlockState var5 = ((Block)BlockTags.CORAL_BLOCKS.getRandomElement(var2)).defaultBlockState();
+      return this.placeFeature(var3, var2, var4, var5);
    }
 
    protected abstract boolean placeFeature(LevelAccessor var1, Random var2, BlockPos var3, BlockState var4);
 
    protected boolean placeCoralBlock(LevelAccessor var1, Random var2, BlockPos var3, BlockState var4) {
