--- net/minecraft/world/level/levelgen/feature/FossilFeature.java
+++ net/minecraft/world/level/levelgen/feature/FossilFeature.java
@@ -47,25 +47,25 @@
       int var7 = var3.nextInt(fossils.length);
       StructureManager var8 = var1.getLevel().getServer().getStructureManager();
       StructureTemplate var9 = var8.getOrCreate(fossils[var7]);
       StructureTemplate var10 = var8.getOrCreate(fossilsCoal[var7]);
       ChunkPos var11 = new ChunkPos(var4);
-      BoundingBox var12 = new BoundingBox(var11.getMinBlockX(), 0, var11.getMinBlockZ(), var11.getMaxBlockX(), 256, var11.getMaxBlockZ());
+      BoundingBox var12 = new BoundingBox(var11.getMinBlockX(), var1.getMinBuildHeight(), var11.getMinBlockZ(), var11.getMaxBlockX(), var1.getMaxBuildHeight(), var11.getMaxBlockZ());
       StructurePlaceSettings var13 = (new StructurePlaceSettings()).setRotation(var6).setBoundingBox(var12).setRandom(var3).addProcessor(BlockIgnoreProcessor.STRUCTURE_AND_AIR);
       BlockPos var14 = var9.getSize(var6);
       int var15 = var3.nextInt(16 - var14.getX());
       int var16 = var3.nextInt(16 - var14.getZ());
-      int var17 = 256;
+      int var17 = var1.getMaxBuildHeight();
 
       int var18;
       for(var18 = 0; var18 < var14.getX(); ++var18) {
          for(int var19 = 0; var19 < var14.getZ(); ++var19) {
             var17 = Math.min(var17, var1.getHeight(Heightmap.Types.OCEAN_FLOOR_WG, var4.getX() + var18 + var15, var4.getZ() + var19 + var16));
          }
       }
 
-      var18 = Math.max(var17 - 15 - var3.nextInt(10), 10);
+      var18 = Math.max(var17 - 15 - var3.nextInt(10), var1.getMinBuildHeight() + 10);
       BlockPos var22 = var9.getZeroPositionWithTransform(var4.offset(var15, var18, var16), Mirror.NONE, var6);
       BlockRotProcessor var20 = new BlockRotProcessor(0.9F);
       var13.clearProcessors().addProcessor(var20);
       var9.placeInWorld(var1, var22, var22, var13, var3, 4);
       var13.popProcessor(var20);
