--- net/minecraft/world/level/levelgen/feature/HugeFungusFeature.java
+++ net/minecraft/world/level/levelgen/feature/HugeFungusFeature.java
@@ -19,12 +19,12 @@
    }
 
    public boolean place(WorldGenLevel var1, ChunkGenerator var2, Random var3, BlockPos var4, HugeFungusConfiguration var5) {
       Block var6 = var5.validBaseState.getBlock();
       BlockPos var7 = null;
-      Block var8 = var1.getBlockState(var4.below()).getBlock();
-      if (var8 == var6) {
+      BlockState var8 = var1.getBlockState(var4.below());
+      if (var8.is(var6)) {
          var7 = var4;
       }
 
       if (var7 == null) {
          return false;
