--- net/minecraft/world/level/levelgen/feature/MonsterRoomFeature.java
+++ net/minecraft/world/level/levelgen/feature/MonsterRoomFeature.java
@@ -75,11 +75,11 @@
                   BlockState var27 = var1.getBlockState(var19);
                   if (var16 != var8 && var17 != -1 && var18 != var13 && var16 != var9 && var17 != 4 && var18 != var14) {
                      if (!var27.is(Blocks.CHEST) && !var27.is(Blocks.SPAWNER)) {
                         var1.setBlock(var19, AIR, 2);
                      }
-                  } else if (var19.getY() >= 0 && !var1.getBlockState(var19.below()).getMaterial().isSolid()) {
+                  } else if (var19.getY() >= var1.getMinBuildHeight() && !var1.getBlockState(var19.below()).getMaterial().isSolid()) {
                      var1.setBlock(var19, AIR, 2);
                   } else if (var27.getMaterial().isSolid() && !var27.is(Blocks.CHEST)) {
                      if (var17 == -1 && var3.nextInt(4) != 0) {
                         var1.setBlock(var19, Blocks.MOSSY_COBBLESTONE.defaultBlockState(), 2);
                      } else {
