--- net/minecraft/world/level/levelgen/feature/RandomBooleanSelectorFeature.java
+++ net/minecraft/world/level/levelgen/feature/RandomBooleanSelectorFeature.java
@@ -10,10 +10,15 @@
 public class RandomBooleanSelectorFeature extends Feature<RandomBooleanFeatureConfiguration> {
    public RandomBooleanSelectorFeature(Codec<RandomBooleanFeatureConfiguration> var1) {
       super(var1);
    }
 
-   public boolean place(WorldGenLevel var1, ChunkGenerator var2, Random var3, BlockPos var4, RandomBooleanFeatureConfiguration var5) {
-      boolean var6 = var3.nextBoolean();
-      return var6 ? ((ConfiguredFeature)var5.featureTrue.get()).place(var1, var2, var3, var4) : ((ConfiguredFeature)var5.featureFalse.get()).place(var1, var2, var3, var4);
+   public boolean place(FeaturePlaceContext<RandomBooleanFeatureConfiguration> var1) {
+      Random var2 = var1.random();
+      RandomBooleanFeatureConfiguration var3 = (RandomBooleanFeatureConfiguration)var1.config();
+      WorldGenLevel var4 = var1.level();
+      ChunkGenerator var5 = var1.chunkGenerator();
+      BlockPos var6 = var1.origin();
+      boolean var7 = var2.nextBoolean();
+      return var7 ? ((ConfiguredFeature)var3.featureTrue.get()).place(var4, var5, var2, var6) : ((ConfiguredFeature)var3.featureFalse.get()).place(var4, var5, var2, var6);
    }
 }
