--- net/minecraft/world/level/levelgen/feature/StructureFeature.java
+++ net/minecraft/world/level/levelgen/feature/StructureFeature.java
@@ -16,10 +16,11 @@
 import net.minecraft.core.SectionPos;
 import net.minecraft.nbt.CompoundTag;
 import net.minecraft.nbt.ListTag;
 import net.minecraft.resources.ResourceLocation;
 import net.minecraft.world.level.ChunkPos;
+import net.minecraft.world.level.LevelHeightAccessor;
 import net.minecraft.world.level.LevelReader;
 import net.minecraft.world.level.StructureFeatureManager;
 import net.minecraft.world.level.biome.Biome;
 import net.minecraft.world.level.biome.BiomeSource;
 import net.minecraft.world.level.biome.MobSpawnSettings;
@@ -150,12 +151,12 @@
    }
 
    @Nullable
    public BlockPos getNearestGeneratedFeature(LevelReader var1, StructureFeatureManager var2, BlockPos var3, int var4, boolean var5, long var6, StructureFeatureConfiguration var8) {
       int var9 = var8.spacing();
-      int var10 = var3.getX() >> 4;
-      int var11 = var3.getZ() >> 4;
+      int var10 = SectionPos.blockToSectionCoord(var3.getX());
+      int var11 = SectionPos.blockToSectionCoord(var3.getZ());
       int var12 = 0;
 
       for(WorldgenRandom var13 = new WorldgenRandom(); var12 <= var4; ++var12) {
          for(int var14 = -var12; var14 <= var12; ++var14) {
             boolean var15 = var14 == -var12 || var14 == var12;
@@ -164,20 +165,23 @@
                boolean var17 = var16 == -var12 || var16 == var12;
                if (var15 || var17) {
                   int var18 = var10 + var9 * var14;
                   int var19 = var11 + var9 * var16;
                   ChunkPos var20 = this.getPotentialFeatureChunk(var8, var6, var13, var18, var19);
-                  ChunkAccess var21 = var1.getChunk(var20.x, var20.z, ChunkStatus.STRUCTURE_STARTS);
-                  StructureStart var22 = var2.getStartForFeature(SectionPos.of(var21.getPos(), 0), this, var21);
-                  if (var22 != null && var22.isValid()) {
-                     if (var5 && var22.canBeReferenced()) {
-                        var22.addReference();
-                        return var22.getLocatePos();
-                     }
+                  boolean var21 = var1.getBiomeManager().getPrimaryBiomeAtChunk(var20).getGenerationSettings().isValidStart(this);
+                  if (var21) {
+                     ChunkAccess var22 = var1.getChunk(var20.x, var20.z, ChunkStatus.STRUCTURE_STARTS);
+                     StructureStart var23 = var2.getStartForFeature(SectionPos.bottomOf(var22), this, var22);
+                     if (var23 != null && var23.isValid()) {
+                        if (var5 && var23.canBeReferenced()) {
+                           var23.addReference();
+                           return var23.getLocatePos();
+                        }
 
-                     if (!var5) {
-                        return var22.getLocatePos();
+                        if (!var5) {
+                           return var23.getLocatePos();
+                        }
                      }
                   }
 
                   if (var12 == 0) {
                      break;
@@ -215,25 +219,25 @@
       }
 
       return new ChunkPos(var9 * var7 + var11, var10 * var7 + var12);
    }
 
-   protected boolean isFeatureChunk(ChunkGenerator var1, BiomeSource var2, long var3, WorldgenRandom var5, int var6, int var7, Biome var8, ChunkPos var9, C var10) {
+   protected boolean isFeatureChunk(ChunkGenerator var1, BiomeSource var2, long var3, WorldgenRandom var5, int var6, int var7, Biome var8, ChunkPos var9, C var10, LevelHeightAccessor var11) {
       return true;
    }
 
    private StructureStart<C> createStart(int var1, int var2, BoundingBox var3, int var4, long var5) {
       return this.getStartFactory().create(this, var1, var2, var3, var4, var5);
    }
 
-   public StructureStart<?> generate(RegistryAccess var1, ChunkGenerator var2, BiomeSource var3, StructureManager var4, long var5, ChunkPos var7, Biome var8, int var9, WorldgenRandom var10, StructureFeatureConfiguration var11, C var12) {
-      ChunkPos var13 = this.getPotentialFeatureChunk(var11, var5, var10, var7.x, var7.z);
-      if (var7.x == var13.x && var7.z == var13.z && this.isFeatureChunk(var2, var3, var5, var10, var7.x, var7.z, var8, var13, var12)) {
-         StructureStart var14 = this.createStart(var7.x, var7.z, BoundingBox.getUnknownBox(), var9, var5);
-         var14.generatePieces(var1, var2, var4, var7.x, var7.z, var8, var12);
-         if (var14.isValid()) {
-            return var14;
+   public StructureStart<?> generate(RegistryAccess var1, ChunkGenerator var2, BiomeSource var3, StructureManager var4, long var5, ChunkPos var7, Biome var8, int var9, WorldgenRandom var10, StructureFeatureConfiguration var11, C var12, LevelHeightAccessor var13) {
+      ChunkPos var14 = this.getPotentialFeatureChunk(var11, var5, var10, var7.x, var7.z);
+      if (var7.x == var14.x && var7.z == var14.z && this.isFeatureChunk(var2, var3, var5, var10, var7.x, var7.z, var8, var14, var12, var13)) {
+         StructureStart var15 = this.createStart(var7.x, var7.z, BoundingBox.getUnknownBox(), var9, var5);
+         var15.generatePieces(var1, var2, var4, var7.x, var7.z, var8, var12, var13);
+         if (var15.isValid()) {
+            return var15;
          }
       }
 
       return StructureStart.INVALID_START;
    }
