--- net/minecraft/world/level/levelgen/feature/WoodlandMansionFeature.java
+++ net/minecraft/world/level/levelgen/feature/WoodlandMansionFeature.java
@@ -7,10 +7,11 @@
 import java.util.Random;
 import java.util.Set;
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.RegistryAccess;
 import net.minecraft.world.level.ChunkPos;
+import net.minecraft.world.level.LevelHeightAccessor;
 import net.minecraft.world.level.StructureFeatureManager;
 import net.minecraft.world.level.WorldGenLevel;
 import net.minecraft.world.level.biome.Biome;
 import net.minecraft.world.level.biome.BiomeSource;
 import net.minecraft.world.level.block.Blocks;
@@ -32,12 +33,12 @@
 
    protected boolean linearSeparation() {
       return false;
    }
 
-   protected boolean isFeatureChunk(ChunkGenerator var1, BiomeSource var2, long var3, WorldgenRandom var5, int var6, int var7, Biome var8, ChunkPos var9, NoneFeatureConfiguration var10) {
-      Set var11 = var2.getBiomesWithin(var6 * 16 + 9, var1.getSeaLevel(), var7 * 16 + 9, 32);
+   protected boolean isFeatureChunk(ChunkGenerator var1, BiomeSource var2, long var3, WorldgenRandom var5, ChunkPos var6, Biome var7, ChunkPos var8, NoneFeatureConfiguration var9, LevelHeightAccessor var10) {
+      Set var11 = var2.getBiomesWithin(var6.getBlockX(9), var1.getSeaLevel(), var6.getBlockZ(9), 32);
       Iterator var12 = var11.iterator();
 
       Biome var13;
       do {
          if (!var12.hasNext()) {
@@ -53,15 +54,15 @@
    public StructureFeature.StructureStartFactory<NoneFeatureConfiguration> getStartFactory() {
       return WoodlandMansionFeature.WoodlandMansionStart::new;
    }
 
    public static class WoodlandMansionStart extends StructureStart<NoneFeatureConfiguration> {
-      public WoodlandMansionStart(StructureFeature<NoneFeatureConfiguration> var1, int var2, int var3, BoundingBox var4, int var5, long var6) {
-         super(var1, var2, var3, var4, var5, var6);
+      public WoodlandMansionStart(StructureFeature<NoneFeatureConfiguration> var1, ChunkPos var2, BoundingBox var3, int var4, long var5) {
+         super(var1, var2, var3, var4, var5);
       }
 
-      public void generatePieces(RegistryAccess var1, ChunkGenerator var2, StructureManager var3, int var4, int var5, Biome var6, NoneFeatureConfiguration var7) {
+      public void generatePieces(RegistryAccess var1, ChunkGenerator var2, StructureManager var3, ChunkPos var4, Biome var5, NoneFeatureConfiguration var6, LevelHeightAccessor var7) {
          Rotation var8 = Rotation.getRandom(this.random);
          byte var9 = 5;
          byte var10 = 5;
          if (var8 == Rotation.CLOCKWISE_90) {
             var9 = -5;
@@ -70,19 +71,19 @@
             var10 = -5;
          } else if (var8 == Rotation.COUNTERCLOCKWISE_90) {
             var10 = -5;
          }
 
-         int var11 = (var4 << 4) + 7;
-         int var12 = (var5 << 4) + 7;
-         int var13 = var2.getFirstOccupiedHeight(var11, var12, Heightmap.Types.WORLD_SURFACE_WG);
-         int var14 = var2.getFirstOccupiedHeight(var11, var12 + var10, Heightmap.Types.WORLD_SURFACE_WG);
-         int var15 = var2.getFirstOccupiedHeight(var11 + var9, var12, Heightmap.Types.WORLD_SURFACE_WG);
-         int var16 = var2.getFirstOccupiedHeight(var11 + var9, var12 + var10, Heightmap.Types.WORLD_SURFACE_WG);
+         int var11 = var4.getBlockX(7);
+         int var12 = var4.getBlockZ(7);
+         int var13 = var2.getFirstOccupiedHeight(var11, var12, Heightmap.Types.WORLD_SURFACE_WG, var7);
+         int var14 = var2.getFirstOccupiedHeight(var11, var12 + var10, Heightmap.Types.WORLD_SURFACE_WG, var7);
+         int var15 = var2.getFirstOccupiedHeight(var11 + var9, var12, Heightmap.Types.WORLD_SURFACE_WG, var7);
+         int var16 = var2.getFirstOccupiedHeight(var11 + var9, var12 + var10, Heightmap.Types.WORLD_SURFACE_WG, var7);
          int var17 = Math.min(Math.min(var13, var14), Math.min(var15, var16));
          if (var17 >= 60) {
-            BlockPos var18 = new BlockPos(var4 * 16 + 8, var17 + 1, var5 * 16 + 8);
+            BlockPos var18 = new BlockPos(var4.getBlockX(8), var17 + 1, var4.getBlockZ(8));
             LinkedList var19 = Lists.newLinkedList();
             WoodlandMansionPieces.generateMansion(var3, var18, var8, var19, this.random);
             this.pieces.addAll(var19);
             this.calculateBoundingBox();
          }
