--- net/minecraft/network/syncher/EntityDataSerializers.java
+++ net/minecraft/network/syncher/EntityDataSerializers.java
@@ -67,11 +67,11 @@
       public void write(FriendlyByteBuf var1, String var2) {
          var1.writeUtf(var2);
       }
 
       public String read(FriendlyByteBuf var1) {
-         return var1.readUtf(32767);
+         return var1.readUtf();
       }
 
       public String copy(String var1) {
          return var1;
       }
