--- net/minecraft/world/effect/MobEffects.java
+++ net/minecraft/world/effect/MobEffects.java
@@ -40,10 +40,13 @@
    public static final MobEffect CONDUIT_POWER;
    public static final MobEffect DOLPHINS_GRACE;
    public static final MobEffect BAD_OMEN;
    public static final MobEffect HERO_OF_THE_VILLAGE;
 
+   public MobEffects() {
+   }
+
    private static MobEffect register(int var0, String var1, MobEffect var2) {
       return (MobEffect)Registry.registerMapping(Registry.MOB_EFFECT, var0, var1, var2);
    }
 
    static {
