--- net/minecraft/world/effect/MobEffectUtil.java
+++ net/minecraft/world/effect/MobEffectUtil.java
@@ -3,10 +3,13 @@
 import net.minecraft.util.Mth;
 import net.minecraft.util.StringUtil;
 import net.minecraft.world.entity.LivingEntity;
 
 public final class MobEffectUtil {
+   public MobEffectUtil() {
+   }
+
    public static String formatDuration(MobEffectInstance var0, float var1) {
       if (var0.isNoCounter()) {
          return "**:**";
       } else {
          int var2 = Mth.floor((float)var0.getDuration() * var1);
