--- net/minecraft/core/particles/BlockParticleOption.java
+++ net/minecraft/core/particles/BlockParticleOption.java
@@ -4,10 +4,11 @@
 import com.mojang.brigadier.exceptions.CommandSyntaxException;
 import com.mojang.serialization.Codec;
 import net.minecraft.commands.arguments.blocks.BlockStateParser;
 import net.minecraft.core.Registry;
 import net.minecraft.network.FriendlyByteBuf;
+import net.minecraft.resources.ResourceLocation;
 import net.minecraft.world.level.block.Block;
 import net.minecraft.world.level.block.state.BlockState;
 
 public class BlockParticleOption implements ParticleOptions {
    public static final ParticleOptions.Deserializer<BlockParticleOption> DESERIALIZER = new ParticleOptions.Deserializer<BlockParticleOption>() {
@@ -39,11 +40,12 @@
    public void writeToNetwork(FriendlyByteBuf var1) {
       var1.writeVarInt(Block.BLOCK_STATE_REGISTRY.getId(this.state));
    }
 
    public String writeToString() {
-      return Registry.PARTICLE_TYPE.getKey(this.getType()) + " " + BlockStateParser.serialize(this.state);
+      ResourceLocation var10000 = Registry.PARTICLE_TYPE.getKey(this.getType());
+      return var10000 + " " + BlockStateParser.serialize(this.state);
    }
 
    public ParticleType<BlockParticleOption> getType() {
       return this.type;
    }
