--- net/minecraft/core/particles/ParticleTypes.java
+++ net/minecraft/core/particles/ParticleTypes.java
@@ -75,10 +75,11 @@
    public static final SimpleParticleType DRIPPING_OBSIDIAN_TEAR;
    public static final SimpleParticleType FALLING_OBSIDIAN_TEAR;
    public static final SimpleParticleType LANDING_OBSIDIAN_TEAR;
    public static final SimpleParticleType REVERSE_PORTAL;
    public static final SimpleParticleType WHITE_ASH;
+   public static final SimpleParticleType SMALL_FLAME;
    public static final Codec<ParticleOptions> CODEC;
 
    private static SimpleParticleType register(String var0, boolean var1) {
       return (SimpleParticleType)Registry.register(Registry.PARTICLE_TYPE, (String)var0, new SimpleParticleType(var1));
    }
@@ -161,8 +162,9 @@
       DRIPPING_OBSIDIAN_TEAR = register("dripping_obsidian_tear", false);
       FALLING_OBSIDIAN_TEAR = register("falling_obsidian_tear", false);
       LANDING_OBSIDIAN_TEAR = register("landing_obsidian_tear", false);
       REVERSE_PORTAL = register("reverse_portal", false);
       WHITE_ASH = register("white_ash", false);
+      SMALL_FLAME = register("small_flame", false);
       CODEC = Registry.PARTICLE_TYPE.dispatch("type", ParticleOptions::getType, ParticleType::codec);
    }
 }
