--- net/minecraft/world/level/biome/BiomeSpecialEffects.java
+++ net/minecraft/world/level/biome/BiomeSpecialEffects.java
@@ -50,11 +50,11 @@
    private final Optional<SoundEvent> ambientLoopSoundEvent;
    private final Optional<AmbientMoodSettings> ambientMoodSettings;
    private final Optional<AmbientAdditionsSettings> ambientAdditionsSettings;
    private final Optional<Music> backgroundMusic;
 
-   private BiomeSpecialEffects(int var1, int var2, int var3, int var4, Optional<Integer> var5, Optional<Integer> var6, BiomeSpecialEffects.GrassColorModifier var7, Optional<AmbientParticleSettings> var8, Optional<SoundEvent> var9, Optional<AmbientMoodSettings> var10, Optional<AmbientAdditionsSettings> var11, Optional<Music> var12) {
+   BiomeSpecialEffects(int var1, int var2, int var3, int var4, Optional<Integer> var5, Optional<Integer> var6, BiomeSpecialEffects.GrassColorModifier var7, Optional<AmbientParticleSettings> var8, Optional<SoundEvent> var9, Optional<AmbientMoodSettings> var10, Optional<AmbientAdditionsSettings> var11, Optional<Music> var12) {
       this.fogColor = var1;
       this.waterColor = var2;
       this.waterFogColor = var3;
       this.skyColor = var4;
       this.foliageColorOverride = var5;
@@ -139,11 +139,11 @@
          return var0;
       }));
 
       public abstract int modifyColor(double var1, double var3, int var5);
 
-      private GrassColorModifier(String var3) {
+      GrassColorModifier(String var3) {
          this.name = var3;
       }
 
       public String getName() {
          return this.name;
