--- net/minecraft/world/level/biome/BiomeManager.java
+++ net/minecraft/world/level/biome/BiomeManager.java
@@ -1,12 +1,15 @@
 package net.minecraft.world.level.biome;
 
 import com.google.common.hash.Hashing;
 import net.minecraft.core.BlockPos;
+import net.minecraft.core.QuartPos;
 import net.minecraft.util.Mth;
+import net.minecraft.world.level.ChunkPos;
 
 public class BiomeManager {
+   static final int CHUNK_CENTER_QUART = QuartPos.fromBlock(8);
    private final BiomeManager.NoiseBiomeSource noiseBiomeSource;
    private final long biomeZoomSeed;
    private final BiomeZoomer zoomer;
 
    public BiomeManager(BiomeManager.NoiseBiomeSource var1, long var2, BiomeZoomer var4) {
@@ -26,26 +29,34 @@
    public Biome getBiome(BlockPos var1) {
       return this.zoomer.getBiome(this.biomeZoomSeed, var1.getX(), var1.getY(), var1.getZ(), this.noiseBiomeSource);
    }
 
    public Biome getNoiseBiomeAtPosition(double var1, double var3, double var5) {
-      int var7 = Mth.floor(var1) >> 2;
-      int var8 = Mth.floor(var3) >> 2;
-      int var9 = Mth.floor(var5) >> 2;
+      int var7 = QuartPos.fromBlock(Mth.floor(var1));
+      int var8 = QuartPos.fromBlock(Mth.floor(var3));
+      int var9 = QuartPos.fromBlock(Mth.floor(var5));
       return this.getNoiseBiomeAtQuart(var7, var8, var9);
    }
 
    public Biome getNoiseBiomeAtPosition(BlockPos var1) {
-      int var2 = var1.getX() >> 2;
-      int var3 = var1.getY() >> 2;
-      int var4 = var1.getZ() >> 2;
+      int var2 = QuartPos.fromBlock(var1.getX());
+      int var3 = QuartPos.fromBlock(var1.getY());
+      int var4 = QuartPos.fromBlock(var1.getZ());
       return this.getNoiseBiomeAtQuart(var2, var3, var4);
    }
 
    public Biome getNoiseBiomeAtQuart(int var1, int var2, int var3) {
       return this.noiseBiomeSource.getNoiseBiome(var1, var2, var3);
    }
 
+   public Biome getPrimaryBiomeAtChunk(ChunkPos var1) {
+      return this.noiseBiomeSource.getPrimaryBiome(var1);
+   }
+
    public interface NoiseBiomeSource {
       Biome getNoiseBiome(int var1, int var2, int var3);
+
+      default Biome getPrimaryBiome(ChunkPos var1) {
+         return this.getNoiseBiome(QuartPos.fromSection(var1.x) + BiomeManager.CHUNK_CENTER_QUART, 0, QuartPos.fromSection(var1.z) + BiomeManager.CHUNK_CENTER_QUART);
+      }
    }
 }
