--- net/minecraft/world/level/biome/MultiNoiseBiomeSource.java
+++ net/minecraft/world/level/biome/MultiNoiseBiomeSource.java
@@ -71,14 +71,14 @@
       this.preset = var8;
       this.temperatureParams = var4;
       this.humidityParams = var5;
       this.altitudeParams = var6;
       this.weirdnessParams = var7;
-      this.temperatureNoise = NormalNoise.create(new WorldgenRandom(var1), var4.firstOctave(), var4.amplitudes());
-      this.humidityNoise = NormalNoise.create(new WorldgenRandom(var1 + 1L), var5.firstOctave(), var5.amplitudes());
-      this.altitudeNoise = NormalNoise.create(new WorldgenRandom(var1 + 2L), var6.firstOctave(), var6.amplitudes());
-      this.weirdnessNoise = NormalNoise.create(new WorldgenRandom(var1 + 3L), var7.firstOctave(), var7.amplitudes());
+      this.temperatureNoise = NormalNoise.create(new WorldgenRandom(var1), var4.firstOctave(), (DoubleList)var4.amplitudes());
+      this.humidityNoise = NormalNoise.create(new WorldgenRandom(var1 + 1L), var5.firstOctave(), (DoubleList)var5.amplitudes());
+      this.altitudeNoise = NormalNoise.create(new WorldgenRandom(var1 + 2L), var6.firstOctave(), (DoubleList)var6.amplitudes());
+      this.weirdnessNoise = NormalNoise.create(new WorldgenRandom(var1 + 3L), var7.firstOctave(), (DoubleList)var7.amplitudes());
       this.parameters = var3;
       this.useY = false;
    }
 
    protected Codec<? extends BiomeSource> codec() {
