--- net/minecraft/world/level/biome/TheEndBiomeSource.java
+++ net/minecraft/world/level/biome/TheEndBiomeSource.java
@@ -16,10 +16,13 @@
          return var0x.biomes;
       }), Codec.LONG.fieldOf("seed").stable().forGetter((var0x) -> {
          return var0x.seed;
       })).apply(var0, var0.stable(TheEndBiomeSource::new));
    });
+   private static final float ISLAND_THRESHOLD = -0.9F;
+   public static final int ISLAND_CHUNK_DISTANCE = 64;
+   private static final long ISLAND_CHUNK_DISTANCE_SQR = 4096L;
    private final SimplexNoise islandNoise;
    private final Registry<Biome> biomes;
    private final long seed;
    private final Biome end;
    private final Biome highlands;
