--- net/minecraft/world/entity/monster/AbstractIllager.java
+++ net/minecraft/world/entity/monster/AbstractIllager.java
@@ -17,11 +17,29 @@
 
    public MobType getMobType() {
       return MobType.ILLAGER;
    }
 
-   public class RaiderOpenDoorGoal extends OpenDoorGoal {
+   public AbstractIllager.IllagerArmPose getArmPose() {
+      return AbstractIllager.IllagerArmPose.CROSSED;
+   }
+
+   public static enum IllagerArmPose {
+      CROSSED,
+      ATTACKING,
+      SPELLCASTING,
+      BOW_AND_ARROW,
+      CROSSBOW_HOLD,
+      CROSSBOW_CHARGE,
+      CELEBRATING,
+      NEUTRAL;
+
+      private IllagerArmPose() {
+      }
+   }
+
+   protected class RaiderOpenDoorGoal extends OpenDoorGoal {
       public RaiderOpenDoorGoal(Raider var2) {
          super(var2, false);
       }
 
       public boolean canUse() {
