--- net/minecraft/world/entity/monster/AbstractSkeleton.java
+++ net/minecraft/world/entity/monster/AbstractSkeleton.java
@@ -152,11 +152,11 @@
    public void reassessWeaponGoal() {
       if (this.level != null && !this.level.isClientSide) {
          this.goalSelector.removeGoal(this.meleeGoal);
          this.goalSelector.removeGoal(this.bowGoal);
          ItemStack var1 = this.getItemInHand(ProjectileUtil.getWeaponHoldingHand(this, Items.BOW));
-         if (var1.getItem() == Items.BOW) {
+         if (var1.is(Items.BOW)) {
             byte var2 = 20;
             if (this.level.getDifficulty() != Difficulty.HARD) {
                var2 = 40;
             }
 
@@ -207,6 +207,10 @@
    }
 
    public double getMyRidingOffset() {
       return -0.6D;
    }
+
+   public boolean isShaking() {
+      return this.isFullyFrozen();
+   }
 }
