--- net/minecraft/world/entity/monster/Blaze.java
+++ net/minecraft/world/entity/monster/Blaze.java
@@ -112,11 +112,11 @@
       }
 
       super.customServerAiStep();
    }
 
-   public boolean causeFallDamage(float var1, float var2) {
+   public boolean causeFallDamage(float var1, float var2, DamageSource var3) {
       return false;
    }
 
    public boolean isOnFire() {
       return this.isCharged();
