--- net/minecraft/world/entity/monster/CrossbowAttackMob.java
+++ net/minecraft/world/entity/monster/CrossbowAttackMob.java
@@ -2,11 +2,10 @@
 
 import com.mojang.math.Quaternion;
 import com.mojang.math.Vector3f;
 import javax.annotation.Nullable;
 import net.minecraft.sounds.SoundEvents;
-import net.minecraft.util.Mth;
 import net.minecraft.world.InteractionHand;
 import net.minecraft.world.entity.LivingEntity;
 import net.minecraft.world.entity.projectile.Projectile;
 import net.minecraft.world.entity.projectile.ProjectileUtil;
 import net.minecraft.world.item.CrossbowItem;
@@ -35,11 +34,11 @@
    }
 
    default void shootCrossbowProjectile(LivingEntity var1, LivingEntity var2, Projectile var3, float var4, float var5) {
       double var7 = var2.getX() - var1.getX();
       double var9 = var2.getZ() - var1.getZ();
-      double var11 = (double)Mth.sqrt(var7 * var7 + var9 * var9);
+      double var11 = Math.sqrt(var7 * var7 + var9 * var9);
       double var13 = var2.getY(0.3333333333333333D) - var3.getY() + var11 * 0.20000000298023224D;
       Vector3f var15 = this.getProjectileShotVector(var1, new Vec3(var7, var13, var9), var4);
       var3.shoot((double)var15.x(), (double)var15.y(), (double)var15.z(), var5, (float)(14 - var1.level.getDifficulty().getId() * 4));
       var1.playSound(SoundEvents.CROSSBOW_SHOOT, 1.0F, 1.0F / (var1.getRandom().nextFloat() * 0.4F + 0.8F));
    }
