--- net/minecraft/world/entity/monster/ElderGuardian.java
+++ net/minecraft/world/entity/monster/ElderGuardian.java
@@ -75,11 +75,11 @@
 
                var8 = (ServerPlayer)var7.next();
             } while(var8.hasEffect(var2) && var8.getEffect(var2).getAmplifier() >= 2 && var8.getEffect(var2).getDuration() >= 1200);
 
             var8.connection.send(new ClientboundGameEventPacket(ClientboundGameEventPacket.GUARDIAN_ELDER_EFFECT, this.isSilent() ? 0.0F : 1.0F));
-            var8.addEffect(new MobEffectInstance(var2, 6000, 2));
+            var8.addEffect(new MobEffectInstance(var2, 6000, 2), this);
          }
       }
 
       if (!this.hasRestriction()) {
          this.restrictTo(this.blockPosition(), 16);
