--- net/minecraft/world/entity/monster/Ghast.java
+++ net/minecraft/world/entity/monster/Ghast.java
@@ -49,10 +49,14 @@
       this.targetSelector.addGoal(1, new NearestAttackableTargetGoal(this, Player.class, 10, true, false, (var1) -> {
          return Math.abs(var1.getY() - this.getY()) <= 4.0D;
       }));
    }
 
+   public boolean isCharging() {
+      return (Boolean)this.entityData.get(DATA_IS_CHARGING);
+   }
+
    public void setCharging(boolean var1) {
       this.entityData.set(DATA_IS_CHARGING, var1);
    }
 
    public int getExplosionPower() {
