--- net/minecraft/world/entity/monster/Husk.java
+++ net/minecraft/world/entity/monster/Husk.java
@@ -47,11 +47,11 @@
 
    public boolean doHurtTarget(Entity var1) {
       boolean var2 = super.doHurtTarget(var1);
       if (var2 && this.getMainHandItem().isEmpty() && var1 instanceof LivingEntity) {
          float var3 = this.level.getCurrentDifficultyAt(this.blockPosition()).getEffectiveDifficulty();
-         ((LivingEntity)var1).addEffect(new MobEffectInstance(MobEffects.HUNGER, 140 * (int)var3));
+         ((LivingEntity)var1).addEffect(new MobEffectInstance(MobEffects.HUNGER, 140 * (int)var3), this);
       }
 
       return var2;
    }
 
