--- net/minecraft/world/entity/monster/MagmaCube.java
+++ net/minecraft/world/entity/monster/MagmaCube.java
@@ -83,11 +83,11 @@
          super.jumpInLiquid(var1);
       }
 
    }
 
-   public boolean causeFallDamage(float var1, float var2) {
+   public boolean causeFallDamage(float var1, float var2, DamageSource var3) {
       return false;
    }
 
    protected boolean isDealsDamage() {
       return this.isEffectiveAi();
