--- net/minecraft/world/entity/monster/Monster.java
+++ net/minecraft/world/entity/monster/Monster.java
@@ -58,14 +58,10 @@
 
    protected SoundEvent getSwimSplashSound() {
       return SoundEvents.HOSTILE_SPLASH;
    }
 
-   public boolean hurt(DamageSource var1, float var2) {
-      return this.isInvulnerableTo(var1) ? false : super.hurt(var1, var2);
-   }
-
    protected SoundEvent getHurtSound(DamageSource var1) {
       return SoundEvents.HOSTILE_HURT;
    }
 
    protected SoundEvent getDeathSound() {
