--- net/minecraft/world/entity/monster/PatrollingMonster.java
+++ net/minecraft/world/entity/monster/PatrollingMonster.java
@@ -129,10 +129,11 @@
    protected void setPatrolling(boolean var1) {
       this.patrolling = var1;
    }
 
    public static class LongDistancePatrolGoal<T extends PatrollingMonster> extends Goal {
+      private static final int NAVIGATION_FAILED_COOLDOWN = 200;
       private final T mob;
       private final double speedModifier;
       private final double leaderSpeedModifier;
       private long cooldownUntil;
 
