--- net/minecraft/world/entity/monster/Silverfish.java
+++ net/minecraft/world/entity/monster/Silverfish.java
@@ -161,11 +161,11 @@
             BlockPos var2 = (new BlockPos(this.mob.getX(), this.mob.getY() + 0.5D, this.mob.getZ())).relative(this.selectedDirection);
             BlockState var3 = var1.getBlockState(var2);
             if (InfestedBlock.isCompatibleHostBlock(var3)) {
                var1.setBlock(var2, InfestedBlock.stateByHostBlock(var3.getBlock()), 3);
                this.mob.spawnAnim();
-               this.mob.remove();
+               this.mob.discard();
             }
 
          }
       }
    }
