--- net/minecraft/world/entity/monster/Zombie.java
+++ net/minecraft/world/entity/monster/Zombie.java
@@ -70,10 +70,15 @@
    private static final UUID SPEED_MODIFIER_BABY_UUID = UUID.fromString("B9766B59-9566-4402-BC1F-2EE2A276D836");
    private static final AttributeModifier SPEED_MODIFIER_BABY;
    private static final EntityDataAccessor<Boolean> DATA_BABY_ID;
    private static final EntityDataAccessor<Integer> DATA_SPECIAL_TYPE_ID;
    private static final EntityDataAccessor<Boolean> DATA_DROWNED_CONVERSION_ID;
+   public static final float ZOMBIE_LEADER_CHANCE = 0.05F;
+   public static final int REINFORCEMENT_ATTEMPTS = 50;
+   public static final int REINFORCEMENT_RANGE_MAX = 40;
+   public static final int REINFORCEMENT_RANGE_MIN = 7;
+   private static final float BREAK_DOOR_CHANCE = 0.1F;
    private static final Predicate<Difficulty> DOOR_BREAKING_PREDICATE;
    private final BreakDoorGoal breakDoorGoal;
    private boolean canBreakDoors;
    private int inWaterTime;
    private int conversionTime;
@@ -391,11 +396,11 @@
    protected float getStandingEyeHeight(Pose var1, EntityDimensions var2) {
       return this.isBaby() ? 0.93F : 1.74F;
    }
 
    public boolean canHoldItem(ItemStack var1) {
-      return var1.getItem() == Items.EGG && this.isBaby() && this.isPassenger() ? false : super.canHoldItem(var1);
+      return var1.is(Items.EGG) && this.isBaby() && this.isPassenger() ? false : super.canHoldItem(var1);
    }
 
    @Nullable
    public SpawnGroupData finalizeSpawn(ServerLevelAccessor var1, DifficultyInstance var2, MobSpawnType var3, @Nullable SpawnGroupData var4, @Nullable CompoundTag var5) {
       Object var10 = super.finalizeSpawn(var1, var2, var3, var4, var5);
@@ -417,11 +422,11 @@
                      var9.setChickenJockey(true);
                      this.startRiding(var9);
                   }
                } else if ((double)var1.getRandom().nextFloat() < 0.05D) {
                   Chicken var12 = (Chicken)EntityType.CHICKEN.create(this.level);
-                  var12.moveTo(this.getX(), this.getY(), this.getZ(), this.yRot, 0.0F);
+                  var12.moveTo(this.getX(), this.getY(), this.getZ(), this.getYRot(), 0.0F);
                   var12.finalizeSpawn(var1, var2, MobSpawnType.JOCKEY, (SpawnGroupData)null, (CompoundTag)null);
                   var12.setChickenJockey(true);
                   this.startRiding(var12);
                   var1.addFreshEntity(var12);
                }
@@ -503,11 +508,11 @@
       DOOR_BREAKING_PREDICATE = (var0) -> {
          return var0 == Difficulty.HARD;
       };
    }
 
-   class ZombieAttackTurtleEggGoal extends RemoveBlockGoal {
+   private class ZombieAttackTurtleEggGoal extends RemoveBlockGoal {
       ZombieAttackTurtleEggGoal(PathfinderMob var2, double var3, int var5) {
          super(Blocks.TURTLE_EGG, var2, var3, var5);
       }
 
       public void playDestroyProgressSound(LevelAccessor var1, BlockPos var2) {
