--- net/minecraft/world/entity/boss/enderdragon/phases/AbstractDragonPhaseInstance.java
+++ net/minecraft/world/entity/boss/enderdragon/phases/AbstractDragonPhaseInstance.java
@@ -1,12 +1,10 @@
 package net.minecraft.world.entity.boss.enderdragon.phases;
 
 import javax.annotation.Nullable;
 import net.minecraft.core.BlockPos;
-import net.minecraft.util.Mth;
 import net.minecraft.world.damagesource.DamageSource;
-import net.minecraft.world.entity.Entity;
 import net.minecraft.world.entity.boss.enderdragon.EndCrystal;
 import net.minecraft.world.entity.boss.enderdragon.EnderDragon;
 import net.minecraft.world.entity.player.Player;
 import net.minecraft.world.phys.Vec3;
 
@@ -48,10 +46,10 @@
    public float onHurt(DamageSource var1, float var2) {
       return var2;
    }
 
    public float getTurnSpeed() {
-      float var1 = Mth.sqrt(Entity.getHorizontalDistanceSqr(this.dragon.getDeltaMovement())) + 1.0F;
+      float var1 = (float)this.dragon.getDeltaMovement().horizontalDistance() + 1.0F;
       float var2 = Math.min(var1, 40.0F);
       return 0.7F / var2 / var1;
    }
 }
