--- net/minecraft/world/entity/boss/enderdragon/phases/DragonLandingApproachPhase.java
+++ net/minecraft/world/entity/boss/enderdragon/phases/DragonLandingApproachPhase.java
@@ -10,11 +10,11 @@
 import net.minecraft.world.level.pathfinder.Node;
 import net.minecraft.world.level.pathfinder.Path;
 import net.minecraft.world.phys.Vec3;
 
 public class DragonLandingApproachPhase extends AbstractDragonPhaseInstance {
-   private static final TargetingConditions NEAR_EGG_TARGETING = (new TargetingConditions()).range(128.0D);
+   private static final TargetingConditions NEAR_EGG_TARGETING = TargetingConditions.forCombat().ignoreLineOfSight();
    private Path currentPath;
    private Vec3 targetLocation;
 
    public DragonLandingApproachPhase(EnderDragon var1) {
       super(var1);
@@ -44,11 +44,11 @@
 
    private void findNewTarget() {
       if (this.currentPath == null || this.currentPath.isDone()) {
          int var1 = this.dragon.findClosestNode();
          BlockPos var2 = this.dragon.level.getHeightmapPos(Heightmap.Types.MOTION_BLOCKING_NO_LEAVES, EndPodiumFeature.END_PODIUM_LOCATION);
-         Player var3 = this.dragon.level.getNearestPlayer(NEAR_EGG_TARGETING, (double)var2.getX(), (double)var2.getY(), (double)var2.getZ());
+         Player var3 = this.dragon.level.getNearestPlayer(NEAR_EGG_TARGETING, this.dragon, (double)var2.getX(), (double)var2.getY(), (double)var2.getZ());
          int var4;
          if (var3 != null) {
             Vec3 var5 = (new Vec3(var3.getX(), 0.0D, var3.getZ())).normalize();
             var4 = this.dragon.findClosestNode(-var5.x * 40.0D, 105.0D, -var5.z * 40.0D);
          } else {
