--- net/minecraft/world/entity/boss/enderdragon/phases/DragonStrafePlayerPhase.java
+++ net/minecraft/world/entity/boss/enderdragon/phases/DragonStrafePlayerPhase.java
@@ -13,10 +13,11 @@
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
 
 public class DragonStrafePlayerPhase extends AbstractDragonPhaseInstance {
    private static final Logger LOGGER = LogManager.getLogger();
+   private static final int FIREBALL_CHARGE_AMOUNT = 5;
    private int fireballCharge;
    private Path currentPath;
    private Vec3 targetLocation;
    private LivingEntity attackTarget;
    private boolean holdingPatternClockwise;
@@ -153,12 +154,12 @@
 
    public void setTarget(LivingEntity var1) {
       this.attackTarget = var1;
       int var2 = this.dragon.findClosestNode();
       int var3 = this.dragon.findClosestNode(this.attackTarget.getX(), this.attackTarget.getY(), this.attackTarget.getZ());
-      int var4 = Mth.floor(this.attackTarget.getX());
-      int var5 = Mth.floor(this.attackTarget.getZ());
+      int var4 = this.attackTarget.getBlockX();
+      int var5 = this.attackTarget.getBlockZ();
       double var6 = (double)var4 - this.dragon.getX();
       double var8 = (double)var5 - this.dragon.getZ();
       double var10 = (double)Mth.sqrt(var6 * var6 + var8 * var8);
       double var12 = Math.min(0.4000000059604645D + var10 / 80.0D - 1.0D, 10.0D);
       int var14 = Mth.floor(this.attackTarget.getY() + var12);
