--- net/minecraft/data/models/blockstates/MultiVariantGenerator.java
+++ net/minecraft/data/models/blockstates/MultiVariantGenerator.java
@@ -8,10 +8,11 @@
 import com.google.gson.JsonObject;
 import com.mojang.datafixers.util.Pair;
 import java.util.Iterator;
 import java.util.List;
 import java.util.Map;
+import java.util.Objects;
 import java.util.Set;
 import java.util.TreeMap;
 import java.util.stream.Stream;
 import net.minecraft.Util;
 import net.minecraft.world.level.block.Block;
@@ -55,14 +56,15 @@
          var4 = var3.getEntries();
       }
 
       TreeMap var5 = new TreeMap();
       var1.forEach((var1x) -> {
-         JsonElement var10000 = (JsonElement)var5.put(((Selector)var1x.getFirst()).getKey(), Variant.convertList((List)var1x.getSecond()));
+         var5.put(((Selector)var1x.getFirst()).getKey(), Variant.convertList((List)var1x.getSecond()));
       });
       JsonObject var6 = new JsonObject();
       var6.add("variants", (JsonElement)Util.make(new JsonObject(), (var1x) -> {
+         Objects.requireNonNull(var1x);
          var5.forEach(var1x::add);
       }));
       return var6;
    }
 
