--- net/minecraft/world/level/levelgen/placement/nether/CountMultiLayerDecorator.java
+++ net/minecraft/world/level/levelgen/placement/nether/CountMultiLayerDecorator.java
@@ -46,11 +46,11 @@
    private static int findOnGroundYPosition(DecorationContext var0, int var1, int var2, int var3, int var4) {
       BlockPos.MutableBlockPos var5 = new BlockPos.MutableBlockPos(var1, var2, var3);
       int var6 = 0;
       BlockState var7 = var0.getBlockState(var5);
 
-      for(int var8 = var2; var8 >= 1; --var8) {
+      for(int var8 = var2; var8 >= var0.getMinBuildHeight() + 1; --var8) {
          var5.setY(var8 - 1);
          BlockState var9 = var0.getBlockState(var5);
          if (!isEmpty(var9) && isEmpty(var7) && !var9.is(Blocks.BEDROCK)) {
             if (var6 == var4) {
                return var5.getY() + 1;
