--- net/minecraft/client/renderer/block/BlockModelShaper.java
+++ net/minecraft/client/renderer/block/BlockModelShaper.java
@@ -44,11 +44,11 @@
       Iterator var1 = Registry.BLOCK.iterator();
 
       while(var1.hasNext()) {
          Block var2 = (Block)var1.next();
          var2.getStateDefinition().getPossibleStates().forEach((var1x) -> {
-            BakedModel var10000 = (BakedModel)this.modelByStateCache.put(var1x, this.modelManager.getModel(stateToModelLocation(var1x)));
+            this.modelByStateCache.put(var1x, this.modelManager.getModel(stateToModelLocation(var1x)));
          });
       }
 
    }
 
