--- net/minecraft/client/renderer/block/LiquidBlockRenderer.java
+++ net/minecraft/client/renderer/block/LiquidBlockRenderer.java
@@ -22,10 +22,11 @@
 import net.minecraft.world.phys.Vec3;
 import net.minecraft.world.phys.shapes.Shapes;
 import net.minecraft.world.phys.shapes.VoxelShape;
 
 public class LiquidBlockRenderer {
+   private static final float MAX_FLUID_HEIGHT = 0.8888889F;
    private final TextureAtlasSprite[] lavaIcons = new TextureAtlasSprite[2];
    private final TextureAtlasSprite[] waterIcons = new TextureAtlasSprite[2];
    private TextureAtlasSprite waterOverlay;
 
    public LiquidBlockRenderer() {
