--- net/minecraft/world/level/levelgen/structure/DesertPyramidPiece.java
+++ net/minecraft/world/level/levelgen/structure/DesertPyramidPiece.java
@@ -3,42 +3,42 @@
 import java.util.Iterator;
 import java.util.Random;
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.Direction;
 import net.minecraft.nbt.CompoundTag;
+import net.minecraft.server.level.ServerLevel;
 import net.minecraft.world.level.ChunkPos;
 import net.minecraft.world.level.StructureFeatureManager;
 import net.minecraft.world.level.WorldGenLevel;
 import net.minecraft.world.level.block.Blocks;
 import net.minecraft.world.level.block.StairBlock;
 import net.minecraft.world.level.block.state.BlockState;
 import net.minecraft.world.level.chunk.ChunkGenerator;
 import net.minecraft.world.level.levelgen.feature.StructurePieceType;
-import net.minecraft.world.level.levelgen.structure.templatesystem.StructureManager;
 import net.minecraft.world.level.storage.loot.BuiltInLootTables;
 
 public class DesertPyramidPiece extends ScatteredFeaturePiece {
    private final boolean[] hasPlacedChest = new boolean[4];
 
    public DesertPyramidPiece(Random var1, int var2, int var3) {
-      super(StructurePieceType.DESERT_PYRAMID_PIECE, var1, var2, 64, var3, 21, 15, 21);
+      super(StructurePieceType.DESERT_PYRAMID_PIECE, var2, 64, var3, 21, 15, 21, getRandomHorizontalDirection(var1));
    }
 
-   public DesertPyramidPiece(StructureManager var1, CompoundTag var2) {
+   public DesertPyramidPiece(ServerLevel var1, CompoundTag var2) {
       super(StructurePieceType.DESERT_PYRAMID_PIECE, var2);
       this.hasPlacedChest[0] = var2.getBoolean("hasPlacedChest0");
       this.hasPlacedChest[1] = var2.getBoolean("hasPlacedChest1");
       this.hasPlacedChest[2] = var2.getBoolean("hasPlacedChest2");
       this.hasPlacedChest[3] = var2.getBoolean("hasPlacedChest3");
    }
 
-   protected void addAdditionalSaveData(CompoundTag var1) {
-      super.addAdditionalSaveData(var1);
-      var1.putBoolean("hasPlacedChest0", this.hasPlacedChest[0]);
-      var1.putBoolean("hasPlacedChest1", this.hasPlacedChest[1]);
-      var1.putBoolean("hasPlacedChest2", this.hasPlacedChest[2]);
-      var1.putBoolean("hasPlacedChest3", this.hasPlacedChest[3]);
+   protected void addAdditionalSaveData(ServerLevel var1, CompoundTag var2) {
+      super.addAdditionalSaveData(var1, var2);
+      var2.putBoolean("hasPlacedChest0", this.hasPlacedChest[0]);
+      var2.putBoolean("hasPlacedChest1", this.hasPlacedChest[1]);
+      var2.putBoolean("hasPlacedChest2", this.hasPlacedChest[2]);
+      var2.putBoolean("hasPlacedChest3", this.hasPlacedChest[3]);
    }
 
    public boolean postProcess(WorldGenLevel var1, StructureFeatureManager var2, ChunkGenerator var3, Random var4, BoundingBox var5, ChunkPos var6, BlockPos var7) {
       this.generateBox(var1, var5, 0, -4, 0, this.width - 1, 0, this.depth - 1, Blocks.SANDSTONE.defaultBlockState(), Blocks.SANDSTONE.defaultBlockState(), false);
 
