--- net/minecraft/world/level/levelgen/structure/JunglePyramidPiece.java
+++ net/minecraft/world/level/levelgen/structure/JunglePyramidPiece.java
@@ -2,10 +2,11 @@
 
 import java.util.Random;
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.Direction;
 import net.minecraft.nbt.CompoundTag;
+import net.minecraft.server.level.ServerLevel;
 import net.minecraft.world.level.ChunkPos;
 import net.minecraft.world.level.StructureFeatureManager;
 import net.minecraft.world.level.WorldGenLevel;
 import net.minecraft.world.level.block.Blocks;
 import net.minecraft.world.level.block.LeverBlock;
@@ -19,38 +20,37 @@
 import net.minecraft.world.level.block.state.BlockState;
 import net.minecraft.world.level.block.state.properties.AttachFace;
 import net.minecraft.world.level.block.state.properties.RedstoneSide;
 import net.minecraft.world.level.chunk.ChunkGenerator;
 import net.minecraft.world.level.levelgen.feature.StructurePieceType;
-import net.minecraft.world.level.levelgen.structure.templatesystem.StructureManager;
 import net.minecraft.world.level.storage.loot.BuiltInLootTables;
 
 public class JunglePyramidPiece extends ScatteredFeaturePiece {
    private boolean placedMainChest;
    private boolean placedHiddenChest;
    private boolean placedTrap1;
    private boolean placedTrap2;
    private static final JunglePyramidPiece.MossStoneSelector STONE_SELECTOR = new JunglePyramidPiece.MossStoneSelector();
 
    public JunglePyramidPiece(Random var1, int var2, int var3) {
-      super(StructurePieceType.JUNGLE_PYRAMID_PIECE, var1, var2, 64, var3, 12, 10, 15);
+      super(StructurePieceType.JUNGLE_PYRAMID_PIECE, var2, 64, var3, 12, 10, 15, getRandomHorizontalDirection(var1));
    }
 
-   public JunglePyramidPiece(StructureManager var1, CompoundTag var2) {
+   public JunglePyramidPiece(ServerLevel var1, CompoundTag var2) {
       super(StructurePieceType.JUNGLE_PYRAMID_PIECE, var2);
       this.placedMainChest = var2.getBoolean("placedMainChest");
       this.placedHiddenChest = var2.getBoolean("placedHiddenChest");
       this.placedTrap1 = var2.getBoolean("placedTrap1");
       this.placedTrap2 = var2.getBoolean("placedTrap2");
    }
 
-   protected void addAdditionalSaveData(CompoundTag var1) {
-      super.addAdditionalSaveData(var1);
-      var1.putBoolean("placedMainChest", this.placedMainChest);
-      var1.putBoolean("placedHiddenChest", this.placedHiddenChest);
-      var1.putBoolean("placedTrap1", this.placedTrap1);
-      var1.putBoolean("placedTrap2", this.placedTrap2);
+   protected void addAdditionalSaveData(ServerLevel var1, CompoundTag var2) {
+      super.addAdditionalSaveData(var1, var2);
+      var2.putBoolean("placedMainChest", this.placedMainChest);
+      var2.putBoolean("placedHiddenChest", this.placedHiddenChest);
+      var2.putBoolean("placedTrap1", this.placedTrap1);
+      var2.putBoolean("placedTrap2", this.placedTrap2);
    }
 
    public boolean postProcess(WorldGenLevel var1, StructureFeatureManager var2, ChunkGenerator var3, Random var4, BoundingBox var5, ChunkPos var6, BlockPos var7) {
       if (!this.updateAverageGroundHeight(var1, var5, 0)) {
          return false;
@@ -232,11 +232,11 @@
          return true;
       }
    }
 
    static class MossStoneSelector extends StructurePiece.BlockSelector {
-      private MossStoneSelector() {
+      MossStoneSelector() {
       }
 
       public void next(Random var1, int var2, int var3, int var4, boolean var5) {
          if (var1.nextFloat() < 0.4F) {
             this.next = Blocks.COBBLESTONE.defaultBlockState();
