--- net/minecraft/world/level/levelgen/structure/LegacyStructureDataHandler.java
+++ net/minecraft/world/level/levelgen/structure/LegacyStructureDataHandler.java
@@ -207,17 +207,15 @@
                   return new Long2ObjectOpenHashMap();
                })).put(var8, var7);
             }
 
             String var14 = var3 + "_index";
-            StructureFeatureIndexSavedData var15 = (StructureFeatureIndexSavedData)var1.computeIfAbsent(() -> {
-               return new StructureFeatureIndexSavedData(var14);
-            }, var14);
+            StructureFeatureIndexSavedData var15 = (StructureFeatureIndexSavedData)var1.computeIfAbsent(StructureFeatureIndexSavedData::load, StructureFeatureIndexSavedData::new, var14);
             if (!var15.getAll().isEmpty()) {
                this.indexMap.put(var3, var15);
             } else {
-               StructureFeatureIndexSavedData var16 = new StructureFeatureIndexSavedData(var14);
+               StructureFeatureIndexSavedData var16 = new StructureFeatureIndexSavedData();
                this.indexMap.put(var3, var16);
                Iterator var17 = var4.getAllKeys().iterator();
 
                while(var17.hasNext()) {
                   String var9 = (String)var17.next();
