--- net/minecraft/world/level/levelgen/structure/OceanMonumentPieces.java
+++ net/minecraft/world/level/levelgen/structure/OceanMonumentPieces.java
@@ -1795,11 +1795,11 @@
       protected static final int GRIDROOM_TOP_CONNECT_INDEX;
       protected static final int GRIDROOM_LEFTWING_CONNECT_INDEX;
       protected static final int GRIDROOM_RIGHTWING_CONNECT_INDEX;
       protected OceanMonumentPieces.RoomDefinition roomDefinition;
 
-      protected static final int getRoomIndex(int var0, int var1, int var2) {
+      protected static int getRoomIndex(int var0, int var1, int var2) {
          return var1 * 25 + var2 * 5 + var0;
       }
 
       public OceanMonumentPiece(StructurePieceType var1, int var2) {
          super(var1, var2);
