--- net/minecraft/world/level/levelgen/structure/StrongholdPieces.java
+++ net/minecraft/world/level/levelgen/structure/StrongholdPieces.java
@@ -26,10 +26,11 @@
 import net.minecraft.world.level.block.entity.SpawnerBlockEntity;
 import net.minecraft.world.level.block.state.BlockState;
 import net.minecraft.world.level.block.state.properties.DoubleBlockHalf;
 import net.minecraft.world.level.block.state.properties.SlabType;
 import net.minecraft.world.level.chunk.ChunkGenerator;
+import net.minecraft.world.level.levelgen.feature.NoiseEffect;
 import net.minecraft.world.level.levelgen.feature.StructurePieceType;
 import net.minecraft.world.level.levelgen.structure.templatesystem.StructureManager;
 import net.minecraft.world.level.storage.loot.BuiltInLootTables;
 
 public class StrongholdPieces {
@@ -968,14 +969,14 @@
          this.generateBox(var1, var5, 0, 0, 0, 4, 4, 6, true, var4, StrongholdPieces.SMOOTH_STONE_SELECTOR);
          this.generateSmallDoor(var1, var4, var5, this.entryDoor, 1, 1, 0);
          this.generateSmallDoor(var1, var4, var5, StrongholdPieces.StrongholdPiece.SmallDoorType.OPENING, 1, 1, 6);
          BlockState var8 = (BlockState)Blocks.WALL_TORCH.defaultBlockState().setValue(WallTorchBlock.FACING, Direction.EAST);
          BlockState var9 = (BlockState)Blocks.WALL_TORCH.defaultBlockState().setValue(WallTorchBlock.FACING, Direction.WEST);
-         this.maybeGenerateBlock(var1, var5, var4, 0.1F, 1, 2, 1, var8);
-         this.maybeGenerateBlock(var1, var5, var4, 0.1F, 3, 2, 1, var9);
-         this.maybeGenerateBlock(var1, var5, var4, 0.1F, 1, 2, 5, var8);
-         this.maybeGenerateBlock(var1, var5, var4, 0.1F, 3, 2, 5, var9);
+         this.maybeGenerateBlock(var1, var5, var4, 0.1F, 1, 2, 1, var8, false);
+         this.maybeGenerateBlock(var1, var5, var4, 0.1F, 3, 2, 1, var9, false);
+         this.maybeGenerateBlock(var1, var5, var4, 0.1F, 1, 2, 5, var8, false);
+         this.maybeGenerateBlock(var1, var5, var4, 0.1F, 3, 2, 5, var9, false);
          if (this.leftChild) {
             this.generateBox(var1, var5, 0, 1, 2, 0, 3, 4, CAVE_AIR, CAVE_AIR, false);
          }
 
          if (this.rightChild) {
@@ -1156,10 +1157,14 @@
          super(var1, var2);
          this.entryDoor = StrongholdPieces.StrongholdPiece.SmallDoorType.OPENING;
          this.entryDoor = StrongholdPieces.StrongholdPiece.SmallDoorType.valueOf(var2.getString("EntryDoor"));
       }
 
+      public NoiseEffect getNoiseEffect() {
+         return NoiseEffect.BURY;
+      }
+
       protected void addAdditionalSaveData(CompoundTag var1) {
          var1.putString("EntryDoor", this.entryDoor.name());
       }
 
       protected void generateSmallDoor(WorldGenLevel var1, Random var2, BoundingBox var3, StrongholdPieces.StrongholdPiece.SmallDoorType var4, int var5, int var6, int var7) {
