--- net/minecraft/world/entity/player/Abilities.java
+++ net/minecraft/world/entity/player/Abilities.java
@@ -47,9 +47,17 @@
 
    public float getFlyingSpeed() {
       return this.flyingSpeed;
    }
 
+   public void setFlyingSpeed(float var1) {
+      this.flyingSpeed = var1;
+   }
+
    public float getWalkingSpeed() {
       return this.walkingSpeed;
    }
+
+   public void setWalkingSpeed(float var1) {
+      this.walkingSpeed = var1;
+   }
 }
