--- net/minecraft/world/entity/player/ChatVisiblity.java
+++ net/minecraft/world/entity/player/ChatVisiblity.java
@@ -1,9 +1,10 @@
 package net.minecraft.world.entity.player;
 
 import java.util.Arrays;
 import java.util.Comparator;
+import net.minecraft.util.Mth;
 
 public enum ChatVisiblity {
    FULL(0, "options.chat.visibility.full"),
    SYSTEM(1, "options.chat.visibility.system"),
    HIDDEN(2, "options.chat.visibility.hidden");
@@ -20,6 +21,14 @@
    }
 
    public int getId() {
       return this.id;
    }
+
+   public String getKey() {
+      return this.key;
+   }
+
+   public static ChatVisiblity byId(int var0) {
+      return BY_ID[Mth.positiveModulo(var0, BY_ID.length)];
+   }
 }
