--- net/minecraft/world/entity/player/StackedContents.java
+++ net/minecraft/world/entity/player/StackedContents.java
@@ -18,10 +18,11 @@
 import net.minecraft.world.item.ItemStack;
 import net.minecraft.world.item.crafting.Ingredient;
 import net.minecraft.world.item.crafting.Recipe;
 
 public class StackedContents {
+   private static final int EMPTY = 0;
    public final Int2IntMap contents = new Int2IntOpenHashMap();
 
    public StackedContents() {
    }
 
