--- net/minecraft/server/players/BanListEntry.java
+++ net/minecraft/server/players/BanListEntry.java
@@ -7,10 +7,11 @@
 import javax.annotation.Nullable;
 import net.minecraft.network.chat.Component;
 
 public abstract class BanListEntry<T> extends StoredUserEntry<T> {
    public static final SimpleDateFormat DATE_FORMAT = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss Z");
+   public static final String EXPIRES_NEVER = "forever";
    protected final Date created;
    protected final String source;
    protected final Date expires;
    protected final String reason;
 
@@ -44,10 +45,14 @@
 
       this.expires = var4;
       this.reason = var2.has("reason") ? var2.get("reason").getAsString() : "Banned by an operator.";
    }
 
+   public Date getCreated() {
+      return this.created;
+   }
+
    public String getSource() {
       return this.source;
    }
 
    public Date getExpires() {
