--- net/minecraft/server/players/OldUsersConverter.java
+++ net/minecraft/server/players/OldUsersConverter.java
@@ -34,10 +34,13 @@
    public static final File OLD_IPBANLIST = new File("banned-ips.txt");
    public static final File OLD_USERBANLIST = new File("banned-players.txt");
    public static final File OLD_OPLIST = new File("ops.txt");
    public static final File OLD_WHITELIST = new File("white-list.txt");
 
+   public OldUsersConverter() {
+   }
+
    static List<String> readOldListFormat(File var0, Map<String, String[]> var1) throws IOException {
       List var2 = Files.readLines(var0, StandardCharsets.UTF_8);
       Iterator var3 = var2.iterator();
 
       while(var3.hasNext()) {
