--- net/minecraft/server/players/ServerOpListEntry.java
+++ net/minecraft/server/players/ServerOpListEntry.java
@@ -1,10 +1,11 @@
 package net.minecraft.server.players;
 
 import com.google.gson.JsonObject;
 import com.mojang.authlib.GameProfile;
 import java.util.UUID;
+import javax.annotation.Nullable;
 
 public class ServerOpListEntry extends StoredUserEntry<GameProfile> {
    private final int level;
    private final boolean bypassesPlayerLimit;
 
@@ -35,10 +36,11 @@
          var1.addProperty("level", (Number)this.level);
          var1.addProperty("bypassesPlayerLimit", this.bypassesPlayerLimit);
       }
    }
 
+   @Nullable
    private static GameProfile createGameProfile(JsonObject var0) {
       if (var0.has("uuid") && var0.has("name")) {
          String var1 = var0.get("uuid").getAsString();
 
          UUID var2;
