--- net/minecraft/server/players/StoredUserList.java
+++ net/minecraft/server/players/StoredUserList.java
@@ -69,11 +69,11 @@
    public void remove(StoredUserEntry<K> var1) {
       this.remove(var1.getUser());
    }
 
    public String[] getUserList() {
-      return (String[])this.map.keySet().toArray(new String[this.map.size()]);
+      return (String[])this.map.keySet().toArray(new String[0]);
    }
 
    public boolean isEmpty() {
       return this.map.size() < 1;
    }
