--- net/minecraft/client/gui/font/glyphs/MissingGlyph.java
+++ net/minecraft/client/gui/font/glyphs/MissingGlyph.java
@@ -5,10 +5,12 @@
 import net.minecraft.Util;
 
 public enum MissingGlyph implements RawGlyph {
    INSTANCE;
 
+   private static final int MISSING_IMAGE_WIDTH = 5;
+   private static final int MISSING_IMAGE_HEIGHT = 8;
    private static final NativeImage IMAGE_DATA = (NativeImage)Util.make(new NativeImage(NativeImage.Format.RGBA, 5, 8, false), (var0) -> {
       for(int var1 = 0; var1 < 8; ++var1) {
          for(int var2 = 0; var2 < 5; ++var2) {
             boolean var3 = var2 == 0 || var2 + 1 == 5 || var1 == 0 || var1 + 1 == 8;
             var0.setPixelRGBA(var2, var1, var3 ? -1 : 0);
