--- net/minecraft/client/gui/font/glyphs/WhiteGlyph.java
+++ net/minecraft/client/gui/font/glyphs/WhiteGlyph.java
@@ -5,10 +5,12 @@
 import net.minecraft.Util;
 
 public enum WhiteGlyph implements RawGlyph {
    INSTANCE;
 
+   private static final int WIDTH = 5;
+   private static final int HEIGHT = 8;
    private static final NativeImage IMAGE_DATA = (NativeImage)Util.make(new NativeImage(NativeImage.Format.RGBA, 5, 8, false), (var0) -> {
       for(int var1 = 0; var1 < 8; ++var1) {
          for(int var2 = 0; var2 < 5; ++var2) {
             boolean var10000;
             if (var2 != 0 && var2 + 1 != 5 && var1 != 0 && var1 + 1 != 8) {
