--- net/minecraft/world/level/storage/loot/LootTables.java
+++ net/minecraft/world/level/storage/loot/LootTables.java
@@ -55,11 +55,11 @@
       ValidationContext var7 = new ValidationContext(var10002, var10003, var6::get);
       var6.forEach((var1x, var2x) -> {
          validate(var7, var1x, var2x);
       });
       var7.getProblems().forEach((var0, var1x) -> {
-         LOGGER.warn("Found validation problem in " + var0 + ": " + var1x);
+         LOGGER.warn("Found validation problem in {}: {}", var0, var1x);
       });
       this.tables = var6;
    }
 
    public static void validate(ValidationContext var0, ResourceLocation var1, LootTable var2) {
