--- net/minecraft/world/level/storage/loot/LootTable.java
+++ net/minecraft/world/level/storage/loot/LootTable.java
@@ -210,11 +210,11 @@
          if (var1.paramSet != LootTable.DEFAULT_PARAM_SET) {
             ResourceLocation var5 = LootContextParamSets.getKey(var1.paramSet);
             if (var5 != null) {
                var4.addProperty("type", var5.toString());
             } else {
-               LootTable.LOGGER.warn("Failed to find id for param set " + var1.paramSet);
+               LootTable.LOGGER.warn("Failed to find id for param set {}", var1.paramSet);
             }
          }
 
          if (var1.pools.length > 0) {
             var4.add("pools", var3.serialize(var1.pools));
