--- net/minecraft/world/level/storage/loot/PredicateManager.java
+++ net/minecraft/world/level/storage/loot/PredicateManager.java
@@ -56,11 +56,11 @@
       });
       var5.forEach((var1x, var2x) -> {
          var2x.validate(var6.enterCondition("{" + var1x + "}", var1x));
       });
       var6.getProblems().forEach((var0, var1x) -> {
-         LOGGER.warn("Found validation problem in " + var0 + ": " + var1x);
+         LOGGER.warn("Found validation problem in {}: {}", var0, var1x);
       });
       this.conditions = var5;
    }
 
    public Set<ResourceLocation> getKeys() {
