--- net/minecraft/world/level/storage/DerivedLevelData.java
+++ net/minecraft/world/level/storage/DerivedLevelData.java
@@ -5,10 +5,11 @@
 import net.minecraft.core.BlockPos;
 import net.minecraft.server.MinecraftServer;
 import net.minecraft.world.Difficulty;
 import net.minecraft.world.level.GameRules;
 import net.minecraft.world.level.GameType;
+import net.minecraft.world.level.LevelHeightAccessor;
 import net.minecraft.world.level.border.WorldBorder;
 import net.minecraft.world.level.timers.TimerQueue;
 
 public class DerivedLevelData implements ServerLevelData {
    private final WorldData worldData;
@@ -160,13 +161,17 @@
    }
 
    public void setWanderingTraderSpawnChance(int var1) {
    }
 
+   public UUID getWanderingTraderId() {
+      return null;
+   }
+
    public void setWanderingTraderId(UUID var1) {
    }
 
-   public void fillCrashReportCategory(CrashReportCategory var1) {
+   public void fillCrashReportCategory(CrashReportCategory var1, LevelHeightAccessor var2) {
       var1.setDetail("Derived", (Object)true);
-      this.wrapped.fillCrashReportCategory(var1);
+      this.wrapped.fillCrashReportCategory(var1, var2);
    }
 }
