--- net/minecraft/world/level/storage/LevelStorageSource.java
+++ net/minecraft/world/level/storage/LevelStorageSource.java
@@ -1,37 +1,51 @@
 package net.minecraft.world.level.storage;
 
 import com.google.common.collect.ImmutableList;
+import com.google.common.collect.Lists;
 import com.google.common.collect.Maps;
 import com.google.common.collect.UnmodifiableIterator;
 import com.mojang.datafixers.DataFixer;
 import com.mojang.datafixers.util.Pair;
 import com.mojang.serialization.DataResult;
 import com.mojang.serialization.Dynamic;
 import com.mojang.serialization.DynamicOps;
 import com.mojang.serialization.Lifecycle;
+import java.io.BufferedOutputStream;
 import java.io.File;
 import java.io.IOException;
+import java.io.OutputStream;
+import java.nio.file.FileVisitResult;
 import java.nio.file.Files;
 import java.nio.file.LinkOption;
 import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.nio.file.SimpleFileVisitor;
+import java.nio.file.attribute.BasicFileAttributes;
+import java.time.LocalDateTime;
 import java.time.format.DateTimeFormatter;
 import java.time.format.DateTimeFormatterBuilder;
 import java.time.format.SignStyle;
 import java.time.temporal.ChronoField;
+import java.util.ArrayList;
+import java.util.List;
 import java.util.Map;
 import java.util.Optional;
 import java.util.function.BiFunction;
+import java.util.zip.ZipEntry;
+import java.util.zip.ZipOutputStream;
 import javax.annotation.Nullable;
+import net.minecraft.FileUtil;
 import net.minecraft.SharedConstants;
 import net.minecraft.Util;
 import net.minecraft.core.Registry;
 import net.minecraft.core.RegistryAccess;
 import net.minecraft.nbt.CompoundTag;
 import net.minecraft.nbt.NbtIo;
 import net.minecraft.nbt.NbtOps;
 import net.minecraft.nbt.Tag;
+import net.minecraft.network.chat.TranslatableComponent;
 import net.minecraft.resources.RegistryLookupCodec;
 import net.minecraft.resources.ResourceKey;
 import net.minecraft.util.DirectoryLock;
 import net.minecraft.util.ProgressListener;
 import net.minecraft.util.datafix.DataFixTypes;
@@ -46,10 +60,11 @@
 import org.apache.logging.log4j.Logger;
 
 public class LevelStorageSource {
    private static final Logger LOGGER = LogManager.getLogger();
    private static final DateTimeFormatter FORMATTER;
+   private static final String ICON_FILENAME = "icon.png";
    private static final ImmutableList<String> OLD_SETTINGS_KEYS;
    private final Path baseDir;
    private final Path backupDir;
    private final DataFixer fixerUpper;
 
@@ -114,10 +129,45 @@
       Logger var10001 = LOGGER;
       var10001.getClass();
       return (DataPackConfig)var10000.resultOrPartial(var10001::error).orElse(DataPackConfig.DEFAULT);
    }
 
+   public String getName() {
+      return "Anvil";
+   }
+
+   public List<LevelSummary> getLevelList() throws LevelStorageException {
+      if (!Files.isDirectory(this.baseDir, new LinkOption[0])) {
+         throw new LevelStorageException((new TranslatableComponent("selectWorld.load_folder_access")).getString());
+      } else {
+         ArrayList var1 = Lists.newArrayList();
+         File[] var2 = this.baseDir.toFile().listFiles();
+         File[] var3 = var2;
+         int var4 = var2.length;
+
+         for(int var5 = 0; var5 < var4; ++var5) {
+            File var6 = var3[var5];
+            if (var6.isDirectory()) {
+               boolean var7;
+               try {
+                  var7 = DirectoryLock.isLocked(var6.toPath());
+               } catch (Exception var9) {
+                  LOGGER.warn((String)"Failed to read {} lock", (Object)var6, (Object)var9);
+                  continue;
+               }
+
+               LevelSummary var8 = (LevelSummary)this.readLevelData(var6, this.levelSummaryReader(var6, var7));
+               if (var8 != null) {
+                  var1.add(var8);
+               }
+            }
+         }
+
+         return var1;
+      }
+   }
+
    private int getStorageVersion() {
       return 19133;
    }
 
    @Nullable
@@ -197,10 +247,33 @@
             return null;
          }
       };
    }
 
+   public boolean isNewLevelIdAcceptable(String var1) {
+      try {
+         Path var2 = this.baseDir.resolve(var1);
+         Files.createDirectory(var2);
+         Files.deleteIfExists(var2);
+         return true;
+      } catch (IOException var3) {
+         return false;
+      }
+   }
+
+   public boolean levelExists(String var1) {
+      return Files.isDirectory(this.baseDir.resolve(var1), new LinkOption[0]);
+   }
+
+   public Path getBaseDir() {
+      return this.baseDir;
+   }
+
+   public Path getBackupPath() {
+      return this.backupDir;
+   }
+
    public LevelStorageSource.LevelStorageAccess createAccess(String var1) throws IOException {
       return new LevelStorageSource.LevelStorageAccess(var1);
    }
 
    static {
@@ -300,10 +373,129 @@
       public File getIconFile() {
          this.checkLock();
          return this.levelPath.resolve("icon.png").toFile();
       }
 
+      public void deleteLevel() throws IOException {
+         this.checkLock();
+         final Path var1 = this.levelPath.resolve("session.lock");
+         int var2 = 1;
+
+         while(var2 <= 5) {
+            LevelStorageSource.LOGGER.info((String)"Attempt {}...", (Object)var2);
+
+            try {
+               Files.walkFileTree(this.levelPath, new SimpleFileVisitor<Path>() {
+                  public FileVisitResult visitFile(Path var1x, BasicFileAttributes var2) throws IOException {
+                     if (!var1x.equals(var1)) {
+                        LevelStorageSource.LOGGER.debug((String)"Deleting {}", (Object)var1x);
+                        Files.delete(var1x);
+                     }
+
+                     return FileVisitResult.CONTINUE;
+                  }
+
+                  public FileVisitResult postVisitDirectory(Path var1x, IOException var2) throws IOException {
+                     if (var2 != null) {
+                        throw var2;
+                     } else {
+                        if (var1x.equals(LevelStorageAccess.this.levelPath)) {
+                           LevelStorageAccess.this.lock.close();
+                           Files.deleteIfExists(var1);
+                        }
+
+                        Files.delete(var1x);
+                        return FileVisitResult.CONTINUE;
+                     }
+                  }
+               });
+               break;
+            } catch (IOException var6) {
+               if (var2 >= 5) {
+                  throw var6;
+               }
+
+               LevelStorageSource.LOGGER.warn((String)"Failed to delete {}", (Object)this.levelPath, (Object)var6);
+
+               try {
+                  Thread.sleep(500L);
+               } catch (InterruptedException var5) {
+               }
+
+               ++var2;
+            }
+         }
+
+      }
+
+      public void renameLevel(String var1) throws IOException {
+         this.checkLock();
+         File var2 = new File(LevelStorageSource.this.baseDir.toFile(), this.levelId);
+         if (var2.exists()) {
+            File var3 = new File(var2, "level.dat");
+            if (var3.exists()) {
+               CompoundTag var4 = NbtIo.readCompressed(var3);
+               CompoundTag var5 = var4.getCompound("Data");
+               var5.putString("LevelName", var1);
+               NbtIo.writeCompressed(var4, var3);
+            }
+
+         }
+      }
+
+      public long makeWorldBackup() throws IOException {
+         this.checkLock();
+         String var1 = LocalDateTime.now().format(LevelStorageSource.FORMATTER) + "_" + this.levelId;
+         Path var2 = LevelStorageSource.this.getBackupPath();
+
+         try {
+            Files.createDirectories(Files.exists(var2, new LinkOption[0]) ? var2.toRealPath() : var2);
+         } catch (IOException var16) {
+            throw new RuntimeException(var16);
+         }
+
+         Path var3 = var2.resolve(FileUtil.findAvailableName(var2, var1, ".zip"));
+         final ZipOutputStream var4 = new ZipOutputStream(new BufferedOutputStream(Files.newOutputStream(var3)));
+         Throwable var5 = null;
+
+         try {
+            final Path var6 = Paths.get(this.levelId);
+            Files.walkFileTree(this.levelPath, new SimpleFileVisitor<Path>() {
+               public FileVisitResult visitFile(Path var1, BasicFileAttributes var2) throws IOException {
+                  if (var1.endsWith("session.lock")) {
+                     return FileVisitResult.CONTINUE;
+                  } else {
+                     String var3 = var6.resolve(LevelStorageAccess.this.levelPath.relativize(var1)).toString().replace('\\', '/');
+                     ZipEntry var4x = new ZipEntry(var3);
+                     var4.putNextEntry(var4x);
+                     com.google.common.io.Files.asByteSource(var1.toFile()).copyTo((OutputStream)var4);
+                     var4.closeEntry();
+                     return FileVisitResult.CONTINUE;
+                  }
+               }
+            });
+         } catch (Throwable var15) {
+            var5 = var15;
+            throw var15;
+         } finally {
+            if (var4 != null) {
+               if (var5 != null) {
+                  try {
+                     var4.close();
+                  } catch (Throwable var14) {
+                     var5.addSuppressed(var14);
+                  }
+               } else {
+                  var4.close();
+               }
+            }
+
+         }
+
+         return Files.size(var3);
+      }
+
       public void close() throws IOException {
          this.lock.close();
       }
    }
 }
