--- net/minecraft/world/level/storage/LevelVersion.java
+++ net/minecraft/world/level/storage/LevelVersion.java
@@ -31,6 +31,18 @@
    }
 
    public long lastPlayed() {
       return this.lastPlayed;
    }
+
+   public String minecraftVersionName() {
+      return this.minecraftVersionName;
+   }
+
+   public int minecraftVersion() {
+      return this.minecraftVersion;
+   }
+
+   public boolean snapshot() {
+      return this.snapshot;
+   }
 }
