--- net/minecraft/world/level/storage/McRegionUpgrader.java
+++ net/minecraft/world/level/storage/McRegionUpgrader.java
@@ -9,11 +9,10 @@
 import java.util.Collection;
 import java.util.Collections;
 import java.util.Iterator;
 import net.minecraft.core.Registry;
 import net.minecraft.core.RegistryAccess;
-import net.minecraft.core.WritableRegistry;
 import net.minecraft.nbt.CompoundTag;
 import net.minecraft.nbt.NbtIo;
 import net.minecraft.nbt.NbtOps;
 import net.minecraft.resources.RegistryReadOps;
 import net.minecraft.server.packs.resources.ResourceManager;
@@ -56,11 +55,11 @@
       LOGGER.info((String)"Total conversion count is {}", (Object)var8);
       RegistryAccess.RegistryHolder var9 = RegistryAccess.builtin();
       RegistryReadOps var10 = RegistryReadOps.create(NbtOps.INSTANCE, (ResourceManager)ResourceManager.Empty.INSTANCE, var9);
       WorldData var11 = var0.getDataTag(var10, DataPackConfig.DEFAULT);
       long var12 = var11 != null ? var11.worldGenSettings().seed() : 0L;
-      WritableRegistry var15 = var9.registryOrThrow(Registry.BIOME_REGISTRY);
+      Registry var15 = var9.registryOrThrow(Registry.BIOME_REGISTRY);
       Object var14;
       if (var11 != null && var11.worldGenSettings().isFlatWorld()) {
          var14 = new FixedBiomeSource((Biome)var15.getOrThrow(Biomes.PLAINS));
       } else {
          var14 = new OverworldBiomeSource(var12, false, false, var15);
