--- net/minecraft/world/level/storage/ServerLevelData.java
+++ net/minecraft/world/level/storage/ServerLevelData.java
@@ -1,11 +1,13 @@
 package net.minecraft.world.level.storage;
 
 import java.util.UUID;
+import javax.annotation.Nullable;
 import net.minecraft.CrashReportCategory;
 import net.minecraft.server.MinecraftServer;
 import net.minecraft.world.level.GameType;
+import net.minecraft.world.level.LevelHeightAccessor;
 import net.minecraft.world.level.border.WorldBorder;
 import net.minecraft.world.level.timers.TimerQueue;
 
 public interface ServerLevelData extends WritableLevelData {
    String getLevelName();
@@ -18,12 +20,12 @@
 
    void setThunderTime(int var1);
 
    int getThunderTime();
 
-   default void fillCrashReportCategory(CrashReportCategory var1) {
-      WritableLevelData.super.fillCrashReportCategory(var1);
+   default void fillCrashReportCategory(CrashReportCategory var1, LevelHeightAccessor var2) {
+      WritableLevelData.super.fillCrashReportCategory(var1, var2);
       var1.setDetail("Level name", this::getLevelName);
       var1.setDetail("Level game mode", () -> {
          return String.format("Game mode: %s (ID %d). Hardcore: %b. Cheats: %b", this.getGameType().getName(), this.getGameType().getId(), this.isHardcore(), this.getAllowCommands());
       });
       var1.setDetail("Level weather", () -> {
@@ -41,10 +43,13 @@
 
    int getWanderingTraderSpawnChance();
 
    void setWanderingTraderSpawnChance(int var1);
 
+   @Nullable
+   UUID getWanderingTraderId();
+
    void setWanderingTraderId(UUID var1);
 
    GameType getGameType();
 
    void setWorldBorder(WorldBorder.Settings var1);
