--- net/minecraft/world/level/storage/WorldData.java
+++ net/minecraft/world/level/storage/WorldData.java
@@ -1,19 +1,24 @@
 package net.minecraft.world.level.storage;
 
+import com.mojang.serialization.Lifecycle;
 import java.util.Set;
 import javax.annotation.Nullable;
 import net.minecraft.CrashReportCategory;
 import net.minecraft.core.RegistryAccess;
 import net.minecraft.nbt.CompoundTag;
 import net.minecraft.world.Difficulty;
 import net.minecraft.world.level.DataPackConfig;
 import net.minecraft.world.level.GameRules;
 import net.minecraft.world.level.GameType;
+import net.minecraft.world.level.LevelSettings;
 import net.minecraft.world.level.levelgen.WorldGenSettings;
 
 public interface WorldData {
+   int ANVIL_VERSION_ID = 19133;
+   int MCREGION_VERSION_ID = 19132;
+
    DataPackConfig getDataPackConfig();
 
    void setDataPackConfig(DataPackConfig var1);
 
    boolean wasModded();
@@ -51,10 +56,12 @@
 
    void setCustomBossEvents(@Nullable CompoundTag var1);
 
    ServerLevelData overworldData();
 
+   LevelSettings getLevelSettings();
+
    CompoundTag createTag(RegistryAccess var1, @Nullable CompoundTag var2);
 
    boolean isHardcore();
 
    int getVersion();
@@ -82,6 +89,8 @@
    CompoundTag endDragonFightData();
 
    void setEndDragonFightData(CompoundTag var1);
 
    WorldGenSettings worldGenSettings();
+
+   Lifecycle worldGenSettingsLifecycle();
 }
