--- net/minecraft/world/level/storage/LevelStorageSource.java
+++ net/minecraft/world/level/storage/LevelStorageSource.java
@@ -59,10 +59,11 @@
 import org.apache.logging.log4j.Logger;
 
 public class LevelStorageSource {
    private static final Logger LOGGER = LogManager.getLogger();
    private static final DateTimeFormatter FORMATTER;
+   private static final String ICON_FILENAME = "icon.png";
    private static final ImmutableList<String> OLD_SETTINGS_KEYS;
    private final Path baseDir;
    private final Path backupDir;
    private final DataFixer fixerUpper;
 
@@ -127,10 +128,14 @@
       Logger var10001 = LOGGER;
       var10001.getClass();
       return (DataPackConfig)var10000.resultOrPartial(var10001::error).orElse(DataPackConfig.DEFAULT);
    }
 
+   public String getName() {
+      return "Anvil";
+   }
+
    public List<LevelSummary> getLevelList() throws LevelStorageException {
       if (!Files.isDirectory(this.baseDir, new LinkOption[0])) {
          throw new LevelStorageException((new TranslatableComponent("selectWorld.load_folder_access")).getString());
       } else {
          ArrayList var1 = Lists.newArrayList();
