--- net/minecraft/world/level/storage/McRegionUpgrader.java
+++ net/minecraft/world/level/storage/McRegionUpgrader.java
@@ -11,11 +11,10 @@
 import java.util.Collection;
 import java.util.Collections;
 import java.util.Iterator;
 import net.minecraft.core.Registry;
 import net.minecraft.core.RegistryAccess;
-import net.minecraft.core.WritableRegistry;
 import net.minecraft.nbt.CompoundTag;
 import net.minecraft.nbt.NbtIo;
 import net.minecraft.nbt.NbtOps;
 import net.minecraft.resources.RegistryReadOps;
 import net.minecraft.server.packs.resources.ResourceManager;
@@ -33,11 +32,15 @@
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
 
 public class McRegionUpgrader {
    private static final Logger LOGGER = LogManager.getLogger();
+   private static final String MCREGION_EXTENSION = ".mcr";
 
+   public McRegionUpgrader() {
+   }
+
    static boolean convertLevel(LevelStorageSource.LevelStorageAccess var0, ProgressListener var1) {
       var1.progressStagePercentage(0);
       ArrayList var2 = Lists.newArrayList();
       ArrayList var3 = Lists.newArrayList();
       ArrayList var4 = Lists.newArrayList();
@@ -58,11 +61,11 @@
       LOGGER.info("Total conversion count is {}", var8);
       RegistryAccess.RegistryHolder var9 = RegistryAccess.builtin();
       RegistryReadOps var10 = RegistryReadOps.create(NbtOps.INSTANCE, (ResourceManager)ResourceManager.Empty.INSTANCE, var9);
       WorldData var11 = var0.getDataTag(var10, DataPackConfig.DEFAULT);
       long var12 = var11 != null ? var11.worldGenSettings().seed() : 0L;
-      WritableRegistry var15 = var9.registryOrThrow(Registry.BIOME_REGISTRY);
+      Registry var15 = var9.registryOrThrow(Registry.BIOME_REGISTRY);
       Object var14;
       if (var11 != null && var11.worldGenSettings().isFlatWorld()) {
          var14 = new FixedBiomeSource((Biome)var15.getOrThrow(Biomes.PLAINS));
       } else {
          var14 = new OverworldBiomeSource(var12, false, false, var15);
