--- net/minecraft/world/level/storage/WorldData.java
+++ net/minecraft/world/level/storage/WorldData.java
@@ -12,10 +12,13 @@
 import net.minecraft.world.level.GameType;
 import net.minecraft.world.level.LevelSettings;
 import net.minecraft.world.level.levelgen.WorldGenSettings;
 
 public interface WorldData {
+   int ANVIL_VERSION_ID = 19133;
+   int MCREGION_VERSION_ID = 19132;
+
    DataPackConfig getDataPackConfig();
 
    void setDataPackConfig(DataPackConfig var1);
 
    boolean wasModded();
