--- net/minecraft/client/multiplayer/ClientLevel.java
+++ net/minecraft/client/multiplayer/ClientLevel.java
@@ -1,14 +1,11 @@
 package net.minecraft.client.multiplayer;
 
+import com.google.common.collect.ImmutableMap;
 import com.google.common.collect.Lists;
 import com.google.common.collect.Maps;
-import it.unimi.dsi.fastutil.ints.Int2ObjectMap;
-import it.unimi.dsi.fastutil.ints.Int2ObjectOpenHashMap;
-import it.unimi.dsi.fastutil.ints.Int2ObjectMap.Entry;
 import it.unimi.dsi.fastutil.objects.Object2ObjectArrayMap;
-import it.unimi.dsi.fastutil.objects.ObjectIterator;
 import java.util.Iterator;
 import java.util.List;
 import java.util.Map;
 import java.util.Random;
 import java.util.function.BooleanSupplier;
@@ -41,50 +38,62 @@
 import net.minecraft.server.MinecraftServer;
 import net.minecraft.sounds.SoundEvent;
 import net.minecraft.sounds.SoundSource;
 import net.minecraft.tags.BlockTags;
 import net.minecraft.tags.TagContainer;
+import net.minecraft.util.CubicSampler;
 import net.minecraft.util.Mth;
 import net.minecraft.util.profiling.ProfilerFiller;
 import net.minecraft.world.Difficulty;
 import net.minecraft.world.entity.Entity;
 import net.minecraft.world.entity.player.Player;
 import net.minecraft.world.item.ItemStack;
+import net.minecraft.world.item.Items;
 import net.minecraft.world.item.crafting.RecipeManager;
+import net.minecraft.world.level.ChunkPos;
 import net.minecraft.world.level.ColorResolver;
 import net.minecraft.world.level.EmptyTickList;
 import net.minecraft.world.level.GameRules;
 import net.minecraft.world.level.GameType;
 import net.minecraft.world.level.Level;
+import net.minecraft.world.level.LevelHeightAccessor;
 import net.minecraft.world.level.TickList;
 import net.minecraft.world.level.biome.Biome;
+import net.minecraft.world.level.biome.BiomeManager;
 import net.minecraft.world.level.biome.Biomes;
 import net.minecraft.world.level.block.Block;
 import net.minecraft.world.level.block.Blocks;
 import net.minecraft.world.level.block.state.BlockState;
-import net.minecraft.world.level.chunk.ChunkStatus;
 import net.minecraft.world.level.chunk.LevelChunk;
 import net.minecraft.world.level.dimension.DimensionType;
+import net.minecraft.world.level.entity.EntityTickList;
+import net.minecraft.world.level.entity.LevelCallback;
+import net.minecraft.world.level.entity.LevelEntityGetter;
+import net.minecraft.world.level.entity.TransientEntitySectionManager;
+import net.minecraft.world.level.gameevent.GameEvent;
 import net.minecraft.world.level.levelgen.Heightmap;
 import net.minecraft.world.level.material.Fluid;
 import net.minecraft.world.level.material.FluidState;
 import net.minecraft.world.level.saveddata.maps.MapItemSavedData;
 import net.minecraft.world.level.storage.WritableLevelData;
 import net.minecraft.world.phys.Vec3;
 import net.minecraft.world.phys.shapes.VoxelShape;
 import net.minecraft.world.scores.Scoreboard;
 
 public class ClientLevel extends Level {
-   private final Int2ObjectMap<Entity> entitiesById = new Int2ObjectOpenHashMap();
+   private static final double FLUID_PARTICLE_SPAWN_OFFSET = 0.05D;
+   final EntityTickList tickingEntities = new EntityTickList();
+   private final TransientEntitySectionManager<Entity> entityStorage = new TransientEntitySectionManager(Entity.class, new ClientLevel.EntityCallbacks());
    private final ClientPacketListener connection;
    private final LevelRenderer levelRenderer;
    private final ClientLevel.ClientLevelData clientLevelData;
    private final DimensionSpecialEffects effects;
    private final Minecraft minecraft = Minecraft.getInstance();
-   private final List<AbstractClientPlayer> players = Lists.newArrayList();
+   final List<AbstractClientPlayer> players = Lists.newArrayList();
    private Scoreboard scoreboard = new Scoreboard();
    private final Map<String, MapItemSavedData> mapData = Maps.newHashMap();
+   private static final long CLOUD_COLOR = 16777215L;
    private int skyFlashTime;
    private final Object2ObjectArrayMap<ColorResolver, BlockTintCache> tintCaches = (Object2ObjectArrayMap)Util.make(new Object2ObjectArrayMap(3), (var0) -> {
       var0.put(BiomeColors.GRASS_COLOR_RESOLVER, new BlockTintCache());
       var0.put(BiomeColors.FOLIAGE_COLOR_RESOLVER, new BlockTintCache());
       var0.put(BiomeColors.WATER_COLOR_RESOLVER, new BlockTintCache());
@@ -137,133 +146,72 @@
 
       this.clientLevelData.setDayTime(var1);
    }
 
    public Iterable<Entity> entitiesForRendering() {
-      return this.entitiesById.values();
+      return this.getEntities().getAll();
    }
 
    public void tickEntities() {
       ProfilerFiller var1 = this.getProfiler();
       var1.push("entities");
-      ObjectIterator var2 = this.entitiesById.int2ObjectEntrySet().iterator();
-
-      while(var2.hasNext()) {
-         Entry var3 = (Entry)var2.next();
-         Entity var4 = (Entity)var3.getValue();
-         if (!var4.isPassenger()) {
-            var1.push("tick");
-            if (!var4.removed) {
-               this.guardEntityTick(this::tickNonPassenger, var4);
-            }
-
-            var1.pop();
-            var1.push("remove");
-            if (var4.removed) {
-               var2.remove();
-               this.onEntityRemoved(var4);
-            }
-
-            var1.pop();
+      this.tickingEntities.forEach((var1x) -> {
+         if (!var1x.isRemoved() && !var1x.isPassenger()) {
+            this.guardEntityTick(this::tickNonPassenger, var1x);
          }
-      }
-
-      this.tickBlockEntities();
+      });
       var1.pop();
+      this.tickBlockEntities();
    }
 
    public void tickNonPassenger(Entity var1) {
-      if (!(var1 instanceof Player) && !this.getChunkSource().isEntityTickingChunk(var1)) {
-         this.updateChunkPos(var1);
-      } else {
-         var1.setPosAndOldPos(var1.getX(), var1.getY(), var1.getZ());
-         var1.yRotO = var1.yRot;
-         var1.xRotO = var1.xRot;
-         if (var1.inChunk || var1.isSpectator()) {
-            ++var1.tickCount;
-            this.getProfiler().push(() -> {
-               return Registry.ENTITY_TYPE.getKey(var1.getType()).toString();
-            });
-            var1.tick();
-            this.getProfiler().pop();
-         }
+      var1.setOldPosAndRot();
+      ++var1.tickCount;
+      this.getProfiler().push(() -> {
+         return Registry.ENTITY_TYPE.getKey(var1.getType()).toString();
+      });
+      var1.tick();
+      this.getProfiler().pop();
+      Iterator var2 = var1.getPassengers().iterator();
 
-         this.updateChunkPos(var1);
-         if (var1.inChunk) {
-            Iterator var2 = var1.getPassengers().iterator();
-
-            while(var2.hasNext()) {
-               Entity var3 = (Entity)var2.next();
-               this.tickPassenger(var1, var3);
-            }
-         }
-
+      while(var2.hasNext()) {
+         Entity var3 = (Entity)var2.next();
+         this.tickPassenger(var1, var3);
       }
+
    }
 
-   public void tickPassenger(Entity var1, Entity var2) {
-      if (!var2.removed && var2.getVehicle() == var1) {
-         if (var2 instanceof Player || this.getChunkSource().isEntityTickingChunk(var2)) {
-            var2.setPosAndOldPos(var2.getX(), var2.getY(), var2.getZ());
-            var2.yRotO = var2.yRot;
-            var2.xRotO = var2.xRot;
-            if (var2.inChunk) {
-               ++var2.tickCount;
-               var2.rideTick();
-            }
+   private void tickPassenger(Entity var1, Entity var2) {
+      if (!var2.isRemoved() && var2.getVehicle() == var1) {
+         if (var2 instanceof Player || this.tickingEntities.contains(var2)) {
+            var2.setOldPosAndRot();
+            ++var2.tickCount;
+            var2.rideTick();
+            Iterator var3 = var2.getPassengers().iterator();
 
-            this.updateChunkPos(var2);
-            if (var2.inChunk) {
-               Iterator var3 = var2.getPassengers().iterator();
-
-               while(var3.hasNext()) {
-                  Entity var4 = (Entity)var3.next();
-                  this.tickPassenger(var2, var4);
-               }
+            while(var3.hasNext()) {
+               Entity var4 = (Entity)var3.next();
+               this.tickPassenger(var2, var4);
             }
 
          }
       } else {
          var2.stopRiding();
       }
    }
 
-   private void updateChunkPos(Entity var1) {
-      if (var1.checkAndResetUpdateChunkPos()) {
-         this.getProfiler().push("chunkCheck");
-         int var2 = Mth.floor(var1.getX() / 16.0D);
-         int var3 = Mth.floor(var1.getY() / 16.0D);
-         int var4 = Mth.floor(var1.getZ() / 16.0D);
-         if (!var1.inChunk || var1.xChunk != var2 || var1.yChunk != var3 || var1.zChunk != var4) {
-            if (var1.inChunk && this.hasChunk(var1.xChunk, var1.zChunk)) {
-               this.getChunk(var1.xChunk, var1.zChunk).removeEntity(var1, var1.yChunk);
-            }
-
-            if (!var1.checkAndResetForcedChunkAdditionFlag() && !this.hasChunk(var2, var4)) {
-               if (var1.inChunk) {
-                  LOGGER.warn("Entity {} left loaded chunk area", var1);
-               }
-
-               var1.inChunk = false;
-            } else {
-               this.getChunk(var2, var4).addEntity(var1);
-            }
-         }
-
-         this.getProfiler().pop();
-      }
-   }
-
    public void unload(LevelChunk var1) {
-      this.blockEntitiesToUnload.addAll(var1.getBlockEntities().values());
+      var1.invalidateAllBlockEntities();
       this.chunkSource.getLightEngine().enableLightSources(var1.getPos(), false);
+      this.entityStorage.stopTicking(var1.getPos());
    }
 
-   public void onChunkLoaded(int var1, int var2) {
-      this.tintCaches.forEach((var2x, var3) -> {
-         var3.invalidateForChunk(var1, var2);
+   public void onChunkLoaded(ChunkPos var1) {
+      this.tintCaches.forEach((var1x, var2) -> {
+         var2.invalidateForChunk(var1.x, var1.z);
       });
+      this.entityStorage.startTicking(var1);
    }
 
    public void clearTintCaches() {
       this.tintCaches.forEach((var0, var1) -> {
          var1.invalidateAll();
@@ -273,64 +221,38 @@
    public boolean hasChunk(int var1, int var2) {
       return true;
    }
 
    public int getEntityCount() {
-      return this.entitiesById.size();
+      return this.entityStorage.count();
    }
 
    public void addPlayer(int var1, AbstractClientPlayer var2) {
       this.addEntity(var1, var2);
-      this.players.add(var2);
    }
 
    public void putNonPlayerEntity(int var1, Entity var2) {
       this.addEntity(var1, var2);
    }
 
    private void addEntity(int var1, Entity var2) {
-      this.removeEntity(var1);
-      this.entitiesById.put(var1, var2);
-      this.getChunkSource().getChunk(Mth.floor(var2.getX() / 16.0D), Mth.floor(var2.getZ() / 16.0D), ChunkStatus.FULL, true).addEntity(var2);
+      this.removeEntity(var1, Entity.RemovalReason.DISCARDED);
+      this.entityStorage.addEntity(var2);
    }
 
-   public void removeEntity(int var1) {
-      Entity var2 = (Entity)this.entitiesById.remove(var1);
-      if (var2 != null) {
-         var2.remove();
-         this.onEntityRemoved(var2);
+   public void removeEntity(int var1, Entity.RemovalReason var2) {
+      Entity var3 = (Entity)this.getEntities().get(var1);
+      if (var3 != null) {
+         var3.setRemoved(var2);
+         var3.onClientRemoval();
       }
 
    }
 
-   private void onEntityRemoved(Entity var1) {
-      var1.unRide();
-      if (var1.inChunk) {
-         this.getChunk(var1.xChunk, var1.zChunk).removeEntity(var1);
-      }
-
-      this.players.remove(var1);
-   }
-
-   public void reAddEntitiesToChunk(LevelChunk var1) {
-      ObjectIterator var2 = this.entitiesById.int2ObjectEntrySet().iterator();
-
-      while(var2.hasNext()) {
-         Entry var3 = (Entry)var2.next();
-         Entity var4 = (Entity)var3.getValue();
-         int var5 = Mth.floor(var4.getX() / 16.0D);
-         int var6 = Mth.floor(var4.getZ() / 16.0D);
-         if (var5 == var1.getPos().x && var6 == var1.getPos().z) {
-            var1.addEntity(var4);
-         }
-      }
-
-   }
-
    @Nullable
    public Entity getEntity(int var1) {
-      return (Entity)this.entitiesById.get(var1);
+      return (Entity)this.getEntities().get(var1);
    }
 
    public void setKnownState(BlockPos var1, BlockState var2) {
       this.setBlock(var1, var2, 19);
    }
@@ -340,33 +262,37 @@
    }
 
    public void animateTick(int var1, int var2, int var3) {
       boolean var4 = true;
       Random var5 = new Random();
-      boolean var6 = false;
-      if (this.minecraft.gameMode.getPlayerMode() == GameType.CREATIVE) {
-         Iterator var7 = this.minecraft.player.getHandSlots().iterator();
+      ClientLevel.MarkerParticleStatus var6 = this.getMarkerParticleStatus();
+      BlockPos.MutableBlockPos var7 = new BlockPos.MutableBlockPos();
 
-         while(var7.hasNext()) {
-            ItemStack var8 = (ItemStack)var7.next();
-            if (var8.getItem() == Blocks.BARRIER.asItem()) {
-               var6 = true;
-               break;
-            }
-         }
+      for(int var8 = 0; var8 < 667; ++var8) {
+         this.doAnimateTick(var1, var2, var3, 16, var5, var6, var7);
+         this.doAnimateTick(var1, var2, var3, 32, var5, var6, var7);
       }
 
-      BlockPos.MutableBlockPos var9 = new BlockPos.MutableBlockPos();
+   }
 
-      for(int var10 = 0; var10 < 667; ++var10) {
-         this.doAnimateTick(var1, var2, var3, 16, var5, var6, var9);
-         this.doAnimateTick(var1, var2, var3, 32, var5, var6, var9);
+   @Nullable
+   private ClientLevel.MarkerParticleStatus getMarkerParticleStatus() {
+      if (this.minecraft.gameMode.getPlayerMode() == GameType.CREATIVE) {
+         ItemStack var1 = this.minecraft.player.getMainHandItem();
+         if (var1.getItem() == Items.BARRIER) {
+            return ClientLevel.MarkerParticleStatus.BARRIER;
+         }
+
+         if (var1.getItem() == Items.LIGHT) {
+            return ClientLevel.MarkerParticleStatus.LIGHT;
+         }
       }
 
+      return null;
    }
 
-   public void doAnimateTick(int var1, int var2, int var3, int var4, Random var5, boolean var6, BlockPos.MutableBlockPos var7) {
+   public void doAnimateTick(int var1, int var2, int var3, int var4, Random var5, @Nullable ClientLevel.MarkerParticleStatus var6, BlockPos.MutableBlockPos var7) {
       int var8 = var1 + this.random.nextInt(var4) - this.random.nextInt(var4);
       int var9 = var2 + this.random.nextInt(var4) - this.random.nextInt(var4);
       int var10 = var3 + this.random.nextInt(var4) - this.random.nextInt(var4);
       var7.set(var8, var9, var10);
       BlockState var11 = this.getBlockState(var7);
@@ -380,12 +306,12 @@
             BlockPos var15 = var7.below();
             this.trySpawnDripParticles(var15, this.getBlockState(var15), var13, var14);
          }
       }
 
-      if (var6 && var11.is(Blocks.BARRIER)) {
-         this.addParticle(ParticleTypes.BARRIER, (double)var8 + 0.5D, (double)var9 + 0.5D, (double)var10 + 0.5D, 0.0D, 0.0D, 0.0D);
+      if (var6 != null && var11.getBlock() == var6.block) {
+         this.addParticle(var6.particle, (double)var8 + 0.5D, (double)var9 + 0.5D, (double)var10 + 0.5D, 0.0D, 0.0D, 0.0D);
       }
 
       if (!var11.isCollisionShapeFullBlock(this, var7)) {
          this.getBiome(var7).getAmbientParticle().ifPresent((var2x) -> {
             if (var2x.canSpawn(this.random)) {
@@ -429,24 +355,10 @@
 
    private void spawnFluidParticle(double var1, double var3, double var5, double var7, double var9, ParticleOptions var11) {
       this.addParticle(var11, Mth.lerp(this.random.nextDouble(), var1, var3), var9, Mth.lerp(this.random.nextDouble(), var5, var7), 0.0D, 0.0D, 0.0D);
    }
 
-   public void removeAllPendingEntityRemovals() {
-      ObjectIterator var1 = this.entitiesById.int2ObjectEntrySet().iterator();
-
-      while(var1.hasNext()) {
-         Entry var2 = (Entry)var1.next();
-         Entity var3 = (Entity)var2.getValue();
-         if (var3.removed) {
-            var1.remove();
-            this.onEntityRemoved(var3);
-         }
-      }
-
-   }
-
    public CrashReportCategory fillReportDetails(CrashReport var1) {
       CrashReportCategory var2 = super.fillReportDetails(var1);
       var2.setDetail("Server brand", () -> {
          return this.minecraft.player.getServerBrand();
       });
@@ -463,11 +375,11 @@
 
    }
 
    public void playSound(@Nullable Player var1, Entity var2, SoundEvent var3, SoundSource var4, float var5, float var6) {
       if (var1 == this.minecraft.player) {
-         this.minecraft.getSoundManager().play(new EntityBoundSoundInstance(var3, var4, var2));
+         this.minecraft.getSoundManager().play(new EntityBoundSoundInstance(var3, var4, var5, var6, var2));
       }
 
    }
 
    public void playLocalSound(BlockPos var1, SoundEvent var2, SoundSource var3, float var4, float var5, boolean var6) {
@@ -517,12 +429,12 @@
    @Nullable
    public MapItemSavedData getMapData(String var1) {
       return (MapItemSavedData)this.mapData.get(var1);
    }
 
-   public void setMapData(MapItemSavedData var1) {
-      this.mapData.put(var1.getId(), var1);
+   public void setMapData(String var1, MapItemSavedData var2) {
+      this.mapData.put(var1, var2);
    }
 
    public int getFreeMapId() {
       return 0;
    }
@@ -563,11 +475,11 @@
       try {
          this.levelRenderer.levelEvent(var1, var2, var3, var4);
       } catch (Throwable var8) {
          CrashReport var6 = CrashReport.forThrowable(var8, "Playing level event");
          CrashReportCategory var7 = var6.addCategory("Level event being played");
-         var7.setDetail("Block coordinates", (Object)CrashReportCategory.formatLocation(var3));
+         var7.setDetail("Block coordinates", (Object)CrashReportCategory.formatLocation(this, var3));
          var7.setDetail("Event source", (Object)var1);
          var7.setDetail("Event type", (Object)var2);
          var7.setDetail("Event data", (Object)var4);
          throw new ReportedException(var6);
       }
@@ -605,55 +517,55 @@
       var3 = (float)((double)var3 * (1.0D - (double)(this.getRainLevel(var1) * 5.0F) / 16.0D));
       var3 = (float)((double)var3 * (1.0D - (double)(this.getThunderLevel(var1) * 5.0F) / 16.0D));
       return var3 * 0.8F + 0.2F;
    }
 
-   public Vec3 getSkyColor(BlockPos var1, float var2) {
+   public Vec3 getSkyColor(Vec3 var1, float var2) {
       float var3 = this.getTimeOfDay(var2);
-      float var4 = Mth.cos(var3 * 6.2831855F) * 2.0F + 0.5F;
-      var4 = Mth.clamp(var4, 0.0F, 1.0F);
-      Biome var5 = this.getBiome(var1);
-      int var6 = var5.getSkyColor();
-      float var7 = (float)(var6 >> 16 & 255) / 255.0F;
-      float var8 = (float)(var6 >> 8 & 255) / 255.0F;
-      float var9 = (float)(var6 & 255) / 255.0F;
-      var7 *= var4;
-      var8 *= var4;
-      var9 *= var4;
-      float var10 = this.getRainLevel(var2);
-      float var11;
+      Vec3 var4 = var1.subtract(2.0D, 2.0D, 2.0D).scale(0.25D);
+      BiomeManager var5 = this.getBiomeManager();
+      Vec3 var6 = CubicSampler.gaussianSampleVec3(var4, (var1x, var2x, var3x) -> {
+         return Vec3.fromRGB24(var5.getNoiseBiomeAtQuart(var1x, var2x, var3x).getSkyColor());
+      });
+      float var7 = Mth.cos(var3 * 6.2831855F) * 2.0F + 0.5F;
+      var7 = Mth.clamp(var7, 0.0F, 1.0F);
+      float var8 = (float)var6.x * var7;
+      float var9 = (float)var6.y * var7;
+      float var10 = (float)var6.z * var7;
+      float var11 = this.getRainLevel(var2);
       float var12;
-      if (var10 > 0.0F) {
-         var11 = (var7 * 0.3F + var8 * 0.59F + var9 * 0.11F) * 0.6F;
-         var12 = 1.0F - var10 * 0.75F;
-         var7 = var7 * var12 + var11 * (1.0F - var12);
-         var8 = var8 * var12 + var11 * (1.0F - var12);
-         var9 = var9 * var12 + var11 * (1.0F - var12);
-      }
-
-      var11 = this.getThunderLevel(var2);
+      float var13;
       if (var11 > 0.0F) {
-         var12 = (var7 * 0.3F + var8 * 0.59F + var9 * 0.11F) * 0.2F;
-         float var13 = 1.0F - var11 * 0.75F;
-         var7 = var7 * var13 + var12 * (1.0F - var13);
+         var12 = (var8 * 0.3F + var9 * 0.59F + var10 * 0.11F) * 0.6F;
+         var13 = 1.0F - var11 * 0.75F;
          var8 = var8 * var13 + var12 * (1.0F - var13);
          var9 = var9 * var13 + var12 * (1.0F - var13);
+         var10 = var10 * var13 + var12 * (1.0F - var13);
       }
 
+      var12 = this.getThunderLevel(var2);
+      if (var12 > 0.0F) {
+         var13 = (var8 * 0.3F + var9 * 0.59F + var10 * 0.11F) * 0.2F;
+         float var14 = 1.0F - var12 * 0.75F;
+         var8 = var8 * var14 + var13 * (1.0F - var14);
+         var9 = var9 * var14 + var13 * (1.0F - var14);
+         var10 = var10 * var14 + var13 * (1.0F - var14);
+      }
+
       if (this.skyFlashTime > 0) {
-         var12 = (float)this.skyFlashTime - var2;
-         if (var12 > 1.0F) {
-            var12 = 1.0F;
+         var13 = (float)this.skyFlashTime - var2;
+         if (var13 > 1.0F) {
+            var13 = 1.0F;
          }
 
-         var12 *= 0.45F;
-         var7 = var7 * (1.0F - var12) + 0.8F * var12;
-         var8 = var8 * (1.0F - var12) + 0.8F * var12;
-         var9 = var9 * (1.0F - var12) + 1.0F * var12;
+         var13 *= 0.45F;
+         var8 = var8 * (1.0F - var13) + 0.8F * var13;
+         var9 = var9 * (1.0F - var13) + 0.8F * var13;
+         var10 = var10 * (1.0F - var13) + 1.0F * var13;
       }
 
-      return new Vec3((double)var7, (double)var8, (double)var9);
+      return new Vec3((double)var8, (double)var9, (double)var10);
    }
 
    public Vec3 getCloudColor(float var1) {
       float var2 = this.getTimeOfDay(var1);
       float var3 = Mth.cos(var2 * 6.2831855F) * 2.0F + 0.5F;
@@ -777,10 +689,65 @@
 
    public ClientLevel.ClientLevelData getLevelData() {
       return this.clientLevelData;
    }
 
+   public void gameEvent(@Nullable Entity var1, GameEvent var2, BlockPos var3) {
+   }
+
+   protected Map<String, MapItemSavedData> getAllMapData() {
+      return ImmutableMap.copyOf(this.mapData);
+   }
+
+   protected void addMapData(Map<String, MapItemSavedData> var1) {
+      this.mapData.putAll(var1);
+   }
+
+   protected LevelEntityGetter<Entity> getEntities() {
+      return this.entityStorage.getEntityGetter();
+   }
+
+   public String gatherChunkSourceStats() {
+      String var10000 = this.chunkSource.gatherStats();
+      return "Chunks[C] W: " + var10000 + " E: " + this.entityStorage.gatherStats();
+   }
+
+   public void addDestroyBlockEffect(BlockPos var1, BlockState var2) {
+      this.minecraft.particleEngine.destroy(var1, var2);
+   }
+
+   private final class EntityCallbacks implements LevelCallback<Entity> {
+      EntityCallbacks() {
+      }
+
+      public void onCreated(Entity var1) {
+      }
+
+      public void onDestroyed(Entity var1) {
+      }
+
+      public void onTickingStart(Entity var1) {
+         ClientLevel.this.tickingEntities.add(var1);
+      }
+
+      public void onTickingEnd(Entity var1) {
+         ClientLevel.this.tickingEntities.remove(var1);
+      }
+
+      public void onTrackingStart(Entity var1) {
+         if (var1 instanceof AbstractClientPlayer) {
+            ClientLevel.this.players.add((AbstractClientPlayer)var1);
+         }
+
+      }
+
+      public void onTrackingEnd(Entity var1) {
+         var1.unRide();
+         ClientLevel.this.players.remove(var1);
+      }
+   }
+
    public static class ClientLevelData implements WritableLevelData {
       private final boolean hardcore;
       private final GameRules gameRules;
       private final boolean isFlat;
       private int xSpawn;
@@ -881,26 +848,39 @@
 
       public boolean isDifficultyLocked() {
          return this.difficultyLocked;
       }
 
-      public void fillCrashReportCategory(CrashReportCategory var1) {
-         WritableLevelData.super.fillCrashReportCategory(var1);
+      public void fillCrashReportCategory(CrashReportCategory var1, LevelHeightAccessor var2) {
+         WritableLevelData.super.fillCrashReportCategory(var1, var2);
       }
 
       public void setDifficulty(Difficulty var1) {
          this.difficulty = var1;
       }
 
       public void setDifficultyLocked(boolean var1) {
          this.difficultyLocked = var1;
       }
 
-      public double getHorizonHeight() {
-         return this.isFlat ? 0.0D : 63.0D;
+      public double getHorizonHeight(LevelHeightAccessor var1) {
+         return this.isFlat ? (double)var1.getMinBuildHeight() : 63.0D;
       }
 
       public double getClearColorScale() {
          return this.isFlat ? 1.0D : 0.03125D;
       }
    }
+
+   static enum MarkerParticleStatus {
+      BARRIER(Blocks.BARRIER, ParticleTypes.BARRIER),
+      LIGHT(Blocks.LIGHT, ParticleTypes.LIGHT);
+
+      final Block block;
+      final ParticleOptions particle;
+
+      private MarkerParticleStatus(Block var3, ParticleOptions var4) {
+         this.block = var3;
+         this.particle = var4;
+      }
+   }
 }
