--- net/minecraft/client/multiplayer/PlayerInfo.java
+++ net/minecraft/client/multiplayer/PlayerInfo.java
@@ -56,10 +56,14 @@
 
    protected void setLatency(int var1) {
       this.latency = var1;
    }
 
+   public boolean isCapeLoaded() {
+      return this.getCapeLocation() != null;
+   }
+
    public boolean isSkinLoaded() {
       return this.getSkinLocation() != null;
    }
 
    public String getModelName() {
