--- net/minecraft/client/multiplayer/PlayerInfo.java
+++ net/minecraft/client/multiplayer/PlayerInfo.java
@@ -56,21 +56,25 @@
 
    protected void setLatency(int var1) {
       this.latency = var1;
    }
 
+   public boolean isCapeLoaded() {
+      return this.getCapeLocation() != null;
+   }
+
    public boolean isSkinLoaded() {
       return this.getSkinLocation() != null;
    }
 
    public String getModelName() {
       return this.skinModel == null ? DefaultPlayerSkin.getSkinModelName(this.profile.getId()) : this.skinModel;
    }
 
    public ResourceLocation getSkinLocation() {
       this.registerTextures();
-      return (ResourceLocation)MoreObjects.firstNonNull(this.textureLocations.get(Type.SKIN), DefaultPlayerSkin.getDefaultSkin(this.profile.getId()));
+      return (ResourceLocation)MoreObjects.firstNonNull((ResourceLocation)this.textureLocations.get(Type.SKIN), DefaultPlayerSkin.getDefaultSkin(this.profile.getId()));
    }
 
    @Nullable
    public ResourceLocation getCapeLocation() {
       this.registerTextures();
