--- net/minecraft/world/inventory/AbstractFurnaceMenu.java
+++ net/minecraft/world/inventory/AbstractFurnaceMenu.java
@@ -1,9 +1,7 @@
 package net.minecraft.world.inventory;
 
-import net.minecraft.recipebook.ServerPlaceSmeltingRecipe;
-import net.minecraft.server.level.ServerPlayer;
 import net.minecraft.world.Container;
 import net.minecraft.world.SimpleContainer;
 import net.minecraft.world.entity.player.Inventory;
 import net.minecraft.world.entity.player.Player;
 import net.minecraft.world.entity.player.StackedContents;
@@ -58,17 +56,14 @@
       }
 
    }
 
    public void clearCraftingContent() {
-      this.container.clearContent();
+      this.getSlot(0).set(ItemStack.EMPTY);
+      this.getSlot(2).set(ItemStack.EMPTY);
    }
 
-   public void handlePlacement(boolean var1, Recipe<?> var2, ServerPlayer var3) {
-      (new ServerPlaceSmeltingRecipe(this)).recipeClicked(var3, var2, var1);
-   }
-
    public boolean recipeMatches(Recipe<? super Container> var1) {
       return var1.matches(this.container, this.level);
    }
 
    public int getResultSlotIndex() {
@@ -167,6 +162,10 @@
    }
 
    public RecipeBookType getRecipeBookType() {
       return this.recipeBookType;
    }
+
+   public boolean shouldMoveToInventory(int var1) {
+      return var1 != 1;
+   }
 }
