--- net/minecraft/world/inventory/BeaconMenu.java
+++ net/minecraft/world/inventory/BeaconMenu.java
@@ -1,9 +1,10 @@
 package net.minecraft.world.inventory;
 
 import javax.annotation.Nullable;
 import net.minecraft.tags.ItemTags;
+import net.minecraft.tags.Tag;
 import net.minecraft.world.Container;
 import net.minecraft.world.SimpleContainer;
 import net.minecraft.world.effect.MobEffect;
 import net.minecraft.world.entity.player.Player;
 import net.minecraft.world.item.ItemStack;
@@ -21,11 +22,11 @@
 
    public BeaconMenu(int var1, Container var2, ContainerData var3, ContainerLevelAccess var4) {
       super(MenuType.BEACON, var1);
       this.beacon = new SimpleContainer(1) {
          public boolean canPlaceItem(int var1, ItemStack var2) {
-            return var2.getItem().is(ItemTags.BEACON_PAYMENT_ITEMS);
+            return var2.is((Tag)ItemTags.BEACON_PAYMENT_ITEMS);
          }
 
          public int getMaxStackSize() {
             return 1;
          }
@@ -147,11 +148,11 @@
       public PaymentSlot(Container var2, int var3, int var4, int var5) {
          super(var2, var3, var4, var5);
       }
 
       public boolean mayPlace(ItemStack var1) {
-         return var1.getItem().is(ItemTags.BEACON_PAYMENT_ITEMS);
+         return var1.is((Tag)ItemTags.BEACON_PAYMENT_ITEMS);
       }
 
       public int getMaxStackSize() {
          return 1;
       }
