--- net/minecraft/world/inventory/BrewingStandMenu.java
+++ net/minecraft/world/inventory/BrewingStandMenu.java
@@ -4,18 +4,27 @@
 import net.minecraft.server.level.ServerPlayer;
 import net.minecraft.world.Container;
 import net.minecraft.world.SimpleContainer;
 import net.minecraft.world.entity.player.Inventory;
 import net.minecraft.world.entity.player.Player;
-import net.minecraft.world.item.Item;
 import net.minecraft.world.item.ItemStack;
 import net.minecraft.world.item.Items;
 import net.minecraft.world.item.alchemy.Potion;
 import net.minecraft.world.item.alchemy.PotionBrewing;
 import net.minecraft.world.item.alchemy.PotionUtils;
 
 public class BrewingStandMenu extends AbstractContainerMenu {
+   private static final int BOTTLE_SLOT_START = 0;
+   private static final int BOTTLE_SLOT_END = 2;
+   private static final int INGREDIENT_SLOT = 3;
+   private static final int FUEL_SLOT = 4;
+   private static final int SLOT_COUNT = 5;
+   private static final int DATA_COUNT = 2;
+   private static final int INV_SLOT_START = 5;
+   private static final int INV_SLOT_END = 32;
+   private static final int USE_ROW_SLOT_START = 32;
+   private static final int USE_ROW_SLOT_END = 41;
    private final Container brewingStand;
    private final ContainerData brewingStandData;
    private final Slot ingredientSlot;
 
    public BrewingStandMenu(int var1, Inventory var2) {
@@ -112,26 +121,35 @@
 
    public int getBrewingTicks() {
       return this.brewingStandData.get(0);
    }
 
-   static class FuelSlot extends Slot {
-      public FuelSlot(Container var1, int var2, int var3, int var4) {
+   private static class PotionSlot extends Slot {
+      public PotionSlot(Container var1, int var2, int var3, int var4) {
          super(var1, var2, var3, var4);
       }
 
       public boolean mayPlace(ItemStack var1) {
          return mayPlaceItem(var1);
       }
 
-      public static boolean mayPlaceItem(ItemStack var0) {
-         return var0.getItem() == Items.BLAZE_POWDER;
+      public int getMaxStackSize() {
+         return 1;
       }
 
-      public int getMaxStackSize() {
-         return 64;
+      public void onTake(Player var1, ItemStack var2) {
+         Potion var3 = PotionUtils.getPotion(var2);
+         if (var1 instanceof ServerPlayer) {
+            CriteriaTriggers.BREWED_POTION.trigger((ServerPlayer)var1, var3);
+         }
+
+         super.onTake(var1, var2);
       }
+
+      public static boolean mayPlaceItem(ItemStack var0) {
+         return var0.is(Items.POTION) || var0.is(Items.SPLASH_POTION) || var0.is(Items.LINGERING_POTION) || var0.is(Items.GLASS_BOTTLE);
+      }
    }
 
    static class IngredientsSlot extends Slot {
       public IngredientsSlot(Container var1, int var2, int var3, int var4) {
          super(var1, var2, var3, var4);
@@ -144,34 +162,23 @@
       public int getMaxStackSize() {
          return 64;
       }
    }
 
-   static class PotionSlot extends Slot {
-      public PotionSlot(Container var1, int var2, int var3, int var4) {
+   static class FuelSlot extends Slot {
+      public FuelSlot(Container var1, int var2, int var3, int var4) {
          super(var1, var2, var3, var4);
       }
 
       public boolean mayPlace(ItemStack var1) {
          return mayPlaceItem(var1);
       }
 
-      public int getMaxStackSize() {
-         return 1;
+      public static boolean mayPlaceItem(ItemStack var0) {
+         return var0.is(Items.BLAZE_POWDER);
       }
 
-      public ItemStack onTake(Player var1, ItemStack var2) {
-         Potion var3 = PotionUtils.getPotion(var2);
-         if (var1 instanceof ServerPlayer) {
-            CriteriaTriggers.BREWED_POTION.trigger((ServerPlayer)var1, var3);
-         }
-
-         super.onTake(var1, var2);
-         return var2;
+      public int getMaxStackSize() {
+         return 64;
       }
-
-      public static boolean mayPlaceItem(ItemStack var0) {
-         Item var1 = var0.getItem();
-         return var1 == Items.POTION || var1 == Items.SPLASH_POTION || var1 == Items.LINGERING_POTION || var1 == Items.GLASS_BOTTLE;
-      }
    }
 }
