--- net/minecraft/world/inventory/BrewingStandMenu.java
+++ net/minecraft/world/inventory/BrewingStandMenu.java
@@ -4,11 +4,10 @@
 import net.minecraft.server.level.ServerPlayer;
 import net.minecraft.world.Container;
 import net.minecraft.world.SimpleContainer;
 import net.minecraft.world.entity.player.Inventory;
 import net.minecraft.world.entity.player.Player;
-import net.minecraft.world.item.Item;
 import net.minecraft.world.item.ItemStack;
 import net.minecraft.world.item.Items;
 import net.minecraft.world.item.alchemy.Potion;
 import net.minecraft.world.item.alchemy.PotionBrewing;
 import net.minecraft.world.item.alchemy.PotionUtils;
@@ -122,11 +121,11 @@
       public boolean mayPlace(ItemStack var1) {
          return mayPlaceItem(var1);
       }
 
       public static boolean mayPlaceItem(ItemStack var0) {
-         return var0.getItem() == Items.BLAZE_POWDER;
+         return var0.is(Items.BLAZE_POWDER);
       }
 
       public int getMaxStackSize() {
          return 64;
       }
@@ -157,21 +156,19 @@
 
       public int getMaxStackSize() {
          return 1;
       }
 
-      public ItemStack onTake(Player var1, ItemStack var2) {
+      public void onTake(Player var1, ItemStack var2) {
          Potion var3 = PotionUtils.getPotion(var2);
          if (var1 instanceof ServerPlayer) {
             CriteriaTriggers.BREWED_POTION.trigger((ServerPlayer)var1, var3);
          }
 
          super.onTake(var1, var2);
-         return var2;
       }
 
       public static boolean mayPlaceItem(ItemStack var0) {
-         Item var1 = var0.getItem();
-         return var1 == Items.POTION || var1 == Items.SPLASH_POTION || var1 == Items.LINGERING_POTION || var1 == Items.GLASS_BOTTLE;
+         return var0.is(Items.POTION) || var0.is(Items.SPLASH_POTION) || var0.is(Items.LINGERING_POTION) || var0.is(Items.GLASS_BOTTLE);
       }
    }
 }
