--- net/minecraft/world/inventory/CraftingMenu.java
+++ net/minecraft/world/inventory/CraftingMenu.java
@@ -89,11 +89,11 @@
    }
 
    public void removed(Player var1) {
       super.removed(var1);
       this.access.execute((var2, var3) -> {
-         this.clearContainer(var1, var2, this.craftSlots);
+         this.clearContainer(var1, this.craftSlots);
       });
    }
 
    public boolean stillValid(Player var1) {
       return stillValid(this.access, var1, Blocks.CRAFTING_TABLE);
@@ -168,6 +168,10 @@
    }
 
    public RecipeBookType getRecipeBookType() {
       return RecipeBookType.CRAFTING;
    }
+
+   public boolean shouldMoveToInventory(int var1) {
+      return var1 != this.getResultSlotIndex();
+   }
 }
