--- net/minecraft/world/inventory/DispenserMenu.java
+++ net/minecraft/world/inventory/DispenserMenu.java
@@ -5,10 +5,15 @@
 import net.minecraft.world.entity.player.Inventory;
 import net.minecraft.world.entity.player.Player;
 import net.minecraft.world.item.ItemStack;
 
 public class DispenserMenu extends AbstractContainerMenu {
+   private static final int SLOT_COUNT = 9;
+   private static final int INV_SLOT_START = 9;
+   private static final int INV_SLOT_END = 36;
+   private static final int USE_ROW_SLOT_START = 36;
+   private static final int USE_ROW_SLOT_END = 45;
    private final Container dispenser;
 
    public DispenserMenu(int var1, Inventory var2) {
       this(var1, var2, new SimpleContainer(9));
    }
