--- net/minecraft/world/inventory/InventoryMenu.java
+++ net/minecraft/world/inventory/InventoryMenu.java
@@ -98,11 +98,11 @@
 
    public void removed(Player var1) {
       super.removed(var1);
       this.resultSlots.clearContent();
       if (!var1.level.isClientSide) {
-         this.clearContainer(var1, var1.level, this.craftSlots);
+         this.clearContainer(var1, this.craftSlots);
       }
    }
 
    public boolean stillValid(Player var1) {
       return true;
