--- net/minecraft/world/inventory/ItemCombinerMenu.java
+++ net/minecraft/world/inventory/ItemCombinerMenu.java
@@ -19,11 +19,11 @@
    protected final ContainerLevelAccess access;
    protected final Player player;
 
    protected abstract boolean mayPickup(Player var1, boolean var2);
 
-   protected abstract ItemStack onTake(Player var1, ItemStack var2);
+   protected abstract void onTake(Player var1, ItemStack var2);
 
    protected abstract boolean isValidBlock(BlockState var1);
 
    public ItemCombinerMenu(@Nullable MenuType<?> var1, int var2, Inventory var3, ContainerLevelAccess var4) {
       super(var1, var2);
@@ -38,12 +38,12 @@
 
          public boolean mayPickup(Player var1) {
             return ItemCombinerMenu.this.mayPickup(var1, this.hasItem());
          }
 
-         public ItemStack onTake(Player var1, ItemStack var2) {
-            return ItemCombinerMenu.this.onTake(var1, var2);
+         public void onTake(Player var1, ItemStack var2) {
+            ItemCombinerMenu.this.onTake(var1, var2);
          }
       });
 
       int var5;
       for(var5 = 0; var5 < 3; ++var5) {
@@ -69,11 +69,11 @@
    }
 
    public void removed(Player var1) {
       super.removed(var1);
       this.access.execute((var2, var3) -> {
-         this.clearContainer(var1, var2, this.inputSlots);
+         this.clearContainer(var1, this.inputSlots);
       });
    }
 
    public boolean stillValid(Player var1) {
       return (Boolean)this.access.evaluate((var2, var3) -> {
