--- net/minecraft/world/inventory/LecternMenu.java
+++ net/minecraft/world/inventory/LecternMenu.java
@@ -49,11 +49,11 @@
                return false;
             }
 
             ItemStack var3 = this.lectern.removeItemNoUpdate(0);
             this.lectern.setChanged();
-            if (!var1.inventory.add(var3)) {
+            if (!var1.getInventory().add(var3)) {
                var1.drop(var3, false);
             }
 
             return true;
          default:
