--- net/minecraft/world/inventory/LecternMenu.java
+++ net/minecraft/world/inventory/LecternMenu.java
@@ -4,10 +4,16 @@
 import net.minecraft.world.SimpleContainer;
 import net.minecraft.world.entity.player.Player;
 import net.minecraft.world.item.ItemStack;
 
 public class LecternMenu extends AbstractContainerMenu {
+   private static final int DATA_COUNT = 1;
+   private static final int SLOT_COUNT = 1;
+   public static final int BUTTON_PREV_PAGE = 1;
+   public static final int BUTTON_NEXT_PAGE = 2;
+   public static final int BUTTON_TAKE_BOOK = 3;
+   public static final int BUTTON_PAGE_JUMP_RANGE_START = 100;
    private final Container lectern;
    private final ContainerData lecternData;
 
    public LecternMenu(int var1) {
       this(var1, new SimpleContainer(1), new SimpleContainerData(1));
@@ -49,11 +55,11 @@
                return false;
             }
 
             ItemStack var3 = this.lectern.removeItemNoUpdate(0);
             this.lectern.setChanged();
-            if (!var1.inventory.add(var3)) {
+            if (!var1.getInventory().add(var3)) {
                var1.drop(var3, false);
             }
 
             return true;
          default:
