--- net/minecraft/world/inventory/LoomMenu.java
+++ net/minecraft/world/inventory/LoomMenu.java
@@ -16,10 +16,14 @@
 import net.minecraft.world.item.ItemStack;
 import net.minecraft.world.level.block.Blocks;
 import net.minecraft.world.level.block.entity.BannerPattern;
 
 public class LoomMenu extends AbstractContainerMenu {
+   private static final int INV_SLOT_START = 4;
+   private static final int INV_SLOT_END = 31;
+   private static final int USE_ROW_SLOT_START = 31;
+   private static final int USE_ROW_SLOT_END = 40;
    private final ContainerLevelAccess access;
    private final DataSlot selectedBannerPatternIndex;
    private Runnable slotUpdateListener;
    private final Slot bannerSlot;
    private final Slot dyeSlot;
@@ -70,11 +74,11 @@
       this.resultSlot = this.addSlot(new Slot(this.outputContainer, 0, 143, 58) {
          public boolean mayPlace(ItemStack var1) {
             return false;
          }
 
-         public ItemStack onTake(Player var1, ItemStack var2) {
+         public void onTake(Player var1, ItemStack var2) {
             LoomMenu.this.bannerSlot.remove(1);
             LoomMenu.this.dyeSlot.remove(1);
             if (!LoomMenu.this.bannerSlot.hasItem() || !LoomMenu.this.dyeSlot.hasItem()) {
                LoomMenu.this.selectedBannerPatternIndex.set(0);
             }
@@ -85,11 +89,11 @@
                   var1x.playSound((Player)null, (BlockPos)var2x, SoundEvents.UI_LOOM_TAKE_RESULT, SoundSource.BLOCKS, 1.0F, 1.0F);
                   LoomMenu.this.lastSoundTime = var3x;
                }
 
             });
-            return super.onTake(var1, var2);
+            super.onTake(var1, var2);
          }
       });
 
       int var4;
       for(var4 = 0; var4 < 3; ++var4) {
@@ -204,11 +208,11 @@
    }
 
    public void removed(Player var1) {
       super.removed(var1);
       this.access.execute((var2, var3) -> {
-         this.clearContainer(var1, var1.level, this.inputContainer);
+         this.clearContainer(var1, this.inputContainer);
       });
    }
 
    private void setupResultSlot() {
       if (this.selectedBannerPatternIndex.get() > 0) {
