--- net/minecraft/world/inventory/MenuType.java
+++ net/minecraft/world/inventory/MenuType.java
@@ -42,9 +42,9 @@
 
    public T create(int var1, Inventory var2) {
       return this.constructor.create(var1, var2);
    }
 
-   interface MenuSupplier<T extends AbstractContainerMenu> {
+   private interface MenuSupplier<T extends AbstractContainerMenu> {
       T create(int var1, Inventory var2);
    }
 }
