--- net/minecraft/world/inventory/MerchantMenu.java
+++ net/minecraft/world/inventory/MerchantMenu.java
@@ -144,23 +144,23 @@
 
    public void removed(Player var1) {
       super.removed(var1);
       this.trader.setTradingPlayer((Player)null);
       if (!this.trader.getLevel().isClientSide) {
-         if (var1.isAlive() && (!(var1 instanceof ServerPlayer) || !((ServerPlayer)var1).hasDisconnected())) {
-            var1.inventory.placeItemBackInInventory(var1.level, this.tradeContainer.removeItemNoUpdate(0));
-            var1.inventory.placeItemBackInInventory(var1.level, this.tradeContainer.removeItemNoUpdate(1));
-         } else {
+         if (!var1.isAlive() || var1 instanceof ServerPlayer && ((ServerPlayer)var1).hasDisconnected()) {
             ItemStack var2 = this.tradeContainer.removeItemNoUpdate(0);
             if (!var2.isEmpty()) {
                var1.drop(var2, false);
             }
 
             var2 = this.tradeContainer.removeItemNoUpdate(1);
             if (!var2.isEmpty()) {
                var1.drop(var2, false);
             }
+         } else if (var1 instanceof ServerPlayer) {
+            var1.getInventory().placeItemBackInInventory(this.tradeContainer.removeItemNoUpdate(0));
+            var1.getInventory().placeItemBackInInventory(this.tradeContainer.removeItemNoUpdate(1));
          }
 
       }
    }
 
@@ -196,11 +196,11 @@
 
    private void moveFromInventoryToPaymentSlot(int var1, ItemStack var2) {
       if (!var2.isEmpty()) {
          for(int var3 = 3; var3 < 39; ++var3) {
             ItemStack var4 = ((Slot)this.slots.get(var3)).getItem();
-            if (!var4.isEmpty() && this.isSameItem(var2, var4)) {
+            if (!var4.isEmpty() && ItemStack.isSameItemSameTags(var2, var4)) {
                ItemStack var5 = this.tradeContainer.getItem(var1);
                int var6 = var5.isEmpty() ? 0 : var5.getCount();
                int var7 = Math.min(var2.getMaxStackSize() - var6, var4.getCount());
                ItemStack var8 = var4.copy();
                int var9 = var6 + var7;
@@ -214,14 +214,10 @@
          }
       }
 
    }
 
-   private boolean isSameItem(ItemStack var1, ItemStack var2) {
-      return var1.getItem() == var2.getItem() && ItemStack.tagMatches(var1, var2);
-   }
-
    public void setOffers(MerchantOffers var1) {
       this.trader.overrideOffers(var1);
    }
 
    public MerchantOffers getOffers() {
