--- net/minecraft/world/inventory/ResultSlot.java
+++ net/minecraft/world/inventory/ResultSlot.java
@@ -66,11 +66,11 @@
             if (var5.isEmpty()) {
                this.craftSlots.setItem(var4, var6);
             } else if (ItemStack.isSame(var5, var6) && ItemStack.tagMatches(var5, var6)) {
                var6.grow(var5.getCount());
                this.craftSlots.setItem(var4, var6);
-            } else if (!this.player.inventory.add(var6)) {
+            } else if (!this.player.getInventory().add(var6)) {
                this.player.drop(var6, false);
             }
          }
       }
 
