--- net/minecraft/world/inventory/ResultSlot.java
+++ net/minecraft/world/inventory/ResultSlot.java
@@ -48,11 +48,11 @@
       }
 
       this.removeCount = 0;
    }
 
-   public ItemStack onTake(Player var1, ItemStack var2) {
+   public void onTake(Player var1, ItemStack var2) {
       this.checkTakeAchievements(var2);
       NonNullList var3 = var1.level.getRecipeManager().getRemainingItemsFor(RecipeType.CRAFTING, this.craftSlots, var1.level);
 
       for(int var4 = 0; var4 < var3.size(); ++var4) {
          ItemStack var5 = this.craftSlots.getItem(var4);
@@ -66,14 +66,13 @@
             if (var5.isEmpty()) {
                this.craftSlots.setItem(var4, var6);
             } else if (ItemStack.isSame(var5, var6) && ItemStack.tagMatches(var5, var6)) {
                var6.grow(var5.getCount());
                this.craftSlots.setItem(var4, var6);
-            } else if (!this.player.inventory.add(var6)) {
+            } else if (!this.player.getInventory().add(var6)) {
                this.player.drop(var6, false);
             }
          }
       }
 
-      return var2;
    }
 }
