--- net/minecraft/world/inventory/SmithingMenu.java
+++ net/minecraft/world/inventory/SmithingMenu.java
@@ -33,19 +33,18 @@
 
    protected boolean mayPickup(Player var1, boolean var2) {
       return this.selectedRecipe != null && this.selectedRecipe.matches(this.inputSlots, this.level);
    }
 
-   protected ItemStack onTake(Player var1, ItemStack var2) {
+   protected void onTake(Player var1, ItemStack var2) {
       var2.onCraftedBy(var1.level, var1, var2.getCount());
       this.resultSlots.awardUsedRecipes(var1);
       this.shrinkStackInSlot(0);
       this.shrinkStackInSlot(1);
       this.access.execute((var0, var1x) -> {
          var0.levelEvent(1044, var1x, 0);
       });
-      return var2;
    }
 
    private void shrinkStackInSlot(int var1) {
       ItemStack var2 = this.inputSlots.getItem(var1);
       var2.shrink(1);
