--- net/minecraft/world/inventory/StonecutterMenu.java
+++ net/minecraft/world/inventory/StonecutterMenu.java
@@ -15,21 +15,27 @@
 import net.minecraft.world.item.crafting.StonecutterRecipe;
 import net.minecraft.world.level.Level;
 import net.minecraft.world.level.block.Blocks;
 
 public class StonecutterMenu extends AbstractContainerMenu {
+   public static final int INPUT_SLOT = 0;
+   public static final int RESULT_SLOT = 1;
+   private static final int INV_SLOT_START = 2;
+   private static final int INV_SLOT_END = 29;
+   private static final int USE_ROW_SLOT_START = 29;
+   private static final int USE_ROW_SLOT_END = 38;
    private final ContainerLevelAccess access;
    private final DataSlot selectedRecipeIndex;
    private final Level level;
    private List<StonecutterRecipe> recipes;
    private ItemStack input;
-   private long lastSoundTime;
+   long lastSoundTime;
    final Slot inputSlot;
    final Slot resultSlot;
-   private Runnable slotUpdateListener;
+   Runnable slotUpdateListener;
    public final Container container;
-   private final ResultContainer resultContainer;
+   final ResultContainer resultContainer;
 
    public StonecutterMenu(int var1, Inventory var2) {
       this(var1, var2, ContainerLevelAccess.NULL);
    }
 
@@ -54,11 +60,11 @@
       this.resultSlot = this.addSlot(new Slot(this.resultContainer, 1, 143, 33) {
          public boolean mayPlace(ItemStack var1) {
             return false;
          }
 
-         public ItemStack onTake(Player var1, ItemStack var2) {
+         public void onTake(Player var1, ItemStack var2) {
             var2.onCraftedBy(var1.level, var1, var2.getCount());
             StonecutterMenu.this.resultContainer.awardUsedRecipes(var1);
             ItemStack var3x = StonecutterMenu.this.inputSlot.remove(1);
             if (!var3x.isEmpty()) {
                StonecutterMenu.this.setupResultSlot();
@@ -70,11 +76,11 @@
                   var1x.playSound((Player)null, (BlockPos)var2x, SoundEvents.UI_STONECUTTER_TAKE_RESULT, SoundSource.BLOCKS, 1.0F, 1.0F);
                   StonecutterMenu.this.lastSoundTime = var3x;
                }
 
             });
-            return super.onTake(var1, var2);
+            super.onTake(var1, var2);
          }
       });
 
       int var4;
       for(var4 = 0; var4 < 3; ++var4) {
@@ -123,11 +129,11 @@
       return var1 >= 0 && var1 < this.recipes.size();
    }
 
    public void slotsChanged(Container var1) {
       ItemStack var2 = this.inputSlot.getItem();
-      if (var2.getItem() != this.input.getItem()) {
+      if (!var2.is(this.input.getItem())) {
          this.input = var2.copy();
          this.setupRecipeList(var1, var2);
       }
 
    }
@@ -140,11 +146,11 @@
          this.recipes = this.level.getRecipeManager().getRecipesFor(RecipeType.STONECUTTING, var1, this.level);
       }
 
    }
 
-   private void setupResultSlot() {
+   void setupResultSlot() {
       if (!this.recipes.isEmpty() && this.isValidRecipeIndex(this.selectedRecipeIndex.get())) {
          StonecutterRecipe var1 = (StonecutterRecipe)this.recipes.get(this.selectedRecipeIndex.get());
          this.resultContainer.setRecipeUsed(var1);
          this.resultSlot.set(var1.assemble(this.container));
       } else {
@@ -214,9 +220,9 @@
 
    public void removed(Player var1) {
       super.removed(var1);
       this.resultContainer.removeItemNoUpdate(1);
       this.access.execute((var2, var3) -> {
-         this.clearContainer(var1, var1.level, this.container);
+         this.clearContainer(var1, this.container);
       });
    }
 }
