--- net/minecraft/world/inventory/StonecutterMenu.java
+++ net/minecraft/world/inventory/StonecutterMenu.java
@@ -54,11 +54,11 @@
       this.resultSlot = this.addSlot(new Slot(this.resultContainer, 1, 143, 33) {
          public boolean mayPlace(ItemStack var1) {
             return false;
          }
 
-         public ItemStack onTake(Player var1, ItemStack var2) {
+         public void onTake(Player var1, ItemStack var2) {
             var2.onCraftedBy(var1.level, var1, var2.getCount());
             StonecutterMenu.this.resultContainer.awardUsedRecipes(var1);
             ItemStack var3x = StonecutterMenu.this.inputSlot.remove(1);
             if (!var3x.isEmpty()) {
                StonecutterMenu.this.setupResultSlot();
@@ -70,11 +70,11 @@
                   var1x.playSound((Player)null, (BlockPos)var2x, SoundEvents.UI_STONECUTTER_TAKE_RESULT, SoundSource.BLOCKS, 1.0F, 1.0F);
                   StonecutterMenu.this.lastSoundTime = var3x;
                }
 
             });
-            return super.onTake(var1, var2);
+            super.onTake(var1, var2);
          }
       });
 
       int var4;
       for(var4 = 0; var4 < 3; ++var4) {
@@ -123,11 +123,11 @@
       return var1 >= 0 && var1 < this.recipes.size();
    }
 
    public void slotsChanged(Container var1) {
       ItemStack var2 = this.inputSlot.getItem();
-      if (var2.getItem() != this.input.getItem()) {
+      if (!var2.is(this.input.getItem())) {
          this.input = var2.copy();
          this.setupRecipeList(var1, var2);
       }
 
    }
@@ -214,9 +214,9 @@
 
    public void removed(Player var1) {
       super.removed(var1);
       this.resultContainer.removeItemNoUpdate(1);
       this.access.execute((var2, var3) -> {
-         this.clearContainer(var1, var1.level, this.container);
+         this.clearContainer(var1, this.container);
       });
    }
 }
