--- net/minecraft/data/recipes/ShapedRecipeBuilder.java
+++ net/minecraft/data/recipes/ShapedRecipeBuilder.java
@@ -25,11 +25,11 @@
 import net.minecraft.world.item.crafting.RecipeSerializer;
 import net.minecraft.world.level.ItemLike;
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
 
-public class ShapedRecipeBuilder {
+public class ShapedRecipeBuilder implements RecipeBuilder {
    private static final Logger LOGGER = LogManager.getLogger();
    private final Item result;
    private final int count;
    private final List<String> rows = Lists.newArrayList();
    private final Map<Character, Ingredient> key = Maps.newLinkedHashMap();
