--- net/minecraft/data/recipes/SingleItemRecipeBuilder.java
+++ net/minecraft/data/recipes/SingleItemRecipeBuilder.java
@@ -41,10 +41,15 @@
    public SingleItemRecipeBuilder unlocks(String var1, CriterionTriggerInstance var2) {
       this.advancement.addCriterion(var1, var2);
       return this;
    }
 
+   public SingleItemRecipeBuilder group(String var1) {
+      this.group = var1;
+      return this;
+   }
+
    public void save(Consumer<FinishedRecipe> var1, String var2) {
       ResourceLocation var3 = Registry.ITEM.getKey(this.result);
       if ((new ResourceLocation(var2)).equals(var3)) {
          throw new IllegalStateException("Single Item Recipe " + var2 + " should remove its 'save' argument");
       } else {
