--- net/minecraft/data/models/ItemModelGenerators.java
+++ net/minecraft/data/models/ItemModelGenerators.java
@@ -30,10 +30,11 @@
       var3.create(ModelLocationUtils.getModelLocation(var1), TextureMapping.layer0(var2), this.output);
    }
 
    public void run() {
       this.generateFlatItem(Items.ACACIA_BOAT, ModelTemplates.FLAT_ITEM);
+      this.generateFlatItem(Items.AMETHYST_SHARD, ModelTemplates.FLAT_ITEM);
       this.generateFlatItem(Items.APPLE, ModelTemplates.FLAT_ITEM);
       this.generateFlatItem(Items.ARMOR_STAND, ModelTemplates.FLAT_ITEM);
       this.generateFlatItem(Items.ARROW, ModelTemplates.FLAT_ITEM);
       this.generateFlatItem(Items.BAKED_POTATO, ModelTemplates.FLAT_ITEM);
       this.generateFlatItem(Items.BAMBOO, ModelTemplates.FLAT_HANDHELD_ITEM);
@@ -85,10 +86,11 @@
       this.generateFlatItem(Items.COOKED_MUTTON, ModelTemplates.FLAT_ITEM);
       this.generateFlatItem(Items.COOKED_PORKCHOP, ModelTemplates.FLAT_ITEM);
       this.generateFlatItem(Items.COOKED_RABBIT, ModelTemplates.FLAT_ITEM);
       this.generateFlatItem(Items.COOKED_SALMON, ModelTemplates.FLAT_ITEM);
       this.generateFlatItem(Items.COOKIE, ModelTemplates.FLAT_ITEM);
+      this.generateFlatItem(Items.COPPER_INGOT, ModelTemplates.FLAT_ITEM);
       this.generateFlatItem(Items.CREEPER_BANNER_PATTERN, ModelTemplates.FLAT_ITEM);
       this.generateFlatItem(Items.CYAN_DYE, ModelTemplates.FLAT_ITEM);
       this.generateFlatItem(Items.DARK_OAK_BOAT, ModelTemplates.FLAT_ITEM);
       this.generateFlatItem(Items.DIAMOND, ModelTemplates.FLAT_ITEM);
       this.generateFlatItem(Items.DIAMOND_AXE, ModelTemplates.FLAT_HANDHELD_ITEM);
@@ -119,11 +121,14 @@
       this.generateFlatItem(Items.FURNACE_MINECART, ModelTemplates.FLAT_ITEM);
       this.generateFlatItem(Items.GHAST_TEAR, ModelTemplates.FLAT_ITEM);
       this.generateFlatItem(Items.GLASS_BOTTLE, ModelTemplates.FLAT_ITEM);
       this.generateFlatItem(Items.GLISTERING_MELON_SLICE, ModelTemplates.FLAT_ITEM);
       this.generateFlatItem(Items.GLOBE_BANNER_PATTER, ModelTemplates.FLAT_ITEM);
+      this.generateFlatItem(Items.GLOW_BERRIES, ModelTemplates.FLAT_ITEM);
       this.generateFlatItem(Items.GLOWSTONE_DUST, ModelTemplates.FLAT_ITEM);
+      this.generateFlatItem(Items.GLOW_INK_SAC, ModelTemplates.FLAT_ITEM);
+      this.generateFlatItem(Items.GLOW_ITEM_FRAME, ModelTemplates.FLAT_ITEM);
       this.generateFlatItem(Items.GOLDEN_APPLE, ModelTemplates.FLAT_ITEM);
       this.generateFlatItem(Items.GOLDEN_AXE, ModelTemplates.FLAT_HANDHELD_ITEM);
       this.generateFlatItem(Items.GOLDEN_BOOTS, ModelTemplates.FLAT_ITEM);
       this.generateFlatItem(Items.GOLDEN_CARROT, ModelTemplates.FLAT_ITEM);
       this.generateFlatItem(Items.GOLDEN_CHESTPLATE, ModelTemplates.FLAT_ITEM);
@@ -211,10 +216,11 @@
       this.generateFlatItem(Items.PIGLIN_BANNER_PATTERN, ModelTemplates.FLAT_ITEM);
       this.generateFlatItem(Items.PINK_DYE, ModelTemplates.FLAT_ITEM);
       this.generateFlatItem(Items.POISONOUS_POTATO, ModelTemplates.FLAT_ITEM);
       this.generateFlatItem(Items.POPPED_CHORUS_FRUIT, ModelTemplates.FLAT_ITEM);
       this.generateFlatItem(Items.PORKCHOP, ModelTemplates.FLAT_ITEM);
+      this.generateFlatItem(Items.POWDER_SNOW_BUCKET, ModelTemplates.FLAT_ITEM);
       this.generateFlatItem(Items.PRISMARINE_CRYSTALS, ModelTemplates.FLAT_ITEM);
       this.generateFlatItem(Items.PRISMARINE_SHARD, ModelTemplates.FLAT_ITEM);
       this.generateFlatItem(Items.PUFFERFISH, ModelTemplates.FLAT_ITEM);
       this.generateFlatItem(Items.PUFFERFISH_BUCKET, ModelTemplates.FLAT_ITEM);
       this.generateFlatItem(Items.PUMPKIN_PIE, ModelTemplates.FLAT_ITEM);
@@ -236,10 +242,11 @@
       this.generateFlatItem(Items.SLIME_BALL, ModelTemplates.FLAT_ITEM);
       this.generateFlatItem(Items.SNOWBALL, ModelTemplates.FLAT_ITEM);
       this.generateFlatItem(Items.SPECTRAL_ARROW, ModelTemplates.FLAT_ITEM);
       this.generateFlatItem(Items.SPIDER_EYE, ModelTemplates.FLAT_ITEM);
       this.generateFlatItem(Items.SPRUCE_BOAT, ModelTemplates.FLAT_ITEM);
+      this.generateFlatItem(Items.SPYGLASS, ModelTemplates.FLAT_ITEM);
       this.generateFlatItem(Items.STICK, ModelTemplates.FLAT_HANDHELD_ITEM);
       this.generateFlatItem(Items.STONE_AXE, ModelTemplates.FLAT_HANDHELD_ITEM);
       this.generateFlatItem(Items.STONE_HOE, ModelTemplates.FLAT_HANDHELD_ITEM);
       this.generateFlatItem(Items.STONE_PICKAXE, ModelTemplates.FLAT_HANDHELD_ITEM);
       this.generateFlatItem(Items.STONE_SHOVEL, ModelTemplates.FLAT_HANDHELD_ITEM);
@@ -249,10 +256,11 @@
       this.generateFlatItem(Items.TNT_MINECART, ModelTemplates.FLAT_ITEM);
       this.generateFlatItem(Items.TOTEM_OF_UNDYING, ModelTemplates.FLAT_ITEM);
       this.generateFlatItem(Items.TRIDENT, ModelTemplates.FLAT_ITEM);
       this.generateFlatItem(Items.TROPICAL_FISH, ModelTemplates.FLAT_ITEM);
       this.generateFlatItem(Items.TROPICAL_FISH_BUCKET, ModelTemplates.FLAT_ITEM);
+      this.generateFlatItem(Items.AXOLOTL_BUCKET, ModelTemplates.FLAT_ITEM);
       this.generateFlatItem(Items.TURTLE_HELMET, ModelTemplates.FLAT_ITEM);
       this.generateFlatItem(Items.WATER_BUCKET, ModelTemplates.FLAT_ITEM);
       this.generateFlatItem(Items.WHEAT, ModelTemplates.FLAT_ITEM);
       this.generateFlatItem(Items.WHITE_DYE, ModelTemplates.FLAT_ITEM);
       this.generateFlatItem(Items.WOODEN_AXE, ModelTemplates.FLAT_HANDHELD_ITEM);
