--- net/minecraft/stats/RecipeBookSettings.java
+++ net/minecraft/stats/RecipeBookSettings.java
@@ -29,14 +29,22 @@
          }
 
       }));
    }
 
+   public boolean isOpen(RecipeBookType var1) {
+      return ((RecipeBookSettings.TypeSettings)this.states.get(var1)).open;
+   }
+
    public void setOpen(RecipeBookType var1, boolean var2) {
       ((RecipeBookSettings.TypeSettings)this.states.get(var1)).open = var2;
    }
 
+   public boolean isFiltering(RecipeBookType var1) {
+      return ((RecipeBookSettings.TypeSettings)this.states.get(var1)).filtering;
+   }
+
    public void setFiltering(RecipeBookType var1, boolean var2) {
       ((RecipeBookSettings.TypeSettings)this.states.get(var1)).filtering = var2;
    }
 
    public static RecipeBookSettings read(FriendlyByteBuf var0) {
