--- net/minecraft/stats/ServerRecipeBook.java
+++ net/minecraft/stats/ServerRecipeBook.java
@@ -20,10 +20,11 @@
 import net.minecraft.world.item.crafting.RecipeManager;
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
 
 public class ServerRecipeBook extends RecipeBook {
+   public static final String RECIPE_BOOK_TAG = "recipeBook";
    private static final Logger LOGGER = LogManager.getLogger();
 
    public ServerRecipeBook() {
    }
 
@@ -111,11 +112,11 @@
             ResourceLocation var6 = new ResourceLocation(var5);
             Optional var7 = var3.byKey(var6);
             if (!var7.isPresent()) {
                LOGGER.error((String)"Tried to load unrecognized recipe: {} removed now.", (Object)var6);
             } else {
-               var2.accept(var7.get());
+               var2.accept((Recipe)var7.get());
             }
          } catch (ResourceLocationException var8) {
             LOGGER.error((String)"Tried to load improperly formatted recipe: {} removed now.", (Object)var5);
          }
       }
