--- net/minecraft/stats/ServerRecipeBook.java
+++ net/minecraft/stats/ServerRecipeBook.java
@@ -20,10 +20,11 @@
 import net.minecraft.world.item.crafting.RecipeManager;
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
 
 public class ServerRecipeBook extends RecipeBook {
+   public static final String RECIPE_BOOK_TAG = "recipeBook";
    private static final Logger LOGGER = LogManager.getLogger();
 
    public ServerRecipeBook() {
    }
 
