--- net/minecraft/stats/StatsCounter.java
+++ net/minecraft/stats/StatsCounter.java
@@ -19,9 +19,13 @@
 
    public void setValue(Player var1, Stat<?> var2, int var3) {
       this.stats.put(var2, var3);
    }
 
+   public <T> int getValue(StatType<T> var1, T var2) {
+      return var1.contains(var2) ? this.getValue(var1.get(var2)) : 0;
+   }
+
    public int getValue(Stat<?> var1) {
       return this.stats.getInt(var1);
    }
 }
