--- net/minecraft/stats/Stats.java
+++ net/minecraft/stats/Stats.java
@@ -89,10 +89,13 @@
    public static final ResourceLocation INTERACT_WITH_ANVIL;
    public static final ResourceLocation INTERACT_WITH_GRINDSTONE;
    public static final ResourceLocation TARGET_HIT;
    public static final ResourceLocation INTERACT_WITH_SMITHING_TABLE;
 
+   public Stats() {
+   }
+
    private static ResourceLocation makeCustomStat(String var0, StatFormatter var1) {
       ResourceLocation var2 = new ResourceLocation(var0);
       Registry.register(Registry.CUSTOM_STAT, (String)var0, var2);
       CUSTOM.get(var2, var1);
       return var2;
