--- net/minecraft/stats/StatFormatter.java
+++ net/minecraft/stats/StatFormatter.java
@@ -13,10 +13,12 @@
    StatFormatter DEFAULT;
    StatFormatter DIVIDE_BY_TEN;
    StatFormatter DISTANCE;
    StatFormatter TIME;
 
+   String format(int var1);
+
    static {
       NumberFormat var10000 = NumberFormat.getIntegerInstance(Locale.US);
       DEFAULT = var10000::format;
       DIVIDE_BY_TEN = (var0) -> {
          return DECIMAL_FORMAT.format((double)var0 * 0.1D);
