--- net/minecraft/stats/Stats.java
+++ net/minecraft/stats/Stats.java
@@ -15,11 +15,12 @@
    public static final StatType<Item> ITEM_DROPPED;
    public static final StatType<EntityType<?>> ENTITY_KILLED;
    public static final StatType<EntityType<?>> ENTITY_KILLED_BY;
    public static final StatType<ResourceLocation> CUSTOM;
    public static final ResourceLocation LEAVE_GAME;
-   public static final ResourceLocation PLAY_ONE_MINUTE;
+   public static final ResourceLocation PLAY_TIME;
+   public static final ResourceLocation TOTAL_WORLD_TIME;
    public static final ResourceLocation TIME_SINCE_DEATH;
    public static final ResourceLocation TIME_SINCE_REST;
    public static final ResourceLocation CROUCH_TIME;
    public static final ResourceLocation WALK_ONE_CM;
    public static final ResourceLocation CROUCH_ONE_CM;
@@ -89,10 +90,13 @@
    public static final ResourceLocation INTERACT_WITH_ANVIL;
    public static final ResourceLocation INTERACT_WITH_GRINDSTONE;
    public static final ResourceLocation TARGET_HIT;
    public static final ResourceLocation INTERACT_WITH_SMITHING_TABLE;
 
+   public Stats() {
+   }
+
    private static ResourceLocation makeCustomStat(String var0, StatFormatter var1) {
       ResourceLocation var2 = new ResourceLocation(var0);
       Registry.register(Registry.CUSTOM_STAT, (String)var0, var2);
       CUSTOM.get(var2, var1);
       return var2;
@@ -111,11 +115,12 @@
       ITEM_DROPPED = makeRegistryStatType("dropped", Registry.ITEM);
       ENTITY_KILLED = makeRegistryStatType("killed", Registry.ENTITY_TYPE);
       ENTITY_KILLED_BY = makeRegistryStatType("killed_by", Registry.ENTITY_TYPE);
       CUSTOM = makeRegistryStatType("custom", Registry.CUSTOM_STAT);
       LEAVE_GAME = makeCustomStat("leave_game", StatFormatter.DEFAULT);
-      PLAY_ONE_MINUTE = makeCustomStat("play_one_minute", StatFormatter.TIME);
+      PLAY_TIME = makeCustomStat("play_time", StatFormatter.TIME);
+      TOTAL_WORLD_TIME = makeCustomStat("total_world_time", StatFormatter.TIME);
       TIME_SINCE_DEATH = makeCustomStat("time_since_death", StatFormatter.TIME);
       TIME_SINCE_REST = makeCustomStat("time_since_rest", StatFormatter.TIME);
       CROUCH_TIME = makeCustomStat("sneak_time", StatFormatter.TIME);
       WALK_ONE_CM = makeCustomStat("walk_one_cm", StatFormatter.DISTANCE);
       CROUCH_ONE_CM = makeCustomStat("crouch_one_cm", StatFormatter.DISTANCE);
