--- net/minecraft/stats/StatFormatter.java
+++ net/minecraft/stats/StatFormatter.java
@@ -2,10 +2,11 @@
 
 import java.text.DecimalFormat;
 import java.text.DecimalFormatSymbols;
 import java.text.NumberFormat;
 import java.util.Locale;
+import java.util.Objects;
 import net.minecraft.Util;
 
 public interface StatFormatter {
    DecimalFormat DECIMAL_FORMAT = (DecimalFormat)Util.make(new DecimalFormat("########0.00"), (var0) -> {
       var0.setDecimalFormatSymbols(DecimalFormatSymbols.getInstance(Locale.ROOT));
@@ -13,12 +14,15 @@
    StatFormatter DEFAULT;
    StatFormatter DIVIDE_BY_TEN;
    StatFormatter DISTANCE;
    StatFormatter TIME;
 
+   String format(int var1);
+
    static {
       NumberFormat var10000 = NumberFormat.getIntegerInstance(Locale.US);
+      Objects.requireNonNull(var10000);
       DEFAULT = var10000::format;
       DIVIDE_BY_TEN = (var0) -> {
          return DECIMAL_FORMAT.format((double)var0 * 0.1D);
       };
       DISTANCE = (var0) -> {
