--- net/minecraft/stats/StatType.java
+++ net/minecraft/stats/StatType.java
@@ -1,20 +1,29 @@
 package net.minecraft.stats;
 
 import java.util.IdentityHashMap;
 import java.util.Iterator;
 import java.util.Map;
+import javax.annotation.Nullable;
 import net.minecraft.core.Registry;
+import net.minecraft.network.chat.Component;
+import net.minecraft.network.chat.TranslatableComponent;
 
 public class StatType<T> implements Iterable<Stat<T>> {
    private final Registry<T> registry;
    private final Map<T, Stat<T>> map = new IdentityHashMap();
+   @Nullable
+   private Component displayName;
 
    public StatType(Registry<T> var1) {
       this.registry = var1;
    }
 
+   public boolean contains(T var1) {
+      return this.map.containsKey(var1);
+   }
+
    public Stat<T> get(T var1, StatFormatter var2) {
       return (Stat)this.map.computeIfAbsent(var1, (var2x) -> {
          return new Stat(this, var2x, var2);
       });
    }
@@ -28,6 +37,18 @@
    }
 
    public Stat<T> get(T var1) {
       return this.get(var1, StatFormatter.DEFAULT);
    }
+
+   public String getTranslationKey() {
+      return "stat_type." + Registry.STAT_TYPE.getKey(this).toString().replace(':', '.');
+   }
+
+   public Component getDisplayName() {
+      if (this.displayName == null) {
+         this.displayName = new TranslatableComponent(this.getTranslationKey());
+      }
+
+      return this.displayName;
+   }
 }
