--- net/minecraft/client/gui/screens/advancements/AdvancementWidget.java
+++ net/minecraft/client/gui/screens/advancements/AdvancementWidget.java
@@ -12,10 +12,11 @@
 import net.minecraft.advancements.DisplayInfo;
 import net.minecraft.client.Minecraft;
 import net.minecraft.client.StringSplitter;
 import net.minecraft.client.gui.Font;
 import net.minecraft.client.gui.GuiComponent;
+import net.minecraft.client.renderer.GameRenderer;
 import net.minecraft.locale.Language;
 import net.minecraft.network.chat.Component;
 import net.minecraft.network.chat.ComponentUtils;
 import net.minecraft.network.chat.FormattedText;
 import net.minecraft.network.chat.Style;
@@ -145,11 +146,12 @@
             var5 = AdvancementWidgetType.OBTAINED;
          } else {
             var5 = AdvancementWidgetType.UNOBTAINED;
          }
 
-         this.minecraft.getTextureManager().bind(WIDGETS_LOCATION);
+         RenderSystem.setShader(GameRenderer::getPositionTexShader);
+         RenderSystem.setShaderTexture(0, WIDGETS_LOCATION);
          this.blit(var1, var2 + this.x + 3, var3 + this.y, this.display.getFrame().getTexture(), 128 + var5.getIndex() * 26, 26, 26);
          this.minecraft.getItemRenderer().renderAndDecorateFakeItem(this.display.getIcon(), var2 + this.x + 8, var3 + this.y + 5);
       }
 
       Iterator var6 = this.children.iterator();
@@ -202,12 +204,13 @@
          var13 = AdvancementWidgetType.UNOBTAINED;
          var14 = AdvancementWidgetType.UNOBTAINED;
       }
 
       int var16 = this.width - var15;
-      this.minecraft.getTextureManager().bind(WIDGETS_LOCATION);
-      RenderSystem.color4f(1.0F, 1.0F, 1.0F, 1.0F);
+      RenderSystem.setShader(GameRenderer::getPositionTexShader);
+      RenderSystem.setShaderTexture(0, WIDGETS_LOCATION);
+      RenderSystem.setShaderColor(1.0F, 1.0F, 1.0F, 1.0F);
       RenderSystem.enableBlend();
       int var17 = var3 + this.y;
       int var18;
       if (var7) {
          var18 = var2 + this.x - this.width + 26 + 6;
