--- net/minecraft/client/gui/screens/advancements/AdvancementWidget.java
+++ net/minecraft/client/gui/screens/advancements/AdvancementWidget.java
@@ -12,10 +12,11 @@
 import net.minecraft.advancements.DisplayInfo;
 import net.minecraft.client.Minecraft;
 import net.minecraft.client.StringSplitter;
 import net.minecraft.client.gui.Font;
 import net.minecraft.client.gui.GuiComponent;
+import net.minecraft.client.renderer.GameRenderer;
 import net.minecraft.locale.Language;
 import net.minecraft.network.chat.Component;
 import net.minecraft.network.chat.ComponentUtils;
 import net.minecraft.network.chat.FormattedText;
 import net.minecraft.network.chat.Style;
@@ -23,10 +24,22 @@
 import net.minecraft.util.FormattedCharSequence;
 import net.minecraft.util.Mth;
 
 public class AdvancementWidget extends GuiComponent {
    private static final ResourceLocation WIDGETS_LOCATION = new ResourceLocation("textures/gui/advancements/widgets.png");
+   private static final int HEIGHT = 26;
+   private static final int BOX_X = 0;
+   private static final int BOX_WIDTH = 200;
+   private static final int FRAME_WIDTH = 26;
+   private static final int ICON_X = 8;
+   private static final int ICON_Y = 5;
+   private static final int ICON_WIDTH = 26;
+   private static final int TITLE_PADDING_LEFT = 3;
+   private static final int TITLE_PADDING_RIGHT = 5;
+   private static final int TITLE_X = 32;
+   private static final int TITLE_Y = 9;
+   private static final int TITLE_MAX_WIDTH = 163;
    private static final int[] TEST_SPLIT_OFFSETS = new int[]{0, 10, -10, 25, -25};
    private final AdvancementTab tab;
    private final Advancement advancement;
    private final DisplayInfo display;
    private final FormattedCharSequence title;
@@ -145,11 +158,12 @@
             var5 = AdvancementWidgetType.OBTAINED;
          } else {
             var5 = AdvancementWidgetType.UNOBTAINED;
          }
 
-         this.minecraft.getTextureManager().bind(WIDGETS_LOCATION);
+         RenderSystem.setShader(GameRenderer::getPositionTexShader);
+         RenderSystem.setShaderTexture(0, WIDGETS_LOCATION);
          this.blit(var1, var2 + this.x + 3, var3 + this.y, this.display.getFrame().getTexture(), 128 + var5.getIndex() * 26, 26, 26);
          this.minecraft.getItemRenderer().renderAndDecorateFakeItem(this.display.getIcon(), var2 + this.x + 8, var3 + this.y + 5);
       }
 
       Iterator var6 = this.children.iterator();
@@ -159,10 +173,14 @@
          var7.draw(var1, var2, var3);
       }
 
    }
 
+   public int getWidth() {
+      return this.width;
+   }
+
    public void setProgress(AdvancementProgress var1) {
       this.progress = var1;
    }
 
    public void addChild(AdvancementWidget var1) {
@@ -202,12 +220,13 @@
          var13 = AdvancementWidgetType.UNOBTAINED;
          var14 = AdvancementWidgetType.UNOBTAINED;
       }
 
       int var16 = this.width - var15;
-      this.minecraft.getTextureManager().bind(WIDGETS_LOCATION);
-      RenderSystem.color4f(1.0F, 1.0F, 1.0F, 1.0F);
+      RenderSystem.setShader(GameRenderer::getPositionTexShader);
+      RenderSystem.setShaderTexture(0, WIDGETS_LOCATION);
+      RenderSystem.setShaderColor(1.0F, 1.0F, 1.0F, 1.0F);
       RenderSystem.enableBlend();
       int var17 = var3 + this.y;
       int var18;
       if (var7) {
          var18 = var2 + this.x - this.width + 26 + 6;
