--- net/minecraft/util/CrudeIncrementalIntIdentityHashBiMap.java
+++ net/minecraft/util/CrudeIncrementalIntIdentityHashBiMap.java
@@ -7,11 +7,13 @@
 import java.util.Iterator;
 import javax.annotation.Nullable;
 import net.minecraft.core.IdMap;
 
 public class CrudeIncrementalIntIdentityHashBiMap<K> implements IdMap<K> {
+   public static final int NOT_FOUND = -1;
    private static final Object EMPTY_SLOT = null;
+   private static final float LOADFACTOR = 0.8F;
    private K[] keys;
    private int[] values;
    private K[] byId;
    private int nextId;
    private int size;
@@ -34,10 +36,18 @@
 
    private int getValue(int var1) {
       return var1 == -1 ? -1 : this.values[var1];
    }
 
+   public boolean contains(K var1) {
+      return this.getId(var1) != -1;
+   }
+
+   public boolean contains(int var1) {
+      return this.byId(var1) != null;
+   }
+
    public int add(K var1) {
       int var2 = this.nextId();
       this.addMapping(var1, var2);
       return var2;
    }
