--- net/minecraft/util/CsvOutput.java
+++ net/minecraft/util/CsvOutput.java
@@ -8,10 +8,12 @@
 import java.util.stream.Stream;
 import javax.annotation.Nullable;
 import org.apache.commons.lang3.StringEscapeUtils;
 
 public class CsvOutput {
+   private static final String LINE_SEPARATOR = "\r\n";
+   private static final String FIELD_SEPARATOR = ",";
    private final Writer output;
    private final int columnCount;
 
    private CsvOutput(Writer var1, List<String> var2) throws IOException {
       this.output = var1;
