--- net/minecraft/util/IntRange.java
+++ net/minecraft/util/IntRange.java
@@ -18,11 +18,11 @@
    public static IntRange of(int var0, int var1) {
       return new IntRange(var0, var1);
    }
 
    public int randomValue(Random var1) {
-      return this.minInclusive == this.maxInclusive ? this.minInclusive : var1.nextInt(this.maxInclusive - this.minInclusive + 1) + this.minInclusive;
+      return Mth.nextInt(var1, this.minInclusive, this.maxInclusive);
    }
 
    public int getMinInclusive() {
       return this.minInclusive;
    }
