--- net/minecraft/util/StringUtil.java
+++ net/minecraft/util/StringUtil.java
@@ -4,10 +4,11 @@
 import javax.annotation.Nullable;
 import org.apache.commons.lang3.StringUtils;
 
 public class StringUtil {
    private static final Pattern STRIP_COLOR_PATTERN = Pattern.compile("(?i)\\u00A7[0-9A-FK-OR]");
+   private static final Pattern LINE_PATTERN = Pattern.compile("\r\n|[\n\r\u2028\u2029\u0085]");
 
    public static boolean isNullOrEmpty(@Nullable String var0) {
       return StringUtils.isEmpty(var0);
    }
 }
