--- net/minecraft/client/main/GameConfig.java
+++ net/minecraft/client/main/GameConfig.java
@@ -22,18 +22,21 @@
       this.location = var3;
       this.game = var4;
       this.server = var5;
    }
 
-   public static class ServerData {
-      @Nullable
-      public final String hostname;
-      public final int port;
+   public static class UserData {
+      public final User user;
+      public final PropertyMap userProperties;
+      public final PropertyMap profileProperties;
+      public final Proxy proxy;
 
-      public ServerData(@Nullable String var1, int var2) {
-         this.hostname = var1;
-         this.port = var2;
+      public UserData(User var1, PropertyMap var2, PropertyMap var3, Proxy var4) {
+         this.user = var1;
+         this.userProperties = var2;
+         this.profileProperties = var3;
+         this.proxy = var4;
       }
    }
 
    public static class FolderData {
       public final File gameDirectory;
@@ -52,24 +55,10 @@
       public AssetIndex getAssetIndex() {
          return (AssetIndex)(this.assetIndex == null ? new DirectAssetIndex(this.assetDirectory) : new AssetIndex(this.assetDirectory, this.assetIndex));
       }
    }
 
-   public static class UserData {
-      public final User user;
-      public final PropertyMap userProperties;
-      public final PropertyMap profileProperties;
-      public final Proxy proxy;
-
-      public UserData(User var1, PropertyMap var2, PropertyMap var3, Proxy var4) {
-         this.user = var1;
-         this.userProperties = var2;
-         this.profileProperties = var3;
-         this.proxy = var4;
-      }
-   }
-
    public static class GameData {
       public final boolean demo;
       public final String launchVersion;
       public final String versionType;
       public final boolean disableMultiplayer;
@@ -81,6 +70,17 @@
          this.versionType = var3;
          this.disableMultiplayer = var4;
          this.disableChat = var5;
       }
    }
+
+   public static class ServerData {
+      @Nullable
+      public final String hostname;
+      public final int port;
+
+      public ServerData(@Nullable String var1, int var2) {
+         this.hostname = var1;
+         this.port = var2;
+      }
+   }
 }
