--- net/minecraft/client/main/Main.java
+++ net/minecraft/client/main/Main.java
@@ -5,10 +5,11 @@
 import com.mojang.authlib.properties.PropertyMap;
 import com.mojang.authlib.properties.PropertyMap.Serializer;
 import com.mojang.blaze3d.pipeline.RenderPipeline;
 import com.mojang.blaze3d.platform.DisplayData;
 import com.mojang.blaze3d.systems.RenderSystem;
+import com.mojang.blaze3d.vertex.BufferUploader;
 import java.io.File;
 import java.net.Authenticator;
 import java.net.InetSocketAddress;
 import java.net.PasswordAuthentication;
 import java.net.Proxy;
@@ -27,10 +28,11 @@
 import net.minecraft.client.Minecraft;
 import net.minecraft.client.Options;
 import net.minecraft.client.User;
 import net.minecraft.client.resources.language.LanguageManager;
 import net.minecraft.client.server.IntegratedServer;
+import net.minecraft.obfuscate.DontObfuscate;
 import net.minecraft.server.Bootstrap;
 import net.minecraft.util.GsonHelper;
 import net.minecraft.world.entity.player.Player;
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
@@ -39,10 +41,11 @@
    private static final Logger LOGGER = LogManager.getLogger();
 
    public Main() {
    }
 
+   @DontObfuscate
    public static void main(String[] var0) {
       OptionParser var1 = new OptionParser();
       var1.allowsUnrecognizedOptions();
       var1.accepts("demo");
       var1.accepts("disableMultiplayer");
@@ -187,10 +190,12 @@
          } catch (Throwable var68) {
             LOGGER.error("Unhandled game exception", var68);
          }
       }
 
+      BufferUploader.reset();
+
       try {
          var56.stop();
          if (var57 != null) {
             var57.join();
          }
