--- net/minecraft/client/main/Main.java
+++ net/minecraft/client/main/Main.java
@@ -5,10 +5,11 @@
 import com.mojang.authlib.properties.PropertyMap;
 import com.mojang.authlib.properties.PropertyMap.Serializer;
 import com.mojang.blaze3d.pipeline.RenderPipeline;
 import com.mojang.blaze3d.platform.DisplayData;
 import com.mojang.blaze3d.systems.RenderSystem;
+import com.mojang.blaze3d.vertex.BufferUploader;
 import java.io.File;
 import java.net.Authenticator;
 import java.net.InetSocketAddress;
 import java.net.PasswordAuthentication;
 import java.net.Proxy;
@@ -187,10 +188,12 @@
          } catch (Throwable var68) {
             LOGGER.error("Unhandled game exception", var68);
          }
       }
 
+      BufferUploader.reset();
+
       try {
          var56.stop();
          if (var57 != null) {
             var57.join();
          }
