--- net/minecraft/client/resources/model/BlockModelRotation.java
+++ net/minecraft/client/resources/model/BlockModelRotation.java
@@ -40,12 +40,12 @@
       return var0 * 360 + var1;
    }
 
    private BlockModelRotation(int var3, int var4) {
       this.index = getIndex(var3, var4);
-      Quaternion var5 = new Quaternion(new Vector3f(0.0F, 1.0F, 0.0F), (float)(-var4), true);
-      var5.mul(new Quaternion(new Vector3f(1.0F, 0.0F, 0.0F), (float)(-var3), true));
+      Quaternion var5 = Vector3f.YP.rotationDegrees((float)(-var4));
+      var5.mul(Vector3f.XP.rotationDegrees((float)(-var3)));
       OctahedralGroup var6 = OctahedralGroup.IDENTITY;
 
       int var7;
       for(var7 = 0; var7 < var4; var7 += 90) {
          var6 = var6.compose(OctahedralGroup.ROT_90_Y_NEG);
