--- net/minecraft/client/resources/model/Material.java
+++ net/minecraft/client/resources/model/Material.java
@@ -64,8 +64,8 @@
    public int hashCode() {
       return Objects.hash(new Object[]{this.atlasLocation, this.texture});
    }
 
    public String toString() {
-      return "Material{atlasLocation=" + this.atlasLocation + ", texture=" + this.texture + '}';
+      return "Material{atlasLocation=" + this.atlasLocation + ", texture=" + this.texture + "}";
    }
 }
