--- net/minecraft/client/resources/model/ModelResourceLocation.java
+++ net/minecraft/client/resources/model/ModelResourceLocation.java
@@ -1,28 +1,35 @@
 package net.minecraft.client.resources.model;
 
+import com.google.common.annotations.VisibleForTesting;
 import java.util.Locale;
 import net.minecraft.resources.ResourceLocation;
 
 public class ModelResourceLocation extends ResourceLocation {
+   @VisibleForTesting
+   static final char VARIANT_SEPARATOR = '#';
    private final String variant;
 
    protected ModelResourceLocation(String[] var1) {
       super(var1);
       this.variant = var1[2].toLowerCase(Locale.ROOT);
    }
 
+   public ModelResourceLocation(String var1, String var2, String var3) {
+      this(new String[]{var1, var2, var3});
+   }
+
    public ModelResourceLocation(String var1) {
       this(decompose(var1));
    }
 
    public ModelResourceLocation(ResourceLocation var1, String var2) {
       this(var1.toString(), var2);
    }
 
    public ModelResourceLocation(String var1, String var2) {
-      this(decompose(var1 + '#' + var2));
+      this(decompose(var1 + "#" + var2));
    }
 
    protected static String[] decompose(String var0) {
       String[] var1 = new String[]{null, var0, ""};
       int var2 = var0.indexOf(35);
@@ -56,8 +63,9 @@
    public int hashCode() {
       return 31 * super.hashCode() + this.variant.hashCode();
    }
 
    public String toString() {
-      return super.toString() + '#' + this.variant;
+      String var10000 = super.toString();
+      return var10000 + "#" + this.variant;
    }
 }
