--- net/minecraft/client/resources/model/ModelResourceLocation.java
+++ net/minecraft/client/resources/model/ModelResourceLocation.java
@@ -1,18 +1,25 @@
 package net.minecraft.client.resources.model;
 
+import com.google.common.annotations.VisibleForTesting;
 import java.util.Locale;
 import net.minecraft.resources.ResourceLocation;
 
 public class ModelResourceLocation extends ResourceLocation {
+   @VisibleForTesting
+   static final char VARIANT_SEPARATOR = '#';
    private final String variant;
 
    protected ModelResourceLocation(String[] var1) {
       super(var1);
       this.variant = var1[2].toLowerCase(Locale.ROOT);
    }
 
+   public ModelResourceLocation(String var1, String var2, String var3) {
+      this(new String[]{var1, var2, var3});
+   }
+
    public ModelResourceLocation(String var1) {
       this(decompose(var1));
    }
 
    public ModelResourceLocation(ResourceLocation var1, String var2) {
