--- net/minecraft/client/resources/model/SimpleBakedModel.java
+++ net/minecraft/client/resources/model/SimpleBakedModel.java
@@ -112,10 +112,14 @@
       public SimpleBakedModel.Builder particle(TextureAtlasSprite var1) {
          this.particleIcon = var1;
          return this;
       }
 
+      public SimpleBakedModel.Builder item() {
+         return this;
+      }
+
       public BakedModel build() {
          if (this.particleIcon == null) {
             throw new RuntimeException("Missing particle!");
          } else {
             return new SimpleBakedModel(this.unculledFaces, this.culledFaces, this.hasAmbientOcclusion, this.usesBlockLight, this.isGui3d, this.particleIcon, this.transforms, this.overrides);
