--- net/minecraft/client/resources/model/WeightedBakedModel.java
+++ net/minecraft/client/resources/model/WeightedBakedModel.java
@@ -1,8 +1,9 @@
 package net.minecraft.client.resources.model;
 
 import com.google.common.collect.Lists;
+import java.util.Collections;
 import java.util.List;
 import java.util.Random;
 import javax.annotation.Nullable;
 import net.minecraft.client.renderer.block.model.BakedQuad;
 import net.minecraft.client.renderer.block.model.ItemOverrides;
@@ -22,11 +23,13 @@
       this.totalWeight = WeighedRandom.getTotalWeight(var1);
       this.wrapped = ((WeightedBakedModel.WeightedModel)var1.get(0)).model;
    }
 
    public List<BakedQuad> getQuads(@Nullable BlockState var1, @Nullable Direction var2, Random var3) {
-      return ((WeightedBakedModel.WeightedModel)WeighedRandom.getWeightedItem(this.list, Math.abs((int)var3.nextLong()) % this.totalWeight)).model.getQuads(var1, var2, var3);
+      return (List)WeighedRandom.getWeightedItem(this.list, Math.abs((int)var3.nextLong()) % this.totalWeight).map((var3x) -> {
+         return var3x.model.getQuads(var1, var2, var3);
+      }).orElse(Collections.emptyList());
    }
 
    public boolean useAmbientOcclusion() {
       return this.wrapped.useAmbientOcclusion();
    }
