--- net/minecraft/world/entity/animal/horse/ZombieHorse.java
+++ net/minecraft/world/entity/animal/horse/ZombieHorse.java
@@ -5,11 +5,11 @@
 import net.minecraft.sounds.SoundEvent;
 import net.minecraft.sounds.SoundEvents;
 import net.minecraft.world.InteractionHand;
 import net.minecraft.world.InteractionResult;
 import net.minecraft.world.damagesource.DamageSource;
-import net.minecraft.world.entity.AgableMob;
+import net.minecraft.world.entity.AgeableMob;
 import net.minecraft.world.entity.EntityType;
 import net.minecraft.world.entity.MobType;
 import net.minecraft.world.entity.ai.attributes.AttributeSupplier;
 import net.minecraft.world.entity.ai.attributes.Attributes;
 import net.minecraft.world.entity.player.Player;
@@ -48,12 +48,12 @@
       super.getHurtSound(var1);
       return SoundEvents.ZOMBIE_HORSE_HURT;
    }
 
    @Nullable
-   public AgableMob getBreedOffspring(ServerLevel var1, AgableMob var2) {
-      return (AgableMob)EntityType.ZOMBIE_HORSE.create(var1);
+   public AgeableMob getBreedOffspring(ServerLevel var1, AgeableMob var2) {
+      return (AgeableMob)EntityType.ZOMBIE_HORSE.create(var1);
    }
 
    public InteractionResult mobInteract(Player var1, InteractionHand var2) {
       ItemStack var3 = var1.getItemInHand(var2);
       if (!this.isTamed()) {
@@ -65,11 +65,11 @@
          return InteractionResult.sidedSuccess(this.level.isClientSide);
       } else if (this.isVehicle()) {
          return super.mobInteract(var1, var2);
       } else {
          if (!var3.isEmpty()) {
-            if (var3.getItem() == Items.SADDLE && !this.isSaddled()) {
+            if (var3.is(Items.SADDLE) && !this.isSaddled()) {
                this.openInventory(var1);
                return InteractionResult.sidedSuccess(this.level.isClientSide);
             }
 
             InteractionResult var4 = var3.interactLivingEntity(var1, this, var2);
