--- net/minecraft/world/level/portal/PortalShape.java
+++ net/minecraft/world/level/portal/PortalShape.java
@@ -27,11 +27,11 @@
    private final Direction rightDir;
    private int numPortalBlocks;
    @Nullable
    private BlockPos bottomLeft;
    private int height;
-   private int width;
+   private final int width;
 
    public static Optional<PortalShape> findEmptyPortalShape(LevelAccessor var0, BlockPos var1, Direction.Axis var2) {
       return findPortalShape(var0, var1, (var0x) -> {
          return var0x.isValid() && var0x.numPortalBlocks == 0;
       }, var2);
@@ -65,11 +65,11 @@
 
    }
 
    @Nullable
    private BlockPos calculateBottomLeft(BlockPos var1) {
-      for(int var2 = Math.max(0, var1.getY() - 21); var1.getY() > var2 && isEmpty(this.level.getBlockState(var1.below())); var1 = var1.below()) {
+      for(int var2 = Math.max(this.level.getMinBuildHeight(), var1.getY() - 21); var1.getY() > var2 && isEmpty(this.level.getBlockState(var1.below())); var1 = var1.below()) {
       }
 
       Direction var3 = this.rightDir.getOpposite();
       int var4 = this.getDistanceUntilEdgeAboveFrame(var1, var3) - 1;
       return var4 < 0 ? null : var1.relative(var3, var4);
