--- net/minecraft/world/level/portal/PortalShape.java
+++ net/minecraft/world/level/portal/PortalShape.java
@@ -17,21 +17,25 @@
 import net.minecraft.world.level.block.state.BlockState;
 import net.minecraft.world.level.block.state.properties.BlockStateProperties;
 import net.minecraft.world.phys.Vec3;
 
 public class PortalShape {
+   private static final int MIN_WIDTH = 2;
+   public static final int MAX_WIDTH = 21;
+   private static final int MIN_HEIGHT = 3;
+   public static final int MAX_HEIGHT = 21;
    private static final BlockBehaviour.StatePredicate FRAME = (var0, var1, var2) -> {
       return var0.is(Blocks.OBSIDIAN);
    };
    private final LevelAccessor level;
    private final Direction.Axis axis;
    private final Direction rightDir;
    private int numPortalBlocks;
    @Nullable
    private BlockPos bottomLeft;
    private int height;
-   private int width;
+   private final int width;
 
    public static Optional<PortalShape> findEmptyPortalShape(LevelAccessor var0, BlockPos var1, Direction.Axis var2) {
       return findPortalShape(var0, var1, (var0x) -> {
          return var0x.isValid() && var0x.numPortalBlocks == 0;
       }, var2);
@@ -65,11 +69,11 @@
 
    }
 
    @Nullable
    private BlockPos calculateBottomLeft(BlockPos var1) {
-      for(int var2 = Math.max(0, var1.getY() - 21); var1.getY() > var2 && isEmpty(this.level.getBlockState(var1.below())); var1 = var1.below()) {
+      for(int var2 = Math.max(this.level.getMinBuildHeight(), var1.getY() - 21); var1.getY() > var2 && isEmpty(this.level.getBlockState(var1.below())); var1 = var1.below()) {
       }
 
       Direction var3 = this.rightDir.getOpposite();
       int var4 = this.getDistanceUntilEdgeAboveFrame(var1, var3) - 1;
       return var4 < 0 ? null : var1.relative(var3, var4);
