--- net/minecraft/commands/arguments/EntityArgument.java
+++ net/minecraft/commands/arguments/EntityArgument.java
@@ -29,12 +29,14 @@
    public static final SimpleCommandExceptionType ERROR_NOT_SINGLE_PLAYER = new SimpleCommandExceptionType(new TranslatableComponent("argument.player.toomany"));
    public static final SimpleCommandExceptionType ERROR_ONLY_PLAYERS_ALLOWED = new SimpleCommandExceptionType(new TranslatableComponent("argument.player.entities"));
    public static final SimpleCommandExceptionType NO_ENTITIES_FOUND = new SimpleCommandExceptionType(new TranslatableComponent("argument.entity.notfound.entity"));
    public static final SimpleCommandExceptionType NO_PLAYERS_FOUND = new SimpleCommandExceptionType(new TranslatableComponent("argument.entity.notfound.player"));
    public static final SimpleCommandExceptionType ERROR_SELECTORS_NOT_ALLOWED = new SimpleCommandExceptionType(new TranslatableComponent("argument.entity.selector.not_allowed"));
-   private final boolean single;
-   private final boolean playersOnly;
+   private static final byte FLAG_SINGLE = 1;
+   private static final byte FLAG_PLAYERS_ONLY = 2;
+   final boolean single;
+   final boolean playersOnly;
 
    protected EntityArgument(boolean var1, boolean var2) {
       this.single = var1;
       this.playersOnly = var2;
    }
