--- net/minecraft/commands/arguments/EntityAnchorArgument.java
+++ net/minecraft/commands/arguments/EntityAnchorArgument.java
@@ -64,11 +64,11 @@
       }),
       EYES("eyes", (var0, var1) -> {
          return new Vec3(var0.x, var0.y + (double)var1.getEyeHeight(), var0.z);
       });
 
-      private static final Map<String, EntityAnchorArgument.Anchor> BY_NAME = (Map)Util.make(Maps.newHashMap(), (var0) -> {
+      static final Map<String, EntityAnchorArgument.Anchor> BY_NAME = (Map)Util.make(Maps.newHashMap(), (var0) -> {
          EntityAnchorArgument.Anchor[] var1 = values();
          int var2 = var1.length;
 
          for(int var3 = 0; var3 < var2; ++var3) {
             EntityAnchorArgument.Anchor var4 = var1[var3];
