--- net/minecraft/commands/arguments/ObjectiveCriteriaArgument.java
+++ net/minecraft/commands/arguments/ObjectiveCriteriaArgument.java
@@ -51,11 +51,11 @@
          return ERROR_INVALID_VALUE.create(var3);
       });
    }
 
    public <S> CompletableFuture<Suggestions> listSuggestions(CommandContext<S> var1, SuggestionsBuilder var2) {
-      ArrayList var3 = Lists.newArrayList((Iterable)ObjectiveCriteria.CRITERIA_BY_NAME.keySet());
+      ArrayList var3 = Lists.newArrayList((Iterable)ObjectiveCriteria.getCustomCriteriaNames());
       Iterator var4 = Registry.STAT_TYPE.iterator();
 
       while(var4.hasNext()) {
          StatType var5 = (StatType)var4.next();
          Iterator var6 = var5.getRegistry().iterator();
