--- net/minecraft/commands/arguments/OperationArgument.java
+++ net/minecraft/commands/arguments/OperationArgument.java
@@ -26,11 +26,11 @@
 
    public static OperationArgument operation() {
       return new OperationArgument();
    }
 
-   public static OperationArgument.Operation getOperation(CommandContext<CommandSourceStack> var0, String var1) throws CommandSyntaxException {
+   public static OperationArgument.Operation getOperation(CommandContext<CommandSourceStack> var0, String var1) {
       return (OperationArgument.Operation)var0.getArgument(var1, OperationArgument.Operation.class);
    }
 
    public OperationArgument.Operation parse(StringReader var1) throws CommandSyntaxException {
       if (!var1.canRead()) {
