--- net/minecraft/commands/arguments/ScoreHolderArgument.java
+++ net/minecraft/commands/arguments/ScoreHolderArgument.java
@@ -41,10 +41,11 @@
          SharedSuggestionProvider.suggest((Iterable)((CommandSourceStack)var0.getSource()).getOnlinePlayerNames(), var1x);
       });
    };
    private static final Collection<String> EXAMPLES = Arrays.asList("Player", "0123", "*", "@e");
    private static final SimpleCommandExceptionType ERROR_NO_RESULTS = new SimpleCommandExceptionType(new TranslatableComponent("argument.scoreHolder.empty"));
+   private static final byte FLAG_MULTIPLE = 1;
    private final boolean multiple;
 
    public ScoreHolderArgument(boolean var1) {
       this.multiple = var1;
    }
