--- net/minecraft/world/BossEvent.java
+++ net/minecraft/world/BossEvent.java
@@ -5,11 +5,11 @@
 import net.minecraft.network.chat.Component;
 
 public abstract class BossEvent {
    private final UUID id;
    protected Component name;
-   protected float percent;
+   protected float progress;
    protected BossEvent.BossBarColor color;
    protected BossEvent.BossBarOverlay overlay;
    protected boolean darkenScreen;
    protected boolean playBossMusic;
    protected boolean createWorldFog;
@@ -17,11 +17,11 @@
    public BossEvent(UUID var1, Component var2, BossEvent.BossBarColor var3, BossEvent.BossBarOverlay var4) {
       this.id = var1;
       this.name = var2;
       this.color = var3;
       this.overlay = var4;
-      this.percent = 1.0F;
+      this.progress = 1.0F;
    }
 
    public UUID getId() {
       return this.id;
    }
@@ -32,16 +32,16 @@
 
    public void setName(Component var1) {
       this.name = var1;
    }
 
-   public float getPercent() {
-      return this.percent;
+   public float getProgress() {
+      return this.progress;
    }
 
-   public void setPercent(float var1) {
-      this.percent = var1;
+   public void setProgress(float var1) {
+      this.progress = var1;
    }
 
    public BossEvent.BossBarColor getColor() {
       return this.color;
    }
