--- net/minecraft/world/Containers.java
+++ net/minecraft/world/Containers.java
@@ -10,10 +10,13 @@
 import net.minecraft.world.level.Level;
 
 public class Containers {
    private static final Random RANDOM = new Random();
 
+   public Containers() {
+   }
+
    public static void dropContents(Level var0, BlockPos var1, Container var2) {
       dropContents(var0, (double)var1.getX(), (double)var1.getY(), (double)var1.getZ(), var2);
    }
 
    public static void dropContents(Level var0, Entity var1, Container var2) {
