--- net/minecraft/world/DifficultyInstance.java
+++ net/minecraft/world/DifficultyInstance.java
@@ -3,10 +3,13 @@
 import javax.annotation.concurrent.Immutable;
 import net.minecraft.util.Mth;
 
 @Immutable
 public class DifficultyInstance {
+   private static final float DIFFICULTY_TIME_GLOBAL_OFFSET = -72000.0F;
+   private static final float MAX_DIFFICULTY_TIME_GLOBAL = 1440000.0F;
+   private static final float MAX_DIFFICULTY_TIME_LOCAL = 3600000.0F;
    private final Difficulty base;
    private final float effectiveDifficulty;
 
    public DifficultyInstance(Difficulty var1, long var2, long var4, float var6) {
       this.base = var1;
@@ -19,10 +22,14 @@
 
    public float getEffectiveDifficulty() {
       return this.effectiveDifficulty;
    }
 
+   public boolean isHard() {
+      return this.effectiveDifficulty >= (float)Difficulty.HARD.ordinal();
+   }
+
    public boolean isHarderThan(float var1) {
       return this.effectiveDifficulty > var1;
    }
 
    public float getSpecialMultiplier() {
