--- net/minecraft/world/level/timers/FunctionTagCallback.java
+++ net/minecraft/world/level/timers/FunctionTagCallback.java
@@ -7,11 +7,11 @@
 import net.minecraft.server.MinecraftServer;
 import net.minecraft.server.ServerFunctionManager;
 import net.minecraft.tags.Tag;
 
 public class FunctionTagCallback implements TimerCallback<MinecraftServer> {
-   private final ResourceLocation tagId;
+   final ResourceLocation tagId;
 
    public FunctionTagCallback(ResourceLocation var1) {
       this.tagId = var1;
    }
 
