--- net/minecraft/world/level/timers/TimerQueue.java
+++ net/minecraft/world/level/timers/TimerQueue.java
@@ -17,10 +17,13 @@
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
 
 public class TimerQueue<T> {
    private static final Logger LOGGER = LogManager.getLogger();
+   private static final String CALLBACK_DATA_TAG = "Callback";
+   private static final String TIMER_NAME_TAG = "Name";
+   private static final String TIMER_TRIGGER_TIME_TAG = "TriggerTime";
    private final TimerCallbacks<T> callbacksRegistry;
    private final Queue<TimerQueue.Event<T>> queue;
    private UnsignedLong sequentialId;
    private final Table<String, Long, TimerQueue.Event<T>> events;
 
