--- net/minecraft/world/damagesource/DamageSource.java
+++ net/minecraft/world/damagesource/DamageSource.java
@@ -213,11 +213,11 @@
       return this;
    }
 
    public boolean isCreativePlayer() {
       Entity var1 = this.getEntity();
-      return var1 instanceof Player && ((Player)var1).abilities.instabuild;
+      return var1 instanceof Player && ((Player)var1).getAbilities().instabuild;
    }
 
    @Nullable
    public Vec3 getSourcePosition() {
       return null;
