--- net/minecraft/world/damagesource/EntityDamageSource.java
+++ net/minecraft/world/damagesource/EntityDamageSource.java
@@ -8,15 +8,14 @@
 import net.minecraft.world.entity.player.Player;
 import net.minecraft.world.item.ItemStack;
 import net.minecraft.world.phys.Vec3;
 
 public class EntityDamageSource extends DamageSource {
-   @Nullable
    protected final Entity entity;
    private boolean isThorns;
 
-   public EntityDamageSource(String var1, @Nullable Entity var2) {
+   public EntityDamageSource(String var1, Entity var2) {
       super(var1);
       this.entity = var2;
    }
 
    public EntityDamageSource setThorns() {
@@ -26,11 +25,10 @@
 
    public boolean isThorns() {
       return this.isThorns;
    }
 
-   @Nullable
    public Entity getEntity() {
       return this.entity;
    }
 
    public Component getLocalizedDeathMessage(LivingEntity var1) {
@@ -38,16 +36,16 @@
       String var3 = "death.attack." + this.msgId;
       return !var2.isEmpty() && var2.hasCustomHoverName() ? new TranslatableComponent(var3 + ".item", new Object[]{var1.getDisplayName(), this.entity.getDisplayName(), var2.getDisplayName()}) : new TranslatableComponent(var3, new Object[]{var1.getDisplayName(), this.entity.getDisplayName()});
    }
 
    public boolean scalesWithDifficulty() {
-      return this.entity != null && this.entity instanceof LivingEntity && !(this.entity instanceof Player);
+      return this.entity instanceof LivingEntity && !(this.entity instanceof Player);
    }
 
    @Nullable
    public Vec3 getSourcePosition() {
-      return this.entity != null ? this.entity.position() : null;
+      return this.entity.position();
    }
 
    public String toString() {
       return "EntityDamageSource (" + this.entity + ")";
    }
