--- net/minecraft/world/scores/Objective.java
+++ net/minecraft/world/scores/Objective.java
@@ -5,10 +5,11 @@
 import net.minecraft.network.chat.HoverEvent;
 import net.minecraft.network.chat.TextComponent;
 import net.minecraft.world.scores.criteria.ObjectiveCriteria;
 
 public class Objective {
+   public static final int MAX_NAME_LENGTH = 16;
    private final Scoreboard scoreboard;
    private final String name;
    private final ObjectiveCriteria criteria;
    private Component displayName;
    private Component formattedDisplayName;
