--- net/minecraft/world/scores/PlayerTeam.java
+++ net/minecraft/world/scores/PlayerTeam.java
@@ -11,10 +11,13 @@
 import net.minecraft.network.chat.MutableComponent;
 import net.minecraft.network.chat.Style;
 import net.minecraft.network.chat.TextComponent;
 
 public class PlayerTeam extends Team {
+   public static final int MAX_NAME_LENGTH = 16;
+   private static final int BIT_FRIENDLY_FIRE = 0;
+   private static final int BIT_SEE_INVISIBLES = 1;
    private final Scoreboard scoreboard;
    private final String name;
    private final Set<String> players = Sets.newHashSet();
    private Component displayName;
    private Component playerPrefix;
@@ -40,10 +43,14 @@
       this.name = var2;
       this.displayName = new TextComponent(var2);
       this.displayNameStyle = Style.EMPTY.withInsertion(var2).withHoverEvent(new HoverEvent(HoverEvent.Action.SHOW_TEXT, new TextComponent(var2)));
    }
 
+   public Scoreboard getScoreboard() {
+      return this.scoreboard;
+   }
+
    public String getName() {
       return this.name;
    }
 
    public Component getDisplayName() {
