--- net/minecraft/world/scores/Team.java
+++ net/minecraft/world/scores/Team.java
@@ -81,10 +81,14 @@
          return var0;
       }));
       public final String name;
       public final int id;
 
+      public static String[] getAllNames() {
+         return (String[])BY_NAME.keySet().toArray(new String[BY_NAME.size()]);
+      }
+
       @Nullable
       public static Team.Visibility byName(String var0) {
          return (Team.Visibility)BY_NAME.get(var0);
       }
 
