--- com/mojang/realmsclient/gui/RealmsWorldSlotButton.java
+++ com/mojang/realmsclient/gui/RealmsWorldSlotButton.java
@@ -140,37 +140,37 @@
       Minecraft var17 = Minecraft.getInstance();
       TextureManager var18 = var17.getTextureManager();
       if (var13) {
          RealmsTextureManager.bindWorldTemplate(String.valueOf(var9), var11);
       } else if (var12) {
-         var18.bind(EMPTY_SLOT_LOCATION);
+         RenderSystem.setShaderTexture(0, EMPTY_SLOT_LOCATION);
       } else if (var11 != null && var9 != -1L) {
          RealmsTextureManager.bindWorldTemplate(String.valueOf(var9), var11);
       } else if (var8 == 1) {
-         var18.bind(DEFAULT_WORLD_SLOT_1);
+         RenderSystem.setShaderTexture(0, DEFAULT_WORLD_SLOT_1);
       } else if (var8 == 2) {
-         var18.bind(DEFAULT_WORLD_SLOT_2);
+         RenderSystem.setShaderTexture(0, DEFAULT_WORLD_SLOT_2);
       } else if (var8 == 3) {
-         var18.bind(DEFAULT_WORLD_SLOT_3);
+         RenderSystem.setShaderTexture(0, DEFAULT_WORLD_SLOT_3);
       }
 
       if (var6) {
          float var19 = 0.85F + 0.15F * Mth.cos((float)this.animTick * 0.2F);
-         RenderSystem.color4f(var19, var19, var19, 1.0F);
+         RenderSystem.setShaderColor(var19, var19, var19, 1.0F);
       } else {
-         RenderSystem.color4f(0.56F, 0.56F, 0.56F, 1.0F);
+         RenderSystem.setShaderColor(0.56F, 0.56F, 0.56F, 1.0F);
       }
 
       blit(var1, var2 + 3, var3 + 3, 0.0F, 0.0F, 74, 74, 74, 74);
-      var18.bind(SLOT_FRAME_LOCATION);
+      RenderSystem.setShaderTexture(0, SLOT_FRAME_LOCATION);
       boolean var20 = var16 && var14 != RealmsWorldSlotButton.Action.NOTHING;
       if (var20) {
-         RenderSystem.color4f(1.0F, 1.0F, 1.0F, 1.0F);
+         RenderSystem.setShaderColor(1.0F, 1.0F, 1.0F, 1.0F);
       } else if (var6) {
-         RenderSystem.color4f(0.8F, 0.8F, 0.8F, 1.0F);
+         RenderSystem.setShaderColor(0.8F, 0.8F, 0.8F, 1.0F);
       } else {
-         RenderSystem.color4f(0.56F, 0.56F, 0.56F, 1.0F);
+         RenderSystem.setShaderColor(0.56F, 0.56F, 0.56F, 1.0F);
       }
 
       blit(var1, var2, var3, 0.0F, 0.0F, 80, 80, 80, 80);
       drawCenteredString(var1, var17.font, var7, var2 + 40, var3 + 66, 16777215);
    }
