--- net/minecraft/ChatFormatting.java
+++ net/minecraft/ChatFormatting.java
@@ -32,10 +32,11 @@
    STRIKETHROUGH("STRIKETHROUGH", 'm', true),
    UNDERLINE("UNDERLINE", 'n', true),
    ITALIC("ITALIC", 'o', true),
    RESET("RESET", 'r', -1, (Integer)null);
 
+   public static final char PREFIX_CODE = '\u00a7';
    private static final Map<String, ChatFormatting> FORMATTING_BY_NAME = (Map)Arrays.stream(values()).collect(Collectors.toMap((var0) -> {
       return cleanName(var0.name);
    }, (var0) -> {
       return var0;
    }));
@@ -67,10 +68,14 @@
       this.id = var6;
       this.color = var7;
       this.toString = "\u00a7" + var4;
    }
 
+   public char getChar() {
+      return this.code;
+   }
+
    public int getId() {
       return this.id;
    }
 
    public boolean isFormat() {
@@ -121,10 +126,26 @@
 
          return null;
       }
    }
 
+   @Nullable
+   public static ChatFormatting getByCode(char var0) {
+      char var1 = Character.toString(var0).toLowerCase(Locale.ROOT).charAt(0);
+      ChatFormatting[] var2 = values();
+      int var3 = var2.length;
+
+      for(int var4 = 0; var4 < var3; ++var4) {
+         ChatFormatting var5 = var2[var4];
+         if (var5.code == var1) {
+            return var5;
+         }
+      }
+
+      return null;
+   }
+
    public static Collection<String> getNames(boolean var0, boolean var1) {
       ArrayList var2 = Lists.newArrayList();
       ChatFormatting[] var3 = values();
       int var4 = var3.length;
 
