--- net/minecraft/CrashReport.java
+++ net/minecraft/CrashReport.java
@@ -74,10 +74,16 @@
 
    public Throwable getException() {
       return this.exception;
    }
 
+   public String getDetails() {
+      StringBuilder var1 = new StringBuilder();
+      this.getDetails(var1);
+      return var1.toString();
+   }
+
    public void getDetails(StringBuilder var1) {
       if ((this.uncategorizedStackTrace == null || this.uncategorizedStackTrace.length <= 0) && !this.details.isEmpty()) {
          this.uncategorizedStackTrace = (StackTraceElement[])ArrayUtils.subarray((Object[])((CrashReportCategory)this.details.get(0)).getStacktrace(), 0, 1);
       }
 
@@ -160,10 +166,14 @@
       var1.append("\n\n");
       this.getDetails(var1);
       return var1.toString();
    }
 
+   public File getSaveFile() {
+      return this.saveFile;
+   }
+
    public boolean saveToFile(File var1) {
       if (this.saveFile != null) {
          return false;
       } else {
          if (var1.getParentFile() != null) {
