--- net/minecraft/Util.java
+++ net/minecraft/Util.java
@@ -337,10 +337,14 @@
 
    public static int getRandom(int[] var0, Random var1) {
       return var0[var1.nextInt(var0.length)];
    }
 
+   public static <T> T getRandom(List<T> var0, Random var1) {
+      return var0.get(var1.nextInt(var0.size()));
+   }
+
    private static BooleanSupplier createRenamer(final Path var0, final Path var1) {
       return new BooleanSupplier() {
          public boolean getAsBoolean() {
             try {
                Files.move(var0, var1);
