--- net/minecraft/world/level/levelgen/feature/treedecorators/TreeDecorator.java
+++ net/minecraft/world/level/levelgen/feature/treedecorators/TreeDecorator.java
@@ -28,11 +28,11 @@
    }
 
    protected void setBlock(LevelWriter var1, BlockPos var2, BlockState var3, Set<BlockPos> var4, BoundingBox var5) {
       var1.setBlock(var2, var3, 19);
       var4.add(var2);
-      var5.expand(new BoundingBox(var2, var2));
+      var5.expand(new BoundingBox(var2));
    }
 
    static {
       CODEC = Registry.TREE_DECORATOR_TYPES.dispatch(TreeDecorator::type, TreeDecoratorType::codec);
    }
