--- net/minecraft/world/entity/ai/sensing/GolemSensor.java
+++ net/minecraft/world/entity/ai/sensing/GolemSensor.java
@@ -8,10 +8,13 @@
 import net.minecraft.world.entity.EntityType;
 import net.minecraft.world.entity.LivingEntity;
 import net.minecraft.world.entity.ai.memory.MemoryModuleType;
 
 public class GolemSensor extends Sensor<LivingEntity> {
+   private static final int GOLEM_SCAN_RATE = 200;
+   private static final int MEMORY_TIME_TO_LIVE = 600;
+
    public GolemSensor() {
       this(200);
    }
 
    public GolemSensor(int var1) {
