--- net/minecraft/world/entity/ai/sensing/HostilesSensor.java
+++ net/minecraft/world/entity/ai/sensing/HostilesSensor.java
@@ -1,0 +1,30 @@
+package net.minecraft.world.entity.ai.sensing;
+
+import com.google.common.collect.ImmutableSet;
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+import net.minecraft.server.level.ServerLevel;
+import net.minecraft.world.entity.LivingEntity;
+import net.minecraft.world.entity.ai.memory.MemoryModuleType;
+
+public abstract class HostilesSensor extends Sensor<LivingEntity> {
+   public HostilesSensor() {
+   }
+
+   protected abstract Optional<LivingEntity> getNearestHostile(LivingEntity var1);
+
+   protected abstract boolean isClose(LivingEntity var1, LivingEntity var2);
+
+   public Set<MemoryModuleType<?>> requires() {
+      return ImmutableSet.of(MemoryModuleType.NEAREST_HOSTILE);
+   }
+
+   protected void doTick(ServerLevel var1, LivingEntity var2) {
+      var2.getBrain().setMemory(MemoryModuleType.NEAREST_HOSTILE, this.getNearestHostile(var2));
+   }
+
+   protected Optional<List<LivingEntity>> getVisibleEntities(LivingEntity var1) {
+      return var1.getBrain().getMemory(MemoryModuleType.VISIBLE_LIVING_ENTITIES);
+   }
+}
