--- net/minecraft/world/entity/ai/sensing/SecondaryPoiSensor.java
+++ net/minecraft/world/entity/ai/sensing/SecondaryPoiSensor.java
@@ -11,10 +11,12 @@
 import net.minecraft.world.entity.ai.Brain;
 import net.minecraft.world.entity.ai.memory.MemoryModuleType;
 import net.minecraft.world.entity.npc.Villager;
 
 public class SecondaryPoiSensor extends Sensor<Villager> {
+   private static final int SCAN_RATE = 40;
+
    public SecondaryPoiSensor() {
       super(40);
    }
 
    protected void doTick(ServerLevel var1, Villager var2) {
