--- net/minecraft/world/entity/ai/sensing/VillagerBabiesSensor.java
+++ net/minecraft/world/entity/ai/sensing/VillagerBabiesSensor.java
@@ -29,8 +29,8 @@
    private boolean isVillagerBaby(LivingEntity var1) {
       return var1.getType() == EntityType.VILLAGER && var1.isBaby();
    }
 
    private List<LivingEntity> getVisibleEntities(LivingEntity var1) {
-      return (List)var1.getBrain().getMemory(MemoryModuleType.VISIBLE_LIVING_ENTITIES).orElse(Lists.newArrayList());
+      return (List)var1.getBrain().getMemory(MemoryModuleType.NEAREST_VISIBLE_LIVING_ENTITIES).orElse(Lists.newArrayList());
    }
 }
