--- net/minecraft/network/Varint21FrameDecoder.java
+++ net/minecraft/network/Varint21FrameDecoder.java
@@ -9,11 +9,11 @@
 
 public class Varint21FrameDecoder extends ByteToMessageDecoder {
    public Varint21FrameDecoder() {
    }
 
-   protected void decode(ChannelHandlerContext var1, ByteBuf var2, List<Object> var3) throws Exception {
+   protected void decode(ChannelHandlerContext var1, ByteBuf var2, List<Object> var3) {
       var2.markReaderIndex();
       byte[] var4 = new byte[3];
 
       for(int var5 = 0; var5 < var4.length; ++var5) {
          if (!var2.isReadable()) {
@@ -25,16 +25,16 @@
          if (var4[var5] >= 0) {
             FriendlyByteBuf var6 = new FriendlyByteBuf(Unpooled.wrappedBuffer(var4));
 
             try {
                int var7 = var6.readVarInt();
-               if (var2.readableBytes() < var7) {
-                  var2.resetReaderIndex();
+               if (var2.readableBytes() >= var7) {
+                  var3.add(var2.readBytes(var7));
                   return;
                }
 
-               var3.add(var2.readBytes(var7));
+               var2.resetReaderIndex();
             } finally {
                var6.release();
             }
 
             return;
