--- net/minecraft/data/DataProvider.java
+++ net/minecraft/data/DataProvider.java
@@ -22,31 +22,28 @@
       String var4 = var0.toJson(var2);
       String var5 = SHA1.hashUnencodedChars(var4).toString();
       if (!Objects.equals(var1.getHash(var3), var5) || !Files.exists(var3, new LinkOption[0])) {
          Files.createDirectories(var3.getParent());
          BufferedWriter var6 = Files.newBufferedWriter(var3);
-         Throwable var7 = null;
 
          try {
             var6.write(var4);
-         } catch (Throwable var16) {
-            var7 = var16;
-            throw var16;
-         } finally {
+         } catch (Throwable var10) {
             if (var6 != null) {
-               if (var7 != null) {
-                  try {
-                     var6.close();
-                  } catch (Throwable var15) {
-                     var7.addSuppressed(var15);
-                  }
-               } else {
+               try {
                   var6.close();
+               } catch (Throwable var9) {
+                  var10.addSuppressed(var9);
                }
             }
 
+            throw var10;
          }
+
+         if (var6 != null) {
+            var6.close();
+         }
       }
 
       var1.putNew(var3, var5);
    }
 }
