--- net/minecraft/data/Main.java
+++ net/minecraft/data/Main.java
@@ -10,10 +10,11 @@
 import joptsimple.ArgumentAcceptingOptionSpec;
 import joptsimple.OptionParser;
 import joptsimple.OptionSet;
 import joptsimple.OptionSpec;
 import joptsimple.OptionSpecBuilder;
+import net.minecraft.SharedConstants;
 import net.minecraft.data.advancements.AdvancementProvider;
 import net.minecraft.data.info.BlockListReport;
 import net.minecraft.data.info.CommandsReport;
 import net.minecraft.data.info.RegistryDumpReport;
 import net.minecraft.data.loot.LootTableProvider;
@@ -23,18 +24,22 @@
 import net.minecraft.data.structures.SnbtToNbt;
 import net.minecraft.data.structures.StructureUpdater;
 import net.minecraft.data.tags.BlockTagsProvider;
 import net.minecraft.data.tags.EntityTypeTagsProvider;
 import net.minecraft.data.tags.FluidTagsProvider;
+import net.minecraft.data.tags.GameEventTagsProvider;
 import net.minecraft.data.tags.ItemTagsProvider;
 import net.minecraft.data.worldgen.biome.BiomeReport;
+import net.minecraft.obfuscate.DontObfuscate;
 
 public class Main {
    public Main() {
    }
 
+   @DontObfuscate
    public static void main(String[] var0) throws IOException {
+      SharedConstants.tryDetectVersion();
       OptionParser var1 = new OptionParser();
       AbstractOptionSpec var2 = var1.accepts("help", "Show the help menu").forHelp();
       OptionSpecBuilder var3 = var1.accepts("server", "Include server generators");
       OptionSpecBuilder var4 = var1.accepts("client", "Include client generators");
       OptionSpecBuilder var5 = var1.accepts("dev", "Include development tools");
@@ -78,10 +83,11 @@
          var7.addProvider(new ItemTagsProvider(var7, var8));
          var7.addProvider(new EntityTypeTagsProvider(var7));
          var7.addProvider(new RecipeProvider(var7));
          var7.addProvider(new AdvancementProvider(var7));
          var7.addProvider(new LootTableProvider(var7));
+         var7.addProvider(new GameEventTagsProvider(var7));
       }
 
       if (var4) {
          var7.addProvider(new NbtToSnbt(var7));
       }
