--- net/minecraft/world/item/trading/MerchantOffer.java
+++ net/minecraft/world/item/trading/MerchantOffer.java
@@ -77,11 +77,11 @@
 
    public ItemStack getCostA() {
       int var1 = this.baseCostA.getCount();
       ItemStack var2 = this.baseCostA.copy();
       int var3 = Math.max(0, Mth.floor((float)(var1 * this.demand) * this.priceMultiplier));
-      var2.setCount(Mth.clamp(var1 + var3 + this.specialPriceDiff, 1, this.baseCostA.getItem().getMaxStackSize()));
+      var2.setCount(Mth.clamp((int)(var1 + var3 + this.specialPriceDiff), (int)1, (int)this.baseCostA.getItem().getMaxStackSize()));
       return var2;
    }
 
    public ItemStack getCostB() {
       return this.costB;
