--- net/minecraft/world/item/crafting/BannerDuplicateRecipe.java
+++ net/minecraft/world/item/crafting/BannerDuplicateRecipe.java
@@ -20,12 +20,16 @@
       ItemStack var4 = null;
       ItemStack var5 = null;
 
       for(int var6 = 0; var6 < var1.getContainerSize(); ++var6) {
          ItemStack var7 = var1.getItem(var6);
-         Item var8 = var7.getItem();
-         if (var8 instanceof BannerItem) {
+         if (!var7.isEmpty()) {
+            Item var8 = var7.getItem();
+            if (!(var8 instanceof BannerItem)) {
+               return false;
+            }
+
             BannerItem var9 = (BannerItem)var8;
             if (var3 == null) {
                var3 = var9.getColor();
             } else if (var3 != var9.getColor()) {
                return false;
