--- net/minecraft/world/item/crafting/RecipeManager.java
+++ net/minecraft/world/item/crafting/RecipeManager.java
@@ -68,10 +68,14 @@
          return ((Builder)var0.getValue()).build();
       }));
       LOGGER.info("Loaded {} recipes", var4.size());
    }
 
+   public boolean hadErrorsLoading() {
+      return this.hasErrors;
+   }
+
    public <C extends Container, T extends Recipe<C>> Optional<T> getRecipeFor(RecipeType<T> var1, C var2, Level var3) {
       return this.byType(var1).values().stream().flatMap((var3x) -> {
          return Util.toStream(var1.tryMatch(var3x, var3, var2));
       }).findFirst();
    }
