--- net/minecraft/world/item/crafting/Recipe.java
+++ net/minecraft/world/item/crafting/Recipe.java
@@ -49,6 +49,13 @@
    ResourceLocation getId();
 
    RecipeSerializer<?> getSerializer();
 
    RecipeType<?> getType();
+
+   default boolean isIncomplete() {
+      NonNullList var1 = this.getIngredients();
+      return var1.isEmpty() || var1.stream().anyMatch((var0) -> {
+         return var0.getItems().length == 0;
+      });
+   }
 }
