--- net/minecraft/world/item/crafting/ShapelessRecipe.java
+++ net/minecraft/world/item/crafting/ShapelessRecipe.java
@@ -80,11 +80,11 @@
          if (var4.isEmpty()) {
             throw new JsonParseException("No ingredients for shapeless recipe");
          } else if (var4.size() > 9) {
             throw new JsonParseException("Too many ingredients for shapeless recipe");
          } else {
-            ItemStack var5 = ShapedRecipe.itemFromJson(GsonHelper.getAsJsonObject(var2, "result"));
+            ItemStack var5 = ShapedRecipe.itemStackFromJson(GsonHelper.getAsJsonObject(var2, "result"));
             return new ShapelessRecipe(var1, var3, var5, var4);
          }
       }
 
       private static NonNullList<Ingredient> itemsFromJson(JsonArray var0) {
@@ -99,11 +99,11 @@
 
          return var1;
       }
 
       public ShapelessRecipe fromNetwork(ResourceLocation var1, FriendlyByteBuf var2) {
-         String var3 = var2.readUtf(32767);
+         String var3 = var2.readUtf();
          int var4 = var2.readVarInt();
          NonNullList var5 = NonNullList.withSize(var4, Ingredient.EMPTY);
 
          for(int var6 = 0; var6 < var5.size(); ++var6) {
             var5.set(var6, Ingredient.fromNetwork(var2));
