--- net/minecraft/world/item/crafting/SimpleCookingSerializer.java
+++ net/minecraft/world/item/crafting/SimpleCookingSerializer.java
@@ -31,11 +31,11 @@
       int var10 = GsonHelper.getAsInt(var2, "cookingtime", this.defaultCookingTime);
       return this.factory.create(var1, var3, var5, var8, var9, var10);
    }
 
    public T fromNetwork(ResourceLocation var1, FriendlyByteBuf var2) {
-      String var3 = var2.readUtf(32767);
+      String var3 = var2.readUtf();
       Ingredient var4 = Ingredient.fromNetwork(var2);
       ItemStack var5 = var2.readItem();
       float var6 = var2.readFloat();
       int var7 = var2.readVarInt();
       return this.factory.create(var1, var3, var4, var5, var6, var7);
