--- net/minecraft/world/item/crafting/SingleItemRecipe.java
+++ net/minecraft/world/item/crafting/SingleItemRecipe.java
@@ -82,11 +82,11 @@
          ItemStack var7 = new ItemStack((ItemLike)Registry.ITEM.get(new ResourceLocation(var5)), var6);
          return this.factory.create(var1, var3, var4, var7);
       }
 
       public T fromNetwork(ResourceLocation var1, FriendlyByteBuf var2) {
-         String var3 = var2.readUtf(32767);
+         String var3 = var2.readUtf();
          Ingredient var4 = Ingredient.fromNetwork(var2);
          ItemStack var5 = var2.readItem();
          return this.factory.create(var1, var3, var4, var5);
       }
 
