--- com/mojang/realmsclient/dto/BackupList.java
+++ com/mojang/realmsclient/dto/BackupList.java
@@ -28,11 +28,11 @@
             while(var4.hasNext()) {
                var2.backups.add(Backup.parse((JsonElement)var4.next()));
             }
          }
       } catch (Exception var5) {
-         LOGGER.error("Could not parse BackupList: " + var5.getMessage());
+         LOGGER.error("Could not parse BackupList: {}", var5.getMessage());
       }
 
       return var2;
    }
 }
