--- com/mojang/realmsclient/dto/Backup.java
+++ com/mojang/realmsclient/dto/Backup.java
@@ -43,11 +43,11 @@
                   var2.metadata.put(format((String)var6.getKey()), ((JsonElement)var6.getValue()).getAsString());
                }
             }
          }
       } catch (Exception var7) {
-         LOGGER.error("Could not parse Backup: " + var7.getMessage());
+         LOGGER.error("Could not parse Backup: {}", var7.getMessage());
       }
 
       return var2;
    }
 
