--- com/mojang/realmsclient/dto/GuardedSerializer.java
+++ com/mojang/realmsclient/dto/GuardedSerializer.java
@@ -1,8 +1,9 @@
 package com.mojang.realmsclient.dto;
 
 import com.google.gson.Gson;
+import javax.annotation.Nullable;
 
 public class GuardedSerializer {
    private final Gson gson = new Gson();
 
    public GuardedSerializer() {
@@ -10,9 +11,10 @@
 
    public String toJson(ReflectionBasedSerialization var1) {
       return this.gson.toJson(var1);
    }
 
+   @Nullable
    public <T extends ReflectionBasedSerialization> T fromJson(String var1, Class<T> var2) {
       return (ReflectionBasedSerialization)this.gson.fromJson(var1, var2);
    }
 }
