--- com/mojang/realmsclient/dto/PendingInvitesList.java
+++ com/mojang/realmsclient/dto/PendingInvitesList.java
@@ -28,11 +28,11 @@
             while(var4.hasNext()) {
                var1.pendingInvites.add(PendingInvite.parse(((JsonElement)var4.next()).getAsJsonObject()));
             }
          }
       } catch (Exception var5) {
-         LOGGER.error("Could not parse PendingInvitesList: " + var5.getMessage());
+         LOGGER.error("Could not parse PendingInvitesList: {}", var5.getMessage());
       }
 
       return var1;
    }
 }
