--- com/mojang/realmsclient/dto/PendingInvite.java
+++ com/mojang/realmsclient/dto/PendingInvite.java
@@ -25,11 +25,11 @@
          var1.worldName = JsonUtils.getStringOr("worldName", var0, "");
          var1.worldOwnerName = JsonUtils.getStringOr("worldOwnerName", var0, "");
          var1.worldOwnerUuid = JsonUtils.getStringOr("worldOwnerUuid", var0, "");
          var1.date = JsonUtils.getDateOr("date", var0);
       } catch (Exception var3) {
-         LOGGER.error("Could not parse PendingInvite: " + var3.getMessage());
+         LOGGER.error("Could not parse PendingInvite: {}", var3.getMessage());
       }
 
       return var1;
    }
 }
