--- com/mojang/realmsclient/dto/RealmsWorldOptions.java
+++ com/mojang/realmsclient/dto/RealmsWorldOptions.java
@@ -1,30 +1,43 @@
 package com.mojang.realmsclient.dto;
 
 import com.google.gson.JsonObject;
 import com.mojang.realmsclient.util.JsonUtils;
 import java.util.Objects;
+import javax.annotation.Nullable;
 import net.minecraft.client.resources.language.I18n;
 
 public class RealmsWorldOptions extends ValueObject {
-   public Boolean pvp;
-   public Boolean spawnAnimals;
-   public Boolean spawnMonsters;
-   public Boolean spawnNPCs;
-   public Integer spawnProtection;
-   public Boolean commandBlocks;
-   public Boolean forceGameMode;
-   public Integer difficulty;
-   public Integer gameMode;
-   public String slotName;
+   public final boolean pvp;
+   public final boolean spawnAnimals;
+   public final boolean spawnMonsters;
+   public final boolean spawnNPCs;
+   public final int spawnProtection;
+   public final boolean commandBlocks;
+   public final boolean forceGameMode;
+   public final int difficulty;
+   public final int gameMode;
+   @Nullable
+   private final String slotName;
    public long templateId;
+   @Nullable
    public String templateImage;
-   public boolean adventureMap;
    public boolean empty;
+   private static final boolean DEFAULT_FORCE_GAME_MODE = false;
+   private static final boolean DEFAULT_PVP = true;
+   private static final boolean DEFAULT_SPAWN_ANIMALS = true;
+   private static final boolean DEFAULT_SPAWN_MONSTERS = true;
+   private static final boolean DEFAULT_SPAWN_NPCS = true;
+   private static final int DEFAULT_SPAWN_PROTECTION = 0;
+   private static final boolean DEFAULT_COMMAND_BLOCKS = false;
+   private static final int DEFAULT_DIFFICULTY = 2;
+   private static final int DEFAULT_GAME_MODE = 0;
+   private static final String DEFAULT_SLOT_NAME = "";
+   private static final long DEFAULT_TEMPLATE_ID = -1L;
    private static final String DEFAULT_TEMPLATE_IMAGE = null;
 
-   public RealmsWorldOptions(Boolean var1, Boolean var2, Boolean var3, Boolean var4, Integer var5, Boolean var6, Integer var7, Integer var8, Boolean var9, String var10) {
+   public RealmsWorldOptions(boolean var1, boolean var2, boolean var3, boolean var4, int var5, boolean var6, int var7, int var8, boolean var9, @Nullable String var10) {
       this.pvp = var1;
       this.spawnAnimals = var2;
       this.spawnMonsters = var3;
       this.spawnNPCs = var4;
       this.spawnProtection = var5;
@@ -51,11 +64,10 @@
 
    public static RealmsWorldOptions parse(JsonObject var0) {
       RealmsWorldOptions var1 = new RealmsWorldOptions(JsonUtils.getBooleanOr("pvp", var0, true), JsonUtils.getBooleanOr("spawnAnimals", var0, true), JsonUtils.getBooleanOr("spawnMonsters", var0, true), JsonUtils.getBooleanOr("spawnNPCs", var0, true), JsonUtils.getIntOr("spawnProtection", var0, 0), JsonUtils.getBooleanOr("commandBlocks", var0, false), JsonUtils.getIntOr("difficulty", var0, 2), JsonUtils.getIntOr("gameMode", var0, 0), JsonUtils.getBooleanOr("forceGameMode", var0, false), JsonUtils.getStringOr("slotName", var0, ""));
       var1.templateId = JsonUtils.getLongOr("worldTemplateId", var0, -1L);
       var1.templateImage = JsonUtils.getStringOr("worldTemplateImage", var0, DEFAULT_TEMPLATE_IMAGE);
-      var1.adventureMap = JsonUtils.getBooleanOr("adventureMap", var0, false);
       return var1;
    }
 
    public String getSlotName(int var1) {
       if (this.slotName != null && !this.slotName.isEmpty()) {
