--- com/mojang/realmsclient/dto/RealmsServerList.java
+++ com/mojang/realmsclient/dto/RealmsServerList.java
@@ -31,11 +31,11 @@
             while(var5.hasNext()) {
                var1.servers.add(RealmsServer.parse(((JsonElement)var5.next()).getAsJsonObject()));
             }
          }
       } catch (Exception var6) {
-         LOGGER.error("Could not parse McoServerList: " + var6.getMessage());
+         LOGGER.error("Could not parse McoServerList: {}", var6.getMessage());
       }
 
       return var1;
    }
 }
