--- com/mojang/realmsclient/dto/RealmsServerPlayerList.java
+++ com/mojang/realmsclient/dto/RealmsServerPlayerList.java
@@ -36,11 +36,11 @@
             }
          } else {
             var1.players = Lists.newArrayList();
          }
       } catch (Exception var4) {
-         LOGGER.error("Could not parse RealmsServerPlayerList: " + var4.getMessage());
+         LOGGER.error("Could not parse RealmsServerPlayerList: {}", var4.getMessage());
       }
 
       return var1;
    }
 
