--- com/mojang/realmsclient/dto/RealmsNews.java
+++ com/mojang/realmsclient/dto/RealmsNews.java
@@ -19,11 +19,11 @@
       try {
          JsonParser var2 = new JsonParser();
          JsonObject var3 = var2.parse(var0).getAsJsonObject();
          var1.newsLink = JsonUtils.getStringOr("newsLink", var3, (String)null);
       } catch (Exception var4) {
-         LOGGER.error("Could not parse RealmsNews: " + var4.getMessage());
+         LOGGER.error("Could not parse RealmsNews: {}", var4.getMessage());
       }
 
       return var1;
    }
 }
