--- com/mojang/realmsclient/dto/Subscription.java
+++ com/mojang/realmsclient/dto/Subscription.java
@@ -24,11 +24,11 @@
          JsonObject var3 = var2.parse(var0).getAsJsonObject();
          var1.startDate = JsonUtils.getLongOr("startDate", var3, 0L);
          var1.daysLeft = JsonUtils.getIntOr("daysLeft", var3, 0);
          var1.type = typeFrom(JsonUtils.getStringOr("subscriptionType", var3, Subscription.SubscriptionType.NORMAL.name()));
       } catch (Exception var4) {
-         LOGGER.error("Could not parse Subscription: " + var4.getMessage());
+         LOGGER.error("Could not parse Subscription: {}", var4.getMessage());
       }
 
       return var1;
    }
 
