--- com/mojang/realmsclient/dto/UploadInfo.java
+++ com/mojang/realmsclient/dto/UploadInfo.java
@@ -40,11 +40,11 @@
                String var7 = JsonUtils.getStringOr("token", var2, (String)null);
                return new UploadInfo(var6, var7, var5);
             }
          }
       } catch (Exception var8) {
-         LOGGER.error("Could not parse UploadInfo: " + var8.getMessage());
+         LOGGER.error("Could not parse UploadInfo: {}", var8.getMessage());
       }
 
       return null;
    }
 
