--- com/mojang/realmsclient/dto/UploadInfo.java
+++ com/mojang/realmsclient/dto/UploadInfo.java
@@ -12,10 +12,12 @@
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
 
 public class UploadInfo extends ValueObject {
    private static final Logger LOGGER = LogManager.getLogger();
+   private static final String DEFAULT_SCHEMA = "http://";
+   private static final int DEFAULT_PORT = 8080;
    private static final Pattern URI_SCHEMA_PATTERN = Pattern.compile("^[a-zA-Z][-a-zA-Z0-9+.]+:");
    private final boolean worldClosed;
    @Nullable
    private final String token;
    private final URI uploadEndpoint;
@@ -40,11 +42,11 @@
                String var7 = JsonUtils.getStringOr("token", var2, (String)null);
                return new UploadInfo(var6, var7, var5);
             }
          }
       } catch (Exception var8) {
-         LOGGER.error("Could not parse UploadInfo: " + var8.getMessage());
+         LOGGER.error("Could not parse UploadInfo: {}", var8.getMessage());
       }
 
       return null;
    }
 
