--- com/mojang/realmsclient/dto/WorldTemplatePaginatedList.java
+++ com/mojang/realmsclient/dto/WorldTemplatePaginatedList.java
@@ -26,10 +26,14 @@
       this.page = 0;
       this.size = var1;
       this.total = -1;
    }
 
+   public boolean isLastPage() {
+      return this.page * this.size >= this.total && this.page > 0 && this.total > 0 && this.size > 0;
+   }
+
    public static WorldTemplatePaginatedList parse(String var0) {
       WorldTemplatePaginatedList var1 = new WorldTemplatePaginatedList();
       var1.templates = Lists.newArrayList();
 
       try {
@@ -45,11 +49,11 @@
 
          var1.page = JsonUtils.getIntOr("page", var3, 0);
          var1.size = JsonUtils.getIntOr("size", var3, 0);
          var1.total = JsonUtils.getIntOr("total", var3, 0);
       } catch (Exception var5) {
-         LOGGER.error("Could not parse WorldTemplatePaginatedList: " + var5.getMessage());
+         LOGGER.error("Could not parse WorldTemplatePaginatedList: {}", var5.getMessage());
       }
 
       return var1;
    }
 }
