--- com/mojang/realmsclient/dto/WorldTemplatePaginatedList.java
+++ com/mojang/realmsclient/dto/WorldTemplatePaginatedList.java
@@ -45,11 +45,11 @@
 
          var1.page = JsonUtils.getIntOr("page", var3, 0);
          var1.size = JsonUtils.getIntOr("size", var3, 0);
          var1.total = JsonUtils.getIntOr("total", var3, 0);
       } catch (Exception var5) {
-         LOGGER.error("Could not parse WorldTemplatePaginatedList: " + var5.getMessage());
+         LOGGER.error("Could not parse WorldTemplatePaginatedList: {}", var5.getMessage());
       }
 
       return var1;
    }
 }
