--- com/mojang/realmsclient/dto/WorldTemplate.java
+++ com/mojang/realmsclient/dto/WorldTemplate.java
@@ -35,11 +35,11 @@
          var1.image = JsonUtils.getStringOr("image", var0, (String)null);
          var1.trailer = JsonUtils.getStringOr("trailer", var0, "");
          var1.recommendedPlayers = JsonUtils.getStringOr("recommendedPlayers", var0, "");
          var1.type = WorldTemplate.WorldTemplateType.valueOf(JsonUtils.getStringOr("type", var0, WorldTemplate.WorldTemplateType.WORLD_TEMPLATE.name()));
       } catch (Exception var3) {
-         LOGGER.error("Could not parse WorldTemplate: " + var3.getMessage());
+         LOGGER.error("Could not parse WorldTemplate: {}", var3.getMessage());
       }
 
       return var1;
    }
 
