--- com/mojang/realmsclient/dto/WorldDownload.java
+++ com/mojang/realmsclient/dto/WorldDownload.java
@@ -23,11 +23,11 @@
       try {
          var3.downloadLink = JsonUtils.getStringOr("downloadLink", var2, "");
          var3.resourcePackUrl = JsonUtils.getStringOr("resourcePackUrl", var2, "");
          var3.resourcePackHash = JsonUtils.getStringOr("resourcePackHash", var2, "");
       } catch (Exception var5) {
-         LOGGER.error("Could not parse WorldDownload: " + var5.getMessage());
+         LOGGER.error("Could not parse WorldDownload: {}", var5.getMessage());
       }
 
       return var3;
    }
 }
