--- net/minecraft/world/entity/vehicle/MinecartTNT.java
+++ net/minecraft/world/entity/vehicle/MinecartTNT.java
@@ -18,10 +18,11 @@
 import net.minecraft.world.level.block.Blocks;
 import net.minecraft.world.level.block.state.BlockState;
 import net.minecraft.world.level.material.FluidState;
 
 public class MinecartTNT extends AbstractMinecart {
+   private static final byte EVENT_PRIME = 10;
    private int fuse = -1;
 
    public MinecartTNT(EntityType<? extends MinecartTNT> var1, Level var2) {
       super(var1, var2);
    }
@@ -91,31 +92,40 @@
          if (var3 > 5.0D) {
             var3 = 5.0D;
          }
 
          this.level.explode(this, this.getX(), this.getY(), this.getZ(), (float)(4.0D + this.random.nextDouble() * 1.5D * var3), Explosion.BlockInteraction.BREAK);
-         this.remove();
+         this.discard();
       }
 
    }
 
-   public boolean causeFallDamage(float var1, float var2) {
+   public boolean causeFallDamage(float var1, float var2, DamageSource var3) {
       if (var1 >= 3.0F) {
-         float var3 = var1 / 10.0F;
-         this.explode((double)(var3 * var3));
+         float var4 = var1 / 10.0F;
+         this.explode((double)(var4 * var4));
       }
 
-      return super.causeFallDamage(var1, var2);
+      return super.causeFallDamage(var1, var2, var3);
    }
 
    public void activateMinecart(int var1, int var2, int var3, boolean var4) {
       if (var4 && this.fuse < 0) {
          this.primeFuse();
       }
 
    }
 
+   public void handleEntityEvent(byte var1) {
+      if (var1 == 10) {
+         this.primeFuse();
+      } else {
+         super.handleEntityEvent(var1);
+      }
+
+   }
+
    public void primeFuse() {
       this.fuse = 80;
       if (!this.level.isClientSide) {
          this.level.broadcastEntityEvent(this, (byte)10);
          if (!this.isSilent()) {
@@ -123,10 +133,14 @@
          }
       }
 
    }
 
+   public int getFuse() {
+      return this.fuse;
+   }
+
    public boolean isPrimed() {
       return this.fuse > -1;
    }
 
    public float getBlockExplosionResistance(Explosion var1, BlockGetter var2, BlockPos var3, BlockState var4, FluidState var5, float var6) {
