--- net/minecraft/world/entity/vehicle/MinecartCommandBlock.java
+++ net/minecraft/world/entity/vehicle/MinecartCommandBlock.java
@@ -14,15 +14,17 @@
 import net.minecraft.world.entity.player.Player;
 import net.minecraft.world.level.BaseCommandBlock;
 import net.minecraft.world.level.Level;
 import net.minecraft.world.level.block.Blocks;
 import net.minecraft.world.level.block.state.BlockState;
+import net.minecraft.world.phys.Vec3;
 
 public class MinecartCommandBlock extends AbstractMinecart {
-   private static final EntityDataAccessor<String> DATA_ID_COMMAND_NAME;
-   private static final EntityDataAccessor<Component> DATA_ID_LAST_OUTPUT;
+   static final EntityDataAccessor<String> DATA_ID_COMMAND_NAME;
+   static final EntityDataAccessor<Component> DATA_ID_LAST_OUTPUT;
    private final BaseCommandBlock commandBlock = new MinecartCommandBlock.MinecartCommandBase();
+   private static final int ACTIVATION_DELAY = 4;
    private int lastActivated;
 
    public MinecartCommandBlock(EntityType<? extends MinecartCommandBlock> var1, Level var2) {
       super(var1, var2);
    }
@@ -106,10 +108,18 @@
       public void onUpdated() {
          MinecartCommandBlock.this.getEntityData().set(MinecartCommandBlock.DATA_ID_COMMAND_NAME, this.getCommand());
          MinecartCommandBlock.this.getEntityData().set(MinecartCommandBlock.DATA_ID_LAST_OUTPUT, this.getLastOutput());
       }
 
+      public Vec3 getPosition() {
+         return MinecartCommandBlock.this.position();
+      }
+
+      public MinecartCommandBlock getMinecart() {
+         return MinecartCommandBlock.this;
+      }
+
       public CommandSourceStack createCommandSourceStack() {
          return new CommandSourceStack(this, MinecartCommandBlock.this.position(), MinecartCommandBlock.this.getRotationVector(), this.getLevel(), 2, this.getName().getString(), MinecartCommandBlock.this.getDisplayName(), this.getLevel().getServer(), MinecartCommandBlock.this);
       }
    }
 }
