--- net/minecraft/world/level/levelgen/heightproviders/HeightProviderType.java
+++ net/minecraft/world/level/levelgen/heightproviders/HeightProviderType.java
@@ -1,0 +1,18 @@
+package net.minecraft.world.level.levelgen.heightproviders;
+
+import com.mojang.serialization.Codec;
+import net.minecraft.core.Registry;
+
+public interface HeightProviderType<P extends HeightProvider> {
+   HeightProviderType<ConstantHeight> CONSTANT = register("constant", ConstantHeight.CODEC);
+   HeightProviderType<UniformHeight> UNIFORM = register("uniform", UniformHeight.CODEC);
+   HeightProviderType<BiasedToBottomHeight> BIASED_TO_BOTTOM = register("biased_to_bottom", BiasedToBottomHeight.CODEC);
+
+   Codec<P> codec();
+
+   static <P extends HeightProvider> HeightProviderType<P> register(String var0, Codec<P> var1) {
+      return (HeightProviderType)Registry.register(Registry.HEIGHT_PROVIDER_TYPES, (String)var0, () -> {
+         return var1;
+      });
+   }
+}
