--- net/minecraft/client/gui/screens/BackupConfirmScreen.java
+++ net/minecraft/client/gui/screens/BackupConfirmScreen.java
@@ -14,10 +14,11 @@
    private final Screen lastScreen;
    protected final BackupConfirmScreen.Listener listener;
    private final Component description;
    private final boolean promptForCacheErase;
    private MultiLineLabel message;
+   protected int id;
    private Checkbox eraseCache;
 
    public BackupConfirmScreen(@Nullable Screen var1, BackupConfirmScreen.Listener var2, Component var3, Component var4, boolean var5) {
       super(var3);
       this.message = MultiLineLabel.EMPTY;
