--- net/minecraft/client/gui/screens/BackupConfirmScreen.java
+++ net/minecraft/client/gui/screens/BackupConfirmScreen.java
@@ -1,8 +1,9 @@
 package net.minecraft.client.gui.screens;
 
 import com.mojang.blaze3d.vertex.PoseStack;
+import java.util.Objects;
 import javax.annotation.Nullable;
 import net.minecraft.client.gui.components.Button;
 import net.minecraft.client.gui.components.Checkbox;
 import net.minecraft.client.gui.components.MultiLineLabel;
 import net.minecraft.network.chat.CommonComponents;
@@ -14,10 +15,11 @@
    private final Screen lastScreen;
    protected final BackupConfirmScreen.Listener listener;
    private final Component description;
    private final boolean promptForCacheErase;
    private MultiLineLabel message;
+   protected int id;
    private Checkbox eraseCache;
 
    public BackupConfirmScreen(@Nullable Screen var1, BackupConfirmScreen.Listener var2, Component var3, Component var4, boolean var5) {
       super(var3);
       this.message = MultiLineLabel.EMPTY;
@@ -29,24 +31,24 @@
 
    protected void init() {
       super.init();
       this.message = MultiLineLabel.create(this.font, this.description, this.width - 50);
       int var10000 = this.message.getLineCount() + 1;
-      this.font.getClass();
+      Objects.requireNonNull(this.font);
       int var1 = var10000 * 9;
-      this.addButton(new Button(this.width / 2 - 155, 100 + var1, 150, 20, new TranslatableComponent("selectWorld.backupJoinConfirmButton"), (var1x) -> {
+      this.addRenderableWidget(new Button(this.width / 2 - 155, 100 + var1, 150, 20, new TranslatableComponent("selectWorld.backupJoinConfirmButton"), (var1x) -> {
          this.listener.proceed(true, this.eraseCache.selected());
       }));
-      this.addButton(new Button(this.width / 2 - 155 + 160, 100 + var1, 150, 20, new TranslatableComponent("selectWorld.backupJoinSkipButton"), (var1x) -> {
+      this.addRenderableWidget(new Button(this.width / 2 - 155 + 160, 100 + var1, 150, 20, new TranslatableComponent("selectWorld.backupJoinSkipButton"), (var1x) -> {
          this.listener.proceed(false, this.eraseCache.selected());
       }));
-      this.addButton(new Button(this.width / 2 - 155 + 80, 124 + var1, 150, 20, CommonComponents.GUI_CANCEL, (var1x) -> {
+      this.addRenderableWidget(new Button(this.width / 2 - 155 + 80, 124 + var1, 150, 20, CommonComponents.GUI_CANCEL, (var1x) -> {
          this.minecraft.setScreen(this.lastScreen);
       }));
       this.eraseCache = new Checkbox(this.width / 2 - 155 + 80, 76 + var1, 150, 20, new TranslatableComponent("selectWorld.backupEraseCache"), false);
       if (this.promptForCacheErase) {
-         this.addButton(this.eraseCache);
+         this.addRenderableWidget(this.eraseCache);
       }
 
    }
 
    public void render(PoseStack var1, int var2, int var3, float var4) {
