--- net/minecraft/client/gui/screens/ChatScreen.java
+++ net/minecraft/client/gui/screens/ChatScreen.java
@@ -10,10 +10,11 @@
 import net.minecraft.network.chat.Style;
 import net.minecraft.network.chat.TranslatableComponent;
 import net.minecraft.util.Mth;
 
 public class ChatScreen extends Screen {
+   public static final int MOUSE_SCROLL_SPEED = 7;
    private String historyBuffer = "";
    private int historyPos = -1;
    protected EditBox input;
    private String initial = "";
    private CommandSuggestions commandSuggestions;
@@ -149,11 +150,11 @@
    }
 
    public void moveInHistory(int var1) {
       int var2 = this.historyPos + var1;
       int var3 = this.minecraft.gui.getChat().getRecentChat().size();
-      var2 = Mth.clamp(var2, 0, var3);
+      var2 = Mth.clamp((int)var2, (int)0, (int)var3);
       if (var2 != this.historyPos) {
          if (var2 == var3) {
             this.historyPos = var3;
             this.input.setValue(this.historyBuffer);
          } else {
