--- net/minecraft/client/gui/screens/CreateBuffetWorldScreen.java
+++ net/minecraft/client/gui/screens/CreateBuffetWorldScreen.java
@@ -9,11 +9,10 @@
 import net.minecraft.client.gui.chat.NarratorChatListener;
 import net.minecraft.client.gui.components.Button;
 import net.minecraft.client.gui.components.ObjectSelectionList;
 import net.minecraft.core.Registry;
 import net.minecraft.core.RegistryAccess;
-import net.minecraft.core.WritableRegistry;
 import net.minecraft.locale.Language;
 import net.minecraft.network.chat.CommonComponents;
 import net.minecraft.network.chat.Component;
 import net.minecraft.network.chat.TextComponent;
 import net.minecraft.network.chat.TranslatableComponent;
@@ -23,11 +22,11 @@
 
 public class CreateBuffetWorldScreen extends Screen {
    private static final Component BIOME_SELECT_INFO = new TranslatableComponent("createWorld.customize.buffet.biome");
    private final Screen parent;
    private final Consumer<Biome> applySettings;
-   private final WritableRegistry<Biome> biomes;
+   private final Registry<Biome> biomes;
    private CreateBuffetWorldScreen.BiomeList list;
    private Biome biome;
    private Button doneButton;
 
    public CreateBuffetWorldScreen(Screen var1, RegistryAccess var2, Consumer<Biome> var3, Biome var4) {
