--- net/minecraft/client/gui/screens/CreateFlatWorldScreen.java
+++ net/minecraft/client/gui/screens/CreateFlatWorldScreen.java
@@ -116,13 +116,13 @@
 
       public void setSelected(@Nullable CreateFlatWorldScreen.DetailsList.Entry var1) {
          super.setSelected(var1);
          if (var1 != null) {
             FlatLayerInfo var2 = (FlatLayerInfo)CreateFlatWorldScreen.this.generator.getLayersInfo().get(CreateFlatWorldScreen.this.generator.getLayersInfo().size() - this.children().indexOf(var1) - 1);
-            Item var3 = var2.getBlockState().getBlock().asItem();
-            if (var3 != Items.AIR) {
-               NarratorChatListener.INSTANCE.sayNow((new TranslatableComponent("narrator.select", new Object[]{var3.getName(new ItemStack(var3))})).getString());
+            ItemStack var3 = new ItemStack(var2.getBlockState().getBlock());
+            if (!var3.isEmpty()) {
+               NarratorChatListener.INSTANCE.sayNow((new TranslatableComponent("narrator.select", new Object[]{var3.getItem().getName(var3)})).getString());
             }
          }
 
          CreateFlatWorldScreen.this.updateButtonValidity();
       }
@@ -190,21 +190,19 @@
             }
          }
 
          private void blitSlot(PoseStack var1, int var2, int var3, ItemStack var4) {
             this.blitSlotBg(var1, var2 + 1, var3 + 1);
-            RenderSystem.enableRescaleNormal();
             if (!var4.isEmpty()) {
                CreateFlatWorldScreen.this.itemRenderer.renderGuiItem(var4, var2 + 2, var3 + 2);
             }
 
-            RenderSystem.disableRescaleNormal();
          }
 
          private void blitSlotBg(PoseStack var1, int var2, int var3) {
-            RenderSystem.color4f(1.0F, 1.0F, 1.0F, 1.0F);
-            DetailsList.this.minecraft.getTextureManager().bind(GuiComponent.STATS_ICON_LOCATION);
+            RenderSystem.setShaderColor(1.0F, 1.0F, 1.0F, 1.0F);
+            RenderSystem.setShaderTexture(0, GuiComponent.STATS_ICON_LOCATION);
             GuiComponent.blit(var1, var2, var3, CreateFlatWorldScreen.this.getBlitOffset(), 0.0F, 0.0F, 18, 18, 128, 128);
          }
       }
    }
 }
