--- net/minecraft/client/gui/screens/DatapackLoadFailureScreen.java
+++ net/minecraft/client/gui/screens/DatapackLoadFailureScreen.java
@@ -16,14 +16,14 @@
    }
 
    protected void init() {
       super.init();
       this.message = MultiLineLabel.create(this.font, this.getTitle(), this.width - 50);
-      this.addButton(new Button(this.width / 2 - 155, this.height / 6 + 96, 150, 20, new TranslatableComponent("datapackFailure.safeMode"), (var1) -> {
+      this.addRenderableWidget(new Button(this.width / 2 - 155, this.height / 6 + 96, 150, 20, new TranslatableComponent("datapackFailure.safeMode"), (var1) -> {
          this.callback.run();
       }));
-      this.addButton(new Button(this.width / 2 - 155 + 160, this.height / 6 + 96, 150, 20, new TranslatableComponent("gui.toTitle"), (var1) -> {
+      this.addRenderableWidget(new Button(this.width / 2 - 155 + 160, this.height / 6 + 96, 150, 20, new TranslatableComponent("gui.toTitle"), (var1) -> {
          this.minecraft.setScreen((Screen)null);
       }));
    }
 
    public void render(PoseStack var1, int var2, int var3, float var4) {
