--- net/minecraft/client/gui/screens/DeathScreen.java
+++ net/minecraft/client/gui/screens/DeathScreen.java
@@ -1,8 +1,7 @@
 package net.minecraft.client.gui.screens;
 
-import com.mojang.blaze3d.systems.RenderSystem;
 import com.mojang.blaze3d.vertex.PoseStack;
 import java.util.Iterator;
 import javax.annotation.Nullable;
 import net.minecraft.ChatFormatting;
 import net.minecraft.client.gui.components.AbstractWidget;
@@ -76,14 +75,14 @@
       this.minecraft.setScreen(new TitleScreen());
    }
 
    public void render(PoseStack var1, int var2, int var3, float var4) {
       this.fillGradient(var1, 0, 0, this.width, this.height, 1615855616, -1602211792);
-      RenderSystem.pushMatrix();
-      RenderSystem.scalef(2.0F, 2.0F, 2.0F);
+      var1.pushPose();
+      var1.scale(2.0F, 2.0F, 2.0F);
       drawCenteredString(var1, this.font, this.title, this.width / 2 / 2, 30, 16777215);
-      RenderSystem.popMatrix();
+      var1.popPose();
       if (this.causeOfDeath != null) {
          drawCenteredString(var1, this.font, this.causeOfDeath, this.width / 2, 85, 16777215);
       }
 
       drawCenteredString(var1, this.font, this.deathScore, this.width / 2, 100, 16777215);
