--- net/minecraft/client/gui/screens/DemoIntroScreen.java
+++ net/minecraft/client/gui/screens/DemoIntroScreen.java
@@ -35,12 +35,12 @@
       this.durationMessage = MultiLineLabel.create(this.font, new TranslatableComponent("demo.help.fullWrapped"), 218);
    }
 
    public void renderBackground(PoseStack var1) {
       super.renderBackground(var1);
-      RenderSystem.color4f(1.0F, 1.0F, 1.0F, 1.0F);
-      this.minecraft.getTextureManager().bind(DEMO_BACKGROUND_LOCATION);
+      RenderSystem.setShaderColor(1.0F, 1.0F, 1.0F, 1.0F);
+      RenderSystem.setShaderTexture(0, DEMO_BACKGROUND_LOCATION);
       int var2 = (this.width - 248) / 2;
       int var3 = (this.height - 166) / 2;
       this.blit(var1, var2, var3, 0, 0, 248, 166);
    }
 
