--- net/minecraft/client/gui/screens/ErrorScreen.java
+++ net/minecraft/client/gui/screens/ErrorScreen.java
@@ -13,11 +13,11 @@
       this.message = var2;
    }
 
    protected void init() {
       super.init();
-      this.addButton(new Button(this.width / 2 - 100, 140, 200, 20, CommonComponents.GUI_CANCEL, (var1) -> {
+      this.addRenderableWidget(new Button(this.width / 2 - 100, 140, 200, 20, CommonComponents.GUI_CANCEL, (var1) -> {
          this.minecraft.setScreen((Screen)null);
       }));
    }
 
    public void render(PoseStack var1, int var2, int var3, float var4) {
