--- net/minecraft/client/gui/screens/InBedChatScreen.java
+++ net/minecraft/client/gui/screens/InBedChatScreen.java
@@ -11,11 +11,11 @@
       super("");
    }
 
    protected void init() {
       super.init();
-      this.addButton(new Button(this.width / 2 - 100, this.height - 40, 200, 20, new TranslatableComponent("multiplayer.stopSleeping"), (var1) -> {
+      this.addRenderableWidget(new Button(this.width / 2 - 100, this.height - 40, 200, 20, new TranslatableComponent("multiplayer.stopSleeping"), (var1) -> {
          this.sendWakeUp();
       }));
    }
 
    public void onClose() {
