--- net/minecraft/client/gui/screens/LoadingOverlay.java
+++ net/minecraft/client/gui/screens/LoadingOverlay.java
@@ -7,10 +7,11 @@
 import java.io.InputStream;
 import java.util.Optional;
 import java.util.function.Consumer;
 import net.minecraft.Util;
 import net.minecraft.client.Minecraft;
+import net.minecraft.client.renderer.GameRenderer;
 import net.minecraft.client.renderer.texture.AbstractTexture;
 import net.minecraft.client.renderer.texture.SimpleTexture;
 import net.minecraft.client.resources.metadata.texture.TextureMetadataSection;
 import net.minecraft.resources.ResourceLocation;
 import net.minecraft.server.packs.PackType;
@@ -80,20 +81,19 @@
       int var13 = (int)((double)this.minecraft.getWindow().getGuiScaledHeight() * 0.5D);
       double var14 = Math.min((double)this.minecraft.getWindow().getGuiScaledWidth() * 0.75D, (double)this.minecraft.getWindow().getGuiScaledHeight()) * 0.25D;
       int var16 = (int)(var14 * 0.5D);
       double var17 = var14 * 4.0D;
       int var19 = (int)(var17 * 0.5D);
-      this.minecraft.getTextureManager().bind(MOJANG_STUDIOS_LOGO_LOCATION);
+      RenderSystem.setShaderTexture(0, MOJANG_STUDIOS_LOGO_LOCATION);
       RenderSystem.enableBlend();
       RenderSystem.blendEquation(32774);
       RenderSystem.blendFunc(770, 1);
-      RenderSystem.alphaFunc(516, 0.0F);
-      RenderSystem.color4f(1.0F, 1.0F, 1.0F, var11);
+      RenderSystem.setShader(GameRenderer::getPositionTexShader);
+      RenderSystem.setShaderColor(1.0F, 1.0F, 1.0F, var11);
       blit(var1, var12 - var19, var13 - var16, var19, (int)var14, -0.0625F, 0.0F, 120, 60, 120, 120);
       blit(var1, var12, var13 - var16, var19, (int)var14, 0.0625F, 60.0F, 120, 60, 120, 120);
       RenderSystem.defaultBlendFunc();
-      RenderSystem.defaultAlphaFunc();
       RenderSystem.disableBlend();
       int var20 = (int)((double)this.minecraft.getWindow().getGuiScaledHeight() * 0.8325D);
       float var21 = this.reload.getActualProgress();
       this.currentProgress = Mth.clamp(this.currentProgress * 0.95F + var21 * 0.050000012F, 0.0F, 1.0F);
       if (var9 < 1.0F) {
