--- net/minecraft/client/gui/screens/MenuScreens.java
+++ net/minecraft/client/gui/screens/MenuScreens.java
@@ -104,11 +104,11 @@
       register(MenuType.STONECUTTER, StonecutterScreen::new);
    }
 
    interface ScreenConstructor<T extends AbstractContainerMenu, U extends Screen & MenuAccess<T>> {
       default void fromPacket(Component var1, MenuType<T> var2, Minecraft var3, int var4) {
-         Screen var5 = this.create(var2.create(var4, var3.player.inventory), var3.player.inventory, var1);
+         Screen var5 = this.create(var2.create(var4, var3.player.getInventory()), var3.player.getInventory(), var1);
          var3.player.containerMenu = ((MenuAccess)var5).getMenu();
          var3.setScreen(var5);
       }
 
       U create(T var1, Inventory var2, Component var3);
