--- net/minecraft/client/gui/screens/OptionsSubScreen.java
+++ net/minecraft/client/gui/screens/OptionsSubScreen.java
@@ -1,10 +1,10 @@
 package net.minecraft.client.gui.screens;
 
+import com.google.common.collect.ImmutableList;
 import java.util.List;
 import java.util.Optional;
-import javax.annotation.Nullable;
 import net.minecraft.client.Options;
 import net.minecraft.client.gui.components.OptionsList;
 import net.minecraft.client.gui.components.TooltipAccessor;
 import net.minecraft.network.chat.Component;
 import net.minecraft.util.FormattedCharSequence;
@@ -25,16 +25,10 @@
 
    public void onClose() {
       this.minecraft.setScreen(this.lastScreen);
    }
 
-   @Nullable
    public static List<FormattedCharSequence> tooltipAt(OptionsList var0, int var1, int var2) {
       Optional var3 = var0.getMouseOver((double)var1, (double)var2);
-      if (var3.isPresent() && var3.get() instanceof TooltipAccessor) {
-         Optional var4 = ((TooltipAccessor)var3.get()).getTooltip();
-         return (List)var4.orElse((Object)null);
-      } else {
-         return null;
-      }
+      return (List)(var3.isPresent() && var3.get() instanceof TooltipAccessor ? ((TooltipAccessor)var3.get()).getTooltip() : ImmutableList.of());
    }
 }
