--- net/minecraft/client/gui/screens/PauseScreen.java
+++ net/minecraft/client/gui/screens/PauseScreen.java
@@ -62,11 +62,12 @@
       }));
       Button var4 = (Button)this.addButton(new Button(this.width / 2 + 4, this.height / 4 + 96 + -16, 98, 20, new TranslatableComponent("menu.shareToLan"), (var1x) -> {
          this.minecraft.setScreen(new ShareToLanScreen(this));
       }));
       var4.active = this.minecraft.hasSingleplayerServer() && !this.minecraft.getSingleplayerServer().isPublished();
-      Button var5 = (Button)this.addButton(new Button(this.width / 2 - 102, this.height / 4 + 120 + -16, 204, 20, new TranslatableComponent("menu.returnToMenu"), (var1x) -> {
+      TranslatableComponent var5 = this.minecraft.isLocalServer() ? new TranslatableComponent("menu.returnToMenu") : new TranslatableComponent("menu.disconnect");
+      this.addButton(new Button(this.width / 2 - 102, this.height / 4 + 120 + -16, 204, 20, var5, (var1x) -> {
          boolean var2 = this.minecraft.isLocalServer();
          boolean var3 = this.minecraft.isConnectedToRealms();
          var1x.active = false;
          this.minecraft.level.disconnect();
          if (var2) {
@@ -83,14 +84,10 @@
          } else {
             this.minecraft.setScreen(new JoinMultiplayerScreen(new TitleScreen()));
          }
 
       }));
-      if (!this.minecraft.isLocalServer()) {
-         var5.setMessage(new TranslatableComponent("menu.disconnect"));
-      }
-
    }
 
    public void tick() {
       super.tick();
    }
