--- net/minecraft/client/gui/screens/PopupScreen.java
+++ net/minecraft/client/gui/screens/PopupScreen.java
@@ -10,10 +10,13 @@
 import net.minecraft.client.gui.components.MultiLineLabel;
 import net.minecraft.network.chat.Component;
 import net.minecraft.network.chat.FormattedText;
 
 public class PopupScreen extends Screen {
+   private static final int BUTTON_PADDING = 20;
+   private static final int BUTTON_MARGIN = 5;
+   private static final int BUTTON_HEIGHT = 20;
    private final FormattedText message;
    private final ImmutableList<PopupScreen.ButtonOption> buttonOptions;
    private MultiLineLabel messageLines;
    private int contentTop;
    private int buttonWidth;
