--- net/minecraft/client/gui/screens/ProgressScreen.java
+++ net/minecraft/client/gui/screens/ProgressScreen.java
@@ -13,13 +13,15 @@
    private Component header;
    @Nullable
    private Component stage;
    private int progress;
    private boolean stop;
+   private final boolean clearScreenAfterStop;
 
-   public ProgressScreen() {
+   public ProgressScreen(boolean var1) {
       super(NarratorChatListener.NO_TITLE);
+      this.clearScreenAfterStop = var1;
    }
 
    public boolean shouldCloseOnEsc() {
       return false;
    }
@@ -46,11 +48,11 @@
       this.stop = true;
    }
 
    public void render(PoseStack var1, int var2, int var3, float var4) {
       if (this.stop) {
-         if (!this.minecraft.isConnectedToRealms()) {
+         if (this.clearScreenAfterStop) {
             this.minecraft.setScreen((Screen)null);
          }
 
       } else {
          this.renderBackground(var1);
