--- net/minecraft/client/gui/screens/VideoSettingsScreen.java
+++ net/minecraft/client/gui/screens/VideoSettingsScreen.java
@@ -14,11 +14,10 @@
 import net.minecraft.client.gui.components.Button;
 import net.minecraft.client.gui.components.OptionsList;
 import net.minecraft.client.renderer.GpuWarnlistManager;
 import net.minecraft.network.chat.CommonComponents;
 import net.minecraft.network.chat.Component;
-import net.minecraft.network.chat.FormattedText;
 import net.minecraft.network.chat.TextComponent;
 import net.minecraft.network.chat.TranslatableComponent;
 
 public class VideoSettingsScreen extends OptionsSubScreen {
    private static final Component FABULOUS;
@@ -46,12 +45,12 @@
    protected void init() {
       this.list = new OptionsList(this.minecraft, this.width, this.height, 32, this.height - 32, 25);
       this.list.addBig(new FullscreenResolutionProgressOption(this.minecraft.getWindow()));
       this.list.addBig(Option.BIOME_BLEND_RADIUS);
       this.list.addSmall(OPTIONS);
-      this.children.add(this.list);
-      this.addButton(new Button(this.width / 2 - 100, this.height - 27, 200, 20, CommonComponents.GUI_DONE, (var1) -> {
+      this.addWidget(this.list);
+      this.addRenderableWidget(new Button(this.width / 2 - 100, this.height - 27, 200, 20, CommonComponents.GUI_DONE, (var1) -> {
          this.minecraft.options.save();
          this.minecraft.getWindow().changeFullscreenVideoMode();
          this.minecraft.setScreen(this.lastScreen);
       }));
    }
@@ -71,11 +70,11 @@
          if (this.options.guiScale != var6) {
             this.minecraft.resizeDisplay();
          }
 
          if (this.gpuWarnlistManager.isShowingWarning()) {
-            ArrayList var7 = Lists.newArrayList(new FormattedText[]{WARNING_MESSAGE, NEW_LINE});
+            ArrayList var7 = Lists.newArrayList(new Component[]{WARNING_MESSAGE, NEW_LINE});
             String var8 = this.gpuWarnlistManager.getRendererWarnings();
             if (var8 != null) {
                var7.add(NEW_LINE);
                var7.add((new TranslatableComponent("options.graphics.warning.renderer", new Object[]{var8})).withStyle(ChatFormatting.GRAY));
             }
