--- net/minecraft/client/gui/screens/WinScreen.java
+++ net/minecraft/client/gui/screens/WinScreen.java
@@ -5,10 +5,11 @@
 import com.mojang.blaze3d.systems.RenderSystem;
 import com.mojang.blaze3d.vertex.BufferBuilder;
 import com.mojang.blaze3d.vertex.DefaultVertexFormat;
 import com.mojang.blaze3d.vertex.PoseStack;
 import com.mojang.blaze3d.vertex.Tesselator;
+import com.mojang.blaze3d.vertex.VertexFormat;
 import it.unimi.dsi.fastutil.ints.IntOpenHashSet;
 import it.unimi.dsi.fastutil.ints.IntSet;
 import java.io.BufferedReader;
 import java.io.InputStream;
 import java.io.InputStreamReader;
@@ -17,10 +18,11 @@
 import java.util.List;
 import java.util.Random;
 import net.minecraft.ChatFormatting;
 import net.minecraft.client.gui.GuiComponent;
 import net.minecraft.client.gui.chat.NarratorChatListener;
+import net.minecraft.client.renderer.GameRenderer;
 import net.minecraft.network.chat.TextComponent;
 import net.minecraft.resources.ResourceLocation;
 import net.minecraft.server.packs.resources.Resource;
 import net.minecraft.util.FormattedCharSequence;
 import org.apache.commons.io.IOUtils;
@@ -154,11 +156,12 @@
 
       }
    }
 
    private void renderBg(int var1, int var2, float var3) {
-      this.minecraft.getTextureManager().bind(GuiComponent.BACKGROUND_LOCATION);
+      RenderSystem.setShader(GameRenderer::getPositionTexColorShader);
+      RenderSystem.setShaderTexture(0, GuiComponent.BACKGROUND_LOCATION);
       int var4 = this.width;
       float var5 = -this.time * 0.5F * this.scrollSpeed;
       float var6 = (float)this.height - this.time * 0.5F * this.scrollSpeed;
       float var7 = 0.015625F;
       float var8 = this.time * 0.02F;
@@ -174,11 +177,11 @@
 
       var8 *= var8;
       var8 = var8 * 96.0F / 255.0F;
       Tesselator var11 = Tesselator.getInstance();
       BufferBuilder var12 = var11.getBuilder();
-      var12.begin(7, DefaultVertexFormat.POSITION_TEX_COLOR);
+      var12.begin(VertexFormat.Mode.QUADS, DefaultVertexFormat.POSITION_TEX_COLOR);
       var12.vertex(0.0D, (double)this.height, (double)this.getBlitOffset()).uv(0.0F, var5 * 0.015625F).color(var8, var8, var8, 1.0F).endVertex();
       var12.vertex((double)var4, (double)this.height, (double)this.getBlitOffset()).uv((float)var4 * 0.015625F, var5 * 0.015625F).color(var8, var8, var8, 1.0F).endVertex();
       var12.vertex((double)var4, 0.0D, (double)this.getBlitOffset()).uv((float)var4 * 0.015625F, var6 * 0.015625F).color(var8, var8, var8, 1.0F).endVertex();
       var12.vertex(0.0D, 0.0D, (double)this.getBlitOffset()).uv(0.0F, var6 * 0.015625F).color(var8, var8, var8, 1.0F).endVertex();
       var11.end();
@@ -189,32 +192,30 @@
       boolean var5 = true;
       int var6 = this.width / 2 - 137;
       int var7 = this.height + 50;
       this.time += var4;
       float var8 = -this.time * this.scrollSpeed;
-      RenderSystem.pushMatrix();
-      RenderSystem.translatef(0.0F, var8, 0.0F);
-      this.minecraft.getTextureManager().bind(LOGO_LOCATION);
-      RenderSystem.color4f(1.0F, 1.0F, 1.0F, 1.0F);
-      RenderSystem.enableAlphaTest();
+      var1.pushPose();
+      var1.translate(0.0D, (double)var8, 0.0D);
+      RenderSystem.setShaderTexture(0, LOGO_LOCATION);
+      RenderSystem.setShaderColor(1.0F, 1.0F, 1.0F, 1.0F);
       RenderSystem.enableBlend();
       this.blitOutlineBlack(var6, var7, (var2x, var3x) -> {
          this.blit(var1, var2x + 0, var3x, 0, 0, 155, 44);
          this.blit(var1, var2x + 155, var3x, 0, 45, 155, 44);
       });
       RenderSystem.disableBlend();
-      this.minecraft.getTextureManager().bind(EDITION_LOCATION);
+      RenderSystem.setShaderTexture(0, EDITION_LOCATION);
       blit(var1, var6 + 88, var7 + 37, 0.0F, 0.0F, 98, 14, 128, 16);
-      RenderSystem.disableAlphaTest();
       int var9 = var7 + 100;
 
       int var10;
       for(var10 = 0; var10 < this.lines.size(); ++var10) {
          if (var10 == this.lines.size() - 1) {
             float var11 = (float)var9 + var8 - (float)(this.height / 2 - 6);
             if (var11 < 0.0F) {
-               RenderSystem.translatef(0.0F, -var11, 0.0F);
+               var1.translate(0.0D, (double)(-var11), 0.0D);
             }
          }
 
          if ((float)var9 + var8 + 12.0F + 8.0F > 0.0F && (float)var9 + var8 < (float)this.height) {
             FormattedCharSequence var14 = (FormattedCharSequence)this.lines.get(var10);
@@ -227,19 +228,20 @@
          }
 
          var9 += 12;
       }
 
-      RenderSystem.popMatrix();
-      this.minecraft.getTextureManager().bind(VIGNETTE_LOCATION);
+      var1.popPose();
+      RenderSystem.setShader(GameRenderer::getPositionTexColorShader);
+      RenderSystem.setShaderTexture(0, VIGNETTE_LOCATION);
       RenderSystem.enableBlend();
       RenderSystem.blendFunc(GlStateManager.SourceFactor.ZERO, GlStateManager.DestFactor.ONE_MINUS_SRC_COLOR);
       var10 = this.width;
       int var15 = this.height;
       Tesselator var12 = Tesselator.getInstance();
       BufferBuilder var13 = var12.getBuilder();
-      var13.begin(7, DefaultVertexFormat.POSITION_TEX_COLOR);
+      var13.begin(VertexFormat.Mode.QUADS, DefaultVertexFormat.POSITION_TEX_COLOR);
       var13.vertex(0.0D, (double)var15, (double)this.getBlitOffset()).uv(0.0F, 1.0F).color(1.0F, 1.0F, 1.0F, 1.0F).endVertex();
       var13.vertex((double)var10, (double)var15, (double)this.getBlitOffset()).uv(1.0F, 1.0F).color(1.0F, 1.0F, 1.0F, 1.0F).endVertex();
       var13.vertex((double)var10, 0.0D, (double)this.getBlitOffset()).uv(1.0F, 0.0F).color(1.0F, 1.0F, 1.0F, 1.0F).endVertex();
       var13.vertex(0.0D, 0.0D, (double)this.getBlitOffset()).uv(0.0F, 0.0F).color(1.0F, 1.0F, 1.0F, 1.0F).endVertex();
       var12.end();
