--- net/minecraft/server/level/progress/LoggerChunkProgressListener.java
+++ net/minecraft/server/level/progress/LoggerChunkProgressListener.java
@@ -32,15 +32,18 @@
       }
 
       int var3 = this.getProgress();
       if (Util.getMillis() > this.nextTickTime) {
          this.nextTickTime += 500L;
-         LOGGER.info((new TranslatableComponent("menu.preparingSpawn", new Object[]{Mth.clamp(var3, 0, 100)})).getString());
+         LOGGER.info((new TranslatableComponent("menu.preparingSpawn", new Object[]{Mth.clamp((int)var3, (int)0, (int)100)})).getString());
       }
 
    }
 
+   public void start() {
+   }
+
    public void stop() {
       LOGGER.info((String)"Time elapsed: {} ms", (Object)(Util.getMillis() - this.startTime));
       this.nextTickTime = Long.MAX_VALUE;
    }
 
