--- net/minecraft/client/gui/screens/controls/ControlsScreen.java
+++ net/minecraft/client/gui/screens/controls/ControlsScreen.java
@@ -22,28 +22,28 @@
    public ControlsScreen(Screen var1, Options var2) {
       super(var1, var2, new TranslatableComponent("controls.title"));
    }
 
    protected void init() {
-      this.addButton(new Button(this.width / 2 - 155, 18, 150, 20, new TranslatableComponent("options.mouse_settings"), (var1) -> {
+      this.addRenderableWidget(new Button(this.width / 2 - 155, 18, 150, 20, new TranslatableComponent("options.mouse_settings"), (var1) -> {
          this.minecraft.setScreen(new MouseSettingsScreen(this, this.options));
       }));
-      this.addButton(Option.AUTO_JUMP.createButton(this.options, this.width / 2 - 155 + 160, 18, 150));
+      this.addRenderableWidget(Option.AUTO_JUMP.createButton(this.options, this.width / 2 - 155 + 160, 18, 150));
       this.controlList = new ControlList(this, this.minecraft);
-      this.children.add(this.controlList);
-      this.resetButton = (Button)this.addButton(new Button(this.width / 2 - 155, this.height - 29, 150, 20, new TranslatableComponent("controls.resetAll"), (var1) -> {
+      this.addWidget(this.controlList);
+      this.resetButton = (Button)this.addRenderableWidget(new Button(this.width / 2 - 155, this.height - 29, 150, 20, new TranslatableComponent("controls.resetAll"), (var1) -> {
          KeyMapping[] var2 = this.options.keyMappings;
          int var3 = var2.length;
 
          for(int var4 = 0; var4 < var3; ++var4) {
             KeyMapping var5 = var2[var4];
             var5.setKey(var5.getDefaultKey());
          }
 
          KeyMapping.resetMapping();
       }));
-      this.addButton(new Button(this.width / 2 - 155 + 160, this.height - 29, 150, 20, CommonComponents.GUI_DONE, (var1) -> {
+      this.addRenderableWidget(new Button(this.width / 2 - 155 + 160, this.height - 29, 150, 20, CommonComponents.GUI_DONE, (var1) -> {
          this.minecraft.setScreen(this.lastScreen);
       }));
    }
 
    public boolean mouseClicked(double var1, double var3, int var5) {
