--- net/minecraft/nbt/NbtOps.java
+++ net/minecraft/nbt/NbtOps.java
@@ -281,11 +281,11 @@
    }
 
    public DataResult<Consumer<Consumer<Tag>>> getList(Tag var1) {
       if (var1 instanceof CollectionTag) {
          CollectionTag var2 = (CollectionTag)var1;
-         var2.getClass();
+         Objects.requireNonNull(var2);
          return DataResult.success(var2::forEach);
       } else {
          return DataResult.error("Not a list: " + var1);
       }
    }
