--- net/minecraft/nbt/Tag.java
+++ net/minecraft/nbt/Tag.java
@@ -1,18 +1,11 @@
 package net.minecraft.nbt;
 
 import java.io.DataOutput;
 import java.io.IOException;
-import net.minecraft.ChatFormatting;
-import net.minecraft.network.chat.Component;
 
 public interface Tag {
-   ChatFormatting SYNTAX_HIGHLIGHTING_KEY = ChatFormatting.AQUA;
-   ChatFormatting SYNTAX_HIGHLIGHTING_STRING = ChatFormatting.GREEN;
-   ChatFormatting SYNTAX_HIGHLIGHTING_NUMBER = ChatFormatting.GOLD;
-   ChatFormatting SYNTAX_HIGHLIGHTING_NUMBER_TYPE = ChatFormatting.RED;
-
    void write(DataOutput var1) throws IOException;
 
    String toString();
 
    byte getId();
@@ -20,14 +13,10 @@
    TagType<?> getType();
 
    Tag copy();
 
    default String getAsString() {
-      return this.toString();
+      return (new StringTagVisitor()).visit(this);
    }
 
-   default Component getPrettyDisplay() {
-      return this.getPrettyDisplay("", 0);
-   }
-
-   Component getPrettyDisplay(String var1, int var2);
+   void accept(TagVisitor var1);
 }
