--- net/minecraft/world/level/portal/PortalForcer.java
+++ net/minecraft/world/level/portal/PortalForcer.java
@@ -54,11 +54,11 @@
       double var4 = -1.0D;
       BlockPos var6 = null;
       double var7 = -1.0D;
       BlockPos var9 = null;
       WorldBorder var10 = this.level.getWorldBorder();
-      int var11 = this.level.getHeight() - 1;
+      int var11 = this.level.getMinBuildHeight() + this.level.getLogicalHeight() - 1;
       BlockPos.MutableBlockPos var12 = var1.mutable();
       Iterator var13 = BlockPos.spiralAround(var1, 16, Direction.EAST, Direction.SOUTH).iterator();
 
       while(true) {
          BlockPos.MutableBlockPos var14;
@@ -71,11 +71,11 @@
                      var4 = var7;
                   }
 
                   int var24;
                   if (var4 == -1.0D) {
-                     var6 = (new BlockPos(var1.getX(), Mth.clamp(var1.getY(), 70, this.level.getHeight() - 10), var1.getZ())).immutable();
+                     var6 = (new BlockPos(var1.getX(), Mth.clamp(var1.getY(), 70, this.level.getMinBuildHeight() + this.level.getLogicalHeight() - 10), var1.getZ())).immutable();
                      Direction var22 = var3.getClockWise();
                      if (!var10.isWithinBounds(var6)) {
                         return Optional.empty();
                      }
 
@@ -117,15 +117,15 @@
             } while(!var10.isWithinBounds((BlockPos)var14));
          } while(!var10.isWithinBounds((BlockPos)var14.move(var3, 1)));
 
          var14.move(var3.getOpposite(), 1);
 
-         for(int var17 = var15; var17 >= 0; --var17) {
+         for(int var17 = var15; var17 >= this.level.getMinBuildHeight(); --var17) {
             var14.setY(var17);
             if (this.level.isEmptyBlock(var14)) {
                int var18;
-               for(var18 = var17; var17 > 0 && this.level.isEmptyBlock(var14.move(Direction.DOWN)); --var17) {
+               for(var18 = var17; var17 > this.level.getMinBuildHeight() && this.level.isEmptyBlock(var14.move(Direction.DOWN)); --var17) {
                }
 
                if (var17 + 4 <= var11) {
                   int var19 = var18 - var17;
                   if (var19 <= 0 || var19 >= 3) {
