--- net/minecraft/world/level/portal/PortalForcer.java
+++ net/minecraft/world/level/portal/PortalForcer.java
@@ -18,10 +18,23 @@
 import net.minecraft.world.level.block.state.properties.BlockStateProperties;
 import net.minecraft.world.level.border.WorldBorder;
 import net.minecraft.world.level.levelgen.Heightmap;
 
 public class PortalForcer {
+   private static final int TICKET_RADIUS = 3;
+   private static final int SEARCH_RADIUS = 128;
+   private static final int CREATE_RADIUS = 16;
+   private static final int FRAME_HEIGHT = 5;
+   private static final int FRAME_WIDTH = 4;
+   private static final int FRAME_BOX = 3;
+   private static final int FRAME_HEIGHT_START = -1;
+   private static final int FRAME_HEIGHT_END = 4;
+   private static final int FRAME_WIDTH_START = -1;
+   private static final int FRAME_WIDTH_END = 3;
+   private static final int FRAME_BOX_START = -1;
+   private static final int FRAME_BOX_END = 2;
+   private static final int NOTHING_FOUND = -1;
    private final ServerLevel level;
 
    public PortalForcer(ServerLevel var1) {
       this.level = var1;
    }
@@ -54,11 +67,11 @@
       double var4 = -1.0D;
       BlockPos var6 = null;
       double var7 = -1.0D;
       BlockPos var9 = null;
       WorldBorder var10 = this.level.getWorldBorder();
-      int var11 = this.level.getHeight() - 1;
+      int var11 = this.level.getMinBuildHeight() + this.level.getLogicalHeight() - 1;
       BlockPos.MutableBlockPos var12 = var1.mutable();
       Iterator var13 = BlockPos.spiralAround(var1, 16, Direction.EAST, Direction.SOUTH).iterator();
 
       while(true) {
          BlockPos.MutableBlockPos var14;
@@ -71,11 +84,11 @@
                      var4 = var7;
                   }
 
                   int var24;
                   if (var4 == -1.0D) {
-                     var6 = (new BlockPos(var1.getX(), Mth.clamp(var1.getY(), 70, this.level.getHeight() - 10), var1.getZ())).immutable();
+                     var6 = (new BlockPos(var1.getX(), Mth.clamp((int)var1.getY(), (int)70, (int)(this.level.getMinBuildHeight() + this.level.getLogicalHeight() - 10)), var1.getZ())).immutable();
                      Direction var22 = var3.getClockWise();
                      if (!var10.isWithinBounds(var6)) {
                         return Optional.empty();
                      }
 
@@ -117,15 +130,15 @@
             } while(!var10.isWithinBounds((BlockPos)var14));
          } while(!var10.isWithinBounds((BlockPos)var14.move(var3, 1)));
 
          var14.move(var3.getOpposite(), 1);
 
-         for(int var17 = var15; var17 >= 0; --var17) {
+         for(int var17 = var15; var17 >= this.level.getMinBuildHeight(); --var17) {
             var14.setY(var17);
             if (this.level.isEmptyBlock(var14)) {
                int var18;
-               for(var18 = var17; var17 > 0 && this.level.isEmptyBlock(var14.move(Direction.DOWN)); --var17) {
+               for(var18 = var17; var17 > this.level.getMinBuildHeight() && this.level.isEmptyBlock(var14.move(Direction.DOWN)); --var17) {
                }
 
                if (var17 + 4 <= var11) {
                   int var19 = var18 - var17;
                   if (var19 <= 0 || var19 >= 3) {
