--- com/mojang/realmsclient/gui/screens/RealmsPlayerScreen.java
+++ com/mojang/realmsclient/gui/screens/RealmsPlayerScreen.java
@@ -3,10 +3,11 @@
 import com.mojang.blaze3d.systems.RenderSystem;
 import com.mojang.blaze3d.vertex.BufferBuilder;
 import com.mojang.blaze3d.vertex.DefaultVertexFormat;
 import com.mojang.blaze3d.vertex.PoseStack;
 import com.mojang.blaze3d.vertex.Tesselator;
+import com.mojang.blaze3d.vertex.VertexFormat;
 import com.mojang.realmsclient.client.RealmsClient;
 import com.mojang.realmsclient.dto.Ops;
 import com.mojang.realmsclient.dto.PlayerInfo;
 import com.mojang.realmsclient.dto.RealmsServer;
 import com.mojang.realmsclient.exception.RealmsServiceException;
@@ -211,11 +212,11 @@
       Tesselator var6 = Tesselator.getInstance();
       BufferBuilder var7 = var6.getBuilder();
       this.minecraft.getTextureManager().bind(OPTIONS_BACKGROUND);
       RenderSystem.color4f(1.0F, 1.0F, 1.0F, 1.0F);
       float var8 = 32.0F;
-      var7.begin(7, DefaultVertexFormat.POSITION_TEX_COLOR);
+      var7.begin(VertexFormat.Mode.QUADS, DefaultVertexFormat.POSITION_TEX_COLOR);
       var7.vertex(0.0D, (double)this.height, 0.0D).uv(0.0F, (float)(this.height - var5) / 32.0F + 0.0F).color(64, 64, 64, 255).endVertex();
       var7.vertex((double)this.width, (double)this.height, 0.0D).uv((float)this.width / 32.0F, (float)(this.height - var5) / 32.0F + 0.0F).color(64, 64, 64, 255).endVertex();
       var7.vertex((double)this.width, (double)var5, 0.0D).uv((float)this.width / 32.0F, 0.0F).color(64, 64, 64, 255).endVertex();
       var7.vertex(0.0D, (double)var5, 0.0D).uv(0.0F, 0.0F).color(64, 64, 64, 255).endVertex();
       var6.end();
