--- com/mojang/realmsclient/gui/screens/RealmsInviteScreen.java
+++ com/mojang/realmsclient/gui/screens/RealmsInviteScreen.java
@@ -2,17 +2,17 @@
 
 import com.mojang.blaze3d.vertex.PoseStack;
 import com.mojang.realmsclient.client.RealmsClient;
 import com.mojang.realmsclient.dto.RealmsServer;
 import javax.annotation.Nullable;
+import net.minecraft.client.gui.chat.NarratorChatListener;
 import net.minecraft.client.gui.components.Button;
 import net.minecraft.client.gui.components.EditBox;
 import net.minecraft.client.gui.screens.Screen;
 import net.minecraft.network.chat.CommonComponents;
 import net.minecraft.network.chat.Component;
 import net.minecraft.network.chat.TranslatableComponent;
-import net.minecraft.realms.NarrationHelper;
 import net.minecraft.realms.RealmsScreen;
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
 
 public class RealmsInviteScreen extends RealmsScreen {
@@ -25,10 +25,11 @@
    private final Screen lastScreen;
    @Nullable
    private Component errorMsg;
 
    public RealmsInviteScreen(RealmsConfigureWorldScreen var1, Screen var2, RealmsServer var3) {
+      super(NarratorChatListener.NO_TITLE);
       this.configureScreen = var1;
       this.lastScreen = var2;
       this.serverData = var3;
    }
 
@@ -39,14 +40,14 @@
    public void init() {
       this.minecraft.keyboardHandler.setSendRepeatsToGui(true);
       this.profileName = new EditBox(this.minecraft.font, this.width / 2 - 100, row(2), 200, 20, (EditBox)null, new TranslatableComponent("mco.configure.world.invite.profile.name"));
       this.addWidget(this.profileName);
       this.setInitialFocus(this.profileName);
-      this.addButton(new Button(this.width / 2 - 100, row(10), 200, 20, new TranslatableComponent("mco.configure.world.buttons.invite"), (var1) -> {
+      this.addRenderableWidget(new Button(this.width / 2 - 100, row(10), 200, 20, new TranslatableComponent("mco.configure.world.buttons.invite"), (var1) -> {
          this.onInvite();
       }));
-      this.addButton(new Button(this.width / 2 - 100, row(12), 200, 20, CommonComponents.GUI_CANCEL, (var1) -> {
+      this.addRenderableWidget(new Button(this.width / 2 - 100, row(12), 200, 20, CommonComponents.GUI_CANCEL, (var1) -> {
          this.minecraft.setScreen(this.lastScreen);
       }));
    }
 
    public void removed() {
@@ -74,11 +75,11 @@
       }
    }
 
    private void showError(Component var1) {
       this.errorMsg = var1;
-      NarrationHelper.now(var1.getString());
+      NarratorChatListener.INSTANCE.sayNow(var1);
    }
 
    public boolean keyPressed(int var1, int var2, int var3) {
       if (var1 == 256) {
          this.minecraft.setScreen(this.lastScreen);
