--- com/mojang/realmsclient/gui/screens/RealmsSubscriptionInfoScreen.java
+++ com/mojang/realmsclient/gui/screens/RealmsSubscriptionInfoScreen.java
@@ -36,10 +36,11 @@
    private final RealmsServer serverData;
    private final Screen mainScreen;
    private Component daysLeft;
    private String startDate;
    private Subscription.SubscriptionType type;
+   private static final String PURCHASE_LINK = "https://aka.ms/ExtendJavaRealms";
 
    public RealmsSubscriptionInfoScreen(Screen var1, RealmsServer var2, Screen var3) {
       this.lastScreen = var1;
       this.serverData = var2;
       this.mainScreen = var3;
