--- com/mojang/realmsclient/gui/screens/RealmsBackupInfoScreen.java
+++ com/mojang/realmsclient/gui/screens/RealmsBackupInfoScreen.java
@@ -11,12 +11,15 @@
 import net.minecraft.client.gui.screens.Screen;
 import net.minecraft.network.chat.CommonComponents;
 import net.minecraft.network.chat.Component;
 import net.minecraft.network.chat.TextComponent;
 import net.minecraft.realms.RealmsScreen;
+import net.minecraft.world.Difficulty;
+import net.minecraft.world.level.GameType;
 
 public class RealmsBackupInfoScreen extends RealmsScreen {
+   private static final Component TEXT_UNKNOWN = new TextComponent("UNKNOWN");
    private final Screen lastScreen;
    private final Backup backup;
    private RealmsBackupInfoScreen.BackupInfoList backupInfoList;
 
    public RealmsBackupInfoScreen(Screen var1, Backup var2) {
@@ -66,21 +69,21 @@
       }
    }
 
    private Component gameDifficultyMetadata(String var1) {
       try {
-         return RealmsSlotOptionsScreen.DIFFICULTIES[Integer.parseInt(var1)];
+         return ((Difficulty)RealmsSlotOptionsScreen.DIFFICULTIES.get(Integer.parseInt(var1))).getDisplayName();
       } catch (Exception var3) {
-         return new TextComponent("UNKNOWN");
+         return TEXT_UNKNOWN;
       }
    }
 
    private Component gameModeMetadata(String var1) {
       try {
-         return RealmsSlotOptionsScreen.GAME_MODES[Integer.parseInt(var1)];
+         return ((GameType)RealmsSlotOptionsScreen.GAME_MODES.get(Integer.parseInt(var1))).getShortDisplayName();
       } catch (Exception var3) {
-         return new TextComponent("UNKNOWN");
+         return TEXT_UNKNOWN;
       }
    }
 
    class BackupInfoList extends ObjectSelectionList<RealmsBackupInfoScreen.BackupInfoListEntry> {
       public BackupInfoList(Minecraft var2) {
