--- com/mojang/realmsclient/gui/screens/RealmsResetNormalWorldScreen.java
+++ com/mojang/realmsclient/gui/screens/RealmsResetNormalWorldScreen.java
@@ -15,11 +15,11 @@
    private final RealmsResetWorldScreen lastScreen;
    private RealmsLabel titleLabel;
    private EditBox seedEdit;
    private Boolean generateStructures = true;
    private Integer levelTypeIndex = 0;
-   private Component buttonTitle;
+   private final Component buttonTitle;
 
    public RealmsResetNormalWorldScreen(RealmsResetWorldScreen var1, Component var2) {
       this.lastScreen = var1;
       this.buttonTitle = var2;
    }
