--- com/mojang/realmsclient/gui/screens/RealmsSettingsScreen.java
+++ com/mojang/realmsclient/gui/screens/RealmsSettingsScreen.java
@@ -9,10 +9,11 @@
 import net.minecraft.network.chat.TranslatableComponent;
 import net.minecraft.realms.RealmsLabel;
 import net.minecraft.realms.RealmsScreen;
 
 public class RealmsSettingsScreen extends RealmsScreen {
+   private static final int COMPONENT_WIDTH = 212;
    private static final Component NAME_LABEL = new TranslatableComponent("mco.configure.world.name");
    private static final Component DESCRIPTION_LABEL = new TranslatableComponent("mco.configure.world.description");
    private final RealmsConfigureWorldScreen configureWorldScreen;
    private final RealmsServer serverData;
    private Button doneButton;
