--- com/mojang/realmsclient/gui/screens/RealmsBackupScreen.java
+++ com/mojang/realmsclient/gui/screens/RealmsBackupScreen.java
@@ -290,32 +290,32 @@
          return DateFormat.getDateTimeInstance(3, 3).format(var1);
       }
 
       private void drawRestore(PoseStack var1, int var2, int var3, int var4, int var5) {
          boolean var6 = var4 >= var2 && var4 <= var2 + 12 && var5 >= var3 && var5 <= var3 + 14 && var5 < RealmsBackupScreen.this.height - 15 && var5 > 32;
-         RealmsBackupScreen.this.minecraft.getTextureManager().bind(RealmsBackupScreen.RESTORE_ICON_LOCATION);
-         RenderSystem.color4f(1.0F, 1.0F, 1.0F, 1.0F);
-         RenderSystem.pushMatrix();
-         RenderSystem.scalef(0.5F, 0.5F, 0.5F);
+         RenderSystem.setShaderTexture(0, RealmsBackupScreen.RESTORE_ICON_LOCATION);
+         RenderSystem.setShaderColor(1.0F, 1.0F, 1.0F, 1.0F);
+         var1.pushPose();
+         var1.scale(0.5F, 0.5F, 0.5F);
          float var7 = var6 ? 28.0F : 0.0F;
          GuiComponent.blit(var1, var2 * 2, var3 * 2, 0.0F, var7, 23, 28, 23, 56);
-         RenderSystem.popMatrix();
+         var1.popPose();
          if (var6) {
             RealmsBackupScreen.this.toolTip = RealmsBackupScreen.RESTORE_TOOLTIP;
          }
 
       }
 
       private void drawInfo(PoseStack var1, int var2, int var3, int var4, int var5) {
          boolean var6 = var4 >= var2 && var4 <= var2 + 8 && var5 >= var3 && var5 <= var3 + 8 && var5 < RealmsBackupScreen.this.height - 15 && var5 > 32;
-         RealmsBackupScreen.this.minecraft.getTextureManager().bind(RealmsBackupScreen.PLUS_ICON_LOCATION);
-         RenderSystem.color4f(1.0F, 1.0F, 1.0F, 1.0F);
-         RenderSystem.pushMatrix();
-         RenderSystem.scalef(0.5F, 0.5F, 0.5F);
+         RenderSystem.setShaderTexture(0, RealmsBackupScreen.PLUS_ICON_LOCATION);
+         RenderSystem.setShaderColor(1.0F, 1.0F, 1.0F, 1.0F);
+         var1.pushPose();
+         var1.scale(0.5F, 0.5F, 0.5F);
          float var7 = var6 ? 15.0F : 0.0F;
          GuiComponent.blit(var1, var2 * 2, var3 * 2, 0.0F, var7, 15, 15, 15, 30);
-         RenderSystem.popMatrix();
+         var1.popPose();
          if (var6) {
             RealmsBackupScreen.this.toolTip = RealmsBackupScreen.HAS_CHANGES_TOOLTIP;
          }
 
       }
