--- com/mojang/realmsclient/gui/screens/RealmsUploadScreen.java
+++ com/mojang/realmsclient/gui/screens/RealmsUploadScreen.java
@@ -5,10 +5,11 @@
 import com.mojang.blaze3d.systems.RenderSystem;
 import com.mojang.blaze3d.vertex.BufferBuilder;
 import com.mojang.blaze3d.vertex.DefaultVertexFormat;
 import com.mojang.blaze3d.vertex.PoseStack;
 import com.mojang.blaze3d.vertex.Tesselator;
+import com.mojang.blaze3d.vertex.VertexFormat;
 import com.mojang.realmsclient.Unit;
 import com.mojang.realmsclient.client.FileUpload;
 import com.mojang.realmsclient.client.RealmsClient;
 import com.mojang.realmsclient.client.UploadStatus;
 import com.mojang.realmsclient.dto.UploadInfo;
@@ -168,11 +169,11 @@
       RenderSystem.disableTexture();
       double var4 = (double)(this.width / 2 - 100);
       double var6 = 0.5D;
       Tesselator var8 = Tesselator.getInstance();
       BufferBuilder var9 = var8.getBuilder();
-      var9.begin(7, DefaultVertexFormat.POSITION_COLOR);
+      var9.begin(VertexFormat.Mode.QUADS, DefaultVertexFormat.POSITION_COLOR);
       var9.vertex(var4 - 0.5D, 95.5D, 0.0D).color(217, 210, 210, 255).endVertex();
       var9.vertex(var4 + 200.0D * var2 + 0.5D, 95.5D, 0.0D).color(217, 210, 210, 255).endVertex();
       var9.vertex(var4 + 200.0D * var2 + 0.5D, 79.5D, 0.0D).color(217, 210, 210, 255).endVertex();
       var9.vertex(var4 - 0.5D, 79.5D, 0.0D).color(217, 210, 210, 255).endVertex();
       var9.vertex(var4, 95.0D, 0.0D).color(128, 128, 128, 255).endVertex();
@@ -346,11 +347,11 @@
                UPLOAD_LOCK.unlock();
                this.showDots = false;
                this.backButton.visible = true;
                this.cancelButton.visible = false;
                if (var1 != null) {
-                  LOGGER.debug("Deleting file " + var1.getAbsolutePath());
+                  LOGGER.debug("Deleting file {}", var1.getAbsolutePath());
                   var1.delete();
                }
 
             } else {
                return;
