--- com/mojang/realmsclient/gui/screens/RealmsPlayerScreen.java
+++ com/mojang/realmsclient/gui/screens/RealmsPlayerScreen.java
@@ -3,10 +3,11 @@
 import com.mojang.blaze3d.systems.RenderSystem;
 import com.mojang.blaze3d.vertex.BufferBuilder;
 import com.mojang.blaze3d.vertex.DefaultVertexFormat;
 import com.mojang.blaze3d.vertex.PoseStack;
 import com.mojang.blaze3d.vertex.Tesselator;
+import com.mojang.blaze3d.vertex.VertexFormat;
 import com.mojang.realmsclient.client.RealmsClient;
 import com.mojang.realmsclient.dto.Ops;
 import com.mojang.realmsclient.dto.PlayerInfo;
 import com.mojang.realmsclient.dto.RealmsServer;
 import com.mojang.realmsclient.exception.RealmsServiceException;
@@ -14,10 +15,11 @@
 import java.util.Iterator;
 import javax.annotation.Nullable;
 import net.minecraft.client.gui.GuiComponent;
 import net.minecraft.client.gui.components.Button;
 import net.minecraft.client.gui.components.ObjectSelectionList;
+import net.minecraft.client.renderer.GameRenderer;
 import net.minecraft.client.resources.language.I18n;
 import net.minecraft.network.chat.CommonComponents;
 import net.minecraft.network.chat.Component;
 import net.minecraft.network.chat.FormattedText;
 import net.minecraft.network.chat.TextComponent;
@@ -208,14 +210,15 @@
       }
 
       int var5 = row(12) + 20;
       Tesselator var6 = Tesselator.getInstance();
       BufferBuilder var7 = var6.getBuilder();
-      this.minecraft.getTextureManager().bind(OPTIONS_BACKGROUND);
-      RenderSystem.color4f(1.0F, 1.0F, 1.0F, 1.0F);
+      RenderSystem.setShader(GameRenderer::getPositionTexColorShader);
+      RenderSystem.setShaderTexture(0, OPTIONS_BACKGROUND);
+      RenderSystem.setShaderColor(1.0F, 1.0F, 1.0F, 1.0F);
       float var8 = 32.0F;
-      var7.begin(7, DefaultVertexFormat.POSITION_TEX_COLOR);
+      var7.begin(VertexFormat.Mode.QUADS, DefaultVertexFormat.POSITION_TEX_COLOR);
       var7.vertex(0.0D, (double)this.height, 0.0D).uv(0.0F, (float)(this.height - var5) / 32.0F + 0.0F).color(64, 64, 64, 255).endVertex();
       var7.vertex((double)this.width, (double)this.height, 0.0D).uv((float)this.width / 32.0F, (float)(this.height - var5) / 32.0F + 0.0F).color(64, 64, 64, 255).endVertex();
       var7.vertex((double)this.width, (double)var5, 0.0D).uv((float)this.width / 32.0F, 0.0F).color(64, 64, 64, 255).endVertex();
       var7.vertex(0.0D, (double)var5, 0.0D).uv(0.0F, 0.0F).color(64, 64, 64, 255).endVertex();
       var6.end();
@@ -242,12 +245,12 @@
       }
    }
 
    private void drawRemoveIcon(PoseStack var1, int var2, int var3, int var4, int var5) {
       boolean var6 = var4 >= var2 && var4 <= var2 + 9 && var5 >= var3 && var5 <= var3 + 9 && var5 < row(12) + 20 && var5 > row(1);
-      this.minecraft.getTextureManager().bind(CROSS_ICON_LOCATION);
-      RenderSystem.color4f(1.0F, 1.0F, 1.0F, 1.0F);
+      RenderSystem.setShaderTexture(0, CROSS_ICON_LOCATION);
+      RenderSystem.setShaderColor(1.0F, 1.0F, 1.0F, 1.0F);
       float var7 = var6 ? 7.0F : 0.0F;
       GuiComponent.blit(var1, var2, var3, 0.0F, var7, 8, 7, 8, 14);
       if (var6) {
          this.toolTip = REMOVE_ENTRY_TOOLTIP;
          this.hoveredUserAction = RealmsPlayerScreen.UserAction.REMOVE;
@@ -255,12 +258,12 @@
 
    }
 
    private void drawOpped(PoseStack var1, int var2, int var3, int var4, int var5) {
       boolean var6 = var4 >= var2 && var4 <= var2 + 9 && var5 >= var3 && var5 <= var3 + 9 && var5 < row(12) + 20 && var5 > row(1);
-      this.minecraft.getTextureManager().bind(OP_ICON_LOCATION);
-      RenderSystem.color4f(1.0F, 1.0F, 1.0F, 1.0F);
+      RenderSystem.setShaderTexture(0, OP_ICON_LOCATION);
+      RenderSystem.setShaderColor(1.0F, 1.0F, 1.0F, 1.0F);
       float var7 = var6 ? 8.0F : 0.0F;
       GuiComponent.blit(var1, var2, var3, 0.0F, var7, 8, 8, 8, 16);
       if (var6) {
          this.toolTip = OP_TOOLTIP;
          this.hoveredUserAction = RealmsPlayerScreen.UserAction.TOGGLE_OP;
@@ -268,12 +271,12 @@
 
    }
 
    private void drawNormal(PoseStack var1, int var2, int var3, int var4, int var5) {
       boolean var6 = var4 >= var2 && var4 <= var2 + 9 && var5 >= var3 && var5 <= var3 + 9 && var5 < row(12) + 20 && var5 > row(1);
-      this.minecraft.getTextureManager().bind(USER_ICON_LOCATION);
-      RenderSystem.color4f(1.0F, 1.0F, 1.0F, 1.0F);
+      RenderSystem.setShaderTexture(0, USER_ICON_LOCATION);
+      RenderSystem.setShaderColor(1.0F, 1.0F, 1.0F, 1.0F);
       float var7 = var6 ? 8.0F : 0.0F;
       GuiComponent.blit(var1, var2, var3, 0.0F, var7, 8, 8, 8, 16);
       if (var6) {
          this.toolTip = NORMAL_USER_TOOLTIP;
          this.hoveredUserAction = RealmsPlayerScreen.UserAction.TOGGLE_OP;
@@ -309,11 +312,11 @@
             RealmsPlayerScreen.this.drawNormal(var1, RealmsPlayerScreen.this.column1X + RealmsPlayerScreen.this.columnWidth - 10, var4 + 1, var5, var6);
          }
 
          RealmsPlayerScreen.this.drawRemoveIcon(var1, RealmsPlayerScreen.this.column1X + RealmsPlayerScreen.this.columnWidth - 22, var4 + 2, var5, var6);
          RealmsTextureManager.withBoundFace(var2.getUuid(), () -> {
-            RenderSystem.color4f(1.0F, 1.0F, 1.0F, 1.0F);
+            RenderSystem.setShaderColor(1.0F, 1.0F, 1.0F, 1.0F);
             GuiComponent.blit(var1, RealmsPlayerScreen.this.column1X + 2 + 2, var4 + 1, 8, 8, 8.0F, 8.0F, 8, 8, 64, 64);
             GuiComponent.blit(var1, RealmsPlayerScreen.this.column1X + 2 + 2, var4 + 1, 8, 8, 40.0F, 8.0F, 8, 8, 64, 64);
          });
       }
    }
