--- com/mojang/realmsclient/gui/screens/RealmsPendingInvitesScreen.java
+++ com/mojang/realmsclient/gui/screens/RealmsPendingInvitesScreen.java
@@ -181,10 +181,11 @@
    private boolean shouldAcceptAndRejectButtonBeVisible(int var1) {
       return var1 != -1;
    }
 
    class Entry extends ObjectSelectionList.Entry<RealmsPendingInvitesScreen.Entry> {
+      private static final int TEXT_LEFT = 38;
       private final PendingInvite pendingInvite;
       private final List<RowButton> rowButtons;
 
       Entry(PendingInvite var2) {
          this.pendingInvite = var2;
@@ -204,11 +205,11 @@
          RealmsPendingInvitesScreen.this.font.draw(var1, var2.worldName, (float)(var3 + 38), (float)(var4 + 1), 16777215);
          RealmsPendingInvitesScreen.this.font.draw(var1, var2.worldOwnerName, (float)(var3 + 38), (float)(var4 + 12), 7105644);
          RealmsPendingInvitesScreen.this.font.draw(var1, RealmsUtil.convertToAgePresentationFromInstant(var2.date), (float)(var3 + 38), (float)(var4 + 24), 7105644);
          RowButton.drawButtonsInRow(var1, this.rowButtons, RealmsPendingInvitesScreen.this.pendingInvitationSelectionList, var3, var4, var5, var6);
          RealmsTextureManager.withBoundFace(var2.worldOwnerUuid, () -> {
-            RenderSystem.color4f(1.0F, 1.0F, 1.0F, 1.0F);
+            RenderSystem.setShaderColor(1.0F, 1.0F, 1.0F, 1.0F);
             GuiComponent.blit(var1, var3, var4, 32, 32, 8.0F, 8.0F, 8, 8, 64, 64);
             GuiComponent.blit(var1, var3, var4, 32, 32, 40.0F, 8.0F, 8, 8, 64, 64);
          });
       }
 
@@ -216,12 +217,12 @@
          RejectRowButton() {
             super(15, 15, 235, 5);
          }
 
          protected void draw(PoseStack var1, int var2, int var3, boolean var4) {
-            RealmsPendingInvitesScreen.this.minecraft.getTextureManager().bind(RealmsPendingInvitesScreen.REJECT_ICON_LOCATION);
-            RenderSystem.color4f(1.0F, 1.0F, 1.0F, 1.0F);
+            RenderSystem.setShaderTexture(0, RealmsPendingInvitesScreen.REJECT_ICON_LOCATION);
+            RenderSystem.setShaderColor(1.0F, 1.0F, 1.0F, 1.0F);
             float var5 = var4 ? 19.0F : 0.0F;
             GuiComponent.blit(var1, var2, var3, var5, 0.0F, 18, 18, 37, 18);
             if (var4) {
                RealmsPendingInvitesScreen.this.toolTip = RealmsPendingInvitesScreen.REJECT_INVITE_TOOLTIP;
             }
@@ -237,12 +238,12 @@
          AcceptRowButton() {
             super(15, 15, 215, 5);
          }
 
          protected void draw(PoseStack var1, int var2, int var3, boolean var4) {
-            RealmsPendingInvitesScreen.this.minecraft.getTextureManager().bind(RealmsPendingInvitesScreen.ACCEPT_ICON_LOCATION);
-            RenderSystem.color4f(1.0F, 1.0F, 1.0F, 1.0F);
+            RenderSystem.setShaderTexture(0, RealmsPendingInvitesScreen.ACCEPT_ICON_LOCATION);
+            RenderSystem.setShaderColor(1.0F, 1.0F, 1.0F, 1.0F);
             float var5 = var4 ? 19.0F : 0.0F;
             GuiComponent.blit(var1, var2, var3, var5, 0.0F, 18, 18, 37, 18);
             if (var4) {
                RealmsPendingInvitesScreen.this.toolTip = RealmsPendingInvitesScreen.ACCEPT_INVITE_TOOLTIP;
             }
