--- com/mojang/realmsclient/gui/screens/RealmsNotificationsScreen.java
+++ com/mojang/realmsclient/gui/screens/RealmsNotificationsScreen.java
@@ -4,20 +4,21 @@
 import com.mojang.blaze3d.vertex.PoseStack;
 import com.mojang.realmsclient.client.RealmsClient;
 import com.mojang.realmsclient.exception.RealmsServiceException;
 import com.mojang.realmsclient.gui.RealmsDataFetcher;
 import net.minecraft.Util;
+import net.minecraft.client.Minecraft;
 import net.minecraft.client.gui.GuiComponent;
 import net.minecraft.client.gui.screens.TitleScreen;
 import net.minecraft.realms.RealmsScreen;
 import net.minecraft.resources.ResourceLocation;
 
 public class RealmsNotificationsScreen extends RealmsScreen {
    private static final ResourceLocation INVITE_ICON_LOCATION = new ResourceLocation("realms", "textures/gui/realms/invite_icon.png");
    private static final ResourceLocation TRIAL_ICON_LOCATION = new ResourceLocation("realms", "textures/gui/realms/trial_icon.png");
    private static final ResourceLocation NEWS_ICON_LOCATION = new ResourceLocation("realms", "textures/gui/realms/news_notification_mainscreen.png");
-   private static final RealmsDataFetcher REALMS_DATA_FETCHER = new RealmsDataFetcher();
+   private static final RealmsDataFetcher REALMS_DATA_FETCHER = new RealmsDataFetcher(Minecraft.getInstance(), RealmsClient.create());
    private volatile int numberOfPendingInvites;
    private static boolean checkedMcoAvailability;
    private static boolean trialAvailable;
    private static boolean validClient;
    private static boolean hasUnreadNews;
@@ -100,29 +101,29 @@
       int var6 = this.height / 4 + 48;
       int var7 = this.width / 2 + 80;
       int var8 = var6 + 48 + 2;
       int var9 = 0;
       if (hasUnreadNews) {
-         this.minecraft.getTextureManager().bind(NEWS_ICON_LOCATION);
-         RenderSystem.color4f(1.0F, 1.0F, 1.0F, 1.0F);
-         RenderSystem.pushMatrix();
-         RenderSystem.scalef(0.4F, 0.4F, 0.4F);
+         RenderSystem.setShaderTexture(0, NEWS_ICON_LOCATION);
+         RenderSystem.setShaderColor(1.0F, 1.0F, 1.0F, 1.0F);
+         var1.pushPose();
+         var1.scale(0.4F, 0.4F, 0.4F);
          GuiComponent.blit(var1, (int)((double)(var7 + 2 - var9) * 2.5D), (int)((double)var8 * 2.5D), 0.0F, 0.0F, 40, 40, 40, 40);
-         RenderSystem.popMatrix();
+         var1.popPose();
          var9 += 14;
       }
 
       if (var4 != 0) {
-         this.minecraft.getTextureManager().bind(INVITE_ICON_LOCATION);
-         RenderSystem.color4f(1.0F, 1.0F, 1.0F, 1.0F);
+         RenderSystem.setShaderTexture(0, INVITE_ICON_LOCATION);
+         RenderSystem.setShaderColor(1.0F, 1.0F, 1.0F, 1.0F);
          GuiComponent.blit(var1, var7 - var9, var8 - 6, 0.0F, 0.0F, 15, 25, 31, 25);
          var9 += 16;
       }
 
       if (trialAvailable) {
-         this.minecraft.getTextureManager().bind(TRIAL_ICON_LOCATION);
-         RenderSystem.color4f(1.0F, 1.0F, 1.0F, 1.0F);
+         RenderSystem.setShaderTexture(0, TRIAL_ICON_LOCATION);
+         RenderSystem.setShaderColor(1.0F, 1.0F, 1.0F, 1.0F);
          byte var10 = 0;
          if ((Util.getMillis() / 800L & 1L) == 1L) {
             var10 = 8;
          }
 
