--- com/mojang/realmsclient/gui/screens/RealmsSelectFileToUploadScreen.java
+++ com/mojang/realmsclient/gui/screens/RealmsSelectFileToUploadScreen.java
@@ -132,11 +132,11 @@
          return super.keyPressed(var1, var2, var3);
       }
    }
 
    private static Component gameModeName(LevelSummary var0) {
-      return var0.getGameMode().getDisplayName();
+      return var0.getGameMode().getLongDisplayName();
    }
 
    private static String formatLastPlayed(LevelSummary var0) {
       return DATE_FORMAT.format(new Date(var0.getLastPlayed()));
    }
