--- com/mojang/realmsclient/gui/screens/RealmsBrokenWorldScreen.java
+++ com/mojang/realmsclient/gui/screens/RealmsBrokenWorldScreen.java
@@ -16,10 +16,11 @@
 import java.util.Iterator;
 import java.util.List;
 import java.util.Map.Entry;
 import java.util.stream.Collectors;
 import java.util.stream.Stream;
+import javax.annotation.Nullable;
 import net.minecraft.client.gui.GuiComponent;
 import net.minecraft.client.gui.components.Button;
 import net.minecraft.client.gui.screens.Screen;
 import net.minecraft.network.chat.CommonComponents;
 import net.minecraft.network.chat.Component;
@@ -238,38 +239,38 @@
 
    private boolean isMinigame() {
       return this.serverData != null && this.serverData.worldType == RealmsServer.WorldType.MINIGAME;
    }
 
-   private void drawSlotFrame(PoseStack var1, int var2, int var3, int var4, int var5, boolean var6, String var7, int var8, long var9, String var11, boolean var12) {
+   private void drawSlotFrame(PoseStack var1, int var2, int var3, int var4, int var5, boolean var6, String var7, int var8, long var9, @Nullable String var11, boolean var12) {
       if (var12) {
-         this.minecraft.getTextureManager().bind(RealmsWorldSlotButton.EMPTY_SLOT_LOCATION);
+         RenderSystem.setShaderTexture(0, RealmsWorldSlotButton.EMPTY_SLOT_LOCATION);
       } else if (var11 != null && var9 != -1L) {
          RealmsTextureManager.bindWorldTemplate(String.valueOf(var9), var11);
       } else if (var8 == 1) {
-         this.minecraft.getTextureManager().bind(RealmsWorldSlotButton.DEFAULT_WORLD_SLOT_1);
+         RenderSystem.setShaderTexture(0, RealmsWorldSlotButton.DEFAULT_WORLD_SLOT_1);
       } else if (var8 == 2) {
-         this.minecraft.getTextureManager().bind(RealmsWorldSlotButton.DEFAULT_WORLD_SLOT_2);
+         RenderSystem.setShaderTexture(0, RealmsWorldSlotButton.DEFAULT_WORLD_SLOT_2);
       } else if (var8 == 3) {
-         this.minecraft.getTextureManager().bind(RealmsWorldSlotButton.DEFAULT_WORLD_SLOT_3);
+         RenderSystem.setShaderTexture(0, RealmsWorldSlotButton.DEFAULT_WORLD_SLOT_3);
       } else {
          RealmsTextureManager.bindWorldTemplate(String.valueOf(this.serverData.minigameId), this.serverData.minigameImage);
       }
 
       if (!var6) {
-         RenderSystem.color4f(0.56F, 0.56F, 0.56F, 1.0F);
+         RenderSystem.setShaderColor(0.56F, 0.56F, 0.56F, 1.0F);
       } else if (var6) {
          float var13 = 0.9F + 0.1F * Mth.cos((float)this.animTick * 0.2F);
-         RenderSystem.color4f(var13, var13, var13, 1.0F);
+         RenderSystem.setShaderColor(var13, var13, var13, 1.0F);
       }
 
       GuiComponent.blit(var1, var2 + 3, var3 + 3, 0.0F, 0.0F, 74, 74, 74, 74);
-      this.minecraft.getTextureManager().bind(RealmsWorldSlotButton.SLOT_FRAME_LOCATION);
+      RenderSystem.setShaderTexture(0, RealmsWorldSlotButton.SLOT_FRAME_LOCATION);
       if (var6) {
-         RenderSystem.color4f(1.0F, 1.0F, 1.0F, 1.0F);
+         RenderSystem.setShaderColor(1.0F, 1.0F, 1.0F, 1.0F);
       } else {
-         RenderSystem.color4f(0.56F, 0.56F, 0.56F, 1.0F);
+         RenderSystem.setShaderColor(0.56F, 0.56F, 0.56F, 1.0F);
       }
 
       GuiComponent.blit(var1, var2, var3, 0.0F, 0.0F, 80, 80, 80, 80);
       drawCenteredString(var1, this.font, var7, var2 + 40, var3 + 66, 16777215);
    }
