--- com/mojang/realmsclient/gui/screens/RealmsBrokenWorldScreen.java
+++ com/mojang/realmsclient/gui/screens/RealmsBrokenWorldScreen.java
@@ -16,10 +16,11 @@
 import java.util.Iterator;
 import java.util.List;
 import java.util.Map.Entry;
 import java.util.stream.Collectors;
 import java.util.stream.Stream;
+import javax.annotation.Nullable;
 import net.minecraft.client.gui.GuiComponent;
 import net.minecraft.client.gui.components.Button;
 import net.minecraft.client.gui.screens.Screen;
 import net.minecraft.network.chat.CommonComponents;
 import net.minecraft.network.chat.Component;
@@ -238,11 +239,11 @@
 
    private boolean isMinigame() {
       return this.serverData != null && this.serverData.worldType == RealmsServer.WorldType.MINIGAME;
    }
 
-   private void drawSlotFrame(PoseStack var1, int var2, int var3, int var4, int var5, boolean var6, String var7, int var8, long var9, String var11, boolean var12) {
+   private void drawSlotFrame(PoseStack var1, int var2, int var3, int var4, int var5, boolean var6, String var7, int var8, long var9, @Nullable String var11, boolean var12) {
       if (var12) {
          this.minecraft.getTextureManager().bind(RealmsWorldSlotButton.EMPTY_SLOT_LOCATION);
       } else if (var11 != null && var9 != -1L) {
          RealmsTextureManager.bindWorldTemplate(String.valueOf(var9), var11);
       } else if (var8 == 1) {
