--- com/mojang/realmsclient/gui/screens/UploadResult.java
+++ com/mojang/realmsclient/gui/screens/UploadResult.java
@@ -2,11 +2,11 @@
 
 public class UploadResult {
    public final int statusCode;
    public final String errorMessage;
 
-   private UploadResult(int var1, String var2) {
+   UploadResult(int var1, String var2) {
       this.statusCode = var1;
       this.errorMessage = var2;
    }
 
    public static class Builder {
