--- net/minecraft/world/level/levelgen/flat/FlatLayerInfo.java
+++ net/minecraft/world/level/levelgen/flat/FlatLayerInfo.java
@@ -4,41 +4,33 @@
 import com.mojang.serialization.codecs.RecordCodecBuilder;
 import net.minecraft.core.Registry;
 import net.minecraft.world.level.block.Block;
 import net.minecraft.world.level.block.Blocks;
 import net.minecraft.world.level.block.state.BlockState;
+import net.minecraft.world.level.dimension.DimensionType;
 
 public class FlatLayerInfo {
    public static final Codec<FlatLayerInfo> CODEC = RecordCodecBuilder.create((var0) -> {
-      return var0.group(Codec.intRange(0, 256).fieldOf("height").forGetter(FlatLayerInfo::getHeight), Registry.BLOCK.fieldOf("block").orElse(Blocks.AIR).forGetter((var0x) -> {
+      return var0.group(Codec.intRange(0, DimensionType.Y_SIZE).fieldOf("height").forGetter(FlatLayerInfo::getHeight), Registry.BLOCK.fieldOf("block").orElse(Blocks.AIR).forGetter((var0x) -> {
          return var0x.getBlockState().getBlock();
       })).apply(var0, FlatLayerInfo::new);
    });
-   private final BlockState blockState;
+   private final Block block;
    private final int height;
-   private int start;
 
    public FlatLayerInfo(int var1, Block var2) {
       this.height = var1;
-      this.blockState = var2.defaultBlockState();
+      this.block = var2;
    }
 
    public int getHeight() {
       return this.height;
    }
 
    public BlockState getBlockState() {
-      return this.blockState;
+      return this.block.defaultBlockState();
    }
 
-   public int getStart() {
-      return this.start;
-   }
-
-   public void setStart(int var1) {
-      this.start = var1;
-   }
-
    public String toString() {
-      return (this.height != 1 ? this.height + "*" : "") + Registry.BLOCK.getKey(this.blockState.getBlock());
+      return (this.height != 1 ? this.height + "*" : "") + Registry.BLOCK.getKey(this.block);
    }
 }
