--- net/minecraft/util/valueproviders/IntProviderType.java
+++ net/minecraft/util/valueproviders/IntProviderType.java
@@ -1,0 +1,17 @@
+package net.minecraft.util.valueproviders;
+
+import com.mojang.serialization.Codec;
+import net.minecraft.core.Registry;
+
+public interface IntProviderType<P extends IntProvider> {
+   IntProviderType<ConstantInt> CONSTANT = register("constant", ConstantInt.CODEC);
+   IntProviderType<UniformInt> UNIFORM = register("uniform", UniformInt.CODEC);
+
+   Codec<P> codec();
+
+   static <P extends IntProvider> IntProviderType<P> register(String var0, Codec<P> var1) {
+      return (IntProviderType)Registry.register(Registry.INT_PROVIDER_TYPES, (String)var0, () -> {
+         return var1;
+      });
+   }
+}
