--- net/minecraft/world/level/storage/loot/functions/SetContainerLootTable.java
+++ net/minecraft/world/level/storage/loot/functions/SetContainerLootTable.java
@@ -54,10 +54,22 @@
          }
 
       }
    }
 
+   public static LootItemConditionalFunction.Builder<?> withLootTable(ResourceLocation var0) {
+      return simpleBuilder((var1) -> {
+         return new SetContainerLootTable(var1, var0, 0L);
+      });
+   }
+
+   public static LootItemConditionalFunction.Builder<?> withLootTable(ResourceLocation var0, long var1) {
+      return simpleBuilder((var3) -> {
+         return new SetContainerLootTable(var3, var0, var1);
+      });
+   }
+
    public static class Serializer extends LootItemConditionalFunction.Serializer<SetContainerLootTable> {
       public Serializer() {
       }
 
       public void serialize(JsonObject var1, SetContainerLootTable var2, JsonSerializationContext var3) {
