--- net/minecraft/world/level/storage/loot/functions/SetNbtFunction.java
+++ net/minecraft/world/level/storage/loot/functions/SetNbtFunction.java
@@ -11,13 +11,13 @@
 import net.minecraft.world.item.ItemStack;
 import net.minecraft.world.level.storage.loot.LootContext;
 import net.minecraft.world.level.storage.loot.predicates.LootItemCondition;
 
 public class SetNbtFunction extends LootItemConditionalFunction {
-   private final CompoundTag tag;
+   final CompoundTag tag;
 
-   private SetNbtFunction(LootItemCondition[] var1, CompoundTag var2) {
+   SetNbtFunction(LootItemCondition[] var1, CompoundTag var2) {
       super(var1);
       this.tag = var2;
    }
 
    public LootItemFunctionType getType() {
