--- net/minecraft/world/entity/monster/AbstractIllager.java
+++ net/minecraft/world/entity/monster/AbstractIllager.java
@@ -21,20 +21,10 @@
 
    public AbstractIllager.IllagerArmPose getArmPose() {
       return AbstractIllager.IllagerArmPose.CROSSED;
    }
 
-   public class RaiderOpenDoorGoal extends OpenDoorGoal {
-      public RaiderOpenDoorGoal(Raider var2) {
-         super(var2, false);
-      }
-
-      public boolean canUse() {
-         return super.canUse() && AbstractIllager.this.hasActiveRaid();
-      }
-   }
-
    public static enum IllagerArmPose {
       CROSSED,
       ATTACKING,
       SPELLCASTING,
       BOW_AND_ARROW,
@@ -44,6 +34,16 @@
       NEUTRAL;
 
       private IllagerArmPose() {
       }
    }
+
+   protected class RaiderOpenDoorGoal extends OpenDoorGoal {
+      public RaiderOpenDoorGoal(Raider var2) {
+         super(var2, false);
+      }
+
+      public boolean canUse() {
+         return super.canUse() && AbstractIllager.this.hasActiveRaid();
+      }
+   }
 }
