--- net/minecraft/world/entity/monster/CaveSpider.java
+++ net/minecraft/world/entity/monster/CaveSpider.java
@@ -36,11 +36,11 @@
             } else if (this.level.getDifficulty() == Difficulty.HARD) {
                var2 = 15;
             }
 
             if (var2 > 0) {
-               ((LivingEntity)var1).addEffect(new MobEffectInstance(MobEffects.POISON, var2 * 20, 0));
+               ((LivingEntity)var1).addEffect(new MobEffectInstance(MobEffects.POISON, var2 * 20, 0), this);
             }
          }
 
          return true;
       } else {
