--- net/minecraft/world/entity/monster/Creeper.java
+++ net/minecraft/world/entity/monster/Creeper.java
@@ -72,18 +72,18 @@
 
    public int getMaxFallDistance() {
       return this.getTarget() == null ? 3 : 3 + (int)(this.getHealth() - 1.0F);
    }
 
-   public boolean causeFallDamage(float var1, float var2) {
-      boolean var3 = super.causeFallDamage(var1, var2);
+   public boolean causeFallDamage(float var1, float var2, DamageSource var3) {
+      boolean var4 = super.causeFallDamage(var1, var2, var3);
       this.swell = (int)((float)this.swell + var1 * 1.5F);
       if (this.swell > this.maxSwell - 5) {
          this.swell = this.maxSwell - 5;
       }
 
-      return var3;
+      return var4;
    }
 
    protected void defineSynchedData() {
       super.defineSynchedData();
       this.entityData.define(DATA_SWELL_DIR, -1);
@@ -191,11 +191,11 @@
       this.entityData.set(DATA_IS_POWERED, true);
    }
 
    protected InteractionResult mobInteract(Player var1, InteractionHand var2) {
       ItemStack var3 = var1.getItemInHand(var2);
-      if (var3.getItem() == Items.FLINT_AND_STEEL) {
+      if (var3.is(Items.FLINT_AND_STEEL)) {
          this.level.playSound(var1, this.getX(), this.getY(), this.getZ(), SoundEvents.FLINTANDSTEEL_USE, this.getSoundSource(), 1.0F, this.random.nextFloat() * 0.4F + 0.8F);
          if (!this.level.isClientSide) {
             this.ignite();
             var3.hurtAndBreak(1, var1, (var1x) -> {
                var1x.broadcastBreakEvent(var2);
@@ -212,11 +212,11 @@
       if (!this.level.isClientSide) {
          Explosion.BlockInteraction var1 = this.level.getGameRules().getBoolean(GameRules.RULE_MOBGRIEFING) ? Explosion.BlockInteraction.DESTROY : Explosion.BlockInteraction.NONE;
          float var2 = this.isPowered() ? 2.0F : 1.0F;
          this.dead = true;
          this.level.explode(this, this.getX(), this.getY(), this.getZ(), (float)this.explosionRadius * var2, var1);
-         this.remove();
+         this.discard();
          this.spawnLingeringCloud();
       }
 
    }
 
