--- net/minecraft/world/entity/monster/Guardian.java
+++ net/minecraft/world/entity/monster/Guardian.java
@@ -34,10 +34,11 @@
 import net.minecraft.world.entity.ai.goal.RandomStrollGoal;
 import net.minecraft.world.entity.ai.goal.target.NearestAttackableTargetGoal;
 import net.minecraft.world.entity.ai.navigation.PathNavigation;
 import net.minecraft.world.entity.ai.navigation.WaterBoundPathNavigation;
 import net.minecraft.world.entity.animal.Squid;
+import net.minecraft.world.entity.animal.axolotl.Axolotl;
 import net.minecraft.world.entity.player.Player;
 import net.minecraft.world.level.Level;
 import net.minecraft.world.level.LevelAccessor;
 import net.minecraft.world.level.LevelReader;
 import net.minecraft.world.level.pathfinder.BlockPathTypes;
@@ -446,9 +447,9 @@
       public GuardianAttackSelector(Guardian var1) {
          this.guardian = var1;
       }
 
       public boolean test(@Nullable LivingEntity var1) {
-         return (var1 instanceof Player || var1 instanceof Squid) && var1.distanceToSqr(this.guardian) > 9.0D;
+         return (var1 instanceof Player || var1 instanceof Squid || var1 != null && Axolotl.NOT_PLAYING_DEAD_SELECTOR.test(var1)) && var1.distanceToSqr(this.guardian) > 9.0D;
       }
    }
 }
