--- net/minecraft/world/entity/monster/Illusioner.java
+++ net/minecraft/world/entity/monster/Illusioner.java
@@ -39,10 +39,13 @@
 import net.minecraft.world.level.ServerLevelAccessor;
 import net.minecraft.world.phys.AABB;
 import net.minecraft.world.phys.Vec3;
 
 public class Illusioner extends SpellcasterIllager implements RangedAttackMob {
+   private static final int NUM_ILLUSIONS = 4;
+   private static final int ILLUSION_TRANSITION_TICKS = 3;
+   private static final int ILLUSION_SPREAD = 3;
    private int clientSideIllusionTicks;
    private final Vec3[][] clientSideIllusionOffsets;
 
    public Illusioner(EntityType<? extends Illusioner> var1, Level var2) {
       super(var1, var2);
