--- net/minecraft/world/entity/monster/Ravager.java
+++ net/minecraft/world/entity/monster/Ravager.java
@@ -37,10 +37,11 @@
 import net.minecraft.world.level.Level;
 import net.minecraft.world.level.LevelReader;
 import net.minecraft.world.level.block.Block;
 import net.minecraft.world.level.block.LeavesBlock;
 import net.minecraft.world.level.block.state.BlockState;
+import net.minecraft.world.level.gameevent.GameEvent;
 import net.minecraft.world.level.pathfinder.BlockPathTypes;
 import net.minecraft.world.level.pathfinder.PathFinder;
 import net.minecraft.world.level.pathfinder.WalkNodeEvaluator;
 import net.minecraft.world.phys.AABB;
 import net.minecraft.world.phys.Vec3;
@@ -119,11 +120,11 @@
       return !this.isNoAi() && this.getControllingPassenger() instanceof LivingEntity;
    }
 
    @Nullable
    public Entity getControllingPassenger() {
-      return this.getPassengers().isEmpty() ? null : (Entity)this.getPassengers().get(0);
+      return this.getFirstPassenger();
    }
 
    public void aiStep() {
       super.aiStep();
       if (this.isAlive()) {
@@ -220,13 +221,13 @@
 
    private void roar() {
       if (this.isAlive()) {
          List var1 = this.level.getEntitiesOfClass(LivingEntity.class, this.getBoundingBox().inflate(4.0D), NO_RAVAGER_AND_ALIVE);
 
-         Entity var3;
+         LivingEntity var3;
          for(Iterator var2 = var1.iterator(); var2.hasNext(); this.strongKnockback(var3)) {
-            var3 = (Entity)var2.next();
+            var3 = (LivingEntity)var2.next();
             if (!(var3 instanceof AbstractIllager)) {
                var3.hurt(DamageSource.mobAttack(this), 6.0F);
             }
          }
 
@@ -236,10 +237,12 @@
             double var4 = this.random.nextGaussian() * 0.2D;
             double var6 = this.random.nextGaussian() * 0.2D;
             double var8 = this.random.nextGaussian() * 0.2D;
             this.level.addParticle(ParticleTypes.POOF, var10.x, var10.y, var10.z, var4, var6, var8);
          }
+
+         this.level.gameEvent(this, GameEvent.RAVAGER_ROAR, this.eyeBlockPosition());
       }
 
    }
 
    private void strongKnockback(Entity var1) {
