--- net/minecraft/world/entity/monster/Spider.java
+++ net/minecraft/world/entity/monster/Spider.java
@@ -42,10 +42,11 @@
 import net.minecraft.world.level.block.state.BlockState;
 import net.minecraft.world.phys.Vec3;
 
 public class Spider extends Monster {
    private static final EntityDataAccessor<Byte> DATA_FLAGS_ID;
+   private static final float SPIDER_SPECIAL_EFFECT_CHANCE = 0.1F;
 
    public Spider(EntityType<? extends Spider> var1, Level var2) {
       super(var1, var2);
    }
 
