--- net/minecraft/world/entity/monster/Stray.java
+++ net/minecraft/world/entity/monster/Stray.java
@@ -12,18 +12,25 @@
 import net.minecraft.world.entity.projectile.AbstractArrow;
 import net.minecraft.world.entity.projectile.Arrow;
 import net.minecraft.world.item.ItemStack;
 import net.minecraft.world.level.Level;
 import net.minecraft.world.level.ServerLevelAccessor;
+import net.minecraft.world.level.block.Blocks;
 
 public class Stray extends AbstractSkeleton {
    public Stray(EntityType<? extends Stray> var1, Level var2) {
       super(var1, var2);
    }
 
    public static boolean checkStraySpawnRules(EntityType<Stray> var0, ServerLevelAccessor var1, MobSpawnType var2, BlockPos var3, Random var4) {
-      return checkMonsterSpawnRules(var0, var1, var2, var3, var4) && (var2 == MobSpawnType.SPAWNER || var1.canSeeSky(var3));
+      BlockPos var5 = var3;
+
+      do {
+         var5 = var5.above();
+      } while(var1.getBlockState(var5).is(Blocks.POWDER_SNOW));
+
+      return checkMonsterSpawnRules(var0, var1, var2, var3, var4) && (var2 == MobSpawnType.SPAWNER || var1.canSeeSky(var5.below()));
    }
 
    protected SoundEvent getAmbientSound() {
       return SoundEvents.STRAY_AMBIENT;
    }
