--- net/minecraft/world/entity/monster/Vindicator.java
+++ net/minecraft/world/entity/monster/Vindicator.java
@@ -44,10 +44,11 @@
 import net.minecraft.world.item.enchantment.Enchantments;
 import net.minecraft.world.level.Level;
 import net.minecraft.world.level.ServerLevelAccessor;
 
 public class Vindicator extends AbstractIllager {
+   private static final String TAG_JOHNNY = "Johnny";
    private static final Predicate<Difficulty> DOOR_BREAKING_PREDICATE = (var0) -> {
       return var0 == Difficulty.NORMAL || var0 == Difficulty.HARD;
    };
    private boolean isJohnny;
 
