--- net/minecraft/world/entity/monster/Witch.java
+++ net/minecraft/world/entity/monster/Witch.java
@@ -112,11 +112,11 @@
          if (this.isDrinkingPotion()) {
             if (this.usingTime-- <= 0) {
                this.setUsingItem(false);
                ItemStack var6 = this.getMainHandItem();
                this.setItemSlot(EquipmentSlot.MAINHAND, ItemStack.EMPTY);
-               if (var6.getItem() == Items.POTION) {
+               if (var6.is(Items.POTION)) {
                   List var5 = PotionUtils.getMobEffects(var6);
                   if (var5 != null) {
                      Iterator var3 = var5.iterator();
 
                      while(var3.hasNext()) {
