--- net/minecraft/world/entity/monster/Witch.java
+++ net/minecraft/world/entity/monster/Witch.java
@@ -9,11 +9,10 @@
 import net.minecraft.network.syncher.EntityDataSerializers;
 import net.minecraft.network.syncher.SynchedEntityData;
 import net.minecraft.sounds.SoundEvent;
 import net.minecraft.sounds.SoundEvents;
 import net.minecraft.tags.FluidTags;
-import net.minecraft.util.Mth;
 import net.minecraft.world.damagesource.DamageSource;
 import net.minecraft.world.effect.MobEffectInstance;
 import net.minecraft.world.effect.MobEffects;
 import net.minecraft.world.entity.EntityDimensions;
 import net.minecraft.world.entity.EntityType;
@@ -112,11 +111,11 @@
          if (this.isDrinkingPotion()) {
             if (this.usingTime-- <= 0) {
                this.setUsingItem(false);
                ItemStack var6 = this.getMainHandItem();
                this.setItemSlot(EquipmentSlot.MAINHAND, ItemStack.EMPTY);
-               if (var6.getItem() == Items.POTION) {
+               if (var6.is(Items.POTION)) {
                   List var5 = PotionUtils.getMobEffects(var6);
                   if (var5 != null) {
                      Iterator var3 = var5.iterator();
 
                      while(var3.hasNext()) {
@@ -194,37 +193,37 @@
       if (!this.isDrinkingPotion()) {
          Vec3 var3 = var1.getDeltaMovement();
          double var4 = var1.getX() + var3.x - this.getX();
          double var6 = var1.getEyeY() - 1.100000023841858D - this.getY();
          double var8 = var1.getZ() + var3.z - this.getZ();
-         float var10 = Mth.sqrt(var4 * var4 + var8 * var8);
-         Potion var11 = Potions.HARMING;
+         double var10 = Math.sqrt(var4 * var4 + var8 * var8);
+         Potion var12 = Potions.HARMING;
          if (var1 instanceof Raider) {
             if (var1.getHealth() <= 4.0F) {
-               var11 = Potions.HEALING;
+               var12 = Potions.HEALING;
             } else {
-               var11 = Potions.REGENERATION;
+               var12 = Potions.REGENERATION;
             }
 
             this.setTarget((LivingEntity)null);
-         } else if (var10 >= 8.0F && !var1.hasEffect(MobEffects.MOVEMENT_SLOWDOWN)) {
-            var11 = Potions.SLOWNESS;
+         } else if (var10 >= 8.0D && !var1.hasEffect(MobEffects.MOVEMENT_SLOWDOWN)) {
+            var12 = Potions.SLOWNESS;
          } else if (var1.getHealth() >= 8.0F && !var1.hasEffect(MobEffects.POISON)) {
-            var11 = Potions.POISON;
-         } else if (var10 <= 3.0F && !var1.hasEffect(MobEffects.WEAKNESS) && this.random.nextFloat() < 0.25F) {
-            var11 = Potions.WEAKNESS;
+            var12 = Potions.POISON;
+         } else if (var10 <= 3.0D && !var1.hasEffect(MobEffects.WEAKNESS) && this.random.nextFloat() < 0.25F) {
+            var12 = Potions.WEAKNESS;
          }
 
-         ThrownPotion var12 = new ThrownPotion(this.level, this);
-         var12.setItem(PotionUtils.setPotion(new ItemStack(Items.SPLASH_POTION), var11));
-         var12.xRot -= -20.0F;
-         var12.shoot(var4, var6 + (double)(var10 * 0.2F), var8, 0.75F, 8.0F);
+         ThrownPotion var13 = new ThrownPotion(this.level, this);
+         var13.setItem(PotionUtils.setPotion(new ItemStack(Items.SPLASH_POTION), var12));
+         var13.setXRot(var13.getXRot() - -20.0F);
+         var13.shoot(var4, var6 + var10 * 0.2D, var8, 0.75F, 8.0F);
          if (!this.isSilent()) {
             this.level.playSound((Player)null, this.getX(), this.getY(), this.getZ(), SoundEvents.WITCH_THROW, this.getSoundSource(), 1.0F, 0.8F + this.random.nextFloat() * 0.4F);
          }
 
-         this.level.addFreshEntity(var12);
+         this.level.addFreshEntity(var13);
       }
    }
 
    protected float getStandingEyeHeight(Pose var1, EntityDimensions var2) {
       return 1.62F;
