--- net/minecraft/world/entity/monster/WitherSkeleton.java
+++ net/minecraft/world/entity/monster/WitherSkeleton.java
@@ -88,11 +88,11 @@
    public boolean doHurtTarget(Entity var1) {
       if (!super.doHurtTarget(var1)) {
          return false;
       } else {
          if (var1 instanceof LivingEntity) {
-            ((LivingEntity)var1).addEffect(new MobEffectInstance(MobEffects.WITHER, 200));
+            ((LivingEntity)var1).addEffect(new MobEffectInstance(MobEffects.WITHER, 200), this);
          }
 
          return true;
       }
    }
