--- net/minecraft/world/entity/ai/goal/FollowParentGoal.java
+++ net/minecraft/world/entity/ai/goal/FollowParentGoal.java
@@ -4,10 +4,13 @@
 import java.util.List;
 import net.minecraft.world.entity.Entity;
 import net.minecraft.world.entity.animal.Animal;
 
 public class FollowParentGoal extends Goal {
+   public static final int HORIZONTAL_SCAN_RANGE = 8;
+   public static final int VERTICAL_SCAN_RANGE = 4;
+   public static final int DONT_FOLLOW_IF_CLOSER_THAN = 3;
    private final Animal animal;
    private Animal parent;
    private final double speedModifier;
    private int timeToRecalcPath;
 
