--- net/minecraft/world/entity/ai/goal/LandOnOwnersShoulderGoal.java
+++ net/minecraft/world/entity/ai/goal/LandOnOwnersShoulderGoal.java
@@ -12,11 +12,11 @@
       this.entity = var1;
    }
 
    public boolean canUse() {
       ServerPlayer var1 = (ServerPlayer)this.entity.getOwner();
-      boolean var2 = var1 != null && !var1.isSpectator() && !var1.abilities.flying && !var1.isInWater();
+      boolean var2 = var1 != null && !var1.isSpectator() && !var1.getAbilities().flying && !var1.isInWater() && !var1.isInPowderSnow;
       return !this.entity.isOrderedToSit() && var2 && this.entity.canSitOnShoulder();
    }
 
    public boolean isInterruptable() {
       return !this.isSittingOnShoulder;
