--- net/minecraft/world/entity/ai/goal/LlamaFollowCaravanGoal.java
+++ net/minecraft/world/entity/ai/goal/LlamaFollowCaravanGoal.java
@@ -10,10 +10,11 @@
 import net.minecraft.world.phys.Vec3;
 
 public class LlamaFollowCaravanGoal extends Goal {
    public final Llama llama;
    private double speedModifier;
+   private static final int CARAVAN_LIMIT = 8;
    private int distCheckCounter;
 
    public LlamaFollowCaravanGoal(Llama var1, double var2) {
       this.llama = var1;
       this.speedModifier = var2;
