--- net/minecraft/world/entity/ai/goal/MeleeAttackGoal.java
+++ net/minecraft/world/entity/ai/goal/MeleeAttackGoal.java
@@ -19,10 +19,11 @@
    private double pathedTargetZ;
    private int ticksUntilNextPathRecalculation;
    private int ticksUntilNextAttack;
    private final int attackInterval = 20;
    private long lastCanUseCheck;
+   private static final long COOLDOWN_BETWEEN_CAN_USE_CHECKS = 20L;
 
    public MeleeAttackGoal(PathfinderMob var1, double var2, boolean var4) {
       this.mob = var1;
       this.speedModifier = var2;
       this.followingTargetEvenIfNotSeen = var4;
