--- net/minecraft/world/entity/ai/goal/PanicGoal.java
+++ net/minecraft/world/entity/ai/goal/PanicGoal.java
@@ -4,11 +4,11 @@
 import javax.annotation.Nullable;
 import net.minecraft.core.BlockPos;
 import net.minecraft.tags.FluidTags;
 import net.minecraft.world.entity.Entity;
 import net.minecraft.world.entity.PathfinderMob;
-import net.minecraft.world.entity.ai.util.RandomPos;
+import net.minecraft.world.entity.ai.util.DefaultRandomPos;
 import net.minecraft.world.level.BlockGetter;
 import net.minecraft.world.phys.Vec3;
 
 public class PanicGoal extends Goal {
    protected final PathfinderMob mob;
@@ -41,11 +41,11 @@
          return this.findRandomPosition();
       }
    }
 
    protected boolean findRandomPosition() {
-      Vec3 var1 = RandomPos.getPos(this.mob, 5, 4);
+      Vec3 var1 = DefaultRandomPos.getPos(this.mob, 5, 4);
       if (var1 == null) {
          return false;
       } else {
          this.posX = var1.x;
          this.posY = var1.y;
