--- net/minecraft/world/entity/ai/goal/PathfindToRaidGoal.java
+++ net/minecraft/world/entity/ai/goal/PathfindToRaidGoal.java
@@ -5,11 +5,11 @@
 import java.util.HashSet;
 import java.util.Iterator;
 import java.util.List;
 import net.minecraft.core.BlockPos;
 import net.minecraft.server.level.ServerLevel;
-import net.minecraft.world.entity.ai.util.RandomPos;
+import net.minecraft.world.entity.ai.util.DefaultRandomPos;
 import net.minecraft.world.entity.raid.Raid;
 import net.minecraft.world.entity.raid.Raider;
 import net.minecraft.world.entity.raid.Raids;
 import net.minecraft.world.phys.Vec3;
 
@@ -35,11 +35,11 @@
          if (this.mob.tickCount % 20 == 0) {
             this.recruitNearby(var1);
          }
 
          if (!this.mob.isPathFinding()) {
-            Vec3 var2 = RandomPos.getPosTowards(this.mob, 15, 4, Vec3.atBottomCenterOf(var1.getCenter()));
+            Vec3 var2 = DefaultRandomPos.getPosTowards(this.mob, 15, 4, Vec3.atBottomCenterOf(var1.getCenter()), 1.5707963705062866D);
             if (var2 != null) {
                this.mob.getNavigation().moveTo(var2.x, var2.y, var2.z, 1.0D);
             }
          }
       }
