--- net/minecraft/world/entity/ai/goal/RangedBowAttackGoal.java
+++ net/minecraft/world/entity/ai/goal/RangedBowAttackGoal.java
@@ -59,11 +59,11 @@
 
    public void tick() {
       LivingEntity var1 = this.mob.getTarget();
       if (var1 != null) {
          double var2 = this.mob.distanceToSqr(var1.getX(), var1.getY(), var1.getZ());
-         boolean var4 = this.mob.getSensing().canSee(var1);
+         boolean var4 = this.mob.getSensing().hasLineOfSight(var1);
          boolean var5 = this.seeTime > 0;
          if (var4 != var5) {
             this.seeTime = 0;
          }
 
