--- net/minecraft/util/thread/ReentrantBlockableEventLoop.java
+++ net/minecraft/util/thread/ReentrantBlockableEventLoop.java
@@ -5,19 +5,19 @@
 
    public ReentrantBlockableEventLoop(String var1) {
       super(var1);
    }
 
-   protected boolean scheduleExecutables() {
+   public boolean scheduleExecutables() {
       return this.runningTask() || super.scheduleExecutables();
    }
 
    protected boolean runningTask() {
       return this.reentrantCount != 0;
    }
 
-   protected void doRunTask(R var1) {
+   public void doRunTask(R var1) {
       ++this.reentrantCount;
 
       try {
          super.doRunTask(var1);
       } finally {
