--- net/minecraft/server/packs/AbstractPackResources.java
+++ net/minecraft/server/packs/AbstractPackResources.java
@@ -40,10 +40,18 @@
       return this.hasResource(getPathFromLocation(var1, var2));
    }
 
    protected abstract InputStream getResource(String var1) throws IOException;
 
+   public InputStream getRootResource(String var1) throws IOException {
+      if (!var1.contains("/") && !var1.contains("\\")) {
+         return this.getResource(var1);
+      } else {
+         throw new IllegalArgumentException("Root resources can only be filenames, not paths (no / allowed!)");
+      }
+   }
+
    protected abstract boolean hasResource(String var1);
 
    protected void logWarning(String var1) {
       LOGGER.warn((String)"ResourcePack: ignored non-lowercase namespace: {} in {}", (Object)var1, (Object)this.file);
    }
