--- net/minecraft/server/packs/FilePackResources.java
+++ net/minecraft/server/packs/FilePackResources.java
@@ -105,11 +105,12 @@
          return Collections.emptySet();
       }
 
       Enumeration var7 = var6.entries();
       ArrayList var8 = Lists.newArrayList();
-      String var9 = var1.getDirectory() + "/" + var2 + "/";
+      String var10000 = var1.getDirectory();
+      String var9 = var10000 + "/" + var2 + "/";
       String var10 = var9 + var3 + "/";
 
       while(var7.hasMoreElements()) {
          ZipEntry var11 = (ZipEntry)var7.nextElement();
          if (!var11.isDirectory()) {
