--- net/minecraft/server/packs/resources/ResourceManager.java
+++ net/minecraft/server/packs/resources/ResourceManager.java
@@ -7,13 +7,11 @@
 import java.util.Collection;
 import java.util.List;
 import java.util.function.Predicate;
 import net.minecraft.resources.ResourceLocation;
 
-public interface ResourceManager {
-   Resource getResource(ResourceLocation var1) throws IOException;
-
+public interface ResourceManager extends ResourceProvider {
    List<Resource> getResources(ResourceLocation var1) throws IOException;
 
    Collection<ResourceLocation> listResources(String var1, Predicate<String> var2);
 
    public static enum Empty implements ResourceManager {
