--- net/minecraft/server/packs/resources/SimpleJsonResourceReloadListener.java
+++ net/minecraft/server/packs/resources/SimpleJsonResourceReloadListener.java
@@ -19,10 +19,11 @@
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
 
 public abstract class SimpleJsonResourceReloadListener extends SimplePreparableReloadListener<Map<ResourceLocation, JsonElement>> {
    private static final Logger LOGGER = LogManager.getLogger();
+   private static final String PATH_SUFFIX = ".json";
    private static final int PATH_SUFFIX_LENGTH = ".json".length();
    private final Gson gson;
    private final String directory;
 
    public SimpleJsonResourceReloadListener(Gson var1, String var2) {
@@ -53,11 +54,11 @@
                try {
                   BufferedReader var13 = new BufferedReader(new InputStreamReader(var11, StandardCharsets.UTF_8));
                   Throwable var14 = null;
 
                   try {
-                     JsonElement var15 = (JsonElement)GsonHelper.fromJson(this.gson, (Reader)var13, JsonElement.class);
+                     JsonElement var15 = (JsonElement)GsonHelper.fromJson(this.gson, (Reader)var13, (Class)JsonElement.class);
                      if (var15 != null) {
                         JsonElement var16 = (JsonElement)var3.put(var8, var15);
                         if (var16 != null) {
                            throw new IllegalStateException("Duplicate data file ignored with ID " + var8);
                         }
