--- net/minecraft/network/CompressionEncoder.java
+++ net/minecraft/network/CompressionEncoder.java
@@ -13,11 +13,11 @@
    public CompressionEncoder(int var1) {
       this.threshold = var1;
       this.deflater = new Deflater();
    }
 
-   protected void encode(ChannelHandlerContext var1, ByteBuf var2, ByteBuf var3) throws Exception {
+   protected void encode(ChannelHandlerContext var1, ByteBuf var2, ByteBuf var3) {
       int var4 = var2.readableBytes();
       FriendlyByteBuf var5 = new FriendlyByteBuf(var3);
       if (var4 < this.threshold) {
          var5.writeVarInt(0);
          var5.writeBytes(var2);
@@ -36,9 +36,13 @@
          this.deflater.reset();
       }
 
    }
 
+   public int getThreshold() {
+      return this.threshold;
+   }
+
    public void setThreshold(int var1) {
       this.threshold = var1;
    }
 }
