--- net/minecraft/network/FriendlyByteBuf.java
+++ net/minecraft/network/FriendlyByteBuf.java
@@ -19,10 +19,11 @@
 import java.nio.channels.FileChannel;
 import java.nio.channels.GatheringByteChannel;
 import java.nio.channels.ScatteringByteChannel;
 import java.nio.charset.Charset;
 import java.nio.charset.StandardCharsets;
+import java.util.BitSet;
 import java.util.Date;
 import java.util.UUID;
 import javax.annotation.Nullable;
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.Direction;
@@ -138,10 +139,14 @@
       }
 
       return this;
    }
 
+   public long[] readLongArray() {
+      return this.readLongArray((long[])null);
+   }
+
    public long[] readLongArray(@Nullable long[] var1) {
       return this.readLongArray(var1, this.readableBytes() / 8);
    }
 
    public long[] readLongArray(@Nullable long[] var1, int var2) {
@@ -397,10 +402,18 @@
       this.writeFloat((float)(var3.y - (double)var2.getY()));
       this.writeFloat((float)(var3.z - (double)var2.getZ()));
       this.writeBoolean(var1.isInside());
    }
 
+   public BitSet readBitSet() {
+      return BitSet.valueOf(this.readLongArray());
+   }
+
+   public void writeBitSet(BitSet var1) {
+      this.writeLongArray(var1.toLongArray());
+   }
+
    public int capacity() {
       return this.source.capacity();
    }
 
    public ByteBuf capacity(int var1) {
