--- net/minecraft/network/PacketEncoder.java
+++ net/minecraft/network/PacketEncoder.java
@@ -35,17 +35,22 @@
          } else {
             FriendlyByteBuf var6 = new FriendlyByteBuf(var3);
             var6.writeVarInt(var5);
 
             try {
+               int var7 = var6.writerIndex();
                var2.write(var6);
-            } catch (Throwable var8) {
-               LOGGER.error(var8);
+               int var8 = var6.writerIndex() - var7;
+               if (var8 > 2097152) {
+                  throw new IllegalArgumentException("Packet too big (is " + var8 + ", should be less than 2097152): " + var2);
+               }
+            } catch (Throwable var9) {
+               LOGGER.error(var9);
                if (var2.isSkippable()) {
-                  throw new SkipPacketException(var8);
+                  throw new SkipPacketException(var9);
                } else {
-                  throw var8;
+                  throw var9;
                }
             }
          }
       }
    }
