--- net/minecraft/data/models/model/ModelTemplate.java
+++ net/minecraft/data/models/model/ModelTemplate.java
@@ -16,11 +16,11 @@
 import net.minecraft.world.level.block.Block;
 
 public class ModelTemplate {
    private final Optional<ResourceLocation> model;
    private final Set<TextureSlot> requiredSlots;
-   private Optional<String> suffix;
+   private final Optional<String> suffix;
 
    public ModelTemplate(Optional<ResourceLocation> var1, Optional<String> var2, TextureSlot... var3) {
       this.model = var1;
       this.suffix = var2;
       this.requiredSlots = ImmutableSet.copyOf((Object[])var3);
