--- net/minecraft/data/models/model/ModelTemplates.java
+++ net/minecraft/data/models/model/ModelTemplates.java
@@ -47,10 +47,11 @@
    public static final ModelTemplate TRAPDOOR_BOTTOM;
    public static final ModelTemplate TRAPDOOR_OPEN;
    public static final ModelTemplate ORIENTABLE_TRAPDOOR_TOP;
    public static final ModelTemplate ORIENTABLE_TRAPDOOR_BOTTOM;
    public static final ModelTemplate ORIENTABLE_TRAPDOOR_OPEN;
+   public static final ModelTemplate POINTED_DRIPSTONE;
    public static final ModelTemplate CROSS;
    public static final ModelTemplate TINTED_CROSS;
    public static final ModelTemplate FLOWER_POT_CROSS;
    public static final ModelTemplate TINTED_FLOWER_POT_CROSS;
    public static final ModelTemplate RAIL_FLAT;
@@ -91,17 +92,26 @@
    public static final ModelTemplate TURTLE_EGG;
    public static final ModelTemplate TWO_TURTLE_EGGS;
    public static final ModelTemplate THREE_TURTLE_EGGS;
    public static final ModelTemplate FOUR_TURTLE_EGGS;
    public static final ModelTemplate SINGLE_FACE;
+   public static final ModelTemplate CAULDRON_LEVEL1;
+   public static final ModelTemplate CAULDRON_LEVEL2;
+   public static final ModelTemplate CAULDRON_FULL;
+   public static final ModelTemplate AZALEA;
    public static final ModelTemplate FLAT_ITEM;
    public static final ModelTemplate FLAT_HANDHELD_ITEM;
    public static final ModelTemplate FLAT_HANDHELD_ROD_ITEM;
    public static final ModelTemplate SHULKER_BOX_INVENTORY;
    public static final ModelTemplate BED_INVENTORY;
    public static final ModelTemplate BANNER_INVENTORY;
    public static final ModelTemplate SKULL_INVENTORY;
+   public static final ModelTemplate CANDLE;
+   public static final ModelTemplate TWO_CANDLES;
+   public static final ModelTemplate THREE_CANDLES;
+   public static final ModelTemplate FOUR_CANDLES;
+   public static final ModelTemplate CANDLE_CAKE;
 
    private static ModelTemplate create(TextureSlot... var0) {
       return new ModelTemplate(Optional.empty(), Optional.empty(), var0);
    }
 
@@ -160,10 +170,11 @@
       TRAPDOOR_BOTTOM = create("template_trapdoor_bottom", "_bottom", TextureSlot.TEXTURE);
       TRAPDOOR_OPEN = create("template_trapdoor_open", "_open", TextureSlot.TEXTURE);
       ORIENTABLE_TRAPDOOR_TOP = create("template_orientable_trapdoor_top", "_top", TextureSlot.TEXTURE);
       ORIENTABLE_TRAPDOOR_BOTTOM = create("template_orientable_trapdoor_bottom", "_bottom", TextureSlot.TEXTURE);
       ORIENTABLE_TRAPDOOR_OPEN = create("template_orientable_trapdoor_open", "_open", TextureSlot.TEXTURE);
+      POINTED_DRIPSTONE = create("pointed_dripstone", TextureSlot.CROSS);
       CROSS = create("cross", TextureSlot.CROSS);
       TINTED_CROSS = create("tinted_cross", TextureSlot.CROSS);
       FLOWER_POT_CROSS = create("flower_pot_cross", TextureSlot.PLANT);
       TINTED_FLOWER_POT_CROSS = create("tinted_flower_pot_cross", TextureSlot.PLANT);
       RAIL_FLAT = create("rail_flat", TextureSlot.RAIL);
@@ -208,14 +219,23 @@
       TURTLE_EGG = create("template_turtle_egg", TextureSlot.ALL);
       TWO_TURTLE_EGGS = create("template_two_turtle_eggs", TextureSlot.ALL);
       THREE_TURTLE_EGGS = create("template_three_turtle_eggs", TextureSlot.ALL);
       FOUR_TURTLE_EGGS = create("template_four_turtle_eggs", TextureSlot.ALL);
       SINGLE_FACE = create("template_single_face", TextureSlot.TEXTURE);
+      CAULDRON_LEVEL1 = create("template_cauldron_level1", TextureSlot.CONTENT, TextureSlot.INSIDE, TextureSlot.PARTICLE, TextureSlot.TOP, TextureSlot.BOTTOM, TextureSlot.SIDE);
+      CAULDRON_LEVEL2 = create("template_cauldron_level2", TextureSlot.CONTENT, TextureSlot.INSIDE, TextureSlot.PARTICLE, TextureSlot.TOP, TextureSlot.BOTTOM, TextureSlot.SIDE);
+      CAULDRON_FULL = create("template_cauldron_full", TextureSlot.CONTENT, TextureSlot.INSIDE, TextureSlot.PARTICLE, TextureSlot.TOP, TextureSlot.BOTTOM, TextureSlot.SIDE);
+      AZALEA = create("template_azalea", TextureSlot.TOP, TextureSlot.SIDE);
       FLAT_ITEM = createItem("generated", TextureSlot.LAYER0);
       FLAT_HANDHELD_ITEM = createItem("handheld", TextureSlot.LAYER0);
       FLAT_HANDHELD_ROD_ITEM = createItem("handheld_rod", TextureSlot.LAYER0);
       SHULKER_BOX_INVENTORY = createItem("template_shulker_box", TextureSlot.PARTICLE);
       BED_INVENTORY = createItem("template_bed", TextureSlot.PARTICLE);
       BANNER_INVENTORY = createItem("template_banner");
       SKULL_INVENTORY = createItem("template_skull");
+      CANDLE = create("template_candle", TextureSlot.ALL, TextureSlot.PARTICLE);
+      TWO_CANDLES = create("template_two_candles", TextureSlot.ALL, TextureSlot.PARTICLE);
+      THREE_CANDLES = create("template_three_candles", TextureSlot.ALL, TextureSlot.PARTICLE);
+      FOUR_CANDLES = create("template_four_candles", TextureSlot.ALL, TextureSlot.PARTICLE);
+      CANDLE_CAKE = create("template_cake_with_candle", TextureSlot.CANDLE, TextureSlot.BOTTOM, TextureSlot.SIDE, TextureSlot.TOP, TextureSlot.PARTICLE);
    }
 }
