--- net/minecraft/data/models/model/TextureSlot.java
+++ net/minecraft/data/models/model/TextureSlot.java
@@ -36,10 +36,13 @@
    public static final TextureSlot PLATFORM;
    public static final TextureSlot UNSTICKY;
    public static final TextureSlot TORCH;
    public static final TextureSlot LAYER0;
    public static final TextureSlot LIT_LOG;
+   public static final TextureSlot CANDLE;
+   public static final TextureSlot INSIDE;
+   public static final TextureSlot CONTENT;
    private final String id;
    @Nullable
    private final TextureSlot parent;
 
    private static TextureSlot create(String var0) {
@@ -101,7 +104,10 @@
       PLATFORM = create("platform");
       UNSTICKY = create("unsticky");
       TORCH = create("torch");
       LAYER0 = create("layer0");
       LIT_LOG = create("lit_log");
+      CANDLE = create("candle");
+      INSIDE = create("inside");
+      CONTENT = create("content");
    }
 }
