--- net/minecraft/data/models/model/TexturedModel.java
+++ net/minecraft/data/models/model/TexturedModel.java
@@ -102,13 +102,7 @@
       }
 
       default ResourceLocation createWithSuffix(Block var1, String var2, BiConsumer<ResourceLocation, Supplier<JsonElement>> var3) {
          return this.get(var1).createWithSuffix(var1, var2, var3);
       }
-
-      default TexturedModel.Provider updateTexture(Consumer<TextureMapping> var1) {
-         return (var2) -> {
-            return this.get(var2).updateTextures(var1);
-         };
-      }
    }
 }
