--- net/minecraft/client/model/geom/builders/MaterialDefinition.java
+++ net/minecraft/client/model/geom/builders/MaterialDefinition.java
@@ -1,0 +1,11 @@
+package net.minecraft.client.model.geom.builders;
+
+public class MaterialDefinition {
+   final int xTexSize;
+   final int yTexSize;
+
+   public MaterialDefinition(int var1, int var2) {
+      this.xTexSize = var1;
+      this.yTexSize = var2;
+   }
+}
