--- net/minecraft/client/model/geom/builders/UVPair.java
+++ net/minecraft/client/model/geom/builders/UVPair.java
@@ -1,0 +1,23 @@
+package net.minecraft.client.model.geom.builders;
+
+public class UVPair {
+   private final float u;
+   private final float v;
+
+   public UVPair(float var1, float var2) {
+      this.u = var1;
+      this.v = var2;
+   }
+
+   public float u() {
+      return this.u;
+   }
+
+   public float v() {
+      return this.v;
+   }
+
+   public String toString() {
+      return "(" + this.u + "," + this.v + ")";
+   }
+}
