--- net/minecraft/client/gui/screens/inventory/AnvilScreen.java
+++ net/minecraft/client/gui/screens/inventory/AnvilScreen.java
@@ -9,22 +9,25 @@
 import net.minecraft.network.chat.TranslatableComponent;
 import net.minecraft.network.protocol.Packet;
 import net.minecraft.network.protocol.game.ServerboundRenameItemPacket;
 import net.minecraft.resources.ResourceLocation;
 import net.minecraft.world.entity.player.Inventory;
+import net.minecraft.world.entity.player.Player;
 import net.minecraft.world.inventory.AbstractContainerMenu;
 import net.minecraft.world.inventory.AnvilMenu;
 import net.minecraft.world.inventory.Slot;
 import net.minecraft.world.item.ItemStack;
 
 public class AnvilScreen extends ItemCombinerScreen<AnvilMenu> {
    private static final ResourceLocation ANVIL_LOCATION = new ResourceLocation("textures/gui/container/anvil.png");
    private static final Component TOO_EXPENSIVE_TEXT = new TranslatableComponent("container.repair.expensive");
    private EditBox name;
+   private final Player player;
 
    public AnvilScreen(AnvilMenu var1, Inventory var2, Component var3) {
       super(var1, var2, var3, ANVIL_LOCATION);
+      this.player = var2.player;
       this.titleLabelX = 60;
    }
 
    public void tick() {
       super.tick();
@@ -40,12 +43,14 @@
       this.name.setTextColor(-1);
       this.name.setTextColorUneditable(-1);
       this.name.setBordered(false);
       this.name.setMaxLength(35);
       this.name.setResponder(this::onNameChanged);
+      this.name.setValue("");
       this.children.add(this.name);
       this.setInitialFocus(this.name);
+      this.name.setEditable(false);
    }
 
    public void resize(Minecraft var1, int var2, int var3) {
       String var4 = this.name.getValue();
       this.init(var1, var2, var3);
@@ -83,18 +88,18 @@
       super.renderLabels(var1, var2, var3);
       int var4 = ((AnvilMenu)this.menu).getCost();
       if (var4 > 0) {
          int var5 = 8453920;
          Object var6;
-         if (var4 >= 40 && !this.minecraft.player.abilities.instabuild) {
+         if (var4 >= 40 && !this.minecraft.player.getAbilities().instabuild) {
             var6 = TOO_EXPENSIVE_TEXT;
             var5 = 16736352;
          } else if (!((AnvilMenu)this.menu).getSlot(2).hasItem()) {
             var6 = null;
          } else {
             var6 = new TranslatableComponent("container.repair.cost", new Object[]{var4});
-            if (!((AnvilMenu)this.menu).getSlot(2).mayPickup(this.inventory.player)) {
+            if (!((AnvilMenu)this.menu).getSlot(2).mayPickup(this.player)) {
                var5 = 16736352;
             }
          }
 
          if (var6 != null) {
