--- net/minecraft/client/gui/screens/inventory/AnvilScreen.java
+++ net/minecraft/client/gui/screens/inventory/AnvilScreen.java
@@ -83,11 +83,11 @@
       super.renderLabels(var1, var2, var3);
       int var4 = ((AnvilMenu)this.menu).getCost();
       if (var4 > 0) {
          int var5 = 8453920;
          Object var6;
-         if (var4 >= 40 && !this.minecraft.player.abilities.instabuild) {
+         if (var4 >= 40 && !this.minecraft.player.getAbilities().instabuild) {
             var6 = TOO_EXPENSIVE_TEXT;
             var5 = 16736352;
          } else if (!((AnvilMenu)this.menu).getSlot(2).hasItem()) {
             var6 = null;
          } else {
