--- net/minecraft/client/gui/screens/inventory/BookEditScreen.java
+++ net/minecraft/client/gui/screens/inventory/BookEditScreen.java
@@ -5,10 +5,11 @@
 import com.mojang.blaze3d.systems.RenderSystem;
 import com.mojang.blaze3d.vertex.BufferBuilder;
 import com.mojang.blaze3d.vertex.DefaultVertexFormat;
 import com.mojang.blaze3d.vertex.PoseStack;
 import com.mojang.blaze3d.vertex.Tesselator;
+import com.mojang.blaze3d.vertex.VertexFormat;
 import it.unimi.dsi.fastutil.ints.IntArrayList;
 import java.util.ArrayList;
 import java.util.Arrays;
 import java.util.List;
 import java.util.ListIterator;
@@ -88,15 +89,12 @@
       this.owner = var1;
       this.book = var2;
       this.hand = var3;
       CompoundTag var4 = var2.getTag();
       if (var4 != null) {
-         ListTag var5 = var4.getList("pages", 8).copy();
-
-         for(int var6 = 0; var6 < var5.size(); ++var6) {
-            this.pages.add(var5.getString(var6));
-         }
+         List var10001 = this.pages;
+         BookViewScreen.loadPages(var4, var10001::add);
       }
 
       if (this.pages.isEmpty()) {
          this.pages.add("");
       }
@@ -218,11 +216,11 @@
          if (var1) {
             this.book.addTagElement("author", StringTag.valueOf(this.owner.getGameProfile().getName()));
             this.book.addTagElement("title", StringTag.valueOf(this.title.trim()));
          }
 
-         int var3 = this.hand == InteractionHand.MAIN_HAND ? this.owner.inventory.selected : 40;
+         int var3 = this.hand == InteractionHand.MAIN_HAND ? this.owner.getInventory().selected : 40;
          this.minecraft.getConnection().send((Packet)(new ServerboundEditBookPacket(this.book, var1, var3)));
       }
    }
 
    private void appendPageToBook() {
@@ -445,11 +443,11 @@
       BufferBuilder var3 = var2.getBuilder();
       RenderSystem.color4f(0.0F, 0.0F, 255.0F, 255.0F);
       RenderSystem.disableTexture();
       RenderSystem.enableColorLogicOp();
       RenderSystem.logicOp(GlStateManager.LogicOp.OR_REVERSE);
-      var3.begin(7, DefaultVertexFormat.POSITION);
+      var3.begin(VertexFormat.Mode.QUADS, DefaultVertexFormat.POSITION);
       Rect2i[] var4 = var1;
       int var5 = var1.length;
 
       for(int var6 = 0; var6 < var5; ++var6) {
          Rect2i var7 = var4[var6];
