--- net/minecraft/client/gui/screens/inventory/CraftingScreen.java
+++ net/minecraft/client/gui/screens/inventory/CraftingScreen.java
@@ -3,10 +3,11 @@
 import com.mojang.blaze3d.systems.RenderSystem;
 import com.mojang.blaze3d.vertex.PoseStack;
 import net.minecraft.client.gui.components.ImageButton;
 import net.minecraft.client.gui.screens.recipebook.RecipeBookComponent;
 import net.minecraft.client.gui.screens.recipebook.RecipeUpdateListener;
+import net.minecraft.client.renderer.GameRenderer;
 import net.minecraft.network.chat.Component;
 import net.minecraft.resources.ResourceLocation;
 import net.minecraft.world.entity.player.Inventory;
 import net.minecraft.world.inventory.ClickType;
 import net.minecraft.world.inventory.CraftingMenu;
@@ -26,18 +27,18 @@
    protected void init() {
       super.init();
       this.widthTooNarrow = this.width < 379;
       this.recipeBookComponent.init(this.width, this.height, this.minecraft, this.widthTooNarrow, (RecipeBookMenu)this.menu);
       this.leftPos = this.recipeBookComponent.updateScreenPosition(this.widthTooNarrow, this.width, this.imageWidth);
-      this.children.add(this.recipeBookComponent);
-      this.setInitialFocus(this.recipeBookComponent);
-      this.addButton(new ImageButton(this.leftPos + 5, this.height / 2 - 49, 20, 18, 0, 0, 19, RECIPE_BUTTON_LOCATION, (var1) -> {
+      this.addRenderableWidget(new ImageButton(this.leftPos + 5, this.height / 2 - 49, 20, 18, 0, 0, 19, RECIPE_BUTTON_LOCATION, (var1) -> {
          this.recipeBookComponent.initVisuals(this.widthTooNarrow);
          this.recipeBookComponent.toggleVisibility();
          this.leftPos = this.recipeBookComponent.updateScreenPosition(this.widthTooNarrow, this.width, this.imageWidth);
          ((ImageButton)var1).setPosition(this.leftPos + 5, this.height / 2 - 49);
       }));
+      this.addWidget(this.recipeBookComponent);
+      this.setInitialFocus(this.recipeBookComponent);
       this.titleLabelX = 29;
    }
 
    public void tick() {
       super.tick();
@@ -58,12 +59,13 @@
       this.renderTooltip(var1, var2, var3);
       this.recipeBookComponent.renderTooltip(var1, this.leftPos, this.topPos, var2, var3);
    }
 
    protected void renderBg(PoseStack var1, float var2, int var3, int var4) {
-      RenderSystem.color4f(1.0F, 1.0F, 1.0F, 1.0F);
-      this.minecraft.getTextureManager().bind(CRAFTING_TABLE_LOCATION);
+      RenderSystem.setShader(GameRenderer::getPositionTexShader);
+      RenderSystem.setShaderColor(1.0F, 1.0F, 1.0F, 1.0F);
+      RenderSystem.setShaderTexture(0, CRAFTING_TABLE_LOCATION);
       int var5 = this.leftPos;
       int var6 = (this.height - this.imageHeight) / 2;
       this.blit(var1, var5, var6, 0, 0, this.imageWidth, this.imageHeight);
    }
 
