--- net/minecraft/client/gui/screens/inventory/CreativeInventoryListener.java
+++ net/minecraft/client/gui/screens/inventory/CreativeInventoryListener.java
@@ -1,9 +1,8 @@
 package net.minecraft.client.gui.screens.inventory;
 
 import net.minecraft.client.Minecraft;
-import net.minecraft.core.NonNullList;
 import net.minecraft.world.inventory.AbstractContainerMenu;
 import net.minecraft.world.inventory.ContainerListener;
 import net.minecraft.world.item.ItemStack;
 
 public class CreativeInventoryListener implements ContainerListener {
@@ -11,15 +10,12 @@
 
    public CreativeInventoryListener(Minecraft var1) {
       this.minecraft = var1;
    }
 
-   public void refreshContainer(AbstractContainerMenu var1, NonNullList<ItemStack> var2) {
-   }
-
    public void slotChanged(AbstractContainerMenu var1, int var2, ItemStack var3) {
       this.minecraft.gameMode.handleCreativeModeItemAdd(var3, var2);
    }
 
-   public void setContainerData(AbstractContainerMenu var1, int var2, int var3) {
+   public void dataChanged(AbstractContainerMenu var1, int var2, int var3) {
    }
 }
