--- net/minecraft/client/gui/screens/inventory/EffectRenderingInventoryScreen.java
+++ net/minecraft/client/gui/screens/inventory/EffectRenderingInventoryScreen.java
@@ -49,11 +49,11 @@
 
    private void renderEffects(PoseStack var1) {
       int var2 = this.leftPos - 124;
       Collection var3 = this.minecraft.player.getActiveEffects();
       if (!var3.isEmpty()) {
-         RenderSystem.color4f(1.0F, 1.0F, 1.0F, 1.0F);
+         RenderSystem.setShaderColor(1.0F, 1.0F, 1.0F, 1.0F);
          int var4 = 33;
          if (var3.size() > 5) {
             var4 = 132 / (var3.size() - 1);
          }
 
@@ -63,16 +63,16 @@
          this.renderLabels(var1, var2, var4, var5);
       }
    }
 
    private void renderBackgrounds(PoseStack var1, int var2, int var3, Iterable<MobEffectInstance> var4) {
-      this.minecraft.getTextureManager().bind(INVENTORY_LOCATION);
+      RenderSystem.setShaderTexture(0, INVENTORY_LOCATION);
       int var5 = this.topPos;
 
       for(Iterator var6 = var4.iterator(); var6.hasNext(); var5 += var3) {
          MobEffectInstance var7 = (MobEffectInstance)var6.next();
-         RenderSystem.color4f(1.0F, 1.0F, 1.0F, 1.0F);
+         RenderSystem.setShaderColor(1.0F, 1.0F, 1.0F, 1.0F);
          this.blit(var1, var2, var5, 0, 166, 140, 32);
       }
 
    }
 
@@ -82,11 +82,11 @@
 
       for(Iterator var7 = var4.iterator(); var7.hasNext(); var6 += var3) {
          MobEffectInstance var8 = (MobEffectInstance)var7.next();
          MobEffect var9 = var8.getEffect();
          TextureAtlasSprite var10 = var5.get(var9);
-         this.minecraft.getTextureManager().bind(var10.atlas().location());
+         RenderSystem.setShaderTexture(0, var10.atlas().location());
          blit(var1, var2 + 6, var6 + 7, this.getBlitOffset(), 18, 18, var10);
       }
 
    }
 
@@ -95,11 +95,11 @@
 
       for(Iterator var6 = var4.iterator(); var6.hasNext(); var5 += var3) {
          MobEffectInstance var7 = (MobEffectInstance)var6.next();
          String var8 = I18n.get(var7.getEffect().getDescriptionId());
          if (var7.getAmplifier() >= 1 && var7.getAmplifier() <= 9) {
-            var8 = var8 + ' ' + I18n.get("enchantment.level." + (var7.getAmplifier() + 1));
+            var8 = var8 + " " + I18n.get("enchantment.level." + (var7.getAmplifier() + 1));
          }
 
          this.font.drawShadow(var1, var8, (float)(var2 + 10 + 18), (float)(var5 + 6), 16777215);
          String var9 = MobEffectUtil.formatDuration(var7, 1.0F);
          this.font.drawShadow(var1, var9, (float)(var2 + 10 + 18), (float)(var5 + 6 + 10), 8355711);
