--- net/minecraft/client/gui/screens/inventory/InventoryScreen.java
+++ net/minecraft/client/gui/screens/inventory/InventoryScreen.java
@@ -27,11 +27,11 @@
    private boolean recipeBookComponentInitialized;
    private boolean widthTooNarrow;
    private boolean buttonClicked;
 
    public InventoryScreen(Player var1) {
-      super(var1.inventoryMenu, var1.inventory, new TranslatableComponent("container.crafting"));
+      super(var1.inventoryMenu, var1.getInventory(), new TranslatableComponent("container.crafting"));
       this.passEvents = true;
       this.titleLabelX = 97;
    }
 
    public void tick() {
