--- net/minecraft/client/gui/screens/inventory/InventoryScreen.java
+++ net/minecraft/client/gui/screens/inventory/InventoryScreen.java
@@ -1,15 +1,17 @@
 package net.minecraft.client.gui.screens.inventory;
 
+import com.mojang.blaze3d.platform.Lighting;
 import com.mojang.blaze3d.systems.RenderSystem;
 import com.mojang.blaze3d.vertex.PoseStack;
 import com.mojang.math.Quaternion;
 import com.mojang.math.Vector3f;
 import net.minecraft.client.Minecraft;
 import net.minecraft.client.gui.components.ImageButton;
 import net.minecraft.client.gui.screens.recipebook.RecipeBookComponent;
 import net.minecraft.client.gui.screens.recipebook.RecipeUpdateListener;
+import net.minecraft.client.renderer.GameRenderer;
 import net.minecraft.client.renderer.MultiBufferSource;
 import net.minecraft.client.renderer.entity.EntityRenderDispatcher;
 import net.minecraft.network.chat.TranslatableComponent;
 import net.minecraft.resources.ResourceLocation;
 import net.minecraft.world.entity.LivingEntity;
@@ -27,11 +29,11 @@
    private boolean recipeBookComponentInitialized;
    private boolean widthTooNarrow;
    private boolean buttonClicked;
 
    public InventoryScreen(Player var1) {
-      super(var1.inventoryMenu, var1.inventory, new TranslatableComponent("container.crafting"));
+      super(var1.inventoryMenu, var1.getInventory(), new TranslatableComponent("container.crafting"));
       this.passEvents = true;
       this.titleLabelX = 97;
    }
 
    public void tick() {
@@ -84,57 +86,63 @@
       this.xMouse = (float)var2;
       this.yMouse = (float)var3;
    }
 
    protected void renderBg(PoseStack var1, float var2, int var3, int var4) {
-      RenderSystem.color4f(1.0F, 1.0F, 1.0F, 1.0F);
-      this.minecraft.getTextureManager().bind(INVENTORY_LOCATION);
+      RenderSystem.setShader(GameRenderer::getPositionTexShader);
+      RenderSystem.setShaderColor(1.0F, 1.0F, 1.0F, 1.0F);
+      RenderSystem.setShaderTexture(0, INVENTORY_LOCATION);
       int var5 = this.leftPos;
       int var6 = this.topPos;
       this.blit(var1, var5, var6, 0, 0, this.imageWidth, this.imageHeight);
       renderEntityInInventory(var5 + 51, var6 + 75, 30, (float)(var5 + 51) - this.xMouse, (float)(var6 + 75 - 50) - this.yMouse, this.minecraft.player);
    }
 
    public static void renderEntityInInventory(int var0, int var1, int var2, float var3, float var4, LivingEntity var5) {
       float var6 = (float)Math.atan((double)(var3 / 40.0F));
       float var7 = (float)Math.atan((double)(var4 / 40.0F));
-      RenderSystem.pushMatrix();
-      RenderSystem.translatef((float)var0, (float)var1, 1050.0F);
-      RenderSystem.scalef(1.0F, 1.0F, -1.0F);
-      PoseStack var8 = new PoseStack();
-      var8.translate(0.0D, 0.0D, 1000.0D);
-      var8.scale((float)var2, (float)var2, (float)var2);
-      Quaternion var9 = Vector3f.ZP.rotationDegrees(180.0F);
-      Quaternion var10 = Vector3f.XP.rotationDegrees(var7 * 20.0F);
-      var9.mul(var10);
-      var8.mulPose(var9);
-      float var11 = var5.yBodyRot;
-      float var12 = var5.yRot;
-      float var13 = var5.xRot;
-      float var14 = var5.yHeadRotO;
-      float var15 = var5.yHeadRot;
+      PoseStack var8 = RenderSystem.getModelViewStack();
+      var8.pushPose();
+      var8.translate((double)var0, (double)var1, 1050.0D);
+      var8.scale(1.0F, 1.0F, -1.0F);
+      RenderSystem.applyModelViewMatrix();
+      PoseStack var9 = new PoseStack();
+      var9.translate(0.0D, 0.0D, 1000.0D);
+      var9.scale((float)var2, (float)var2, (float)var2);
+      Quaternion var10 = Vector3f.ZP.rotationDegrees(180.0F);
+      Quaternion var11 = Vector3f.XP.rotationDegrees(var7 * 20.0F);
+      var10.mul(var11);
+      var9.mulPose(var10);
+      float var12 = var5.yBodyRot;
+      float var13 = var5.yRot;
+      float var14 = var5.xRot;
+      float var15 = var5.yHeadRotO;
+      float var16 = var5.yHeadRot;
       var5.yBodyRot = 180.0F + var6 * 20.0F;
       var5.yRot = 180.0F + var6 * 40.0F;
       var5.xRot = -var7 * 20.0F;
       var5.yHeadRot = var5.yRot;
       var5.yHeadRotO = var5.yRot;
-      EntityRenderDispatcher var16 = Minecraft.getInstance().getEntityRenderDispatcher();
-      var10.conj();
-      var16.overrideCameraOrientation(var10);
-      var16.setRenderShadow(false);
-      MultiBufferSource.BufferSource var17 = Minecraft.getInstance().renderBuffers().bufferSource();
+      Lighting.setupForEntityInInventory();
+      EntityRenderDispatcher var17 = Minecraft.getInstance().getEntityRenderDispatcher();
+      var11.conj();
+      var17.overrideCameraOrientation(var11);
+      var17.setRenderShadow(false);
+      MultiBufferSource.BufferSource var18 = Minecraft.getInstance().renderBuffers().bufferSource();
       RenderSystem.runAsFancy(() -> {
-         var16.render(var5, 0.0D, 0.0D, 0.0D, 0.0F, 1.0F, var8, var17, 15728880);
+         var17.render(var5, 0.0D, 0.0D, 0.0D, 0.0F, 1.0F, var9, var18, 15728880);
       });
-      var17.endBatch();
-      var16.setRenderShadow(true);
-      var5.yBodyRot = var11;
-      var5.yRot = var12;
-      var5.xRot = var13;
-      var5.yHeadRotO = var14;
-      var5.yHeadRot = var15;
-      RenderSystem.popMatrix();
+      var18.endBatch();
+      var17.setRenderShadow(true);
+      var5.yBodyRot = var12;
+      var5.yRot = var13;
+      var5.xRot = var14;
+      var5.yHeadRotO = var15;
+      var5.yHeadRot = var16;
+      var8.popPose();
+      RenderSystem.applyModelViewMatrix();
+      Lighting.setupFor3DItems();
    }
 
    protected boolean isHovering(int var1, int var2, int var3, int var4, double var5, double var7) {
       return (!this.widthTooNarrow || !this.recipeBookComponent.isVisible()) && super.isHovering(var1, var2, var3, var4, var5, var7);
    }
