--- net/minecraft/client/gui/screens/inventory/ItemCombinerScreen.java
+++ net/minecraft/client/gui/screens/inventory/ItemCombinerScreen.java
@@ -10,11 +10,11 @@
 import net.minecraft.world.inventory.ContainerListener;
 import net.minecraft.world.inventory.ItemCombinerMenu;
 import net.minecraft.world.item.ItemStack;
 
 public class ItemCombinerScreen<T extends ItemCombinerMenu> extends AbstractContainerScreen<T> implements ContainerListener {
-   private ResourceLocation menuResource;
+   private final ResourceLocation menuResource;
 
    public ItemCombinerScreen(T var1, Inventory var2, Component var3, ResourceLocation var4) {
       super(var1, var2, var3);
       this.menuResource = var4;
    }
