--- net/minecraft/client/gui/screens/inventory/JigsawBlockEditScreen.java
+++ net/minecraft/client/gui/screens/inventory/JigsawBlockEditScreen.java
@@ -3,10 +3,11 @@
 import com.mojang.blaze3d.vertex.PoseStack;
 import net.minecraft.client.Minecraft;
 import net.minecraft.client.gui.chat.NarratorChatListener;
 import net.minecraft.client.gui.components.AbstractSliderButton;
 import net.minecraft.client.gui.components.Button;
+import net.minecraft.client.gui.components.CycleButton;
 import net.minecraft.client.gui.components.EditBox;
 import net.minecraft.client.gui.screens.Screen;
 import net.minecraft.network.chat.CommonComponents;
 import net.minecraft.network.chat.Component;
 import net.minecraft.network.chat.FormattedText;
@@ -31,11 +32,11 @@
    private EditBox targetEdit;
    private EditBox poolEdit;
    private EditBox finalStateEdit;
    private int levels;
    private boolean keepJigsaws = true;
-   private Button jointButton;
+   private CycleButton<JigsawBlockEntity.JointType> jointButton;
    private Button doneButton;
    private JigsawBlockEntity.JointType joint;
 
    public JigsawBlockEditScreen(JigsawBlockEntity var1) {
       super(NarratorChatListener.NO_TITLE);
@@ -97,15 +98,12 @@
       this.finalStateEdit.setMaxLength(256);
       this.finalStateEdit.setValue(this.jigsawEntity.getFinalState());
       this.children.add(this.finalStateEdit);
       this.joint = this.jigsawEntity.getJoint();
       int var1 = this.font.width((FormattedText)JOINT_LABEL) + 10;
-      this.jointButton = (Button)this.addButton(new Button(this.width / 2 - 152 + var1, 150, 300 - var1, 20, this.getJointText(), (var1x) -> {
-         JigsawBlockEntity.JointType[] var2 = JigsawBlockEntity.JointType.values();
-         int var3 = (this.joint.ordinal() + 1) % var2.length;
-         this.joint = var2[var3];
-         var1x.setMessage(this.getJointText());
+      this.jointButton = (CycleButton)this.addButton(CycleButton.builder(JigsawBlockEntity.JointType::getTranslatedName).withValues((Object[])JigsawBlockEntity.JointType.values()).withInitialValue(this.joint).displayOnlyValue().create(this.width / 2 - 152 + var1, 150, 300 - var1, 20, JOINT_LABEL, (var1x, var2x) -> {
+         this.joint = var2x;
       }));
       boolean var2 = JigsawBlock.getFrontFacing(this.jigsawEntity.getBlockState()).getAxis().isVertical();
       this.jointButton.active = var2;
       this.jointButton.visible = var2;
       this.addButton(new AbstractSliderButton(this.width / 2 - 154, 180, 100, 20, TextComponent.EMPTY, 0.0D) {
@@ -119,18 +117,13 @@
 
          protected void applyValue() {
             JigsawBlockEditScreen.this.levels = Mth.floor(Mth.clampedLerp(0.0D, 7.0D, this.value));
          }
       });
-      this.addButton(new Button(this.width / 2 - 50, 180, 100, 20, new TranslatableComponent("jigsaw_block.keep_jigsaws"), (var1x) -> {
-         this.keepJigsaws = !this.keepJigsaws;
-         var1x.queueNarration(250);
-      }) {
-         public Component getMessage() {
-            return CommonComponents.optionStatus(super.getMessage(), JigsawBlockEditScreen.this.keepJigsaws);
-         }
-      });
+      this.addButton(CycleButton.onOffBuilder(this.keepJigsaws).create(this.width / 2 - 50, 180, 100, 20, new TranslatableComponent("jigsaw_block.keep_jigsaws"), (var1x, var2x) -> {
+         this.keepJigsaws = var2x;
+      }));
       this.addButton(new Button(this.width / 2 + 54, 180, 100, 20, new TranslatableComponent("jigsaw_block.generate"), (var1x) -> {
          this.onDone();
          this.sendGenerate();
       }));
       this.doneButton = (Button)this.addButton(new Button(this.width / 2 - 4 - 150, 210, 150, 20, CommonComponents.GUI_DONE, (var1x) -> {
@@ -159,17 +152,13 @@
       this.targetEdit.setValue(var5);
       this.poolEdit.setValue(var6);
       this.finalStateEdit.setValue(var7);
       this.levels = var8;
       this.joint = var9;
-      this.jointButton.setMessage(this.getJointText());
+      this.jointButton.setValue(var9);
    }
 
-   private Component getJointText() {
-      return new TranslatableComponent("jigsaw_block.joint." + this.joint.getSerializedName());
-   }
-
    public void removed() {
       this.minecraft.keyboardHandler.setSendRepeatsToGui(false);
    }
 
    public boolean keyPressed(int var1, int var2, int var3) {
