--- net/minecraft/client/gui/screens/inventory/LecternScreen.java
+++ net/minecraft/client/gui/screens/inventory/LecternScreen.java
@@ -1,10 +1,9 @@
 package net.minecraft.client.gui.screens.inventory;
 
 import net.minecraft.client.gui.components.Button;
 import net.minecraft.client.gui.screens.Screen;
-import net.minecraft.core.NonNullList;
 import net.minecraft.network.chat.CommonComponents;
 import net.minecraft.network.chat.Component;
 import net.minecraft.network.chat.TranslatableComponent;
 import net.minecraft.world.entity.player.Inventory;
 import net.minecraft.world.inventory.AbstractContainerMenu;
@@ -13,19 +12,15 @@
 import net.minecraft.world.item.ItemStack;
 
 public class LecternScreen extends BookViewScreen implements MenuAccess<LecternMenu> {
    private final LecternMenu menu;
    private final ContainerListener listener = new ContainerListener() {
-      public void refreshContainer(AbstractContainerMenu var1, NonNullList<ItemStack> var2) {
-         LecternScreen.this.bookChanged();
-      }
-
       public void slotChanged(AbstractContainerMenu var1, int var2, ItemStack var3) {
          LecternScreen.this.bookChanged();
       }
 
-      public void setContainerData(AbstractContainerMenu var1, int var2, int var3) {
+      public void dataChanged(AbstractContainerMenu var1, int var2, int var3) {
          if (var2 == 0) {
             LecternScreen.this.pageChanged();
          }
 
       }
@@ -99,6 +94,10 @@
    }
 
    private void pageChanged() {
       this.setPage(this.menu.getPage());
    }
+
+   protected void closeScreen() {
+      this.minecraft.player.closeContainer();
+   }
 }
