--- net/minecraft/client/gui/screens/inventory/LecternScreen.java
+++ net/minecraft/client/gui/screens/inventory/LecternScreen.java
@@ -1,10 +1,8 @@
 package net.minecraft.client.gui.screens.inventory;
 
 import net.minecraft.client.gui.components.Button;
-import net.minecraft.client.gui.screens.Screen;
-import net.minecraft.core.NonNullList;
 import net.minecraft.network.chat.CommonComponents;
 import net.minecraft.network.chat.Component;
 import net.minecraft.network.chat.TranslatableComponent;
 import net.minecraft.world.entity.player.Inventory;
 import net.minecraft.world.inventory.AbstractContainerMenu;
@@ -13,19 +11,15 @@
 import net.minecraft.world.item.ItemStack;
 
 public class LecternScreen extends BookViewScreen implements MenuAccess<LecternMenu> {
    private final LecternMenu menu;
    private final ContainerListener listener = new ContainerListener() {
-      public void refreshContainer(AbstractContainerMenu var1, NonNullList<ItemStack> var2) {
-         LecternScreen.this.bookChanged();
-      }
-
       public void slotChanged(AbstractContainerMenu var1, int var2, ItemStack var3) {
          LecternScreen.this.bookChanged();
       }
 
-      public void setContainerData(AbstractContainerMenu var1, int var2, int var3) {
+      public void dataChanged(AbstractContainerMenu var1, int var2, int var3) {
          if (var2 == 0) {
             LecternScreen.this.pageChanged();
          }
 
       }
@@ -54,14 +48,14 @@
       this.menu.removeSlotListener(this.listener);
    }
 
    protected void createMenuControls() {
       if (this.minecraft.player.mayBuild()) {
-         this.addButton(new Button(this.width / 2 - 100, 196, 98, 20, CommonComponents.GUI_DONE, (var1) -> {
-            this.minecraft.setScreen((Screen)null);
+         this.addRenderableWidget(new Button(this.width / 2 - 100, 196, 98, 20, CommonComponents.GUI_DONE, (var1) -> {
+            this.onClose();
          }));
-         this.addButton(new Button(this.width / 2 + 2, 196, 98, 20, new TranslatableComponent("lectern.take_book"), (var1) -> {
+         this.addRenderableWidget(new Button(this.width / 2 + 2, 196, 98, 20, new TranslatableComponent("lectern.take_book"), (var1) -> {
             this.sendButtonClick(3);
          }));
       } else {
          super.createMenuControls();
       }
@@ -91,14 +85,18 @@
 
    public boolean isPauseScreen() {
       return false;
    }
 
-   private void bookChanged() {
+   void bookChanged() {
       ItemStack var1 = this.menu.getBook();
       this.setBookAccess(BookViewScreen.BookAccess.fromItem(var1));
    }
 
-   private void pageChanged() {
+   void pageChanged() {
       this.setPage(this.menu.getPage());
    }
+
+   protected void closeScreen() {
+      this.minecraft.player.closeContainer();
+   }
 }
