--- net/minecraft/client/gui/screens/inventory/LoomScreen.java
+++ net/minecraft/client/gui/screens/inventory/LoomScreen.java
@@ -4,10 +4,11 @@
 import com.mojang.blaze3d.vertex.PoseStack;
 import com.mojang.datafixers.util.Pair;
 import java.util.List;
 import javax.annotation.Nullable;
 import net.minecraft.client.Minecraft;
+import net.minecraft.client.model.geom.ModelLayers;
 import net.minecraft.client.model.geom.ModelPart;
 import net.minecraft.client.renderer.MultiBufferSource;
 import net.minecraft.client.renderer.blockentity.BannerRenderer;
 import net.minecraft.client.renderer.texture.OverlayTexture;
 import net.minecraft.client.resources.model.ModelBakery;
@@ -29,11 +30,11 @@
 import net.minecraft.world.level.block.entity.BannerPattern;
 
 public class LoomScreen extends AbstractContainerScreen<LoomMenu> {
    private static final ResourceLocation BG_LOCATION = new ResourceLocation("textures/gui/container/loom.png");
    private static final int TOTAL_PATTERN_ROWS;
-   private final ModelPart flag;
+   private ModelPart flag;
    @Nullable
    private List<Pair<BannerPattern, DyeColor>> resultBannerPatterns;
    private ItemStack bannerStack;
    private ItemStack dyeStack;
    private ItemStack patternStack;
@@ -48,15 +49,19 @@
       super(var1, var2, var3);
       this.bannerStack = ItemStack.EMPTY;
       this.dyeStack = ItemStack.EMPTY;
       this.patternStack = ItemStack.EMPTY;
       this.startIndex = 1;
-      this.flag = BannerRenderer.makeFlag();
       var1.registerUpdateListener(this::containerChanged);
       this.titleLabelY -= 2;
    }
 
+   protected void init() {
+      super.init();
+      this.flag = this.minecraft.getEntityModels().getLayer(ModelLayers.BANNER).getChild("flag");
+   }
+
    public void render(PoseStack var1, int var2, int var3, float var4) {
       super.render(var1, var2, var3, var4);
       this.renderTooltip(var1, var2, var3);
    }
 
