--- net/minecraft/client/gui/screens/inventory/MinecartCommandBlockEditScreen.java
+++ net/minecraft/client/gui/screens/inventory/MinecartCommandBlockEditScreen.java
@@ -20,12 +20,10 @@
       return 150;
    }
 
    protected void init() {
       super.init();
-      this.trackOutput = this.getCommandBlock().isTrackOutput();
-      this.updateCommandOutput();
       this.commandEdit.setValue(this.getCommandBlock().getCommand());
    }
 
    protected void populateAndSendPacket(BaseCommandBlock var1) {
       if (var1 instanceof MinecartCommandBlock.MinecartCommandBase) {
