--- net/minecraft/client/gui/screens/inventory/PageButton.java
+++ net/minecraft/client/gui/screens/inventory/PageButton.java
@@ -1,11 +1,11 @@
 package net.minecraft.client.gui.screens.inventory;
 
 import com.mojang.blaze3d.systems.RenderSystem;
 import com.mojang.blaze3d.vertex.PoseStack;
-import net.minecraft.client.Minecraft;
 import net.minecraft.client.gui.components.Button;
+import net.minecraft.client.renderer.GameRenderer;
 import net.minecraft.client.resources.sounds.SimpleSoundInstance;
 import net.minecraft.client.sounds.SoundManager;
 import net.minecraft.network.chat.TextComponent;
 import net.minecraft.sounds.SoundEvents;
 
@@ -18,12 +18,13 @@
       this.isForward = var3;
       this.playTurnSound = var5;
    }
 
    public void renderButton(PoseStack var1, int var2, int var3, float var4) {
-      RenderSystem.color4f(1.0F, 1.0F, 1.0F, 1.0F);
-      Minecraft.getInstance().getTextureManager().bind(BookViewScreen.BOOK_LOCATION);
+      RenderSystem.setShader(GameRenderer::getPositionTexShader);
+      RenderSystem.setShaderColor(1.0F, 1.0F, 1.0F, 1.0F);
+      RenderSystem.setShaderTexture(0, BookViewScreen.BOOK_LOCATION);
       int var5 = 0;
       int var6 = 192;
       if (this.isHovered()) {
          var5 += 23;
       }
