--- net/minecraft/client/gui/screens/inventory/ShulkerBoxScreen.java
+++ net/minecraft/client/gui/screens/inventory/ShulkerBoxScreen.java
@@ -1,9 +1,10 @@
 package net.minecraft.client.gui.screens.inventory;
 
 import com.mojang.blaze3d.systems.RenderSystem;
 import com.mojang.blaze3d.vertex.PoseStack;
+import net.minecraft.client.renderer.GameRenderer;
 import net.minecraft.network.chat.Component;
 import net.minecraft.resources.ResourceLocation;
 import net.minecraft.world.entity.player.Inventory;
 import net.minecraft.world.inventory.ShulkerBoxMenu;
 
@@ -20,12 +21,13 @@
       super.render(var1, var2, var3, var4);
       this.renderTooltip(var1, var2, var3);
    }
 
    protected void renderBg(PoseStack var1, float var2, int var3, int var4) {
-      RenderSystem.color4f(1.0F, 1.0F, 1.0F, 1.0F);
-      this.minecraft.getTextureManager().bind(CONTAINER_TEXTURE);
+      RenderSystem.setShader(GameRenderer::getPositionTexShader);
+      RenderSystem.setShaderColor(1.0F, 1.0F, 1.0F, 1.0F);
+      RenderSystem.setShaderTexture(0, CONTAINER_TEXTURE);
       int var5 = (this.width - this.imageWidth) / 2;
       int var6 = (this.height - this.imageHeight) / 2;
       this.blit(var1, var5, var6, 0, 0, this.imageWidth, this.imageHeight);
    }
 }
