--- net/minecraft/world/entity/ai/behavior/AcquirePoi.java
+++ net/minecraft/world/entity/ai/behavior/AcquirePoi.java
@@ -19,10 +19,13 @@
 import net.minecraft.world.entity.ai.village.poi.PoiManager;
 import net.minecraft.world.entity.ai.village.poi.PoiType;
 import net.minecraft.world.level.pathfinder.Path;
 
 public class AcquirePoi extends Behavior<PathfinderMob> {
+   private static final int BATCH_SIZE = 5;
+   private static final int RATE = 20;
+   public static final int SCAN_RANGE = 48;
    private final PoiType poiType;
    private final MemoryModuleType<GlobalPos> memoryToAcquire;
    private final boolean onlyIfAdult;
    private final Optional<Byte> onPoiAcquisitionEvent;
    private long nextScheduledStart;
@@ -106,10 +109,13 @@
       }
 
    }
 
    static class JitteredLinearRetry {
+      private static final int MIN_INTERVAL_INCREASE = 40;
+      private static final int MAX_INTERVAL_INCREASE = 80;
+      private static final int MAX_RETRY_PATHFINDING_INTERVAL = 400;
       private final Random random;
       private long previousAttemptTimestamp;
       private long nextScheduledAttemptTimestamp;
       private int currentDelay;
 
