--- net/minecraft/world/entity/ai/behavior/AnimalMakeLove.java
+++ net/minecraft/world/entity/ai/behavior/AnimalMakeLove.java
@@ -3,11 +3,11 @@
 import com.google.common.collect.ImmutableMap;
 import java.util.List;
 import java.util.Optional;
 import java.util.stream.Stream;
 import net.minecraft.server.level.ServerLevel;
-import net.minecraft.world.entity.AgableMob;
+import net.minecraft.world.entity.AgeableMob;
 import net.minecraft.world.entity.EntityType;
 import net.minecraft.world.entity.ai.Brain;
 import net.minecraft.world.entity.ai.memory.MemoryModuleType;
 import net.minecraft.world.entity.ai.memory.MemoryStatus;
 import net.minecraft.world.entity.animal.Animal;
@@ -69,11 +69,11 @@
       return (Animal)var1.getBrain().getMemory(MemoryModuleType.BREED_TARGET).get();
    }
 
    private boolean hasBreedTargetOfRightType(Animal var1) {
       Brain var2 = var1.getBrain();
-      return var2.hasMemoryValue(MemoryModuleType.BREED_TARGET) && ((AgableMob)var2.getMemory(MemoryModuleType.BREED_TARGET).get()).getType() == this.partnerType;
+      return var2.hasMemoryValue(MemoryModuleType.BREED_TARGET) && ((AgeableMob)var2.getMemory(MemoryModuleType.BREED_TARGET).get()).getType() == this.partnerType;
    }
 
    private Optional<? extends Animal> findValidBreedPartner(Animal var1) {
       Stream var10000 = ((List)var1.getBrain().getMemory(MemoryModuleType.VISIBLE_LIVING_ENTITIES).get()).stream().filter((var1x) -> {
          return var1x.getType() == this.partnerType;
