--- net/minecraft/world/entity/ai/behavior/BlockPosTracker.java
+++ net/minecraft/world/entity/ai/behavior/BlockPosTracker.java
@@ -24,8 +24,8 @@
    public boolean isVisibleBy(LivingEntity var1) {
       return true;
    }
 
    public String toString() {
-      return "BlockPosTracker{blockPos=" + this.blockPos + ", centerPosition=" + this.centerPosition + '}';
+      return "BlockPosTracker{blockPos=" + this.blockPos + ", centerPosition=" + this.centerPosition + "}";
    }
 }
