--- net/minecraft/world/entity/ai/behavior/CrossbowAttack.java
+++ net/minecraft/world/entity/ai/behavior/CrossbowAttack.java
@@ -12,10 +12,11 @@
 import net.minecraft.world.item.CrossbowItem;
 import net.minecraft.world.item.ItemStack;
 import net.minecraft.world.item.Items;
 
 public class CrossbowAttack<E extends Mob & CrossbowAttackMob, T extends LivingEntity> extends Behavior<E> {
+   private static final int TIMEOUT = 1200;
    private int attackDelay;
    private CrossbowAttack.CrossbowState crossbowState;
 
    public CrossbowAttack() {
       super(ImmutableMap.of(MemoryModuleType.LOOK_TARGET, MemoryStatus.REGISTERED, MemoryModuleType.ATTACK_TARGET, MemoryStatus.VALUE_PRESENT), 1200);
@@ -87,11 +88,11 @@
 
    private static LivingEntity getAttackTarget(LivingEntity var0) {
       return (LivingEntity)var0.getBrain().getMemory(MemoryModuleType.ATTACK_TARGET).get();
    }
 
-   static enum CrossbowState {
+   private static enum CrossbowState {
       UNCHARGED,
       CHARGING,
       CHARGED,
       READY_TO_ATTACK;
 
