--- net/minecraft/world/entity/ai/behavior/GateBehavior.java
+++ net/minecraft/world/entity/ai/behavior/GateBehavior.java
@@ -69,11 +69,11 @@
          return var0.getStatus() == Behavior.Status.RUNNING;
       }).collect(Collectors.toSet());
       return "(" + this.getClass().getSimpleName() + "): " + var1;
    }
 
-   static enum RunningPolicy {
+   public static enum RunningPolicy {
       RUN_ONE {
          public <E extends LivingEntity> void apply(WeightedList<Behavior<? super E>> var1, ServerLevel var2, E var3, long var4) {
             var1.stream().filter((var0) -> {
                return var0.getStatus() == Behavior.Status.STOPPED;
             }).filter((var4x) -> {
@@ -95,11 +95,11 @@
       }
 
       public abstract <E extends LivingEntity> void apply(WeightedList<Behavior<? super E>> var1, ServerLevel var2, E var3, long var4);
    }
 
-   static enum OrderPolicy {
+   public static enum OrderPolicy {
       ORDERED((var0) -> {
       }),
       SHUFFLED(WeightedList::shuffle);
 
       private final Consumer<WeightedList<?>> consumer;
