--- net/minecraft/world/entity/ai/behavior/GoToClosestVillage.java
+++ net/minecraft/world/entity/ai/behavior/GoToClosestVillage.java
@@ -5,11 +5,11 @@
 import net.minecraft.core.SectionPos;
 import net.minecraft.server.level.ServerLevel;
 import net.minecraft.world.entity.ai.memory.MemoryModuleType;
 import net.minecraft.world.entity.ai.memory.MemoryStatus;
 import net.minecraft.world.entity.ai.memory.WalkTarget;
-import net.minecraft.world.entity.ai.util.RandomPos;
+import net.minecraft.world.entity.ai.util.LandRandomPos;
 import net.minecraft.world.entity.ai.village.poi.PoiManager;
 import net.minecraft.world.entity.npc.Villager;
 import net.minecraft.world.phys.Vec3;
 
 public class GoToClosestVillage extends Behavior<Villager> {
@@ -30,11 +30,11 @@
       PoiManager var5 = var1.getPoiManager();
       int var6 = var5.sectionsToVillage(SectionPos.of(var2.blockPosition()));
       Vec3 var7 = null;
 
       for(int var8 = 0; var8 < 5; ++var8) {
-         Vec3 var9 = RandomPos.getLandPos(var2, 15, 7, (var1x) -> {
+         Vec3 var9 = LandRandomPos.getPos(var2, 15, 7, (var1x) -> {
             return (double)(-var1.sectionsToVillage(SectionPos.of(var1x)));
          });
          if (var9 != null) {
             int var10 = var5.sectionsToVillage(SectionPos.of(new BlockPos(var9)));
             if (var10 < var6) {
