--- net/minecraft/world/entity/ai/behavior/JumpOnBed.java
+++ net/minecraft/world/entity/ai/behavior/JumpOnBed.java
@@ -10,10 +10,14 @@
 import net.minecraft.world.entity.ai.memory.MemoryModuleType;
 import net.minecraft.world.entity.ai.memory.MemoryStatus;
 import net.minecraft.world.entity.ai.memory.WalkTarget;
 
 public class JumpOnBed extends Behavior<Mob> {
+   private static final int MAX_TIME_TO_REACH_BED = 100;
+   private static final int MIN_JUMPS = 3;
+   private static final int MAX_JUMPS = 6;
+   private static final int COOLDOWN_BETWEEN_JUMPS = 5;
    private final float speedModifier;
    @Nullable
    private BlockPos targetBed;
    private int remainingTimeToReachBed;
    private int remainingJumps;
