--- net/minecraft/world/entity/ai/behavior/Mount.java
+++ net/minecraft/world/entity/ai/behavior/Mount.java
@@ -6,10 +6,11 @@
 import net.minecraft.world.entity.LivingEntity;
 import net.minecraft.world.entity.ai.memory.MemoryModuleType;
 import net.minecraft.world.entity.ai.memory.MemoryStatus;
 
 public class Mount<E extends LivingEntity> extends Behavior<E> {
+   private static final int CLOSE_ENOUGH_TO_START_RIDING_DIST = 1;
    private final float speedModifier;
 
    public Mount(float var1) {
       super(ImmutableMap.of(MemoryModuleType.LOOK_TARGET, MemoryStatus.REGISTERED, MemoryModuleType.WALK_TARGET, MemoryStatus.VALUE_ABSENT, MemoryModuleType.RIDE_TARGET, MemoryStatus.VALUE_PRESENT));
       this.speedModifier = var1;
