--- net/minecraft/world/entity/ai/behavior/RandomStroll.java
+++ net/minecraft/world/entity/ai/behavior/RandomStroll.java
@@ -1,20 +1,24 @@
 package net.minecraft.world.entity.ai.behavior;
 
 import com.google.common.collect.ImmutableMap;
 import java.util.Optional;
+import javax.annotation.Nullable;
 import net.minecraft.server.level.ServerLevel;
 import net.minecraft.world.entity.PathfinderMob;
 import net.minecraft.world.entity.ai.memory.MemoryModuleType;
 import net.minecraft.world.entity.ai.memory.MemoryStatus;
 import net.minecraft.world.entity.ai.memory.WalkTarget;
-import net.minecraft.world.entity.ai.util.RandomPos;
+import net.minecraft.world.entity.ai.util.LandRandomPos;
+import net.minecraft.world.phys.Vec3;
 
 public class RandomStroll extends Behavior<PathfinderMob> {
+   private static final int MAX_XZ_DIST = 10;
+   private static final int MAX_Y_DIST = 7;
    private final float speedModifier;
-   private final int maxHorizontalDistance;
-   private final int maxVerticalDistance;
+   protected final int maxHorizontalDistance;
+   protected final int maxVerticalDistance;
 
    public RandomStroll(float var1) {
       this(var1, 10, 7);
    }
 
@@ -24,11 +28,16 @@
       this.maxHorizontalDistance = var2;
       this.maxVerticalDistance = var3;
    }
 
    protected void start(ServerLevel var1, PathfinderMob var2, long var3) {
-      Optional var5 = Optional.ofNullable(RandomPos.getLandPos(var2, this.maxHorizontalDistance, this.maxVerticalDistance));
+      Optional var5 = Optional.ofNullable(this.getTargetPos(var2));
       var2.getBrain().setMemory(MemoryModuleType.WALK_TARGET, var5.map((var1x) -> {
          return new WalkTarget(var1x, this.speedModifier, 0);
       }));
    }
+
+   @Nullable
+   protected Vec3 getTargetPos(PathfinderMob var1) {
+      return LandRandomPos.getPos(var1, this.maxHorizontalDistance, this.maxVerticalDistance);
+   }
 }
