--- net/minecraft/world/entity/ai/behavior/RandomStroll.java
+++ net/minecraft/world/entity/ai/behavior/RandomStroll.java
@@ -5,11 +5,11 @@
 import net.minecraft.server.level.ServerLevel;
 import net.minecraft.world.entity.PathfinderMob;
 import net.minecraft.world.entity.ai.memory.MemoryModuleType;
 import net.minecraft.world.entity.ai.memory.MemoryStatus;
 import net.minecraft.world.entity.ai.memory.WalkTarget;
-import net.minecraft.world.entity.ai.util.RandomPos;
+import net.minecraft.world.entity.ai.util.LandRandomPos;
 
 public class RandomStroll extends Behavior<PathfinderMob> {
    private final float speedModifier;
    private final int maxHorizontalDistance;
    private final int maxVerticalDistance;
@@ -24,11 +24,11 @@
       this.maxHorizontalDistance = var2;
       this.maxVerticalDistance = var3;
    }
 
    protected void start(ServerLevel var1, PathfinderMob var2, long var3) {
-      Optional var5 = Optional.ofNullable(RandomPos.getLandPos(var2, this.maxHorizontalDistance, this.maxVerticalDistance));
+      Optional var5 = Optional.ofNullable(LandRandomPos.getPos(var2, this.maxHorizontalDistance, this.maxVerticalDistance));
       var2.getBrain().setMemory(MemoryModuleType.WALK_TARGET, var5.map((var1x) -> {
          return new WalkTarget(var1x, this.speedModifier, 0);
       }));
    }
 }
