--- net/minecraft/world/entity/ai/behavior/RunSometimes.java
+++ net/minecraft/world/entity/ai/behavior/RunSometimes.java
@@ -1,23 +1,23 @@
 package net.minecraft.world.entity.ai.behavior;
 
 import net.minecraft.server.level.ServerLevel;
-import net.minecraft.util.IntRange;
+import net.minecraft.util.valueproviders.UniformInt;
 import net.minecraft.world.entity.LivingEntity;
 
 public class RunSometimes<E extends LivingEntity> extends Behavior<E> {
    private boolean resetTicks;
    private boolean wasRunning;
-   private final IntRange interval;
+   private final UniformInt interval;
    private final Behavior<? super E> wrappedBehavior;
    private int ticksUntilNextStart;
 
-   public RunSometimes(Behavior<? super E> var1, IntRange var2) {
+   public RunSometimes(Behavior<? super E> var1, UniformInt var2) {
       this(var1, false, var2);
    }
 
-   public RunSometimes(Behavior<? super E> var1, boolean var2, IntRange var3) {
+   public RunSometimes(Behavior<? super E> var1, boolean var2, UniformInt var3) {
       super(var1.entryCondition);
       this.wrappedBehavior = var1;
       this.resetTicks = !var2;
       this.interval = var3;
    }
@@ -56,11 +56,11 @@
       this.resetTicksUntilNextStart(var1);
       this.wrappedBehavior.stop(var1, var2, var3);
    }
 
    private void resetTicksUntilNextStart(ServerLevel var1) {
-      this.ticksUntilNextStart = this.interval.randomValue(var1.random);
+      this.ticksUntilNextStart = this.interval.sample(var1.random);
    }
 
    protected boolean timedOut(long var1) {
       return false;
    }
