--- net/minecraft/world/entity/ai/behavior/SetHiddenState.java
+++ net/minecraft/world/entity/ai/behavior/SetHiddenState.java
@@ -9,10 +9,11 @@
 import net.minecraft.world.entity.ai.Brain;
 import net.minecraft.world.entity.ai.memory.MemoryModuleType;
 import net.minecraft.world.entity.ai.memory.MemoryStatus;
 
 public class SetHiddenState extends Behavior<LivingEntity> {
+   private static final int HIDE_TIMEOUT = 300;
    private final int closeEnoughDist;
    private final int stayHiddenTicks;
    private int ticksHidden;
 
    public SetHiddenState(int var1, int var2) {
