--- net/minecraft/world/entity/ai/behavior/StartAttacking.java
+++ net/minecraft/world/entity/ai/behavior/StartAttacking.java
@@ -29,11 +29,16 @@
    protected boolean checkExtraStartConditions(ServerLevel var1, E var2) {
       if (!this.canAttackPredicate.test(var2)) {
          return false;
       } else {
          Optional var3 = (Optional)this.targetFinderFunction.apply(var2);
-         return var3.isPresent() && ((LivingEntity)var3.get()).isAlive();
+         if (!var3.isPresent()) {
+            return false;
+         } else {
+            LivingEntity var4 = (LivingEntity)var3.get();
+            return var4.isAlive() && var4.canBeTargeted();
+         }
       }
    }
 
    protected void start(ServerLevel var1, E var2, long var3) {
       ((Optional)this.targetFinderFunction.apply(var2)).ifPresent((var2x) -> {
