--- net/minecraft/world/entity/ai/behavior/StopAttackingIfTargetInvalid.java
+++ net/minecraft/world/entity/ai/behavior/StopAttackingIfTargetInvalid.java
@@ -23,12 +23,15 @@
          return false;
       });
    }
 
    protected void start(ServerLevel var1, E var2, long var3) {
-      if (isTiredOfTryingToReachTarget(var2)) {
+      LivingEntity var5 = this.getAttackTarget(var2);
+      if (!var5.canBeTargeted()) {
          this.clearAttackTarget(var2);
+      } else if (isTiredOfTryingToReachTarget(var2)) {
+         this.clearAttackTarget(var2);
       } else if (this.isCurrentTargetDeadOrRemoved(var2)) {
          this.clearAttackTarget(var2);
       } else if (this.isCurrentTargetInDifferentLevel(var2)) {
          this.clearAttackTarget(var2);
       } else if (!EntitySelector.ATTACK_ALLOWED.test(this.getAttackTarget(var2))) {
