--- net/minecraft/world/entity/ai/behavior/TradeWithVillager.java
+++ net/minecraft/world/entity/ai/behavior/TradeWithVillager.java
@@ -15,14 +15,16 @@
 import net.minecraft.world.item.Item;
 import net.minecraft.world.item.ItemStack;
 import net.minecraft.world.item.Items;
 
 public class TradeWithVillager extends Behavior<Villager> {
+   private static final int INTERACT_DIST_SQR = 5;
+   private static final float SPEED_MODIFIER = 0.5F;
    private Set<Item> trades = ImmutableSet.of();
 
    public TradeWithVillager() {
-      super(ImmutableMap.of(MemoryModuleType.INTERACTION_TARGET, MemoryStatus.VALUE_PRESENT, MemoryModuleType.VISIBLE_LIVING_ENTITIES, MemoryStatus.VALUE_PRESENT));
+      super(ImmutableMap.of(MemoryModuleType.INTERACTION_TARGET, MemoryStatus.VALUE_PRESENT, MemoryModuleType.NEAREST_VISIBLE_LIVING_ENTITIES, MemoryStatus.VALUE_PRESENT));
    }
 
    protected boolean checkExtraStartConditions(ServerLevel var1, Villager var2) {
       return BehaviorUtils.targetIsValid(var2.getBrain(), MemoryModuleType.INTERACTION_TARGET, EntityType.VILLAGER);
    }
