--- net/minecraft/world/entity/ai/behavior/TradeWithVillager.java
+++ net/minecraft/world/entity/ai/behavior/TradeWithVillager.java
@@ -15,10 +15,12 @@
 import net.minecraft.world.item.Item;
 import net.minecraft.world.item.ItemStack;
 import net.minecraft.world.item.Items;
 
 public class TradeWithVillager extends Behavior<Villager> {
+   private static final int INTERACT_DIST_SQR = 5;
+   private static final float SPEED_MODIFIER = 0.5F;
    private Set<Item> trades = ImmutableSet.of();
 
    public TradeWithVillager() {
       super(ImmutableMap.of(MemoryModuleType.INTERACTION_TARGET, MemoryStatus.VALUE_PRESENT, MemoryModuleType.VISIBLE_LIVING_ENTITIES, MemoryStatus.VALUE_PRESENT));
    }
