--- net/minecraft/world/entity/ai/behavior/WorkAtPoi.java
+++ net/minecraft/world/entity/ai/behavior/WorkAtPoi.java
@@ -8,10 +8,12 @@
 import net.minecraft.world.entity.ai.memory.MemoryModuleType;
 import net.minecraft.world.entity.ai.memory.MemoryStatus;
 import net.minecraft.world.entity.npc.Villager;
 
 public class WorkAtPoi extends Behavior<Villager> {
+   private static final int CHECK_COOLDOWN = 300;
+   private static final double DISTANCE = 1.73D;
    private long lastCheck;
 
    public WorkAtPoi() {
       super(ImmutableMap.of(MemoryModuleType.JOB_SITE, MemoryStatus.VALUE_PRESENT, MemoryModuleType.LOOK_TARGET, MemoryStatus.REGISTERED));
    }
