--- net/minecraft/world/entity/ai/behavior/YieldJobSite.java
+++ net/minecraft/world/entity/ai/behavior/YieldJobSite.java
@@ -15,11 +15,11 @@
 
 public class YieldJobSite extends Behavior<Villager> {
    private final float speedModifier;
 
    public YieldJobSite(float var1) {
-      super(ImmutableMap.of(MemoryModuleType.POTENTIAL_JOB_SITE, MemoryStatus.VALUE_PRESENT, MemoryModuleType.JOB_SITE, MemoryStatus.VALUE_ABSENT, MemoryModuleType.LIVING_ENTITIES, MemoryStatus.VALUE_PRESENT));
+      super(ImmutableMap.of(MemoryModuleType.POTENTIAL_JOB_SITE, MemoryStatus.VALUE_PRESENT, MemoryModuleType.JOB_SITE, MemoryStatus.VALUE_ABSENT, MemoryModuleType.NEAREST_LIVING_ENTITIES, MemoryStatus.VALUE_PRESENT));
       this.speedModifier = var1;
    }
 
    protected boolean checkExtraStartConditions(ServerLevel var1, Villager var2) {
       if (var2.isBaby()) {
