--- net/minecraft/data/tags/BlockTagsProvider.java
+++ net/minecraft/data/tags/BlockTagsProvider.java
@@ -98,10 +98,17 @@
       this.tag(BlockTags.INFINIBURN_OVERWORLD).add((Object[])(Blocks.NETHERRACK, Blocks.MAGMA_BLOCK));
       this.tag(BlockTags.INFINIBURN_NETHER).addTag(BlockTags.INFINIBURN_OVERWORLD);
       this.tag(BlockTags.INFINIBURN_END).addTag(BlockTags.INFINIBURN_OVERWORLD).add((Object)Blocks.BEDROCK);
       this.tag(BlockTags.BASE_STONE_OVERWORLD).add((Object)Blocks.STONE).add((Object)Blocks.GRANITE).add((Object)Blocks.DIORITE).add((Object)Blocks.ANDESITE);
       this.tag(BlockTags.BASE_STONE_NETHER).add((Object)Blocks.NETHERRACK).add((Object)Blocks.BASALT).add((Object)Blocks.BLACKSTONE);
+      this.tag(BlockTags.CANDLES).add((Object[])(Blocks.CANDLE, Blocks.WHITE_CANDLE, Blocks.ORANGE_CANDLE, Blocks.MAGENTA_CANDLE, Blocks.LIGHT_BLUE_CANDLE, Blocks.YELLOW_CANDLE, Blocks.LIME_CANDLE, Blocks.PINK_CANDLE, Blocks.GRAY_CANDLE, Blocks.LIGHT_GRAY_CANDLE, Blocks.CYAN_CANDLE, Blocks.PURPLE_CANDLE, Blocks.BLUE_CANDLE, Blocks.BROWN_CANDLE, Blocks.GREEN_CANDLE, Blocks.RED_CANDLE, Blocks.BLACK_CANDLE));
+      this.tag(BlockTags.CANDLE_CAKES).add((Object[])(Blocks.CANDLE_CAKE, Blocks.WHITE_CANDLE_CAKE, Blocks.ORANGE_CANDLE_CAKE, Blocks.MAGENTA_CANDLE_CAKE, Blocks.LIGHT_BLUE_CANDLE_CAKE, Blocks.YELLOW_CANDLE_CAKE, Blocks.LIME_CANDLE_CAKE, Blocks.PINK_CANDLE_CAKE, Blocks.GRAY_CANDLE_CAKE, Blocks.LIGHT_GRAY_CANDLE_CAKE, Blocks.CYAN_CANDLE_CAKE, Blocks.PURPLE_CANDLE_CAKE, Blocks.BLUE_CANDLE_CAKE, Blocks.BROWN_CANDLE_CAKE, Blocks.GREEN_CANDLE_CAKE, Blocks.RED_CANDLE_CAKE, Blocks.BLACK_CANDLE_CAKE));
+      this.tag(BlockTags.CRYSTAL_SOUND_BLOCKS).add((Object[])(Blocks.AMETHYST_BLOCK, Blocks.BUDDING_AMETHYST));
+      this.tag(BlockTags.CAULDRONS).add((Object[])(Blocks.CAULDRON, Blocks.WATER_CAULDRON, Blocks.LAVA_CAULDRON, Blocks.POWDER_SNOW_CAULDRON));
+      this.tag(BlockTags.INSIDE_STEP_SOUND_BLOCKS).add((Object[])(Blocks.SNOW, Blocks.POWDER_SNOW));
+      this.tag(BlockTags.DRIPSTONE_REPLACEABLE).addTag(BlockTags.BASE_STONE_OVERWORLD).add((Object)Blocks.DIRT);
+      this.tag(BlockTags.OCCLUDES_VIBRATION_SIGNALS).addTag(BlockTags.WOOL);
    }
 
    protected Path getPath(ResourceLocation var1) {
       return this.generator.getOutputFolder().resolve("data/" + var1.getNamespace() + "/tags/blocks/" + var1.getPath() + ".json");
    }
