--- net/minecraft/data/tags/FluidTagsProvider.java
+++ net/minecraft/data/tags/FluidTagsProvider.java
@@ -17,11 +17,13 @@
       this.tag(FluidTags.WATER).add((Object[])(Fluids.WATER, Fluids.FLOWING_WATER));
       this.tag(FluidTags.LAVA).add((Object[])(Fluids.LAVA, Fluids.FLOWING_LAVA));
    }
 
    protected Path getPath(ResourceLocation var1) {
-      return this.generator.getOutputFolder().resolve("data/" + var1.getNamespace() + "/tags/fluids/" + var1.getPath() + ".json");
+      Path var10000 = this.generator.getOutputFolder();
+      String var10001 = var1.getNamespace();
+      return var10000.resolve("data/" + var10001 + "/tags/fluids/" + var1.getPath() + ".json");
    }
 
    public String getName() {
       return "Fluid Tags";
    }
