--- net/minecraft/world/level/dimension/end/EndDragonFight.java
+++ net/minecraft/world/level/dimension/end/EndDragonFight.java
@@ -57,10 +57,18 @@
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
 
 public class EndDragonFight {
    private static final Logger LOGGER = LogManager.getLogger();
+   private static final int MAX_TICKS_BEFORE_DRAGON_RESPAWN = 1200;
+   private static final int TIME_BETWEEN_CRYSTAL_SCANS = 100;
+   private static final int TIME_BETWEEN_PLAYER_SCANS = 20;
+   private static final int ARENA_SIZE_CHUNKS = 8;
+   public static final int ARENA_TICKET_LEVEL = 9;
+   private static final int GATEWAY_COUNT = 20;
+   private static final int GATEWAY_DISTANCE = 96;
+   public static final int DRAGON_SPAWN_Y = 128;
    private static final Predicate<Entity> VALID_PLAYER;
    private final ServerBossEvent dragonEvent;
    private final ServerLevel level;
    private final List<Integer> gateways;
    private final BlockPattern exitPortalPattern;
@@ -202,11 +210,11 @@
          this.dragonUUID = var3.getUUID();
          LOGGER.info("Found that there's a dragon still alive ({})", var3);
          this.dragonKilled = false;
          if (!var1) {
             LOGGER.info("But we didn't have a portal, let's remove it.");
-            var3.remove();
+            var3.discard();
             this.dragonUUID = null;
          }
       }
 
       if (!this.previouslyKilled && this.dragonKilled) {
@@ -293,11 +301,11 @@
          }
       }
 
       var1 = this.level.getHeightmapPos(Heightmap.Types.MOTION_BLOCKING, EndPodiumFeature.END_PODIUM_LOCATION).getY();
 
-      for(var2 = var1; var2 >= 0; --var2) {
+      for(var2 = var1; var2 >= this.level.getMinBuildHeight(); --var2) {
          BlockPattern.BlockPatternMatch var8 = this.exitPortalPattern.find(this.level, new BlockPos(EndPodiumFeature.END_PODIUM_LOCATION.getX(), var2, EndPodiumFeature.END_PODIUM_LOCATION.getZ()));
          if (var8 != null) {
             if (this.portalLocation == null) {
                this.portalLocation = var8.getBlock(3, 3, 3).getPos();
             }
@@ -360,11 +368,11 @@
       LOGGER.debug("Found {} end crystals still alive", this.crystalsAlive);
    }
 
    public void setDragonKilled(EnderDragon var1) {
       if (var1.getUUID().equals(this.dragonUUID)) {
-         this.dragonEvent.setPercent(0.0F);
+         this.dragonEvent.setProgress(0.0F);
          this.dragonEvent.setVisible(false);
          this.spawnExitPortal(true);
          this.spawnNewGateway();
          if (!this.previouslyKilled) {
             this.level.setBlockAndUpdate(this.level.getHeightmapPos(Heightmap.Types.MOTION_BLOCKING, EndPodiumFeature.END_PODIUM_LOCATION), Blocks.DRAGON_EGG.defaultBlockState());
@@ -410,11 +418,11 @@
       return var1;
    }
 
    public void updateDragon(EnderDragon var1) {
       if (var1.getUUID().equals(this.dragonUUID)) {
-         this.dragonEvent.setPercent(var1.getHealth() / var1.getMaxHealth());
+         this.dragonEvent.setProgress(var1.getHealth() / var1.getMaxHealth());
          this.ticksSinceDragonSeen = 0;
          if (var1.hasCustomName()) {
             this.dragonEvent.setName(var1.getDisplayName());
          }
       }
