--- net/minecraft/world/entity/ai/Brain.java
+++ net/minecraft/world/entity/ai/Brain.java
@@ -360,11 +360,11 @@
 
    public void addActivityWithConditions(Activity var1, ImmutableList<? extends Pair<Integer, ? extends Behavior<? super E>>> var2, Set<Pair<MemoryModuleType<?>, MemoryStatus>> var3) {
       this.addActivityAndRemoveMemoriesWhenStopped(var1, var2, var3, Sets.newHashSet());
    }
 
-   private void addActivityAndRemoveMemoriesWhenStopped(Activity var1, ImmutableList<? extends Pair<Integer, ? extends Behavior<? super E>>> var2, Set<Pair<MemoryModuleType<?>, MemoryStatus>> var3, Set<MemoryModuleType<?>> var4) {
+   public void addActivityAndRemoveMemoriesWhenStopped(Activity var1, ImmutableList<? extends Pair<Integer, ? extends Behavior<? super E>>> var2, Set<Pair<MemoryModuleType<?>, MemoryStatus>> var3, Set<MemoryModuleType<?>> var4) {
       this.activityRequirements.put(var1, var3);
       if (!var4.isEmpty()) {
          this.activityMemoriesToEraseWhenStopped.put(var1, var4);
       }
 
