--- net/minecraft/world/entity/ai/Brain.java
+++ net/minecraft/world/entity/ai/Brain.java
@@ -1,7 +1,8 @@
 package net.minecraft.world.entity.ai;
 
+import com.google.common.annotations.VisibleForTesting;
 import com.google.common.collect.ImmutableList;
 import com.google.common.collect.ImmutableSet;
 import com.google.common.collect.Maps;
 import com.google.common.collect.Sets;
 import com.google.common.collect.UnmodifiableIterator;
@@ -16,19 +17,21 @@
 import it.unimi.dsi.fastutil.objects.ObjectArrayList;
 import java.util.Collection;
 import java.util.Iterator;
 import java.util.List;
 import java.util.Map;
+import java.util.Objects;
 import java.util.Optional;
 import java.util.Set;
 import java.util.Map.Entry;
 import java.util.function.Supplier;
 import java.util.stream.Stream;
 import javax.annotation.Nullable;
 import net.minecraft.Util;
 import net.minecraft.core.Registry;
 import net.minecraft.server.level.ServerLevel;
+import net.minecraft.util.VisibleForDebug;
 import net.minecraft.world.entity.LivingEntity;
 import net.minecraft.world.entity.ai.behavior.Behavior;
 import net.minecraft.world.entity.ai.memory.ExpirableValue;
 import net.minecraft.world.entity.ai.memory.MemoryModuleType;
 import net.minecraft.world.entity.ai.memory.MemoryStatus;
@@ -39,12 +42,13 @@
 import org.apache.commons.lang3.mutable.MutableObject;
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
 
 public class Brain<E extends LivingEntity> {
-   private static final Logger LOGGER = LogManager.getLogger();
+   static final Logger LOGGER = LogManager.getLogger();
    private final Supplier<Codec<Brain<E>>> codec;
+   private static final int SCHEDULE_UPDATE_DELAY = 20;
    private final Map<MemoryModuleType<?>, Optional<? extends ExpirableValue<?>>> memories = Maps.newHashMap();
    private final Map<SensorType<? extends Sensor<? super E>>, Sensor<? super E>> sensors = Maps.newLinkedHashMap();
    private final Map<Integer, Map<Activity, Set<Behavior<? super E>>>> availableBehaviorsByPriority = Maps.newTreeMap();
    private Schedule schedule;
    private final Map<Activity, Set<Pair<MemoryModuleType<?>, MemoryStatus>>> activityRequirements;
@@ -80,16 +84,16 @@
                });
                var3.setValue(((DataResult)var3.getValue()).apply2(ImmutableList.Builder::add, var5));
             });
             DataResult var10000 = (DataResult)var3.getValue();
             Logger var10001 = Brain.LOGGER;
-            var10001.getClass();
+            Objects.requireNonNull(var10001);
             ImmutableList var4 = (ImmutableList)var10000.resultOrPartial(var10001::error).map(ImmutableList.Builder::build).orElseGet(ImmutableList::of);
             Collection var10002 = var0;
             Collection var10003 = var1;
             MutableObject var10005 = var2;
-            var10005.getClass();
+            Objects.requireNonNull(var10005);
             return DataResult.success(new Brain(var10002, var10003, var4, var10005::getValue));
          }
 
          private <T, U> DataResult<Brain.MemoryValue<U>> captureRead(MemoryModuleType<U> var1x, DynamicOps<T> var2x, T var3) {
             return ((DataResult)var1x.getCodec().map(DataResult::success).orElseGet(() -> {
@@ -157,11 +161,11 @@
 
    public <T> DataResult<T> serializeStart(DynamicOps<T> var1) {
       return ((Codec)this.codec.get()).encodeStart(var1, this);
    }
 
-   private Stream<Brain.MemoryValue<?>> memories() {
+   Stream<Brain.MemoryValue<?>> memories() {
       return this.memories.entrySet().stream().map((var0) -> {
          return Brain.MemoryValue.createUnchecked((MemoryModuleType)var0.getKey(), (Optional)var0.getValue());
       });
    }
 
@@ -183,11 +187,11 @@
 
    public <U> void setMemory(MemoryModuleType<U> var1, Optional<? extends U> var2) {
       this.setMemoryInternal(var1, var2.map(ExpirableValue::of));
    }
 
-   private <U> void setMemoryInternal(MemoryModuleType<U> var1, Optional<? extends ExpirableValue<?>> var2) {
+   <U> void setMemoryInternal(MemoryModuleType<U> var1, Optional<? extends ExpirableValue<?>> var2) {
       if (this.memories.containsKey(var1)) {
          if (var2.isPresent() && this.isEmptyCollection(((ExpirableValue)var2.get()).getValue())) {
             this.eraseMemory(var1);
          } else {
             this.memories.put(var1, var2);
@@ -198,10 +202,21 @@
 
    public <U> Optional<U> getMemory(MemoryModuleType<U> var1) {
       return ((Optional)this.memories.get(var1)).map(ExpirableValue::getValue);
    }
 
+   public <U> long getTimeUntilExpiry(MemoryModuleType<U> var1) {
+      Optional var2 = (Optional)this.memories.get(var1);
+      return (Long)var2.map(ExpirableValue::getTimeToLive).orElse(0L);
+   }
+
+   @Deprecated
+   @VisibleForDebug
+   public Map<MemoryModuleType<?>, Optional<? extends ExpirableValue<?>>> getMemories() {
+      return this.memories;
+   }
+
    public <U> boolean isMemoryValue(MemoryModuleType<U> var1, U var2) {
       return !this.hasMemoryValue(var1) ? false : this.getMemory(var1).filter((var1x) -> {
          return var1x.equals(var2);
       }).isPresent();
    }
@@ -226,10 +241,17 @@
    public void setCoreActivities(Set<Activity> var1) {
       this.coreActivities = var1;
    }
 
    @Deprecated
+   @VisibleForDebug
+   public Set<Activity> getActiveActivities() {
+      return this.activeActivities;
+   }
+
+   @Deprecated
+   @VisibleForDebug
    public List<Behavior<? super E>> getRunningBehaviors() {
       ObjectArrayList var1 = new ObjectArrayList();
       Iterator var2 = this.availableBehaviorsByPriority.values().iterator();
 
       while(var2.hasNext()) {
@@ -360,29 +382,34 @@
 
    public void addActivityWithConditions(Activity var1, ImmutableList<? extends Pair<Integer, ? extends Behavior<? super E>>> var2, Set<Pair<MemoryModuleType<?>, MemoryStatus>> var3) {
       this.addActivityAndRemoveMemoriesWhenStopped(var1, var2, var3, Sets.newHashSet());
    }
 
-   private void addActivityAndRemoveMemoriesWhenStopped(Activity var1, ImmutableList<? extends Pair<Integer, ? extends Behavior<? super E>>> var2, Set<Pair<MemoryModuleType<?>, MemoryStatus>> var3, Set<MemoryModuleType<?>> var4) {
+   public void addActivityAndRemoveMemoriesWhenStopped(Activity var1, ImmutableList<? extends Pair<Integer, ? extends Behavior<? super E>>> var2, Set<Pair<MemoryModuleType<?>, MemoryStatus>> var3, Set<MemoryModuleType<?>> var4) {
       this.activityRequirements.put(var1, var3);
       if (!var4.isEmpty()) {
          this.activityMemoriesToEraseWhenStopped.put(var1, var4);
       }
 
       UnmodifiableIterator var5 = var2.iterator();
 
       while(var5.hasNext()) {
          Pair var6 = (Pair)var5.next();
-         ((Set)((Map)this.availableBehaviorsByPriority.computeIfAbsent(var6.getFirst(), (var0) -> {
+         ((Set)((Map)this.availableBehaviorsByPriority.computeIfAbsent((Integer)var6.getFirst(), (var0) -> {
             return Maps.newHashMap();
          })).computeIfAbsent(var1, (var0) -> {
             return Sets.newLinkedHashSet();
-         })).add(var6.getSecond());
+         })).add((Behavior)var6.getSecond());
       }
 
    }
 
+   @VisibleForTesting
+   public void removeAllBehaviors() {
+      this.availableBehaviorsByPriority.clear();
+   }
+
    public boolean isActive(Activity var1) {
       return this.activeActivities.contains(var1);
    }
 
    public Brain<E> copyWithoutBehaviors() {
@@ -391,11 +418,11 @@
 
       while(var2.hasNext()) {
          Entry var3 = (Entry)var2.next();
          MemoryModuleType var4 = (MemoryModuleType)var3.getKey();
          if (((Optional)var3.getValue()).isPresent()) {
-            var1.memories.put(var4, var3.getValue());
+            var1.memories.put(var4, (Optional)var3.getValue());
          }
       }
 
       return var1;
    }
@@ -527,24 +554,47 @@
       }
 
       return var4.build();
    }
 
-   static final class MemoryValue<U> {
+   public static final class Provider<E extends LivingEntity> {
+      private final Collection<? extends MemoryModuleType<?>> memoryTypes;
+      private final Collection<? extends SensorType<? extends Sensor<? super E>>> sensorTypes;
+      private final Codec<Brain<E>> codec;
+
+      Provider(Collection<? extends MemoryModuleType<?>> var1, Collection<? extends SensorType<? extends Sensor<? super E>>> var2) {
+         this.memoryTypes = var1;
+         this.sensorTypes = var2;
+         this.codec = Brain.codec(var1, var2);
+      }
+
+      public Brain<E> makeBrain(Dynamic<?> var1) {
+         DataResult var10000 = this.codec.parse(var1);
+         Logger var10001 = Brain.LOGGER;
+         Objects.requireNonNull(var10001);
+         return (Brain)var10000.resultOrPartial(var10001::error).orElseGet(() -> {
+            return new Brain(this.memoryTypes, this.sensorTypes, ImmutableList.of(), () -> {
+               return this.codec;
+            });
+         });
+      }
+   }
+
+   private static final class MemoryValue<U> {
       private final MemoryModuleType<U> type;
       private final Optional<? extends ExpirableValue<U>> value;
 
-      private static <U> Brain.MemoryValue<U> createUnchecked(MemoryModuleType<U> var0, Optional<? extends ExpirableValue<?>> var1) {
+      static <U> Brain.MemoryValue<U> createUnchecked(MemoryModuleType<U> var0, Optional<? extends ExpirableValue<?>> var1) {
          return new Brain.MemoryValue(var0, var1);
       }
 
-      private MemoryValue(MemoryModuleType<U> var1, Optional<? extends ExpirableValue<U>> var2) {
+      MemoryValue(MemoryModuleType<U> var1, Optional<? extends ExpirableValue<U>> var2) {
          this.type = var1;
          this.value = var2;
       }
 
-      private void setMemoryInternal(Brain<?> var1) {
+      void setMemoryInternal(Brain<?> var1) {
          var1.setMemoryInternal(this.type, this.value);
       }
 
       public <T> void serialize(DynamicOps<T> var1, RecordBuilder<T> var2) {
          this.type.getCodec().ifPresent((var3) -> {
@@ -552,29 +602,6 @@
                var2.add(Registry.MEMORY_MODULE_TYPE.encodeStart(var1, this.type), var3.encodeStart(var1, var4));
             });
          });
       }
    }
-
-   public static final class Provider<E extends LivingEntity> {
-      private final Collection<? extends MemoryModuleType<?>> memoryTypes;
-      private final Collection<? extends SensorType<? extends Sensor<? super E>>> sensorTypes;
-      private final Codec<Brain<E>> codec;
-
-      private Provider(Collection<? extends MemoryModuleType<?>> var1, Collection<? extends SensorType<? extends Sensor<? super E>>> var2) {
-         this.memoryTypes = var1;
-         this.sensorTypes = var2;
-         this.codec = Brain.codec(var1, var2);
-      }
-
-      public Brain<E> makeBrain(Dynamic<?> var1) {
-         DataResult var10000 = this.codec.parse(var1);
-         Logger var10001 = Brain.LOGGER;
-         var10001.getClass();
-         return (Brain)var10000.resultOrPartial(var10001::error).orElseGet(() -> {
-            return new Brain(this.memoryTypes, this.sensorTypes, ImmutableList.of(), () -> {
-               return this.codec;
-            });
-         });
-      }
-   }
 }
