--- net/minecraft/client/renderer/blockentity/BannerRenderer.java
+++ net/minecraft/client/renderer/blockentity/BannerRenderer.java
@@ -3,11 +3,17 @@
 import com.mojang.blaze3d.vertex.PoseStack;
 import com.mojang.blaze3d.vertex.VertexConsumer;
 import com.mojang.datafixers.util.Pair;
 import com.mojang.math.Vector3f;
 import java.util.List;
+import net.minecraft.client.model.geom.ModelLayers;
 import net.minecraft.client.model.geom.ModelPart;
+import net.minecraft.client.model.geom.PartPose;
+import net.minecraft.client.model.geom.builders.CubeListBuilder;
+import net.minecraft.client.model.geom.builders.LayerDefinition;
+import net.minecraft.client.model.geom.builders.MeshDefinition;
+import net.minecraft.client.model.geom.builders.PartDefinition;
 import net.minecraft.client.renderer.MultiBufferSource;
 import net.minecraft.client.renderer.RenderType;
 import net.minecraft.client.renderer.Sheets;
 import net.minecraft.client.resources.model.Material;
 import net.minecraft.client.resources.model.ModelBakery;
@@ -19,26 +25,35 @@
 import net.minecraft.world.level.block.WallBannerBlock;
 import net.minecraft.world.level.block.entity.BannerBlockEntity;
 import net.minecraft.world.level.block.entity.BannerPattern;
 import net.minecraft.world.level.block.state.BlockState;
 
-public class BannerRenderer extends BlockEntityRenderer<BannerBlockEntity> {
-   private final ModelPart flag = makeFlag();
-   private final ModelPart pole = new ModelPart(64, 64, 44, 0);
+public class BannerRenderer implements BlockEntityRenderer<BannerBlockEntity> {
+   private static final int BANNER_WIDTH = 20;
+   private static final int BANNER_HEIGHT = 40;
+   private static final int MAX_PATTERNS = 16;
+   public static final String FLAG = "flag";
+   private static final String POLE = "pole";
+   private static final String BAR = "bar";
+   private final ModelPart flag;
+   private final ModelPart pole;
    private final ModelPart bar;
 
-   public BannerRenderer(BlockEntityRenderDispatcher var1) {
-      super(var1);
-      this.pole.addBox(-1.0F, -30.0F, -1.0F, 2.0F, 42.0F, 2.0F, 0.0F);
-      this.bar = new ModelPart(64, 64, 0, 42);
-      this.bar.addBox(-10.0F, -32.0F, -1.0F, 20.0F, 2.0F, 2.0F, 0.0F);
+   public BannerRenderer(BlockEntityRendererProvider.Context var1) {
+      ModelPart var2 = var1.bakeLayer(ModelLayers.BANNER);
+      this.flag = var2.getChild("flag");
+      this.pole = var2.getChild("pole");
+      this.bar = var2.getChild("bar");
    }
 
-   public static ModelPart makeFlag() {
-      ModelPart var0 = new ModelPart(64, 64, 0, 0);
-      var0.addBox(-10.0F, 0.0F, -2.0F, 20.0F, 40.0F, 1.0F, 0.0F);
-      return var0;
+   public static LayerDefinition createBodyLayer() {
+      MeshDefinition var0 = new MeshDefinition();
+      PartDefinition var1 = var0.getRoot();
+      var1.addOrReplaceChild("flag", CubeListBuilder.create().texOffs(0, 0).addBox(-10.0F, 0.0F, -2.0F, 20.0F, 40.0F, 1.0F), PartPose.ZERO);
+      var1.addOrReplaceChild("pole", CubeListBuilder.create().texOffs(44, 0).addBox(-1.0F, -30.0F, -1.0F, 2.0F, 42.0F, 2.0F), PartPose.ZERO);
+      var1.addOrReplaceChild("bar", CubeListBuilder.create().texOffs(0, 42).addBox(-10.0F, -32.0F, -1.0F, 20.0F, 2.0F, 2.0F), PartPose.ZERO);
+      return LayerDefinition.create(var0, 64, 64);
    }
 
    public void render(BannerBlockEntity var1, float var2, PoseStack var3, MultiBufferSource var4, int var5, int var6) {
       List var7 = var1.getPatterns();
       if (var7 != null) {
@@ -91,11 +106,12 @@
       var4.render(var0, var5.buffer(var1, RenderType::entitySolid, var8), var2, var3);
 
       for(int var9 = 0; var9 < 17 && var9 < var7.size(); ++var9) {
          Pair var10 = (Pair)var7.get(var9);
          float[] var11 = ((DyeColor)var10.getSecond()).getTextureDiffuseColors();
-         Material var12 = new Material(var6 ? Sheets.BANNER_SHEET : Sheets.SHIELD_SHEET, ((BannerPattern)var10.getFirst()).location(var6));
-         var4.render(var0, var12.buffer(var1, RenderType::entityNoOutline), var2, var3, var11[0], var11[1], var11[2], 1.0F);
+         BannerPattern var12 = (BannerPattern)var10.getFirst();
+         Material var13 = var6 ? Sheets.getBannerMaterial(var12) : Sheets.getShieldMaterial(var12);
+         var4.render(var0, var13.buffer(var1, RenderType::entityNoOutline), var2, var3, var11[0], var11[1], var11[2], 1.0F);
       }
 
    }
 }
