--- net/minecraft/client/renderer/blockentity/BeaconRenderer.java
+++ net/minecraft/client/renderer/blockentity/BeaconRenderer.java
@@ -10,16 +10,16 @@
 import net.minecraft.client.renderer.RenderType;
 import net.minecraft.client.renderer.texture.OverlayTexture;
 import net.minecraft.resources.ResourceLocation;
 import net.minecraft.util.Mth;
 import net.minecraft.world.level.block.entity.BeaconBlockEntity;
+import net.minecraft.world.phys.Vec3;
 
-public class BeaconRenderer extends BlockEntityRenderer<BeaconBlockEntity> {
+public class BeaconRenderer implements BlockEntityRenderer<BeaconBlockEntity> {
    public static final ResourceLocation BEAM_LOCATION = new ResourceLocation("textures/entity/beacon_beam.png");
 
-   public BeaconRenderer(BlockEntityRenderDispatcher var1) {
-      super(var1);
+   public BeaconRenderer(BlockEntityRendererProvider.Context var1) {
    }
 
    public void render(BeaconBlockEntity var1, float var2, PoseStack var3, MultiBufferSource var4, int var5, int var6) {
       long var7 = var1.getLevel().getGameTime();
       List var9 = var1.getBeamSections();
@@ -95,6 +95,14 @@
    }
 
    public boolean shouldRenderOffScreen(BeaconBlockEntity var1) {
       return true;
    }
+
+   public int getViewDistance() {
+      return 256;
+   }
+
+   public boolean shouldRender(BeaconBlockEntity var1, Vec3 var2) {
+      return Vec3.atCenterOf(var1.getBlockPos()).multiply(1.0D, 0.0D, 1.0D).closerThan(var2.multiply(1.0D, 0.0D, 1.0D), (double)this.getViewDistance());
+   }
 }
