--- net/minecraft/client/renderer/blockentity/BlockEntityRenderer.java
+++ net/minecraft/client/renderer/blockentity/BlockEntityRenderer.java
@@ -1,19 +1,22 @@
 package net.minecraft.client.renderer.blockentity;
 
 import com.mojang.blaze3d.vertex.PoseStack;
 import net.minecraft.client.renderer.MultiBufferSource;
 import net.minecraft.world.level.block.entity.BlockEntity;
+import net.minecraft.world.phys.Vec3;
 
-public abstract class BlockEntityRenderer<T extends BlockEntity> {
-   protected final BlockEntityRenderDispatcher renderer;
+public interface BlockEntityRenderer<T extends BlockEntity> {
+   void render(T var1, float var2, PoseStack var3, MultiBufferSource var4, int var5, int var6);
 
-   public BlockEntityRenderer(BlockEntityRenderDispatcher var1) {
-      this.renderer = var1;
+   default boolean shouldRenderOffScreen(T var1) {
+      return false;
    }
 
-   public abstract void render(T var1, float var2, PoseStack var3, MultiBufferSource var4, int var5, int var6);
+   default int getViewDistance() {
+      return 64;
+   }
 
-   public boolean shouldRenderOffScreen(T var1) {
-      return false;
+   default boolean shouldRender(T var1, Vec3 var2) {
+      return Vec3.atCenterOf(var1.getBlockPos()).closerThan(var2, (double)this.getViewDistance());
    }
 }
