--- net/minecraft/client/renderer/blockentity/CampfireRenderer.java
+++ net/minecraft/client/renderer/blockentity/CampfireRenderer.java
@@ -9,13 +9,12 @@
 import net.minecraft.core.NonNullList;
 import net.minecraft.world.item.ItemStack;
 import net.minecraft.world.level.block.CampfireBlock;
 import net.minecraft.world.level.block.entity.CampfireBlockEntity;
 
-public class CampfireRenderer extends BlockEntityRenderer<CampfireBlockEntity> {
-   public CampfireRenderer(BlockEntityRenderDispatcher var1) {
-      super(var1);
+public class CampfireRenderer implements BlockEntityRenderer<CampfireBlockEntity> {
+   public CampfireRenderer(BlockEntityRendererProvider.Context var1) {
    }
 
    public void render(CampfireBlockEntity var1, float var2, PoseStack var3, MultiBufferSource var4, int var5, int var6) {
       Direction var7 = (Direction)var1.getBlockState().getValue(CampfireBlock.FACING);
       NonNullList var8 = var1.getItems();
