--- net/minecraft/client/renderer/blockentity/CampfireRenderer.java
+++ net/minecraft/client/renderer/blockentity/CampfireRenderer.java
@@ -9,31 +9,33 @@
 import net.minecraft.core.NonNullList;
 import net.minecraft.world.item.ItemStack;
 import net.minecraft.world.level.block.CampfireBlock;
 import net.minecraft.world.level.block.entity.CampfireBlockEntity;
 
-public class CampfireRenderer extends BlockEntityRenderer<CampfireBlockEntity> {
-   public CampfireRenderer(BlockEntityRenderDispatcher var1) {
-      super(var1);
+public class CampfireRenderer implements BlockEntityRenderer<CampfireBlockEntity> {
+   private static final float SIZE = 0.375F;
+
+   public CampfireRenderer(BlockEntityRendererProvider.Context var1) {
    }
 
    public void render(CampfireBlockEntity var1, float var2, PoseStack var3, MultiBufferSource var4, int var5, int var6) {
       Direction var7 = (Direction)var1.getBlockState().getValue(CampfireBlock.FACING);
       NonNullList var8 = var1.getItems();
+      int var9 = (int)var1.getBlockPos().asLong();
 
-      for(int var9 = 0; var9 < var8.size(); ++var9) {
-         ItemStack var10 = (ItemStack)var8.get(var9);
-         if (var10 != ItemStack.EMPTY) {
+      for(int var10 = 0; var10 < var8.size(); ++var10) {
+         ItemStack var11 = (ItemStack)var8.get(var10);
+         if (var11 != ItemStack.EMPTY) {
             var3.pushPose();
             var3.translate(0.5D, 0.44921875D, 0.5D);
-            Direction var11 = Direction.from2DDataValue((var9 + var7.get2DDataValue()) % 4);
-            float var12 = -var11.toYRot();
-            var3.mulPose(Vector3f.YP.rotationDegrees(var12));
+            Direction var12 = Direction.from2DDataValue((var10 + var7.get2DDataValue()) % 4);
+            float var13 = -var12.toYRot();
+            var3.mulPose(Vector3f.YP.rotationDegrees(var13));
             var3.mulPose(Vector3f.XP.rotationDegrees(90.0F));
             var3.translate(-0.3125D, -0.3125D, 0.0D);
             var3.scale(0.375F, 0.375F, 0.375F);
-            Minecraft.getInstance().getItemRenderer().renderStatic(var10, ItemTransforms.TransformType.FIXED, var5, var6, var3, var4);
+            Minecraft.getInstance().getItemRenderer().renderStatic(var11, ItemTransforms.TransformType.FIXED, var5, var6, var3, var4, var9 + var10);
             var3.popPose();
          }
       }
 
    }
