--- net/minecraft/client/renderer/blockentity/LecternRenderer.java
+++ net/minecraft/client/renderer/blockentity/LecternRenderer.java
@@ -2,22 +2,23 @@
 
 import com.mojang.blaze3d.vertex.PoseStack;
 import com.mojang.blaze3d.vertex.VertexConsumer;
 import com.mojang.math.Vector3f;
 import net.minecraft.client.model.BookModel;
+import net.minecraft.client.model.geom.ModelLayers;
 import net.minecraft.client.renderer.MultiBufferSource;
 import net.minecraft.client.renderer.RenderType;
 import net.minecraft.core.Direction;
 import net.minecraft.world.level.block.LecternBlock;
 import net.minecraft.world.level.block.entity.LecternBlockEntity;
 import net.minecraft.world.level.block.state.BlockState;
 
-public class LecternRenderer extends BlockEntityRenderer<LecternBlockEntity> {
-   private final BookModel bookModel = new BookModel();
+public class LecternRenderer implements BlockEntityRenderer<LecternBlockEntity> {
+   private final BookModel bookModel;
 
-   public LecternRenderer(BlockEntityRenderDispatcher var1) {
-      super(var1);
+   public LecternRenderer(BlockEntityRendererProvider.Context var1) {
+      this.bookModel = new BookModel(var1.bakeLayer(ModelLayers.BOOK));
    }
 
    public void render(LecternBlockEntity var1, float var2, PoseStack var3, MultiBufferSource var4, int var5, int var6) {
       BlockState var7 = var1.getBlockState();
       if ((Boolean)var7.getValue(LecternBlock.HAS_BOOK)) {
