--- net/minecraft/client/renderer/blockentity/PistonHeadRenderer.java
+++ net/minecraft/client/renderer/blockentity/PistonHeadRenderer.java
@@ -1,11 +1,10 @@
 package net.minecraft.client.renderer.blockentity;
 
 import com.mojang.blaze3d.vertex.PoseStack;
 import com.mojang.blaze3d.vertex.VertexConsumer;
 import java.util.Random;
-import net.minecraft.client.Minecraft;
 import net.minecraft.client.renderer.ItemBlockRenderTypes;
 import net.minecraft.client.renderer.MultiBufferSource;
 import net.minecraft.client.renderer.RenderType;
 import net.minecraft.client.renderer.block.BlockRenderDispatcher;
 import net.minecraft.client.renderer.block.ModelBlockRenderer;
@@ -16,15 +15,15 @@
 import net.minecraft.world.level.block.piston.PistonHeadBlock;
 import net.minecraft.world.level.block.piston.PistonMovingBlockEntity;
 import net.minecraft.world.level.block.state.BlockState;
 import net.minecraft.world.level.block.state.properties.PistonType;
 
-public class PistonHeadRenderer extends BlockEntityRenderer<PistonMovingBlockEntity> {
-   private final BlockRenderDispatcher blockRenderer = Minecraft.getInstance().getBlockRenderer();
+public class PistonHeadRenderer implements BlockEntityRenderer<PistonMovingBlockEntity> {
+   private final BlockRenderDispatcher blockRenderer;
 
-   public PistonHeadRenderer(BlockEntityRenderDispatcher var1) {
-      super(var1);
+   public PistonHeadRenderer(BlockEntityRendererProvider.Context var1) {
+      this.blockRenderer = var1.getBlockRenderDispatcher();
    }
 
    public void render(PistonMovingBlockEntity var1, float var2, PoseStack var3, MultiBufferSource var4, int var5, int var6) {
       Level var7 = var1.getLevel();
       if (var7 != null) {
@@ -60,6 +59,10 @@
    private void renderBlock(BlockPos var1, BlockState var2, PoseStack var3, MultiBufferSource var4, Level var5, boolean var6, int var7) {
       RenderType var8 = ItemBlockRenderTypes.getMovingBlockRenderType(var2);
       VertexConsumer var9 = var4.getBuffer(var8);
       this.blockRenderer.getModelRenderer().tesselateBlock(var5, this.blockRenderer.getBlockModel(var2), var2, var1, var3, var9, var6, new Random(), var2.getSeed(var1), var7);
    }
+
+   public int getViewDistance() {
+      return 68;
+   }
 }
