--- net/minecraft/client/renderer/blockentity/StructureBlockRenderer.java
+++ net/minecraft/client/renderer/blockentity/StructureBlockRenderer.java
@@ -12,13 +12,12 @@
 import net.minecraft.world.level.block.Blocks;
 import net.minecraft.world.level.block.entity.StructureBlockEntity;
 import net.minecraft.world.level.block.state.BlockState;
 import net.minecraft.world.level.block.state.properties.StructureMode;
 
-public class StructureBlockRenderer extends BlockEntityRenderer<StructureBlockEntity> {
-   public StructureBlockRenderer(BlockEntityRenderDispatcher var1) {
-      super(var1);
+public class StructureBlockRenderer implements BlockEntityRenderer<StructureBlockEntity> {
+   public StructureBlockRenderer(BlockEntityRendererProvider.Context var1) {
    }
 
    public void render(StructureBlockEntity var1, float var2, PoseStack var3, MultiBufferSource var4, int var5, int var6) {
       if (Minecraft.getInstance().player.canUseGameMasterBlocks() || Minecraft.getInstance().player.isSpectator()) {
          BlockPos var7 = var1.getStructurePos();
@@ -132,6 +131,10 @@
    }
 
    public boolean shouldRenderOffScreen(StructureBlockEntity var1) {
       return true;
    }
+
+   public int getViewDistance() {
+      return 96;
+   }
 }
