--- net/minecraft/client/renderer/blockentity/StructureBlockRenderer.java
+++ net/minecraft/client/renderer/blockentity/StructureBlockRenderer.java
@@ -6,25 +6,25 @@
 import net.minecraft.client.Minecraft;
 import net.minecraft.client.renderer.LevelRenderer;
 import net.minecraft.client.renderer.MultiBufferSource;
 import net.minecraft.client.renderer.RenderType;
 import net.minecraft.core.BlockPos;
+import net.minecraft.core.Vec3i;
 import net.minecraft.world.level.Level;
 import net.minecraft.world.level.block.Blocks;
 import net.minecraft.world.level.block.entity.StructureBlockEntity;
 import net.minecraft.world.level.block.state.BlockState;
 import net.minecraft.world.level.block.state.properties.StructureMode;
 
-public class StructureBlockRenderer extends BlockEntityRenderer<StructureBlockEntity> {
-   public StructureBlockRenderer(BlockEntityRenderDispatcher var1) {
-      super(var1);
+public class StructureBlockRenderer implements BlockEntityRenderer<StructureBlockEntity> {
+   public StructureBlockRenderer(BlockEntityRendererProvider.Context var1) {
    }
 
    public void render(StructureBlockEntity var1, float var2, PoseStack var3, MultiBufferSource var4, int var5, int var6) {
       if (Minecraft.getInstance().player.canUseGameMasterBlocks() || Minecraft.getInstance().player.isSpectator()) {
          BlockPos var7 = var1.getStructurePos();
-         BlockPos var8 = var1.getStructureSize();
+         Vec3i var8 = var1.getStructureSize();
          if (var8.getX() >= 1 && var8.getY() >= 1 && var8.getZ() >= 1) {
             if (var1.getMode() == StructureMode.SAVE || var1.getMode() == StructureMode.LOAD) {
                double var9 = (double)var7.getX();
                double var11 = (double)var7.getZ();
                double var19 = (double)var7.getY();
@@ -132,6 +132,10 @@
    }
 
    public boolean shouldRenderOffScreen(StructureBlockEntity var1) {
       return true;
    }
+
+   public int getViewDistance() {
+      return 96;
+   }
 }
