--- net/minecraft/client/renderer/blockentity/TheEndGatewayRenderer.java
+++ net/minecraft/client/renderer/blockentity/TheEndGatewayRenderer.java
@@ -9,24 +9,23 @@
 import net.minecraft.world.level.block.entity.TheEndPortalBlockEntity;
 
 public class TheEndGatewayRenderer extends TheEndPortalRenderer<TheEndGatewayBlockEntity> {
    private static final ResourceLocation BEAM_LOCATION = new ResourceLocation("textures/entity/end_gateway_beam.png");
 
-   public TheEndGatewayRenderer(BlockEntityRenderDispatcher var1) {
+   public TheEndGatewayRenderer(BlockEntityRendererProvider.Context var1) {
       super(var1);
    }
 
    public void render(TheEndGatewayBlockEntity var1, float var2, PoseStack var3, MultiBufferSource var4, int var5, int var6) {
       if (var1.isSpawning() || var1.isCoolingDown()) {
          float var7 = var1.isSpawning() ? var1.getSpawnPercent(var2) : var1.getCooldownPercent(var2);
-         double var8 = var1.isSpawning() ? 256.0D : 50.0D;
+         double var8 = var1.isSpawning() ? (double)var1.getLevel().getMaxBuildHeight() : 50.0D;
          var7 = Mth.sin(var7 * 3.1415927F);
          int var10 = Mth.floor((double)var7 * var8);
          float[] var11 = var1.isSpawning() ? DyeColor.MAGENTA.getTextureDiffuseColors() : DyeColor.PURPLE.getTextureDiffuseColors();
          long var12 = var1.getLevel().getGameTime();
-         BeaconRenderer.renderBeaconBeam(var3, var4, BEAM_LOCATION, var2, var7, var12, 0, var10, var11, 0.15F, 0.175F);
-         BeaconRenderer.renderBeaconBeam(var3, var4, BEAM_LOCATION, var2, var7, var12, 0, -var10, var11, 0.15F, 0.175F);
+         BeaconRenderer.renderBeaconBeam(var3, var4, BEAM_LOCATION, var2, var7, var12, -var10, var10 * 2, var11, 0.15F, 0.175F);
       }
 
       super.render((TheEndPortalBlockEntity)var1, var2, var3, var4, var5, var6);
    }
 
