--- net/minecraft/commands/arguments/selector/options/EntitySelectorOptions.java
+++ net/minecraft/commands/arguments/selector/options/EntitySelectorOptions.java
@@ -250,11 +250,11 @@
                GameType[] var6 = GameType.values();
                int var7 = var6.length;
 
                for(int var8 = 0; var8 < var7; ++var8) {
                   GameType var9 = var6[var8];
-                  if (var9 != GameType.NOT_SET && var9.getName().toLowerCase(Locale.ROOT).startsWith(var3)) {
+                  if (var9.getName().toLowerCase(Locale.ROOT).startsWith(var3)) {
                      if (var5) {
                         var1x.suggest('!' + var9.getName());
                      }
 
                      if (var4) {
@@ -270,12 +270,12 @@
             if (var0.hasGamemodeNotEquals() && !var2) {
                var0.getReader().setCursor(var1);
                throw ERROR_INAPPLICABLE_OPTION.createWithContext(var0.getReader(), "gamemode");
             } else {
                String var3 = var0.getReader().readUnquotedString();
-               GameType var4 = GameType.byName(var3, GameType.NOT_SET);
-               if (var4 == GameType.NOT_SET) {
+               GameType var4 = GameType.byName(var3, (GameType)null);
+               if (var4 == null) {
                   var0.getReader().setCursor(var1);
                   throw ERROR_GAME_MODE_INVALID.createWithContext(var0.getReader(), var3);
                } else {
                   var0.setIncludesEntities(false);
                   var0.addPredicate((var2x) -> {
@@ -341,11 +341,11 @@
 
                ResourceLocation var3;
                if (var0.isTag()) {
                   var3 = ResourceLocation.read(var0.getReader());
                   var0.addPredicate((var2x) -> {
-                     return var2x.getServer().getTags().getEntityTypes().getTagOrEmpty(var3).contains(var2x.getType()) != var2;
+                     return var2x.getType().is(var2x.getServer().getTags().getOrEmpty(Registry.ENTITY_TYPE_REGISTRY).getTagOrEmpty(var3)) != var2;
                   });
                } else {
                   var3 = ResourceLocation.read(var0.getReader());
                   EntityType var4 = (EntityType)Registry.ENTITY_TYPE.getOptional(var3).orElseThrow(() -> {
                      var0.getReader().setCursor(var1);
@@ -384,11 +384,11 @@
             boolean var1 = var0.shouldInvertValue();
             CompoundTag var2 = (new TagParser(var0.getReader())).readStruct();
             var0.addPredicate((var2x) -> {
                CompoundTag var3 = var2x.saveWithoutId(new CompoundTag());
                if (var2x instanceof ServerPlayer) {
-                  ItemStack var4 = ((ServerPlayer)var2x).inventory.getSelected();
+                  ItemStack var4 = ((ServerPlayer)var2x).getInventory().getSelected();
                   if (!var4.isEmpty()) {
                      var3.put("SelectedItem", var4.save(new CompoundTag()));
                   }
                }
 
