--- com/mojang/blaze3d/audio/Channel.java
+++ com/mojang/blaze3d/audio/Channel.java
@@ -11,10 +11,12 @@
 import org.apache.logging.log4j.Logger;
 import org.lwjgl.openal.AL10;
 
 public class Channel {
    private static final Logger LOGGER = LogManager.getLogger();
+   private static final int QUEUED_BUFFER_COUNT = 4;
+   public static final int BUFFER_DURATION_SECONDS = 1;
    private final int source;
    private final AtomicBoolean initialized = new AtomicBoolean(true);
    private int streamingBufferSize = 16384;
    @Nullable
    private AudioStream stream;
@@ -79,10 +81,14 @@
          OpenAlUtil.checkALError("Stop");
       }
 
    }
 
+   public boolean playing() {
+      return this.getState() == 4114;
+   }
+
    public boolean stopped() {
       return this.getState() == 4116;
    }
 
    public void setSelfPosition(Vec3 var1) {
