--- com/mojang/blaze3d/audio/Library.java
+++ com/mojang/blaze3d/audio/Library.java
@@ -88,33 +88,28 @@
 
       try {
          int var3 = ALC10.alcGetInteger(this.device, 4098);
          if (OpenAlUtil.checkALCError(this.device, "Get attributes size")) {
             throw new IllegalStateException("Failed to get OpenAL attributes");
-         } else {
-            IntBuffer var4 = var1.mallocInt(var3);
-            ALC10.alcGetIntegerv(this.device, 4099, var4);
-            if (OpenAlUtil.checkALCError(this.device, "Get attributes")) {
-               throw new IllegalStateException("Failed to get OpenAL attributes");
-            } else {
-               int var5 = 0;
+         }
 
-               int var6;
-               int var7;
-               do {
-                  if (var5 >= var3) {
-                     return 30;
-                  }
+         IntBuffer var4 = var1.mallocInt(var3);
+         ALC10.alcGetIntegerv(this.device, 4099, var4);
+         if (OpenAlUtil.checkALCError(this.device, "Get attributes")) {
+            throw new IllegalStateException("Failed to get OpenAL attributes");
+         }
 
-                  var6 = var4.get(var5++);
-                  if (var6 == 0) {
-                     return 30;
-                  }
+         int var5 = 0;
 
-                  var7 = var4.get(var5++);
-               } while(var6 != 4112);
+         while(var5 < var3) {
+            int var6 = var4.get(var5++);
+            if (var6 == 0) {
+               break;
+            }
 
+            int var7 = var4.get(var5++);
+            if (var6 == 4112) {
                int var8 = var7;
                return var8;
             }
          }
       } catch (Throwable var18) {
@@ -132,10 +127,12 @@
                var1.close();
             }
          }
 
       }
+
+      return 30;
    }
 
    private static long tryOpenDevice() {
       for(int var0 = 0; var0 < 3; ++var0) {
          long var1 = ALC10.alcOpenDevice((ByteBuffer)null);
