--- com/mojang/blaze3d/audio/OggAudioStream.java
+++ com/mojang/blaze3d/audio/OggAudioStream.java
@@ -17,10 +17,11 @@
 import org.lwjgl.stb.STBVorbisInfo;
 import org.lwjgl.system.MemoryStack;
 import org.lwjgl.system.MemoryUtil;
 
 public class OggAudioStream implements AudioStream {
+   private static final int EXPECTED_MAX_FRAME_SIZE = 8192;
    private long handle;
    private final AudioFormat audioFormat;
    private final InputStream input;
    private ByteBuffer buffer = MemoryUtil.memAlloc(8192);
 
@@ -246,11 +247,11 @@
             this.currentBuffer.flip();
             this.buffers.add(this.currentBuffer);
             this.createNewBuffer();
          }
 
-         int var2 = Mth.clamp((int)(var1 * 32767.5F - 0.5F), -32768, 32767);
+         int var2 = Mth.clamp((int)((int)(var1 * 32767.5F - 0.5F)), (int)-32768, (int)32767);
          this.currentBuffer.putShort((short)var2);
          this.byteCount += 2;
       }
 
       public ByteBuffer get() {
