--- com/mojang/blaze3d/audio/OggAudioStream.java
+++ com/mojang/blaze3d/audio/OggAudioStream.java
@@ -147,17 +147,17 @@
                         this.convertMono(var11.getFloatBuffer(0, var9), var1);
                         var12 = true;
                         return var12;
                      }
 
-                     if (var10 == 2) {
-                        this.convertStereo(var11.getFloatBuffer(0, var9), var11.getFloatBuffer(1, var9), var1);
-                        var12 = true;
-                        return var12;
+                     if (var10 != 2) {
+                        throw new IllegalStateException("Invalid number of channels: " + var10);
                      }
 
-                     throw new IllegalStateException("Invalid number of channels: " + var10);
+                     this.convertStereo(var11.getFloatBuffer(0, var9), var11.getFloatBuffer(1, var9), var1);
+                     var12 = true;
+                     return var12;
                   }
                }
             }
          } catch (Throwable var23) {
             var3 = var23;
