--- com/mojang/blaze3d/audio/OpenAlUtil.java
+++ com/mojang/blaze3d/audio/OpenAlUtil.java
@@ -8,10 +8,13 @@
 import org.lwjgl.openal.ALC10;
 
 public class OpenAlUtil {
    private static final Logger LOGGER = LogManager.getLogger();
 
+   public OpenAlUtil() {
+   }
+
    private static String alErrorToString(int var0) {
       switch(var0) {
       case 40961:
          return "Invalid name parameter.";
       case 40962:
