--- net/minecraft/world/level/storage/loot/providers/score/LootScoreProviderType.java
+++ net/minecraft/world/level/storage/loot/providers/score/LootScoreProviderType.java
@@ -1,0 +1,10 @@
+package net.minecraft.world.level.storage.loot.providers.score;
+
+import net.minecraft.world.level.storage.loot.Serializer;
+import net.minecraft.world.level.storage.loot.SerializerType;
+
+public class LootScoreProviderType extends SerializerType<ScoreboardNameProvider> {
+   public LootScoreProviderType(Serializer<? extends ScoreboardNameProvider> var1) {
+      super(var1);
+   }
+}
