--- com/mojang/blaze3d/pipeline/RenderTarget.java
+++ com/mojang/blaze3d/pipeline/RenderTarget.java
@@ -5,10 +5,11 @@
 import com.mojang.blaze3d.platform.TextureUtil;
 import com.mojang.blaze3d.systems.RenderSystem;
 import com.mojang.blaze3d.vertex.BufferBuilder;
 import com.mojang.blaze3d.vertex.DefaultVertexFormat;
 import com.mojang.blaze3d.vertex.Tesselator;
+import com.mojang.blaze3d.vertex.VertexFormat;
 import java.nio.IntBuffer;
 
 public class RenderTarget {
    public int width;
    public int height;
@@ -252,11 +253,11 @@
       float var5 = (float)var2;
       float var6 = (float)this.viewWidth / (float)this.width;
       float var7 = (float)this.viewHeight / (float)this.height;
       Tesselator var8 = RenderSystem.renderThreadTesselator();
       BufferBuilder var9 = var8.getBuilder();
-      var9.begin(7, DefaultVertexFormat.POSITION_TEX_COLOR);
+      var9.begin(VertexFormat.Mode.QUADS, DefaultVertexFormat.POSITION_TEX_COLOR);
       var9.vertex(0.0D, (double)var5, 0.0D).uv(0.0F, 0.0F).color(255, 255, 255, 255).endVertex();
       var9.vertex((double)var4, (double)var5, 0.0D).uv(var6, 0.0F).color(255, 255, 255, 255).endVertex();
       var9.vertex((double)var4, 0.0D, 0.0D).uv(var6, var7).color(255, 255, 255, 255).endVertex();
       var9.vertex(0.0D, 0.0D, 0.0D).uv(0.0F, var7).color(255, 255, 255, 255).endVertex();
       var8.end();
