--- com/mojang/blaze3d/pipeline/TextureTarget.java
+++ com/mojang/blaze3d/pipeline/TextureTarget.java
@@ -1,0 +1,11 @@
+package com.mojang.blaze3d.pipeline;
+
+import com.mojang.blaze3d.systems.RenderSystem;
+
+public class TextureTarget extends RenderTarget {
+   public TextureTarget(int var1, int var2, boolean var3, boolean var4) {
+      super(var3);
+      RenderSystem.assertThread(RenderSystem::isOnRenderThreadOrInit);
+      this.resize(var1, var2, var4);
+   }
+}
