--- net/minecraft/sounds/Music.java
+++ net/minecraft/sounds/Music.java
@@ -25,6 +25,22 @@
       this.event = var1;
       this.minDelay = var2;
       this.maxDelay = var3;
       this.replaceCurrentMusic = var4;
    }
+
+   public SoundEvent getEvent() {
+      return this.event;
+   }
+
+   public int getMinDelay() {
+      return this.minDelay;
+   }
+
+   public int getMaxDelay() {
+      return this.maxDelay;
+   }
+
+   public boolean replaceCurrentMusic() {
+      return this.replaceCurrentMusic;
+   }
 }
