--- net/minecraft/sounds/Musics.java
+++ net/minecraft/sounds/Musics.java
@@ -1,16 +1,24 @@
 package net.minecraft.sounds;
 
 public class Musics {
+   private static final int ONE_SECOND = 20;
+   private static final int THIRTY_SECONDS = 600;
+   private static final int TEN_MINUTES = 12000;
+   private static final int TWENTY_MINUTES = 24000;
+   private static final int FIVE_MINUTES = 6000;
    public static final Music MENU;
    public static final Music CREATIVE;
    public static final Music CREDITS;
    public static final Music END_BOSS;
    public static final Music END;
    public static final Music UNDER_WATER;
    public static final Music GAME;
 
+   public Musics() {
+   }
+
    public static Music createGameMusic(SoundEvent var0) {
       return new Music(var0, 12000, 24000, false);
    }
 
    static {
