--- net/minecraft/sounds/SoundEvent.java
+++ net/minecraft/sounds/SoundEvent.java
@@ -9,10 +9,14 @@
 
    public SoundEvent(ResourceLocation var1) {
       this.location = var1;
    }
 
+   public ResourceLocation getLocation() {
+      return this.location;
+   }
+
    static {
       CODEC = ResourceLocation.CODEC.xmap(SoundEvent::new, (var0) -> {
          return var0.location;
       });
    }
