--- net/minecraft/sounds/SoundSource.java
+++ net/minecraft/sounds/SoundSource.java
@@ -1,12 +1,7 @@
 package net.minecraft.sounds;
 
-import java.util.Arrays;
-import java.util.Map;
-import java.util.function.Function;
-import java.util.stream.Collectors;
-
 public enum SoundSource {
    MASTER("master"),
    MUSIC("music"),
    RECORDS("record"),
    WEATHER("weather"),
@@ -15,11 +10,10 @@
    NEUTRAL("neutral"),
    PLAYERS("player"),
    AMBIENT("ambient"),
    VOICE("voice");
 
-   private static final Map<String, SoundSource> BY_NAME = (Map)Arrays.stream(values()).collect(Collectors.toMap(SoundSource::getName, Function.identity()));
    private final String name;
 
    private SoundSource(String var3) {
       this.name = var3;
    }
