--- net/minecraft/client/gui/screens/worldselection/CreateWorldScreen.java
+++ net/minecraft/client/gui/screens/worldselection/CreateWorldScreen.java
@@ -34,10 +34,11 @@
 import net.minecraft.network.chat.MutableComponent;
 import net.minecraft.network.chat.TextComponent;
 import net.minecraft.network.chat.TranslatableComponent;
 import net.minecraft.server.MinecraftServer;
 import net.minecraft.server.ServerResources;
+import net.minecraft.server.packs.PackType;
 import net.minecraft.server.packs.repository.FolderRepositorySource;
 import net.minecraft.server.packs.repository.PackRepository;
 import net.minecraft.server.packs.repository.PackSource;
 import net.minecraft.server.packs.repository.RepositorySource;
 import net.minecraft.server.packs.repository.ServerPacksSource;
@@ -626,11 +627,11 @@
    private Pair<File, PackRepository> getDataPackSelectionSettings() {
       Path var1 = this.getTempDataPackDir();
       if (var1 != null) {
          File var2 = var1.toFile();
          if (this.tempDataPackRepository == null) {
-            this.tempDataPackRepository = new PackRepository(new RepositorySource[]{new ServerPacksSource(), new FolderRepositorySource(var2, PackSource.DEFAULT)});
+            this.tempDataPackRepository = new PackRepository(PackType.SERVER_DATA, new RepositorySource[]{new ServerPacksSource(), new FolderRepositorySource(var2, PackSource.DEFAULT)});
             this.tempDataPackRepository.reload();
          }
 
          this.tempDataPackRepository.setSelected(this.dataPacks.getEnabled());
          return Pair.of(var2, this.tempDataPackRepository);
