--- net/minecraft/client/gui/screens/worldselection/EditGameRulesScreen.java
+++ net/minecraft/client/gui/screens/worldselection/EditGameRulesScreen.java
@@ -18,10 +18,11 @@
 import net.minecraft.ChatFormatting;
 import net.minecraft.client.gui.GuiComponent;
 import net.minecraft.client.gui.components.AbstractSelectionList;
 import net.minecraft.client.gui.components.Button;
 import net.minecraft.client.gui.components.ContainerObjectSelectionList;
+import net.minecraft.client.gui.components.CycleButton;
 import net.minecraft.client.gui.components.EditBox;
 import net.minecraft.client.gui.components.events.GuiEventListener;
 import net.minecraft.client.gui.screens.Screen;
 import net.minecraft.client.resources.language.I18n;
 import net.minecraft.network.chat.CommonComponents;
@@ -187,23 +188,19 @@
          this.input.render(var1, var7, var8, var10);
       }
    }
 
    public class BooleanRuleEntry extends EditGameRulesScreen.GameRuleEntry {
-      private final Button checkbox;
+      private final CycleButton<Boolean> checkbox;
 
-      public BooleanRuleEntry(final Component var2, List<FormattedCharSequence> var3, final String var4, final GameRules.BooleanValue var5) {
+      public BooleanRuleEntry(Component var2, List<FormattedCharSequence> var3, String var4, GameRules.BooleanValue var5) {
          super(var3, var2);
-         this.checkbox = new Button(10, 5, 44, 20, CommonComponents.optionStatus(var5.get()), (var1x) -> {
-            boolean var2 = !var5.get();
-            var5.set(var2, (MinecraftServer)null);
-            var1x.setMessage(CommonComponents.optionStatus(var5.get()));
-         }) {
-            protected MutableComponent createNarrationMessage() {
-               return CommonComponents.optionStatus(var2, var5.get()).append("\n").append(var4);
-            }
-         };
+         this.checkbox = CycleButton.onOffBuilder(var5.get()).displayOnlyValue().withCustomNarration((var1x) -> {
+            return var1x.createDefaultNarrationMessage().append("\n").append(var4);
+         }).create(10, 5, 44, 20, var2, (var1x, var2x) -> {
+            var5.set(var2x, (MinecraftServer)null);
+         });
          this.children.add(this.checkbox);
       }
 
       public void render(PoseStack var1, int var2, int var3, int var4, int var5, int var6, int var7, int var8, boolean var9, float var10) {
          this.renderLabel(var1, var3, var4);
