--- net/minecraft/client/gui/screens/worldselection/SelectWorldScreen.java
+++ net/minecraft/client/gui/screens/worldselection/SelectWorldScreen.java
@@ -6,12 +6,15 @@
 import net.minecraft.client.gui.components.EditBox;
 import net.minecraft.client.gui.screens.Screen;
 import net.minecraft.network.chat.CommonComponents;
 import net.minecraft.network.chat.TranslatableComponent;
 import net.minecraft.util.FormattedCharSequence;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
 
 public class SelectWorldScreen extends Screen {
+   private static final Logger LOGGER = LogManager.getLogger();
    protected final Screen lastScreen;
    private List<FormattedCharSequence> toolTip;
    private Button deleteButton;
    private Button selectButton;
    private Button renameButton;
