--- net/minecraft/client/gui/screens/worldselection/SelectWorldScreen.java
+++ net/minecraft/client/gui/screens/worldselection/SelectWorldScreen.java
@@ -6,12 +6,15 @@
 import net.minecraft.client.gui.components.EditBox;
 import net.minecraft.client.gui.screens.Screen;
 import net.minecraft.network.chat.CommonComponents;
 import net.minecraft.network.chat.TranslatableComponent;
 import net.minecraft.util.FormattedCharSequence;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
 
 public class SelectWorldScreen extends Screen {
+   private static final Logger LOGGER = LogManager.getLogger();
    protected final Screen lastScreen;
    private List<FormattedCharSequence> toolTip;
    private Button deleteButton;
    private Button selectButton;
    private Button renameButton;
@@ -41,28 +44,28 @@
          }, false);
       });
       this.list = new WorldSelectionList(this, this.minecraft, this.width, this.height, 48, this.height - 64, 36, () -> {
          return this.searchBox.getValue();
       }, this.list);
-      this.children.add(this.searchBox);
-      this.children.add(this.list);
-      this.selectButton = (Button)this.addButton(new Button(this.width / 2 - 154, this.height - 52, 150, 20, new TranslatableComponent("selectWorld.select"), (var1) -> {
+      this.addWidget(this.searchBox);
+      this.addWidget(this.list);
+      this.selectButton = (Button)this.addRenderableWidget(new Button(this.width / 2 - 154, this.height - 52, 150, 20, new TranslatableComponent("selectWorld.select"), (var1) -> {
          this.list.getSelectedOpt().ifPresent(WorldSelectionList.WorldListEntry::joinWorld);
       }));
-      this.addButton(new Button(this.width / 2 + 4, this.height - 52, 150, 20, new TranslatableComponent("selectWorld.create"), (var1) -> {
+      this.addRenderableWidget(new Button(this.width / 2 + 4, this.height - 52, 150, 20, new TranslatableComponent("selectWorld.create"), (var1) -> {
          this.minecraft.setScreen(CreateWorldScreen.create(this));
       }));
-      this.renameButton = (Button)this.addButton(new Button(this.width / 2 - 154, this.height - 28, 72, 20, new TranslatableComponent("selectWorld.edit"), (var1) -> {
+      this.renameButton = (Button)this.addRenderableWidget(new Button(this.width / 2 - 154, this.height - 28, 72, 20, new TranslatableComponent("selectWorld.edit"), (var1) -> {
          this.list.getSelectedOpt().ifPresent(WorldSelectionList.WorldListEntry::editWorld);
       }));
-      this.deleteButton = (Button)this.addButton(new Button(this.width / 2 - 76, this.height - 28, 72, 20, new TranslatableComponent("selectWorld.delete"), (var1) -> {
+      this.deleteButton = (Button)this.addRenderableWidget(new Button(this.width / 2 - 76, this.height - 28, 72, 20, new TranslatableComponent("selectWorld.delete"), (var1) -> {
          this.list.getSelectedOpt().ifPresent(WorldSelectionList.WorldListEntry::deleteWorld);
       }));
-      this.copyButton = (Button)this.addButton(new Button(this.width / 2 + 4, this.height - 28, 72, 20, new TranslatableComponent("selectWorld.recreate"), (var1) -> {
+      this.copyButton = (Button)this.addRenderableWidget(new Button(this.width / 2 + 4, this.height - 28, 72, 20, new TranslatableComponent("selectWorld.recreate"), (var1) -> {
          this.list.getSelectedOpt().ifPresent(WorldSelectionList.WorldListEntry::recreateWorld);
       }));
-      this.addButton(new Button(this.width / 2 + 82, this.height - 28, 72, 20, CommonComponents.GUI_CANCEL, (var1) -> {
+      this.addRenderableWidget(new Button(this.width / 2 + 82, this.height - 28, 72, 20, CommonComponents.GUI_CANCEL, (var1) -> {
          this.minecraft.setScreen(this.lastScreen);
       }));
       this.updateButtonStatus(false);
       this.setInitialFocus(this.searchBox);
    }
