--- net/minecraft/client/gui/screens/worldselection/WorldSelectionList.java
+++ net/minecraft/client/gui/screens/worldselection/WorldSelectionList.java
@@ -15,30 +15,31 @@
 import java.util.Collections;
 import java.util.Date;
 import java.util.Iterator;
 import java.util.List;
 import java.util.Locale;
+import java.util.Objects;
 import java.util.Optional;
 import java.util.function.Supplier;
 import javax.annotation.Nullable;
 import net.minecraft.ChatFormatting;
 import net.minecraft.SharedConstants;
 import net.minecraft.Util;
 import net.minecraft.client.Minecraft;
 import net.minecraft.client.gui.Font;
 import net.minecraft.client.gui.GuiComponent;
-import net.minecraft.client.gui.chat.NarratorChatListener;
 import net.minecraft.client.gui.components.AbstractSelectionList;
 import net.minecraft.client.gui.components.ObjectSelectionList;
 import net.minecraft.client.gui.components.toasts.SystemToast;
 import net.minecraft.client.gui.screens.AlertScreen;
 import net.minecraft.client.gui.screens.BackupConfirmScreen;
 import net.minecraft.client.gui.screens.ConfirmScreen;
 import net.minecraft.client.gui.screens.ErrorScreen;
 import net.minecraft.client.gui.screens.GenericDirtMessageScreen;
 import net.minecraft.client.gui.screens.ProgressScreen;
 import net.minecraft.client.gui.screens.Screen;
+import net.minecraft.client.renderer.GameRenderer;
 import net.minecraft.client.renderer.texture.AbstractTexture;
 import net.minecraft.client.renderer.texture.DynamicTexture;
 import net.minecraft.client.resources.language.I18n;
 import net.minecraft.client.resources.sounds.SimpleSoundInstance;
 import net.minecraft.core.RegistryAccess;
@@ -59,19 +60,20 @@
 import org.apache.commons.lang3.Validate;
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
 
 public class WorldSelectionList extends ObjectSelectionList<WorldSelectionList.WorldListEntry> {
-   private static final Logger LOGGER = LogManager.getLogger();
-   private static final DateFormat DATE_FORMAT = new SimpleDateFormat();
-   private static final ResourceLocation ICON_MISSING = new ResourceLocation("textures/misc/unknown_server.png");
-   private static final ResourceLocation ICON_OVERLAY_LOCATION = new ResourceLocation("textures/gui/world_selection.png");
-   private static final Component FROM_NEWER_TOOLTIP_1;
-   private static final Component FROM_NEWER_TOOLTIP_2;
-   private static final Component SNAPSHOT_TOOLTIP_1;
-   private static final Component SNAPSHOT_TOOLTIP_2;
-   private static final Component WORLD_LOCKED_TOOLTIP;
+   static final Logger LOGGER = LogManager.getLogger();
+   static final DateFormat DATE_FORMAT = new SimpleDateFormat();
+   static final ResourceLocation ICON_MISSING = new ResourceLocation("textures/misc/unknown_server.png");
+   static final ResourceLocation ICON_OVERLAY_LOCATION = new ResourceLocation("textures/gui/world_selection.png");
+   static final Component FROM_NEWER_TOOLTIP_1;
+   static final Component FROM_NEWER_TOOLTIP_2;
+   static final Component SNAPSHOT_TOOLTIP_1;
+   static final Component SNAPSHOT_TOOLTIP_2;
+   static final Component WORLD_LOCKED_TOOLTIP;
+   static final Component WORLD_PRE_WORLDHEIGHT_TOOLTIP;
    private final SelectWorldScreen screen;
    @Nullable
    private List<LevelSummary> cachedList;
 
    public WorldSelectionList(SelectWorldScreen var1, Minecraft var2, int var3, int var4, int var5, int var6, int var7, Supplier<String> var8, @Nullable WorldSelectionList var9) {
@@ -132,26 +134,21 @@
       return this.screen.getFocused() == this;
    }
 
    public void setSelected(@Nullable WorldSelectionList.WorldListEntry var1) {
       super.setSelected(var1);
-      if (var1 != null) {
-         LevelSummary var2 = var1.summary;
-         NarratorChatListener.INSTANCE.sayNow((new TranslatableComponent("narrator.select", new Object[]{new TranslatableComponent("narrator.select.world", new Object[]{var2.getLevelName(), new Date(var2.getLastPlayed()), var2.isHardcore() ? new TranslatableComponent("gameMode.hardcore") : new TranslatableComponent("gameMode." + var2.getGameMode().getName()), var2.hasCheats() ? new TranslatableComponent("selectWorld.cheats") : TextComponent.EMPTY, var2.getWorldVersionName()})})).getString());
-      }
-
-      this.screen.updateButtonStatus(var1 != null && !var1.summary.isLocked());
+      this.screen.updateButtonStatus(var1 != null && !var1.summary.isDisabled());
    }
 
    protected void moveSelection(AbstractSelectionList.SelectionDirection var1) {
       this.moveSelection(var1, (var0) -> {
-         return !var0.summary.isLocked();
+         return !var0.summary.isDisabled();
       });
    }
 
    public Optional<WorldSelectionList.WorldListEntry> getSelectedOpt() {
-      return Optional.ofNullable(this.getSelected());
+      return Optional.ofNullable((WorldSelectionList.WorldListEntry)this.getSelected());
    }
 
    public SelectWorldScreen getScreen() {
       return this.screen;
    }
@@ -160,16 +157,25 @@
       FROM_NEWER_TOOLTIP_1 = (new TranslatableComponent("selectWorld.tooltip.fromNewerVersion1")).withStyle(ChatFormatting.RED);
       FROM_NEWER_TOOLTIP_2 = (new TranslatableComponent("selectWorld.tooltip.fromNewerVersion2")).withStyle(ChatFormatting.RED);
       SNAPSHOT_TOOLTIP_1 = (new TranslatableComponent("selectWorld.tooltip.snapshot1")).withStyle(ChatFormatting.GOLD);
       SNAPSHOT_TOOLTIP_2 = (new TranslatableComponent("selectWorld.tooltip.snapshot2")).withStyle(ChatFormatting.GOLD);
       WORLD_LOCKED_TOOLTIP = (new TranslatableComponent("selectWorld.locked")).withStyle(ChatFormatting.RED);
+      WORLD_PRE_WORLDHEIGHT_TOOLTIP = (new TranslatableComponent("selectWorld.pre_worldheight")).withStyle(ChatFormatting.RED);
    }
 
    public final class WorldListEntry extends ObjectSelectionList.Entry<WorldSelectionList.WorldListEntry> implements AutoCloseable {
+      private static final int ICON_WIDTH = 32;
+      private static final int ICON_HEIGHT = 32;
+      private static final int ICON_OVERLAY_X_JOIN = 0;
+      private static final int ICON_OVERLAY_X_JOIN_WITH_NOTIFY = 32;
+      private static final int ICON_OVERLAY_X_WARNING = 64;
+      private static final int ICON_OVERLAY_X_ERROR = 96;
+      private static final int ICON_OVERLAY_Y_UNSELECTED = 0;
+      private static final int ICON_OVERLAY_Y_SELECTED = 32;
       private final Minecraft minecraft;
       private final SelectWorldScreen screen;
-      private final LevelSummary summary;
+      final LevelSummary summary;
       private final ResourceLocation iconLocation;
       private File iconFile;
       @Nullable
       private final DynamicTexture icon;
       private long lastClickTime;
@@ -177,55 +183,79 @@
       public WorldListEntry(WorldSelectionList var2, LevelSummary var3) {
          this.screen = var2.getScreen();
          this.summary = var3;
          this.minecraft = Minecraft.getInstance();
          String var4 = var3.getLevelId();
-         this.iconLocation = new ResourceLocation("minecraft", "worlds/" + Util.sanitizeName(var4, ResourceLocation::validPathChar) + "/" + Hashing.sha1().hashUnencodedChars(var4) + "/icon");
+         String var10004 = Util.sanitizeName(var4, ResourceLocation::validPathChar);
+         this.iconLocation = new ResourceLocation("minecraft", "worlds/" + var10004 + "/" + Hashing.sha1().hashUnencodedChars(var4) + "/icon");
          this.iconFile = var3.getIcon();
          if (!this.iconFile.isFile()) {
             this.iconFile = null;
          }
 
          this.icon = this.loadServerIcon();
       }
 
+      public Component getNarration() {
+         TranslatableComponent var1 = new TranslatableComponent("narrator.select.world", new Object[]{this.summary.getLevelName(), new Date(this.summary.getLastPlayed()), this.summary.isHardcore() ? new TranslatableComponent("gameMode.hardcore") : new TranslatableComponent("gameMode." + this.summary.getGameMode().getName()), this.summary.hasCheats() ? new TranslatableComponent("selectWorld.cheats") : TextComponent.EMPTY, this.summary.getWorldVersionName()});
+         Object var2;
+         if (this.summary.isLocked()) {
+            var2 = CommonComponents.joinForNarration(var1, WorldSelectionList.WORLD_LOCKED_TOOLTIP);
+         } else if (this.summary.isIncompatibleWorldHeight()) {
+            var2 = CommonComponents.joinForNarration(var1, WorldSelectionList.WORLD_PRE_WORLDHEIGHT_TOOLTIP);
+         } else {
+            var2 = var1;
+         }
+
+         return new TranslatableComponent("narrator.select", new Object[]{var2});
+      }
+
       public void render(PoseStack var1, int var2, int var3, int var4, int var5, int var6, int var7, int var8, boolean var9, float var10) {
          String var11 = this.summary.getLevelName();
-         String var12 = this.summary.getLevelId() + " (" + WorldSelectionList.DATE_FORMAT.format(new Date(this.summary.getLastPlayed())) + ")";
+         String var10000 = this.summary.getLevelId();
+         String var12 = var10000 + " (" + WorldSelectionList.DATE_FORMAT.format(new Date(this.summary.getLastPlayed())) + ")";
          if (StringUtils.isEmpty(var11)) {
-            var11 = I18n.get("selectWorld.world") + " " + (var2 + 1);
+            var10000 = I18n.get("selectWorld.world");
+            var11 = var10000 + " " + (var2 + 1);
          }
 
          Component var13 = this.summary.getInfo();
          this.minecraft.font.draw(var1, var11, (float)(var4 + 32 + 3), (float)(var3 + 1), 16777215);
-         Font var10000 = this.minecraft.font;
+         Font var17 = this.minecraft.font;
          float var10003 = (float)(var4 + 32 + 3);
-         this.minecraft.font.getClass();
-         var10000.draw(var1, var12, var10003, (float)(var3 + 9 + 3), 8421504);
-         var10000 = this.minecraft.font;
+         Objects.requireNonNull(this.minecraft.font);
+         var17.draw(var1, var12, var10003, (float)(var3 + 9 + 3), 8421504);
+         var17 = this.minecraft.font;
          var10003 = (float)(var4 + 32 + 3);
-         this.minecraft.font.getClass();
+         Objects.requireNonNull(this.minecraft.font);
          int var10004 = var3 + 9;
-         this.minecraft.font.getClass();
-         var10000.draw(var1, var13, var10003, (float)(var10004 + 9 + 3), 8421504);
-         RenderSystem.color4f(1.0F, 1.0F, 1.0F, 1.0F);
-         this.minecraft.getTextureManager().bind(this.icon != null ? this.iconLocation : WorldSelectionList.ICON_MISSING);
+         Objects.requireNonNull(this.minecraft.font);
+         var17.draw(var1, var13, var10003, (float)(var10004 + 9 + 3), 8421504);
+         RenderSystem.setShader(GameRenderer::getPositionTexShader);
+         RenderSystem.setShaderColor(1.0F, 1.0F, 1.0F, 1.0F);
+         RenderSystem.setShaderTexture(0, this.icon != null ? this.iconLocation : WorldSelectionList.ICON_MISSING);
          RenderSystem.enableBlend();
          GuiComponent.blit(var1, var4, var3, 0.0F, 0.0F, 32, 32, 32, 32);
          RenderSystem.disableBlend();
          if (this.minecraft.options.touchscreen || var9) {
-            this.minecraft.getTextureManager().bind(WorldSelectionList.ICON_OVERLAY_LOCATION);
+            RenderSystem.setShaderTexture(0, WorldSelectionList.ICON_OVERLAY_LOCATION);
             GuiComponent.fill(var1, var4, var3, var4 + 32, var3 + 32, -1601138544);
-            RenderSystem.color4f(1.0F, 1.0F, 1.0F, 1.0F);
+            RenderSystem.setShader(GameRenderer::getPositionTexShader);
+            RenderSystem.setShaderColor(1.0F, 1.0F, 1.0F, 1.0F);
             int var14 = var7 - var4;
             boolean var15 = var14 < 32;
             int var16 = var15 ? 32 : 0;
             if (this.summary.isLocked()) {
                GuiComponent.blit(var1, var4, var3, 96.0F, (float)var16, 32, 32, 256, 256);
                if (var15) {
                   this.screen.setToolTip(this.minecraft.font.split(WorldSelectionList.WORLD_LOCKED_TOOLTIP, 175));
                }
+            } else if (this.summary.isIncompatibleWorldHeight()) {
+               GuiComponent.blit(var1, var4, var3, 96.0F, 32.0F, 32, 32, 256, 256);
+               if (var15) {
+                  this.screen.setToolTip(this.minecraft.font.split(WorldSelectionList.WORLD_PRE_WORLDHEIGHT_TOOLTIP, 175));
+               }
             } else if (this.summary.markVersionInList()) {
                GuiComponent.blit(var1, var4, var3, 32.0F, (float)var16, 32, 32, 256, 256);
                if (this.summary.askToOpenWorld()) {
                   GuiComponent.blit(var1, var4, var3, 96.0F, (float)var16, 32, 32, 256, 256);
                   if (var15) {
@@ -243,11 +273,11 @@
          }
 
       }
 
       public boolean mouseClicked(double var1, double var3, int var5) {
-         if (this.summary.isLocked()) {
+         if (this.summary.isDisabled()) {
             return true;
          } else {
             WorldSelectionList.this.setSelected(this);
             this.screen.updateButtonStatus(WorldSelectionList.this.getSelectedOpt().isPresent());
             if (var1 - (double)WorldSelectionList.this.getRowLeft() <= 32.0D) {
@@ -262,49 +292,53 @@
             }
          }
       }
 
       public void joinWorld() {
-         if (!this.summary.isLocked()) {
-            if (this.summary.shouldBackup()) {
-               TranslatableComponent var1 = new TranslatableComponent("selectWorld.backupQuestion");
-               TranslatableComponent var2 = new TranslatableComponent("selectWorld.backupWarning", new Object[]{this.summary.getWorldVersionName(), SharedConstants.getCurrentVersion().getName()});
+         if (!this.summary.isDisabled()) {
+            LevelSummary.BackupStatus var1 = this.summary.backupStatus();
+            if (var1.shouldBackup()) {
+               String var2 = "selectWorld.backupQuestion." + var1.getTranslationKey();
+               String var3 = "selectWorld.backupWarning." + var1.getTranslationKey();
+               TranslatableComponent var4 = new TranslatableComponent(var2);
+               if (var1.isSevere()) {
+                  var4.withStyle(ChatFormatting.BOLD, ChatFormatting.RED);
+               }
+
+               TranslatableComponent var5 = new TranslatableComponent(var3, new Object[]{this.summary.getWorldVersionName(), SharedConstants.getCurrentVersion().getName()});
                this.minecraft.setScreen(new BackupConfirmScreen(this.screen, (var1x, var2x) -> {
                   if (var1x) {
                      String var3 = this.summary.getLevelId();
 
                      try {
                         LevelStorageSource.LevelStorageAccess var4 = this.minecraft.getLevelSource().createAccess(var3);
-                        Throwable var5 = null;
 
                         try {
                            EditWorldScreen.makeBackupAndShowToast(var4);
-                        } catch (Throwable var15) {
-                           var5 = var15;
-                           throw var15;
-                        } finally {
+                        } catch (Throwable var8) {
                            if (var4 != null) {
-                              if (var5 != null) {
-                                 try {
-                                    var4.close();
-                                 } catch (Throwable var14) {
-                                    var5.addSuppressed(var14);
-                                 }
-                              } else {
+                              try {
                                  var4.close();
+                              } catch (Throwable var7) {
+                                 var8.addSuppressed(var7);
                               }
                            }
 
+                           throw var8;
                         }
-                     } catch (IOException var17) {
+
+                        if (var4 != null) {
+                           var4.close();
+                        }
+                     } catch (IOException var9) {
                         SystemToast.onWorldAccessFailure(this.minecraft, var3);
-                        WorldSelectionList.LOGGER.error("Failed to backup level {}", var3, var17);
+                        WorldSelectionList.LOGGER.error("Failed to backup level {}", var3, var9);
                      }
                   }
 
                   this.loadWorld();
-               }, var1, var2, false));
+               }, var4, var5, false));
             } else if (this.summary.askToOpenWorld()) {
                this.minecraft.setScreen(new ConfirmScreen((var1x) -> {
                   if (var1x) {
                      try {
                         this.loadWorld();
@@ -316,60 +350,61 @@
                      }
                   } else {
                      this.minecraft.setScreen(this.screen);
                   }
 
-               }, new TranslatableComponent("selectWorld.versionQuestion"), new TranslatableComponent("selectWorld.versionWarning", new Object[]{this.summary.getWorldVersionName(), new TranslatableComponent("selectWorld.versionJoinButton"), CommonComponents.GUI_CANCEL})));
+               }, new TranslatableComponent("selectWorld.versionQuestion"), new TranslatableComponent("selectWorld.versionWarning", new Object[]{this.summary.getWorldVersionName()}), new TranslatableComponent("selectWorld.versionJoinButton"), CommonComponents.GUI_CANCEL));
             } else {
                this.loadWorld();
             }
 
          }
       }
 
       public void deleteWorld() {
          this.minecraft.setScreen(new ConfirmScreen((var1) -> {
             if (var1) {
-               this.minecraft.setScreen(new ProgressScreen());
-               LevelStorageSource var2 = this.minecraft.getLevelSource();
-               String var3 = this.summary.getLevelId();
+               this.minecraft.setScreen(new ProgressScreen(true));
+               this.doDeleteWorld();
+            }
 
-               try {
-                  LevelStorageSource.LevelStorageAccess var4 = var2.createAccess(var3);
-                  Throwable var5 = null;
+            this.minecraft.setScreen(this.screen);
+         }, new TranslatableComponent("selectWorld.deleteQuestion"), new TranslatableComponent("selectWorld.deleteWarning", new Object[]{this.summary.getLevelName()}), new TranslatableComponent("selectWorld.deleteButton"), CommonComponents.GUI_CANCEL));
+      }
 
-                  try {
-                     var4.deleteLevel();
-                  } catch (Throwable var15) {
-                     var5 = var15;
-                     throw var15;
-                  } finally {
-                     if (var4 != null) {
-                        if (var5 != null) {
-                           try {
-                              var4.close();
-                           } catch (Throwable var14) {
-                              var5.addSuppressed(var14);
-                           }
-                        } else {
-                           var4.close();
-                        }
-                     }
+      public void doDeleteWorld() {
+         LevelStorageSource var1 = this.minecraft.getLevelSource();
+         String var2 = this.summary.getLevelId();
 
+         try {
+            LevelStorageSource.LevelStorageAccess var3 = var1.createAccess(var2);
+
+            try {
+               var3.deleteLevel();
+            } catch (Throwable var7) {
+               if (var3 != null) {
+                  try {
+                     var3.close();
+                  } catch (Throwable var6) {
+                     var7.addSuppressed(var6);
                   }
-               } catch (IOException var17) {
-                  SystemToast.onWorldDeleteFailure(this.minecraft, var3);
-                  WorldSelectionList.LOGGER.error("Failed to delete world {}", var3, var17);
                }
 
-               WorldSelectionList.this.refreshList(() -> {
-                  return this.screen.searchBox.getValue();
-               }, true);
+               throw var7;
             }
 
-            this.minecraft.setScreen(this.screen);
-         }, new TranslatableComponent("selectWorld.deleteQuestion"), new TranslatableComponent("selectWorld.deleteWarning", new Object[]{this.summary.getLevelName()}), new TranslatableComponent("selectWorld.deleteButton"), CommonComponents.GUI_CANCEL));
+            if (var3 != null) {
+               var3.close();
+            }
+         } catch (IOException var8) {
+            SystemToast.onWorldDeleteFailure(this.minecraft, var2);
+            WorldSelectionList.LOGGER.error("Failed to delete world {}", var2, var8);
+         }
+
+         WorldSelectionList.this.refreshList(() -> {
+            return this.screen.searchBox.getValue();
+         }, true);
       }
 
       public void editWorld() {
          String var1 = this.summary.getLevelId();
 
@@ -404,64 +439,58 @@
          this.queueLoadScreen();
          RegistryAccess.RegistryHolder var1 = RegistryAccess.builtin();
 
          try {
             LevelStorageSource.LevelStorageAccess var2 = this.minecraft.getLevelSource().createAccess(this.summary.getLevelId());
-            Throwable var3 = null;
 
             try {
-               Minecraft.ServerStem var4 = this.minecraft.makeServerStem(var1, Minecraft::loadDataPacks, Minecraft::loadWorldData, false, var2);
-               Throwable var5 = null;
+               Minecraft.ServerStem var3 = this.minecraft.makeServerStem(var1, Minecraft::loadDataPacks, Minecraft::loadWorldData, false, var2);
 
                try {
-                  LevelSettings var6 = var4.worldData().getLevelSettings();
-                  DataPackConfig var7 = var6.getDataPackConfig();
-                  WorldGenSettings var8 = var4.worldData().worldGenSettings();
-                  Path var9 = CreateWorldScreen.createTempDataPackDirFromExistingWorld(var2.getLevelPath(LevelResource.DATAPACK_DIR), this.minecraft);
-                  if (var8.isOldCustomizedWorld()) {
+                  LevelSettings var4 = var3.worldData().getLevelSettings();
+                  DataPackConfig var5 = var4.getDataPackConfig();
+                  WorldGenSettings var6 = var3.worldData().worldGenSettings();
+                  Path var7 = CreateWorldScreen.createTempDataPackDirFromExistingWorld(var2.getLevelPath(LevelResource.DATAPACK_DIR), this.minecraft);
+                  if (var6.isOldCustomizedWorld()) {
                      this.minecraft.setScreen(new ConfirmScreen((var6x) -> {
-                        this.minecraft.setScreen((Screen)(var6x ? new CreateWorldScreen(this.screen, var6, var8, var9, var7, var1) : this.screen));
+                        this.minecraft.setScreen((Screen)(var6x ? new CreateWorldScreen(this.screen, var4, var6, var7, var5, var1) : this.screen));
                      }, new TranslatableComponent("selectWorld.recreate.customized.title"), new TranslatableComponent("selectWorld.recreate.customized.text"), CommonComponents.GUI_PROCEED, CommonComponents.GUI_CANCEL));
                   } else {
-                     this.minecraft.setScreen(new CreateWorldScreen(this.screen, var6, var8, var9, var7, var1));
+                     this.minecraft.setScreen(new CreateWorldScreen(this.screen, var4, var6, var7, var5, var1));
                   }
-               } catch (Throwable var33) {
-                  var5 = var33;
-                  throw var33;
-               } finally {
-                  if (var4 != null) {
-                     if (var5 != null) {
-                        try {
-                           var4.close();
-                        } catch (Throwable var32) {
-                           var5.addSuppressed(var32);
-                        }
-                     } else {
-                        var4.close();
+               } catch (Throwable var10) {
+                  if (var3 != null) {
+                     try {
+                        var3.close();
+                     } catch (Throwable var9) {
+                        var10.addSuppressed(var9);
                      }
                   }
 
+                  throw var10;
                }
-            } catch (Throwable var35) {
-               var3 = var35;
-               throw var35;
-            } finally {
+
+               if (var3 != null) {
+                  var3.close();
+               }
+            } catch (Throwable var11) {
                if (var2 != null) {
-                  if (var3 != null) {
-                     try {
-                        var2.close();
-                     } catch (Throwable var31) {
-                        var3.addSuppressed(var31);
-                     }
-                  } else {
+                  try {
                      var2.close();
+                  } catch (Throwable var8) {
+                     var11.addSuppressed(var8);
                   }
                }
 
+               throw var11;
             }
-         } catch (Exception var37) {
-            WorldSelectionList.LOGGER.error("Unable to recreate world", var37);
+
+            if (var2 != null) {
+               var2.close();
+            }
+         } catch (Exception var12) {
+            WorldSelectionList.LOGGER.error("Unable to recreate world", var12);
             this.minecraft.setScreen(new AlertScreen(() -> {
                this.minecraft.setScreen(this.screen);
             }, new TranslatableComponent("selectWorld.recreate.error.title"), new TranslatableComponent("selectWorld.recreate.error.text")));
          }
 
@@ -484,41 +513,33 @@
       private DynamicTexture loadServerIcon() {
          boolean var1 = this.iconFile != null && this.iconFile.isFile();
          if (var1) {
             try {
                FileInputStream var2 = new FileInputStream(this.iconFile);
-               Throwable var3 = null;
 
-               DynamicTexture var6;
+               DynamicTexture var5;
                try {
-                  NativeImage var4 = NativeImage.read((InputStream)var2);
-                  Validate.validState(var4.getWidth() == 64, "Must be 64 pixels wide", new Object[0]);
-                  Validate.validState(var4.getHeight() == 64, "Must be 64 pixels high", new Object[0]);
-                  DynamicTexture var5 = new DynamicTexture(var4);
-                  this.minecraft.getTextureManager().register((ResourceLocation)this.iconLocation, (AbstractTexture)var5);
-                  var6 = var5;
-               } catch (Throwable var16) {
-                  var3 = var16;
-                  throw var16;
-               } finally {
-                  if (var2 != null) {
-                     if (var3 != null) {
-                        try {
-                           var2.close();
-                        } catch (Throwable var15) {
-                           var3.addSuppressed(var15);
-                        }
-                     } else {
-                        var2.close();
-                     }
+                  NativeImage var3 = NativeImage.read((InputStream)var2);
+                  Validate.validState(var3.getWidth() == 64, "Must be 64 pixels wide", new Object[0]);
+                  Validate.validState(var3.getHeight() == 64, "Must be 64 pixels high", new Object[0]);
+                  DynamicTexture var4 = new DynamicTexture(var3);
+                  this.minecraft.getTextureManager().register((ResourceLocation)this.iconLocation, (AbstractTexture)var4);
+                  var5 = var4;
+               } catch (Throwable var7) {
+                  try {
+                     var2.close();
+                  } catch (Throwable var6) {
+                     var7.addSuppressed(var6);
                   }
 
+                  throw var7;
                }
 
-               return var6;
-            } catch (Throwable var18) {
-               WorldSelectionList.LOGGER.error("Invalid icon for world {}", this.summary.getLevelId(), var18);
+               var2.close();
+               return var5;
+            } catch (Throwable var8) {
+               WorldSelectionList.LOGGER.error("Invalid icon for world {}", this.summary.getLevelId(), var8);
                this.iconFile = null;
                return null;
             }
          } else {
             this.minecraft.getTextureManager().release(this.iconLocation);
@@ -530,7 +551,11 @@
          if (this.icon != null) {
             this.icon.close();
          }
 
       }
+
+      public String getLevelName() {
+         return this.summary.getLevelName();
+      }
    }
 }
