--- net/minecraft/client/gui/screens/worldselection/WorldSelectionList.java
+++ net/minecraft/client/gui/screens/worldselection/WorldSelectionList.java
@@ -316,11 +316,11 @@
                      }
                   } else {
                      this.minecraft.setScreen(this.screen);
                   }
 
-               }, new TranslatableComponent("selectWorld.versionQuestion"), new TranslatableComponent("selectWorld.versionWarning", new Object[]{this.summary.getWorldVersionName(), new TranslatableComponent("selectWorld.versionJoinButton"), CommonComponents.GUI_CANCEL})));
+               }, new TranslatableComponent("selectWorld.versionQuestion"), new TranslatableComponent("selectWorld.versionWarning", new Object[]{this.summary.getWorldVersionName()}), new TranslatableComponent("selectWorld.versionJoinButton"), CommonComponents.GUI_CANCEL));
             } else {
                this.loadWorld();
             }
 
          }
