--- net/minecraft/client/gui/screens/worldselection/WorldSelectionList.java
+++ net/minecraft/client/gui/screens/worldselection/WorldSelectionList.java
@@ -263,13 +263,20 @@
          }
       }
 
       public void joinWorld() {
          if (!this.summary.isLocked()) {
-            if (this.summary.shouldBackup()) {
-               TranslatableComponent var1 = new TranslatableComponent("selectWorld.backupQuestion");
-               TranslatableComponent var2 = new TranslatableComponent("selectWorld.backupWarning", new Object[]{this.summary.getWorldVersionName(), SharedConstants.getCurrentVersion().getName()});
+            LevelSummary.BackupStatus var1 = this.summary.backupStatus();
+            if (var1.shouldBackup()) {
+               String var2 = "selectWorld.backupQuestion." + var1.getTranslationKey();
+               String var3 = "selectWorld.backupWarning." + var1.getTranslationKey();
+               TranslatableComponent var4 = new TranslatableComponent(var2);
+               if (var1.isSevere()) {
+                  var4.withStyle(ChatFormatting.BOLD, ChatFormatting.RED);
+               }
+
+               TranslatableComponent var5 = new TranslatableComponent(var3, new Object[]{this.summary.getWorldVersionName(), SharedConstants.getCurrentVersion().getName()});
                this.minecraft.setScreen(new BackupConfirmScreen(this.screen, (var1x, var2x) -> {
                   if (var1x) {
                      String var3 = this.summary.getLevelId();
 
                      try {
@@ -300,11 +307,11 @@
                         WorldSelectionList.LOGGER.error("Failed to backup level {}", var3, var17);
                      }
                   }
 
                   this.loadWorld();
-               }, var1, var2, false));
+               }, var4, var5, false));
             } else if (this.summary.askToOpenWorld()) {
                this.minecraft.setScreen(new ConfirmScreen((var1x) -> {
                   if (var1x) {
                      try {
                         this.loadWorld();
@@ -316,11 +323,11 @@
                      }
                   } else {
                      this.minecraft.setScreen(this.screen);
                   }
 
-               }, new TranslatableComponent("selectWorld.versionQuestion"), new TranslatableComponent("selectWorld.versionWarning", new Object[]{this.summary.getWorldVersionName(), new TranslatableComponent("selectWorld.versionJoinButton"), CommonComponents.GUI_CANCEL})));
+               }, new TranslatableComponent("selectWorld.versionQuestion"), new TranslatableComponent("selectWorld.versionWarning", new Object[]{this.summary.getWorldVersionName()}), new TranslatableComponent("selectWorld.versionJoinButton"), CommonComponents.GUI_CANCEL));
             } else {
                this.loadWorld();
             }
 
          }
@@ -328,48 +335,52 @@
 
       public void deleteWorld() {
          this.minecraft.setScreen(new ConfirmScreen((var1) -> {
             if (var1) {
                this.minecraft.setScreen(new ProgressScreen());
-               LevelStorageSource var2 = this.minecraft.getLevelSource();
-               String var3 = this.summary.getLevelId();
+               this.doDeleteWorld();
+            }
 
-               try {
-                  LevelStorageSource.LevelStorageAccess var4 = var2.createAccess(var3);
-                  Throwable var5 = null;
+            this.minecraft.setScreen(this.screen);
+         }, new TranslatableComponent("selectWorld.deleteQuestion"), new TranslatableComponent("selectWorld.deleteWarning", new Object[]{this.summary.getLevelName()}), new TranslatableComponent("selectWorld.deleteButton"), CommonComponents.GUI_CANCEL));
+      }
 
-                  try {
-                     var4.deleteLevel();
-                  } catch (Throwable var15) {
-                     var5 = var15;
-                     throw var15;
-                  } finally {
-                     if (var4 != null) {
-                        if (var5 != null) {
-                           try {
-                              var4.close();
-                           } catch (Throwable var14) {
-                              var5.addSuppressed(var14);
-                           }
-                        } else {
-                           var4.close();
-                        }
-                     }
+      public void doDeleteWorld() {
+         LevelStorageSource var1 = this.minecraft.getLevelSource();
+         String var2 = this.summary.getLevelId();
 
+         try {
+            LevelStorageSource.LevelStorageAccess var3 = var1.createAccess(var2);
+            Throwable var4 = null;
+
+            try {
+               var3.deleteLevel();
+            } catch (Throwable var14) {
+               var4 = var14;
+               throw var14;
+            } finally {
+               if (var3 != null) {
+                  if (var4 != null) {
+                     try {
+                        var3.close();
+                     } catch (Throwable var13) {
+                        var4.addSuppressed(var13);
+                     }
+                  } else {
+                     var3.close();
                   }
-               } catch (IOException var17) {
-                  SystemToast.onWorldDeleteFailure(this.minecraft, var3);
-                  WorldSelectionList.LOGGER.error("Failed to delete world {}", var3, var17);
                }
 
-               WorldSelectionList.this.refreshList(() -> {
-                  return this.screen.searchBox.getValue();
-               }, true);
             }
+         } catch (IOException var16) {
+            SystemToast.onWorldDeleteFailure(this.minecraft, var2);
+            WorldSelectionList.LOGGER.error("Failed to delete world {}", var2, var16);
+         }
 
-            this.minecraft.setScreen(this.screen);
-         }, new TranslatableComponent("selectWorld.deleteQuestion"), new TranslatableComponent("selectWorld.deleteWarning", new Object[]{this.summary.getLevelName()}), new TranslatableComponent("selectWorld.deleteButton"), CommonComponents.GUI_CANCEL));
+         WorldSelectionList.this.refreshList(() -> {
+            return this.screen.searchBox.getValue();
+         }, true);
       }
 
       public void editWorld() {
          String var1 = this.summary.getLevelId();
 
