--- com/mojang/realmsclient/KeyCombo.java
+++ com/mojang/realmsclient/KeyCombo.java
@@ -14,10 +14,15 @@
       } else {
          this.chars = var1;
       }
    }
 
+   public KeyCombo(char[] var1) {
+      this(var1, () -> {
+      });
+   }
+
    public boolean keyPressed(char var1) {
       if (var1 == this.chars[this.matchIndex++]) {
          if (this.matchIndex == this.chars.length) {
             this.reset();
             this.onCompletion.run();
