--- com/mojang/realmsclient/KeyCombo.java
+++ com/mojang/realmsclient/KeyCombo.java
@@ -14,10 +14,15 @@
       } else {
          this.chars = var1;
       }
    }
 
+   public KeyCombo(char[] var1) {
+      this(var1, () -> {
+      });
+   }
+
    public boolean keyPressed(char var1) {
       if (var1 == this.chars[this.matchIndex++]) {
          if (this.matchIndex == this.chars.length) {
             this.reset();
             this.onCompletion.run();
@@ -33,8 +38,9 @@
    public void reset() {
       this.matchIndex = 0;
    }
 
    public String toString() {
-      return "KeyCombo{chars=" + Arrays.toString(this.chars) + ", matchIndex=" + this.matchIndex + '}';
+      String var10000 = Arrays.toString(this.chars);
+      return "KeyCombo{chars=" + var10000 + ", matchIndex=" + this.matchIndex + "}";
    }
 }
