--- com/mojang/realmsclient/RealmsMainScreen.java
+++ com/mojang/realmsclient/RealmsMainScreen.java
@@ -568,11 +568,11 @@
                      RealmsClient.switchToStage();
                      RealmsMainScreen.LOGGER.info("Switched to stage");
                      RealmsMainScreen.REALMS_DATA_FETCHER.forceUpdate();
                   }
                } catch (RealmsServiceException var3) {
-                  RealmsMainScreen.LOGGER.error("Couldn't connect to Realms: " + var3);
+                  RealmsMainScreen.LOGGER.error("Couldn't connect to Realms: {}", var3.toString());
                }
 
             }
          }).start();
       }
@@ -591,11 +591,11 @@
                      RealmsClient.switchToLocal();
                      RealmsMainScreen.LOGGER.info("Switched to local");
                      RealmsMainScreen.REALMS_DATA_FETCHER.forceUpdate();
                   }
                } catch (RealmsServiceException var3) {
-                  RealmsMainScreen.LOGGER.error("Couldn't connect to Realms: " + var3);
+                  RealmsMainScreen.LOGGER.error("Couldn't connect to Realms: {}", var3.toString());
                }
 
             }
          }).start();
       }
@@ -802,11 +802,11 @@
          this.hasSwitchedCarouselImage = true;
          this.updateButtonStates((RealmsServer)null);
          if (this.children.contains(this.realmSelectionList)) {
             RealmsMainScreen.RealmSelectionList var6 = this.realmSelectionList;
             if (!this.children.remove(var6)) {
-               LOGGER.error("Unable to remove widget: " + var6);
+               LOGGER.error("Unable to remove widget: {}", var6);
             }
          }
 
          NarrationHelper.now(POPUP_TEXT.getString());
       }
@@ -1395,11 +1395,11 @@
 
       public boolean keyPressed(int var1, int var2, int var3) {
          if (var1 != 257 && var1 != 32 && var1 != 335) {
             return super.keyPressed(var1, var2, var3);
          } else {
-            ObjectSelectionList.Entry var4 = (ObjectSelectionList.Entry)this.getSelected();
+            RealmsMainScreen.Entry var4 = (RealmsMainScreen.Entry)this.getSelected();
             return var4 == null ? super.keyPressed(var1, var2, var3) : var4.mouseClicked(0.0D, 0.0D, 0);
          }
       }
 
       public boolean mouseClicked(double var1, double var3, int var5) {
