--- com/mojang/realmsclient/Unit.java
+++ com/mojang/realmsclient/Unit.java
@@ -6,10 +6,12 @@
    B,
    KB,
    MB,
    GB;
 
+   private static final int BASE_UNIT = 1024;
+
    private Unit() {
    }
 
    public static Unit getLargest(long var0) {
       if (var0 < 1024L) {
@@ -33,11 +35,11 @@
       boolean var2 = true;
       if (var0 < 1024L) {
          return var0 + " B";
       } else {
          int var3 = (int)(Math.log((double)var0) / Math.log(1024.0D));
-         String var4 = "KMGTPE".charAt(var3 - 1) + "";
+         String var4 = "KMGTPE".charAt(var3 - 1).makeConcatWithConstants<invokedynamic>("KMGTPE".charAt(var3 - 1));
          return String.format(Locale.ROOT, "%.1f %sB", (double)var0 / Math.pow(1024.0D, (double)var3), var4);
       }
    }
 
    public static String humanReadable(long var0, Unit var2) {
