--- net/minecraft/world/level/saveddata/maps/MapFrame.java
+++ net/minecraft/world/level/saveddata/maps/MapFrame.java
@@ -45,8 +45,9 @@
    public String getId() {
       return frameId(this.pos);
    }
 
    public static String frameId(BlockPos var0) {
-      return "frame-" + var0.getX() + "," + var0.getY() + "," + var0.getZ();
+      int var10000 = var0.getX();
+      return "frame-" + var10000 + "," + var0.getY() + "," + var0.getZ();
    }
 }
