--- net/minecraft/world/entity/ai/attributes/AttributeMap.java
+++ net/minecraft/world/entity/ai/attributes/AttributeMap.java
@@ -4,10 +4,11 @@
 import com.google.common.collect.Multimap;
 import com.google.common.collect.Sets;
 import java.util.Collection;
 import java.util.Iterator;
 import java.util.Map;
+import java.util.Objects;
 import java.util.Set;
 import java.util.UUID;
 import java.util.stream.Collectors;
 import javax.annotation.Nullable;
 import net.minecraft.Util;
@@ -78,10 +79,11 @@
 
    public void removeAttributeModifiers(Multimap<Attribute, AttributeModifier> var1) {
       var1.asMap().forEach((var1x, var2) -> {
          AttributeInstance var3 = (AttributeInstance)this.attributes.get(var1x);
          if (var3 != null) {
+            Objects.requireNonNull(var3);
             var2.forEach(var3::removeModifier);
          }
 
       });
    }
@@ -95,10 +97,20 @@
          }
 
       });
    }
 
+   public void assignValues(AttributeMap var1) {
+      var1.attributes.values().forEach((var1x) -> {
+         AttributeInstance var2 = this.getInstance(var1x.getAttribute());
+         if (var2 != null) {
+            var2.replaceFrom(var1x);
+         }
+
+      });
+   }
+
    public ListTag save() {
       ListTag var1 = new ListTag();
       Iterator var2 = this.attributes.values().iterator();
 
       while(var2.hasNext()) {
