--- net/minecraft/world/entity/ai/attributes/AttributeInstance.java
+++ net/minecraft/world/entity/ai/attributes/AttributeInstance.java
@@ -114,10 +114,20 @@
       } else {
          return false;
       }
    }
 
+   public void removeModifiers() {
+      Iterator var1 = this.getModifiers().iterator();
+
+      while(var1.hasNext()) {
+         AttributeModifier var2 = (AttributeModifier)var1.next();
+         this.removeModifier(var2);
+      }
+
+   }
+
    public double getValue() {
       if (this.dirty) {
          this.cachedValue = this.calculateValue();
          this.dirty = false;
       }
