--- net/minecraft/world/entity/ai/attributes/RangedAttribute.java
+++ net/minecraft/world/entity/ai/attributes/RangedAttribute.java
@@ -17,10 +17,18 @@
       } else if (var2 > var6) {
          throw new IllegalArgumentException("Default value cannot be bigger than maximum value!");
       }
    }
 
+   public double getMinValue() {
+      return this.minValue;
+   }
+
+   public double getMaxValue() {
+      return this.maxValue;
+   }
+
    public double sanitizeValue(double var1) {
       var1 = Mth.clamp(var1, this.minValue, this.maxValue);
       return var1;
    }
 }
