--- net/minecraft/server/commands/CloneCommands.java
+++ net/minecraft/server/commands/CloneCommands.java
@@ -174,11 +174,11 @@
                   BlockEntity var35 = var10.getBlockEntity(var34.pos);
                   if (var34.tag != null && var35 != null) {
                      var34.tag.putInt("x", var34.pos.getX());
                      var34.tag.putInt("y", var34.pos.getY());
                      var34.tag.putInt("z", var34.pos.getZ());
-                     var35.load(var34.state, var34.tag);
+                     var35.load(var34.tag);
                      var35.setChanged();
                   }
                }
 
                var32 = var28.iterator();
