--- net/minecraft/server/commands/DifficultyCommand.java
+++ net/minecraft/server/commands/DifficultyCommand.java
@@ -13,10 +13,13 @@
 public class DifficultyCommand {
    private static final DynamicCommandExceptionType ERROR_ALREADY_DIFFICULT = new DynamicCommandExceptionType((var0) -> {
       return new TranslatableComponent("commands.difficulty.failure", new Object[]{var0});
    });
 
+   public DifficultyCommand() {
+   }
+
    public static void register(CommandDispatcher<CommandSourceStack> var0) {
       LiteralArgumentBuilder var1 = Commands.literal("difficulty");
       Difficulty[] var2 = Difficulty.values();
       int var3 = var2.length;
 
