--- net/minecraft/server/commands/ExecuteCommand.java
+++ net/minecraft/server/commands/ExecuteCommand.java
@@ -396,12 +396,12 @@
    private static OptionalInt checkRegions(CommandContext<CommandSourceStack> var0, boolean var1) throws CommandSyntaxException {
       return checkRegions(((CommandSourceStack)var0.getSource()).getLevel(), BlockPosArgument.getLoadedBlockPos(var0, "start"), BlockPosArgument.getLoadedBlockPos(var0, "end"), BlockPosArgument.getLoadedBlockPos(var0, "destination"), var1);
    }
 
    private static OptionalInt checkRegions(ServerLevel var0, BlockPos var1, BlockPos var2, BlockPos var3, boolean var4) throws CommandSyntaxException {
-      BoundingBox var5 = new BoundingBox(var1, var2);
-      BoundingBox var6 = new BoundingBox(var3, var3.offset(var5.getLength()));
+      BoundingBox var5 = BoundingBox.createProper(var1, var2);
+      BoundingBox var6 = BoundingBox.createProper(var3, var3.offset(var5.getLength()));
       BlockPos var7 = new BlockPos(var6.x0 - var5.x0, var6.y0 - var5.y0, var6.z0 - var5.z0);
       int var8 = var5.getXSpan() * var5.getYSpan() * var5.getZSpan();
       if (var8 > 32768) {
          throw ERROR_AREA_TOO_LARGE.create(32768, var8);
       } else {
