--- net/minecraft/server/commands/ExperienceCommand.java
+++ net/minecraft/server/commands/ExperienceCommand.java
@@ -21,10 +21,13 @@
 import net.minecraft.world.entity.player.Player;
 
 public class ExperienceCommand {
    private static final SimpleCommandExceptionType ERROR_SET_POINTS_INVALID = new SimpleCommandExceptionType(new TranslatableComponent("commands.experience.set.points.invalid"));
 
+   public ExperienceCommand() {
+   }
+
    public static void register(CommandDispatcher<CommandSourceStack> var0) {
       LiteralCommandNode var1 = var0.register((LiteralArgumentBuilder)((LiteralArgumentBuilder)((LiteralArgumentBuilder)((LiteralArgumentBuilder)Commands.literal("experience").requires((var0x) -> {
          return var0x.hasPermission(2);
       })).then(Commands.literal("add").then(Commands.argument("targets", EntityArgument.players()).then(((RequiredArgumentBuilder)((RequiredArgumentBuilder)Commands.argument("amount", IntegerArgumentType.integer()).executes((var0x) -> {
          return addExperience((CommandSourceStack)var0x.getSource(), EntityArgument.getPlayers(var0x, "targets"), IntegerArgumentType.getInteger(var0x, "amount"), ExperienceCommand.Type.POINTS);
@@ -93,11 +96,11 @@
 
          return var1.size();
       }
    }
 
-   static enum Type {
+   private static enum Type {
       POINTS("points", Player::giveExperiencePoints, (var0, var1) -> {
          if (var1 >= var0.getXpNeededForNextLevel()) {
             return false;
          } else {
             var0.setExperiencePoints(var1);
@@ -114,11 +117,11 @@
       });
 
       public final BiConsumer<ServerPlayer, Integer> add;
       public final BiPredicate<ServerPlayer, Integer> set;
       public final String name;
-      private final ToIntFunction<ServerPlayer> query;
+      final ToIntFunction<ServerPlayer> query;
 
       private Type(String var3, BiConsumer<ServerPlayer, Integer> var4, BiPredicate<ServerPlayer, Integer> var5, ToIntFunction<ServerPlayer> var6) {
          this.add = var4;
          this.name = var3;
          this.set = var5;
