--- net/minecraft/server/commands/FunctionCommand.java
+++ net/minecraft/server/commands/FunctionCommand.java
@@ -18,10 +18,13 @@
       ServerFunctionManager var2 = ((CommandSourceStack)var0.getSource()).getServer().getFunctions();
       SharedSuggestionProvider.suggestResource(var2.getTagNames(), var1, "#");
       return SharedSuggestionProvider.suggestResource(var2.getFunctionNames(), var1);
    };
 
+   public FunctionCommand() {
+   }
+
    public static void register(CommandDispatcher<CommandSourceStack> var0) {
       var0.register((LiteralArgumentBuilder)((LiteralArgumentBuilder)Commands.literal("function").requires((var0x) -> {
          return var0x.hasPermission(2);
       })).then(Commands.argument("name", FunctionArgument.functions()).suggests(SUGGEST_FUNCTION).executes((var0x) -> {
          return runFunction((CommandSourceStack)var0x.getSource(), FunctionArgument.getFunctions(var0x, "name"));
