--- net/minecraft/server/commands/GameRuleCommand.java
+++ net/minecraft/server/commands/GameRuleCommand.java
@@ -7,10 +7,13 @@
 import net.minecraft.commands.Commands;
 import net.minecraft.network.chat.TranslatableComponent;
 import net.minecraft.world.level.GameRules;
 
 public class GameRuleCommand {
+   public GameRuleCommand() {
+   }
+
    public static void register(CommandDispatcher<CommandSourceStack> var0) {
       final LiteralArgumentBuilder var1 = (LiteralArgumentBuilder)Commands.literal("gamerule").requires((var0x) -> {
          return var0x.hasPermission(2);
       });
       GameRules.visitGameRuleTypes(new GameRules.GameRuleTypeVisitor() {
@@ -23,19 +26,19 @@
          }
       });
       var0.register(var1);
    }
 
-   private static <T extends GameRules.Value<T>> int setRule(CommandContext<CommandSourceStack> var0, GameRules.Key<T> var1) {
+   static <T extends GameRules.Value<T>> int setRule(CommandContext<CommandSourceStack> var0, GameRules.Key<T> var1) {
       CommandSourceStack var2 = (CommandSourceStack)var0.getSource();
       GameRules.Value var3 = var2.getServer().getGameRules().getRule(var1);
       var3.setFromArgument(var0, "value");
       var2.sendSuccess(new TranslatableComponent("commands.gamerule.set", new Object[]{var1.getId(), var3.toString()}), true);
       return var3.getCommandResult();
    }
 
-   private static <T extends GameRules.Value<T>> int queryRule(CommandSourceStack var0, GameRules.Key<T> var1) {
+   static <T extends GameRules.Value<T>> int queryRule(CommandSourceStack var0, GameRules.Key<T> var1) {
       GameRules.Value var2 = var0.getServer().getGameRules().getRule(var1);
       var0.sendSuccess(new TranslatableComponent("commands.gamerule.query", new Object[]{var1.getId(), var2.toString()}), false);
       return var2.getCommandResult();
    }
 }
