--- net/minecraft/server/commands/GameRuleCommand.java
+++ net/minecraft/server/commands/GameRuleCommand.java
@@ -7,10 +7,13 @@
 import net.minecraft.commands.Commands;
 import net.minecraft.network.chat.TranslatableComponent;
 import net.minecraft.world.level.GameRules;
 
 public class GameRuleCommand {
+   public GameRuleCommand() {
+   }
+
    public static void register(CommandDispatcher<CommandSourceStack> var0) {
       final LiteralArgumentBuilder var1 = (LiteralArgumentBuilder)Commands.literal("gamerule").requires((var0x) -> {
          return var0x.hasPermission(2);
       });
       GameRules.visitGameRuleTypes(new GameRules.GameRuleTypeVisitor() {
