--- net/minecraft/server/commands/MsgCommand.java
+++ net/minecraft/server/commands/MsgCommand.java
@@ -17,10 +17,13 @@
 import net.minecraft.network.chat.TranslatableComponent;
 import net.minecraft.server.level.ServerPlayer;
 import net.minecraft.world.entity.Entity;
 
 public class MsgCommand {
+   public MsgCommand() {
+   }
+
    public static void register(CommandDispatcher<CommandSourceStack> var0) {
       LiteralCommandNode var1 = var0.register((LiteralArgumentBuilder)Commands.literal("msg").then(Commands.argument("targets", EntityArgument.players()).then(Commands.argument("message", MessageArgument.message()).executes((var0x) -> {
          return sendMessage((CommandSourceStack)var0x.getSource(), EntityArgument.getPlayers(var0x, "targets"), MessageArgument.getMessage(var0x, "message"));
       }))));
       var0.register((LiteralArgumentBuilder)Commands.literal("tell").redirect(var1));
