--- net/minecraft/server/commands/OpCommand.java
+++ net/minecraft/server/commands/OpCommand.java
@@ -15,10 +15,13 @@
 import net.minecraft.server.players.PlayerList;
 
 public class OpCommand {
    private static final SimpleCommandExceptionType ERROR_ALREADY_OP = new SimpleCommandExceptionType(new TranslatableComponent("commands.op.failed"));
 
+   public OpCommand() {
+   }
+
    public static void register(CommandDispatcher<CommandSourceStack> var0) {
       var0.register((LiteralArgumentBuilder)((LiteralArgumentBuilder)Commands.literal("op").requires((var0x) -> {
          return var0x.hasPermission(3);
       })).then(Commands.argument("targets", GameProfileArgument.gameProfile()).suggests((var0x, var1) -> {
          PlayerList var2 = ((CommandSourceStack)var0x.getSource()).getServer().getPlayerList();
