--- net/minecraft/server/commands/ParticleCommand.java
+++ net/minecraft/server/commands/ParticleCommand.java
@@ -21,10 +21,13 @@
 import net.minecraft.world.phys.Vec3;
 
 public class ParticleCommand {
    private static final SimpleCommandExceptionType ERROR_FAILED = new SimpleCommandExceptionType(new TranslatableComponent("commands.particle.failed"));
 
+   public ParticleCommand() {
+   }
+
    public static void register(CommandDispatcher<CommandSourceStack> var0) {
       var0.register((LiteralArgumentBuilder)((LiteralArgumentBuilder)Commands.literal("particle").requires((var0x) -> {
          return var0x.hasPermission(2);
       })).then(((RequiredArgumentBuilder)Commands.argument("name", ParticleArgument.particle()).executes((var0x) -> {
          return sendParticles((CommandSourceStack)var0x.getSource(), ParticleArgument.getParticle(var0x, "name"), ((CommandSourceStack)var0x.getSource()).getPosition(), Vec3.ZERO, 0.0F, 0, false, ((CommandSourceStack)var0x.getSource()).getServer().getPlayerList().getPlayers());
