--- net/minecraft/server/commands/ReloadCommand.java
+++ net/minecraft/server/commands/ReloadCommand.java
@@ -17,10 +17,13 @@
 import org.apache.logging.log4j.Logger;
 
 public class ReloadCommand {
    private static final Logger LOGGER = LogManager.getLogger();
 
+   public ReloadCommand() {
+   }
+
    public static void reloadPacks(Collection<String> var0, CommandSourceStack var1) {
       var1.getServer().reloadResources(var0).exceptionally((var1x) -> {
          LOGGER.warn("Failed to execute reload", var1x);
          var1.sendFailure(new TranslatableComponent("commands.reload.failure"));
          return null;
