--- net/minecraft/server/commands/SaveOffCommand.java
+++ net/minecraft/server/commands/SaveOffCommand.java
@@ -10,10 +10,13 @@
 import net.minecraft.server.level.ServerLevel;
 
 public class SaveOffCommand {
    private static final SimpleCommandExceptionType ERROR_ALREADY_OFF = new SimpleCommandExceptionType(new TranslatableComponent("commands.save.alreadyOff"));
 
+   public SaveOffCommand() {
+   }
+
    public static void register(CommandDispatcher<CommandSourceStack> var0) {
       var0.register((LiteralArgumentBuilder)((LiteralArgumentBuilder)Commands.literal("save-off").requires((var0x) -> {
          return var0x.hasPermission(4);
       })).executes((var0x) -> {
          CommandSourceStack var1 = (CommandSourceStack)var0x.getSource();
