--- net/minecraft/server/commands/SaveAllCommand.java
+++ net/minecraft/server/commands/SaveAllCommand.java
@@ -10,10 +10,13 @@
 import net.minecraft.server.MinecraftServer;
 
 public class SaveAllCommand {
    private static final SimpleCommandExceptionType ERROR_FAILED = new SimpleCommandExceptionType(new TranslatableComponent("commands.save.failed"));
 
+   public SaveAllCommand() {
+   }
+
    public static void register(CommandDispatcher<CommandSourceStack> var0) {
       var0.register((LiteralArgumentBuilder)((LiteralArgumentBuilder)((LiteralArgumentBuilder)Commands.literal("save-all").requires((var0x) -> {
          return var0x.hasPermission(4);
       })).executes((var0x) -> {
          return saveAll((CommandSourceStack)var0x.getSource(), false);
