--- net/minecraft/server/commands/ScoreboardCommand.java
+++ net/minecraft/server/commands/ScoreboardCommand.java
@@ -44,10 +44,13 @@
    private static final SimpleCommandExceptionType ERROR_NOT_TRIGGER = new SimpleCommandExceptionType(new TranslatableComponent("commands.scoreboard.players.enable.invalid"));
    private static final Dynamic2CommandExceptionType ERROR_NO_VALUE = new Dynamic2CommandExceptionType((var0, var1) -> {
       return new TranslatableComponent("commands.scoreboard.players.get.null", new Object[]{var0, var1});
    });
 
+   public ScoreboardCommand() {
+   }
+
    public static void register(CommandDispatcher<CommandSourceStack> var0) {
       var0.register((LiteralArgumentBuilder)((LiteralArgumentBuilder)((LiteralArgumentBuilder)Commands.literal("scoreboard").requires((var0x) -> {
          return var0x.hasPermission(2);
       })).then(((LiteralArgumentBuilder)((LiteralArgumentBuilder)((LiteralArgumentBuilder)((LiteralArgumentBuilder)Commands.literal("objectives").then(Commands.literal("list").executes((var0x) -> {
          return listObjectives((CommandSourceStack)var0x.getSource());
