--- net/minecraft/server/commands/SetSpawnCommand.java
+++ net/minecraft/server/commands/SetSpawnCommand.java
@@ -23,13 +23,13 @@
       })).executes((var0x) -> {
          return setSpawn((CommandSourceStack)var0x.getSource(), Collections.singleton(((CommandSourceStack)var0x.getSource()).getPlayerOrException()), new BlockPos(((CommandSourceStack)var0x.getSource()).getPosition()), 0.0F);
       })).then(((RequiredArgumentBuilder)Commands.argument("targets", EntityArgument.players()).executes((var0x) -> {
          return setSpawn((CommandSourceStack)var0x.getSource(), EntityArgument.getPlayers(var0x, "targets"), new BlockPos(((CommandSourceStack)var0x.getSource()).getPosition()), 0.0F);
       })).then(((RequiredArgumentBuilder)Commands.argument("pos", BlockPosArgument.blockPos()).executes((var0x) -> {
-         return setSpawn((CommandSourceStack)var0x.getSource(), EntityArgument.getPlayers(var0x, "targets"), BlockPosArgument.getOrLoadBlockPos(var0x, "pos"), 0.0F);
+         return setSpawn((CommandSourceStack)var0x.getSource(), EntityArgument.getPlayers(var0x, "targets"), BlockPosArgument.getSpawnablePos(var0x, "pos"), 0.0F);
       })).then(Commands.argument("angle", AngleArgument.angle()).executes((var0x) -> {
-         return setSpawn((CommandSourceStack)var0x.getSource(), EntityArgument.getPlayers(var0x, "targets"), BlockPosArgument.getOrLoadBlockPos(var0x, "pos"), AngleArgument.getAngle(var0x, "angle"));
+         return setSpawn((CommandSourceStack)var0x.getSource(), EntityArgument.getPlayers(var0x, "targets"), BlockPosArgument.getSpawnablePos(var0x, "pos"), AngleArgument.getAngle(var0x, "angle"));
       })))));
    }
 
    private static int setSpawn(CommandSourceStack var0, Collection<ServerPlayer> var1, BlockPos var2, float var3) {
       ResourceKey var4 = var0.getLevel().dimension();
