--- net/minecraft/server/commands/SetWorldSpawnCommand.java
+++ net/minecraft/server/commands/SetWorldSpawnCommand.java
@@ -9,19 +9,22 @@
 import net.minecraft.commands.arguments.coordinates.BlockPosArgument;
 import net.minecraft.core.BlockPos;
 import net.minecraft.network.chat.TranslatableComponent;
 
 public class SetWorldSpawnCommand {
+   public SetWorldSpawnCommand() {
+   }
+
    public static void register(CommandDispatcher<CommandSourceStack> var0) {
       var0.register((LiteralArgumentBuilder)((LiteralArgumentBuilder)((LiteralArgumentBuilder)Commands.literal("setworldspawn").requires((var0x) -> {
          return var0x.hasPermission(2);
       })).executes((var0x) -> {
          return setSpawn((CommandSourceStack)var0x.getSource(), new BlockPos(((CommandSourceStack)var0x.getSource()).getPosition()), 0.0F);
       })).then(((RequiredArgumentBuilder)Commands.argument("pos", BlockPosArgument.blockPos()).executes((var0x) -> {
-         return setSpawn((CommandSourceStack)var0x.getSource(), BlockPosArgument.getOrLoadBlockPos(var0x, "pos"), 0.0F);
+         return setSpawn((CommandSourceStack)var0x.getSource(), BlockPosArgument.getSpawnablePos(var0x, "pos"), 0.0F);
       })).then(Commands.argument("angle", AngleArgument.angle()).executes((var0x) -> {
-         return setSpawn((CommandSourceStack)var0x.getSource(), BlockPosArgument.getOrLoadBlockPos(var0x, "pos"), AngleArgument.getAngle(var0x, "angle"));
+         return setSpawn((CommandSourceStack)var0x.getSource(), BlockPosArgument.getSpawnablePos(var0x, "pos"), AngleArgument.getAngle(var0x, "angle"));
       }))));
    }
 
    private static int setSpawn(CommandSourceStack var0, BlockPos var1, float var2) {
       var0.getLevel().setDefaultSpawnPos(var1, var2);
