--- net/minecraft/server/commands/SpectateCommand.java
+++ net/minecraft/server/commands/SpectateCommand.java
@@ -19,10 +19,13 @@
    private static final SimpleCommandExceptionType ERROR_SELF = new SimpleCommandExceptionType(new TranslatableComponent("commands.spectate.self"));
    private static final DynamicCommandExceptionType ERROR_NOT_SPECTATOR = new DynamicCommandExceptionType((var0) -> {
       return new TranslatableComponent("commands.spectate.not_spectator", new Object[]{var0});
    });
 
+   public SpectateCommand() {
+   }
+
    public static void register(CommandDispatcher<CommandSourceStack> var0) {
       var0.register((LiteralArgumentBuilder)((LiteralArgumentBuilder)((LiteralArgumentBuilder)Commands.literal("spectate").requires((var0x) -> {
          return var0x.hasPermission(2);
       })).executes((var0x) -> {
          return spectate((CommandSourceStack)var0x.getSource(), (Entity)null, ((CommandSourceStack)var0x.getSource()).getPlayerOrException());
