--- net/minecraft/server/commands/TeamMsgCommand.java
+++ net/minecraft/server/commands/TeamMsgCommand.java
@@ -22,10 +22,13 @@
 
 public class TeamMsgCommand {
    private static final Style SUGGEST_STYLE;
    private static final SimpleCommandExceptionType ERROR_NOT_ON_TEAM;
 
+   public TeamMsgCommand() {
+   }
+
    public static void register(CommandDispatcher<CommandSourceStack> var0) {
       LiteralCommandNode var1 = var0.register((LiteralArgumentBuilder)Commands.literal("teammsg").then(Commands.argument("message", MessageArgument.message()).executes((var0x) -> {
          return sendMessage((CommandSourceStack)var0x.getSource(), MessageArgument.getMessage(var0x, "message"));
       })));
       var0.register((LiteralArgumentBuilder)Commands.literal("tm").redirect(var1));
