--- com/mojang/realmsclient/client/FileDownload.java
+++ com/mojang/realmsclient/client/FileDownload.java
@@ -120,11 +120,11 @@
                      var9.setListener(var8);
                      IOUtils.copy(var6.getEntity().getContent(), var9);
                      var90 = false;
                      break label1403;
                   } catch (Exception var103) {
-                     LOGGER.error("Caught exception while downloading: " + var103.getMessage());
+                     LOGGER.error("Caught exception while downloading: {}", var103.getMessage());
                      this.error = true;
                      var90 = false;
                   } finally {
                      if (var90) {
                         this.request.releaseConnection();
@@ -149,11 +149,11 @@
                                  FileDownload.ResourcePackProgressListener var17 = new FileDownload.ResourcePackProgressListener(this.tempFile, var3, var1);
                                  FileDownload.DownloadCountingOutputStream var18 = new FileDownload.DownloadCountingOutputStream(var16);
                                  var18.setListener(var17);
                                  IOUtils.copy(var15.getEntity().getContent(), var18);
                               } catch (Exception var95) {
-                                 LOGGER.error("Caught exception while downloading: " + var95.getMessage());
+                                 LOGGER.error("Caught exception while downloading: {}", var95.getMessage());
                                  this.error = true;
                               } finally {
                                  this.request.releaseConnection();
                                  if (this.tempFile != null) {
                                     this.tempFile.delete();
@@ -198,11 +198,11 @@
                            var106 = new FileDownload.ResourcePackProgressListener(this.tempFile, var3, var1);
                            var9 = new FileDownload.DownloadCountingOutputStream(var7);
                            var9.setListener(var106);
                            IOUtils.copy(var6.getEntity().getContent(), var9);
                         } catch (Exception var99) {
-                           LOGGER.error("Caught exception while downloading: " + var99.getMessage());
+                           LOGGER.error("Caught exception while downloading: {}", var99.getMessage());
                            this.error = true;
                         } finally {
                            this.request.releaseConnection();
                            if (this.tempFile != null) {
                               this.tempFile.delete();
@@ -249,11 +249,11 @@
                         var106 = new FileDownload.ResourcePackProgressListener(this.tempFile, var3, var1);
                         var9 = new FileDownload.DownloadCountingOutputStream(var7);
                         var9.setListener(var106);
                         IOUtils.copy(var6.getEntity().getContent(), var9);
                      } catch (Exception var101) {
-                        LOGGER.error("Caught exception while downloading: " + var101.getMessage());
+                        LOGGER.error("Caught exception while downloading: {}", var101.getMessage());
                         this.error = true;
                      } finally {
                         this.request.releaseConnection();
                         if (this.tempFile != null) {
                            this.tempFile.delete();
@@ -298,11 +298,11 @@
                      FileDownload.ResourcePackProgressListener var108 = new FileDownload.ResourcePackProgressListener(this.tempFile, var3, var1);
                      FileDownload.DownloadCountingOutputStream var10 = new FileDownload.DownloadCountingOutputStream(var107);
                      var10.setListener(var108);
                      IOUtils.copy(var105.getEntity().getContent(), var10);
                   } catch (Exception var97) {
-                     LOGGER.error("Caught exception while downloading: " + var97.getMessage());
+                     LOGGER.error("Caught exception while downloading: {}", var97.getMessage());
                      this.error = true;
                   } finally {
                      this.request.releaseConnection();
                      if (this.tempFile != null) {
                         this.tempFile.delete();
@@ -650,16 +650,16 @@
                String var2 = Hashing.sha1().hashBytes(Files.toByteArray(this.tempFile)).toString();
                if (var2.equals(this.worldDownload.resourcePackHash)) {
                   FileUtils.copyFile(this.tempFile, FileDownload.this.resourcePackPath);
                   FileDownload.this.finished = true;
                } else {
-                  FileDownload.LOGGER.error("Resourcepack had wrong hash (expected " + this.worldDownload.resourcePackHash + ", found " + var2 + "). Deleting it.");
+                  FileDownload.LOGGER.error("Resourcepack had wrong hash (expected {}, found {}). Deleting it.", this.worldDownload.resourcePackHash, var2);
                   FileUtils.deleteQuietly(this.tempFile);
                   FileDownload.this.error = true;
                }
             } catch (IOException var3) {
-               FileDownload.LOGGER.error("Error copying resourcepack file", var3.getMessage());
+               FileDownload.LOGGER.error("Error copying resourcepack file: {}", var3.getMessage());
                FileDownload.this.error = true;
             }
          }
 
       }
