--- com/mojang/realmsclient/client/FileUpload.java
+++ com/mojang/realmsclient/client/FileUpload.java
@@ -125,11 +125,11 @@
    }
 
    private void handleResponse(HttpResponse var1, UploadResult.Builder var2) throws IOException {
       int var3 = var1.getStatusLine().getStatusCode();
       if (var3 == 401) {
-         LOGGER.debug("Realms server returned 401: " + var1.getFirstHeader("WWW-Authenticate"));
+         LOGGER.debug("Realms server returned 401: {}", var1.getFirstHeader("WWW-Authenticate"));
       }
 
       var2.withStatusCode(var3);
       if (var1.getEntity() != null) {
          String var4 = EntityUtils.toString(var1.getEntity(), "UTF-8");
