--- com/mojang/realmsclient/client/FileDownload.java
+++ com/mojang/realmsclient/client/FileDownload.java
@@ -92,16 +92,16 @@
       if (this.currentThread == null) {
          this.currentThread = new Thread(() -> {
             CloseableHttpClient var5 = null;
             boolean var90 = false;
 
-            label1409: {
+            label1408: {
                CloseableHttpResponse var6;
                FileOutputStream var7;
                FileDownload.DownloadCountingOutputStream var9;
                FileDownload.ResourcePackProgressListener var106;
-               label1403: {
+               label1402: {
                   try {
                      var90 = true;
                      this.tempFile = File.createTempFile("backup", ".tar.gz");
                      this.request = new HttpGet(var1.downloadLink);
                      var5 = HttpClientBuilder.create().setDefaultRequestConfig(this.requestConfig).build();
@@ -109,22 +109,22 @@
                      var3.totalBytes = Long.parseLong(var6.getFirstHeader("Content-Length").getValue());
                      if (var6.getStatusLine().getStatusCode() != 200) {
                         this.error = true;
                         this.request.abort();
                         var90 = false;
-                        break label1409;
+                        break label1408;
                      }
 
                      var7 = new FileOutputStream(this.tempFile);
                      FileDownload.ProgressListener var8 = new FileDownload.ProgressListener(var2.trim(), this.tempFile, var4, var3);
                      var9 = new FileDownload.DownloadCountingOutputStream(var7);
                      var9.setListener(var8);
                      IOUtils.copy(var6.getEntity().getContent(), var9);
                      var90 = false;
-                     break label1403;
+                     break label1402;
                   } catch (Exception var103) {
-                     LOGGER.error("Caught exception while downloading: " + var103.getMessage());
+                     LOGGER.error("Caught exception while downloading: {}", var103.getMessage());
                      this.error = true;
                      var90 = false;
                   } finally {
                      if (var90) {
                         this.request.releaseConnection();
@@ -149,11 +149,11 @@
                                  FileDownload.ResourcePackProgressListener var17 = new FileDownload.ResourcePackProgressListener(this.tempFile, var3, var1);
                                  FileDownload.DownloadCountingOutputStream var18 = new FileDownload.DownloadCountingOutputStream(var16);
                                  var18.setListener(var17);
                                  IOUtils.copy(var15.getEntity().getContent(), var18);
                               } catch (Exception var95) {
-                                 LOGGER.error("Caught exception while downloading: " + var95.getMessage());
+                                 LOGGER.error("Caught exception while downloading: {}", var95.getMessage());
                                  this.error = true;
                               } finally {
                                  this.request.releaseConnection();
                                  if (this.tempFile != null) {
                                     this.tempFile.delete();
@@ -198,11 +198,11 @@
                            var106 = new FileDownload.ResourcePackProgressListener(this.tempFile, var3, var1);
                            var9 = new FileDownload.DownloadCountingOutputStream(var7);
                            var9.setListener(var106);
                            IOUtils.copy(var6.getEntity().getContent(), var9);
                         } catch (Exception var99) {
-                           LOGGER.error("Caught exception while downloading: " + var99.getMessage());
+                           LOGGER.error("Caught exception while downloading: {}", var99.getMessage());
                            this.error = true;
                         } finally {
                            this.request.releaseConnection();
                            if (this.tempFile != null) {
                               this.tempFile.delete();
@@ -249,11 +249,11 @@
                         var106 = new FileDownload.ResourcePackProgressListener(this.tempFile, var3, var1);
                         var9 = new FileDownload.DownloadCountingOutputStream(var7);
                         var9.setListener(var106);
                         IOUtils.copy(var6.getEntity().getContent(), var9);
                      } catch (Exception var101) {
-                        LOGGER.error("Caught exception while downloading: " + var101.getMessage());
+                        LOGGER.error("Caught exception while downloading: {}", var101.getMessage());
                         this.error = true;
                      } finally {
                         this.request.releaseConnection();
                         if (this.tempFile != null) {
                            this.tempFile.delete();
@@ -298,11 +298,11 @@
                      FileDownload.ResourcePackProgressListener var108 = new FileDownload.ResourcePackProgressListener(this.tempFile, var3, var1);
                      FileDownload.DownloadCountingOutputStream var10 = new FileDownload.DownloadCountingOutputStream(var107);
                      var10.setListener(var108);
                      IOUtils.copy(var105.getEntity().getContent(), var10);
                   } catch (Exception var97) {
-                     LOGGER.error("Caught exception while downloading: " + var97.getMessage());
+                     LOGGER.error("Caught exception while downloading: {}", var97.getMessage());
                      this.error = true;
                   } finally {
                      this.request.releaseConnection();
                      if (this.tempFile != null) {
                         this.tempFile.delete();
@@ -650,16 +650,16 @@
                String var2 = Hashing.sha1().hashBytes(Files.toByteArray(this.tempFile)).toString();
                if (var2.equals(this.worldDownload.resourcePackHash)) {
                   FileUtils.copyFile(this.tempFile, FileDownload.this.resourcePackPath);
                   FileDownload.this.finished = true;
                } else {
-                  FileDownload.LOGGER.error("Resourcepack had wrong hash (expected " + this.worldDownload.resourcePackHash + ", found " + var2 + "). Deleting it.");
+                  FileDownload.LOGGER.error("Resourcepack had wrong hash (expected {}, found {}). Deleting it.", this.worldDownload.resourcePackHash, var2);
                   FileUtils.deleteQuietly(this.tempFile);
                   FileDownload.this.error = true;
                }
             } catch (IOException var3) {
-               FileDownload.LOGGER.error("Error copying resourcepack file", var3.getMessage());
+               FileDownload.LOGGER.error("Error copying resourcepack file: {}", var3.getMessage());
                FileDownload.this.error = true;
             }
          }
 
       }
