--- com/mojang/realmsclient/client/Ping.java
+++ com/mojang/realmsclient/client/Ping.java
@@ -8,10 +8,13 @@
 import java.util.Comparator;
 import java.util.List;
 import net.minecraft.Util;
 
 public class Ping {
+   public Ping() {
+   }
+
    public static List<RegionPingResult> ping(Ping.Region... var0) {
       Ping.Region[] var1 = var0;
       int var2 = var0.length;
 
       int var3;
@@ -81,12 +84,12 @@
       AP_SOUTHEAST_1("ap-southeast-1", "ec2.ap-southeast-1.amazonaws.com"),
       AP_SOUTHEAST_2("ap-southeast-2", "ec2.ap-southeast-2.amazonaws.com"),
       AP_NORTHEAST_1("ap-northeast-1", "ec2.ap-northeast-1.amazonaws.com"),
       SA_EAST_1("sa-east-1", "ec2.sa-east-1.amazonaws.com");
 
-      private final String name;
-      private final String endpoint;
+      final String name;
+      final String endpoint;
 
       private Region(String var3, String var4) {
          this.name = var3;
          this.endpoint = var4;
       }
