--- com/mojang/realmsclient/client/RealmsError.java
+++ com/mojang/realmsclient/client/RealmsError.java
@@ -22,12 +22,12 @@
          JsonObject var2 = var1.parse(var0).getAsJsonObject();
          String var3 = JsonUtils.getStringOr("errorMsg", var2, "");
          int var4 = JsonUtils.getIntOr("errorCode", var2, -1);
          return new RealmsError(var3, var4);
       } catch (Exception var5) {
-         LOGGER.error("Could not parse RealmsError: " + var5.getMessage());
-         LOGGER.error("The error was: " + var0);
+         LOGGER.error("Could not parse RealmsError: {}", var5.getMessage());
+         LOGGER.error("The error was: {}", var0);
          return new RealmsError("Failed to parse response from server", -1);
       }
    }
 
    public String getErrorMessage() {
