--- net/minecraft/world/level/border/WorldBorder.java
+++ net/minecraft/world/level/border/WorldBorder.java
@@ -36,10 +36,14 @@
 
    public boolean isWithinBounds(ChunkPos var1) {
       return (double)var1.getMaxBlockX() > this.getMinX() && (double)var1.getMinBlockX() < this.getMaxX() && (double)var1.getMaxBlockZ() > this.getMinZ() && (double)var1.getMinBlockZ() < this.getMaxZ();
    }
 
+   public boolean isWithinBounds(double var1, double var3) {
+      return var1 > this.getMinX() && var1 < this.getMaxX() && var3 > this.getMinZ() && var3 < this.getMaxZ();
+   }
+
    public boolean isWithinBounds(AABB var1) {
       return var1.maxX > this.getMinX() && var1.minX < this.getMaxX() && var1.maxZ > this.getMinZ() && var1.minZ < this.getMaxZ();
    }
 
    public double getDistanceToBorder(Entity var1) {
@@ -141,10 +145,14 @@
 
    public void addListener(BorderChangeListener var1) {
       this.listeners.add(var1);
    }
 
+   public void removeListener(BorderChangeListener var1) {
+      this.listeners.remove(var1);
+   }
+
    public void setAbsoluteMaxSize(int var1) {
       this.absoluteMaxSize = var1;
       this.extent.onAbsoluteMaxSizeChange();
    }
 
