--- net/minecraft/world/level/BaseCommandBlock.java
+++ net/minecraft/world/level/BaseCommandBlock.java
@@ -162,11 +162,13 @@
 
    }
 
    public void sendMessage(Component var1, UUID var2) {
       if (this.trackOutput) {
-         this.lastOutput = (new TextComponent("[" + TIME_FORMAT.format(new Date()) + "] ")).append(var1);
+         SimpleDateFormat var10003 = TIME_FORMAT;
+         Date var10004 = new Date();
+         this.lastOutput = (new TextComponent("[" + var10003.format(var10004) + "] ")).append(var1);
          this.onUpdated();
       }
 
    }
 
