--- net/minecraft/world/level/ForcedChunksSavedData.java
+++ net/minecraft/world/level/ForcedChunksSavedData.java
@@ -4,18 +4,24 @@
 import it.unimi.dsi.fastutil.longs.LongSet;
 import net.minecraft.nbt.CompoundTag;
 import net.minecraft.world.level.saveddata.SavedData;
 
 public class ForcedChunksSavedData extends SavedData {
-   private LongSet chunks = new LongOpenHashSet();
+   public static final String FILE_ID = "chunks";
+   private static final String TAG_FORCED = "Forced";
+   private final LongSet chunks;
 
+   private ForcedChunksSavedData(LongSet var1) {
+      this.chunks = var1;
+   }
+
    public ForcedChunksSavedData() {
-      super("chunks");
+      this(new LongOpenHashSet());
    }
 
-   public void load(CompoundTag var1) {
-      this.chunks = new LongOpenHashSet(var1.getLongArray("Forced"));
+   public static ForcedChunksSavedData load(CompoundTag var0) {
+      return new ForcedChunksSavedData(new LongOpenHashSet(var0.getLongArray("Forced")));
    }
 
    public CompoundTag save(CompoundTag var1) {
       var1.putLongArray("Forced", this.chunks.toLongArray());
       return var1;
