--- net/minecraft/world/level/PotentialCalculator.java
+++ net/minecraft/world/level/PotentialCalculator.java
@@ -31,11 +31,11 @@
 
          return var4 * var2;
       }
    }
 
-   static class PointCharge {
+   private static class PointCharge {
       private final BlockPos pos;
       private final double charge;
 
       public PointCharge(BlockPos var1, double var2) {
          this.pos = var1;
