--- net/minecraft/world/level/StructureFeatureManager.java
+++ net/minecraft/world/level/StructureFeatureManager.java
@@ -28,12 +28,12 @@
          return new StructureFeatureManager(var1, this.worldGenSettings);
       }
    }
 
    public Stream<? extends StructureStart<?>> startsForFeature(SectionPos var1, StructureFeature<?> var2) {
-      return this.level.getChunk(var1.x(), var1.z(), ChunkStatus.STRUCTURE_REFERENCES).getReferencesForFeature(var2).stream().map((var0) -> {
-         return SectionPos.of(new ChunkPos(var0), 0);
+      return this.level.getChunk(var1.x(), var1.z(), ChunkStatus.STRUCTURE_REFERENCES).getReferencesForFeature(var2).stream().map((var1x) -> {
+         return SectionPos.of(new ChunkPos(var1x), this.level.getMinSection());
       }).map((var2x) -> {
          return this.getStartForFeature(var2x, var2, this.level.getChunk(var2x.x(), var2x.z(), ChunkStatus.STRUCTURE_STARTS));
       }).filter((var0) -> {
          return var0 != null && var0.isValid();
       });
