--- net/minecraft/world/level/levelgen/feature/configurations/CountConfiguration.java
+++ net/minecraft/world/level/levelgen/feature/configurations/CountConfiguration.java
@@ -1,21 +1,22 @@
 package net.minecraft.world.level.levelgen.feature.configurations;
 
 import com.mojang.serialization.Codec;
-import net.minecraft.util.UniformInt;
+import net.minecraft.util.valueproviders.ConstantInt;
+import net.minecraft.util.valueproviders.IntProvider;
 
 public class CountConfiguration implements DecoratorConfiguration, FeatureConfiguration {
-   public static final Codec<CountConfiguration> CODEC = UniformInt.codec(-10, 128, 128).fieldOf("count").xmap(CountConfiguration::new, CountConfiguration::count).codec();
-   private final UniformInt count;
+   public static final Codec<CountConfiguration> CODEC = IntProvider.codec(-10, 256).fieldOf("count").xmap(CountConfiguration::new, CountConfiguration::count).codec();
+   private final IntProvider count;
 
    public CountConfiguration(int var1) {
-      this.count = UniformInt.fixed(var1);
+      this.count = ConstantInt.of(var1);
    }
 
-   public CountConfiguration(UniformInt var1) {
+   public CountConfiguration(IntProvider var1) {
       this.count = var1;
    }
 
-   public UniformInt count() {
+   public IntProvider count() {
       return this.count;
    }
 }
