--- net/minecraft/world/level/levelgen/feature/configurations/LayerConfiguration.java
+++ net/minecraft/world/level/levelgen/feature/configurations/LayerConfiguration.java
@@ -1,14 +1,15 @@
 package net.minecraft.world.level.levelgen.feature.configurations;
 
 import com.mojang.serialization.Codec;
 import com.mojang.serialization.codecs.RecordCodecBuilder;
 import net.minecraft.world.level.block.state.BlockState;
+import net.minecraft.world.level.dimension.DimensionType;
 
 public class LayerConfiguration implements FeatureConfiguration {
    public static final Codec<LayerConfiguration> CODEC = RecordCodecBuilder.create((var0) -> {
-      return var0.group(Codec.intRange(0, 255).fieldOf("height").forGetter((var0x) -> {
+      return var0.group(Codec.intRange(DimensionType.MIN_Y, DimensionType.MAX_Y).fieldOf("height").forGetter((var0x) -> {
          return var0x.height;
       }), BlockState.CODEC.fieldOf("state").forGetter((var0x) -> {
          return var0x.state;
       })).apply(var0, LayerConfiguration::new);
    });
