--- net/minecraft/world/level/levelgen/feature/configurations/RandomPatchConfiguration.java
+++ net/minecraft/world/level/levelgen/feature/configurations/RandomPatchConfiguration.java
@@ -78,11 +78,11 @@
       private int xspread = 7;
       private int yspread = 3;
       private int zspread = 7;
       private boolean canReplace;
       private boolean project = true;
-      private boolean needWater = false;
+      private boolean needWater;
 
       public GrassConfigurationBuilder(BlockStateProvider var1, BlockPlacer var2) {
          this.stateProvider = var1;
          this.blockPlacer = var2;
       }
