--- net/minecraft/world/level/levelgen/feature/configurations/RangeDecoratorConfiguration.java
+++ net/minecraft/world/level/levelgen/feature/configurations/RangeDecoratorConfiguration.java
@@ -1,25 +1,26 @@
 package net.minecraft.world.level.levelgen.feature.configurations;
 
 import com.mojang.serialization.Codec;
 import com.mojang.serialization.codecs.RecordCodecBuilder;
+import net.minecraft.world.level.levelgen.VerticalAnchor;
 
 public class RangeDecoratorConfiguration implements DecoratorConfiguration {
    public static final Codec<RangeDecoratorConfiguration> CODEC = RecordCodecBuilder.create((var0) -> {
-      return var0.group(Codec.INT.fieldOf("bottom_offset").orElse(0).forGetter((var0x) -> {
-         return var0x.bottomOffset;
-      }), Codec.INT.fieldOf("top_offset").orElse(0).forGetter((var0x) -> {
-         return var0x.topOffset;
-      }), Codec.INT.fieldOf("maximum").orElse(0).forGetter((var0x) -> {
-         return var0x.maximum;
-      })).apply(var0, RangeDecoratorConfiguration::new);
+      return var0.group(VerticalAnchor.CODEC.fieldOf("bottom_inclusive").forGetter(RangeDecoratorConfiguration::bottomInclusive), VerticalAnchor.CODEC.fieldOf("top_inclusive").forGetter(RangeDecoratorConfiguration::topInclusive)).apply(var0, RangeDecoratorConfiguration::new);
    });
-   public final int bottomOffset;
-   public final int topOffset;
-   public final int maximum;
+   private final VerticalAnchor bottomInclusive;
+   private final VerticalAnchor topInclusive;
 
-   public RangeDecoratorConfiguration(int var1, int var2, int var3) {
-      this.bottomOffset = var1;
-      this.topOffset = var2;
-      this.maximum = var3;
+   public RangeDecoratorConfiguration(VerticalAnchor var1, VerticalAnchor var2) {
+      this.bottomInclusive = var1;
+      this.topInclusive = var2;
    }
+
+   public VerticalAnchor bottomInclusive() {
+      return this.bottomInclusive;
+   }
+
+   public VerticalAnchor topInclusive() {
+      return this.topInclusive;
+   }
 }
