--- net/minecraft/world/level/lighting/BlockLightEngine.java
+++ net/minecraft/world/level/lighting/BlockLightEngine.java
@@ -22,11 +22,11 @@
 
    private int getLightEmission(long var1) {
       int var3 = BlockPos.getX(var1);
       int var4 = BlockPos.getY(var1);
       int var5 = BlockPos.getZ(var1);
-      BlockGetter var6 = this.chunkSource.getChunkForLighting(var3 >> 4, var5 >> 4);
+      BlockGetter var6 = this.chunkSource.getChunkForLighting(SectionPos.blockToSectionCoord(var3), SectionPos.blockToSectionCoord(var5));
       return var6 != null ? var6.getLightEmission(this.pos.set(var3, var4, var5)) : 0;
    }
 
    protected int computeLevelFromNeighbor(long var1, long var3, int var5) {
       if (var3 == Long.MAX_VALUE) {
