--- net/minecraft/world/entity/projectile/DragonFireball.java
+++ net/minecraft/world/entity/projectile/DragonFireball.java
@@ -18,14 +18,10 @@
 public class DragonFireball extends AbstractHurtingProjectile {
    public DragonFireball(EntityType<? extends DragonFireball> var1, Level var2) {
       super(var1, var2);
    }
 
-   public DragonFireball(Level var1, double var2, double var4, double var6, double var8, double var10, double var12) {
-      super(EntityType.DRAGON_FIREBALL, var2, var4, var6, var8, var10, var12, var1);
-   }
-
    public DragonFireball(Level var1, LivingEntity var2, double var3, double var5, double var7) {
       super(EntityType.DRAGON_FIREBALL, var2, var3, var5, var7, var1);
    }
 
    protected void onHit(HitResult var1) {
@@ -57,11 +53,11 @@
                }
             }
 
             this.level.levelEvent(2006, this.blockPosition(), this.isSilent() ? -1 : 1);
             this.level.addFreshEntity(var4);
-            this.remove();
+            this.discard();
          }
 
       }
    }
 
