--- net/minecraft/world/entity/projectile/EvokerFangs.java
+++ net/minecraft/world/entity/projectile/EvokerFangs.java
@@ -15,10 +15,13 @@
 import net.minecraft.world.entity.EntityType;
 import net.minecraft.world.entity.LivingEntity;
 import net.minecraft.world.level.Level;
 
 public class EvokerFangs extends Entity {
+   public static final int ATTACK_DURATION = 20;
+   public static final int LIFE_OFFSET = 2;
+   public static final int ATTACK_TRIGGER_TICKS = 14;
    private int warmupDelayTicks;
    private boolean sentSpikeEvent;
    private int lifeTicks;
    private boolean clientSideAttackStarted;
    private LivingEntity owner;
@@ -105,11 +108,11 @@
             this.level.broadcastEntityEvent(this, (byte)4);
             this.sentSpikeEvent = true;
          }
 
          if (--this.lifeTicks < 0) {
-            this.remove();
+            this.discard();
          }
       }
 
    }
 
