--- net/minecraft/world/entity/projectile/EvokerFangs.java
+++ net/minecraft/world/entity/projectile/EvokerFangs.java
@@ -15,15 +15,20 @@
 import net.minecraft.world.entity.EntityType;
 import net.minecraft.world.entity.LivingEntity;
 import net.minecraft.world.level.Level;
 
 public class EvokerFangs extends Entity {
+   public static final int ATTACK_DURATION = 20;
+   public static final int LIFE_OFFSET = 2;
+   public static final int ATTACK_TRIGGER_TICKS = 14;
    private int warmupDelayTicks;
    private boolean sentSpikeEvent;
    private int lifeTicks;
    private boolean clientSideAttackStarted;
+   @Nullable
    private LivingEntity owner;
+   @Nullable
    private UUID ownerUUID;
 
    public EvokerFangs(EntityType<? extends EvokerFangs> var1, Level var2) {
       super(var1, var2);
       this.lifeTicks = 22;
@@ -31,11 +36,11 @@
 
    public EvokerFangs(Level var1, double var2, double var4, double var6, float var8, int var9, LivingEntity var10) {
       this(EntityType.EVOKER_FANGS, var1);
       this.warmupDelayTicks = var9;
       this.setOwner(var10);
-      this.yRot = var8 * 57.295776F;
+      this.setYRot(var8 * 57.295776F);
       this.setPos(var2, var4, var6);
    }
 
    protected void defineSynchedData() {
    }
@@ -105,11 +110,11 @@
             this.level.broadcastEntityEvent(this, (byte)4);
             this.sentSpikeEvent = true;
          }
 
          if (--this.lifeTicks < 0) {
-            this.remove();
+            this.discard();
          }
       }
 
    }
 
