--- net/minecraft/world/entity/projectile/LargeFireball.java
+++ net/minecraft/world/entity/projectile/LargeFireball.java
@@ -16,24 +16,20 @@
 
    public LargeFireball(EntityType<? extends LargeFireball> var1, Level var2) {
       super(var1, var2);
    }
 
-   public LargeFireball(Level var1, double var2, double var4, double var6, double var8, double var10, double var12) {
-      super(EntityType.FIREBALL, var2, var4, var6, var8, var10, var12, var1);
-   }
-
    public LargeFireball(Level var1, LivingEntity var2, double var3, double var5, double var7) {
       super(EntityType.FIREBALL, var2, var3, var5, var7, var1);
    }
 
    protected void onHit(HitResult var1) {
       super.onHit(var1);
       if (!this.level.isClientSide) {
          boolean var2 = this.level.getGameRules().getBoolean(GameRules.RULE_MOBGRIEFING);
          this.level.explode((Entity)null, this.getX(), this.getY(), this.getZ(), (float)this.explosionPower, var2, var2 ? Explosion.BlockInteraction.DESTROY : Explosion.BlockInteraction.NONE);
-         this.remove();
+         this.discard();
       }
 
    }
 
    protected void onHitEntity(EntityHitResult var1) {
