--- net/minecraft/world/entity/projectile/SmallFireball.java
+++ net/minecraft/world/entity/projectile/SmallFireball.java
@@ -47,24 +47,24 @@
 
    protected void onHitBlock(BlockHitResult var1) {
       super.onHitBlock(var1);
       if (!this.level.isClientSide) {
          Entity var2 = this.getOwner();
-         if (var2 == null || !(var2 instanceof Mob) || this.level.getGameRules().getBoolean(GameRules.RULE_MOBGRIEFING)) {
-            BlockPos var4 = var1.getBlockPos().relative(var1.getDirection());
-            if (this.level.isEmptyBlock(var4)) {
-               this.level.setBlockAndUpdate(var4, BaseFireBlock.getState(this.level, var4));
+         if (!(var2 instanceof Mob) || this.level.getGameRules().getBoolean(GameRules.RULE_MOBGRIEFING)) {
+            BlockPos var3 = var1.getBlockPos().relative(var1.getDirection());
+            if (this.level.isEmptyBlock(var3)) {
+               this.level.setBlockAndUpdate(var3, BaseFireBlock.getState(this.level, var3));
             }
          }
 
       }
    }
 
    protected void onHit(HitResult var1) {
       super.onHit(var1);
       if (!this.level.isClientSide) {
-         this.remove();
+         this.discard();
       }
 
    }
 
    public boolean isPickable() {
