--- net/minecraft/world/entity/projectile/SpectralArrow.java
+++ net/minecraft/world/entity/projectile/SpectralArrow.java
@@ -38,11 +38,11 @@
    }
 
    protected void doPostHurtEffects(LivingEntity var1) {
       super.doPostHurtEffects(var1);
       MobEffectInstance var2 = new MobEffectInstance(MobEffects.GLOWING, this.duration, 0);
-      var1.addEffect(var2);
+      var1.addEffect(var2, this.getEffectSource());
    }
 
    public void readAdditionalSaveData(CompoundTag var1) {
       super.readAdditionalSaveData(var1);
       if (var1.contains("Duration")) {
