--- net/minecraft/world/entity/projectile/ThrownTrident.java
+++ net/minecraft/world/entity/projectile/ThrownTrident.java
@@ -41,15 +41,10 @@
       this.tridentItem = var3.copy();
       this.entityData.set(ID_LOYALTY, (byte)EnchantmentHelper.getLoyalty(var3));
       this.entityData.set(ID_FOIL, var3.hasFoil());
    }
 
-   public ThrownTrident(Level var1, double var2, double var4, double var6) {
-      super(EntityType.TRIDENT, var2, var4, var6, var1);
-      this.tridentItem = new ItemStack(Items.TRIDENT);
-   }
-
    protected void defineSynchedData() {
       super.defineSynchedData();
       this.entityData.define(ID_LOYALTY, (byte)0);
       this.entityData.define(ID_FOIL, false);
    }
@@ -58,21 +53,21 @@
       if (this.inGroundTime > 4) {
          this.dealtDamage = true;
       }
 
       Entity var1 = this.getOwner();
-      if ((this.dealtDamage || this.isNoPhysics()) && var1 != null) {
-         byte var2 = (Byte)this.entityData.get(ID_LOYALTY);
-         if (var2 > 0 && !this.isAcceptibleReturnOwner()) {
+      byte var2 = (Byte)this.entityData.get(ID_LOYALTY);
+      if (var2 > 0 && (this.dealtDamage || this.isNoPhysics()) && var1 != null) {
+         if (!this.isAcceptibleReturnOwner()) {
             if (!this.level.isClientSide && this.pickup == AbstractArrow.Pickup.ALLOWED) {
                this.spawnAtLocation(this.getPickupItem(), 0.1F);
             }
 
-            this.remove();
-         } else if (var2 > 0) {
+            this.discard();
+         } else {
             this.setNoPhysics(true);
-            Vec3 var3 = new Vec3(var1.getX() - this.getX(), var1.getEyeY() - this.getY(), var1.getZ() - this.getZ());
+            Vec3 var3 = var1.getEyePosition().subtract(this.position());
             this.setPosRaw(this.getX(), this.getY() + var3.y * 0.015D * (double)var2, this.getZ());
             if (this.level.isClientSide) {
                this.yOld = this.getY();
             }
 
@@ -139,11 +134,11 @@
          }
       }
 
       this.setDeltaMovement(this.getDeltaMovement().multiply(-0.01D, -0.1D, -0.01D));
       float var11 = 1.0F;
-      if (this.level instanceof ServerLevel && this.level.isThundering() && EnchantmentHelper.hasChanneling(this.tridentItem)) {
+      if (this.level instanceof ServerLevel && this.level.isThundering() && this.isChanneling()) {
          BlockPos var8 = var2.blockPosition();
          if (this.level.canSeeSky(var8)) {
             LightningBolt var9 = (LightningBolt)EntityType.LIGHTNING_BOLT.create(this.level);
             var9.moveTo(Vec3.atBottomCenterOf(var8));
             var9.setCause(var10 instanceof ServerPlayer ? (ServerPlayer)var10 : null);
@@ -154,19 +149,27 @@
       }
 
       this.playSound(var6, var11, 1.0F);
    }
 
+   public boolean isChanneling() {
+      return EnchantmentHelper.hasChanneling(this.tridentItem);
+   }
+
+   protected boolean tryPickup(Player var1) {
+      return super.tryPickup(var1) || this.isNoPhysics() && this.ownedBy(var1);
+   }
+
    protected SoundEvent getDefaultHitGroundSoundEvent() {
       return SoundEvents.TRIDENT_HIT_GROUND;
    }
 
    public void playerTouch(Player var1) {
-      Entity var2 = this.getOwner();
-      if (var2 == null || var2.getUUID() == var1.getUUID()) {
+      if (this.ownedBy(var1) || this.getOwner() == null) {
          super.playerTouch(var1);
       }
+
    }
 
    public void readAdditionalSaveData(CompoundTag var1) {
       super.readAdditionalSaveData(var1);
       if (var1.contains("Trident", 10)) {
