--- net/minecraft/world/entity/projectile/ThrownPotion.java
+++ net/minecraft/world/entity/projectile/ThrownPotion.java
@@ -94,11 +94,11 @@
             }
          }
 
          int var6 = var3.hasInstantEffects() ? 2007 : 2002;
          this.level.levelEvent(var6, this.blockPosition(), PotionUtils.getColor(var2));
-         this.remove();
+         this.discard();
       }
    }
 
    private void applyWater() {
       AABB var1 = this.getBoundingBox().inflate(4.0D, 2.0D, 4.0D);
@@ -187,11 +187,11 @@
 
       this.level.addFreshEntity(var3);
    }
 
    private boolean isLingering() {
-      return this.getItem().getItem() == Items.LINGERING_POTION;
+      return this.getItem().is(Items.LINGERING_POTION);
    }
 
    private void dowseFire(BlockPos var1, Direction var2) {
       BlockState var3 = this.level.getBlockState(var1);
       if (var3.is(BlockTags.FIRE)) {
