--- net/minecraft/world/entity/projectile/ThrownEgg.java
+++ net/minecraft/world/entity/projectile/ThrownEgg.java
@@ -51,17 +51,17 @@
             }
 
             for(int var3 = 0; var3 < var2; ++var3) {
                Chicken var4 = (Chicken)EntityType.CHICKEN.create(this.level);
                var4.setAge(-24000);
-               var4.moveTo(this.getX(), this.getY(), this.getZ(), this.yRot, 0.0F);
+               var4.moveTo(this.getX(), this.getY(), this.getZ(), this.getYRot(), 0.0F);
                this.level.addFreshEntity(var4);
             }
          }
 
          this.level.broadcastEntityEvent(this, (byte)3);
-         this.remove();
+         this.discard();
       }
 
    }
 
    protected Item getDefaultItem() {
