--- net/minecraft/world/entity/projectile/ThrownEnderpearl.java
+++ net/minecraft/world/entity/projectile/ThrownEnderpearl.java
@@ -24,14 +24,10 @@
 
    public ThrownEnderpearl(Level var1, LivingEntity var2) {
       super(EntityType.ENDER_PEARL, var2, var1);
    }
 
-   public ThrownEnderpearl(Level var1, double var2, double var4, double var6) {
-      super(EntityType.ENDER_PEARL, var2, var4, var6, var1);
-   }
-
    protected Item getDefaultItem() {
       return Items.ENDER_PEARL;
    }
 
    protected void onHitEntity(EntityHitResult var1) {
@@ -45,11 +41,11 @@
 
       for(int var3 = 0; var3 < 32; ++var3) {
          this.level.addParticle(ParticleTypes.PORTAL, this.getX(), this.getY() + this.random.nextDouble() * 2.0D, this.getZ(), this.random.nextGaussian(), 0.0D, this.random.nextGaussian());
       }
 
-      if (!this.level.isClientSide && !this.removed) {
+      if (!this.level.isClientSide && !this.isRemoved()) {
          if (var2 instanceof ServerPlayer) {
             ServerPlayer var5 = (ServerPlayer)var2;
             if (var5.connection.getConnection().isConnected() && var5.level == this.level && !var5.isSleeping()) {
                if (this.random.nextFloat() < 0.05F && this.level.getGameRules().getBoolean(GameRules.RULE_DOMOBSPAWNING)) {
                   Endermite var4 = (Endermite)EntityType.ENDERMITE.create(this.level);
@@ -69,19 +65,19 @@
          } else if (var2 != null) {
             var2.teleportTo(this.getX(), this.getY(), this.getZ());
             var2.fallDistance = 0.0F;
          }
 
-         this.remove();
+         this.discard();
       }
 
    }
 
    public void tick() {
       Entity var1 = this.getOwner();
       if (var1 instanceof Player && !var1.isAlive()) {
-         this.remove();
+         this.discard();
       } else {
          super.tick();
       }
 
    }
