--- net/minecraft/network/protocol/status/ServerStatus.java
+++ net/minecraft/network/protocol/status/ServerStatus.java
@@ -13,10 +13,12 @@
 import java.util.UUID;
 import net.minecraft.network.chat.Component;
 import net.minecraft.util.GsonHelper;
 
 public class ServerStatus {
+   public static final int FAVICON_WIDTH = 64;
+   public static final int FAVICON_HEIGHT = 64;
    private Component description;
    private ServerStatus.Players players;
    private ServerStatus.Version version;
    private String favicon;
 
