--- net/minecraft/client/model/AgeableListModel.java
+++ net/minecraft/client/model/AgeableListModel.java
@@ -8,12 +8,12 @@
 import net.minecraft.resources.ResourceLocation;
 import net.minecraft.world.entity.Entity;
 
 public abstract class AgeableListModel<E extends Entity> extends EntityModel<E> {
    private final boolean scaleHead;
-   private final float yHeadOffset;
-   private final float zHeadOffset;
+   private final float babyYHeadOffset;
+   private final float babyZHeadOffset;
    private final float babyHeadScale;
    private final float babyBodyScale;
    private final float bodyYOffset;
 
    protected AgeableListModel(boolean var1, float var2, float var3) {
@@ -25,12 +25,12 @@
    }
 
    protected AgeableListModel(Function<ResourceLocation, RenderType> var1, boolean var2, float var3, float var4, float var5, float var6, float var7) {
       super(var1);
       this.scaleHead = var2;
-      this.yHeadOffset = var3;
-      this.zHeadOffset = var4;
+      this.babyYHeadOffset = var3;
+      this.babyZHeadOffset = var4;
       this.babyHeadScale = var5;
       this.babyBodyScale = var6;
       this.bodyYOffset = var7;
    }
 
@@ -45,11 +45,11 @@
          if (this.scaleHead) {
             var9 = 1.5F / this.babyHeadScale;
             var1.scale(var9, var9, var9);
          }
 
-         var1.translate(0.0D, (double)(this.yHeadOffset / 16.0F), (double)(this.zHeadOffset / 16.0F));
+         var1.translate(0.0D, (double)(this.babyYHeadOffset / 16.0F), (double)(this.babyZHeadOffset / 16.0F));
          this.headParts().forEach((var8x) -> {
             var8x.render(var1, var2, var3, var4, var5, var6, var7, var8);
          });
          var1.popPose();
          var1.pushPose();
