--- net/minecraft/client/model/ChestedHorseModel.java
+++ net/minecraft/client/model/ChestedHorseModel.java
@@ -1,50 +1,48 @@
 package net.minecraft.client.model;
 
 import net.minecraft.client.model.geom.ModelPart;
+import net.minecraft.client.model.geom.PartPose;
+import net.minecraft.client.model.geom.builders.CubeDeformation;
+import net.minecraft.client.model.geom.builders.CubeListBuilder;
+import net.minecraft.client.model.geom.builders.LayerDefinition;
+import net.minecraft.client.model.geom.builders.MeshDefinition;
+import net.minecraft.client.model.geom.builders.PartDefinition;
 import net.minecraft.world.entity.animal.horse.AbstractChestedHorse;
 import net.minecraft.world.entity.animal.horse.AbstractHorse;
 
 public class ChestedHorseModel<T extends AbstractChestedHorse> extends HorseModel<T> {
-   private final ModelPart boxL = new ModelPart(this, 26, 21);
-   private final ModelPart boxR;
+   private final ModelPart leftChest;
+   private final ModelPart rightChest;
 
-   public ChestedHorseModel(float var1) {
+   public ChestedHorseModel(ModelPart var1) {
       super(var1);
-      this.boxL.addBox(-4.0F, 0.0F, -2.0F, 8.0F, 8.0F, 3.0F);
-      this.boxR = new ModelPart(this, 26, 21);
-      this.boxR.addBox(-4.0F, 0.0F, -2.0F, 8.0F, 8.0F, 3.0F);
-      this.boxL.yRot = -1.5707964F;
-      this.boxR.yRot = 1.5707964F;
-      this.boxL.setPos(6.0F, -8.0F, 0.0F);
-      this.boxR.setPos(-6.0F, -8.0F, 0.0F);
-      this.body.addChild(this.boxL);
-      this.body.addChild(this.boxR);
+      this.leftChest = this.body.getChild("left_chest");
+      this.rightChest = this.body.getChild("right_chest");
    }
 
-   protected void addEarModels(ModelPart var1) {
-      ModelPart var2 = new ModelPart(this, 0, 12);
-      var2.addBox(-1.0F, -7.0F, 0.0F, 2.0F, 7.0F, 1.0F);
-      var2.setPos(1.25F, -10.0F, 4.0F);
-      ModelPart var3 = new ModelPart(this, 0, 12);
-      var3.addBox(-1.0F, -7.0F, 0.0F, 2.0F, 7.0F, 1.0F);
-      var3.setPos(-1.25F, -10.0F, 4.0F);
-      var2.xRot = 0.2617994F;
-      var2.zRot = 0.2617994F;
-      var3.xRot = 0.2617994F;
-      var3.zRot = -0.2617994F;
-      var1.addChild(var2);
-      var1.addChild(var3);
+   public static LayerDefinition createBodyLayer() {
+      MeshDefinition var0 = HorseModel.createBodyMesh(CubeDeformation.NONE);
+      PartDefinition var1 = var0.getRoot();
+      PartDefinition var2 = var1.getChild("body");
+      CubeListBuilder var3 = CubeListBuilder.create().texOffs(26, 21).addBox(-4.0F, 0.0F, -2.0F, 8.0F, 8.0F, 3.0F);
+      var2.addOrReplaceChild("left_chest", var3, PartPose.offsetAndRotation(6.0F, -8.0F, 0.0F, 0.0F, -1.5707964F, 0.0F));
+      var2.addOrReplaceChild("right_chest", var3, PartPose.offsetAndRotation(-6.0F, -8.0F, 0.0F, 0.0F, 1.5707964F, 0.0F));
+      PartDefinition var4 = var1.getChild("head_parts").getChild("head");
+      CubeListBuilder var5 = CubeListBuilder.create().texOffs(0, 12).addBox(-1.0F, -7.0F, 0.0F, 2.0F, 7.0F, 1.0F);
+      var4.addOrReplaceChild("left_ear", var5, PartPose.offsetAndRotation(1.25F, -10.0F, 4.0F, 0.2617994F, 0.0F, 0.2617994F));
+      var4.addOrReplaceChild("right_ear", var5, PartPose.offsetAndRotation(-1.25F, -10.0F, 4.0F, 0.2617994F, 0.0F, -0.2617994F));
+      return LayerDefinition.create(var0, 64, 64);
    }
 
    public void setupAnim(T var1, float var2, float var3, float var4, float var5, float var6) {
       super.setupAnim((AbstractHorse)var1, var2, var3, var4, var5, var6);
       if (var1.hasChest()) {
-         this.boxL.visible = true;
-         this.boxR.visible = true;
+         this.leftChest.visible = true;
+         this.rightChest.visible = true;
       } else {
-         this.boxL.visible = false;
-         this.boxR.visible = false;
+         this.leftChest.visible = false;
+         this.rightChest.visible = false;
       }
 
    }
 }
