--- net/minecraft/client/model/CreeperModel.java
+++ net/minecraft/client/model/CreeperModel.java
@@ -1,58 +1,56 @@
 package net.minecraft.client.model;
 
-import com.google.common.collect.ImmutableList;
 import net.minecraft.client.model.geom.ModelPart;
+import net.minecraft.client.model.geom.PartPose;
+import net.minecraft.client.model.geom.builders.CubeDeformation;
+import net.minecraft.client.model.geom.builders.CubeListBuilder;
+import net.minecraft.client.model.geom.builders.LayerDefinition;
+import net.minecraft.client.model.geom.builders.MeshDefinition;
+import net.minecraft.client.model.geom.builders.PartDefinition;
 import net.minecraft.util.Mth;
 import net.minecraft.world.entity.Entity;
 
-public class CreeperModel<T extends Entity> extends ListModel<T> {
+public class CreeperModel<T extends Entity> extends HierarchicalModel<T> {
+   private final ModelPart root;
    private final ModelPart head;
-   private final ModelPart hair;
-   private final ModelPart body;
-   private final ModelPart leg0;
-   private final ModelPart leg1;
-   private final ModelPart leg2;
-   private final ModelPart leg3;
+   private final ModelPart rightHindLeg;
+   private final ModelPart leftHindLeg;
+   private final ModelPart rightFrontLeg;
+   private final ModelPart leftFrontLeg;
+   private static final int Y_OFFSET = 6;
 
-   public CreeperModel() {
-      this(0.0F);
+   public CreeperModel(ModelPart var1) {
+      this.root = var1;
+      this.head = var1.getChild("head");
+      this.leftHindLeg = var1.getChild("right_hind_leg");
+      this.rightHindLeg = var1.getChild("left_hind_leg");
+      this.leftFrontLeg = var1.getChild("right_front_leg");
+      this.rightFrontLeg = var1.getChild("left_front_leg");
    }
 
-   public CreeperModel(float var1) {
-      boolean var2 = true;
-      this.head = new ModelPart(this, 0, 0);
-      this.head.addBox(-4.0F, -8.0F, -4.0F, 8.0F, 8.0F, 8.0F, var1);
-      this.head.setPos(0.0F, 6.0F, 0.0F);
-      this.hair = new ModelPart(this, 32, 0);
-      this.hair.addBox(-4.0F, -8.0F, -4.0F, 8.0F, 8.0F, 8.0F, var1 + 0.5F);
-      this.hair.setPos(0.0F, 6.0F, 0.0F);
-      this.body = new ModelPart(this, 16, 16);
-      this.body.addBox(-4.0F, 0.0F, -2.0F, 8.0F, 12.0F, 4.0F, var1);
-      this.body.setPos(0.0F, 6.0F, 0.0F);
-      this.leg0 = new ModelPart(this, 0, 16);
-      this.leg0.addBox(-2.0F, 0.0F, -2.0F, 4.0F, 6.0F, 4.0F, var1);
-      this.leg0.setPos(-2.0F, 18.0F, 4.0F);
-      this.leg1 = new ModelPart(this, 0, 16);
-      this.leg1.addBox(-2.0F, 0.0F, -2.0F, 4.0F, 6.0F, 4.0F, var1);
-      this.leg1.setPos(2.0F, 18.0F, 4.0F);
-      this.leg2 = new ModelPart(this, 0, 16);
-      this.leg2.addBox(-2.0F, 0.0F, -2.0F, 4.0F, 6.0F, 4.0F, var1);
-      this.leg2.setPos(-2.0F, 18.0F, -4.0F);
-      this.leg3 = new ModelPart(this, 0, 16);
-      this.leg3.addBox(-2.0F, 0.0F, -2.0F, 4.0F, 6.0F, 4.0F, var1);
-      this.leg3.setPos(2.0F, 18.0F, -4.0F);
+   public static LayerDefinition createBodyLayer(CubeDeformation var0) {
+      MeshDefinition var1 = new MeshDefinition();
+      PartDefinition var2 = var1.getRoot();
+      var2.addOrReplaceChild("head", CubeListBuilder.create().texOffs(0, 0).addBox(-4.0F, -8.0F, -4.0F, 8.0F, 8.0F, 8.0F, var0), PartPose.offset(0.0F, 6.0F, 0.0F));
+      var2.addOrReplaceChild("body", CubeListBuilder.create().texOffs(16, 16).addBox(-4.0F, 0.0F, -2.0F, 8.0F, 12.0F, 4.0F, var0), PartPose.offset(0.0F, 6.0F, 0.0F));
+      CubeListBuilder var3 = CubeListBuilder.create().texOffs(0, 16).addBox(-2.0F, 0.0F, -2.0F, 4.0F, 6.0F, 4.0F, var0);
+      var2.addOrReplaceChild("right_hind_leg", var3, PartPose.offset(-2.0F, 18.0F, 4.0F));
+      var2.addOrReplaceChild("left_hind_leg", var3, PartPose.offset(2.0F, 18.0F, 4.0F));
+      var2.addOrReplaceChild("right_front_leg", var3, PartPose.offset(-2.0F, 18.0F, -4.0F));
+      var2.addOrReplaceChild("left_front_leg", var3, PartPose.offset(2.0F, 18.0F, -4.0F));
+      return LayerDefinition.create(var1, 64, 32);
    }
 
-   public Iterable<ModelPart> parts() {
-      return ImmutableList.of(this.head, this.body, this.leg0, this.leg1, this.leg2, this.leg3);
+   public ModelPart root() {
+      return this.root;
    }
 
    public void setupAnim(T var1, float var2, float var3, float var4, float var5, float var6) {
       this.head.yRot = var5 * 0.017453292F;
       this.head.xRot = var6 * 0.017453292F;
-      this.leg0.xRot = Mth.cos(var2 * 0.6662F) * 1.4F * var3;
-      this.leg1.xRot = Mth.cos(var2 * 0.6662F + 3.1415927F) * 1.4F * var3;
-      this.leg2.xRot = Mth.cos(var2 * 0.6662F + 3.1415927F) * 1.4F * var3;
-      this.leg3.xRot = Mth.cos(var2 * 0.6662F) * 1.4F * var3;
+      this.rightHindLeg.xRot = Mth.cos(var2 * 0.6662F) * 1.4F * var3;
+      this.leftHindLeg.xRot = Mth.cos(var2 * 0.6662F + 3.1415927F) * 1.4F * var3;
+      this.rightFrontLeg.xRot = Mth.cos(var2 * 0.6662F + 3.1415927F) * 1.4F * var3;
+      this.leftFrontLeg.xRot = Mth.cos(var2 * 0.6662F) * 1.4F * var3;
    }
 }
