--- net/minecraft/client/model/HoglinModel.java
+++ net/minecraft/client/model/HoglinModel.java
@@ -1,79 +1,68 @@
 package net.minecraft.client.model;
 
 import com.google.common.collect.ImmutableList;
 import net.minecraft.client.model.geom.ModelPart;
+import net.minecraft.client.model.geom.PartPose;
+import net.minecraft.client.model.geom.builders.CubeDeformation;
+import net.minecraft.client.model.geom.builders.CubeListBuilder;
+import net.minecraft.client.model.geom.builders.LayerDefinition;
+import net.minecraft.client.model.geom.builders.MeshDefinition;
+import net.minecraft.client.model.geom.builders.PartDefinition;
 import net.minecraft.util.Mth;
 import net.minecraft.world.entity.Mob;
 import net.minecraft.world.entity.monster.hoglin.HoglinBase;
 
 public class HoglinModel<T extends Mob & HoglinBase> extends AgeableListModel<T> {
    private final ModelPart head;
    private final ModelPart rightEar;
    private final ModelPart leftEar;
    private final ModelPart body;
-   private final ModelPart frontRightLeg;
-   private final ModelPart frontLeftLeg;
-   private final ModelPart backRightLeg;
-   private final ModelPart backLeftLeg;
+   private final ModelPart rightFrontLeg;
+   private final ModelPart leftFrontLeg;
+   private final ModelPart rightHindLeg;
+   private final ModelPart leftHindLeg;
    private final ModelPart mane;
 
-   public HoglinModel() {
+   public HoglinModel(ModelPart var1) {
       super(true, 8.0F, 6.0F, 1.9F, 2.0F, 24.0F);
-      this.texWidth = 128;
-      this.texHeight = 64;
-      this.body = new ModelPart(this);
-      this.body.setPos(0.0F, 7.0F, 0.0F);
-      this.body.texOffs(1, 1).addBox(-8.0F, -7.0F, -13.0F, 16.0F, 14.0F, 26.0F);
-      this.mane = new ModelPart(this);
-      this.mane.setPos(0.0F, -14.0F, -5.0F);
-      this.mane.texOffs(90, 33).addBox(0.0F, 0.0F, -9.0F, 0.0F, 10.0F, 19.0F, 0.001F);
-      this.body.addChild(this.mane);
-      this.head = new ModelPart(this);
-      this.head.setPos(0.0F, 2.0F, -12.0F);
-      this.head.texOffs(61, 1).addBox(-7.0F, -3.0F, -19.0F, 14.0F, 6.0F, 19.0F);
-      this.rightEar = new ModelPart(this);
-      this.rightEar.setPos(-6.0F, -2.0F, -3.0F);
-      this.rightEar.texOffs(1, 1).addBox(-6.0F, -1.0F, -2.0F, 6.0F, 1.0F, 4.0F);
-      this.rightEar.zRot = -0.6981317F;
-      this.head.addChild(this.rightEar);
-      this.leftEar = new ModelPart(this);
-      this.leftEar.setPos(6.0F, -2.0F, -3.0F);
-      this.leftEar.texOffs(1, 6).addBox(0.0F, -1.0F, -2.0F, 6.0F, 1.0F, 4.0F);
-      this.leftEar.zRot = 0.6981317F;
-      this.head.addChild(this.leftEar);
-      ModelPart var1 = new ModelPart(this);
-      var1.setPos(-7.0F, 2.0F, -12.0F);
-      var1.texOffs(10, 13).addBox(-1.0F, -11.0F, -1.0F, 2.0F, 11.0F, 2.0F);
-      this.head.addChild(var1);
-      ModelPart var2 = new ModelPart(this);
-      var2.setPos(7.0F, 2.0F, -12.0F);
-      var2.texOffs(1, 13).addBox(-1.0F, -11.0F, -1.0F, 2.0F, 11.0F, 2.0F);
-      this.head.addChild(var2);
-      this.head.xRot = 0.87266463F;
-      boolean var3 = true;
+      this.body = var1.getChild("body");
+      this.mane = this.body.getChild("mane");
+      this.head = var1.getChild("head");
+      this.rightEar = this.head.getChild("right_ear");
+      this.leftEar = this.head.getChild("left_ear");
+      this.rightFrontLeg = var1.getChild("right_front_leg");
+      this.leftFrontLeg = var1.getChild("left_front_leg");
+      this.rightHindLeg = var1.getChild("right_hind_leg");
+      this.leftHindLeg = var1.getChild("left_hind_leg");
+   }
+
+   public static LayerDefinition createBodyLayer() {
+      MeshDefinition var0 = new MeshDefinition();
+      PartDefinition var1 = var0.getRoot();
+      PartDefinition var2 = var1.addOrReplaceChild("body", CubeListBuilder.create().texOffs(1, 1).addBox(-8.0F, -7.0F, -13.0F, 16.0F, 14.0F, 26.0F), PartPose.offset(0.0F, 7.0F, 0.0F));
+      var2.addOrReplaceChild("mane", CubeListBuilder.create().texOffs(90, 33).addBox(0.0F, 0.0F, -9.0F, 0.0F, 10.0F, 19.0F, new CubeDeformation(0.001F)), PartPose.offset(0.0F, -14.0F, -5.0F));
+      PartDefinition var3 = var1.addOrReplaceChild("head", CubeListBuilder.create().texOffs(61, 1).addBox(-7.0F, -3.0F, -19.0F, 14.0F, 6.0F, 19.0F), PartPose.offsetAndRotation(0.0F, 2.0F, -12.0F, 0.87266463F, 0.0F, 0.0F));
+      var3.addOrReplaceChild("right_ear", CubeListBuilder.create().texOffs(1, 1).addBox(-6.0F, -1.0F, -2.0F, 6.0F, 1.0F, 4.0F), PartPose.offsetAndRotation(-6.0F, -2.0F, -3.0F, 0.0F, 0.0F, -0.6981317F));
+      var3.addOrReplaceChild("left_ear", CubeListBuilder.create().texOffs(1, 6).addBox(0.0F, -1.0F, -2.0F, 6.0F, 1.0F, 4.0F), PartPose.offsetAndRotation(6.0F, -2.0F, -3.0F, 0.0F, 0.0F, 0.6981317F));
+      var3.addOrReplaceChild("right_horn", CubeListBuilder.create().texOffs(10, 13).addBox(-1.0F, -11.0F, -1.0F, 2.0F, 11.0F, 2.0F), PartPose.offset(-7.0F, 2.0F, -12.0F));
+      var3.addOrReplaceChild("left_horn", CubeListBuilder.create().texOffs(1, 13).addBox(-1.0F, -11.0F, -1.0F, 2.0F, 11.0F, 2.0F), PartPose.offset(7.0F, 2.0F, -12.0F));
       boolean var4 = true;
-      this.frontRightLeg = new ModelPart(this);
-      this.frontRightLeg.setPos(-4.0F, 10.0F, -8.5F);
-      this.frontRightLeg.texOffs(66, 42).addBox(-3.0F, 0.0F, -3.0F, 6.0F, 14.0F, 6.0F);
-      this.frontLeftLeg = new ModelPart(this);
-      this.frontLeftLeg.setPos(4.0F, 10.0F, -8.5F);
-      this.frontLeftLeg.texOffs(41, 42).addBox(-3.0F, 0.0F, -3.0F, 6.0F, 14.0F, 6.0F);
-      this.backRightLeg = new ModelPart(this);
-      this.backRightLeg.setPos(-5.0F, 13.0F, 10.0F);
-      this.backRightLeg.texOffs(21, 45).addBox(-2.5F, 0.0F, -2.5F, 5.0F, 11.0F, 5.0F);
-      this.backLeftLeg = new ModelPart(this);
-      this.backLeftLeg.setPos(5.0F, 13.0F, 10.0F);
-      this.backLeftLeg.texOffs(0, 45).addBox(-2.5F, 0.0F, -2.5F, 5.0F, 11.0F, 5.0F);
+      boolean var5 = true;
+      var1.addOrReplaceChild("right_front_leg", CubeListBuilder.create().texOffs(66, 42).addBox(-3.0F, 0.0F, -3.0F, 6.0F, 14.0F, 6.0F), PartPose.offset(-4.0F, 10.0F, -8.5F));
+      var1.addOrReplaceChild("left_front_leg", CubeListBuilder.create().texOffs(41, 42).addBox(-3.0F, 0.0F, -3.0F, 6.0F, 14.0F, 6.0F), PartPose.offset(4.0F, 10.0F, -8.5F));
+      var1.addOrReplaceChild("right_hind_leg", CubeListBuilder.create().texOffs(21, 45).addBox(-2.5F, 0.0F, -2.5F, 5.0F, 11.0F, 5.0F), PartPose.offset(-5.0F, 13.0F, 10.0F));
+      var1.addOrReplaceChild("left_hind_leg", CubeListBuilder.create().texOffs(0, 45).addBox(-2.5F, 0.0F, -2.5F, 5.0F, 11.0F, 5.0F), PartPose.offset(5.0F, 13.0F, 10.0F));
+      return LayerDefinition.create(var0, 128, 64);
    }
 
    protected Iterable<ModelPart> headParts() {
       return ImmutableList.of(this.head);
    }
 
    protected Iterable<ModelPart> bodyParts() {
-      return ImmutableList.of(this.body, this.frontRightLeg, this.frontLeftLeg, this.backRightLeg, this.backLeftLeg);
+      return ImmutableList.of(this.body, this.rightFrontLeg, this.leftFrontLeg, this.rightHindLeg, this.leftHindLeg);
    }
 
    public void setupAnim(T var1, float var2, float var3, float var4, float var5, float var6) {
       this.rightEar.zRot = -0.6981317F - var3 * Mth.sin(var2);
       this.leftEar.zRot = 0.6981317F + var3 * Mth.sin(var2);
@@ -88,11 +77,11 @@
          this.head.y = 2.0F;
          this.mane.z = -7.0F;
       }
 
       float var9 = 1.2F;
-      this.frontRightLeg.xRot = Mth.cos(var2) * 1.2F * var3;
-      this.frontLeftLeg.xRot = Mth.cos(var2 + 3.1415927F) * 1.2F * var3;
-      this.backRightLeg.xRot = this.frontLeftLeg.xRot;
-      this.backLeftLeg.xRot = this.frontRightLeg.xRot;
+      this.rightFrontLeg.xRot = Mth.cos(var2) * 1.2F * var3;
+      this.leftFrontLeg.xRot = Mth.cos(var2 + 3.1415927F) * 1.2F * var3;
+      this.rightHindLeg.xRot = this.leftFrontLeg.xRot;
+      this.leftHindLeg.xRot = this.rightFrontLeg.xRot;
    }
 }
