--- net/minecraft/client/model/LeashKnotModel.java
+++ net/minecraft/client/model/LeashKnotModel.java
@@ -1,26 +1,36 @@
 package net.minecraft.client.model;
 
-import com.google.common.collect.ImmutableList;
 import net.minecraft.client.model.geom.ModelPart;
+import net.minecraft.client.model.geom.PartPose;
+import net.minecraft.client.model.geom.builders.CubeListBuilder;
+import net.minecraft.client.model.geom.builders.LayerDefinition;
+import net.minecraft.client.model.geom.builders.MeshDefinition;
+import net.minecraft.client.model.geom.builders.PartDefinition;
 import net.minecraft.world.entity.Entity;
 
-public class LeashKnotModel<T extends Entity> extends ListModel<T> {
+public class LeashKnotModel<T extends Entity> extends HierarchicalModel<T> {
+   private static final String KNOT = "knot";
+   private final ModelPart root;
    private final ModelPart knot;
 
-   public LeashKnotModel() {
-      this.texWidth = 32;
-      this.texHeight = 32;
-      this.knot = new ModelPart(this, 0, 0);
-      this.knot.addBox(-3.0F, -6.0F, -3.0F, 6.0F, 8.0F, 6.0F, 0.0F);
-      this.knot.setPos(0.0F, 0.0F, 0.0F);
+   public LeashKnotModel(ModelPart var1) {
+      this.root = var1;
+      this.knot = var1.getChild("knot");
    }
 
-   public Iterable<ModelPart> parts() {
-      return ImmutableList.of(this.knot);
+   public static LayerDefinition createBodyLayer() {
+      MeshDefinition var0 = new MeshDefinition();
+      PartDefinition var1 = var0.getRoot();
+      var1.addOrReplaceChild("knot", CubeListBuilder.create().texOffs(0, 0).addBox(-3.0F, -8.0F, -3.0F, 6.0F, 8.0F, 6.0F), PartPose.ZERO);
+      return LayerDefinition.create(var0, 32, 32);
    }
 
+   public ModelPart root() {
+      return this.root;
+   }
+
    public void setupAnim(T var1, float var2, float var3, float var4, float var5, float var6) {
       this.knot.yRot = var5 * 0.017453292F;
       this.knot.xRot = var6 * 0.017453292F;
    }
 }
