--- net/minecraft/client/model/RavagerModel.java
+++ net/minecraft/client/model/RavagerModel.java
@@ -1,83 +1,66 @@
 package net.minecraft.client.model;
 
-import com.google.common.collect.ImmutableList;
 import net.minecraft.client.model.geom.ModelPart;
+import net.minecraft.client.model.geom.PartPose;
+import net.minecraft.client.model.geom.builders.CubeListBuilder;
+import net.minecraft.client.model.geom.builders.LayerDefinition;
+import net.minecraft.client.model.geom.builders.MeshDefinition;
+import net.minecraft.client.model.geom.builders.PartDefinition;
 import net.minecraft.util.Mth;
 import net.minecraft.world.entity.monster.Ravager;
 
-public class RavagerModel extends ListModel<Ravager> {
+public class RavagerModel extends HierarchicalModel<Ravager> {
+   private final ModelPart root;
    private final ModelPart head;
    private final ModelPart mouth;
-   private final ModelPart body;
-   private final ModelPart leg0;
-   private final ModelPart leg1;
-   private final ModelPart leg2;
-   private final ModelPart leg3;
+   private final ModelPart rightHindLeg;
+   private final ModelPart leftHindLeg;
+   private final ModelPart rightFrontLeg;
+   private final ModelPart leftFrontLeg;
    private final ModelPart neck;
 
-   public RavagerModel() {
-      this.texWidth = 128;
-      this.texHeight = 128;
-      boolean var1 = true;
-      float var2 = 0.0F;
-      this.neck = new ModelPart(this);
-      this.neck.setPos(0.0F, -7.0F, -1.5F);
-      this.neck.texOffs(68, 73).addBox(-5.0F, -1.0F, -18.0F, 10.0F, 10.0F, 18.0F, 0.0F);
-      this.head = new ModelPart(this);
-      this.head.setPos(0.0F, 16.0F, -17.0F);
-      this.head.texOffs(0, 0).addBox(-8.0F, -20.0F, -14.0F, 16.0F, 20.0F, 16.0F, 0.0F);
-      this.head.texOffs(0, 0).addBox(-2.0F, -6.0F, -18.0F, 4.0F, 8.0F, 4.0F, 0.0F);
-      ModelPart var3 = new ModelPart(this);
-      var3.setPos(-10.0F, -14.0F, -8.0F);
-      var3.texOffs(74, 55).addBox(0.0F, -14.0F, -2.0F, 2.0F, 14.0F, 4.0F, 0.0F);
-      var3.xRot = 1.0995574F;
-      this.head.addChild(var3);
-      ModelPart var4 = new ModelPart(this);
-      var4.mirror = true;
-      var4.setPos(8.0F, -14.0F, -8.0F);
-      var4.texOffs(74, 55).addBox(0.0F, -14.0F, -2.0F, 2.0F, 14.0F, 4.0F, 0.0F);
-      var4.xRot = 1.0995574F;
-      this.head.addChild(var4);
-      this.mouth = new ModelPart(this);
-      this.mouth.setPos(0.0F, -2.0F, 2.0F);
-      this.mouth.texOffs(0, 36).addBox(-8.0F, 0.0F, -16.0F, 16.0F, 3.0F, 16.0F, 0.0F);
-      this.head.addChild(this.mouth);
-      this.neck.addChild(this.head);
-      this.body = new ModelPart(this);
-      this.body.texOffs(0, 55).addBox(-7.0F, -10.0F, -7.0F, 14.0F, 16.0F, 20.0F, 0.0F);
-      this.body.texOffs(0, 91).addBox(-6.0F, 6.0F, -7.0F, 12.0F, 13.0F, 18.0F, 0.0F);
-      this.body.setPos(0.0F, 1.0F, 2.0F);
-      this.leg0 = new ModelPart(this, 96, 0);
-      this.leg0.addBox(-4.0F, 0.0F, -4.0F, 8.0F, 37.0F, 8.0F, 0.0F);
-      this.leg0.setPos(-8.0F, -13.0F, 18.0F);
-      this.leg1 = new ModelPart(this, 96, 0);
-      this.leg1.mirror = true;
-      this.leg1.addBox(-4.0F, 0.0F, -4.0F, 8.0F, 37.0F, 8.0F, 0.0F);
-      this.leg1.setPos(8.0F, -13.0F, 18.0F);
-      this.leg2 = new ModelPart(this, 64, 0);
-      this.leg2.addBox(-4.0F, 0.0F, -4.0F, 8.0F, 37.0F, 8.0F, 0.0F);
-      this.leg2.setPos(-8.0F, -13.0F, -5.0F);
-      this.leg3 = new ModelPart(this, 64, 0);
-      this.leg3.mirror = true;
-      this.leg3.addBox(-4.0F, 0.0F, -4.0F, 8.0F, 37.0F, 8.0F, 0.0F);
-      this.leg3.setPos(8.0F, -13.0F, -5.0F);
+   public RavagerModel(ModelPart var1) {
+      this.root = var1;
+      this.neck = var1.getChild("neck");
+      this.head = this.neck.getChild("head");
+      this.mouth = this.head.getChild("mouth");
+      this.rightHindLeg = var1.getChild("right_hind_leg");
+      this.leftHindLeg = var1.getChild("left_hind_leg");
+      this.rightFrontLeg = var1.getChild("right_front_leg");
+      this.leftFrontLeg = var1.getChild("left_front_leg");
    }
 
-   public Iterable<ModelPart> parts() {
-      return ImmutableList.of(this.neck, this.body, this.leg0, this.leg1, this.leg2, this.leg3);
+   public static LayerDefinition createBodyLayer() {
+      MeshDefinition var0 = new MeshDefinition();
+      PartDefinition var1 = var0.getRoot();
+      boolean var2 = true;
+      PartDefinition var3 = var1.addOrReplaceChild("neck", CubeListBuilder.create().texOffs(68, 73).addBox(-5.0F, -1.0F, -18.0F, 10.0F, 10.0F, 18.0F), PartPose.offset(0.0F, -7.0F, 5.5F));
+      PartDefinition var4 = var3.addOrReplaceChild("head", CubeListBuilder.create().texOffs(0, 0).addBox(-8.0F, -20.0F, -14.0F, 16.0F, 20.0F, 16.0F).texOffs(0, 0).addBox(-2.0F, -6.0F, -18.0F, 4.0F, 8.0F, 4.0F), PartPose.offset(0.0F, 16.0F, -17.0F));
+      var4.addOrReplaceChild("right_horn", CubeListBuilder.create().texOffs(74, 55).addBox(0.0F, -14.0F, -2.0F, 2.0F, 14.0F, 4.0F), PartPose.offsetAndRotation(-10.0F, -14.0F, -8.0F, 1.0995574F, 0.0F, 0.0F));
+      var4.addOrReplaceChild("left_horn", CubeListBuilder.create().texOffs(74, 55).mirror().addBox(0.0F, -14.0F, -2.0F, 2.0F, 14.0F, 4.0F), PartPose.offsetAndRotation(8.0F, -14.0F, -8.0F, 1.0995574F, 0.0F, 0.0F));
+      var4.addOrReplaceChild("mouth", CubeListBuilder.create().texOffs(0, 36).addBox(-8.0F, 0.0F, -16.0F, 16.0F, 3.0F, 16.0F), PartPose.offset(0.0F, -2.0F, 2.0F));
+      var1.addOrReplaceChild("body", CubeListBuilder.create().texOffs(0, 55).addBox(-7.0F, -10.0F, -7.0F, 14.0F, 16.0F, 20.0F).texOffs(0, 91).addBox(-6.0F, 6.0F, -7.0F, 12.0F, 13.0F, 18.0F), PartPose.offsetAndRotation(0.0F, 1.0F, 2.0F, 1.5707964F, 0.0F, 0.0F));
+      var1.addOrReplaceChild("right_hind_leg", CubeListBuilder.create().texOffs(96, 0).addBox(-4.0F, 0.0F, -4.0F, 8.0F, 37.0F, 8.0F), PartPose.offset(-8.0F, -13.0F, 18.0F));
+      var1.addOrReplaceChild("left_hind_leg", CubeListBuilder.create().texOffs(96, 0).mirror().addBox(-4.0F, 0.0F, -4.0F, 8.0F, 37.0F, 8.0F), PartPose.offset(8.0F, -13.0F, 18.0F));
+      var1.addOrReplaceChild("right_front_leg", CubeListBuilder.create().texOffs(64, 0).addBox(-4.0F, 0.0F, -4.0F, 8.0F, 37.0F, 8.0F), PartPose.offset(-8.0F, -13.0F, -5.0F));
+      var1.addOrReplaceChild("left_front_leg", CubeListBuilder.create().texOffs(64, 0).mirror().addBox(-4.0F, 0.0F, -4.0F, 8.0F, 37.0F, 8.0F), PartPose.offset(8.0F, -13.0F, -5.0F));
+      return LayerDefinition.create(var0, 128, 128);
    }
 
+   public ModelPart root() {
+      return this.root;
+   }
+
    public void setupAnim(Ravager var1, float var2, float var3, float var4, float var5, float var6) {
       this.head.xRot = var6 * 0.017453292F;
       this.head.yRot = var5 * 0.017453292F;
-      this.body.xRot = 1.5707964F;
       float var7 = 0.4F * var3;
-      this.leg0.xRot = Mth.cos(var2 * 0.6662F) * var7;
-      this.leg1.xRot = Mth.cos(var2 * 0.6662F + 3.1415927F) * var7;
-      this.leg2.xRot = Mth.cos(var2 * 0.6662F + 3.1415927F) * var7;
-      this.leg3.xRot = Mth.cos(var2 * 0.6662F) * var7;
+      this.rightHindLeg.xRot = Mth.cos(var2 * 0.6662F) * var7;
+      this.leftHindLeg.xRot = Mth.cos(var2 * 0.6662F + 3.1415927F) * var7;
+      this.rightFrontLeg.xRot = Mth.cos(var2 * 0.6662F + 3.1415927F) * var7;
+      this.leftFrontLeg.xRot = Mth.cos(var2 * 0.6662F) * var7;
    }
 
    public void prepareMobModel(Ravager var1, float var2, float var3, float var4) {
       super.prepareMobModel(var1, var2, var3, var4);
       int var5 = var1.getStunnedTick();
