--- net/minecraft/client/model/SheepFurModel.java
+++ net/minecraft/client/model/SheepFurModel.java
@@ -1,36 +1,36 @@
 package net.minecraft.client.model;
 
 import net.minecraft.client.model.geom.ModelPart;
+import net.minecraft.client.model.geom.PartPose;
+import net.minecraft.client.model.geom.builders.CubeDeformation;
+import net.minecraft.client.model.geom.builders.CubeListBuilder;
+import net.minecraft.client.model.geom.builders.LayerDefinition;
+import net.minecraft.client.model.geom.builders.MeshDefinition;
+import net.minecraft.client.model.geom.builders.PartDefinition;
 import net.minecraft.world.entity.animal.Sheep;
 
 public class SheepFurModel<T extends Sheep> extends QuadrupedModel<T> {
    private float headXRot;
 
-   public SheepFurModel() {
-      super(12, 0.0F, false, 8.0F, 4.0F, 2.0F, 2.0F, 24);
-      this.head = new ModelPart(this, 0, 0);
-      this.head.addBox(-3.0F, -4.0F, -4.0F, 6.0F, 6.0F, 6.0F, 0.6F);
-      this.head.setPos(0.0F, 6.0F, -8.0F);
-      this.body = new ModelPart(this, 28, 8);
-      this.body.addBox(-4.0F, -10.0F, -7.0F, 8.0F, 16.0F, 6.0F, 1.75F);
-      this.body.setPos(0.0F, 5.0F, 2.0F);
-      float var1 = 0.5F;
-      this.leg0 = new ModelPart(this, 0, 16);
-      this.leg0.addBox(-2.0F, 0.0F, -2.0F, 4.0F, 6.0F, 4.0F, 0.5F);
-      this.leg0.setPos(-3.0F, 12.0F, 7.0F);
-      this.leg1 = new ModelPart(this, 0, 16);
-      this.leg1.addBox(-2.0F, 0.0F, -2.0F, 4.0F, 6.0F, 4.0F, 0.5F);
-      this.leg1.setPos(3.0F, 12.0F, 7.0F);
-      this.leg2 = new ModelPart(this, 0, 16);
-      this.leg2.addBox(-2.0F, 0.0F, -2.0F, 4.0F, 6.0F, 4.0F, 0.5F);
-      this.leg2.setPos(-3.0F, 12.0F, -5.0F);
-      this.leg3 = new ModelPart(this, 0, 16);
-      this.leg3.addBox(-2.0F, 0.0F, -2.0F, 4.0F, 6.0F, 4.0F, 0.5F);
-      this.leg3.setPos(3.0F, 12.0F, -5.0F);
+   public SheepFurModel(ModelPart var1) {
+      super(var1, false, 8.0F, 4.0F, 2.0F, 2.0F, 24);
    }
 
+   public static LayerDefinition createFurLayer() {
+      MeshDefinition var0 = new MeshDefinition();
+      PartDefinition var1 = var0.getRoot();
+      var1.addOrReplaceChild("head", CubeListBuilder.create().texOffs(0, 0).addBox(-3.0F, -4.0F, -4.0F, 6.0F, 6.0F, 6.0F, new CubeDeformation(0.6F)), PartPose.offset(0.0F, 6.0F, -8.0F));
+      var1.addOrReplaceChild("body", CubeListBuilder.create().texOffs(28, 8).addBox(-4.0F, -10.0F, -7.0F, 8.0F, 16.0F, 6.0F, new CubeDeformation(1.75F)), PartPose.offsetAndRotation(0.0F, 5.0F, 2.0F, 1.5707964F, 0.0F, 0.0F));
+      CubeListBuilder var2 = CubeListBuilder.create().texOffs(0, 16).addBox(-2.0F, 0.0F, -2.0F, 4.0F, 6.0F, 4.0F, new CubeDeformation(0.5F));
+      var1.addOrReplaceChild("right_hind_leg", var2, PartPose.offset(-3.0F, 12.0F, 7.0F));
+      var1.addOrReplaceChild("left_hind_leg", var2, PartPose.offset(3.0F, 12.0F, 7.0F));
+      var1.addOrReplaceChild("right_front_leg", var2, PartPose.offset(-3.0F, 12.0F, -5.0F));
+      var1.addOrReplaceChild("left_front_leg", var2, PartPose.offset(3.0F, 12.0F, -5.0F));
+      return LayerDefinition.create(var0, 64, 32);
+   }
+
    public void prepareMobModel(T var1, float var2, float var3, float var4) {
       super.prepareMobModel(var1, var2, var3, var4);
       this.head.y = 6.0F + var1.getHeadEatPositionScale(var4) * 9.0F;
       this.headXRot = var1.getHeadEatAngleScale(var4);
    }
