--- net/minecraft/client/model/SheepModel.java
+++ net/minecraft/client/model/SheepModel.java
@@ -1,23 +1,31 @@
 package net.minecraft.client.model;
 
 import net.minecraft.client.model.geom.ModelPart;
+import net.minecraft.client.model.geom.PartPose;
+import net.minecraft.client.model.geom.builders.CubeDeformation;
+import net.minecraft.client.model.geom.builders.CubeListBuilder;
+import net.minecraft.client.model.geom.builders.LayerDefinition;
+import net.minecraft.client.model.geom.builders.MeshDefinition;
+import net.minecraft.client.model.geom.builders.PartDefinition;
 import net.minecraft.world.entity.animal.Sheep;
 
 public class SheepModel<T extends Sheep> extends QuadrupedModel<T> {
    private float headXRot;
 
-   public SheepModel() {
-      super(12, 0.0F, false, 8.0F, 4.0F, 2.0F, 2.0F, 24);
-      this.head = new ModelPart(this, 0, 0);
-      this.head.addBox(-3.0F, -4.0F, -6.0F, 6.0F, 6.0F, 8.0F, 0.0F);
-      this.head.setPos(0.0F, 6.0F, -8.0F);
-      this.body = new ModelPart(this, 28, 8);
-      this.body.addBox(-4.0F, -10.0F, -7.0F, 8.0F, 16.0F, 6.0F, 0.0F);
-      this.body.setPos(0.0F, 5.0F, 2.0F);
+   public SheepModel(ModelPart var1) {
+      super(var1, false, 8.0F, 4.0F, 2.0F, 2.0F, 24);
    }
 
+   public static LayerDefinition createBodyLayer() {
+      MeshDefinition var0 = QuadrupedModel.createBodyMesh(12, CubeDeformation.NONE);
+      PartDefinition var1 = var0.getRoot();
+      var1.addOrReplaceChild("head", CubeListBuilder.create().texOffs(0, 0).addBox(-3.0F, -4.0F, -6.0F, 6.0F, 6.0F, 8.0F), PartPose.offset(0.0F, 6.0F, -8.0F));
+      var1.addOrReplaceChild("body", CubeListBuilder.create().texOffs(28, 8).addBox(-4.0F, -10.0F, -7.0F, 8.0F, 16.0F, 6.0F), PartPose.offsetAndRotation(0.0F, 5.0F, 2.0F, 1.5707964F, 0.0F, 0.0F));
+      return LayerDefinition.create(var0, 64, 32);
+   }
+
    public void prepareMobModel(T var1, float var2, float var3, float var4) {
       super.prepareMobModel(var1, var2, var3, var4);
       this.head.y = 6.0F + var1.getHeadEatPositionScale(var4) * 9.0F;
       this.headXRot = var1.getHeadEatAngleScale(var4);
    }
