--- net/minecraft/client/model/SlimeModel.java
+++ net/minecraft/client/model/SlimeModel.java
@@ -1,35 +1,41 @@
 package net.minecraft.client.model;
 
-import com.google.common.collect.ImmutableList;
 import net.minecraft.client.model.geom.ModelPart;
+import net.minecraft.client.model.geom.PartPose;
+import net.minecraft.client.model.geom.builders.CubeListBuilder;
+import net.minecraft.client.model.geom.builders.LayerDefinition;
+import net.minecraft.client.model.geom.builders.MeshDefinition;
+import net.minecraft.client.model.geom.builders.PartDefinition;
 import net.minecraft.world.entity.Entity;
 
-public class SlimeModel<T extends Entity> extends ListModel<T> {
-   private final ModelPart cube;
-   private final ModelPart eye0;
-   private final ModelPart eye1;
-   private final ModelPart mouth;
+public class SlimeModel<T extends Entity> extends HierarchicalModel<T> {
+   private final ModelPart root;
 
-   public SlimeModel(int var1) {
-      this.cube = new ModelPart(this, 0, var1);
-      this.eye0 = new ModelPart(this, 32, 0);
-      this.eye1 = new ModelPart(this, 32, 4);
-      this.mouth = new ModelPart(this, 32, 8);
-      if (var1 > 0) {
-         this.cube.addBox(-3.0F, 17.0F, -3.0F, 6.0F, 6.0F, 6.0F);
-         this.eye0.addBox(-3.25F, 18.0F, -3.5F, 2.0F, 2.0F, 2.0F);
-         this.eye1.addBox(1.25F, 18.0F, -3.5F, 2.0F, 2.0F, 2.0F);
-         this.mouth.addBox(0.0F, 21.0F, -3.5F, 1.0F, 1.0F, 1.0F);
-      } else {
-         this.cube.addBox(-4.0F, 16.0F, -4.0F, 8.0F, 8.0F, 8.0F);
-      }
+   public SlimeModel(ModelPart var1) {
+      this.root = var1;
+   }
 
+   public static LayerDefinition createOuterBodyLayer() {
+      MeshDefinition var0 = new MeshDefinition();
+      PartDefinition var1 = var0.getRoot();
+      var1.addOrReplaceChild("cube", CubeListBuilder.create().texOffs(0, 0).addBox(-4.0F, 16.0F, -4.0F, 8.0F, 8.0F, 8.0F), PartPose.ZERO);
+      return LayerDefinition.create(var0, 64, 32);
    }
 
+   public static LayerDefinition createInnerBodyLayer() {
+      MeshDefinition var0 = new MeshDefinition();
+      PartDefinition var1 = var0.getRoot();
+      var1.addOrReplaceChild("cube", CubeListBuilder.create().texOffs(0, 16).addBox(-3.0F, 17.0F, -3.0F, 6.0F, 6.0F, 6.0F), PartPose.ZERO);
+      var1.addOrReplaceChild("right_eye", CubeListBuilder.create().texOffs(32, 0).addBox(-3.25F, 18.0F, -3.5F, 2.0F, 2.0F, 2.0F), PartPose.ZERO);
+      var1.addOrReplaceChild("left_eye", CubeListBuilder.create().texOffs(32, 4).addBox(1.25F, 18.0F, -3.5F, 2.0F, 2.0F, 2.0F), PartPose.ZERO);
+      var1.addOrReplaceChild("mouth", CubeListBuilder.create().texOffs(32, 8).addBox(0.0F, 21.0F, -3.5F, 1.0F, 1.0F, 1.0F), PartPose.ZERO);
+      return LayerDefinition.create(var0, 64, 32);
+   }
+
    public void setupAnim(T var1, float var2, float var3, float var4, float var5, float var6) {
    }
 
-   public Iterable<ModelPart> parts() {
-      return ImmutableList.of(this.cube, this.eye0, this.eye1, this.mouth);
+   public ModelPart root() {
+      return this.root;
    }
 }
