--- net/minecraft/client/model/WitherBossModel.java
+++ net/minecraft/client/model/WitherBossModel.java
@@ -1,66 +1,72 @@
 package net.minecraft.client.model;
 
-import com.google.common.collect.ImmutableList;
-import com.google.common.collect.ImmutableList.Builder;
-import java.util.Arrays;
 import net.minecraft.client.model.geom.ModelPart;
+import net.minecraft.client.model.geom.PartPose;
+import net.minecraft.client.model.geom.builders.CubeDeformation;
+import net.minecraft.client.model.geom.builders.CubeListBuilder;
+import net.minecraft.client.model.geom.builders.LayerDefinition;
+import net.minecraft.client.model.geom.builders.MeshDefinition;
+import net.minecraft.client.model.geom.builders.PartDefinition;
 import net.minecraft.util.Mth;
 import net.minecraft.world.entity.boss.wither.WitherBoss;
 
-public class WitherBossModel<T extends WitherBoss> extends ListModel<T> {
-   private final ModelPart[] upperBodyParts;
-   private final ModelPart[] heads;
-   private final ImmutableList<ModelPart> parts;
+public class WitherBossModel<T extends WitherBoss> extends HierarchicalModel<T> {
+   private static final String RIBCAGE = "ribcage";
+   private static final String CENTER_HEAD = "center_head";
+   private static final String RIGHT_HEAD = "right_head";
+   private static final String LEFT_HEAD = "left_head";
+   private static final float RIBCAGE_X_ROT_OFFSET = 0.065F;
+   private static final float TAIL_X_ROT_OFFSET = 0.265F;
+   private final ModelPart root;
+   private final ModelPart centerHead;
+   private final ModelPart rightHead;
+   private final ModelPart leftHead;
+   private final ModelPart ribcage;
+   private final ModelPart tail;
 
-   public WitherBossModel(float var1) {
-      this.texWidth = 64;
-      this.texHeight = 64;
-      this.upperBodyParts = new ModelPart[3];
-      this.upperBodyParts[0] = new ModelPart(this, 0, 16);
-      this.upperBodyParts[0].addBox(-10.0F, 3.9F, -0.5F, 20.0F, 3.0F, 3.0F, var1);
-      this.upperBodyParts[1] = (new ModelPart(this)).setTexSize(this.texWidth, this.texHeight);
-      this.upperBodyParts[1].setPos(-2.0F, 6.9F, -0.5F);
-      this.upperBodyParts[1].texOffs(0, 22).addBox(0.0F, 0.0F, 0.0F, 3.0F, 10.0F, 3.0F, var1);
-      this.upperBodyParts[1].texOffs(24, 22).addBox(-4.0F, 1.5F, 0.5F, 11.0F, 2.0F, 2.0F, var1);
-      this.upperBodyParts[1].texOffs(24, 22).addBox(-4.0F, 4.0F, 0.5F, 11.0F, 2.0F, 2.0F, var1);
-      this.upperBodyParts[1].texOffs(24, 22).addBox(-4.0F, 6.5F, 0.5F, 11.0F, 2.0F, 2.0F, var1);
-      this.upperBodyParts[2] = new ModelPart(this, 12, 22);
-      this.upperBodyParts[2].addBox(0.0F, 0.0F, 0.0F, 3.0F, 6.0F, 3.0F, var1);
-      this.heads = new ModelPart[3];
-      this.heads[0] = new ModelPart(this, 0, 0);
-      this.heads[0].addBox(-4.0F, -4.0F, -4.0F, 8.0F, 8.0F, 8.0F, var1);
-      this.heads[1] = new ModelPart(this, 32, 0);
-      this.heads[1].addBox(-4.0F, -4.0F, -4.0F, 6.0F, 6.0F, 6.0F, var1);
-      this.heads[1].x = -8.0F;
-      this.heads[1].y = 4.0F;
-      this.heads[2] = new ModelPart(this, 32, 0);
-      this.heads[2].addBox(-4.0F, -4.0F, -4.0F, 6.0F, 6.0F, 6.0F, var1);
-      this.heads[2].x = 10.0F;
-      this.heads[2].y = 4.0F;
-      Builder var2 = ImmutableList.builder();
-      var2.addAll(Arrays.asList(this.heads));
-      var2.addAll(Arrays.asList(this.upperBodyParts));
-      this.parts = var2.build();
+   public WitherBossModel(ModelPart var1) {
+      this.root = var1;
+      this.ribcage = var1.getChild("ribcage");
+      this.tail = var1.getChild("tail");
+      this.centerHead = var1.getChild("center_head");
+      this.rightHead = var1.getChild("right_head");
+      this.leftHead = var1.getChild("left_head");
    }
 
-   public ImmutableList<ModelPart> parts() {
-      return this.parts;
+   public static LayerDefinition createBodyLayer(CubeDeformation var0) {
+      MeshDefinition var1 = new MeshDefinition();
+      PartDefinition var2 = var1.getRoot();
+      var2.addOrReplaceChild("shoulders", CubeListBuilder.create().texOffs(0, 16).addBox(-10.0F, 3.9F, -0.5F, 20.0F, 3.0F, 3.0F, var0), PartPose.ZERO);
+      float var3 = 0.20420352F;
+      var2.addOrReplaceChild("ribcage", CubeListBuilder.create().texOffs(0, 22).addBox(0.0F, 0.0F, 0.0F, 3.0F, 10.0F, 3.0F, var0).texOffs(24, 22).addBox(-4.0F, 1.5F, 0.5F, 11.0F, 2.0F, 2.0F, var0).texOffs(24, 22).addBox(-4.0F, 4.0F, 0.5F, 11.0F, 2.0F, 2.0F, var0).texOffs(24, 22).addBox(-4.0F, 6.5F, 0.5F, 11.0F, 2.0F, 2.0F, var0), PartPose.offsetAndRotation(-2.0F, 6.9F, -0.5F, 0.20420352F, 0.0F, 0.0F));
+      var2.addOrReplaceChild("tail", CubeListBuilder.create().texOffs(12, 22).addBox(0.0F, 0.0F, 0.0F, 3.0F, 6.0F, 3.0F, var0), PartPose.offsetAndRotation(-2.0F, 6.9F + Mth.cos(0.20420352F) * 10.0F, -0.5F + Mth.sin(0.20420352F) * 10.0F, 0.83252203F, 0.0F, 0.0F));
+      var2.addOrReplaceChild("center_head", CubeListBuilder.create().texOffs(0, 0).addBox(-4.0F, -4.0F, -4.0F, 8.0F, 8.0F, 8.0F, var0), PartPose.ZERO);
+      CubeListBuilder var4 = CubeListBuilder.create().texOffs(32, 0).addBox(-4.0F, -4.0F, -4.0F, 6.0F, 6.0F, 6.0F, var0);
+      var2.addOrReplaceChild("right_head", var4, PartPose.offset(-8.0F, 4.0F, 0.0F));
+      var2.addOrReplaceChild("left_head", var4, PartPose.offset(10.0F, 4.0F, 0.0F));
+      return LayerDefinition.create(var1, 64, 64);
    }
 
+   public ModelPart root() {
+      return this.root;
+   }
+
    public void setupAnim(T var1, float var2, float var3, float var4, float var5, float var6) {
       float var7 = Mth.cos(var4 * 0.1F);
-      this.upperBodyParts[1].xRot = (0.065F + 0.05F * var7) * 3.1415927F;
-      this.upperBodyParts[2].setPos(-2.0F, 6.9F + Mth.cos(this.upperBodyParts[1].xRot) * 10.0F, -0.5F + Mth.sin(this.upperBodyParts[1].xRot) * 10.0F);
-      this.upperBodyParts[2].xRot = (0.265F + 0.1F * var7) * 3.1415927F;
-      this.heads[0].yRot = var5 * 0.017453292F;
-      this.heads[0].xRot = var6 * 0.017453292F;
+      this.ribcage.xRot = (0.065F + 0.05F * var7) * 3.1415927F;
+      this.tail.setPos(-2.0F, 6.9F + Mth.cos(this.ribcage.xRot) * 10.0F, -0.5F + Mth.sin(this.ribcage.xRot) * 10.0F);
+      this.tail.xRot = (0.265F + 0.1F * var7) * 3.1415927F;
+      this.centerHead.yRot = var5 * 0.017453292F;
+      this.centerHead.xRot = var6 * 0.017453292F;
    }
 
    public void prepareMobModel(T var1, float var2, float var3, float var4) {
-      for(int var5 = 1; var5 < 3; ++var5) {
-         this.heads[var5].yRot = (var1.getHeadYRot(var5 - 1) - var1.yBodyRot) * 0.017453292F;
-         this.heads[var5].xRot = var1.getHeadXRot(var5 - 1) * 0.017453292F;
-      }
+      setupHeadRotation(var1, this.rightHead, 0);
+      setupHeadRotation(var1, this.leftHead, 1);
+   }
 
+   private static <T extends WitherBoss> void setupHeadRotation(T var0, ModelPart var1, int var2) {
+      var1.yRot = (var0.getHeadYRot(var2) - var0.yBodyRot) * 0.017453292F;
+      var1.xRot = var0.getHeadXRot(var2) * 0.017453292F;
    }
 }
