--- net/minecraft/world/level/storage/loot/functions/ApplyExplosionDecay.java
+++ net/minecraft/world/level/storage/loot/functions/ApplyExplosionDecay.java
@@ -7,11 +7,11 @@
 import net.minecraft.world.level.storage.loot.LootContext;
 import net.minecraft.world.level.storage.loot.parameters.LootContextParams;
 import net.minecraft.world.level.storage.loot.predicates.LootItemCondition;
 
 public class ApplyExplosionDecay extends LootItemConditionalFunction {
-   private ApplyExplosionDecay(LootItemCondition[] var1) {
+   ApplyExplosionDecay(LootItemCondition[] var1) {
       super(var1);
    }
 
    public LootItemFunctionType getType() {
       return LootItemFunctions.EXPLOSION_DECAY;
