--- net/minecraft/world/level/storage/loot/functions/FillPlayerHead.java
+++ net/minecraft/world/level/storage/loot/functions/FillPlayerHead.java
@@ -32,21 +32,27 @@
    public Set<LootContextParam<?>> getReferencedContextParams() {
       return ImmutableSet.of(this.entityTarget.getParam());
    }
 
    public ItemStack run(ItemStack var1, LootContext var2) {
-      if (var1.getItem() == Items.PLAYER_HEAD) {
+      if (var1.is(Items.PLAYER_HEAD)) {
          Entity var3 = (Entity)var2.getParamOrNull(this.entityTarget.getParam());
          if (var3 instanceof Player) {
             GameProfile var4 = ((Player)var3).getGameProfile();
             var1.getOrCreateTag().put("SkullOwner", NbtUtils.writeGameProfile(new CompoundTag(), var4));
          }
       }
 
       return var1;
    }
 
+   public static LootItemConditionalFunction.Builder<?> fillPlayerHead(LootContext.EntityTarget var0) {
+      return simpleBuilder((var1) -> {
+         return new FillPlayerHead(var1, var0);
+      });
+   }
+
    public static class Serializer extends LootItemConditionalFunction.Serializer<FillPlayerHead> {
       public Serializer() {
       }
 
       public void serialize(JsonObject var1, FillPlayerHead var2, JsonSerializationContext var3) {
