--- net/minecraft/world/level/storage/loot/functions/LootItemFunctions.java
+++ net/minecraft/world/level/storage/loot/functions/LootItemFunctions.java
@@ -31,10 +31,11 @@
    public static final LootItemFunctionType EXPLOSION_DECAY = register("explosion_decay", new ApplyExplosionDecay.Serializer());
    public static final LootItemFunctionType SET_LORE = register("set_lore", new SetLoreFunction.Serializer());
    public static final LootItemFunctionType FILL_PLAYER_HEAD = register("fill_player_head", new FillPlayerHead.Serializer());
    public static final LootItemFunctionType COPY_NBT = register("copy_nbt", new CopyNbtFunction.Serializer());
    public static final LootItemFunctionType COPY_STATE = register("copy_state", new CopyBlockState.Serializer());
+   public static final LootItemFunctionType SET_BANNER_PATTERN = register("set_banner_pattern", new SetBannerPatternFunction.Serializer());
 
    private static LootItemFunctionType register(String var0, Serializer<? extends LootItemFunction> var1) {
       return (LootItemFunctionType)Registry.register(Registry.LOOT_FUNCTION_TYPE, (ResourceLocation)(new ResourceLocation(var0)), new LootItemFunctionType(var1));
    }
 
