--- net/minecraft/world/level/storage/loot/functions/SetNameFunction.java
+++ net/minecraft/world/level/storage/loot/functions/SetNameFunction.java
@@ -68,10 +68,22 @@
       }
 
       return var1;
    }
 
+   public static LootItemConditionalFunction.Builder<?> setName(Component var0) {
+      return simpleBuilder((var1) -> {
+         return new SetNameFunction(var1, var0, (LootContext.EntityTarget)null);
+      });
+   }
+
+   public static LootItemConditionalFunction.Builder<?> setName(Component var0, LootContext.EntityTarget var1) {
+      return simpleBuilder((var2) -> {
+         return new SetNameFunction(var2, var0, var1);
+      });
+   }
+
    public static class Serializer extends LootItemConditionalFunction.Serializer<SetNameFunction> {
       public Serializer() {
       }
 
       public void serialize(JsonObject var1, SetNameFunction var2, JsonSerializationContext var3) {
