--- net/minecraft/world/level/storage/loot/functions/SetNameFunction.java
+++ net/minecraft/world/level/storage/loot/functions/SetNameFunction.java
@@ -21,15 +21,15 @@
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
 
 public class SetNameFunction extends LootItemConditionalFunction {
    private static final Logger LOGGER = LogManager.getLogger();
-   private final Component name;
+   final Component name;
    @Nullable
-   private final LootContext.EntityTarget resolutionContext;
+   final LootContext.EntityTarget resolutionContext;
 
-   private SetNameFunction(LootItemCondition[] var1, @Nullable Component var2, @Nullable LootContext.EntityTarget var3) {
+   SetNameFunction(LootItemCondition[] var1, @Nullable Component var2, @Nullable LootContext.EntityTarget var3) {
       super(var1);
       this.name = var2;
       this.resolutionContext = var3;
    }
 
@@ -46,11 +46,11 @@
          Entity var2 = (Entity)var0.getParamOrNull(var1.getParam());
          if (var2 != null) {
             CommandSourceStack var3 = var2.createCommandSourceStack().withPermission(2);
             return (var2x) -> {
                try {
-                  return ComponentUtils.updateForEntity(var3, var2x, var2, 0);
+                  return ComponentUtils.updateForEntity(var3, (Component)var2x, var2, 0);
                } catch (CommandSyntaxException var4) {
                   LOGGER.warn("Failed to resolve text component", var4);
                   return var2x;
                }
             };
@@ -68,10 +68,22 @@
       }
 
       return var1;
    }
 
+   public static LootItemConditionalFunction.Builder<?> setName(Component var0) {
+      return simpleBuilder((var1) -> {
+         return new SetNameFunction(var1, var0, (LootContext.EntityTarget)null);
+      });
+   }
+
+   public static LootItemConditionalFunction.Builder<?> setName(Component var0, LootContext.EntityTarget var1) {
+      return simpleBuilder((var2) -> {
+         return new SetNameFunction(var2, var0, var1);
+      });
+   }
+
    public static class Serializer extends LootItemConditionalFunction.Serializer<SetNameFunction> {
       public Serializer() {
       }
 
       public void serialize(JsonObject var1, SetNameFunction var2, JsonSerializationContext var3) {
