--- net/minecraft/world/level/storage/loot/functions/SetStewEffectFunction.java
+++ net/minecraft/world/level/storage/loot/functions/SetStewEffectFunction.java
@@ -37,11 +37,11 @@
    public LootItemFunctionType getType() {
       return LootItemFunctions.SET_STEW_EFFECT;
    }
 
    public ItemStack run(ItemStack var1, LootContext var2) {
-      if (var1.getItem() == Items.SUSPICIOUS_STEW && !this.effectDurationMap.isEmpty()) {
+      if (var1.is(Items.SUSPICIOUS_STEW) && !this.effectDurationMap.isEmpty()) {
          Random var3 = var2.getRandom();
          int var4 = var3.nextInt(this.effectDurationMap.size());
          Entry var5 = (Entry)Iterables.get(this.effectDurationMap.entrySet(), var4);
          MobEffect var6 = (MobEffect)var5.getKey();
          int var7 = ((RandomValueBounds)var5.getValue()).getInt(var3);
