--- net/minecraft/world/level/storage/loot/functions/SmeltItemFunction.java
+++ net/minecraft/world/level/storage/loot/functions/SmeltItemFunction.java
@@ -13,11 +13,11 @@
 import org.apache.logging.log4j.Logger;
 
 public class SmeltItemFunction extends LootItemConditionalFunction {
    private static final Logger LOGGER = LogManager.getLogger();
 
-   private SmeltItemFunction(LootItemCondition[] var1) {
+   SmeltItemFunction(LootItemCondition[] var1) {
       super(var1);
    }
 
    public LootItemFunctionType getType() {
       return LootItemFunctions.FURNACE_SMELT;
